---
output: html_document
---

# Biology & Bioinformatics

## R in Biology {#bio-r}

R became popular in biological data analysis in the early to mid 2000s, when
[microarray
technology](https://www.genome.gov/genetics-glossary/Microarray-Technology) came
into widespread use enabling researchers to look for statistical differences in
gene expression for thousands of genes across large numbers of samples. As a
result of this popularity, a community of biological researchers and data
analysts created a collection of software packages called
[Bioconductor](https://www.bioconductor.org/), which made a vast array of
cutting edge statistical and bioinformatic methodologies widely available.

Due to its ease of use, widespread community support, rich ecosystem of
biological data analysis packages, and to it being free, R remains one of the
primary languages of biological data analysis. The language's early focus on
statistical analysis, and later transition to data science in general, makes it
very useful and accessible to perform the kinds of analyses the new data science
of biology required. It is also a bridge between biologists without a
computational background and statisticians and bioinformaticians, who invent and
implement methods into R packages that are easily accessible by all. The
language and the package ecosystems and communities it supports continue to be a
major source of biological discovery and innovation.

As a data science, biology benefits from the main strengths of R and [the
tidyverse](#dw-tidyverse) when combined with the powerful analytical techniques
available in Bioconductor packages, namely to manipulate, visualize, analyze,
and communicate biological data.

## CSV Files

The most common, convenient, and flexible data file format in biology and
bioinformatics is the *character-delimited* or *character-separated text file*.
These files are plain text files (i.e. not the native file format of any
specific program, like Excel) that generally contain rectangles of data. When
formatted correctly, you can think of these files as containing a grid or matrix
of data values with some number of columns, each of which has the same number of
values. Each line of these files has some number of data values separated by a
consistent character, most commonly the comma which are called [comma-separated
value, or "CSV", files](https://en.wikipedia.org/wiki/Comma-separated_values)
and filenames typically end with the extension `.csv`. Note that
other characters, especially the <Tab> character, may be used to create valid
files in this format, and all the same general principles apply. This is an
example of a simple CSV file:

```
id,somevalue,category,genes
1,2.123,A,APOE
4,5.123,B,"HOXA1,HOXB1"
7,8.123,,SNCA
```

Some properties and principles of CSV files:

* The first line often but not always contains the column names of each column
* Each value is delimited by the same character, in this case `,`
* Values can be any value, including numbers and characters
* When a value contains the delimiting character (e.g. HOXA1,HOXB1 contains a
  `,`), the value is wrapped in double quotes
* Values can be missing, indicated by sequential delimiters (i.e. `,,` or one
  `,` at the end of the line, if the last column value is missing)
* There is no delimiter at the end of the lines
* To be well-formatted *every line must have the same number of delimited
  values*

These same properties and principles apply to all character-separated files,
regardless of the specific delimiter used. If a file follows these principles,
they can be loaded very easily into R or any other data analysis setting.

## Types of Biological Data {#bio-data-types}

In very general terms, there are five types of data used in biological data
analysis: raw/primary data, processed data, analysis results, metadata, and
annotation data.

**Raw/primary data.** These data are the primary observations made by whatever
instruments/techniques we are using for our experiment. These may include
high-throughput sequencing data, mass/charge ratio data from mass spectrometry,
16S rRNA sequencing data from metagenomic studies, SNPs from genotyping assays,
etc. These data can be very large and are often not efficiently processed using
R. Instead, specialized tools built outside of R are used to first process the
primary data into a form that is amenable to analysis.

**Processed data.** Processed data is the result of any analysis or
transformation of primary data into an intermediate, more interpretable form.
For example, when performing gene expression analysis with RNASeq, short reads
that form the primary data are typically aligned against a genome and then
counted against annotated genes, resulting in a count for each gene in the
genome. Typically, these counts are then combined for many samples into a single
matrix and subject to downstream analyses like differential expression.

**Analysis results.** Analysis results aren't data *per se*, but are the results
of analysis of primary data or processed results. For example, in gene
expression studies, a first analysis is often differential expression, where
each gene in the genome is tested for expression associated with some outcome
of interest across many samples. Each gene then has a number of statistics and
related values, like log2 fold change, nominal and adjusted p-value, etc. These
forms of data are usually what we use to form interpretations of our datasets
and therefore we must manipulate them in much the same way as any other dataset.

**Metadata.** In most experimental settings, multiple samples have been
processed and had the same primary data type collected. These samples also have
information associated with them, which is here termed *metadata*, or "data
about data". In our gene expression of post mortem brain experiments mentioned
above, the information about the individuals themselves, including age at death,
whether the person had a disease, the measurements of tissue quality, etc. is
the metadata. The primary and processed data and metadata are usually stored in
different files, where the metadata (or *sample information* or *sample data*,
etc) will have one column indicating the unique identifier (ID) of each sample.
The processed data will typically have columns named for each of the sample IDs.

**Annotation data.** A tremendous amount of knowledge has been generated about
biological entities, e.g. genes, especially since the publication of the human
genome. Annotation data is publicly available information about the features we
measure in our experiments. This may come in the form of the coordinates in a
genome where genes exist, any known functions of those genes, the domains found
in proteins and their relative sequence, gene identifier cross references across
different gene naming systems (e.g. symbol vs Ensembl ID), single nucleotide
polymorphism genomic locations and associations with traits or diseases, etc.
This is information that we use to aid in interpretation of our experimental
data but generally do not generate ourselves. Annotation data comes in many
forms, some of which are in CSV format.

The figure below contains a simple diagram of how these different types of data
work together for a typical biological data analysis.

![Information flow in biological data analysis](information_flow.png)

## Bioconductor {#bio-bioconductor}

[Bioconductor](https://www.bioconductor.org/) is an organized collection of
strictly biological analysis methods packages for R. These packages are hosted
and maintained outside of [CRAN](https://cran.r-project.org/) because the
maintainers enforce [a more rigorous set of coding quality, testing, and
documentation
standards](https://bioconductor.org/developers/package-guidelines/) than is
required by normal R packages. These stricter requirements arose from a
recognition that software packages are generally only as usable as their
documentation allows them to be, and also that many if not most of the users of
these packages are not statisticians or experienced computer programmers.
Rather, they are people like us: biological analysis practitioners who may or
may not have substantial coding experience but must analyze our data
nonetheless. The excellent documentation and community support of the
bioconductor ecosystem is a primary reason why R is such a popular language in
biological analysis.

Bioconductor is divided into roughly two sets of packages: *core maintainer*
packages and user contributed packages. The core maintainer packages are among
the most critical, because they define a set of common objects and classes (e.g.
the [ExpressionSet class](https://www.bioconductor.org/packages/devel/bioc/vignettes/Biobase/inst/doc/ExpressionSetIntroduction.pdf)
in the [Biobase package](https://bioconductor.org/packages/release/bioc/html/Biobase.html))
that all Bioconductor packages know how to work with. This common base provides
consistency among all Bioconductor packages thereby streamlining the learning
process. User contributed and maintained packages provide specialized
functionality for specific types of analysis.

Bioconductor itself must be installed prior to installing other Bioconductor
packages. To [install bioconductor], you can run the following:

```r
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install(version = "3.14")
```

::: {.box .warning}
Bioconductor undergoes regular releases indicated by a version, e.g. `version =
"3.14"` at the time of writing. Every bioconductor package also has a version,
and each of those versions may or may not be compatible with a specific version
of Bioconductor. To make matters worse, Bioconductor versions are only
compatible with certain versions of R. Bioconductor version 3.14 requires R
version 4.1.0, and will not work with older R versions. These versions can cause
major version compatibility issues when you are forced to use older versions of
R, as may sometimes be the case on managed compute clusters. There is not a good
general solution for ensuring all your packages work together, but the general
best rule of thumb is to use the most current version of R and all packages at
the time when you write your code.
:::

The `BiocManager` package is the only Bioconductor package installable using
`install.packages()`. After installing the BiocManager package, you may then
install bioconductor packages:

```r
# installs the affy bioconductor package for microarray analysis
BiocManager::install("affy")
```

One key aspect of Bioconductor packages is consistent and helpful documentation;
every package page on the Bioconductor.org site lists a block of code that will
install the package, e.g. for the [affy
package](https://www.bioconductor.org/packages/release/bioc/html/affy.html). In
addition, Biconductor provides three types of documentation:

* [Workflow tutorials](https://bioconductor.org/packages/release/workflows/) on
  how to perform specific analysis use cases
* [Package vignettes](https://bioconductor.org/help/package-vignettes/) for
  every package, which provide a worked example of how to use the package that
  can be adapted by the user to their specific case
* Detailed, consistently formatted reference documentation that gives precise
  information on functionality and use of each package

The thorough and useful documentation of packages in Bioconductor is one of the
reasons why the package ecosystem, and R, is so widely used in biology and
bioinformatics.

The base Bioconductor packages define convenient data structures for storing and
analyzing many types of data. Recall earlier in the [Types of Biological Data]
section, we described five types of biological data: primary, processed,
analysis, metadata, and annotation data. Bioconductor provides a convenient
class for storing many of these different data types together in one place,
specifically the SummarizedExperiment class in the [package of the same
name](https://bioconductor.org/packages/release/bioc/html/SummarizedExperiment.html)
package. An illustration of what a SummarizedExperiment object stores is below,
from the Bioconductor maintainer's paper in Nature:

![SummarizedExperiment Schematic - Huber, et al. 2015. “Orchestrating High-Throughput Genomic Analysis with Bioconductor.” Nature Methods 12 (2): 115–21.](https://media.springernature.com/full/springer-static/image/art%3A10.1038%2Fnmeth.3252/MediaObjects/41592_2015_Article_BFnmeth3252_Fig2_HTML.jpg?as=webp)

As you can see from the figure, this class stores processed data (`assays`),
metadata (`colData` and `exptData`), and annotation data (`rowData`). The
SummarizedExperiment class is used ubiquitously throughout the Bioconductor
package ecosystem, as are other core classes some of which we will cover later
in this chapter.

## Gene Identifiers

Since the first genetic sequence for a gene was determined in 1965 - [an alanine
tRNA in yeast](https://pubmed.ncbi.nlm.nih.gov/14263761/)[@Holley1965-tu] -
scientists have been trying to give them names. In 1979, [formal guidelines for
the nomenclature used for human genes was
proposed](https://pubmed.ncbi.nlm.nih.gov/295268/)[@Shows1979-iw] along with the
founding of the Human Gene Nomenclature Committee [@Bruford2020-zs], so that
researchers would have a common vocabulary to refer to genes in a consistent
way. In 1989, this committee was placed under the auspices of the Human Genome
Organisation (HUGO), becoming the HUGO Gene Nomenclature Committee (HGNC). The
HGNC has been the official body providing guidelines and gene naming authority
since, and as such HGNC gene names are often called *gene symbols*.

::: {.box .note}
As per @Bruford2020-zs, the official human gene symbol guidelines are as
follows:

1. Each gene is assigned a unique symbol, HGNC ID and descriptive name.
2. Symbols contain only uppercase Latin letters and Arabic numerals.
3. Symbols should not be the same as commonly used abbreviations
4. Nomenclature should not contain reference to any species or “G” for gene.
5. Nomenclature should not be offensive or pejorative.

:::

Gene symbols are the most human-readable system for naming genes. The gene
symbols *BRCA2*, *APOE*, and *ACE2* may be familiar to you as they are involved
in common diseases, namely breast cancer, Alzheimer's Disease risk, and
SARS-COV-2, respectively. Typically, the gene symbol is an acronym that roughly
represents a label of what the gene does (or was originally found to do), e.g.
*APOE* represents the gene [apolipoprotein E](https://www.ensembl.org/Homo_sapiens/Gene/Summary?db=core;g=ENSG00000130203;r=19:44905791-44909393). This convention is
convenient for humans when reading and identifying genes. However,



## BiomaRt

## Gene Expression {#bio-geneexp}

## Differential Expression: Microarrays {#bio-micro}

## Gene Set Enrichment Analysis {#bio-gsea}

## High Throughput Sequencing {#bio-hts}

## Differential Expression: RNASeq {#bio-rnaseq}

## Biological Network Analysis {#bio-networks}

## Single Cell Sequencing Analysis {#bio-scseq}
