<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta name="viewport" content="width=device-width, initial-scale=1" />
<meta property="og:title" content="5.7 Arranging Data | BF591 - R for Biological Sciences" />
<meta property="og:type" content="book" />







<script type="text/x-mathjax-config">
MathJax.Hub.Config({
  TeX: { equationNumbers: { autoNumber: "AMS" } }
});
</script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<meta name="description" content="5.7 Arranging Data | BF591 - R for Biological Sciences">

<title>5.7 Arranging Data | BF591 - R for Biological Sciences</title>

<script src="libs/header-attrs-2.13/header-attrs.js"></script>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="libs/bootstrap-3.3.5/css/darkly.min.css" rel="stylesheet" />
<script src="libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="libs/navigation-1.1/tabsets.js"></script>


<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>


<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />

<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
/* show arrow before summary tag as in bootstrap
TODO: remove if boostrap in updated in html_document (rmarkdown#1485) */
details > summary {
  display: list-item;
  cursor: pointer;
}
</style>
</head>

<body>

<div class="container-fluid main-container">


<div class="row">
<div class="col-sm-12">
<div id="TOC">
<ul>
<li><a href="index.html#syllabus">Syllabus</a>
<ul>
<li><a href="course-schedule.html#course-schedule">Course Schedule</a></li>
<li><a href="instructors.html#instructors">Instructors</a></li>
<li><a href="office-hours.html#office-hours">Office Hours</a></li>
<li><a href="course-values-and-policies.html#course-values-and-policies">Course Values and Policies</a></li>
</ul></li>
<li><a href="introduction.html#introduction"><span class="toc-section-number">1</span> Introduction</a>
<ul>
<li><a href="who-this-book-is-for.html#who-this-book-is-for"><span class="toc-section-number">1.1</span> Who This Book Is For</a></li>
<li><a href="a-note-about-reinventing-the-wheel.html#a-note-about-reinventing-the-wheel"><span class="toc-section-number">1.2</span> A Note About Reinventing the Wheel</a></li>
<li><a href="sources-and-references.html#sources-and-references"><span class="toc-section-number">1.3</span> Sources and References</a></li>
</ul></li>
<li><a href="data-bio.html#data-bio"><span class="toc-section-number">2</span> Data in Biology</a>
<ul>
<li><a href="a-brief-history-of-data-in-molecular-biology.html#a-brief-history-of-data-in-molecular-biology"><span class="toc-section-number">2.1</span> A Brief History of Data in Molecular Biology</a></li>
<li><a href="biology-as-a-mature-data-science.html#biology-as-a-mature-data-science"><span class="toc-section-number">2.2</span> Biology as a Mature Data Science</a></li>
</ul></li>
<li><a href="preliminaries.html#preliminaries"><span class="toc-section-number">3</span> Preliminaries</a>
<ul>
<li><a href="prelim-r.html#prelim-r"><span class="toc-section-number">3.1</span> The R Language</a></li>
<li><a href="prelim-rstudio.html#prelim-rstudio"><span class="toc-section-number">3.2</span> RStudio</a></li>
<li><a href="prog-the-r-script.html#prog-the-r-script"><span class="toc-section-number">3.3</span> The R Script</a></li>
<li><a href="prog-workflow.html#prog-workflow"><span class="toc-section-number">3.4</span> The Scripting Workflow</a></li>
<li><a href="prelim-git.html#prelim-git"><span class="toc-section-number">3.5</span> git + github</a>
<ul>
<li><a href="prelim-git.html#motivation"><span class="toc-section-number">3.5.1</span> Motivation</a></li>
<li><a href="prelim-git.html#git"><span class="toc-section-number">3.5.2</span> git</a></li>
<li><a href="prelim-git.html#git-hosting-platforms-github"><span class="toc-section-number">3.5.3</span> Git hosting platforms (GitHub)</a></li>
</ul></li>
</ul></li>
<li><a href="prog-basics.html#prog-basics"><span class="toc-section-number">4</span> R Programming</a>
<ul>
<li><a href="before-you-begin.html#before-you-begin"><span class="toc-section-number">4.1</span> Before you begin</a></li>
<li><a href="introduction-1.html#introduction-1"><span class="toc-section-number">4.2</span> Introduction</a></li>
<li><a href="prog-r-syntax.html#prog-r-syntax"><span class="toc-section-number">4.3</span> R Syntax Basics</a></li>
<li><a href="prog-types.html#prog-types"><span class="toc-section-number">4.4</span> Basic Types of Values</a>
<ul>
<li><a href="prog-types.html#factors"><span class="toc-section-number">4.4.1</span> Factors</a></li>
</ul></li>
<li><a href="prog-struct.html#prog-struct"><span class="toc-section-number">4.5</span> Data Structures</a>
<ul>
<li><a href="prog-struct.html#vectors"><span class="toc-section-number">4.5.1</span> Vectors</a></li>
<li><a href="prog-struct.html#matrices"><span class="toc-section-number">4.5.2</span> Matrices</a></li>
<li><a href="prog-struct.html#lists-and-data-frames"><span class="toc-section-number">4.5.3</span> Lists and data frames</a></li>
</ul></li>
<li><a href="logical-tests-and-comparators.html#logical-tests-and-comparators"><span class="toc-section-number">4.6</span> Logical Tests and Comparators</a></li>
<li><a href="functions.html#functions"><span class="toc-section-number">4.7</span> Functions</a>
<ul>
<li><a href="functions.html#prog-dry"><span class="toc-section-number">4.7.1</span> DRY: Don’t Repeat Yourself</a></li>
<li><a href="functions.html#writing-your-own-functions"><span class="toc-section-number">4.7.2</span> Writing your own functions</a></li>
<li><a href="functions.html#scope"><span class="toc-section-number">4.7.3</span> Scope</a></li>
</ul></li>
<li><a href="iteration.html#iteration"><span class="toc-section-number">4.8</span> Iteration</a>
<ul>
<li><a href="iteration.html#vectorized-operations"><span class="toc-section-number">4.8.1</span> Vectorized operations</a></li>
<li><a href="iteration.html#functional-programming"><span class="toc-section-number">4.8.2</span> Functional programming</a></li>
<li><a href="iteration.html#apply-and-friends"><span class="toc-section-number">4.8.3</span> <code>apply()</code> and friends</a></li>
</ul></li>
<li><a href="installing-packages.html#installing-packages"><span class="toc-section-number">4.9</span> Installing Packages</a></li>
<li><a href="saving-and-loading-r-data.html#saving-and-loading-r-data"><span class="toc-section-number">4.10</span> Saving and Loading R Data</a></li>
<li><a href="prog-debugging.html#prog-debugging"><span class="toc-section-number">4.11</span> Troubleshooting and Debugging</a></li>
<li><a href="prog-style.html#prog-style"><span class="toc-section-number">4.12</span> Coding Style and Conventions</a>
<ul>
<li><a href="prog-style.html#is-my-code-correct"><span class="toc-section-number">4.12.1</span> Is my code correct?</a></li>
<li><a href="prog-style.html#does-my-code-follow-the-dry-principle"><span class="toc-section-number">4.12.2</span> Does my code follow the <span>DRY principle</span>?</a></li>
<li><a href="prog-style.html#did-i-choose-concise-but-descriptive-variable-and-function-names"><span class="toc-section-number">4.12.3</span> Did I choose concise but descriptive variable and function names?</a></li>
<li><a href="prog-style.html#did-i-use-indentation-and-naming-conventions-consistently-throughout-my-code"><span class="toc-section-number">4.12.4</span> Did I use indentation and naming conventions consistently throughout my code?</a></li>
<li><a href="prog-style.html#did-i-write-comments-especially-when-what-the-code-does-is-not-obvious"><span class="toc-section-number">4.12.5</span> Did I write comments, especially when what the code does is not obvious?</a></li>
<li><a href="prog-style.html#how-easy-would-it-be-for-someone-else-to-understand-my-code"><span class="toc-section-number">4.12.6</span> How easy would it be for someone else to understand my code?</a></li>
<li><a href="prog-style.html#is-my-code-easy-to-maintainchange"><span class="toc-section-number">4.12.7</span> Is my code easy to maintain/change?</a></li>
<li><a href="prog-style.html#the-styler-package"><span class="toc-section-number">4.12.8</span> The <code>styler</code> package</a></li>
</ul></li>
</ul></li>
<li><a href="data-wrangling.html#data-wrangling"><span class="toc-section-number">5</span> Data Wrangling</a>
<ul>
<li><a href="the-tidyverse.html#the-tidyverse"><span class="toc-section-number">5.1</span> The Tidyverse</a></li>
<li><a href="dw-basics.html#dw-basics"><span class="toc-section-number">5.2</span> Tidyverse Basics</a></li>
<li><a href="importing-data.html#importing-data"><span class="toc-section-number">5.3</span> Importing Data</a></li>
<li><a href="data-tibble.html#data-tibble"><span class="toc-section-number">5.4</span> The tibble</a></li>
<li><a href="tidy-data.html#tidy-data"><span class="toc-section-number">5.5</span> Tidy Data</a></li>
<li><a href="pipes.html#pipes"><span class="toc-section-number">5.6</span> pipes</a></li>
<li><a href="arranging-data.html#arranging-data"><span class="toc-section-number">5.7</span> Arranging Data</a>
<ul>
<li><a href="arranging-data.html#dplyrmutate---create-new-columns-using-other-columns"><span class="toc-section-number">5.7.1</span> <code>dplyr::mutate()</code> - Create new columns using other columns</a></li>
<li><a href="arranging-data.html#stringr---working-with-character-values"><span class="toc-section-number">5.7.2</span> <code>stringr</code> - Working with character values</a></li>
<li><a href="arranging-data.html#dplyrselect---subset-columns-by-name"><span class="toc-section-number">5.7.3</span> <code>dplyr::select()</code> - Subset Columns by Name</a></li>
<li><a href="arranging-data.html#dplyrfilter---pick-rows-out-of-a-data-set"><span class="toc-section-number">5.7.4</span> <code>dplyr::filter()</code> - Pick rows out of a data set</a></li>
<li><a href="arranging-data.html#dplyrarrange---order-rows-based-on-their-values"><span class="toc-section-number">5.7.5</span> <code>dplyr::arrange()</code> - Order rows based on their values</a></li>
<li><a href="arranging-data.html#putting-it-all-together"><span class="toc-section-number">5.7.6</span> Putting it all together</a></li>
</ul></li>
<li><a href="grouping-data.html#grouping-data"><span class="toc-section-number">5.8</span> Grouping Data</a></li>
<li><a href="rearranging-data.html#rearranging-data"><span class="toc-section-number">5.9</span> Rearranging Data</a></li>
<li><a href="relational-data.html#relational-data"><span class="toc-section-number">5.10</span> Relational Data</a>
<ul>
<li><a href="relational-data.html#dealing-with-multiple-matches"><span class="toc-section-number">5.10.1</span> Dealing with multiple matches</a></li>
</ul></li>
</ul></li>
<li><a href="data-science.html#data-science"><span class="toc-section-number">6</span> Data Science</a>
<ul>
<li><a href="data-modeling.html#data-modeling"><span class="toc-section-number">6.1</span> Data Modeling</a>
<ul>
<li><a href="data-modeling.html#a-worked-modeling-example"><span class="toc-section-number">6.1.1</span> A Worked Modeling Example</a></li>
<li><a href="data-modeling.html#data-summarization"><span class="toc-section-number">6.1.2</span> Data Summarization</a></li>
<li><a href="data-modeling.html#linear-models"><span class="toc-section-number">6.1.3</span> Linear Models</a></li>
<li><a href="data-modeling.html#flavors-of-linear-models"><span class="toc-section-number">6.1.4</span> Flavors of Linear Models</a></li>
<li><a href="data-modeling.html#assessing-model-accuracy-."><span class="toc-section-number">6.1.5</span> Assessing Model Accuracy .</a></li>
</ul></li>
<li><a href="statistical-distributions.html#statistical-distributions"><span class="toc-section-number">6.2</span> Statistical Distributions</a>
<ul>
<li><a href="statistical-distributions.html#random-variables"><span class="toc-section-number">6.2.1</span> Random Variables</a></li>
<li><a href="statistical-distributions.html#statistical-distribution-basics"><span class="toc-section-number">6.2.2</span> Statistical Distribution Basics</a></li>
<li><a href="statistical-distributions.html#distributions-in-r"><span class="toc-section-number">6.2.3</span> Distributions in R</a></li>
<li><a href="statistical-distributions.html#discrete-distributions"><span class="toc-section-number">6.2.4</span> Discrete Distributions</a></li>
<li><a href="statistical-distributions.html#continuous-distributions"><span class="toc-section-number">6.2.5</span> Continuous Distributions</a></li>
<li><a href="statistical-distributions.html#empirical-distributions"><span class="toc-section-number">6.2.6</span> Empirical Distributions</a></li>
</ul></li>
<li><a href="statistical-tests.html#statistical-tests"><span class="toc-section-number">6.3</span> Statistical Tests</a>
<ul>
<li><a href="statistical-tests.html#what-is-a-statistical-test"><span class="toc-section-number">6.3.1</span> What is a statistical test</a></li>
<li><a href="statistical-tests.html#common-statistical-tests"><span class="toc-section-number">6.3.2</span> Common Statistical Tests</a></li>
<li><a href="statistical-tests.html#choosing-a-test"><span class="toc-section-number">6.3.3</span> Choosing a Test</a></li>
<li><a href="statistical-tests.html#p-values"><span class="toc-section-number">6.3.4</span> P-values</a></li>
<li><a href="statistical-tests.html#multiple-hypothesis-testing-."><span class="toc-section-number">6.3.5</span> Multiple Hypothesis Testing .</a></li>
<li><a href="statistical-tests.html#statistical-power"><span class="toc-section-number">6.3.6</span> Statistical power</a></li>
</ul></li>
<li><a href="exploratory-data-analysis.html#exploratory-data-analysis"><span class="toc-section-number">6.4</span> Exploratory Data Analysis</a>
<ul>
<li><a href="exploratory-data-analysis.html#principal-component-analysis"><span class="toc-section-number">6.4.1</span> Principal Component Analysis</a></li>
<li><a href="exploratory-data-analysis.html#cluster-analysis"><span class="toc-section-number">6.4.2</span> Cluster Analysis</a></li>
<li><a href="exploratory-data-analysis.html#k-means-."><span class="toc-section-number">6.4.3</span> k-means .</a></li>
<li><a href="exploratory-data-analysis.html#others"><span class="toc-section-number">6.4.4</span> Others</a></li>
</ul></li>
<li><a href="network-analysis-..html#network-analysis-."><span class="toc-section-number">6.5</span> Network Analysis .</a>
<ul>
<li><a href="network-analysis-..html#basic-network-analysis"><span class="toc-section-number">6.5.1</span> Basic network analysis</a></li>
<li><a href="network-analysis-..html#create-your-own"><span class="toc-section-number">6.5.2</span> Create your own</a></li>
</ul></li>
</ul></li>
<li><a href="data-visualization.html#data-visualization"><span class="toc-section-number">7</span> Data Visualization</a>
<ul>
<li><a href="responsible-plotting.html#responsible-plotting"><span class="toc-section-number">7.1</span> Responsible Plotting</a>
<ul>
<li><a href="responsible-plotting.html#visualization-principles"><span class="toc-section-number">7.1.1</span> Visualization Principles</a></li>
<li><a href="responsible-plotting.html#human-visual-perception"><span class="toc-section-number">7.1.2</span> Human Visual Perception</a></li>
<li><a href="responsible-plotting.html#visual-encodings"><span class="toc-section-number">7.1.3</span> Visual Encodings</a></li>
<li><a href="responsible-plotting.html#some-opinionated-rules-of-thumb"><span class="toc-section-number">7.1.4</span> Some Opinionated Rules of Thumb</a></li>
</ul></li>
<li><a href="dv-gg.html#dv-gg"><span class="toc-section-number">7.2</span> Grammar of Graphics</a>
<ul>
<li><a href="dv-gg.html#ggplot-mechanics"><span class="toc-section-number">7.2.1</span> <code>ggplot</code> mechanics</a></li>
</ul></li>
<li><a href="plotting-one-dimension.html#plotting-one-dimension"><span class="toc-section-number">7.3</span> Plotting One Dimension</a>
<ul>
<li><a href="plotting-one-dimension.html#dv-bar"><span class="toc-section-number">7.3.1</span> Bar chart</a></li>
<li><a href="plotting-one-dimension.html#lollipop-plots"><span class="toc-section-number">7.3.2</span> Lollipop plots</a></li>
<li><a href="plotting-one-dimension.html#stacked-area-charts"><span class="toc-section-number">7.3.3</span> Stacked Area charts</a></li>
<li><a href="plotting-one-dimension.html#parallel-coordinate-plots"><span class="toc-section-number">7.3.4</span> Parallel Coordinate plots</a></li>
</ul></li>
<li><a href="visualizing-distributions.html#visualizing-distributions"><span class="toc-section-number">7.4</span> Visualizing Distributions</a>
<ul>
<li><a href="visualizing-distributions.html#histogram"><span class="toc-section-number">7.4.1</span> Histogram</a></li>
<li><a href="visualizing-distributions.html#density"><span class="toc-section-number">7.4.2</span> Density</a></li>
<li><a href="visualizing-distributions.html#boxplot"><span class="toc-section-number">7.4.3</span> Boxplot</a></li>
<li><a href="visualizing-distributions.html#violin-plot"><span class="toc-section-number">7.4.4</span> Violin plot</a></li>
<li><a href="visualizing-distributions.html#beeswarm-plot"><span class="toc-section-number">7.4.5</span> Beeswarm plot</a></li>
<li><a href="visualizing-distributions.html#ridgeline"><span class="toc-section-number">7.4.6</span> Ridgeline</a></li>
</ul></li>
<li><a href="plotting-two-or-more-dimensions.html#plotting-two-or-more-dimensions"><span class="toc-section-number">7.5</span> Plotting Two or More Dimensions</a>
<ul>
<li><a href="plotting-two-or-more-dimensions.html#scatter-plots"><span class="toc-section-number">7.5.1</span> Scatter Plots</a></li>
<li><a href="plotting-two-or-more-dimensions.html#bubble-plots"><span class="toc-section-number">7.5.2</span> Bubble Plots</a></li>
<li><a href="plotting-two-or-more-dimensions.html#connected-scatter-plots"><span class="toc-section-number">7.5.3</span> Connected Scatter Plots</a></li>
<li><a href="plotting-two-or-more-dimensions.html#line-plots"><span class="toc-section-number">7.5.4</span> Line Plots</a></li>
<li><a href="plotting-two-or-more-dimensions.html#parallel-coordinate-plots-1"><span class="toc-section-number">7.5.5</span> Parallel Coordinate Plots</a></li>
<li><a href="plotting-two-or-more-dimensions.html#heatmaps"><span class="toc-section-number">7.5.6</span> Heatmaps</a></li>
<li><a href="plotting-two-or-more-dimensions.html#how-to-use-heatmaps-responsibly"><span class="toc-section-number">7.5.7</span> How To Use Heatmaps Responsibly</a></li>
</ul></li>
<li><a href="other-kind-of-plots.html#other-kind-of-plots"><span class="toc-section-number">7.6</span> Other Kind of Plots</a>
<ul>
<li><a href="other-kind-of-plots.html#dendrograms"><span class="toc-section-number">7.6.1</span> Dendrograms</a></li>
<li><a href="other-kind-of-plots.html#chord-diagrams-and-circos-plots"><span class="toc-section-number">7.6.2</span> Chord Diagrams and Circos Plots</a></li>
</ul></li>
<li><a href="multiple-plots.html#multiple-plots"><span class="toc-section-number">7.7</span> Multiple Plots</a>
<ul>
<li><a href="multiple-plots.html#facet-wrapping"><span class="toc-section-number">7.7.1</span> Facet wrapping</a></li>
<li><a href="multiple-plots.html#multipanel-figures"><span class="toc-section-number">7.7.2</span> Multipanel Figures</a></li>
</ul></li>
<li><a href="publication-ready-plots.html#publication-ready-plots"><span class="toc-section-number">7.8</span> Publication Ready Plots</a>
<ul>
<li><a href="publication-ready-plots.html#ggplot-themes"><span class="toc-section-number">7.8.1</span> <code>ggplot</code> Themes</a></li>
<li><a href="publication-ready-plots.html#exporting-to-svg"><span class="toc-section-number">7.8.2</span> Exporting to SVG</a></li>
</ul></li>
<li><a href="network-visualization.html#network-visualization"><span class="toc-section-number">7.9</span> Network visualization</a>
<ul>
<li><a href="network-visualization.html#interactive-network-plots"><span class="toc-section-number">7.9.1</span> Interactive network plots</a></li>
</ul></li>
</ul></li>
<li><a href="biology-bioinformatics.html#biology-bioinformatics"><span class="toc-section-number">8</span> Biology &amp; Bioinformatics</a>
<ul>
<li><a href="bio-r.html#bio-r"><span class="toc-section-number">8.1</span> R in Biology</a></li>
<li><a href="biological-data-overview.html#biological-data-overview"><span class="toc-section-number">8.2</span> Biological Data Overview</a>
<ul>
<li><a href="biological-data-overview.html#bio-data-types"><span class="toc-section-number">8.2.1</span> Types of Biological Data</a></li>
<li><a href="biological-data-overview.html#csv-files"><span class="toc-section-number">8.2.2</span> CSV Files</a></li>
<li><a href="biological-data-overview.html#biological-data-is-not-tidy"><span class="toc-section-number">8.2.3</span> Biological data is NOT Tidy!</a></li>
</ul></li>
<li><a href="bio-bioconductor.html#bio-bioconductor"><span class="toc-section-number">8.3</span> Bioconductor</a></li>
<li><a href="microarrays.html#microarrays"><span class="toc-section-number">8.4</span> Microarrays</a></li>
<li><a href="high-throughput-sequencing.html#high-throughput-sequencing"><span class="toc-section-number">8.5</span> High Throughput Sequencing</a>
<ul>
<li><a href="high-throughput-sequencing.html#raw-hts-data"><span class="toc-section-number">8.5.1</span> Raw HTS Data</a></li>
<li><a href="high-throughput-sequencing.html#preliminary-hts-data-analysis"><span class="toc-section-number">8.5.2</span> Preliminary HTS Data Analysis</a></li>
<li><a href="high-throughput-sequencing.html#count-data"><span class="toc-section-number">8.5.3</span> Count Data</a></li>
</ul></li>
<li><a href="gene-identifiers.html#gene-identifiers"><span class="toc-section-number">8.6</span> Gene Identifiers</a>
<ul>
<li><a href="gene-identifiers.html#gene-identifier-systems"><span class="toc-section-number">8.6.1</span> Gene Identifier Systems</a></li>
<li><a href="gene-identifiers.html#mapping-between-identifier-systems"><span class="toc-section-number">8.6.2</span> Mapping Between Identifier Systems</a></li>
<li><a href="gene-identifiers.html#mapping-homologs"><span class="toc-section-number">8.6.3</span> Mapping Homologs</a></li>
</ul></li>
<li><a href="gene-expression.html#gene-expression"><span class="toc-section-number">8.7</span> Gene Expression</a>
<ul>
<li><a href="gene-expression.html#gene-expression-data-in-bioconductor"><span class="toc-section-number">8.7.1</span> Gene Expression Data in Bioconductor</a></li>
<li><a href="gene-expression.html#differential-expression-analysis"><span class="toc-section-number">8.7.2</span> Differential Expression Analysis</a></li>
<li><a href="gene-expression.html#microarray-gene-expression-data"><span class="toc-section-number">8.7.3</span> Microarray Gene Expression Data</a></li>
<li><a href="gene-expression.html#differential-expression-microarrays-limma"><span class="toc-section-number">8.7.4</span> Differential Expression: Microarrays (limma)</a></li>
<li><a href="gene-expression.html#rnaseq"><span class="toc-section-number">8.7.5</span> RNASeq</a></li>
<li><a href="gene-expression.html#rnaseq-gene-expression-data"><span class="toc-section-number">8.7.6</span> RNASeq Gene Expression Data</a></li>
<li><a href="gene-expression.html#filtering-counts"><span class="toc-section-number">8.7.7</span> Filtering Counts</a></li>
<li><a href="gene-expression.html#count-distributions"><span class="toc-section-number">8.7.8</span> Count Distributions</a></li>
<li><a href="gene-expression.html#differential-expression-rnaseq"><span class="toc-section-number">8.7.9</span> Differential Expression: RNASeq</a></li>
</ul></li>
<li><a href="gene-set-enrichment-analysis.html#gene-set-enrichment-analysis"><span class="toc-section-number">8.8</span> Gene Set Enrichment Analysis</a>
<ul>
<li><a href="gene-set-enrichment-analysis.html#gene-sets"><span class="toc-section-number">8.8.1</span> Gene Sets</a></li>
<li><a href="gene-set-enrichment-analysis.html#working-with-gene-sets-in-r"><span class="toc-section-number">8.8.2</span> Working with gene sets in R</a></li>
<li><a href="gene-set-enrichment-analysis.html#over-representation-analysis"><span class="toc-section-number">8.8.3</span> Over-representation Analysis</a></li>
<li><a href="gene-set-enrichment-analysis.html#rank-based-analysis"><span class="toc-section-number">8.8.4</span> Rank-based Analysis</a></li>
<li><a href="gene-set-enrichment-analysis.html#fgsea"><span class="toc-section-number">8.8.5</span> fgsea</a></li>
</ul></li>
<li><a href="biological-networks-..html#biological-networks-."><span class="toc-section-number">8.9</span> Biological Networks .</a>
<ul>
<li><a href="biological-networks-..html#biological-pathways-."><span class="toc-section-number">8.9.1</span> Biological Pathways .</a></li>
<li><a href="biological-networks-..html#gene-regulatory-networks-."><span class="toc-section-number">8.9.2</span> Gene Regulatory Networks .</a></li>
<li><a href="biological-networks-..html#protein-protein-interaction-network-."><span class="toc-section-number">8.9.3</span> Protein-Protein Interaction Network .</a></li>
<li><a href="biological-networks-..html#wgcna-."><span class="toc-section-number">8.9.4</span> WGCNA .</a></li>
</ul></li>
</ul></li>
<li><a href="engineering.html#engineering"><span class="toc-section-number">9</span> EngineeRing</a>
<ul>
<li><a href="unit-testing.html#unit-testing"><span class="toc-section-number">9.1</span> Unit Testing</a></li>
<li><a href="toolification.html#toolification"><span class="toc-section-number">9.2</span> Toolification</a>
<ul>
<li><a href="toolification.html#the-r-interpreter"><span class="toc-section-number">9.2.1</span> The R Interpreter</a></li>
<li><a href="toolification.html#rscript"><span class="toc-section-number">9.2.2</span> <code>Rscript</code></a></li>
<li><a href="toolification.html#commandargs"><span class="toc-section-number">9.2.3</span> <code>commandArgs()</code></a></li>
</ul></li>
<li><a href="parallel-processing.html#parallel-processing"><span class="toc-section-number">9.3</span> Parallel Processing</a>
<ul>
<li><a href="parallel-processing.html#brief-introduction-to-parallelization"><span class="toc-section-number">9.3.1</span> Brief Introduction to Parallelization</a></li>
<li><a href="parallel-processing.html#apply-and-friends-are-pleasingly-parallel"><span class="toc-section-number">9.3.2</span> <code>apply</code> and Friends Are Pleasingly Parallel</a></li>
<li><a href="parallel-processing.html#the-parallel-package"><span class="toc-section-number">9.3.3</span> The <code>parallel</code> package</a></li>
</ul></li>
<li><a href="object-orientation-in-r.html#object-orientation-in-r"><span class="toc-section-number">9.4</span> Object Orientation in R</a></li>
<li><a href="r-packages.html#r-packages"><span class="toc-section-number">9.5</span> R Packages</a></li>
</ul></li>
<li><a href="rshiny.html#rshiny"><span class="toc-section-number">10</span> RShiny</a>
<ul>
<li><a href="rshiny-introduction.html#rshiny-introduction"><span class="toc-section-number">10.1</span> RShiny Introduction</a>
<ul>
<li><a href="rshiny-introduction.html#how-does-a-web-page-work"><span class="toc-section-number">10.1.1</span> How does a web page work?</a></li>
</ul></li>
<li><a href="application-structure.html#application-structure"><span class="toc-section-number">10.2</span> Application Structure</a>
<ul>
<li><a href="application-structure.html#ui-design"><span class="toc-section-number">10.2.1</span> UI Design</a></li>
<li><a href="application-structure.html#server-functionality"><span class="toc-section-number">10.2.2</span> Server Functionality</a></li>
</ul></li>
<li><a href="reactivity.html#reactivity"><span class="toc-section-number">10.3</span> Reactivity</a></li>
<li><a href="publishing.html#publishing"><span class="toc-section-number">10.4</span> Publishing</a></li>
</ul></li>
<li><a href="communicating-with-r.html#communicating-with-r"><span class="toc-section-number">11</span> Communicating with R</a>
<ul>
<li><a href="rmarkdown-knitr.html#rmarkdown-knitr"><span class="toc-section-number">11.1</span> RMarkdown &amp; knitr</a></li>
<li><a href="bookdown.html#bookdown"><span class="toc-section-number">11.2</span> bookdown</a></li>
</ul></li>
<li><a href="contribution-guide.html#contribution-guide"><span class="toc-section-number">12</span> Contribution Guide</a>
<ul>
<li><a href="custom-blocks.html#custom-blocks"><span class="toc-section-number">12.1</span> Custom blocks</a></li>
</ul></li>
<li><a href="assignments.html#assignments">Assignments</a>
<ul>
<li><a href="assignment-format.html#assignment-format">Assignment Format</a></li>
<li><a href="starting-an-assignment.html#starting-an-assignment">Starting an Assignment</a>
<ul>
<li><a href="starting-an-assignment.html#starting-your-own-git-repository-and-cloning-it">Starting your own <code>git</code> repository and cloning it</a></li>
<li><a href="starting-an-assignment.html#contents">Contents</a></li>
<li><a href="starting-an-assignment.html#github-tutorial">GitHub tutorial</a></li>
<li><a href="starting-an-assignment.html#committing-and-pushing-with-r-and-without">Committing and Pushing with R (and without)</a></li>
<li><a href="starting-an-assignment.html#using-rstudio">Using RStudio</a></li>
<li><a href="starting-an-assignment.html#using-the-command-line">Using the command line</a></li>
</ul></li>
<li><a href="assignment-1.html#assignment-1">Assignment 1</a>
<ul>
<li><a href="assignment-1.html#problem-statement">Problem Statement</a></li>
<li><a href="assignment-1.html#learning-objectives-and-skill-list">Learning Objectives and Skill List</a></li>
<li><a href="assignment-1.html#instructions">Instructions</a></li>
<li><a href="assignment-1.html#hints">Hints</a></li>
</ul></li>
<li><a href="assignment-2.html#assignment-2">Assignment 2</a>
<ul>
<li><a href="assignment-2.html#problem-statement-1">Problem Statement</a></li>
<li><a href="assignment-2.html#learning-objectives">Learning Objectives</a></li>
<li><a href="assignment-2.html#skill-list">Skill List</a></li>
<li><a href="assignment-2.html#instructions-1">Instructions</a></li>
<li><a href="assignment-2.html#function-details">Function Details</a></li>
</ul></li>
<li><a href="assignment-3.html#assignment-3">Assignment 3</a>
<ul>
<li><a href="assignment-3.html#problem-statement-2">Problem Statement</a></li>
<li><a href="assignment-3.html#learning-objectives-1">Learning Objectives</a></li>
<li><a href="assignment-3.html#skill-list-1">Skill List</a></li>
<li><a href="assignment-3.html#background-on-microarrays">Background on Microarrays</a></li>
<li><a href="assignment-3.html#background-on-principal-component-analysis">Background on Principal Component Analysis</a></li>
<li><a href="assignment-3.html#marisa-et-al.-gene-expression-classification-of-colon-cancer-into-molecular-subtypes-characterization-validation-and-prognostic-value.-plos-medicine-may-2013.-pmid-23700391">Marisa et al. Gene Expression Classification of Colon Cancer into Molecular Subtypes: Characterization, Validation, and Prognostic Value. PLoS Medicine, May 2013. PMID: 23700391</a></li>
<li><a href="assignment-3.html#scaling-data-using-r-scale">Scaling data using R <code>scale()</code></a></li>
<li><a href="assignment-3.html#proportion-of-variance-explained">Proportion of variance explained</a></li>
<li><a href="assignment-3.html#plotting-and-visualization-of-pca">Plotting and visualization of PCA</a></li>
<li><a href="assignment-3.html#hierarchical-clustering-and-heatmaps">Hierarchical Clustering and Heatmaps</a></li>
<li><a href="assignment-3.html#references">References</a></li>
</ul></li>
<li><a href="assignment-4.html#assignment-4">Assignment 4</a>
<ul>
<li><a href="assignment-4.html#problem-statement-3">Problem Statement</a></li>
<li><a href="assignment-4.html#learning-objectives-2">Learning Objectives</a></li>
<li><a href="assignment-4.html#skill-list-2">Skill List</a></li>
<li><a href="assignment-4.html#instructions-2">Instructions</a></li>
<li><a href="assignment-4.html#deliverables">Deliverables</a></li>
<li><a href="assignment-4.html#function-details-1">Function Details</a></li>
<li><a href="assignment-4.html#hints-1">Hints</a></li>
</ul></li>
<li><a href="assignment-5.html#assignment-5">Assignment 5</a>
<ul>
<li><a href="assignment-5.html#problem-statement-4">Problem Statement</a></li>
<li><a href="assignment-5.html#learning-objectives-3">Learning Objectives</a></li>
<li><a href="assignment-5.html#skill-list-3">Skill List</a></li>
<li><a href="assignment-5.html#deseq2-background">DESeq2 Background</a></li>
<li><a href="assignment-5.html#generating-a-counts-matrix">Generating a counts matrix</a></li>
<li><a href="assignment-5.html#prefiltering-counts-matrix">Prefiltering Counts matrix</a></li>
<li><a href="assignment-5.html#median-of-ratios-normalization">Median-of-ratios normalization</a></li>
<li><a href="assignment-5.html#deseq2-preparation">DESeq2 preparation</a></li>
<li><a href="assignment-5.html#omeara-et-al.-transcriptional-reversion-of-cardiac-myocyte-fate-during-mammalian-cardiac-regeneration.-circ-res.-feb-2015.-pmid-25477501l">O’Meara et al. Transcriptional Reversion of Cardiac Myocyte Fate During Mammalian Cardiac Regeneration. Circ Res. Feb 2015. PMID: 25477501l</a></li>
<li><a href="assignment-5.html#reading-and-subsetting-the-data-from-verse_counts.tsv-and-sample_metadata.csv">1. Reading and subsetting the data from verse_counts.tsv and sample_metadata.csv</a></li>
<li><a href="assignment-5.html#running-deseq2">2. Running DESeq2</a></li>
<li><a href="assignment-5.html#annotating-results-to-construct-a-labeled-volcano-plot">3. Annotating results to construct a labeled volcano plot</a></li>
<li><a href="assignment-5.html#diagnostic-plot-of-the-raw-p-values-for-all-genes">4. Diagnostic plot of the raw p-values for all genes</a></li>
<li><a href="assignment-5.html#plotting-the-logfoldchanges-for-differentially-expressed-genes">5. Plotting the LogFoldChanges for differentially expressed genes</a></li>
<li><a href="assignment-5.html#the-choice-of-fdr-cutoff-depends-on-cost">The choice of FDR cutoff depends on cost</a></li>
<li><a href="assignment-5.html#plotting-the-normalized-counts-of-differentially-expressed-genes">6. Plotting the normalized counts of differentially expressed genes</a></li>
<li><a href="assignment-5.html#volcano-plot-to-visualize-differential-expression-results">7. Volcano Plot to visualize differential expression results</a></li>
<li><a href="assignment-5.html#running-fgsea-vignette">8. Running fgsea <span>vignette</span></a></li>
<li><a href="assignment-5.html#plotting-the-top-ten-positive-nes-and-top-ten-negative-nes-pathways">9. Plotting the top ten positive NES and top ten negative NES pathways</a></li>
<li><a href="assignment-5.html#references-1">References</a></li>
</ul></li>
<li><a href="assignment-6.html#assignment-6">Assignment 6</a>
<ul>
<li><a href="assignment-6.html#problem-statement-5">Problem Statement</a></li>
<li><a href="assignment-6.html#learning-objectives-4">Learning Objectives</a></li>
<li><a href="assignment-6.html#skill-list-4">Skill List</a></li>
<li><a href="assignment-6.html#instructions-3">Instructions</a></li>
<li><a href="assignment-6.html#function-details-2">Function Details</a></li>
</ul></li>
<li><a href="assignment-7.html#assignment-7">Assignment 7</a>
<ul>
<li><a href="assignment-7.html#problem-statement-6">Problem Statement</a></li>
<li><a href="assignment-7.html#learning-objectives-5">Learning Objectives</a></li>
<li><a href="assignment-7.html#skill-list-5">Skill List</a></li>
<li><a href="assignment-7.html#instructions-4">Instructions</a></li>
<li><a href="assignment-7.html#function-details-3">“Function” Details</a></li>
</ul></li>
</ul></li>
<li class="appendix"><span><b>Appendix</b></span></li>
<li><a href="class-outlines.html#class-outlines"><span class="toc-section-number">A</span> Class Outlines</a>
<ul>
<li><a href="week-1.html#week-1"><span class="toc-section-number">A.1</span> Week 1</a></li>
<li><a href="week-2.html#week-2"><span class="toc-section-number">A.2</span> Week 2</a></li>
<li><a href="week-3.html#week-3"><span class="toc-section-number">A.3</span> Week 3</a></li>
<li><a href="week-4.html#week-4"><span class="toc-section-number">A.4</span> Week 4</a></li>
<li><a href="week-5.html#week-5"><span class="toc-section-number">A.5</span> Week 5</a></li>
<li><a href="week-6.html#week-6"><span class="toc-section-number">A.6</span> Week 6</a></li>
<li><a href="week-7.html#week-7"><span class="toc-section-number">A.7</span> Week 7</a></li>
<li><a href="week-8.html#week-8"><span class="toc-section-number">A.8</span> Week 8</a></li>
<li><a href="week-9.html#week-9"><span class="toc-section-number">A.9</span> Week 9</a></li>
<li><a href="week-10.html#week-10"><span class="toc-section-number">A.10</span> Week 10</a></li>
<li><a href="week-11.html#week-11"><span class="toc-section-number">A.11</span> Week 11</a></li>
<li><a href="week-12.html#week-12"><span class="toc-section-number">A.12</span> Week 12</a></li>
<li><a href="week-13.html#week-13"><span class="toc-section-number">A.13</span> Week 13</a></li>
<li><a href="week-14.html#week-14"><span class="toc-section-number">A.14</span> Week 14</a></li>
</ul></li>
</ul>
</div>
</div>
</div>
<div class="row">
<div class="col-sm-12">
<div id="arranging-data" class="section level2" number="5.7">
<h2><span class="header-section-number">5.7</span> Arranging Data</h2>
<p>After we have loaded our data from a file into a <code>tibble</code>, we often need to
manipulate it in various ways to make the values amenable to our desired
analysis. Such manipulations might include renaming poorly named columns,
filtering out certain records, deriving new columns using the values in others,
changing the order of rows etc. These operations may collectively be termed
<em>arranging</em> the data and many are provided in the
*<a href="https://dplyr.tidyverse.org/"><code>dplyr</code> package</a>. We will cover some of the most
common data arranging functions here, but there are many more in the <code>dplyr</code>
package worth knowing.</p>
<p>In the examples below, we will make use of the following made-up tibble that
contains fake statistics and p-values for three human genes:</p>
<div class="sourceCode" id="cb74"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb74-1"><a href="arranging-data.html#cb74-1" aria-hidden="true" tabindex="-1"></a>gene_stats</span></code></pre></div>
<pre><code>## # A tibble: 3 x 5
##   gene  test1_stat      test1_p test2_stat  test2_p
##   &lt;chr&gt;      &lt;dbl&gt;        &lt;dbl&gt;      &lt;dbl&gt;    &lt;dbl&gt;
## 1 apoe       12.5  0.103            34.2   0.000013
## 2 hoxd1       4.40 0.632            16.3   0.0421  
## 3 snca       45.7  0.0000000042      0.757 0.915</code></pre>
<p>The <code>gene_stats</code> tibble above is a simple example of a very common type of
data we work with in biology; namely instead of raw data, we work with
statistics that have been computed using raw data that help us interpret the
results. While these statistics may not be ‘data’ per se, we can still use all
the functions and strategies in the tidyverse to work with them.</p>
<div class="box note">
<p>The tidyverse is a very big place. RStudio created many helpful
<a href="https://github.com/rstudio/cheatsheets">cheatsheets</a> to aid in looking up how
do to certain operations. The <a href="https://github.com/rstudio/cheatsheets/blob/main/data-transformation.pdf">cheatsheet on
dplyr</a>
has lots of useful information on how to use the many functions in the package.</p>
</div>
<div id="dplyrmutate---create-new-columns-using-other-columns" class="section level3" number="5.7.1">
<h3><span class="header-section-number">5.7.1</span> <code>dplyr::mutate()</code> - Create new columns using other columns</h3>
<p>Many biological analysis procedures perform some kind of statistical test on a
collection of features (e.g. genes) and produce p-values that indicate how
“surprising” each feature is according to the test. The p-values in our tibble
are nominal, i.e. they have not been adjusted for <a href="#ds-multiple-hypotheses">multiple
hypotheses</a>. Briefly, when we run multiple tests like
we are on each of our three genes, there is a chance that some of the tests will
have a small p-value simply by chance. <a href="https://en.wikipedia.org/wiki/Multiple_comparisons_problem">Multiple testing
correction</a>
procedures adjust nominal p-values to account for this possibility in a number
of different ways, but the most common procedure in biological analysis is the
<a href="https://en.wikipedia.org/wiki/False_discovery_rate">Benjamini-Hochberg or False Discovery Rate
(FDR)</a> procedure. In R, the
<code>p.adjust</code> function can perform several of these procedures, including FDR:</p>
<div class="sourceCode" id="cb76"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb76-1"><a href="arranging-data.html#cb76-1" aria-hidden="true" tabindex="-1"></a>dplyr<span class="sc">::</span><span class="fu">mutate</span>(gene_stats,</span>
<span id="cb76-2"><a href="arranging-data.html#cb76-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">test1_padj=</span><span class="fu">p.adjust</span>(test1_p,<span class="at">method=</span><span class="st">&quot;fdr&quot;</span>)</span>
<span id="cb76-3"><a href="arranging-data.html#cb76-3" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div>
<pre><code>## # A tibble: 3 x 6
##   gene  test1_stat      test1_p test2_stat  test2_p   test1_padj
##   &lt;chr&gt;      &lt;dbl&gt;        &lt;dbl&gt;      &lt;dbl&gt;    &lt;dbl&gt;        &lt;dbl&gt;
## 1 apoe       12.5  0.103            34.2   0.000013 0.155       
## 2 hoxd1       4.40 0.632            16.3   0.0421   0.632       
## 3 snca       45.7  0.0000000042      0.757 0.915    0.0000000126</code></pre>
<p>Notice how the adjusted p-values are larger than the nominal ones; this is the
effect of the multiple testing procedure. Since we have two sets of p-values,
we must compute the FDR on each of them, which we can do in the same call to
<code>mutate()</code>:</p>
<div class="sourceCode" id="cb78"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb78-1"><a href="arranging-data.html#cb78-1" aria-hidden="true" tabindex="-1"></a>gene_stats_mutated <span class="ot">&lt;-</span> dplyr<span class="sc">::</span><span class="fu">mutate</span>(gene_stats,</span>
<span id="cb78-2"><a href="arranging-data.html#cb78-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">test1_padj=</span><span class="fu">p.adjust</span>(test1_p,<span class="at">method=</span><span class="st">&quot;fdr&quot;</span>),</span>
<span id="cb78-3"><a href="arranging-data.html#cb78-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">test2_padj=</span><span class="fu">p.adjust</span>(test2_p,<span class="at">method=</span><span class="st">&quot;fdr&quot;</span>)</span>
<span id="cb78-4"><a href="arranging-data.html#cb78-4" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb78-5"><a href="arranging-data.html#cb78-5" aria-hidden="true" tabindex="-1"></a>gene_stats_mutated</span></code></pre></div>
<pre><code>## # A tibble: 3 x 7
##   gene  test1_stat      test1_p test2_stat  test2_p   test1_padj test2_padj
##   &lt;chr&gt;      &lt;dbl&gt;        &lt;dbl&gt;      &lt;dbl&gt;    &lt;dbl&gt;        &lt;dbl&gt;      &lt;dbl&gt;
## 1 apoe       12.5  0.103            34.2   0.000013 0.155          0.000039
## 2 hoxd1       4.40 0.632            16.3   0.0421   0.632          0.0632  
## 3 snca       45.7  0.0000000042      0.757 0.915    0.0000000126   0.915</code></pre>
<p>Another common operation is to create new columns derived from the values in
multiple other columns. We (or our wetlab colleagues) might decide it is
convenient to have a new column with <code>TRUE</code> or <code>FALSE</code> based on whether the
gene was significant in either test. Such a column would make it easy to filter
genes down to just ones that might be interesting in tools like Excel. We can
create new columns from multiple columns just as easily using the <code>mutate()</code>
function:</p>
<div class="sourceCode" id="cb80"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb80-1"><a href="arranging-data.html#cb80-1" aria-hidden="true" tabindex="-1"></a>dplyr<span class="sc">::</span><span class="fu">mutate</span>(gene_stats_mutated,</span>
<span id="cb80-2"><a href="arranging-data.html#cb80-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">signif_either=</span>(test1_padj <span class="sc">&lt;</span> <span class="fl">0.05</span> <span class="sc">|</span> test2_padj <span class="sc">&lt;</span> <span class="fl">0.05</span>),</span>
<span id="cb80-3"><a href="arranging-data.html#cb80-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">signif_both=</span>(test1_padj <span class="sc">&lt;</span> <span class="fl">0.05</span> <span class="sc">&amp;</span> test2_padj <span class="sc">&lt;</span> <span class="fl">0.05</span>)</span>
<span id="cb80-4"><a href="arranging-data.html#cb80-4" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div>
<pre><code>## # A tibble: 3 x 9
##   gene  test1_stat      test1_p test2_stat  test2_p   test1_padj test2_padj
##   &lt;chr&gt;      &lt;dbl&gt;        &lt;dbl&gt;      &lt;dbl&gt;    &lt;dbl&gt;        &lt;dbl&gt;      &lt;dbl&gt;
## 1 apoe       12.5  0.103            34.2   0.000013 0.155          0.000039
## 2 hoxd1       4.40 0.632            16.3   0.0421   0.632          0.0632  
## 3 snca       45.7  0.0000000042      0.757 0.915    0.0000000126   0.915   
## # ... with 2 more variables: signif_either &lt;lgl&gt;, signif_both &lt;lgl&gt;</code></pre>
<p>Recall that the <code>|</code> and <code>&amp;</code> operators execute ‘or’ and ‘and’ logic,
respectively. The above example required the creation of a new variable
<code>gene_stats_mutated</code> because the columns <code>test1_padj</code> and
<code>test2_padj</code> need to be in the tibble before computing the new fields. However,
in <code>mutate()</code>, columns created first in the function call are available to later
columns. In the following example, note that <code>test1_padj</code> is created first and
then used to create the <code>signif</code> columns:</p>
<div class="sourceCode" id="cb82"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb82-1"><a href="arranging-data.html#cb82-1" aria-hidden="true" tabindex="-1"></a>dplyr<span class="sc">::</span><span class="fu">mutate</span>(gene_stats,</span>
<span id="cb82-2"><a href="arranging-data.html#cb82-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">test1_padj=</span><span class="fu">p.adjust</span>(test1_p,<span class="at">method=</span><span class="st">&quot;fdr&quot;</span>), <span class="co"># test1_padj created</span></span>
<span id="cb82-3"><a href="arranging-data.html#cb82-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">test2_padj=</span><span class="fu">p.adjust</span>(test2_p,<span class="at">method=</span><span class="st">&quot;fdr&quot;</span>),</span>
<span id="cb82-4"><a href="arranging-data.html#cb82-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">signif_either=</span>(test1_padj <span class="sc">&lt;</span> <span class="fl">0.05</span> <span class="sc">|</span> test2_padj <span class="sc">&lt;</span> <span class="fl">0.05</span>), <span class="co">#test1_padj used</span></span>
<span id="cb82-5"><a href="arranging-data.html#cb82-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">signif_both=</span>(test1_padj <span class="sc">&lt;</span> <span class="fl">0.05</span> <span class="sc">&amp;</span> test2_padj <span class="sc">&lt;</span> <span class="fl">0.05</span>)</span>
<span id="cb82-6"><a href="arranging-data.html#cb82-6" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div>
<pre><code>## # A tibble: 3 x 9
##   gene  test1_stat      test1_p test2_stat  test2_p   test1_padj test2_padj
##   &lt;chr&gt;      &lt;dbl&gt;        &lt;dbl&gt;      &lt;dbl&gt;    &lt;dbl&gt;        &lt;dbl&gt;      &lt;dbl&gt;
## 1 apoe       12.5  0.103            34.2   0.000013 0.155          0.000039
## 2 hoxd1       4.40 0.632            16.3   0.0421   0.632          0.0632  
## 3 snca       45.7  0.0000000042      0.757 0.915    0.0000000126   0.915   
## # ... with 2 more variables: signif_either &lt;lgl&gt;, signif_both &lt;lgl&gt;</code></pre>
<p>The alternative would be to split this into two <code>mutate()</code> commands, the first
creating the adjusted p-value columns and the second creating the significance
columns. <code>dplyr</code> recognizes how common it is to build new variables off of other
new variables in a <code>mutate()</code> command, and therefore provides this convenient
behavior.</p>
<p><code>mutate()</code> can also be used to modify columns in place. The official convention
for human gene symbols is that they are upper case, but for some reason our
tibble contains lower case gene symbols. We can correct this using <code>mutate()</code>
but first we should talk about the <a href="https://stringr.tidyverse.org/">stringr
package</a> which makes working with strings much
easier than with base R functions.</p>
<div class="box readmore">
<ul>
<li><a href="https://r4ds.had.co.nz/transform.html#add-new-variables-with-mutate">R for Data Science - Add new variables with <code>mutate()</code></a></li>
<li><a href="https://dplyr.tidyverse.org/reference/mutate.html">dplyr <code>mutate()</code> reference</a></li>
</ul>
</div>
</div>
<div id="stringr---working-with-character-values" class="section level3" number="5.7.2">
<h3><span class="header-section-number">5.7.2</span> <code>stringr</code> - Working with character values</h3>
<p>Base R does not have very convenient functions for working with character
strings (or just <em>strings</em>). This is due to its original intent a statistical
programming language, where string manipulation is not (in principle) a common
operation. However, in practice, we must frequently manipulate strings while
loading, cleaning, and analyzing datasets. The <a href="https://stringr.tidyverse.org/">stringr
package</a> aims to make working with strings “as
easy as possible.”</p>
<p>The package includes many useful functions for operating on strings, including
searching for patterns, mutating strings, lexicographical sorting, combining
multiple strings together (i.e. concatenation), and performing complex
search/replace operations. There are far too many useful functions to cover here
and you should become comfortable reading the <a href="https://stringr.tidyverse.org/reference/index.html">stringr
documentation</a> and the very
helpful <a href="https://github.com/rstudio/cheatsheets/blob/main/strings.pdf">stringr
cheatsheet</a>.</p>
<p>In the previous section, we noted that the gene symbols in our tibble were lower
case while official gene symbols are in upper case. We can use the stringr
function <code>stringr::str_to_upper()</code> with the <code>dplyr::mutate()</code> function to
perform this adjustment easily:</p>
<div class="sourceCode" id="cb84"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb84-1"><a href="arranging-data.html#cb84-1" aria-hidden="true" tabindex="-1"></a>dplyr<span class="sc">::</span><span class="fu">mutate</span>(gene_stats,</span>
<span id="cb84-2"><a href="arranging-data.html#cb84-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">gene=</span>stringr<span class="sc">::</span><span class="fu">str_to_upper</span>(gene)</span>
<span id="cb84-3"><a href="arranging-data.html#cb84-3" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div>
<pre><code>## # A tibble: 3 x 5
##   gene  test1_stat      test1_p test2_stat  test2_p
##   &lt;chr&gt;      &lt;dbl&gt;        &lt;dbl&gt;      &lt;dbl&gt;    &lt;dbl&gt;
## 1 APOE       12.5  0.103            34.2   0.000013
## 2 HOXD1       4.40 0.632            16.3   0.0421  
## 3 SNCA       45.7  0.0000000042      0.757 0.915</code></pre>
<p>Now are gene symbols have the appropriate case, and our wetlab colleagues won’t
complain about it. :)</p>
<div id="regular-expressions" class="section level4" number="5.7.2.1">
<h4><span class="header-section-number">5.7.2.1</span> Regular expressions</h4>
<p>Many of the string operations in the <code>stringr</code> package use <a href="https://en.wikipedia.org/wiki/Regular_expression">regular
expression</a> syntax. A regular
expression is a sequence of characters that describes patterns in text. Regular
expressions are written in a sort of mini programming language where certain
characters have special meaning that help in defining <em>search patterns</em> that
identifies the location of sequences of characters in text that follow a
particular pattern specified by the regular expression. This is similar to the
“Find” functionality in many word processors, but is more powerful due to the
flexibility of the patterns that can be found.</p>
<p>A simple example will be helpful. Let’s say we have a tibble containing the
result of a (made-up) statistical test for all the genes in a genome:</p>
<div class="sourceCode" id="cb86"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb86-1"><a href="arranging-data.html#cb86-1" aria-hidden="true" tabindex="-1"></a>de_genes</span></code></pre></div>
<pre><code>## # A tibble: 39,535 x 4
##    hgnc_symbol  mean        p    padj
##    &lt;chr&gt;       &lt;int&gt;    &lt;dbl&gt;   &lt;dbl&gt;
##  1 MT-TF        5799 0.000910 0.00941
##  2 MT-RNR1       153 0.0272   0.0342 
##  3 MT-TV         115 0.0228   0.0301 
##  4 MT-RNR2       495 0.00318  0.0123 
##  5 MT-TL1      20201 0.000377 0.00841
##  6 MT-ND1        160 0.0179   0.0258 
##  7 MT-TI        3511 0.00247  0.0115 
##  8 MT-TQ         772 0.00376  0.0129 
##  9 MT-TM         301 0.00325  0.0124 
## 10 MT-ND2         12 0.107    0.111  
## # ... with 39,525 more rows</code></pre>
<p>Now let’s say we’re interested in examining the results for the BRCA family of
genes, BRCA1 and BRCA2. We can use <code>filter()</code> on the data frame to look for them
individually:</p>
<div class="sourceCode" id="cb88"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb88-1"><a href="arranging-data.html#cb88-1" aria-hidden="true" tabindex="-1"></a>de_genes <span class="sc">%&gt;%</span> <span class="fu">filter</span>(hgnc_symbol <span class="sc">==</span> <span class="st">&quot;BRCA1&quot;</span> <span class="sc">|</span> hgnc_symbol <span class="sc">==</span> <span class="st">&quot;BRCA2&quot;</span>)</span></code></pre></div>
<pre><code>## # A tibble: 2 x 4
##   hgnc_symbol  mean      p   padj
##   &lt;chr&gt;       &lt;int&gt;  &lt;dbl&gt;  &lt;dbl&gt;
## 1 BRCA1          41 0.0321 0.0386
## 2 BRCA2         447 0.0140 0.0223</code></pre>
<p>This isn’t so bad, but we can do the same thing with
<a href="https://stringr.tidyverse.org/reference/str_detect.html"><code>stringr::str_detect()</code></a>
, which returns <code>TRUE</code> if the provided pattern matches the input and <code>FALSE</code>
otherwise, a regular expression, and the [<code>dplyr::filter()</code> function], which is
described in greater detail in a later section:</p>
<div class="sourceCode" id="cb90"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb90-1"><a href="arranging-data.html#cb90-1" aria-hidden="true" tabindex="-1"></a>dplyr<span class="sc">::</span><span class="fu">filter</span>(de_genes, <span class="fu">str_detect</span>(hgnc_symbol,<span class="st">&quot;^BRCA[12]$&quot;</span>))</span></code></pre></div>
<pre><code>## # A tibble: 2 x 4
##   hgnc_symbol  mean      p   padj
##   &lt;chr&gt;       &lt;int&gt;  &lt;dbl&gt;  &lt;dbl&gt;
## 1 BRCA1          41 0.0321 0.0386
## 2 BRCA2         447 0.0140 0.0223</code></pre>
<p>The argument <code>"^BRCA[12]$"</code> is a regular expression that searches for the
following:</p>
<ul>
<li>Search for genes that start with the characters <code>BRCA</code> - the <code>^</code> at the
beginning of the pattern stands for the start of the string</li>
<li>For genes that start with <code>BRCA</code>, then look for genes where the next character
is either <code>1</code> or <code>2</code> with <code>[12]</code> - the characters between the <code>[]</code> are searched
for explicitly, and any character encountered that is not listed between them
results in a non-match</li>
<li>For genes that start with <code>BRCA</code> followed with either a <code>1</code> or a <code>2</code>, match
successfully if the number is at the end of the name - the <code>$</code> at the end of the
pattern stands for the end of the string</li>
</ul>
<p>We can use these principles to find genes with more complex naming conventions.
The <a href="https://www.nature.com/articles/pr19972506">Homeobox (HOX) genes</a> encode
DNA binding proteins that regulate gene expression of genes involved in
morphgenesis and cell differentiation in vertebrates. In humans, HOX genes are
organized into 4 clusters of paralogs that were the result of three DNA
duplication events in the distant evolutionary past<span class="citation">(<a href="#ref-Abbasi2015-rn" role="doc-biblioref">Abbasi 2015</a>)</span>, where each
cluster encodes a subset of 13 distinct HOX genes placed next to each other.
Each of these clusters has been assigned a letter identifier A-D (e.g. HOXA,
HOXB, HOXC, and HOXD) and each paralogous gene within each cluster is assigned
the same number (e.g. HOXA4, HOXB4, HOXC4, and HOXD4 are paralogs). There are 13
HOX genes in total, though not all genes remain in all clusters (e.g. HOXA1,
HOXB1, and HOXD1 exist but HOXC1 was lost over time). The following figure
depicts the human HOX gene clusters:</p>
<div class="figure">
<img src="human_hox.jpg" alt="" />
<p class="caption">Human HOX gene clusters - Veraksa, A.; Del Campo, M.; McGinnis, W. Developmental Patterning Genes and their Conserved Functions: From Model Organisms to Humans. Mol. Genet. Metab. 2000, 69 (2), 85–100.</p>
</div>
<p>Let’s say we want to extract out all the HOX genes from our gene statistics. We
can write a regular expression that matches the pattern described above:</p>
<div class="sourceCode" id="cb92"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb92-1"><a href="arranging-data.html#cb92-1" aria-hidden="true" tabindex="-1"></a>dplyr<span class="sc">::</span><span class="fu">filter</span>(de_genes, <span class="fu">str_detect</span>(hgnc_symbol,<span class="st">&quot;^HOX[A-D][0-9]+$&quot;</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb92-2"><a href="arranging-data.html#cb92-2" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">arrange</span>(hgnc_symbol)</span></code></pre></div>
<pre><code>## # A tibble: 39 x 4
##    hgnc_symbol  mean        p    padj
##    &lt;chr&gt;       &lt;int&gt;    &lt;dbl&gt;   &lt;dbl&gt;
##  1 HOXA1        8734 0.000858 0.00934
##  2 HOXA10       4149 0.00152  0.0102 
##  3 HOXA11        567 0.0101   0.0188 
##  4 HOXA13        411 0.0105   0.0191 
##  5 HOXA2         554 0.00600  0.0151 
##  6 HOXA3          18 0.0919   0.0959 
##  7 HOXA4         475 0.0113   0.0199 
##  8 HOXA5         434 0.0127   0.0211 
##  9 HOXA6        3983 0.00252  0.0115 
## 10 HOXA7         897 0.00961  0.0183 
## # ... with 29 more rows</code></pre>
<p>In this query we used two new regular expression features:</p>
<ul>
<li>within the <code>[]</code> we specified a range of characters <code>A-D</code> and <code>0-9</code> which will
match any of the characters between A and D (i.e. A, B, C, or D) and 0 and 9
respectively</li>
<li>the <code>+</code> character means “match one or more of the preceding expression,” which
in our case is the <code>[0-9]</code>. This allows us to match genes with only a single
number (e.g. <code>HOXA1</code>) as well as double digit numbers (e.g. <code>HOXA10</code>).</li>
</ul>
<p>Since we know the cluster identifier part of the HOX gene names (i.e. the
<code>[A-D]</code> part) is exactly one character long, we could alternatively write the
regular expression as follows, using the special <code>.</code> character:</p>
<div class="sourceCode" id="cb94"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb94-1"><a href="arranging-data.html#cb94-1" aria-hidden="true" tabindex="-1"></a>dplyr<span class="sc">::</span><span class="fu">filter</span>(de_genes, <span class="fu">str_detect</span>(hgnc_symbol,<span class="st">&quot;^HOX.[0-9]+$&quot;</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb94-2"><a href="arranging-data.html#cb94-2" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">arrange</span>(hgnc_symbol)</span></code></pre></div>
<pre><code>## # A tibble: 39 x 4
##    hgnc_symbol  mean        p    padj
##    &lt;chr&gt;       &lt;int&gt;    &lt;dbl&gt;   &lt;dbl&gt;
##  1 HOXA1        8734 0.000858 0.00934
##  2 HOXA10       4149 0.00152  0.0102 
##  3 HOXA11        567 0.0101   0.0188 
##  4 HOXA13        411 0.0105   0.0191 
##  5 HOXA2         554 0.00600  0.0151 
##  6 HOXA3          18 0.0919   0.0959 
##  7 HOXA4         475 0.0113   0.0199 
##  8 HOXA5         434 0.0127   0.0211 
##  9 HOXA6        3983 0.00252  0.0115 
## 10 HOXA7         897 0.00961  0.0183 
## # ... with 29 more rows</code></pre>
<p>Here, the <code>.</code> character is interpreted by the regex to match <em>any single
character</em>, regardless of what it is, between the <code>HOX</code> part and the number
part. This also requires that there exist exactly one character between the two
parts; a gene symbol <code>HOX1</code> would <em>not</em> be matched, because the <code>1</code> would match
to the <code>.</code>, but no number remains to match to the <code>[0-9]+</code> part.</p>
<div class="box note">
<p>Sometimes you want to search text for characters that are considered special in
the regular expression language. For example, if you had a list of filenames:</p>
<div class="sourceCode" id="cb96"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb96-1"><a href="arranging-data.html#cb96-1" aria-hidden="true" tabindex="-1"></a>filenames <span class="ot">&lt;-</span> <span class="fu">tribble</span>(</span>
<span id="cb96-2"><a href="arranging-data.html#cb96-2" aria-hidden="true" tabindex="-1"></a>  <span class="sc">~</span>name,</span>
<span id="cb96-3"><a href="arranging-data.html#cb96-3" aria-hidden="true" tabindex="-1"></a>  <span class="st">&quot;annotation.csv&quot;</span>,</span>
<span id="cb96-4"><a href="arranging-data.html#cb96-4" aria-hidden="true" tabindex="-1"></a>  <span class="st">&quot;file1.txt&quot;</span>,</span>
<span id="cb96-5"><a href="arranging-data.html#cb96-5" aria-hidden="true" tabindex="-1"></a>  <span class="st">&quot;file2.txt&quot;</span>,</span>
<span id="cb96-6"><a href="arranging-data.html#cb96-6" aria-hidden="true" tabindex="-1"></a>  <span class="st">&quot;results.csv&quot;</span></span>
<span id="cb96-7"><a href="arranging-data.html#cb96-7" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div>
<p>and wanted to limit to just those with the <code>.txt</code> extension, you need to match
using a literal <code>.</code> character:</p>
<div class="sourceCode" id="cb97"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb97-1"><a href="arranging-data.html#cb97-1" aria-hidden="true" tabindex="-1"></a><span class="fu">filter</span>(filenames, stringr<span class="sc">::</span><span class="fu">str_detect</span>(name,<span class="st">&quot;[.]txt$&quot;</span>))</span></code></pre></div>
<pre><code>## # A tibble: 2 x 1
##   name     
##   &lt;chr&gt;    
## 1 file1.txt
## 2 file2.txt</code></pre>
<p>Inside a <code>[]</code>, characters do not have their usual regular expression meaning,
and therefore <code>[.]</code> will match a literal <code>.</code> character. Instead of using the
<code>[]</code> syntax, you may also escape these literal characters using two back
slashes:</p>
<div class="sourceCode" id="cb99"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb99-1"><a href="arranging-data.html#cb99-1" aria-hidden="true" tabindex="-1"></a><span class="fu">filter</span>(filenames, stringr<span class="sc">::</span><span class="fu">str_detect</span>(name,<span class="st">&quot;</span><span class="sc">\\</span><span class="st">.txt$&quot;</span>))</span></code></pre></div>
<pre><code>## # A tibble: 2 x 1
##   name     
##   &lt;chr&gt;    
## 1 file1.txt
## 2 file2.txt</code></pre>
</div>
<p>Regular expressions are very powerful, and can do much more than what is
described here. See the regular expression tutorial linked in the readmore box
to learn more details.</p>
<div class="box readmore">
<ul>
<li><a href="https://r4ds.had.co.nz/strings.html">R for Data Science - Strings</a></li>
<li><a href="https://stringr.tidyverse.org/reference/index.html">stringr
documentation</a></li>
<li><a href="https://github.com/rstudio/cheatsheets/blob/main/strings.pdf">stringr
cheatsheet</a></li>
<li><a href="https://regexone.com/">RegexOne - regular expression tutorial</a></li>
</ul>
</div>
</div>
</div>
<div id="dplyrselect---subset-columns-by-name" class="section level3" number="5.7.3">
<h3><span class="header-section-number">5.7.3</span> <code>dplyr::select()</code> - Subset Columns by Name</h3>
<p>Our <code>mutate()</code> operations above created a number of new columns in our tibble,
but we did not specify where in the tibble the new columns should go. Let’s
consider the mutated tibble we created with all four new columns:</p>
<div class="sourceCode" id="cb101"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb101-1"><a href="arranging-data.html#cb101-1" aria-hidden="true" tabindex="-1"></a>dplyr<span class="sc">::</span><span class="fu">mutate</span>(gene_stats,</span>
<span id="cb101-2"><a href="arranging-data.html#cb101-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">test1_padj=</span><span class="fu">p.adjust</span>(test1_p,<span class="at">method=</span><span class="st">&quot;fdr&quot;</span>),</span>
<span id="cb101-3"><a href="arranging-data.html#cb101-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">test2_padj=</span><span class="fu">p.adjust</span>(test2_p,<span class="at">method=</span><span class="st">&quot;fdr&quot;</span>),</span>
<span id="cb101-4"><a href="arranging-data.html#cb101-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">signif_either=</span>(test1_padj <span class="sc">&lt;</span> <span class="fl">0.05</span> <span class="sc">|</span> test2_padj <span class="sc">&lt;</span> <span class="fl">0.05</span>),</span>
<span id="cb101-5"><a href="arranging-data.html#cb101-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">signif_both=</span>(test1_padj <span class="sc">&lt;</span> <span class="fl">0.05</span> <span class="sc">&amp;</span> test2_padj <span class="sc">&lt;</span> <span class="fl">0.05</span>),</span>
<span id="cb101-6"><a href="arranging-data.html#cb101-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">gene=</span>stringr<span class="sc">::</span><span class="fu">str_to_upper</span>(gene)</span>
<span id="cb101-7"><a href="arranging-data.html#cb101-7" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div>
<pre><code>## # A tibble: 3 x 9
##   gene  test1_stat      test1_p test2_stat  test2_p   test1_padj test2_padj
##   &lt;chr&gt;      &lt;dbl&gt;        &lt;dbl&gt;      &lt;dbl&gt;    &lt;dbl&gt;        &lt;dbl&gt;      &lt;dbl&gt;
## 1 APOE       12.5  0.103            34.2   0.000013 0.155          0.000039
## 2 HOXD1       4.40 0.632            16.3   0.0421   0.632          0.0632  
## 3 SNCA       45.7  0.0000000042      0.757 0.915    0.0000000126   0.915   
## # ... with 2 more variables: signif_either &lt;lgl&gt;, signif_both &lt;lgl&gt;</code></pre>
<p>From a readability standpoint, it might be helpful if all the columns that are
about each test were grouped together, rather than having to look at the end of
the tibble to find them.</p>
<p>The <a href="https://dplyr.tidyverse.org/reference/select.html"><code>dplyr::select()</code>
function</a> allows you to pick
specific columns out of a larger <code>tibble</code> in whatever order you choose:</p>
<div class="sourceCode" id="cb103"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb103-1"><a href="arranging-data.html#cb103-1" aria-hidden="true" tabindex="-1"></a>stats <span class="ot">&lt;-</span> dplyr<span class="sc">::</span><span class="fu">select</span>(gene_stats, test1_stat, test2_stat)</span>
<span id="cb103-2"><a href="arranging-data.html#cb103-2" aria-hidden="true" tabindex="-1"></a>stats</span></code></pre></div>
<pre><code>## # A tibble: 3 x 2
##   test1_stat test2_stat
##        &lt;dbl&gt;      &lt;dbl&gt;
## 1      12.5      34.2  
## 2       4.40     16.3  
## 3      45.7       0.757</code></pre>
<p>Here we have explicitly selected the statistics columns. dplyr also has <a href="https://tidyselect.r-lib.org/reference/starts_with.html">helper
functions</a> that allow
for more flexible selection of columns. For example, if all of the columns we
wished to select ended with <code>_stat</code>, we could use the <code>ends_with()</code> helper
function:</p>
<div class="sourceCode" id="cb105"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb105-1"><a href="arranging-data.html#cb105-1" aria-hidden="true" tabindex="-1"></a>stats <span class="ot">&lt;-</span> dplyr<span class="sc">::</span><span class="fu">select</span>(gene_stats, <span class="fu">ends_with</span>(<span class="st">&quot;_stat&quot;</span>))</span>
<span id="cb105-2"><a href="arranging-data.html#cb105-2" aria-hidden="true" tabindex="-1"></a>stats</span></code></pre></div>
<pre><code>## # A tibble: 3 x 2
##   test1_stat test2_stat
##        &lt;dbl&gt;      &lt;dbl&gt;
## 1      12.5      34.2  
## 2       4.40     16.3  
## 3      45.7       0.757</code></pre>
<p>If you so desire, <code>select()</code> allows for the renaming of selected columns:</p>
<div class="sourceCode" id="cb107"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb107-1"><a href="arranging-data.html#cb107-1" aria-hidden="true" tabindex="-1"></a>stats <span class="ot">&lt;-</span> dplyr<span class="sc">::</span><span class="fu">select</span>(gene_stats,</span>
<span id="cb107-2"><a href="arranging-data.html#cb107-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">t=</span>test1_stat,</span>
<span id="cb107-3"><a href="arranging-data.html#cb107-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">chisq=</span>test2_stat</span>
<span id="cb107-4"><a href="arranging-data.html#cb107-4" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb107-5"><a href="arranging-data.html#cb107-5" aria-hidden="true" tabindex="-1"></a>stats</span></code></pre></div>
<pre><code>## # A tibble: 3 x 2
##       t  chisq
##   &lt;dbl&gt;  &lt;dbl&gt;
## 1 12.5  34.2  
## 2  4.40 16.3  
## 3 45.7   0.757</code></pre>
<p>If we knew that these test statistics actually corresponded to some kind of
t-test and a <span class="math inline">\(\chi\)</span>-squared test, naming the columns of the tibble appropriately
may help others (and possibly you) understand your code better.</p>
<p>We can use the <code>dplyr::select()</code> function to obtain our desired column order:</p>
<div class="sourceCode" id="cb109"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb109-1"><a href="arranging-data.html#cb109-1" aria-hidden="true" tabindex="-1"></a>dplyr<span class="sc">::</span><span class="fu">mutate</span>(gene_stats,</span>
<span id="cb109-2"><a href="arranging-data.html#cb109-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">test1_padj=</span><span class="fu">p.adjust</span>(test1_p,<span class="at">method=</span><span class="st">&quot;fdr&quot;</span>),</span>
<span id="cb109-3"><a href="arranging-data.html#cb109-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">test2_padj=</span><span class="fu">p.adjust</span>(test2_p,<span class="at">method=</span><span class="st">&quot;fdr&quot;</span>),</span>
<span id="cb109-4"><a href="arranging-data.html#cb109-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">signif_either=</span>(test1_padj <span class="sc">&lt;</span> <span class="fl">0.05</span> <span class="sc">|</span> test2_padj <span class="sc">&lt;</span> <span class="fl">0.05</span>),</span>
<span id="cb109-5"><a href="arranging-data.html#cb109-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">signif_both=</span>(test1_padj <span class="sc">&lt;</span> <span class="fl">0.05</span> <span class="sc">&amp;</span> test2_padj <span class="sc">&lt;</span> <span class="fl">0.05</span>),</span>
<span id="cb109-6"><a href="arranging-data.html#cb109-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">gene=</span>stringr<span class="sc">::</span><span class="fu">str_to_upper</span>(gene)</span>
<span id="cb109-7"><a href="arranging-data.html#cb109-7" aria-hidden="true" tabindex="-1"></a>) <span class="sc">%&gt;%</span></span>
<span id="cb109-8"><a href="arranging-data.html#cb109-8" aria-hidden="true" tabindex="-1"></a>dplyr<span class="sc">::</span><span class="fu">select</span>(</span>
<span id="cb109-9"><a href="arranging-data.html#cb109-9" aria-hidden="true" tabindex="-1"></a>    gene,</span>
<span id="cb109-10"><a href="arranging-data.html#cb109-10" aria-hidden="true" tabindex="-1"></a>    test1_stat, test1_p, test1_padj,</span>
<span id="cb109-11"><a href="arranging-data.html#cb109-11" aria-hidden="true" tabindex="-1"></a>    test2_stat, test2_p, test2_padj,</span>
<span id="cb109-12"><a href="arranging-data.html#cb109-12" aria-hidden="true" tabindex="-1"></a>    signif_either,</span>
<span id="cb109-13"><a href="arranging-data.html#cb109-13" aria-hidden="true" tabindex="-1"></a>    signif_both</span>
<span id="cb109-14"><a href="arranging-data.html#cb109-14" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div>
<pre><code>## # A tibble: 3 x 9
##   gene  test1_stat      test1_p   test1_padj test2_stat  test2_p test2_padj
##   &lt;chr&gt;      &lt;dbl&gt;        &lt;dbl&gt;        &lt;dbl&gt;      &lt;dbl&gt;    &lt;dbl&gt;      &lt;dbl&gt;
## 1 APOE       12.5  0.103        0.155            34.2   0.000013   0.000039
## 2 HOXD1       4.40 0.632        0.632            16.3   0.0421     0.0632  
## 3 SNCA       45.7  0.0000000042 0.0000000126      0.757 0.915      0.915   
## # ... with 2 more variables: signif_either &lt;lgl&gt;, signif_both &lt;lgl&gt;</code></pre>
<p>Now the order of our columns is clear and convenient. It is not necessary to
list the columns for each test statistic on the same line, but the author thinks
this makes the code easier to read and understand.</p>
<div class="box readmore">
<ul>
<li><a href="https://r4ds.had.co.nz/transform.html#select">R for Data Science - Select columns with <code>select()</code></a></li>
<li><a href="https://dplyr.tidyverse.org/reference/index.html">dplyr reference</a></li>
<li><a href="https://tidyselect.r-lib.org/reference/starts_with.html">select helper functions</a></li>
</ul>
</div>
</div>
<div id="dplyrfilter---pick-rows-out-of-a-data-set" class="section level3" number="5.7.4">
<h3><span class="header-section-number">5.7.4</span> <code>dplyr::filter()</code> - Pick rows out of a data set</h3>
<p>Often, the first step in interpreting an analysis is to identify the features
that are significant at some adjusted p-value threshold. First we will
save our mutated tibble to another variable, to aid in demonstration:</p>
<div class="sourceCode" id="cb111"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb111-1"><a href="arranging-data.html#cb111-1" aria-hidden="true" tabindex="-1"></a>gene_stats_mutated <span class="ot">&lt;-</span> dplyr<span class="sc">::</span><span class="fu">mutate</span>(gene_stats,</span>
<span id="cb111-2"><a href="arranging-data.html#cb111-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">test1_padj=</span><span class="fu">p.adjust</span>(test1_p,<span class="at">method=</span><span class="st">&quot;fdr&quot;</span>),</span>
<span id="cb111-3"><a href="arranging-data.html#cb111-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">test2_padj=</span><span class="fu">p.adjust</span>(test2_p,<span class="at">method=</span><span class="st">&quot;fdr&quot;</span>),</span>
<span id="cb111-4"><a href="arranging-data.html#cb111-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">signif_either=</span>(test1_padj <span class="sc">&lt;</span> <span class="fl">0.05</span> <span class="sc">|</span> test2_padj <span class="sc">&lt;</span> <span class="fl">0.05</span>),</span>
<span id="cb111-5"><a href="arranging-data.html#cb111-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">signif_both=</span>(test1_padj <span class="sc">&lt;</span> <span class="fl">0.05</span> <span class="sc">&amp;</span> test2_padj <span class="sc">&lt;</span> <span class="fl">0.05</span>),</span>
<span id="cb111-6"><a href="arranging-data.html#cb111-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">gene=</span>stringr<span class="sc">::</span><span class="fu">str_to_upper</span>(gene)</span>
<span id="cb111-7"><a href="arranging-data.html#cb111-7" aria-hidden="true" tabindex="-1"></a>) <span class="sc">%&gt;%</span></span>
<span id="cb111-8"><a href="arranging-data.html#cb111-8" aria-hidden="true" tabindex="-1"></a>dplyr<span class="sc">::</span><span class="fu">select</span>(</span>
<span id="cb111-9"><a href="arranging-data.html#cb111-9" aria-hidden="true" tabindex="-1"></a>    gene,</span>
<span id="cb111-10"><a href="arranging-data.html#cb111-10" aria-hidden="true" tabindex="-1"></a>    test1_stat, test1_p, test1_padj,</span>
<span id="cb111-11"><a href="arranging-data.html#cb111-11" aria-hidden="true" tabindex="-1"></a>    test2_stat, test2_p, test2_padj,</span>
<span id="cb111-12"><a href="arranging-data.html#cb111-12" aria-hidden="true" tabindex="-1"></a>    signif_either,</span>
<span id="cb111-13"><a href="arranging-data.html#cb111-13" aria-hidden="true" tabindex="-1"></a>    signif_both</span>
<span id="cb111-14"><a href="arranging-data.html#cb111-14" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div>
<p>Now we can use the <code>dplyr::filter()</code> function to select rows based on whether
they are significant in either test this with our above example.</p>
<div class="sourceCode" id="cb112"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb112-1"><a href="arranging-data.html#cb112-1" aria-hidden="true" tabindex="-1"></a>dplyr<span class="sc">::</span><span class="fu">filter</span>(gene_stats_mutated, test1_padj <span class="sc">&lt;</span> <span class="fl">0.05</span>)</span></code></pre></div>
<pre><code>## # A tibble: 1 x 9
##   gene  test1_stat      test1_p   test1_padj test2_stat test2_p test2_padj
##   &lt;chr&gt;      &lt;dbl&gt;        &lt;dbl&gt;        &lt;dbl&gt;      &lt;dbl&gt;   &lt;dbl&gt;      &lt;dbl&gt;
## 1 SNCA        45.7 0.0000000042 0.0000000126      0.757   0.915      0.915
## # ... with 2 more variables: signif_either &lt;lgl&gt;, signif_both &lt;lgl&gt;</code></pre>
<div class="sourceCode" id="cb114"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb114-1"><a href="arranging-data.html#cb114-1" aria-hidden="true" tabindex="-1"></a>dplyr<span class="sc">::</span><span class="fu">filter</span>(gene_stats_mutated, test2_padj <span class="sc">&lt;</span> <span class="fl">0.05</span>)</span></code></pre></div>
<pre><code>## # A tibble: 1 x 9
##   gene  test1_stat test1_p test1_padj test2_stat  test2_p test2_padj
##   &lt;chr&gt;      &lt;dbl&gt;   &lt;dbl&gt;      &lt;dbl&gt;      &lt;dbl&gt;    &lt;dbl&gt;      &lt;dbl&gt;
## 1 APOE        12.5   0.103      0.155       34.2 0.000013   0.000039
## # ... with 2 more variables: signif_either &lt;lgl&gt;, signif_both &lt;lgl&gt;</code></pre>
<p>Here we are filtering the result so that only genes with nominal p-value less
than 0.05 remain. Note we filter on the two tests separately, but we can also
combine these tests using <a href="https://r4ds.had.co.nz/transform.html#logical-operators">logical
operators</a> to achieve
different results:</p>
<div class="sourceCode" id="cb116"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb116-1"><a href="arranging-data.html#cb116-1" aria-hidden="true" tabindex="-1"></a><span class="co"># | means &quot;logical or&quot;, meaning the row is retained if either condition is true</span></span>
<span id="cb116-2"><a href="arranging-data.html#cb116-2" aria-hidden="true" tabindex="-1"></a>dplyr<span class="sc">::</span><span class="fu">filter</span>(gene_stats_mutated, test1_padj <span class="sc">&lt;</span> <span class="fl">0.05</span> <span class="sc">|</span> test2_padj <span class="sc">&lt;</span> <span class="fl">0.05</span>)</span></code></pre></div>
<pre><code>## # A tibble: 2 x 9
##   gene  test1_stat      test1_p   test1_padj test2_stat  test2_p test2_padj
##   &lt;chr&gt;      &lt;dbl&gt;        &lt;dbl&gt;        &lt;dbl&gt;      &lt;dbl&gt;    &lt;dbl&gt;      &lt;dbl&gt;
## 1 APOE        12.5 0.103        0.155            34.2   0.000013   0.000039
## 2 SNCA        45.7 0.0000000042 0.0000000126      0.757 0.915      0.915   
## # ... with 2 more variables: signif_either &lt;lgl&gt;, signif_both &lt;lgl&gt;</code></pre>
<p>Only APOE and SCNA are significant in at least one of the tests.</p>
<div class="sourceCode" id="cb118"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb118-1"><a href="arranging-data.html#cb118-1" aria-hidden="true" tabindex="-1"></a><span class="co"># &amp; means &quot;logical and&quot;, meaning the row is retained only if both conditions are true</span></span>
<span id="cb118-2"><a href="arranging-data.html#cb118-2" aria-hidden="true" tabindex="-1"></a>dplyr<span class="sc">::</span><span class="fu">filter</span>(gene_stats_mutated, test1_padj <span class="sc">&lt;</span> <span class="fl">0.05</span> <span class="sc">&amp;</span> test2_padj <span class="sc">&lt;</span> <span class="fl">0.05</span>)</span></code></pre></div>
<pre><code>## # A tibble: 0 x 9
## # ... with 9 variables: gene &lt;chr&gt;, test1_stat &lt;dbl&gt;, test1_p &lt;dbl&gt;,
## #   test1_padj &lt;dbl&gt;, test2_stat &lt;dbl&gt;, test2_p &lt;dbl&gt;, test2_padj &lt;dbl&gt;,
## #   signif_either &lt;lgl&gt;, signif_both &lt;lgl&gt;</code></pre>
<p>It looks like we don’t have any genes that are significant by both tests.
Filtering results like this is one of the most common operations we do on the
results of biological analyses.</p>
<div class="box readmore">
<ul>
<li><a href="https://r4ds.had.co.nz/transform.html#filter-rows-with-filter">R for Data Science - Filter rows with <code>filter()</code></a></li>
<li><a href="https://dplyr.tidyverse.org/reference/filter.html">dplyr - <code>filter()</code> reference</a></li>
</ul>
</div>
</div>
<div id="dplyrarrange---order-rows-based-on-their-values" class="section level3" number="5.7.5">
<h3><span class="header-section-number">5.7.5</span> <code>dplyr::arrange()</code> - Order rows based on their values</h3>
<p>Another common operation when working with biological analysis results is
ordering them by some meaningful value. Like above, p-values are often used to
prioritize results by simply sorting them in ascending order. The <code>arrange()</code>
function is how to perform this sorting in tidyverse:</p>
<div class="sourceCode" id="cb120"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb120-1"><a href="arranging-data.html#cb120-1" aria-hidden="true" tabindex="-1"></a>stats_sorted_by_test1_p <span class="ot">&lt;-</span> dplyr<span class="sc">::</span><span class="fu">arrange</span>(gene_stats, test1_p)</span>
<span id="cb120-2"><a href="arranging-data.html#cb120-2" aria-hidden="true" tabindex="-1"></a>stats_sorted_by_test1_p</span></code></pre></div>
<pre><code>## # A tibble: 3 x 5
##   gene  test1_stat      test1_p test2_stat  test2_p
##   &lt;chr&gt;      &lt;dbl&gt;        &lt;dbl&gt;      &lt;dbl&gt;    &lt;dbl&gt;
## 1 snca       45.7  0.0000000042      0.757 0.915   
## 2 apoe       12.5  0.103            34.2   0.000013
## 3 hoxd1       4.40 0.632            16.3   0.0421</code></pre>
<div class="box note">
<p>Note we are sorting by nominal p-value here, not adjusted p-value. In general,
sorting by nominal or adjusted p-value results in the same order of results. The
only exception is when, due to the way the FDR procedure works, some adjusted
p-values will be identical, making the relative order of those tests with the
same FDR meaningless. In contrast, it is very rare that nominal p-values will be
identical, and since they induce the same ordering of results, when sorting
analysis results there are advantages to using nominal p-value, rather than
adjusted p-value.</p>
</div>
<p>In general, the larger the magnitude of the statistic, the smaller the p-value
(for two-tailed tests), so if we so desired we could induce a similar ranking by
arranging the data by the statistic in descending order:</p>
<div class="sourceCode" id="cb122"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb122-1"><a href="arranging-data.html#cb122-1" aria-hidden="true" tabindex="-1"></a><span class="co"># desc() is a helper function that causes the results to be sorted in descending</span></span>
<span id="cb122-2"><a href="arranging-data.html#cb122-2" aria-hidden="true" tabindex="-1"></a><span class="co"># order for the given column</span></span>
<span id="cb122-3"><a href="arranging-data.html#cb122-3" aria-hidden="true" tabindex="-1"></a>dplyr<span class="sc">::</span><span class="fu">arrange</span>(gene_stats, <span class="fu">desc</span>(<span class="fu">abs</span>(test1_stat)))</span></code></pre></div>
<pre><code>## # A tibble: 3 x 5
##   gene  test1_stat      test1_p test2_stat  test2_p
##   &lt;chr&gt;      &lt;dbl&gt;        &lt;dbl&gt;      &lt;dbl&gt;    &lt;dbl&gt;
## 1 snca       45.7  0.0000000042      0.757 0.915   
## 2 apoe       12.5  0.103            34.2   0.000013
## 3 hoxd1       4.40 0.632            16.3   0.0421</code></pre>
<p>Here we first apply the base R <code>abs()</code> function to compute the absolute value of
the test 1 statistic and then specify that we want to sort largest first. Note
although we don’t have any negative values in our dataset, we should not assume
that in general, so it is safer for us to be complete and add the absolute value
call in case later we decide to copy and paste this code into another analysis.
That’s pretty much all there is to <code>arrange()</code>.</p>
<div class="box readmore">
<ul>
<li><a href="https://r4ds.had.co.nz/transform.html#arrange-rows-with-arrange">R for Data Science - Arrange rows with<code>arrange()</code></a></li>
<li><a href="https://dplyr.tidyverse.org/reference/arrange.html">dplyr <code>arrange()</code> reference</a></li>
</ul>
</div>
</div>
<div id="putting-it-all-together" class="section level3" number="5.7.6">
<h3><span class="header-section-number">5.7.6</span> Putting it all together</h3>
<p>In the previous sections, we performed the following operations:</p>
<ol style="list-style-type: decimal">
<li>Created new columns by computing false discovery rate on the nominal p-values
using the <code>dplyr::mutate()</code> and <code>p.adjust</code> functions</li>
<li>Created new columns that indicate the patterns of significance for each gene
using <code>dplyr::mutate()</code></li>
<li>Mutated the gene symbol case using <code>stringr::str_to_upper</code> and
<code>dplyr::mutate()</code></li>
<li>Reordered the columns to group related variables with <code>select()</code></li>
<li>Filtered genes based on whether they have an adjusted p-value less than 0.05
for either and both statistical tests using <code>dplyr::filter()</code></li>
<li>Sorted the results by p-value using <code>dplyr::arrange()</code></li>
</ol>
<p>For the sake of illustration, these steps were presented separately, but
together they represent a single unit of data processing and thus might
profitably be done in the same R command using <code>%&gt;%</code>:</p>
<div class="sourceCode" id="cb124"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb124-1"><a href="arranging-data.html#cb124-1" aria-hidden="true" tabindex="-1"></a>gene_stats <span class="ot">&lt;-</span> dplyr<span class="sc">::</span><span class="fu">mutate</span>(gene_stats,</span>
<span id="cb124-2"><a href="arranging-data.html#cb124-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">test1_padj=</span><span class="fu">p.adjust</span>(test1_p,<span class="at">method=</span><span class="st">&quot;fdr&quot;</span>),</span>
<span id="cb124-3"><a href="arranging-data.html#cb124-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">test2_padj=</span><span class="fu">p.adjust</span>(test2_p,<span class="at">method=</span><span class="st">&quot;fdr&quot;</span>),</span>
<span id="cb124-4"><a href="arranging-data.html#cb124-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">signif_either=</span>(test1_padj <span class="sc">&lt;</span> <span class="fl">0.05</span> <span class="sc">|</span> test2_padj <span class="sc">&lt;</span> <span class="fl">0.05</span>),</span>
<span id="cb124-5"><a href="arranging-data.html#cb124-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">signif_both=</span>(test1_padj <span class="sc">&lt;</span> <span class="fl">0.05</span> <span class="sc">&amp;</span> test2_padj <span class="sc">&lt;</span> <span class="fl">0.05</span>),</span>
<span id="cb124-6"><a href="arranging-data.html#cb124-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">gene=</span>stringr<span class="sc">::</span><span class="fu">str_to_upper</span>(gene)</span>
<span id="cb124-7"><a href="arranging-data.html#cb124-7" aria-hidden="true" tabindex="-1"></a>) <span class="sc">%&gt;%</span></span>
<span id="cb124-8"><a href="arranging-data.html#cb124-8" aria-hidden="true" tabindex="-1"></a>dplyr<span class="sc">::</span><span class="fu">select</span>(</span>
<span id="cb124-9"><a href="arranging-data.html#cb124-9" aria-hidden="true" tabindex="-1"></a>    gene,</span>
<span id="cb124-10"><a href="arranging-data.html#cb124-10" aria-hidden="true" tabindex="-1"></a>    test1_stat, test1_p, test1_padj,</span>
<span id="cb124-11"><a href="arranging-data.html#cb124-11" aria-hidden="true" tabindex="-1"></a>    test2_stat, test2_p, test2_padj,</span>
<span id="cb124-12"><a href="arranging-data.html#cb124-12" aria-hidden="true" tabindex="-1"></a>    signif_either,</span>
<span id="cb124-13"><a href="arranging-data.html#cb124-13" aria-hidden="true" tabindex="-1"></a>    signif_both</span>
<span id="cb124-14"><a href="arranging-data.html#cb124-14" aria-hidden="true" tabindex="-1"></a>) <span class="sc">%&gt;%</span></span>
<span id="cb124-15"><a href="arranging-data.html#cb124-15" aria-hidden="true" tabindex="-1"></a>dplyr<span class="sc">::</span><span class="fu">filter</span>(</span>
<span id="cb124-16"><a href="arranging-data.html#cb124-16" aria-hidden="true" tabindex="-1"></a>  test1_padj <span class="sc">&lt;</span> <span class="fl">0.05</span> <span class="sc">|</span> test2_padj <span class="sc">&lt;</span> <span class="fl">0.05</span></span>
<span id="cb124-17"><a href="arranging-data.html#cb124-17" aria-hidden="true" tabindex="-1"></a>) <span class="sc">%&gt;%</span></span>
<span id="cb124-18"><a href="arranging-data.html#cb124-18" aria-hidden="true" tabindex="-1"></a>dplyr<span class="sc">::</span><span class="fu">arrange</span>(</span>
<span id="cb124-19"><a href="arranging-data.html#cb124-19" aria-hidden="true" tabindex="-1"></a>  test1_p</span>
<span id="cb124-20"><a href="arranging-data.html#cb124-20" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb124-21"><a href="arranging-data.html#cb124-21" aria-hidden="true" tabindex="-1"></a>gene_stats</span></code></pre></div>
<pre><code>## # A tibble: 2 x 9
##   gene  test1_stat      test1_p   test1_padj test2_stat  test2_p test2_padj
##   &lt;chr&gt;      &lt;dbl&gt;        &lt;dbl&gt;        &lt;dbl&gt;      &lt;dbl&gt;    &lt;dbl&gt;      &lt;dbl&gt;
## 1 SNCA        45.7 0.0000000042 0.0000000126      0.757 0.915      0.915   
## 2 APOE        12.5 0.103        0.155            34.2   0.000013   0.000039
## # ... with 2 more variables: signif_either &lt;lgl&gt;, signif_both &lt;lgl&gt;</code></pre>
<p>This complete pipeline now contains all of our manipulations and our mutated
tibble can be passed on to downstream analysis or collaborators.</p>
</div>
</div>
<h3>References</h3>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-Abbasi2015-rn" class="csl-entry">
Abbasi, Amir Ali. 2015. <span>“Diversification of Four Human <span>HOX</span> Gene Clusters by Step-Wise Evolution Rather Than Ancient Whole-Genome Duplications.”</span> <em>Dev. Genes Evol.</em> 225 (6): 353–57.
</div>
</div>
<p style="text-align: center;">
<a href="pipes.html"><button class="btn btn-default">Previous</button></a>
<a href="grouping-data.html"><button class="btn btn-default">Next</button></a>
</p>
</div>
</div>


</div>

<script>

// add bootstrap table styles to pandoc tables
$(document).ready(function () {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
});

</script>

</body>
</html>
