<!DOCTYPE html>
<html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Assignments | BF591 - R for Biological Sciences</title>
<meta name="generator" content="bookdown 0.24 with bs4_book()">
<meta property="og:title" content="Assignments | BF591 - R for Biological Sciences">
<meta property="og:type" content="book">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="Assignments | BF591 - R for Biological Sciences">
<!-- JS --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://kit.fontawesome.com/6ecbd6c532.js" crossorigin="anonymous"></script><script src="libs/header-attrs-2.13/header-attrs.js"></script><script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="libs/bootstrap-4.6.0/bootstrap.min.css" rel="stylesheet">
<script src="libs/bootstrap-4.6.0/bootstrap.bundle.min.js"></script><script src="libs/bs3compat-0.3.1/transition.js"></script><script src="libs/bs3compat-0.3.1/tabs.js"></script><script src="libs/bs3compat-0.3.1/bs3compat.js"></script><link href="libs/bs4_book-1.0.0/bs4_book.css" rel="stylesheet">
<script src="libs/bs4_book-1.0.0/bs4_book.js"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- CSS --><link rel="stylesheet" href="style.css">
<meta name="description" content="Assignment Format GitHub repo templates here  Starting an Assignment  :^) good luck When starting a project there are a small number of steps to get things set up just right in order to have a...">
<meta property="og:description" content="Assignment Format GitHub repo templates here  Starting an Assignment  :^) good luck When starting a project there are a small number of steps to get things set up just right in order to have a...">
<meta name="twitter:description" content="Assignment Format GitHub repo templates here  Starting an Assignment  :^) good luck When starting a project there are a small number of steps to get things set up just right in order to have a...">
</head>
<body data-spy="scroll" data-target="#toc">

<div class="container-fluid">
<div class="row">
  <header class="col-sm-12 col-lg-3 sidebar sidebar-book"><a class="sr-only sr-only-focusable" href="#content">Skip to main content</a>

    <div class="d-flex align-items-start justify-content-between">
      <h1>
        <a href="index.html" title="">BF591 - R for Biological Sciences</a>
      </h1>
      <button class="btn btn-outline-primary d-lg-none ml-2 mt-1" type="button" data-toggle="collapse" data-target="#main-nav" aria-expanded="true" aria-controls="main-nav"><i class="fas fa-bars"></i><span class="sr-only">Show table of contents</span></button>
    </div>

    <div id="main-nav" class="collapse-lg">
      <form role="search">
        <input id="search" class="form-control" type="search" placeholder="Search" aria-label="Search">
</form>

      <nav aria-label="Table of contents"><h2>Table of contents</h2>
        <ul class="book-toc list-unstyled">
<li><a class="" href="index.html">R for Biological Sciences</a></li>
<li><a class="" href="introduction.html"><span class="header-section-number">1</span> Introduction</a></li>
<li><a class="" href="data-bio.html"><span class="header-section-number">2</span> Data in Biology</a></li>
<li><a class="" href="preliminaries.html"><span class="header-section-number">3</span> Preliminaries</a></li>
<li><a class="" href="prog-basics.html"><span class="header-section-number">4</span> R Programming</a></li>
<li><a class="" href="data-wrangling.html"><span class="header-section-number">5</span> Data Wrangling</a></li>
<li><a class="" href="data-science.html"><span class="header-section-number">6</span> Data Science</a></li>
<li><a class="" href="data-visualization.html"><span class="header-section-number">7</span> Data Visualization</a></li>
<li><a class="" href="biology-bioinformatics.html"><span class="header-section-number">8</span> Biology &amp; Bioinformatics</a></li>
<li><a class="" href="engineering.html"><span class="header-section-number">9</span> EngineeRing</a></li>
<li><a class="" href="rshiny.html"><span class="header-section-number">10</span> RShiny</a></li>
<li><a class="" href="communicating-with-r.html"><span class="header-section-number">11</span> Communicating with R</a></li>
<li><a class="" href="contribution-guide.html"><span class="header-section-number">12</span> Contribution Guide</a></li>
<li><a class="active" href="assignments.html">Assignments</a></li>
<li class="book-part">Appendix</li>
<li><a class="" href="class-outlines.html"><span class="header-section-number">A</span> Class Outlines</a></li>
</ul>

        <div class="book-extra">
          
        </div>
      </nav>
</div>
  </header><main class="col-sm-12 col-md-9 col-lg-7" id="content"><div id="assignments" class="section level1 unnumbered">
<h1>Assignments<a class="anchor" aria-label="anchor" href="#assignments"><i class="fas fa-link"></i></a>
</h1>
<div id="assignment-format" class="section level2 unnumbered">
<h2>Assignment Format<a class="anchor" aria-label="anchor" href="#assignment-format"><i class="fas fa-link"></i></a>
</h2>
<p>GitHub repo templates here</p>

</div>
<div id="starting-an-assignment" class="section level2 unnumbered">
<h2>Starting an Assignment<a class="anchor" aria-label="anchor" href="#starting-an-assignment"><i class="fas fa-link"></i></a>
</h2>
<blockquote>
<p>:^) good luck</p>
</blockquote>
<p>When starting a project there are a small number of steps to get things set up <em>just</em> right in order to have a smooth and seamless experience. While it is absolutely possible to complete these projects on your own machine and your own installation of R and RStudio, it will be easier and faster to do your work on BU’s SCC. The experience will be more straightforward and basically identical to doing it on your own.</p>
<div id="starting-your-own-git-repository-and-cloning-it" class="section level3 unnumbered">
<h3>Starting your own <code>git</code> repository and cloning it<a class="anchor" aria-label="anchor" href="#starting-your-own-git-repository-and-cloning-it"><i class="fas fa-link"></i></a>
</h3>
<p>In our development of these assignments we used git and <a href="github.com">github.com</a> to manage our updates and allow for easy sharing. We can now pass these repositories (repos) along to you using GitHub classroom so you can follow the steps within.</p>
</div>
<div id="contents" class="section level3 unnumbered">
<h3>Contents<a class="anchor" aria-label="anchor" href="#contents"><i class="fas fa-link"></i></a>
</h3>
<p>Each assignment will have these following files within, roughly:</p>
<pre><code>├── reference_report.html
├── main.R
├── README.md
├── report.Rmd
└── test_main.R</code></pre>
<ul>
<li>
<strong>reference_report.html</strong> - This is the completed “report” of the assignment. You will be endeavoring to replicate this while completing the assignment, with some room for creative differences in elements like plotting.</li>
<li>
<strong>main.R</strong> - This is the main R script in the assignment, where you will be doing a lot of your <em>programming</em>. This script contains function definitions and descriptions, but with functions that are empty. Complete the functions as described in the script and ensure they pass the tests.</li>
<li>
<strong>README.md</strong> - Every repository <em>should</em> have a README. Usually it contains useful information about installation, usage, and licensing. In this case, it will have helpful links back to assignment instructions.</li>
<li>
<strong>report.Rmd</strong> - Another empty file, this markdown script will <code><a href="https://rdrr.io/r/base/source.html">source()</a></code> your <code>main.R</code> script, loading the functions into its environment. There will be empty code blocks throughout this document, use the functions you developed in <code>main.R</code> to replicate the figures or match the captions seen in <code>finalized_report.pdf</code>.</li>
<li>
<strong>test_main.R</strong> - This is a completed testing script produced by us to help you quickly determine how will your code is running. This file <strong>sources</strong> your <code>main.R</code> file, meaning it runs it to completion and stores any functions and variables in the active environment. It then runs tests to ensure your functions are working in a predictable way. Tests are incredibly valuable tools, and potentially one of the strongest in a programmer’s toolkit.</li>
</ul>
<div class="box important">
<p>Most assignments beyond assignment 1 have a complementary guide hosted on the website page. These go into more background about how functions and tests work so use them if you’re feeling stuck or would like more context. <a href="Assignment%20Format">https://bu-bioinfo.github.io/r_for_biological_sciences/assignment-format.html</a></p>
</div>
</div>
<div id="github-tutorial" class="section level3 unnumbered">
<h3>GitHub tutorial<a class="anchor" aria-label="anchor" href="#github-tutorial"><i class="fas fa-link"></i></a>
</h3>
<ol style="list-style-type: decimal">
<li>Find your assignment link</li>
<li>Clone your repository, you will see a new repo in the form: <code>myusername/assignment-X</code>
</li>
<li>Inside this repo, click the big <font color="green"><strong>g r e e n</strong></font> button and copy the HTTP link to your repository.</li>
<li>Login to <a href="SCC%20interactive">https://scc-ondemand.bu.edu/</a>. Start an interactive R session (Interactive Apps &gt; RStudio Server), use the default settings but increase the hours to like 12 or something. Or if you like high pressure situations set it to like 20 minutes.</li>
<li>Once your session is loaded and you are login in, select the “New Project” button below the “Edit” button. <img src="img/how_to_1-new_proj.png">
</li>
<li>With the prompt open, select <code>Version Control</code> &gt; <code>git</code>, and paste your repo’s url into the first prompt. Feel free to rename the project, but also like don’t? That’s confusing. You can change where this folder is setup on SCC, but I find <code>~/Documents</code> or <code>~/Documents/BF591</code> to be a good home for organizing my various projects.<br><img src="img/how_to_2-prompt.png">
</li>
<li>You will need to login to complete this process, enter your GitHub username and password when prompted.</li>
<li>R will download all of the repository contents, and add an <code>.RProj</code> file to the directory. This RProject file is useful for switching between R assignments in RStudio. This context menu is in the top right of the RStudio interface:<br><img src="img/how_to_3-switch-proj.png">
</li>
</ol>
</div>
<div id="committing-and-pushing-with-r-and-without" class="section level3 unnumbered">
<h3>Committing and Pushing with R (and without)<a class="anchor" aria-label="anchor" href="#committing-and-pushing-with-r-and-without"><i class="fas fa-link"></i></a>
</h3>
<p>With the RProject set up, you now have a directory on SCC that has cloned your remote repository in GitHub. At this moment all of the files are the same, since you have made no progress on your assignment (depressing, I know). In a burst of productive energy, you will start to make changes to <code>main.R</code> and write your beautiful, buggy functions. Once you’re done for a little bit, what next? You could leave it on SCC and it would probably be pretty safe, but what if you want to share this code easily? Or if a cataclysm strikes Western Massachusetts? Your changes are gone and life will have lost all meaning.</p>
<p>To prevent this cataclysm, you must <strong>add</strong>, <strong>commit</strong>, and <strong>push</strong> your code back to GitHub.</p>
</div>
<div id="using-rstudio" class="section level3 unnumbered">
<h3>Using RStudio<a class="anchor" aria-label="anchor" href="#using-rstudio"><i class="fas fa-link"></i></a>
</h3>
<ol style="list-style-type: decimal">
<li>Save the files you want to commit, and select the Git tab in the top right viewer (by default).<br><img src="img/how_to_4-git_tab.png">
</li>
<li>You’ll notice the files you changed are listed here, along with varying symbols. A question mark <strong>?</strong> means the file is untracked, an <strong>M</strong> means a tracked file has been modified. A <strong>D</strong> means a tracked file has been deleted. If I select a file I want to add back to GitHub, I can tick the checkbox and it will switch to the left column. This is called <strong>staging</strong> a file.<br><img src="img/how_to_5-checkbox.png">
</li>
<li>Once I’ve staged all the files I want to add, I can click on the <strong>Commit</strong> button. This brings up another window, which will again list the files you are staging and actually show you the differences you’re making (this is called the <strong>diff</strong>, based on the GNU utility of the same name). The most important part is the <strong>Commit</strong> box. A commit is a series of file updates made together, typically to solve a task. It is <strong>paramount</strong> that you include commit messages, so you can track what it is you’re doing with your work! Just a sentence is usually enough to describe what the changes in this commit are doing.
<img src="img/how_to_6-commit_menu.png">
</li>
<li>Once you’ve pressed the commit button, all that’s left to do is to <strong>push</strong> your commit. This sends the commit to GitHub where it can safely live forever, duplicated across the world and safe from most cataclysms.<br><img src="img/how_to_7-push.png">
</li>
<li>In the same vein, the <strong>pull</strong> button is useful if your GitHub has been updated more recently than your current SCC directory. This can happen if you’re editing these files from another computer or on <a href="GitHub.com">GitHub.com</a> directly.</li>
</ol>
</div>
<div id="using-the-command-line" class="section level3 unnumbered">
<h3>Using the command line<a class="anchor" aria-label="anchor" href="#using-the-command-line"><i class="fas fa-link"></i></a>
</h3>
<p>That was kind of a lot of clicking and funny little menus, and my doctor says I am at high risk for a repetitive stress injury, so can’t I just use my keyboard instead? Yes, of course. We’ll use the terminal, which is also built-in to RStudio but really just emulates the standard bash terminal that you would use to interact with SCC normally.</p>
<ol style="list-style-type: decimal">
<li>Click on the terminal tab at the bottom of your layout (default).<br><img src="img/how_to_8-terminal_tab.png"><br>
</li>
<li>Marvel at the wonderful monospace font and rest your hands upon the keyboard. To see how the directory is doing, we can type <code>git status</code> to see what files have been changed. Wow! Looks like I have a lot of stuff to deal with. What’s nice is the <code>git</code> command offers a lot of helpful tips for getting me on my way.
<img src="img/how_to_9-git_status.png">
</li>
<li>I want to add these files, and I really don’t care how many I had so I am going to use <code>git add *</code>, which basically adds everything that can be found in the current directory. If I only wanted one or two things, I could specify <code>git add file1.txt file2.txt</code> and leave the other files alone.</li>
<li>Once I have added (or removed <code>git rm file1.txt</code>) the files I want to commit, I simply use <code>git commit -m "Added a lot more content to tutorials!</code> to create the commit and add my <strong>very important message</strong>. Messages are not mandatory but if you slack on being descriptive you will regret it when you come back later to figure out what in God’s name you were doing that day.</li>
<li>Finally, with my commit done and files added, I simply say <code>git push</code> to send my wonderful code back to GitHub, and I can sit back and marvel at my handiwork.</li>
</ol>
<p>Put simply, the three commands <code>git add *</code>, <code>git commit -m "message"</code>, and <code>git push</code> will save the changes to your code and keep your repository up to date. Easier than clicking around RStudio but both ways are super valid!</p>

</div>
</div>
<div id="assignment-1" class="section level2 unnumbered">
<h2>Assignment 1<a class="anchor" aria-label="anchor" href="#assignment-1"><i class="fas fa-link"></i></a>
</h2>
<div id="problem-statement" class="section level3 unnumbered">
<h3>Problem Statement<a class="anchor" aria-label="anchor" href="#problem-statement"><i class="fas fa-link"></i></a>
</h3>
<p>This assignment will focus on basic functions of R with an emphasis on
<code>tidyverse</code> implementations. <code>tidyverse</code> is a collection of packages, pioneered
by Hadley Wickham and RStudio, that looks to standardize procedures,
functionality, and syntax in R.</p>
<p>To gain familiarity with R, we will be working with a microarray dataset that
contains gene probe expression data for various samples collected from cancer
patients. In bioinformatics, it is common to have multiple datasets for your
different modes of data (i.e. microarray expression data is kept separate
from clinical data detailing the samples). You will get an opportunity to work
with both of these datasets, and be required to cross reference between the two.</p>
</div>
<div id="learning-objectives-and-skill-list" class="section level3 unnumbered">
<h3>Learning Objectives and Skill List<a class="anchor" aria-label="anchor" href="#learning-objectives-and-skill-list"><i class="fas fa-link"></i></a>
</h3>
<ul>
<li>Install various packages needed for analysis</li>
<li>Load Data</li>
<li>Gain familiarity with common <code>tidyverse</code> operations such as <code>groupby()</code>,
<code>mutate</code>, and <code>summarize</code>.</li>
<li>Create a small plot to display results</li>
<li>Utilize R Markdown to create an attractive format for sharing data.</li>
</ul>
</div>
<div id="instructions" class="section level3 unnumbered">
<h3>Instructions<a class="anchor" aria-label="anchor" href="#instructions"><i class="fas fa-link"></i></a>
</h3>
<p>Our main focus for this assignment are installing packages, manipulating data,
and summarizing important statistics across both samples and features (genes). An
empty project repo can be found here:</p>
<p><a href="https://github.com/bu-bioinfo/bf591-assignment-1">https://github.com/bu-bioinfo/bf591-assignment-1</a></p>
<p>The project is laid out as such:</p>
<pre><code>main.R
report.Rmd
reference_report.html</code></pre>
<p>Each step of the assignment is explained in the R markdown file, <code>report.Rmd</code>.
There you will find a list of tasks to explicity implement functions in your
empty <code>main.R</code> script. The <code>main.R</code> script contains skeletons of each function
you’ll need to implement, explaining what each function should do, the parameters
it expects to receive, and what type of output is expected to be returned. A
reference report, <code>reference_report.html</code> is also provided. Assuming you successfully
implement all the functions in <code>main.R</code>, your generated report should look
identical to the information displayed in <code>reference_report.html</code>. In this way,
you can use <code>reference_report.html</code> as a guide to determine if you are correctly
implementing your functions.</p>
<p>Here is the suggested workflow for developing and checking your code in this
assignment:</p>
<ol style="list-style-type: decimal">
<li>
<code>main.R</code> contains function definitions, including signature descriptions, for
a number of functions, but the bodies of those functions are currently blank</li>
<li>
<code>report.Rmd</code> has code chunks that call functions defined in <code>main.R</code> - you do
not need to write anything in the Rmd file (but you may)</li>
<li>Your task is to read the function descriptions and the text in the Rmarkdown
document and fill in the function bodies to produce the desired behavior in
main.R</li>
<li>You can test your work by executing individual code chunks in report.Rmd and
comparing your output to the example compiled report in the repo</li>
<li>In the workflow, you will go back and forth between developing code in main.R
and running code chunks in report.Rmd</li>
<li>When you have developed function bodies for all the functions and executed
all the code chunks in the report successfully, you should be able to knit the
entire report</li>
</ol>
</div>
<div id="hints" class="section level3 unnumbered">
<h3>Hints<a class="anchor" aria-label="anchor" href="#hints"><i class="fas fa-link"></i></a>
</h3>
<ul>
<li>When developing the <code>period_to_underscore()</code> function, you might find the
<a href="https://stringr.tidyverse.org/reference/str_replace.html"><code>stringr::str_replace_all()</code></a>
function helpful. The <code>pattern</code> argument to these functions is interpreted as
a <a href="data-wrangling.html#regular-expressions">regular expression</a> or “regex” for short. A regular
expression is a sequence of characters (i.e. a string) written in a language
that describes patterns in text, similar to “Find and Replace” operations in
word processing software, but is more powerful and flexible in the kinds of
patterns it can detect. Some characters have special meaning in regular
expressions, one of which is the <code>.</code> character. In order to identify the literal
period character like we are trying to do, we must instruct the regular
expression to do so by either escaping the character with <code>\\.</code> or place it in a
range with <code>[.]</code>. Either of these two methods will work to replace a literal <code>.</code>
with <code>_</code>. See the section on <a href="data-wrangling.html#regular-expressions">Regular expressions</a> for more information.</li>
<li>If you are getting <strong>type conversion</strong> errors when loading in your expression
CSV file, check to make sure you aren’t supplying your own column names. The
CSV file has column headers already, and supplying your own will cause the first
line to be read in as data. Since the first row are character values in this
case, all of the other values in the columns will be coerced to characters as
well, instead of reading them in as numbers.</li>
</ul>
</div>
</div>
<div id="assignment-2" class="section level2 unnumbered">
<h2>Assignment 2<a class="anchor" aria-label="anchor" href="#assignment-2"><i class="fas fa-link"></i></a>
</h2>
<div id="problem-statement-1" class="section level3 unnumbered">
<h3>Problem Statement<a class="anchor" aria-label="anchor" href="#problem-statement-1"><i class="fas fa-link"></i></a>
</h3>
<p>Arranging the structure of our data inputs is vital to using R, as is creating
the correct environment for analysis by installing packages. It is also
important to share data and results easily using R markdown.</p>
</div>
<div id="learning-objectives" class="section level3 unnumbered">
<h3>Learning Objectives<a class="anchor" aria-label="anchor" href="#learning-objectives"><i class="fas fa-link"></i></a>
</h3>
<ul>
<li>Install various packages needed for analysis</li>
<li>Load data, filter that data, and retrieve HGNC ids for the data</li>
<li>Create a small plot to display results</li>
<li>Utilize R Markdown to create an attractive format for sharing data.</li>
</ul>
</div>
<div id="skill-list" class="section level3 unnumbered">
<h3>Skill List<a class="anchor" aria-label="anchor" href="#skill-list"><i class="fas fa-link"></i></a>
</h3>
<ul>
<li>How to utilize R markdown to create a report of data analysis</li>
<li>Installing and loading packages in R</li>
<li>Utilizing Bioconductor to equate affy ids to HGNC ids (gene names)</li>
</ul>
</div>
<div id="instructions-1" class="section level3 unnumbered">
<h3>Instructions<a class="anchor" aria-label="anchor" href="#instructions-1"><i class="fas fa-link"></i></a>
</h3>
<p>Our main focus for this assignment are installing packages, manipulating data,
and plotting our manipulated data. We will be borrowing from the same CSV of
expression data from BF528’s Project 1, available
<a href="https://bf528.readthedocs.io/en/latest/content/projects/project_1_microarrays/project_1_microarrays.html">here</a>.<br>
An empty project repo can be found here:</p>
<p><a href="https://github.com/bu-bioinfo/bf591-assignment-2">https://github.com/bu-bioinfo/bf591-assignment-2</a></p>
<p>The project is laid out as such:</p>
<pre><code>main.R
test_main.R
report.Rmd</code></pre>
<p>A skeleton of the functions you need to complete is in <code>main.R</code>. Tests have been
pre-written to test your code and help you ensure it is running correctly, these
are in <code>test_main.R</code>. Finally, we are also introducing the concept of R
Markdown, which for this assignment is <code>report.Rmd</code>. The document itself goes
into greater detail, but you will:</p>
<ol style="list-style-type: decimal">
<li>Complete the functions in <code>main.R</code> and use
<code>testthat:test_file('test_main.R')</code> to ensure they work correctly.<br>
</li>
<li>Read the R Markdown file and complete the section called <strong>“Assignment”</strong>. To
do this, you can <code>source('main.R')</code> to bring over the functions you wrote in
step one.<br>
</li>
<li>Finally, annotate the functions you wrote and <strong>Knit</strong> the R Markdown report,
complete with your additional comments and code execution.</li>
</ol>
<p>This page will go into detail on how the functions and their associated tests
should work.</p>
</div>
<div id="function-details" class="section level3 unnumbered">
<h3>Function Details<a class="anchor" aria-label="anchor" href="#function-details"><i class="fas fa-link"></i></a>
</h3>
<div id="bioconductor" class="section level4 unnumbered">
<h4>1. Bioconductor<a class="anchor" aria-label="anchor" href="#bioconductor"><i class="fas fa-link"></i></a>
</h4>
<p>While many useful R packages can be loaded through
<a href="https://cran.r-project.org/">CRAN</a> using the <code><a href="https://rdrr.io/r/utils/install.packages.html">install.packages()</a></code> syntax, a lot
of specifically bioinformatics packages are exclusively released on
<a href="https://bioconductor.org/install/">Bioconductor</a>. For this assignment we only
need the package called
<a href="https://bioconductor.org/packages/release/bioc/html/biomaRt.html">bioma<strong>R</strong>t</a>.
R programmers fancy themselves very clever, so <strong>R</strong>s show up a lot.</p>
<p>Naturally, we want to load our packages at the beginning of our script so all of
the code we write beneath can access it as it runs. However, if a user or
ourselves <em>already</em> has this package installed we don’t want to waste their time
installing it again. The function <code><a href="https://rdrr.io/r/base/library.html">require()</a></code> can help us avoid unnecessary
installation time and will help us develop faster. The Bioconductor link above
has an example of this method.</p>
<p>This section is untested.</p>
</div>
<div id="load_expression" class="section level4 unnumbered">
<h4>2. <code>load_expression()</code><a class="anchor" aria-label="anchor" href="#load_expression"><i class="fas fa-link"></i></a>
</h4>
<p>Perhaps the most integral part of using the many data wrangling abilities of R
is actually entering your data into the R environment. While there are many ways
to do this in R, we ultimately want this data to be in a <strong>tibble</strong>, which means
the current form of the CSV will make R very angry if you attempt to load it in.
This is because tibbles don’t support row names very well, and the first column
of our data doesn’t have a name. Try to use the <code>load_expression()</code> data to load
data from a <code>filename</code> parameter and <strong>return</strong> a tibble of that information. I
called my firs column “probeids.”</p>
<div id="tests" class="section level5 unnumbered">
<h5>Tests<a class="anchor" aria-label="anchor" href="#tests"><i class="fas fa-link"></i></a>
</h5>
<p>The tests for this function are:</p>
<div class="sourceCode" id="cb519"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">test_that</span><span class="op">(</span><span class="st">"loading csv correctly"</span>, <span class="op">{</span>
  <span class="va">result_tib</span> <span class="op">&lt;-</span> <span class="fu">load_expression</span><span class="op">(</span><span class="st">"/project/bf528/project_1/data/example_intensity_data.csv"</span><span class="op">)</span>
  <span class="fu">expect_equal</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">result_tib</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html">c</a></span><span class="op">(</span><span class="fl">54675</span>, <span class="fl">36</span><span class="op">)</span><span class="op">)</span>
  <span class="fu">expect_true</span><span class="op">(</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/is_tibble.html">is_tibble</a></span><span class="op">(</span><span class="va">result_tib</span><span class="op">)</span><span class="op">)</span>
<span class="op">}</span><span class="op">)</span></code></pre></div>
<p>This test uses the <code>load_expression()</code> you write to store the returned tibble in <code>result_tib</code>.</p>
<div class="box important">
<p>While this test is using the same file for data input as you are, this may not always be the case.</p>
</div>
<p>The test then compares the dimensions of that result, and expects 54,675 rows
and 36 columns. These are the dimensions of the input CSV. It also checks to
confirm it is a tibble object (because tibbles are better than dataframes).</p>
</div>
</div>
<div id="filter_15" class="section level4 unnumbered">
<h4>3. <code>filter_15()</code><a class="anchor" aria-label="anchor" href="#filter_15"><i class="fas fa-link"></i></a>
</h4>
<p>In order to filter the numerous rows we have for this data, we introduce a
function that filters the probe IDs in the tibble our data is stored in. We want
to capture probes that have a suitably high level of expression, so we are
setting <code>log2(15)</code> as the cutoff for an expression level. We will keep a row if
15% of the values in that row exceed <code>log2(15)</code> (about 3.9). Since we may want
to examine the probe IDs we find, the function simple returns the values of the
probe IDs (column 1) instead of returning the entire tibble.</p>
<p>This function presents an important concept in R: using built-ins to speed up
our code. Built-ins are functions and packages that are optimized to process
data in a certain way. Since we’re looking at each row of a table, we could
simply use a for loop to iterate one row at a time. This is slow, though, and
for this function might take 5-10 seconds to run (a long time for a program like
this!). Instead, you could use a function like <code><a href="https://rdrr.io/r/base/apply.html">apply()</a></code> or <code><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html">lapply()</a></code> to filter
every row at once. This solution takes mere moments.</p>
<div id="tests-1" class="section level5 unnumbered">
<h5>Tests<a class="anchor" aria-label="anchor" href="#tests-1"><i class="fas fa-link"></i></a>
</h5>
<div class="sourceCode" id="cb520"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://tibble.tidyverse.org/">tibble</a></span><span class="op">)</span>
<span class="fu">test_that</span><span class="op">(</span><span class="st">"the correct rows are filtered out in filter_15()"</span>, <span class="op">{</span>
  <span class="va">test_tib</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html">tibble</a></span><span class="op">(</span>probeids<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html">c</a></span><span class="op">(</span><span class="st">'1_s_at'</span>, <span class="st">'2_s_at'</span>, <span class="st">'3_s_at'</span>, <span class="st">'4_s_at'</span><span class="op">)</span>,
                     GSM1<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html">c</a></span><span class="op">(</span><span class="fl">1.0</span>, <span class="fl">3.95</span>, <span class="fl">4.05</span>, <span class="fl">0.5</span><span class="op">)</span>,
                     GSM2<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html">rep</a></span><span class="op">(</span><span class="fl">1.6</span>, <span class="fl">4</span><span class="op">)</span>,
                     GSM3<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html">rep</a></span><span class="op">(</span><span class="fl">2.5</span>, <span class="fl">4</span><span class="op">)</span>,
                     GSM4<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html">rep</a></span><span class="op">(</span><span class="fl">3.99</span>, <span class="fl">4</span><span class="op">)</span>,
                     GSM5<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html">rep</a></span><span class="op">(</span><span class="fl">3.0</span>, <span class="fl">4</span><span class="op">)</span>,
                     GSM6<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html">rep</a></span><span class="op">(</span><span class="fl">1.0</span>, <span class="fl">4</span><span class="op">)</span>,
                     GSM7<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html">rep</a></span><span class="op">(</span><span class="fl">0.5</span>, <span class="fl">4</span><span class="op">)</span><span class="op">)</span>
  <span class="fu">expect_equal</span><span class="op">(</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html">pull</a></span><span class="op">(</span><span class="fu">filter_15</span><span class="op">(</span><span class="va">test_tib</span><span class="op">)</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html">c</a></span><span class="op">(</span><span class="st">"2_s_at"</span>, <span class="st">"3_s_at"</span><span class="op">)</span><span class="op">)</span>
<span class="op">}</span><span class="op">)</span></code></pre></div>
<p>In order to test this function, we create a small sample tibble of expression
data containing only seven samples and four IDs. Two of the rows <em>do</em> have more
than 15% of their values exceeding <code>log2(15)</code>, the other two do not. This test
ensures that <code>filter_15()</code> selects the correct rows. Creating a small sample
table like this can be very useful when testing your own code since you don’t
need to look at a large amount of data to see if it’s working correctly or not.</p>
</div>
</div>
<div id="affy_to_hgnc" class="section level4 unnumbered">
<h4>4. <code>affy_to_hgnc()</code><a class="anchor" aria-label="anchor" href="#affy_to_hgnc"><i class="fas fa-link"></i></a>
</h4>
<p>This is an important, but sometimes painful, part of using R. There is a great
built-in package for connecting to Ensembl (a database of genomic information
for many species) called <code>biomaRt</code>. We will use <code>biomaRt</code> to connect the
affymetrix probe IDs to more recognizable HGNC gene IDs. The problem is that
<code>biomaRt</code> depends on an external API (application program interface) to retrieve
data, and this connection sometimes (oftentimes) doesn’t work. While their may
be more nuanced approaches to an unstable resource like this like automatically
retrying failed connections, the best advice for the time being is to try
running this function a few times if it doesn’t work at first. The errors are
clear when it comes to a failed connection, so know that when you get to this
stage it likely isn’t your code’s fault.</p>
<p>To build a <code>biomaRt</code> query, read the documentation in section 3
<a href="https://bioconductor.org/packages/release/bioc/vignettes/biomaRt/inst/doc/accessing_ensembl.html#how-to-build-a-biomart-query">here</a>.
The <code>biomart</code> you should use is <code>ENSEMBL_MART_ENSEMBL</code>, the data set
<code>hsapiens_gene_ensembl</code>, and you want to find the attributes
<code>c("affy_hg_u133_plus_2", "hgnc_symbol")</code>. The data you filter using
<code>filter_15()</code> returns a list of <code>affy_hg_u133_plus_2</code> probe IDs, and the gene
names we’re interested in are stored in <code>hgnc_symbol</code>.</p>
<p>This function should return a tibble, but <code>biomaRt</code>’s <code>getBM()</code> will only accept
and return a <code>data.frame</code>. You can use <code><a href="https://dplyr.tidyverse.org/reference/pull.html">dplyr::pull()</a></code> to turn a tibble into a
simple character vector, and <code><a href="https://tibble.tidyverse.org/reference/as_tibble.html">dplyr::as_tibble()</a></code> to go from a data frame to a
tibble.</p>
<div id="tests-2" class="section level5 unnumbered">
<h5>Tests<a class="anchor" aria-label="anchor" href="#tests-2"><i class="fas fa-link"></i></a>
</h5>
<div class="sourceCode" id="cb521"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">test_that</span><span class="op">(</span><span class="st">"affy ids can be converted to HGNC names properly using affy_to_hgnc()"</span>, <span class="op">{</span>
  <span class="co"># biomaRt super buggy so we can try to account for not connecting well</span>
  <span class="va">response</span> <span class="op">&lt;-</span> <span class="kw"><a href="https://rdrr.io/r/base/try.html">try</a></span><span class="op">(</span><span class="fu">affy_to_hgnc</span><span class="op">(</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html">tibble</a></span><span class="op">(</span><span class="st">'1553551_s_at'</span><span class="op">)</span><span class="op">)</span>, <span class="cn">TRUE</span><span class="op">)</span>
  <span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/grep.html">grepl</a></span><span class="op">(</span><span class="st">"Error"</span>, <span class="va">response</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span>
    <span class="fu">expect_warning</span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/warning.html">warning</a></span><span class="op">(</span><span class="st">"Could not connect to ENSEMBL."</span><span class="op">)</span><span class="op">)</span>
  <span class="op">}</span>
  <span class="kw">else</span> <span class="op">{</span>
    <span class="fu">expect_equal</span><span class="op">(</span><span class="va">response</span><span class="op">$</span><span class="va">hgnc_symbol</span>, <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html">c</a></span><span class="op">(</span><span class="st">"MT-ND1"</span>, <span class="st">"MT-TI"</span>, <span class="st">"MT-TM"</span>, <span class="st">"MT-ND2"</span><span class="op">)</span><span class="op">)</span>
  <span class="op">}</span>
<span class="op">}</span><span class="op">)</span></code></pre></div>
<p>As fun as it is to try to get <code>biomaRt</code> function to connect correctly, it is
even more fun to test them. Since a failure to connect doesn’t indicate an
actual failure in <em>our</em> code we must use a <code><a href="https://rdrr.io/r/base/try.html">try()</a></code> block in order to capture if
there is a connection error.</p>
<div class="box note">
<p><strong>Note</strong> that the <code><a href="https://rdrr.io/r/base/try.html">try()</a></code> function is a part of programming called
<em><strong>error-handling</strong> which extends to many other languages. We often expect
</em>errors when running our programs (such as right now) but don’t want to shut
<em>down our entire operation if it’s an error we can expect. Using <code>try</code>,
</em><code>except</code>, and <code>finally</code> (the latter two not appearing here) we can account for
*issues outside of our control and adapt our code to change the outcome.</p>
<p>Using <code>try except</code> is <strong><em>not</em></strong> a replacement for writing code that doesn’t
generate errors. If you can avoid an error in the first place, that is far
better than using error-handling.</p>
</div>
<p>In this case we <code><a href="https://rdrr.io/r/base/try.html">try()</a></code> to use <code>affy_to_hgnc()</code> to connect to Ensembl and store
the resulting error in <code>response</code>. We then check: if there is an “Error” then we
throw a <code><a href="https://rdrr.io/r/base/warning.html">warning()</a></code> to our testing output. This doesn’t stop further testing
from happening, but it ensures we know that something isn’t quite right. If the
<code>response</code> does not contain an error, we simply test that it returned the
correct gene symbols for our random affy probe ID of choice.</p>
</div>
</div>
<div id="reduce_data" class="section level4 unnumbered">
<h4>5. <code>reduce_data()</code><a class="anchor" aria-label="anchor" href="#reduce_data"><i class="fas fa-link"></i></a>
</h4>
<p>We have one final step in manipulating our data before we plot it. We have our
original data, the probe IDs and their associated HGNC symbols, and a list of
good gene names and bad gene names. <code>reduce_data</code> takes these four inputs and
returns a tibble that has reduced our expression data to only the genes of
interest and has a column describing which set of genes it belongs to (“good” or
“bad”). Changing the shape of the data is incredibly useful for <code>ggplot</code>, the
tidyverse package we will use for plotting. While there is flexibility when
using <code>ggplot</code> to plot data, having data in
<a href="https://www.datacamp.com/community/tutorials/long-wide-data-R">long</a> format is
typically ideal.</p>
<p>Once again, there are multiple ways to reorganize data in this way. We used the
base function <code><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html">match()</a></code> to connect our probe IDs to with our HGNC IDs in
<code>name_ids</code>. We then used <code><a href="https://tibble.tidyverse.org/reference/add_column.html">tibble::add_column()</a></code> to insert the new data in the
correct location. Finally, we created two tibbles of good and bad genes using
<code><a href="https://rdrr.io/pkg/BiocGenerics/man/which.html">which()</a></code> and the <code>%in%</code> modifier. Which evaluates true conditions across a
range of data, so we can pass the list of genes we want and select the correct
ones. For instance:</p>
<div class="sourceCode" id="cb522"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://tibble.tidyverse.org/">tibble</a></span><span class="op">)</span>
<span class="va">tib</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html">tibble</a></span><span class="op">(</span>gene <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html">c</a></span><span class="op">(</span><span class="st">"gene1"</span>, <span class="st">"gene2"</span>, <span class="st">"gene3"</span>, <span class="st">"gene4"</span><span class="op">)</span>,
              affy <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html">c</a></span><span class="op">(</span><span class="st">"a_s_1"</span>, <span class="st">"a_s_2"</span>, <span class="st">"a_s_3"</span>, <span class="st">"a_s_4"</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/which.html">which</a></span><span class="op">(</span><span class="va">tib</span><span class="op">$</span><span class="va">gene</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html">%in%</a></span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html">c</a></span><span class="op">(</span><span class="st">"gene2"</span>, <span class="st">"gene3"</span><span class="op">)</span><span class="op">)</span> <span class="co"># returns the index of the TRUE rows</span></code></pre></div>
<pre><code>## [1] 2 3</code></pre>
<div class="sourceCode" id="cb524"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">tib</span><span class="op">$</span><span class="va">affy</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/which.html">which</a></span><span class="op">(</span><span class="va">tib</span><span class="op">$</span><span class="va">gene</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html">%in%</a></span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html">c</a></span><span class="op">(</span><span class="st">"gene2"</span>, <span class="st">"gene3"</span><span class="op">)</span><span class="op">)</span><span class="op">]</span> <span class="co"># use [] to get data back</span></code></pre></div>
<pre><code>## [1] "a_s_2" "a_s_3"</code></pre>
<p>Once again, there are many ways to reshape this data (some maybe more elegant
than this!) and all we need is the data to be correctly shaped when it is
returned.</p>
<div id="tests-3" class="section level5 unnumbered">
<h5>Tests<a class="anchor" aria-label="anchor" href="#tests-3"><i class="fas fa-link"></i></a>
</h5>
<div class="sourceCode" id="cb526"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">test_that</span><span class="op">(</span><span class="st">"reduce_data() is correctly changing the size and shape of the tibble"</span>, <span class="op">{</span>
  <span class="va">t_tibble</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html">tibble</a></span><span class="op">(</span>probeids <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html">c</a></span><span class="op">(</span><span class="st">"1_s_at"</span>, <span class="st">"2_s_at"</span>, <span class="st">"3_s_at"</span><span class="op">)</span>,
                     GSM1 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html">c</a></span><span class="op">(</span><span class="fl">9.5</span>, <span class="fl">7.6</span>, <span class="fl">5.5</span><span class="op">)</span>,
                     GSM2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html">c</a></span><span class="op">(</span><span class="fl">9.7</span>, <span class="fl">7.2</span>, <span class="fl">2.9</span><span class="op">)</span>,
                     GSM3 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html">c</a></span><span class="op">(</span><span class="fl">6.9</span>, <span class="fl">4.3</span>, <span class="fl">6.8</span><span class="op">)</span><span class="op">)</span>
  <span class="va">names</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html">tibble</a></span><span class="op">(</span>affy_hg_u133_plus_2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html">c</a></span><span class="op">(</span><span class="st">"1_s_at"</span>, <span class="st">"2_s_at"</span>, <span class="st">"3_s_at"</span><span class="op">)</span>,
                  hgnc_symbol <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html">c</a></span><span class="op">(</span><span class="st">"A-REAL-GENE"</span>, <span class="st">"SONIC"</span>, <span class="st">"UTWPU"</span><span class="op">)</span><span class="op">)</span>
  <span class="va">good</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html">c</a></span><span class="op">(</span><span class="st">"A-REAL-GENE"</span><span class="op">)</span>
  <span class="va">bad</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html">c</a></span><span class="op">(</span><span class="st">"SONIC"</span><span class="op">)</span>
  <span class="va">reduce_test</span> <span class="op">&lt;-</span> <span class="fu">reduce_data</span><span class="op">(</span><span class="va">t_tibble</span>, <span class="va">names</span>, <span class="va">good</span>, <span class="va">bad</span><span class="op">)</span>
  <span class="va">result</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html">tibble</a></span><span class="op">(</span>probeids <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html">c</a></span><span class="op">(</span><span class="st">"1_s_at"</span>, <span class="st">"2_s_at"</span><span class="op">)</span>,
                   hgnc_symbol <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html">c</a></span><span class="op">(</span><span class="st">"A-REAL-GENE"</span>, <span class="st">"SONIC"</span><span class="op">)</span>,
                   gene_set <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html">c</a></span><span class="op">(</span><span class="st">"good"</span>, <span class="st">"bad"</span><span class="op">)</span>,
                   GSM1 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html">c</a></span><span class="op">(</span><span class="fl">9.5</span>, <span class="fl">7.6</span><span class="op">)</span>,
                   GSM2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html">c</a></span><span class="op">(</span><span class="fl">9.7</span>, <span class="fl">7.2</span><span class="op">)</span>,
                   GSM3 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html">c</a></span><span class="op">(</span><span class="fl">6.9</span>, <span class="fl">4.3</span><span class="op">)</span><span class="op">)</span>
  <span class="fu">expect_equal</span><span class="op">(</span><span class="va">reduce_test</span>, <span class="va">result</span><span class="op">)</span>
<span class="op">}</span><span class="op">)</span></code></pre></div>
<p>In order to test a function that changes a tibble, we need to use a tibble. We
only create a test table that is three rows by four columns, but that is enough
to get the gist of the function. We simply pass the four parameters to
<code>reduce_data()</code> and we expect it to create a tibble like <code>result</code>. <strong>Ensure your
output column names are the same here or your tests my fail.</strong> While not crucial
to the success of your assignment, maintaining correct column names across
multiple data transformations is an important skill.</p>

</div>
</div>
</div>
</div>
<div id="assignment-3" class="section level2 unnumbered">
<h2>Assignment 3<a class="anchor" aria-label="anchor" href="#assignment-3"><i class="fas fa-link"></i></a>
</h2>
<div id="problem-statement-2" class="section level3 unnumbered">
<h3>Problem Statement<a class="anchor" aria-label="anchor" href="#problem-statement-2"><i class="fas fa-link"></i></a>
</h3>
<p>High dimensional data is typically time and resource-intensive to analyze,
interpret and visualize. Gene expression data (including microarray data) often
consists of measurements for tens of thousands of known genes for every sample.
Learning methods to inspect, reduce and display high dimensional data is
necessary for many machine learning and bioinformatics problems.</p>
</div>
<div id="learning-objectives-1" class="section level3 unnumbered">
<h3>Learning Objectives<a class="anchor" aria-label="anchor" href="#learning-objectives-1"><i class="fas fa-link"></i></a>
</h3>
<ol style="list-style-type: decimal">
<li>Understand the basic principles of Principal Component Analysis (PCA) and hierarchical clustering / heatmaps</li>
<li>Perform PCA in R and evaluate the basic outputs of PCA</li>
<li>Generate basic clustered heatmaps</li>
</ol>
</div>
<div id="skill-list-1" class="section level3 unnumbered">
<h3>Skill List<a class="anchor" aria-label="anchor" href="#skill-list-1"><i class="fas fa-link"></i></a>
</h3>
<ol style="list-style-type: decimal">
<li>R: <code><a href="https://rdrr.io/r/base/scale.html">scale()</a></code>, transpose <code><a href="https://rdrr.io/pkg/BiocGenerics/man/t.html">t()</a></code>, <code><a href="https://rdrr.io/r/stats/prcomp.html">prcomp()</a></code>, <code><a href="https://rdrr.io/r/stats/heatmap.html">heatmap()</a></code>, ggplot2</li>
</ol>
</div>
<div id="background-on-microarrays" class="section level3 unnumbered">
<h3>Background on Microarrays<a class="anchor" aria-label="anchor" href="#background-on-microarrays"><i class="fas fa-link"></i></a>
</h3>
<p>A <a href="biology-bioinformatics.html#microarrays">microarray</a> consists of thousands of specifically
designed single stranded DNA sequences affixed or bound to a solid surface
(glass, nylon, etc.). Extracted RNA or DNA from samples of interest are labeled
with fluorescent dyes, and hybridized with the bound probes by flowing across
the surface. Non-hybridized molecules are washed away and b a laser will excite
the attached dye to produce light to be detected by a scanner and converted to a
digital image. Finally, image processing will transform each spot (bound probe)
to a numerical measurement that can be used to infer expression levels for
genes.</p>
</div>
<div id="background-on-principal-component-analysis" class="section level3 unnumbered">
<h3>Background on Principal Component Analysis<a class="anchor" aria-label="anchor" href="#background-on-principal-component-analysis"><i class="fas fa-link"></i></a>
</h3>
<p><a href="data-science.html#principal-component-analysis">Principal component analysis</a> is an exploratory
data analysis technique commonly used to reduce the dimensionality of data while
trying to simultaneously minimize information loss. To understand the inner
workings of PCA will require an in-depth review of linear algebra and is beyond
the scope of this specific assignment. We will provide several links and
references if you wish to do so on your own. At its core, PCA creates new
uncorrelated linear combinations of the original variables that successively
maximize variance. Thus, the first principle component (PC) represents the
direction of the data that explains a maximal amount of variance. The second PC
represents the direction that captures the second most variance and so on and so
forth. As you can infer, if a small number of PCs capture a majority of variance
contained within a dataset, they can be used as a lower dimensional
representation of the data.</p>
</div>
<div id="marisa-et-al.-gene-expression-classification-of-colon-cancer-into-molecular-subtypes-characterization-validation-and-prognostic-value.-plos-medicine-may-2013.-pmid-23700391" class="section level3 unnumbered">
<h3>Marisa et al. Gene Expression Classification of Colon Cancer into Molecular Subtypes: Characterization, Validation, and Prognostic Value. PLoS Medicine, May 2013. PMID: 23700391<a class="anchor" aria-label="anchor" href="#marisa-et-al.-gene-expression-classification-of-colon-cancer-into-molecular-subtypes-characterization-validation-and-prognostic-value.-plos-medicine-may-2013.-pmid-23700391"><i class="fas fa-link"></i></a>
</h3>
<p>The example intensity data was taken from the listed publication. In it, the
authors proposed the use of gene expression profiling (through microarray
technology) to generate a robust and reproducible classifier to identify
subtypes of colorectal cancer samples. They identified six subtypes that they
demonstrated to be significantly associated with distinct molecular pathways and
clinical pathologies. We have provided you a sample expression matrix that was
normalized and subjected to batch correction. We will use the
<code>example_intensity_data.csv</code> as our data matrix going forward.</p>
</div>
<div id="scaling-data-using-r-scale" class="section level3 unnumbered">
<h3>Scaling data using R <code>scale()</code><a class="anchor" aria-label="anchor" href="#scaling-data-using-r-scale"><i class="fas fa-link"></i></a>
</h3>
<p>Oftentimes, it is necessary to analyze or plot data that contains variables that
differ by multiple orders of magnitude. In bioinformatics, this is an extremely
common situation as different genes may be expressed at wildly different levels.
For instance, in typical HTS experiments and as you will see later on in the
course, you may detect genes with counts ranging from the single digits to those
with counts in the tens of thousands. Heatmaps of gene expression data are
commonly transformed by standard scaling in order to improve the visualization
while retaining the underlying pattern of expression between samples.</p>
<p>One such way to standardize data is the <code><a href="https://rdrr.io/r/base/scale.html">scale()</a></code> function in R. Scale
essentially takes a vector of values and determines the mean and standard
deviation of the entire vector. Scale() will then subtract the mean from each
element in the vector and divide each by the standard deviation. Z-scores thus
correspond to the number of standard deviations by which a data point is above
or below the mean of all the observations (i.e. a Z-score of 0 represents a
value equal to the mean and a Z-score of 1 represents a value one standard
deviation greater than the mean value). This has the effect of representing all
of your data points on the same “scale” while preserving the pattern or profile
of differences inherent between values across different observations.</p>
<p>The <a href="https://bu-bioinfo.github.io/r_for_biological_sciences/prog-basics.html#matrices">transpose function <code>t()</code></a>
is a function covered in the textbook for this course. As covered earlier, it
converts a <span class="math inline">\(m \times n\)</span> matrix to <span class="math inline">\(n \times m\)</span>. The built-in R function
<code><a href="https://rdrr.io/r/base/scale.html">scale()</a></code> operates on a column-wise basis. Transpose the matrix such that
scaling occurs within genes rather than samples. You may also use the tidyverse
<a href="data-wrangling.html#rearranging-data">pivot functions</a> to perform this transpose operation if you
prefer.</p>
<p><strong>Using what you’ve learned in the course and prior assignments, read in the
<code>example_intensity_data.csv</code> to a proper format. Return the scaled matrix.</strong></p>
<div class="box hint">
<p>#1: We want to both center and scale our matrix.</p>
</div>
<div class="box important">
<p><strong>Deliverables</strong>
1. The example_intensity_data.csv read in as a <em>dataframe</em>.</p>
</div>
</div>
<div id="proportion-of-variance-explained" class="section level3 unnumbered">
<h3>Proportion of variance explained<a class="anchor" aria-label="anchor" href="#proportion-of-variance-explained"><i class="fas fa-link"></i></a>
</h3>
<p>As mentioned earlier, principal components represent a successively maximal
amount of variance in your dataset. It is often helpful to visualize the
percentage of total variance explained by each principal component. <strong>Write a
function to determine the proportion of variance explained by each principal
component. Use the results generated by this function to create a bar chart
displaying the variance explained by successive PCs. On this same plot, make a
scatter and line plot showing the cumulative proportion of variance explained by
each principal component.</strong></p>
<div class="box hint">
<p>#1: The summary() function in R provides top-level information about model fitting and statistical objects.
#2: You may access the standard deviation of the PCA results object by $sdev</p>
</div>
<div class="box important">
<p><strong>Deliverables</strong>
1. A vector with the variance explained for each PC.
2. A tibble with the labels for each PC, the variance explained and the
cumulative variance explained.
3. A barchart of the proportion of individual variance explained by each
principal component overlayed with a scatter plot of the cumulative sum of the
variance explained for each successive principal component. Label all relevant
axes and provide a descriptive caption for the figure.</p>
</div>
</div>
<div id="plotting-and-visualization-of-pca" class="section level3 unnumbered">
<h3>Plotting and visualization of PCA<a class="anchor" aria-label="anchor" href="#plotting-and-visualization-of-pca"><i class="fas fa-link"></i></a>
</h3>
<p>In typical HTS experiments, PCA is a common tool used to analyze the similarity
between samples and determine if the experimental variable of interest
(genotype, knockout, etc.) represents a large source of variance. One of the
most common visualization is to plot the first two principal components (which
may not always represent a majority of variance in your dataset) against each
other. Typically, one will examine the pattern of clustering in this plot to see
if samples are grouped together in any meaningful pattern according to important
experimental variables. The value or score of each sample in terms of its
principal components may be found in the pca results object accessed by $x.</p>
<div class="box warning">
<p>When plotting individual PCs against one another, remember that different
principal components explain different amounts of variance and so it will often
be the case that the scale of “importance” across the axes will be
different.</p>
</div>
<div class="box important">
<p><strong>Deliverables</strong>
1. A scatterplot of PC1 vs PC2 from the pca_results$x values. Read in the
metadata CSV and use it to annotate each sample with its corresponding
assignment to the SixSubtypesClassification made in the original publication
(samples belong to either c3 or c4). Label the points in the plot by color for
their corresponding Six Subtypes Classification.</p>
</div>
</div>
<div id="hierarchical-clustering-and-heatmaps" class="section level3 unnumbered">
<h3>Hierarchical Clustering and Heatmaps<a class="anchor" aria-label="anchor" href="#hierarchical-clustering-and-heatmaps"><i class="fas fa-link"></i></a>
</h3>
<p>Now that we’ve discussed the basics of PCA, we will move on to discuss the use
of hierarchical clustering and heatmaps. Heatmaps are a common plot used to
quickly visualize the expression and pattern of many relevant observations
across samples. For HTS data, they essentially depict the expression of genes
(rows) for each sample (columns) as colors that denote the magnitude of
expression to enable quick identification of patterns and changes between
samples/experimental groups. Prior to constructing a heatmap, it is common to
use some form of clustering algorithm to learn potential groups and patterns in
the expression data.</p>
<p>One such method is known as hierarchical clustering (specifically focusing on
agglomerative clustering), a method which attempts to cluster data into
hierarchies where single observations begin as separate points and are
successively merged into larger clusters until all points have been grouped.
Typically, agglomerative clustering will begin with a dissimilarity matrix that
defines how “close” all observations are by calculating some “distance” metric
for every pair of observations. There are multiple “distance” metrics that can
be chosen, and one of the most commonly used is simple euclidean distance. A
linkage function will then use these distance metrics to define how clusters are
formed based on these distance values. For example, complete linkage will define
clusters by the maximum value of all pairwise distances between two different
clusters.</p>
<p>We will be using the base R function <a href="data-visualization.html#heatmaps"><code>heatmap()</code></a> which by default
uses euclidean distance and the <code>hclust()`` function to produce a clustered heatmap. The</code>heatmap()` function will also scale your data row-wise to aid in
visualization.</p>
<p>To begin, we have provided you with a representative output of differentially
expressed probes (<code>differential_expression_results.csv</code>) between the C3 and C4
subtypes. <strong>Filter this table to return probes with an adjusted p-value of &lt;
.01. Use what you’ve learned to extract out the normalized intensity values for
this subset of DE probes. Use the <code><a href="https://rdrr.io/r/stats/heatmap.html">heatmap()</a></code> function in R to create a heatmap
of the normalized intensity values for these DE genes.</strong></p>
<div class="box hint">
<p>#1: It is important to remember that commonly used color combinations in biology
(red-green) are quite difficult to discern for those with color blindness. To
simplify this design decision, we are going to use RColorBrewer, which both
generates appropriate color palettes and has a number of built-in color blind
friendly palettes. You may use the command
display.brewer.all(colorblindFriendly=TRUE) to view these curated palettes.</p>
</div>
<div class="box important">
<p><strong>Deliverables</strong>
1. A list of the significant probes from the differential_expression_results.csv.
2. A matrix of the example_intensity_data.csv filtered to only contain probes
contained within the list of significant probes.
3. A heatmap of the normalized intensity values for the differentially expressed probes (adjusted p-value &lt; .01).</p>
</div>
</div>
<div id="references" class="section level3 unnumbered">
<h3>References<a class="anchor" aria-label="anchor" href="#references"><i class="fas fa-link"></i></a>
</h3>
<p><small>
<a href="https://www.genome.gov/about-genomics/fact-sheets/DNA-Microarray-Technology" class="uri">https://www.genome.gov/about-genomics/fact-sheets/DNA-Microarray-Technology</a></small></p>
<p>Govindarajan, R., Duraiyan, J., Kaliyappan, K. &amp; Palanisamy, M. Microarray and its applications. Journal of Pharmacy &amp; Bioallied Sciences 4, S310 (2012).</p>
<p>Trevino, V., Falciani, F. &amp; Barrera-Saldaña, H. A. DNA microarrays: A powerful genomic tool for biomedical and clinical research. Molecular Medicine 13, 527–541 (2007).</p>
<p>Shlens, J. A Tutorial on Principal Component Analysis.(<a href="https://arxiv.org/abs/1404.1100" class="uri">https://arxiv.org/abs/1404.1100</a>)</p>
<p>Lever, J., Krzywinski, M. &amp; Altman, N. Points of Significance: Principal component analysis. Nature Methods 14, 641–642 (2017).</p>
<p>Hastie, T., Hastie, T., Tibshirani, R., &amp; Friedman, J. H. (2001). The elements of statistical learning: Data mining, inference, and prediction. New York: Springer.https://hastie.su.domains/ElemStatLearn/printings/ESLII_print12_toc.pdf</p>
<p></p>

<div class="inline-table"><table class="table table-sm"><tbody>
<tr class="odd">
<td>output:</td>
</tr>
<tr class="even">
<td>html_document:</td>
</tr>
<tr class="odd">
<td>css: style.css</td>
</tr>
</tbody></table></div>
</div>
</div>
<div id="assignment-4" class="section level2 unnumbered">
<h2>Assignment 4<a class="anchor" aria-label="anchor" href="#assignment-4"><i class="fas fa-link"></i></a>
</h2>
<div id="problem-statement-3" class="section level3 unnumbered">
<h3>Problem Statement<a class="anchor" aria-label="anchor" href="#problem-statement-3"><i class="fas fa-link"></i></a>
</h3>
<p>When dealing with gene counts in an mRNA-seq dataset, it is important to
normalize the data before performing any analyses so you can make accurate
comparisons of gene expression between samples. There are a number of different
methods you can use to accomplish and the method you use will depend on the
kinds of samples you have and the analyses you want to perform. This first part
of this assignment will guide you through two commonly used methods: Counts Per
Million and DESeq2 Normalization.</p>
<p>Once you normalize your data you can do exploratory analysis and construct the
appropriate graphs. As bioinformaticians, you will need to be able present your
data in tidy reports with generated plots- sometimes between blocks of text. R
Markdown is a good tool to accomplish this with, allowing you to display tables
and plots alongside the bulk of your writing. You’ve had some exposure to R
Markdown in previous assignments, now you will start writing your own code to
generate the visualizations for your report. It will be important to remember
that you need to keep your code separate depending on its function: the function
declarations and implementation- your code that will work behind-the-scenes to
perform data manipulation and process inputs without user interaction on the
“back end”- should stay in your <code>main.R</code> file while your calls to construct and
display the tibbles and visuals created in the back end should be put in your
<code>report.Rmd</code> file- despite the user not directly manipulating or editing the
outputs you generate, they are still interacting with the display by viewing it
and therefore the lines of code that call the functions to create these outputs
belong in your “front end.”</p>
</div>
<div id="learning-objectives-2" class="section level3 unnumbered">
<h3>Learning Objectives<a class="anchor" aria-label="anchor" href="#learning-objectives-2"><i class="fas fa-link"></i></a>
</h3>
<ul>
<li>Data normalization methods (CPM, DESeq2)</li>
<li>Plotting data in Tidyverse</li>
<li>Displaying results and compiling reports in R Markdown</li>
<li>Application of Tidyverse functions</li>
</ul>
</div>
<div id="skill-list-2" class="section level3 unnumbered">
<h3>Skill List<a class="anchor" aria-label="anchor" href="#skill-list-2"><i class="fas fa-link"></i></a>
</h3>
<ul>
<li>DESeq normalization, referencing the linked Bioconductor vignette if needed</li>
<li>Tibble manipulation</li>
<li>Creating different types of graphs in ggplot2</li>
<li>Running PCA</li>
</ul>
</div>
<div id="instructions-2" class="section level3 unnumbered">
<h3>Instructions<a class="anchor" aria-label="anchor" href="#instructions-2"><i class="fas fa-link"></i></a>
</h3>
<p>An empty project repo can be found here:</p>
<p><a href="https://github.com/BF591-R/bf591-assignment-4">https://github.com/BF591-R/bf591-assignment-4</a></p>
<p>You will be given</p>
<pre><code>main.R
test_main.R
report.Rmd
verse_counts.tsv
example_report.html
</code></pre>
<p>Like previous assignments, the bulk of the work will be done in the skeleton
file <code>main.R</code>, but unlike previous assignments the provided <code>report.Rmd</code> is also
a skeleton file. Both documents will provide the information needed to complete
the assignment. Please ensure that your inputs and outputs match the
specifications listed- or at the very least can properly handle inputs as
instructed and return the expected output since that is how your functions will
be tested.</p>
<p>To help you complete the assignment, you are being provided a sample report and
a set of test functions that are similar to the ones we will use to grade. The
sample report is provided in the appropriately named <code>sample_report.html</code> and
the tests for you to use can be found in <code>test_main.R</code>. Please note that while
the sample report does not display any code, it is okay and- preferred, really-
for your report to <em>display the code blocks</em> like they have done in previous
assignments.</p>
<div id="tasks" class="section level4 unnumbered">
<h4>Tasks<a class="anchor" aria-label="anchor" href="#tasks"><i class="fas fa-link"></i></a>
</h4>
<ol style="list-style-type: decimal">
<li>Implement the functions as described in <code>main.R</code>.<br>
</li>
<li>[optional but highly recommended] Test your functions using `test_main.R</li>
<li>Fill in <code>report.Rmd</code> as instructed. You will need the functions you implemented in <code>main.R</code>
</li>
<li>Knit <code>report.Rmd</code> and create your <code>report.html</code>
</li>
</ol>
</div>
</div>
<div id="deliverables" class="section level3 unnumbered">
<h3>Deliverables<a class="anchor" aria-label="anchor" href="#deliverables"><i class="fas fa-link"></i></a>
</h3>
<ol style="list-style-type: decimal">
<li>
<code>main.R</code> with all of the functions completed</li>
<li>
<code>report.Rmd</code> filled in as instructed</li>
<li>
<code>report.html</code> knitted from your <code>report.Rmd</code>
</li>
</ol>
</div>
<div id="function-details-1" class="section level3 unnumbered">
<h3>Function Details<a class="anchor" aria-label="anchor" href="#function-details-1"><i class="fas fa-link"></i></a>
</h3>
<div id="read_data" class="section level4 unnumbered">
<h4>1. <code>read_data()</code><a class="anchor" aria-label="anchor" href="#read_data"><i class="fas fa-link"></i></a>
</h4>
<p>Load a tsv located at a specified location</p>
<p>Input (1): string path to file</p>
<p>Output: (g x m) tibble</p>
<p>Details: This function will be tested for handling various input strings and
returning the proper output: dimensions, column names, if column ‘gene’ is
located in the first column, and output type tibble. Your test_main will include
additional testing of output column types and lack of row names for your
reference, to help catch errors early on.</p>
</div>
<div id="filter_zero_var_genes" class="section level4 unnumbered">
<h4>2. <code>filter_zero_var_genes()</code><a class="anchor" aria-label="anchor" href="#filter_zero_var_genes"><i class="fas fa-link"></i></a>
</h4>
<p>Filter out genes with zero variance</p>
<p>Input (1): (g x m) tibble</p>
<p>Output: (n x m) tibble</p>
<p>Details: This function will be tested for handling an input (g x m) tibble and
returning the proper output: column names, if column ‘gene’ is located in the
first column, names of ‘genes’ returned, and output type tibble. Your test_main
will include additional testing for row consistency- ensuring that the sample
data still correlates with the gene names- for your reference and to help catch
errors early on</p>
</div>
<div id="timepoint_from_sample" class="section level4 unnumbered">
<h4>3. <code>timepoint_from_sample()</code><a class="anchor" aria-label="anchor" href="#timepoint_from_sample"><i class="fas fa-link"></i></a>
</h4>
<p>Extract time point information from sample name</p>
<p>Input (1): string (length 5) of sample name in format <code>v[A-Z][a-z,1-9]_[1-9]</code>
(In other words: <code>v[α][β]_[γ]</code>, where α is any capital modern English letter, β
is any lower case modern English letter OR any Arabic number 0-9, and γ is any
Arabic number from 0-9)</p>
<p>Output: string (length 2) of substring <code>[A-Z][a-z,1-9]</code> from sample name:</p>
<p>Details: This function will be tested for handling various strings of length 5
in the form <code>v[A-Z][a-z,1-9]_[1-9]</code> and outputting the proper string, preserving
letter case where letter case is provided.</p>
</div>
<div id="sample_replicate" class="section level4 unnumbered">
<h4>4. <code>sample_replicate()</code><a class="anchor" aria-label="anchor" href="#sample_replicate"><i class="fas fa-link"></i></a>
</h4>
<p>Grab sample replicate number from sample name</p>
<p>Input (1): Input (1): string (length 5) of sample name in format
<code>v[A-Z][a-z,1-9]_[1-9]</code> (In other words: <code>v[α][β]_[γ]</code>, where α is any capital
modern English letter, β is any lower case modern English letter OR any Arabic
number 0-9, and γ is any Arabic number from 0-9)</p>
<p>Output: string (length 1 of substring <code>[1-9]</code> from sample name:
<code>v[A-Z][a-z,1-9]_[1-9]</code> (In other words: [γ] from <code>v[α][β]_[γ]</code>)</p>
<p>Details: This function will be tested for handling various strings of length 5
in the form <code>v[A-Z][a-z,1-9]_[1-9]</code> and outputting the proper character string.</p>
</div>
<div id="meta_info_from_labels" class="section level4 unnumbered">
<h4>5. <code>meta_info_from_labels()</code><a class="anchor" aria-label="anchor" href="#meta_info_from_labels"><i class="fas fa-link"></i></a>
</h4>
<p>Generate sample-level metadata from sample names and stores the data into a
tibble. Will include columns named “sample,” “timepoint,” and “replicate” that
store sample names, sample time points, and sample replicate, respectively.</p>
<p>Input (1): Character vector of length <code>_S_</code> of sample names with column names “sample,” “timepoint,” and “replicate”</p>
<p>Output: a <code>(_S_ x 3)</code> tibble</p>
<p>Details: This function will be tested for handling of a character vector of
length <code>_S_</code>, where each element in the vector is a string with a length of 5 in
the form of <code>v[A-Z][a-z,1-9]_[1-9]</code>, and properly outputting a <code>(_S_ x 3)</code>
tibble with columns named “sample,” “timepoint,” and “replicate” and rows that
correspond with the input samples. Your test_main will include additional
testing for the order of elements in column ‘sample’’s correspondence with the
order of elements in the input vector for your reference. The column types of
your output tibble will not be tested.</p>
</div>
<div id="get_library_size" class="section level4 unnumbered">
<h4>6. <code>get_library_size()</code><a class="anchor" aria-label="anchor" href="#get_library_size"><i class="fas fa-link"></i></a>
</h4>
<p>Calculate total read counts for each sample in a counts dataset.</p>
<p>Input (1): a (n x m) tibble of raw read counts</p>
<p>Output: tibble or named vector of read totals from each sample. Vectors must be
length <code>_S_</code>, a tibble can be <code>(1 x _S_)</code> with sample names as columns names OR
<code>(_S_ x 2)</code> where sample name is in the first column and library size is the
second column</p>
<p>Details: This function will be tested for the return of a tibble or named vector
that have sample names which correspond with the appropriate library size.</p>
</div>
<div id="normalize_by_cpm" class="section level4 unnumbered">
<h4>7. <code>normalize_by_cpm()</code><a class="anchor" aria-label="anchor" href="#normalize_by_cpm"><i class="fas fa-link"></i></a>
</h4>
<p>Normalize raw counts data to counts per million using (counts) /
(sample_library_size) * 10^6</p>
<p>Input (1): a (n x m) tibble of raw read counts</p>
<p>Output: a (n x m) tibble with read count normalized to counts per million</p>
<p>Details: This function will be tested to handle a (n x m) tibble. Its output
will be tested for dimensions, column names, location of string and numeric
column(s), performance of the desired equation on numeric columns, and that gene
names still correspond to their rows.</p>
</div>
<div id="deseq_normalize" class="section level4 unnumbered">
<h4>8. <code>deseq_normalize()</code><a class="anchor" aria-label="anchor" href="#deseq_normalize"><i class="fas fa-link"></i></a>
</h4>
<p>Normalize raw counts data using DESeq2</p>
<p>Input (1): a (n x m) tibble of raw read counts</p>
<p>Output: a (n x m) tibble of DESeq2 normalized counts data</p>
<p>Details:This function will be tested to handle a (n x m) tibble. Its output will
be tested for dimensions, column names, location of string and numeric
column(s), performance of the desired equation on numeric columns, and that gene
names still correspond to their rows.</p>
</div>
<div id="plot_pca" class="section level4 unnumbered">
<h4>9. <code>plot_pca()</code><a class="anchor" aria-label="anchor" href="#plot_pca"><i class="fas fa-link"></i></a>
</h4>
<p>Input (3): a <code>(n x _S_)</code> tibble of data, a <code>(_S_ x 3)</code> tibble of sample-level
meta information, and a string</p>
<p>Output: a ggplot scatter plot showing each sample, with PC1 on x-axis and PC2 on y-axis</p>
<p>Details: The output of this will be tested for the appropriate test run and PCs
used to plot. It may be visually inspected as part of your grade</p>
</div>
<div id="plot_sample_distributions" class="section level4 unnumbered">
<h4>10. <code>plot_sample_distributions()</code><a class="anchor" aria-label="anchor" href="#plot_sample_distributions"><i class="fas fa-link"></i></a>
</h4>
<p>Input (3): a <code>(n x _S_)</code> tibble of data, a boolean to determine whether to scale
the ‘y’ axis to log10 values, and a string</p>
<p>Output: a ggplot boxplot that shows gene count distributions</p>
<p>Details: This function will be tested on a <code>(n x _S_)</code> tibble. It will be tested
for functionality of its inputs, handling of data, expected graph elements, and
graph type. It may also be visually inspected as part of your grade</p>
</div>
<div id="plot_variance_vs_mean" class="section level4 unnumbered">
<h4>11. <code>plot_variance_vs_mean()</code><a class="anchor" aria-label="anchor" href="#plot_variance_vs_mean"><i class="fas fa-link"></i></a>
</h4>
<p>Input (3): a <code>(n x _S_)</code> tibble of data, a boolean to determine whether to scale
the ‘y’ axis to log10 values, and a string</p>
<p>Output: a ggplot scatter plot where the x-axis is the rank of gene ordered by
mean count over all samples, and the y-axis is the observed variance of the
given gene. Each dot should have their transparency increased. The scatter plot
should also be accompanied by a line representing the average mean and variance
values</p>
<p>Details: This function will be tested on a <code>(n x _S_)</code> tibble. It will be tested
for functionality of its inputs, handling of data, expected graph elements, and
graph type. It may also be visually inspected as part of your grade</p>
</div>
</div>
<div id="hints-1" class="section level3 unnumbered">
<h3>Hints<a class="anchor" aria-label="anchor" href="#hints-1"><i class="fas fa-link"></i></a>
</h3>
<ul>
<li><p>Make sure you don’t have any rownames on any of the tibbles your function(s)</p></li>
<li><p>return. You should get a warning in your R console if you do.</p></li>
<li><p>Sometimes Tydyverse is the best tool for the job, sometimes it isn’t</p></li>
<li><p>Some tasks are easier to accomplish with dataframes than tibbles. Not everyone
will use these methods, but if you do you are welcome to use dataframes within
a function as long as the inputs and outputs are the ones specified (returning
a dataframe instead of a tibble may points).</p></li>
<li><p>The Bioconductor vignette for DESeq2 is linked in part 3 of your <code>report.Rmd</code>.
<a href="http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#data-transformations-and-visualization">It can also be found
here</a>)</p></li>
<li><p>In a previous version of this assignment, the output of
<code>filter_zero_var_genes()</code> was transformed from <code>(n x m)</code> to <code>(n x _s_)</code> and
called the <code>counts_matrix</code>. One of the TAs thought use of ‘matrix’ would be
confusing (since we want you to use datasets of type <code>tibble</code>) so it was
re-worded, but this piece of trivia might be helpful for interpreting the
Bioconductor vignette- or if you see any sort of ‘matrix’ referenced in
<code>report.Rmd</code> because it got overlooked during editing.</p></li>
<li><p>There may be occasions where you will need to reshape your data from a wide</p></li>
<li><p>format to a long format</p></li>
<li><p>Symbols used in <code>main.R</code>~ <code>g</code>: initial number of Genes, <code>m</code>: initial number of
columns expected when you import <code>verse_counts.tsv</code>, <code>n</code>: number of genes
expected after you filter in part 1b, <code>_S_</code>: number of Samples</p></li>
</ul>
</div>
</div>
<div id="assignment-5" class="section level2 unnumbered">
<h2>Assignment 5<a class="anchor" aria-label="anchor" href="#assignment-5"><i class="fas fa-link"></i></a>
</h2>
<div id="problem-statement-4" class="section level3 unnumbered">
<h3>Problem Statement<a class="anchor" aria-label="anchor" href="#problem-statement-4"><i class="fas fa-link"></i></a>
</h3>
<p>One of the main questions asked by high throughput sequencing is whether some
variable or condition (e.g. treatment/disease status) induces a measurable,
distinguishable change. In the context of mRNAseq, we are interested in whether
we can determine if mRNA levels (what we actually measure) and genes (what we
infer) are significantly up- or downregulated between conditions and if we can
relate these changes back to a biological phenotype either directly or
indirectly.</p>
</div>
<div id="learning-objectives-3" class="section level3 unnumbered">
<h3>Learning Objectives<a class="anchor" aria-label="anchor" href="#learning-objectives-3"><i class="fas fa-link"></i></a>
</h3>
<ol style="list-style-type: decimal">
<li>Understand the methodology and rationale for how DESeq2 normalizes data,
models counts, and performs differential expression</li>
<li>Perform differential expression analysis using DESeq2 and learn one method to
analyze time course data</li>
<li>Understand how to do basic inspection and evaluation of differential
expression (DE) results</li>
</ol>
</div>
<div id="skill-list-3" class="section level3 unnumbered">
<h3>Skill List<a class="anchor" aria-label="anchor" href="#skill-list-3"><i class="fas fa-link"></i></a>
</h3>
<ol style="list-style-type: decimal">
<li>DESeq2</li>
<li>Basic diagnostic plots of DE results</li>
<li>High-quality figures and plots using R and ggplot</li>
</ol>
</div>
<div id="deseq2-background" class="section level3 unnumbered">
<h3>DESeq2 Background<a class="anchor" aria-label="anchor" href="#deseq2-background"><i class="fas fa-link"></i></a>
</h3>
<p>DESeq2 is a well validated and highly cited method for differential expression
analysis of HTS data. The original paper describing the methodology in more
detail can be found here:</p>
<blockquote>
<p>Love, M.I., Huber, W. &amp; Anders, S. Moderated estimation of fold change and
dispersion for RNA-seq data with DESeq2. Genome Biol 15, 550 (2014).
<a href="https://doi.org/10.1186/s13059-014-0550-8" class="uri">https://doi.org/10.1186/s13059-014-0550-8</a></p>
</blockquote>
<p>We will link back to the original DESeq2 vignette at the head of each applicable
section for reference. We strongly encourage you to read the associated
documentation as it succinctly summarizes the important concepts and methodology
underlying each step of the process. It also provides example code snippets and
answers to many common questions regarding the usage of DESeq2 can already be
found in this documentation.</p>
<p>Briefly, at its core, DESeq2 models read counts and tests for differential
expression through negative binomial generalized linear models. DESeq2
normalizes count data by scaling by size factors calculated by the
median-of-ratios (developed in the original publication for DESeq). For
significance testing, DESeq2 performs a Wald test for model coefficients to
generate p-values and these are adjusted for multiple testing correction through
the Benjamini-Hochberg method.</p>
</div>
<div id="generating-a-counts-matrix" class="section level3 unnumbered">
<h3>Generating a counts matrix<a class="anchor" aria-label="anchor" href="#generating-a-counts-matrix"><i class="fas fa-link"></i></a>
</h3>
<p>DESeq2 takes as input a counts matrix where genes correspond to rows and samples
to columns. For one of the more common use cases of differential expression
analysis, mRNAseq experiments, we are interested in counts associated with
genes; however, DESeq2 is broadly applicable to many kinds of high-throughout
sequencing (HTS) count data.</p>
<p>Returning to the analysis of mRNAseq specifically, there are three main
approaches to map reads:</p>
<ol style="list-style-type: decimal">
<li>Alignment to a reference genome by a splice-aware aligner and counting of
reads falling into exonic regions using a reference annotation file</li>
<li>Transcript-based assembly approaches (TopHat, Cufflinks) or transcript abundance
quantification (Salmon, Kallisto)</li>
<li>Reference free assembly followed by mapping
and counting</li>
</ol>
<p>DESeq2 expects the values of the input matrix to be counts (integers) and for
the purposes of this assignment, you have been provided with a count matrix of
samples generated by the first method. Reads were aligned to the mouse genome
(GENCODE GRCm39) with STAR and quantified to gene counts by Verse using the
matching GTF file.</p>
</div>
<div id="prefiltering-counts-matrix" class="section level3 unnumbered">
<h3>Prefiltering Counts matrix<a class="anchor" aria-label="anchor" href="#prefiltering-counts-matrix"><i class="fas fa-link"></i></a>
</h3>
<p><a href="http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#pre-filtering">DESeq2 Vignette - Pre-filtering</a></p>
<p>Unless your experiment is quite large, filtering your counts matrix prior to
DESeq2 is not strictly required. If your experiment incorporates hundreds of
samples and a complicated model design with many interaction terms, you may wish
to perform a combination of manual filtering and further optimization of DESeq2
settings. Otherwise, for most datasets, default settings in DESeq2 will run in a
reasonably short amount of time on relatively modest hardware even without
pre-filtering.</p>
<p>There are many different strategies to filter count datasets and which you
choose to employ should be informed by your objective, and your data. Some
common filters include removing genes where the mean count across all samples is
below some threshold or removing genes based on how many samples have a zero
count. In general, extensive pre-filtering is not required and DESeq2
incorporates downstream methods to increase statistical power through
independent filtering (described in methods). <strong>For this assignment, we will not
be filtering the dataset prior to running DE analysis.</strong></p>
</div>
<div id="median-of-ratios-normalization" class="section level3 unnumbered">
<h3>Median-of-ratios normalization<a class="anchor" aria-label="anchor" href="#median-of-ratios-normalization"><i class="fas fa-link"></i></a>
</h3>
<p><a href="https://genomebiology.biomedcentral.com/articles/10.1186/gb-2010-11-10-r106">Original DESeq Publication</a></p>
<p>As mentioned earlier, DESeq2 expects a non-transformed, non-normalized matrix of
integer “counts” as input. DESeq2 performs its own normalization based on the
underlying assumption that not all or a small amount of genes are truly
differentially expressed. This is encapsulated in the median-of-ratios method
which accounts for library size as well as RNA composition. Although DESeq2
performs this normalization in the background, it can help conceptually to
understand the process by which it does so.</p>
<p>First, a pseudo-reference sample is created and set equal to the geometric mean
across all samples for a given gene. Then, ratios are calculated comparing each
sample to this pseudo-reference on gene-wise basis. Next, on a sample-wise
basis, the median value for all ratios is taken as that sample’s normalization
factor and for a given sample, all counts are divided by this factor. You have
learned how to extract these normalized counts by size factors in the last
assignment.</p>
</div>
<div id="deseq2-preparation" class="section level3 unnumbered">
<h3>DESeq2 preparation<a class="anchor" aria-label="anchor" href="#deseq2-preparation"><i class="fas fa-link"></i></a>
</h3>
<p><a href="http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#count-matrix-input">DESeq2 Vignette - Preparation</a></p>
<p>The DESeqDataSet is the object that holds the read counts and associated
statistical measures generated through the DESeq2 algorithm. Although there are
several ways to generate this object, we will focus on the DESeqDataSet method
which directly takes a SummarizedExperiment object.</p>
</div>
<div id="omeara-et-al.-transcriptional-reversion-of-cardiac-myocyte-fate-during-mammalian-cardiac-regeneration.-circ-res.-feb-2015.-pmid-25477501l" class="section level3 unnumbered">
<h3>O’Meara et al. Transcriptional Reversion of Cardiac Myocyte Fate During Mammalian Cardiac Regeneration. Circ Res. Feb 2015. PMID: 25477501l<a class="anchor" aria-label="anchor" href="#omeara-et-al.-transcriptional-reversion-of-cardiac-myocyte-fate-during-mammalian-cardiac-regeneration.-circ-res.-feb-2015.-pmid-25477501l"><i class="fas fa-link"></i></a>
</h3>
<p>The counts matrix provided was generated by data made available by the listed
publication. We have only taken a subset of their samples, focusing specifically
on the mRNAseq experiment during <em>in vivo</em> heart development from 0,
4, 7 days after birth as well as in adult mice.</p>
<p>To start, we are going to consider the simplest use case for differential
expression and subset our data to contain only the samples from postnatal day 0
and adult hearts. In addition to subsetting the counts matrix, we also need to
construct the sample data information as a dataframe that lists the samplenames
(columns in our counts matrix) as well as their associated information (in our
case, timepoint).</p>
</div>
<div id="reading-and-subsetting-the-data-from-verse_counts.tsv-and-sample_metadata.csv" class="section level3 unnumbered">
<h3>1. Reading and subsetting the data from verse_counts.tsv and sample_metadata.csv<a class="anchor" aria-label="anchor" href="#reading-and-subsetting-the-data-from-verse_counts.tsv-and-sample_metadata.csv"><i class="fas fa-link"></i></a>
</h3>
<p>We will use this as a demonstration of one of the simplest use cases for DESeq2,
comparing two groups of samples based on a single experimental variable. Subset
the samples to only include those from timepoints vP0 and vAd, which should
correspond to a total of 4 samples (vP0_1, vP0_2, vAd_1, vAd_2). Store both
the counts matrix and sample dataframe in a SummarizedExperiment object.</p>
<div class="box hint">
<ol style="list-style-type: decimal">
<li><p>You may ignore most of the columns in the metadata CSV as they were used in
the generation of the data itself. Focus on the samplenames and timepoints
columns.</p></li>
<li><p>The columns (sample names) in your counts matrix and sample dataframe need
to be in the same order</p></li>
<li><p>You may need to convert your counts dataframe to a matrix</p></li>
<li><p>The counts matrix is verse_counts.tsv and the sample metadata is
sample_metadata.csv</p></li>
</ol>
</div>
</div>
<div id="running-deseq2" class="section level3 unnumbered">
<h3>2. Running DESeq2<a class="anchor" aria-label="anchor" href="#running-deseq2"><i class="fas fa-link"></i></a>
</h3>
<div id="understanding-factor-levels" class="section level4 unnumbered">
<h4>Understanding Factor Levels<a class="anchor" aria-label="anchor" href="#understanding-factor-levels"><i class="fas fa-link"></i></a>
</h4>
<p><a href="http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#note-on-factor-levels">DESeq2 Vignette - Factor Levels</a></p>
<p>The factor level in DESeq2 determines which level represents the “control” group
you want to compare against. By default, the reference level for factors will be
chosen alphabetically. You can either manually change and set the reference
factor level or specify a comparison of interest using DESeq2 Contrast() after
performing the first steps of the DE analysis.</p>
</div>
<div id="performing-differential-expression-analysis" class="section level4 unnumbered">
<h4>Performing Differential Expression Analysis<a class="anchor" aria-label="anchor" href="#performing-differential-expression-analysis"><i class="fas fa-link"></i></a>
</h4>
<p>Here you may find the vignette for example usage of DESeq2. Please read the
following section to instruct you on how to perform differential expression
analysis in DESeq2: <a href="http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#differential-expression-analysis">Running
DESeq2</a></p>
<p>We will be running DESeq2 with a model design of <code>~timepoint</code> with vP0 set as
our reference level. This will test for differentially expressed genes in adult
mouse heart ventricles vs. postnatal day 0. Return both the results as a
dataframe and the dds object from running DESeq2 as a list. Assign the results
dataframe and dds object to variables so you can use them later in your Rmd.</p>
</div>
</div>
<div id="annotating-results-to-construct-a-labeled-volcano-plot" class="section level3 unnumbered">
<h3>3. Annotating results to construct a labeled volcano plot<a class="anchor" aria-label="anchor" href="#annotating-results-to-construct-a-labeled-volcano-plot"><i class="fas fa-link"></i></a>
</h3>
<p>Later in the analysis, we will construct a volcano plot. Typically, volcano
plots will label genes by their significance as well as their direction of
change. Convert the results dataframe to a tibble and add a column labeled
<code>volc_plot_status</code> that you can easily input in ggplot2 to color points by the
below criteria.</p>
<div class="box hint">
<ol style="list-style-type: decimal">
<li>The labels will encompass three groups: 1. Positive log fold change and
significant at a given padj threshold, 2. Negative log fold change and
significant at a given padj threshold, 3. Remaining non-significant genes</li>
<li>Ensure your column is named exactly <code>volc_plot_status</code> or the test will not
work without modification.</li>
<li>Have the values for these labels be `UP’, ‘DOWN,’ ‘NS’ respectively.</li>
</ol>
</div>
</div>
<div id="diagnostic-plot-of-the-raw-p-values-for-all-genes" class="section level3 unnumbered">
<h3>4. Diagnostic plot of the raw p-values for all genes<a class="anchor" aria-label="anchor" href="#diagnostic-plot-of-the-raw-p-values-for-all-genes"><i class="fas fa-link"></i></a>
</h3>
<p>Separate from differential expression analysis, it is always an important
diagnostic to plot the unadjusted distribution of p-values obtained from an
experiment. Built in to the definition of a p-value is the idea that under the
null distribution and given all other assumptions made are true, p-values follow
a uniform distribution. In general, you should typically observe a peak of
values close to 0 and a roughly uniform distribution as you approach 1. The
values closer to 0 will be a mix of situations where the alternative hypothesis
is true as well as potential false positives. Significant deviations from this
general pattern likely require close examination of your data and the
significance test you are employing. Make a histogram of the raw p-values from
all the genes discovered in the experiment.</p>
</div>
<div id="plotting-the-logfoldchanges-for-differentially-expressed-genes" class="section level3 unnumbered">
<h3>5. Plotting the LogFoldChanges for differentially expressed genes<a class="anchor" aria-label="anchor" href="#plotting-the-logfoldchanges-for-differentially-expressed-genes"><i class="fas fa-link"></i></a>
</h3>
<p>It is often helpful visualize the log2FoldChanges for your differentially
expressed genes to gain a sense for the distribution of up- and downregulated
genes and gain a global view of how genes are changing. Subset your results to
only include genes significant at a padj threshold of &lt; .10. Plot a histogram of
the log2FoldChange values for these genes.</p>
</div>
<div id="the-choice-of-fdr-cutoff-depends-on-cost" class="section level3 unnumbered box note">
<h3>The choice of FDR cutoff depends on cost<a class="anchor" aria-label="anchor" href="#the-choice-of-fdr-cutoff-depends-on-cost"><i class="fas fa-link"></i></a>
</h3>
<p>The choice of a FDR cutoff or any p-value cutoff is always subjective and based
on the objectives of your experiment and the cost of false discoveries. For
example, if your initial sequencing experiment is exploratory and meant to
generate hypotheses to be followed up <em>in vitro</em> or <em>in vivo</em>, it may be more
appropriate to set a permissive cutoff (padj &lt; .20). P-value thresholds are set
on an experiment by experiment basis and should not be adjusted simply to
increase or decrease the number of DE genes retroactively.</p>
</div>
<div id="plotting-the-normalized-counts-of-differentially-expressed-genes" class="section level3 unnumbered">
<h3>6. Plotting the normalized counts of differentially expressed genes<a class="anchor" aria-label="anchor" href="#plotting-the-normalized-counts-of-differentially-expressed-genes"><i class="fas fa-link"></i></a>
</h3>
<p>It can be helpful to visualize the normalized counts for a few differentially
expressed genes as a quick diagnostic that your analysis is working as intended.
Although you will likely not need to do this for every RNAseq analysis, it can
provide confidence in your results and also help you remember the meaning of the
directionality of your fold change values. Make a scatter / jitter plot of the
DESeq2 normalized counts for the top ten significantly differentially expressed
genes ranked by ascending padj</p>
<div class="box hint">
<ol style="list-style-type: decimal">
<li><p>If you are plotting all of the genes on the same plot, you may find it
helpful to take the logarithm of the counts as different genes often have counts
that differ by several orders of magnitude.</p></li>
<li><p>You may find it helpful to add a slight horizontal “jitter” to the points to
aid in visualization for cases where counts for each replicate are similar.</p></li>
<li><p>You have already used DESeq2 to extract normalized counts. There are also
directions available in the vignette.</p></li>
</ol>
</div>
</div>
<div id="volcano-plot-to-visualize-differential-expression-results" class="section level3 unnumbered">
<h3>7. Volcano Plot to visualize differential expression results<a class="anchor" aria-label="anchor" href="#volcano-plot-to-visualize-differential-expression-results"><i class="fas fa-link"></i></a>
</h3>
<p>A volcano plot is a common data plot that is intended to quickly display points
of interest by plotting statistical significance against a magnitude of change.
In the specific case of mRNAseq data, this entails plotting the -log(p-values /
adjusted p-values) against the estimated log fold changes reported by most
differential expression methods.</p>
<p>Statistically significant genes with low p-values will appear higher on the
y-axis of the plot and they will be separated by the magnitude and
directionality of their change (i.e. Upregulated genes with larger fold change
estimates will be further along to the right of the origin of the x-axis, and
downregulated genes will be farther to the left of the origin on the x-axis).
Typically, genes of interest that have additional biological relevance to the
experiment will be directly annotated using domain knowledge.</p>
<p>Make a volcano plot (log2FoldChange vs. -log10(padj)) for all the genes
discovered in the experiment. Use the column generated in part 3 to color the
genes with the appropriate label.</p>
</div>
<div id="running-fgsea-vignette" class="section level3 unnumbered">
<h3>8. Running fgsea <a href="http://bioconductor.org/packages/release/bioc/vignettes/fgsea/inst/doc/fgsea-tutorial.html">vignette</a><a class="anchor" aria-label="anchor" href="#running-fgsea-vignette"><i class="fas fa-link"></i></a>
</h3>
<p>We will be performing a GSEA on the results generated by running DESeq2 using
the fgsea package. We will use log2FoldChange as our ranking metric and test
against all the pathways in the C2 Canonical Pathways gene set collection
provided by MSigDB and already downloaded here (c2.cp.v7.5.1.symbols.gmt). Read
the section in the textbook or the fgsea documentation for the appropriate
formats for the ranked gene list and the gene sets.</p>
<div class="box hint">
<ol style="list-style-type: decimal">
<li><p>You may read in the gene sets in the proper format using GSEABase or fgsea</p></li>
<li><p>You will need to convert the mouse ensembl gene IDs to human HGNC symbols to
match the genes provided in the C2 Canonical Pathways gene sets. If you are
encountering issues, you may need to remove the version number of the annotation.</p></li>
<li><p>As you have discovered, biomaRt can be difficult to work with. If you are
encountering frequent issues, you may consider mapping your gene IDs and saving
the results locally to avoid having to re-run the biomaRt query multiple times.</p></li>
<li><p>You will inevitably have genes with duplicate names and/or genes with
duplicate values for log2FoldChange. Choose a method to remove these values.</p></li>
</ol>
</div>
</div>
<div id="plotting-the-top-ten-positive-nes-and-top-ten-negative-nes-pathways" class="section level3 unnumbered">
<h3>9. Plotting the top ten positive NES and top ten negative NES pathways<a class="anchor" aria-label="anchor" href="#plotting-the-top-ten-positive-nes-and-top-ten-negative-nes-pathways"><i class="fas fa-link"></i></a>
</h3>
<div class="box hint">
<ol style="list-style-type: decimal">
<li><p>Ensure that the pathway labels are visible</p></li>
<li><p>You will probably get warnings about duplicate values or ranks, do your best
to address these but don’t get hung up on trying to find them all.</p></li>
</ol>
</div>
</div>
<div id="references-1" class="section level3 unnumbered">
<h3>References<a class="anchor" aria-label="anchor" href="#references-1"><i class="fas fa-link"></i></a>
</h3>
<p><small> Dobin A, Davis CA, Schlesinger F, Drenkow J, Zaleski C, Jha S, Batut P,
Chaisson M, Gingeras TR. STAR: ultrafast universal RNA-seq aligner.
Bioinformatics. 2013 Jan 1;29(1):15-21. doi: 10.1093/bioinformatics/bts635. Epub
2012 Oct 25. PMID: 23104886; PMCID: PMC3530905.</small></p>
<p>Zhu, Q., Fisher, S.A., Shallcross, J., Kim, J. (Preprint). VERSE: a versatile
and efficient RNA-Seq read counting tool. bioRxiv 053306. <a href="doi:http://dx.doi.org/10.1101/053306" class="uri">doi:http://dx.doi.org/10.1101/053306</a></p>
<p><a href="https://www.gencodegenes.org/mouse/" class="uri">https://www.gencodegenes.org/mouse/</a></p>
<p>Anders, S. &amp; Huber, W. Differential expression analysis for sequence count data.
Genome Biology 11, 1–12 (2010).</p>
<p>Love, M. I., Huber, W. &amp; Anders, S. Moderated estimation of fold change and
dispersion for RNA-seq data with DESeq2. Genome Biology 15, 550 (2014). </p>

</div>
</div>
<div id="assignment-6" class="section level2 unnumbered">
<h2>Assignment 6<a class="anchor" aria-label="anchor" href="#assignment-6"><i class="fas fa-link"></i></a>
</h2>
<blockquote>
<p>Volcanoes don’t really look like that, do they?</p>
</blockquote>
<div id="problem-statement-5" class="section level3 unnumbered">
<h3>Problem Statement<a class="anchor" aria-label="anchor" href="#problem-statement-5"><i class="fas fa-link"></i></a>
</h3>
<p>Because R has been utilized in bioinformatics communities for years at this point, there are many different approaches to do one task. While in the last assignment we just used DESeq2, we will be performing differential expression analysis with two additional packages (edgeR and Limma) and comparing the results we find.</p>
</div>
<div id="learning-objectives-4" class="section level3 unnumbered">
<h3>Learning Objectives<a class="anchor" aria-label="anchor" href="#learning-objectives-4"><i class="fas fa-link"></i></a>
</h3>
<ol style="list-style-type: decimal">
<li>The basic operations of <code>ggplot</code> and “The Grammar of Graphics.”</li>
<li>Using differential gene expression analysis packages to generate data for plotting.</li>
<li>Creating simple plots and <strong>not</strong> using the default colors and themes.</li>
<li>Combining multiple plots into one image using <code><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap()</a></code>
</li>
</ol>
</div>
<div id="skill-list-4" class="section level3 unnumbered">
<h3>Skill List<a class="anchor" aria-label="anchor" href="#skill-list-4"><i class="fas fa-link"></i></a>
</h3>
<ul>
<li>A tempered heart and mind that understands a plot may not ever look <em>exactly</em> how you want it to.</li>
<li>An intermediate understanding of R’s most popular plotting package, <code>ggplot</code>.</li>
<li>Further understanding of differential expression analysis and its plotting.</li>
<li>A sense of superiority whenever you see a publication use the base R plotting package for figures.</li>
</ul>
</div>
<div id="instructions-3" class="section level3 unnumbered">
<h3>Instructions<a class="anchor" aria-label="anchor" href="#instructions-3"><i class="fas fa-link"></i></a>
</h3>
<p>Complete <code>main.R</code> in a way that satisfies all of the tests in <code>test_main.R</code>. Follow the instructions in the function descriptions for <code>main.R</code> and read here for details on the tests your code should satisfy. You will use these functions to complete figures in <code>report.Rmd</code>.</p>
<p>You can use <code>testthat:test_file('test_main.R')</code> in your R console to run all tests in <code>test_main.R</code>. Do not modify the tests. While it is difficult to write tests that cover every solution to a problem, our tests are written to ensure the outputs of a function are aligned with what an analysis should perform. If you feel a test is working incorrectly, let a TA know.</p>
</div>
<div id="function-details-2" class="section level3 unnumbered">
<h3>Function Details<a class="anchor" aria-label="anchor" href="#function-details-2"><i class="fas fa-link"></i></a>
</h3>
<div id="load_n_trim" class="section level4 unnumbered">
<h4>1. <code>load_n_trim()</code><a class="anchor" aria-label="anchor" href="#load_n_trim"><i class="fas fa-link"></i></a>
</h4>
<p>Once again we have some data to load into R in order to manipulate. We’ll be loading a counts file, which is a matrix of genes (rows) by samples (columns). Each cell is the number of that gene found in that sample. This is a good opportunity to use a data frame or a tibble, but because of package restrictions we will be sticking to data frames for this assignment. We’re using a couple different packages to process our data, and they input/output <em>exclusively</em> in data frames, so we will be consistent and save ourselves some trouble with conversions.</p>
<p>We want to return a data frame that is essentially identical to our input file (always a good idea to use the command line to see what you’re working with). Ensure the gene names are stored as <code>row.names</code> and <em>not</em> as a separate column, data frames have a distinction between the two. This function should also reduce the columns to just those of interest (P0 and Adult).</p>
<div class="box hint">
<p>Row names can be changed using the <code><a href="https://rdrr.io/r/base/row.names.html">row.names()</a></code> function. The new row names must be a character vector of exactly the same length.</p>
<div class="sourceCode" id="cb528"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb528-1"><a href="assignments.html#cb528-1" aria-hidden="true" tabindex="-1"></a>df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">a =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">0</span>, <span class="dv">0</span>), <span class="at">b =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">3</span>, <span class="dv">5</span>))</span>
<span id="cb528-2"><a href="assignments.html#cb528-2" aria-hidden="true" tabindex="-1"></a><span class="fu">row.names</span>(df) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Row1"</span>, <span class="st">"Row2"</span>, <span class="st">"Row3"</span>)</span>
<span id="cb528-3"><a href="assignments.html#cb528-3" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(df)</span>
<span id="cb528-4"><a href="assignments.html#cb528-4" aria-hidden="true" tabindex="-1"></a>     a b</span>
<span id="cb528-5"><a href="assignments.html#cb528-5" aria-hidden="true" tabindex="-1"></a>Row1 <span class="dv">0</span> <span class="dv">1</span></span>
<span id="cb528-6"><a href="assignments.html#cb528-6" aria-hidden="true" tabindex="-1"></a>Row2 <span class="dv">0</span> <span class="dv">3</span></span>
<span id="cb528-7"><a href="assignments.html#cb528-7" aria-hidden="true" tabindex="-1"></a>Row3 <span class="dv">0</span> <span class="dv">5</span></span></code></pre></div>
</div>
<div id="tests-4" class="section level5 unnumbered">
<h5>Tests<a class="anchor" aria-label="anchor" href="#tests-4"><i class="fas fa-link"></i></a>
</h5>
<div class="sourceCode" id="cb529"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">test_that</span><span class="op">(</span><span class="st">"test data is loading correctly and reduced"</span>, <span class="op">{</span>
  <span class="va">test_df</span> <span class="op">&lt;-</span> <span class="fu">load_n_trim</span><span class="op">(</span><span class="va">csv</span><span class="op">)</span>
  <span class="fu">expect_equal</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">test_df</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html">c</a></span><span class="op">(</span><span class="st">"vP0_1"</span>, <span class="st">"vP0_2"</span>, <span class="st">"vAd_1"</span>, <span class="st">"vAd_2"</span><span class="op">)</span><span class="op">)</span>
  <span class="fu">expect_equal</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">test_df</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html">c</a></span><span class="op">(</span><span class="fl">55416</span>, <span class="fl">4</span><span class="op">)</span><span class="op">)</span>
  <span class="fu">expect_equal</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/class.html">class</a></span><span class="op">(</span><span class="va">test_df</span><span class="op">)</span>, <span class="st">"data.frame"</span><span class="op">)</span>
<span class="op">}</span><span class="op">)</span></code></pre></div>
<p>The tests for this assignment are relatively straightforward, mostly ensuring that data is loaded and ends up in the right shape and format. The first test ensures the columns were filtered, the second ensures the size is correct, and the final ensures it is a data frame and <em>not</em> a tibble. No tibbles. Tibble-less?</p>
</div>
</div>
<div id="run_deseq" class="section level4 unnumbered">
<h4>2. <code>run_deseq()</code><a class="anchor" aria-label="anchor" href="#run_deseq"><i class="fas fa-link"></i></a>
</h4>
<p>This is the first of three functions that technically do identical things (but in different ways). One of the most popular Bioconductor packages, DESeq2 has a number of options available for differential expression analysis. We will load in the counts data, select our variables of interest, and use <code><a href="https://rdrr.io/pkg/DESeq2/man/DESeq.html">DESeq()</a></code> to process them. Links are included in the function description to DESeq2 documentation, you will find answers to most questions there, especially since those were the documents used to <em>write</em> this assignment. This function will return the results of the analysis.</p>
<div id="tests-5" class="section level5 unnumbered">
<h5>Tests<a class="anchor" aria-label="anchor" href="#tests-5"><i class="fas fa-link"></i></a>
</h5>
<div class="sourceCode" id="cb530"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">test_that</span><span class="op">(</span><span class="st">"deseq2 function is returning correct results"</span>, <span class="op">{</span>
  <span class="fu"><a href="https://rdrr.io/r/base/load.html">load</a></span><span class="op">(</span><span class="st">"mock_counts_df.RData"</span><span class="op">)</span> <span class="co"># loads the counts_df object into env</span>
  <span class="va">coldata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>condition <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html">c</a></span><span class="op">(</span><span class="st">"day4"</span>, <span class="st">"day7"</span><span class="op">)</span>, each<span class="op">=</span><span class="fl">2</span><span class="op">)</span>,
                        type<span class="op">=</span><span class="st">"paired-end"</span><span class="op">)</span>
  <span class="fu"><a href="https://rdrr.io/r/base/row.names.html">row.names</a></span><span class="op">(</span><span class="va">coldata</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html">c</a></span><span class="op">(</span><span class="st">"vP4_1"</span>, <span class="st">"vP4_2"</span>, <span class="st">"vP7_1"</span>, <span class="st">"vP7_2"</span><span class="op">)</span>
  <span class="fu">expect_warning</span><span class="op">(</span><span class="va">deseq</span> <span class="op">&lt;-</span> <span class="fu">run_deseq</span><span class="op">(</span><span class="va">counts_df</span>,
                                    <span class="va">coldata</span>,
                                    <span class="fl">10</span>, <span class="st">"condition_day7_vs_day4"</span><span class="op">)</span><span class="op">)</span>
  <span class="fu">expect_equal</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">deseq</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html">c</a></span><span class="op">(</span><span class="fl">19127</span>, <span class="fl">6</span><span class="op">)</span><span class="op">)</span>
  <span class="fu">expect_equal</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/class.html">class</a></span><span class="op">(</span><span class="va">deseq</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="st">"DESeqResults"</span><span class="op">)</span>
  <span class="fu">expect_equal</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html">c</a></span><span class="op">(</span><span class="st">"pvalue"</span>, <span class="st">"padj"</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">deseq</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html">c</a></span><span class="op">(</span><span class="cn">TRUE</span>, <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span>
<span class="op">}</span><span class="op">)</span></code></pre></div>
<p>Less concerning for you all but writing tests for larger data sets and more complicated functions like this is…tricky. The problem is it can’t be written like a script, we can’t use the results from <code>load_n_trim()</code> to test <em>this</em> function, because what if one fails independently of the other? So in this case, we create an RData object of correctly loaded data that works <em>even if</em> <code>load_n_trim()</code> is absolutely broken. Also, in this case, we are using different sets to test this (days 4 and 7 instead of 0 and adult) so you can’t just load the test data and say you wrote the code! That would be disingenuous of you, and being disingenuous is for after you graduate.</p>
<p>This test loads the sample data (a data frame of counts) and creates the <code>coldata</code> object according to DESeq2 specifications. We wrap the <code>run_deseq()</code> in <code>expect_warning()</code> because DESeq throws a warning when it performs an expected behavior: converting strings to factors (this is bad programming! this is not what warnings are for, the package is functioning as expected). We then do a similar series of tests: ensure the dimensions are correct, ensure the results object are from DESeq, and ensure that the column names we need are present.</p>
</div>
</div>
<div id="run_edger" class="section level4 unnumbered">
<h4>3. <code>run_edger()</code><a class="anchor" aria-label="anchor" href="#run_edger"><i class="fas fa-link"></i></a>
</h4>
<p>As mentioned in the previous function, this is the EdgeR implementation of running the differential expression analysis based on our counts file. Again, this function should follow roughly the portion of the documentation mentioned in the function description. EdgeR has a number of additional plotting functions (such as <code>plotMDS()</code>) you may like to include. These are optional but pretty fun to see if your data is working out as expected (similar days clustered close together and so on). You will return the <em>entire</em> results object, no filtering necessary yet.</p>
<div id="tests-6" class="section level5 unnumbered">
<h5>Tests<a class="anchor" aria-label="anchor" href="#tests-6"><i class="fas fa-link"></i></a>
</h5>
<div class="sourceCode" id="cb531"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">test_that</span><span class="op">(</span><span class="st">"edger function is returning correct results"</span>, <span class="op">{</span>
  <span class="fu"><a href="https://rdrr.io/r/base/load.html">load</a></span><span class="op">(</span><span class="st">"mock_counts_df.RData"</span><span class="op">)</span>
  <span class="va">group</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span>, each<span class="op">=</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span>
  <span class="va">edger_res</span> <span class="op">&lt;-</span> <span class="fu">run_edger</span><span class="op">(</span><span class="va">counts_df</span>, <span class="va">group</span><span class="op">)</span>
  <span class="fu">expect_equal</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">edger_res</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html">c</a></span><span class="op">(</span><span class="st">"logFC"</span>, <span class="st">"logCPM"</span>, <span class="st">"PValue"</span><span class="op">)</span><span class="op">)</span>
  <span class="fu">expect_equal</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">edger_res</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html">c</a></span><span class="op">(</span><span class="fl">15026</span>, <span class="fl">3</span><span class="op">)</span><span class="op">)</span>
<span class="op">}</span><span class="op">)</span></code></pre></div>
<p>Similar to the DESeq2 tests, we aren’t necessarily looking for a carbon copy of the results. While the process <em>should</em> be entirely deterministic, it is better to allow some wiggle room and merely look that the columns we’re interested in are present and that the data has the right shape. Note that EdgeR doesn’t perform a BH correction (the p-adjusted column is non-existant). It’s either an option that’s off by default or just not a feature! Not a problem, we have the p-values so we can do it ourselves (see the markdown).</p>
</div>
</div>
<div id="run_limma-and-run_voom" class="section level4 unnumbered">
<h4>4. <code>run_limma()</code> and <code>run_voom()</code><a class="anchor" aria-label="anchor" href="#run_limma-and-run_voom"><i class="fas fa-link"></i></a>
</h4>
<p>Our final in this set of three, we’re testing two related functions in the Limma package. While you can create an analysis that <em>only</em> uses Limma, in our case we can apply the <strong>voom</strong> workflow to refine our data a little more. More info on how these work in the cited documentation. A little technique that can be useful for optional workflows like this is including a boolean flag in the parameters of the function. If I am running a function <code>add_numbers(1, -4, abs=TRUE)</code>, the parameter <code>abs</code> might change the functions operation so that it takes the absolute value of numbers before adding them together. I could create this behavior like so:</p>
<div class="sourceCode" id="cb532"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">add_numbers</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span>, <span class="va">y</span>, <span class="va">abs</span><span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span>
  <span class="kw">if</span> <span class="op">(</span><span class="va">abs</span><span class="op">)</span> <span class="op">{</span> <span class="co"># code enters here if abs is TRUE</span>
    <span class="kw"><a href="https://rdrr.io/r/base/function.html">return</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html">abs</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html">abs</a></span><span class="op">(</span><span class="va">y</span><span class="op">)</span><span class="op">)</span> <span class="co"># return means function exits, doesn't go to "else {}"</span>
  <span class="op">}</span> <span class="kw">else</span> <span class="op">{</span> <span class="co"># code enters here if abs was false</span>
    <span class="kw"><a href="https://rdrr.io/r/base/function.html">return</a></span><span class="op">(</span><span class="va">x</span> <span class="op">+</span> <span class="va">y</span><span class="op">)</span>
  <span class="op">}</span>
<span class="op">}</span></code></pre></div>
<p>This is a very useful technique for re-using functions, meaning I don’t need to write two functions that do similar things when I can change their behavior with that boolean parameter.</p>
<div id="tests-7" class="section level5 unnumbered">
<h5>Tests<a class="anchor" aria-label="anchor" href="#tests-7"><i class="fas fa-link"></i></a>
</h5>
<div class="sourceCode" id="cb533"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">test_that</span><span class="op">(</span><span class="st">"test limma + voom work + work together"</span>, <span class="op">{</span>
  <span class="fu"><a href="https://rdrr.io/r/base/load.html">load</a></span><span class="op">(</span><span class="st">"data/mock_counts_df.RData"</span><span class="op">)</span>
  <span class="va">group</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span>, each<span class="op">=</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span>
  <span class="co"># design</span>
  <span class="va">design</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>day4<span class="op">=</span><span class="fl">1</span>, day4vsday7<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">1</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span>
  <span class="fu"><a href="https://rdrr.io/r/base/row.names.html">row.names</a></span><span class="op">(</span><span class="va">design</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html">c</a></span><span class="op">(</span><span class="st">"vP4_1"</span>, <span class="st">"vP4_2"</span>, <span class="st">"vP7_1"</span>, <span class="st">"vP7_2"</span><span class="op">)</span>
  <span class="co"># limma + voom</span>
  <span class="fu">expect_warning</span><span class="op">(</span><span class="va">voom_res</span> <span class="op">&lt;-</span> <span class="fu">run_limma</span><span class="op">(</span><span class="va">counts_df</span>, <span class="va">design</span>, <span class="va">group</span><span class="op">)</span><span class="op">)</span> <span class="co"># voom yelling, ignore</span>
  <span class="fu">expect_equal</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">voom_res</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html">c</a></span><span class="op">(</span><span class="fl">15026</span>, <span class="fl">6</span><span class="op">)</span><span class="op">)</span>
  <span class="fu">expect_equal</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">voom_res</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html">c</a></span><span class="op">(</span><span class="st">"logFC"</span>, <span class="st">"AveExpr"</span>, <span class="st">"t"</span>, <span class="st">"P.Value"</span>, 
                                  <span class="st">"adj.P.Val"</span>, <span class="st">"B"</span><span class="op">)</span><span class="op">)</span>
<span class="op">}</span><span class="op">)</span></code></pre></div>
<p>This ultimately functions similarly to the first two series of tests: load test data, apply the function to it, make sure the results are the right size, shape, and type. This includes the voom procedure, so using <em>only</em> limma results will yield failures.</p>
<div class="box note">
<p>Note that Limma uses EdgeR to input data, so your first couple steps will look pretty familiar.</p>
</div>
<p>We create a design object to use, defining the experimental conditions. These are again using the day 4 and day 7 trials from the heart data.</p>
</div>
</div>
<div id="combine_pval" class="section level4 unnumbered">
<h4>5. <code>combine_pval()</code><a class="anchor" aria-label="anchor" href="#combine_pval"><i class="fas fa-link"></i></a>
</h4>
<p>The final three functions concern modifying our differential expression results and creating plots of these combined results. While it would be trivial to make three plots using the same code or function, have R generate one consistent plot after changes in your data or method is extremely useful.</p>
<p>We use <code>combine_pval()</code> to coerce our data into long format, which is more applicable for facet wrapping. <code><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap()</a></code> is a <code>ggplot2</code> function that uses a property of the data to create separate plots in one image. In our case we will use the <code>package</code> property (DESeq2, edgeR, or Limma) to differentiate our data. There are a couple ways to reorganize data in this way, but I would suggest using <code><a href="https://tidyr.tidyverse.org/reference/gather.html">tidyr::gather()</a></code>. Our <code>key</code> will be the package and the <code>value</code> will be the p-value.</p>
<div id="tests-8" class="section level5 unnumbered">
<h5>Tests<a class="anchor" aria-label="anchor" href="#tests-8"><i class="fas fa-link"></i></a>
</h5>
<p>These tests are organized into one larger testing function because they use the same, randomly generated input data (a bunch of data frames with uniform distributions but with ranges that approximate actual data). If you would like to test only parts of this testing function, use an octothorpe (<code>#</code>) to comment out tests you haven’t reached yet.</p>
<div class="sourceCode" id="cb534"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb534-1"><a href="assignments.html#cb534-1" aria-hidden="true" tabindex="-1"></a><span class="fu">test_that</span>(<span class="st">"ggplot and data formatting works"</span>, {</span>
<span id="cb534-2"><a href="assignments.html#cb534-2" aria-hidden="true" tabindex="-1"></a>  deseq <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">pvalue =</span> <span class="fu">runif</span>(<span class="dv">1000</span>, <span class="fl">1e-100</span>, <span class="fl">1e-5</span>),</span>
<span id="cb534-3"><a href="assignments.html#cb534-3" aria-hidden="true" tabindex="-1"></a>                      <span class="at">log2FoldChange =</span> <span class="fu">runif</span>(<span class="dv">1000</span>, <span class="sc">-</span><span class="dv">9</span>, <span class="dv">9</span>),</span>
<span id="cb534-4"><a href="assignments.html#cb534-4" aria-hidden="true" tabindex="-1"></a>                      <span class="at">padj =</span> <span class="fu">runif</span>(<span class="dv">1000</span>, <span class="fl">1e-300</span>, <span class="fl">1e-15</span>))</span>
<span id="cb534-5"><a href="assignments.html#cb534-5" aria-hidden="true" tabindex="-1"></a>  edger <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">PValue =</span> <span class="fu">runif</span>(<span class="dv">1000</span>, <span class="fl">1e-100</span>, <span class="fl">1e-5</span>),</span>
<span id="cb534-6"><a href="assignments.html#cb534-6" aria-hidden="true" tabindex="-1"></a>                      <span class="at">logFC =</span> <span class="fu">runif</span>(<span class="dv">1000</span>, <span class="sc">-</span><span class="dv">9</span>, <span class="dv">9</span>),</span>
<span id="cb534-7"><a href="assignments.html#cb534-7" aria-hidden="true" tabindex="-1"></a>                      <span class="at">padj =</span> <span class="fu">runif</span>(<span class="dv">1000</span>, <span class="fl">1e-100</span>, <span class="fl">1e-14</span>))</span>
<span id="cb534-8"><a href="assignments.html#cb534-8" aria-hidden="true" tabindex="-1"></a>  limma <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">P.Value =</span> <span class="fu">runif</span>(<span class="dv">1000</span>, <span class="fl">1e-100</span>, <span class="fl">1e-5</span>),</span>
<span id="cb534-9"><a href="assignments.html#cb534-9" aria-hidden="true" tabindex="-1"></a>                      <span class="at">logFC =</span> <span class="fu">runif</span>(<span class="dv">1000</span>, <span class="sc">-</span><span class="dv">9</span>, <span class="dv">9</span>),</span>
<span id="cb534-10"><a href="assignments.html#cb534-10" aria-hidden="true" tabindex="-1"></a>                      <span class="at">adj.P.Val =</span> <span class="fu">runif</span>(<span class="dv">1000</span>, <span class="fl">1e-6</span>, <span class="fl">1e-4</span>))</span>
<span id="cb534-11"><a href="assignments.html#cb534-11" aria-hidden="true" tabindex="-1"></a>  <span class="co"># combine_pval</span></span>
<span id="cb534-12"><a href="assignments.html#cb534-12" aria-hidden="true" tabindex="-1"></a>  res <span class="ot">&lt;-</span> <span class="fu">combine_pval</span>(deseq, edger, limma)</span>
<span id="cb534-13"><a href="assignments.html#cb534-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">expect_equal</span>(<span class="fu">dim</span>(res), <span class="fu">c</span>(<span class="dv">3000</span>, <span class="dv">2</span>))</span>
<span id="cb534-14"><a href="assignments.html#cb534-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">expect_true</span>(<span class="fu">mean</span>(res<span class="sc">$</span>pval) <span class="sc">&lt;</span> <span class="fl">5e-5</span> <span class="sc">&amp;&amp;</span> <span class="fu">mean</span>(res<span class="sc">$</span>pval) <span class="sc">&gt;</span> <span class="fl">5e-7</span>)</span>
<span id="cb534-15"><a href="assignments.html#cb534-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">expect_equal</span>(<span class="fu">table</span>(res<span class="sc">$</span>package)[[<span class="dv">1</span>]], <span class="dv">1000</span>)</span>
<span id="cb534-16"><a href="assignments.html#cb534-16" aria-hidden="true" tabindex="-1"></a>  ...</span></code></pre></div>
<p>The first nine lines establish the test data frames. While random, they will let us test if functions are working correctly.</p>
<div class="box important">
<p>The test data frames do not mimic the normal results exactly. The rows we are interested in have the same name, but have different numbers of columns. For that reason, I suggest your function uses <code>$</code>-notation when referencing columns, instead of position. <code>deseq$pvalue</code> will work more better than <code>deseq[,1]</code> for this test data.</p>
</div>
<p>We test that we only have 2 columns and 3,000 rows, and that the fake p-values are within the expected range. Finally, we use the <code><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html">table()</a></code> function to test that there is only 1,000 of any one package in this data frame.</p>
</div>
</div>
<div id="create_facets" class="section level4 unnumbered">
<h4>6. <code>create_facets()</code><a class="anchor" aria-label="anchor" href="#create_facets"><i class="fas fa-link"></i></a>
</h4>
<p>Now that we have examined the p-values, we can look at the adjusted p-values. We’ll be constructing volcano plots, which compare the log<sub>2</sub> fold-change and the adjusted p-value, so the <code>create_facets()</code> function will be somewhat more complicated. Rather than use <code><a href="https://tidyr.tidyverse.org/reference/gather.html">gather()</a></code> again, it may be easier to create three tibbles or data frames of the variables of interest and combine them with <code><a href="https://rdrr.io/pkg/BiocGenerics/man/cbind.html">rbind()</a></code>.</p>
<div id="tests-9" class="section level5 unnumbered">
<h5>Tests<a class="anchor" aria-label="anchor" href="#tests-9"><i class="fas fa-link"></i></a>
</h5>
<div class="sourceCode" id="cb535"><pre class="downlit sourceCode r">
<code class="sourceCode R">  <span class="va">...</span>
  <span class="co"># create_facets</span>
  <span class="va">facets</span> <span class="op">&lt;-</span> <span class="fu">create_facets</span><span class="op">(</span><span class="va">deseq</span>, <span class="va">edger</span>, <span class="va">limma</span><span class="op">)</span>
  <span class="fu">expect_equal</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">facets</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html">c</a></span><span class="op">(</span><span class="fl">3000</span>, <span class="fl">3</span><span class="op">)</span><span class="op">)</span>
  <span class="fu">expect_equal</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html">table</a></span><span class="op">(</span><span class="va">facets</span><span class="op">$</span><span class="va">package</span><span class="op">)</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span>, <span class="fl">1000</span><span class="op">)</span>
  <span class="va">...</span></code></pre></div>
<p>Another simple set of tests, this time we want to ensure there are 3,000 rows and <em>3</em> columns. We also want to ensure there aren’t more than 1,000 rows of any package type.</p>
</div>
</div>
<div id="theme_plot" class="section level4 unnumbered">
<h4>7. <code>theme_plot()</code><a class="anchor" aria-label="anchor" href="#theme_plot"><i class="fas fa-link"></i></a>
</h4>
<p>Finally, we have all of our data in the right shape and we can do some plotting. There is a lot to learn about ggplot, and I mention a few resources in the function description, but don’t be afraid to experiment! Try different colors, themes, and styles. The easiest part is getting your data on the plot, the hard part is making it look attractive.</p>
<div id="tests-10" class="section level5 unnumbered">
<h5>Tests<a class="anchor" aria-label="anchor" href="#tests-10"><i class="fas fa-link"></i></a>
</h5>
<div class="sourceCode" id="cb536"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb536-1"><a href="assignments.html#cb536-1" aria-hidden="true" tabindex="-1"></a>  ...</span>
<span id="cb536-2"><a href="assignments.html#cb536-2" aria-hidden="true" tabindex="-1"></a>  <span class="co"># theme_plot</span></span>
<span id="cb536-3"><a href="assignments.html#cb536-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tibble</span>(<span class="at">logFC =</span> <span class="fu">runif</span>(<span class="dv">3000</span>, <span class="sc">-</span><span class="dv">9</span>, <span class="dv">9</span>),</span>
<span id="cb536-4"><a href="assignments.html#cb536-4" aria-hidden="true" tabindex="-1"></a>         <span class="at">padj =</span> <span class="fu">runif</span>(<span class="dv">3000</span>, <span class="fl">1e-300</span>, <span class="fl">1e-5</span>),</span>
<span id="cb536-5"><a href="assignments.html#cb536-5" aria-hidden="true" tabindex="-1"></a>         <span class="at">package =</span> <span class="fu">rep</span>(<span class="fu">c</span>(<span class="st">"DESeq2"</span>, <span class="st">"edgeR"</span>, <span class="st">"Limma"</span>), <span class="at">each=</span><span class="dv">1000</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb536-6"><a href="assignments.html#cb536-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_plot</span>() <span class="ot">-&gt;</span> volc_plot</span>
<span id="cb536-7"><a href="assignments.html#cb536-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">expect_equal</span>(<span class="fu">dim</span>(volc_plot<span class="sc">$</span>data), <span class="fu">c</span>(<span class="dv">3000</span>, <span class="dv">3</span>))</span>
<span id="cb536-8"><a href="assignments.html#cb536-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">expect_equal</span>(<span class="fu">class</span>(volc_plot<span class="sc">$</span>layers[[<span class="dv">1</span>]]<span class="sc">$</span>geom)[<span class="dv">1</span>], <span class="st">"GeomPoint"</span>)</span>
<span id="cb536-9"><a href="assignments.html#cb536-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">expect_equal</span>(<span class="fu">class</span>(volc_plot<span class="sc">$</span>facet<span class="sc">$</span>params), <span class="st">"list"</span>)</span>
<span id="cb536-10"><a href="assignments.html#cb536-10" aria-hidden="true" tabindex="-1"></a><span class="er">})</span></span></code></pre></div>
<p>We create a little more fake data, and then pipe it into our <code>theme_plot()</code> function, and save that ggplot object. Our tests simply look inside the ggplot object to confirm the data is the right shape, that you’re plotting using <code><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point()</a></code>, and that facets are present.</p>
<p>Once your tests are passing, take a look at the report and see how everything can work together!</p>

</div>
</div>
</div>
</div>
<div id="assignment-7" class="section level2 unnumbered">
<h2>Assignment 7<a class="anchor" aria-label="anchor" href="#assignment-7"><i class="fas fa-link"></i></a>
</h2>
<blockquote>
<p>Now working on my companion package: R Dully</p>
</blockquote>
<div id="problem-statement-6" class="section level3 unnumbered">
<h3>Problem Statement<a class="anchor" aria-label="anchor" href="#problem-statement-6"><i class="fas fa-link"></i></a>
</h3>
<p>Communicating data and results is one of the most nefariously troubling aspects of data science. In bioinformatics we often have colleagues who are less experienced when it comes to viewing and manipulating data, such as clinicians or biologists who think computers are living objects of evil (they are). We must utilize better ways to communicate our work to them.</p>
</div>
<div id="learning-objectives-5" class="section level3 unnumbered">
<h3>Learning Objectives<a class="anchor" aria-label="anchor" href="#learning-objectives-5"><i class="fas fa-link"></i></a>
</h3>
<ol style="list-style-type: decimal">
<li>Develop the two sides of an R Shiny application: server and UI.</li>
<li>Understand reactivity when it comes to an application.</li>
<li>Take in user input and use it to change the state of a web application.</li>
</ol>
</div>
<div id="skill-list-5" class="section level3 unnumbered">
<h3>Skill List<a class="anchor" aria-label="anchor" href="#skill-list-5"><i class="fas fa-link"></i></a>
</h3>
<ul>
<li>R Shiny</li>
<li>Troubleshooting R Shiny</li>
<li>Testing R Shiny? Like, <em>nobody</em> does this so it could be a pretty good edge for you.</li>
</ul>
</div>
<div id="instructions-4" class="section level3 unnumbered">
<h3>Instructions<a class="anchor" aria-label="anchor" href="#instructions-4"><i class="fas fa-link"></i></a>
</h3>
<p>Complete <code>app.R</code> (not <code>main.R</code>), including the <code>server</code> and <code>ui</code> functions. We are not looking for exact parity when it comes to the <em>style</em> of the application, just that you replicate the user experience and inputs and create the two methods of output: a table and a volcano plot.</p>
<p>The sample application is hosted here: <a href="https://taylorfalk.shinyapps.io/bf591-assignment-7/">https://taylorfalk.shinyapps.io/bf591-assignment-7/</a></p>
<p>View this document along with the sample application and try to determine which parts of the UI and output mentioned in this document are associated with the functioning application.</p>
<p>Test are available for this assignment, because of course they are. You can run them using <code>testthat::test_file("test_app.R")</code>. You can also use <code>source("test_app.R")</code> (but it’s less informative). Shiny apps add another dimensionality of complexity, so we are relatively limited in what we can test for. As always, if you feel a test isn’t working correctly please let a TA know.</p>
<p><span style="color:red">Note:</span> By design, these instructions are rather sparse. This is likely entirely new territory for you, but you have gotten this far and there is a lot of information available for R Shiny. Please read documentation or look for help online to help yourself understand how this application works.</p>
</div>
<div id="function-details-3" class="section level3 unnumbered">
<h3>“Function” Details<a class="anchor" aria-label="anchor" href="#function-details-3"><i class="fas fa-link"></i></a>
</h3>
<p>Without writing an entire treatise on R Shiny (see the book), a Shiny application is broken into two components: the UI function and the Server function. In web development parlance, these represent the “front end” and the “back end,” respectively. The front end is all the code that controls what a user sees and does (the goop that is in the web browser), and the back end is what connects that front end to the server where most of the code, processing power, and data will live.</p>
<p>You can use two files for this, but we have set it up for <code>app.R</code> to contain the entire Shiny application. It’s general format is like so:</p>
<div class="sourceCode" id="cb537"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://shiny.rstudio.com/">shiny</a></span><span class="op">)</span>
<span class="va">ui</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/shiny/man/fluidPage.html">fluidPage</a></span><span class="op">(</span>
  <span class="co"># front end</span>
<span class="op">)</span> 
<span class="va">server</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">input</span>, <span class="va">output</span>, <span class="va">session</span><span class="op">)</span> <span class="op">{</span>
   <span class="co"># back end</span>
<span class="op">}</span>
<span class="fu"><a href="https://rdrr.io/pkg/shiny/man/shinyApp.html">shinyApp</a></span><span class="op">(</span>ui <span class="op">=</span> <span class="va">ui</span>, server <span class="op">=</span> <span class="va">server</span><span class="op">)</span> <span class="co"># run app</span></code></pre></div>
<p>After loading the Shiny library, we create the front end parameters with the <code><a href="https://rdrr.io/pkg/shiny/man/fluidPage.html">fluidPage()</a></code> function. Fluid page is a simple wrapper for designing a Shiny page, and inside we can use functions like <code><a href="https://rdrr.io/pkg/shiny/man/titlePanel.html">titlePanel()</a></code>, <code><a href="https://rdrr.io/pkg/shiny/man/sidebarLayout.html">sidebarLayout()</a></code>, and <code><a href="https://rdrr.io/pkg/shiny/man/tabPanel.html">tabPanel()</a></code> to organize what is displayed.</p>
<p>Next, we create the <code>server</code> function which is much more similar to writing base R. <code>input</code>, <code>output</code>, and <code>server</code> are automatically handled by Shiny, and we use these objects to pass data between front end (UI) and back end (server). A user can select a color for their graph, we can store that color in <code>input$color1</code>, read that <code>input</code> inside server, and pass the resulting graph to <code>output</code> to place onto the page.</p>
<p>Finally, the <code><a href="https://rdrr.io/pkg/shiny/man/shinyApp.html">shinyApp()</a></code> is called to run everything. It does most of the heavy lifting so we can worry more about themes and pretty colors for our app (<code><a href="https://rstudio.github.io/bslib/">library(bslib)</a></code>, themeing will not be graded). It goes without saying, much like working on the fonts and headers in an essay, you should wait until your app is mostly complete before you start changing the aesthetics.</p>
<div id="ui---fluidpage..." class="section level4 unnumbered">
<h4>1. <code>ui &lt;- fluidPage(...)</code><a class="anchor" aria-label="anchor" href="#ui---fluidpage..."><i class="fas fa-link"></i></a>
</h4>
<p>R Shiny is a <em>highly</em> customization package, so we will describe what inputs and outputs we’re expecting, but leave the theme, placement, arrangement, and design of these elements up to you. R Shiny’s default layout is the very usable <code>sidebayLayout()</code>, but if you want more control feel free to use <code><a href="https://rdrr.io/pkg/shiny/man/fluidPage.html">fluidRow()</a></code>.</p>
<p>In our sidebar, we have six elements for user input. There is the <code><a href="https://rdrr.io/pkg/shiny/man/fileInput.html">fileInput()</a></code>, followed by two <code><a href="https://rdrr.io/pkg/shiny/man/radioButtons.html">radioButtons()</a></code>, followed by two <code>colourInputs()</code>, and finally a <code><a href="https://rdrr.io/pkg/shiny/man/sliderInput.html">sliderInput()</a></code>. The most important part for these inputs is the <code>inputId</code> argument, which later is used inside the server function’s <code>input</code>. If I make my <code>inputId=fileupload</code>, then I can access that upload using <code>input$fileupload</code> later on.</p>
<div id="file-input" class="section level5 unnumbered">
<h5>File Input<a class="anchor" aria-label="anchor" href="#file-input"><i class="fas fa-link"></i></a>
</h5>
<p>The file input function will take in a CSV file, helpfully included in this repository in the <code>data/</code> folder. Not vital, but a nice method we can do is <em>only</em> allow CSV file types using <code>accept=</code>. This is useful if your users are easily confused and try to enter other data.</p>
</div>
<div id="radio-buttons" class="section level5 unnumbered">
<h5>Radio Buttons<a class="anchor" aria-label="anchor" href="#radio-buttons"><i class="fas fa-link"></i></a>
</h5>
<p>Since we are ultimately plotting a graph based on the data uploaded, we want to select what variables are being plotted. We set up radio buttons using the names from the column header, and can match those later on in ggplot when selecting data to plot. Make sure your two radio button functions have different <code>inputIds</code>.</p>
</div>
<div id="colour-inputs" class="section level5 unnumbered">
<h5>Colour Inputs<a class="anchor" aria-label="anchor" href="#colour-inputs"><i class="fas fa-link"></i></a>
</h5>
<p>This is a bonus library that is very useful when working with color. Installing <code>colourpicker</code> and using <code>colourInput()</code> inside your side panel layout lets us choose two colors to place on our graph. Feel free to choose your own default colors, and again make sure these have separate input IDs.</p>
</div>
<div id="slider-input" class="section level5 unnumbered">
<h5>Slider Input<a class="anchor" aria-label="anchor" href="#slider-input"><i class="fas fa-link"></i></a>
</h5>
<p>Out last piece of input will be the slider input. This allows the user to select the magnitude of adjusted p-value they would like to color on their graph. This is the <span class="math inline">\(x\)</span> in <span class="math inline">\(1 \times 10^{x}\)</span>.</p>
</div>
<div id="the-button" class="section level5 unnumbered">
<h5>The Button<a class="anchor" aria-label="anchor" href="#the-button"><i class="fas fa-link"></i></a>
</h5>
<p>Finally, there is a button to hit when everything is ready. You can use a <code><a href="https://rdrr.io/pkg/shiny/man/submitButton.html">submitButton()</a></code> or an <code><a href="https://rdrr.io/pkg/shiny/man/actionButton.html">actionButton()</a></code>. Both work slightly differently but having a button just lets us control when a plot is created. You can use <code>icon = icon()</code> to choose a different icon for the button.</p>
</div>
</div>
<div id="tabs-and-outputs" class="section level4" number="12.1.0.1">
<h4>
<span class="header-section-number">12.1.0.1</span> 2. Tabs and Outputs<a class="anchor" aria-label="anchor" href="#tabs-and-outputs"><i class="fas fa-link"></i></a>
</h4>
<p>Once the inputs are established on the page, we want somewhere to put everything once it’s done being run! This is accomplished by the <code>*Output()</code> functions. We have two outputs to consider, a plot and a table, so we can use <code><a href="https://rdrr.io/pkg/shiny/man/plotOutput.html">plotOutput()</a></code> and <code><a href="https://rdrr.io/pkg/shiny/man/renderTable.html">tableOutput()</a></code> respectively, using the names inside <code>output</code> from the server function. These can be placed simply on the <code><a href="https://rdrr.io/pkg/shiny/man/sidebarLayout.html">mainPanel()</a></code>.</p>
<p>There is one additional step if you would like to use tabs instead, which lets us create more space for our outputs. Instead of using the output functions directly, we can first use <code><a href="https://rdrr.io/pkg/shiny/man/tabsetPanel.html">tabsetPanel()</a></code> and place two <code><a href="https://rdrr.io/pkg/shiny/man/tabPanel.html">tabPanel()</a></code> functions inside. We then finally place our outputs. A tab setup will look something like this:</p>
<div class="sourceCode" id="cb538"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb538-1"><a href="assignments.html#cb538-1" aria-hidden="true" tabindex="-1"></a>ui <span class="ot">&lt;-</span> <span class="fu">fluidPage</span>(</span>
<span id="cb538-2"><a href="assignments.html#cb538-2" aria-hidden="true" tabindex="-1"></a>  ...</span>
<span id="cb538-3"><a href="assignments.html#cb538-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mainPanel</span>(</span>
<span id="cb538-4"><a href="assignments.html#cb538-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">tabsetPanel</span>(</span>
<span id="cb538-5"><a href="assignments.html#cb538-5" aria-hidden="true" tabindex="-1"></a>      <span class="fu">tabPanel</span>(<span class="st">"Tab1 Name"</span></span>
<span id="cb538-6"><a href="assignments.html#cb538-6" aria-hidden="true" tabindex="-1"></a>        <span class="fu">tableOutput</span>(...)</span>
<span id="cb538-7"><a href="assignments.html#cb538-7" aria-hidden="true" tabindex="-1"></a>      ),</span>
<span id="cb538-8"><a href="assignments.html#cb538-8" aria-hidden="true" tabindex="-1"></a>      <span class="fu">tabPanel</span>(<span class="st">"Tab2 Name"</span></span>
<span id="cb538-9"><a href="assignments.html#cb538-9" aria-hidden="true" tabindex="-1"></a>        <span class="fu">plotOutput</span>(...)</span>
<span id="cb538-10"><a href="assignments.html#cb538-10" aria-hidden="true" tabindex="-1"></a>      )</span>
<span id="cb538-11"><a href="assignments.html#cb538-11" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb538-12"><a href="assignments.html#cb538-12" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb538-13"><a href="assignments.html#cb538-13" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div>
<p>It might look a little messy but if you can keep an eye on hierarchy it will start to make more sense.</p>
</div>
<div id="server---function..." class="section level4 unnumbered">
<h4>2. <code>server &lt;- function(...</code><a class="anchor" aria-label="anchor" href="#server---function..."><i class="fas fa-link"></i></a>
</h4>
<p>Our second main component, and where we will be doing a lot of the “work,” is the server function. This is written like a normal function in R, with three specific parameters or arguments. These arguments can be used <em>inside</em> server to communicate with the elements we wrote above. The <code>input</code> object can be used to take in user selections, such as what color was chosen. The <code>output</code> object is then use to assign results to the output functions we included above.</p>
<div id="load_data" class="section level5 unnumbered">
<h5>
<code>load_data()</code><a class="anchor" aria-label="anchor" href="#load_data"><i class="fas fa-link"></i></a>
</h5>
<p>Much like a normal R script, we can save ourselves some trouble by splitting our code blocks into functions. In this case, we can load the user select data into different parts of our app, like displaying a plot and then a table.</p>
<p>This <code>load_data()</code> function should be written as a reactive expression, covered in more detail in the book chapter. Because the user is being asked to upload data we do not want to process that data twice, which would slow down our application.</p>
<p>This reaction function will update when the inputs inside it change, in our case the file name from our file uploader. Ultimately, we just want this function to return the data frame from loading whatever is in the upload box.</p>
<p><span style="color:red">Note</span> You will need to specify the <strong>datapath</strong> from the <code><a href="https://rdrr.io/pkg/shiny/man/fileInput.html">fileInput()</a></code> function’s input ID. If my input ID was “taylorsfile” then my actual input will be <code>input$taylorsfile$datapath</code>.</p>
</div>
<div id="volcano_plot" class="section level5 unnumbered">
<h5>
<code>volcano_plot()</code><a class="anchor" aria-label="anchor" href="#volcano_plot"><i class="fas fa-link"></i></a>
</h5>
<p>This should feel like meeting an old friend: make a volcano plot. The function takes in a data frame, column names to choose as axes, a negative integer to use as a cutoff, and two colors to make points above and below that cutoff. Simple and exciting. Please do all the normal ggplot things like labeling axes and changing the theme.</p>
<div class="box hint">
<p>A little hint, sometimes tidy elements don’t play nice with string names (like this function asks for). You can use <code>!!sym(x_name)</code> to access those columns. There are also tidy friendly ways to do this, but uhh I like this way!</p>
</div>
</div>
<div id="draw_table" class="section level5 unnumbered">
<h5>
<code>draw_table()</code><a class="anchor" aria-label="anchor" href="#draw_table"><i class="fas fa-link"></i></a>
</h5>
<p>Same as above, this function is <em>even</em> simpler. Take in a data frame and the slider magnitude, and cut off any rows that aren’t lower than the slider threshold. See the second tab on the example app for how the slider value interacts with what populates the table. We also want to format some of our p-value columns to display more digits, as otherwise they will show up as “0” in the table (not informative). You can use tidyverse functions or <code><a href="https://rdrr.io/r/base/formatc.html">formatC()</a></code> to accomplish this.</p>
</div>
<div id="output..." class="section level5 unnumbered">
<h5>
<code>output$...</code><a class="anchor" aria-label="anchor" href="#output..."><i class="fas fa-link"></i></a>
</h5>
<p>Finally we will cram our results onto the web page. Use the functions <code><a href="https://rdrr.io/pkg/shiny/man/renderPlot.html">renderPlot()</a></code> and <code><a href="https://rdrr.io/pkg/shiny/man/renderTable.html">renderTable()</a></code> to <strong>return</strong> your volcano plot and filtered data table using the above two functions. These outputs <em>are</em> checked by the testing script, so there is a bare minimum as to what we are looking for. Please reference the test file <code>test_app.R</code> to see what is being tested. The table’s p-values should have a magnitude, not just <code>0</code>.</p>
<p>If the outputs show up correctly when you load the CSV file, congrats! You’ve built your app correctly. If it didn’t work, I am so sorry. R Shiny can be difficult to troubleshoot so do not be afraid to look up the strange errors you see or ask for help.</p>

</div>
</div>
</div>
</div>
</div>



<div class="chapter-nav">
<div class="prev"><a href="contribution-guide.html"><span class="header-section-number">12</span> Contribution Guide</a></div>
<div class="next"><a href="class-outlines.html"><span class="header-section-number">A</span> Class Outlines</a></div>
</div></main><div class="col-md-3 col-lg-2 d-none d-md-block sidebar sidebar-chapter">
    <nav id="toc" data-toggle="toc" aria-label="On this page"><h2>On this page</h2>
      <ul class="nav navbar-nav">
<li><a class="nav-link" href="#assignments">Assignments</a></li>
<li><a class="nav-link" href="#assignment-format">Assignment Format</a></li>
<li>
<a class="nav-link" href="#starting-an-assignment">Starting an Assignment</a><ul class="nav navbar-nav">
<li><a class="nav-link" href="#starting-your-own-git-repository-and-cloning-it">Starting your own git repository and cloning it</a></li>
<li><a class="nav-link" href="#contents">Contents</a></li>
<li><a class="nav-link" href="#github-tutorial">GitHub tutorial</a></li>
<li><a class="nav-link" href="#committing-and-pushing-with-r-and-without">Committing and Pushing with R (and without)</a></li>
<li><a class="nav-link" href="#using-rstudio">Using RStudio</a></li>
<li><a class="nav-link" href="#using-the-command-line">Using the command line</a></li>
</ul>
</li>
<li>
<a class="nav-link" href="#assignment-1">Assignment 1</a><ul class="nav navbar-nav">
<li><a class="nav-link" href="#problem-statement">Problem Statement</a></li>
<li><a class="nav-link" href="#learning-objectives-and-skill-list">Learning Objectives and Skill List</a></li>
<li><a class="nav-link" href="#instructions">Instructions</a></li>
<li><a class="nav-link" href="#hints">Hints</a></li>
</ul>
</li>
<li>
<a class="nav-link" href="#assignment-2">Assignment 2</a><ul class="nav navbar-nav">
<li><a class="nav-link" href="#problem-statement-1">Problem Statement</a></li>
<li><a class="nav-link" href="#learning-objectives">Learning Objectives</a></li>
<li><a class="nav-link" href="#skill-list">Skill List</a></li>
<li><a class="nav-link" href="#instructions-1">Instructions</a></li>
<li><a class="nav-link" href="#function-details">Function Details</a></li>
</ul>
</li>
<li>
<a class="nav-link" href="#assignment-3">Assignment 3</a><ul class="nav navbar-nav">
<li><a class="nav-link" href="#problem-statement-2">Problem Statement</a></li>
<li><a class="nav-link" href="#learning-objectives-1">Learning Objectives</a></li>
<li><a class="nav-link" href="#skill-list-1">Skill List</a></li>
<li><a class="nav-link" href="#background-on-microarrays">Background on Microarrays</a></li>
<li><a class="nav-link" href="#background-on-principal-component-analysis">Background on Principal Component Analysis</a></li>
<li><a class="nav-link" href="#marisa-et-al.-gene-expression-classification-of-colon-cancer-into-molecular-subtypes-characterization-validation-and-prognostic-value.-plos-medicine-may-2013.-pmid-23700391">Marisa et al. Gene Expression Classification of Colon Cancer into Molecular Subtypes: Characterization, Validation, and Prognostic Value. PLoS Medicine, May 2013. PMID: 23700391</a></li>
<li><a class="nav-link" href="#scaling-data-using-r-scale">Scaling data using R scale()</a></li>
<li><a class="nav-link" href="#proportion-of-variance-explained">Proportion of variance explained</a></li>
<li><a class="nav-link" href="#plotting-and-visualization-of-pca">Plotting and visualization of PCA</a></li>
<li><a class="nav-link" href="#hierarchical-clustering-and-heatmaps">Hierarchical Clustering and Heatmaps</a></li>
<li><a class="nav-link" href="#references">References</a></li>
</ul>
</li>
<li>
<a class="nav-link" href="#assignment-4">Assignment 4</a><ul class="nav navbar-nav">
<li><a class="nav-link" href="#problem-statement-3">Problem Statement</a></li>
<li><a class="nav-link" href="#learning-objectives-2">Learning Objectives</a></li>
<li><a class="nav-link" href="#skill-list-2">Skill List</a></li>
<li><a class="nav-link" href="#instructions-2">Instructions</a></li>
<li><a class="nav-link" href="#deliverables">Deliverables</a></li>
<li><a class="nav-link" href="#function-details-1">Function Details</a></li>
<li><a class="nav-link" href="#hints-1">Hints</a></li>
</ul>
</li>
<li>
<a class="nav-link" href="#assignment-5">Assignment 5</a><ul class="nav navbar-nav">
<li><a class="nav-link" href="#problem-statement-4">Problem Statement</a></li>
<li><a class="nav-link" href="#learning-objectives-3">Learning Objectives</a></li>
<li><a class="nav-link" href="#skill-list-3">Skill List</a></li>
<li><a class="nav-link" href="#deseq2-background">DESeq2 Background</a></li>
<li><a class="nav-link" href="#generating-a-counts-matrix">Generating a counts matrix</a></li>
<li><a class="nav-link" href="#prefiltering-counts-matrix">Prefiltering Counts matrix</a></li>
<li><a class="nav-link" href="#median-of-ratios-normalization">Median-of-ratios normalization</a></li>
<li><a class="nav-link" href="#deseq2-preparation">DESeq2 preparation</a></li>
<li><a class="nav-link" href="#omeara-et-al.-transcriptional-reversion-of-cardiac-myocyte-fate-during-mammalian-cardiac-regeneration.-circ-res.-feb-2015.-pmid-25477501l">O’Meara et al. Transcriptional Reversion of Cardiac Myocyte Fate During Mammalian Cardiac Regeneration. Circ Res. Feb 2015. PMID: 25477501l</a></li>
<li><a class="nav-link" href="#reading-and-subsetting-the-data-from-verse_counts.tsv-and-sample_metadata.csv">1. Reading and subsetting the data from verse_counts.tsv and sample_metadata.csv</a></li>
<li><a class="nav-link" href="#running-deseq2">2. Running DESeq2</a></li>
<li><a class="nav-link" href="#annotating-results-to-construct-a-labeled-volcano-plot">3. Annotating results to construct a labeled volcano plot</a></li>
<li><a class="nav-link" href="#diagnostic-plot-of-the-raw-p-values-for-all-genes">4. Diagnostic plot of the raw p-values for all genes</a></li>
<li><a class="nav-link" href="#plotting-the-logfoldchanges-for-differentially-expressed-genes">5. Plotting the LogFoldChanges for differentially expressed genes</a></li>
<li><a class="nav-link" href="#the-choice-of-fdr-cutoff-depends-on-cost">The choice of FDR cutoff depends on cost</a></li>
<li><a class="nav-link" href="#plotting-the-normalized-counts-of-differentially-expressed-genes">6. Plotting the normalized counts of differentially expressed genes</a></li>
<li><a class="nav-link" href="#volcano-plot-to-visualize-differential-expression-results">7. Volcano Plot to visualize differential expression results</a></li>
<li><a class="nav-link" href="#running-fgsea-vignette">8. Running fgsea vignette</a></li>
<li><a class="nav-link" href="#plotting-the-top-ten-positive-nes-and-top-ten-negative-nes-pathways">9. Plotting the top ten positive NES and top ten negative NES pathways</a></li>
<li><a class="nav-link" href="#references-1">References</a></li>
</ul>
</li>
<li>
<a class="nav-link" href="#assignment-6">Assignment 6</a><ul class="nav navbar-nav">
<li><a class="nav-link" href="#problem-statement-5">Problem Statement</a></li>
<li><a class="nav-link" href="#learning-objectives-4">Learning Objectives</a></li>
<li><a class="nav-link" href="#skill-list-4">Skill List</a></li>
<li><a class="nav-link" href="#instructions-3">Instructions</a></li>
<li><a class="nav-link" href="#function-details-2">Function Details</a></li>
</ul>
</li>
<li>
<a class="nav-link" href="#assignment-7">Assignment 7</a><ul class="nav navbar-nav">
<li><a class="nav-link" href="#problem-statement-6">Problem Statement</a></li>
<li><a class="nav-link" href="#learning-objectives-5">Learning Objectives</a></li>
<li><a class="nav-link" href="#skill-list-5">Skill List</a></li>
<li><a class="nav-link" href="#instructions-4">Instructions</a></li>
<li><a class="nav-link" href="#function-details-3">“Function” Details</a></li>
</ul>
</li>
</ul>

      <div class="book-extra">
        <ul class="list-unstyled">
          
        </ul>
</div>
    </nav>
</div>
</div>
  

  

</div>
 <!-- .container -->

<footer class="bg-primary text-light mt-5"><div class="container"><div class="row">

  <div class="col-12 col-md-6 mt-3">
    <p>"<strong>BF591 - R for Biological Sciences</strong>" was written by . </p>
  </div>

  <div class="col-12 col-md-6 mt-3">
    <p>This book was built by the <a class="text-light" href="https://bookdown.org">bookdown</a> R package.</p>
  </div>

</div></div>
</footer><!-- dynamically load mathjax for compatibility with self-contained --><script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script><script type="text/x-mathjax-config">const popovers = document.querySelectorAll('a.footnote-ref[data-toggle="popover"]');
for (let popover of popovers) {
  const div = document.createElement('div');
  div.setAttribute('style', 'position: absolute; top: 0, left:0; width:0, height:0, overflow: hidden; visibility: hidden;');
  div.innerHTML = popover.getAttribute('data-content');

  var has_math = div.querySelector("span.math");
  if (has_math) {
    document.body.appendChild(div);
    MathJax.Hub.Queue(["Typeset", MathJax.Hub, div]);
    MathJax.Hub.Queue(function() {
      popover.setAttribute('data-content', div.innerHTML);
      document.body.removeChild(div);
    })
  }
}
</script>
</body>
</html>
