<!DOCTYPE html>
<html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Assignment 4 | BF591 - R for Biological Sciences</title>
<meta name="generator" content="bookdown 0.32 with bs4_book()">
<meta property="og:title" content="Assignment 4 | BF591 - R for Biological Sciences">
<meta property="og:type" content="book">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="Assignment 4 | BF591 - R for Biological Sciences">
<!-- JS --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://kit.fontawesome.com/6ecbd6c532.js" crossorigin="anonymous"></script><script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="libs/bootstrap-4.6.0/bootstrap.min.css" rel="stylesheet">
<script src="libs/bootstrap-4.6.0/bootstrap.bundle.min.js"></script><script src="libs/bs3compat-0.4.2/transition.js"></script><script src="libs/bs3compat-0.4.2/tabs.js"></script><script src="libs/bs3compat-0.4.2/bs3compat.js"></script><link href="libs/bs4_book-1.0.0/bs4_book.css" rel="stylesheet">
<script src="libs/bs4_book-1.0.0/bs4_book.js"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- CSS --><style type="text/css">
    
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  </style>
<style type="text/css">
    /* Used with Pandoc 2.11+ new --citeproc when CSL is used */
    div.csl-bib-body { }
    div.csl-entry {
      clear: both;
        }
    .hanging div.csl-entry {
      margin-left:2em;
      text-indent:-2em;
    }
    div.csl-left-margin {
      min-width:2em;
      float:left;
    }
    div.csl-right-inline {
      margin-left:2em;
      padding-left:1em;
    }
    div.csl-indent {
      margin-left: 2em;
    }
  </style>
<link rel="stylesheet" href="style.css">
<meta name="description" content="Problem Statement When dealing with gene counts in an mRNA-seq dataset, it is important to normalize the data before performing any analyses so you can make accurate comparisons of gene expression...">
<meta property="og:description" content="Problem Statement When dealing with gene counts in an mRNA-seq dataset, it is important to normalize the data before performing any analyses so you can make accurate comparisons of gene expression...">
<meta name="twitter:description" content="Problem Statement When dealing with gene counts in an mRNA-seq dataset, it is important to normalize the data before performing any analyses so you can make accurate comparisons of gene expression...">
</head>
<body data-spy="scroll" data-target="#toc">

<div class="container-fluid">
<div class="row">
  <header class="col-sm-12 col-lg-3 sidebar sidebar-book"><a class="sr-only sr-only-focusable" href="#content">Skip to main content</a>

    <div class="d-flex align-items-start justify-content-between">
      <h1>
        <a href="index.html" title="">BF591 - R for Biological Sciences</a>
      </h1>
      <button class="btn btn-outline-primary d-lg-none ml-2 mt-1" type="button" data-toggle="collapse" data-target="#main-nav" aria-expanded="true" aria-controls="main-nav"><i class="fas fa-bars"></i><span class="sr-only">Show table of contents</span></button>
    </div>

    <div id="main-nav" class="collapse-lg">
      <form role="search">
        <input id="search" class="form-control" type="search" placeholder="Search" aria-label="Search">
</form>

      <nav aria-label="Table of contents"><h2>Table of contents</h2>
        <ul class="book-toc list-unstyled">
<li><a class="" href="index.html">R for Biological Sciences</a></li>
<li><a class="" href="introduction.html"><span class="header-section-number">1</span> Introduction</a></li>
<li><a class="" href="data-bio.html"><span class="header-section-number">2</span> Data in Biology</a></li>
<li><a class="" href="preliminaries.html"><span class="header-section-number">3</span> Preliminaries</a></li>
<li><a class="" href="prog-basics.html"><span class="header-section-number">4</span> R Programming</a></li>
<li><a class="" href="data-wrangling.html"><span class="header-section-number">5</span> Data Wrangling</a></li>
<li><a class="" href="data-science.html"><span class="header-section-number">6</span> Data Science</a></li>
<li><a class="" href="data-visualization.html"><span class="header-section-number">7</span> Data Visualization</a></li>
<li><a class="" href="biology-bioinformatics.html"><span class="header-section-number">8</span> Biology &amp; Bioinformatics</a></li>
<li><a class="" href="engineering.html"><span class="header-section-number">9</span> EngineeRing</a></li>
<li><a class="" href="rshiny.html"><span class="header-section-number">10</span> RShiny</a></li>
<li><a class="" href="communicating-with-r.html"><span class="header-section-number">11</span> Communicating with R</a></li>
<li><a class="" href="contribution-guide.html"><span class="header-section-number">12</span> Contribution Guide</a></li>
<li class="book-part">Assignments</li>
<li><a class="" href="starting-an-assignment.html">Starting an Assignment</a></li>
<li><a class="" href="assignment-overview.html">Assignment Overview</a></li>
<li><a class="" href="assignment-0.html">Assignment 0</a></li>
<li><a class="" href="assignment-1.html">Assignment 1</a></li>
<li><a class="" href="assignment-2.html">Assignment 2</a></li>
<li><a class="" href="assignment-3.html">Assignment 3</a></li>
<li><a class="active" href="assignment-4.html">Assignment 4</a></li>
<li><a class="" href="assignment-5.html">Assignment 5</a></li>
<li><a class="" href="assignment-6.html">Assignment 6</a></li>
<li><a class="" href="assignment-7.html">Assignment 7</a></li>
<li><a class="" href="final-project.html">Final Project</a></li>
<li class="book-part">Appendix</li>
<li><a class="" href="detailed-class-schedule.html"><span class="header-section-number">A</span> Detailed Class Schedule</a></li>
</ul>

        <div class="book-extra">
          
        </div>
      </nav>
</div>
  </header><main class="col-sm-12 col-md-9 col-lg-7" id="content"><div id="assignment-4" class="section level1 unnumbered">
<h1>Assignment 4<a class="anchor" aria-label="anchor" href="#assignment-4"><i class="fas fa-link"></i></a>
</h1>
<div id="problem-statement-4" class="section level2 unnumbered">
<h2>Problem Statement<a class="anchor" aria-label="anchor" href="#problem-statement-4"><i class="fas fa-link"></i></a>
</h2>
<p>When dealing with gene counts in an mRNA-seq dataset, it is important to
normalize the data before performing any analyses so you can make accurate
comparisons of gene expression between samples. There are a number of different
methods you can use to accomplish and the method you use will depend on the
kinds of samples you have and the analyses you want to perform. This first part
of this assignment will guide you through two commonly used methods: Counts Per
Million and DESeq2 Normalization.</p>
<p>Once you normalize your data you can do exploratory analysis and construct the
appropriate graphs. As bioinformaticians, you will need to be able present your
data in tidy reports with generated plots- sometimes between blocks of text. R
Markdown is a good tool to accomplish this with, allowing you to display tables
and plots alongside the bulk of your writing. You’ve had some exposure to R
Markdown in previous assignments, now you will start writing your own code to
generate the visualizations for your report. It will be important to remember
that you need to keep your code separate depending on its function: the function
declarations and implementation- your code that will work behind-the-scenes to
perform data manipulation and process inputs without user interaction on the
“back end”- should stay in your <code>main.R</code> file while your calls to construct and
display the tibbles and visuals created in the back end should be put in your
<code>report.Rmd</code> file- despite the user not directly manipulating or editing the
outputs you generate, they are still interacting with the display by viewing it
and therefore the lines of code that call the functions to create these outputs
belong in your “front end.”</p>
</div>
<div id="learning-objectives-2" class="section level2 unnumbered">
<h2>Learning Objectives<a class="anchor" aria-label="anchor" href="#learning-objectives-2"><i class="fas fa-link"></i></a>
</h2>
<ul>
<li>Data normalization methods (CPM, DESeq2)</li>
<li>Plotting data in Tidyverse</li>
<li>Displaying results and compiling reports in R Markdown</li>
<li>Application of Tidyverse functions</li>
</ul>
</div>
<div id="skill-list-2" class="section level2 unnumbered">
<h2>Skill List<a class="anchor" aria-label="anchor" href="#skill-list-2"><i class="fas fa-link"></i></a>
</h2>
<ul>
<li>DESeq normalization, referencing the linked Bioconductor vignette if needed</li>
<li>Tibble manipulation</li>
<li>Creating different types of graphs in ggplot2</li>
<li>Running PCA</li>
</ul>
</div>
<div id="instructions-3" class="section level2 unnumbered">
<h2>Instructions<a class="anchor" aria-label="anchor" href="#instructions-3"><i class="fas fa-link"></i></a>
</h2>
<p>An empty project repo can be found here:</p>
<p><a href="https://github.com/BF591-R/bf591-assignment-4">https://github.com/BF591-R/bf591-assignment-4</a></p>
<p>You will be given</p>
<pre><code>main.R
test_main.R
report.Rmd
verse_counts.tsv
example_report.html
</code></pre>
<p>Like previous assignments, the bulk of the work will be done in the skeleton
file <code>main.R</code>, but unlike previous assignments the provided <code>report.Rmd</code> is also
a skeleton file. Both documents will provide the information needed to complete
the assignment. Please ensure that your inputs and outputs match the
specifications listed- or at the very least can properly handle inputs as
instructed and return the expected output since that is how your functions will
be tested.</p>
<p>To help you complete the assignment, you are being provided a sample report and
a set of test functions that are similar to the ones we will use to grade. The
sample report is provided in the appropriately named <code>sample_report.html</code> and
the tests for you to use can be found in <code>test_main.R</code>. Please note that while
the sample report does not display any code, it is okay and- preferred, really-
for your report to <em>display the code blocks</em> like they have done in previous
assignments.</p>
<div id="tasks" class="section level3 unnumbered">
<h3>Tasks<a class="anchor" aria-label="anchor" href="#tasks"><i class="fas fa-link"></i></a>
</h3>
<ol style="list-style-type: decimal">
<li>Implement the functions as described in <code>main.R</code>.<br>
</li>
<li>[optional but highly recommended] Test your functions using `test_main.R</li>
<li>Fill in <code>report.Rmd</code> as instructed. You will need the functions you implemented in <code>main.R</code>
</li>
<li>Knit <code>report.Rmd</code> and create your <code>report.html</code>
</li>
</ol>
</div>
</div>
<div id="deliverables" class="section level2 unnumbered">
<h2>Deliverables<a class="anchor" aria-label="anchor" href="#deliverables"><i class="fas fa-link"></i></a>
</h2>
<ol style="list-style-type: decimal">
<li>
<code>main.R</code> with all of the functions completed</li>
<li>
<code>report.Rmd</code> filled in as instructed</li>
<li>
<code>report.html</code> knitted from your <code>report.Rmd</code>
</li>
</ol>
</div>
<div id="function-details-1" class="section level2 unnumbered">
<h2>Function Details<a class="anchor" aria-label="anchor" href="#function-details-1"><i class="fas fa-link"></i></a>
</h2>
<div id="read_data" class="section level3 unnumbered">
<h3>1. <code>read_data()</code><a class="anchor" aria-label="anchor" href="#read_data"><i class="fas fa-link"></i></a>
</h3>
<p>Load a tsv located at a specified location</p>
<p>Input (1): string path to file</p>
<p>Output: (g x m) tibble</p>
<p>Details: This function will be tested for handling various input strings and
returning the proper output: dimensions, column names, if column ‘gene’ is
located in the first column, and output type tibble. Your test_main will include
additional testing of output column types and lack of row names for your
reference, to help catch errors early on.</p>
</div>
<div id="filter_zero_var_genes" class="section level3 unnumbered">
<h3>2. <code>filter_zero_var_genes()</code><a class="anchor" aria-label="anchor" href="#filter_zero_var_genes"><i class="fas fa-link"></i></a>
</h3>
<p>Filter out genes with zero variance</p>
<p>Input (1): (g x m) tibble</p>
<p>Output: (n x m) tibble</p>
<p>Details: This function will be tested for handling an input (g x m) tibble and
returning the proper output: column names, if column ‘gene’ is located in the
first column, names of ‘genes’ returned, and output type tibble. Your test_main
will include additional testing for row consistency- ensuring that the sample
data still correlates with the gene names- for your reference and to help catch
errors early on</p>
</div>
<div id="timepoint_from_sample" class="section level3 unnumbered">
<h3>3. <code>timepoint_from_sample()</code><a class="anchor" aria-label="anchor" href="#timepoint_from_sample"><i class="fas fa-link"></i></a>
</h3>
<p>Extract time point information from sample name</p>
<p>Input (1): string (length 5) of sample name in format <code>v[A-Z][a-z,1-9]_[1-9]</code>
(In other words: <code>v[α][β]_[γ]</code>, where α is any capital modern English letter, β
is any lower case modern English letter OR any Arabic number 0-9, and γ is any
Arabic number from 0-9)</p>
<p>Output: string (length 2) of substring <code>[A-Z][a-z,1-9]</code> from sample name:</p>
<p>Details: This function will be tested for handling various strings of length 5
in the form <code>v[A-Z][a-z,1-9]_[1-9]</code> and outputting the proper string, preserving
letter case where letter case is provided.</p>
</div>
<div id="sample_replicate" class="section level3 unnumbered">
<h3>4. <code>sample_replicate()</code><a class="anchor" aria-label="anchor" href="#sample_replicate"><i class="fas fa-link"></i></a>
</h3>
<p>Grab sample replicate number from sample name</p>
<p>Input (1): Input (1): string (length 5) of sample name in format
<code>v[A-Z][a-z,1-9]_[1-9]</code> (In other words: <code>v[α][β]_[γ]</code>, where α is any capital
modern English letter, β is any lower case modern English letter OR any Arabic
number 0-9, and γ is any Arabic number from 0-9)</p>
<p>Output: string (length 1 of substring <code>[1-9]</code> from sample name:
<code>v[A-Z][a-z,1-9]_[1-9]</code> (In other words: [γ] from <code>v[α][β]_[γ]</code>)</p>
<p>Details: This function will be tested for handling various strings of length 5
in the form <code>v[A-Z][a-z,1-9]_[1-9]</code> and outputting the proper character string.</p>
</div>
<div id="meta_info_from_labels" class="section level3 unnumbered">
<h3>5. <code>meta_info_from_labels()</code><a class="anchor" aria-label="anchor" href="#meta_info_from_labels"><i class="fas fa-link"></i></a>
</h3>
<p>Generate sample-level metadata from sample names and stores the data into a
tibble. Will include columns named “sample”, “timepoint”, and “replicate” that
store sample names, sample time points, and sample replicate, respectively.</p>
<p>Input (1): Character vector of length <code>_S_</code> of sample names with column names “sample”, “timepoint”, and “replicate”</p>
<p>Output: a <code>(_S_ x 3)</code> tibble</p>
<p>Details: This function will be tested for handling of a character vector of
length <code>_S_</code>, where each element in the vector is a string with a length of 5 in
the form of <code>v[A-Z][a-z,1-9]_[1-9]</code>, and properly outputting a <code>(_S_ x 3)</code>
tibble with columns named “sample”, “timepoint”, and “replicate” and rows that
correspond with the input samples. Your test_main will include additional
testing for the order of elements in column ‘sample’’s correspondence with the
order of elements in the input vector for your reference. The column types of
your output tibble will not be tested.</p>
</div>
<div id="get_library_size" class="section level3 unnumbered">
<h3>6. <code>get_library_size()</code><a class="anchor" aria-label="anchor" href="#get_library_size"><i class="fas fa-link"></i></a>
</h3>
<p>Calculate total read counts for each sample in a counts dataset.</p>
<p>Input (1): a (n x m) tibble of raw read counts</p>
<p>Output: tibble or named vector of read totals from each sample. Vectors must be
length <code>_S_</code>, a tibble can be <code>(1 x _S_)</code> with sample names as columns names OR
<code>(_S_ x 2)</code> where sample name is in the first column and library size is the
second column</p>
<p>Details: This function will be tested for the return of a tibble or named vector
that have sample names which correspond with the appropriate library size.</p>
</div>
<div id="normalize_by_cpm" class="section level3 unnumbered">
<h3>7. <code>normalize_by_cpm()</code><a class="anchor" aria-label="anchor" href="#normalize_by_cpm"><i class="fas fa-link"></i></a>
</h3>
<p>Normalize raw counts data to counts per million using (counts) /
(sample_library_size) * 10^6</p>
<p>Input (1): a (n x m) tibble of raw read counts</p>
<p>Output: a (n x m) tibble with read count normalized to counts per million</p>
<p>Details: This function will be tested to handle a (n x m) tibble. Its output
will be tested for dimensions, column names, location of string and numeric
column(s), performance of the desired equation on numeric columns, and that gene
names still correspond to their rows.</p>
</div>
<div id="deseq_normalize" class="section level3 unnumbered">
<h3>8. <code>deseq_normalize()</code><a class="anchor" aria-label="anchor" href="#deseq_normalize"><i class="fas fa-link"></i></a>
</h3>
<p>Normalize raw counts data using DESeq2</p>
<p>Input (1): a (n x m) tibble of raw read counts</p>
<p>Output: a (n x m) tibble of DESeq2 normalized counts data</p>
<p>Details:This function will be tested to handle a (n x m) tibble. Its output will
be tested for dimensions, column names, location of string and numeric
column(s), performance of the desired equation on numeric columns, and that gene
names still correspond to their rows.</p>
</div>
<div id="plot_pca" class="section level3 unnumbered">
<h3>9. <code>plot_pca()</code><a class="anchor" aria-label="anchor" href="#plot_pca"><i class="fas fa-link"></i></a>
</h3>
<p>Input (3): a <code>(n x _S_)</code> tibble of data, a <code>(_S_ x 3)</code> tibble of sample-level
meta information, and a string</p>
<p>Output: a ggplot scatter plot showing each sample, with PC1 on x-axis and PC2 on y-axis</p>
<p>Details: The output of this will be tested for the appropriate test run and PCs
used to plot. It may be visually inspected as part of your grade</p>
</div>
<div id="plot_sample_distributions" class="section level3 unnumbered">
<h3>10. <code>plot_sample_distributions()</code><a class="anchor" aria-label="anchor" href="#plot_sample_distributions"><i class="fas fa-link"></i></a>
</h3>
<p>Input (3): a <code>(n x _S_)</code> tibble of data, a boolean to determine whether to scale
the ‘y’ axis to log10 values, and a string</p>
<p>Output: a ggplot boxplot that shows gene count distributions</p>
<p>Details: This function will be tested on a <code>(n x _S_)</code> tibble. It will be tested
for functionality of its inputs, handling of data, expected graph elements, and
graph type. It may also be visually inspected as part of your grade</p>
</div>
<div id="plot_variance_vs_mean" class="section level3 unnumbered">
<h3>11. <code>plot_variance_vs_mean()</code><a class="anchor" aria-label="anchor" href="#plot_variance_vs_mean"><i class="fas fa-link"></i></a>
</h3>
<p>Input (3): a <code>(n x _S_)</code> tibble of data, a boolean to determine whether to scale
the ‘y’ axis to log10 values, and a string</p>
<p>Output: a ggplot scatter plot where the x-axis is the rank of gene ordered by
mean count over all samples, and the y-axis is the observed variance of the
given gene. Each dot should have their transparency increased. The scatter plot
should also be accompanied by a line representing the average mean and variance
values</p>
<p>Details: This function will be tested on a <code>(n x _S_)</code> tibble. It will be tested
for functionality of its inputs, handling of data, expected graph elements, and
graph type. It may also be visually inspected as part of your grade</p>
</div>
</div>
<div id="hints-2" class="section level2 unnumbered">
<h2>Hints<a class="anchor" aria-label="anchor" href="#hints-2"><i class="fas fa-link"></i></a>
</h2>
<ul>
<li><p>Make sure you don’t have any rownames on any of the tibbles your function(s)</p></li>
<li><p>return. You should get a warning in your R console if you do.</p></li>
<li><p>Sometimes Tydyverse is the best tool for the job, sometimes it isn’t</p></li>
<li><p>Some tasks are easier to accomplish with dataframes than tibbles. Not everyone
will use these methods, but if you do you are welcome to use dataframes within
a function as long as the inputs and outputs are the ones specified (returning
a dataframe instead of a tibble may points).</p></li>
<li><p>The Bioconductor vignette for DESeq2 is linked in part 3 of your <code>report.Rmd</code>.
<a href="http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#data-transformations-and-visualization">It can also be found
here</a>)</p></li>
<li><p>In a previous version of this assignment, the output of
<code>filter_zero_var_genes()</code> was transformed from <code>(n x m)</code> to <code>(n x _s_)</code> and
called the <code>counts_matrix</code>. One of the TAs thought use of ‘matrix’ would be
confusing (since we want you to use datasets of type <code>tibble</code>) so it was
re-worded, but this piece of trivia might be helpful for interpreting the
Bioconductor vignette- or if you see any sort of ‘matrix’ referenced in
<code>report.Rmd</code> because it got overlooked during editing.</p></li>
<li><p>There may be occasions where you will need to reshape your data from a wide</p></li>
<li><p>format to a long format</p></li>
<li><p>Symbols used in <code>main.R</code>~ <code>g</code>: initial number of Genes, <code>m</code>: initial number of
columns expected when you import <code>verse_counts.tsv</code>, <code>n</code>: number of genes
expected after you filter in part 1b, <code>_S_</code>: number of Samples</p></li>
</ul>
</div>
</div>
  <div class="chapter-nav">
<div class="prev"><a href="assignment-3.html">Assignment 3</a></div>
<div class="next"><a href="assignment-5.html">Assignment 5</a></div>
</div></main><div class="col-md-3 col-lg-2 d-none d-md-block sidebar sidebar-chapter">
    <nav id="toc" data-toggle="toc" aria-label="On this page"><h2>On this page</h2>
      <ul class="nav navbar-nav">
<li><a class="nav-link" href="#assignment-4">Assignment 4</a></li>
<li><a class="nav-link" href="#problem-statement-4">Problem Statement</a></li>
<li><a class="nav-link" href="#learning-objectives-2">Learning Objectives</a></li>
<li><a class="nav-link" href="#skill-list-2">Skill List</a></li>
<li>
<a class="nav-link" href="#instructions-3">Instructions</a><ul class="nav navbar-nav"><li><a class="nav-link" href="#tasks">Tasks</a></li></ul>
</li>
<li><a class="nav-link" href="#deliverables">Deliverables</a></li>
<li>
<a class="nav-link" href="#function-details-1">Function Details</a><ul class="nav navbar-nav">
<li><a class="nav-link" href="#read_data">1. read_data()</a></li>
<li><a class="nav-link" href="#filter_zero_var_genes">2. filter_zero_var_genes()</a></li>
<li><a class="nav-link" href="#timepoint_from_sample">3. timepoint_from_sample()</a></li>
<li><a class="nav-link" href="#sample_replicate">4. sample_replicate()</a></li>
<li><a class="nav-link" href="#meta_info_from_labels">5. meta_info_from_labels()</a></li>
<li><a class="nav-link" href="#get_library_size">6. get_library_size()</a></li>
<li><a class="nav-link" href="#normalize_by_cpm">7. normalize_by_cpm()</a></li>
<li><a class="nav-link" href="#deseq_normalize">8. deseq_normalize()</a></li>
<li><a class="nav-link" href="#plot_pca">9. plot_pca()</a></li>
<li><a class="nav-link" href="#plot_sample_distributions">10. plot_sample_distributions()</a></li>
<li><a class="nav-link" href="#plot_variance_vs_mean">11. plot_variance_vs_mean()</a></li>
</ul>
</li>
<li><a class="nav-link" href="#hints-2">Hints</a></li>
</ul>

      <div class="book-extra">
        <ul class="list-unstyled">
          
        </ul>
</div>
    </nav>
</div>

</div>
</div> <!-- .container -->

<footer class="bg-primary text-light mt-5"><div class="container"><div class="row">

  <div class="col-12 col-md-6 mt-3">
    <p>"<strong>BF591 - R for Biological Sciences</strong>" was written by . </p>
  </div>

  <div class="col-12 col-md-6 mt-3">
    <p>This book was built by the <a class="text-light" href="https://bookdown.org">bookdown</a> R package.</p>
  </div>

</div></div>
</footer><!-- dynamically load mathjax for compatibility with self-contained --><script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/latest.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script><script type="text/x-mathjax-config">const popovers = document.querySelectorAll('a.footnote-ref[data-toggle="popover"]');
for (let popover of popovers) {
  const div = document.createElement('div');
  div.setAttribute('style', 'position: absolute; top: 0, left:0; width:0, height:0, overflow: hidden; visibility: hidden;');
  div.innerHTML = popover.getAttribute('data-content');

  var has_math = div.querySelector("span.math");
  if (has_math) {
    document.body.appendChild(div);
    MathJax.Hub.Queue(["Typeset", MathJax.Hub, div]);
    MathJax.Hub.Queue(function() {
      popover.setAttribute('data-content', div.innerHTML);
      document.body.removeChild(div);
    })
  }
}
</script>
</body>
</html>
