<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta name="viewport" content="width=device-width, initial-scale=1" />
<meta property="og:title" content="Assignment 6 | BF591 - R for Biological Sciences" />
<meta property="og:type" content="book" />







<script type="text/x-mathjax-config">
MathJax.Hub.Config({
  TeX: { equationNumbers: { autoNumber: "AMS" } }
});
</script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<meta name="description" content="Assignment 6 | BF591 - R for Biological Sciences">

<title>Assignment 6 | BF591 - R for Biological Sciences</title>

<script src="libs/header-attrs-2.13/header-attrs.js"></script>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="libs/bootstrap-3.3.5/css/darkly.min.css" rel="stylesheet" />
<script src="libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="libs/navigation-1.1/tabsets.js"></script>


<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>


<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />

<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
/* show arrow before summary tag as in bootstrap
TODO: remove if boostrap in updated in html_document (rmarkdown#1485) */
details > summary {
  display: list-item;
  cursor: pointer;
}
</style>
</head>

<body>

<div class="container-fluid main-container">


<div class="row">
<div class="col-sm-12">
<div id="TOC">
<ul>
<li><a href="index.html#syllabus">Syllabus</a>
<ul>
<li><a href="course-schedule.html#course-schedule">Course Schedule</a></li>
<li><a href="instructors.html#instructors">Instructors</a></li>
<li><a href="office-hours.html#office-hours">Office Hours</a></li>
<li><a href="course-values-and-policies.html#course-values-and-policies">Course Values and Policies</a></li>
</ul></li>
<li><a href="introduction.html#introduction"><span class="toc-section-number">1</span> Introduction</a>
<ul>
<li><a href="who-this-book-is-for.html#who-this-book-is-for"><span class="toc-section-number">1.1</span> Who This Book Is For</a></li>
<li><a href="a-note-about-reinventing-the-wheel.html#a-note-about-reinventing-the-wheel"><span class="toc-section-number">1.2</span> A Note About Reinventing the Wheel</a></li>
<li><a href="sources-and-references.html#sources-and-references"><span class="toc-section-number">1.3</span> Sources and References</a></li>
</ul></li>
<li><a href="data-bio.html#data-bio"><span class="toc-section-number">2</span> Data in Biology</a>
<ul>
<li><a href="a-brief-history-of-data-in-molecular-biology.html#a-brief-history-of-data-in-molecular-biology"><span class="toc-section-number">2.1</span> A Brief History of Data in Molecular Biology</a></li>
<li><a href="biology-as-a-mature-data-science.html#biology-as-a-mature-data-science"><span class="toc-section-number">2.2</span> Biology as a Mature Data Science</a></li>
</ul></li>
<li><a href="preliminaries.html#preliminaries"><span class="toc-section-number">3</span> Preliminaries</a>
<ul>
<li><a href="prelim-r.html#prelim-r"><span class="toc-section-number">3.1</span> The R Language</a></li>
<li><a href="prelim-rstudio.html#prelim-rstudio"><span class="toc-section-number">3.2</span> RStudio</a></li>
<li><a href="prog-the-r-script.html#prog-the-r-script"><span class="toc-section-number">3.3</span> The R Script</a></li>
<li><a href="prog-workflow.html#prog-workflow"><span class="toc-section-number">3.4</span> The Scripting Workflow</a></li>
<li><a href="prelim-git.html#prelim-git"><span class="toc-section-number">3.5</span> git + github</a>
<ul>
<li><a href="prelim-git.html#motivation"><span class="toc-section-number">3.5.1</span> Motivation</a></li>
<li><a href="prelim-git.html#git"><span class="toc-section-number">3.5.2</span> git</a></li>
<li><a href="prelim-git.html#git-hosting-platforms-github"><span class="toc-section-number">3.5.3</span> Git hosting platforms (GitHub)</a></li>
</ul></li>
</ul></li>
<li><a href="prog-basics.html#prog-basics"><span class="toc-section-number">4</span> R Programming</a>
<ul>
<li><a href="before-you-begin.html#before-you-begin"><span class="toc-section-number">4.1</span> Before you begin</a></li>
<li><a href="introduction-1.html#introduction-1"><span class="toc-section-number">4.2</span> Introduction</a></li>
<li><a href="prog-r-syntax.html#prog-r-syntax"><span class="toc-section-number">4.3</span> R Syntax Basics</a></li>
<li><a href="prog-types.html#prog-types"><span class="toc-section-number">4.4</span> Basic Types of Values</a>
<ul>
<li><a href="prog-types.html#factors"><span class="toc-section-number">4.4.1</span> Factors</a></li>
</ul></li>
<li><a href="prog-struct.html#prog-struct"><span class="toc-section-number">4.5</span> Data Structures</a>
<ul>
<li><a href="prog-struct.html#vectors"><span class="toc-section-number">4.5.1</span> Vectors</a></li>
<li><a href="prog-struct.html#matrices"><span class="toc-section-number">4.5.2</span> Matrices</a></li>
<li><a href="prog-struct.html#lists-and-data-frames"><span class="toc-section-number">4.5.3</span> Lists and data frames</a></li>
</ul></li>
<li><a href="logical-tests-and-comparators.html#logical-tests-and-comparators"><span class="toc-section-number">4.6</span> Logical Tests and Comparators</a></li>
<li><a href="functions.html#functions"><span class="toc-section-number">4.7</span> Functions</a>
<ul>
<li><a href="functions.html#prog-dry"><span class="toc-section-number">4.7.1</span> DRY: Don’t Repeat Yourself</a></li>
<li><a href="functions.html#writing-your-own-functions"><span class="toc-section-number">4.7.2</span> Writing your own functions</a></li>
<li><a href="functions.html#scope"><span class="toc-section-number">4.7.3</span> Scope</a></li>
</ul></li>
<li><a href="iteration.html#iteration"><span class="toc-section-number">4.8</span> Iteration</a>
<ul>
<li><a href="iteration.html#vectorized-operations"><span class="toc-section-number">4.8.1</span> Vectorized operations</a></li>
<li><a href="iteration.html#functional-programming"><span class="toc-section-number">4.8.2</span> Functional programming</a></li>
<li><a href="iteration.html#apply-and-friends"><span class="toc-section-number">4.8.3</span> <code>apply()</code> and friends</a></li>
</ul></li>
<li><a href="installing-packages.html#installing-packages"><span class="toc-section-number">4.9</span> Installing Packages</a></li>
<li><a href="saving-and-loading-r-data.html#saving-and-loading-r-data"><span class="toc-section-number">4.10</span> Saving and Loading R Data</a></li>
<li><a href="prog-debugging.html#prog-debugging"><span class="toc-section-number">4.11</span> Troubleshooting and Debugging</a></li>
<li><a href="prog-style.html#prog-style"><span class="toc-section-number">4.12</span> Coding Style and Conventions</a>
<ul>
<li><a href="prog-style.html#is-my-code-correct"><span class="toc-section-number">4.12.1</span> Is my code correct?</a></li>
<li><a href="prog-style.html#does-my-code-follow-the-dry-principle"><span class="toc-section-number">4.12.2</span> Does my code follow the <span>DRY principle</span>?</a></li>
<li><a href="prog-style.html#did-i-choose-concise-but-descriptive-variable-and-function-names"><span class="toc-section-number">4.12.3</span> Did I choose concise but descriptive variable and function names?</a></li>
<li><a href="prog-style.html#did-i-use-indentation-and-naming-conventions-consistently-throughout-my-code"><span class="toc-section-number">4.12.4</span> Did I use indentation and naming conventions consistently throughout my code?</a></li>
<li><a href="prog-style.html#did-i-write-comments-especially-when-what-the-code-does-is-not-obvious"><span class="toc-section-number">4.12.5</span> Did I write comments, especially when what the code does is not obvious?</a></li>
<li><a href="prog-style.html#how-easy-would-it-be-for-someone-else-to-understand-my-code"><span class="toc-section-number">4.12.6</span> How easy would it be for someone else to understand my code?</a></li>
<li><a href="prog-style.html#is-my-code-easy-to-maintainchange"><span class="toc-section-number">4.12.7</span> Is my code easy to maintain/change?</a></li>
<li><a href="prog-style.html#the-styler-package"><span class="toc-section-number">4.12.8</span> The <code>styler</code> package</a></li>
</ul></li>
</ul></li>
<li><a href="data-wrangling.html#data-wrangling"><span class="toc-section-number">5</span> Data Wrangling</a>
<ul>
<li><a href="the-tidyverse.html#the-tidyverse"><span class="toc-section-number">5.1</span> The Tidyverse</a></li>
<li><a href="dw-basics.html#dw-basics"><span class="toc-section-number">5.2</span> Tidyverse Basics</a></li>
<li><a href="importing-data.html#importing-data"><span class="toc-section-number">5.3</span> Importing Data</a></li>
<li><a href="data-tibble.html#data-tibble"><span class="toc-section-number">5.4</span> The tibble</a></li>
<li><a href="tidy-data.html#tidy-data"><span class="toc-section-number">5.5</span> Tidy Data</a></li>
<li><a href="pipes.html#pipes"><span class="toc-section-number">5.6</span> pipes</a></li>
<li><a href="arranging-data.html#arranging-data"><span class="toc-section-number">5.7</span> Arranging Data</a>
<ul>
<li><a href="arranging-data.html#dplyrmutate---create-new-columns-using-other-columns"><span class="toc-section-number">5.7.1</span> <code>dplyr::mutate()</code> - Create new columns using other columns</a></li>
<li><a href="arranging-data.html#stringr---working-with-character-values"><span class="toc-section-number">5.7.2</span> <code>stringr</code> - Working with character values</a></li>
<li><a href="arranging-data.html#dplyrselect---subset-columns-by-name"><span class="toc-section-number">5.7.3</span> <code>dplyr::select()</code> - Subset Columns by Name</a></li>
<li><a href="arranging-data.html#dplyrfilter---pick-rows-out-of-a-data-set"><span class="toc-section-number">5.7.4</span> <code>dplyr::filter()</code> - Pick rows out of a data set</a></li>
<li><a href="arranging-data.html#dplyrarrange---order-rows-based-on-their-values"><span class="toc-section-number">5.7.5</span> <code>dplyr::arrange()</code> - Order rows based on their values</a></li>
<li><a href="arranging-data.html#putting-it-all-together"><span class="toc-section-number">5.7.6</span> Putting it all together</a></li>
</ul></li>
<li><a href="grouping-data.html#grouping-data"><span class="toc-section-number">5.8</span> Grouping Data</a></li>
<li><a href="rearranging-data.html#rearranging-data"><span class="toc-section-number">5.9</span> Rearranging Data</a></li>
<li><a href="relational-data.html#relational-data"><span class="toc-section-number">5.10</span> Relational Data</a>
<ul>
<li><a href="relational-data.html#dealing-with-multiple-matches"><span class="toc-section-number">5.10.1</span> Dealing with multiple matches</a></li>
</ul></li>
</ul></li>
<li><a href="data-science.html#data-science"><span class="toc-section-number">6</span> Data Science</a>
<ul>
<li><a href="data-modeling.html#data-modeling"><span class="toc-section-number">6.1</span> Data Modeling</a>
<ul>
<li><a href="data-modeling.html#a-worked-modeling-example"><span class="toc-section-number">6.1.1</span> A Worked Modeling Example</a></li>
<li><a href="data-modeling.html#data-summarization"><span class="toc-section-number">6.1.2</span> Data Summarization</a></li>
<li><a href="data-modeling.html#linear-models"><span class="toc-section-number">6.1.3</span> Linear Models</a></li>
<li><a href="data-modeling.html#flavors-of-linear-models"><span class="toc-section-number">6.1.4</span> Flavors of Linear Models</a></li>
<li><a href="data-modeling.html#assessing-model-accuracy-."><span class="toc-section-number">6.1.5</span> Assessing Model Accuracy .</a></li>
</ul></li>
<li><a href="statistical-distributions.html#statistical-distributions"><span class="toc-section-number">6.2</span> Statistical Distributions</a>
<ul>
<li><a href="statistical-distributions.html#random-variables"><span class="toc-section-number">6.2.1</span> Random Variables</a></li>
<li><a href="statistical-distributions.html#statistical-distribution-basics"><span class="toc-section-number">6.2.2</span> Statistical Distribution Basics</a></li>
<li><a href="statistical-distributions.html#distributions-in-r"><span class="toc-section-number">6.2.3</span> Distributions in R</a></li>
<li><a href="statistical-distributions.html#discrete-distributions"><span class="toc-section-number">6.2.4</span> Discrete Distributions</a></li>
<li><a href="statistical-distributions.html#continuous-distributions"><span class="toc-section-number">6.2.5</span> Continuous Distributions</a></li>
<li><a href="statistical-distributions.html#empirical-distributions"><span class="toc-section-number">6.2.6</span> Empirical Distributions</a></li>
</ul></li>
<li><a href="statistical-tests.html#statistical-tests"><span class="toc-section-number">6.3</span> Statistical Tests</a>
<ul>
<li><a href="statistical-tests.html#what-is-a-statistical-test"><span class="toc-section-number">6.3.1</span> What is a statistical test</a></li>
<li><a href="statistical-tests.html#common-statistical-tests"><span class="toc-section-number">6.3.2</span> Common Statistical Tests</a></li>
<li><a href="statistical-tests.html#choosing-a-test"><span class="toc-section-number">6.3.3</span> Choosing a Test</a></li>
<li><a href="statistical-tests.html#p-values"><span class="toc-section-number">6.3.4</span> P-values</a></li>
<li><a href="statistical-tests.html#multiple-hypothesis-testing-."><span class="toc-section-number">6.3.5</span> Multiple Hypothesis Testing .</a></li>
<li><a href="statistical-tests.html#statistical-power"><span class="toc-section-number">6.3.6</span> Statistical power</a></li>
</ul></li>
<li><a href="exploratory-data-analysis.html#exploratory-data-analysis"><span class="toc-section-number">6.4</span> Exploratory Data Analysis</a>
<ul>
<li><a href="exploratory-data-analysis.html#principal-component-analysis"><span class="toc-section-number">6.4.1</span> Principal Component Analysis</a></li>
<li><a href="exploratory-data-analysis.html#cluster-analysis"><span class="toc-section-number">6.4.2</span> Cluster Analysis</a></li>
<li><a href="exploratory-data-analysis.html#k-means-."><span class="toc-section-number">6.4.3</span> k-means .</a></li>
<li><a href="exploratory-data-analysis.html#others"><span class="toc-section-number">6.4.4</span> Others</a></li>
</ul></li>
<li><a href="network-analysis-..html#network-analysis-."><span class="toc-section-number">6.5</span> Network Analysis .</a>
<ul>
<li><a href="network-analysis-..html#basic-network-analysis"><span class="toc-section-number">6.5.1</span> Basic network analysis</a></li>
<li><a href="network-analysis-..html#create-your-own"><span class="toc-section-number">6.5.2</span> Create your own</a></li>
</ul></li>
</ul></li>
<li><a href="data-visualization.html#data-visualization"><span class="toc-section-number">7</span> Data Visualization</a>
<ul>
<li><a href="responsible-plotting.html#responsible-plotting"><span class="toc-section-number">7.1</span> Responsible Plotting</a>
<ul>
<li><a href="responsible-plotting.html#visualization-principles"><span class="toc-section-number">7.1.1</span> Visualization Principles</a></li>
<li><a href="responsible-plotting.html#human-visual-perception"><span class="toc-section-number">7.1.2</span> Human Visual Perception</a></li>
<li><a href="responsible-plotting.html#visual-encodings"><span class="toc-section-number">7.1.3</span> Visual Encodings</a></li>
<li><a href="responsible-plotting.html#some-opinionated-rules-of-thumb"><span class="toc-section-number">7.1.4</span> Some Opinionated Rules of Thumb</a></li>
</ul></li>
<li><a href="dv-gg.html#dv-gg"><span class="toc-section-number">7.2</span> Grammar of Graphics</a>
<ul>
<li><a href="dv-gg.html#ggplot-mechanics"><span class="toc-section-number">7.2.1</span> <code>ggplot</code> mechanics</a></li>
</ul></li>
<li><a href="plotting-one-dimension.html#plotting-one-dimension"><span class="toc-section-number">7.3</span> Plotting One Dimension</a>
<ul>
<li><a href="plotting-one-dimension.html#dv-bar"><span class="toc-section-number">7.3.1</span> Bar chart</a></li>
<li><a href="plotting-one-dimension.html#lollipop-plots"><span class="toc-section-number">7.3.2</span> Lollipop plots</a></li>
<li><a href="plotting-one-dimension.html#stacked-area-charts"><span class="toc-section-number">7.3.3</span> Stacked Area charts</a></li>
<li><a href="plotting-one-dimension.html#parallel-coordinate-plots"><span class="toc-section-number">7.3.4</span> Parallel Coordinate plots</a></li>
</ul></li>
<li><a href="visualizing-distributions.html#visualizing-distributions"><span class="toc-section-number">7.4</span> Visualizing Distributions</a>
<ul>
<li><a href="visualizing-distributions.html#histogram"><span class="toc-section-number">7.4.1</span> Histogram</a></li>
<li><a href="visualizing-distributions.html#density"><span class="toc-section-number">7.4.2</span> Density</a></li>
<li><a href="visualizing-distributions.html#boxplot"><span class="toc-section-number">7.4.3</span> Boxplot</a></li>
<li><a href="visualizing-distributions.html#violin-plot"><span class="toc-section-number">7.4.4</span> Violin plot</a></li>
<li><a href="visualizing-distributions.html#beeswarm-plot"><span class="toc-section-number">7.4.5</span> Beeswarm plot</a></li>
<li><a href="visualizing-distributions.html#ridgeline"><span class="toc-section-number">7.4.6</span> Ridgeline</a></li>
</ul></li>
<li><a href="plotting-two-or-more-dimensions.html#plotting-two-or-more-dimensions"><span class="toc-section-number">7.5</span> Plotting Two or More Dimensions</a>
<ul>
<li><a href="plotting-two-or-more-dimensions.html#scatter-plots"><span class="toc-section-number">7.5.1</span> Scatter Plots</a></li>
<li><a href="plotting-two-or-more-dimensions.html#bubble-plots"><span class="toc-section-number">7.5.2</span> Bubble Plots</a></li>
<li><a href="plotting-two-or-more-dimensions.html#connected-scatter-plots"><span class="toc-section-number">7.5.3</span> Connected Scatter Plots</a></li>
<li><a href="plotting-two-or-more-dimensions.html#line-plots"><span class="toc-section-number">7.5.4</span> Line Plots</a></li>
<li><a href="plotting-two-or-more-dimensions.html#parallel-coordinate-plots-1"><span class="toc-section-number">7.5.5</span> Parallel Coordinate Plots</a></li>
<li><a href="plotting-two-or-more-dimensions.html#heatmaps"><span class="toc-section-number">7.5.6</span> Heatmaps</a></li>
<li><a href="plotting-two-or-more-dimensions.html#how-to-use-heatmaps-responsibly"><span class="toc-section-number">7.5.7</span> How To Use Heatmaps Responsibly</a></li>
</ul></li>
<li><a href="other-kind-of-plots.html#other-kind-of-plots"><span class="toc-section-number">7.6</span> Other Kind of Plots</a>
<ul>
<li><a href="other-kind-of-plots.html#dendrograms"><span class="toc-section-number">7.6.1</span> Dendrograms</a></li>
<li><a href="other-kind-of-plots.html#chord-diagrams-and-circos-plots"><span class="toc-section-number">7.6.2</span> Chord Diagrams and Circos Plots</a></li>
</ul></li>
<li><a href="multiple-plots.html#multiple-plots"><span class="toc-section-number">7.7</span> Multiple Plots</a>
<ul>
<li><a href="multiple-plots.html#facet-wrapping"><span class="toc-section-number">7.7.1</span> Facet wrapping</a></li>
<li><a href="multiple-plots.html#multipanel-figures"><span class="toc-section-number">7.7.2</span> Multipanel Figures</a></li>
</ul></li>
<li><a href="publication-ready-plots.html#publication-ready-plots"><span class="toc-section-number">7.8</span> Publication Ready Plots</a>
<ul>
<li><a href="publication-ready-plots.html#ggplot-themes"><span class="toc-section-number">7.8.1</span> <code>ggplot</code> Themes</a></li>
<li><a href="publication-ready-plots.html#exporting-to-svg"><span class="toc-section-number">7.8.2</span> Exporting to SVG</a></li>
</ul></li>
<li><a href="network-visualization.html#network-visualization"><span class="toc-section-number">7.9</span> Network visualization</a>
<ul>
<li><a href="network-visualization.html#interactive-network-plots"><span class="toc-section-number">7.9.1</span> Interactive network plots</a></li>
</ul></li>
</ul></li>
<li><a href="biology-bioinformatics.html#biology-bioinformatics"><span class="toc-section-number">8</span> Biology &amp; Bioinformatics</a>
<ul>
<li><a href="bio-r.html#bio-r"><span class="toc-section-number">8.1</span> R in Biology</a></li>
<li><a href="biological-data-overview.html#biological-data-overview"><span class="toc-section-number">8.2</span> Biological Data Overview</a>
<ul>
<li><a href="biological-data-overview.html#bio-data-types"><span class="toc-section-number">8.2.1</span> Types of Biological Data</a></li>
<li><a href="biological-data-overview.html#csv-files"><span class="toc-section-number">8.2.2</span> CSV Files</a></li>
<li><a href="biological-data-overview.html#biological-data-is-not-tidy"><span class="toc-section-number">8.2.3</span> Biological data is NOT Tidy!</a></li>
</ul></li>
<li><a href="bio-bioconductor.html#bio-bioconductor"><span class="toc-section-number">8.3</span> Bioconductor</a></li>
<li><a href="microarrays.html#microarrays"><span class="toc-section-number">8.4</span> Microarrays</a></li>
<li><a href="high-throughput-sequencing.html#high-throughput-sequencing"><span class="toc-section-number">8.5</span> High Throughput Sequencing</a>
<ul>
<li><a href="high-throughput-sequencing.html#raw-hts-data"><span class="toc-section-number">8.5.1</span> Raw HTS Data</a></li>
<li><a href="high-throughput-sequencing.html#preliminary-hts-data-analysis"><span class="toc-section-number">8.5.2</span> Preliminary HTS Data Analysis</a></li>
<li><a href="high-throughput-sequencing.html#count-data"><span class="toc-section-number">8.5.3</span> Count Data</a></li>
</ul></li>
<li><a href="gene-identifiers.html#gene-identifiers"><span class="toc-section-number">8.6</span> Gene Identifiers</a>
<ul>
<li><a href="gene-identifiers.html#gene-identifier-systems"><span class="toc-section-number">8.6.1</span> Gene Identifier Systems</a></li>
<li><a href="gene-identifiers.html#mapping-between-identifier-systems"><span class="toc-section-number">8.6.2</span> Mapping Between Identifier Systems</a></li>
<li><a href="gene-identifiers.html#mapping-homologs"><span class="toc-section-number">8.6.3</span> Mapping Homologs</a></li>
</ul></li>
<li><a href="gene-expression.html#gene-expression"><span class="toc-section-number">8.7</span> Gene Expression</a>
<ul>
<li><a href="gene-expression.html#gene-expression-data-in-bioconductor"><span class="toc-section-number">8.7.1</span> Gene Expression Data in Bioconductor</a></li>
<li><a href="gene-expression.html#differential-expression-analysis"><span class="toc-section-number">8.7.2</span> Differential Expression Analysis</a></li>
<li><a href="gene-expression.html#microarray-gene-expression-data"><span class="toc-section-number">8.7.3</span> Microarray Gene Expression Data</a></li>
<li><a href="gene-expression.html#differential-expression-microarrays-limma"><span class="toc-section-number">8.7.4</span> Differential Expression: Microarrays (limma)</a></li>
<li><a href="gene-expression.html#rnaseq"><span class="toc-section-number">8.7.5</span> RNASeq</a></li>
<li><a href="gene-expression.html#rnaseq-gene-expression-data"><span class="toc-section-number">8.7.6</span> RNASeq Gene Expression Data</a></li>
<li><a href="gene-expression.html#filtering-counts"><span class="toc-section-number">8.7.7</span> Filtering Counts</a></li>
<li><a href="gene-expression.html#count-distributions"><span class="toc-section-number">8.7.8</span> Count Distributions</a></li>
<li><a href="gene-expression.html#differential-expression-rnaseq"><span class="toc-section-number">8.7.9</span> Differential Expression: RNASeq</a></li>
</ul></li>
<li><a href="gene-set-enrichment-analysis.html#gene-set-enrichment-analysis"><span class="toc-section-number">8.8</span> Gene Set Enrichment Analysis</a>
<ul>
<li><a href="gene-set-enrichment-analysis.html#gene-sets"><span class="toc-section-number">8.8.1</span> Gene Sets</a></li>
<li><a href="gene-set-enrichment-analysis.html#working-with-gene-sets-in-r"><span class="toc-section-number">8.8.2</span> Working with gene sets in R</a></li>
<li><a href="gene-set-enrichment-analysis.html#over-representation-analysis"><span class="toc-section-number">8.8.3</span> Over-representation Analysis</a></li>
<li><a href="gene-set-enrichment-analysis.html#rank-based-analysis"><span class="toc-section-number">8.8.4</span> Rank-based Analysis</a></li>
<li><a href="gene-set-enrichment-analysis.html#fgsea"><span class="toc-section-number">8.8.5</span> fgsea</a></li>
</ul></li>
<li><a href="biological-networks-..html#biological-networks-."><span class="toc-section-number">8.9</span> Biological Networks .</a>
<ul>
<li><a href="biological-networks-..html#biological-pathways-."><span class="toc-section-number">8.9.1</span> Biological Pathways .</a></li>
<li><a href="biological-networks-..html#gene-regulatory-networks-."><span class="toc-section-number">8.9.2</span> Gene Regulatory Networks .</a></li>
<li><a href="biological-networks-..html#protein-protein-interaction-network-."><span class="toc-section-number">8.9.3</span> Protein-Protein Interaction Network .</a></li>
<li><a href="biological-networks-..html#wgcna-."><span class="toc-section-number">8.9.4</span> WGCNA .</a></li>
</ul></li>
</ul></li>
<li><a href="engineering.html#engineering"><span class="toc-section-number">9</span> EngineeRing</a>
<ul>
<li><a href="unit-testing.html#unit-testing"><span class="toc-section-number">9.1</span> Unit Testing</a></li>
<li><a href="toolification.html#toolification"><span class="toc-section-number">9.2</span> Toolification</a>
<ul>
<li><a href="toolification.html#the-r-interpreter"><span class="toc-section-number">9.2.1</span> The R Interpreter</a></li>
<li><a href="toolification.html#rscript"><span class="toc-section-number">9.2.2</span> <code>Rscript</code></a></li>
<li><a href="toolification.html#commandargs"><span class="toc-section-number">9.2.3</span> <code>commandArgs()</code></a></li>
</ul></li>
<li><a href="parallel-processing.html#parallel-processing"><span class="toc-section-number">9.3</span> Parallel Processing</a>
<ul>
<li><a href="parallel-processing.html#brief-introduction-to-parallelization"><span class="toc-section-number">9.3.1</span> Brief Introduction to Parallelization</a></li>
<li><a href="parallel-processing.html#apply-and-friends-are-pleasingly-parallel"><span class="toc-section-number">9.3.2</span> <code>apply</code> and Friends Are Pleasingly Parallel</a></li>
<li><a href="parallel-processing.html#the-parallel-package"><span class="toc-section-number">9.3.3</span> The <code>parallel</code> package</a></li>
</ul></li>
<li><a href="object-orientation-in-r.html#object-orientation-in-r"><span class="toc-section-number">9.4</span> Object Orientation in R</a></li>
<li><a href="r-packages.html#r-packages"><span class="toc-section-number">9.5</span> R Packages</a></li>
</ul></li>
<li><a href="rshiny.html#rshiny"><span class="toc-section-number">10</span> RShiny</a>
<ul>
<li><a href="rshiny-introduction.html#rshiny-introduction"><span class="toc-section-number">10.1</span> RShiny Introduction</a>
<ul>
<li><a href="rshiny-introduction.html#how-does-a-web-page-work"><span class="toc-section-number">10.1.1</span> How does a web page work?</a></li>
</ul></li>
<li><a href="application-structure.html#application-structure"><span class="toc-section-number">10.2</span> Application Structure</a>
<ul>
<li><a href="application-structure.html#ui-design"><span class="toc-section-number">10.2.1</span> UI Design</a></li>
<li><a href="application-structure.html#server-functionality"><span class="toc-section-number">10.2.2</span> Server Functionality</a></li>
</ul></li>
<li><a href="reactivity.html#reactivity"><span class="toc-section-number">10.3</span> Reactivity</a></li>
<li><a href="publishing.html#publishing"><span class="toc-section-number">10.4</span> Publishing</a></li>
</ul></li>
<li><a href="communicating-with-r.html#communicating-with-r"><span class="toc-section-number">11</span> Communicating with R</a>
<ul>
<li><a href="rmarkdown-knitr.html#rmarkdown-knitr"><span class="toc-section-number">11.1</span> RMarkdown &amp; knitr</a></li>
<li><a href="bookdown.html#bookdown"><span class="toc-section-number">11.2</span> bookdown</a></li>
</ul></li>
<li><a href="contribution-guide.html#contribution-guide"><span class="toc-section-number">12</span> Contribution Guide</a>
<ul>
<li><a href="custom-blocks.html#custom-blocks"><span class="toc-section-number">12.1</span> Custom blocks</a></li>
</ul></li>
<li><a href="assignments.html#assignments">Assignments</a>
<ul>
<li><a href="assignment-format.html#assignment-format">Assignment Format</a></li>
<li><a href="starting-an-assignment.html#starting-an-assignment">Starting an Assignment</a>
<ul>
<li><a href="starting-an-assignment.html#starting-your-own-git-repository-and-cloning-it">Starting your own <code>git</code> repository and cloning it</a></li>
<li><a href="starting-an-assignment.html#contents">Contents</a></li>
<li><a href="starting-an-assignment.html#github-tutorial">GitHub tutorial</a></li>
<li><a href="starting-an-assignment.html#committing-and-pushing-with-r-and-without">Committing and Pushing with R (and without)</a></li>
<li><a href="starting-an-assignment.html#using-rstudio">Using RStudio</a></li>
<li><a href="starting-an-assignment.html#using-the-command-line">Using the command line</a></li>
</ul></li>
<li><a href="assignment-1.html#assignment-1">Assignment 1</a>
<ul>
<li><a href="assignment-1.html#problem-statement">Problem Statement</a></li>
<li><a href="assignment-1.html#learning-objectives-and-skill-list">Learning Objectives and Skill List</a></li>
<li><a href="assignment-1.html#instructions">Instructions</a></li>
<li><a href="assignment-1.html#hints">Hints</a></li>
</ul></li>
<li><a href="assignment-2.html#assignment-2">Assignment 2</a>
<ul>
<li><a href="assignment-2.html#problem-statement-1">Problem Statement</a></li>
<li><a href="assignment-2.html#learning-objectives">Learning Objectives</a></li>
<li><a href="assignment-2.html#skill-list">Skill List</a></li>
<li><a href="assignment-2.html#instructions-1">Instructions</a></li>
<li><a href="assignment-2.html#function-details">Function Details</a></li>
</ul></li>
<li><a href="assignment-3.html#assignment-3">Assignment 3</a>
<ul>
<li><a href="assignment-3.html#problem-statement-2">Problem Statement</a></li>
<li><a href="assignment-3.html#learning-objectives-1">Learning Objectives</a></li>
<li><a href="assignment-3.html#skill-list-1">Skill List</a></li>
<li><a href="assignment-3.html#background-on-microarrays">Background on Microarrays</a></li>
<li><a href="assignment-3.html#background-on-principal-component-analysis">Background on Principal Component Analysis</a></li>
<li><a href="assignment-3.html#marisa-et-al.-gene-expression-classification-of-colon-cancer-into-molecular-subtypes-characterization-validation-and-prognostic-value.-plos-medicine-may-2013.-pmid-23700391">Marisa et al. Gene Expression Classification of Colon Cancer into Molecular Subtypes: Characterization, Validation, and Prognostic Value. PLoS Medicine, May 2013. PMID: 23700391</a></li>
<li><a href="assignment-3.html#scaling-data-using-r-scale">Scaling data using R <code>scale()</code></a></li>
<li><a href="assignment-3.html#proportion-of-variance-explained">Proportion of variance explained</a></li>
<li><a href="assignment-3.html#plotting-and-visualization-of-pca">Plotting and visualization of PCA</a></li>
<li><a href="assignment-3.html#hierarchical-clustering-and-heatmaps">Hierarchical Clustering and Heatmaps</a></li>
<li><a href="assignment-3.html#references">References</a></li>
</ul></li>
<li><a href="assignment-4.html#assignment-4">Assignment 4</a>
<ul>
<li><a href="assignment-4.html#problem-statement-3">Problem Statement</a></li>
<li><a href="assignment-4.html#learning-objectives-2">Learning Objectives</a></li>
<li><a href="assignment-4.html#skill-list-2">Skill List</a></li>
<li><a href="assignment-4.html#instructions-2">Instructions</a></li>
<li><a href="assignment-4.html#deliverables">Deliverables</a></li>
<li><a href="assignment-4.html#function-details-1">Function Details</a></li>
<li><a href="assignment-4.html#hints-1">Hints</a></li>
</ul></li>
<li><a href="assignment-5.html#assignment-5">Assignment 5</a>
<ul>
<li><a href="assignment-5.html#problem-statement-4">Problem Statement</a></li>
<li><a href="assignment-5.html#learning-objectives-3">Learning Objectives</a></li>
<li><a href="assignment-5.html#skill-list-3">Skill List</a></li>
<li><a href="assignment-5.html#deseq2-background">DESeq2 Background</a></li>
<li><a href="assignment-5.html#generating-a-counts-matrix">Generating a counts matrix</a></li>
<li><a href="assignment-5.html#prefiltering-counts-matrix">Prefiltering Counts matrix</a></li>
<li><a href="assignment-5.html#median-of-ratios-normalization">Median-of-ratios normalization</a></li>
<li><a href="assignment-5.html#deseq2-preparation">DESeq2 preparation</a></li>
<li><a href="assignment-5.html#omeara-et-al.-transcriptional-reversion-of-cardiac-myocyte-fate-during-mammalian-cardiac-regeneration.-circ-res.-feb-2015.-pmid-25477501l">O’Meara et al. Transcriptional Reversion of Cardiac Myocyte Fate During Mammalian Cardiac Regeneration. Circ Res. Feb 2015. PMID: 25477501l</a></li>
<li><a href="assignment-5.html#reading-and-subsetting-the-data-from-verse_counts.tsv-and-sample_metadata.csv">1. Reading and subsetting the data from verse_counts.tsv and sample_metadata.csv</a></li>
<li><a href="assignment-5.html#running-deseq2">2. Running DESeq2</a></li>
<li><a href="assignment-5.html#annotating-results-to-construct-a-labeled-volcano-plot">3. Annotating results to construct a labeled volcano plot</a></li>
<li><a href="assignment-5.html#diagnostic-plot-of-the-raw-p-values-for-all-genes">4. Diagnostic plot of the raw p-values for all genes</a></li>
<li><a href="assignment-5.html#plotting-the-logfoldchanges-for-differentially-expressed-genes">5. Plotting the LogFoldChanges for differentially expressed genes</a></li>
<li><a href="assignment-5.html#the-choice-of-fdr-cutoff-depends-on-cost">The choice of FDR cutoff depends on cost</a></li>
<li><a href="assignment-5.html#plotting-the-normalized-counts-of-differentially-expressed-genes">6. Plotting the normalized counts of differentially expressed genes</a></li>
<li><a href="assignment-5.html#volcano-plot-to-visualize-differential-expression-results">7. Volcano Plot to visualize differential expression results</a></li>
<li><a href="assignment-5.html#running-fgsea-vignette">8. Running fgsea <span>vignette</span></a></li>
<li><a href="assignment-5.html#plotting-the-top-ten-positive-nes-and-top-ten-negative-nes-pathways">9. Plotting the top ten positive NES and top ten negative NES pathways</a></li>
<li><a href="assignment-5.html#references-1">References</a></li>
</ul></li>
<li><a href="assignment-6.html#assignment-6">Assignment 6</a>
<ul>
<li><a href="assignment-6.html#problem-statement-5">Problem Statement</a></li>
<li><a href="assignment-6.html#learning-objectives-4">Learning Objectives</a></li>
<li><a href="assignment-6.html#skill-list-4">Skill List</a></li>
<li><a href="assignment-6.html#instructions-3">Instructions</a></li>
<li><a href="assignment-6.html#function-details-2">Function Details</a></li>
</ul></li>
<li><a href="assignment-7.html#assignment-7">Assignment 7</a>
<ul>
<li><a href="assignment-7.html#problem-statement-6">Problem Statement</a></li>
<li><a href="assignment-7.html#learning-objectives-5">Learning Objectives</a></li>
<li><a href="assignment-7.html#skill-list-5">Skill List</a></li>
<li><a href="assignment-7.html#instructions-4">Instructions</a></li>
<li><a href="assignment-7.html#function-details-3">“Function” Details</a></li>
</ul></li>
</ul></li>
<li class="appendix"><span><b>Appendix</b></span></li>
<li><a href="class-outlines.html#class-outlines"><span class="toc-section-number">A</span> Class Outlines</a>
<ul>
<li><a href="week-1.html#week-1"><span class="toc-section-number">A.1</span> Week 1</a></li>
<li><a href="week-2.html#week-2"><span class="toc-section-number">A.2</span> Week 2</a></li>
<li><a href="week-3.html#week-3"><span class="toc-section-number">A.3</span> Week 3</a></li>
<li><a href="week-4.html#week-4"><span class="toc-section-number">A.4</span> Week 4</a></li>
<li><a href="week-5.html#week-5"><span class="toc-section-number">A.5</span> Week 5</a></li>
<li><a href="week-6.html#week-6"><span class="toc-section-number">A.6</span> Week 6</a></li>
<li><a href="week-7.html#week-7"><span class="toc-section-number">A.7</span> Week 7</a></li>
<li><a href="week-8.html#week-8"><span class="toc-section-number">A.8</span> Week 8</a></li>
<li><a href="week-9.html#week-9"><span class="toc-section-number">A.9</span> Week 9</a></li>
<li><a href="week-10.html#week-10"><span class="toc-section-number">A.10</span> Week 10</a></li>
<li><a href="week-11.html#week-11"><span class="toc-section-number">A.11</span> Week 11</a></li>
<li><a href="week-12.html#week-12"><span class="toc-section-number">A.12</span> Week 12</a></li>
<li><a href="week-13.html#week-13"><span class="toc-section-number">A.13</span> Week 13</a></li>
<li><a href="week-14.html#week-14"><span class="toc-section-number">A.14</span> Week 14</a></li>
</ul></li>
</ul>
</div>
</div>
</div>
<div class="row">
<div class="col-sm-12">
<div id="assignment-6" class="section level2 unnumbered">
<h2>Assignment 6</h2>
<blockquote>
<p>Volcanoes don’t really look like that, do they?</p>
</blockquote>
<div id="problem-statement-5" class="section level3 unnumbered">
<h3>Problem Statement</h3>
<p>Because R has been utilized in bioinformatics communities for years at this point, there are many different approaches to do one task. While in the last assignment we just used DESeq2, we will be performing differential expression analysis with two additional packages (edgeR and Limma) and comparing the results we find.</p>
</div>
<div id="learning-objectives-4" class="section level3 unnumbered">
<h3>Learning Objectives</h3>
<ol style="list-style-type: decimal">
<li>The basic operations of <code>ggplot</code> and “The Grammar of Graphics.”</li>
<li>Using differential gene expression analysis packages to generate data for plotting.</li>
<li>Creating simple plots and <strong>not</strong> using the default colors and themes.</li>
<li>Combining multiple plots into one image using <code>facet_wrap()</code></li>
</ol>
</div>
<div id="skill-list-4" class="section level3 unnumbered">
<h3>Skill List</h3>
<ul>
<li>A tempered heart and mind that understands a plot may not ever look <em>exactly</em> how you want it to.</li>
<li>An intermediate understanding of R’s most popular plotting package, <code>ggplot</code>.</li>
<li>Further understanding of differential expression analysis and its plotting.</li>
<li>A sense of superiority whenever you see a publication use the base R plotting package for figures.</li>
</ul>
</div>
<div id="instructions-3" class="section level3 unnumbered">
<h3>Instructions</h3>
<p>Complete <code>main.R</code> in a way that satisfies all of the tests in <code>test_main.R</code>. Follow the instructions in the function descriptions for <code>main.R</code> and read here for details on the tests your code should satisfy. You will use these functions to complete figures in <code>report.Rmd</code>.</p>
<p>You can use <code>testthat:test_file('test_main.R')</code> in your R console to run all tests in <code>test_main.R</code>. Do not modify the tests. While it is difficult to write tests that cover every solution to a problem, our tests are written to ensure the outputs of a function are aligned with what an analysis should perform. If you feel a test is working incorrectly, let a TA know.</p>
</div>
<div id="function-details-2" class="section level3 unnumbered">
<h3>Function Details</h3>
<div id="load_n_trim" class="section level4 unnumbered">
<h4>1. <code>load_n_trim()</code></h4>
<p>Once again we have some data to load into R in order to manipulate. We’ll be loading a counts file, which is a matrix of genes (rows) by samples (columns). Each cell is the number of that gene found in that sample. This is a good opportunity to use a data frame or a tibble, but because of package restrictions we will be sticking to data frames for this assignment. We’re using a couple different packages to process our data, and they input/output <em>exclusively</em> in data frames, so we will be consistent and save ourselves some trouble with conversions.</p>
<p>We want to return a data frame that is essentially identical to our input file (always a good idea to use the command line to see what you’re working with). Ensure the gene names are stored as <code>row.names</code> and <em>not</em> as a separate column, data frames have a distinction between the two. This function should also reduce the columns to just those of interest (P0 and Adult).</p>
<div class="box hint">
<p>Row names can be changed using the <code>row.names()</code> function. The new row names must be a character vector of exactly the same length.</p>
<div class="sourceCode" id="cb528"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb528-1"><a href="assignment-6.html#cb528-1" aria-hidden="true" tabindex="-1"></a>df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">a =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">0</span>, <span class="dv">0</span>), <span class="at">b =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">3</span>, <span class="dv">5</span>))</span>
<span id="cb528-2"><a href="assignment-6.html#cb528-2" aria-hidden="true" tabindex="-1"></a><span class="fu">row.names</span>(df) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;Row1&quot;</span>, <span class="st">&quot;Row2&quot;</span>, <span class="st">&quot;Row3&quot;</span>)</span>
<span id="cb528-3"><a href="assignment-6.html#cb528-3" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(df)</span>
<span id="cb528-4"><a href="assignment-6.html#cb528-4" aria-hidden="true" tabindex="-1"></a>     a b</span>
<span id="cb528-5"><a href="assignment-6.html#cb528-5" aria-hidden="true" tabindex="-1"></a>Row1 <span class="dv">0</span> <span class="dv">1</span></span>
<span id="cb528-6"><a href="assignment-6.html#cb528-6" aria-hidden="true" tabindex="-1"></a>Row2 <span class="dv">0</span> <span class="dv">3</span></span>
<span id="cb528-7"><a href="assignment-6.html#cb528-7" aria-hidden="true" tabindex="-1"></a>Row3 <span class="dv">0</span> <span class="dv">5</span></span></code></pre></div>
</div>
<div id="tests-4" class="section level5 unnumbered">
<h5>Tests</h5>
<div class="sourceCode" id="cb529"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb529-1"><a href="assignment-6.html#cb529-1" aria-hidden="true" tabindex="-1"></a><span class="fu">test_that</span>(<span class="st">&quot;test data is loading correctly and reduced&quot;</span>, {</span>
<span id="cb529-2"><a href="assignment-6.html#cb529-2" aria-hidden="true" tabindex="-1"></a>  test_df <span class="ot">&lt;-</span> <span class="fu">load_n_trim</span>(csv)</span>
<span id="cb529-3"><a href="assignment-6.html#cb529-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">expect_equal</span>(<span class="fu">names</span>(test_df), <span class="fu">c</span>(<span class="st">&quot;vP0_1&quot;</span>, <span class="st">&quot;vP0_2&quot;</span>, <span class="st">&quot;vAd_1&quot;</span>, <span class="st">&quot;vAd_2&quot;</span>))</span>
<span id="cb529-4"><a href="assignment-6.html#cb529-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">expect_equal</span>(<span class="fu">dim</span>(test_df), <span class="fu">c</span>(<span class="dv">55416</span>, <span class="dv">4</span>))</span>
<span id="cb529-5"><a href="assignment-6.html#cb529-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">expect_equal</span>(<span class="fu">class</span>(test_df), <span class="st">&quot;data.frame&quot;</span>)</span>
<span id="cb529-6"><a href="assignment-6.html#cb529-6" aria-hidden="true" tabindex="-1"></a>})</span></code></pre></div>
<p>The tests for this assignment are relatively straightforward, mostly ensuring that data is loaded and ends up in the right shape and format. The first test ensures the columns were filtered, the second ensures the size is correct, and the final ensures it is a data frame and <em>not</em> a tibble. No tibbles. Tibble-less?</p>
</div>
</div>
<div id="run_deseq" class="section level4 unnumbered">
<h4>2. <code>run_deseq()</code></h4>
<p>This is the first of three functions that technically do identical things (but in different ways). One of the most popular Bioconductor packages, DESeq2 has a number of options available for differential expression analysis. We will load in the counts data, select our variables of interest, and use <code>DESeq()</code> to process them. Links are included in the function description to DESeq2 documentation, you will find answers to most questions there, especially since those were the documents used to <em>write</em> this assignment. This function will return the results of the analysis.</p>
<div id="tests-5" class="section level5 unnumbered">
<h5>Tests</h5>
<div class="sourceCode" id="cb530"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb530-1"><a href="assignment-6.html#cb530-1" aria-hidden="true" tabindex="-1"></a><span class="fu">test_that</span>(<span class="st">&quot;deseq2 function is returning correct results&quot;</span>, {</span>
<span id="cb530-2"><a href="assignment-6.html#cb530-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">load</span>(<span class="st">&quot;mock_counts_df.RData&quot;</span>) <span class="co"># loads the counts_df object into env</span></span>
<span id="cb530-3"><a href="assignment-6.html#cb530-3" aria-hidden="true" tabindex="-1"></a>  coldata <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">condition =</span> <span class="fu">rep</span>(<span class="fu">c</span>(<span class="st">&quot;day4&quot;</span>, <span class="st">&quot;day7&quot;</span>), <span class="at">each=</span><span class="dv">2</span>),</span>
<span id="cb530-4"><a href="assignment-6.html#cb530-4" aria-hidden="true" tabindex="-1"></a>                        <span class="at">type=</span><span class="st">&quot;paired-end&quot;</span>)</span>
<span id="cb530-5"><a href="assignment-6.html#cb530-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">row.names</span>(coldata) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;vP4_1&quot;</span>, <span class="st">&quot;vP4_2&quot;</span>, <span class="st">&quot;vP7_1&quot;</span>, <span class="st">&quot;vP7_2&quot;</span>)</span>
<span id="cb530-6"><a href="assignment-6.html#cb530-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">expect_warning</span>(deseq <span class="ot">&lt;-</span> <span class="fu">run_deseq</span>(counts_df,</span>
<span id="cb530-7"><a href="assignment-6.html#cb530-7" aria-hidden="true" tabindex="-1"></a>                                    coldata,</span>
<span id="cb530-8"><a href="assignment-6.html#cb530-8" aria-hidden="true" tabindex="-1"></a>                                    <span class="dv">10</span>, <span class="st">&quot;condition_day7_vs_day4&quot;</span>))</span>
<span id="cb530-9"><a href="assignment-6.html#cb530-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">expect_equal</span>(<span class="fu">dim</span>(deseq), <span class="fu">c</span>(<span class="dv">19127</span>, <span class="dv">6</span>))</span>
<span id="cb530-10"><a href="assignment-6.html#cb530-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">expect_equal</span>(<span class="fu">class</span>(deseq)[<span class="dv">1</span>], <span class="st">&quot;DESeqResults&quot;</span>)</span>
<span id="cb530-11"><a href="assignment-6.html#cb530-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">expect_equal</span>(<span class="fu">c</span>(<span class="st">&quot;pvalue&quot;</span>, <span class="st">&quot;padj&quot;</span>) <span class="sc">%in%</span> <span class="fu">names</span>(deseq), <span class="fu">c</span>(<span class="cn">TRUE</span>, <span class="cn">TRUE</span>))</span>
<span id="cb530-12"><a href="assignment-6.html#cb530-12" aria-hidden="true" tabindex="-1"></a>})</span></code></pre></div>
<p>Less concerning for you all but writing tests for larger data sets and more complicated functions like this is…tricky. The problem is it can’t be written like a script, we can’t use the results from <code>load_n_trim()</code> to test <em>this</em> function, because what if one fails independently of the other? So in this case, we create an RData object of correctly loaded data that works <em>even if</em> <code>load_n_trim()</code> is absolutely broken. Also, in this case, we are using different sets to test this (days 4 and 7 instead of 0 and adult) so you can’t just load the test data and say you wrote the code! That would be disingenuous of you, and being disingenuous is for after you graduate.</p>
<p>This test loads the sample data (a data frame of counts) and creates the <code>coldata</code> object according to DESeq2 specifications. We wrap the <code>run_deseq()</code> in <code>expect_warning()</code> because DESeq throws a warning when it performs an expected behavior: converting strings to factors (this is bad programming! this is not what warnings are for, the package is functioning as expected). We then do a similar series of tests: ensure the dimensions are correct, ensure the results object are from DESeq, and ensure that the column names we need are present.</p>
</div>
</div>
<div id="run_edger" class="section level4 unnumbered">
<h4>3. <code>run_edger()</code></h4>
<p>As mentioned in the previous function, this is the EdgeR implementation of running the differential expression analysis based on our counts file. Again, this function should follow roughly the portion of the documentation mentioned in the function description. EdgeR has a number of additional plotting functions (such as <code>plotMDS()</code>) you may like to include. These are optional but pretty fun to see if your data is working out as expected (similar days clustered close together and so on). You will return the <em>entire</em> results object, no filtering necessary yet.</p>
<div id="tests-6" class="section level5 unnumbered">
<h5>Tests</h5>
<div class="sourceCode" id="cb531"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb531-1"><a href="assignment-6.html#cb531-1" aria-hidden="true" tabindex="-1"></a><span class="fu">test_that</span>(<span class="st">&quot;edger function is returning correct results&quot;</span>, {</span>
<span id="cb531-2"><a href="assignment-6.html#cb531-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">load</span>(<span class="st">&quot;mock_counts_df.RData&quot;</span>)</span>
<span id="cb531-3"><a href="assignment-6.html#cb531-3" aria-hidden="true" tabindex="-1"></a>  group <span class="ot">&lt;-</span> <span class="fu">factor</span>(<span class="fu">rep</span>(<span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>), <span class="at">each=</span><span class="dv">2</span>))</span>
<span id="cb531-4"><a href="assignment-6.html#cb531-4" aria-hidden="true" tabindex="-1"></a>  edger_res <span class="ot">&lt;-</span> <span class="fu">run_edger</span>(counts_df, group)</span>
<span id="cb531-5"><a href="assignment-6.html#cb531-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">expect_equal</span>(<span class="fu">names</span>(edger_res), <span class="fu">c</span>(<span class="st">&quot;logFC&quot;</span>, <span class="st">&quot;logCPM&quot;</span>, <span class="st">&quot;PValue&quot;</span>))</span>
<span id="cb531-6"><a href="assignment-6.html#cb531-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">expect_equal</span>(<span class="fu">dim</span>(edger_res), <span class="fu">c</span>(<span class="dv">15026</span>, <span class="dv">3</span>))</span>
<span id="cb531-7"><a href="assignment-6.html#cb531-7" aria-hidden="true" tabindex="-1"></a>})</span></code></pre></div>
<p>Similar to the DESeq2 tests, we aren’t necessarily looking for a carbon copy of the results. While the process <em>should</em> be entirely deterministic, it is better to allow some wiggle room and merely look that the columns we’re interested in are present and that the data has the right shape. Note that EdgeR doesn’t perform a BH correction (the p-adjusted column is non-existant). It’s either an option that’s off by default or just not a feature! Not a problem, we have the p-values so we can do it ourselves (see the markdown).</p>
</div>
</div>
<div id="run_limma-and-run_voom" class="section level4 unnumbered">
<h4>4. <code>run_limma()</code> and <code>run_voom()</code></h4>
<p>Our final in this set of three, we’re testing two related functions in the Limma package. While you can create an analysis that <em>only</em> uses Limma, in our case we can apply the <strong>voom</strong> workflow to refine our data a little more. More info on how these work in the cited documentation. A little technique that can be useful for optional workflows like this is including a boolean flag in the parameters of the function. If I am running a function <code>add_numbers(1, -4, abs=TRUE)</code>, the parameter <code>abs</code> might change the functions operation so that it takes the absolute value of numbers before adding them together. I could create this behavior like so:</p>
<div class="sourceCode" id="cb532"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb532-1"><a href="assignment-6.html#cb532-1" aria-hidden="true" tabindex="-1"></a>add_numbers <span class="ot">&lt;-</span> <span class="cf">function</span>(x, y, <span class="at">abs=</span><span class="cn">FALSE</span>) {</span>
<span id="cb532-2"><a href="assignment-6.html#cb532-2" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (abs) { <span class="co"># code enters here if abs is TRUE</span></span>
<span id="cb532-3"><a href="assignment-6.html#cb532-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">return</span>(<span class="fu">abs</span>(x) <span class="sc">+</span> <span class="fu">abs</span>(y)) <span class="co"># return means function exits, doesn&#39;t go to &quot;else {}&quot;</span></span>
<span id="cb532-4"><a href="assignment-6.html#cb532-4" aria-hidden="true" tabindex="-1"></a>  } <span class="cf">else</span> { <span class="co"># code enters here if abs was false</span></span>
<span id="cb532-5"><a href="assignment-6.html#cb532-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">return</span>(x <span class="sc">+</span> y)</span>
<span id="cb532-6"><a href="assignment-6.html#cb532-6" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb532-7"><a href="assignment-6.html#cb532-7" aria-hidden="true" tabindex="-1"></a>}</span></code></pre></div>
<p>This is a very useful technique for re-using functions, meaning I don’t need to write two functions that do similar things when I can change their behavior with that boolean parameter.</p>
<div id="tests-7" class="section level5 unnumbered">
<h5>Tests</h5>
<div class="sourceCode" id="cb533"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb533-1"><a href="assignment-6.html#cb533-1" aria-hidden="true" tabindex="-1"></a><span class="fu">test_that</span>(<span class="st">&quot;test limma + voom work + work together&quot;</span>, {</span>
<span id="cb533-2"><a href="assignment-6.html#cb533-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">load</span>(<span class="st">&quot;data/mock_counts_df.RData&quot;</span>)</span>
<span id="cb533-3"><a href="assignment-6.html#cb533-3" aria-hidden="true" tabindex="-1"></a>  group <span class="ot">&lt;-</span> <span class="fu">factor</span>(<span class="fu">rep</span>(<span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>), <span class="at">each=</span><span class="dv">2</span>))</span>
<span id="cb533-4"><a href="assignment-6.html#cb533-4" aria-hidden="true" tabindex="-1"></a>  <span class="co"># design</span></span>
<span id="cb533-5"><a href="assignment-6.html#cb533-5" aria-hidden="true" tabindex="-1"></a>  design <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">day4=</span><span class="dv">1</span>, <span class="at">day4vsday7=</span><span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">0</span>, <span class="dv">1</span>, <span class="dv">1</span>))</span>
<span id="cb533-6"><a href="assignment-6.html#cb533-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">row.names</span>(design) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;vP4_1&quot;</span>, <span class="st">&quot;vP4_2&quot;</span>, <span class="st">&quot;vP7_1&quot;</span>, <span class="st">&quot;vP7_2&quot;</span>)</span>
<span id="cb533-7"><a href="assignment-6.html#cb533-7" aria-hidden="true" tabindex="-1"></a>  <span class="co"># limma + voom</span></span>
<span id="cb533-8"><a href="assignment-6.html#cb533-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">expect_warning</span>(voom_res <span class="ot">&lt;-</span> <span class="fu">run_limma</span>(counts_df, design, group)) <span class="co"># voom yelling, ignore</span></span>
<span id="cb533-9"><a href="assignment-6.html#cb533-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">expect_equal</span>(<span class="fu">dim</span>(voom_res), <span class="fu">c</span>(<span class="dv">15026</span>, <span class="dv">6</span>))</span>
<span id="cb533-10"><a href="assignment-6.html#cb533-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">expect_equal</span>(<span class="fu">names</span>(voom_res), <span class="fu">c</span>(<span class="st">&quot;logFC&quot;</span>, <span class="st">&quot;AveExpr&quot;</span>, <span class="st">&quot;t&quot;</span>, <span class="st">&quot;P.Value&quot;</span>, </span>
<span id="cb533-11"><a href="assignment-6.html#cb533-11" aria-hidden="true" tabindex="-1"></a>                                  <span class="st">&quot;adj.P.Val&quot;</span>, <span class="st">&quot;B&quot;</span>))</span>
<span id="cb533-12"><a href="assignment-6.html#cb533-12" aria-hidden="true" tabindex="-1"></a>})</span></code></pre></div>
<p>This ultimately functions similarly to the first two series of tests: load test data, apply the function to it, make sure the results are the right size, shape, and type. This includes the voom procedure, so using <em>only</em> limma results will yield failures.</p>
<div class="box note">
<p>Note that Limma uses EdgeR to input data, so your first couple steps will look pretty familiar.</p>
</div>
<p>We create a design object to use, defining the experimental conditions. These are again using the day 4 and day 7 trials from the heart data.</p>
</div>
</div>
<div id="combine_pval" class="section level4 unnumbered">
<h4>5. <code>combine_pval()</code></h4>
<p>The final three functions concern modifying our differential expression results and creating plots of these combined results. While it would be trivial to make three plots using the same code or function, have R generate one consistent plot after changes in your data or method is extremely useful.</p>
<p>We use <code>combine_pval()</code> to coerce our data into long format, which is more applicable for facet wrapping. <code>facet_wrap()</code> is a <code>ggplot2</code> function that uses a property of the data to create separate plots in one image. In our case we will use the <code>package</code> property (DESeq2, edgeR, or Limma) to differentiate our data. There are a couple ways to reorganize data in this way, but I would suggest using <code>tidyr::gather()</code>. Our <code>key</code> will be the package and the <code>value</code> will be the p-value.</p>
<div id="tests-8" class="section level5 unnumbered">
<h5>Tests</h5>
<p>These tests are organized into one larger testing function because they use the same, randomly generated input data (a bunch of data frames with uniform distributions but with ranges that approximate actual data). If you would like to test only parts of this testing function, use an octothorpe (<code>#</code>) to comment out tests you haven’t reached yet.</p>
<div class="sourceCode" id="cb534"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb534-1"><a href="assignment-6.html#cb534-1" aria-hidden="true" tabindex="-1"></a><span class="fu">test_that</span>(<span class="st">&quot;ggplot and data formatting works&quot;</span>, {</span>
<span id="cb534-2"><a href="assignment-6.html#cb534-2" aria-hidden="true" tabindex="-1"></a>  deseq <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">pvalue =</span> <span class="fu">runif</span>(<span class="dv">1000</span>, <span class="fl">1e-100</span>, <span class="fl">1e-5</span>),</span>
<span id="cb534-3"><a href="assignment-6.html#cb534-3" aria-hidden="true" tabindex="-1"></a>                      <span class="at">log2FoldChange =</span> <span class="fu">runif</span>(<span class="dv">1000</span>, <span class="sc">-</span><span class="dv">9</span>, <span class="dv">9</span>),</span>
<span id="cb534-4"><a href="assignment-6.html#cb534-4" aria-hidden="true" tabindex="-1"></a>                      <span class="at">padj =</span> <span class="fu">runif</span>(<span class="dv">1000</span>, <span class="fl">1e-300</span>, <span class="fl">1e-15</span>))</span>
<span id="cb534-5"><a href="assignment-6.html#cb534-5" aria-hidden="true" tabindex="-1"></a>  edger <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">PValue =</span> <span class="fu">runif</span>(<span class="dv">1000</span>, <span class="fl">1e-100</span>, <span class="fl">1e-5</span>),</span>
<span id="cb534-6"><a href="assignment-6.html#cb534-6" aria-hidden="true" tabindex="-1"></a>                      <span class="at">logFC =</span> <span class="fu">runif</span>(<span class="dv">1000</span>, <span class="sc">-</span><span class="dv">9</span>, <span class="dv">9</span>),</span>
<span id="cb534-7"><a href="assignment-6.html#cb534-7" aria-hidden="true" tabindex="-1"></a>                      <span class="at">padj =</span> <span class="fu">runif</span>(<span class="dv">1000</span>, <span class="fl">1e-100</span>, <span class="fl">1e-14</span>))</span>
<span id="cb534-8"><a href="assignment-6.html#cb534-8" aria-hidden="true" tabindex="-1"></a>  limma <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">P.Value =</span> <span class="fu">runif</span>(<span class="dv">1000</span>, <span class="fl">1e-100</span>, <span class="fl">1e-5</span>),</span>
<span id="cb534-9"><a href="assignment-6.html#cb534-9" aria-hidden="true" tabindex="-1"></a>                      <span class="at">logFC =</span> <span class="fu">runif</span>(<span class="dv">1000</span>, <span class="sc">-</span><span class="dv">9</span>, <span class="dv">9</span>),</span>
<span id="cb534-10"><a href="assignment-6.html#cb534-10" aria-hidden="true" tabindex="-1"></a>                      <span class="at">adj.P.Val =</span> <span class="fu">runif</span>(<span class="dv">1000</span>, <span class="fl">1e-6</span>, <span class="fl">1e-4</span>))</span>
<span id="cb534-11"><a href="assignment-6.html#cb534-11" aria-hidden="true" tabindex="-1"></a>  <span class="co"># combine_pval</span></span>
<span id="cb534-12"><a href="assignment-6.html#cb534-12" aria-hidden="true" tabindex="-1"></a>  res <span class="ot">&lt;-</span> <span class="fu">combine_pval</span>(deseq, edger, limma)</span>
<span id="cb534-13"><a href="assignment-6.html#cb534-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">expect_equal</span>(<span class="fu">dim</span>(res), <span class="fu">c</span>(<span class="dv">3000</span>, <span class="dv">2</span>))</span>
<span id="cb534-14"><a href="assignment-6.html#cb534-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">expect_true</span>(<span class="fu">mean</span>(res<span class="sc">$</span>pval) <span class="sc">&lt;</span> <span class="fl">5e-5</span> <span class="sc">&amp;&amp;</span> <span class="fu">mean</span>(res<span class="sc">$</span>pval) <span class="sc">&gt;</span> <span class="fl">5e-7</span>)</span>
<span id="cb534-15"><a href="assignment-6.html#cb534-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">expect_equal</span>(<span class="fu">table</span>(res<span class="sc">$</span>package)[[<span class="dv">1</span>]], <span class="dv">1000</span>)</span>
<span id="cb534-16"><a href="assignment-6.html#cb534-16" aria-hidden="true" tabindex="-1"></a>  ...</span></code></pre></div>
<p>The first nine lines establish the test data frames. While random, they will let us test if functions are working correctly.</p>
<div class="box important">
<p>The test data frames do not mimic the normal results exactly. The rows we are interested in have the same name, but have different numbers of columns. For that reason, I suggest your function uses <code>$</code>-notation when referencing columns, instead of position. <code>deseq$pvalue</code> will work more better than <code>deseq[,1]</code> for this test data.</p>
</div>
<p>We test that we only have 2 columns and 3,000 rows, and that the fake p-values are within the expected range. Finally, we use the <code>table()</code> function to test that there is only 1,000 of any one package in this data frame.</p>
</div>
</div>
<div id="create_facets" class="section level4 unnumbered">
<h4>6. <code>create_facets()</code></h4>
<p>Now that we have examined the p-values, we can look at the adjusted p-values. We’ll be constructing volcano plots, which compare the log<sub>2</sub> fold-change and the adjusted p-value, so the <code>create_facets()</code> function will be somewhat more complicated. Rather than use <code>gather()</code> again, it may be easier to create three tibbles or data frames of the variables of interest and combine them with <code>rbind()</code>.</p>
<div id="tests-9" class="section level5 unnumbered">
<h5>Tests</h5>
<div class="sourceCode" id="cb535"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb535-1"><a href="assignment-6.html#cb535-1" aria-hidden="true" tabindex="-1"></a>  ...</span>
<span id="cb535-2"><a href="assignment-6.html#cb535-2" aria-hidden="true" tabindex="-1"></a>  <span class="co"># create_facets</span></span>
<span id="cb535-3"><a href="assignment-6.html#cb535-3" aria-hidden="true" tabindex="-1"></a>  facets <span class="ot">&lt;-</span> <span class="fu">create_facets</span>(deseq, edger, limma)</span>
<span id="cb535-4"><a href="assignment-6.html#cb535-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">expect_equal</span>(<span class="fu">dim</span>(facets), <span class="fu">c</span>(<span class="dv">3000</span>, <span class="dv">3</span>))</span>
<span id="cb535-5"><a href="assignment-6.html#cb535-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">expect_equal</span>(<span class="fu">table</span>(facets<span class="sc">$</span>package)[[<span class="dv">2</span>]], <span class="dv">1000</span>)</span>
<span id="cb535-6"><a href="assignment-6.html#cb535-6" aria-hidden="true" tabindex="-1"></a>  ...</span></code></pre></div>
<p>Another simple set of tests, this time we want to ensure there are 3,000 rows and <em>3</em> columns. We also want to ensure there aren’t more than 1,000 rows of any package type.</p>
</div>
</div>
<div id="theme_plot" class="section level4 unnumbered">
<h4>7. <code>theme_plot()</code></h4>
<p>Finally, we have all of our data in the right shape and we can do some plotting. There is a lot to learn about ggplot, and I mention a few resources in the function description, but don’t be afraid to experiment! Try different colors, themes, and styles. The easiest part is getting your data on the plot, the hard part is making it look attractive.</p>
<div id="tests-10" class="section level5 unnumbered">
<h5>Tests</h5>
<div class="sourceCode" id="cb536"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb536-1"><a href="assignment-6.html#cb536-1" aria-hidden="true" tabindex="-1"></a>  ...</span>
<span id="cb536-2"><a href="assignment-6.html#cb536-2" aria-hidden="true" tabindex="-1"></a>  <span class="co"># theme_plot</span></span>
<span id="cb536-3"><a href="assignment-6.html#cb536-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tibble</span>(<span class="at">logFC =</span> <span class="fu">runif</span>(<span class="dv">3000</span>, <span class="sc">-</span><span class="dv">9</span>, <span class="dv">9</span>),</span>
<span id="cb536-4"><a href="assignment-6.html#cb536-4" aria-hidden="true" tabindex="-1"></a>         <span class="at">padj =</span> <span class="fu">runif</span>(<span class="dv">3000</span>, <span class="fl">1e-300</span>, <span class="fl">1e-5</span>),</span>
<span id="cb536-5"><a href="assignment-6.html#cb536-5" aria-hidden="true" tabindex="-1"></a>         <span class="at">package =</span> <span class="fu">rep</span>(<span class="fu">c</span>(<span class="st">&quot;DESeq2&quot;</span>, <span class="st">&quot;edgeR&quot;</span>, <span class="st">&quot;Limma&quot;</span>), <span class="at">each=</span><span class="dv">1000</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb536-6"><a href="assignment-6.html#cb536-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_plot</span>() <span class="ot">-&gt;</span> volc_plot</span>
<span id="cb536-7"><a href="assignment-6.html#cb536-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">expect_equal</span>(<span class="fu">dim</span>(volc_plot<span class="sc">$</span>data), <span class="fu">c</span>(<span class="dv">3000</span>, <span class="dv">3</span>))</span>
<span id="cb536-8"><a href="assignment-6.html#cb536-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">expect_equal</span>(<span class="fu">class</span>(volc_plot<span class="sc">$</span>layers[[<span class="dv">1</span>]]<span class="sc">$</span>geom)[<span class="dv">1</span>], <span class="st">&quot;GeomPoint&quot;</span>)</span>
<span id="cb536-9"><a href="assignment-6.html#cb536-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">expect_equal</span>(<span class="fu">class</span>(volc_plot<span class="sc">$</span>facet<span class="sc">$</span>params), <span class="st">&quot;list&quot;</span>)</span>
<span id="cb536-10"><a href="assignment-6.html#cb536-10" aria-hidden="true" tabindex="-1"></a><span class="er">})</span></span></code></pre></div>
<p>We create a little more fake data, and then pipe it into our <code>theme_plot()</code> function, and save that ggplot object. Our tests simply look inside the ggplot object to confirm the data is the right shape, that you’re plotting using <code>geom_point()</code>, and that facets are present.</p>
<p>Once your tests are passing, take a look at the report and see how everything can work together!</p>

</div>
</div>
</div>
</div>
<p style="text-align: center;">
<a href="assignment-5.html"><button class="btn btn-default">Previous</button></a>
<a href="assignment-7.html"><button class="btn btn-default">Next</button></a>
</p>
</div>
</div>


</div>

<script>

// add bootstrap table styles to pandoc tables
$(document).ready(function () {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
});

</script>

</body>
</html>
