<!DOCTYPE html>
<html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Assignment 2 | BF591 - R for Biological Sciences</title>
<meta name="generator" content="bookdown 0.32 with bs4_book()">
<meta property="og:title" content="Assignment 2 | BF591 - R for Biological Sciences">
<meta property="og:type" content="book">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="Assignment 2 | BF591 - R for Biological Sciences">
<!-- JS --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://kit.fontawesome.com/6ecbd6c532.js" crossorigin="anonymous"></script><script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="libs/bootstrap-4.6.0/bootstrap.min.css" rel="stylesheet">
<script src="libs/bootstrap-4.6.0/bootstrap.bundle.min.js"></script><script src="libs/bs3compat-0.4.2/transition.js"></script><script src="libs/bs3compat-0.4.2/tabs.js"></script><script src="libs/bs3compat-0.4.2/bs3compat.js"></script><link href="libs/bs4_book-1.0.0/bs4_book.css" rel="stylesheet">
<script src="libs/bs4_book-1.0.0/bs4_book.js"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- CSS --><style type="text/css">
    
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  </style>
<style type="text/css">
    /* Used with Pandoc 2.11+ new --citeproc when CSL is used */
    div.csl-bib-body { }
    div.csl-entry {
      clear: both;
        }
    .hanging div.csl-entry {
      margin-left:2em;
      text-indent:-2em;
    }
    div.csl-left-margin {
      min-width:2em;
      float:left;
    }
    div.csl-right-inline {
      margin-left:2em;
      padding-left:1em;
    }
    div.csl-indent {
      margin-left: 2em;
    }
  </style>
<link rel="stylesheet" href="style.css">
<meta name="description" content="Problem Statement Arranging the structure of our data inputs is vital to using R, as is creating the correct environment for analysis by installing packages. It is also important to share data and...">
<meta property="og:description" content="Problem Statement Arranging the structure of our data inputs is vital to using R, as is creating the correct environment for analysis by installing packages. It is also important to share data and...">
<meta name="twitter:description" content="Problem Statement Arranging the structure of our data inputs is vital to using R, as is creating the correct environment for analysis by installing packages. It is also important to share data and...">
</head>
<body data-spy="scroll" data-target="#toc">

<div class="container-fluid">
<div class="row">
  <header class="col-sm-12 col-lg-3 sidebar sidebar-book"><a class="sr-only sr-only-focusable" href="#content">Skip to main content</a>

    <div class="d-flex align-items-start justify-content-between">
      <h1>
        <a href="index.html" title="">BF591 - R for Biological Sciences</a>
      </h1>
      <button class="btn btn-outline-primary d-lg-none ml-2 mt-1" type="button" data-toggle="collapse" data-target="#main-nav" aria-expanded="true" aria-controls="main-nav"><i class="fas fa-bars"></i><span class="sr-only">Show table of contents</span></button>
    </div>

    <div id="main-nav" class="collapse-lg">
      <form role="search">
        <input id="search" class="form-control" type="search" placeholder="Search" aria-label="Search">
</form>

      <nav aria-label="Table of contents"><h2>Table of contents</h2>
        <ul class="book-toc list-unstyled">
<li><a class="" href="index.html">R for Biological Sciences</a></li>
<li><a class="" href="introduction.html"><span class="header-section-number">1</span> Introduction</a></li>
<li><a class="" href="data-bio.html"><span class="header-section-number">2</span> Data in Biology</a></li>
<li><a class="" href="preliminaries.html"><span class="header-section-number">3</span> Preliminaries</a></li>
<li><a class="" href="prog-basics.html"><span class="header-section-number">4</span> R Programming</a></li>
<li><a class="" href="data-wrangling.html"><span class="header-section-number">5</span> Data Wrangling</a></li>
<li><a class="" href="data-science.html"><span class="header-section-number">6</span> Data Science</a></li>
<li><a class="" href="data-visualization.html"><span class="header-section-number">7</span> Data Visualization</a></li>
<li><a class="" href="biology-bioinformatics.html"><span class="header-section-number">8</span> Biology &amp; Bioinformatics</a></li>
<li><a class="" href="engineering.html"><span class="header-section-number">9</span> EngineeRing</a></li>
<li><a class="" href="rshiny.html"><span class="header-section-number">10</span> RShiny</a></li>
<li><a class="" href="communicating-with-r.html"><span class="header-section-number">11</span> Communicating with R</a></li>
<li><a class="" href="contribution-guide.html"><span class="header-section-number">12</span> Contribution Guide</a></li>
<li class="book-part">Assignments</li>
<li><a class="" href="starting-an-assignment.html">Starting an Assignment</a></li>
<li><a class="" href="assignment-overview.html">Assignment Overview</a></li>
<li><a class="" href="assignment-0.html">Assignment 0</a></li>
<li><a class="" href="assignment-1.html">Assignment 1</a></li>
<li><a class="active" href="assignment-2.html">Assignment 2</a></li>
<li><a class="" href="assignment-3.html">Assignment 3</a></li>
<li><a class="" href="assignment-4.html">Assignment 4</a></li>
<li><a class="" href="assignment-5.html">Assignment 5</a></li>
<li><a class="" href="assignment-6.html">Assignment 6</a></li>
<li><a class="" href="assignment-7.html">Assignment 7</a></li>
<li><a class="" href="final-project.html">Final Project</a></li>
<li class="book-part">Appendix</li>
<li><a class="" href="detailed-class-schedule.html"><span class="header-section-number">A</span> Detailed Class Schedule</a></li>
</ul>

        <div class="book-extra">
          
        </div>
      </nav>
</div>
  </header><main class="col-sm-12 col-md-9 col-lg-7" id="content"><div id="assignment-2" class="section level1 unnumbered">
<h1>Assignment 2<a class="anchor" aria-label="anchor" href="#assignment-2"><i class="fas fa-link"></i></a>
</h1>
<div id="problem-statement-2" class="section level2 unnumbered">
<h2>Problem Statement<a class="anchor" aria-label="anchor" href="#problem-statement-2"><i class="fas fa-link"></i></a>
</h2>
<p>Arranging the structure of our data inputs is vital to using R, as is creating
the correct environment for analysis by installing packages. It is also
important to share data and results easily using R markdown.</p>
</div>
<div id="learning-objectives" class="section level2 unnumbered">
<h2>Learning Objectives<a class="anchor" aria-label="anchor" href="#learning-objectives"><i class="fas fa-link"></i></a>
</h2>
<ul>
<li>Install various packages needed for analysis</li>
<li>Load data, filter that data, and retrieve HGNC ids for the data</li>
<li>Create a small plot to display results</li>
<li>Utilize R Markdown to create an attractive format for sharing data.</li>
</ul>
</div>
<div id="skill-list" class="section level2 unnumbered">
<h2>Skill List<a class="anchor" aria-label="anchor" href="#skill-list"><i class="fas fa-link"></i></a>
</h2>
<ul>
<li>How to utilize R markdown to create a report of data analysis</li>
<li>Installing and loading packages in R</li>
<li>Utilizing Bioconductor to equate affy ids to HGNC ids (gene names)</li>
</ul>
</div>
<div id="instructions-2" class="section level2 unnumbered">
<h2>Instructions<a class="anchor" aria-label="anchor" href="#instructions-2"><i class="fas fa-link"></i></a>
</h2>
<p>Our main focus for this assignment are installing packages, manipulating data,
and plotting our manipulated data. We will be borrowing from the same CSV of
expression data from BF528’s Project 1, available
<a href="https://bu-bioinfo.github.io/bf528/project-1-microarray-based-tumor-classification.html#project-1-microarray-based-tumor-classification">here</a>.<br>
An empty project repo can be found here:</p>
<p><a href="https://github.com/bu-bioinfo/bf591-assignment-2">https://github.com/bu-bioinfo/bf591-assignment-2</a></p>
<p>The project is laid out as such:</p>
<pre><code>main.R
test_main.R
report.Rmd</code></pre>
<p>A skeleton of the functions you need to complete is in <code>main.R</code>. Tests have been
pre-written to test your code and help you ensure it is running correctly, these
are in <code>test_main.R</code>. Finally, we are also introducing the concept of R
Markdown, which for this assignment is <code>report.Rmd</code>. The document itself goes
into greater detail, but you will:</p>
<ol style="list-style-type: decimal">
<li>Complete the functions in <code>main.R</code> and use
<code>testthat:test_file('test_main.R')</code> to ensure they work correctly.<br>
</li>
<li>Read the R Markdown file and complete the section called <strong>“Assignment”</strong>. To
do this, you can <code>source('main.R')</code> to bring over the functions you wrote in
step one.<br>
</li>
<li>Finally, annotate the functions you wrote and <strong>Knit</strong> the R Markdown report,
complete with your additional comments and code execution.</li>
</ol>
<p>This page will go into detail on how the functions and their associated tests
should work.</p>
</div>
<div id="function-details" class="section level2 unnumbered">
<h2>Function Details<a class="anchor" aria-label="anchor" href="#function-details"><i class="fas fa-link"></i></a>
</h2>
<div id="bioconductor" class="section level3 unnumbered">
<h3>1. Bioconductor<a class="anchor" aria-label="anchor" href="#bioconductor"><i class="fas fa-link"></i></a>
</h3>
<p>While many useful R packages can be loaded through
<a href="https://cran.r-project.org/">CRAN</a> using the <code><a href="https://rdrr.io/r/utils/install.packages.html">install.packages()</a></code> syntax, a lot
of specifically bioinformatics packages are exclusively released on
<a href="https://bioconductor.org/install/">Bioconductor</a>. For this assignment we only
need the package called
<a href="https://bioconductor.org/packages/release/bioc/html/biomaRt.html">bioma<strong>R</strong>t</a>.
R programmers fancy themselves very clever, so <strong>R</strong>s show up a lot.</p>
<p>Naturally, we want to load our packages at the beginning of our script so all of
the code we write beneath can access it as it runs. However, if a user or
ourselves <em>already</em> has this package installed we don’t want to waste their time
installing it again. The function <code><a href="https://rdrr.io/r/base/library.html">require()</a></code> can help us avoid unnecessary
installation time and will help us develop faster. The Bioconductor link above
has an example of this method.</p>
<p>This section is untested.</p>
</div>
<div id="load_expression" class="section level3 unnumbered">
<h3>2. <code>load_expression()</code><a class="anchor" aria-label="anchor" href="#load_expression"><i class="fas fa-link"></i></a>
</h3>
<p>Perhaps the most integral part of using the many data wrangling abilities of R
is actually entering your data into the R environment. While there are many ways
to do this in R, we ultimately want this data to be in a <strong>tibble</strong>, which means
the current form of the CSV will make R very angry if you attempt to load it in.
This is because tibbles don’t support row names very well, and the first column
of our data doesn’t have a name. Try to use the <code>load_expression()</code> data to load
data from a <code>filename</code> parameter and <strong>return</strong> a tibble of that information. I
called my firs column “probeids”.</p>
<div id="tests" class="section level4 unnumbered">
<h4>Tests<a class="anchor" aria-label="anchor" href="#tests"><i class="fas fa-link"></i></a>
</h4>
<p>The tests for this function are:</p>
<div class="sourceCode" id="cb518"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">test_that</span><span class="op">(</span><span class="st">"loading csv correctly"</span>, <span class="op">{</span></span>
<span>  <span class="va">result_tib</span> <span class="op">&lt;-</span> <span class="fu">load_expression</span><span class="op">(</span><span class="st">"/project/bf528/project_1/data/example_intensity_data.csv"</span><span class="op">)</span></span>
<span>  <span class="fu">expect_equal</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">result_tib</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">54675</span>, <span class="fl">36</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="fu">expect_true</span><span class="op">(</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/is_tibble.html">is_tibble</a></span><span class="op">(</span><span class="va">result_tib</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span></code></pre></div>
<p>This test uses the <code>load_expression()</code> you write to store the returned tibble in <code>result_tib</code>.</p>
<div class="box important">
<p>While this test is using the same file for data input as you are, this may not always be the case.</p>
</div>
<p>The test then compares the dimensions of that result, and expects 54,675 rows
and 36 columns. These are the dimensions of the input CSV. It also checks to
confirm it is a tibble object (because tibbles are better than dataframes).</p>
</div>
</div>
<div id="filter_15" class="section level3 unnumbered">
<h3>3. <code>filter_15()</code><a class="anchor" aria-label="anchor" href="#filter_15"><i class="fas fa-link"></i></a>
</h3>
<p>In order to filter the numerous rows we have for this data, we introduce a
function that filters the probe IDs in the tibble our data is stored in. We want
to capture probes that have a suitably high level of expression, so we are
setting <code>log2(15)</code> as the cutoff for an expression level. We will keep a row if
15% of the values in that row exceed <code>log2(15)</code> (about 3.9). Since we may want
to examine the probe IDs we find, the function simple returns the values of the
probe IDs (column 1) instead of returning the entire tibble.</p>
<p>This function presents an important concept in R: using built-ins to speed up
our code. Built-ins are functions and packages that are optimized to process
data in a certain way. Since we’re looking at each row of a table, we could
simply use a for loop to iterate one row at a time. This is slow, though, and
for this function might take 5-10 seconds to run (a long time for a program like
this!). Instead, you could use a function like <code><a href="https://rdrr.io/r/base/apply.html">apply()</a></code> or <code><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html">lapply()</a></code> to filter
every row at once. This solution takes mere moments.</p>
<div id="tests-1" class="section level4 unnumbered">
<h4>Tests<a class="anchor" aria-label="anchor" href="#tests-1"><i class="fas fa-link"></i></a>
</h4>
<div class="sourceCode" id="cb519"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://tibble.tidyverse.org/">tibble</a></span><span class="op">)</span></span>
<span><span class="fu">test_that</span><span class="op">(</span><span class="st">"the correct rows are filtered out in filter_15()"</span>, <span class="op">{</span></span>
<span>  <span class="va">test_tib</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html">tibble</a></span><span class="op">(</span>probeids<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'1_s_at'</span>, <span class="st">'2_s_at'</span>, <span class="st">'3_s_at'</span>, <span class="st">'4_s_at'</span><span class="op">)</span>,</span>
<span>                     GSM1<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1.0</span>, <span class="fl">3.95</span>, <span class="fl">4.05</span>, <span class="fl">0.5</span><span class="op">)</span>,</span>
<span>                     GSM2<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="fl">1.6</span>, <span class="fl">4</span><span class="op">)</span>,</span>
<span>                     GSM3<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="fl">2.5</span>, <span class="fl">4</span><span class="op">)</span>,</span>
<span>                     GSM4<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="fl">3.99</span>, <span class="fl">4</span><span class="op">)</span>,</span>
<span>                     GSM5<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="fl">3.0</span>, <span class="fl">4</span><span class="op">)</span>,</span>
<span>                     GSM6<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="fl">1.0</span>, <span class="fl">4</span><span class="op">)</span>,</span>
<span>                     GSM7<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="fl">0.5</span>, <span class="fl">4</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="fu">expect_equal</span><span class="op">(</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html">pull</a></span><span class="op">(</span><span class="fu">filter_15</span><span class="op">(</span><span class="va">test_tib</span><span class="op">)</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"2_s_at"</span>, <span class="st">"3_s_at"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span></code></pre></div>
<p>In order to test this function, we create a small sample tibble of expression
data containing only seven samples and four IDs. Two of the rows <em>do</em> have more
than 15% of their values exceeding <code>log2(15)</code>, the other two do not. This test
ensures that <code>filter_15()</code> selects the correct rows. Creating a small sample
table like this can be very useful when testing your own code since you don’t
need to look at a large amount of data to see if it’s working correctly or not.</p>
</div>
</div>
<div id="affy_to_hgnc" class="section level3 unnumbered">
<h3>4. <code>affy_to_hgnc()</code><a class="anchor" aria-label="anchor" href="#affy_to_hgnc"><i class="fas fa-link"></i></a>
</h3>
<p>This is an important, but sometimes painful, part of using R. There is a great
built-in package for connecting to Ensembl (a database of genomic information
for many species) called <code>biomaRt</code>. We will use <code>biomaRt</code> to connect the
affymetrix probe IDs to more recognizable HGNC gene IDs. The problem is that
<code>biomaRt</code> depends on an external API (application program interface) to retrieve
data, and this connection sometimes (oftentimes) doesn’t work. While their may
be more nuanced approaches to an unstable resource like this like automatically
retrying failed connections, the best advice for the time being is to try
running this function a few times if it doesn’t work at first. The errors are
clear when it comes to a failed connection, so know that when you get to this
stage it likely isn’t your code’s fault.</p>
<p>To build a <code>biomaRt</code> query, read the documentation in section 3
<a href="https://bioconductor.org/packages/release/bioc/vignettes/biomaRt/inst/doc/accessing_ensembl.html#how-to-build-a-biomart-query">here</a>.
The <code>biomart</code> you should use is <code>ENSEMBL_MART_ENSEMBL</code>, the data set
<code>hsapiens_gene_ensembl</code>, and you want to find the attributes
<code>c("affy_hg_u133_plus_2", "hgnc_symbol")</code>. The data you filter using
<code>filter_15()</code> returns a list of <code>affy_hg_u133_plus_2</code> probe IDs, and the gene
names we’re interested in are stored in <code>hgnc_symbol</code>.</p>
<p>This function should return a tibble, but <code>biomaRt</code>’s <code>getBM()</code> will only accept
and return a <code>data.frame</code>. You can use <code><a href="https://dplyr.tidyverse.org/reference/pull.html">dplyr::pull()</a></code> to turn a tibble into a
simple character vector, and <code><a href="https://tibble.tidyverse.org/reference/as_tibble.html">dplyr::as_tibble()</a></code> to go from a data frame to a
tibble.</p>
<div id="tests-2" class="section level4 unnumbered">
<h4>Tests<a class="anchor" aria-label="anchor" href="#tests-2"><i class="fas fa-link"></i></a>
</h4>
<div class="sourceCode" id="cb520"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">test_that</span><span class="op">(</span><span class="st">"affy ids can be converted to HGNC names properly using affy_to_hgnc()"</span>, <span class="op">{</span></span>
<span>  <span class="co"># biomaRt super buggy so we can try to account for not connecting well</span></span>
<span>  <span class="va">response</span> <span class="op">&lt;-</span> <span class="kw"><a href="https://rdrr.io/r/base/try.html">try</a></span><span class="op">(</span><span class="fu">affy_to_hgnc</span><span class="op">(</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html">tibble</a></span><span class="op">(</span><span class="st">'1553551_s_at'</span><span class="op">)</span><span class="op">)</span>, <span class="cn">TRUE</span><span class="op">)</span></span>
<span>  <span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/grep.html">grepl</a></span><span class="op">(</span><span class="st">"Error"</span>, <span class="va">response</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu">expect_warning</span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/warning.html">warning</a></span><span class="op">(</span><span class="st">"Could not connect to ENSEMBL."</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="op">}</span></span>
<span>  <span class="kw">else</span> <span class="op">{</span></span>
<span>    <span class="fu">expect_equal</span><span class="op">(</span><span class="va">response</span><span class="op">$</span><span class="va">hgnc_symbol</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"MT-ND1"</span>, <span class="st">"MT-TI"</span>, <span class="st">"MT-TM"</span>, <span class="st">"MT-ND2"</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="op">}</span></span>
<span><span class="op">}</span><span class="op">)</span></span></code></pre></div>
<p>As fun as it is to try to get <code>biomaRt</code> function to connect correctly, it is
even more fun to test them. Since a failure to connect doesn’t indicate an
actual failure in <em>our</em> code we must use a <code><a href="https://rdrr.io/r/base/try.html">try()</a></code> block in order to capture if
there is a connection error.</p>
<div class="box note">
<p><strong>Note</strong> that the <code><a href="https://rdrr.io/r/base/try.html">try()</a></code> function is a part of programming called
<em><strong>error-handling</strong> which extends to many other languages. We often expect
</em>errors when running our programs (such as right now) but don’t want to shut
<em>down our entire operation if it’s an error we can expect. Using <code>try</code>,
</em><code>except</code>, and <code>finally</code> (the latter two not appearing here) we can account for
*issues outside of our control and adapt our code to change the outcome.</p>
<p>Using <code>try except</code> is <strong><em>not</em></strong> a replacement for writing code that doesn’t
generate errors. If you can avoid an error in the first place, that is far
better than using error-handling.</p>
</div>
<p>In this case we <code><a href="https://rdrr.io/r/base/try.html">try()</a></code> to use <code>affy_to_hgnc()</code> to connect to Ensembl and store
the resulting error in <code>response</code>. We then check: if there is an “Error” then we
throw a <code><a href="https://rdrr.io/r/base/warning.html">warning()</a></code> to our testing output. This doesn’t stop further testing
from happening, but it ensures we know that something isn’t quite right. If the
<code>response</code> does not contain an error, we simply test that it returned the
correct gene symbols for our random affy probe ID of choice.</p>
</div>
</div>
<div id="reduce_data" class="section level3 unnumbered">
<h3>5. <code>reduce_data()</code><a class="anchor" aria-label="anchor" href="#reduce_data"><i class="fas fa-link"></i></a>
</h3>
<p>We have one final step in manipulating our data before we plot it. We have our
original data, the probe IDs and their associated HGNC symbols, and a list of
good gene names and bad gene names. <code>reduce_data</code> takes these four inputs and
returns a tibble that has reduced our expression data to only the genes of
interest and has a column describing which set of genes it belongs to (“good” or
“bad”). Changing the shape of the data is incredibly useful for <code>ggplot</code>, the
tidyverse package we will use for plotting. While there is flexibility when
using <code>ggplot</code> to plot data, having data in
<a href="https://www.datacamp.com/community/tutorials/long-wide-data-R">long</a> format is
typically ideal.</p>
<p>Once again, there are multiple ways to reorganize data in this way. We used the
base function <code><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html">match()</a></code> to connect our probe IDs to with our HGNC IDs in
<code>name_ids</code>. We then used <code><a href="https://tibble.tidyverse.org/reference/add_column.html">tibble::add_column()</a></code> to insert the new data in the
correct location. Finally, we created two tibbles of good and bad genes using
<code><a href="https://rdrr.io/pkg/BiocGenerics/man/which.html">which()</a></code> and the <code>%in%</code> modifier. Which evaluates true conditions across a
range of data, so we can pass the list of genes we want and select the correct
ones. For instance:</p>
<div class="sourceCode" id="cb521"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://tibble.tidyverse.org/">tibble</a></span><span class="op">)</span></span>
<span><span class="va">tib</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html">tibble</a></span><span class="op">(</span>gene <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"gene1"</span>, <span class="st">"gene2"</span>, <span class="st">"gene3"</span>, <span class="st">"gene4"</span><span class="op">)</span>,</span>
<span>              affy <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"a_s_1"</span>, <span class="st">"a_s_2"</span>, <span class="st">"a_s_3"</span>, <span class="st">"a_s_4"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/which.html">which</a></span><span class="op">(</span><span class="va">tib</span><span class="op">$</span><span class="va">gene</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"gene2"</span>, <span class="st">"gene3"</span><span class="op">)</span><span class="op">)</span> <span class="co"># returns the index of the TRUE rows</span></span></code></pre></div>
<pre><code>## [1] 2 3</code></pre>
<div class="sourceCode" id="cb523"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tib</span><span class="op">$</span><span class="va">affy</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/which.html">which</a></span><span class="op">(</span><span class="va">tib</span><span class="op">$</span><span class="va">gene</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"gene2"</span>, <span class="st">"gene3"</span><span class="op">)</span><span class="op">)</span><span class="op">]</span> <span class="co"># use [] to get data back</span></span></code></pre></div>
<pre><code>## [1] "a_s_2" "a_s_3"</code></pre>
<p>Once again, there are many ways to reshape this data (some maybe more elegant
than this!) and all we need is the data to be correctly shaped when it is
returned.</p>
<div id="tests-3" class="section level4 unnumbered">
<h4>Tests<a class="anchor" aria-label="anchor" href="#tests-3"><i class="fas fa-link"></i></a>
</h4>
<div class="sourceCode" id="cb525"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">test_that</span><span class="op">(</span><span class="st">"reduce_data() is correctly changing the size and shape of the tibble"</span>, <span class="op">{</span></span>
<span>  <span class="va">t_tibble</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html">tibble</a></span><span class="op">(</span>probeids <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"1_s_at"</span>, <span class="st">"2_s_at"</span>, <span class="st">"3_s_at"</span><span class="op">)</span>,</span>
<span>                     GSM1 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">9.5</span>, <span class="fl">7.6</span>, <span class="fl">5.5</span><span class="op">)</span>,</span>
<span>                     GSM2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">9.7</span>, <span class="fl">7.2</span>, <span class="fl">2.9</span><span class="op">)</span>,</span>
<span>                     GSM3 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">6.9</span>, <span class="fl">4.3</span>, <span class="fl">6.8</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="va">names</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html">tibble</a></span><span class="op">(</span>affy_hg_u133_plus_2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"1_s_at"</span>, <span class="st">"2_s_at"</span>, <span class="st">"3_s_at"</span><span class="op">)</span>,</span>
<span>                  hgnc_symbol <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"A-REAL-GENE"</span>, <span class="st">"SONIC"</span>, <span class="st">"UTWPU"</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="va">good</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"A-REAL-GENE"</span><span class="op">)</span></span>
<span>  <span class="va">bad</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"SONIC"</span><span class="op">)</span></span>
<span>  <span class="va">reduce_test</span> <span class="op">&lt;-</span> <span class="fu">reduce_data</span><span class="op">(</span><span class="va">t_tibble</span>, <span class="va">names</span>, <span class="va">good</span>, <span class="va">bad</span><span class="op">)</span></span>
<span>  <span class="va">result</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html">tibble</a></span><span class="op">(</span>probeids <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"1_s_at"</span>, <span class="st">"2_s_at"</span><span class="op">)</span>,</span>
<span>                   hgnc_symbol <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"A-REAL-GENE"</span>, <span class="st">"SONIC"</span><span class="op">)</span>,</span>
<span>                   gene_set <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"good"</span>, <span class="st">"bad"</span><span class="op">)</span>,</span>
<span>                   GSM1 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">9.5</span>, <span class="fl">7.6</span><span class="op">)</span>,</span>
<span>                   GSM2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">9.7</span>, <span class="fl">7.2</span><span class="op">)</span>,</span>
<span>                   GSM3 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">6.9</span>, <span class="fl">4.3</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="fu">expect_equal</span><span class="op">(</span><span class="va">reduce_test</span>, <span class="va">result</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span></code></pre></div>
<p>In order to test a function that changes a tibble, we need to use a tibble. We
only create a test table that is three rows by four columns, but that is enough
to get the gist of the function. We simply pass the four parameters to
<code>reduce_data()</code> and we expect it to create a tibble like <code>result</code>. <strong>Ensure your
output column names are the same here or your tests my fail.</strong> While not crucial
to the success of your assignment, maintaining correct column names across
multiple data transformations is an important skill.</p>

</div>
</div>
</div>
</div>
  <div class="chapter-nav">
<div class="prev"><a href="assignment-1.html">Assignment 1</a></div>
<div class="next"><a href="assignment-3.html">Assignment 3</a></div>
</div></main><div class="col-md-3 col-lg-2 d-none d-md-block sidebar sidebar-chapter">
    <nav id="toc" data-toggle="toc" aria-label="On this page"><h2>On this page</h2>
      <ul class="nav navbar-nav">
<li><a class="nav-link" href="#assignment-2">Assignment 2</a></li>
<li><a class="nav-link" href="#problem-statement-2">Problem Statement</a></li>
<li><a class="nav-link" href="#learning-objectives">Learning Objectives</a></li>
<li><a class="nav-link" href="#skill-list">Skill List</a></li>
<li><a class="nav-link" href="#instructions-2">Instructions</a></li>
<li>
<a class="nav-link" href="#function-details">Function Details</a><ul class="nav navbar-nav">
<li><a class="nav-link" href="#bioconductor">1. Bioconductor</a></li>
<li><a class="nav-link" href="#load_expression">2. load_expression()</a></li>
<li><a class="nav-link" href="#filter_15">3. filter_15()</a></li>
<li><a class="nav-link" href="#affy_to_hgnc">4. affy_to_hgnc()</a></li>
<li><a class="nav-link" href="#reduce_data">5. reduce_data()</a></li>
</ul>
</li>
</ul>

      <div class="book-extra">
        <ul class="list-unstyled">
          
        </ul>
</div>
    </nav>
</div>

</div>
</div> <!-- .container -->

<footer class="bg-primary text-light mt-5"><div class="container"><div class="row">

  <div class="col-12 col-md-6 mt-3">
    <p>"<strong>BF591 - R for Biological Sciences</strong>" was written by . </p>
  </div>

  <div class="col-12 col-md-6 mt-3">
    <p>This book was built by the <a class="text-light" href="https://bookdown.org">bookdown</a> R package.</p>
  </div>

</div></div>
</footer><!-- dynamically load mathjax for compatibility with self-contained --><script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/latest.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script><script type="text/x-mathjax-config">const popovers = document.querySelectorAll('a.footnote-ref[data-toggle="popover"]');
for (let popover of popovers) {
  const div = document.createElement('div');
  div.setAttribute('style', 'position: absolute; top: 0, left:0; width:0, height:0, overflow: hidden; visibility: hidden;');
  div.innerHTML = popover.getAttribute('data-content');

  var has_math = div.querySelector("span.math");
  if (has_math) {
    document.body.appendChild(div);
    MathJax.Hub.Queue(["Typeset", MathJax.Hub, div]);
    MathJax.Hub.Queue(function() {
      popover.setAttribute('data-content', div.innerHTML);
      document.body.removeChild(div);
    })
  }
}
</script>
</body>
</html>
