<!DOCTYPE html>
<html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Assignment 6 | BF591 - R for Biological Sciences</title>
<meta name="generator" content="bookdown 0.32 with bs4_book()">
<meta property="og:title" content="Assignment 6 | BF591 - R for Biological Sciences">
<meta property="og:type" content="book">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="Assignment 6 | BF591 - R for Biological Sciences">
<!-- JS --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://kit.fontawesome.com/6ecbd6c532.js" crossorigin="anonymous"></script><script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="libs/bootstrap-4.6.0/bootstrap.min.css" rel="stylesheet">
<script src="libs/bootstrap-4.6.0/bootstrap.bundle.min.js"></script><script src="libs/bs3compat-0.4.2/transition.js"></script><script src="libs/bs3compat-0.4.2/tabs.js"></script><script src="libs/bs3compat-0.4.2/bs3compat.js"></script><link href="libs/bs4_book-1.0.0/bs4_book.css" rel="stylesheet">
<script src="libs/bs4_book-1.0.0/bs4_book.js"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- CSS --><style type="text/css">
    
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  </style>
<style type="text/css">
    /* Used with Pandoc 2.11+ new --citeproc when CSL is used */
    div.csl-bib-body { }
    div.csl-entry {
      clear: both;
        }
    .hanging div.csl-entry {
      margin-left:2em;
      text-indent:-2em;
    }
    div.csl-left-margin {
      min-width:2em;
      float:left;
    }
    div.csl-right-inline {
      margin-left:2em;
      padding-left:1em;
    }
    div.csl-indent {
      margin-left: 2em;
    }
  </style>
<link rel="stylesheet" href="style.css">
<meta name="description" content="Volcanoes don’t really look like that, do they?  Problem Statement Because R has been utilized in bioinformatics communities for years at this point, there are many different approaches to do one...">
<meta property="og:description" content="Volcanoes don’t really look like that, do they?  Problem Statement Because R has been utilized in bioinformatics communities for years at this point, there are many different approaches to do one...">
<meta name="twitter:description" content="Volcanoes don’t really look like that, do they?  Problem Statement Because R has been utilized in bioinformatics communities for years at this point, there are many different approaches to do one...">
</head>
<body data-spy="scroll" data-target="#toc">

<div class="container-fluid">
<div class="row">
  <header class="col-sm-12 col-lg-3 sidebar sidebar-book"><a class="sr-only sr-only-focusable" href="#content">Skip to main content</a>

    <div class="d-flex align-items-start justify-content-between">
      <h1>
        <a href="index.html" title="">BF591 - R for Biological Sciences</a>
      </h1>
      <button class="btn btn-outline-primary d-lg-none ml-2 mt-1" type="button" data-toggle="collapse" data-target="#main-nav" aria-expanded="true" aria-controls="main-nav"><i class="fas fa-bars"></i><span class="sr-only">Show table of contents</span></button>
    </div>

    <div id="main-nav" class="collapse-lg">
      <form role="search">
        <input id="search" class="form-control" type="search" placeholder="Search" aria-label="Search">
</form>

      <nav aria-label="Table of contents"><h2>Table of contents</h2>
        <ul class="book-toc list-unstyled">
<li><a class="" href="index.html">R for Biological Sciences</a></li>
<li><a class="" href="introduction.html"><span class="header-section-number">1</span> Introduction</a></li>
<li><a class="" href="data-bio.html"><span class="header-section-number">2</span> Data in Biology</a></li>
<li><a class="" href="preliminaries.html"><span class="header-section-number">3</span> Preliminaries</a></li>
<li><a class="" href="prog-basics.html"><span class="header-section-number">4</span> R Programming</a></li>
<li><a class="" href="data-wrangling.html"><span class="header-section-number">5</span> Data Wrangling</a></li>
<li><a class="" href="data-science.html"><span class="header-section-number">6</span> Data Science</a></li>
<li><a class="" href="data-visualization.html"><span class="header-section-number">7</span> Data Visualization</a></li>
<li><a class="" href="biology-bioinformatics.html"><span class="header-section-number">8</span> Biology &amp; Bioinformatics</a></li>
<li><a class="" href="engineering.html"><span class="header-section-number">9</span> EngineeRing</a></li>
<li><a class="" href="rshiny.html"><span class="header-section-number">10</span> RShiny</a></li>
<li><a class="" href="communicating-with-r.html"><span class="header-section-number">11</span> Communicating with R</a></li>
<li><a class="" href="contribution-guide.html"><span class="header-section-number">12</span> Contribution Guide</a></li>
<li class="book-part">Assignments</li>
<li><a class="" href="starting-an-assignment.html">Starting an Assignment</a></li>
<li><a class="" href="assignment-overview.html">Assignment Overview</a></li>
<li><a class="" href="assignment-0.html">Assignment 0</a></li>
<li><a class="" href="assignment-1.html">Assignment 1</a></li>
<li><a class="" href="assignment-2.html">Assignment 2</a></li>
<li><a class="" href="assignment-3.html">Assignment 3</a></li>
<li><a class="" href="assignment-4.html">Assignment 4</a></li>
<li><a class="" href="assignment-5.html">Assignment 5</a></li>
<li><a class="active" href="assignment-6.html">Assignment 6</a></li>
<li><a class="" href="assignment-7.html">Assignment 7</a></li>
<li><a class="" href="final-project.html">Final Project</a></li>
<li class="book-part">Appendix</li>
<li><a class="" href="detailed-class-schedule.html"><span class="header-section-number">A</span> Detailed Class Schedule</a></li>
</ul>

        <div class="book-extra">
          
        </div>
      </nav>
</div>
  </header><main class="col-sm-12 col-md-9 col-lg-7" id="content"><div id="assignment-6" class="section level1 unnumbered">
<h1>Assignment 6<a class="anchor" aria-label="anchor" href="#assignment-6"><i class="fas fa-link"></i></a>
</h1>
<blockquote>
<p>Volcanoes don’t really look like that, do they?</p>
</blockquote>
<div id="problem-statement-6" class="section level2 unnumbered">
<h2>Problem Statement<a class="anchor" aria-label="anchor" href="#problem-statement-6"><i class="fas fa-link"></i></a>
</h2>
<p>Because R has been utilized in bioinformatics communities for years at this point, there are many different approaches to do one task. While in the last assignment we just used DESeq2, we will be performing differential expression analysis with two additional packages (edgeR and Limma) and comparing the results we find.</p>
</div>
<div id="learning-objectives-4" class="section level2 unnumbered">
<h2>Learning Objectives<a class="anchor" aria-label="anchor" href="#learning-objectives-4"><i class="fas fa-link"></i></a>
</h2>
<ol style="list-style-type: decimal">
<li>The basic operations of <code>ggplot</code> and “The Grammar of Graphics”.</li>
<li>Using differential gene expression analysis packages to generate data for plotting.</li>
<li>Creating simple plots and <strong>not</strong> using the default colors and themes.</li>
<li>Combining multiple plots into one image using <code><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap()</a></code>
</li>
</ol>
</div>
<div id="skill-list-4" class="section level2 unnumbered">
<h2>Skill List<a class="anchor" aria-label="anchor" href="#skill-list-4"><i class="fas fa-link"></i></a>
</h2>
<ul>
<li>A tempered heart and mind that understands a plot may not ever look <em>exactly</em> how you want it to.</li>
<li>An intermediate understanding of R’s most popular plotting package, <code>ggplot</code>.</li>
<li>Further understanding of differential expression analysis and its plotting.</li>
<li>A sense of superiority whenever you see a publication use the base R plotting package for figures.</li>
</ul>
</div>
<div id="instructions-4" class="section level2 unnumbered">
<h2>Instructions<a class="anchor" aria-label="anchor" href="#instructions-4"><i class="fas fa-link"></i></a>
</h2>
<p>Complete <code>main.R</code> in a way that satisfies all of the tests in <code>test_main.R</code>. Follow the instructions in the function descriptions for <code>main.R</code> and read here for details on the tests your code should satisfy. You will use these functions to complete figures in <code>report.Rmd</code>.</p>
<p>You can use <code>testthat:test_file('test_main.R')</code> in your R console to run all tests in <code>test_main.R</code>. Do not modify the tests. While it is difficult to write tests that cover every solution to a problem, our tests are written to ensure the outputs of a function are aligned with what an analysis should perform. If you feel a test is working incorrectly, let a TA know.</p>
</div>
<div id="function-details-2" class="section level2 unnumbered">
<h2>Function Details<a class="anchor" aria-label="anchor" href="#function-details-2"><i class="fas fa-link"></i></a>
</h2>
<div id="load_n_trim" class="section level3 unnumbered">
<h3>1. <code>load_n_trim()</code><a class="anchor" aria-label="anchor" href="#load_n_trim"><i class="fas fa-link"></i></a>
</h3>
<p>Once again we have some data to load into R in order to manipulate. We’ll be loading a counts file, which is a matrix of genes (rows) by samples (columns). Each cell is the number of that gene found in that sample. This is a good opportunity to use a data frame or a tibble, but because of package restrictions we will be sticking to data frames for this assignment. We’re using a couple different packages to process our data, and they input/output <em>exclusively</em> in data frames, so we will be consistent and save ourselves some trouble with conversions.</p>
<p>We want to return a data frame that is essentially identical to our input file (always a good idea to use the command line to see what you’re working with). Ensure the gene names are stored as <code>row.names</code> and <em>not</em> as a separate column, data frames have a distinction between the two. This function should also reduce the columns to just those of interest (P0 and Adult).</p>
<div class="box hint">
<p>Row names can be changed using the <code><a href="https://rdrr.io/r/base/row.names.html">row.names()</a></code> function. The new row names must be a character vector of exactly the same length.</p>
<div class="sourceCode" id="cb541"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb541-1"><a href="assignment-6.html#cb541-1" tabindex="-1"></a>df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">a =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">0</span>, <span class="dv">0</span>), <span class="at">b =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">3</span>, <span class="dv">5</span>))</span>
<span id="cb541-2"><a href="assignment-6.html#cb541-2" tabindex="-1"></a><span class="fu">row.names</span>(df) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Row1"</span>, <span class="st">"Row2"</span>, <span class="st">"Row3"</span>)</span>
<span id="cb541-3"><a href="assignment-6.html#cb541-3" tabindex="-1"></a><span class="fu">print</span>(df)</span>
<span id="cb541-4"><a href="assignment-6.html#cb541-4" tabindex="-1"></a>     a b</span>
<span id="cb541-5"><a href="assignment-6.html#cb541-5" tabindex="-1"></a>Row1 <span class="dv">0</span> <span class="dv">1</span></span>
<span id="cb541-6"><a href="assignment-6.html#cb541-6" tabindex="-1"></a>Row2 <span class="dv">0</span> <span class="dv">3</span></span>
<span id="cb541-7"><a href="assignment-6.html#cb541-7" tabindex="-1"></a>Row3 <span class="dv">0</span> <span class="dv">5</span></span></code></pre></div>
</div>
<div id="tests-4" class="section level4 unnumbered">
<h4>Tests<a class="anchor" aria-label="anchor" href="#tests-4"><i class="fas fa-link"></i></a>
</h4>
<div class="sourceCode" id="cb542"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">test_that</span><span class="op">(</span><span class="st">"test data is loading correctly and reduced"</span>, <span class="op">{</span></span>
<span>  <span class="va">test_df</span> <span class="op">&lt;-</span> <span class="fu">load_n_trim</span><span class="op">(</span><span class="va">csv</span><span class="op">)</span></span>
<span>  <span class="fu">expect_equal</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">test_df</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"vP0_1"</span>, <span class="st">"vP0_2"</span>, <span class="st">"vAd_1"</span>, <span class="st">"vAd_2"</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="fu">expect_equal</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">test_df</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">55416</span>, <span class="fl">4</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="fu">expect_equal</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/class.html">class</a></span><span class="op">(</span><span class="va">test_df</span><span class="op">)</span>, <span class="st">"data.frame"</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span></code></pre></div>
<p>The tests for this assignment are relatively straightforward, mostly ensuring that data is loaded and ends up in the right shape and format. The first test ensures the columns were filtered, the second ensures the size is correct, and the final ensures it is a data frame and <em>not</em> a tibble. No tibbles. Tibble-less?</p>
</div>
</div>
<div id="run_deseq" class="section level3 unnumbered">
<h3>2. <code>run_deseq()</code><a class="anchor" aria-label="anchor" href="#run_deseq"><i class="fas fa-link"></i></a>
</h3>
<p>This is the first of three functions that technically do identical things (but in different ways). One of the most popular Bioconductor packages, DESeq2 has a number of options available for differential expression analysis. We will load in the counts data, select our variables of interest, and use <code><a href="https://rdrr.io/pkg/DESeq2/man/DESeq.html">DESeq()</a></code> to process them. Links are included in the function description to DESeq2 documentation, you will find answers to most questions there, especially since those were the documents used to <em>write</em> this assignment. This function will return the results of the analysis.</p>
<div id="tests-5" class="section level4 unnumbered">
<h4>Tests<a class="anchor" aria-label="anchor" href="#tests-5"><i class="fas fa-link"></i></a>
</h4>
<div class="sourceCode" id="cb543"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">test_that</span><span class="op">(</span><span class="st">"deseq2 function is returning correct results"</span>, <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/load.html">load</a></span><span class="op">(</span><span class="st">"mock_counts_df.RData"</span><span class="op">)</span> <span class="co"># loads the counts_df object into env</span></span>
<span>  <span class="va">coldata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>condition <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"day4"</span>, <span class="st">"day7"</span><span class="op">)</span>, each<span class="op">=</span><span class="fl">2</span><span class="op">)</span>,</span>
<span>                        type<span class="op">=</span><span class="st">"paired-end"</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/row.names.html">row.names</a></span><span class="op">(</span><span class="va">coldata</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"vP4_1"</span>, <span class="st">"vP4_2"</span>, <span class="st">"vP7_1"</span>, <span class="st">"vP7_2"</span><span class="op">)</span></span>
<span>  <span class="fu">expect_warning</span><span class="op">(</span><span class="va">deseq</span> <span class="op">&lt;-</span> <span class="fu">run_deseq</span><span class="op">(</span><span class="va">counts_df</span>,</span>
<span>                                    <span class="va">coldata</span>,</span>
<span>                                    <span class="fl">10</span>, <span class="st">"condition_day7_vs_day4"</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="fu">expect_equal</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">deseq</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">19127</span>, <span class="fl">6</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="fu">expect_equal</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/class.html">class</a></span><span class="op">(</span><span class="va">deseq</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="st">"DESeqResults"</span><span class="op">)</span></span>
<span>  <span class="fu">expect_equal</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"pvalue"</span>, <span class="st">"padj"</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">deseq</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="cn">TRUE</span>, <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span></code></pre></div>
<p>Less concerning for you all but writing tests for larger data sets and more complicated functions like this is…tricky. The problem is it can’t be written like a script, we can’t use the results from <code>load_n_trim()</code> to test <em>this</em> function, because what if one fails independently of the other? So in this case, we create an RData object of correctly loaded data that works <em>even if</em> <code>load_n_trim()</code> is absolutely broken. Also, in this case, we are using different sets to test this (days 4 and 7 instead of 0 and adult) so you can’t just load the test data and say you wrote the code! That would be disingenuous of you, and being disingenuous is for after you graduate.</p>
<p>This test loads the sample data (a data frame of counts) and creates the <code>coldata</code> object according to DESeq2 specifications. We wrap the <code>run_deseq()</code> in <code>expect_warning()</code> because DESeq throws a warning when it performs an expected behavior: converting strings to factors (this is bad programming! this is not what warnings are for, the package is functioning as expected). We then do a similar series of tests: ensure the dimensions are correct, ensure the results object are from DESeq, and ensure that the column names we need are present.</p>
</div>
</div>
<div id="run_edger" class="section level3 unnumbered">
<h3>3. <code>run_edger()</code><a class="anchor" aria-label="anchor" href="#run_edger"><i class="fas fa-link"></i></a>
</h3>
<p>As mentioned in the previous function, this is the EdgeR implementation of running the differential expression analysis based on our counts file. Again, this function should follow roughly the portion of the documentation mentioned in the function description. EdgeR has a number of additional plotting functions (such as <code>plotMDS()</code>) you may like to include. These are optional but pretty fun to see if your data is working out as expected (similar days clustered close together and so on). You will return the <em>entire</em> results object, no filtering necessary yet.</p>
<div id="tests-6" class="section level4 unnumbered">
<h4>Tests<a class="anchor" aria-label="anchor" href="#tests-6"><i class="fas fa-link"></i></a>
</h4>
<div class="sourceCode" id="cb544"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">test_that</span><span class="op">(</span><span class="st">"edger function is returning correct results"</span>, <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/load.html">load</a></span><span class="op">(</span><span class="st">"mock_counts_df.RData"</span><span class="op">)</span></span>
<span>  <span class="va">group</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span>, each<span class="op">=</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="va">edger_res</span> <span class="op">&lt;-</span> <span class="fu">run_edger</span><span class="op">(</span><span class="va">counts_df</span>, <span class="va">group</span><span class="op">)</span></span>
<span>  <span class="fu">expect_equal</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">edger_res</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"logFC"</span>, <span class="st">"logCPM"</span>, <span class="st">"PValue"</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="fu">expect_equal</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">edger_res</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">15026</span>, <span class="fl">3</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span></code></pre></div>
<p>Similar to the DESeq2 tests, we aren’t necessarily looking for a carbon copy of the results. While the process <em>should</em> be entirely deterministic, it is better to allow some wiggle room and merely look that the columns we’re interested in are present and that the data has the right shape. Note that EdgeR doesn’t perform a BH correction (the p-adjusted column is non-existant). It’s either an option that’s off by default or just not a feature! Not a problem, we have the p-values so we can do it ourselves (see the markdown).</p>
</div>
</div>
<div id="run_limma-and-run_voom" class="section level3 unnumbered">
<h3>4. <code>run_limma()</code> and <code>run_voom()</code><a class="anchor" aria-label="anchor" href="#run_limma-and-run_voom"><i class="fas fa-link"></i></a>
</h3>
<p>Our final in this set of three, we’re testing two related functions in the Limma package. While you can create an analysis that <em>only</em> uses Limma, in our case we can apply the <strong>voom</strong> workflow to refine our data a little more. More info on how these work in the cited documentation. A little technique that can be useful for optional workflows like this is including a boolean flag in the parameters of the function. If I am running a function <code>add_numbers(1, -4, abs=TRUE)</code>, the parameter <code>abs</code> might change the functions operation so that it takes the absolute value of numbers before adding them together. I could create this behavior like so:</p>
<div class="sourceCode" id="cb545"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">add_numbers</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span>, <span class="va">y</span>, <span class="va">abs</span><span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="kw">if</span> <span class="op">(</span><span class="va">abs</span><span class="op">)</span> <span class="op">{</span> <span class="co"># code enters here if abs is TRUE</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/function.html">return</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html">abs</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html">abs</a></span><span class="op">(</span><span class="va">y</span><span class="op">)</span><span class="op">)</span> <span class="co"># return means function exits, doesn't go to "else {}"</span></span>
<span>  <span class="op">}</span> <span class="kw">else</span> <span class="op">{</span> <span class="co"># code enters here if abs was false</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/function.html">return</a></span><span class="op">(</span><span class="va">x</span> <span class="op">+</span> <span class="va">y</span><span class="op">)</span></span>
<span>  <span class="op">}</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>This is a very useful technique for re-using functions, meaning I don’t need to write two functions that do similar things when I can change their behavior with that boolean parameter.</p>
<div id="tests-7" class="section level4 unnumbered">
<h4>Tests<a class="anchor" aria-label="anchor" href="#tests-7"><i class="fas fa-link"></i></a>
</h4>
<div class="sourceCode" id="cb546"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">test_that</span><span class="op">(</span><span class="st">"test limma + voom work + work together"</span>, <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/load.html">load</a></span><span class="op">(</span><span class="st">"data/mock_counts_df.RData"</span><span class="op">)</span></span>
<span>  <span class="va">group</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span>, each<span class="op">=</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="co"># design</span></span>
<span>  <span class="va">design</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>day4<span class="op">=</span><span class="fl">1</span>, day4vsday7<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">1</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/row.names.html">row.names</a></span><span class="op">(</span><span class="va">design</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"vP4_1"</span>, <span class="st">"vP4_2"</span>, <span class="st">"vP7_1"</span>, <span class="st">"vP7_2"</span><span class="op">)</span></span>
<span>  <span class="co"># limma + voom</span></span>
<span>  <span class="fu">expect_warning</span><span class="op">(</span><span class="va">voom_res</span> <span class="op">&lt;-</span> <span class="fu">run_limma</span><span class="op">(</span><span class="va">counts_df</span>, <span class="va">design</span>, <span class="va">group</span><span class="op">)</span><span class="op">)</span> <span class="co"># voom yelling, ignore</span></span>
<span>  <span class="fu">expect_equal</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">voom_res</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">15026</span>, <span class="fl">6</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="fu">expect_equal</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">voom_res</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"logFC"</span>, <span class="st">"AveExpr"</span>, <span class="st">"t"</span>, <span class="st">"P.Value"</span>,</span>
<span>                                  <span class="st">"adj.P.Val"</span>, <span class="st">"B"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span></code></pre></div>
<p>This ultimately functions similarly to the first two series of tests: load test data, apply the function to it, make sure the results are the right size, shape, and type. This includes the voom procedure, so using <em>only</em> limma results will yield failures.</p>
<div class="box note">
<p>Note that Limma uses EdgeR to input data, so your first couple steps will look pretty familiar.</p>
</div>
<p>We create a design object to use, defining the experimental conditions. These are again using the day 4 and day 7 trials from the heart data.</p>
</div>
</div>
<div id="combine_pval" class="section level3 unnumbered">
<h3>5. <code>combine_pval()</code><a class="anchor" aria-label="anchor" href="#combine_pval"><i class="fas fa-link"></i></a>
</h3>
<p>The final three functions concern modifying our differential expression results and creating plots of these combined results. While it would be trivial to make three plots using the same code or function, have R generate one consistent plot after changes in your data or method is extremely useful.</p>
<p>We use <code>combine_pval()</code> to coerce our data into long format, which is more applicable for facet wrapping. <code><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap()</a></code> is a <code>ggplot2</code> function that uses a property of the data to create separate plots in one image. In our case we will use the <code>package</code> property (DESeq2, edgeR, or Limma) to differentiate our data. There are a couple ways to reorganize data in this way, but I would suggest using <code><a href="https://tidyr.tidyverse.org/reference/gather.html">tidyr::gather()</a></code>. Our <code>key</code> will be the package and the <code>value</code> will be the p-value.</p>
<div id="tests-8" class="section level4 unnumbered">
<h4>Tests<a class="anchor" aria-label="anchor" href="#tests-8"><i class="fas fa-link"></i></a>
</h4>
<p>These tests are organized into one larger testing function because they use the same, randomly generated input data (a bunch of data frames with uniform distributions but with ranges that approximate actual data). If you would like to test only parts of this testing function, use an octothorpe (<code>#</code>) to comment out tests you haven’t reached yet.</p>
<div class="sourceCode" id="cb547"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb547-1"><a href="assignment-6.html#cb547-1" tabindex="-1"></a><span class="fu">test_that</span>(<span class="st">"ggplot and data formatting works"</span>, {</span>
<span id="cb547-2"><a href="assignment-6.html#cb547-2" tabindex="-1"></a>  deseq <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">pvalue =</span> <span class="fu">runif</span>(<span class="dv">1000</span>, <span class="fl">1e-100</span>, <span class="fl">1e-5</span>),</span>
<span id="cb547-3"><a href="assignment-6.html#cb547-3" tabindex="-1"></a>                      <span class="at">log2FoldChange =</span> <span class="fu">runif</span>(<span class="dv">1000</span>, <span class="sc">-</span><span class="dv">9</span>, <span class="dv">9</span>),</span>
<span id="cb547-4"><a href="assignment-6.html#cb547-4" tabindex="-1"></a>                      <span class="at">padj =</span> <span class="fu">runif</span>(<span class="dv">1000</span>, <span class="fl">1e-300</span>, <span class="fl">1e-15</span>))</span>
<span id="cb547-5"><a href="assignment-6.html#cb547-5" tabindex="-1"></a>  edger <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">PValue =</span> <span class="fu">runif</span>(<span class="dv">1000</span>, <span class="fl">1e-100</span>, <span class="fl">1e-5</span>),</span>
<span id="cb547-6"><a href="assignment-6.html#cb547-6" tabindex="-1"></a>                      <span class="at">logFC =</span> <span class="fu">runif</span>(<span class="dv">1000</span>, <span class="sc">-</span><span class="dv">9</span>, <span class="dv">9</span>),</span>
<span id="cb547-7"><a href="assignment-6.html#cb547-7" tabindex="-1"></a>                      <span class="at">padj =</span> <span class="fu">runif</span>(<span class="dv">1000</span>, <span class="fl">1e-100</span>, <span class="fl">1e-14</span>))</span>
<span id="cb547-8"><a href="assignment-6.html#cb547-8" tabindex="-1"></a>  limma <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">P.Value =</span> <span class="fu">runif</span>(<span class="dv">1000</span>, <span class="fl">1e-100</span>, <span class="fl">1e-5</span>),</span>
<span id="cb547-9"><a href="assignment-6.html#cb547-9" tabindex="-1"></a>                      <span class="at">logFC =</span> <span class="fu">runif</span>(<span class="dv">1000</span>, <span class="sc">-</span><span class="dv">9</span>, <span class="dv">9</span>),</span>
<span id="cb547-10"><a href="assignment-6.html#cb547-10" tabindex="-1"></a>                      <span class="at">adj.P.Val =</span> <span class="fu">runif</span>(<span class="dv">1000</span>, <span class="fl">1e-6</span>, <span class="fl">1e-4</span>))</span>
<span id="cb547-11"><a href="assignment-6.html#cb547-11" tabindex="-1"></a>  <span class="co"># combine_pval</span></span>
<span id="cb547-12"><a href="assignment-6.html#cb547-12" tabindex="-1"></a>  res <span class="ot">&lt;-</span> <span class="fu">combine_pval</span>(deseq, edger, limma)</span>
<span id="cb547-13"><a href="assignment-6.html#cb547-13" tabindex="-1"></a>  <span class="fu">expect_equal</span>(<span class="fu">dim</span>(res), <span class="fu">c</span>(<span class="dv">3000</span>, <span class="dv">2</span>))</span>
<span id="cb547-14"><a href="assignment-6.html#cb547-14" tabindex="-1"></a>  <span class="fu">expect_true</span>(<span class="fu">mean</span>(res<span class="sc">$</span>pval) <span class="sc">&lt;</span> <span class="fl">5e-5</span> <span class="sc">&amp;&amp;</span> <span class="fu">mean</span>(res<span class="sc">$</span>pval) <span class="sc">&gt;</span> <span class="fl">5e-7</span>)</span>
<span id="cb547-15"><a href="assignment-6.html#cb547-15" tabindex="-1"></a>  <span class="fu">expect_equal</span>(<span class="fu">table</span>(res<span class="sc">$</span>package)[[<span class="dv">1</span>]], <span class="dv">1000</span>)</span>
<span id="cb547-16"><a href="assignment-6.html#cb547-16" tabindex="-1"></a>  ...</span></code></pre></div>
<p>The first nine lines establish the test data frames. While random, they will let us test if functions are working correctly.</p>
<div class="box important">
<p>The test data frames do not mimic the normal results exactly. The rows we are interested in have the same name, but have different numbers of columns. For that reason, I suggest your function uses <code>$</code>-notation when referencing columns, instead of position. <code>deseq$pvalue</code> will work more better than <code>deseq[,1]</code> for this test data.</p>
</div>
<p>We test that we only have 2 columns and 3,000 rows, and that the fake p-values are within the expected range. Finally, we use the <code><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html">table()</a></code> function to test that there is only 1,000 of any one package in this data frame.</p>
</div>
</div>
<div id="create_facets" class="section level3 unnumbered">
<h3>6. <code>create_facets()</code><a class="anchor" aria-label="anchor" href="#create_facets"><i class="fas fa-link"></i></a>
</h3>
<p>Now that we have examined the p-values, we can look at the adjusted p-values. We’ll be constructing volcano plots, which compare the log<sub>2</sub> fold-change and the adjusted p-value, so the <code>create_facets()</code> function will be somewhat more complicated. Rather than use <code><a href="https://tidyr.tidyverse.org/reference/gather.html">gather()</a></code> again, it may be easier to create three tibbles or data frames of the variables of interest and combine them with <code><a href="https://rdrr.io/pkg/BiocGenerics/man/cbind.html">rbind()</a></code>.</p>
<div id="tests-9" class="section level4 unnumbered">
<h4>Tests<a class="anchor" aria-label="anchor" href="#tests-9"><i class="fas fa-link"></i></a>
</h4>
<div class="sourceCode" id="cb548"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span>  <span class="va">...</span></span>
<span>  <span class="co"># create_facets</span></span>
<span>  <span class="va">facets</span> <span class="op">&lt;-</span> <span class="fu">create_facets</span><span class="op">(</span><span class="va">deseq</span>, <span class="va">edger</span>, <span class="va">limma</span><span class="op">)</span></span>
<span>  <span class="fu">expect_equal</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">facets</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">3000</span>, <span class="fl">3</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="fu">expect_equal</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html">table</a></span><span class="op">(</span><span class="va">facets</span><span class="op">$</span><span class="va">package</span><span class="op">)</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span>, <span class="fl">1000</span><span class="op">)</span></span>
<span>  <span class="va">...</span></span></code></pre></div>
<p>Another simple set of tests, this time we want to ensure there are 3,000 rows and <em>3</em> columns. We also want to ensure there aren’t more than 1,000 rows of any package type.</p>
</div>
</div>
<div id="theme_plot" class="section level3 unnumbered">
<h3>7. <code>theme_plot()</code><a class="anchor" aria-label="anchor" href="#theme_plot"><i class="fas fa-link"></i></a>
</h3>
<p>Finally, we have all of our data in the right shape and we can do some plotting. There is a lot to learn about ggplot, and I mention a few resources in the function description, but don’t be afraid to experiment! Try different colors, themes, and styles. The easiest part is getting your data on the plot, the hard part is making it look attractive.</p>
<div id="tests-10" class="section level4 unnumbered">
<h4>Tests<a class="anchor" aria-label="anchor" href="#tests-10"><i class="fas fa-link"></i></a>
</h4>
<div class="sourceCode" id="cb549"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb549-1"><a href="assignment-6.html#cb549-1" tabindex="-1"></a>  ...</span>
<span id="cb549-2"><a href="assignment-6.html#cb549-2" tabindex="-1"></a>  <span class="co"># theme_plot</span></span>
<span id="cb549-3"><a href="assignment-6.html#cb549-3" tabindex="-1"></a>  <span class="fu">tibble</span>(<span class="at">logFC =</span> <span class="fu">runif</span>(<span class="dv">3000</span>, <span class="sc">-</span><span class="dv">9</span>, <span class="dv">9</span>),</span>
<span id="cb549-4"><a href="assignment-6.html#cb549-4" tabindex="-1"></a>         <span class="at">padj =</span> <span class="fu">runif</span>(<span class="dv">3000</span>, <span class="fl">1e-300</span>, <span class="fl">1e-5</span>),</span>
<span id="cb549-5"><a href="assignment-6.html#cb549-5" tabindex="-1"></a>         <span class="at">package =</span> <span class="fu">rep</span>(<span class="fu">c</span>(<span class="st">"DESeq2"</span>, <span class="st">"edgeR"</span>, <span class="st">"Limma"</span>), <span class="at">each=</span><span class="dv">1000</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb549-6"><a href="assignment-6.html#cb549-6" tabindex="-1"></a>  <span class="fu">theme_plot</span>() <span class="ot">-&gt;</span> volc_plot</span>
<span id="cb549-7"><a href="assignment-6.html#cb549-7" tabindex="-1"></a>  <span class="fu">expect_equal</span>(<span class="fu">dim</span>(volc_plot<span class="sc">$</span>data), <span class="fu">c</span>(<span class="dv">3000</span>, <span class="dv">3</span>))</span>
<span id="cb549-8"><a href="assignment-6.html#cb549-8" tabindex="-1"></a>  <span class="fu">expect_equal</span>(<span class="fu">class</span>(volc_plot<span class="sc">$</span>layers[[<span class="dv">1</span>]]<span class="sc">$</span>geom)[<span class="dv">1</span>], <span class="st">"GeomPoint"</span>)</span>
<span id="cb549-9"><a href="assignment-6.html#cb549-9" tabindex="-1"></a>  <span class="fu">expect_equal</span>(<span class="fu">class</span>(volc_plot<span class="sc">$</span>facet<span class="sc">$</span>params), <span class="st">"list"</span>)</span>
<span id="cb549-10"><a href="assignment-6.html#cb549-10" tabindex="-1"></a><span class="er">})</span></span></code></pre></div>
<p>We create a little more fake data, and then pipe it into our <code>theme_plot()</code> function, and save that ggplot object. Our tests simply look inside the ggplot object to confirm the data is the right shape, that you’re plotting using <code><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point()</a></code>, and that facets are present.</p>
<p>Once your tests are passing, take a look at the report and see how everything can work together!</p>

</div>
</div>
</div>
</div>
  <div class="chapter-nav">
<div class="prev"><a href="assignment-5.html">Assignment 5</a></div>
<div class="next"><a href="assignment-7.html">Assignment 7</a></div>
</div></main><div class="col-md-3 col-lg-2 d-none d-md-block sidebar sidebar-chapter">
    <nav id="toc" data-toggle="toc" aria-label="On this page"><h2>On this page</h2>
      <ul class="nav navbar-nav">
<li><a class="nav-link" href="#assignment-6">Assignment 6</a></li>
<li><a class="nav-link" href="#problem-statement-6">Problem Statement</a></li>
<li><a class="nav-link" href="#learning-objectives-4">Learning Objectives</a></li>
<li><a class="nav-link" href="#skill-list-4">Skill List</a></li>
<li><a class="nav-link" href="#instructions-4">Instructions</a></li>
<li>
<a class="nav-link" href="#function-details-2">Function Details</a><ul class="nav navbar-nav">
<li><a class="nav-link" href="#load_n_trim">1. load_n_trim()</a></li>
<li><a class="nav-link" href="#run_deseq">2. run_deseq()</a></li>
<li><a class="nav-link" href="#run_edger">3. run_edger()</a></li>
<li><a class="nav-link" href="#run_limma-and-run_voom">4. run_limma() and run_voom()</a></li>
<li><a class="nav-link" href="#combine_pval">5. combine_pval()</a></li>
<li><a class="nav-link" href="#create_facets">6. create_facets()</a></li>
<li><a class="nav-link" href="#theme_plot">7. theme_plot()</a></li>
</ul>
</li>
</ul>

      <div class="book-extra">
        <ul class="list-unstyled">
          
        </ul>
</div>
    </nav>
</div>

</div>
</div> <!-- .container -->

<footer class="bg-primary text-light mt-5"><div class="container"><div class="row">

  <div class="col-12 col-md-6 mt-3">
    <p>"<strong>BF591 - R for Biological Sciences</strong>" was written by . </p>
  </div>

  <div class="col-12 col-md-6 mt-3">
    <p>This book was built by the <a class="text-light" href="https://bookdown.org">bookdown</a> R package.</p>
  </div>

</div></div>
</footer><!-- dynamically load mathjax for compatibility with self-contained --><script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/latest.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script><script type="text/x-mathjax-config">const popovers = document.querySelectorAll('a.footnote-ref[data-toggle="popover"]');
for (let popover of popovers) {
  const div = document.createElement('div');
  div.setAttribute('style', 'position: absolute; top: 0, left:0; width:0, height:0, overflow: hidden; visibility: hidden;');
  div.innerHTML = popover.getAttribute('data-content');

  var has_math = div.querySelector("span.math");
  if (has_math) {
    document.body.appendChild(div);
    MathJax.Hub.Queue(["Typeset", MathJax.Hub, div]);
    MathJax.Hub.Queue(function() {
      popover.setAttribute('data-content', div.innerHTML);
      document.body.removeChild(div);
    })
  }
}
</script>
</body>
</html>
