<!DOCTYPE html>
<html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Assignment 5 | BF591 - R for Biological Sciences</title>
<meta name="generator" content="bookdown 0.29 with bs4_book()">
<meta property="og:title" content="Assignment 5 | BF591 - R for Biological Sciences">
<meta property="og:type" content="book">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="Assignment 5 | BF591 - R for Biological Sciences">
<!-- JS --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://kit.fontawesome.com/6ecbd6c532.js" crossorigin="anonymous"></script><script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="libs/bootstrap-4.6.0/bootstrap.min.css" rel="stylesheet">
<script src="libs/bootstrap-4.6.0/bootstrap.bundle.min.js"></script><script src="libs/bs3compat-0.3.1/transition.js"></script><script src="libs/bs3compat-0.3.1/tabs.js"></script><script src="libs/bs3compat-0.3.1/bs3compat.js"></script><link href="libs/bs4_book-1.0.0/bs4_book.css" rel="stylesheet">
<script src="libs/bs4_book-1.0.0/bs4_book.js"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- CSS --><style type="text/css">
    
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  </style>
<style type="text/css">
    /* Used with Pandoc 2.11+ new --citeproc when CSL is used */
    div.csl-bib-body { }
    div.csl-entry {
      clear: both;
        }
    .hanging div.csl-entry {
      margin-left:2em;
      text-indent:-2em;
    }
    div.csl-left-margin {
      min-width:2em;
      float:left;
    }
    div.csl-right-inline {
      margin-left:2em;
      padding-left:1em;
    }
    div.csl-indent {
      margin-left: 2em;
    }
  </style>
<link rel="stylesheet" href="style.css">
<meta name="description" content="Problem Statement One of the main questions asked by high throughput sequencing is whether some variable or condition (e.g. treatment/disease status) induces a measurable, distinguishable change....">
<meta property="og:description" content="Problem Statement One of the main questions asked by high throughput sequencing is whether some variable or condition (e.g. treatment/disease status) induces a measurable, distinguishable change....">
<meta name="twitter:description" content="Problem Statement One of the main questions asked by high throughput sequencing is whether some variable or condition (e.g. treatment/disease status) induces a measurable, distinguishable change....">
</head>
<body data-spy="scroll" data-target="#toc">

<div class="container-fluid">
<div class="row">
  <header class="col-sm-12 col-lg-3 sidebar sidebar-book"><a class="sr-only sr-only-focusable" href="#content">Skip to main content</a>

    <div class="d-flex align-items-start justify-content-between">
      <h1>
        <a href="index.html" title="">BF591 - R for Biological Sciences</a>
      </h1>
      <button class="btn btn-outline-primary d-lg-none ml-2 mt-1" type="button" data-toggle="collapse" data-target="#main-nav" aria-expanded="true" aria-controls="main-nav"><i class="fas fa-bars"></i><span class="sr-only">Show table of contents</span></button>
    </div>

    <div id="main-nav" class="collapse-lg">
      <form role="search">
        <input id="search" class="form-control" type="search" placeholder="Search" aria-label="Search">
</form>

      <nav aria-label="Table of contents"><h2>Table of contents</h2>
        <ul class="book-toc list-unstyled">
<li><a class="" href="index.html">R for Biological Sciences</a></li>
<li><a class="" href="introduction.html"><span class="header-section-number">1</span> Introduction</a></li>
<li><a class="" href="data-bio.html"><span class="header-section-number">2</span> Data in Biology</a></li>
<li><a class="" href="preliminaries.html"><span class="header-section-number">3</span> Preliminaries</a></li>
<li><a class="" href="prog-basics.html"><span class="header-section-number">4</span> R Programming</a></li>
<li><a class="" href="data-wrangling.html"><span class="header-section-number">5</span> Data Wrangling</a></li>
<li><a class="" href="data-science.html"><span class="header-section-number">6</span> Data Science</a></li>
<li><a class="" href="data-visualization.html"><span class="header-section-number">7</span> Data Visualization</a></li>
<li><a class="" href="biology-bioinformatics.html"><span class="header-section-number">8</span> Biology &amp; Bioinformatics</a></li>
<li><a class="" href="engineering.html"><span class="header-section-number">9</span> EngineeRing</a></li>
<li><a class="" href="rshiny.html"><span class="header-section-number">10</span> RShiny</a></li>
<li><a class="" href="communicating-with-r.html"><span class="header-section-number">11</span> Communicating with R</a></li>
<li><a class="" href="contribution-guide.html"><span class="header-section-number">12</span> Contribution Guide</a></li>
<li class="book-part">Assignments</li>
<li><a class="" href="starting-an-assignment.html">Starting an Assignment</a></li>
<li><a class="" href="assignment-overview.html">Assignment Overview</a></li>
<li><a class="" href="assignment-1.html">Assignment 1</a></li>
<li><a class="" href="assignment-2.html">Assignment 2</a></li>
<li><a class="" href="assignment-3.html">Assignment 3</a></li>
<li><a class="" href="assignment-4.html">Assignment 4</a></li>
<li><a class="active" href="assignment-5.html">Assignment 5</a></li>
<li><a class="" href="assignment-6.html">Assignment 6</a></li>
<li><a class="" href="assignment-7.html">Assignment 7</a></li>
<li><a class="" href="final-project.html">Final Project</a></li>
<li class="book-part">Appendix</li>
<li><a class="" href="class-outlines.html"><span class="header-section-number">A</span> Class Outlines</a></li>
</ul>

        <div class="book-extra">
          
        </div>
      </nav>
</div>
  </header><main class="col-sm-12 col-md-9 col-lg-7" id="content"><div id="assignment-5" class="section level1 unnumbered">
<h1>Assignment 5<a class="anchor" aria-label="anchor" href="#assignment-5"><i class="fas fa-link"></i></a>
</h1>
<div id="problem-statement-4" class="section level2 unnumbered">
<h2>Problem Statement<a class="anchor" aria-label="anchor" href="#problem-statement-4"><i class="fas fa-link"></i></a>
</h2>
<p>One of the main questions asked by high throughput sequencing is whether some
variable or condition (e.g. treatment/disease status) induces a measurable,
distinguishable change. In the context of mRNAseq, we are interested in whether
we can determine if mRNA levels (what we actually measure) and genes (what we
infer) are significantly up- or downregulated between conditions and if we can
relate these changes back to a biological phenotype either directly or
indirectly.</p>
</div>
<div id="learning-objectives-3" class="section level2 unnumbered">
<h2>Learning Objectives<a class="anchor" aria-label="anchor" href="#learning-objectives-3"><i class="fas fa-link"></i></a>
</h2>
<ol style="list-style-type: decimal">
<li>Understand the methodology and rationale for how DESeq2 normalizes data,
models counts, and performs differential expression</li>
<li>Perform differential expression analysis using DESeq2 and learn one method to
analyze time course data</li>
<li>Understand how to do basic inspection and evaluation of differential
expression (DE) results</li>
</ol>
</div>
<div id="skill-list-3" class="section level2 unnumbered">
<h2>Skill List<a class="anchor" aria-label="anchor" href="#skill-list-3"><i class="fas fa-link"></i></a>
</h2>
<ol style="list-style-type: decimal">
<li>DESeq2</li>
<li>Basic diagnostic plots of DE results</li>
<li>High-quality figures and plots using R and ggplot</li>
</ol>
</div>
<div id="deseq2-background" class="section level2 unnumbered">
<h2>DESeq2 Background<a class="anchor" aria-label="anchor" href="#deseq2-background"><i class="fas fa-link"></i></a>
</h2>
<p>DESeq2 is a well validated and highly cited method for differential expression
analysis of HTS data. The original paper describing the methodology in more
detail can be found here:</p>
<blockquote>
<p>Love, M.I., Huber, W. &amp; Anders, S. Moderated estimation of fold change and
dispersion for RNA-seq data with DESeq2. Genome Biol 15, 550 (2014).
<a href="https://doi.org/10.1186/s13059-014-0550-8" class="uri">https://doi.org/10.1186/s13059-014-0550-8</a></p>
</blockquote>
<p>We will link back to the original DESeq2 vignette at the head of each applicable
section for reference. We strongly encourage you to read the associated
documentation as it succinctly summarizes the important concepts and methodology
underlying each step of the process. It also provides example code snippets and
answers to many common questions regarding the usage of DESeq2 can already be
found in this documentation.</p>
<p>Briefly, at its core, DESeq2 models read counts and tests for differential
expression through negative binomial generalized linear models. DESeq2
normalizes count data by scaling by size factors calculated by the
median-of-ratios (developed in the original publication for DESeq). For
significance testing, DESeq2 performs a Wald test for model coefficients to
generate p-values and these are adjusted for multiple testing correction through
the Benjamini-Hochberg method.</p>
</div>
<div id="generating-a-counts-matrix" class="section level2 unnumbered">
<h2>Generating a counts matrix<a class="anchor" aria-label="anchor" href="#generating-a-counts-matrix"><i class="fas fa-link"></i></a>
</h2>
<p>DESeq2 takes as input a counts matrix where genes correspond to rows and samples
to columns. For one of the more common use cases of differential expression
analysis, mRNAseq experiments, we are interested in counts associated with
genes; however, DESeq2 is broadly applicable to many kinds of high-throughout
sequencing (HTS) count data.</p>
<p>Returning to the analysis of mRNAseq specifically, there are three main
approaches to map reads:</p>
<ol style="list-style-type: decimal">
<li>Alignment to a reference genome by a splice-aware aligner and counting of
reads falling into exonic regions using a reference annotation file</li>
<li>Transcript-based assembly approaches (TopHat, Cufflinks) or transcript abundance
quantification (Salmon, Kallisto)</li>
<li>Reference free assembly followed by mapping
and counting</li>
</ol>
<p>DESeq2 expects the values of the input matrix to be counts (integers) and for
the purposes of this assignment, you have been provided with a count matrix of
samples generated by the first method. Reads were aligned to the mouse genome
(GENCODE GRCm39) with STAR and quantified to gene counts by Verse using the
matching GTF file.</p>
</div>
<div id="prefiltering-counts-matrix" class="section level2 unnumbered">
<h2>Prefiltering Counts matrix<a class="anchor" aria-label="anchor" href="#prefiltering-counts-matrix"><i class="fas fa-link"></i></a>
</h2>
<p><a href="http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#pre-filtering">DESeq2 Vignette - Pre-filtering</a></p>
<p>Unless your experiment is quite large, filtering your counts matrix prior to
DESeq2 is not strictly required. If your experiment incorporates hundreds of
samples and a complicated model design with many interaction terms, you may wish
to perform a combination of manual filtering and further optimization of DESeq2
settings. Otherwise, for most datasets, default settings in DESeq2 will run in a
reasonably short amount of time on relatively modest hardware even without
pre-filtering.</p>
<p>There are many different strategies to filter count datasets and which you
choose to employ should be informed by your objective, and your data. Some
common filters include removing genes where the mean count across all samples is
below some threshold or removing genes based on how many samples have a zero
count. In general, extensive pre-filtering is not required and DESeq2
incorporates downstream methods to increase statistical power through
independent filtering (described in methods). <strong>For this assignment, we will not
be filtering the dataset prior to running DE analysis.</strong></p>
</div>
<div id="median-of-ratios-normalization" class="section level2 unnumbered">
<h2>Median-of-ratios normalization<a class="anchor" aria-label="anchor" href="#median-of-ratios-normalization"><i class="fas fa-link"></i></a>
</h2>
<p><a href="https://genomebiology.biomedcentral.com/articles/10.1186/gb-2010-11-10-r106">Original DESeq Publication</a></p>
<p>As mentioned earlier, DESeq2 expects a non-transformed, non-normalized matrix of
integer “counts” as input. DESeq2 performs its own normalization based on the
underlying assumption that not all or a small amount of genes are truly
differentially expressed. This is encapsulated in the median-of-ratios method
which accounts for library size as well as RNA composition. Although DESeq2
performs this normalization in the background, it can help conceptually to
understand the process by which it does so.</p>
<p>First, a pseudo-reference sample is created and set equal to the geometric mean
across all samples for a given gene. Then, ratios are calculated comparing each
sample to this pseudo-reference on gene-wise basis. Next, on a sample-wise
basis, the median value for all ratios is taken as that sample’s normalization
factor and for a given sample, all counts are divided by this factor. You have
learned how to extract these normalized counts by size factors in the last
assignment.</p>
</div>
<div id="deseq2-preparation" class="section level2 unnumbered">
<h2>DESeq2 preparation<a class="anchor" aria-label="anchor" href="#deseq2-preparation"><i class="fas fa-link"></i></a>
</h2>
<p><a href="http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#count-matrix-input">DESeq2 Vignette - Preparation</a></p>
<p>The DESeqDataSet is the object that holds the read counts and associated
statistical measures generated through the DESeq2 algorithm. Although there are
several ways to generate this object, we will focus on the DESeqDataSet method
which directly takes a SummarizedExperiment object.</p>
</div>
<div id="omeara-et-al.-transcriptional-reversion-of-cardiac-myocyte-fate-during-mammalian-cardiac-regeneration.-circ-res.-feb-2015.-pmid-25477501l" class="section level2 unnumbered">
<h2>O’Meara et al. Transcriptional Reversion of Cardiac Myocyte Fate During Mammalian Cardiac Regeneration. Circ Res. Feb 2015. PMID: 25477501l<a class="anchor" aria-label="anchor" href="#omeara-et-al.-transcriptional-reversion-of-cardiac-myocyte-fate-during-mammalian-cardiac-regeneration.-circ-res.-feb-2015.-pmid-25477501l"><i class="fas fa-link"></i></a>
</h2>
<p>The counts matrix provided was generated by data made available by the listed
publication. We have only taken a subset of their samples, focusing specifically
on the mRNAseq experiment during <em>in vivo</em> heart development from 0,
4, 7 days after birth as well as in adult mice.</p>
<p>To start, we are going to consider the simplest use case for differential
expression and subset our data to contain only the samples from postnatal day 0
and adult hearts. In addition to subsetting the counts matrix, we also need to
construct the sample data information as a dataframe that lists the samplenames
(columns in our counts matrix) as well as their associated information (in our
case, timepoint).</p>
</div>
<div id="reading-and-subsetting-the-data-from-verse_counts.tsv-and-sample_metadata.csv" class="section level2 unnumbered">
<h2>1. Reading and subsetting the data from verse_counts.tsv and sample_metadata.csv<a class="anchor" aria-label="anchor" href="#reading-and-subsetting-the-data-from-verse_counts.tsv-and-sample_metadata.csv"><i class="fas fa-link"></i></a>
</h2>
<p>We will use this as a demonstration of one of the simplest use cases for DESeq2,
comparing two groups of samples based on a single experimental variable. Subset
the samples to only include those from timepoints vP0 and vAd, which should
correspond to a total of 4 samples (vP0_1, vP0_2, vAd_1, vAd_2). Store both
the counts matrix and sample dataframe in a SummarizedExperiment object.</p>
<div class="box hint">
<ol style="list-style-type: decimal">
<li><p>You may ignore most of the columns in the metadata CSV as they were used in
the generation of the data itself. Focus on the samplenames and timepoints
columns.</p></li>
<li><p>The columns (sample names) in your counts matrix and sample dataframe need
to be in the same order</p></li>
<li><p>You may need to convert your counts dataframe to a matrix</p></li>
<li><p>The counts matrix is verse_counts.tsv and the sample metadata is
sample_metadata.csv</p></li>
</ol>
</div>
</div>
<div id="running-deseq2" class="section level2 unnumbered">
<h2>2. Running DESeq2<a class="anchor" aria-label="anchor" href="#running-deseq2"><i class="fas fa-link"></i></a>
</h2>
<div id="understanding-factor-levels" class="section level3 unnumbered">
<h3>Understanding Factor Levels<a class="anchor" aria-label="anchor" href="#understanding-factor-levels"><i class="fas fa-link"></i></a>
</h3>
<p><a href="http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#note-on-factor-levels">DESeq2 Vignette - Factor Levels</a></p>
<p>The factor level in DESeq2 determines which level represents the “control” group
you want to compare against. By default, the reference level for factors will be
chosen alphabetically. You can either manually change and set the reference
factor level or specify a comparison of interest using DESeq2 Contrast() after
performing the first steps of the DE analysis.</p>
</div>
<div id="performing-differential-expression-analysis" class="section level3 unnumbered">
<h3>Performing Differential Expression Analysis<a class="anchor" aria-label="anchor" href="#performing-differential-expression-analysis"><i class="fas fa-link"></i></a>
</h3>
<p>Here you may find the vignette for example usage of DESeq2. Please read the
following section to instruct you on how to perform differential expression
analysis in DESeq2: <a href="http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#differential-expression-analysis">Running
DESeq2</a></p>
<p>We will be running DESeq2 with a model design of <code>~timepoint</code> with vP0 set as
our reference level. This will test for differentially expressed genes in adult
mouse heart ventricles vs. postnatal day 0. Return both the results as a
dataframe and the dds object from running DESeq2 as a list. Assign the results
dataframe and dds object to variables so you can use them later in your Rmd.</p>
</div>
</div>
<div id="annotating-results-to-construct-a-labeled-volcano-plot" class="section level2 unnumbered">
<h2>3. Annotating results to construct a labeled volcano plot<a class="anchor" aria-label="anchor" href="#annotating-results-to-construct-a-labeled-volcano-plot"><i class="fas fa-link"></i></a>
</h2>
<p>Later in the analysis, we will construct a volcano plot. Typically, volcano
plots will label genes by their significance as well as their direction of
change. Convert the results dataframe to a tibble and add a column labeled
<code>volc_plot_status</code> that you can easily input in ggplot2 to color points by the
below criteria.</p>
<div class="box hint">
<ol style="list-style-type: decimal">
<li>The labels will encompass three groups: 1. Positive log fold change and
significant at a given padj threshold, 2. Negative log fold change and
significant at a given padj threshold, 3. Remaining non-significant genes</li>
<li>Ensure your column is named exactly <code>volc_plot_status</code> or the test will not
work without modification.</li>
<li>Have the values for these labels be `UP’, ‘DOWN’, ‘NS’ respectively.</li>
</ol>
</div>
</div>
<div id="diagnostic-plot-of-the-raw-p-values-for-all-genes" class="section level2 unnumbered">
<h2>4. Diagnostic plot of the raw p-values for all genes<a class="anchor" aria-label="anchor" href="#diagnostic-plot-of-the-raw-p-values-for-all-genes"><i class="fas fa-link"></i></a>
</h2>
<p>Separate from differential expression analysis, it is always an important
diagnostic to plot the unadjusted distribution of p-values obtained from an
experiment. Built in to the definition of a p-value is the idea that under the
null distribution and given all other assumptions made are true, p-values follow
a uniform distribution. In general, you should typically observe a peak of
values close to 0 and a roughly uniform distribution as you approach 1. The
values closer to 0 will be a mix of situations where the alternative hypothesis
is true as well as potential false positives. Significant deviations from this
general pattern likely require close examination of your data and the
significance test you are employing. Make a histogram of the raw p-values from
all the genes discovered in the experiment.</p>
</div>
<div id="plotting-the-logfoldchanges-for-differentially-expressed-genes" class="section level2 unnumbered">
<h2>5. Plotting the LogFoldChanges for differentially expressed genes<a class="anchor" aria-label="anchor" href="#plotting-the-logfoldchanges-for-differentially-expressed-genes"><i class="fas fa-link"></i></a>
</h2>
<p>It is often helpful visualize the log2FoldChanges for your differentially
expressed genes to gain a sense for the distribution of up- and downregulated
genes and gain a global view of how genes are changing. Subset your results to
only include genes significant at a padj threshold of &lt; .10. Plot a histogram of
the log2FoldChange values for these genes.</p>
</div>
<div id="the-choice-of-fdr-cutoff-depends-on-cost" class="section level2 unnumbered box note">
<h2>The choice of FDR cutoff depends on cost<a class="anchor" aria-label="anchor" href="#the-choice-of-fdr-cutoff-depends-on-cost"><i class="fas fa-link"></i></a>
</h2>
<p>The choice of a FDR cutoff or any p-value cutoff is always subjective and based
on the objectives of your experiment and the cost of false discoveries. For
example, if your initial sequencing experiment is exploratory and meant to
generate hypotheses to be followed up <em>in vitro</em> or <em>in vivo</em>, it may be more
appropriate to set a permissive cutoff (padj &lt; .20). P-value thresholds are set
on an experiment by experiment basis and should not be adjusted simply to
increase or decrease the number of DE genes retroactively.</p>
</div>
<div id="plotting-the-normalized-counts-of-differentially-expressed-genes" class="section level2 unnumbered">
<h2>6. Plotting the normalized counts of differentially expressed genes<a class="anchor" aria-label="anchor" href="#plotting-the-normalized-counts-of-differentially-expressed-genes"><i class="fas fa-link"></i></a>
</h2>
<p>It can be helpful to visualize the normalized counts for a few differentially
expressed genes as a quick diagnostic that your analysis is working as intended.
Although you will likely not need to do this for every RNAseq analysis, it can
provide confidence in your results and also help you remember the meaning of the
directionality of your fold change values. Make a scatter / jitter plot of the
DESeq2 normalized counts for the top ten significantly differentially expressed
genes ranked by ascending padj</p>
<div class="box hint">
<ol style="list-style-type: decimal">
<li><p>If you are plotting all of the genes on the same plot, you may find it
helpful to take the logarithm of the counts as different genes often have counts
that differ by several orders of magnitude.</p></li>
<li><p>You may find it helpful to add a slight horizontal “jitter” to the points to
aid in visualization for cases where counts for each replicate are similar.</p></li>
<li><p>You have already used DESeq2 to extract normalized counts. There are also
directions available in the vignette.</p></li>
</ol>
</div>
</div>
<div id="volcano-plot-to-visualize-differential-expression-results" class="section level2 unnumbered">
<h2>7. Volcano Plot to visualize differential expression results<a class="anchor" aria-label="anchor" href="#volcano-plot-to-visualize-differential-expression-results"><i class="fas fa-link"></i></a>
</h2>
<p>A volcano plot is a common data plot that is intended to quickly display points
of interest by plotting statistical significance against a magnitude of change.
In the specific case of mRNAseq data, this entails plotting the -log(p-values /
adjusted p-values) against the estimated log fold changes reported by most
differential expression methods.</p>
<p>Statistically significant genes with low p-values will appear higher on the
y-axis of the plot and they will be separated by the magnitude and
directionality of their change (i.e. Upregulated genes with larger fold change
estimates will be further along to the right of the origin of the x-axis, and
downregulated genes will be farther to the left of the origin on the x-axis).
Typically, genes of interest that have additional biological relevance to the
experiment will be directly annotated using domain knowledge.</p>
<p>Make a volcano plot (log2FoldChange vs. -log10(padj)) for all the genes
discovered in the experiment. Use the column generated in part 3 to color the
genes with the appropriate label.</p>
</div>
<div id="running-fgsea-vignette" class="section level2 unnumbered">
<h2>8. Running fgsea <a href="http://bioconductor.org/packages/release/bioc/vignettes/fgsea/inst/doc/fgsea-tutorial.html">vignette</a><a class="anchor" aria-label="anchor" href="#running-fgsea-vignette"><i class="fas fa-link"></i></a>
</h2>
<p>We will be performing a GSEA on the results generated by running DESeq2 using
the fgsea package. We will use log2FoldChange as our ranking metric and test
against all the pathways in the C2 Canonical Pathways gene set collection
provided by MSigDB and already downloaded here (c2.cp.v7.5.1.symbols.gmt). Read
the section in the textbook or the fgsea documentation for the appropriate
formats for the ranked gene list and the gene sets.</p>
<div class="box hint">
<ol style="list-style-type: decimal">
<li><p>You may read in the gene sets in the proper format using GSEABase or fgsea</p></li>
<li><p>You will need to convert the mouse ensembl gene IDs to human HGNC symbols to
match the genes provided in the C2 Canonical Pathways gene sets. If you are
encountering issues, you may need to remove the version number of the annotation.</p></li>
<li><p>As you have discovered, biomaRt can be difficult to work with. If you are
encountering frequent issues, you may consider mapping your gene IDs and saving
the results locally to avoid having to re-run the biomaRt query multiple times.</p></li>
<li><p>You will inevitably have genes with duplicate names and/or genes with
duplicate values for log2FoldChange. Choose a method to remove these values.</p></li>
</ol>
</div>
</div>
<div id="plotting-the-top-ten-positive-nes-and-top-ten-negative-nes-pathways" class="section level2 unnumbered">
<h2>9. Plotting the top ten positive NES and top ten negative NES pathways<a class="anchor" aria-label="anchor" href="#plotting-the-top-ten-positive-nes-and-top-ten-negative-nes-pathways"><i class="fas fa-link"></i></a>
</h2>
<div class="box hint">
<ol style="list-style-type: decimal">
<li><p>Ensure that the pathway labels are visible</p></li>
<li><p>You will probably get warnings about duplicate values or ranks, do your best
to address these but don’t get hung up on trying to find them all.</p></li>
</ol>
</div>
</div>
<div id="references-1" class="section level2 unnumbered">
<h2>References<a class="anchor" aria-label="anchor" href="#references-1"><i class="fas fa-link"></i></a>
</h2>
<p><small> Dobin A, Davis CA, Schlesinger F, Drenkow J, Zaleski C, Jha S, Batut P,
Chaisson M, Gingeras TR. STAR: ultrafast universal RNA-seq aligner.
Bioinformatics. 2013 Jan 1;29(1):15-21. doi: 10.1093/bioinformatics/bts635. Epub
2012 Oct 25. PMID: 23104886; PMCID: PMC3530905.</small></p>
<p>Zhu, Q., Fisher, S.A., Shallcross, J., Kim, J. (Preprint). VERSE: a versatile
and efficient RNA-Seq read counting tool. bioRxiv 053306. <a href="doi:http://dx.doi.org/10.1101/053306" class="uri">doi:http://dx.doi.org/10.1101/053306</a></p>
<p><a href="https://www.gencodegenes.org/mouse/" class="uri">https://www.gencodegenes.org/mouse/</a></p>
<p>Anders, S. &amp; Huber, W. Differential expression analysis for sequence count data.
Genome Biology 11, 1–12 (2010).</p>
<p>Love, M. I., Huber, W. &amp; Anders, S. Moderated estimation of fold change and
dispersion for RNA-seq data with DESeq2. Genome Biology 15, 550 (2014). </p>

</div>
</div>
  <div class="chapter-nav">
<div class="prev"><a href="assignment-4.html">Assignment 4</a></div>
<div class="next"><a href="assignment-6.html">Assignment 6</a></div>
</div></main><div class="col-md-3 col-lg-2 d-none d-md-block sidebar sidebar-chapter">
    <nav id="toc" data-toggle="toc" aria-label="On this page"><h2>On this page</h2>
      <ul class="nav navbar-nav">
<li><a class="nav-link" href="#assignment-5">Assignment 5</a></li>
<li><a class="nav-link" href="#problem-statement-4">Problem Statement</a></li>
<li><a class="nav-link" href="#learning-objectives-3">Learning Objectives</a></li>
<li><a class="nav-link" href="#skill-list-3">Skill List</a></li>
<li><a class="nav-link" href="#deseq2-background">DESeq2 Background</a></li>
<li><a class="nav-link" href="#generating-a-counts-matrix">Generating a counts matrix</a></li>
<li><a class="nav-link" href="#prefiltering-counts-matrix">Prefiltering Counts matrix</a></li>
<li><a class="nav-link" href="#median-of-ratios-normalization">Median-of-ratios normalization</a></li>
<li><a class="nav-link" href="#deseq2-preparation">DESeq2 preparation</a></li>
<li><a class="nav-link" href="#omeara-et-al.-transcriptional-reversion-of-cardiac-myocyte-fate-during-mammalian-cardiac-regeneration.-circ-res.-feb-2015.-pmid-25477501l">O’Meara et al. Transcriptional Reversion of Cardiac Myocyte Fate During Mammalian Cardiac Regeneration. Circ Res. Feb 2015. PMID: 25477501l</a></li>
<li><a class="nav-link" href="#reading-and-subsetting-the-data-from-verse_counts.tsv-and-sample_metadata.csv">1. Reading and subsetting the data from verse_counts.tsv and sample_metadata.csv</a></li>
<li>
<a class="nav-link" href="#running-deseq2">2. Running DESeq2</a><ul class="nav navbar-nav">
<li><a class="nav-link" href="#understanding-factor-levels">Understanding Factor Levels</a></li>
<li><a class="nav-link" href="#performing-differential-expression-analysis">Performing Differential Expression Analysis</a></li>
</ul>
</li>
<li><a class="nav-link" href="#annotating-results-to-construct-a-labeled-volcano-plot">3. Annotating results to construct a labeled volcano plot</a></li>
<li><a class="nav-link" href="#diagnostic-plot-of-the-raw-p-values-for-all-genes">4. Diagnostic plot of the raw p-values for all genes</a></li>
<li><a class="nav-link" href="#plotting-the-logfoldchanges-for-differentially-expressed-genes">5. Plotting the LogFoldChanges for differentially expressed genes</a></li>
<li><a class="nav-link" href="#the-choice-of-fdr-cutoff-depends-on-cost">The choice of FDR cutoff depends on cost</a></li>
<li><a class="nav-link" href="#plotting-the-normalized-counts-of-differentially-expressed-genes">6. Plotting the normalized counts of differentially expressed genes</a></li>
<li><a class="nav-link" href="#volcano-plot-to-visualize-differential-expression-results">7. Volcano Plot to visualize differential expression results</a></li>
<li><a class="nav-link" href="#running-fgsea-vignette">8. Running fgsea vignette</a></li>
<li><a class="nav-link" href="#plotting-the-top-ten-positive-nes-and-top-ten-negative-nes-pathways">9. Plotting the top ten positive NES and top ten negative NES pathways</a></li>
<li><a class="nav-link" href="#references-1">References</a></li>
</ul>

      <div class="book-extra">
        <ul class="list-unstyled">
          
        </ul>
</div>
    </nav>
</div>

</div>
</div> <!-- .container -->

<footer class="bg-primary text-light mt-5"><div class="container"><div class="row">

  <div class="col-12 col-md-6 mt-3">
    <p>"<strong>BF591 - R for Biological Sciences</strong>" was written by . </p>
  </div>

  <div class="col-12 col-md-6 mt-3">
    <p>This book was built by the <a class="text-light" href="https://bookdown.org">bookdown</a> R package.</p>
  </div>

</div></div>
</footer><!-- dynamically load mathjax for compatibility with self-contained --><script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/latest.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script><script type="text/x-mathjax-config">const popovers = document.querySelectorAll('a.footnote-ref[data-toggle="popover"]');
for (let popover of popovers) {
  const div = document.createElement('div');
  div.setAttribute('style', 'position: absolute; top: 0, left:0; width:0, height:0, overflow: hidden; visibility: hidden;');
  div.innerHTML = popover.getAttribute('data-content');

  var has_math = div.querySelector("span.math");
  if (has_math) {
    document.body.appendChild(div);
    MathJax.Hub.Queue(["Typeset", MathJax.Hub, div]);
    MathJax.Hub.Queue(function() {
      popover.setAttribute('data-content', div.innerHTML);
      document.body.removeChild(div);
    })
  }
}
</script>
</body>
</html>
