<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>7 Biology &amp; Bioinformatics | BF591 - R for Biological Sciences</title>
  <meta name="description" content="7 Biology &amp; Bioinformatics | BF591 - R for Biological Sciences" />
  <meta name="generator" content="bookdown 0.24 and GitBook 2.6.7" />

  <meta property="og:title" content="7 Biology &amp; Bioinformatics | BF591 - R for Biological Sciences" />
  <meta property="og:type" content="book" />
  
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="7 Biology &amp; Bioinformatics | BF591 - R for Biological Sciences" />
  
  
  




  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="data-science.html"/>
<link rel="next" href="data-visualization.html"/>
<script src="libs/header-attrs-2.11/header-attrs.js"></script>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.0.1/anchor-sections.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.0.1/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Syllabus</a>
<ul>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#course-schedule"><i class="fa fa-check"></i>Course Schedule</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#course-values-and-policies"><i class="fa fa-check"></i>Course Values and Policies</a></li>
</ul></li>
<li class="chapter" data-level="1" data-path="introduction.html"><a href="introduction.html"><i class="fa fa-check"></i><b>1</b> Introduction</a>
<ul>
<li class="chapter" data-level="1.1" data-path="introduction.html"><a href="introduction.html#who-this-book-is-for"><i class="fa fa-check"></i><b>1.1</b> Who This Book Is For</a></li>
<li class="chapter" data-level="1.2" data-path="introduction.html"><a href="introduction.html#a-note-about-reinventing-the-wheel"><i class="fa fa-check"></i><b>1.2</b> A Note About Reinventing the Wheel</a></li>
<li class="chapter" data-level="1.3" data-path="introduction.html"><a href="introduction.html#sources-and-references"><i class="fa fa-check"></i><b>1.3</b> Sources and References</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="data-bio.html"><a href="data-bio.html"><i class="fa fa-check"></i><b>2</b> Data in Biology</a>
<ul>
<li class="chapter" data-level="2.1" data-path="data-bio.html"><a href="data-bio.html#a-brief-history-of-data-in-molecular-biology"><i class="fa fa-check"></i><b>2.1</b> A Brief History of Data in Molecular Biology</a></li>
<li class="chapter" data-level="2.2" data-path="data-bio.html"><a href="data-bio.html#biology-as-a-mature-data-science"><i class="fa fa-check"></i><b>2.2</b> Biology as a Mature Data Science</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="preliminaries.html"><a href="preliminaries.html"><i class="fa fa-check"></i><b>3</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="3.1" data-path="preliminaries.html"><a href="preliminaries.html#prelim-r"><i class="fa fa-check"></i><b>3.1</b> The R Language</a></li>
<li class="chapter" data-level="3.2" data-path="preliminaries.html"><a href="preliminaries.html#prelim-rstudio"><i class="fa fa-check"></i><b>3.2</b> RStudio</a></li>
<li class="chapter" data-level="3.3" data-path="preliminaries.html"><a href="preliminaries.html#prog-the-r-script"><i class="fa fa-check"></i><b>3.3</b> The R Script</a></li>
<li class="chapter" data-level="3.4" data-path="preliminaries.html"><a href="preliminaries.html#prog-workflow"><i class="fa fa-check"></i><b>3.4</b> The Scripting Workflow</a></li>
<li class="chapter" data-level="3.5" data-path="preliminaries.html"><a href="preliminaries.html#prelim-git"><i class="fa fa-check"></i><b>3.5</b> git + github</a>
<ul>
<li class="chapter" data-level="3.5.1" data-path="preliminaries.html"><a href="preliminaries.html#motivation"><i class="fa fa-check"></i><b>3.5.1</b> Motivation</a></li>
<li class="chapter" data-level="3.5.2" data-path="preliminaries.html"><a href="preliminaries.html#git"><i class="fa fa-check"></i><b>3.5.2</b> git</a></li>
<li class="chapter" data-level="3.5.3" data-path="preliminaries.html"><a href="preliminaries.html#git-hosting-platforms-github"><i class="fa fa-check"></i><b>3.5.3</b> Git hosting platforms (GitHub)</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="prog-basics.html"><a href="prog-basics.html"><i class="fa fa-check"></i><b>4</b> R Programming</a>
<ul>
<li class="chapter" data-level="4.1" data-path="prog-basics.html"><a href="prog-basics.html#before-you-begin"><i class="fa fa-check"></i><b>4.1</b> Before you begin</a></li>
<li class="chapter" data-level="4.2" data-path="prog-basics.html"><a href="prog-basics.html#introduction-1"><i class="fa fa-check"></i><b>4.2</b> Introduction</a></li>
<li class="chapter" data-level="4.3" data-path="prog-basics.html"><a href="prog-basics.html#prog-r-syntax"><i class="fa fa-check"></i><b>4.3</b> R Syntax Basics</a></li>
<li class="chapter" data-level="4.4" data-path="prog-basics.html"><a href="prog-basics.html#prog-types"><i class="fa fa-check"></i><b>4.4</b> Basic Types of Values</a></li>
<li class="chapter" data-level="4.5" data-path="prog-basics.html"><a href="prog-basics.html#prog-struct"><i class="fa fa-check"></i><b>4.5</b> Data Structures</a>
<ul>
<li class="chapter" data-level="4.5.1" data-path="prog-basics.html"><a href="prog-basics.html#vectors"><i class="fa fa-check"></i><b>4.5.1</b> Vectors</a></li>
<li class="chapter" data-level="4.5.2" data-path="prog-basics.html"><a href="prog-basics.html#matrices"><i class="fa fa-check"></i><b>4.5.2</b> Matrices</a></li>
<li class="chapter" data-level="4.5.3" data-path="prog-basics.html"><a href="prog-basics.html#lists-and-data-frames"><i class="fa fa-check"></i><b>4.5.3</b> Lists and data frames</a></li>
</ul></li>
<li class="chapter" data-level="4.6" data-path="prog-basics.html"><a href="prog-basics.html#logical-tests-and-comparators"><i class="fa fa-check"></i><b>4.6</b> Logical Tests and Comparators</a></li>
<li class="chapter" data-level="4.7" data-path="prog-basics.html"><a href="prog-basics.html#functions"><i class="fa fa-check"></i><b>4.7</b> Functions</a>
<ul>
<li class="chapter" data-level="4.7.1" data-path="prog-basics.html"><a href="prog-basics.html#prog-dry"><i class="fa fa-check"></i><b>4.7.1</b> DRY: Don’t Repeat Yourself</a></li>
<li class="chapter" data-level="4.7.2" data-path="prog-basics.html"><a href="prog-basics.html#writing-your-own-functions"><i class="fa fa-check"></i><b>4.7.2</b> Writing your own functions</a></li>
<li class="chapter" data-level="4.7.3" data-path="prog-basics.html"><a href="prog-basics.html#scope"><i class="fa fa-check"></i><b>4.7.3</b> Scope</a></li>
</ul></li>
<li class="chapter" data-level="4.8" data-path="prog-basics.html"><a href="prog-basics.html#iteration"><i class="fa fa-check"></i><b>4.8</b> Iteration</a>
<ul>
<li class="chapter" data-level="4.8.1" data-path="prog-basics.html"><a href="prog-basics.html#vectorized-operations"><i class="fa fa-check"></i><b>4.8.1</b> Vectorized operations</a></li>
<li class="chapter" data-level="4.8.2" data-path="prog-basics.html"><a href="prog-basics.html#functional-programming"><i class="fa fa-check"></i><b>4.8.2</b> Functional programming</a></li>
<li class="chapter" data-level="4.8.3" data-path="prog-basics.html"><a href="prog-basics.html#apply-and-friends"><i class="fa fa-check"></i><b>4.8.3</b> <code>apply()</code> and friends</a></li>
</ul></li>
<li class="chapter" data-level="4.9" data-path="prog-basics.html"><a href="prog-basics.html#installing-packages"><i class="fa fa-check"></i><b>4.9</b> Installing Packages</a></li>
<li class="chapter" data-level="4.10" data-path="prog-basics.html"><a href="prog-basics.html#prog-debugging"><i class="fa fa-check"></i><b>4.10</b> Troubleshooting and Debugging</a></li>
<li class="chapter" data-level="4.11" data-path="prog-basics.html"><a href="prog-basics.html#prog-style"><i class="fa fa-check"></i><b>4.11</b> Coding Style and Conventions</a>
<ul>
<li class="chapter" data-level="4.11.1" data-path="prog-basics.html"><a href="prog-basics.html#is-my-code-correct"><i class="fa fa-check"></i><b>4.11.1</b> Is my code correct?</a></li>
<li class="chapter" data-level="4.11.2" data-path="prog-basics.html"><a href="prog-basics.html#does-my-code-follow-the-dry-principle"><i class="fa fa-check"></i><b>4.11.2</b> Does my code follow the <span>DRY principle</span>?</a></li>
<li class="chapter" data-level="4.11.3" data-path="prog-basics.html"><a href="prog-basics.html#did-i-choose-concise-but-descriptive-variable-and-function-names"><i class="fa fa-check"></i><b>4.11.3</b> Did I choose concise but descriptive variable and function names?</a></li>
<li class="chapter" data-level="4.11.4" data-path="prog-basics.html"><a href="prog-basics.html#did-i-use-indentation-and-naming-conventions-consistently-throughout-my-code"><i class="fa fa-check"></i><b>4.11.4</b> Did I use indentation and naming conventions consistently throughout my code?</a></li>
<li class="chapter" data-level="4.11.5" data-path="prog-basics.html"><a href="prog-basics.html#did-i-write-comments-especially-when-what-the-code-does-is-not-obvious"><i class="fa fa-check"></i><b>4.11.5</b> Did I write comments, especially when what the code does is not obvious?</a></li>
<li class="chapter" data-level="4.11.6" data-path="prog-basics.html"><a href="prog-basics.html#how-easy-would-it-be-for-someone-else-to-understand-my-code"><i class="fa fa-check"></i><b>4.11.6</b> How easy would it be for someone else to understand my code?</a></li>
<li class="chapter" data-level="4.11.7" data-path="prog-basics.html"><a href="prog-basics.html#is-my-code-easy-to-maintainchange"><i class="fa fa-check"></i><b>4.11.7</b> Is my code easy to maintain/change?</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="5" data-path="data-wrangling.html"><a href="data-wrangling.html"><i class="fa fa-check"></i><b>5</b> Data Wrangling</a>
<ul>
<li class="chapter" data-level="5.1" data-path="data-wrangling.html"><a href="data-wrangling.html#the-tidyverse"><i class="fa fa-check"></i><b>5.1</b> The Tidyverse</a></li>
<li class="chapter" data-level="5.2" data-path="data-wrangling.html"><a href="data-wrangling.html#dw-basics"><i class="fa fa-check"></i><b>5.2</b> Tidyverse Basics</a></li>
<li class="chapter" data-level="5.3" data-path="data-wrangling.html"><a href="data-wrangling.html#importing-data"><i class="fa fa-check"></i><b>5.3</b> Importing Data</a></li>
<li class="chapter" data-level="5.4" data-path="data-wrangling.html"><a href="data-wrangling.html#data-tibble"><i class="fa fa-check"></i><b>5.4</b> The tibble</a></li>
<li class="chapter" data-level="5.5" data-path="data-wrangling.html"><a href="data-wrangling.html#pipes"><i class="fa fa-check"></i><b>5.5</b> pipes</a></li>
<li class="chapter" data-level="5.6" data-path="data-wrangling.html"><a href="data-wrangling.html#tidying-data"><i class="fa fa-check"></i><b>5.6</b> Tidying Data</a></li>
<li class="chapter" data-level="5.7" data-path="data-wrangling.html"><a href="data-wrangling.html#arranging-data"><i class="fa fa-check"></i><b>5.7</b> Arranging Data</a>
<ul>
<li class="chapter" data-level="5.7.1" data-path="data-wrangling.html"><a href="data-wrangling.html#dplyrmutate---create-new-columns-using-other-columns"><i class="fa fa-check"></i><b>5.7.1</b> <code>dplyr::mutate()</code> - Create new columns using other columns</a></li>
<li class="chapter" data-level="5.7.2" data-path="data-wrangling.html"><a href="data-wrangling.html#stringr---working-with-character-values"><i class="fa fa-check"></i><b>5.7.2</b> <code>stringr</code> - Working with character values</a></li>
<li class="chapter" data-level="5.7.3" data-path="data-wrangling.html"><a href="data-wrangling.html#dplyrselect---subset-columns-by-name"><i class="fa fa-check"></i><b>5.7.3</b> <code>dplyr::select()</code> - Subset Columns by Name</a></li>
<li class="chapter" data-level="5.7.4" data-path="data-wrangling.html"><a href="data-wrangling.html#dplyrfilter---pick-rows-out-of-a-data-set"><i class="fa fa-check"></i><b>5.7.4</b> <code>dplyr::filter()</code> - Pick rows out of a data set</a></li>
<li class="chapter" data-level="5.7.5" data-path="data-wrangling.html"><a href="data-wrangling.html#dplyrarrange---order-rows-based-on-their-values"><i class="fa fa-check"></i><b>5.7.5</b> <code>dplyr::arrange()</code> - Order rows based on their values</a></li>
<li class="chapter" data-level="5.7.6" data-path="data-wrangling.html"><a href="data-wrangling.html#putting-it-all-together"><i class="fa fa-check"></i><b>5.7.6</b> Putting it all together</a></li>
</ul></li>
<li class="chapter" data-level="5.8" data-path="data-wrangling.html"><a href="data-wrangling.html#grouping-data"><i class="fa fa-check"></i><b>5.8</b> Grouping Data</a></li>
<li class="chapter" data-level="5.9" data-path="data-wrangling.html"><a href="data-wrangling.html#relational-data"><i class="fa fa-check"></i><b>5.9</b> Relational Data</a>
<ul>
<li class="chapter" data-level="5.9.1" data-path="data-wrangling.html"><a href="data-wrangling.html#dealing-with-multiple-matches"><i class="fa fa-check"></i><b>5.9.1</b> Dealing with multiple matches</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="6" data-path="data-science.html"><a href="data-science.html"><i class="fa fa-check"></i><b>6</b> Data Science</a>
<ul>
<li class="chapter" data-level="6.1" data-path="data-science.html"><a href="data-science.html#ds-dists"><i class="fa fa-check"></i><b>6.1</b> Distributions &amp; Tests</a></li>
<li class="chapter" data-level="6.2" data-path="data-science.html"><a href="data-science.html#ds-summ"><i class="fa fa-check"></i><b>6.2</b> Data summarization &amp; cleaning</a></li>
<li class="chapter" data-level="6.3" data-path="data-science.html"><a href="data-science.html#ds-clust"><i class="fa fa-check"></i><b>6.3</b> Clustering</a></li>
<li class="chapter" data-level="6.4" data-path="data-science.html"><a href="data-science.html#ds-model"><i class="fa fa-check"></i><b>6.4</b> Data Modeling</a></li>
<li class="chapter" data-level="6.5" data-path="data-science.html"><a href="data-science.html#ds-networks"><i class="fa fa-check"></i><b>6.5</b> Network Analysis</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html"><i class="fa fa-check"></i><b>7</b> Biology &amp; Bioinformatics</a>
<ul>
<li class="chapter" data-level="7.1" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#bio-r"><i class="fa fa-check"></i><b>7.1</b> R in Biology</a></li>
<li class="chapter" data-level="7.2" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#csv-files"><i class="fa fa-check"></i><b>7.2</b> CSV Files</a></li>
<li class="chapter" data-level="7.3" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#bio-data-types"><i class="fa fa-check"></i><b>7.3</b> Types of Biological Data</a></li>
<li class="chapter" data-level="7.4" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#bio-bioconductor"><i class="fa fa-check"></i><b>7.4</b> Bioconductor</a></li>
<li class="chapter" data-level="7.5" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#bio-geneexp"><i class="fa fa-check"></i><b>7.5</b> Gene Expression</a></li>
<li class="chapter" data-level="7.6" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#bio-micro"><i class="fa fa-check"></i><b>7.6</b> Differential Expression: Microarrays</a></li>
<li class="chapter" data-level="7.7" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#bio-gsea"><i class="fa fa-check"></i><b>7.7</b> Gene Set Enrichment Analysis</a></li>
<li class="chapter" data-level="7.8" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#bio-hts"><i class="fa fa-check"></i><b>7.8</b> High Throughput Sequencing</a></li>
<li class="chapter" data-level="7.9" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#bio-rnaseq"><i class="fa fa-check"></i><b>7.9</b> Differential Expression: RNASeq</a></li>
<li class="chapter" data-level="7.10" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#bio-networks"><i class="fa fa-check"></i><b>7.10</b> Biological Network Analysis</a></li>
<li class="chapter" data-level="7.11" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#bio-scseq"><i class="fa fa-check"></i><b>7.11</b> Single Cell Sequencing Analysis</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="data-visualization.html"><a href="data-visualization.html"><i class="fa fa-check"></i><b>8</b> Data Visualization</a>
<ul>
<li class="chapter" data-level="8.1" data-path="data-visualization.html"><a href="data-visualization.html#dv-gg"><i class="fa fa-check"></i><b>8.1</b> Grammar of Graphics</a></li>
<li class="chapter" data-level="8.2" data-path="data-visualization.html"><a href="data-visualization.html#plotting-one-dimension"><i class="fa fa-check"></i><b>8.2</b> Plotting One Dimension</a></li>
<li class="chapter" data-level="8.3" data-path="data-visualization.html"><a href="data-visualization.html#plotting-two-or-more-dimensions"><i class="fa fa-check"></i><b>8.3</b> Plotting Two or More Dimensions</a></li>
<li class="chapter" data-level="8.4" data-path="data-visualization.html"><a href="data-visualization.html#visualizing-distributions"><i class="fa fa-check"></i><b>8.4</b> Visualizing Distributions</a></li>
<li class="chapter" data-level="8.5" data-path="data-visualization.html"><a href="data-visualization.html#other-kind-of-plots"><i class="fa fa-check"></i><b>8.5</b> Other Kind of Plots</a></li>
<li class="chapter" data-level="8.6" data-path="data-visualization.html"><a href="data-visualization.html#ds-network-viz"><i class="fa fa-check"></i><b>8.6</b> Network Viz</a></li>
<li class="chapter" data-level="8.7" data-path="data-visualization.html"><a href="data-visualization.html#tips-and-tricks"><i class="fa fa-check"></i><b>8.7</b> Tips and Tricks</a></li>
<li class="chapter" data-level="8.8" data-path="data-visualization.html"><a href="data-visualization.html#multiple-plots"><i class="fa fa-check"></i><b>8.8</b> Multiple Plots</a></li>
<li class="chapter" data-level="8.9" data-path="data-visualization.html"><a href="data-visualization.html#dv-resp"><i class="fa fa-check"></i><b>8.9</b> Responsible Plotting</a></li>
<li class="chapter" data-level="8.10" data-path="data-visualization.html"><a href="data-visualization.html#publication-ready-plots"><i class="fa fa-check"></i><b>8.10</b> Publication Ready Plots</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="engineering.html"><a href="engineering.html"><i class="fa fa-check"></i><b>9</b> EngineeRing</a>
<ul>
<li class="chapter" data-level="9.1" data-path="engineering.html"><a href="engineering.html#unit-testing"><i class="fa fa-check"></i><b>9.1</b> Unit Testing</a></li>
<li class="chapter" data-level="9.2" data-path="engineering.html"><a href="engineering.html#toolification"><i class="fa fa-check"></i><b>9.2</b> Toolification</a></li>
<li class="chapter" data-level="9.3" data-path="engineering.html"><a href="engineering.html#pipelines-workflows"><i class="fa fa-check"></i><b>9.3</b> Pipelines &amp; Workflows</a></li>
<li class="chapter" data-level="9.4" data-path="engineering.html"><a href="engineering.html#parallel-processing"><i class="fa fa-check"></i><b>9.4</b> Parallel Processing</a></li>
<li class="chapter" data-level="9.5" data-path="engineering.html"><a href="engineering.html#r-packages"><i class="fa fa-check"></i><b>9.5</b> R Packages</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="rshiny.html"><a href="rshiny.html"><i class="fa fa-check"></i><b>10</b> RShiny</a>
<ul>
<li class="chapter" data-level="10.1" data-path="rshiny.html"><a href="rshiny.html#overview"><i class="fa fa-check"></i><b>10.1</b> Overview</a></li>
<li class="chapter" data-level="10.2" data-path="rshiny.html"><a href="rshiny.html#learning-objectives"><i class="fa fa-check"></i><b>10.2</b> Learning Objectives</a></li>
<li class="chapter" data-level="10.3" data-path="rshiny.html"><a href="rshiny.html#skill-list"><i class="fa fa-check"></i><b>10.3</b> Skill List</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="communicating-with-r.html"><a href="communicating-with-r.html"><i class="fa fa-check"></i><b>11</b> Communicating with R</a>
<ul>
<li class="chapter" data-level="11.1" data-path="communicating-with-r.html"><a href="communicating-with-r.html#rmarkdown-knitr"><i class="fa fa-check"></i><b>11.1</b> RMarkdown &amp; knitr</a></li>
<li class="chapter" data-level="11.2" data-path="communicating-with-r.html"><a href="communicating-with-r.html#bookdown"><i class="fa fa-check"></i><b>11.2</b> bookdown</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="contribution-guide.html"><a href="contribution-guide.html"><i class="fa fa-check"></i><b>12</b> Contribution Guide</a>
<ul>
<li class="chapter" data-level="12.1" data-path="contribution-guide.html"><a href="contribution-guide.html#custom-blocks"><i class="fa fa-check"></i><b>12.1</b> Custom blocks</a></li>
</ul></li>
<li class="part"><span><b>I Assignments</b></span></li>
<li class="chapter" data-level="" data-path="assignment-format.html"><a href="assignment-format.html"><i class="fa fa-check"></i>Assignment Format</a></li>
<li class="chapter" data-level="" data-path="starting-an-assignment.html"><a href="starting-an-assignment.html"><i class="fa fa-check"></i>Starting an Assignment</a>
<ul>
<li><a href="starting-an-assignment.html#starting-your-own-git-repository-and-cloning-it">Starting your own <code>git</code> repository and cloning it</a>
<ul>
<li class="chapter" data-level="" data-path="starting-an-assignment.html"><a href="starting-an-assignment.html#contents"><i class="fa fa-check"></i>Contents</a></li>
<li class="chapter" data-level="" data-path="starting-an-assignment.html"><a href="starting-an-assignment.html#uhhh-various-github-classroom-details-idk.-todo-or-whatever."><i class="fa fa-check"></i>Uhhh various github classroom details, idk. TODO: or whatever.</a></li>
<li class="chapter" data-level="12.1.1" data-path="starting-an-assignment.html"><a href="starting-an-assignment.html#committing-and-pushing-with-r-and-without"><i class="fa fa-check"></i><b>12.1.1</b> Committing and Pushing with R (and without)</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="assignment-2.html"><a href="assignment-2.html"><i class="fa fa-check"></i>Assignment 2</a>
<ul>
<li class="chapter" data-level="" data-path="assignment-2.html"><a href="assignment-2.html#problem-statement"><i class="fa fa-check"></i>Problem Statement</a></li>
<li class="chapter" data-level="" data-path="assignment-2.html"><a href="assignment-2.html#learning-objectives-1"><i class="fa fa-check"></i>Learning Objectives</a></li>
<li class="chapter" data-level="" data-path="assignment-2.html"><a href="assignment-2.html#skill-list-1"><i class="fa fa-check"></i>Skill List</a></li>
<li class="chapter" data-level="" data-path="assignment-2.html"><a href="assignment-2.html#instructions"><i class="fa fa-check"></i>Instructions</a></li>
<li class="chapter" data-level="" data-path="assignment-2.html"><a href="assignment-2.html#function-details"><i class="fa fa-check"></i>Function Details</a>
<ul>
<li class="chapter" data-level="" data-path="assignment-2.html"><a href="assignment-2.html#bioconductor"><i class="fa fa-check"></i>1. Bioconductor</a></li>
<li><a href="assignment-2.html#load_expression">2. <code>load_expression()</code></a></li>
<li><a href="assignment-2.html#filter_15">3. <code>filter_15()</code></a></li>
<li><a href="assignment-2.html#affy_to_hgnc">4. <code>affy_to_hgnc()</code></a></li>
<li><a href="assignment-2.html#reduce_data">5. <code>reduce_data()</code></a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="assignment-5---deseq2.html"><a href="assignment-5---deseq2.html"><i class="fa fa-check"></i>Assignment 5 - DESeq2</a>
<ul>
<li class="chapter" data-level="" data-path="assignment-5---deseq2.html"><a href="assignment-5---deseq2.html#problem-statement-1"><i class="fa fa-check"></i>Problem Statement</a></li>
<li class="chapter" data-level="" data-path="assignment-5---deseq2.html"><a href="assignment-5---deseq2.html#learning-objectives-2"><i class="fa fa-check"></i>Learning Objectives</a></li>
<li class="chapter" data-level="" data-path="assignment-5---deseq2.html"><a href="assignment-5---deseq2.html#skill-list-2"><i class="fa fa-check"></i>Skill List</a></li>
<li class="chapter" data-level="" data-path="assignment-5---deseq2.html"><a href="assignment-5---deseq2.html#deseq2-background"><i class="fa fa-check"></i>DESeq2 Background</a></li>
<li class="chapter" data-level="" data-path="assignment-5---deseq2.html"><a href="assignment-5---deseq2.html#generating-a-counts-matrix"><i class="fa fa-check"></i>Generating a counts matrix</a></li>
<li class="chapter" data-level="" data-path="assignment-5---deseq2.html"><a href="assignment-5---deseq2.html#prefiltering-counts-matrix"><i class="fa fa-check"></i>Prefiltering Counts matrix</a></li>
<li class="chapter" data-level="" data-path="assignment-5---deseq2.html"><a href="assignment-5---deseq2.html#median-of-ratios-normalization"><i class="fa fa-check"></i>Median-of-ratios normalization</a></li>
<li class="chapter" data-level="" data-path="assignment-5---deseq2.html"><a href="assignment-5---deseq2.html#deseq2-preparation"><i class="fa fa-check"></i>DESeq2 preparation</a></li>
<li class="chapter" data-level="" data-path="assignment-5---deseq2.html"><a href="assignment-5---deseq2.html#omeara-et-al.-transcriptional-reversion-of-cardiac-myocyte-fate-during-mammalian-cardiac-regeneration.-circ-res.-feb-2015.-pmid-25477501l"><i class="fa fa-check"></i>O’Meara et al. Transcriptional Reversion of Cardiac Myocyte Fate During Mammalian Cardiac Regeneration. Circ Res. Feb 2015. PMID: 25477501l</a></li>
<li class="chapter" data-level="" data-path="assignment-5---deseq2.html"><a href="assignment-5---deseq2.html#understanding-factor-levels"><i class="fa fa-check"></i>Understanding Factor Levels</a></li>
<li class="chapter" data-level="" data-path="assignment-5---deseq2.html"><a href="assignment-5---deseq2.html#performing-differential-expression-analysis"><i class="fa fa-check"></i>Performing Differential Expression Analysis</a></li>
<li class="chapter" data-level="" data-path="assignment-5---deseq2.html"><a href="assignment-5---deseq2.html#p-value-distribution-and-multiple-test-correction"><i class="fa fa-check"></i>P-value distribution and multiple test correction</a></li>
<li class="chapter" data-level="" data-path="assignment-5---deseq2.html"><a href="assignment-5---deseq2.html#histogram-of-log-fold-changes"><i class="fa fa-check"></i>Histogram of Log Fold Changes</a></li>
<li class="chapter" data-level="" data-path="assignment-5---deseq2.html"><a href="assignment-5---deseq2.html#make-a-volcano-plot"><i class="fa fa-check"></i>Make a volcano plot</a></li>
<li class="chapter" data-level="" data-path="assignment-5---deseq2.html"><a href="assignment-5---deseq2.html#visualize-normalized-counts-for-the-ten-differentially-expressed-genes-with-the-lowest-adjusted-p-values"><i class="fa fa-check"></i>Visualize normalized counts for the ten differentially expressed genes with the lowest adjusted p-values</a></li>
<li class="chapter" data-level="" data-path="assignment-5---deseq2.html"><a href="assignment-5---deseq2.html#time-series-with-deseq2-likelihood-ratio-test-lrt"><i class="fa fa-check"></i>Time series with DESeq2 Likelihood Ratio Test (LRT)</a></li>
<li class="chapter" data-level="" data-path="assignment-5---deseq2.html"><a href="assignment-5---deseq2.html#plotting-the-normalized-counts-over-time-for-the-top-differentially-expressed-gene-from-the-lrt-analysis"><i class="fa fa-check"></i>Plotting the normalized counts over time for the top differentially expressed gene from the LRT analysis</a></li>
<li class="chapter" data-level="" data-path="assignment-5---deseq2.html"><a href="assignment-5---deseq2.html#optional-exercises"><i class="fa fa-check"></i>Optional exercises</a></li>
<li class="chapter" data-level="" data-path="assignment-5---deseq2.html"><a href="assignment-5---deseq2.html#references"><i class="fa fa-check"></i>References</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="assignment-6.html"><a href="assignment-6.html"><i class="fa fa-check"></i>Assignment 6</a>
<ul>
<li class="chapter" data-level="" data-path="assignment-6.html"><a href="assignment-6.html#problem-statement-2"><i class="fa fa-check"></i>Problem Statement</a></li>
<li class="chapter" data-level="" data-path="assignment-6.html"><a href="assignment-6.html#learning-objectives-3"><i class="fa fa-check"></i>Learning Objectives</a></li>
<li class="chapter" data-level="" data-path="assignment-6.html"><a href="assignment-6.html#skill-list-3"><i class="fa fa-check"></i>Skill List</a></li>
<li class="chapter" data-level="" data-path="assignment-6.html"><a href="assignment-6.html#instructions-1"><i class="fa fa-check"></i>Instructions</a></li>
</ul></li>
<li class="appendix"><span><b>Appendix</b></span></li>
<li class="chapter" data-level="A" data-path="class-outlines.html"><a href="class-outlines.html"><i class="fa fa-check"></i><b>A</b> Class Outlines</a>
<ul>
<li class="chapter" data-level="A.1" data-path="class-outlines.html"><a href="class-outlines.html#week-1"><i class="fa fa-check"></i><b>A.1</b> Week 1</a></li>
<li class="chapter" data-level="A.2" data-path="class-outlines.html"><a href="class-outlines.html#week-2"><i class="fa fa-check"></i><b>A.2</b> Week 2</a></li>
</ul></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">BF591 - R for Biological Sciences</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="biology-bioinformatics" class="section level1" number="7">
<h1><span class="header-section-number">7</span> Biology &amp; Bioinformatics</h1>
<div id="bio-r" class="section level2" number="7.1">
<h2><span class="header-section-number">7.1</span> R in Biology</h2>
</div>
<div id="csv-files" class="section level2" number="7.2">
<h2><span class="header-section-number">7.2</span> CSV Files</h2>
<p>The most common, convenient, and flexible data file format in biology and
bioinformatics is the <em>character-delimited</em> or <em>character-separated text file</em>.
These files are plain text files (i.e. not the native file format of any
specific program, like Excel) that generally contain rectangles of data. When
formatted correctly, you can think of these files as containing a grid or matrix
of data values with some number of columns, each of which has the same number of
values. Each line of these files has some number of data values separated by a
consistent character, most commonly the comma which are called <a href="https://en.wikipedia.org/wiki/Comma-separated_values">comma-separated
value, or “CSV”, files</a>
and filenames typically end with the extension <code>.csv</code>. Note that
other characters, especially the <Tab> character, may be used to create valid
files in this format, and all the same general principles apply. This is an
example of a simple CSV file:</p>
<pre><code>id,somevalue,category,genes
1,2.123,A,APOE
4,5.123,B,&quot;HOXA1,HOXB1&quot;
7,8.123,,SNCA</code></pre>
<p>Some properties and principles of CSV files:</p>
<ul>
<li>The first line often but not always contains the column names of each column</li>
<li>Each value is delimited by the same character, in this case <code>,</code></li>
<li>Values can be any value, including numbers and characters</li>
<li>When a value contains the delimiting character (e.g. HOXA1,HOXB1 contains a
<code>,</code>), the value is wrapped in double quotes</li>
<li>Values can be missing, indicated by sequential delimiters (i.e. <code>,,</code> or one
<code>,</code> at the end of the line, if the last column value is missing)</li>
<li>There is no delimiter at the end of the lines</li>
<li>To be well-formatted <em>every line must have the same number of delimited
values</em></li>
</ul>
<p>These same properties and principles apply to all character-separated files,
regardless of the specific delimiter used. If a file follows these principles,
they can be loaded very easily into R or any other data analysis setting.</p>
</div>
<div id="bio-data-types" class="section level2" number="7.3">
<h2><span class="header-section-number">7.3</span> Types of Biological Data</h2>
<p>In very general terms, there are five types of data used in biological data
analysis: raw/primary data, processed data, analysis results, metadata, and
annotation data.</p>
<p><strong>Raw/primary data.</strong> These data are the primary observations made by whatever
instruments/techniques we are using for our experiment. These may include
high-throughput sequencing data, mass/charge ratio data from mass spectrometry,
16S rRNA sequencing data from metagenomic studies, SNPs from genotyping assays,
etc. These data can be very large and are often not efficiently processed using
R. Instead, specialized tools built outside of R are used to first process the
primary data into a form that is amenable to analysis.</p>
<p><strong>Processed data.</strong> Processed data is the result of any analysis or
transformation of primary data into an intermediate, more interpretable form.
For example, when performing gene expression analysis with RNASeq, short reads
that form the primary data are typically aligned against a genome and then
counted against annotated genes, resulting in a count for each gene in the
genome. Typically, these counts are then combined for many samples into a single
matrix and subject to downstream analyses like differential expression.</p>
<p><strong>Analysis results.</strong> Analysis results aren’t data <em>per se</em>, but are the results
of analysis of primary data or processed results. For example, in gene
expression studies, a first analysis is often differential expression, where
each gene in the genome is tested for expression associated with some outcome
of interest across many samples. Each gene then has a number of statistics and
related values, like log2 fold change, nominal and adjusted p-value, etc. These
forms of data are usually what we use to form interpretations of our datasets
and therefore we must manipulate them in much the same way as any other dataset.</p>
<p><strong>Metadata.</strong> In most experimental settings, multiple samples have been
processed and had the same primary data type collected. These samples also have
information associated with them, which is here termed <em>metadata</em>, or “data
about data”. In our gene expression of post mortem brain experiments mentioned
above, the information about the individuals themselves, including age at death,
whether the person had a disease, the measurements of tissue quality, etc. is
the metadata. The primary and processed data and metadata are usually stored in
different files, where the metadata (or <em>sample information</em> or <em>sample data</em>,
etc) will have one column indicating the unique identifier (ID) of each sample.
The processed data will typically have columns named for each of the sample IDs.</p>
<p><strong>Annotation data.</strong> A tremendous amount of knowledge has been generated about
biological entities, e.g. genes, especially since the publication of the human
genome. Annotation data is publicly available information about the features we
measure in our experiments. This may come in the form of the coordinates in a
genome where genes exist, any known functions of those genes, the domains found
in proteins and their relative sequence, gene identifier cross references across
different gene naming systems (e.g. symbol vs Ensembl ID), single nucleotide
polymorphism genomic locations and associations with traits or diseases, etc.
This is information that we use to aid in interpretation of our experimental
data but generally do not generate ourselves. Annotation data comes in many
forms, some of which are in CSV format.</p>
</div>
<div id="bio-bioconductor" class="section level2" number="7.4">
<h2><span class="header-section-number">7.4</span> Bioconductor</h2>
</div>
<div id="bio-geneexp" class="section level2" number="7.5">
<h2><span class="header-section-number">7.5</span> Gene Expression</h2>
</div>
<div id="bio-micro" class="section level2" number="7.6">
<h2><span class="header-section-number">7.6</span> Differential Expression: Microarrays</h2>
</div>
<div id="bio-gsea" class="section level2" number="7.7">
<h2><span class="header-section-number">7.7</span> Gene Set Enrichment Analysis</h2>
</div>
<div id="bio-hts" class="section level2" number="7.8">
<h2><span class="header-section-number">7.8</span> High Throughput Sequencing</h2>
</div>
<div id="bio-rnaseq" class="section level2" number="7.9">
<h2><span class="header-section-number">7.9</span> Differential Expression: RNASeq</h2>
</div>
<div id="bio-networks" class="section level2" number="7.10">
<h2><span class="header-section-number">7.10</span> Biological Network Analysis</h2>
</div>
<div id="bio-scseq" class="section level2" number="7.11">
<h2><span class="header-section-number">7.11</span> Single Cell Sequencing Analysis</h2>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="data-science.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="data-visualization.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
