<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>8 Biology &amp; Bioinformatics | BF591 - R for Biological Sciences</title>
  <meta name="description" content="8 Biology &amp; Bioinformatics | BF591 - R for Biological Sciences" />
  <meta name="generator" content="bookdown 0.24 and GitBook 2.6.7" />

  <meta property="og:title" content="8 Biology &amp; Bioinformatics | BF591 - R for Biological Sciences" />
  <meta property="og:type" content="book" />
  
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="8 Biology &amp; Bioinformatics | BF591 - R for Biological Sciences" />
  
  
  




  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="data-visualization.html"/>
<link rel="next" href="engineering.html"/>
<script src="libs/header-attrs-2.11/header-attrs.js"></script>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.0.1/anchor-sections.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.0.1/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Syllabus</a>
<ul>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#course-schedule"><i class="fa fa-check"></i>Course Schedule</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#instructors"><i class="fa fa-check"></i>Instructors</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#office-hours"><i class="fa fa-check"></i>Office Hours</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#course-values-and-policies"><i class="fa fa-check"></i>Course Values and Policies</a></li>
</ul></li>
<li class="chapter" data-level="1" data-path="introduction.html"><a href="introduction.html"><i class="fa fa-check"></i><b>1</b> Introduction</a>
<ul>
<li class="chapter" data-level="1.1" data-path="introduction.html"><a href="introduction.html#who-this-book-is-for"><i class="fa fa-check"></i><b>1.1</b> Who This Book Is For</a></li>
<li class="chapter" data-level="1.2" data-path="introduction.html"><a href="introduction.html#a-note-about-reinventing-the-wheel"><i class="fa fa-check"></i><b>1.2</b> A Note About Reinventing the Wheel</a></li>
<li class="chapter" data-level="1.3" data-path="introduction.html"><a href="introduction.html#sources-and-references"><i class="fa fa-check"></i><b>1.3</b> Sources and References</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="data-bio.html"><a href="data-bio.html"><i class="fa fa-check"></i><b>2</b> Data in Biology</a>
<ul>
<li class="chapter" data-level="2.1" data-path="data-bio.html"><a href="data-bio.html#a-brief-history-of-data-in-molecular-biology"><i class="fa fa-check"></i><b>2.1</b> A Brief History of Data in Molecular Biology</a></li>
<li class="chapter" data-level="2.2" data-path="data-bio.html"><a href="data-bio.html#biology-as-a-mature-data-science"><i class="fa fa-check"></i><b>2.2</b> Biology as a Mature Data Science</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="preliminaries.html"><a href="preliminaries.html"><i class="fa fa-check"></i><b>3</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="3.1" data-path="preliminaries.html"><a href="preliminaries.html#prelim-r"><i class="fa fa-check"></i><b>3.1</b> The R Language</a></li>
<li class="chapter" data-level="3.2" data-path="preliminaries.html"><a href="preliminaries.html#prelim-rstudio"><i class="fa fa-check"></i><b>3.2</b> RStudio</a></li>
<li class="chapter" data-level="3.3" data-path="preliminaries.html"><a href="preliminaries.html#prog-the-r-script"><i class="fa fa-check"></i><b>3.3</b> The R Script</a></li>
<li class="chapter" data-level="3.4" data-path="preliminaries.html"><a href="preliminaries.html#prog-workflow"><i class="fa fa-check"></i><b>3.4</b> The Scripting Workflow</a></li>
<li class="chapter" data-level="3.5" data-path="preliminaries.html"><a href="preliminaries.html#prelim-git"><i class="fa fa-check"></i><b>3.5</b> git + github</a>
<ul>
<li class="chapter" data-level="3.5.1" data-path="preliminaries.html"><a href="preliminaries.html#motivation"><i class="fa fa-check"></i><b>3.5.1</b> Motivation</a></li>
<li class="chapter" data-level="3.5.2" data-path="preliminaries.html"><a href="preliminaries.html#git"><i class="fa fa-check"></i><b>3.5.2</b> git</a></li>
<li class="chapter" data-level="3.5.3" data-path="preliminaries.html"><a href="preliminaries.html#git-hosting-platforms-github"><i class="fa fa-check"></i><b>3.5.3</b> Git hosting platforms (GitHub)</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="prog-basics.html"><a href="prog-basics.html"><i class="fa fa-check"></i><b>4</b> R Programming</a>
<ul>
<li class="chapter" data-level="4.1" data-path="prog-basics.html"><a href="prog-basics.html#before-you-begin"><i class="fa fa-check"></i><b>4.1</b> Before you begin</a></li>
<li class="chapter" data-level="4.2" data-path="prog-basics.html"><a href="prog-basics.html#introduction-1"><i class="fa fa-check"></i><b>4.2</b> Introduction</a></li>
<li class="chapter" data-level="4.3" data-path="prog-basics.html"><a href="prog-basics.html#prog-r-syntax"><i class="fa fa-check"></i><b>4.3</b> R Syntax Basics</a></li>
<li class="chapter" data-level="4.4" data-path="prog-basics.html"><a href="prog-basics.html#prog-types"><i class="fa fa-check"></i><b>4.4</b> Basic Types of Values</a>
<ul>
<li class="chapter" data-level="4.4.1" data-path="prog-basics.html"><a href="prog-basics.html#factors"><i class="fa fa-check"></i><b>4.4.1</b> Factors</a></li>
</ul></li>
<li class="chapter" data-level="4.5" data-path="prog-basics.html"><a href="prog-basics.html#prog-struct"><i class="fa fa-check"></i><b>4.5</b> Data Structures</a>
<ul>
<li class="chapter" data-level="4.5.1" data-path="prog-basics.html"><a href="prog-basics.html#vectors"><i class="fa fa-check"></i><b>4.5.1</b> Vectors</a></li>
<li class="chapter" data-level="4.5.2" data-path="prog-basics.html"><a href="prog-basics.html#matrices"><i class="fa fa-check"></i><b>4.5.2</b> Matrices</a></li>
<li class="chapter" data-level="4.5.3" data-path="prog-basics.html"><a href="prog-basics.html#lists-and-data-frames"><i class="fa fa-check"></i><b>4.5.3</b> Lists and data frames</a></li>
</ul></li>
<li class="chapter" data-level="4.6" data-path="prog-basics.html"><a href="prog-basics.html#logical-tests-and-comparators"><i class="fa fa-check"></i><b>4.6</b> Logical Tests and Comparators</a></li>
<li class="chapter" data-level="4.7" data-path="prog-basics.html"><a href="prog-basics.html#functions"><i class="fa fa-check"></i><b>4.7</b> Functions</a>
<ul>
<li class="chapter" data-level="4.7.1" data-path="prog-basics.html"><a href="prog-basics.html#prog-dry"><i class="fa fa-check"></i><b>4.7.1</b> DRY: Don’t Repeat Yourself</a></li>
<li class="chapter" data-level="4.7.2" data-path="prog-basics.html"><a href="prog-basics.html#writing-your-own-functions"><i class="fa fa-check"></i><b>4.7.2</b> Writing your own functions</a></li>
<li class="chapter" data-level="4.7.3" data-path="prog-basics.html"><a href="prog-basics.html#scope"><i class="fa fa-check"></i><b>4.7.3</b> Scope</a></li>
</ul></li>
<li class="chapter" data-level="4.8" data-path="prog-basics.html"><a href="prog-basics.html#iteration"><i class="fa fa-check"></i><b>4.8</b> Iteration</a>
<ul>
<li class="chapter" data-level="4.8.1" data-path="prog-basics.html"><a href="prog-basics.html#vectorized-operations"><i class="fa fa-check"></i><b>4.8.1</b> Vectorized operations</a></li>
<li class="chapter" data-level="4.8.2" data-path="prog-basics.html"><a href="prog-basics.html#functional-programming"><i class="fa fa-check"></i><b>4.8.2</b> Functional programming</a></li>
<li class="chapter" data-level="4.8.3" data-path="prog-basics.html"><a href="prog-basics.html#apply-and-friends"><i class="fa fa-check"></i><b>4.8.3</b> <code>apply()</code> and friends</a></li>
</ul></li>
<li class="chapter" data-level="4.9" data-path="prog-basics.html"><a href="prog-basics.html#installing-packages"><i class="fa fa-check"></i><b>4.9</b> Installing Packages</a></li>
<li class="chapter" data-level="4.10" data-path="prog-basics.html"><a href="prog-basics.html#saving-and-loading-r-data"><i class="fa fa-check"></i><b>4.10</b> Saving and Loading R Data</a></li>
<li class="chapter" data-level="4.11" data-path="prog-basics.html"><a href="prog-basics.html#prog-debugging"><i class="fa fa-check"></i><b>4.11</b> Troubleshooting and Debugging</a></li>
<li class="chapter" data-level="4.12" data-path="prog-basics.html"><a href="prog-basics.html#prog-style"><i class="fa fa-check"></i><b>4.12</b> Coding Style and Conventions</a>
<ul>
<li class="chapter" data-level="4.12.1" data-path="prog-basics.html"><a href="prog-basics.html#is-my-code-correct"><i class="fa fa-check"></i><b>4.12.1</b> Is my code correct?</a></li>
<li class="chapter" data-level="4.12.2" data-path="prog-basics.html"><a href="prog-basics.html#does-my-code-follow-the-dry-principle"><i class="fa fa-check"></i><b>4.12.2</b> Does my code follow the <span>DRY principle</span>?</a></li>
<li class="chapter" data-level="4.12.3" data-path="prog-basics.html"><a href="prog-basics.html#did-i-choose-concise-but-descriptive-variable-and-function-names"><i class="fa fa-check"></i><b>4.12.3</b> Did I choose concise but descriptive variable and function names?</a></li>
<li class="chapter" data-level="4.12.4" data-path="prog-basics.html"><a href="prog-basics.html#did-i-use-indentation-and-naming-conventions-consistently-throughout-my-code"><i class="fa fa-check"></i><b>4.12.4</b> Did I use indentation and naming conventions consistently throughout my code?</a></li>
<li class="chapter" data-level="4.12.5" data-path="prog-basics.html"><a href="prog-basics.html#did-i-write-comments-especially-when-what-the-code-does-is-not-obvious"><i class="fa fa-check"></i><b>4.12.5</b> Did I write comments, especially when what the code does is not obvious?</a></li>
<li class="chapter" data-level="4.12.6" data-path="prog-basics.html"><a href="prog-basics.html#how-easy-would-it-be-for-someone-else-to-understand-my-code"><i class="fa fa-check"></i><b>4.12.6</b> How easy would it be for someone else to understand my code?</a></li>
<li class="chapter" data-level="4.12.7" data-path="prog-basics.html"><a href="prog-basics.html#is-my-code-easy-to-maintainchange"><i class="fa fa-check"></i><b>4.12.7</b> Is my code easy to maintain/change?</a></li>
<li class="chapter" data-level="4.12.8" data-path="prog-basics.html"><a href="prog-basics.html#the-styler-package-."><i class="fa fa-check"></i><b>4.12.8</b> The <code>styler</code> package .</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="5" data-path="data-wrangling.html"><a href="data-wrangling.html"><i class="fa fa-check"></i><b>5</b> Data Wrangling</a>
<ul>
<li class="chapter" data-level="5.1" data-path="data-wrangling.html"><a href="data-wrangling.html#the-tidyverse"><i class="fa fa-check"></i><b>5.1</b> The Tidyverse</a></li>
<li class="chapter" data-level="5.2" data-path="data-wrangling.html"><a href="data-wrangling.html#dw-basics"><i class="fa fa-check"></i><b>5.2</b> Tidyverse Basics</a></li>
<li class="chapter" data-level="5.3" data-path="data-wrangling.html"><a href="data-wrangling.html#importing-data"><i class="fa fa-check"></i><b>5.3</b> Importing Data</a></li>
<li class="chapter" data-level="5.4" data-path="data-wrangling.html"><a href="data-wrangling.html#data-tibble"><i class="fa fa-check"></i><b>5.4</b> The tibble</a></li>
<li class="chapter" data-level="5.5" data-path="data-wrangling.html"><a href="data-wrangling.html#tidy-data"><i class="fa fa-check"></i><b>5.5</b> Tidy Data</a></li>
<li class="chapter" data-level="5.6" data-path="data-wrangling.html"><a href="data-wrangling.html#pipes"><i class="fa fa-check"></i><b>5.6</b> pipes</a></li>
<li class="chapter" data-level="5.7" data-path="data-wrangling.html"><a href="data-wrangling.html#arranging-data"><i class="fa fa-check"></i><b>5.7</b> Arranging Data</a>
<ul>
<li class="chapter" data-level="5.7.1" data-path="data-wrangling.html"><a href="data-wrangling.html#dplyrmutate---create-new-columns-using-other-columns"><i class="fa fa-check"></i><b>5.7.1</b> <code>dplyr::mutate()</code> - Create new columns using other columns</a></li>
<li class="chapter" data-level="5.7.2" data-path="data-wrangling.html"><a href="data-wrangling.html#stringr---working-with-character-values"><i class="fa fa-check"></i><b>5.7.2</b> <code>stringr</code> - Working with character values</a></li>
<li class="chapter" data-level="5.7.3" data-path="data-wrangling.html"><a href="data-wrangling.html#dplyrselect---subset-columns-by-name"><i class="fa fa-check"></i><b>5.7.3</b> <code>dplyr::select()</code> - Subset Columns by Name</a></li>
<li class="chapter" data-level="5.7.4" data-path="data-wrangling.html"><a href="data-wrangling.html#dplyrfilter---pick-rows-out-of-a-data-set"><i class="fa fa-check"></i><b>5.7.4</b> <code>dplyr::filter()</code> - Pick rows out of a data set</a></li>
<li class="chapter" data-level="5.7.5" data-path="data-wrangling.html"><a href="data-wrangling.html#dplyrarrange---order-rows-based-on-their-values"><i class="fa fa-check"></i><b>5.7.5</b> <code>dplyr::arrange()</code> - Order rows based on their values</a></li>
<li class="chapter" data-level="5.7.6" data-path="data-wrangling.html"><a href="data-wrangling.html#putting-it-all-together"><i class="fa fa-check"></i><b>5.7.6</b> Putting it all together</a></li>
</ul></li>
<li class="chapter" data-level="5.8" data-path="data-wrangling.html"><a href="data-wrangling.html#grouping-data"><i class="fa fa-check"></i><b>5.8</b> Grouping Data</a></li>
<li class="chapter" data-level="5.9" data-path="data-wrangling.html"><a href="data-wrangling.html#rearranging-data"><i class="fa fa-check"></i><b>5.9</b> Rearranging Data</a></li>
<li class="chapter" data-level="5.10" data-path="data-wrangling.html"><a href="data-wrangling.html#relational-data"><i class="fa fa-check"></i><b>5.10</b> Relational Data</a>
<ul>
<li class="chapter" data-level="5.10.1" data-path="data-wrangling.html"><a href="data-wrangling.html#dealing-with-multiple-matches"><i class="fa fa-check"></i><b>5.10.1</b> Dealing with multiple matches</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="6" data-path="data-science.html"><a href="data-science.html"><i class="fa fa-check"></i><b>6</b> Data Science</a>
<ul>
<li class="chapter" data-level="6.1" data-path="data-science.html"><a href="data-science.html#data-modeling"><i class="fa fa-check"></i><b>6.1</b> Data Modeling</a>
<ul>
<li class="chapter" data-level="6.1.1" data-path="data-science.html"><a href="data-science.html#a-worked-modeling-example"><i class="fa fa-check"></i><b>6.1.1</b> A Worked Modeling Example</a></li>
<li class="chapter" data-level="6.1.2" data-path="data-science.html"><a href="data-science.html#data-summarization"><i class="fa fa-check"></i><b>6.1.2</b> Data Summarization</a></li>
<li class="chapter" data-level="6.1.3" data-path="data-science.html"><a href="data-science.html#linear-models"><i class="fa fa-check"></i><b>6.1.3</b> Linear Models</a></li>
<li class="chapter" data-level="6.1.4" data-path="data-science.html"><a href="data-science.html#flavors-of-linear-models"><i class="fa fa-check"></i><b>6.1.4</b> Flavors of Linear Models</a></li>
<li class="chapter" data-level="6.1.5" data-path="data-science.html"><a href="data-science.html#assessing-model-accuracy-."><i class="fa fa-check"></i><b>6.1.5</b> Assessing Model Accuracy .</a></li>
</ul></li>
<li class="chapter" data-level="6.2" data-path="data-science.html"><a href="data-science.html#statistical-distributions-tests-."><i class="fa fa-check"></i><b>6.2</b> Statistical Distributions &amp; Tests .</a>
<ul>
<li class="chapter" data-level="6.2.1" data-path="data-science.html"><a href="data-science.html#statistical-distributions"><i class="fa fa-check"></i><b>6.2.1</b> Statistical Distributions</a></li>
<li class="chapter" data-level="6.2.2" data-path="data-science.html"><a href="data-science.html#statistical-tests-."><i class="fa fa-check"></i><b>6.2.2</b> Statistical Tests .</a></li>
<li class="chapter" data-level="6.2.3" data-path="data-science.html"><a href="data-science.html#p-values-."><i class="fa fa-check"></i><b>6.2.3</b> p-values .</a></li>
<li class="chapter" data-level="6.2.4" data-path="data-science.html"><a href="data-science.html#multiple-hypothesis-testing-."><i class="fa fa-check"></i><b>6.2.4</b> Multiple Hypothesis Testing .</a></li>
<li class="chapter" data-level="6.2.5" data-path="data-science.html"><a href="data-science.html#statistical-power-."><i class="fa fa-check"></i><b>6.2.5</b> Statistical power .</a></li>
</ul></li>
<li class="chapter" data-level="6.3" data-path="data-science.html"><a href="data-science.html#exploratory-analysis"><i class="fa fa-check"></i><b>6.3</b> Exploratory Analysis</a>
<ul>
<li class="chapter" data-level="6.3.1" data-path="data-science.html"><a href="data-science.html#principal-component-analysis"><i class="fa fa-check"></i><b>6.3.1</b> Principal Component Analysis</a></li>
<li class="chapter" data-level="6.3.2" data-path="data-science.html"><a href="data-science.html#cluster-analysis"><i class="fa fa-check"></i><b>6.3.2</b> Cluster Analysis</a></li>
<li class="chapter" data-level="6.3.3" data-path="data-science.html"><a href="data-science.html#k-means-."><i class="fa fa-check"></i><b>6.3.3</b> k-means .</a></li>
<li class="chapter" data-level="6.3.4" data-path="data-science.html"><a href="data-science.html#others"><i class="fa fa-check"></i><b>6.3.4</b> Others</a></li>
</ul></li>
<li class="chapter" data-level="6.4" data-path="data-science.html"><a href="data-science.html#network-analysis-."><i class="fa fa-check"></i><b>6.4</b> Network Analysis .</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="data-visualization.html"><a href="data-visualization.html"><i class="fa fa-check"></i><b>7</b> Data Visualization</a>
<ul>
<li class="chapter" data-level="7.1" data-path="data-visualization.html"><a href="data-visualization.html#dv-gg"><i class="fa fa-check"></i><b>7.1</b> Grammar of Graphics</a></li>
<li class="chapter" data-level="7.2" data-path="data-visualization.html"><a href="data-visualization.html#plotting-one-dimension"><i class="fa fa-check"></i><b>7.2</b> Plotting One Dimension</a>
<ul>
<li class="chapter" data-level="7.2.1" data-path="data-visualization.html"><a href="data-visualization.html#dv-bar"><i class="fa fa-check"></i><b>7.2.1</b> Bar chart</a></li>
<li class="chapter" data-level="7.2.2" data-path="data-visualization.html"><a href="data-visualization.html#lollipop-plots"><i class="fa fa-check"></i><b>7.2.2</b> Lollipop plots</a></li>
<li class="chapter" data-level="7.2.3" data-path="data-visualization.html"><a href="data-visualization.html#stacked-area-charts"><i class="fa fa-check"></i><b>7.2.3</b> Stacked Area charts</a></li>
<li class="chapter" data-level="7.2.4" data-path="data-visualization.html"><a href="data-visualization.html#parallel-coordinate-plots"><i class="fa fa-check"></i><b>7.2.4</b> Parallel Coordinate plots</a></li>
</ul></li>
<li class="chapter" data-level="7.3" data-path="data-visualization.html"><a href="data-visualization.html#visualizing-distributions"><i class="fa fa-check"></i><b>7.3</b> Visualizing Distributions</a>
<ul>
<li class="chapter" data-level="7.3.1" data-path="data-visualization.html"><a href="data-visualization.html#histogram"><i class="fa fa-check"></i><b>7.3.1</b> Histogram</a></li>
<li class="chapter" data-level="7.3.2" data-path="data-visualization.html"><a href="data-visualization.html#density"><i class="fa fa-check"></i><b>7.3.2</b> Density</a></li>
<li class="chapter" data-level="7.3.3" data-path="data-visualization.html"><a href="data-visualization.html#boxplot"><i class="fa fa-check"></i><b>7.3.3</b> Boxplot</a></li>
<li class="chapter" data-level="7.3.4" data-path="data-visualization.html"><a href="data-visualization.html#violin-plot"><i class="fa fa-check"></i><b>7.3.4</b> Violin plot</a></li>
<li class="chapter" data-level="7.3.5" data-path="data-visualization.html"><a href="data-visualization.html#beeswarm-plot"><i class="fa fa-check"></i><b>7.3.5</b> Beeswarm plot</a></li>
<li class="chapter" data-level="7.3.6" data-path="data-visualization.html"><a href="data-visualization.html#ridgeline"><i class="fa fa-check"></i><b>7.3.6</b> Ridgeline</a></li>
</ul></li>
<li class="chapter" data-level="7.4" data-path="data-visualization.html"><a href="data-visualization.html#plotting-two-or-more-dimensions-."><i class="fa fa-check"></i><b>7.4</b> Plotting Two or More Dimensions .</a>
<ul>
<li class="chapter" data-level="7.4.1" data-path="data-visualization.html"><a href="data-visualization.html#scatter-plot-."><i class="fa fa-check"></i><b>7.4.1</b> Scatter plot .</a></li>
<li class="chapter" data-level="7.4.2" data-path="data-visualization.html"><a href="data-visualization.html#line-plot-."><i class="fa fa-check"></i><b>7.4.2</b> Line plot .</a></li>
<li class="chapter" data-level="7.4.3" data-path="data-visualization.html"><a href="data-visualization.html#bubble-plot-."><i class="fa fa-check"></i><b>7.4.3</b> Bubble plot .</a></li>
<li class="chapter" data-level="7.4.4" data-path="data-visualization.html"><a href="data-visualization.html#heatmaps"><i class="fa fa-check"></i><b>7.4.4</b> Heatmaps</a></li>
</ul></li>
<li class="chapter" data-level="7.5" data-path="data-visualization.html"><a href="data-visualization.html#other-kind-of-plots-."><i class="fa fa-check"></i><b>7.5</b> Other Kind of Plots .</a>
<ul>
<li class="chapter" data-level="7.5.1" data-path="data-visualization.html"><a href="data-visualization.html#chord-."><i class="fa fa-check"></i><b>7.5.1</b> Chord .</a></li>
<li class="chapter" data-level="7.5.2" data-path="data-visualization.html"><a href="data-visualization.html#sankey-."><i class="fa fa-check"></i><b>7.5.2</b> Sankey .</a></li>
<li class="chapter" data-level="7.5.3" data-path="data-visualization.html"><a href="data-visualization.html#dendrograms"><i class="fa fa-check"></i><b>7.5.3</b> Dendrograms</a></li>
</ul></li>
<li class="chapter" data-level="7.6" data-path="data-visualization.html"><a href="data-visualization.html#tips-and-tricks-."><i class="fa fa-check"></i><b>7.6</b> Tips and Tricks .</a>
<ul>
<li class="chapter" data-level="7.6.1" data-path="data-visualization.html"><a href="data-visualization.html#reordering-1-d-data-elements-."><i class="fa fa-check"></i><b>7.6.1</b> Reordering 1-D Data Elements .</a></li>
<li class="chapter" data-level="7.6.2" data-path="data-visualization.html"><a href="data-visualization.html#confidence-intervals-."><i class="fa fa-check"></i><b>7.6.2</b> Confidence Intervals .</a></li>
<li class="chapter" data-level="7.6.3" data-path="data-visualization.html"><a href="data-visualization.html#annotations-."><i class="fa fa-check"></i><b>7.6.3</b> Annotations .</a></li>
</ul></li>
<li class="chapter" data-level="7.7" data-path="data-visualization.html"><a href="data-visualization.html#multiple-plots-."><i class="fa fa-check"></i><b>7.7</b> Multiple Plots .</a>
<ul>
<li class="chapter" data-level="7.7.1" data-path="data-visualization.html"><a href="data-visualization.html#secondary-axes-."><i class="fa fa-check"></i><b>7.7.1</b> Secondary Axes .</a></li>
<li class="chapter" data-level="7.7.2" data-path="data-visualization.html"><a href="data-visualization.html#facet-wrapping-."><i class="fa fa-check"></i><b>7.7.2</b> Facet wrapping .</a></li>
<li class="chapter" data-level="7.7.3" data-path="data-visualization.html"><a href="data-visualization.html#multipanel-figures-."><i class="fa fa-check"></i><b>7.7.3</b> Multipanel Figures .</a></li>
</ul></li>
<li class="chapter" data-level="7.8" data-path="data-visualization.html"><a href="data-visualization.html#responsible-plotting-."><i class="fa fa-check"></i><b>7.8</b> Responsible Plotting .</a></li>
<li class="chapter" data-level="7.9" data-path="data-visualization.html"><a href="data-visualization.html#publication-ready-plots-."><i class="fa fa-check"></i><b>7.9</b> Publication Ready Plots .</a></li>
<li class="chapter" data-level="7.10" data-path="data-visualization.html"><a href="data-visualization.html#network-visualization-."><i class="fa fa-check"></i><b>7.10</b> Network visualization .</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html"><i class="fa fa-check"></i><b>8</b> Biology &amp; Bioinformatics</a>
<ul>
<li class="chapter" data-level="8.1" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#bio-r"><i class="fa fa-check"></i><b>8.1</b> R in Biology</a></li>
<li class="chapter" data-level="8.2" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#biological-data-overview"><i class="fa fa-check"></i><b>8.2</b> Biological Data Overview</a>
<ul>
<li class="chapter" data-level="8.2.1" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#csv-files"><i class="fa fa-check"></i><b>8.2.1</b> CSV Files</a></li>
<li class="chapter" data-level="8.2.2" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#bio-data-types"><i class="fa fa-check"></i><b>8.2.2</b> Types of Biological Data</a></li>
<li class="chapter" data-level="8.2.3" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#common-biological-data-matrices"><i class="fa fa-check"></i><b>8.2.3</b> Common Biological Data Matrices</a></li>
<li class="chapter" data-level="8.2.4" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#biological-data-is-not-tidy"><i class="fa fa-check"></i><b>8.2.4</b> Biological data is NOT Tidy!</a></li>
</ul></li>
<li class="chapter" data-level="8.3" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#bio-bioconductor"><i class="fa fa-check"></i><b>8.3</b> Bioconductor</a></li>
<li class="chapter" data-level="8.4" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#gene-identifiers"><i class="fa fa-check"></i><b>8.4</b> Gene Identifiers</a>
<ul>
<li class="chapter" data-level="8.4.1" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#gene-identifier-systems"><i class="fa fa-check"></i><b>8.4.1</b> Gene Identifier Systems</a></li>
<li class="chapter" data-level="8.4.2" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#mapping-between-identifier-systems"><i class="fa fa-check"></i><b>8.4.2</b> Mapping Between Identifier Systems</a></li>
<li class="chapter" data-level="8.4.3" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#mapping-homologs"><i class="fa fa-check"></i><b>8.4.3</b> Mapping Homologs</a></li>
</ul></li>
<li class="chapter" data-level="8.5" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#gene-expression"><i class="fa fa-check"></i><b>8.5</b> Gene Expression</a>
<ul>
<li class="chapter" data-level="8.5.1" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#gene-expression-data-in-bioconductor"><i class="fa fa-check"></i><b>8.5.1</b> Gene Expression Data in Bioconductor</a></li>
<li class="chapter" data-level="8.5.2" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#microarrays"><i class="fa fa-check"></i><b>8.5.2</b> Microarrays</a></li>
<li class="chapter" data-level="8.5.3" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#microarray-data-in-bioconductor"><i class="fa fa-check"></i><b>8.5.3</b> Microarray Data in Bioconductor</a></li>
<li class="chapter" data-level="8.5.4" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#differential-expression-analysis"><i class="fa fa-check"></i><b>8.5.4</b> Differential Expression Analysis</a></li>
<li class="chapter" data-level="8.5.5" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#differential-expression-microarrays-limma"><i class="fa fa-check"></i><b>8.5.5</b> Differential Expression: Microarrays (limma)</a></li>
<li class="chapter" data-level="8.5.6" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#high-throughput-sequencing-."><i class="fa fa-check"></i><b>8.5.6</b> High Throughput Sequencing .</a></li>
<li class="chapter" data-level="8.5.7" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#count-data-."><i class="fa fa-check"></i><b>8.5.7</b> Count Data .</a></li>
<li class="chapter" data-level="8.5.8" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#rnaseq-."><i class="fa fa-check"></i><b>8.5.8</b> RNASeq .</a></li>
<li class="chapter" data-level="8.5.9" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#differential-expression-rnaseq-."><i class="fa fa-check"></i><b>8.5.9</b> Differential Expression: RNASeq .</a></li>
</ul></li>
<li class="chapter" data-level="8.6" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#gene-set-enrichment-analysis"><i class="fa fa-check"></i><b>8.6</b> Gene Set Enrichment Analysis</a>
<ul>
<li class="chapter" data-level="8.6.1" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#gene-sets"><i class="fa fa-check"></i><b>8.6.1</b> Gene Sets</a></li>
<li class="chapter" data-level="8.6.2" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#working-with-gene-sets-in-r"><i class="fa fa-check"></i><b>8.6.2</b> Working with gene sets in R</a></li>
<li class="chapter" data-level="8.6.3" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#over-representation-analysis"><i class="fa fa-check"></i><b>8.6.3</b> Over-representation Analysis</a></li>
<li class="chapter" data-level="8.6.4" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#rank-based-analysis-."><i class="fa fa-check"></i><b>8.6.4</b> Rank-based Analysis .</a></li>
<li class="chapter" data-level="8.6.5" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#fgsea"><i class="fa fa-check"></i><b>8.6.5</b> fgsea</a></li>
</ul></li>
<li class="chapter" data-level="8.7" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#biological-networks-."><i class="fa fa-check"></i><b>8.7</b> Biological Networks .</a>
<ul>
<li class="chapter" data-level="8.7.1" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#biological-pathways-."><i class="fa fa-check"></i><b>8.7.1</b> Biological Pathways .</a></li>
<li class="chapter" data-level="8.7.2" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#gene-regulatory-networks-."><i class="fa fa-check"></i><b>8.7.2</b> Gene Regulatory Networks .</a></li>
<li class="chapter" data-level="8.7.3" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#protein-protein-interaction-network-."><i class="fa fa-check"></i><b>8.7.3</b> Protein-Protein Interaction Network .</a></li>
<li class="chapter" data-level="8.7.4" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#wgcna-."><i class="fa fa-check"></i><b>8.7.4</b> WGCNA .</a></li>
</ul></li>
<li class="chapter" data-level="8.8" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#single-cell-sequencing-analysis-."><i class="fa fa-check"></i><b>8.8</b> Single Cell Sequencing Analysis .</a>
<ul>
<li class="chapter" data-level="8.8.1" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#single-cell-sequencing-."><i class="fa fa-check"></i><b>8.8.1</b> Single Cell Sequencing .</a></li>
<li class="chapter" data-level="8.8.2" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#single-cell-analysis-."><i class="fa fa-check"></i><b>8.8.2</b> Single Cell Analysis .</a></li>
<li class="chapter" data-level="8.8.3" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#single-cell-clustering-."><i class="fa fa-check"></i><b>8.8.3</b> Single Cell Clustering .</a></li>
<li class="chapter" data-level="8.8.4" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#dimensionality-reduction-projection-."><i class="fa fa-check"></i><b>8.8.4</b> Dimensionality Reduction &amp; Projection .</a></li>
<li class="chapter" data-level="8.8.5" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#single-cell-data-visualization-."><i class="fa fa-check"></i><b>8.8.5</b> Single Cell Data Visualization .</a></li>
<li class="chapter" data-level="8.8.6" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#single-cell-marker-analysis-."><i class="fa fa-check"></i><b>8.8.6</b> Single Cell Marker Analysis .</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="9" data-path="engineering.html"><a href="engineering.html"><i class="fa fa-check"></i><b>9</b> EngineeRing</a>
<ul>
<li class="chapter" data-level="9.1" data-path="engineering.html"><a href="engineering.html#unit-testing"><i class="fa fa-check"></i><b>9.1</b> Unit Testing</a></li>
<li class="chapter" data-level="9.2" data-path="engineering.html"><a href="engineering.html#object-orientation-in-r-."><i class="fa fa-check"></i><b>9.2</b> Object Orientation in R .</a></li>
<li class="chapter" data-level="9.3" data-path="engineering.html"><a href="engineering.html#toolification-."><i class="fa fa-check"></i><b>9.3</b> Toolification .</a></li>
<li class="chapter" data-level="9.4" data-path="engineering.html"><a href="engineering.html#pipelines-workflows-."><i class="fa fa-check"></i><b>9.4</b> Pipelines &amp; Workflows .</a></li>
<li class="chapter" data-level="9.5" data-path="engineering.html"><a href="engineering.html#parallel-processing-."><i class="fa fa-check"></i><b>9.5</b> Parallel Processing .</a></li>
<li class="chapter" data-level="9.6" data-path="engineering.html"><a href="engineering.html#r-packages-."><i class="fa fa-check"></i><b>9.6</b> R Packages .</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="rshiny.html"><a href="rshiny.html"><i class="fa fa-check"></i><b>10</b> RShiny</a>
<ul>
<li class="chapter" data-level="10.1" data-path="rshiny.html"><a href="rshiny.html#overview"><i class="fa fa-check"></i><b>10.1</b> Overview</a></li>
<li class="chapter" data-level="10.2" data-path="rshiny.html"><a href="rshiny.html#learning-objectives"><i class="fa fa-check"></i><b>10.2</b> Learning Objectives</a></li>
<li class="chapter" data-level="10.3" data-path="rshiny.html"><a href="rshiny.html#skill-list"><i class="fa fa-check"></i><b>10.3</b> Skill List</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="communicating-with-r.html"><a href="communicating-with-r.html"><i class="fa fa-check"></i><b>11</b> Communicating with R</a>
<ul>
<li class="chapter" data-level="11.1" data-path="communicating-with-r.html"><a href="communicating-with-r.html#rmarkdown-knitr"><i class="fa fa-check"></i><b>11.1</b> RMarkdown &amp; knitr</a></li>
<li class="chapter" data-level="11.2" data-path="communicating-with-r.html"><a href="communicating-with-r.html#bookdown"><i class="fa fa-check"></i><b>11.2</b> bookdown</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="contribution-guide.html"><a href="contribution-guide.html"><i class="fa fa-check"></i><b>12</b> Contribution Guide</a>
<ul>
<li class="chapter" data-level="12.1" data-path="contribution-guide.html"><a href="contribution-guide.html#custom-blocks"><i class="fa fa-check"></i><b>12.1</b> Custom blocks</a></li>
</ul></li>
<li class="part"><span><b>I Assignments</b></span></li>
<li class="chapter" data-level="" data-path="assignment-format.html"><a href="assignment-format.html"><i class="fa fa-check"></i>Assignment Format</a></li>
<li class="chapter" data-level="" data-path="starting-an-assignment.html"><a href="starting-an-assignment.html"><i class="fa fa-check"></i>Starting an Assignment</a>
<ul>
<li><a href="starting-an-assignment.html#starting-your-own-git-repository-and-cloning-it">Starting your own <code>git</code> repository and cloning it</a>
<ul>
<li class="chapter" data-level="" data-path="starting-an-assignment.html"><a href="starting-an-assignment.html#contents"><i class="fa fa-check"></i>Contents</a></li>
<li class="chapter" data-level="" data-path="starting-an-assignment.html"><a href="starting-an-assignment.html#github-tutorial"><i class="fa fa-check"></i>GitHub tutorial</a></li>
<li class="chapter" data-level="" data-path="starting-an-assignment.html"><a href="starting-an-assignment.html#committing-and-pushing-with-r-and-without"><i class="fa fa-check"></i>Committing and Pushing with R (and without)</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="assignment-1.html"><a href="assignment-1.html"><i class="fa fa-check"></i>Assignment 1</a>
<ul>
<li class="chapter" data-level="" data-path="assignment-1.html"><a href="assignment-1.html#problem-statement"><i class="fa fa-check"></i>Problem Statement</a></li>
<li class="chapter" data-level="" data-path="assignment-1.html"><a href="assignment-1.html#learning-objectives-and-skill-list"><i class="fa fa-check"></i>Learning Objectives and Skill List</a></li>
<li class="chapter" data-level="" data-path="assignment-1.html"><a href="assignment-1.html#instructions"><i class="fa fa-check"></i>Instructions</a></li>
<li class="chapter" data-level="" data-path="assignment-1.html"><a href="assignment-1.html#hints"><i class="fa fa-check"></i>Hints</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="assignment-2.html"><a href="assignment-2.html"><i class="fa fa-check"></i>Assignment 2</a>
<ul>
<li class="chapter" data-level="" data-path="assignment-2.html"><a href="assignment-2.html#problem-statement-1"><i class="fa fa-check"></i>Problem Statement</a></li>
<li class="chapter" data-level="" data-path="assignment-2.html"><a href="assignment-2.html#learning-objectives-1"><i class="fa fa-check"></i>Learning Objectives</a></li>
<li class="chapter" data-level="" data-path="assignment-2.html"><a href="assignment-2.html#skill-list-1"><i class="fa fa-check"></i>Skill List</a></li>
<li class="chapter" data-level="" data-path="assignment-2.html"><a href="assignment-2.html#instructions-1"><i class="fa fa-check"></i>Instructions</a></li>
<li class="chapter" data-level="" data-path="assignment-2.html"><a href="assignment-2.html#function-details"><i class="fa fa-check"></i>Function Details</a>
<ul>
<li class="chapter" data-level="" data-path="assignment-2.html"><a href="assignment-2.html#bioconductor"><i class="fa fa-check"></i>1. Bioconductor</a></li>
<li><a href="assignment-2.html#load_expression">2. <code>load_expression()</code></a></li>
<li><a href="assignment-2.html#filter_15">3. <code>filter_15()</code></a></li>
<li><a href="assignment-2.html#affy_to_hgnc">4. <code>affy_to_hgnc()</code></a></li>
<li><a href="assignment-2.html#reduce_data">5. <code>reduce_data()</code></a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="assignment-3.html"><a href="assignment-3.html"><i class="fa fa-check"></i>Assignment 3</a>
<ul>
<li class="chapter" data-level="" data-path="assignment-3.html"><a href="assignment-3.html#problem-statement-2"><i class="fa fa-check"></i>Problem Statement</a></li>
<li class="chapter" data-level="" data-path="assignment-3.html"><a href="assignment-3.html#learning-objectives-2"><i class="fa fa-check"></i>Learning Objectives</a></li>
<li class="chapter" data-level="" data-path="assignment-3.html"><a href="assignment-3.html#skill-list-2"><i class="fa fa-check"></i>Skill List</a></li>
<li class="chapter" data-level="" data-path="assignment-3.html"><a href="assignment-3.html#background-on-microarrays"><i class="fa fa-check"></i>Background on Microarrays</a></li>
<li class="chapter" data-level="" data-path="assignment-3.html"><a href="assignment-3.html#background-on-principal-component-analysis"><i class="fa fa-check"></i>Background on Principal Component Analysis</a></li>
<li class="chapter" data-level="" data-path="assignment-3.html"><a href="assignment-3.html#marisa-et-al.-gene-expression-classification-of-colon-cancer-into-molecular-subtypes-characterization-validation-and-prognostic-value.-plos-medicine-may-2013.-pmid-23700391"><i class="fa fa-check"></i>Marisa et al. Gene Expression Classification of Colon Cancer into Molecular Subtypes: Characterization, Validation, and Prognostic Value. PLoS Medicine, May 2013. PMID: 23700391</a></li>
<li><a href="assignment-3.html#scaling-data-using-r-scale">Scaling data using R <code>scale()</code></a></li>
<li class="chapter" data-level="" data-path="assignment-3.html"><a href="assignment-3.html#proportion-of-variance-explained"><i class="fa fa-check"></i>Proportion of variance explained</a></li>
<li class="chapter" data-level="" data-path="assignment-3.html"><a href="assignment-3.html#plotting-and-visualization-of-pca"><i class="fa fa-check"></i>Plotting and visualization of PCA</a></li>
<li class="chapter" data-level="" data-path="assignment-3.html"><a href="assignment-3.html#hierarchical-clustering-and-heatmaps"><i class="fa fa-check"></i>Hierarchical Clustering and Heatmaps</a></li>
<li class="chapter" data-level="" data-path="assignment-3.html"><a href="assignment-3.html#references"><i class="fa fa-check"></i>References</a></li>
</ul></li>
<li class="appendix"><span><b>Appendix</b></span></li>
<li class="chapter" data-level="A" data-path="class-outlines.html"><a href="class-outlines.html"><i class="fa fa-check"></i><b>A</b> Class Outlines</a>
<ul>
<li class="chapter" data-level="A.1" data-path="class-outlines.html"><a href="class-outlines.html#week-1"><i class="fa fa-check"></i><b>A.1</b> Week 1</a></li>
<li class="chapter" data-level="A.2" data-path="class-outlines.html"><a href="class-outlines.html#week-2"><i class="fa fa-check"></i><b>A.2</b> Week 2</a></li>
<li class="chapter" data-level="A.3" data-path="class-outlines.html"><a href="class-outlines.html#week-3"><i class="fa fa-check"></i><b>A.3</b> Week 3</a></li>
<li class="chapter" data-level="A.4" data-path="class-outlines.html"><a href="class-outlines.html#week-4"><i class="fa fa-check"></i><b>A.4</b> Week 4</a></li>
<li class="chapter" data-level="A.5" data-path="class-outlines.html"><a href="class-outlines.html#week-5"><i class="fa fa-check"></i><b>A.5</b> Week 5</a></li>
<li class="chapter" data-level="A.6" data-path="class-outlines.html"><a href="class-outlines.html#week-6"><i class="fa fa-check"></i><b>A.6</b> Week 6</a></li>
<li class="chapter" data-level="A.7" data-path="class-outlines.html"><a href="class-outlines.html#week-7"><i class="fa fa-check"></i><b>A.7</b> Week 7</a></li>
<li class="chapter" data-level="A.8" data-path="class-outlines.html"><a href="class-outlines.html#week-8"><i class="fa fa-check"></i><b>A.8</b> Week 8</a></li>
<li class="chapter" data-level="A.9" data-path="class-outlines.html"><a href="class-outlines.html#week-9"><i class="fa fa-check"></i><b>A.9</b> Week 9</a></li>
<li class="chapter" data-level="A.10" data-path="class-outlines.html"><a href="class-outlines.html#week-10"><i class="fa fa-check"></i><b>A.10</b> Week 10</a></li>
<li class="chapter" data-level="A.11" data-path="class-outlines.html"><a href="class-outlines.html#week-11"><i class="fa fa-check"></i><b>A.11</b> Week 11</a></li>
<li class="chapter" data-level="A.12" data-path="class-outlines.html"><a href="class-outlines.html#week-12"><i class="fa fa-check"></i><b>A.12</b> Week 12</a></li>
<li class="chapter" data-level="A.13" data-path="class-outlines.html"><a href="class-outlines.html#week-13---no-class-for-project-work"><i class="fa fa-check"></i><b>A.13</b> Week 13 - no class for project work</a></li>
<li class="chapter" data-level="A.14" data-path="class-outlines.html"><a href="class-outlines.html#week-14---no-class-for-project-work"><i class="fa fa-check"></i><b>A.14</b> Week 14 - no class for project work</a></li>
</ul></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">BF591 - R for Biological Sciences</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="biology-bioinformatics" class="section level1" number="8">
<h1><span class="header-section-number">8</span> Biology &amp; Bioinformatics</h1>
<div id="bio-r" class="section level2" number="8.1">
<h2><span class="header-section-number">8.1</span> R in Biology</h2>
<p>R became popular in biological data analysis in the early to mid 2000s, when
<a href="https://www.genome.gov/genetics-glossary/Microarray-Technology">microarray
technology</a> came
into widespread use enabling researchers to look for statistical differences in
gene expression for thousands of genes across large numbers of samples. As a
result of this popularity, a community of biological researchers and data
analysts created a collection of software packages called
<a href="https://www.bioconductor.org/">Bioconductor</a>, which made a vast array of
cutting edge statistical and bioinformatic methodologies widely available.</p>
<p>Due to its ease of use, widespread community support, rich ecosystem of
biological data analysis packages, and to it being free, R remains one of the
primary languages of biological data analysis. The language’s early focus on
statistical analysis, and later transition to data science in general, makes it
very useful and accessible to perform the kinds of analyses the new data science
of biology required. It is also a bridge between biologists without a
computational background and statisticians and bioinformaticians, who invent
new methods and implement them as R packages that are easily accessible by all.
The language and the package ecosystems and communities it supports continue to
be a major source of biological discovery and innovation.</p>
<p>As a data science, biology benefits from the main strengths of R and <a href="#dw-tidyverse">the
tidyverse</a> when combined with the powerful analytical techniques
available in Bioconductor packages, namely to manipulate, visualize, analyze,
and communicate biological data.</p>

</div>
<div id="biological-data-overview" class="section level2" number="8.2">
<h2><span class="header-section-number">8.2</span> Biological Data Overview</h2>
<div id="csv-files" class="section level3" number="8.2.1">
<h3><span class="header-section-number">8.2.1</span> CSV Files</h3>
<p>The most common, convenient, and flexible data file format in biology and
bioinformatics is the <em>character-delimited</em> or <em>character-separated text file</em>.
These files are plain text files (i.e. not the native file format of any
specific program, like Excel) that generally contain rectangles of data. When
formatted correctly, you can think of these files as containing a grid or matrix
of data values with some number of columns, each of which has the same number of
values. Each line of these files has some number of data values separated by a
consistent character, most commonly the comma which are called <a href="https://en.wikipedia.org/wiki/Comma-separated_values">comma-separated
value, or “CSV,” files</a>
and filenames typically end with the extension <code>.csv</code>. Note that
other characters, especially the <Tab> character, may be used to create valid
files in this format, and all the same general principles apply. This is an
example of a simple CSV file:</p>
<pre><code>id,somevalue,category,genes
1,2.123,A,APOE
4,5.123,B,&quot;HOXA1,HOXB1&quot;
7,8.123,,SNCA</code></pre>
<p>Some properties and principles of CSV files:</p>
<ul>
<li>The first line often but not always contains the column names of each column</li>
<li>Each value is delimited by the same character, in this case <code>,</code></li>
<li>Values can be any value, including numbers and characters</li>
<li>When a value contains the delimiting character (e.g. HOXA1,HOXB1 contains a
<code>,</code>), the value is wrapped in double quotes</li>
<li>Values can be missing, indicated by sequential delimiters (i.e. <code>,,</code> or one
<code>,</code> at the end of the line, if the last column value is missing)</li>
<li>There is no delimiter at the end of the lines</li>
<li>To be well-formatted <em>every line must have the same number of delimited
values</em></li>
</ul>
<p>These same properties and principles apply to all character-separated files,
regardless of the specific delimiter used. If a file follows these principles,
they can be loaded very easily into R or any other data analysis setting.</p>
</div>
<div id="bio-data-types" class="section level3" number="8.2.2">
<h3><span class="header-section-number">8.2.2</span> Types of Biological Data</h3>
<p>In very general terms, there are five types of data used in biological data
analysis: raw/primary data, processed data, analysis results, metadata, and
annotation data.</p>
<p><strong>Raw/primary data.</strong> These data are the primary observations made by whatever
instruments/techniques we are using for our experiment. These may include
high-throughput sequencing data, mass/charge ratio data from mass spectrometry,
16S rRNA sequencing data from metagenomic studies, SNPs from genotyping assays,
etc. These data can be very large and are often not efficiently processed using
R. Instead, specialized tools built outside of R are used to first process the
primary data into a form that is amenable to analysis.</p>
<p><strong>Processed data.</strong> Processed data is the result of any analysis or
transformation of primary data into an intermediate, more interpretable form.
For example, when performing gene expression analysis with RNASeq, short reads
that form the primary data are typically aligned against a genome and then
counted against annotated genes, resulting in a count for each gene in the
genome. Typically, these counts are then combined for many samples into a single
matrix and subject to downstream analyses like differential expression.</p>
<p><strong>Analysis results.</strong> Analysis results aren’t data <em>per se</em>, but are the results
of analysis of primary data or processed results. For example, in gene
expression studies, a first analysis is often differential expression, where
each gene in the genome is tested for expression associated with some outcome
of interest across many samples. Each gene then has a number of statistics and
related values, like log2 fold change, nominal and adjusted p-value, etc. These
forms of data are usually what we use to form interpretations of our datasets
and therefore we must manipulate them in much the same way as any other dataset.</p>
<p><strong>Metadata.</strong> In most experimental settings, multiple samples have been
processed and had the same primary data type collected. These samples also have
information associated with them, which is here termed <em>metadata</em>, or “data
about data.” In our gene expression of post mortem brain experiments mentioned
above, the information about the individuals themselves, including age at death,
whether the person had a disease, the measurements of tissue quality, etc. is
the metadata. The primary and processed data and metadata are usually stored in
different files, where the metadata (or <em>sample information</em> or <em>sample data</em>,
etc) will have one column indicating the unique identifier (ID) of each sample.
The processed data will typically have columns named for each of the sample IDs.</p>
<p><strong>Annotation data.</strong> A tremendous amount of knowledge has been generated about
biological entities, e.g. genes, especially since the publication of the human
genome. Annotation data is publicly available information about the features we
measure in our experiments. This may come in the form of the coordinates in a
genome where genes exist, any known functions of those genes, the domains found
in proteins and their relative sequence, gene identifier cross references across
different gene naming systems (e.g. symbol vs Ensembl ID), single nucleotide
polymorphism genomic locations and associations with traits or diseases, etc.
This is information that we use to aid in interpretation of our experimental
data but generally do not generate ourselves. Annotation data comes in many
forms, some of which are in CSV format.</p>
<p>The figure below contains a simple diagram of how these different types of data
work together for a typical biological data analysis.</p>
<div class="figure">
<img src="information_flow.png" alt="" />
<p class="caption">Information flow in biological data analysis</p>
</div>
</div>
<div id="common-biological-data-matrices" class="section level3" number="8.2.3">
<h3><span class="header-section-number">8.2.3</span> Common Biological Data Matrices</h3>
<p>Typically the first data set you will work with in R is <em>processed data</em> as
described in the previous section. This data has been transformed from primary
data in some way such that it (usually) forms a numeric matrix with features
as rows and samples as columns. The first column of these files usually contains
a feature identifier, e.g. gene identifier, genomic locus, probe set ID, etc and
the remaining columns have numerical values, one per sample. The first row is
usually column names for all the columns in the file. Below is an example of one
of these files from a microarray gene expression dataset loaded into R:</p>
<div class="sourceCode" id="cb266"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb266-1"><a href="biology-bioinformatics.html#cb266-1" aria-hidden="true" tabindex="-1"></a>intensities <span class="ot">&lt;-</span> <span class="fu">read_csv</span>(<span class="st">&quot;example_intensity_data.csv&quot;</span>)</span>
<span id="cb266-2"><a href="biology-bioinformatics.html#cb266-2" aria-hidden="true" tabindex="-1"></a>intensities</span>
<span id="cb266-3"><a href="biology-bioinformatics.html#cb266-3" aria-hidden="true" tabindex="-1"></a><span class="co"># A tibble: 54,675 x 36</span></span>
<span id="cb266-4"><a href="biology-bioinformatics.html#cb266-4" aria-hidden="true" tabindex="-1"></a>   probe     GSM972389 GSM972390 GSM972396 GSM972401 GSM972409 GSM972412 GSM972413 GSM972422 GSM972429</span>
<span id="cb266-5"><a href="biology-bioinformatics.html#cb266-5" aria-hidden="true" tabindex="-1"></a>   <span class="sc">&lt;</span>chr<span class="sc">&gt;</span>         <span class="er">&lt;</span>dbl<span class="sc">&gt;</span>     <span class="er">&lt;</span>dbl<span class="sc">&gt;</span>     <span class="er">&lt;</span>dbl<span class="sc">&gt;</span>     <span class="er">&lt;</span>dbl<span class="sc">&gt;</span>     <span class="er">&lt;</span>dbl<span class="sc">&gt;</span>     <span class="er">&lt;</span>dbl<span class="sc">&gt;</span>     <span class="er">&lt;</span>dbl<span class="sc">&gt;</span>     <span class="er">&lt;</span>dbl<span class="sc">&gt;</span>     <span class="er">&lt;</span>dbl<span class="sc">&gt;</span></span>
<span id="cb266-6"><a href="biology-bioinformatics.html#cb266-6" aria-hidden="true" tabindex="-1"></a> <span class="dv">1</span> 1007_s_at      <span class="fl">9.54</span>     <span class="fl">10.2</span>       <span class="fl">9.72</span>      <span class="fl">9.68</span>      <span class="fl">9.35</span>      <span class="fl">9.89</span>      <span class="fl">9.70</span>      <span class="fl">9.67</span>      <span class="fl">9.87</span></span>
<span id="cb266-7"><a href="biology-bioinformatics.html#cb266-7" aria-hidden="true" tabindex="-1"></a> <span class="dv">2</span> 1053_at        <span class="fl">7.62</span>      <span class="fl">7.92</span>      <span class="fl">7.17</span>      <span class="fl">7.24</span>      <span class="fl">8.20</span>      <span class="fl">6.87</span>      <span class="fl">6.62</span>      <span class="fl">7.23</span>      <span class="fl">7.45</span></span>
<span id="cb266-8"><a href="biology-bioinformatics.html#cb266-8" aria-hidden="true" tabindex="-1"></a> <span class="dv">3</span> 117_at         <span class="fl">5.50</span>      <span class="fl">5.56</span>      <span class="fl">5.06</span>      <span class="fl">7.44</span>      <span class="fl">5.19</span>      <span class="fl">5.72</span>      <span class="fl">5.87</span>      <span class="fl">6.15</span>      <span class="fl">5.46</span></span>
<span id="cb266-9"><a href="biology-bioinformatics.html#cb266-9" aria-hidden="true" tabindex="-1"></a> <span class="dv">4</span> 121_at         <span class="fl">7.27</span>      <span class="fl">7.96</span>      <span class="fl">7.42</span>      <span class="fl">7.34</span>      <span class="fl">7.49</span>      <span class="fl">7.76</span>      <span class="fl">7.44</span>      <span class="fl">7.66</span>      <span class="fl">8.02</span></span>
<span id="cb266-10"><a href="biology-bioinformatics.html#cb266-10" aria-hidden="true" tabindex="-1"></a> <span class="dv">5</span> 1255_g_at      <span class="fl">2.79</span>      <span class="fl">3.10</span>      <span class="fl">2.78</span>      <span class="fl">2.91</span>      <span class="fl">3.02</span>      <span class="fl">2.73</span>      <span class="fl">2.78</span>      <span class="fl">3.56</span>      <span class="fl">2.83</span></span>
<span id="cb266-11"><a href="biology-bioinformatics.html#cb266-11" aria-hidden="true" tabindex="-1"></a> <span class="dv">6</span> 1294_at        <span class="fl">7.51</span>      <span class="fl">7.28</span>      <span class="fl">7.00</span>      <span class="fl">7.18</span>      <span class="fl">7.38</span>      <span class="fl">6.98</span>      <span class="fl">6.90</span>      <span class="fl">7.54</span>      <span class="fl">7.66</span></span>
<span id="cb266-12"><a href="biology-bioinformatics.html#cb266-12" aria-hidden="true" tabindex="-1"></a> <span class="dv">7</span> 1316_at        <span class="fl">3.89</span>      <span class="fl">4.36</span>      <span class="fl">4.24</span>      <span class="fl">3.94</span>      <span class="fl">4.20</span>      <span class="fl">4.34</span>      <span class="fl">4.06</span>      <span class="fl">4.24</span>      <span class="fl">4.11</span></span>
<span id="cb266-13"><a href="biology-bioinformatics.html#cb266-13" aria-hidden="true" tabindex="-1"></a> <span class="dv">8</span> 1320_at        <span class="fl">4.65</span>      <span class="fl">4.91</span>      <span class="fl">4.70</span>      <span class="fl">4.78</span>      <span class="fl">5.06</span>      <span class="fl">4.71</span>      <span class="fl">4.55</span>      <span class="fl">4.58</span>      <span class="fl">5.10</span></span>
<span id="cb266-14"><a href="biology-bioinformatics.html#cb266-14" aria-hidden="true" tabindex="-1"></a> <span class="dv">9</span> 1405_i_at      <span class="fl">8.03</span>      <span class="fl">7.47</span>      <span class="fl">5.42</span>      <span class="fl">7.21</span>      <span class="fl">9.48</span>      <span class="fl">6.79</span>      <span class="fl">6.57</span>      <span class="fl">8.50</span>      <span class="fl">6.36</span></span>
<span id="cb266-15"><a href="biology-bioinformatics.html#cb266-15" aria-hidden="true" tabindex="-1"></a><span class="dv">10</span> 1431_at        <span class="fl">3.09</span>      <span class="fl">3.78</span>      <span class="fl">3.33</span>      <span class="fl">3.12</span>      <span class="fl">3.21</span>      <span class="fl">3.27</span>      <span class="fl">3.37</span>      <span class="fl">3.84</span>      <span class="fl">3.32</span></span>
<span id="cb266-16"><a href="biology-bioinformatics.html#cb266-16" aria-hidden="true" tabindex="-1"></a><span class="co"># ... with 54,665 more rows, and 26 more variables: GSM972433 &lt;dbl&gt;, GSM972438 &lt;dbl&gt;, GSM972440 &lt;dbl&gt;,</span></span>
<span id="cb266-17"><a href="biology-bioinformatics.html#cb266-17" aria-hidden="true" tabindex="-1"></a><span class="co">#   GSM972443 &lt;dbl&gt;, GSM972444 &lt;dbl&gt;, GSM972446 &lt;dbl&gt;, GSM972453 &lt;dbl&gt;, GSM972457 &lt;dbl&gt;,</span></span>
<span id="cb266-18"><a href="biology-bioinformatics.html#cb266-18" aria-hidden="true" tabindex="-1"></a><span class="co">#   GSM972459 &lt;dbl&gt;, GSM972463 &lt;dbl&gt;, GSM972467 &lt;dbl&gt;, GSM972472 &lt;dbl&gt;, GSM972473 &lt;dbl&gt;,</span></span>
<span id="cb266-19"><a href="biology-bioinformatics.html#cb266-19" aria-hidden="true" tabindex="-1"></a><span class="co">#   GSM972475 &lt;dbl&gt;, GSM972476 &lt;dbl&gt;, GSM972477 &lt;dbl&gt;, GSM972479 &lt;dbl&gt;, GSM972480 &lt;dbl&gt;,</span></span>
<span id="cb266-20"><a href="biology-bioinformatics.html#cb266-20" aria-hidden="true" tabindex="-1"></a><span class="co">#   GSM972487 &lt;dbl&gt;, GSM972488 &lt;dbl&gt;, GSM972489 &lt;dbl&gt;, GSM972496 &lt;dbl&gt;, GSM972502 &lt;dbl&gt;,</span></span>
<span id="cb266-21"><a href="biology-bioinformatics.html#cb266-21" aria-hidden="true" tabindex="-1"></a><span class="co">#   GSM972510 &lt;dbl&gt;, GSM972512 &lt;dbl&gt;, GSM972521 &lt;dbl&gt;</span></span></code></pre></div>
<p>The file has 54,676 rows, consisting of one header row which R loads in as the
column names, and the remaining are probe sets, one per row. There are 36
columns, where the first contains the probe set ID (e.g. <code>1007_s_at</code>) and the
remaining 35 columns correspond to samples.</p>
</div>
<div id="biological-data-is-not-tidy" class="section level3" number="8.2.4">
<h3><span class="header-section-number">8.2.4</span> Biological data is NOT Tidy!</h3>
<p>As mentioned in the <a href="#tidy%20data">tidy data</a> section, the tidyverse packages
assume data to be in so-called “tidy format,” with variables as columns and
observations as rows. Unfortunately, certain forms of biological data are
typically available in the opposite orientation - variables are in rows and
observations are in columns. This is primarily true in feature data matrices,
e.g. gene expression counts matrices, where the number of variables (e.g. genes)
is much larger than the number of samples, which tend to be small very small
compared with the number of features. This format is convenient for humans to
interact with using, e.g. spreadsheet programs like Microsoft Excel, but can
unfortunately make performing certain operations on them challenging in
tidyverse.</p>
<p>For example, consider the microarray expression dataset in the previous section.
Each of the 54,676 rows is a probeset, and each of the 35 numeric columns is a
sample. This is a very large number of probesets to consider, especially if we
plan to conduct a statistical test on each, which would impose a substantial
<a href="#multiple-hypothesis-testing">multiple hypothesis testing</a> burden on our
results. We may therefore wish to eliminate probesets that have very low
variance from the overall dataset, since these probesets are not likely to have
a detectable statistical difference in our tests. However, computing the
variance for each probeset is a computation <em>across all columns</em>, not on columns
themselves, and this is not what tidyverse is designed to do well. Said
differently, R and tidyverse do not operate by default on the rows of a data
frame, tibble, or matrix.</p>
<p>Both base R and tidyverse are optimized to perform computations on columns, not
rows. The reasons for this are buried in the details of how the R program itself
was written to organize data internally and are beyond the scope of this book.
The consequence of this design choice is that, while we can perform operations
on the rows rather than the columns of a data structure, our code may perform
very poorly (i.e. take a very long time to run).</p>
<p>When working with these datasets, we have a couple options to deal with this
issue:</p>
<ul>
<li><p><strong>Pivot into long format.</strong> As described in the <a href="data-wrangling.html#rearranging-data">Rearranging Data</a> section, we
can rearrange our tibble to be more amenable to certain computations. In our
earlier example, we wish to group all of our measurements by probeset and
compute the variance of each, then possibly filter out probesets based on low
variance. We can therefore combine <code>pivot_longer()</code>, <code>group_by()</code>,
<code>summarize()</code>, and finally <code>left_join()</code> to perform this operation. Exactly
how to do this is left as an exercise in <a href="assignment-1.html#assignment-1">Assignment 1</a>.</p></li>
<li><p><strong>Compute row-wise statistics using <code>apply()</code>.</strong> As described in <a href="prog-basics.html#iteration">Iteration</a>,
R is a functional programming language and implements iteration in a
functional style using the <code>apply()</code> function. The <a href="prog-basics.html#apply-and-friends"><code>apply()   function</code></a> accepts a <code>MARGIN</code> argument of 1 or 2 if the
provided function is to be applied to the columns or rows, respectively. This
method can be used to compute a summary statistic on each row of a tibble and
the result saved back into the tibble using the column set operator:</p>
<div class="sourceCode" id="cb267"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb267-1"><a href="biology-bioinformatics.html#cb267-1" aria-hidden="true" tabindex="-1"></a>intensity_variance <span class="ot">&lt;-</span> <span class="fu">apply</span>(intensities, <span class="dv">2</span>, var)</span>
<span id="cb267-2"><a href="biology-bioinformatics.html#cb267-2" aria-hidden="true" tabindex="-1"></a>intensities<span class="sc">$</span>variance <span class="ot">&lt;-</span> intensity_variance</span></code></pre></div></li>
</ul>

</div>
</div>
<div id="bio-bioconductor" class="section level2" number="8.3">
<h2><span class="header-section-number">8.3</span> Bioconductor</h2>
<p><a href="https://www.bioconductor.org/">Bioconductor</a> is an organized collection of
strictly biological analysis methods packages for R. These packages are hosted
and maintained outside of <a href="https://cran.r-project.org/">CRAN</a> because the
maintainers enforce <a href="https://bioconductor.org/developers/package-guidelines/">a more rigorous set of coding quality, testing, and
documentation
standards</a> than is
required by normal R packages. These stricter requirements arose from a
recognition that software packages are generally only as usable as their
documentation allows them to be, and also that many if not most of the users of
these packages are not statisticians or experienced computer programmers.
Rather, they are people like us: biological analysis practitioners who may or
may not have substantial coding experience but must analyze our data
nonetheless. The excellent documentation and community support of the
bioconductor ecosystem is a primary reason why R is such a popular language in
biological analysis.</p>
<p>Bioconductor is divided into roughly two sets of packages: <em>core maintainer</em>
packages and user contributed packages. The core maintainer packages are among
the most critical, because they define a set of common objects and classes (e.g.
the <a href="https://www.bioconductor.org/packages/devel/bioc/vignettes/Biobase/inst/doc/ExpressionSetIntroduction.pdf">ExpressionSet class</a>
in the <a href="https://bioconductor.org/packages/release/bioc/html/Biobase.html">Biobase package</a>)
that all Bioconductor packages know how to work with. This common base provides
consistency among all Bioconductor packages thereby streamlining the learning
process. User contributed and maintained packages provide specialized
functionality for specific types of analysis.</p>
<p>Bioconductor itself must be installed prior to installing other Bioconductor
packages. To [install bioconductor], you can run the following:</p>
<div class="sourceCode" id="cb268"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb268-1"><a href="biology-bioinformatics.html#cb268-1" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">require</span>(<span class="st">&quot;BiocManager&quot;</span>, <span class="at">quietly =</span> <span class="cn">TRUE</span>))</span>
<span id="cb268-2"><a href="biology-bioinformatics.html#cb268-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">install.packages</span>(<span class="st">&quot;BiocManager&quot;</span>)</span>
<span id="cb268-3"><a href="biology-bioinformatics.html#cb268-3" aria-hidden="true" tabindex="-1"></a>BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="at">version =</span> <span class="st">&quot;3.14&quot;</span>)</span></code></pre></div>
<div class="box warning">
<p>Bioconductor undergoes regular releases indicated by a version, e.g. <code>version = "3.14"</code> at the time of writing. Every bioconductor package also has a version,
and each of those versions may or may not be compatible with a specific version
of Bioconductor. To make matters worse, Bioconductor versions are only
compatible with certain versions of R. Bioconductor version 3.14 requires R
version 4.1.0, and will not work with older R versions. These versions can cause
major version compatibility issues when you are forced to use older versions of
R, as may sometimes be the case on managed compute clusters. There is not a good
general solution for ensuring all your packages work together, but the general
best rule of thumb is to use the most current version of R and all packages at
the time when you write your code.</p>
</div>
<p>The <code>BiocManager</code> package is the only Bioconductor package installable using
<code>install.packages()</code>. After installing the BiocManager package, you may then
install bioconductor packages:</p>
<div class="sourceCode" id="cb269"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb269-1"><a href="biology-bioinformatics.html#cb269-1" aria-hidden="true" tabindex="-1"></a><span class="co"># installs the affy bioconductor package for microarray analysis</span></span>
<span id="cb269-2"><a href="biology-bioinformatics.html#cb269-2" aria-hidden="true" tabindex="-1"></a>BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="st">&quot;affy&quot;</span>)</span></code></pre></div>
<p>One key aspect of Bioconductor packages is consistent and helpful documentation;
every package page on the Bioconductor.org site lists a block of code that will
install the package, e.g. for the <a href="https://www.bioconductor.org/packages/release/bioc/html/affy.html">affy
package</a>. In
addition, Biconductor provides three types of documentation:</p>
<ul>
<li><a href="https://bioconductor.org/packages/release/workflows/">Workflow tutorials</a> on
how to perform specific analysis use cases</li>
<li><a href="https://bioconductor.org/help/package-vignettes/">Package vignettes</a> for
every package, which provide a worked example of how to use the package that
can be adapted by the user to their specific case</li>
<li>Detailed, consistently formatted reference documentation that gives precise
information on functionality and use of each package</li>
</ul>
<p>The thorough and useful documentation of packages in Bioconductor is one of the
reasons why the package ecosystem, and R, is so widely used in biology and
bioinformatics.</p>
<p>The base Bioconductor packages define convenient data structures for storing and
analyzing many types of data. Recall earlier in the <a href="biology-bioinformatics.html#bio-data-types">Types of Biological Data</a>
section, we described five types of biological data: primary, processed,
analysis, metadata, and annotation data. Bioconductor provides a convenient
class for storing many of these different data types together in one place,
specifically the SummarizedExperiment class in the <a href="https://bioconductor.org/packages/release/bioc/html/SummarizedExperiment.html">package of the same
name</a>
package. An illustration of what a SummarizedExperiment object stores is below,
from the Bioconductor maintainer’s paper in Nature:</p>
<div class="figure">
<img src="https://media.springernature.com/full/springer-static/image/art%3A10.1038%2Fnmeth.3252/MediaObjects/41592_2015_Article_BFnmeth3252_Fig2_HTML.jpg?as=webp" alt="" />
<p class="caption">SummarizedExperiment Schematic - Huber, et al. 2015. “Orchestrating High-Throughput Genomic Analysis with Bioconductor.” Nature Methods 12 (2): 115–21.</p>
</div>
<p>As you can see from the figure, this class stores processed data (<code>assays</code>),
metadata (<code>colData</code> and <code>exptData</code>), and annotation data (<code>rowData</code>). The
SummarizedExperiment class is used ubiquitously throughout the Bioconductor
package ecosystem, as are other core classes some of which we will cover later
in this chapter.</p>

</div>
<div id="gene-identifiers" class="section level2" number="8.4">
<h2><span class="header-section-number">8.4</span> Gene Identifiers</h2>
<div id="gene-identifier-systems" class="section level3" number="8.4.1">
<h3><span class="header-section-number">8.4.1</span> Gene Identifier Systems</h3>
<p>Since the first genetic sequence for a gene was determined in 1965 - <a href="https://pubmed.ncbi.nlm.nih.gov/14263761/">an alanine
tRNA in yeast</a><span class="citation">(<a href="#ref-Holley1965-tu" role="doc-biblioref">Holley et al. 1965</a>)</span> -
scientists have been trying to give them names. In 1979, <a href="https://pubmed.ncbi.nlm.nih.gov/295268/">formal guidelines for
the nomenclature used for human genes was
proposed</a><span class="citation">(<a href="#ref-Shows1979-iw" role="doc-biblioref">Shows et al. 1979</a>)</span> along with the
founding of the Human Gene Nomenclature Committee <span class="citation">(<a href="#ref-Bruford2020-zs" role="doc-biblioref">Bruford et al. 2020</a>)</span>, so that
researchers would have a common vocabulary to refer to genes in a consistent
way. In 1989, this committee was placed under the auspices of the Human Genome
Organisation (HUGO), becoming the HUGO Gene Nomenclature Committee (HGNC). The
HGNC has been the official body providing guidelines and gene naming authority
since, and as such HGNC gene names are often called <em>gene symbols</em>.</p>
<div class="box note">
<p>As per <span class="citation"><a href="#ref-Bruford2020-zs" role="doc-biblioref">Bruford et al.</a> (<a href="#ref-Bruford2020-zs" role="doc-biblioref">2020</a>)</span>, the official human gene symbol guidelines are as
follows:</p>
<ol style="list-style-type: decimal">
<li>Each gene is assigned a unique symbol, HGNC ID and descriptive name.</li>
<li>Symbols contain only uppercase Latin letters and Arabic numerals.</li>
<li>Symbols should not be the same as commonly used abbreviations</li>
<li>Nomenclature should not contain reference to any species or “G” for gene.</li>
<li>Nomenclature should not be offensive or pejorative.</li>
</ol>
</div>
<p>Gene symbols are the most human-readable system for naming genes. The gene
symbols <em>BRCA1</em>, <em>APOE</em>, and <em>ACE2</em> may be familiar to you as they are involved
in common diseases, namely breast cancer, Alzheimer’s Disease risk, and
SARS-COV-2, respectively. Typically, the gene symbol is an acronym that roughly
represents a label of what the gene is or does (or was originally found to be or
do, as many genes are subsequently discovered to be involved in entirely
separate processes as well), e.g. <em>APOE</em> represents the gene <a href="https://www.ensembl.org/Homo_sapiens/Gene/Summary?db=core;g=ENSG00000130203;r=19:44905791-44909393">apolipoprotein
E</a>.
This convention is convenient for humans when reading and identifying genes.
However, standardized though the symbol conventions may be, they are not always
easy for computers to work with, and ambiguities can cause mapping problems.</p>
<p>Other gene identifier systems developed along with other gene information
databases. In 2000, the <a href="https://www.ensembl.org/">Ensembl</a> genome browser was
launched by the <a href="https://wellcome.org/">Wellcome Trust</a>, a charitable foundation
based in the United Kingdom, with the goal of providing automated annotation of
the human genome. The Ensembl Project, which supports the genome browser,
recognized even before the publication of the human genome that manual
annotation and curation of genes is a slow and labor intensive process that
would not provide researchers around the world timely access to information. The
project therefore created a set of automated tools and pipelines to collect,
process, and publish rapid and consistent annotation of genes in the human
genome. Since its initial release, Ensembl now supports over 50,000 different
genomes.</p>
<p>Ensembl assigns a automatic, systematic ID called the Ensembl Gene ID to every
gene in its database. Human Ensembl Gene IDs take the form <code>ENSG</code> plus an 11
digit number, optionally followed by a period delimited version number. For
example, at the time of writing the BRCA1 gene has an Ensembl Gene ID of
<code>ENSG00000012048.23</code>. The stable ID portion (i.e. <code>ENSGNNNNNNNNNNN</code>) will remain
associated with the gene forever (unless the gene annotation <a href="http://useast.ensembl.org/Help/View?id=181">changes
“dramatically” in which case it is
retired</a>). The <code>.23</code> is the version
number of the gene annotation, meaning this gene has been updated 22 times (plus
its initial version) since addition to the database. The additional version
information is very important for reproducibility of biological analysis, since
conclusions drawn by results of these analyses are usually based on the most
current information about a gene which are continually updated over time.</p>
<div class="box note">
<p>Ensembl maintains annotations for many different organisms, and the gene
identifiers for each genome contain codes that indicate which organism the gene
is for. Here are the codes for genes in several different species:</p>
<table style="width:92%;">
<colgroup>
<col width="23%" />
<col width="36%" />
<col width="31%" />
</colgroup>
<thead>
<tr class="header">
<th>Gene ID Prefix</th>
<th>Organism</th>
<th>Example</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><code>ENSG</code></td>
<td>Homo sapiens</td>
<td><a href="https://useast.ensembl.org/Homo_sapiens/Gene/Summary?db=core;g=ENSG00000139618;r=13:32315086-32400268">ENSG00000139618</a></td>
</tr>
<tr class="even">
<td><code>ENSMUSG</code></td>
<td>Mus musculus (mouse)</td>
<td><a href="https://useast.ensembl.org/Mus_musculus/Gene/Summary?db=core;g=ENSMUSG00000017167;r=11:101061349-101081550">ENSMUSG00000017167</a></td>
</tr>
<tr class="odd">
<td><code>ENSDARG</code></td>
<td>Danio rerio (zebrafish)</td>
<td><a href="https://useast.ensembl.org/Danio_rerio/Gene/Summary?db=core;g=ENSDARG00000024771;r=18:5213338-5227420;t=ENSDART00000033574">ENSDARG00000024771</a></td>
</tr>
<tr class="even">
<td><code>ENSMMUG</code></td>
<td>Macaca mulata (Macaque)</td>
<td><a href="https://useast.ensembl.org/Macaca_mulatta/Gene/Summary?db=core;g=ENSMMUG00000020312;r=5:65376507-65505566">ENSMMUG00000020312</a></td>
</tr>
<tr class="odd">
<td><code>ENSVPAG</code></td>
<td>Vicugna pacos (Alpaca)</td>
<td><a href="https://useast.ensembl.org/Vicugna_pacos/Gene/Summary?db=core;g=ENSVPAG00000001990;r=GeneScaffold_2392:178775-179671;t=ENSVPAT00000001989">ENSVPAG00000001990</a></td>
</tr>
</tbody>
</table>
</div>
<p>Ensembl Gene IDs and gene symbols are the most commonly used gene identifiers.
The <a href="https://www.gencodegenes.org/">GENCODE project</a> which provides consistent
and stable genome annotation releases for human and mouse genomes, uses these
two types of identifiers with its official annotations.</p>
<div class="box note">
<p>Ensembl provides stable identifiers for transcripts as well as genes. Transcript
identifiers correspond to distinct patterns of exons/introns that have been
identified for a gene, and each gene has one or more distinct transcripts.
Ensembl Transcript IDs have the form <code>ENSTNNNNNNNNNNN.NN</code> similar to Gene IDs.</p>
</div>
<div class="box note">
<p>Ensembl is not the only gene identifier system besides gene symbol. Several
other databases have devised and maintain their own identifiers, most notably
[Entrez Gene IDs](<a href="">UCSC Gene IDs</a> used by the <a href="https://www.ncbi.nlm.nih.gov/gene">NCBI Gene
database</a> which assigns a unique integer to
each gene in each organism, and the <a href="https://www.omim.org/">Online Mendelian Inheritance in Man
(OMIM)</a> database, which has identifiers that look like
<code>OMIM:NNNNN</code>, where each OMIM ID refers to a unique gene or human phenotype.
However, the primary identifiers used in modern human biological research remain
Ensembl IDs and official HGNC gene symbols.</p>
</div>
</div>
<div id="mapping-between-identifier-systems" class="section level3" number="8.4.2">
<h3><span class="header-section-number">8.4.2</span> Mapping Between Identifier Systems</h3>
<p>A very common operation in biological analysis is to map between gene identifier
systems, e.g. you are given Ensembl Gene ID and want to map to the more
human-recognizable gene symbols. Ensembl provides a service called
<a href="https://www.ensembl.org/biomart/martview">BioMart</a> that allows you to download
annotation information for genes, transcripts, and other information maintained
in their databases. It also provides limited access to external sources of
information on their genes, including cross references to HGNC gene symbols and
some other gene identifier systems. The Ensembl website has <a href="https://useast.ensembl.org/info/data/biomart/index.html">helpful
documentation</a> on how
to use BioMart to download annotation data using the web interface.</p>
<p>In addition to downloading annotations as a CSV file from the web interface,
Ensembl also provides <a href="https://useast.ensembl.org/info/data/biomart/biomart_r_package.html">the biomaRt Bioconductor
package</a> to
allow programmatic access to the same information directly from R. There is <a href="https://useast.ensembl.org/info/genome/index.html">much
more information in the Ensembl
databases</a> than gene
annotation data that can be accessed via BioMart, but we will provide a brief
example of how to extract gene information here:</p>
<div class="sourceCode" id="cb270"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb270-1"><a href="biology-bioinformatics.html#cb270-1" aria-hidden="true" tabindex="-1"></a><span class="co"># this assumes biomaRt is already installed through bioconductor</span></span>
<span id="cb270-2"><a href="biology-bioinformatics.html#cb270-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(biomaRt)</span>
<span id="cb270-3"><a href="biology-bioinformatics.html#cb270-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb270-4"><a href="biology-bioinformatics.html#cb270-4" aria-hidden="true" tabindex="-1"></a><span class="co"># the human biomaRt database is named &quot;hsapiens_gene_ensembl&quot;</span></span>
<span id="cb270-5"><a href="biology-bioinformatics.html#cb270-5" aria-hidden="true" tabindex="-1"></a>ensembl <span class="ot">&lt;-</span> <span class="fu">useEnsembl</span>(<span class="at">biomart=</span><span class="st">&quot;ensembl&quot;</span>, <span class="at">dataset=</span><span class="st">&quot;hsapiens_gene_ensembl&quot;</span>)</span>
<span id="cb270-6"><a href="biology-bioinformatics.html#cb270-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb270-7"><a href="biology-bioinformatics.html#cb270-7" aria-hidden="true" tabindex="-1"></a><span class="co"># listAttributes() returns a data frame, turn into a tibble to help with formatting</span></span>
<span id="cb270-8"><a href="biology-bioinformatics.html#cb270-8" aria-hidden="true" tabindex="-1"></a><span class="fu">as_tibble</span>(<span class="fu">listAttributes</span>(ensembl))</span>
<span id="cb270-9"><a href="biology-bioinformatics.html#cb270-9" aria-hidden="true" tabindex="-1"></a><span class="co"># A tibble: 3,143 x 3</span></span>
<span id="cb270-10"><a href="biology-bioinformatics.html#cb270-10" aria-hidden="true" tabindex="-1"></a>   name                          description                  page        </span>
<span id="cb270-11"><a href="biology-bioinformatics.html#cb270-11" aria-hidden="true" tabindex="-1"></a>   <span class="sc">&lt;</span>chr<span class="sc">&gt;</span>                         <span class="er">&lt;</span>chr<span class="sc">&gt;</span>                        <span class="er">&lt;</span>chr<span class="sc">&gt;</span>       </span>
<span id="cb270-12"><a href="biology-bioinformatics.html#cb270-12" aria-hidden="true" tabindex="-1"></a> <span class="dv">1</span> ensembl_gene_id               Gene stable ID               feature_page</span>
<span id="cb270-13"><a href="biology-bioinformatics.html#cb270-13" aria-hidden="true" tabindex="-1"></a> <span class="dv">2</span> ensembl_gene_id_version       Gene stable ID version       feature_page</span>
<span id="cb270-14"><a href="biology-bioinformatics.html#cb270-14" aria-hidden="true" tabindex="-1"></a> <span class="dv">3</span> ensembl_transcript_id         Transcript stable ID         feature_page</span>
<span id="cb270-15"><a href="biology-bioinformatics.html#cb270-15" aria-hidden="true" tabindex="-1"></a> <span class="dv">4</span> ensembl_transcript_id_version Transcript stable ID version feature_page</span>
<span id="cb270-16"><a href="biology-bioinformatics.html#cb270-16" aria-hidden="true" tabindex="-1"></a> <span class="dv">5</span> ensembl_peptide_id            Protein stable ID            feature_page</span>
<span id="cb270-17"><a href="biology-bioinformatics.html#cb270-17" aria-hidden="true" tabindex="-1"></a> <span class="dv">6</span> ensembl_peptide_id_version    Protein stable ID version    feature_page</span>
<span id="cb270-18"><a href="biology-bioinformatics.html#cb270-18" aria-hidden="true" tabindex="-1"></a> <span class="dv">7</span> ensembl_exon_id               Exon stable ID               feature_page</span>
<span id="cb270-19"><a href="biology-bioinformatics.html#cb270-19" aria-hidden="true" tabindex="-1"></a> <span class="dv">8</span> description                   Gene description             feature_page</span>
<span id="cb270-20"><a href="biology-bioinformatics.html#cb270-20" aria-hidden="true" tabindex="-1"></a> <span class="dv">9</span> chromosome_name               Chromosome<span class="sc">/</span>scaffold name     feature_page</span>
<span id="cb270-21"><a href="biology-bioinformatics.html#cb270-21" aria-hidden="true" tabindex="-1"></a><span class="dv">10</span> start_position                Gene <span class="fu">start</span> (bp)              feature_page</span>
<span id="cb270-22"><a href="biology-bioinformatics.html#cb270-22" aria-hidden="true" tabindex="-1"></a><span class="co"># ... with 3,133 more rows</span></span></code></pre></div>
<p>The <code>name</code> column provides the programmatic name associated with the attribute
that can be used to retrieve that annotation information using the <code>getBM()</code>
function:</p>
<div class="sourceCode" id="cb271"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb271-1"><a href="biology-bioinformatics.html#cb271-1" aria-hidden="true" tabindex="-1"></a><span class="co"># create a tibble with ensembl gene ID, HGNC gene symbol, and gene description</span></span>
<span id="cb271-2"><a href="biology-bioinformatics.html#cb271-2" aria-hidden="true" tabindex="-1"></a>gene_map <span class="ot">&lt;-</span> <span class="fu">as_tibble</span>(</span>
<span id="cb271-3"><a href="biology-bioinformatics.html#cb271-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">getBM</span>(</span>
<span id="cb271-4"><a href="biology-bioinformatics.html#cb271-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">attributes=</span><span class="fu">c</span>(<span class="st">&quot;ensembl_gene_id&quot;</span>, <span class="st">&quot;hgnc_gene_symbol&quot;</span>, <span class="st">&quot;description&quot;</span>),</span>
<span id="cb271-5"><a href="biology-bioinformatics.html#cb271-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">mart=</span>ensembl</span>
<span id="cb271-6"><a href="biology-bioinformatics.html#cb271-6" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb271-7"><a href="biology-bioinformatics.html#cb271-7" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb271-8"><a href="biology-bioinformatics.html#cb271-8" aria-hidden="true" tabindex="-1"></a>gene_map</span>
<span id="cb271-9"><a href="biology-bioinformatics.html#cb271-9" aria-hidden="true" tabindex="-1"></a><span class="co"># A tibble: 68,012 x 3</span></span>
<span id="cb271-10"><a href="biology-bioinformatics.html#cb271-10" aria-hidden="true" tabindex="-1"></a>   ensembl_gene_id hgnc_symbol description                                                               </span>
<span id="cb271-11"><a href="biology-bioinformatics.html#cb271-11" aria-hidden="true" tabindex="-1"></a>   <span class="sc">&lt;</span>chr<span class="sc">&gt;</span>           <span class="er">&lt;</span>chr<span class="sc">&gt;</span>       <span class="er">&lt;</span>chr<span class="sc">&gt;</span>                                                                     </span>
<span id="cb271-12"><a href="biology-bioinformatics.html#cb271-12" aria-hidden="true" tabindex="-1"></a> <span class="dv">1</span> ENSG00000210049 MT<span class="sc">-</span>TF       mitochondrially encoded tRNA<span class="sc">-</span><span class="fu">Phe</span> (UUU<span class="sc">/</span>C) [Source<span class="sc">:</span>HGNC Symbol;Acc<span class="sc">:</span>HGNC<span class="sc">:</span><span class="dv">748</span><span class="sc">~</span></span>
<span id="cb271-13"><a href="biology-bioinformatics.html#cb271-13" aria-hidden="true" tabindex="-1"></a> <span class="dv">2</span> ENSG00000211459 MT<span class="sc">-</span>RNR1     mitochondrially encoded 12S rRNA [Source<span class="sc">:</span>HGNC Symbol;Acc<span class="sc">:</span>HGNC<span class="sc">:</span><span class="dv">7470</span>]       </span>
<span id="cb271-14"><a href="biology-bioinformatics.html#cb271-14" aria-hidden="true" tabindex="-1"></a> <span class="dv">3</span> ENSG00000210077 MT<span class="sc">-</span>TV       mitochondrially encoded tRNA<span class="sc">-</span><span class="fu">Val</span> (GUN) [Source<span class="sc">:</span>HGNC Symbol;Acc<span class="sc">:</span>HGNC<span class="sc">:</span><span class="dv">7500</span>]</span>
<span id="cb271-15"><a href="biology-bioinformatics.html#cb271-15" aria-hidden="true" tabindex="-1"></a> <span class="dv">4</span> ENSG00000210082 MT<span class="sc">-</span>RNR2     mitochondrially encoded 16S rRNA [Source<span class="sc">:</span>HGNC Symbol;Acc<span class="sc">:</span>HGNC<span class="sc">:</span><span class="dv">7471</span>]       </span>
<span id="cb271-16"><a href="biology-bioinformatics.html#cb271-16" aria-hidden="true" tabindex="-1"></a> <span class="dv">5</span> ENSG00000209082 MT<span class="sc">-</span>TL1      mitochondrially encoded tRNA<span class="sc">-</span><span class="fu">Leu</span> (UUA<span class="sc">/</span>G) <span class="dv">1</span> [Source<span class="sc">:</span>HGNC Symbol;Acc<span class="sc">:</span>HGNC<span class="sc">:</span><span class="dv">7</span><span class="sc">~</span></span>
<span id="cb271-17"><a href="biology-bioinformatics.html#cb271-17" aria-hidden="true" tabindex="-1"></a> <span class="dv">6</span> ENSG00000198888 MT<span class="sc">-</span>ND1      mitochondrially encoded NADH<span class="sc">:</span>ubiquinone oxidoreductase core subunit <span class="dv">1</span> [So<span class="sc">~</span></span>
<span id="cb271-18"><a href="biology-bioinformatics.html#cb271-18" aria-hidden="true" tabindex="-1"></a> <span class="dv">7</span> ENSG00000210100 MT<span class="sc">-</span>TI       mitochondrially encoded tRNA<span class="sc">-</span><span class="fu">Ile</span> (AUU<span class="sc">/</span>C) [Source<span class="sc">:</span>HGNC Symbol;Acc<span class="sc">:</span>HGNC<span class="sc">:</span><span class="dv">748</span><span class="sc">~</span></span>
<span id="cb271-19"><a href="biology-bioinformatics.html#cb271-19" aria-hidden="true" tabindex="-1"></a> <span class="dv">8</span> ENSG00000210107 MT<span class="sc">-</span>TQ       mitochondrially encoded tRNA<span class="sc">-</span><span class="fu">Gln</span> (CAA<span class="sc">/</span>G) [Source<span class="sc">:</span>HGNC Symbol;Acc<span class="sc">:</span>HGNC<span class="sc">:</span><span class="dv">749</span><span class="sc">~</span></span>
<span id="cb271-20"><a href="biology-bioinformatics.html#cb271-20" aria-hidden="true" tabindex="-1"></a> <span class="dv">9</span> ENSG00000210112 MT<span class="sc">-</span>TM       mitochondrially encoded tRNA<span class="sc">-</span><span class="fu">Met</span> (AUA<span class="sc">/</span>G) [Source<span class="sc">:</span>HGNC Symbol;Acc<span class="sc">:</span>HGNC<span class="sc">:</span><span class="dv">749</span><span class="sc">~</span></span>
<span id="cb271-21"><a href="biology-bioinformatics.html#cb271-21" aria-hidden="true" tabindex="-1"></a><span class="dv">10</span> ENSG00000198763 MT<span class="sc">-</span>ND2      mitochondrially encoded NADH<span class="sc">:</span>ubiquinone oxidoreductase core subunit <span class="dv">2</span> [So<span class="sc">~</span></span>
<span id="cb271-22"><a href="biology-bioinformatics.html#cb271-22" aria-hidden="true" tabindex="-1"></a><span class="co"># ... with 68,002 more rows</span></span></code></pre></div>
<p>With this Ensembl Gene ID to HGNC symbol mapping in hand, you can combine the
tibble above with other tibbles containing gene information by <a href="data-wrangling.html#relational-data">joining them
together</a>.</p>
<div class="box note">
<p>BioMart/biomaRt is not the only ways to map different gene identifiers. The
Bioconductor package
<a href="https://bioconductor.org/packages/release/bioc/html/AnnotationDbi.html">AnnotateDbi</a>
also provides this functionality in a flexible format independent of the Ensembl
databases. This package includes not only gene identifier mapping and
information, but also identifiers from technology platforms, e.g. probe set IDs
from <a href="biology-bioinformatics.html#microarrays">microarrays</a>, that can help when working with these types of
data. The package also allows comprehensive and flexible <a href="biology-bioinformatics.html#mapping-homologs">homolog
mapping</a>. As with all Bioconductor packages, the
AnnotationDbi documentation is well written and comprehensive, though knowledge
of <a href="https://en.wikipedia.org/wiki/Relational_database">relational databases</a> is
helpful in understanding how the packages work.</p>
</div>
<div class="box readmore">
<ul>
<li><a href="https://www.ensembl.org/biomart/martview">Ensembl BioMart portal</a></li>
<li><a href="https://useast.ensembl.org/info/data/biomart/index.html">Ensembl BioMart web portal documentation</a></li>
<li><a href="http://www.bioconductor.org/packages/release/bioc/html/biomaRt.html">biomaRt Bioconductor documentation</a></li>
</ul>
</div>
</div>
<div id="mapping-homologs" class="section level3" number="8.4.3">
<h3><span class="header-section-number">8.4.3</span> Mapping Homologs</h3>
<p>Sometimes it is necessary to link datasets from different organisms together by
orthology. For example, in an experiment performed in mice, we might be
interested in comparing gene expression patterns observed in the mouse samples
to a publicly available human dataset. In these contexts, we must link gene
identifiers from one organism to their corresponding homologs in the other.
BioMart enables us to extract these linked identifiers by explicitly <a href="https://bioconductor.org/packages/release/bioc/vignettes/biomaRt/inst/doc/accessing_ensembl.html#given-the-human-gene-tp53-retrieve-the-human-chromosomal-location-of-this-gene-and-also-retrieve-the-chromosomal-location-and-refseq-id-of-its-homolog-in-mouse.">connecting
different biomaRt
databases</a>
with the <code>getLDS()</code> function:</p>
<div class="sourceCode" id="cb272"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb272-1"><a href="biology-bioinformatics.html#cb272-1" aria-hidden="true" tabindex="-1"></a>human_db <span class="ot">&lt;-</span> <span class="fu">useEnsembl</span>(<span class="st">&quot;ensembl&quot;</span>, <span class="at">dataset =</span> <span class="st">&quot;hsapiens_gene_ensembl&quot;</span>)</span>
<span id="cb272-2"><a href="biology-bioinformatics.html#cb272-2" aria-hidden="true" tabindex="-1"></a>mouse_db <span class="ot">&lt;-</span> <span class="fu">useEnsembl</span>(<span class="st">&quot;ensembl&quot;</span>, <span class="at">dataset =</span> <span class="st">&quot;mmusculus_gene_ensembl&quot;</span>)</span>
<span id="cb272-3"><a href="biology-bioinformatics.html#cb272-3" aria-hidden="true" tabindex="-1"></a>orth_map <span class="ot">&lt;-</span> <span class="fu">as_tibble</span>(</span>
<span id="cb272-4"><a href="biology-bioinformatics.html#cb272-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">getLDS</span>(<span class="at">attributes =</span> <span class="fu">c</span>(<span class="st">&quot;ensembl_gene_id&quot;</span>, <span class="st">&quot;hgnc_symbol&quot;</span>),</span>
<span id="cb272-5"><a href="biology-bioinformatics.html#cb272-5" aria-hidden="true" tabindex="-1"></a>       <span class="at">mart =</span> human_db,</span>
<span id="cb272-6"><a href="biology-bioinformatics.html#cb272-6" aria-hidden="true" tabindex="-1"></a>       <span class="at">attributesL =</span> <span class="fu">c</span>(<span class="st">&quot;ensembl_gene_id&quot;</span>, <span class="st">&quot;mgi_symbol&quot;</span>),</span>
<span id="cb272-7"><a href="biology-bioinformatics.html#cb272-7" aria-hidden="true" tabindex="-1"></a>       <span class="at">martL =</span> mouse_db</span>
<span id="cb272-8"><a href="biology-bioinformatics.html#cb272-8" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb272-9"><a href="biology-bioinformatics.html#cb272-9" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb272-10"><a href="biology-bioinformatics.html#cb272-10" aria-hidden="true" tabindex="-1"></a>orth_map</span>
<span id="cb272-11"><a href="biology-bioinformatics.html#cb272-11" aria-hidden="true" tabindex="-1"></a><span class="co"># A tibble: 26,390 x 4</span></span>
<span id="cb272-12"><a href="biology-bioinformatics.html#cb272-12" aria-hidden="true" tabindex="-1"></a>   Gene.stable.ID  HGNC.symbol Gene.stable.ID<span class="fl">.1</span>   MGI.symbol</span>
<span id="cb272-13"><a href="biology-bioinformatics.html#cb272-13" aria-hidden="true" tabindex="-1"></a>   <span class="sc">&lt;</span>chr<span class="sc">&gt;</span>           <span class="er">&lt;</span>chr<span class="sc">&gt;</span>       <span class="er">&lt;</span>chr<span class="sc">&gt;</span>              <span class="er">&lt;</span>chr<span class="sc">&gt;</span>     </span>
<span id="cb272-14"><a href="biology-bioinformatics.html#cb272-14" aria-hidden="true" tabindex="-1"></a> <span class="dv">1</span> ENSG00000198695 MT<span class="sc">-</span>ND6      ENSMUSG00000064368 <span class="st">&quot;mt-Nd6&quot;</span>  </span>
<span id="cb272-15"><a href="biology-bioinformatics.html#cb272-15" aria-hidden="true" tabindex="-1"></a> <span class="dv">2</span> ENSG00000212907 MT<span class="sc">-</span>ND4L     ENSMUSG00000065947 <span class="st">&quot;mt-Nd4l&quot;</span></span>
<span id="cb272-16"><a href="biology-bioinformatics.html#cb272-16" aria-hidden="true" tabindex="-1"></a> <span class="dv">3</span> ENSG00000279169 PRAMEF13    ENSMUSG00000094303 <span class="st">&quot;&quot;</span>        </span>
<span id="cb272-17"><a href="biology-bioinformatics.html#cb272-17" aria-hidden="true" tabindex="-1"></a> <span class="dv">4</span> ENSG00000279169 PRAMEF13    ENSMUSG00000094722 <span class="st">&quot;&quot;</span>        </span>
<span id="cb272-18"><a href="biology-bioinformatics.html#cb272-18" aria-hidden="true" tabindex="-1"></a> <span class="dv">5</span> ENSG00000279169 PRAMEF13    ENSMUSG00000095666 <span class="st">&quot;&quot;</span>        </span>
<span id="cb272-19"><a href="biology-bioinformatics.html#cb272-19" aria-hidden="true" tabindex="-1"></a> <span class="dv">6</span> ENSG00000279169 PRAMEF13    ENSMUSG00000094741 <span class="st">&quot;&quot;</span>        </span>
<span id="cb272-20"><a href="biology-bioinformatics.html#cb272-20" aria-hidden="true" tabindex="-1"></a> <span class="dv">7</span> ENSG00000279169 PRAMEF13    ENSMUSG00000094836 <span class="st">&quot;&quot;</span>        </span>
<span id="cb272-21"><a href="biology-bioinformatics.html#cb272-21" aria-hidden="true" tabindex="-1"></a> <span class="dv">8</span> ENSG00000279169 PRAMEF13    ENSMUSG00000074720 <span class="st">&quot;&quot;</span>        </span>
<span id="cb272-22"><a href="biology-bioinformatics.html#cb272-22" aria-hidden="true" tabindex="-1"></a> <span class="dv">9</span> ENSG00000279169 PRAMEF13    ENSMUSG00000096236 <span class="st">&quot;&quot;</span>        </span>
<span id="cb272-23"><a href="biology-bioinformatics.html#cb272-23" aria-hidden="true" tabindex="-1"></a><span class="dv">10</span> ENSG00000198763 MT<span class="sc">-</span>ND2      ENSMUSG00000064345 <span class="st">&quot;mt-Nd2&quot;</span>  </span>
<span id="cb272-24"><a href="biology-bioinformatics.html#cb272-24" aria-hidden="true" tabindex="-1"></a><span class="co"># ... with 26,380 more rows</span></span></code></pre></div>
<p>The <code>mgi_symbol</code> field refers to the gene symbol assigned in the <a href="https://www.jax.org/research-and-faculty/resources/mouse-genome-informatics">Mouse Genome
Informatics database</a>
maintained by <a href="https://www.jax.org/">The Jackson Laboratory</a>.</p>

</div>
</div>
<div id="gene-expression" class="section level2" number="8.5">
<h2><span class="header-section-number">8.5</span> Gene Expression</h2>
<p>Gene expression is the process by which information from a gene is used in the
synthesis of functional gene products that affect a phenotype. While gene
expression studies are often focused on protein coding genes, there are many
other functional molecules, e.g. <a href="https://www.genome.gov/genetics-glossary/Transfer-RNA">transfer
RNAs</a>,
<a href="https://en.wikipedia.org/wiki/Long_non-coding_RNA">lncRNAs</a>, etc, that are
produced by this process. The gene expression process has many steps and
intermediate products, as depicted in the following figure:</p>
<div class="figure">
<img src="gene_expression_nature.jpg" alt="" />
<p class="caption">The Gene Expression Process - An overview of the flow of information from DNA
to protein in a eukaryote</p>
</div>
<p>The specific parts of the genome that code for genes are copied, or
<em>transcribed</em> into RNA molecules called <em>transcripts</em>. In lower organisms like
bacteria, these RNA molecules are passed on directly to ribosomes, which
translate them into proteins. In most higher organisms like eukaryotes, these
initial transcripts, called <em>pre-messenger RNA (pre-mRNA) transcripts</em> are
further processed such that certain parts of the transcript, called <em>introns</em>,
are spliced out and the flanking sequences, called <em>exons</em>, are concatenated
together. After this splicing process is complete, the pre-mRNA transcripts
become mature <em>messenger RNA (mRNA) transcripts</em> which go on to be exported from
the nucleus and loaded into ribosomes in the cytoplasm to undergo translation
into proteins.</p>
<p>In gene expression studies, the relative abundance, or number of copies, of RNA
or mRNA transcripts in a sample is measured. The measurements are non-negative
numbers that are proportional to the relative abundance of a transcript with
respect to some reference, either another gene, or in the case of high
throughput assays like microarrays or high throughput sequencing, relative to
measurements of all distinct transcripts examined in the experiment.
Conceptually, the larger the magnitude of a transcript abundance measurement,
the more copies of that transcript were in the original sample.</p>
<p>There are many ways to measure the abundance of RNA transcripts; the following
are some of the common methods at the time of writing:</p>
<ul>
<li>Light absorbance - RNA absorbs ultraviolet light with wavelength 260 nm, which
can be used to determine RNA concentration in a sample using specialized equipment</li>
<li><a href="https://www.genome.gov/genetics-glossary/Northern-Blot">Northern blot</a> -
measures relative abundance of an RNA with a specific sequence</li>
<li><a href="https://www.thermofisher.com/blog/ask-a-scientist/what-is-qpcr/">quantitative polymerase chain reaction
(qPCR)</a> -
measures relative abundance of an RNA with a specific sequence using <a href="https://en.wikipedia.org/wiki/Polymerase_chain_reaction">PCR
amplification</a></li>
<li><a href="biology-bioinformatics.html#microarrays">Oligonucleotide and microarrays</a> - measures relative abundance
of thousands of genes simultaneously using known DNA probe sequences and
fluorescently tagged RNA molecules</li>
<li><a href="#rnaseq">High throughput RNA sequencing (RNASeq)</a> - measures thousands to
millions of RNA fragments simultaneously in proportion to their relative
abundance</li>
</ul>
<p>While any of the measurement methods above may be analyzed in R, the high
throughput methods (i.e. microarray, high throughput sequencing) are the primary
concern of this chapter. These methods generate measurements for thousands of
transcripts or genes simultaneously, requiring the power and flexibility of
programmatic analysis to process in a practical amount of time. The remainder of
this chapter is devoted to understanding the specific technologies, data types,
and analytical strategies involved in working with these data.</p>
<div class="box important">
<p>Gene expression measurements are almost always inherently relative, either due
to limitations of the measurement methods (e.g. microarrays, described below) or
because measuring all molecules in a sample will almost always be prohibitively
expensive, have diminishing returns, and it is very difficult if not impossible
to determine if all the molecules have been measured. This means we cannot in
general associate the numbers associated with the measurements with an absolute
molecular copy number. An important implication of the inherent relativity of
these measurements is: <strong>absence of evidence is not evidence of absence</strong>. In
other words, if a transcript has an abundance measurement of zero, this does
<em>not necessarily imply that the gene is not expressed</em>. It may be that the gene
is indeed expressed, but the copy number is sufficiently small that it was not
detected by the assay.</p>
</div>
<div id="gene-expression-data-in-bioconductor" class="section level3" number="8.5.1">
<h3><span class="header-section-number">8.5.1</span> Gene Expression Data in Bioconductor</h3>
<p>The <a href="https://bioconductor.org/packages/release/bioc/html/SummarizedExperiment.html">SummarizedExperiment
container</a>
is the standard way to load and work with gene expression data in Bioconductor.
This container requires the following information:</p>
<ul>
<li><code>assays</code> - one or more measurement assays (e.g. gene expression) in the form of a feature by sample matrix</li>
<li><code>colData</code> - metadata associated with the samples (i.e. columns) of the assays</li>
<li><code>rowData</code> - metadata associated with the features (i.e. rows) of the assays</li>
<li><code>exptData</code> - additional metadata about the experiment itself, like protocol, project name, etc</li>
</ul>
<p>The figure below illustrates how the SummarizedExperiment container is
structured and how the different data elements are accessed:</p>
<div class="figure">
<img src="https://media.springernature.com/full/springer-static/image/art%3A10.1038%2Fnmeth.3252/MediaObjects/41592_2015_Article_BFnmeth3252_Fig2_HTML.jpg?as=webp" alt="" />
<p class="caption">SummarizedExperiment Schematic - Huber, et al. 2015. “Orchestrating High-Throughput Genomic Analysis with Bioconductor.” Nature Methods 12 (2): 115–21.</p>
</div>
<p>Many Bioconductor packages for specific types of data, e.g.
<a href="#differential-expression--microarrays-limma">limma</a> create these
SummarizedExperiment objects for you, but you may also create your own by
assembling each of these data into data frames manually:</p>
<div class="sourceCode" id="cb273"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb273-1"><a href="biology-bioinformatics.html#cb273-1" aria-hidden="true" tabindex="-1"></a><span class="co"># microarray expression dataset intensities</span></span>
<span id="cb273-2"><a href="biology-bioinformatics.html#cb273-2" aria-hidden="true" tabindex="-1"></a>intensities <span class="ot">&lt;-</span> readr<span class="sc">::</span><span class="fu">read_delim</span>(<span class="st">&quot;example_intensity_data.csv&quot;</span>,<span class="at">delim=</span><span class="st">&quot; &quot;</span>)</span>
<span id="cb273-3"><a href="biology-bioinformatics.html#cb273-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb273-4"><a href="biology-bioinformatics.html#cb273-4" aria-hidden="true" tabindex="-1"></a><span class="co"># the first column of intensities tibble is the probesetId, extract to pass as rowData</span></span>
<span id="cb273-5"><a href="biology-bioinformatics.html#cb273-5" aria-hidden="true" tabindex="-1"></a>rowData <span class="ot">&lt;-</span> intensities[<span class="st">&quot;probeset_id&quot;</span>]</span>
<span id="cb273-6"><a href="biology-bioinformatics.html#cb273-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb273-7"><a href="biology-bioinformatics.html#cb273-7" aria-hidden="true" tabindex="-1"></a><span class="co"># remove probeset IDs from tibble and turn into a R dataframe so that we can assign rownames</span></span>
<span id="cb273-8"><a href="biology-bioinformatics.html#cb273-8" aria-hidden="true" tabindex="-1"></a><span class="co"># since tibbles don&#39;t support row names</span></span>
<span id="cb273-9"><a href="biology-bioinformatics.html#cb273-9" aria-hidden="true" tabindex="-1"></a>intensities <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(</span>
<span id="cb273-10"><a href="biology-bioinformatics.html#cb273-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(intensities, <span class="sc">-</span>probeset_id)</span>
<span id="cb273-11"><a href="biology-bioinformatics.html#cb273-11" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb273-12"><a href="biology-bioinformatics.html#cb273-12" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(intensities) <span class="ot">&lt;-</span> rowData<span class="sc">$</span>probeset_id</span>
<span id="cb273-13"><a href="biology-bioinformatics.html#cb273-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb273-14"><a href="biology-bioinformatics.html#cb273-14" aria-hidden="true" tabindex="-1"></a><span class="co"># these column data are made up, but you would have a sample metadata file to use</span></span>
<span id="cb273-15"><a href="biology-bioinformatics.html#cb273-15" aria-hidden="true" tabindex="-1"></a>colData <span class="ot">&lt;-</span> <span class="fu">tibble</span>(</span>
<span id="cb273-16"><a href="biology-bioinformatics.html#cb273-16" aria-hidden="true" tabindex="-1"></a>  <span class="at">sample_name=</span><span class="fu">colnames</span>(intensities),</span>
<span id="cb273-17"><a href="biology-bioinformatics.html#cb273-17" aria-hidden="true" tabindex="-1"></a>  <span class="at">condition=</span><span class="fu">sample</span>(<span class="fu">c</span>(<span class="st">&quot;case&quot;</span>,<span class="st">&quot;control&quot;</span>),<span class="fu">ncol</span>(intensities),<span class="at">replace=</span><span class="cn">TRUE</span>)</span>
<span id="cb273-18"><a href="biology-bioinformatics.html#cb273-18" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb273-19"><a href="biology-bioinformatics.html#cb273-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb273-20"><a href="biology-bioinformatics.html#cb273-20" aria-hidden="true" tabindex="-1"></a>se <span class="ot">&lt;-</span> <span class="fu">SummarizedExperiment</span>(</span>
<span id="cb273-21"><a href="biology-bioinformatics.html#cb273-21" aria-hidden="true" tabindex="-1"></a>   <span class="at">assays=</span><span class="fu">list</span>(<span class="at">intensities=</span>intensities),</span>
<span id="cb273-22"><a href="biology-bioinformatics.html#cb273-22" aria-hidden="true" tabindex="-1"></a>   <span class="at">colData=</span>colData,</span>
<span id="cb273-23"><a href="biology-bioinformatics.html#cb273-23" aria-hidden="true" tabindex="-1"></a>   <span class="at">rowData=</span>rowData</span>
<span id="cb273-24"><a href="biology-bioinformatics.html#cb273-24" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb273-25"><a href="biology-bioinformatics.html#cb273-25" aria-hidden="true" tabindex="-1"></a>se</span></code></pre></div>
<pre><code>## class: SummarizedExperiment
## dim: 54675 35
## metadata(0):
## assays(1): intensities
## rownames(54675): 1007_s_at 1053_at ... AFFX-TrpnX-5_at AFFX-TrpnX-M_at
## rowData names(1): probeset_id
## colnames(35): GSM972389 GSM972390 ... GSM972512 GSM972521
## colData names(2): sample_name condition</code></pre>
<p>Detailed documentation of how to create and use the SummarizedExperiment is available in the <a href="https://bioconductor.org/packages/release/bioc/vignettes/SummarizedExperiment/inst/doc/SummarizedExperiment.html#constructing-a-summarizedexperiment">SummarizedExperiment vignette</a>.</p>
<div class="box note">
<p>SummarizedExperiment is the successor to the older
<a href="https://www.bioconductor.org/packages/devel/bioc/vignettes/Biobase/inst/doc/ExpressionSetIntroduction.pdf">ExpressionSet</a>
container. Both are still used by Bioconductor packages, but
SummarizedExperiment is more modern and flexible, so it is suggested for use
whenever possible.</p>
</div>
</div>
<div id="microarrays" class="section level3" number="8.5.2">
<h3><span class="header-section-number">8.5.2</span> Microarrays</h3>
<div class="figure">
<img src="microarray_chip.png" alt="" />
<p class="caption">A Microarray Device - Thermo Fisher Scientific</p>
</div>
<p>Microarrays are devices that measure the relative abundance of thousands of
distinct DNA sequences simultaneously. Short (~25 nucleotide) single-stranded
DNA molecules called <em>probes</em> are deposited on a small glass slide in a grid of
spots, where each spot contains many copies of a probe with identical sequence.
The probe sequences are selected <em>a priori</em> based on the purpose of the array.
For example, gene expression microarrays have probes that correspond to the
coding regions of the genome for a species of interest, while genotyping
microarrays use sequences with known variants found in a population of genomes,
most often the human population. Microarrays of the same type (e.g. human gene
expression) all have the same set of probes. The choice of DNA sequence probes
therefore determines what the microarray measures and how to interpret the data.
The design of a microarray is illustrated in the following figure.</p>
<div class="figure">
<img src="microarray_design.png" alt="" />
<p class="caption">Illustration of Microarray Design</p>
</div>
<p>A microarray device generates data by applying a specially prepared DNA sample
to it; the sample usually corresponds to a single biological specimen, e.g. an
individual patient. The preparation method for the sample depends on what is
being measured:</p>
<ul>
<li>When measuring DNA directly, e.g. genetic variants, the DNA itself is
biochemically extracted</li>
<li>When measuring gene expression via RNA abundance, RNA is first extracted and
then reverse transcribed to complementary DNA (cDNA)</li>
</ul>
<p>In either case, the molecules that will be applied to the microarray are DNA
molecules. After extraction and preparation, the DNA molecules are then randomly
cut up into shorter molecules (i.e. <em>sheared</em>) and each molecule has a molecular
tag biochemically ligated to it that will emit fluorescence when excited by a
specific wavelength of light. This tagged DNA sample is then washed over the
microarray chip, where DNA molecules that share sequence complementarity with
the probes on the array pair together. After this treatment, the microarray is
washed to remove DNA molecules that did not have a match on the array, leaving
only those molecules with a sequence match that remain bound to probes.</p>
<p>The microarray chip is then loaded into a scanning device, where a laser with a
specific wavelength of light is then shone onto the array, causing the spots
with tagged DNA molecules associated with probes to fluoresce, and other spots
remain dark. A high resolution image is taken of the fluorescent array, and the
image is analyzed to map the intensity of the light on each spot to a numeric
value proportional to its intensity. The reason for this is that, since each
spot has many individual probe molecules contained within it, the more copies of
the corresponding DNA molecule were in the sample, the more light the spot
emits. In this way, the relative abundance of all probes on the array are
measured simultaneously. The process of generating microarray data from a sample
is illustrated in the following figure.</p>
<div class="figure">
<img src="microarray_data_gen.png" alt="" />
<p class="caption">Illustration of Microarray Data Generation Process</p>
</div>
<p>After the microarray has been scanned, the relative copy number of the DNA in
the sample matching the probes on the microarray are expressed as the intensity
of fluorescence of each probe. The raw intensity data from the scan has been
processed and analyzed by the scanner software to account for technical biases
and artefacts of the scanning instrument and data generation process. The data
from a single scan is processed and stored in a file in <a href="https://www.affymetrix.com/support/developer/powertools/changelog/gcos-agcc/cel.html">CEL
format</a>,
a proprietary data format that stores the raw probe intensity data that can be
loaded for downstream analysis.</p>
<p>On a high level, there are three steps involved in analyzing gene expression
microarray data:</p>
<ol style="list-style-type: decimal">
<li><strong>Summarization of probes to probesets.</strong> Each gene is represented by
multiple probes with different sequences. Summarization is a statistical
procedure that computes a single value for each probeset from its corresponding
probes.</li>
<li><strong>Normalization.</strong> This includes removing background signal from individual
arrays as well as adjusting probeset intensities so that they are comparable
across multiple sample arrays.</li>
<li><strong>Quality control.</strong> Compare all normalized samples within a sample set to
identify, mitigate, or eliminate potential outlier samples.</li>
<li><strong>Analysis.</strong> Using the quality controlled expression data, Implement
statistical analysis to answer research questions.</li>
</ol>
<p>The full details of microarray analysis are beyond the scope of this book.
However in the following sections we cover some of the basic entry points to
performing these steps in R and Bioconductor.</p>
</div>
<div id="microarray-data-in-bioconductor" class="section level3" number="8.5.3">
<h3><span class="header-section-number">8.5.3</span> Microarray Data in Bioconductor</h3>
<p>The CEL data files from a set of microarrays can be loaded into R for analysis
using the <a href="https://www.bioconductor.org/packages/release/bioc/html/affy.html"><code>affy</code> Bioconductor
package</a>.
This package provides all the functions necessary for loading the data and
performing key preprocessing operations. Typically, two or more samples were
processed in this way, resulting in a set of CEL files that should be processed
together. These CEL files will typically be all stored in the same directory,
and may be loaded using the <code>affy::ReadAffy</code> function:</p>
<div class="sourceCode" id="cb275"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb275-1"><a href="biology-bioinformatics.html#cb275-1" aria-hidden="true" tabindex="-1"></a><span class="co"># read all CEL files in a single directory</span></span>
<span id="cb275-2"><a href="biology-bioinformatics.html#cb275-2" aria-hidden="true" tabindex="-1"></a>affy_batch <span class="ot">&lt;-</span> affy<span class="sc">::</span><span class="fu">ReadAffy</span>(<span class="at">celfile.path=</span><span class="st">&quot;directory/of/CELfiles&quot;</span>)</span>
<span id="cb275-3"><a href="biology-bioinformatics.html#cb275-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb275-4"><a href="biology-bioinformatics.html#cb275-4" aria-hidden="true" tabindex="-1"></a><span class="co"># or individual files in different directories</span></span>
<span id="cb275-5"><a href="biology-bioinformatics.html#cb275-5" aria-hidden="true" tabindex="-1"></a>cel_filenames <span class="ot">&lt;-</span> <span class="fu">c</span>(</span>
<span id="cb275-6"><a href="biology-bioinformatics.html#cb275-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">list.files</span>( <span class="at">path=</span><span class="st">&quot;CELfile_dir1&quot;</span>, <span class="at">full.names=</span><span class="cn">TRUE</span>, <span class="at">pattern=</span><span class="st">&quot;.CEL$&quot;</span> ),</span>
<span id="cb275-7"><a href="biology-bioinformatics.html#cb275-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">list.files</span>( <span class="at">path=</span><span class="st">&quot;CELfile_dir2&quot;</span>, <span class="at">full.names=</span><span class="cn">TRUE</span>, <span class="at">pattern=</span><span class="st">&quot;.CEL$&quot;</span> )</span>
<span id="cb275-8"><a href="biology-bioinformatics.html#cb275-8" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb275-9"><a href="biology-bioinformatics.html#cb275-9" aria-hidden="true" tabindex="-1"></a>affy_batch <span class="ot">&lt;-</span> affy<span class="sc">::</span><span class="fu">ReadAffy</span>(<span class="at">filenames=</span>cel_filenames)</span></code></pre></div>
<p>The <code>affy_batch</code> variable is a <code>AffyBatch</code> container, which stores information
on the probe definitions based on the type of microarray, probe-level intensity
for each sample, and other information about the experiment contained within the
CEL files.</p>
<p>The <code>affy</code> package provides functions to perform probe summarization and
normalization. The most popular method to accomplish this at the time of writing
is called Robust Multi-array Average or RMA <span class="citation">(<a href="#ref-Irizarry2003-hn" role="doc-biblioref">Irizarry et al. 2003</a>)</span>, which performs
summarization and normalization of multiple arrays simultaneously. Below, the
RMA algorithm is used to normalize an example dataset provided by the
<a href="https://bioconductor.org/packages/release/data/experiment/html/affydata.html"><code>affydata</code></a>
Bioconductor package.</p>
<div class="box note">
<p>Certain Bioconductor packages provide example datasets for use with companion
analysis packages. For example, the <code>affydata</code> package provides the <code>Dilution</code>
dataset, which was generated using two concentrations of cDNA from human liver
tissue and a central nervous system cell line. To load a data package into R,
first run <code>library(&lt;data package&gt;)</code> and then <code>data(&lt;dataset name&gt;)</code>.</p>
</div>
<div class="sourceCode" id="cb276"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb276-1"><a href="biology-bioinformatics.html#cb276-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(affy)</span>
<span id="cb276-2"><a href="biology-bioinformatics.html#cb276-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(affydata)</span></code></pre></div>
<pre><code>##      Package    LibPath                                     Item      
## [1,] &quot;affydata&quot; &quot;C:/Users/Adam/Documents/R/win-library/4.1&quot; &quot;Dilution&quot;
##      Title                        
## [1,] &quot;AffyBatch instance Dilution&quot;</code></pre>
<div class="sourceCode" id="cb278"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb278-1"><a href="biology-bioinformatics.html#cb278-1" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(Dilution)</span>
<span id="cb278-2"><a href="biology-bioinformatics.html#cb278-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb278-3"><a href="biology-bioinformatics.html#cb278-3" aria-hidden="true" tabindex="-1"></a><span class="co"># normalize the Dilution microarray expression values</span></span>
<span id="cb278-4"><a href="biology-bioinformatics.html#cb278-4" aria-hidden="true" tabindex="-1"></a><span class="co"># note Dilution is an AffyBatch object</span></span>
<span id="cb278-5"><a href="biology-bioinformatics.html#cb278-5" aria-hidden="true" tabindex="-1"></a>eset_rma <span class="ot">&lt;-</span> affy<span class="sc">::</span><span class="fu">rma</span>(Dilution,<span class="at">verbose=</span><span class="cn">FALSE</span>)</span>
<span id="cb278-6"><a href="biology-bioinformatics.html#cb278-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb278-7"><a href="biology-bioinformatics.html#cb278-7" aria-hidden="true" tabindex="-1"></a><span class="co"># plot distribution of non-normalized probes</span></span>
<span id="cb278-8"><a href="biology-bioinformatics.html#cb278-8" aria-hidden="true" tabindex="-1"></a><span class="co"># note rma normalization takes the log2 of the expression values,</span></span>
<span id="cb278-9"><a href="biology-bioinformatics.html#cb278-9" aria-hidden="true" tabindex="-1"></a><span class="co"># so we must do so on the raw data to compare</span></span>
<span id="cb278-10"><a href="biology-bioinformatics.html#cb278-10" aria-hidden="true" tabindex="-1"></a>raw_intensity <span class="ot">&lt;-</span> <span class="fu">as_tibble</span>(<span class="fu">exprs</span>(Dilution)) <span class="sc">%&gt;%</span></span>
<span id="cb278-11"><a href="biology-bioinformatics.html#cb278-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">probeset_id=</span><span class="fu">rownames</span>(<span class="fu">exprs</span>(Dilution))) <span class="sc">%&gt;%</span></span>
<span id="cb278-12"><a href="biology-bioinformatics.html#cb278-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(<span class="sc">-</span>probeset_id, <span class="at">names_to=</span><span class="st">&quot;Sample&quot;</span>, <span class="at">values_to =</span> <span class="st">&quot;Intensity&quot;</span>) <span class="sc">%&gt;%</span></span>
<span id="cb278-13"><a href="biology-bioinformatics.html#cb278-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(</span>
<span id="cb278-14"><a href="biology-bioinformatics.html#cb278-14" aria-hidden="true" tabindex="-1"></a>    <span class="st">`</span><span class="at">log2 Intensity</span><span class="st">`</span><span class="ot">=</span><span class="fu">log2</span>(Intensity),</span>
<span id="cb278-15"><a href="biology-bioinformatics.html#cb278-15" aria-hidden="true" tabindex="-1"></a>    <span class="at">Category=</span><span class="st">&quot;Before Normalization&quot;</span></span>
<span id="cb278-16"><a href="biology-bioinformatics.html#cb278-16" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb278-17"><a href="biology-bioinformatics.html#cb278-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb278-18"><a href="biology-bioinformatics.html#cb278-18" aria-hidden="true" tabindex="-1"></a><span class="co"># plot distribution of normalized probes</span></span>
<span id="cb278-19"><a href="biology-bioinformatics.html#cb278-19" aria-hidden="true" tabindex="-1"></a>rma_intensity <span class="ot">&lt;-</span> <span class="fu">as_tibble</span>(<span class="fu">exprs</span>(eset_rma)) <span class="sc">%&gt;%</span></span>
<span id="cb278-20"><a href="biology-bioinformatics.html#cb278-20" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">probesetid=</span><span class="fu">featureNames</span>(eset_rma)) <span class="sc">%&gt;%</span></span>
<span id="cb278-21"><a href="biology-bioinformatics.html#cb278-21" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(<span class="sc">-</span>probesetid, <span class="at">names_to=</span><span class="st">&quot;Sample&quot;</span>, <span class="at">values_to =</span> <span class="st">&quot;log2 Intensity&quot;</span>) <span class="sc">%&gt;%</span></span>
<span id="cb278-22"><a href="biology-bioinformatics.html#cb278-22" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">Category=</span><span class="st">&quot;RMA Normalized&quot;</span>)</span>
<span id="cb278-23"><a href="biology-bioinformatics.html#cb278-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb278-24"><a href="biology-bioinformatics.html#cb278-24" aria-hidden="true" tabindex="-1"></a>dplyr<span class="sc">::</span><span class="fu">bind_rows</span>(raw_intensity, rma_intensity) <span class="sc">%&gt;%</span></span>
<span id="cb278-25"><a href="biology-bioinformatics.html#cb278-25" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x=</span>Sample, <span class="at">y=</span><span class="st">`</span><span class="at">log2 Intensity</span><span class="st">`</span>)) <span class="sc">+</span></span>
<span id="cb278-26"><a href="biology-bioinformatics.html#cb278-26" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>() <span class="sc">+</span></span>
<span id="cb278-27"><a href="biology-bioinformatics.html#cb278-27" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(<span class="fu">vars</span>(Category))</span></code></pre></div>
<p><img src="main_files/microarray%20norm-1.png" width="672" /></p>
<p>Above, we first apply RMA to the <code>Dilution</code> microarray dataset using the
<code>affy::rma()</code> function. Then we plot the log 2 intensities of the probes in each
array, first using the raw intensities and then afysis seeking to
identify how associated gene expression is with one or more variables of
interest. For example,</p>
</div>
<div id="differential-expression-analysis" class="section level3" number="8.5.4">
<h3><span class="header-section-number">8.5.4</span> Differential Expression Analysis</h3>
<p>Differential expression analysis seeks to identify to what extent gene
expression is associated with one or more variables of interest. For example,
we might be interested in genes that have higher expression in subjects with a
disease, or which genes change in response to a treatment.</p>
<p>Typically, gene expression analysis requires two types of data to run: an
expression matrix and a <em>design matrix</em>. The expression matrix will generally
have features (e.g. genes) as rows and samples as columns. The design matrix is
a numeric matrix that contains the variables we wish to model and any covariates
or confounders we wish to adjust for. The variables in the design matrix then
must be encoded in a way that statistical procedures can understand. The full
details of how design matrices are constructed is beyond the scope of this book.
Fortunately, R makes it very easy to construct these matrices from a tibble with
the <a href="https://www.rdocumentation.org/packages/stats/versions/3.6.2/topics/model.matrix"><code>model.matrix()</code>
function</a>.
Consider for the following imaginary sample information tibble that has 10 AD
and 10 Control subjects with different clinical and protein histology
measurements assayed on their brain tissue:</p>
<div class="sourceCode" id="cb279"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb279-1"><a href="biology-bioinformatics.html#cb279-1" aria-hidden="true" tabindex="-1"></a>ad_metadata</span></code></pre></div>
<pre><code>## # A tibble: 20 x 8
##    ID    age_at_death condition    tau  abeta   iba1   gfap braak_stage
##    &lt;chr&gt;        &lt;dbl&gt; &lt;fct&gt;      &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt; &lt;fct&gt;      
##  1 A1              81 AD         98543 166078  77299  75277 4          
##  2 A2              80 AD         32711  56919  49571 117441 1          
##  3 A3              91 AD        134427 295407 106724  67719 6          
##  4 A4              90 AD         81515 221801 171212  56346 4          
##  5 A5              80 AD         42442  97918   8321 109406 2          
##  6 A6              81 AD         54802  49358  82374  42797 2          
##  7 A7              77 AD        127166   8417  62675  85377 6          
##  8 A8              74 AD        103015  28478 131168 133892 5          
##  9 A9              82 AD        114270 188208 194818  61935 5          
## 10 A10             83 AD         90785  46279 121555 131853 4          
## 11 C1              79 Control      991    237   1062  64752 0          
## 12 C2              76 Control     2804   4423   3441  23384 0          
## 13 C3              73 Control     3120   7073   1402 135049 0          
## 14 C4              73 Control     7167    798   8365  18961 0          
## 15 C5              69 Control     6591   2590   4885  65094 0          
## 16 C6              77 Control    38697 103664  33430  23136 2          
## 17 C7              73 Control    58381  72071  16864  17295 3          
## 18 C8              75 Control     5102  10533   5321  89479 0          
## 19 C9              76 Control    59906  47967  19103 124651 3          
## 20 C10             81 Control    46553  49670  90763  52684 2</code></pre>
<p>We might be interested in identifying genes that are increased or decreased in
people with AD compared with Controls. We can create a model matrix for this as
follows:</p>
<div class="sourceCode" id="cb281"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb281-1"><a href="biology-bioinformatics.html#cb281-1" aria-hidden="true" tabindex="-1"></a><span class="fu">model.matrix</span>(<span class="sc">~</span> condition, <span class="at">data=</span>ad_metadata)</span></code></pre></div>
<pre><code>##    (Intercept) conditionAD
## 1            1           1
## 2            1           1
## 3            1           1
## 4            1           1
## 5            1           1
## 6            1           1
## 7            1           1
## 8            1           1
## 9            1           1
## 10           1           1
## 11           1           0
## 12           1           0
## 13           1           0
## 14           1           0
## 15           1           0
## 16           1           0
## 17           1           0
## 18           1           0
## 19           1           0
## 20           1           0
## attr(,&quot;assign&quot;)
## [1] 0 1
## attr(,&quot;contrasts&quot;)
## attr(,&quot;contrasts&quot;)$condition
## [1] &quot;contr.treatment&quot;</code></pre>
<p>The <code>~ condition</code> argument is an an R
<a href="https://rviews.rstudio.com/2017/02/01/the-r-formula-method-the-good-parts/">forumla</a>,
which is a concise description of how variable should be included in the model.
The general format of a formula is as follows:</p>
<pre><code># portions in [] are optional
[&lt;outcome variable&gt;] ~ &lt;predictor variable 1&gt; [+ &lt;predictor variable 2&gt;]...

# examples

# model Gene 3 expression as a function of disease status
`Gene 3` ~ condition

# model the amount of tau pathology as a function of abeta pathology,
# adjusting for age at death
tau ~ age_at_death + abeta

# create a model without an outcome variable that can be used to create the
# model matrix to test for differences with disease status adjusting for age at death
~ age_at_death + condition</code></pre>
<p>For most models, the design matrix will have an intercept term of all ones and
additional colums for the other variables in the model. We might wish to adjust
out the effect of age by including <code>age_at_death</code> as a covariate in the model:</p>
<div class="sourceCode" id="cb284"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb284-1"><a href="biology-bioinformatics.html#cb284-1" aria-hidden="true" tabindex="-1"></a><span class="fu">model.matrix</span>(<span class="sc">~</span> age_at_death <span class="sc">+</span> condition, <span class="at">data=</span>ad_metadata)</span></code></pre></div>
<pre><code>##    (Intercept) age_at_death conditionAD
## 1            1           81           1
## 2            1           80           1
## 3            1           91           1
## 4            1           90           1
## 5            1           80           1
## 6            1           81           1
## 7            1           77           1
## 8            1           74           1
## 9            1           82           1
## 10           1           83           1
## 11           1           79           0
## 12           1           76           0
## 13           1           73           0
## 14           1           73           0
## 15           1           69           0
## 16           1           77           0
## 17           1           73           0
## 18           1           75           0
## 19           1           76           0
## 20           1           81           0
## attr(,&quot;assign&quot;)
## [1] 0 1 2
## attr(,&quot;contrasts&quot;)
## attr(,&quot;contrasts&quot;)$condition
## [1] &quot;contr.treatment&quot;</code></pre>
<p>The model matrix now includes a column for age at death as well. This model
model matrix is now suitable to pass to differential expression software
packages to look for genes associated with disease status.</p>
<p>As mentioned above, modern gene expression assays measure thousands of genes
simultaneously. This means each gene must be tested for differential expression
individually. In general, each gene is tested using the same statistical model,
so a differential expression analysis package will perform something equivalent
to the following:</p>
<pre><code>Gene 1 ~ age_at_death + condition
Gene 2 ~ age_at_death + condition
Gene 3 ~ age_at_death + condition
...
Gene N ~ age_at_death + condition</code></pre>
<p>Each gene model will have its own statistics associated with it that we must
process and interpret after the analysis is complete.</p>
</div>
<div id="differential-expression-microarrays-limma" class="section level3" number="8.5.5">
<h3><span class="header-section-number">8.5.5</span> Differential Expression: Microarrays (limma)</h3>
<p><a href="https://bioconductor.org/packages/release/bioc/html/limma.html">limma</a>, which
is short for <code>li</code>near <code>m</code>odels of <code>mi</code>croarrays, is one of the <a href="https://bioconductor.org/packages/stats/">top most
downloaded</a> Bioconductor packages.
limma is utilized for analyzing microarray gene expression data, with a focus on
analyses using linear models to integrate <em>all</em> of the data from an experiment.
limma was developed for microarray analysis prior to the development of
sequencing based gene expression methods (i.e. [RNASeq]) but has since added
functionality to analyze other types of gene expression data.</p>
<p>limma excels at analyzing these types of data as it can support arbitrarily
complex experimental designs while maintaining strong statistical power. An
experiment with a large number of conditions or predictors can still be analyze
even with small sample sizes. The method accomplishes this by using an
<a href="https://en.wikipedia.org/wiki/Empirical_Bayes_method">empirical Bayes</a> approach
that borrows information across all the genes in the dataset to better control
error in individual gene models. See <span class="citation"><a href="#ref-Ritchie2015-bj" role="doc-biblioref">Ritchie et al.</a> (<a href="#ref-Ritchie2015-bj" role="doc-biblioref">2015</a>)</span> for more details on how
limma works.</p>
<p>limma requres an expression matrix like that stored in an ExpressionSet or
SummarizedExperiment container and a design matrix:</p>
<div class="sourceCode" id="cb287"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb287-1"><a href="biology-bioinformatics.html#cb287-1" aria-hidden="true" tabindex="-1"></a><span class="co"># intensities is an example gene expression matrix that corresponds to our</span></span>
<span id="cb287-2"><a href="biology-bioinformatics.html#cb287-2" aria-hidden="true" tabindex="-1"></a><span class="co"># AD sample metadata of 10 AD and 10 Control individuals</span></span>
<span id="cb287-3"><a href="biology-bioinformatics.html#cb287-3" aria-hidden="true" tabindex="-1"></a>ad_se  <span class="ot">&lt;-</span> <span class="fu">SummarizedExperiment</span>(</span>
<span id="cb287-4"><a href="biology-bioinformatics.html#cb287-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">assays=</span><span class="fu">list</span>(<span class="at">intensities=</span>intensities),</span>
<span id="cb287-5"><a href="biology-bioinformatics.html#cb287-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">colData=</span>ad_metadata,</span>
<span id="cb287-6"><a href="biology-bioinformatics.html#cb287-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">rowData=</span><span class="fu">rownames</span>(intensities)</span>
<span id="cb287-7"><a href="biology-bioinformatics.html#cb287-7" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb287-8"><a href="biology-bioinformatics.html#cb287-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb287-9"><a href="biology-bioinformatics.html#cb287-9" aria-hidden="true" tabindex="-1"></a><span class="co"># define our design matrix for AD vs control, adjusting for age at death</span></span>
<span id="cb287-10"><a href="biology-bioinformatics.html#cb287-10" aria-hidden="true" tabindex="-1"></a>ad_vs_control_model <span class="ot">&lt;-</span> <span class="fu">model.matrix</span>(<span class="sc">~</span> age_at_death <span class="sc">+</span> condition, <span class="at">data=</span>ad_metadata)</span>
<span id="cb287-11"><a href="biology-bioinformatics.html#cb287-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb287-12"><a href="biology-bioinformatics.html#cb287-12" aria-hidden="true" tabindex="-1"></a><span class="co"># now run limma</span></span>
<span id="cb287-13"><a href="biology-bioinformatics.html#cb287-13" aria-hidden="true" tabindex="-1"></a><span class="co"># first fit all genes with the model</span></span>
<span id="cb287-14"><a href="biology-bioinformatics.html#cb287-14" aria-hidden="true" tabindex="-1"></a>fit <span class="ot">&lt;-</span> limma<span class="sc">::</span><span class="fu">lmFit</span>(</span>
<span id="cb287-15"><a href="biology-bioinformatics.html#cb287-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">assays</span>(se)<span class="sc">$</span>intensities,</span>
<span id="cb287-16"><a href="biology-bioinformatics.html#cb287-16" aria-hidden="true" tabindex="-1"></a>  ad_vs_control_model</span>
<span id="cb287-17"><a href="biology-bioinformatics.html#cb287-17" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb287-18"><a href="biology-bioinformatics.html#cb287-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb287-19"><a href="biology-bioinformatics.html#cb287-19" aria-hidden="true" tabindex="-1"></a><span class="co"># now better control fit errors with the empirical Bayes method</span></span>
<span id="cb287-20"><a href="biology-bioinformatics.html#cb287-20" aria-hidden="true" tabindex="-1"></a>fit <span class="ot">&lt;-</span> limma<span class="sc">::</span><span class="fu">eBayes</span>(fit)</span></code></pre></div>
<p>We have now conducted a differential expression analysis with limma. We can
extract out the results for our question of interest - which genes are
associated with AD - using the <code>limma::topTable()</code> function:</p>
<div class="sourceCode" id="cb288"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb288-1"><a href="biology-bioinformatics.html#cb288-1" aria-hidden="true" tabindex="-1"></a><span class="co"># the coefficient name conditionAD is the column name in the design matrix</span></span>
<span id="cb288-2"><a href="biology-bioinformatics.html#cb288-2" aria-hidden="true" tabindex="-1"></a><span class="co"># adjust=&quot;BH&quot; means perform Benjamini-Hochberg multiple testing adjustment</span></span>
<span id="cb288-3"><a href="biology-bioinformatics.html#cb288-3" aria-hidden="true" tabindex="-1"></a><span class="co"># on the nominal p-values from each gene test</span></span>
<span id="cb288-4"><a href="biology-bioinformatics.html#cb288-4" aria-hidden="true" tabindex="-1"></a><span class="co"># topTable only returns the top 10 results sorted by ascending p-value by default</span></span>
<span id="cb288-5"><a href="biology-bioinformatics.html#cb288-5" aria-hidden="true" tabindex="-1"></a><span class="fu">topTable</span>(fit, <span class="at">coef=</span><span class="st">&quot;conditionAD&quot;</span>, <span class="at">adjust=</span><span class="st">&quot;BH&quot;</span>)</span></code></pre></div>
<pre><code>##                   logFC  AveExpr         t      P.Value adj.P.Val         B
## 234942_s_at   1.1231942 8.397425  4.890779 9.613196e-05  0.770605 -2.987573
## 202423_at    -0.7499088 8.646687 -4.785281 1.221646e-04  0.770605 -3.025525
## 1557538_at   -0.8057310 4.348859 -4.697520 1.492072e-04  0.770605 -3.057737
## 1558290_a_at -0.8580750 7.278878 -4.658725 1.630238e-04  0.770605 -3.072162
## 234596_at     0.4925167 3.038448  4.564888 2.020432e-04  0.770605 -3.107523
## 225986_x_at  -0.6030385 6.958092 -4.502976 2.328354e-04  0.770605 -3.131216
## 203124_s_at  -1.2857166 8.565022 -4.488589 2.406451e-04  0.770605 -3.136763
## 209346_s_at   0.6578475 6.010219  4.421869 2.804637e-04  0.770605 -3.162688
## 224825_at     1.0266849 8.147104  4.421597 2.806391e-04  0.770605 -3.162795
## 205647_at    -0.8569475 4.399029 -4.399700 2.951151e-04  0.770605 -3.171376</code></pre>
<p>From the table, none of the probesets in this experiment are significant at FDR
&lt; 0.05.</p>
<div class="box readmore">
<ul>
<li><a href="https://bioinf.wehi.edu.au/limma/">Homepage</a></li>
<li><a href="https://bioconductor.org/packages/release/bioc/html/limma.html">BioConductor</a></li>
<li><a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4402510/">Publication</a><br />
</li>
</ul>
</div>
</div>
<div id="high-throughput-sequencing-." class="section level3" number="8.5.6">
<h3><span class="header-section-number">8.5.6</span> High Throughput Sequencing .</h3>
<p>What is high throughput sequencing (HTS)? What does HTS data look like? What
does HTS measure, and what kinds of questions can we answer with it? How is HTS
data analyzed? (How) do we analyze it in R?</p>
</div>
<div id="count-data-." class="section level3" number="8.5.7">
<h3><span class="header-section-number">8.5.7</span> Count Data .</h3>
<p>What is count data? Where does it come from? What are the statistical properties
of counts and how do those properties influence our choice of analysis methods (
i.e. Poisson/negative binomial distribution, regularized log transformation to
normality, etc)?</p>
</div>
<div id="rnaseq-." class="section level3" number="8.5.8">
<h3><span class="header-section-number">8.5.8</span> RNASeq .</h3>
<p>What is RNASeq? What does RNASeq measure (steady state RNA copy number)?</p>
</div>
<div id="differential-expression-rnaseq-." class="section level3" number="8.5.9">
<h3><span class="header-section-number">8.5.9</span> Differential Expression: RNASeq .</h3>
<p>How do we analyze RNASeq data (i.e. differential expression) in R? What are
DESeq2/EdgeR, why are they different than other differential expression methods?
How do we interpret DE results?</p>
<div id="deseq2edger-." class="section level4" number="8.5.9.1">
<h4><span class="header-section-number">8.5.9.1</span> DESeq2/EdgeR .</h4>
</div>
<div id="limmavoom-." class="section level4" number="8.5.9.2">
<h4><span class="header-section-number">8.5.9.2</span> limma/voom .</h4>
<p>his brief example is from the limma User Guide chapter 15, and covers loading
and processing data from an RNA-seq experiment. We go into more depth while
working with limma in assignment 6.</p>
<p>Without going into too much detail, the <code>design</code> variable is how we inform limma
of our experimental conditions, and where limma draws from to construct its
linear models correctly. This design is relatively simple, just four samples
belonging to two different conditions (the <code>swirls</code> here refer to the swirl of
zebra fish, you can just see them as a phenotypic difference)</p>
<div class="sourceCode" id="cb290"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb290-1"><a href="biology-bioinformatics.html#cb290-1" aria-hidden="true" tabindex="-1"></a>design <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">swirl =</span> <span class="fu">c</span>(<span class="st">&quot;swirl.1&quot;</span>, <span class="st">&quot;swirl.2&quot;</span>, <span class="st">&quot;swirl.3&quot;</span>, <span class="st">&quot;swirl.4&quot;</span>),</span>
<span id="cb290-2"><a href="biology-bioinformatics.html#cb290-2" aria-hidden="true" tabindex="-1"></a>                     <span class="at">condition =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="sc">-</span><span class="dv">1</span>, <span class="dv">1</span>, <span class="sc">-</span><span class="dv">1</span>))</span>
<span id="cb290-3"><a href="biology-bioinformatics.html#cb290-3" aria-hidden="true" tabindex="-1"></a>dge <span class="ot">&lt;-</span> <span class="fu">DGEList</span>(<span class="at">counts=</span>counts)</span>
<span id="cb290-4"><a href="biology-bioinformatics.html#cb290-4" aria-hidden="true" tabindex="-1"></a>keep <span class="ot">&lt;-</span> <span class="fu">filterByExpr</span>(dge, design)</span>
<span id="cb290-5"><a href="biology-bioinformatics.html#cb290-5" aria-hidden="true" tabindex="-1"></a>dge <span class="ot">&lt;-</span> dge[keep,,keep.lib.sizes<span class="ot">=</span><span class="cn">FALSE</span>]</span></code></pre></div>
<p><code>DGEList()</code> is a function from edgeR, of which limma borrows some loading and
filtering functions. This experiment filters by expression level, and uses
square bracket notation (<code>[]</code>) to reduce the number of rows.</p>
<p>Finally, the expression data is transformed into CPM, counts per million, and
a linear model is applied to the data with <code>lmFit()</code>. <code>topTable()</code> is used to view the most
differentially expressed data.</p>
<div class="sourceCode" id="cb291"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb291-1"><a href="biology-bioinformatics.html#cb291-1" aria-hidden="true" tabindex="-1"></a><span class="co"># limma trend</span></span>
<span id="cb291-2"><a href="biology-bioinformatics.html#cb291-2" aria-hidden="true" tabindex="-1"></a>logCPM <span class="ot">&lt;-</span> <span class="fu">cpm</span>(dge, <span class="at">log=</span><span class="cn">TRUE</span>, <span class="at">prior.count=</span><span class="dv">3</span>)</span>
<span id="cb291-3"><a href="biology-bioinformatics.html#cb291-3" aria-hidden="true" tabindex="-1"></a>fit <span class="ot">&lt;-</span> <span class="fu">lmFit</span>(logCPM, design)</span>
<span id="cb291-4"><a href="biology-bioinformatics.html#cb291-4" aria-hidden="true" tabindex="-1"></a>fit <span class="ot">&lt;-</span> <span class="fu">eBayes</span>(fit, <span class="at">trend=</span><span class="cn">TRUE</span>)</span>
<span id="cb291-5"><a href="biology-bioinformatics.html#cb291-5" aria-hidden="true" tabindex="-1"></a><span class="fu">topTable</span>(fit, <span class="at">coef=</span><span class="fu">ncol</span>(design))</span></code></pre></div>

</div>
</div>
</div>
<div id="gene-set-enrichment-analysis" class="section level2" number="8.6">
<h2><span class="header-section-number">8.6</span> Gene Set Enrichment Analysis</h2>
<p>With the constant evolution of high-throughput sequencing (HTS) technologies,
the size and dimensionality of data generated has been ever increasing. The
question of interest has shifted from how do we generate the data to how do we
make meaningful biological interpretations on a genome-wide level. The simplest
and most common output of HTS experiments is a list of “interesting” genes. In
the specific case of differential gene expression analysis, it is possible and
quite common to obtain hundreds or even thousands of differentially expressed
genes in a single experiment that may be directly or indirectly related to the
phenotype of interest.</p>
<p>While it can be helpful and fruitful to research these genes individually, this
form of personal inspection is limited by one’s domain knowledge and by the size
of the results. On a biological level, it is complicated by the fact that most
processes are often coordinated by the actions of many genes/gene products
working in concert. Cellular signaling pathways, phenotypic differences or
responses to various stimuli are typically associated with changes in the
expression pattern of many genes that share common biological functions or
regulation.</p>
<p>Gene set enrichment analysis is an umbrella term for methods designed to analyze
expression data and capture changes in higher-level biological processes and
pathways by organizing genes into biologically relevant groups or gene sets. We
will discuss the background of two common gene set enrichment analyses
(over-representation analysis and rank-based Gene Set Enrichment Analysis),
their advantages and disadvantages, and walk-through an example of how they can
each be implemented in R. Before this, we will briefly touch upon the definition
of a gene set and describe how to construct or obtain gene sets.</p>
<div id="gene-sets" class="section level3" number="8.6.1">
<h3><span class="header-section-number">8.6.1</span> Gene Sets</h3>
<p>Gene sets are curated lists of genes that are grouped together by some logical
association or pre-existing domain knowledge. Their primary use is to facilitate
the biological interpretation of expression data by capturing high-level
interactions between biologically relevant groups of genes. These sets are
highly flexible and may be constructed based on any <em>a priori</em> knowledge or
classification. For example, one could define a gene set that includes all genes
that are members of the same biochemical pathway or a gene set that consists of
all genes that are induced upon treatment with a particular pharmacological
agent. While it is perfectly valid to construct new gene sets, there are many
existing and curated collections of gene sets that are maintained and
contributed to by communities of scientists throughout the world. Below, we will
highlight some of the major collections that are commonly used and cited in
published work:</p>
<p><strong><a href="https://www.genome.jp/kegg/pathway.html">KEGG Pathways</a></strong></p>
<p>The Kyoto Encyclopedia of Genes and Genomes (KEGG) is a repository for
biological pathway information that the authors describe as a means to
“computerize functional interpretations as part of the pathway reconstruction
process based on the hierarchically structured knowledge about the genomic,
chemical, and network spaces.” The KEGG Pathways database consists of maps
displaying the functional and regulatory relationships of genes within various
metabolic and cell signaling pathways. The KEGG offers a web service that allows
for the extraction of the genes and other information belonging to specific
pathways.</p>
<p><strong><a href="http://geneontology.org/docs/go-annotations/">GO Annotations</a></strong></p>
<p>The Gene Ontology provides a network representation of biological systems from
the molecular-level to the organismal level by defining a graph based
representation of connected terms using a controlled vocabulary. At a high
level, GO annotations consist of three broad ontologies termed molecular
function, cellular component, and biological process. The molecular function is
the specific activity a gene product performs on a molecular level (i.e. a
protein kinase would be annotated with <em>protein kinase activity</em>). The
biological process defines the higher-level programs and pathways that’s
accomplished by the activities of the gene product (i.e. if we take our previous
example of a protein kinase, it’s biological process might be assigned as
<em>signal transduction</em>). The cellular component refers to the localization of the
gene product within the cell (i.e. the cellular component of a protein kinase
might be <em>cytosol</em>). A gene product can be annotated with zero or more terms
from each ontology, and these annotations are based on multiple levels of
evidence from published work. The GO annotations may be accessed or downloaded
directly from their own website, and there exist a number of web services that
use GO annotations in the background (<a href="https://david.ncifcrf.gov">DAVID</a>,
<a href="https://maayanlab.cloud/Enrichr/">enrichR</a>, etc.)</p>
<p><strong><a href="http://www.gsea-msigdb.org/gsea/msigdb/">Molecular Signatures Database</a></strong></p>
<p>The Molecular Signatures Database (MSigDB) is a collection of gene sets curated,
maintained and provided by the Broad Institute and UC San Diego. Although
intended and designed for specific use with the Gene Set Enrichment Analysis
Methodology, they are freely available and only require proper attribution for
other uses. The MSigDB consists of 9 major collections of gene sets: H (hallmark
gene sets), C1 (positional gene sets), C2 (curated gene sets), c3 (regulatory
target gene sets), c4 (computational gene sets), c5 (ontology gene sets), c6
(oncogenic signature gene sets), c7 (immunologic signature gene sets), c8 (cell
type signature gene sets). They are available in formats ready for use in the
GSEA methodology and other formats that are easily imported into various
settings for custom use. The gene sets as well as the GSEA methodology are
available directly from their website. There are a number of R-specific packages
that have been developed for working directly with these gene sets such as
<a href="https://bioconductor.org/packages/release/bioc/html/GSEABase.html">GSEABase</a> or
<a href="http://bioconductor.org/packages/release/bioc/html/fgsea.html">fgsea</a>, which we
will discuss later.</p>
</div>
<div id="working-with-gene-sets-in-r" class="section level3" number="8.6.2">
<h3><span class="header-section-number">8.6.2</span> Working with gene sets in R</h3>
<p>We will walk through two quick examples of how to read in an example collection
of gene sets. We will utilize the Hallmarks gene set collection, downloaded
directly from the MSigDB website, which consists of 50 gene sets representing
well-defined biological processes and generated by aggregating together many
pre-existing gene sets. The MSigDB provides these gene set collections in the
GMT format. These files are tab-delimited and each row in the format represents
one gene set with the first column being the name of the gene set, and the
second column a short description. The remaining columns each represent a gene
and unequal lengths of columns per row is allowed. As is, the GMT format is
designed to work specifically with the GSEA methodology developed and provided
by the Broad Institute. However, we will also show two ways to manually parse
these gene collections for exploration and further use in R.</p>
<p>The first way to parse these gene sets would be to use various tidyverse
functions that you should be familiar with already to construct a tibble.
Essentially, we read the file, rename the first two columns for convenience, and
use a combination of <code>pivot_longer()</code> and <code>group_by()</code> to quickly access genes
by pathway. Below, we have the results of these operations and used them to
display a summary of the number of genes in all the pathways contained with the
hallmark pathways gene collection.</p>
<div class="sourceCode" id="cb292"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb292-1"><a href="biology-bioinformatics.html#cb292-1" aria-hidden="true" tabindex="-1"></a><span class="fu">read_delim</span>(<span class="st">&#39;h.all.v7.5.1.symbols.gmt&#39;</span>, <span class="at">delim=</span><span class="st">&#39;</span><span class="sc">\t</span><span class="st">&#39;</span>, <span class="at">col_names=</span><span class="cn">FALSE</span>) <span class="sc">%&gt;%</span></span>
<span id="cb292-2"><a href="biology-bioinformatics.html#cb292-2" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">rename</span>(<span class="at">Pathway =</span> X1, <span class="at">Desc=</span>X2) <span class="sc">%&gt;%</span></span>
<span id="cb292-3"><a href="biology-bioinformatics.html#cb292-3" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">select</span>(<span class="sc">-</span>Desc) <span class="sc">%&gt;%</span></span>
<span id="cb292-4"><a href="biology-bioinformatics.html#cb292-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(<span class="sc">!</span>Pathway, <span class="at">values_to=</span><span class="st">&#39;genes&#39;</span>, <span class="at">values_drop_na=</span><span class="cn">TRUE</span>, <span class="at">names_to =</span> <span class="cn">NULL</span>) <span class="sc">%&gt;%</span></span>
<span id="cb292-5"><a href="biology-bioinformatics.html#cb292-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(Pathway) <span class="sc">%&gt;%</span></span>
<span id="cb292-6"><a href="biology-bioinformatics.html#cb292-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="at">n=</span><span class="fu">n</span>())</span></code></pre></div>
<pre><code>## # A tibble: 50 x 2
##    Pathway                              n
##    &lt;chr&gt;                            &lt;int&gt;
##  1 HALLMARK_ADIPOGENESIS              200
##  2 HALLMARK_ALLOGRAFT_REJECTION       200
##  3 HALLMARK_ANDROGEN_RESPONSE         100
##  4 HALLMARK_ANGIOGENESIS               36
##  5 HALLMARK_APICAL_JUNCTION           200
##  6 HALLMARK_APICAL_SURFACE             44
##  7 HALLMARK_APOPTOSIS                 161
##  8 HALLMARK_BILE_ACID_METABOLISM      112
##  9 HALLMARK_CHOLESTEROL_HOMEOSTASIS    74
## 10 HALLMARK_COAGULATION               138
## # ... with 40 more rows</code></pre>
<p>We can see that there are 50 total pathways in the hallmarks gene collection
that contain varying numbers of genes. We discarded the description column and
if we were to save the results after using <code>pivot_longer()</code> instead of piping
them to <code>group_by</code> and <code>summarise()</code>, we would have a tibble in the “long”
format with each row representing a single pathway and single gene.</p>
<p>However, there do exist various packages that have been developed to
specifically handle gene set data in R such as the previously mentioned
GSEABase. This is a collection of functions and class-based objects that
facilitate working with gene sets. The foundation of the GSEABase package is the
GeneSet and GeneSetCollection classes which store gene sets and metadata or a
collection of GeneSet objects, respectively. We will use the GSEABase package to
read in the collection of gene sets we previously downloaded.</p>
<div class="sourceCode" id="cb294"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb294-1"><a href="biology-bioinformatics.html#cb294-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">&#39;GSEABase&#39;</span>)</span>
<span id="cb294-2"><a href="biology-bioinformatics.html#cb294-2" aria-hidden="true" tabindex="-1"></a>hallmarks_gmt <span class="ot">&lt;-</span> <span class="fu">getGmt</span>(<span class="at">con=</span><span class="st">&#39;h.all.v7.5.1.symbols.gmt&#39;</span>)</span>
<span id="cb294-3"><a href="biology-bioinformatics.html#cb294-3" aria-hidden="true" tabindex="-1"></a>hallmarks_gmt</span></code></pre></div>
<pre><code>## GeneSetCollection
##   names: HALLMARK_TNFA_SIGNALING_VIA_NFKB, HALLMARK_HYPOXIA, ..., HALLMARK_PANCREAS_BETA_CELLS (50 total)
##   unique identifiers: JUNB, CXCL2, ..., SRP14 (4383 total)
##   types in collection:
##     geneIdType: NullIdentifier (1 total)
##     collectionType: NullCollection (1 total)</code></pre>
<p>If we simply access the hallmarks_gmt variable, we can see that it is a
GeneSetCollection object containing 0 total gene
sets which encompass 4383 unique identifiers or HGNC symbols. Although this
package supports a range of functions, we will focus on the basics. For a more
thorough description of the classes and methods, please read their extended
documentation available
<a href="https://bioconductor.org/packages/release/bioc/manuals/GSEABase/man/GSEABase.pdf">here</a>.</p>
<p>The geneIds method will return a list with each pathway as a named vector of associated gene ids:</p>
<div class="sourceCode" id="cb296"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb296-1"><a href="biology-bioinformatics.html#cb296-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">geneIds</span>(hallmarks_gmt), <span class="dv">2</span>)</span></code></pre></div>
<pre><code>## $HALLMARK_TNFA_SIGNALING_VIA_NFKB
##   [1] &quot;JUNB&quot;     &quot;CXCL2&quot;    &quot;ATF3&quot;     &quot;NFKBIA&quot;   &quot;TNFAIP3&quot;  &quot;PTGS2&quot;   
##   [7] &quot;CXCL1&quot;    &quot;IER3&quot;     &quot;CD83&quot;     &quot;CCL20&quot;    &quot;CXCL3&quot;    &quot;MAFF&quot;    
##  [13] &quot;NFKB2&quot;    &quot;TNFAIP2&quot;  &quot;HBEGF&quot;    &quot;KLF6&quot;     &quot;BIRC3&quot;    &quot;PLAUR&quot;   
##  [19] &quot;ZFP36&quot;    &quot;ICAM1&quot;    &quot;JUN&quot;      &quot;EGR3&quot;     &quot;IL1B&quot;     &quot;BCL2A1&quot;  
##  [25] &quot;PPP1R15A&quot; &quot;ZC3H12A&quot;  &quot;SOD2&quot;     &quot;NR4A2&quot;    &quot;IL1A&quot;     &quot;RELB&quot;    
##  [31] &quot;TRAF1&quot;    &quot;BTG2&quot;     &quot;DUSP1&quot;    &quot;MAP3K8&quot;   &quot;ETS2&quot;     &quot;F3&quot;      
##  [37] &quot;SDC4&quot;     &quot;EGR1&quot;     &quot;IL6&quot;      &quot;TNF&quot;      &quot;KDM6B&quot;    &quot;NFKB1&quot;   
##  [43] &quot;LIF&quot;      &quot;PTX3&quot;     &quot;FOSL1&quot;    &quot;NR4A1&quot;    &quot;JAG1&quot;     &quot;CCL4&quot;    
##  [49] &quot;GCH1&quot;     &quot;CCL2&quot;     &quot;RCAN1&quot;    &quot;DUSP2&quot;    &quot;EHD1&quot;     &quot;IER2&quot;    
##  [55] &quot;REL&quot;      &quot;CFLAR&quot;    &quot;RIPK2&quot;    &quot;NFKBIE&quot;   &quot;NR4A3&quot;    &quot;PHLDA1&quot;  
##  [61] &quot;IER5&quot;     &quot;TNFSF9&quot;   &quot;GEM&quot;      &quot;GADD45A&quot;  &quot;CXCL10&quot;   &quot;PLK2&quot;    
##  [67] &quot;BHLHE40&quot;  &quot;EGR2&quot;     &quot;SOCS3&quot;    &quot;SLC2A6&quot;   &quot;PTGER4&quot;   &quot;DUSP5&quot;   
##  [73] &quot;SERPINB2&quot; &quot;NFIL3&quot;    &quot;SERPINE1&quot; &quot;TRIB1&quot;    &quot;TIPARP&quot;   &quot;RELA&quot;    
##  [79] &quot;BIRC2&quot;    &quot;CXCL6&quot;    &quot;LITAF&quot;    &quot;TNFAIP6&quot;  &quot;CD44&quot;     &quot;INHBA&quot;   
##  [85] &quot;PLAU&quot;     &quot;MYC&quot;      &quot;TNFRSF9&quot;  &quot;SGK1&quot;     &quot;TNIP1&quot;    &quot;NAMPT&quot;   
##  [91] &quot;FOSL2&quot;    &quot;PNRC1&quot;    &quot;ID2&quot;      &quot;CD69&quot;     &quot;IL7R&quot;     &quot;EFNA1&quot;   
##  [97] &quot;PHLDA2&quot;   &quot;PFKFB3&quot;   &quot;CCL5&quot;     &quot;YRDC&quot;     &quot;IFNGR2&quot;   &quot;SQSTM1&quot;  
## [103] &quot;BTG3&quot;     &quot;GADD45B&quot;  &quot;KYNU&quot;     &quot;G0S2&quot;     &quot;BTG1&quot;     &quot;MCL1&quot;    
## [109] &quot;VEGFA&quot;    &quot;MAP2K3&quot;   &quot;CDKN1A&quot;   &quot;CCN1&quot;     &quot;TANK&quot;     &quot;IFIT2&quot;   
## [115] &quot;IL18&quot;     &quot;TUBB2A&quot;   &quot;IRF1&quot;     &quot;FOS&quot;      &quot;OLR1&quot;     &quot;RHOB&quot;    
## [121] &quot;AREG&quot;     &quot;NINJ1&quot;    &quot;ZBTB10&quot;   &quot;PLPP3&quot;    &quot;KLF4&quot;     &quot;CXCL11&quot;  
## [127] &quot;SAT1&quot;     &quot;CSF1&quot;     &quot;GPR183&quot;   &quot;PMEPA1&quot;   &quot;PTPRE&quot;    &quot;TLR2&quot;    
## [133] &quot;ACKR3&quot;    &quot;KLF10&quot;    &quot;MARCKS&quot;   &quot;LAMB3&quot;    &quot;CEBPB&quot;    &quot;TRIP10&quot;  
## [139] &quot;F2RL1&quot;    &quot;KLF9&quot;     &quot;LDLR&quot;     &quot;TGIF1&quot;    &quot;RNF19B&quot;   &quot;DRAM1&quot;   
## [145] &quot;B4GALT1&quot;  &quot;DNAJB4&quot;   &quot;CSF2&quot;     &quot;PDE4B&quot;    &quot;SNN&quot;      &quot;PLEK&quot;    
## [151] &quot;STAT5A&quot;   &quot;DENND5A&quot;  &quot;CCND1&quot;    &quot;DDX58&quot;    &quot;SPHK1&quot;    &quot;CD80&quot;    
## [157] &quot;TNFAIP8&quot;  &quot;CCNL1&quot;    &quot;FUT4&quot;     &quot;CCRL2&quot;    &quot;SPSB1&quot;    &quot;TSC22D1&quot; 
## [163] &quot;B4GALT5&quot;  &quot;SIK1&quot;     &quot;CLCF1&quot;    &quot;NFE2L2&quot;   &quot;FOSB&quot;     &quot;PER1&quot;    
## [169] &quot;NFAT5&quot;    &quot;ATP2B1&quot;   &quot;IL12B&quot;    &quot;IL6ST&quot;    &quot;SLC16A6&quot;  &quot;ABCA1&quot;   
## [175] &quot;HES1&quot;     &quot;BCL6&quot;     &quot;IRS2&quot;     &quot;SLC2A3&quot;   &quot;CEBPD&quot;    &quot;IL23A&quot;   
## [181] &quot;SMAD3&quot;    &quot;TAP1&quot;     &quot;MSC&quot;      &quot;IFIH1&quot;    &quot;IL15RA&quot;   &quot;TNIP2&quot;   
## [187] &quot;BCL3&quot;     &quot;PANX1&quot;    &quot;FJX1&quot;     &quot;EDN1&quot;     &quot;EIF1&quot;     &quot;BMP2&quot;    
## [193] &quot;DUSP4&quot;    &quot;PDLIM5&quot;   &quot;ICOSLG&quot;   &quot;GFPT2&quot;    &quot;KLF2&quot;     &quot;TNC&quot;     
## [199] &quot;SERPINB8&quot; &quot;MXD1&quot;    
## 
## $HALLMARK_HYPOXIA
##   [1] &quot;PGK1&quot;     &quot;PDK1&quot;     &quot;GBE1&quot;     &quot;PFKL&quot;     &quot;ALDOA&quot;    &quot;ENO2&quot;    
##   [7] &quot;PGM1&quot;     &quot;NDRG1&quot;    &quot;HK2&quot;      &quot;ALDOC&quot;    &quot;GPI&quot;      &quot;MXI1&quot;    
##  [13] &quot;SLC2A1&quot;   &quot;P4HA1&quot;    &quot;ADM&quot;      &quot;P4HA2&quot;    &quot;ENO1&quot;     &quot;PFKP&quot;    
##  [19] &quot;AK4&quot;      &quot;FAM162A&quot;  &quot;PFKFB3&quot;   &quot;VEGFA&quot;    &quot;BNIP3L&quot;   &quot;TPI1&quot;    
##  [25] &quot;ERO1A&quot;    &quot;KDM3A&quot;    &quot;CCNG2&quot;    &quot;LDHA&quot;     &quot;GYS1&quot;     &quot;GAPDH&quot;   
##  [31] &quot;BHLHE40&quot;  &quot;ANGPTL4&quot;  &quot;JUN&quot;      &quot;SERPINE1&quot; &quot;LOX&quot;      &quot;GCK&quot;     
##  [37] &quot;PPFIA4&quot;   &quot;MAFF&quot;     &quot;DDIT4&quot;    &quot;SLC2A3&quot;   &quot;IGFBP3&quot;   &quot;NFIL3&quot;   
##  [43] &quot;FOS&quot;      &quot;RBPJ&quot;     &quot;HK1&quot;      &quot;CITED2&quot;   &quot;ISG20&quot;    &quot;GALK1&quot;   
##  [49] &quot;WSB1&quot;     &quot;PYGM&quot;     &quot;STC1&quot;     &quot;ZNF292&quot;   &quot;BTG1&quot;     &quot;PLIN2&quot;   
##  [55] &quot;CSRP2&quot;    &quot;VLDLR&quot;    &quot;JMJD6&quot;    &quot;EXT1&quot;     &quot;F3&quot;       &quot;PDK3&quot;    
##  [61] &quot;ANKZF1&quot;   &quot;UGP2&quot;     &quot;ALDOB&quot;    &quot;STC2&quot;     &quot;ERRFI1&quot;   &quot;ENO3&quot;    
##  [67] &quot;PNRC1&quot;    &quot;HMOX1&quot;    &quot;PGF&quot;      &quot;GAPDHS&quot;   &quot;CHST2&quot;    &quot;TMEM45A&quot; 
##  [73] &quot;BCAN&quot;     &quot;ATF3&quot;     &quot;CAV1&quot;     &quot;AMPD3&quot;    &quot;GPC3&quot;     &quot;NDST1&quot;   
##  [79] &quot;IRS2&quot;     &quot;SAP30&quot;    &quot;GAA&quot;      &quot;SDC4&quot;     &quot;STBD1&quot;    &quot;IER3&quot;    
##  [85] &quot;PKLR&quot;     &quot;IGFBP1&quot;   &quot;PLAUR&quot;    &quot;CAVIN3&quot;   &quot;CCN5&quot;     &quot;LARGE1&quot;  
##  [91] &quot;NOCT&quot;     &quot;S100A4&quot;   &quot;RRAGD&quot;    &quot;ZFP36&quot;    &quot;EGFR&quot;     &quot;EDN2&quot;    
##  [97] &quot;IDS&quot;      &quot;CDKN1A&quot;   &quot;RORA&quot;     &quot;DUSP1&quot;    &quot;MIF&quot;      &quot;PPP1R3C&quot; 
## [103] &quot;DPYSL4&quot;   &quot;KDELR3&quot;   &quot;DTNA&quot;     &quot;ADORA2B&quot;  &quot;HS3ST1&quot;   &quot;CAVIN1&quot;  
## [109] &quot;NR3C1&quot;    &quot;KLF6&quot;     &quot;GPC4&quot;     &quot;CCN1&quot;     &quot;TNFAIP3&quot;  &quot;CA12&quot;    
## [115] &quot;HEXA&quot;     &quot;BGN&quot;      &quot;PPP1R15A&quot; &quot;PGM2&quot;     &quot;PIM1&quot;     &quot;PRDX5&quot;   
## [121] &quot;NAGK&quot;     &quot;CDKN1B&quot;   &quot;BRS3&quot;     &quot;TKTL1&quot;    &quot;MT1E&quot;     &quot;ATP7A&quot;   
## [127] &quot;MT2A&quot;     &quot;SDC3&quot;     &quot;TIPARP&quot;   &quot;PKP1&quot;     &quot;ANXA2&quot;    &quot;PGAM2&quot;   
## [133] &quot;DDIT3&quot;    &quot;PRKCA&quot;    &quot;SLC37A4&quot;  &quot;CXCR4&quot;    &quot;EFNA3&quot;    &quot;CP&quot;      
## [139] &quot;KLF7&quot;     &quot;CCN2&quot;     &quot;CHST3&quot;    &quot;TPD52&quot;    &quot;LXN&quot;      &quot;B4GALNT2&quot;
## [145] &quot;PPARGC1A&quot; &quot;BCL2&quot;     &quot;GCNT2&quot;    &quot;HAS1&quot;     &quot;KLHL24&quot;   &quot;SCARB1&quot;  
## [151] &quot;SLC25A1&quot;  &quot;SDC2&quot;     &quot;CASP6&quot;    &quot;VHL&quot;      &quot;FOXO3&quot;    &quot;PDGFB&quot;   
## [157] &quot;B3GALT6&quot;  &quot;SLC2A5&quot;   &quot;SRPX&quot;     &quot;EFNA1&quot;    &quot;GLRX&quot;     &quot;ACKR3&quot;   
## [163] &quot;PAM&quot;      &quot;TGFBI&quot;    &quot;DCN&quot;      &quot;SIAH2&quot;    &quot;PLAC8&quot;    &quot;FBP1&quot;    
## [169] &quot;TPST2&quot;    &quot;PHKG1&quot;    &quot;MYH9&quot;     &quot;CDKN1C&quot;   &quot;GRHPR&quot;    &quot;PCK1&quot;    
## [175] &quot;INHA&quot;     &quot;HSPA5&quot;    &quot;NDST2&quot;    &quot;NEDD4L&quot;   &quot;TPBG&quot;     &quot;XPNPEP1&quot; 
## [181] &quot;IL6&quot;      &quot;SLC6A6&quot;   &quot;MAP3K1&quot;   &quot;LDHC&quot;     &quot;AKAP12&quot;   &quot;TES&quot;     
## [187] &quot;KIF5A&quot;    &quot;LALBA&quot;    &quot;COL5A1&quot;   &quot;GPC1&quot;     &quot;HDLBP&quot;    &quot;ILVBL&quot;   
## [193] &quot;NCAN&quot;     &quot;TGM2&quot;     &quot;ETS1&quot;     &quot;HOXB9&quot;    &quot;SELENBP1&quot; &quot;FOSL2&quot;   
## [199] &quot;SULT2B1&quot;  &quot;TGFB3&quot;</code></pre>
<p>The names method will return all of the gene set names contained within a specific collection:</p>
<div class="sourceCode" id="cb298"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb298-1"><a href="biology-bioinformatics.html#cb298-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">names</span>(hallmarks_gmt),<span class="dv">2</span>)</span></code></pre></div>
<pre><code>## [1] &quot;HALLMARK_TNFA_SIGNALING_VIA_NFKB&quot; &quot;HALLMARK_HYPOXIA&quot;</code></pre>
<p>We can access a specific gene set contained within this collection by referring
to its name and using the following notation:</p>
<div class="sourceCode" id="cb300"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb300-1"><a href="biology-bioinformatics.html#cb300-1" aria-hidden="true" tabindex="-1"></a>hallmarks_gmt[[<span class="st">&#39;HALLMARK_ANGIOGENESIS&#39;</span>]]</span></code></pre></div>
<pre><code>## setName: HALLMARK_ANGIOGENESIS 
## geneIds: VCAN, POSTN, ..., CXCL6 (total: 36)
## geneIdType: Null
## collectionType: Null 
## details: use &#39;details(object)&#39;</code></pre>
<p>Simply accessing the object will provide a high-level summary of the information
contained within. To access a specific value of this GeneSet object, we would
call one of the slots (a core concept in object-oriented programming). In our
particular case, we could extract the gene names contained assigned to this
GeneSet by calling the <code>geneIds</code> slot as shown below to return a vector of the
gene names. We can see the first five below and also the length of the returned
vector by using the base R <code>length()</code> function:</p>
<div class="sourceCode" id="cb302"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb302-1"><a href="biology-bioinformatics.html#cb302-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(hallmarks_gmt[[<span class="st">&#39;HALLMARK_ANGIOGENESIS&#39;</span>]]<span class="sc">@</span>geneIds)</span></code></pre></div>
<pre><code>## [1] &quot;VCAN&quot;   &quot;POSTN&quot;  &quot;FSTL1&quot;  &quot;LRPAP1&quot; &quot;STC1&quot;   &quot;LPL&quot;</code></pre>
<div class="sourceCode" id="cb304"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb304-1"><a href="biology-bioinformatics.html#cb304-1" aria-hidden="true" tabindex="-1"></a><span class="fu">length</span>(hallmarks_gmt[[<span class="st">&#39;HALLMARK_ANGIOGENESIS&#39;</span>]]<span class="sc">@</span>geneIds)</span></code></pre></div>
<pre><code>## [1] 36</code></pre>
<p>GSEABase includes a number of built-in functions for reading gene sets in from
various sources and performing common operations such as set intersections, set
differences, and ID conversions. We will demonstrate the usage of some of these
in the next section covering Over-representation analysis.</p>
</div>
<div id="over-representation-analysis" class="section level3" number="8.6.3">
<h3><span class="header-section-number">8.6.3</span> Over-representation Analysis</h3>
<p>One of the most common ways to utilize gene sets to evaluate gene expression
results is to perform Over-representation Analysis (ORA). Let us assume that we
have obtained a list of differentially expressed genes from an experiment. We
are curious to know if within this list of differentially expressed genes, do we
see an <em>over-representation</em> or <em>enrichment</em> of genes belonging to gene sets of
interest? In more general terms, the goal of ORA is to determine how likely it
is that there is a non-random association between a gene being differentially
expressed and having membership in a chosen (and ideally relevant) gene set. In
R, we can do a simple ORA by utilizing a Fisher’s exact test and a contingency
table.</p>
<p>For a purely hypothetical example, let us assume that we have performed
differential gene expression analysis between two different cell lines. We
obtain a list of 10,000 total genes (our background) discovered in the
experiment and find that at our chosen statistical threshold, 1,000 of these are
differentially expressed. To keep things simple, we will perform a single ORA
test against the Hallmarks Angiogenesis gene set using a sample list of 1000
differentially expressed genes selected from data generated by <a href="https://pubmed.ncbi.nlm.nih.gov/23700391/">Marisa et al.
2013</a>. The Hallmarks Angiogenesis
gene set consists of 36 genes and we find that 13 of these are also present in
our list of differentially expressed genes.</p>
<div class="box warning">
<p>Please note that we are removing this list of genes from its original meaning
and context found in the publication and simply using it to demonstrate the
basic steps occurring during ORA. All of these numbers and lists were
arbitrarily chosen and this experimental setup is purely hypothetical.</p>
<p>Another issue to note is that typically the “background” list should represent
the entire pool of genes from which any differentially expressed genes could
have been selected. For expression experiments, it is typical to choose all of
the detected genes (regardless of significance) as the background. The number of
genes in the organism’s genome could potentially also be an appropriate
“background”</p>
<p>Also, it is important to keep in mind that in reality, ORA is nearly always
performed on a larger scale against a variety of different gene sets. This
allows for the unbiased discovery of potentially novel and unexpected enrichment
in other biological areas of interest. It also necessitates the need for
multiple-testing correction, which we have discussed in <a href="#multiple-hypothesis-testing">multiple hypothesis
testing</a>.</p>
</div>
<p>To begin, we would want to prepare a contingency table which describes the
various overlaps between our sets of interest. For a 2x2 contingency table,
these four values will be:</p>
<ol style="list-style-type: decimal">
<li>Genes present in our list of differentially expressed genes and present in our gene set</li>
<li>Genes present in our list of differentially expressed genes and not present in our gene set</li>
<li>Genes not present in our list of differentially expressed genes and present in our gene set</li>
<li>Genes not present in our list of differentially expressed genes and not present in our gene set</li>
</ol>
<p>To demonstrate what this would look like, we have manually constructed a
contingency table with labels and totals added below. If you look at the margins
of the table and recall the previously given values above, you can reconstruct
the logic used to generate each of the values in all the cells.</p>
<table>
<thead>
<tr class="header">
<th align="left"></th>
<th align="right">Differentially Expressed</th>
<th align="right">Not Differentially Expressed</th>
<th align="right">Total</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">In Gene Set</td>
<td align="right">13</td>
<td align="right">23</td>
<td align="right">36</td>
</tr>
<tr class="even">
<td align="left">Not in Gene Set</td>
<td align="right">987</td>
<td align="right">8977</td>
<td align="right">9964</td>
</tr>
<tr class="odd">
<td align="left">Total</td>
<td align="right">1000</td>
<td align="right">9000</td>
<td align="right">10000</td>
</tr>
</tbody>
</table>
<p>For the purposes of this example, we are reading in our differentially expressed
genes from an external file, but this vector could be generated in any number of
ways depending upon where and how your results are stored. Following good coding
practices, we will write a small function that takes this list of DE genes and a
GeneSet object to programmatically generate a contingency table:</p>
<div class="box note">
<p>There are many ways to construct a contingency table. This is just one way that
was chosen to make calculations of the values contained within the table
transparent and easy to understand.</p>
</div>
<div class="sourceCode" id="cb306"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb306-1"><a href="biology-bioinformatics.html#cb306-1" aria-hidden="true" tabindex="-1"></a><span class="co">#load and read our list of DE genes contained within a newline delimited txt file</span></span>
<span id="cb306-2"><a href="biology-bioinformatics.html#cb306-2" aria-hidden="true" tabindex="-1"></a>de_genes <span class="ot">&lt;-</span> <span class="fu">scan</span>(<span class="st">&#39;example_de_list.txt&#39;</span>, <span class="at">what=</span><span class="fu">character</span>(), <span class="at">sep=</span><span class="st">&#39;</span><span class="sc">\n</span><span class="st">&#39;</span>)</span>
<span id="cb306-3"><a href="biology-bioinformatics.html#cb306-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb306-4"><a href="biology-bioinformatics.html#cb306-4" aria-hidden="true" tabindex="-1"></a><span class="co">#define a function that takes a list of DE genes, and a specific GeneSet from a GeneSetCollection to generate a contingency table</span></span>
<span id="cb306-5"><a href="biology-bioinformatics.html#cb306-5" aria-hidden="true" tabindex="-1"></a><span class="co">#using set operations in GSEABase</span></span>
<span id="cb306-6"><a href="biology-bioinformatics.html#cb306-6" aria-hidden="true" tabindex="-1"></a>make_contingency <span class="ot">&lt;-</span> <span class="cf">function</span>(de_list, GeneSetobj) {</span>
<span id="cb306-7"><a href="biology-bioinformatics.html#cb306-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb306-8"><a href="biology-bioinformatics.html#cb306-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb306-9"><a href="biology-bioinformatics.html#cb306-9" aria-hidden="true" tabindex="-1"></a>  <span class="co">#make our de list into a simple GeneSet object using GSEABase</span></span>
<span id="cb306-10"><a href="biology-bioinformatics.html#cb306-10" aria-hidden="true" tabindex="-1"></a>  de_geneset <span class="ot">&lt;-</span> <span class="fu">GeneSet</span>(de_list, <span class="at">setName=</span><span class="st">&#39;1000 DE genes&#39;</span>)</span>
<span id="cb306-11"><a href="biology-bioinformatics.html#cb306-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb306-12"><a href="biology-bioinformatics.html#cb306-12" aria-hidden="true" tabindex="-1"></a>  <span class="co">#If we had the full results, we could determine this value without manually setting it</span></span>
<span id="cb306-13"><a href="biology-bioinformatics.html#cb306-13" aria-hidden="true" tabindex="-1"></a>  background_len <span class="ot">&lt;-</span> <span class="dv">10000</span></span>
<span id="cb306-14"><a href="biology-bioinformatics.html#cb306-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb306-15"><a href="biology-bioinformatics.html#cb306-15" aria-hidden="true" tabindex="-1"></a>  <span class="co">#Calculate the values inside the contingency table using set operations</span></span>
<span id="cb306-16"><a href="biology-bioinformatics.html#cb306-16" aria-hidden="true" tabindex="-1"></a>  de_in <span class="ot">&lt;-</span> <span class="fu">length</span>((de_geneset <span class="sc">&amp;</span> GeneSetobj)<span class="sc">@</span>geneIds)</span>
<span id="cb306-17"><a href="biology-bioinformatics.html#cb306-17" aria-hidden="true" tabindex="-1"></a>  de_notin <span class="ot">&lt;-</span> <span class="fu">length</span>(<span class="fu">setdiff</span>(de_geneset, GeneSetobj)<span class="sc">@</span>geneIds)</span>
<span id="cb306-18"><a href="biology-bioinformatics.html#cb306-18" aria-hidden="true" tabindex="-1"></a>  set_notde <span class="ot">&lt;-</span> <span class="fu">length</span>(<span class="fu">setdiff</span>(GeneSetobj, de_geneset)<span class="sc">@</span>geneIds)</span>
<span id="cb306-19"><a href="biology-bioinformatics.html#cb306-19" aria-hidden="true" tabindex="-1"></a>  notin_notde <span class="ot">&lt;-</span> background_len <span class="sc">-</span> (de_in <span class="sc">+</span> de_notin <span class="sc">+</span> set_notde)</span>
<span id="cb306-20"><a href="biology-bioinformatics.html#cb306-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb306-21"><a href="biology-bioinformatics.html#cb306-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb306-22"><a href="biology-bioinformatics.html#cb306-22" aria-hidden="true" tabindex="-1"></a>  <span class="co">#return a matrix of the contingency values</span></span>
<span id="cb306-23"><a href="biology-bioinformatics.html#cb306-23" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(<span class="fu">matrix</span>(<span class="fu">c</span>(de_in, de_notin, set_notde, notin_notde), <span class="at">nrow=</span><span class="dv">2</span>))</span>
<span id="cb306-24"><a href="biology-bioinformatics.html#cb306-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb306-25"><a href="biology-bioinformatics.html#cb306-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb306-26"><a href="biology-bioinformatics.html#cb306-26" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb306-27"><a href="biology-bioinformatics.html#cb306-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb306-28"><a href="biology-bioinformatics.html#cb306-28" aria-hidden="true" tabindex="-1"></a>contingency_table <span class="ot">&lt;-</span> <span class="fu">make_contingency</span>(de_genes, hallmarks_gmt[[<span class="st">&#39;HALLMARK_ANGIOGENESIS&#39;</span>]])</span>
<span id="cb306-29"><a href="biology-bioinformatics.html#cb306-29" aria-hidden="true" tabindex="-1"></a>contingency_table</span></code></pre></div>
<pre><code>##      [,1] [,2]
## [1,]   13   23
## [2,]  987 8977</code></pre>
<p>We perform the Fisher’s Exact test using the built-in R function <code>fisher.test()</code>
and view the summarized output by simply calling the variable where we stored
the test results:</p>
<div class="sourceCode" id="cb308"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb308-1"><a href="biology-bioinformatics.html#cb308-1" aria-hidden="true" tabindex="-1"></a>fisher_results <span class="ot">&lt;-</span> <span class="fu">fisher.test</span>(contingency_table, <span class="at">alternative=</span><span class="st">&#39;greater&#39;</span>)</span>
<span id="cb308-2"><a href="biology-bioinformatics.html#cb308-2" aria-hidden="true" tabindex="-1"></a>fisher_results</span></code></pre></div>
<pre><code>## 
##  Fisher&#39;s Exact Test for Count Data
## 
## data:  contingency_table
## p-value = 2.382e-05
## alternative hypothesis: true odds ratio is greater than 1
## 95 percent confidence interval:
##  2.685749      Inf
## sample estimates:
## odds ratio 
##   5.139308</code></pre>
<div class="box note">
<p>Specific values of the results can be accessed by the $notation (i.e.
fisher_results$p.value). The full list of returned values may be found in the R
documentation for <code>fisher.test()</code></p>
</div>
<p>Back to our hypothetical example and focusing on the p-value returned of <code>r fisher_results$p.value</code>, we can interpret this as the probability of randomly
obtaining results as or more extreme than what we observed assuming the null
hypothesis that there is no association between differential expression and gene
set membership is true. Based on these results and if this p-value was below our
pre-determined statistical threshold, we could make the conclusion that there is
an <em>enrichment</em> or <em>over-representation</em> of our differentially expressed genes
from this experiment in the Hallmark Angiogenesis gene set. Relating this back
to the experiment, we might hypothesize that the differences between our two
cell lines might be driving gene expression changes that result in alterations
in genes involved in angiogenesis. This <em>might</em> motivate potential further <em>in
vitro</em> experiments on these cell lines, including migration and proliferation
assays, that could reveal if this enrichment of angiogenesis genes is reflected
at a phenotypic or functional level.</p>
<p>ORA is a quick and useful way to generate further hypotheses to investigate
specific mechanisms of action or regulation. For example, after identifying a
gene set as being <em>enriched</em> or <em>over-represented</em>, one could further test the
specific genes in the set by examining their directionality of change or asking
if dependent pathways/networks are also perturbed.</p>
<p>One major limitation of ORA is that it relies on the choice of arbitrary
statistical thresholds to define “interesting” or “differentially expressed”
genes. To reiterate again, p-value thresholds hold no inherent biological
meaning and are subjectively determined. Changing the p-value threshold may
result in dramatic differences in the outcome of ORA. Additionally, expression
datasets may measure tens of thousands of genes in a single experiment, and
filtering by a p-value threshold may discard potentially useful information.</p>
<p>ORA (though often modified with slightly different statistical methodologies) is
implemented in a number of different R packages such as
<a href="https://bioconductor.org/packages/release/bioc/html/topGO.html">topGO</a> or
various web services including <a href="https://david.ncifcrf.gov">DAVID</a> and
<a href="https://maayanlab.cloud/Enrichr/">enrichR</a></p>
</div>
<div id="rank-based-analysis-." class="section level3" number="8.6.4">
<h3><span class="header-section-number">8.6.4</span> Rank-based Analysis .</h3>
<p>We will refer to the specific method developed by the Broad Institute and UC San
Diego as Gene Set Enrichment Analysis (note the capitalization). This specific
methodology should not be confused with gene set enrichment analysis, which we
use as an umbrella term covering the many statistical methodologies used to
analyze gene sets.</p>
<p><a href="biology-bioinformatics.html#gene-set-enrichment-analysis">Gene Set Enrichment Analysis</a> (<a href="https://www.gsea-msigdb.org/gsea/index.jsp" class="uri">https://www.gsea-msigdb.org/gsea/index.jsp</a>) was
a method developed to facilitate the biological analysis of genome-wide
expression experiments. It is a rank-based method which utilizes all of the
information from an expression dataset instead of relying on pre-determined
statistical thresholds. In the simplest case, GSEA organizes expression data
into two classes, and ranks all discovered genes by a chosen metric correlating
their expression to these classes. Then, for a pre-defined set of genes, GSEA
tests whether the members of this particular gene set occur more frequently at
the top or bottom of the ranked list or are randomly distributed throughout.
This ranking can be done by a number of different measures, but common ones
include signal-to-noise ratio, signed log p-value or log fold change estimates.
These pre-defined gene sets are flexible and may be manually constructed or
drawn from the many curated databases of gene sets.</p>
<p>Behind the scenes, GSEA functions by descending through the ranked list and
increasing a cumulative score when a gene is encountered that’s contained within
the chosen gene set and decreasing that score when it encounters a gene that is
not within that gene set. This incrementing is weighted to put more emphasis on
genes that are found at the extremes of the ranked list and the final score
taken as the enrichment score (ES) is the maximum of the deviation from zero.
From this score, a p-value is determined by permuting the phenotype labels to
generate a null distribution to compare against. To account for multiple
hypothesis correction, the ES is normalized by the size of the gene set to
create a normalized enrichment score (NES) and the p-values are subjected to
Benjamini-Hochberg correction to generate FDR values for each NES.</p>
<p>To make this explanation more transparent, let us assume that we have performed
a RNAseq experiment to determine which genes are changed when we knock out our
gene of interest in a model cell line. We have obtained our results, which
consists of a list of all genes discovered in the experiment and associated
statistical measures such as p-value and log fold change estimates. To perform a
basic GSEA, we would sort this list of genes, without respect to significance,
by descending fold change to generate a ranked list where genes that are
upregulated (positive fold change) are at the “top” of our list and genes that
are downregulated (negative fold change) are at the “bottom” of our list. We
would use this ranked list and any number or combination of gene sets to perform
GSEA using one of the available means.</p>
<p>GSEA is available as a java application with a graphical user interface as well
as a command-line utility for use in HPC cluster environments. There also exist
several packages in R that also implement the core GSEA algorithm along with
various changes and refinements to the underlying statistical methodology. In
the following section, we will refer to the usage of one such package, fgsea.
Under most circumstances, GSEA can be run with default parameters but please see
the official documentation for a list of customizable parameters and situations
in which to adjust them.</p>
<p>GSEA will return a list of results for gene sets tested consisting of their
associated enrichment scores, and various statistical measures for a given
ranked input With GSEA results, it is typical to set a permissive FDR threshold
to consider which gene sets are significant. Further inspection of individual
gene sets may be performed by analyzing the leading edge subset for each gene
set. This is the group of genes that occur before and contribute to the maximum
deviation from zero score for a given gene set during the analysis. These genes
are typically interpreted as being the more relevant subset of genes that likely
contribute the most to the enrichment signal, and are typically used in further
investigations comparing leading edge subsets from other gene sets or applying
domain knowledge to predict their regulatory effects and functions with respect
to their annotated biological pathway or gene set.</p>
<p>GSEA is a highly flexible method that incorporates information from all of the
data generated in HTS experiments. The choice of ranking methods and gene sets
can all be tailored to the specific experiment or question at hand, and it is a
simple yet fast method that can detect more subtle changes in gene networks than
ORA.</p>
<div class="box warning">
<p>The interpretation of a GSEA result needs to be made with care. Let’s assume
that we have performed GSEA using fold Change as our ranking metric and we find
that the “KEGG Glycolysis” pathway is found to be enriched with a positive
normalized enrichment score and is statistically significant. Alone, this does
not necessarily imply that glycolysis is more active between the conditions of
interest. It simply allows for the conclusion that there is an enrichment of
genes belonging to this gene set among the genes with a positive log fold change
in the experiment. This <em>might</em> imply that your experimental conditions are
affecting some underlying transcriptional regulation that results in these genes
being up-regulated. However, it could also be explained by feedback mechanisms,
direct or indirect, from related biological pathways or gene networks. It <em>does
not</em> allow you to directly conclude that the glycolysis pathway itself is more
active at a functional level. To make that assertion, you would need to perform
further <em>in vitro</em> or <em>in vivo</em> functional assays that directly measure the
activity and outputs of glycolysis given your experimental conditions.</p>
<p>It is important to remember that unless specified, gene sets will often contain
a mix of genes with different regulatory activities including both activators
and inhibitors of that pathway. In addition, for many biological pathways, there
are other factors that influence its regulation that may not be apparent at
a gene expression level. In the specific example of metabolic pathways, it is
common for the enzymatic products of subsequent reactions to feedback and
regulate the activity and flux through the pathway or connected pathways. This
regulation is not easily discerned at the gene expression level alone. All of
this is to say that the results from GSEA must be interpreted with care and
integrated with other analyses and knowledge to gain a holistic understanding
of its meaning.</p>
</div>
</div>
<div id="fgsea" class="section level3" number="8.6.5">
<h3><span class="header-section-number">8.6.5</span> fgsea</h3>
<p>To perform GSEA analysis in R, we will be using the <a href="biology-bioinformatics.html#fgsea">fgsea</a>
(<a href="http://bioconductor.org/packages/release/bioc/html/fgsea.html" class="uri">http://bioconductor.org/packages/release/bioc/html/fgsea.html</a>) package. fgsea
uses the core algorithm behind rank-based GSEA but with additional statistical
methodologies meant to provide better estimations of small p-values. For this
example, we have generated simulated and completely artificial data for the
ranked list. We have chosen 1000 genes to represent all of the genes discovered
in our experiment, annotated them with random values for their fold change and
sorted the list in descending order. The top of our list corresponds to genes
with a positive fold change and the bottom those with a negative fold change.
For this hypothetical experiment, let us assume that genes that have a positive
fold change are upregulated in our condition of interest vs. the reference and
genes that have a negative fold change are downregulated in our condition of
interest vs. the reference.</p>
<div class="box note">
<p>The genes that served as our ranked list input were taken from an experiment
looking at differences in cancer subtypes. Although our fold change values
associated with them were generated randomly, there is likely a strong residual
signal / bias towards various cancer pathways as we will see in our results
below</p>
</div>
<div class="sourceCode" id="cb310"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb310-1"><a href="biology-bioinformatics.html#cb310-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">&#39;fgsea&#39;</span>)</span></code></pre></div>
<p>fgsea takes as input, a named vector of gene-level statistics (a ranked list)
and a named list of gene sets. The gene names in the ranked list must exactly
match how they appear in the gene sets (see the note below for additional
considerations to keep in mind). By default, the <code>fgsea()</code> function will run a
pre-ranked GSEA meaning that you must generate the ranked list manually
beforehand. Importantly, it will not check your ranked list input and will
assume it is sorted numerically in descending order of your ranking metric.</p>
<p>:::{.box .warning}
Gene sets downloaded from the MSigDB are sets of <strong>human</strong> genes and are
represented by either NCBI Entrez IDs or HGNC symbols. For convenience, we have
been working with gene sets containing HGNC symbols. It is important to remember
to check what identifier system and what species the gene set is provided as.
This also becomes relevant when performing rank-based GSEA on data generated in
non-human species. In these situations, you will need to convert the gene IDs
between species and assuming you’re using a pre-defined gene set, match the
identifiers from your ranked list to the format found in the gene set.</p>
<p>This issue typically becomes relevant when trying to use the MSigDB gene sets to
compare against a ranked list of genes generated from HTS experiments in <em>Mus
musculus</em>. MGI symbols typically begin with an uppercase letter and are followed
by all lowercase letters or numbers. While there are examples of exact
concordance where true orthologous mouse and human genes have matching MGI and
HGNC symbols only differing in case (i.e. Gzmb and GZMB), you should <strong>not</strong>
simply convert MGI symbols to all uppercase. This will work for the subset of
genes where the orthologs have the same symbol but will misidentify many genes
where they do not share the same base name. One way to properly perform this
kind of ‘orthology’ mapping is to use <code>biomaRt</code>, which we have discussed
previously. :::</p>
<p>For this example, we have already made simulated vectors containing the gene
names and their associated fold change values. We will generate the appropriate
format using the <code>setNames</code> function on our two vectors and we can use both
<code>head</code> and <code>tail</code> to quickly check the sorting. For non-simulated data, you
could generate these vectors in a number of ways depending upon the format
of your results data.</p>
<p>These are the top 5 genes in our ranked list:</p>
<div class="sourceCode" id="cb311"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb311-1"><a href="biology-bioinformatics.html#cb311-1" aria-hidden="true" tabindex="-1"></a>rnk_list <span class="ot">&lt;-</span> <span class="fu">setNames</span>(rnks, de_genes)</span>
<span id="cb311-2"><a href="biology-bioinformatics.html#cb311-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(rnk_list)</span></code></pre></div>
<pre><code>##  RBMS1P1    RBMS1     GAS1    SFRP2   CCDC80      MGP 
## 9.965527 9.907958 9.888917 9.828172 9.826594 9.802787</code></pre>
<p>And these are the bottom 5 genes in the list:</p>
<div class="sourceCode" id="cb313"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb313-1"><a href="biology-bioinformatics.html#cb313-1" aria-hidden="true" tabindex="-1"></a><span class="fu">tail</span>(rnk_list)</span></code></pre></div>
<pre><code>##      KLF7   ANKDD1A    RIMKLB     RGS19    PLPPR4     TENM4 
## -9.865186 -9.922858 -9.930453 -9.931198 -9.959923 -9.984908</code></pre>
<p>For our gene sets, we will use the same Hallmarks gene set provided by MSigDB.
There are two ways we can load these gene sets in the appropriate format
expected. The first is to use one of the built-in functions in fgsea,
<code>gmtPathways</code>, which will read directly from the GMT file. We have displayed
the contents of a randomly selected gene set below:</p>
<div class="sourceCode" id="cb315"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb315-1"><a href="biology-bioinformatics.html#cb315-1" aria-hidden="true" tabindex="-1"></a>hallmark_pathways_fgsea <span class="ot">&lt;-</span> fgsea<span class="sc">::</span><span class="fu">gmtPathways</span>(<span class="st">&#39;h.all.v7.5.1.symbols.gmt&#39;</span>)</span>
<span id="cb315-2"><a href="biology-bioinformatics.html#cb315-2" aria-hidden="true" tabindex="-1"></a>hallmark_pathways_fgsea<span class="sc">$</span>HALLMARK_TGF_BETA_SIGNALING</span></code></pre></div>
<pre><code>##  [1] &quot;TGFBR1&quot;   &quot;SMAD7&quot;    &quot;TGFB1&quot;    &quot;SMURF2&quot;   &quot;SMURF1&quot;   &quot;BMPR2&quot;   
##  [7] &quot;SKIL&quot;     &quot;SKI&quot;      &quot;ACVR1&quot;    &quot;PMEPA1&quot;   &quot;NCOR2&quot;    &quot;SERPINE1&quot;
## [13] &quot;JUNB&quot;     &quot;SMAD1&quot;    &quot;SMAD6&quot;    &quot;PPP1R15A&quot; &quot;TGIF1&quot;    &quot;FURIN&quot;   
## [19] &quot;SMAD3&quot;    &quot;FKBP1A&quot;   &quot;MAP3K7&quot;   &quot;BMPR1A&quot;   &quot;CTNNB1&quot;   &quot;HIPK2&quot;   
## [25] &quot;KLF10&quot;    &quot;BMP2&quot;     &quot;ENG&quot;      &quot;APC&quot;      &quot;PPM1A&quot;    &quot;XIAP&quot;    
## [31] &quot;CDH1&quot;     &quot;ID1&quot;      &quot;LEFTY2&quot;   &quot;CDKN1C&quot;   &quot;TRIM33&quot;   &quot;RAB31&quot;   
## [37] &quot;TJP1&quot;     &quot;SLC20A1&quot;  &quot;CDK9&quot;     &quot;ID3&quot;      &quot;NOG&quot;      &quot;ARID4B&quot;  
## [43] &quot;IFNGR2&quot;   &quot;ID2&quot;      &quot;PPP1CA&quot;   &quot;SPTBN1&quot;   &quot;WWTR1&quot;    &quot;BCAR3&quot;   
## [49] &quot;THBS1&quot;    &quot;FNTA&quot;     &quot;HDAC1&quot;    &quot;UBE2D3&quot;   &quot;LTBP2&quot;    &quot;RHOA&quot;</code></pre>
<p>If we had previously loaded these gene sets in using GSEABase, we could also
simply do the following:</p>
<div class="sourceCode" id="cb317"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb317-1"><a href="biology-bioinformatics.html#cb317-1" aria-hidden="true" tabindex="-1"></a>hallmark_pathways_GSEABase <span class="ot">&lt;-</span> <span class="fu">geneIds</span>(hallmarks_gmt)</span>
<span id="cb317-2"><a href="biology-bioinformatics.html#cb317-2" aria-hidden="true" tabindex="-1"></a>hallmark_pathways_GSEABase<span class="sc">$</span>HALLMARK_TGF_BETA_SIGNALING</span></code></pre></div>
<pre><code>##  [1] &quot;TGFBR1&quot;   &quot;SMAD7&quot;    &quot;TGFB1&quot;    &quot;SMURF2&quot;   &quot;SMURF1&quot;   &quot;BMPR2&quot;   
##  [7] &quot;SKIL&quot;     &quot;SKI&quot;      &quot;ACVR1&quot;    &quot;PMEPA1&quot;   &quot;NCOR2&quot;    &quot;SERPINE1&quot;
## [13] &quot;JUNB&quot;     &quot;SMAD1&quot;    &quot;SMAD6&quot;    &quot;PPP1R15A&quot; &quot;TGIF1&quot;    &quot;FURIN&quot;   
## [19] &quot;SMAD3&quot;    &quot;FKBP1A&quot;   &quot;MAP3K7&quot;   &quot;BMPR1A&quot;   &quot;CTNNB1&quot;   &quot;HIPK2&quot;   
## [25] &quot;KLF10&quot;    &quot;BMP2&quot;     &quot;ENG&quot;      &quot;APC&quot;      &quot;PPM1A&quot;    &quot;XIAP&quot;    
## [31] &quot;CDH1&quot;     &quot;ID1&quot;      &quot;LEFTY2&quot;   &quot;CDKN1C&quot;   &quot;TRIM33&quot;   &quot;RAB31&quot;   
## [37] &quot;TJP1&quot;     &quot;SLC20A1&quot;  &quot;CDK9&quot;     &quot;ID3&quot;      &quot;NOG&quot;      &quot;ARID4B&quot;  
## [43] &quot;IFNGR2&quot;   &quot;ID2&quot;      &quot;PPP1CA&quot;   &quot;SPTBN1&quot;   &quot;WWTR1&quot;    &quot;BCAR3&quot;   
## [49] &quot;THBS1&quot;    &quot;FNTA&quot;     &quot;HDAC1&quot;    &quot;UBE2D3&quot;   &quot;LTBP2&quot;    &quot;RHOA&quot;</code></pre>
<p>Now to run fgsea, we will use default parameters besides manually setting
thresholds to ignore gene sets in the analysis based on minimum and maximum
values for their size:</p>
<div class="sourceCode" id="cb319"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb319-1"><a href="biology-bioinformatics.html#cb319-1" aria-hidden="true" tabindex="-1"></a>fgsea_results <span class="ot">&lt;-</span> <span class="fu">fgsea</span>(hallmark_pathways_GSEABase, rnk_list, <span class="at">minSize =</span> <span class="dv">15</span>, <span class="at">maxSize=</span><span class="dv">500</span>)</span>
<span id="cb319-2"><a href="biology-bioinformatics.html#cb319-2" aria-hidden="true" tabindex="-1"></a>fgsea_results <span class="ot">&lt;-</span> fgsea_results <span class="sc">%&gt;%</span> <span class="fu">as_tibble</span>()</span></code></pre></div>
<p>As noted in the official documentation for the original GSEA, very small gene
sets or very large gene sets may lead to issues with artificially high scores
or poor normalization, respectively. These values can be adjusted as needed, but
we have used the range suggested by the official documentation.</p>
<p>We will convert the fgsea results into a tibble, and perform some basic
exploration of our results. We can see below the results sorted by ascending
<code>padj</code> and the pathways with the lowest adjusted p-values are largely associated
with cancer pathways including Epithelial Mesenchymal Transition, and Apical
Junctions.</p>
<div class="sourceCode" id="cb320"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb320-1"><a href="biology-bioinformatics.html#cb320-1" aria-hidden="true" tabindex="-1"></a>fgsea_results <span class="sc">%&gt;%</span> <span class="fu">arrange</span>(padj)</span></code></pre></div>
<pre><code>## # A tibble: 15 x 8
##    pathway                 pval     padj log2err     ES    NES  size leadingEdge
##    &lt;chr&gt;                  &lt;dbl&gt;    &lt;dbl&gt;   &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt; &lt;int&gt; &lt;list&gt;     
##  1 HALLMARK_EPITHELIA~ 4.86e-11 7.28e-10  0.851   0.509  2.60    107 &lt;chr [56]&gt; 
##  2 HALLMARK_APICAL_JU~ 1.50e- 3 1.12e- 2  0.455   0.479  1.87     32 &lt;chr [16]&gt; 
##  3 HALLMARK_MYOGENESIS 3.61e- 3 1.81e- 2  0.432   0.428  1.76     40 &lt;chr [24]&gt; 
##  4 HALLMARK_COAGULATI~ 5.16e- 2 1.55e- 1  0.277   0.383  1.46     30 &lt;chr [13]&gt; 
##  5 HALLMARK_COMPLEMENT 5.15e- 2 1.55e- 1  0.277   0.380  1.49     33 &lt;chr [11]&gt; 
##  6 HALLMARK_KRAS_SIGN~ 1.08e- 1 2.37e- 1  0.188   0.329  1.30     35 &lt;chr [13]&gt; 
##  7 HALLMARK_UV_RESPON~ 1.11e- 1 2.37e- 1  0.186   0.325  1.31     37 &lt;chr [17]&gt; 
##  8 HALLMARK_APOPTOSIS  1.28e- 1 2.39e- 1  0.171   0.378  1.31     21 &lt;chr [10]&gt; 
##  9 HALLMARK_HYPOXIA    2.24e- 1 3.74e- 1  0.124   0.298  1.16     31 &lt;chr [12]&gt; 
## 10 HALLMARK_ADIPOGENE~ 2.73e- 1 4.09e- 1  0.111   0.329  1.13     20 &lt;chr [11]&gt; 
## 11 HALLMARK_IL2_STAT5~ 5.71e- 1 7.14e- 1  0.0747 -0.253 -0.925    24 &lt;chr [6]&gt;  
## 12 HALLMARK_INTERFERO~ 5.63e- 1 7.14e- 1  0.0749 -0.254 -0.924    23 &lt;chr [7]&gt;  
## 13 HALLMARK_ALLOGRAFT~ 6.92e- 1 7.99e- 1  0.0646 -0.236 -0.857    23 &lt;chr [4]&gt;  
## 14 HALLMARK_INFLAMMAT~ 9.10e- 1 9.72e- 1  0.0479  0.184  0.683    26 &lt;chr [7]&gt;  
## 15 HALLMARK_TNFA_SIGN~ 9.72e- 1 9.72e- 1  0.0501 -0.171 -0.591    20 &lt;chr [4]&gt;</code></pre>
<p>To begin to explore these results, we could filter by our FDR threshold and
subset significant gene sets by the direction of their NES (positive indicating
an enrichment at the “top” of our list and negative indicating an enrichment at
the “bottom”).</p>
<div class="sourceCode" id="cb322"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb322-1"><a href="biology-bioinformatics.html#cb322-1" aria-hidden="true" tabindex="-1"></a>top_positive_nes <span class="ot">&lt;-</span> fgsea_results <span class="sc">%&gt;%</span></span>
<span id="cb322-2"><a href="biology-bioinformatics.html#cb322-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(padj <span class="sc">&lt;</span> .<span class="dv">25</span> <span class="sc">&amp;</span> NES <span class="sc">&gt;</span> <span class="dv">0</span>) <span class="sc">%&gt;%</span></span>
<span id="cb322-3"><a href="biology-bioinformatics.html#cb322-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">slice_max</span>(NES, <span class="at">n=</span><span class="dv">5</span>)</span>
<span id="cb322-4"><a href="biology-bioinformatics.html#cb322-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb322-5"><a href="biology-bioinformatics.html#cb322-5" aria-hidden="true" tabindex="-1"></a>top_positive_nes</span></code></pre></div>
<pre><code>## # A tibble: 5 x 8
##   pathway                    pval     padj log2err    ES   NES  size leadingEdge
##   &lt;chr&gt;                     &lt;dbl&gt;    &lt;dbl&gt;   &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;int&gt; &lt;list&gt;     
## 1 HALLMARK_EPITHELIAL_M~ 4.86e-11 7.28e-10   0.851 0.509  2.60   107 &lt;chr [56]&gt; 
## 2 HALLMARK_APICAL_JUNCT~ 1.50e- 3 1.12e- 2   0.455 0.479  1.87    32 &lt;chr [16]&gt; 
## 3 HALLMARK_MYOGENESIS    3.61e- 3 1.81e- 2   0.432 0.428  1.76    40 &lt;chr [24]&gt; 
## 4 HALLMARK_COMPLEMENT    5.15e- 2 1.55e- 1   0.277 0.380  1.49    33 &lt;chr [11]&gt; 
## 5 HALLMARK_COAGULATION   5.16e- 2 1.55e- 1   0.277 0.383  1.46    30 &lt;chr [13]&gt;</code></pre>
<p>To provide a basic visualization of these results, we can display the normalized
enrichment scores for all of our pathways in a bar chart and <code>fill</code> the bars by
whether or not they meet our <code>padj</code> threshold.</p>
<div class="box note">
<p>Rank-based GSEA is often used as a hypothesis-generating experiment as it can
quickly capture potentially interesting global patterns of regulation among many
biological pathways and processes while considering all the information
generated in a HTS experiment. It is typical to set a permissive FDR for
considering “significant” gene sets. The results of rank-based GSEA need to be
further inspected and most often confirmed through complementary or follow-up
experiments. Thus, we will choose to use a relaxed FDR threshold of &lt; .25.</p>
</div>
<div class="sourceCode" id="cb324"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb324-1"><a href="biology-bioinformatics.html#cb324-1" aria-hidden="true" tabindex="-1"></a>fgsea_results <span class="sc">%&gt;%</span></span>
<span id="cb324-2"><a href="biology-bioinformatics.html#cb324-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">pathway =</span> forcats<span class="sc">::</span><span class="fu">fct_reorder</span>(pathway, NES)) <span class="sc">%&gt;%</span></span>
<span id="cb324-3"><a href="biology-bioinformatics.html#cb324-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb324-4"><a href="biology-bioinformatics.html#cb324-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_bar</span>(<span class="fu">aes</span>(<span class="at">x=</span>pathway, <span class="at">y=</span>NES, <span class="at">fill =</span> padj <span class="sc">&lt;</span> .<span class="dv">25</span>), <span class="at">stat=</span><span class="st">&#39;identity&#39;</span>) <span class="sc">+</span></span>
<span id="cb324-5"><a href="biology-bioinformatics.html#cb324-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="st">&#39;TRUE&#39;</span> <span class="ot">=</span> <span class="st">&#39;red&#39;</span>, <span class="st">&#39;FALSE&#39;</span> <span class="ot">=</span> <span class="st">&#39;blue&#39;</span>)) <span class="sc">+</span> </span>
<span id="cb324-6"><a href="biology-bioinformatics.html#cb324-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_minimal</span>() <span class="sc">+</span></span>
<span id="cb324-7"><a href="biology-bioinformatics.html#cb324-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">&#39;fgsea results for Hallmark MSigDB gene sets&#39;</span>) <span class="sc">+</span></span>
<span id="cb324-8"><a href="biology-bioinformatics.html#cb324-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">&#39;Normalized Enrichment Score (NES)&#39;</span>) <span class="sc">+</span></span>
<span id="cb324-9"><a href="biology-bioinformatics.html#cb324-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">&#39;&#39;</span>) <span class="sc">+</span></span>
<span id="cb324-10"><a href="biology-bioinformatics.html#cb324-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_flip</span>()</span></code></pre></div>
<div class="figure">
<img src="main_files/top%20level%20results%20plot-1.png" alt="GSEA results suggest an enrichment of cancer-related pathways 
amongst upregulated genes. GSEA was performed using a list of genes
generated through RNAseq and ranked by descending fold change. GSEA was run 
using default parameters with minSize = 15 and maxSize = 500. Gene sets with
a Benjamini-Hochberg adjusted p-value &lt; .25 are considered significant." width="672" />
<p class="caption">
(#fig:top level results plot)GSEA results suggest an enrichment of cancer-related pathways
amongst upregulated genes. GSEA was performed using a list of genes
generated through RNAseq and ranked by descending fold change. GSEA was run
using default parameters with minSize = 15 and maxSize = 500. Gene sets with
a Benjamini-Hochberg adjusted p-value &lt; .25 are considered significant.
</p>
</div>
<p>Given the artificial nature of our data and our choice of a small gene set
collection, we have very few results and are able to display all of them
intelligibly on one plot. In real experiments, there may be several hundred
significant gene sets of interest. In cases like this, you may need to restrict
the number of gene sets by taking the top ten ranked by NES (both positive
and negative) and plotting those in a figure like the one above. This is only a
suggestion and there are many ways to choose ‘interesting’ gene sets to plot
together depending upon the experimental context and future questions of
interest.</p>
<p>If we wished to investigate a single gene set from this list of results, we
could create an enrichment plot displaying the Enrichment Score as it’s
calculated through the ranked list. For convenience, we have wrapped the
pre-built function, <code>plotEnrichment</code> from fgsea, in a user-defined function that
allows us to specify a specific pathway. The <code>plotEnrichment</code> function expects
at minimum the list of genes contained within the pathway to plot and the input
ranked list of gene-level statistics. We have displayed the enrichment plot for
one of the significantly enriched gene sets below:</p>
<div class="sourceCode" id="cb325"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb325-1"><a href="biology-bioinformatics.html#cb325-1" aria-hidden="true" tabindex="-1"></a>make_gsea_plot <span class="ot">&lt;-</span> <span class="cf">function</span>(pathway_name, rnks) {</span>
<span id="cb325-2"><a href="biology-bioinformatics.html#cb325-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plotEnrichment</span>(hallmark_pathways_fgsea[[pathway_name]], rnks) <span class="sc">+</span></span>
<span id="cb325-3"><a href="biology-bioinformatics.html#cb325-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">labs</span>(<span class="at">title=</span>pathway_name)</span>
<span id="cb325-4"><a href="biology-bioinformatics.html#cb325-4" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb325-5"><a href="biology-bioinformatics.html#cb325-5" aria-hidden="true" tabindex="-1"></a><span class="fu">make_gsea_plot</span>(<span class="st">&#39;HALLMARK_EPITHELIAL_MESENCHYMAL_TRANSITION&#39;</span>, rnk_list)</span></code></pre></div>
<div class="figure">
<img src="main_files/enrichment%20plot-1.png" alt="The Hallmarks EPITHELIAL_MESENCHYMAL_TRANSITION Pathway
is significantly enriched amongst the upregulated genes. There are 56 genes in the 
leading edge subset including GAS1, POSTN, and LOX. The Normalized Enrichment
Score calculated for this gene set was 2.59 and it has an adjusted p-value of 
7.28e-10" width="672" />
<p class="caption">
(#fig:enrichment plot)The Hallmarks EPITHELIAL_MESENCHYMAL_TRANSITION Pathway
is significantly enriched amongst the upregulated genes. There are 56 genes in the
leading edge subset including GAS1, POSTN, and LOX. The Normalized Enrichment
Score calculated for this gene set was 2.59 and it has an adjusted p-value of
7.28e-10
</p>
</div>
<p>As mentioned in the prior section, it is often useful for further investigation
to inspect the leading edge subset of genes contained within ‘interesting’ gene
sets. fgsea stores these genes in the form of a list associated with each
pathway in the returned table. We could convert and save these values in a named
list for convenience. This can be accomplished by making use of the <code>deframe()</code>
function. Below we have saved all of the genes contained within the leading edge
subset of significant gene sets, and displayed those belonging to the
HALLMARK_EPITHELIAL_MESENCHYMAL_TRANSITION gene set as determined in this
experiment:</p>
<div class="sourceCode" id="cb326"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb326-1"><a href="biology-bioinformatics.html#cb326-1" aria-hidden="true" tabindex="-1"></a>gather_leadingedge <span class="ot">&lt;-</span> <span class="cf">function</span>(results, threshold) {</span>
<span id="cb326-2"><a href="biology-bioinformatics.html#cb326-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb326-3"><a href="biology-bioinformatics.html#cb326-3" aria-hidden="true" tabindex="-1"></a>  genes <span class="ot">&lt;-</span> results <span class="sc">%&gt;%</span> </span>
<span id="cb326-4"><a href="biology-bioinformatics.html#cb326-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(padj <span class="sc">&lt;</span> threshold) <span class="sc">%&gt;%</span></span>
<span id="cb326-5"><a href="biology-bioinformatics.html#cb326-5" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">select</span>(pathway, leadingEdge) <span class="sc">%&gt;%</span></span>
<span id="cb326-6"><a href="biology-bioinformatics.html#cb326-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">deframe</span>()</span>
<span id="cb326-7"><a href="biology-bioinformatics.html#cb326-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(genes)</span>
<span id="cb326-8"><a href="biology-bioinformatics.html#cb326-8" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb326-9"><a href="biology-bioinformatics.html#cb326-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb326-10"><a href="biology-bioinformatics.html#cb326-10" aria-hidden="true" tabindex="-1"></a>leading_edge_genes <span class="ot">&lt;-</span> <span class="fu">gather_leadingedge</span>(fgsea_results, .<span class="dv">25</span>)</span>
<span id="cb326-11"><a href="biology-bioinformatics.html#cb326-11" aria-hidden="true" tabindex="-1"></a>leading_edge_genes<span class="sc">$</span>HALLMARK_EPITHELIAL_MESENCHYMAL_TRANSITION</span></code></pre></div>
<pre><code>##  [1] &quot;GAS1&quot;   &quot;MGP&quot;    &quot;SPOCK1&quot; &quot;EFEMP2&quot; &quot;FERMT2&quot; &quot;FBN1&quot;   &quot;TAGLN&quot;  &quot;FSTL1&quot; 
##  [9] &quot;TIMP3&quot;  &quot;VIM&quot;    &quot;VCAN&quot;   &quot;THBS2&quot;  &quot;ACTA2&quot;  &quot;DCN&quot;    &quot;BGN&quot;    &quot;NNMT&quot;  
## [17] &quot;CALD1&quot;  &quot;NTM&quot;    &quot;FBLN1&quot;  &quot;SPARC&quot;  &quot;COL5A2&quot; &quot;MYL9&quot;   &quot;SLIT2&quot;  &quot;DPYSL3&quot;
## [25] &quot;VEGFC&quot;  &quot;COL3A1&quot; &quot;LGALS1&quot; &quot;THY1&quot;   &quot;PCOLCE&quot; &quot;ADAM12&quot; &quot;SFRP4&quot;  &quot;WIPF1&quot; 
## [33] &quot;COL6A3&quot; &quot;CTHRC1&quot; &quot;TIMP1&quot;  &quot;TPM2&quot;   &quot;FN1&quot;    &quot;COL5A1&quot; &quot;PMP22&quot;  &quot;MMP2&quot;  
## [41] &quot;HTRA1&quot;  &quot;COL1A2&quot; &quot;POSTN&quot;  &quot;PDGFRB&quot; &quot;EMP3&quot;   &quot;MXRA5&quot;  &quot;FLNA&quot;   &quot;PRRX1&quot; 
## [49] &quot;FAP&quot;    &quot;LOX&quot;    &quot;BASP1&quot;  &quot;GREM1&quot;  &quot;CCN2&quot;   &quot;CDH11&quot;  &quot;LUM&quot;    &quot;LAMC1&quot;</code></pre>
<p>This named list contains all of the leading edge genes for gene sets meeting a
certain <code>padj</code> threshold. We could import these as a GSEABase GeneSetCollection
object for further analyses in R or write them to a file to pass on to
collaborators.</p>
<p>These examples are just some of the ways you can utilize and explore the GSEA
results returned by fgsea. There are also other methods such as the one found in
the built-in function, <code>fgsea::plotGseaTable</code>, that provide other ways to
visualize these results for groups of different gene sets. You can also perform
any number of different filtering and sorting options on the table of results
depending upon your interests.</p>

</div>
</div>
<div id="biological-networks-." class="section level2" number="8.7">
<h2><span class="header-section-number">8.7</span> Biological Networks .</h2>
<p>What are biological networks? What do they represent (networks encode
relationships between entities)? Why are biological networks useful and what can
we learn from them?</p>
<div id="biological-pathways-." class="section level3" number="8.7.1">
<h3><span class="header-section-number">8.7.1</span> Biological Pathways .</h3>
<p>What is a biological pathway? Why are they important, and how are they useful?</p>
</div>
<div id="gene-regulatory-networks-." class="section level3" number="8.7.2">
<h3><span class="header-section-number">8.7.2</span> Gene Regulatory Networks .</h3>
<p>What are gene regulatory networks? Why are they important? How do we identify
them (data driven correlation, wetlab experiments, others?)? How are they
useful?</p>
</div>
<div id="protein-protein-interaction-network-." class="section level3" number="8.7.3">
<h3><span class="header-section-number">8.7.3</span> Protein-Protein Interaction Network .</h3>
<p>What are protein-protein interaction (PPI) networks? What information do they
represent (direct association, functional association, etc)? Where does PPI
information come from (datasets, databases, etc)? What are some ways we can use
PPI information when interpreting other biological data (like differential
expression? not sure)?</p>
</div>
<div id="wgcna-." class="section level3" number="8.7.4">
<h3><span class="header-section-number">8.7.4</span> WGCNA .</h3>
<p>What is WGCNA? What problem does WGCNA attempt to solve, and how is it different
than other classes of analysis methods (e.g. differential expression)? What data
are appropriate for use in WGCNA? How do we run an interpret a WGCNA analysis in
R?</p>

</div>
</div>
<div id="single-cell-sequencing-analysis-." class="section level2" number="8.8">
<h2><span class="header-section-number">8.8</span> Single Cell Sequencing Analysis .</h2>
<div id="single-cell-sequencing-." class="section level3" number="8.8.1">
<h3><span class="header-section-number">8.8.1</span> Single Cell Sequencing .</h3>
<p>What is the goal of single cell sequencing? How is it generated, and what does
it measure? What questions can we ask of single cell data that we cannot ask of
other types of sequencing data? How is single cell data loaded into R?</p>
</div>
<div id="single-cell-analysis-." class="section level3" number="8.8.2">
<h3><span class="header-section-number">8.8.2</span> Single Cell Analysis .</h3>
<p>What is the general analytical workflow for single cell analysis? Why do we
perform each step along the analysis path? What are key parameters that we must
choose that influence the results? How do we perform single cell analysis in R
(Seurat)?</p>
</div>
<div id="single-cell-clustering-." class="section level3" number="8.8.3">
<h3><span class="header-section-number">8.8.3</span> Single Cell Clustering .</h3>
</div>
<div id="dimensionality-reduction-projection-." class="section level3" number="8.8.4">
<h3><span class="header-section-number">8.8.4</span> Dimensionality Reduction &amp; Projection .</h3>
<p>PCA, tSNE, UMAP</p>
</div>
<div id="single-cell-data-visualization-." class="section level3" number="8.8.5">
<h3><span class="header-section-number">8.8.5</span> Single Cell Data Visualization .</h3>
</div>
<div id="single-cell-marker-analysis-." class="section level3" number="8.8.6">
<h3><span class="header-section-number">8.8.6</span> Single Cell Marker Analysis .</h3>

</div>
</div>
</div>
<h3>References</h3>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-Bruford2020-zs" class="csl-entry">
Bruford, Elspeth A, Bryony Braschi, Paul Denny, Tamsin E M Jones, Ruth L Seal, and Susan Tweedie. 2020. <span>“Guidelines for Human Gene Nomenclature.”</span> <em>Nat. Genet.</em> 52 (8): 754–58.
</div>
<div id="ref-Holley1965-tu" class="csl-entry">
Holley, R W, J Apgar, G A Everett, J T Madison, M Marquisee, S H Merrill, J R Penswick, and A Zamir. 1965. <span>“<span>STRUCTURE</span> <span>OF</span> a <span>RIBONUCLEIC</span> <span>ACID</span>.”</span> <em>Science</em> 147 (3664): 1462–65.
</div>
<div id="ref-Irizarry2003-hn" class="csl-entry">
Irizarry, Rafael A, Bridget Hobbs, Francois Collin, Yasmin D Beazer-Barclay, Kristen J Antonellis, Uwe Scherf, and Terence P Speed. 2003. <span>“Exploration, Normalization, and Summaries of High Density Oligonucleotide Array Probe Level Data.”</span> <em>Biostatistics</em> 4 (2): 249–64.
</div>
<div id="ref-Ritchie2015-bj" class="csl-entry">
Ritchie, Matthew E, Belinda Phipson, Di Wu, Yifang Hu, Charity W Law, Wei Shi, and Gordon K Smyth. 2015. <span>“Limma Powers Differential Expression Analyses for <span class="nocase">RNA-sequencing</span> and Microarray Studies.”</span> <em>Nucleic Acids Res.</em>, January, gkv007.
</div>
<div id="ref-Shows1979-iw" class="csl-entry">
Shows, T B, C A Alper, D Bootsma, M Dorf, T Douglas, T Huisman, S Kit, et al. 1979. <span>“International System for Human Gene Nomenclature (1979) <span>ISGN</span> (1979).”</span> <em>Cytogenet. Cell Genet.</em> 25 (1-4): 96–116.
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="data-visualization.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="engineering.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
