<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>7 Biology &amp; Bioinformatics | BF591 - R for Biological Sciences</title>
  <meta name="description" content="7 Biology &amp; Bioinformatics | BF591 - R for Biological Sciences" />
  <meta name="generator" content="bookdown 0.24 and GitBook 2.6.7" />

  <meta property="og:title" content="7 Biology &amp; Bioinformatics | BF591 - R for Biological Sciences" />
  <meta property="og:type" content="book" />
  
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="7 Biology &amp; Bioinformatics | BF591 - R for Biological Sciences" />
  
  
  




  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="data-science.html"/>
<link rel="next" href="data-visualization.html"/>
<script src="libs/header-attrs-2.11/header-attrs.js"></script>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.0.1/anchor-sections.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.0.1/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Syllabus</a>
<ul>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#course-schedule"><i class="fa fa-check"></i>Course Schedule</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#instructors"><i class="fa fa-check"></i>Instructors</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#course-values-and-policies"><i class="fa fa-check"></i>Course Values and Policies</a></li>
</ul></li>
<li class="chapter" data-level="1" data-path="introduction.html"><a href="introduction.html"><i class="fa fa-check"></i><b>1</b> Introduction</a>
<ul>
<li class="chapter" data-level="1.1" data-path="introduction.html"><a href="introduction.html#who-this-book-is-for"><i class="fa fa-check"></i><b>1.1</b> Who This Book Is For</a></li>
<li class="chapter" data-level="1.2" data-path="introduction.html"><a href="introduction.html#a-note-about-reinventing-the-wheel"><i class="fa fa-check"></i><b>1.2</b> A Note About Reinventing the Wheel</a></li>
<li class="chapter" data-level="1.3" data-path="introduction.html"><a href="introduction.html#sources-and-references"><i class="fa fa-check"></i><b>1.3</b> Sources and References</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="data-bio.html"><a href="data-bio.html"><i class="fa fa-check"></i><b>2</b> Data in Biology</a>
<ul>
<li class="chapter" data-level="2.1" data-path="data-bio.html"><a href="data-bio.html#a-brief-history-of-data-in-molecular-biology"><i class="fa fa-check"></i><b>2.1</b> A Brief History of Data in Molecular Biology</a></li>
<li class="chapter" data-level="2.2" data-path="data-bio.html"><a href="data-bio.html#biology-as-a-mature-data-science"><i class="fa fa-check"></i><b>2.2</b> Biology as a Mature Data Science</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="preliminaries.html"><a href="preliminaries.html"><i class="fa fa-check"></i><b>3</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="3.1" data-path="preliminaries.html"><a href="preliminaries.html#prelim-r"><i class="fa fa-check"></i><b>3.1</b> The R Language</a></li>
<li class="chapter" data-level="3.2" data-path="preliminaries.html"><a href="preliminaries.html#prelim-rstudio"><i class="fa fa-check"></i><b>3.2</b> RStudio</a></li>
<li class="chapter" data-level="3.3" data-path="preliminaries.html"><a href="preliminaries.html#prog-the-r-script"><i class="fa fa-check"></i><b>3.3</b> The R Script</a></li>
<li class="chapter" data-level="3.4" data-path="preliminaries.html"><a href="preliminaries.html#prog-workflow"><i class="fa fa-check"></i><b>3.4</b> The Scripting Workflow</a></li>
<li class="chapter" data-level="3.5" data-path="preliminaries.html"><a href="preliminaries.html#prelim-git"><i class="fa fa-check"></i><b>3.5</b> git + github</a>
<ul>
<li class="chapter" data-level="3.5.1" data-path="preliminaries.html"><a href="preliminaries.html#motivation"><i class="fa fa-check"></i><b>3.5.1</b> Motivation</a></li>
<li class="chapter" data-level="3.5.2" data-path="preliminaries.html"><a href="preliminaries.html#git"><i class="fa fa-check"></i><b>3.5.2</b> git</a></li>
<li class="chapter" data-level="3.5.3" data-path="preliminaries.html"><a href="preliminaries.html#git-hosting-platforms-github"><i class="fa fa-check"></i><b>3.5.3</b> Git hosting platforms (GitHub)</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="prog-basics.html"><a href="prog-basics.html"><i class="fa fa-check"></i><b>4</b> R Programming</a>
<ul>
<li class="chapter" data-level="4.1" data-path="prog-basics.html"><a href="prog-basics.html#before-you-begin"><i class="fa fa-check"></i><b>4.1</b> Before you begin</a></li>
<li class="chapter" data-level="4.2" data-path="prog-basics.html"><a href="prog-basics.html#introduction-1"><i class="fa fa-check"></i><b>4.2</b> Introduction</a></li>
<li class="chapter" data-level="4.3" data-path="prog-basics.html"><a href="prog-basics.html#prog-r-syntax"><i class="fa fa-check"></i><b>4.3</b> R Syntax Basics</a></li>
<li class="chapter" data-level="4.4" data-path="prog-basics.html"><a href="prog-basics.html#prog-types"><i class="fa fa-check"></i><b>4.4</b> Basic Types of Values</a></li>
<li class="chapter" data-level="4.5" data-path="prog-basics.html"><a href="prog-basics.html#prog-struct"><i class="fa fa-check"></i><b>4.5</b> Data Structures</a>
<ul>
<li class="chapter" data-level="4.5.1" data-path="prog-basics.html"><a href="prog-basics.html#vectors"><i class="fa fa-check"></i><b>4.5.1</b> Vectors</a></li>
<li class="chapter" data-level="4.5.2" data-path="prog-basics.html"><a href="prog-basics.html#matrices"><i class="fa fa-check"></i><b>4.5.2</b> Matrices</a></li>
<li class="chapter" data-level="4.5.3" data-path="prog-basics.html"><a href="prog-basics.html#lists-and-data-frames"><i class="fa fa-check"></i><b>4.5.3</b> Lists and data frames</a></li>
</ul></li>
<li class="chapter" data-level="4.6" data-path="prog-basics.html"><a href="prog-basics.html#logical-tests-and-comparators"><i class="fa fa-check"></i><b>4.6</b> Logical Tests and Comparators</a></li>
<li class="chapter" data-level="4.7" data-path="prog-basics.html"><a href="prog-basics.html#functions"><i class="fa fa-check"></i><b>4.7</b> Functions</a>
<ul>
<li class="chapter" data-level="4.7.1" data-path="prog-basics.html"><a href="prog-basics.html#prog-dry"><i class="fa fa-check"></i><b>4.7.1</b> DRY: Don’t Repeat Yourself</a></li>
<li class="chapter" data-level="4.7.2" data-path="prog-basics.html"><a href="prog-basics.html#writing-your-own-functions"><i class="fa fa-check"></i><b>4.7.2</b> Writing your own functions</a></li>
<li class="chapter" data-level="4.7.3" data-path="prog-basics.html"><a href="prog-basics.html#scope"><i class="fa fa-check"></i><b>4.7.3</b> Scope</a></li>
</ul></li>
<li class="chapter" data-level="4.8" data-path="prog-basics.html"><a href="prog-basics.html#iteration"><i class="fa fa-check"></i><b>4.8</b> Iteration</a>
<ul>
<li class="chapter" data-level="4.8.1" data-path="prog-basics.html"><a href="prog-basics.html#vectorized-operations"><i class="fa fa-check"></i><b>4.8.1</b> Vectorized operations</a></li>
<li class="chapter" data-level="4.8.2" data-path="prog-basics.html"><a href="prog-basics.html#functional-programming"><i class="fa fa-check"></i><b>4.8.2</b> Functional programming</a></li>
<li class="chapter" data-level="4.8.3" data-path="prog-basics.html"><a href="prog-basics.html#apply-and-friends"><i class="fa fa-check"></i><b>4.8.3</b> <code>apply()</code> and friends</a></li>
</ul></li>
<li class="chapter" data-level="4.9" data-path="prog-basics.html"><a href="prog-basics.html#installing-packages"><i class="fa fa-check"></i><b>4.9</b> Installing Packages</a></li>
<li class="chapter" data-level="4.10" data-path="prog-basics.html"><a href="prog-basics.html#saving-and-loading-r-data"><i class="fa fa-check"></i><b>4.10</b> Saving and Loading R Data</a></li>
<li class="chapter" data-level="4.11" data-path="prog-basics.html"><a href="prog-basics.html#prog-debugging"><i class="fa fa-check"></i><b>4.11</b> Troubleshooting and Debugging</a></li>
<li class="chapter" data-level="4.12" data-path="prog-basics.html"><a href="prog-basics.html#prog-style"><i class="fa fa-check"></i><b>4.12</b> Coding Style and Conventions</a>
<ul>
<li class="chapter" data-level="4.12.1" data-path="prog-basics.html"><a href="prog-basics.html#is-my-code-correct"><i class="fa fa-check"></i><b>4.12.1</b> Is my code correct?</a></li>
<li class="chapter" data-level="4.12.2" data-path="prog-basics.html"><a href="prog-basics.html#does-my-code-follow-the-dry-principle"><i class="fa fa-check"></i><b>4.12.2</b> Does my code follow the <span>DRY principle</span>?</a></li>
<li class="chapter" data-level="4.12.3" data-path="prog-basics.html"><a href="prog-basics.html#did-i-choose-concise-but-descriptive-variable-and-function-names"><i class="fa fa-check"></i><b>4.12.3</b> Did I choose concise but descriptive variable and function names?</a></li>
<li class="chapter" data-level="4.12.4" data-path="prog-basics.html"><a href="prog-basics.html#did-i-use-indentation-and-naming-conventions-consistently-throughout-my-code"><i class="fa fa-check"></i><b>4.12.4</b> Did I use indentation and naming conventions consistently throughout my code?</a></li>
<li class="chapter" data-level="4.12.5" data-path="prog-basics.html"><a href="prog-basics.html#did-i-write-comments-especially-when-what-the-code-does-is-not-obvious"><i class="fa fa-check"></i><b>4.12.5</b> Did I write comments, especially when what the code does is not obvious?</a></li>
<li class="chapter" data-level="4.12.6" data-path="prog-basics.html"><a href="prog-basics.html#how-easy-would-it-be-for-someone-else-to-understand-my-code"><i class="fa fa-check"></i><b>4.12.6</b> How easy would it be for someone else to understand my code?</a></li>
<li class="chapter" data-level="4.12.7" data-path="prog-basics.html"><a href="prog-basics.html#is-my-code-easy-to-maintainchange"><i class="fa fa-check"></i><b>4.12.7</b> Is my code easy to maintain/change?</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="5" data-path="data-wrangling.html"><a href="data-wrangling.html"><i class="fa fa-check"></i><b>5</b> Data Wrangling</a>
<ul>
<li class="chapter" data-level="5.1" data-path="data-wrangling.html"><a href="data-wrangling.html#the-tidyverse"><i class="fa fa-check"></i><b>5.1</b> The Tidyverse</a></li>
<li class="chapter" data-level="5.2" data-path="data-wrangling.html"><a href="data-wrangling.html#dw-basics"><i class="fa fa-check"></i><b>5.2</b> Tidyverse Basics</a></li>
<li class="chapter" data-level="5.3" data-path="data-wrangling.html"><a href="data-wrangling.html#importing-data"><i class="fa fa-check"></i><b>5.3</b> Importing Data</a></li>
<li class="chapter" data-level="5.4" data-path="data-wrangling.html"><a href="data-wrangling.html#data-tibble"><i class="fa fa-check"></i><b>5.4</b> The tibble</a></li>
<li class="chapter" data-level="5.5" data-path="data-wrangling.html"><a href="data-wrangling.html#tidy-data"><i class="fa fa-check"></i><b>5.5</b> Tidy Data</a></li>
<li class="chapter" data-level="5.6" data-path="data-wrangling.html"><a href="data-wrangling.html#pipes"><i class="fa fa-check"></i><b>5.6</b> pipes</a></li>
<li class="chapter" data-level="5.7" data-path="data-wrangling.html"><a href="data-wrangling.html#arranging-data"><i class="fa fa-check"></i><b>5.7</b> Arranging Data</a>
<ul>
<li class="chapter" data-level="5.7.1" data-path="data-wrangling.html"><a href="data-wrangling.html#dplyrmutate---create-new-columns-using-other-columns"><i class="fa fa-check"></i><b>5.7.1</b> <code>dplyr::mutate()</code> - Create new columns using other columns</a></li>
<li class="chapter" data-level="5.7.2" data-path="data-wrangling.html"><a href="data-wrangling.html#stringr---working-with-character-values"><i class="fa fa-check"></i><b>5.7.2</b> <code>stringr</code> - Working with character values</a></li>
<li class="chapter" data-level="5.7.3" data-path="data-wrangling.html"><a href="data-wrangling.html#dplyrselect---subset-columns-by-name"><i class="fa fa-check"></i><b>5.7.3</b> <code>dplyr::select()</code> - Subset Columns by Name</a></li>
<li class="chapter" data-level="5.7.4" data-path="data-wrangling.html"><a href="data-wrangling.html#dplyrfilter---pick-rows-out-of-a-data-set"><i class="fa fa-check"></i><b>5.7.4</b> <code>dplyr::filter()</code> - Pick rows out of a data set</a></li>
<li class="chapter" data-level="5.7.5" data-path="data-wrangling.html"><a href="data-wrangling.html#dplyrarrange---order-rows-based-on-their-values"><i class="fa fa-check"></i><b>5.7.5</b> <code>dplyr::arrange()</code> - Order rows based on their values</a></li>
<li class="chapter" data-level="5.7.6" data-path="data-wrangling.html"><a href="data-wrangling.html#putting-it-all-together"><i class="fa fa-check"></i><b>5.7.6</b> Putting it all together</a></li>
</ul></li>
<li class="chapter" data-level="5.8" data-path="data-wrangling.html"><a href="data-wrangling.html#grouping-data"><i class="fa fa-check"></i><b>5.8</b> Grouping Data</a></li>
<li class="chapter" data-level="5.9" data-path="data-wrangling.html"><a href="data-wrangling.html#rearranging-data"><i class="fa fa-check"></i><b>5.9</b> Rearranging Data</a></li>
<li class="chapter" data-level="5.10" data-path="data-wrangling.html"><a href="data-wrangling.html#relational-data"><i class="fa fa-check"></i><b>5.10</b> Relational Data</a>
<ul>
<li class="chapter" data-level="5.10.1" data-path="data-wrangling.html"><a href="data-wrangling.html#dealing-with-multiple-matches"><i class="fa fa-check"></i><b>5.10.1</b> Dealing with multiple matches</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="6" data-path="data-science.html"><a href="data-science.html"><i class="fa fa-check"></i><b>6</b> Data Science</a>
<ul>
<li class="chapter" data-level="6.1" data-path="data-science.html"><a href="data-science.html#ds-model"><i class="fa fa-check"></i><b>6.1</b> Data Modeling</a></li>
<li class="chapter" data-level="6.2" data-path="data-science.html"><a href="data-science.html#ds-dists"><i class="fa fa-check"></i><b>6.2</b> Distributions &amp; Tests</a></li>
<li class="chapter" data-level="6.3" data-path="data-science.html"><a href="data-science.html#multiple-hypothesis-testing"><i class="fa fa-check"></i><b>6.3</b> Multiple Hypothesis Testing</a></li>
<li class="chapter" data-level="6.4" data-path="data-science.html"><a href="data-science.html#ds-summ"><i class="fa fa-check"></i><b>6.4</b> Data summarization &amp; cleaning</a></li>
<li class="chapter" data-level="6.5" data-path="data-science.html"><a href="data-science.html#ds-clust"><i class="fa fa-check"></i><b>6.5</b> Clustering</a></li>
<li class="chapter" data-level="6.6" data-path="data-science.html"><a href="data-science.html#ds-networks"><i class="fa fa-check"></i><b>6.6</b> Network Analysis</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html"><i class="fa fa-check"></i><b>7</b> Biology &amp; Bioinformatics</a>
<ul>
<li class="chapter" data-level="7.1" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#bio-r"><i class="fa fa-check"></i><b>7.1</b> R in Biology</a></li>
<li class="chapter" data-level="7.2" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#csv-files"><i class="fa fa-check"></i><b>7.2</b> CSV Files</a></li>
<li class="chapter" data-level="7.3" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#bio-data-types"><i class="fa fa-check"></i><b>7.3</b> Types of Biological Data</a></li>
<li class="chapter" data-level="7.4" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#common-biological-data-matrices"><i class="fa fa-check"></i><b>7.4</b> Common Biological Data Matrices</a>
<ul>
<li class="chapter" data-level="7.4.1" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#biological-data-is-not-tidy"><i class="fa fa-check"></i><b>7.4.1</b> Biological data is NOT Tidy!</a></li>
</ul></li>
<li class="chapter" data-level="7.5" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#bio-bioconductor"><i class="fa fa-check"></i><b>7.5</b> Bioconductor</a></li>
<li class="chapter" data-level="7.6" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#gene-identifiers"><i class="fa fa-check"></i><b>7.6</b> Gene Identifiers</a>
<ul>
<li class="chapter" data-level="7.6.1" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#gene-identifier-systems"><i class="fa fa-check"></i><b>7.6.1</b> Gene Identifier Systems</a></li>
<li class="chapter" data-level="7.6.2" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#mapping-between-identifier-systems"><i class="fa fa-check"></i><b>7.6.2</b> Mapping Between Identifier Systems</a></li>
<li class="chapter" data-level="7.6.3" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#mapping-homologs"><i class="fa fa-check"></i><b>7.6.3</b> Mapping Homologs</a></li>
</ul></li>
<li class="chapter" data-level="7.7" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#bio-geneexp"><i class="fa fa-check"></i><b>7.7</b> Gene Expression</a></li>
<li class="chapter" data-level="7.8" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#microarrays"><i class="fa fa-check"></i><b>7.8</b> Microarrays</a></li>
<li class="chapter" data-level="7.9" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#bio-micro"><i class="fa fa-check"></i><b>7.9</b> Differential Expression: Microarrays</a></li>
<li class="chapter" data-level="7.10" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#bio-gsea"><i class="fa fa-check"></i><b>7.10</b> Gene Set Enrichment Analysis</a></li>
<li class="chapter" data-level="7.11" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#bio-hts"><i class="fa fa-check"></i><b>7.11</b> High Throughput Sequencing</a></li>
<li class="chapter" data-level="7.12" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#bio-rnaseq"><i class="fa fa-check"></i><b>7.12</b> Differential Expression: RNASeq</a></li>
<li class="chapter" data-level="7.13" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#bio-networks"><i class="fa fa-check"></i><b>7.13</b> Biological Network Analysis</a></li>
<li class="chapter" data-level="7.14" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#bio-scseq"><i class="fa fa-check"></i><b>7.14</b> Single Cell Sequencing Analysis</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="data-visualization.html"><a href="data-visualization.html"><i class="fa fa-check"></i><b>8</b> Data Visualization</a>
<ul>
<li class="chapter" data-level="8.1" data-path="data-visualization.html"><a href="data-visualization.html#dv-gg"><i class="fa fa-check"></i><b>8.1</b> Grammar of Graphics</a></li>
<li class="chapter" data-level="8.2" data-path="data-visualization.html"><a href="data-visualization.html#plotting-one-dimension"><i class="fa fa-check"></i><b>8.2</b> Plotting One Dimension</a>
<ul>
<li class="chapter" data-level="8.2.1" data-path="data-visualization.html"><a href="data-visualization.html#dv-bar"><i class="fa fa-check"></i><b>8.2.1</b> Bar chart</a></li>
<li class="chapter" data-level="8.2.2" data-path="data-visualization.html"><a href="data-visualization.html#lollipop-plots"><i class="fa fa-check"></i><b>8.2.2</b> Lollipop plots</a></li>
<li class="chapter" data-level="8.2.3" data-path="data-visualization.html"><a href="data-visualization.html#parallel-coordinate-plots"><i class="fa fa-check"></i><b>8.2.3</b> Parallel Coordinate plots</a></li>
</ul></li>
<li class="chapter" data-level="8.3" data-path="data-visualization.html"><a href="data-visualization.html#visualizing-distributions"><i class="fa fa-check"></i><b>8.3</b> Visualizing Distributions</a>
<ul>
<li class="chapter" data-level="8.3.1" data-path="data-visualization.html"><a href="data-visualization.html#histogram"><i class="fa fa-check"></i><b>8.3.1</b> Histogram</a></li>
<li class="chapter" data-level="8.3.2" data-path="data-visualization.html"><a href="data-visualization.html#density"><i class="fa fa-check"></i><b>8.3.2</b> Density</a></li>
<li class="chapter" data-level="8.3.3" data-path="data-visualization.html"><a href="data-visualization.html#boxplot"><i class="fa fa-check"></i><b>8.3.3</b> Boxplot</a></li>
<li class="chapter" data-level="8.3.4" data-path="data-visualization.html"><a href="data-visualization.html#violin-plot"><i class="fa fa-check"></i><b>8.3.4</b> Violin plot</a></li>
<li class="chapter" data-level="8.3.5" data-path="data-visualization.html"><a href="data-visualization.html#beeswarm-plot"><i class="fa fa-check"></i><b>8.3.5</b> Beeswarm plot</a></li>
<li class="chapter" data-level="8.3.6" data-path="data-visualization.html"><a href="data-visualization.html#ridgeline"><i class="fa fa-check"></i><b>8.3.6</b> Ridgeline</a></li>
</ul></li>
<li class="chapter" data-level="8.4" data-path="data-visualization.html"><a href="data-visualization.html#plotting-two-or-more-dimensions"><i class="fa fa-check"></i><b>8.4</b> Plotting Two or More Dimensions</a>
<ul>
<li class="chapter" data-level="8.4.1" data-path="data-visualization.html"><a href="data-visualization.html#scatter-plot"><i class="fa fa-check"></i><b>8.4.1</b> Scatter plot</a></li>
<li class="chapter" data-level="8.4.2" data-path="data-visualization.html"><a href="data-visualization.html#line-plot"><i class="fa fa-check"></i><b>8.4.2</b> Line plot</a></li>
<li class="chapter" data-level="8.4.3" data-path="data-visualization.html"><a href="data-visualization.html#bubble-plot"><i class="fa fa-check"></i><b>8.4.3</b> Bubble plot</a></li>
<li class="chapter" data-level="8.4.4" data-path="data-visualization.html"><a href="data-visualization.html#heatmaps"><i class="fa fa-check"></i><b>8.4.4</b> Heatmaps</a></li>
</ul></li>
<li class="chapter" data-level="8.5" data-path="data-visualization.html"><a href="data-visualization.html#other-kind-of-plots"><i class="fa fa-check"></i><b>8.5</b> Other Kind of Plots</a></li>
<li class="chapter" data-level="8.6" data-path="data-visualization.html"><a href="data-visualization.html#ds-network-viz"><i class="fa fa-check"></i><b>8.6</b> Network Viz</a></li>
<li class="chapter" data-level="8.7" data-path="data-visualization.html"><a href="data-visualization.html#tips-and-tricks"><i class="fa fa-check"></i><b>8.7</b> Tips and Tricks</a></li>
<li class="chapter" data-level="8.8" data-path="data-visualization.html"><a href="data-visualization.html#multiple-plots"><i class="fa fa-check"></i><b>8.8</b> Multiple Plots</a>
<ul>
<li class="chapter" data-level="8.8.1" data-path="data-visualization.html"><a href="data-visualization.html#facet-wrapping"><i class="fa fa-check"></i><b>8.8.1</b> Facet wrapping</a></li>
</ul></li>
<li class="chapter" data-level="8.9" data-path="data-visualization.html"><a href="data-visualization.html#dv-resp"><i class="fa fa-check"></i><b>8.9</b> Responsible Plotting</a></li>
<li class="chapter" data-level="8.10" data-path="data-visualization.html"><a href="data-visualization.html#publication-ready-plots"><i class="fa fa-check"></i><b>8.10</b> Publication Ready Plots</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="engineering.html"><a href="engineering.html"><i class="fa fa-check"></i><b>9</b> EngineeRing</a>
<ul>
<li class="chapter" data-level="9.1" data-path="engineering.html"><a href="engineering.html#unit-testing"><i class="fa fa-check"></i><b>9.1</b> Unit Testing</a></li>
<li class="chapter" data-level="9.2" data-path="engineering.html"><a href="engineering.html#toolification"><i class="fa fa-check"></i><b>9.2</b> Toolification</a></li>
<li class="chapter" data-level="9.3" data-path="engineering.html"><a href="engineering.html#pipelines-workflows"><i class="fa fa-check"></i><b>9.3</b> Pipelines &amp; Workflows</a></li>
<li class="chapter" data-level="9.4" data-path="engineering.html"><a href="engineering.html#parallel-processing"><i class="fa fa-check"></i><b>9.4</b> Parallel Processing</a></li>
<li class="chapter" data-level="9.5" data-path="engineering.html"><a href="engineering.html#r-packages"><i class="fa fa-check"></i><b>9.5</b> R Packages</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="rshiny.html"><a href="rshiny.html"><i class="fa fa-check"></i><b>10</b> RShiny</a>
<ul>
<li class="chapter" data-level="10.1" data-path="rshiny.html"><a href="rshiny.html#overview"><i class="fa fa-check"></i><b>10.1</b> Overview</a></li>
<li class="chapter" data-level="10.2" data-path="rshiny.html"><a href="rshiny.html#learning-objectives"><i class="fa fa-check"></i><b>10.2</b> Learning Objectives</a></li>
<li class="chapter" data-level="10.3" data-path="rshiny.html"><a href="rshiny.html#skill-list"><i class="fa fa-check"></i><b>10.3</b> Skill List</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="communicating-with-r.html"><a href="communicating-with-r.html"><i class="fa fa-check"></i><b>11</b> Communicating with R</a>
<ul>
<li class="chapter" data-level="11.1" data-path="communicating-with-r.html"><a href="communicating-with-r.html#rmarkdown-knitr"><i class="fa fa-check"></i><b>11.1</b> RMarkdown &amp; knitr</a></li>
<li class="chapter" data-level="11.2" data-path="communicating-with-r.html"><a href="communicating-with-r.html#bookdown"><i class="fa fa-check"></i><b>11.2</b> bookdown</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="contribution-guide.html"><a href="contribution-guide.html"><i class="fa fa-check"></i><b>12</b> Contribution Guide</a>
<ul>
<li class="chapter" data-level="12.1" data-path="contribution-guide.html"><a href="contribution-guide.html#custom-blocks"><i class="fa fa-check"></i><b>12.1</b> Custom blocks</a></li>
</ul></li>
<li class="part"><span><b>I Assignments</b></span></li>
<li class="chapter" data-level="" data-path="assignment-format.html"><a href="assignment-format.html"><i class="fa fa-check"></i>Assignment Format</a></li>
<li class="chapter" data-level="" data-path="starting-an-assignment.html"><a href="starting-an-assignment.html"><i class="fa fa-check"></i>Starting an Assignment</a>
<ul>
<li><a href="starting-an-assignment.html#starting-your-own-git-repository-and-cloning-it">Starting your own <code>git</code> repository and cloning it</a>
<ul>
<li class="chapter" data-level="" data-path="starting-an-assignment.html"><a href="starting-an-assignment.html#contents"><i class="fa fa-check"></i>Contents</a></li>
<li class="chapter" data-level="" data-path="starting-an-assignment.html"><a href="starting-an-assignment.html#uhhh-various-github-classroom-details-idk.-todo-or-whatever."><i class="fa fa-check"></i>Uhhh various github classroom details, idk. TODO: or whatever.</a></li>
<li class="chapter" data-level="" data-path="starting-an-assignment.html"><a href="starting-an-assignment.html#committing-and-pushing-with-r-and-without"><i class="fa fa-check"></i>Committing and Pushing with R (and without)</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="assignment-2.html"><a href="assignment-2.html"><i class="fa fa-check"></i>Assignment 2</a>
<ul>
<li class="chapter" data-level="" data-path="assignment-2.html"><a href="assignment-2.html#problem-statement"><i class="fa fa-check"></i>Problem Statement</a></li>
<li class="chapter" data-level="" data-path="assignment-2.html"><a href="assignment-2.html#learning-objectives-1"><i class="fa fa-check"></i>Learning Objectives</a></li>
<li class="chapter" data-level="" data-path="assignment-2.html"><a href="assignment-2.html#skill-list-1"><i class="fa fa-check"></i>Skill List</a></li>
<li class="chapter" data-level="" data-path="assignment-2.html"><a href="assignment-2.html#instructions"><i class="fa fa-check"></i>Instructions</a></li>
<li class="chapter" data-level="" data-path="assignment-2.html"><a href="assignment-2.html#function-details"><i class="fa fa-check"></i>Function Details</a>
<ul>
<li class="chapter" data-level="" data-path="assignment-2.html"><a href="assignment-2.html#bioconductor"><i class="fa fa-check"></i>1. Bioconductor</a></li>
<li><a href="assignment-2.html#load_expression">2. <code>load_expression()</code></a></li>
<li><a href="assignment-2.html#filter_15">3. <code>filter_15()</code></a></li>
<li><a href="assignment-2.html#affy_to_hgnc">4. <code>affy_to_hgnc()</code></a></li>
<li><a href="assignment-2.html#reduce_data">5. <code>reduce_data()</code></a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="assignment-5---deseq2.html"><a href="assignment-5---deseq2.html"><i class="fa fa-check"></i>Assignment 5 - DESeq2</a>
<ul>
<li class="chapter" data-level="" data-path="assignment-5---deseq2.html"><a href="assignment-5---deseq2.html#problem-statement-1"><i class="fa fa-check"></i>Problem Statement</a></li>
<li class="chapter" data-level="" data-path="assignment-5---deseq2.html"><a href="assignment-5---deseq2.html#learning-objectives-2"><i class="fa fa-check"></i>Learning Objectives</a></li>
<li class="chapter" data-level="" data-path="assignment-5---deseq2.html"><a href="assignment-5---deseq2.html#skill-list-2"><i class="fa fa-check"></i>Skill List</a></li>
<li class="chapter" data-level="" data-path="assignment-5---deseq2.html"><a href="assignment-5---deseq2.html#deseq2-background"><i class="fa fa-check"></i>DESeq2 Background</a></li>
<li class="chapter" data-level="" data-path="assignment-5---deseq2.html"><a href="assignment-5---deseq2.html#generating-a-counts-matrix"><i class="fa fa-check"></i>Generating a counts matrix</a></li>
<li class="chapter" data-level="" data-path="assignment-5---deseq2.html"><a href="assignment-5---deseq2.html#prefiltering-counts-matrix"><i class="fa fa-check"></i>Prefiltering Counts matrix</a></li>
<li class="chapter" data-level="" data-path="assignment-5---deseq2.html"><a href="assignment-5---deseq2.html#median-of-ratios-normalization"><i class="fa fa-check"></i>Median-of-ratios normalization</a></li>
<li class="chapter" data-level="" data-path="assignment-5---deseq2.html"><a href="assignment-5---deseq2.html#deseq2-preparation"><i class="fa fa-check"></i>DESeq2 preparation</a></li>
<li class="chapter" data-level="" data-path="assignment-5---deseq2.html"><a href="assignment-5---deseq2.html#omeara-et-al.-transcriptional-reversion-of-cardiac-myocyte-fate-during-mammalian-cardiac-regeneration.-circ-res.-feb-2015.-pmid-25477501l"><i class="fa fa-check"></i>O’Meara et al. Transcriptional Reversion of Cardiac Myocyte Fate During Mammalian Cardiac Regeneration. Circ Res. Feb 2015. PMID: 25477501l</a></li>
<li class="chapter" data-level="" data-path="assignment-5---deseq2.html"><a href="assignment-5---deseq2.html#understanding-factor-levels"><i class="fa fa-check"></i>Understanding Factor Levels</a></li>
<li class="chapter" data-level="" data-path="assignment-5---deseq2.html"><a href="assignment-5---deseq2.html#performing-differential-expression-analysis"><i class="fa fa-check"></i>Performing Differential Expression Analysis</a></li>
<li class="chapter" data-level="" data-path="assignment-5---deseq2.html"><a href="assignment-5---deseq2.html#p-value-distribution-and-multiple-test-correction"><i class="fa fa-check"></i>P-value distribution and multiple test correction</a></li>
<li class="chapter" data-level="" data-path="assignment-5---deseq2.html"><a href="assignment-5---deseq2.html#histogram-of-log-fold-changes"><i class="fa fa-check"></i>Histogram of Log Fold Changes</a></li>
<li class="chapter" data-level="" data-path="assignment-5---deseq2.html"><a href="assignment-5---deseq2.html#make-a-volcano-plot"><i class="fa fa-check"></i>Make a volcano plot</a></li>
<li class="chapter" data-level="" data-path="assignment-5---deseq2.html"><a href="assignment-5---deseq2.html#visualize-normalized-counts-for-the-ten-differentially-expressed-genes-with-the-lowest-adjusted-p-values"><i class="fa fa-check"></i>Visualize normalized counts for the ten differentially expressed genes with the lowest adjusted p-values</a></li>
<li class="chapter" data-level="" data-path="assignment-5---deseq2.html"><a href="assignment-5---deseq2.html#time-series-with-deseq2-likelihood-ratio-test-lrt"><i class="fa fa-check"></i>Time series with DESeq2 Likelihood Ratio Test (LRT)</a></li>
<li class="chapter" data-level="" data-path="assignment-5---deseq2.html"><a href="assignment-5---deseq2.html#plotting-the-normalized-counts-over-time-for-the-top-differentially-expressed-gene-from-the-lrt-analysis"><i class="fa fa-check"></i>Plotting the normalized counts over time for the top differentially expressed gene from the LRT analysis</a></li>
<li class="chapter" data-level="" data-path="assignment-5---deseq2.html"><a href="assignment-5---deseq2.html#optional-exercises"><i class="fa fa-check"></i>Optional exercises</a></li>
<li class="chapter" data-level="" data-path="assignment-5---deseq2.html"><a href="assignment-5---deseq2.html#references"><i class="fa fa-check"></i>References</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="assignment-6.html"><a href="assignment-6.html"><i class="fa fa-check"></i>Assignment 6</a>
<ul>
<li class="chapter" data-level="" data-path="assignment-6.html"><a href="assignment-6.html#problem-statement-2"><i class="fa fa-check"></i>Problem Statement</a></li>
<li class="chapter" data-level="" data-path="assignment-6.html"><a href="assignment-6.html#learning-objectives-3"><i class="fa fa-check"></i>Learning Objectives</a></li>
<li class="chapter" data-level="" data-path="assignment-6.html"><a href="assignment-6.html#skill-list-3"><i class="fa fa-check"></i>Skill List</a></li>
<li class="chapter" data-level="" data-path="assignment-6.html"><a href="assignment-6.html#instructions-1"><i class="fa fa-check"></i>Instructions</a></li>
<li class="chapter" data-level="" data-path="assignment-6.html"><a href="assignment-6.html#function-details-1"><i class="fa fa-check"></i>Function Details</a>
<ul>
<li><a href="assignment-6.html#load_n_trim">1. <code>load_n_trim()</code></a></li>
<li><a href="assignment-6.html#run_deseq">2. <code>run_deseq()</code></a></li>
<li><a href="assignment-6.html#run_edger">3. <code>run_edger()</code></a></li>
<li><a href="assignment-6.html#run_limma-and-run_voom">4. <code>run_limma()</code> and <code>run_voom()</code></a></li>
<li><a href="assignment-6.html#combine_pval">5. <code>combine_pval()</code></a></li>
<li><a href="assignment-6.html#create_facets">6. <code>create_facets()</code></a></li>
<li><a href="assignment-6.html#theme_plot">7. <code>theme_plot()</code></a></li>
</ul></li>
</ul></li>
<li class="appendix"><span><b>Appendix</b></span></li>
<li class="chapter" data-level="A" data-path="class-outlines.html"><a href="class-outlines.html"><i class="fa fa-check"></i><b>A</b> Class Outlines</a>
<ul>
<li class="chapter" data-level="A.1" data-path="class-outlines.html"><a href="class-outlines.html#week-1"><i class="fa fa-check"></i><b>A.1</b> Week 1</a></li>
<li class="chapter" data-level="A.2" data-path="class-outlines.html"><a href="class-outlines.html#week-2"><i class="fa fa-check"></i><b>A.2</b> Week 2</a></li>
<li class="chapter" data-level="A.3" data-path="class-outlines.html"><a href="class-outlines.html#week-3"><i class="fa fa-check"></i><b>A.3</b> Week 3</a></li>
<li class="chapter" data-level="A.4" data-path="class-outlines.html"><a href="class-outlines.html#week-4"><i class="fa fa-check"></i><b>A.4</b> Week 4</a></li>
<li class="chapter" data-level="A.5" data-path="class-outlines.html"><a href="class-outlines.html#week-5"><i class="fa fa-check"></i><b>A.5</b> Week 5</a></li>
<li class="chapter" data-level="A.6" data-path="class-outlines.html"><a href="class-outlines.html#week-6"><i class="fa fa-check"></i><b>A.6</b> Week 6</a></li>
<li class="chapter" data-level="A.7" data-path="class-outlines.html"><a href="class-outlines.html#week-7"><i class="fa fa-check"></i><b>A.7</b> Week 7</a></li>
<li class="chapter" data-level="A.8" data-path="class-outlines.html"><a href="class-outlines.html#week-8"><i class="fa fa-check"></i><b>A.8</b> Week 8</a></li>
<li class="chapter" data-level="A.9" data-path="class-outlines.html"><a href="class-outlines.html#week-9"><i class="fa fa-check"></i><b>A.9</b> Week 9</a></li>
<li class="chapter" data-level="A.10" data-path="class-outlines.html"><a href="class-outlines.html#week-10"><i class="fa fa-check"></i><b>A.10</b> Week 10</a></li>
<li class="chapter" data-level="A.11" data-path="class-outlines.html"><a href="class-outlines.html#week-11"><i class="fa fa-check"></i><b>A.11</b> Week 11</a></li>
<li class="chapter" data-level="A.12" data-path="class-outlines.html"><a href="class-outlines.html#week-12"><i class="fa fa-check"></i><b>A.12</b> Week 12</a></li>
<li class="chapter" data-level="A.13" data-path="class-outlines.html"><a href="class-outlines.html#week-13---no-class-for-project-work"><i class="fa fa-check"></i><b>A.13</b> Week 13 - no class for project work</a></li>
<li class="chapter" data-level="A.14" data-path="class-outlines.html"><a href="class-outlines.html#week-14---no-class-for-project-work"><i class="fa fa-check"></i><b>A.14</b> Week 14 - no class for project work</a></li>
</ul></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">BF591 - R for Biological Sciences</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="biology-bioinformatics" class="section level1" number="7">
<h1><span class="header-section-number">7</span> Biology &amp; Bioinformatics</h1>
<div id="bio-r" class="section level2" number="7.1">
<h2><span class="header-section-number">7.1</span> R in Biology</h2>
<p>R became popular in biological data analysis in the early to mid 2000s, when
<a href="https://www.genome.gov/genetics-glossary/Microarray-Technology">microarray
technology</a> came
into widespread use enabling researchers to look for statistical differences in
gene expression for thousands of genes across large numbers of samples. As a
result of this popularity, a community of biological researchers and data
analysts created a collection of software packages called
<a href="https://www.bioconductor.org/">Bioconductor</a>, which made a vast array of
cutting edge statistical and bioinformatic methodologies widely available.</p>
<p>Due to its ease of use, widespread community support, rich ecosystem of
biological data analysis packages, and to it being free, R remains one of the
primary languages of biological data analysis. The language’s early focus on
statistical analysis, and later transition to data science in general, makes it
very useful and accessible to perform the kinds of analyses the new data science
of biology required. It is also a bridge between biologists without a
computational background and statisticians and bioinformaticians, who invent and
implement methods into R packages that are easily accessible by all. The
language and the package ecosystems and communities it supports continue to be a
major source of biological discovery and innovation.</p>
<p>As a data science, biology benefits from the main strengths of R and <a href="#dw-tidyverse">the
tidyverse</a> when combined with the powerful analytical techniques
available in Bioconductor packages, namely to manipulate, visualize, analyze,
and communicate biological data.</p>
</div>
<div id="csv-files" class="section level2" number="7.2">
<h2><span class="header-section-number">7.2</span> CSV Files</h2>
<p>The most common, convenient, and flexible data file format in biology and
bioinformatics is the <em>character-delimited</em> or <em>character-separated text file</em>.
These files are plain text files (i.e. not the native file format of any
specific program, like Excel) that generally contain rectangles of data. When
formatted correctly, you can think of these files as containing a grid or matrix
of data values with some number of columns, each of which has the same number of
values. Each line of these files has some number of data values separated by a
consistent character, most commonly the comma which are called <a href="https://en.wikipedia.org/wiki/Comma-separated_values">comma-separated
value, or “CSV,” files</a>
and filenames typically end with the extension <code>.csv</code>. Note that
other characters, especially the <Tab> character, may be used to create valid
files in this format, and all the same general principles apply. This is an
example of a simple CSV file:</p>
<pre><code>id,somevalue,category,genes
1,2.123,A,APOE
4,5.123,B,&quot;HOXA1,HOXB1&quot;
7,8.123,,SNCA</code></pre>
<p>Some properties and principles of CSV files:</p>
<ul>
<li>The first line often but not always contains the column names of each column</li>
<li>Each value is delimited by the same character, in this case <code>,</code></li>
<li>Values can be any value, including numbers and characters</li>
<li>When a value contains the delimiting character (e.g. HOXA1,HOXB1 contains a
<code>,</code>), the value is wrapped in double quotes</li>
<li>Values can be missing, indicated by sequential delimiters (i.e. <code>,,</code> or one
<code>,</code> at the end of the line, if the last column value is missing)</li>
<li>There is no delimiter at the end of the lines</li>
<li>To be well-formatted <em>every line must have the same number of delimited
values</em></li>
</ul>
<p>These same properties and principles apply to all character-separated files,
regardless of the specific delimiter used. If a file follows these principles,
they can be loaded very easily into R or any other data analysis setting.</p>
</div>
<div id="bio-data-types" class="section level2" number="7.3">
<h2><span class="header-section-number">7.3</span> Types of Biological Data</h2>
<p>In very general terms, there are five types of data used in biological data
analysis: raw/primary data, processed data, analysis results, metadata, and
annotation data.</p>
<p><strong>Raw/primary data.</strong> These data are the primary observations made by whatever
instruments/techniques we are using for our experiment. These may include
high-throughput sequencing data, mass/charge ratio data from mass spectrometry,
16S rRNA sequencing data from metagenomic studies, SNPs from genotyping assays,
etc. These data can be very large and are often not efficiently processed using
R. Instead, specialized tools built outside of R are used to first process the
primary data into a form that is amenable to analysis.</p>
<p><strong>Processed data.</strong> Processed data is the result of any analysis or
transformation of primary data into an intermediate, more interpretable form.
For example, when performing gene expression analysis with RNASeq, short reads
that form the primary data are typically aligned against a genome and then
counted against annotated genes, resulting in a count for each gene in the
genome. Typically, these counts are then combined for many samples into a single
matrix and subject to downstream analyses like differential expression.</p>
<p><strong>Analysis results.</strong> Analysis results aren’t data <em>per se</em>, but are the results
of analysis of primary data or processed results. For example, in gene
expression studies, a first analysis is often differential expression, where
each gene in the genome is tested for expression associated with some outcome
of interest across many samples. Each gene then has a number of statistics and
related values, like log2 fold change, nominal and adjusted p-value, etc. These
forms of data are usually what we use to form interpretations of our datasets
and therefore we must manipulate them in much the same way as any other dataset.</p>
<p><strong>Metadata.</strong> In most experimental settings, multiple samples have been
processed and had the same primary data type collected. These samples also have
information associated with them, which is here termed <em>metadata</em>, or “data
about data.” In our gene expression of post mortem brain experiments mentioned
above, the information about the individuals themselves, including age at death,
whether the person had a disease, the measurements of tissue quality, etc. is
the metadata. The primary and processed data and metadata are usually stored in
different files, where the metadata (or <em>sample information</em> or <em>sample data</em>,
etc) will have one column indicating the unique identifier (ID) of each sample.
The processed data will typically have columns named for each of the sample IDs.</p>
<p><strong>Annotation data.</strong> A tremendous amount of knowledge has been generated about
biological entities, e.g. genes, especially since the publication of the human
genome. Annotation data is publicly available information about the features we
measure in our experiments. This may come in the form of the coordinates in a
genome where genes exist, any known functions of those genes, the domains found
in proteins and their relative sequence, gene identifier cross references across
different gene naming systems (e.g. symbol vs Ensembl ID), single nucleotide
polymorphism genomic locations and associations with traits or diseases, etc.
This is information that we use to aid in interpretation of our experimental
data but generally do not generate ourselves. Annotation data comes in many
forms, some of which are in CSV format.</p>
<p>The figure below contains a simple diagram of how these different types of data
work together for a typical biological data analysis.</p>
<div class="figure">
<img src="information_flow.png" alt="" />
<p class="caption">Information flow in biological data analysis</p>
</div>
</div>
<div id="common-biological-data-matrices" class="section level2" number="7.4">
<h2><span class="header-section-number">7.4</span> Common Biological Data Matrices</h2>
<p>Typically the first data set you will work with in R is <em>processed data</em> as
described in the previous section. This data has been transformed from primary
data in some way such that it (usually) forms a numeric matrix with features
as rows and samples as columns. The first column of these files usually contains
a feature identifier, e.g. gene identifier, genomic locus, probe set ID, etc and
the remaining columns have numerical values, one per sample. The first row is
usually column names for all the columns in the file. Below is an example of one
of these files from a microarray gene expression dataset loaded into R:</p>
<div class="sourceCode" id="cb147"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb147-1"><a href="biology-bioinformatics.html#cb147-1" aria-hidden="true" tabindex="-1"></a>intensities <span class="ot">&lt;-</span> <span class="fu">read_csv</span>(<span class="st">&quot;example_intensity_data.csv&quot;</span>)</span>
<span id="cb147-2"><a href="biology-bioinformatics.html#cb147-2" aria-hidden="true" tabindex="-1"></a>intensities</span>
<span id="cb147-3"><a href="biology-bioinformatics.html#cb147-3" aria-hidden="true" tabindex="-1"></a><span class="co"># A tibble: 54,675 x 36</span></span>
<span id="cb147-4"><a href="biology-bioinformatics.html#cb147-4" aria-hidden="true" tabindex="-1"></a>   probe     GSM972389 GSM972390 GSM972396 GSM972401 GSM972409 GSM972412 GSM972413 GSM972422 GSM972429</span>
<span id="cb147-5"><a href="biology-bioinformatics.html#cb147-5" aria-hidden="true" tabindex="-1"></a>   <span class="sc">&lt;</span>chr<span class="sc">&gt;</span>         <span class="er">&lt;</span>dbl<span class="sc">&gt;</span>     <span class="er">&lt;</span>dbl<span class="sc">&gt;</span>     <span class="er">&lt;</span>dbl<span class="sc">&gt;</span>     <span class="er">&lt;</span>dbl<span class="sc">&gt;</span>     <span class="er">&lt;</span>dbl<span class="sc">&gt;</span>     <span class="er">&lt;</span>dbl<span class="sc">&gt;</span>     <span class="er">&lt;</span>dbl<span class="sc">&gt;</span>     <span class="er">&lt;</span>dbl<span class="sc">&gt;</span>     <span class="er">&lt;</span>dbl<span class="sc">&gt;</span></span>
<span id="cb147-6"><a href="biology-bioinformatics.html#cb147-6" aria-hidden="true" tabindex="-1"></a> <span class="dv">1</span> 1007_s_at      <span class="fl">9.54</span>     <span class="fl">10.2</span>       <span class="fl">9.72</span>      <span class="fl">9.68</span>      <span class="fl">9.35</span>      <span class="fl">9.89</span>      <span class="fl">9.70</span>      <span class="fl">9.67</span>      <span class="fl">9.87</span></span>
<span id="cb147-7"><a href="biology-bioinformatics.html#cb147-7" aria-hidden="true" tabindex="-1"></a> <span class="dv">2</span> 1053_at        <span class="fl">7.62</span>      <span class="fl">7.92</span>      <span class="fl">7.17</span>      <span class="fl">7.24</span>      <span class="fl">8.20</span>      <span class="fl">6.87</span>      <span class="fl">6.62</span>      <span class="fl">7.23</span>      <span class="fl">7.45</span></span>
<span id="cb147-8"><a href="biology-bioinformatics.html#cb147-8" aria-hidden="true" tabindex="-1"></a> <span class="dv">3</span> 117_at         <span class="fl">5.50</span>      <span class="fl">5.56</span>      <span class="fl">5.06</span>      <span class="fl">7.44</span>      <span class="fl">5.19</span>      <span class="fl">5.72</span>      <span class="fl">5.87</span>      <span class="fl">6.15</span>      <span class="fl">5.46</span></span>
<span id="cb147-9"><a href="biology-bioinformatics.html#cb147-9" aria-hidden="true" tabindex="-1"></a> <span class="dv">4</span> 121_at         <span class="fl">7.27</span>      <span class="fl">7.96</span>      <span class="fl">7.42</span>      <span class="fl">7.34</span>      <span class="fl">7.49</span>      <span class="fl">7.76</span>      <span class="fl">7.44</span>      <span class="fl">7.66</span>      <span class="fl">8.02</span></span>
<span id="cb147-10"><a href="biology-bioinformatics.html#cb147-10" aria-hidden="true" tabindex="-1"></a> <span class="dv">5</span> 1255_g_at      <span class="fl">2.79</span>      <span class="fl">3.10</span>      <span class="fl">2.78</span>      <span class="fl">2.91</span>      <span class="fl">3.02</span>      <span class="fl">2.73</span>      <span class="fl">2.78</span>      <span class="fl">3.56</span>      <span class="fl">2.83</span></span>
<span id="cb147-11"><a href="biology-bioinformatics.html#cb147-11" aria-hidden="true" tabindex="-1"></a> <span class="dv">6</span> 1294_at        <span class="fl">7.51</span>      <span class="fl">7.28</span>      <span class="fl">7.00</span>      <span class="fl">7.18</span>      <span class="fl">7.38</span>      <span class="fl">6.98</span>      <span class="fl">6.90</span>      <span class="fl">7.54</span>      <span class="fl">7.66</span></span>
<span id="cb147-12"><a href="biology-bioinformatics.html#cb147-12" aria-hidden="true" tabindex="-1"></a> <span class="dv">7</span> 1316_at        <span class="fl">3.89</span>      <span class="fl">4.36</span>      <span class="fl">4.24</span>      <span class="fl">3.94</span>      <span class="fl">4.20</span>      <span class="fl">4.34</span>      <span class="fl">4.06</span>      <span class="fl">4.24</span>      <span class="fl">4.11</span></span>
<span id="cb147-13"><a href="biology-bioinformatics.html#cb147-13" aria-hidden="true" tabindex="-1"></a> <span class="dv">8</span> 1320_at        <span class="fl">4.65</span>      <span class="fl">4.91</span>      <span class="fl">4.70</span>      <span class="fl">4.78</span>      <span class="fl">5.06</span>      <span class="fl">4.71</span>      <span class="fl">4.55</span>      <span class="fl">4.58</span>      <span class="fl">5.10</span></span>
<span id="cb147-14"><a href="biology-bioinformatics.html#cb147-14" aria-hidden="true" tabindex="-1"></a> <span class="dv">9</span> 1405_i_at      <span class="fl">8.03</span>      <span class="fl">7.47</span>      <span class="fl">5.42</span>      <span class="fl">7.21</span>      <span class="fl">9.48</span>      <span class="fl">6.79</span>      <span class="fl">6.57</span>      <span class="fl">8.50</span>      <span class="fl">6.36</span></span>
<span id="cb147-15"><a href="biology-bioinformatics.html#cb147-15" aria-hidden="true" tabindex="-1"></a><span class="dv">10</span> 1431_at        <span class="fl">3.09</span>      <span class="fl">3.78</span>      <span class="fl">3.33</span>      <span class="fl">3.12</span>      <span class="fl">3.21</span>      <span class="fl">3.27</span>      <span class="fl">3.37</span>      <span class="fl">3.84</span>      <span class="fl">3.32</span></span>
<span id="cb147-16"><a href="biology-bioinformatics.html#cb147-16" aria-hidden="true" tabindex="-1"></a><span class="co"># ... with 54,665 more rows, and 26 more variables: GSM972433 &lt;dbl&gt;, GSM972438 &lt;dbl&gt;, GSM972440 &lt;dbl&gt;,</span></span>
<span id="cb147-17"><a href="biology-bioinformatics.html#cb147-17" aria-hidden="true" tabindex="-1"></a><span class="co">#   GSM972443 &lt;dbl&gt;, GSM972444 &lt;dbl&gt;, GSM972446 &lt;dbl&gt;, GSM972453 &lt;dbl&gt;, GSM972457 &lt;dbl&gt;,</span></span>
<span id="cb147-18"><a href="biology-bioinformatics.html#cb147-18" aria-hidden="true" tabindex="-1"></a><span class="co">#   GSM972459 &lt;dbl&gt;, GSM972463 &lt;dbl&gt;, GSM972467 &lt;dbl&gt;, GSM972472 &lt;dbl&gt;, GSM972473 &lt;dbl&gt;,</span></span>
<span id="cb147-19"><a href="biology-bioinformatics.html#cb147-19" aria-hidden="true" tabindex="-1"></a><span class="co">#   GSM972475 &lt;dbl&gt;, GSM972476 &lt;dbl&gt;, GSM972477 &lt;dbl&gt;, GSM972479 &lt;dbl&gt;, GSM972480 &lt;dbl&gt;,</span></span>
<span id="cb147-20"><a href="biology-bioinformatics.html#cb147-20" aria-hidden="true" tabindex="-1"></a><span class="co">#   GSM972487 &lt;dbl&gt;, GSM972488 &lt;dbl&gt;, GSM972489 &lt;dbl&gt;, GSM972496 &lt;dbl&gt;, GSM972502 &lt;dbl&gt;,</span></span>
<span id="cb147-21"><a href="biology-bioinformatics.html#cb147-21" aria-hidden="true" tabindex="-1"></a><span class="co">#   GSM972510 &lt;dbl&gt;, GSM972512 &lt;dbl&gt;, GSM972521 &lt;dbl&gt;</span></span></code></pre></div>
<p>The file has 54,676 rows, consisting of one header row which R loads in as the
column names, and the remaining are probe sets, one per row. There are 36
columns, where the first contains the probe set ID (e.g. <code>1007_s_at</code>) and the
remaining 35 columns correspond to samples.</p>
<div id="biological-data-is-not-tidy" class="section level3" number="7.4.1">
<h3><span class="header-section-number">7.4.1</span> Biological data is NOT Tidy!</h3>
<p>As mentioned in the <a href="#tidy%20data">tidy data</a> section, the tidyverse packages
assume data to be in so-called “tidy format,” with variables as columns and
observations as rows. Unfortunately, certain forms of biological data are
typically available in the opposite orientation - variables are in rows and
observations are in columns. This is primarily true in feature data matrices,
e.g. gene expression counts matrices, where the number of variables (e.g. genes)
is much larger than the number of samples, which tend to be small very small
compared with the number of features. This format is convenient for humans to
interact with using, e.g. spreadsheet programs like Microsoft Excel, but can
unfortunately make performing certain operations on them challenging in
tidyverse.</p>
<p>For example, consider the microarray expression dataset in the previous section.
Each of the 54,676 rows is a probeset, and each of the 35 numeric columns is a
sample. This is a very large number of probesets to consider, especially if we
plan to conduct a statistical test on each, which would impose a substantial
<a href="data-science.html#multiple-hypothesis-testing">multiple hypothesis testing</a> burden on our
results. We may therefore wish to eliminate probesets that have very low
variance from the overall dataset, since these probesets are not likely to have
a detectable statistical difference in our tests. However, computing the
variance for each probeset is a computation <em>across all columns</em>, not on columns
themselves, and this is not what tidyverse is designed to do well. Said
differently, R and tidyverse do not operate by default on the rows of a data
frame, tibble, or matrix.</p>
<p>Both base R and tidyverse are optimized to perform computations on columns, not
rows. The reasons for this are buried in the details of how the R program itself
was written to organize data internally and are beyond the scope of this book.
The consequence of this design choice is that, while we can perform operations
on the rows rather than the columns of a data structure, our code may perform
very poorly (i.e. take a very long time to run).</p>
<p>When working with these datasets, we have a couple options to deal with this
issue:</p>
<ul>
<li><p><strong>Pivot into long format.</strong> As described in the <a href="data-wrangling.html#rearranging-data">Rearranging Data</a> section, we
can rearrange our tibble to be more amenable to certain computations. In our
earlier example, we wish to group all of our measurements by probeset and
compute the variance of each, then possibly filter out probesets based on low
variance. We can therefore combine <code>pivot_longer()</code>, <code>group_by()</code>,
<code>summarize()</code>, and finally <code>left_join()</code> to perform this operation. Exactly
how to do this is left as an exercise in [Assignment 1].</p></li>
<li><p><strong>Compute row-wise statistics using <code>apply()</code>.</strong> As described in <a href="prog-basics.html#iteration">Iteration</a>,
R is a functional programming language and implements iteration in a
functional style using the <code>apply()</code> function. The <a href="prog-basics.html#apply-and-friends"><code>apply()   function</code></a> accepts a <code>MARGIN</code> argument of 1 or 2 if the
provided function is to be applied to the columns or rows, respectively. This
method can be used to compute a summary statistic on each row of a tibble and
the result saved back into the tibble using the column set operator:</p>
<div class="sourceCode" id="cb148"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb148-1"><a href="biology-bioinformatics.html#cb148-1" aria-hidden="true" tabindex="-1"></a>intensity_variance <span class="ot">&lt;-</span> <span class="fu">apply</span>(intensities, <span class="dv">2</span>, var)</span>
<span id="cb148-2"><a href="biology-bioinformatics.html#cb148-2" aria-hidden="true" tabindex="-1"></a>intensities<span class="sc">$</span>variance <span class="ot">&lt;-</span> intensity_variance</span></code></pre></div></li>
</ul>
</div>
</div>
<div id="bio-bioconductor" class="section level2" number="7.5">
<h2><span class="header-section-number">7.5</span> Bioconductor</h2>
<p><a href="https://www.bioconductor.org/">Bioconductor</a> is an organized collection of
strictly biological analysis methods packages for R. These packages are hosted
and maintained outside of <a href="https://cran.r-project.org/">CRAN</a> because the
maintainers enforce <a href="https://bioconductor.org/developers/package-guidelines/">a more rigorous set of coding quality, testing, and
documentation
standards</a> than is
required by normal R packages. These stricter requirements arose from a
recognition that software packages are generally only as usable as their
documentation allows them to be, and also that many if not most of the users of
these packages are not statisticians or experienced computer programmers.
Rather, they are people like us: biological analysis practitioners who may or
may not have substantial coding experience but must analyze our data
nonetheless. The excellent documentation and community support of the
bioconductor ecosystem is a primary reason why R is such a popular language in
biological analysis.</p>
<p>Bioconductor is divided into roughly two sets of packages: <em>core maintainer</em>
packages and user contributed packages. The core maintainer packages are among
the most critical, because they define a set of common objects and classes (e.g.
the <a href="https://www.bioconductor.org/packages/devel/bioc/vignettes/Biobase/inst/doc/ExpressionSetIntroduction.pdf">ExpressionSet class</a>
in the <a href="https://bioconductor.org/packages/release/bioc/html/Biobase.html">Biobase package</a>)
that all Bioconductor packages know how to work with. This common base provides
consistency among all Bioconductor packages thereby streamlining the learning
process. User contributed and maintained packages provide specialized
functionality for specific types of analysis.</p>
<p>Bioconductor itself must be installed prior to installing other Bioconductor
packages. To [install bioconductor], you can run the following:</p>
<div class="sourceCode" id="cb149"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb149-1"><a href="biology-bioinformatics.html#cb149-1" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">require</span>(<span class="st">&quot;BiocManager&quot;</span>, <span class="at">quietly =</span> <span class="cn">TRUE</span>))</span>
<span id="cb149-2"><a href="biology-bioinformatics.html#cb149-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">install.packages</span>(<span class="st">&quot;BiocManager&quot;</span>)</span>
<span id="cb149-3"><a href="biology-bioinformatics.html#cb149-3" aria-hidden="true" tabindex="-1"></a>BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="at">version =</span> <span class="st">&quot;3.14&quot;</span>)</span></code></pre></div>
<div class="box warning">
<p>Bioconductor undergoes regular releases indicated by a version, e.g. <code>version = "3.14"</code> at the time of writing. Every bioconductor package also has a version,
and each of those versions may or may not be compatible with a specific version
of Bioconductor. To make matters worse, Bioconductor versions are only
compatible with certain versions of R. Bioconductor version 3.14 requires R
version 4.1.0, and will not work with older R versions. These versions can cause
major version compatibility issues when you are forced to use older versions of
R, as may sometimes be the case on managed compute clusters. There is not a good
general solution for ensuring all your packages work together, but the general
best rule of thumb is to use the most current version of R and all packages at
the time when you write your code.</p>
</div>
<p>The <code>BiocManager</code> package is the only Bioconductor package installable using
<code>install.packages()</code>. After installing the BiocManager package, you may then
install bioconductor packages:</p>
<div class="sourceCode" id="cb150"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb150-1"><a href="biology-bioinformatics.html#cb150-1" aria-hidden="true" tabindex="-1"></a><span class="co"># installs the affy bioconductor package for microarray analysis</span></span>
<span id="cb150-2"><a href="biology-bioinformatics.html#cb150-2" aria-hidden="true" tabindex="-1"></a>BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="st">&quot;affy&quot;</span>)</span></code></pre></div>
<p>One key aspect of Bioconductor packages is consistent and helpful documentation;
every package page on the Bioconductor.org site lists a block of code that will
install the package, e.g. for the <a href="https://www.bioconductor.org/packages/release/bioc/html/affy.html">affy
package</a>. In
addition, Biconductor provides three types of documentation:</p>
<ul>
<li><a href="https://bioconductor.org/packages/release/workflows/">Workflow tutorials</a> on
how to perform specific analysis use cases</li>
<li><a href="https://bioconductor.org/help/package-vignettes/">Package vignettes</a> for
every package, which provide a worked example of how to use the package that
can be adapted by the user to their specific case</li>
<li>Detailed, consistently formatted reference documentation that gives precise
information on functionality and use of each package</li>
</ul>
<p>The thorough and useful documentation of packages in Bioconductor is one of the
reasons why the package ecosystem, and R, is so widely used in biology and
bioinformatics.</p>
<p>The base Bioconductor packages define convenient data structures for storing and
analyzing many types of data. Recall earlier in the <a href="biology-bioinformatics.html#bio-data-types">Types of Biological Data</a>
section, we described five types of biological data: primary, processed,
analysis, metadata, and annotation data. Bioconductor provides a convenient
class for storing many of these different data types together in one place,
specifically the SummarizedExperiment class in the <a href="https://bioconductor.org/packages/release/bioc/html/SummarizedExperiment.html">package of the same
name</a>
package. An illustration of what a SummarizedExperiment object stores is below,
from the Bioconductor maintainer’s paper in Nature:</p>
<div class="figure">
<img src="https://media.springernature.com/full/springer-static/image/art%3A10.1038%2Fnmeth.3252/MediaObjects/41592_2015_Article_BFnmeth3252_Fig2_HTML.jpg?as=webp" alt="" />
<p class="caption">SummarizedExperiment Schematic - Huber, et al. 2015. “Orchestrating High-Throughput Genomic Analysis with Bioconductor.” Nature Methods 12 (2): 115–21.</p>
</div>
<p>As you can see from the figure, this class stores processed data (<code>assays</code>),
metadata (<code>colData</code> and <code>exptData</code>), and annotation data (<code>rowData</code>). The
SummarizedExperiment class is used ubiquitously throughout the Bioconductor
package ecosystem, as are other core classes some of which we will cover later
in this chapter.</p>
</div>
<div id="gene-identifiers" class="section level2" number="7.6">
<h2><span class="header-section-number">7.6</span> Gene Identifiers</h2>
<div id="gene-identifier-systems" class="section level3" number="7.6.1">
<h3><span class="header-section-number">7.6.1</span> Gene Identifier Systems</h3>
<p>Since the first genetic sequence for a gene was determined in 1965 - <a href="https://pubmed.ncbi.nlm.nih.gov/14263761/">an alanine
tRNA in yeast</a><span class="citation">(<a href="#ref-Holley1965-tu" role="doc-biblioref">Holley et al. 1965</a>)</span> -
scientists have been trying to give them names. In 1979, <a href="https://pubmed.ncbi.nlm.nih.gov/295268/">formal guidelines for
the nomenclature used for human genes was
proposed</a><span class="citation">(<a href="#ref-Shows1979-iw" role="doc-biblioref">Shows et al. 1979</a>)</span> along with the
founding of the Human Gene Nomenclature Committee <span class="citation">(<a href="#ref-Bruford2020-zs" role="doc-biblioref">Bruford et al. 2020</a>)</span>, so that
researchers would have a common vocabulary to refer to genes in a consistent
way. In 1989, this committee was placed under the auspices of the Human Genome
Organisation (HUGO), becoming the HUGO Gene Nomenclature Committee (HGNC). The
HGNC has been the official body providing guidelines and gene naming authority
since, and as such HGNC gene names are often called <em>gene symbols</em>.</p>
<div class="box note">
<p>As per <span class="citation"><a href="#ref-Bruford2020-zs" role="doc-biblioref">Bruford et al.</a> (<a href="#ref-Bruford2020-zs" role="doc-biblioref">2020</a>)</span>, the official human gene symbol guidelines are as
follows:</p>
<ol style="list-style-type: decimal">
<li>Each gene is assigned a unique symbol, HGNC ID and descriptive name.</li>
<li>Symbols contain only uppercase Latin letters and Arabic numerals.</li>
<li>Symbols should not be the same as commonly used abbreviations</li>
<li>Nomenclature should not contain reference to any species or “G” for gene.</li>
<li>Nomenclature should not be offensive or pejorative.</li>
</ol>
</div>
<p>Gene symbols are the most human-readable system for naming genes. The gene
symbols <em>BRCA1</em>, <em>APOE</em>, and <em>ACE2</em> may be familiar to you as they are involved
in common diseases, namely breast cancer, Alzheimer’s Disease risk, and
SARS-COV-2, respectively. Typically, the gene symbol is an acronym that roughly
represents a label of what the gene is or does (or was originally found to be or
do, as many genes are subsequently discovered to be involved in entirely
separate processes as well), e.g. <em>APOE</em> represents the gene <a href="https://www.ensembl.org/Homo_sapiens/Gene/Summary?db=core;g=ENSG00000130203;r=19:44905791-44909393">apolipoprotein
E</a>.
This convention is convenient for humans when reading and identifying genes.
However, standardized though the symbol conventions may be, they are not always
easy for computers to work with, and ambiguities can cause mapping problems.</p>
<p>Other gene identifier systems developed along with other gene information
databases. In 2000, the <a href="https://www.ensembl.org/">Ensembl</a> genome browser was
launched by the <a href="https://wellcome.org/">Wellcome Trust</a>, a charitable foundation
based in the United Kingdom, with the goal of providing automated annotation of
the human genome. The Ensembl Project, which supports the genome browser,
recognized even before the publication of the human genome that manual
annotation and curation of genes is a slow and labor intensive process that
would not provide researchers around the world timely access to information. The
project therefore created a set of automated tools and pipelines to collect,
process, and publish rapid and consistent annotation of genes in the human
genome. Since its initial release, Ensembl now supports over 50,000 different
genomes.</p>
<p>Ensembl assigns a automatic, systematic ID called the Ensembl Gene ID to every
gene in its database. Human Ensembl Gene IDs take the form <code>ENSG</code> plus an 11
digit number, optionally followed by a period delimited version number. For
example, at the time of writing the BRCA1 gene has an Ensembl Gene ID of
<code>ENSG00000012048.23</code>. The stable ID portion (i.e. <code>ENSGNNNNNNNNNNN</code>) will remain
associated with the gene forever (unless the gene annotation <a href="http://useast.ensembl.org/Help/View?id=181">changes
“dramatically” in which case it is
retired</a>). The <code>.23</code> is the version
number of the gene annotation, meaning this gene has been updated 22 times (plus
its initial version) since addition to the database. The additional version
information is very important for reproducibility of biological analysis, since
conclusions drawn by results of these analyses are usually based on the most
current information about a gene which are continually updated over time.</p>
<div class="box note">
<p>Ensembl maintains annotations for many different organisms, and the gene
identifiers for each genome contain codes that indicate which organism the gene
is for. Here are the codes for genes in several different species:</p>
<table style="width:92%;">
<colgroup>
<col width="23%" />
<col width="36%" />
<col width="31%" />
</colgroup>
<thead>
<tr class="header">
<th>Gene ID Prefix</th>
<th>Organism</th>
<th>Example</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><code>ENSG</code></td>
<td>Homo sapiens</td>
<td><a href="https://useast.ensembl.org/Homo_sapiens/Gene/Summary?db=core;g=ENSG00000139618;r=13:32315086-32400268">ENSG00000139618</a></td>
</tr>
<tr class="even">
<td><code>ENSMUSG</code></td>
<td>Mus musculus (mouse)</td>
<td><a href="https://useast.ensembl.org/Mus_musculus/Gene/Summary?db=core;g=ENSMUSG00000017167;r=11:101061349-101081550">ENSMUSG00000017167</a></td>
</tr>
<tr class="odd">
<td><code>ENSDARG</code></td>
<td>Danio rerio (zebrafish)</td>
<td><a href="https://useast.ensembl.org/Danio_rerio/Gene/Summary?db=core;g=ENSDARG00000024771;r=18:5213338-5227420;t=ENSDART00000033574">ENSDARG00000024771</a></td>
</tr>
<tr class="even">
<td><code>ENSMMUG</code></td>
<td>Macaca mulata (Macaque)</td>
<td><a href="https://useast.ensembl.org/Macaca_mulatta/Gene/Summary?db=core;g=ENSMMUG00000020312;r=5:65376507-65505566">ENSMMUG00000020312</a></td>
</tr>
<tr class="odd">
<td><code>ENSVPAG</code></td>
<td>Vicugna pacos (Alpaca)</td>
<td><a href="https://useast.ensembl.org/Vicugna_pacos/Gene/Summary?db=core;g=ENSVPAG00000001990;r=GeneScaffold_2392:178775-179671;t=ENSVPAT00000001989">ENSVPAG00000001990</a></td>
</tr>
</tbody>
</table>
</div>
<p>Ensembl Gene IDs and gene symbols are the most commonly used gene identifiers.
The <a href="https://www.gencodegenes.org/">GENCODE project</a> which provides consistent
and stable genome annotation releases for human and mouse genomes, uses these
two types of identifiers with its official annotations.</p>
<div class="box note">
<p>Ensembl provides stable identifiers for transcripts as well as genes. Transcript
identifiers correspond to distinct patterns of exons/introns that have been
identified for a gene, and each gene has one or more distinct transcripts.
Ensembl Transcript IDs have the form <code>ENSTNNNNNNNNNNN.NN</code> similar to Gene IDs.</p>
</div>
<div class="box note">
<p>Ensembl is not the only gene identifier system besides gene symbol. Several
other databases have devised and maintain their own identifiers, most notably
[Entrez Gene IDs](<a href="">UCSC Gene IDs</a> used by the <a href="https://www.ncbi.nlm.nih.gov/gene">NCBI Gene
database</a> which assigns a unique integer to
each gene in each organism, and the <a href="https://www.omim.org/">Online Mendelian Inheritance in Man
(OMIM)</a> database, which has identifiers that look like
<code>OMIM:NNNNN</code>, where each OMIM ID refers to a unique gene or human phenotype.
However, the primary identifiers used in modern human biological research remain
Ensembl IDs and official HGNC gene symbols.</p>
</div>
</div>
<div id="mapping-between-identifier-systems" class="section level3" number="7.6.2">
<h3><span class="header-section-number">7.6.2</span> Mapping Between Identifier Systems</h3>
<p>A very common operation in biological analysis is to map between gene identifier
systems, e.g. you are given Ensembl Gene ID and want to map to the more
human-recognizable gene symbols. Ensembl provides a service called
<a href="https://www.ensembl.org/biomart/martview">BioMart</a> that allows you to download
annotation information for genes, transcripts, and other information maintained
in their databases. It also provides limited access to external sources of
information on their genes, including cross references to HGNC gene symbols and
some other gene identifier systems. The Ensembl website has <a href="https://useast.ensembl.org/info/data/biomart/index.html">helpful
documentation</a> on how
to use BioMart to download annotation data using the web interface.</p>
<p>In addition to downloading annotations as a CSV file from the web interface,
Ensembl also provides <a href="https://useast.ensembl.org/info/data/biomart/biomart_r_package.html">the biomaRt Bioconductor
package</a> to
allow programmatic access to the same information directly from R. There is <a href="https://useast.ensembl.org/info/genome/index.html">much
more information in the Ensembl
databases</a> than gene
annotation data that can be accessed via BioMart, but we will provide a brief
example of how to extract gene information here:</p>
<div class="sourceCode" id="cb151"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb151-1"><a href="biology-bioinformatics.html#cb151-1" aria-hidden="true" tabindex="-1"></a><span class="co"># this assumes biomaRt is already installed through bioconductor</span></span>
<span id="cb151-2"><a href="biology-bioinformatics.html#cb151-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(biomaRt)</span>
<span id="cb151-3"><a href="biology-bioinformatics.html#cb151-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb151-4"><a href="biology-bioinformatics.html#cb151-4" aria-hidden="true" tabindex="-1"></a><span class="co"># the human biomaRt database is named &quot;hsapiens_gene_ensembl&quot;</span></span>
<span id="cb151-5"><a href="biology-bioinformatics.html#cb151-5" aria-hidden="true" tabindex="-1"></a>ensembl <span class="ot">&lt;-</span> <span class="fu">useEnsembl</span>(<span class="at">biomart=</span><span class="st">&quot;ensembl&quot;</span>, <span class="at">dataset=</span><span class="st">&quot;hsapiens_gene_ensembl&quot;</span>)</span>
<span id="cb151-6"><a href="biology-bioinformatics.html#cb151-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb151-7"><a href="biology-bioinformatics.html#cb151-7" aria-hidden="true" tabindex="-1"></a><span class="co"># listAttributes() returns a data frame, turn into a tibble to help with formatting</span></span>
<span id="cb151-8"><a href="biology-bioinformatics.html#cb151-8" aria-hidden="true" tabindex="-1"></a><span class="fu">as_tibble</span>(<span class="fu">listAttributes</span>(ensembl))</span>
<span id="cb151-9"><a href="biology-bioinformatics.html#cb151-9" aria-hidden="true" tabindex="-1"></a><span class="co"># A tibble: 3,143 x 3</span></span>
<span id="cb151-10"><a href="biology-bioinformatics.html#cb151-10" aria-hidden="true" tabindex="-1"></a>   name                          description                  page        </span>
<span id="cb151-11"><a href="biology-bioinformatics.html#cb151-11" aria-hidden="true" tabindex="-1"></a>   <span class="sc">&lt;</span>chr<span class="sc">&gt;</span>                         <span class="er">&lt;</span>chr<span class="sc">&gt;</span>                        <span class="er">&lt;</span>chr<span class="sc">&gt;</span>       </span>
<span id="cb151-12"><a href="biology-bioinformatics.html#cb151-12" aria-hidden="true" tabindex="-1"></a> <span class="dv">1</span> ensembl_gene_id               Gene stable ID               feature_page</span>
<span id="cb151-13"><a href="biology-bioinformatics.html#cb151-13" aria-hidden="true" tabindex="-1"></a> <span class="dv">2</span> ensembl_gene_id_version       Gene stable ID version       feature_page</span>
<span id="cb151-14"><a href="biology-bioinformatics.html#cb151-14" aria-hidden="true" tabindex="-1"></a> <span class="dv">3</span> ensembl_transcript_id         Transcript stable ID         feature_page</span>
<span id="cb151-15"><a href="biology-bioinformatics.html#cb151-15" aria-hidden="true" tabindex="-1"></a> <span class="dv">4</span> ensembl_transcript_id_version Transcript stable ID version feature_page</span>
<span id="cb151-16"><a href="biology-bioinformatics.html#cb151-16" aria-hidden="true" tabindex="-1"></a> <span class="dv">5</span> ensembl_peptide_id            Protein stable ID            feature_page</span>
<span id="cb151-17"><a href="biology-bioinformatics.html#cb151-17" aria-hidden="true" tabindex="-1"></a> <span class="dv">6</span> ensembl_peptide_id_version    Protein stable ID version    feature_page</span>
<span id="cb151-18"><a href="biology-bioinformatics.html#cb151-18" aria-hidden="true" tabindex="-1"></a> <span class="dv">7</span> ensembl_exon_id               Exon stable ID               feature_page</span>
<span id="cb151-19"><a href="biology-bioinformatics.html#cb151-19" aria-hidden="true" tabindex="-1"></a> <span class="dv">8</span> description                   Gene description             feature_page</span>
<span id="cb151-20"><a href="biology-bioinformatics.html#cb151-20" aria-hidden="true" tabindex="-1"></a> <span class="dv">9</span> chromosome_name               Chromosome<span class="sc">/</span>scaffold name     feature_page</span>
<span id="cb151-21"><a href="biology-bioinformatics.html#cb151-21" aria-hidden="true" tabindex="-1"></a><span class="dv">10</span> start_position                Gene <span class="fu">start</span> (bp)              feature_page</span>
<span id="cb151-22"><a href="biology-bioinformatics.html#cb151-22" aria-hidden="true" tabindex="-1"></a><span class="co"># ... with 3,133 more rows</span></span></code></pre></div>
<p>The <code>name</code> column provides the programmatic name associated with the attribute
that can be used to retrieve that annotation information using the <code>getBM()</code>
function:</p>
<div class="sourceCode" id="cb152"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb152-1"><a href="biology-bioinformatics.html#cb152-1" aria-hidden="true" tabindex="-1"></a><span class="co"># create a tibble with ensembl gene ID, HGNC gene symbol, and gene description</span></span>
<span id="cb152-2"><a href="biology-bioinformatics.html#cb152-2" aria-hidden="true" tabindex="-1"></a>gene_map <span class="ot">&lt;-</span> <span class="fu">as_tibble</span>(</span>
<span id="cb152-3"><a href="biology-bioinformatics.html#cb152-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">getBM</span>(</span>
<span id="cb152-4"><a href="biology-bioinformatics.html#cb152-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">attributes=</span><span class="fu">c</span>(<span class="st">&quot;ensembl_gene_id&quot;</span>, <span class="st">&quot;hgnc_gene_symbol&quot;</span>, <span class="st">&quot;description&quot;</span>),</span>
<span id="cb152-5"><a href="biology-bioinformatics.html#cb152-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">mart=</span>ensembl</span>
<span id="cb152-6"><a href="biology-bioinformatics.html#cb152-6" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb152-7"><a href="biology-bioinformatics.html#cb152-7" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb152-8"><a href="biology-bioinformatics.html#cb152-8" aria-hidden="true" tabindex="-1"></a>gene_map</span>
<span id="cb152-9"><a href="biology-bioinformatics.html#cb152-9" aria-hidden="true" tabindex="-1"></a><span class="co"># A tibble: 68,012 x 3</span></span>
<span id="cb152-10"><a href="biology-bioinformatics.html#cb152-10" aria-hidden="true" tabindex="-1"></a>   ensembl_gene_id hgnc_symbol description                                                               </span>
<span id="cb152-11"><a href="biology-bioinformatics.html#cb152-11" aria-hidden="true" tabindex="-1"></a>   <span class="sc">&lt;</span>chr<span class="sc">&gt;</span>           <span class="er">&lt;</span>chr<span class="sc">&gt;</span>       <span class="er">&lt;</span>chr<span class="sc">&gt;</span>                                                                     </span>
<span id="cb152-12"><a href="biology-bioinformatics.html#cb152-12" aria-hidden="true" tabindex="-1"></a> <span class="dv">1</span> ENSG00000210049 MT<span class="sc">-</span>TF       mitochondrially encoded tRNA<span class="sc">-</span><span class="fu">Phe</span> (UUU<span class="sc">/</span>C) [Source<span class="sc">:</span>HGNC Symbol;Acc<span class="sc">:</span>HGNC<span class="sc">:</span><span class="dv">748</span><span class="sc">~</span></span>
<span id="cb152-13"><a href="biology-bioinformatics.html#cb152-13" aria-hidden="true" tabindex="-1"></a> <span class="dv">2</span> ENSG00000211459 MT<span class="sc">-</span>RNR1     mitochondrially encoded 12S rRNA [Source<span class="sc">:</span>HGNC Symbol;Acc<span class="sc">:</span>HGNC<span class="sc">:</span><span class="dv">7470</span>]       </span>
<span id="cb152-14"><a href="biology-bioinformatics.html#cb152-14" aria-hidden="true" tabindex="-1"></a> <span class="dv">3</span> ENSG00000210077 MT<span class="sc">-</span>TV       mitochondrially encoded tRNA<span class="sc">-</span><span class="fu">Val</span> (GUN) [Source<span class="sc">:</span>HGNC Symbol;Acc<span class="sc">:</span>HGNC<span class="sc">:</span><span class="dv">7500</span>]</span>
<span id="cb152-15"><a href="biology-bioinformatics.html#cb152-15" aria-hidden="true" tabindex="-1"></a> <span class="dv">4</span> ENSG00000210082 MT<span class="sc">-</span>RNR2     mitochondrially encoded 16S rRNA [Source<span class="sc">:</span>HGNC Symbol;Acc<span class="sc">:</span>HGNC<span class="sc">:</span><span class="dv">7471</span>]       </span>
<span id="cb152-16"><a href="biology-bioinformatics.html#cb152-16" aria-hidden="true" tabindex="-1"></a> <span class="dv">5</span> ENSG00000209082 MT<span class="sc">-</span>TL1      mitochondrially encoded tRNA<span class="sc">-</span><span class="fu">Leu</span> (UUA<span class="sc">/</span>G) <span class="dv">1</span> [Source<span class="sc">:</span>HGNC Symbol;Acc<span class="sc">:</span>HGNC<span class="sc">:</span><span class="dv">7</span><span class="sc">~</span></span>
<span id="cb152-17"><a href="biology-bioinformatics.html#cb152-17" aria-hidden="true" tabindex="-1"></a> <span class="dv">6</span> ENSG00000198888 MT<span class="sc">-</span>ND1      mitochondrially encoded NADH<span class="sc">:</span>ubiquinone oxidoreductase core subunit <span class="dv">1</span> [So<span class="sc">~</span></span>
<span id="cb152-18"><a href="biology-bioinformatics.html#cb152-18" aria-hidden="true" tabindex="-1"></a> <span class="dv">7</span> ENSG00000210100 MT<span class="sc">-</span>TI       mitochondrially encoded tRNA<span class="sc">-</span><span class="fu">Ile</span> (AUU<span class="sc">/</span>C) [Source<span class="sc">:</span>HGNC Symbol;Acc<span class="sc">:</span>HGNC<span class="sc">:</span><span class="dv">748</span><span class="sc">~</span></span>
<span id="cb152-19"><a href="biology-bioinformatics.html#cb152-19" aria-hidden="true" tabindex="-1"></a> <span class="dv">8</span> ENSG00000210107 MT<span class="sc">-</span>TQ       mitochondrially encoded tRNA<span class="sc">-</span><span class="fu">Gln</span> (CAA<span class="sc">/</span>G) [Source<span class="sc">:</span>HGNC Symbol;Acc<span class="sc">:</span>HGNC<span class="sc">:</span><span class="dv">749</span><span class="sc">~</span></span>
<span id="cb152-20"><a href="biology-bioinformatics.html#cb152-20" aria-hidden="true" tabindex="-1"></a> <span class="dv">9</span> ENSG00000210112 MT<span class="sc">-</span>TM       mitochondrially encoded tRNA<span class="sc">-</span><span class="fu">Met</span> (AUA<span class="sc">/</span>G) [Source<span class="sc">:</span>HGNC Symbol;Acc<span class="sc">:</span>HGNC<span class="sc">:</span><span class="dv">749</span><span class="sc">~</span></span>
<span id="cb152-21"><a href="biology-bioinformatics.html#cb152-21" aria-hidden="true" tabindex="-1"></a><span class="dv">10</span> ENSG00000198763 MT<span class="sc">-</span>ND2      mitochondrially encoded NADH<span class="sc">:</span>ubiquinone oxidoreductase core subunit <span class="dv">2</span> [So<span class="sc">~</span></span>
<span id="cb152-22"><a href="biology-bioinformatics.html#cb152-22" aria-hidden="true" tabindex="-1"></a><span class="co"># ... with 68,002 more rows</span></span></code></pre></div>
<p>With this Ensembl Gene ID to HGNC symbol mapping in hand, you can combine the
tibble above with other tibbles containing gene information by <a href="data-wrangling.html#relational-data">joining them
together</a>.</p>
<div class="box note">
<p>BioMart/biomaRt is not the only ways to map different gene identifiers. The
Bioconductor package
<a href="https://bioconductor.org/packages/release/bioc/html/AnnotationDbi.html">AnnotateDbi</a>
also provides this functionality in a flexible format independent of the Ensembl
databases. This package includes not only gene identifier mapping and
information, but also identifiers from technology platforms, e.g. probe set IDs
from <a href="biology-bioinformatics.html#microarrays">microarrays</a>, that can help when working with these types of
data. The package also allows comprehensive and flexible <a href="biology-bioinformatics.html#mapping-homologs">homolog
mapping</a>. As with all Bioconductor packages, the
AnnotationDbi documentation is well written and comprehensive, though knowledge
of <a href="https://en.wikipedia.org/wiki/Relational_database">relational databases</a> is
helpful in understanding how the packages work.</p>
</div>
<div class="box readmore">
<ul>
<li><a href="https://www.ensembl.org/biomart/martview">Ensembl BioMart portal</a></li>
<li><a href="https://useast.ensembl.org/info/data/biomart/index.html">Ensembl BioMart web portal documentation</a></li>
<li><a href="http://www.bioconductor.org/packages/release/bioc/html/biomaRt.html">biomaRt Bioconductor documentation</a></li>
</ul>
</div>
</div>
<div id="mapping-homologs" class="section level3" number="7.6.3">
<h3><span class="header-section-number">7.6.3</span> Mapping Homologs</h3>
<p>Sometimes it is necessary to link datasets from different organisms together by
orthology. For example, in an experiment performed in mice, we might be
interested in comparing gene expression patterns observed in the mouse samples
to a publicly available human dataset. In these contexts, we must link gene
identifiers from one organism to their corresponding homologs in the other.
BioMart enables us to extract these linked identifiers by explicitly <a href="https://bioconductor.org/packages/release/bioc/vignettes/biomaRt/inst/doc/accessing_ensembl.html#given-the-human-gene-tp53-retrieve-the-human-chromosomal-location-of-this-gene-and-also-retrieve-the-chromosomal-location-and-refseq-id-of-its-homolog-in-mouse.">connecting
different biomaRt
databases</a>
with the <code>getLDS()</code> function:</p>
<div class="sourceCode" id="cb153"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb153-1"><a href="biology-bioinformatics.html#cb153-1" aria-hidden="true" tabindex="-1"></a>human_db <span class="ot">&lt;-</span> <span class="fu">useEnsembl</span>(<span class="st">&quot;ensembl&quot;</span>, <span class="at">dataset =</span> <span class="st">&quot;hsapiens_gene_ensembl&quot;</span>)</span>
<span id="cb153-2"><a href="biology-bioinformatics.html#cb153-2" aria-hidden="true" tabindex="-1"></a>mouse_db <span class="ot">&lt;-</span> <span class="fu">useEnsembl</span>(<span class="st">&quot;ensembl&quot;</span>, <span class="at">dataset =</span> <span class="st">&quot;mmusculus_gene_ensembl&quot;</span>)</span>
<span id="cb153-3"><a href="biology-bioinformatics.html#cb153-3" aria-hidden="true" tabindex="-1"></a>orth_map <span class="ot">&lt;-</span> <span class="fu">as_tibble</span>(</span>
<span id="cb153-4"><a href="biology-bioinformatics.html#cb153-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">getLDS</span>(<span class="at">attributes =</span> <span class="fu">c</span>(<span class="st">&quot;ensembl_gene_id&quot;</span>, <span class="st">&quot;hgnc_symbol&quot;</span>),</span>
<span id="cb153-5"><a href="biology-bioinformatics.html#cb153-5" aria-hidden="true" tabindex="-1"></a>       <span class="at">mart =</span> human_db,</span>
<span id="cb153-6"><a href="biology-bioinformatics.html#cb153-6" aria-hidden="true" tabindex="-1"></a>       <span class="at">attributesL =</span> <span class="fu">c</span>(<span class="st">&quot;ensembl_gene_id&quot;</span>, <span class="st">&quot;mgi_symbol&quot;</span>),</span>
<span id="cb153-7"><a href="biology-bioinformatics.html#cb153-7" aria-hidden="true" tabindex="-1"></a>       <span class="at">martL =</span> mouse_db</span>
<span id="cb153-8"><a href="biology-bioinformatics.html#cb153-8" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb153-9"><a href="biology-bioinformatics.html#cb153-9" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb153-10"><a href="biology-bioinformatics.html#cb153-10" aria-hidden="true" tabindex="-1"></a>orth_map</span>
<span id="cb153-11"><a href="biology-bioinformatics.html#cb153-11" aria-hidden="true" tabindex="-1"></a><span class="co"># A tibble: 26,390 x 4</span></span>
<span id="cb153-12"><a href="biology-bioinformatics.html#cb153-12" aria-hidden="true" tabindex="-1"></a>   Gene.stable.ID  HGNC.symbol Gene.stable.ID<span class="fl">.1</span>   MGI.symbol</span>
<span id="cb153-13"><a href="biology-bioinformatics.html#cb153-13" aria-hidden="true" tabindex="-1"></a>   <span class="sc">&lt;</span>chr<span class="sc">&gt;</span>           <span class="er">&lt;</span>chr<span class="sc">&gt;</span>       <span class="er">&lt;</span>chr<span class="sc">&gt;</span>              <span class="er">&lt;</span>chr<span class="sc">&gt;</span>     </span>
<span id="cb153-14"><a href="biology-bioinformatics.html#cb153-14" aria-hidden="true" tabindex="-1"></a> <span class="dv">1</span> ENSG00000198695 MT<span class="sc">-</span>ND6      ENSMUSG00000064368 <span class="st">&quot;mt-Nd6&quot;</span>  </span>
<span id="cb153-15"><a href="biology-bioinformatics.html#cb153-15" aria-hidden="true" tabindex="-1"></a> <span class="dv">2</span> ENSG00000212907 MT<span class="sc">-</span>ND4L     ENSMUSG00000065947 <span class="st">&quot;mt-Nd4l&quot;</span></span>
<span id="cb153-16"><a href="biology-bioinformatics.html#cb153-16" aria-hidden="true" tabindex="-1"></a> <span class="dv">3</span> ENSG00000279169 PRAMEF13    ENSMUSG00000094303 <span class="st">&quot;&quot;</span>        </span>
<span id="cb153-17"><a href="biology-bioinformatics.html#cb153-17" aria-hidden="true" tabindex="-1"></a> <span class="dv">4</span> ENSG00000279169 PRAMEF13    ENSMUSG00000094722 <span class="st">&quot;&quot;</span>        </span>
<span id="cb153-18"><a href="biology-bioinformatics.html#cb153-18" aria-hidden="true" tabindex="-1"></a> <span class="dv">5</span> ENSG00000279169 PRAMEF13    ENSMUSG00000095666 <span class="st">&quot;&quot;</span>        </span>
<span id="cb153-19"><a href="biology-bioinformatics.html#cb153-19" aria-hidden="true" tabindex="-1"></a> <span class="dv">6</span> ENSG00000279169 PRAMEF13    ENSMUSG00000094741 <span class="st">&quot;&quot;</span>        </span>
<span id="cb153-20"><a href="biology-bioinformatics.html#cb153-20" aria-hidden="true" tabindex="-1"></a> <span class="dv">7</span> ENSG00000279169 PRAMEF13    ENSMUSG00000094836 <span class="st">&quot;&quot;</span>        </span>
<span id="cb153-21"><a href="biology-bioinformatics.html#cb153-21" aria-hidden="true" tabindex="-1"></a> <span class="dv">8</span> ENSG00000279169 PRAMEF13    ENSMUSG00000074720 <span class="st">&quot;&quot;</span>        </span>
<span id="cb153-22"><a href="biology-bioinformatics.html#cb153-22" aria-hidden="true" tabindex="-1"></a> <span class="dv">9</span> ENSG00000279169 PRAMEF13    ENSMUSG00000096236 <span class="st">&quot;&quot;</span>        </span>
<span id="cb153-23"><a href="biology-bioinformatics.html#cb153-23" aria-hidden="true" tabindex="-1"></a><span class="dv">10</span> ENSG00000198763 MT<span class="sc">-</span>ND2      ENSMUSG00000064345 <span class="st">&quot;mt-Nd2&quot;</span>  </span>
<span id="cb153-24"><a href="biology-bioinformatics.html#cb153-24" aria-hidden="true" tabindex="-1"></a><span class="co"># ... with 26,380 more rows</span></span></code></pre></div>
<p>The <code>mgi_symbol</code> field refers to the gene symbol assigned in the <a href="https://www.jax.org/research-and-faculty/resources/mouse-genome-informatics">Mouse Genome
Informatics database</a>
maintained by <a href="https://www.jax.org/">The Jackson Laboratory</a>.</p>
</div>
</div>
<div id="bio-geneexp" class="section level2" number="7.7">
<h2><span class="header-section-number">7.7</span> Gene Expression</h2>
</div>
<div id="microarrays" class="section level2" number="7.8">
<h2><span class="header-section-number">7.8</span> Microarrays</h2>
</div>
<div id="bio-micro" class="section level2" number="7.9">
<h2><span class="header-section-number">7.9</span> Differential Expression: Microarrays</h2>
</div>
<div id="bio-gsea" class="section level2" number="7.10">
<h2><span class="header-section-number">7.10</span> Gene Set Enrichment Analysis</h2>
</div>
<div id="bio-hts" class="section level2" number="7.11">
<h2><span class="header-section-number">7.11</span> High Throughput Sequencing</h2>
</div>
<div id="bio-rnaseq" class="section level2" number="7.12">
<h2><span class="header-section-number">7.12</span> Differential Expression: RNASeq</h2>
</div>
<div id="bio-networks" class="section level2" number="7.13">
<h2><span class="header-section-number">7.13</span> Biological Network Analysis</h2>
</div>
<div id="bio-scseq" class="section level2" number="7.14">
<h2><span class="header-section-number">7.14</span> Single Cell Sequencing Analysis</h2>

</div>
</div>
<h3>References</h3>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-Bruford2020-zs" class="csl-entry">
Bruford, Elspeth A, Bryony Braschi, Paul Denny, Tamsin E M Jones, Ruth L Seal, and Susan Tweedie. 2020. <span>“Guidelines for Human Gene Nomenclature.”</span> <em>Nat. Genet.</em> 52 (8): 754–58.
</div>
<div id="ref-Holley1965-tu" class="csl-entry">
Holley, R W, J Apgar, G A Everett, J T Madison, M Marquisee, S H Merrill, J R Penswick, and A Zamir. 1965. <span>“<span>STRUCTURE</span> <span>OF</span> a <span>RIBONUCLEIC</span> <span>ACID</span>.”</span> <em>Science</em> 147 (3664): 1462–65.
</div>
<div id="ref-Shows1979-iw" class="csl-entry">
Shows, T B, C A Alper, D Bootsma, M Dorf, T Douglas, T Huisman, S Kit, et al. 1979. <span>“International System for Human Gene Nomenclature (1979) <span>ISGN</span> (1979).”</span> <em>Cytogenet. Cell Genet.</em> 25 (1-4): 96–116.
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="data-science.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="data-visualization.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
