<!DOCTYPE html>
<html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>8 Biology &amp; Bioinformatics | BF591 - R for Biological Sciences</title>
<meta name="generator" content="bookdown 0.32 with bs4_book()">
<meta property="og:title" content="8 Biology &amp; Bioinformatics | BF591 - R for Biological Sciences">
<meta property="og:type" content="book">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="8 Biology &amp; Bioinformatics | BF591 - R for Biological Sciences">
<!-- JS --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://kit.fontawesome.com/6ecbd6c532.js" crossorigin="anonymous"></script><script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="libs/bootstrap-4.6.0/bootstrap.min.css" rel="stylesheet">
<script src="libs/bootstrap-4.6.0/bootstrap.bundle.min.js"></script><script src="libs/bs3compat-0.4.2/transition.js"></script><script src="libs/bs3compat-0.4.2/tabs.js"></script><script src="libs/bs3compat-0.4.2/bs3compat.js"></script><link href="libs/bs4_book-1.0.0/bs4_book.css" rel="stylesheet">
<script src="libs/bs4_book-1.0.0/bs4_book.js"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- CSS --><style type="text/css">
    
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  </style>
<style type="text/css">
    /* Used with Pandoc 2.11+ new --citeproc when CSL is used */
    div.csl-bib-body { }
    div.csl-entry {
      clear: both;
        }
    .hanging div.csl-entry {
      margin-left:2em;
      text-indent:-2em;
    }
    div.csl-left-margin {
      min-width:2em;
      float:left;
    }
    div.csl-right-inline {
      margin-left:2em;
      padding-left:1em;
    }
    div.csl-indent {
      margin-left: 2em;
    }
  </style>
<link rel="stylesheet" href="style.css">
<meta name="description" content="8.1 R in Biology R became popular in biological data analysis in the early to mid 2000s, when microarray technology came into widespread use enabling researchers to look for statistical...">
<meta property="og:description" content="8.1 R in Biology R became popular in biological data analysis in the early to mid 2000s, when microarray technology came into widespread use enabling researchers to look for statistical...">
<meta name="twitter:description" content="8.1 R in Biology R became popular in biological data analysis in the early to mid 2000s, when microarray technology came into widespread use enabling researchers to look for statistical...">
</head>
<body data-spy="scroll" data-target="#toc">

<div class="container-fluid">
<div class="row">
  <header class="col-sm-12 col-lg-3 sidebar sidebar-book"><a class="sr-only sr-only-focusable" href="#content">Skip to main content</a>

    <div class="d-flex align-items-start justify-content-between">
      <h1>
        <a href="index.html" title="">BF591 - R for Biological Sciences</a>
      </h1>
      <button class="btn btn-outline-primary d-lg-none ml-2 mt-1" type="button" data-toggle="collapse" data-target="#main-nav" aria-expanded="true" aria-controls="main-nav"><i class="fas fa-bars"></i><span class="sr-only">Show table of contents</span></button>
    </div>

    <div id="main-nav" class="collapse-lg">
      <form role="search">
        <input id="search" class="form-control" type="search" placeholder="Search" aria-label="Search">
</form>

      <nav aria-label="Table of contents"><h2>Table of contents</h2>
        <ul class="book-toc list-unstyled">
<li><a class="" href="index.html">R for Biological Sciences</a></li>
<li><a class="" href="introduction.html"><span class="header-section-number">1</span> Introduction</a></li>
<li><a class="" href="data-bio.html"><span class="header-section-number">2</span> Data in Biology</a></li>
<li><a class="" href="preliminaries.html"><span class="header-section-number">3</span> Preliminaries</a></li>
<li><a class="" href="prog-basics.html"><span class="header-section-number">4</span> R Programming</a></li>
<li><a class="" href="data-wrangling.html"><span class="header-section-number">5</span> Data Wrangling</a></li>
<li><a class="" href="data-science.html"><span class="header-section-number">6</span> Data Science</a></li>
<li><a class="" href="data-visualization.html"><span class="header-section-number">7</span> Data Visualization</a></li>
<li><a class="active" href="biology-bioinformatics.html"><span class="header-section-number">8</span> Biology &amp; Bioinformatics</a></li>
<li><a class="" href="engineering.html"><span class="header-section-number">9</span> EngineeRing</a></li>
<li><a class="" href="rshiny.html"><span class="header-section-number">10</span> RShiny</a></li>
<li><a class="" href="communicating-with-r.html"><span class="header-section-number">11</span> Communicating with R</a></li>
<li><a class="" href="contribution-guide.html"><span class="header-section-number">12</span> Contribution Guide</a></li>
<li class="book-part">Assignments</li>
<li><a class="" href="starting-an-assignment.html">Starting an Assignment</a></li>
<li><a class="" href="assignment-overview.html">Assignment Overview</a></li>
<li><a class="" href="assignment-0.html">Assignment 0</a></li>
<li><a class="" href="assignment-1.html">Assignment 1</a></li>
<li><a class="" href="assignment-2.html">Assignment 2</a></li>
<li><a class="" href="assignment-3.html">Assignment 3</a></li>
<li><a class="" href="assignment-4.html">Assignment 4</a></li>
<li><a class="" href="assignment-5.html">Assignment 5</a></li>
<li><a class="" href="assignment-6.html">Assignment 6</a></li>
<li><a class="" href="assignment-7.html">Assignment 7</a></li>
<li><a class="" href="final-project.html">Final Project</a></li>
<li class="book-part">Appendix</li>
<li><a class="" href="detailed-class-schedule.html"><span class="header-section-number">A</span> Detailed Class Schedule</a></li>
</ul>

        <div class="book-extra">
          
        </div>
      </nav>
</div>
  </header><main class="col-sm-12 col-md-9 col-lg-7" id="content"><div id="biology-bioinformatics" class="section level1" number="8">
<h1>
<span class="header-section-number">8</span> Biology &amp; Bioinformatics<a class="anchor" aria-label="anchor" href="#biology-bioinformatics"><i class="fas fa-link"></i></a>
</h1>
<div id="bio-r" class="section level2" number="8.1">
<h2>
<span class="header-section-number">8.1</span> R in Biology<a class="anchor" aria-label="anchor" href="#bio-r"><i class="fas fa-link"></i></a>
</h2>
<p>R became popular in biological data analysis in the early to mid 2000s, when
<a href="https://www.genome.gov/genetics-glossary/Microarray-Technology">microarray
technology</a> came
into widespread use enabling researchers to look for statistical differences in
gene expression for thousands of genes across large numbers of samples. As a
result of this popularity, a community of biological researchers and data
analysts created a collection of software packages called
<a href="https://www.bioconductor.org/">Bioconductor</a>, which made a vast array of
cutting edge statistical and bioinformatic methodologies widely available.</p>
<p>Due to its ease of use, widespread community support, rich ecosystem of
biological data analysis packages, and to it being free, R remains one of the
primary languages of biological data analysis. The language’s early focus on
statistical analysis, and later transition to data science in general, makes it
very useful and accessible to perform the kinds of analyses the new data science
of biology required. It is also a bridge between biologists without a
computational background and statisticians and bioinformaticians, who invent
new methods and implement them as R packages that are easily accessible by all.
The language and the package ecosystems and communities it supports continue to
be a major source of biological discovery and innovation.</p>
<p>As a data science, biology benefits from the main strengths of R and <a href="#dw-tidyverse">the
tidyverse</a> when combined with the powerful analytical techniques
available in Bioconductor packages, namely to manipulate, visualize, analyze,
and communicate biological data.</p>

</div>
<div id="biological-data-overview" class="section level2" number="8.2">
<h2>
<span class="header-section-number">8.2</span> Biological Data Overview<a class="anchor" aria-label="anchor" href="#biological-data-overview"><i class="fas fa-link"></i></a>
</h2>
<div id="bio-data-types" class="section level3" number="8.2.1">
<h3>
<span class="header-section-number">8.2.1</span> Types of Biological Data<a class="anchor" aria-label="anchor" href="#bio-data-types"><i class="fas fa-link"></i></a>
</h3>
<p>In very general terms, there are five types of data used in biological data
analysis: raw/primary data, processed data, analysis results, metadata, and
annotation data.</p>
<p><strong>Raw/primary data.</strong> These data are the primary observations made by whatever
instruments/techniques we are using for our experiment. These may include
high-throughput sequencing data, mass/charge ratio data from mass spectrometry,
16S rRNA sequencing data from metagenomic studies, SNPs from genotyping assays,
etc. These data can be very large and are often not efficiently processed using
R. Instead, specialized tools built outside of R are used to first process the
primary data into a form that is amenable to analysis. The most common primary
biological data types include <a href="biology-bioinformatics.html#microarrays">Microarrays</a>, <a href="biology-bioinformatics.html#high-throughput-sequencing">High Throughput Sequencing</a> data,
and <a href="https://en.wikipedia.org/wiki/Mass_spectrometry">mass spectrometry</a> data.</p>
<p><strong>Processed data.</strong> Processed data is the result of any analysis or
transformation of primary data into an intermediate, more interpretable form.
For example, when performing gene expression analysis with RNASeq, short reads
that form the primary data are typically aligned against a genome and then
counted against annotated genes, resulting in a count for each gene in the
genome. Typically, these counts are then combined for many samples into a single
matrix and subject to downstream analyses like differential expression.</p>
<p><strong>Analysis results.</strong> Analysis results aren’t data <em>per se</em>, but are the results
of analysis of primary data or processed results. For example, in gene
expression studies, a first analysis is often differential expression, where
each gene in the genome is tested for expression associated with some outcome
of interest across many samples. Each gene then has a number of statistics and
related values, like log2 fold change, nominal and adjusted p-value, etc. These
forms of data are usually what we use to form interpretations of our datasets
and therefore we must manipulate them in much the same way as any other dataset.</p>
<p><strong>Metadata.</strong> In most experimental settings, multiple samples have been
processed and had the same primary data type collected. These samples also have
information associated with them, which is here termed <em>metadata</em>, or “data
about data”. In our gene expression of post mortem brain experiments mentioned
above, the information about the individuals themselves, including age at death,
whether the person had a disease, the measurements of tissue quality, etc. is
the metadata. The primary and processed data and metadata are usually stored in
different files, where the metadata (or <em>sample information</em> or <em>sample data</em>,
etc) will have one column indicating the unique identifier (ID) of each sample.
The processed data will typically have columns named for each of the sample IDs.</p>
<p><strong>Annotation data.</strong> A tremendous amount of knowledge has been generated about
biological entities, e.g. genes, especially since the publication of the human
genome. Annotation data is publicly available information about the features we
measure in our experiments. This may come in the form of the coordinates in a
genome where genes exist, any known functions of those genes, the domains found
in proteins and their relative sequence, gene identifier cross references across
different gene naming systems (e.g. symbol vs Ensembl ID), single nucleotide
polymorphism genomic locations and associations with traits or diseases, etc.
This is information that we use to aid in interpretation of our experimental
data but generally do not generate ourselves. Annotation data comes in many
forms, some of which are in CSV format.</p>
<p>The figure below contains a simple diagram of how these different types of data
work together for a typical biological data analysis.</p>
<div class="float">
<img src="information_flow.png" alt="Information flow in biological data analysis"><div class="figcaption">Information flow in biological data analysis</div>
</div>
</div>
<div id="csv-files" class="section level3" number="8.2.2">
<h3>
<span class="header-section-number">8.2.2</span> CSV Files<a class="anchor" aria-label="anchor" href="#csv-files"><i class="fas fa-link"></i></a>
</h3>
<p>The most common, convenient, and flexible data file format in biology and
bioinformatics is the <em>character-delimited</em> or <em>character-separated text file</em>.
These files are plain text files (i.e. not the native file format of any
specific program, like Excel) that generally contain rectangles of data. When
formatted correctly, you can think of these files as containing a grid or matrix
of data values with some number of columns, each of which has the same number of
values. Each line of these files has some number of data values separated by a
consistent character, most commonly the comma which are called <a href="https://en.wikipedia.org/wiki/Comma-separated_values">comma-separated
value, or “CSV”, files</a>
and filenames typically end with the extension <code>.csv</code>. Note that
other characters, especially the <tab> character, may be used to create valid
files in this format, and all the same general principles apply. This is an
example of a simple CSV file:</tab></p>
<pre><code>id,somevalue,category,genes
1,2.123,A,APOE
4,5.123,B,"HOXA1,HOXB1"
7,8.123,,SNCA</code></pre>
<p>Some properties and principles of CSV files:</p>
<ul>
<li>The first line often but not always contains the column names of each column</li>
<li>Each value is delimited by the same character, in this case <code>,</code>
</li>
<li>Values can be any value, including numbers and characters</li>
<li>When a value contains the delimiting character (e.g. HOXA1,HOXB1 contains a
<code>,</code>), the value is wrapped in double quotes</li>
<li>Values can be missing, indicated by sequential delimiters (i.e. <code>,,</code> or one
<code>,</code> at the end of the line, if the last column value is missing)</li>
<li>There is no delimiter at the end of the lines</li>
<li>To be well-formatted <em>every line must have the same number of delimited
values</em>
</li>
</ul>
<p>These same properties and principles apply to all character-separated files,
regardless of the specific delimiter used. If a file follows these principles,
they can be loaded very easily into R or any other data analysis setting.
### Common Biological Data Matrices</p>
<p>Typically the first data set you will work with in R is <em>processed data</em> as
described in the previous section. This data has been transformed from primary
data in some way such that it (usually) forms a numeric matrix with features
as rows and samples as columns. The first column of these files usually contains
a feature identifier, e.g. gene identifier, genomic locus, probe set ID, etc and
the remaining columns have numerical values, one per sample. The first row is
usually column names for all the columns in the file. Below is an example of one
of these files from a microarray gene expression dataset loaded into R:</p>
<div class="sourceCode" id="cb372"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb372-1"><a href="biology-bioinformatics.html#cb372-1" tabindex="-1"></a>intensities <span class="ot">&lt;-</span> <span class="fu">read_csv</span>(<span class="st">"example_intensity_data.csv"</span>)</span>
<span id="cb372-2"><a href="biology-bioinformatics.html#cb372-2" tabindex="-1"></a>intensities</span>
<span id="cb372-3"><a href="biology-bioinformatics.html#cb372-3" tabindex="-1"></a><span class="co"># A tibble: 54,675 x 36</span></span>
<span id="cb372-4"><a href="biology-bioinformatics.html#cb372-4" tabindex="-1"></a>   probe     GSM972389 GSM972390 GSM972396 GSM972401 GSM972409 GSM972412 GSM972413 GSM972422 GSM972429</span>
<span id="cb372-5"><a href="biology-bioinformatics.html#cb372-5" tabindex="-1"></a>   <span class="sc">&lt;</span>chr<span class="sc">&gt;</span>         <span class="er">&lt;</span>dbl<span class="sc">&gt;</span>     <span class="er">&lt;</span>dbl<span class="sc">&gt;</span>     <span class="er">&lt;</span>dbl<span class="sc">&gt;</span>     <span class="er">&lt;</span>dbl<span class="sc">&gt;</span>     <span class="er">&lt;</span>dbl<span class="sc">&gt;</span>     <span class="er">&lt;</span>dbl<span class="sc">&gt;</span>     <span class="er">&lt;</span>dbl<span class="sc">&gt;</span>     <span class="er">&lt;</span>dbl<span class="sc">&gt;</span>     <span class="er">&lt;</span>dbl<span class="sc">&gt;</span></span>
<span id="cb372-6"><a href="biology-bioinformatics.html#cb372-6" tabindex="-1"></a> <span class="dv">1</span> 1007_s_at      <span class="fl">9.54</span>     <span class="fl">10.2</span>       <span class="fl">9.72</span>      <span class="fl">9.68</span>      <span class="fl">9.35</span>      <span class="fl">9.89</span>      <span class="fl">9.70</span>      <span class="fl">9.67</span>      <span class="fl">9.87</span></span>
<span id="cb372-7"><a href="biology-bioinformatics.html#cb372-7" tabindex="-1"></a> <span class="dv">2</span> 1053_at        <span class="fl">7.62</span>      <span class="fl">7.92</span>      <span class="fl">7.17</span>      <span class="fl">7.24</span>      <span class="fl">8.20</span>      <span class="fl">6.87</span>      <span class="fl">6.62</span>      <span class="fl">7.23</span>      <span class="fl">7.45</span></span>
<span id="cb372-8"><a href="biology-bioinformatics.html#cb372-8" tabindex="-1"></a> <span class="dv">3</span> 117_at         <span class="fl">5.50</span>      <span class="fl">5.56</span>      <span class="fl">5.06</span>      <span class="fl">7.44</span>      <span class="fl">5.19</span>      <span class="fl">5.72</span>      <span class="fl">5.87</span>      <span class="fl">6.15</span>      <span class="fl">5.46</span></span>
<span id="cb372-9"><a href="biology-bioinformatics.html#cb372-9" tabindex="-1"></a> <span class="dv">4</span> 121_at         <span class="fl">7.27</span>      <span class="fl">7.96</span>      <span class="fl">7.42</span>      <span class="fl">7.34</span>      <span class="fl">7.49</span>      <span class="fl">7.76</span>      <span class="fl">7.44</span>      <span class="fl">7.66</span>      <span class="fl">8.02</span></span>
<span id="cb372-10"><a href="biology-bioinformatics.html#cb372-10" tabindex="-1"></a> <span class="dv">5</span> 1255_g_at      <span class="fl">2.79</span>      <span class="fl">3.10</span>      <span class="fl">2.78</span>      <span class="fl">2.91</span>      <span class="fl">3.02</span>      <span class="fl">2.73</span>      <span class="fl">2.78</span>      <span class="fl">3.56</span>      <span class="fl">2.83</span></span>
<span id="cb372-11"><a href="biology-bioinformatics.html#cb372-11" tabindex="-1"></a> <span class="dv">6</span> 1294_at        <span class="fl">7.51</span>      <span class="fl">7.28</span>      <span class="fl">7.00</span>      <span class="fl">7.18</span>      <span class="fl">7.38</span>      <span class="fl">6.98</span>      <span class="fl">6.90</span>      <span class="fl">7.54</span>      <span class="fl">7.66</span></span>
<span id="cb372-12"><a href="biology-bioinformatics.html#cb372-12" tabindex="-1"></a> <span class="dv">7</span> 1316_at        <span class="fl">3.89</span>      <span class="fl">4.36</span>      <span class="fl">4.24</span>      <span class="fl">3.94</span>      <span class="fl">4.20</span>      <span class="fl">4.34</span>      <span class="fl">4.06</span>      <span class="fl">4.24</span>      <span class="fl">4.11</span></span>
<span id="cb372-13"><a href="biology-bioinformatics.html#cb372-13" tabindex="-1"></a> <span class="dv">8</span> 1320_at        <span class="fl">4.65</span>      <span class="fl">4.91</span>      <span class="fl">4.70</span>      <span class="fl">4.78</span>      <span class="fl">5.06</span>      <span class="fl">4.71</span>      <span class="fl">4.55</span>      <span class="fl">4.58</span>      <span class="fl">5.10</span></span>
<span id="cb372-14"><a href="biology-bioinformatics.html#cb372-14" tabindex="-1"></a> <span class="dv">9</span> 1405_i_at      <span class="fl">8.03</span>      <span class="fl">7.47</span>      <span class="fl">5.42</span>      <span class="fl">7.21</span>      <span class="fl">9.48</span>      <span class="fl">6.79</span>      <span class="fl">6.57</span>      <span class="fl">8.50</span>      <span class="fl">6.36</span></span>
<span id="cb372-15"><a href="biology-bioinformatics.html#cb372-15" tabindex="-1"></a><span class="dv">10</span> 1431_at        <span class="fl">3.09</span>      <span class="fl">3.78</span>      <span class="fl">3.33</span>      <span class="fl">3.12</span>      <span class="fl">3.21</span>      <span class="fl">3.27</span>      <span class="fl">3.37</span>      <span class="fl">3.84</span>      <span class="fl">3.32</span></span>
<span id="cb372-16"><a href="biology-bioinformatics.html#cb372-16" tabindex="-1"></a><span class="co"># ... with 54,665 more rows, and 26 more variables: GSM972433 &lt;dbl&gt;, GSM972438 &lt;dbl&gt;, GSM972440 &lt;dbl&gt;,</span></span>
<span id="cb372-17"><a href="biology-bioinformatics.html#cb372-17" tabindex="-1"></a><span class="co">#   GSM972443 &lt;dbl&gt;, GSM972444 &lt;dbl&gt;, GSM972446 &lt;dbl&gt;, GSM972453 &lt;dbl&gt;, GSM972457 &lt;dbl&gt;,</span></span>
<span id="cb372-18"><a href="biology-bioinformatics.html#cb372-18" tabindex="-1"></a><span class="co">#   GSM972459 &lt;dbl&gt;, GSM972463 &lt;dbl&gt;, GSM972467 &lt;dbl&gt;, GSM972472 &lt;dbl&gt;, GSM972473 &lt;dbl&gt;,</span></span>
<span id="cb372-19"><a href="biology-bioinformatics.html#cb372-19" tabindex="-1"></a><span class="co">#   GSM972475 &lt;dbl&gt;, GSM972476 &lt;dbl&gt;, GSM972477 &lt;dbl&gt;, GSM972479 &lt;dbl&gt;, GSM972480 &lt;dbl&gt;,</span></span>
<span id="cb372-20"><a href="biology-bioinformatics.html#cb372-20" tabindex="-1"></a><span class="co">#   GSM972487 &lt;dbl&gt;, GSM972488 &lt;dbl&gt;, GSM972489 &lt;dbl&gt;, GSM972496 &lt;dbl&gt;, GSM972502 &lt;dbl&gt;,</span></span>
<span id="cb372-21"><a href="biology-bioinformatics.html#cb372-21" tabindex="-1"></a><span class="co">#   GSM972510 &lt;dbl&gt;, GSM972512 &lt;dbl&gt;, GSM972521 &lt;dbl&gt;</span></span></code></pre></div>
<p>The file has 54,676 rows, consisting of one header row which R loads in as the
column names, and the remaining are probe sets, one per row. There are 36
columns, where the first contains the probe set ID (e.g. <code>1007_s_at</code>) and the
remaining 35 columns correspond to samples.</p>
</div>
<div id="biological-data-is-not-tidy" class="section level3" number="8.2.3">
<h3>
<span class="header-section-number">8.2.3</span> Biological data is NOT Tidy!<a class="anchor" aria-label="anchor" href="#biological-data-is-not-tidy"><i class="fas fa-link"></i></a>
</h3>
<p>As mentioned in the <a href="#tidy%20data">tidy data</a> section, the tidyverse packages
assume data to be in so-called “tidy format”, with variables as columns and
observations as rows. Unfortunately, certain forms of biological data are
typically available in the opposite orientation - variables are in rows and
observations are in columns. This is primarily true in feature data matrices,
e.g. gene expression counts matrices, where the number of variables (e.g. genes)
is much larger than the number of samples, which tend to be small very small
compared with the number of features. This format is convenient for humans to
interact with using, e.g. spreadsheet programs like Microsoft Excel, but can
unfortunately make performing certain operations on them challenging in
tidyverse.</p>
<p>For example, consider the microarray expression dataset in the previous section.
Each of the 54,676 rows is a probeset, and each of the 35 numeric columns is a
sample. This is a very large number of probesets to consider, especially if we
plan to conduct a statistical test on each, which would impose a substantial
<a href="#multiple-hypothesis-testing">multiple hypothesis testing</a> burden on our
results. We may therefore wish to eliminate probesets that have very low
variance from the overall dataset, since these probesets are not likely to have
a detectable statistical difference in our tests. However, computing the
variance for each probeset is a computation <em>across all columns</em>, not on columns
themselves, and this is not what tidyverse is designed to do well. Said
differently, R and tidyverse do not operate by default on the rows of a data
frame, tibble, or matrix.</p>
<p>Both base R and tidyverse are optimized to perform computations on columns, not
rows. The reasons for this are buried in the details of how the R program itself
was written to organize data internally and are beyond the scope of this book.
The consequence of this design choice is that, while we can perform operations
on the rows rather than the columns of a data structure, our code may perform
very poorly (i.e. take a very long time to run).</p>
<p>When working with these datasets, we have a couple options to deal with this
issue:</p>
<ul>
<li><p><strong>Pivot into long format.</strong> As described in the <a href="data-wrangling.html#rearranging-data">Rearranging Data</a> section, we
can rearrange our tibble to be more amenable to certain computations. In our
earlier example, we wish to group all of our measurements by probeset and
compute the variance of each, then possibly filter out probesets based on low
variance. We can therefore combine <code>pivot_longer()</code>, <code><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by()</a></code>,
<code><a href="https://dplyr.tidyverse.org/reference/summarise.html">summarize()</a></code>, and finally <code><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html">left_join()</a></code> to perform this operation. Exactly
how to do this is left as an exercise in <a href="assignment-1.html#assignment-1">Assignment 1</a>.</p></li>
<li>
<p><strong>Compute row-wise statistics using <code><a href="https://rdrr.io/r/base/apply.html">apply()</a></code>.</strong> As described in <a href="prog-basics.html#iteration">Iteration</a>,
R is a functional programming language and implements iteration in a
functional style using the <code><a href="https://rdrr.io/r/base/apply.html">apply()</a></code> function. The <a href="prog-basics.html#apply-and-friends"><code>apply()   function</code></a> accepts a <code>MARGIN</code> argument of 1 or 2 if the
provided function is to be applied to the columns or rows, respectively. This
method can be used to compute a summary statistic on each row of a tibble and
the result saved back into the tibble using the column set operator:</p>
<div class="sourceCode" id="cb373"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">intensity_variance</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html">apply</a></span><span class="op">(</span><span class="va">intensities</span>, <span class="fl">2</span>, <span class="va">var</span><span class="op">)</span></span>
<span><span class="va">intensities</span><span class="op">$</span><span class="va">variance</span> <span class="op">&lt;-</span> <span class="va">intensity_variance</span></span></code></pre></div>
</li>
</ul>
</div>
</div>
<div id="bio-bioconductor" class="section level2" number="8.3">
<h2>
<span class="header-section-number">8.3</span> Bioconductor<a class="anchor" aria-label="anchor" href="#bio-bioconductor"><i class="fas fa-link"></i></a>
</h2>
<p><a href="https://www.bioconductor.org/">Bioconductor</a> is an organized collection of
strictly biological analysis methods packages for R. These packages are hosted
and maintained outside of <a href="https://cran.r-project.org/">CRAN</a> because the
maintainers enforce <a href="https://bioconductor.org/developers/package-guidelines/">a more rigorous set of coding quality, testing, and
documentation
standards</a> than is
required by normal R packages. These stricter requirements arose from a
recognition that software packages are generally only as usable as their
documentation allows them to be, and also that many if not most of the users of
these packages are not statisticians or experienced computer programmers.
Rather, they are people like us: biological analysis practitioners who may or
may not have substantial coding experience but must analyze our data
nonetheless. The excellent documentation and community support of the
bioconductor ecosystem is a primary reason why R is such a popular language in
biological analysis.</p>
<p>Bioconductor is divided into roughly two sets of packages: <em>core maintainer</em>
packages and user contributed packages. The core maintainer packages are among
the most critical, because they define a set of common objects and classes (e.g.
the <a href="https://www.bioconductor.org/packages/devel/bioc/vignettes/Biobase/inst/doc/ExpressionSetIntroduction.pdf">ExpressionSet class</a>
in the <a href="https://bioconductor.org/packages/release/bioc/html/Biobase.html">Biobase package</a>)
that all Bioconductor packages know how to work with. This common base provides
consistency among all Bioconductor packages thereby streamlining the learning
process. User contributed and maintained packages provide specialized
functionality for specific types of analysis.</p>
<p>Bioconductor itself must be installed prior to installing other Bioconductor
packages. To [install bioconductor], you can run the following:</p>
<div class="sourceCode" id="cb374"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html">require</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span><span class="op">(</span>version <span class="op">=</span> <span class="st">"3.14"</span><span class="op">)</span></span></code></pre></div>
<div class="box warning">
<p>Bioconductor undergoes regular releases indicated by a version, e.g. <code>version = "3.14"</code> at the time of writing. Every bioconductor package also has a version,
and each of those versions may or may not be compatible with a specific version
of Bioconductor. To make matters worse, Bioconductor versions are only
compatible with certain versions of R. Bioconductor version 3.14 requires R
version 4.1.0, and will not work with older R versions. These versions can cause
major version compatibility issues when you are forced to use older versions of
R, as may sometimes be the case on managed compute clusters. There is not a good
general solution for ensuring all your packages work together, but the general
best rule of thumb is to use the most current version of R and all packages at
the time when you write your code.</p>
</div>
<p>The <code>BiocManager</code> package is the only Bioconductor package installable using
<code><a href="https://rdrr.io/r/utils/install.packages.html">install.packages()</a></code>. After installing the BiocManager package, you may then
install bioconductor packages:</p>
<div class="sourceCode" id="cb375"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># installs the affy bioconductor package for microarray analysis</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span><span class="op">(</span><span class="st">"affy"</span><span class="op">)</span></span></code></pre></div>
<p>One key aspect of Bioconductor packages is consistent and helpful documentation;
every package page on the Bioconductor.org site lists a block of code that will
install the package, e.g. for the <a href="https://www.bioconductor.org/packages/release/bioc/html/affy.html">affy
package</a>. In
addition, Biconductor provides three types of documentation:</p>
<ul>
<li>
<a href="https://bioconductor.org/packages/release/workflows/">Workflow tutorials</a> on
how to perform specific analysis use cases</li>
<li>
<a href="https://bioconductor.org/help/package-vignettes/">Package vignettes</a> for
every package, which provide a worked example of how to use the package that
can be adapted by the user to their specific case</li>
<li>Detailed, consistently formatted reference documentation that gives precise
information on functionality and use of each package</li>
</ul>
<p>The thorough and useful documentation of packages in Bioconductor is one of the
reasons why the package ecosystem, and R, is so widely used in biology and
bioinformatics.</p>
<p>The base Bioconductor packages define convenient data structures for storing and
analyzing many types of data. Recall earlier in the <a href="biology-bioinformatics.html#bio-data-types">Types of Biological Data</a>
section, we described five types of biological data: primary, processed,
analysis, metadata, and annotation data. Bioconductor provides a convenient
class for storing many of these different data types together in one place,
specifically the SummarizedExperiment class in the <a href="https://bioconductor.org/packages/release/bioc/html/SummarizedExperiment.html">package of the same
name</a>
package. An illustration of what a SummarizedExperiment object stores is below,
from the Bioconductor maintainer’s paper in Nature:</p>
<div class="float">
<img src="https://media.springernature.com/full/springer-static/image/art%3A10.1038%2Fnmeth.3252/MediaObjects/41592_2015_Article_BFnmeth3252_Fig2_HTML.jpg?as=webp" alt="SummarizedExperiment Schematic - Huber, et al. 2015. “Orchestrating High-Throughput Genomic Analysis with Bioconductor.” Nature Methods 12 (2): 115–21."><div class="figcaption">SummarizedExperiment Schematic - Huber, et al. 2015. “Orchestrating High-Throughput Genomic Analysis with Bioconductor.” Nature Methods 12 (2): 115–21.</div>
</div>
<p>As you can see from the figure, this class stores processed data (<code>assays</code>),
metadata (<code>colData</code> and <code>exptData</code>), and annotation data (<code>rowData</code>). The
SummarizedExperiment class is used ubiquitously throughout the Bioconductor
package ecosystem, as are other core classes some of which we will cover later
in this chapter.</p>

</div>
<div id="microarrays" class="section level2" number="8.4">
<h2>
<span class="header-section-number">8.4</span> Microarrays<a class="anchor" aria-label="anchor" href="#microarrays"><i class="fas fa-link"></i></a>
</h2>
<div class="float">
<img src="microarray_chip.png" alt="A Microarray Device - Thermo Fisher Scientific"><div class="figcaption">A Microarray Device - Thermo Fisher Scientific</div>
</div>
<p>Microarrays are devices that measure the relative abundance of thousands of
distinct DNA sequences simultaneously. Short (~25 nucleotide) single-stranded
DNA molecules called <em>probes</em> are deposited on a small glass slide in a grid of
spots, where each spot contains many copies of a probe with identical sequence.
The probe sequences are selected <em>a priori</em> based on the purpose of the array.
For example, gene expression microarrays have probes that correspond to the
coding regions of the genome for a species of interest, while genotyping
microarrays use sequences with known variants found in a population of genomes,
most often the human population. Microarrays of the same type (e.g. human gene
expression) all have the same set of probes. The choice of DNA sequence probes
therefore determines what the microarray measures and how to interpret the data.
The design of a microarray is illustrated in the following figure.</p>
<div class="float">
<img src="microarray_design.png" alt="Illustration of Microarray Design"><div class="figcaption">Illustration of Microarray Design</div>
</div>
<p>A microarray device generates data by applying a specially prepared DNA sample
to it; the sample usually corresponds to a single biological specimen, e.g. an
individual patient. The preparation method for the sample depends on what is
being measured:</p>
<ul>
<li>When measuring DNA directly, e.g. genetic variants, the DNA itself is
biochemically extracted</li>
<li>When measuring gene expression via RNA abundance, RNA is first extracted and
then reverse transcribed to complementary DNA (cDNA)</li>
</ul>
<p>In either case, the molecules that will be applied to the microarray are DNA
molecules. After extraction and preparation, the DNA molecules are then randomly
cut up into shorter molecules (i.e. <em>sheared</em>) and each molecule has a molecular
tag biochemically ligated to it that will emit fluorescence when excited by a
specific wavelength of light. This tagged DNA sample is then washed over the
microarray chip, where DNA molecules that share sequence complementarity with
the probes on the array pair together. After this treatment, the microarray is
washed to remove DNA molecules that did not have a match on the array, leaving
only those molecules with a sequence match that remain bound to probes.</p>
<p>The microarray chip is then loaded into a scanning device, where a laser with a
specific wavelength of light is then shone onto the array, causing the spots
with tagged DNA molecules associated with probes to fluoresce, and other spots
remain dark. A high resolution image is taken of the fluorescent array, and the
image is analyzed to map the intensity of the light on each spot to a numeric
value proportional to its intensity. The reason for this is that, since each
spot has many individual probe molecules contained within it, the more copies of
the corresponding DNA molecule were in the sample, the more light the spot
emits. In this way, the relative abundance of all probes on the array are
measured simultaneously. The process of generating microarray data from a sample
is illustrated in the following figure.</p>
<div class="float">
<img src="microarray_data_gen.png" alt="Illustration of Microarray Data Generation Process"><div class="figcaption">Illustration of Microarray Data Generation Process</div>
</div>
<p>After the microarray has been scanned, the relative copy number of the DNA in
the sample matching the probes on the microarray are expressed as the intensity
of fluorescence of each probe. The raw intensity data from the scan has been
processed and analyzed by the scanner software to account for technical biases
and artefacts of the scanning instrument and data generation process. The data
from a single scan is processed and stored in a file in <a href="https://www.affymetrix.com/support/developer/powertools/changelog/gcos-agcc/cel.html">CEL
format</a>,
a proprietary data format that stores the raw probe intensity data that can be
loaded for downstream analysis.</p>

</div>
<div id="high-throughput-sequencing" class="section level2" number="8.5">
<h2>
<span class="header-section-number">8.5</span> High Throughput Sequencing<a class="anchor" aria-label="anchor" href="#high-throughput-sequencing"><i class="fas fa-link"></i></a>
</h2>
<p>High throughput sequencing (HTS) technologies measure and digitize the
nucleotide sequences of thousands to billions of DNA molecules simultaneously.
In contrast to microarrays, which cannot identify new sequences due to their
<em>a priori</em> probe based design, HTS instruments can in principle determine the
sequence of any DNA molecule in a sample. For this reason, HTS datasets are
sometimes called <em>unbiased</em> assays, because the sequences they identify are not
predetermined.</p>
<p>The most popular HTS technology at the time of writing is provided by the
<a href="https://en.wikipedia.org/wiki/Illumina,_Inc.">Illumina biotechnology company</a>.
This technology was first developed and commercialized by the company
<a href="https://www.illumina.com/science/technology/next-generation-sequencing/illumina-sequencing-history.html">Solexa</a>
during the late 1990s and was instrumental in the completion of the human
genome. The cost of generating data with this technology has reduced by several
orders of magnitude since then, resulting in an exponential growth of biological
data and fundamentally changed our understanding of biology and health. Because
of this, the analysis of these data, and not the time and cost of generation,
has become the primary bottleneck in biological and biomedical research.</p>
<p>The Illumina HTS technology uses a biotechnological technique called <em>sequencing
by synthesis</em>. Briefly, this technique entails biochemically ligating millions
to billions of short (~300 nucleotide long) DNA molecules to a glass slide,
denatured to become single stranded, and then used to synthesize the
complementary strand by incorporating fluorescently tagged nucleotides. The
tagged nucleotides that are incorporated on each nucleotide addition are excited
with a laser and a high resolution photograph is taken of the flow cell. This
process is repeated many times until the DNA molecules on the flow cell have
been completely synthesized, and the collective images are computationally
analyzed to build up the complete sequence of each molecule.</p>
<div class="box note">
<p>Other HTS methods have been developed that use different strategies, some of
which do not utilize the sequencing by synthesis method. These alternative
technologies differ in the length, number, confidence and per base cost of the
DNA molecules they sequence. At the time of writing, the two most common
technologies include:</p>
<ul>
<li>
<a href="https://www.pacb.com/smrt-science/smrt-sequencing/">Pacific Biosciences (PacBio) Long Read
Sequencing</a> - this
technology uses an engineered polymerase that performs run-on amplification of
circularized DNA molecules and records a signature of each different base as it
is incorporated into the amplification product. PacBio sequencing can generate
reads that are thousands of nucleotides in length.</li>
<li>
<a href="https://nanoporetech.com/">Oxford Nanopore Sequencing</a> (ONS) - this technology uses
engineered proteins to form pores in an electrically conductive membrane,
through which single stranded DNA molecules pass. As the individual nucleotides
of a DNA molecule pass through the pore, the electrical current induced by the
different nucleotide molecular structures is recorded and analyzed with signal
processing algorithms to recover the original sequence of the molecule. No
synthesis occurs in ONS technology.</li>
</ul>
<p>Due to its widespread use, only Illumina sequencing data analysis is covered in
this book, though additions in the future may be warranted.</p>
</div>
<p>Any DNA molecules may be subjected to sequencing via this method; the meaning of
the resulting datasets and the goal of subsequent analysis are therefore
dependent upon the material being studied. Below are some of the most common
types of HTS experiments:</p>
<ul>
<li>
<em>Whole genome sequencing (WGS)</em> - the input to sequencing is randomly sheared
genomic DNA molecules, from single organisms or a collection of organisms, with
the goal of reconstructing or further refining the genome sequence of the
constituent organisms</li>
<li>
<em>RNA Sequencing (RNASeq)</em> - the input is complementary DNA (cDNA) that has
been <a href="https://www.ncbi.nlm.nih.gov/books/NBK19424/">reverse transcribed</a> from
RNA extracted from a sample, with the goal of detecting which RNA transcripts
(and therefore genes) are expressed in that sample</li>
<li>
<em>Whole genome bisulfite sequencing (WGBS)</em> - the input to sequencing is
randomly sheared genomic DNA that has undergone bisulfite treatment, which
converts methylated cytosines to thymines, thereby enabling detection of
epigenetic methylation marks in the DNA</li>
<li>
<em>Chromatin immunoprecipitation followed by sequencing (ChIPSeq)</em> - the input
is genomic DNA that was bound by specific DNA-binding proteins (e.g.
transcription factors), that allows identification of locations in the genome
where those proteins were bound</li>
</ul>
<p>A complete coverage of the many different types of sequencing assays is beyond
the scope of this book, but analysis techniques used for of some of these
different types of experiments are covered later in this chapter.</p>
<div id="raw-hts-data" class="section level3" number="8.5.1">
<h3>
<span class="header-section-number">8.5.1</span> Raw HTS Data<a class="anchor" aria-label="anchor" href="#raw-hts-data"><i class="fas fa-link"></i></a>
</h3>
<p>The raw data produced by the sequencing instrument, or <em>sequencer</em>, are the
digitized DNA sequences for each of the billions of molecules that were captured
by the flow cell. The digitized nucleotide sequence for each molecule is called
a <em>read</em>, and therefore every run of a sequencing instrument produces millions
to billions of reads. The data is stored in a standardized format called the
<a href="https://en.wikipedia.org/wiki/FASTQ_format">FASTQ file format</a>. Data for a
single read in FASTQ format is shown below:</p>
<pre><code>@SEQ_ID
GATTTGGGGTTCAAAGCAGTATCGATCAAATAGTAAATCCATTTGTTCAACTCACAGTTT
+
!''*((((***+))%%%++)(%%%%).1***-+*''))**55CCF&gt;&gt;&gt;&gt;&gt;&gt;CCCCCCC65</code></pre>
<p>Each read has four lines in the FASTQ file. The first beginning with <code>@</code> is the
sequencing <em>header</em> or name, which provides a unique identifier for the read in
the file. The second line is the nucleotide sequence as detected by the
sequencer. The third line is a secondary header line that starts with <code>+</code> and
may or may not include further information. The last line contains the <a href="https://en.wikipedia.org/wiki/Phred_quality_score">phred
quality scores</a> for the base
calls in the corresponding position of the read. Briefly, each character in the
phred quality score line has a corresponding integer is proportional to the
confidence that the base in the corresponding position of the read was called
correctly. This information is used to assess the quality of the data. See <a href="https://en.wikipedia.org/wiki/FASTQ_format#Quality">this
more in-depth explanation of phred
score</a> for more information.</p>
</div>
<div id="preliminary-hts-data-analysis" class="section level3" number="8.5.2">
<h3>
<span class="header-section-number">8.5.2</span> Preliminary HTS Data Analysis<a class="anchor" aria-label="anchor" href="#preliminary-hts-data-analysis"><i class="fas fa-link"></i></a>
</h3>
<p>Since all HTS experiments generate the same type of data (i.e. reads), all data
analyses begin with processing of the sequences via bioinformatic algorithms;
the choice of algorithm is determined by the experiment type and the biological
question being asked. A full treatment of the variety of bioinformatic
approaches to processing these data is beyond the scope of this book. We will
focus on one analysis approach, <em>alignment</em>, that is commonly employed to
produce derived data that is then further processed using R.</p>
<p>It is very common for HTS data, also called <em>short read sequencing</em> data, to be
generated from organisms that have already had their genome sequenced, e.g.
human or mouse. The reads in a dataset can therefore be compared with the
appropriate genome sequence to aid in interpretation. This comparison amounts to
searching in the organism’s genome, or the so-called <em>reference genome</em>, for the
location or locations from which the molecule captured by each read originated.
This search process is called <em>alignment</em>, where each short read sequence is
<em>aligned</em> against the usually much larger reference genome to identify locations
with identical or nearly identical sequence. Said differently, the alignment
process attempts to assign each read to one or more locations in a reference
genome, where only reads that have a match in the target genome will be
assigned.</p>
<div class="box important">
<p>The alignment process is computationally intensive and requires specialized
software to perform efficiently. While it may be technically capable of
performing alignment, R is not optimized to process character data in the volume
typical of these datasets. Therefore, it is not common to work with short read
data (i.e. FASTQ files) directly in R, and other processing steps are required
to transform the sequencing data into other forms that are then amenable to
analysis in R.</p>
</div>
<p>The end result of aligning all the reads in a short read dataset is assignments
of reads to positions across an entire reference genome. The distribution of
those read alignment locations across the genome forms the basis of the data
used to interpret the dataset. Specifically, each <em>locus</em>, or location spanning
one or more bases, has zero or more reads aligned to it. The number of reads
aligned to a locus, or the <em>read count</em> is commonly interpreted as being
proportional to the number of those molecules found in the original sample.
Therefore, the analysis and interpretation of HTS data commonly involves the
analysis of <em>count data</em>, which we discuss in detail in the next section. The
following figure illustrates the concepts described in this section.</p>
<div class="float">
<img src="hts_examples.png" alt="Illustration of WGS and RNASeq read alignment process"><div class="figcaption">Illustration of WGS and RNASeq read alignment process</div>
</div>
</div>
<div id="count-data" class="section level3" number="8.5.3">
<h3>
<span class="header-section-number">8.5.3</span> Count Data<a class="anchor" aria-label="anchor" href="#count-data"><i class="fas fa-link"></i></a>
</h3>
<p>As mentioned above, the number of reads, or <em>read counts</em> that align to each
locus of a genome form a distribution we can use to interpret the data. Count
data have two important properties that require special attention:</p>
<ul>
<li>
<strong>Count data are integers</strong> - the read counts assigned to a particular locus
are whole numbers</li>
<li>
<strong>Counts are non-negative</strong> - counts can either be zero or positive; they
cannot be negative</li>
</ul>
<p>These two properties have the important consequence that <em>count data are not
normally distributed</em>. Unlike with microarray data, common statistical
techniques like linear regression are not directly applicable to count data. We
must therefore account for this property of the data prior to statistical
analysis in one of two ways:</p>
<ul>
<li>
<strong>Use statistical methods that model counts data</strong> - generalized linear models
that model data using Poisson and Negative Binomial distributions</li>
<li>
<strong>Transform the data to have be normally distributed</strong> - certain <a href="https://en.wikipedia.org/wiki/Data_transformation_(statistics)">statistical
transformations</a>,
such as <a href="https://rdrr.io/bioc/DESeq2/man/rlog.html">regularized log</a> or rank
transformations can make counts data more closely follow a normal distribution</li>
</ul>
<p>We will discuss these strategies in greater detail in the <a href="biology-bioinformatics.html#rnaseq-gene-expression-data">RNASeq Gene
Expression Data</a> section.</p>

</div>
</div>
<div id="gene-identifiers" class="section level2" number="8.6">
<h2>
<span class="header-section-number">8.6</span> Gene Identifiers<a class="anchor" aria-label="anchor" href="#gene-identifiers"><i class="fas fa-link"></i></a>
</h2>
<div id="gene-identifier-systems" class="section level3" number="8.6.1">
<h3>
<span class="header-section-number">8.6.1</span> Gene Identifier Systems<a class="anchor" aria-label="anchor" href="#gene-identifier-systems"><i class="fas fa-link"></i></a>
</h3>
<p>Since the first genetic sequence for a gene was determined in 1965 - <a href="https://pubmed.ncbi.nlm.nih.gov/14263761/">an alanine
tRNA in yeast</a><span class="citation">(<a href="detailed-class-schedule.html#ref-Holley1965-tu">Holley et al. 1965</a>)</span> -
scientists have been trying to give them names. In 1979, <a href="https://pubmed.ncbi.nlm.nih.gov/295268/">formal guidelines for
the nomenclature used for human genes was
proposed</a><span class="citation">(<a href="detailed-class-schedule.html#ref-Shows1979-iw">Shows et al. 1979</a>)</span> along with the
founding of the Human Gene Nomenclature Committee <span class="citation">(<a href="detailed-class-schedule.html#ref-Bruford2020-zs">Bruford et al. 2020</a>)</span>, so that
researchers would have a common vocabulary to refer to genes in a consistent
way. In 1989, this committee was placed under the auspices of the Human Genome
Organisation (HUGO), becoming the HUGO Gene Nomenclature Committee (HGNC). The
HGNC has been the official body providing guidelines and gene naming authority
since, and as such HGNC gene names are often called <em>gene symbols</em>.</p>
<div class="box note">
<p>As per <span class="citation">Bruford et al. (<a href="detailed-class-schedule.html#ref-Bruford2020-zs">2020</a>)</span>, the official human gene symbol guidelines are as
follows:</p>
<ol style="list-style-type: decimal">
<li>Each gene is assigned a unique symbol, HGNC ID and descriptive name.</li>
<li>Symbols contain only uppercase Latin letters and Arabic numerals.</li>
<li>Symbols should not be the same as commonly used abbreviations</li>
<li>Nomenclature should not contain reference to any species or “G” for gene.</li>
<li>Nomenclature should not be offensive or pejorative.</li>
</ol>
</div>
<p>Gene symbols are the most human-readable system for naming genes. The gene
symbols <em>BRCA1</em>, <em>APOE</em>, and <em>ACE2</em> may be familiar to you as they are involved
in common diseases, namely breast cancer, Alzheimer’s Disease risk, and
SARS-COV-2, respectively. Typically, the gene symbol is an acronym that roughly
represents a label of what the gene is or does (or was originally found to be or
do, as many genes are subsequently discovered to be involved in entirely
separate processes as well), e.g. <em>APOE</em> represents the gene <a href="https://www.ensembl.org/Homo_sapiens/Gene/Summary?db=core;g=ENSG00000130203;r=19:44905791-44909393">apolipoprotein
E</a>.
This convention is convenient for humans when reading and identifying genes.
However, standardized though the symbol conventions may be, they are not always
easy for computers to work with, and ambiguities can cause mapping problems.</p>
<p>Other gene identifier systems developed along with other gene information
databases. In 2000, the <a href="https://www.ensembl.org/">Ensembl</a> genome browser was
launched by the <a href="https://wellcome.org/">Wellcome Trust</a>, a charitable foundation
based in the United Kingdom, with the goal of providing automated annotation of
the human genome. The Ensembl Project, which supports the genome browser,
recognized even before the publication of the human genome that manual
annotation and curation of genes is a slow and labor intensive process that
would not provide researchers around the world timely access to information. The
project therefore created a set of automated tools and pipelines to collect,
process, and publish rapid and consistent annotation of genes in the human
genome. Since its initial release, Ensembl now supports over 50,000 different
genomes.</p>
<p>Ensembl assigns a automatic, systematic ID called the Ensembl Gene ID to every
gene in its database. Human Ensembl Gene IDs take the form <code>ENSG</code> plus an 11
digit number, optionally followed by a period delimited version number. For
example, at the time of writing the BRCA1 gene has an Ensembl Gene ID of
<code>ENSG00000012048.23</code>. The stable ID portion (i.e. <code>ENSGNNNNNNNNNNN</code>) will remain
associated with the gene forever (unless the gene annotation <a href="http://useast.ensembl.org/Help/View?id=181">changes
“dramatically” in which case it is
retired</a>). The <code>.23</code> is the version
number of the gene annotation, meaning this gene has been updated 22 times (plus
its initial version) since addition to the database. The additional version
information is very important for reproducibility of biological analysis, since
conclusions drawn by results of these analyses are usually based on the most
current information about a gene which are continually updated over time.</p>
<div class="box note">
<p>Ensembl maintains annotations for many different organisms, and the gene
identifiers for each genome contain codes that indicate which organism the gene
is for. Here are the codes for orthologs of the body plan developmental gene
<a href="https://www.genecards.org/cgi-bin/carddisp.pl?gene=HOXA1">HOXA1</a> in several
different species:</p>
<div class="inline-table"><table class="table table-sm">
<thead><tr class="header">
<th align="left">Gene ID Prefix</th>
<th align="left">Organism</th>
<th align="left">Symbol</th>
<th align="left">Example</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left"><code>ENSG</code></td>
<td align="left">Homo sapiens</td>
<td align="left">HOXA1</td>
<td align="left">ENSG00000105991</td>
</tr>
<tr class="even">
<td align="left"><code>ENSMUSG</code></td>
<td align="left">Mus musculus (mouse)</td>
<td align="left">Hoxa1</td>
<td align="left">ENSMUSG00000029844</td>
</tr>
<tr class="odd">
<td align="left"><code>ENSDARG</code></td>
<td align="left">Danio rario (zebrafish)</td>
<td align="left">hoxa1a</td>
<td align="left">ENSDARG00000104307</td>
</tr>
<tr class="even">
<td align="left"><code>ENSFCAG</code></td>
<td align="left">Felis catus (cat)</td>
<td align="left">HOXA1</td>
<td align="left">ENSFCAG00000007937</td>
</tr>
<tr class="odd">
<td align="left"><code>ENSMMUG</code></td>
<td align="left">Macaca mulata (macaque)</td>
<td align="left">HOXA1</td>
<td align="left">ENSMMUG00000012807</td>
</tr>
</tbody>
</table></div>
</div>
<p>Ensembl Gene IDs and gene symbols are the most commonly used gene identifiers.
The <a href="https://www.gencodegenes.org/">GENCODE project</a> which provides consistent
and stable genome annotation releases for human and mouse genomes, uses these
two types of identifiers with its official annotations.</p>
<div class="box note">
<p>Ensembl provides stable identifiers for transcripts as well as genes. Transcript
identifiers correspond to distinct patterns of exons/introns that have been
identified for a gene, and each gene has one or more distinct transcripts.
Ensembl Transcript IDs have the form <code>ENSTNNNNNNNNNNN.NN</code> similar to Gene IDs.</p>
</div>
<div class="box note">
<p>Ensembl is not the only gene identifier system besides gene symbol. Several
other databases have devised and maintain their own identifiers, most notably
[Entrez Gene IDs](<a href="">UCSC Gene IDs</a> used by the <a href="https://www.ncbi.nlm.nih.gov/gene">NCBI Gene
database</a> which assigns a unique integer to
each gene in each organism, and the <a href="https://www.omim.org/">Online Mendelian Inheritance in Man
(OMIM)</a> database, which has identifiers that look like
<code>OMIM:NNNNN</code>, where each OMIM ID refers to a unique gene or human phenotype.
However, the primary identifiers used in modern human biological research remain
Ensembl IDs and official HGNC gene symbols.</p>
</div>
</div>
<div id="mapping-between-identifier-systems" class="section level3" number="8.6.2">
<h3>
<span class="header-section-number">8.6.2</span> Mapping Between Identifier Systems<a class="anchor" aria-label="anchor" href="#mapping-between-identifier-systems"><i class="fas fa-link"></i></a>
</h3>
<p>A very common operation in biological analysis is to map between gene identifier
systems, e.g. you are given Ensembl Gene ID and want to map to the more
human-recognizable gene symbols. Ensembl provides a service called
<a href="https://www.ensembl.org/biomart/martview">BioMart</a> that allows you to download
annotation information for genes, transcripts, and other information maintained
in their databases. It also provides limited access to external sources of
information on their genes, including cross references to HGNC gene symbols and
some other gene identifier systems. The Ensembl website has <a href="https://useast.ensembl.org/info/data/biomart/index.html">helpful
documentation</a> on how
to use BioMart to download annotation data using the web interface.</p>
<p>In addition to downloading annotations as a CSV file from the web interface,
Ensembl also provides <a href="https://useast.ensembl.org/info/data/biomart/biomart_r_package.html">the biomaRt Bioconductor
package</a> to
allow programmatic access to the same information directly from R. There is <a href="https://useast.ensembl.org/info/genome/index.html">much
more information in the Ensembl
databases</a> than gene
annotation data that can be accessed via BioMart, but we will provide a brief
example of how to extract gene information here:</p>
<div class="sourceCode" id="cb377"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb377-1"><a href="biology-bioinformatics.html#cb377-1" tabindex="-1"></a><span class="co"># this assumes biomaRt is already installed through bioconductor</span></span>
<span id="cb377-2"><a href="biology-bioinformatics.html#cb377-2" tabindex="-1"></a><span class="fu">library</span>(biomaRt)</span>
<span id="cb377-3"><a href="biology-bioinformatics.html#cb377-3" tabindex="-1"></a></span>
<span id="cb377-4"><a href="biology-bioinformatics.html#cb377-4" tabindex="-1"></a><span class="co"># the human biomaRt database is named "hsapiens_gene_ensembl"</span></span>
<span id="cb377-5"><a href="biology-bioinformatics.html#cb377-5" tabindex="-1"></a>ensembl <span class="ot">&lt;-</span> <span class="fu">useEnsembl</span>(<span class="at">biomart=</span><span class="st">"ensembl"</span>, <span class="at">dataset=</span><span class="st">"hsapiens_gene_ensembl"</span>)</span>
<span id="cb377-6"><a href="biology-bioinformatics.html#cb377-6" tabindex="-1"></a></span>
<span id="cb377-7"><a href="biology-bioinformatics.html#cb377-7" tabindex="-1"></a><span class="co"># listAttributes() returns a data frame, turn into a tibble to help with formatting</span></span>
<span id="cb377-8"><a href="biology-bioinformatics.html#cb377-8" tabindex="-1"></a><span class="fu">as_tibble</span>(<span class="fu">listAttributes</span>(ensembl))</span>
<span id="cb377-9"><a href="biology-bioinformatics.html#cb377-9" tabindex="-1"></a><span class="co"># A tibble: 3,143 x 3</span></span>
<span id="cb377-10"><a href="biology-bioinformatics.html#cb377-10" tabindex="-1"></a>   name                          description                  page        </span>
<span id="cb377-11"><a href="biology-bioinformatics.html#cb377-11" tabindex="-1"></a>   <span class="sc">&lt;</span>chr<span class="sc">&gt;</span>                         <span class="er">&lt;</span>chr<span class="sc">&gt;</span>                        <span class="er">&lt;</span>chr<span class="sc">&gt;</span>       </span>
<span id="cb377-12"><a href="biology-bioinformatics.html#cb377-12" tabindex="-1"></a> <span class="dv">1</span> ensembl_gene_id               Gene stable ID               feature_page</span>
<span id="cb377-13"><a href="biology-bioinformatics.html#cb377-13" tabindex="-1"></a> <span class="dv">2</span> ensembl_gene_id_version       Gene stable ID version       feature_page</span>
<span id="cb377-14"><a href="biology-bioinformatics.html#cb377-14" tabindex="-1"></a> <span class="dv">3</span> ensembl_transcript_id         Transcript stable ID         feature_page</span>
<span id="cb377-15"><a href="biology-bioinformatics.html#cb377-15" tabindex="-1"></a> <span class="dv">4</span> ensembl_transcript_id_version Transcript stable ID version feature_page</span>
<span id="cb377-16"><a href="biology-bioinformatics.html#cb377-16" tabindex="-1"></a> <span class="dv">5</span> ensembl_peptide_id            Protein stable ID            feature_page</span>
<span id="cb377-17"><a href="biology-bioinformatics.html#cb377-17" tabindex="-1"></a> <span class="dv">6</span> ensembl_peptide_id_version    Protein stable ID version    feature_page</span>
<span id="cb377-18"><a href="biology-bioinformatics.html#cb377-18" tabindex="-1"></a> <span class="dv">7</span> ensembl_exon_id               Exon stable ID               feature_page</span>
<span id="cb377-19"><a href="biology-bioinformatics.html#cb377-19" tabindex="-1"></a> <span class="dv">8</span> description                   Gene description             feature_page</span>
<span id="cb377-20"><a href="biology-bioinformatics.html#cb377-20" tabindex="-1"></a> <span class="dv">9</span> chromosome_name               Chromosome<span class="sc">/</span>scaffold name     feature_page</span>
<span id="cb377-21"><a href="biology-bioinformatics.html#cb377-21" tabindex="-1"></a><span class="dv">10</span> start_position                Gene <span class="fu">start</span> (bp)              feature_page</span>
<span id="cb377-22"><a href="biology-bioinformatics.html#cb377-22" tabindex="-1"></a><span class="co"># ... with 3,133 more rows</span></span></code></pre></div>
<p>The <code>name</code> column provides the programmatic name associated with the attribute
that can be used to retrieve that annotation information using the <code>getBM()</code>
function:</p>
<div class="sourceCode" id="cb378"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb378-1"><a href="biology-bioinformatics.html#cb378-1" tabindex="-1"></a><span class="co"># create a tibble with ensembl gene ID, HGNC gene symbol, and gene description</span></span>
<span id="cb378-2"><a href="biology-bioinformatics.html#cb378-2" tabindex="-1"></a>gene_map <span class="ot">&lt;-</span> <span class="fu">as_tibble</span>(</span>
<span id="cb378-3"><a href="biology-bioinformatics.html#cb378-3" tabindex="-1"></a>  <span class="fu">getBM</span>(</span>
<span id="cb378-4"><a href="biology-bioinformatics.html#cb378-4" tabindex="-1"></a>    <span class="at">attributes=</span><span class="fu">c</span>(<span class="st">"ensembl_gene_id"</span>, <span class="st">"hgnc_symbol"</span>, <span class="st">"description"</span>),</span>
<span id="cb378-5"><a href="biology-bioinformatics.html#cb378-5" tabindex="-1"></a>    <span class="at">mart=</span>ensembl</span>
<span id="cb378-6"><a href="biology-bioinformatics.html#cb378-6" tabindex="-1"></a>  )</span>
<span id="cb378-7"><a href="biology-bioinformatics.html#cb378-7" tabindex="-1"></a>)</span>
<span id="cb378-8"><a href="biology-bioinformatics.html#cb378-8" tabindex="-1"></a>gene_map</span>
<span id="cb378-9"><a href="biology-bioinformatics.html#cb378-9" tabindex="-1"></a><span class="co"># A tibble: 68,012 x 3</span></span>
<span id="cb378-10"><a href="biology-bioinformatics.html#cb378-10" tabindex="-1"></a>   ensembl_gene_id hgnc_symbol description                                                               </span>
<span id="cb378-11"><a href="biology-bioinformatics.html#cb378-11" tabindex="-1"></a>   <span class="sc">&lt;</span>chr<span class="sc">&gt;</span>           <span class="er">&lt;</span>chr<span class="sc">&gt;</span>       <span class="er">&lt;</span>chr<span class="sc">&gt;</span>                                                                     </span>
<span id="cb378-12"><a href="biology-bioinformatics.html#cb378-12" tabindex="-1"></a> <span class="dv">1</span> ENSG00000210049 MT<span class="sc">-</span>TF       mitochondrially encoded tRNA<span class="sc">-</span><span class="fu">Phe</span> (UUU<span class="sc">/</span>C) [Source<span class="sc">:</span>HGNC Symbol;Acc<span class="sc">:</span>HGNC<span class="sc">:</span><span class="dv">748</span><span class="sc">~</span></span>
<span id="cb378-13"><a href="biology-bioinformatics.html#cb378-13" tabindex="-1"></a> <span class="dv">2</span> ENSG00000211459 MT<span class="sc">-</span>RNR1     mitochondrially encoded 12S rRNA [Source<span class="sc">:</span>HGNC Symbol;Acc<span class="sc">:</span>HGNC<span class="sc">:</span><span class="dv">7470</span>]       </span>
<span id="cb378-14"><a href="biology-bioinformatics.html#cb378-14" tabindex="-1"></a> <span class="dv">3</span> ENSG00000210077 MT<span class="sc">-</span>TV       mitochondrially encoded tRNA<span class="sc">-</span><span class="fu">Val</span> (GUN) [Source<span class="sc">:</span>HGNC Symbol;Acc<span class="sc">:</span>HGNC<span class="sc">:</span><span class="dv">7500</span>]</span>
<span id="cb378-15"><a href="biology-bioinformatics.html#cb378-15" tabindex="-1"></a> <span class="dv">4</span> ENSG00000210082 MT<span class="sc">-</span>RNR2     mitochondrially encoded 16S rRNA [Source<span class="sc">:</span>HGNC Symbol;Acc<span class="sc">:</span>HGNC<span class="sc">:</span><span class="dv">7471</span>]       </span>
<span id="cb378-16"><a href="biology-bioinformatics.html#cb378-16" tabindex="-1"></a> <span class="dv">5</span> ENSG00000209082 MT<span class="sc">-</span>TL1      mitochondrially encoded tRNA<span class="sc">-</span><span class="fu">Leu</span> (UUA<span class="sc">/</span>G) <span class="dv">1</span> [Source<span class="sc">:</span>HGNC Symbol;Acc<span class="sc">:</span>HGNC<span class="sc">:</span><span class="dv">7</span><span class="sc">~</span></span>
<span id="cb378-17"><a href="biology-bioinformatics.html#cb378-17" tabindex="-1"></a> <span class="dv">6</span> ENSG00000198888 MT<span class="sc">-</span>ND1      mitochondrially encoded NADH<span class="sc">:</span>ubiquinone oxidoreductase core subunit <span class="dv">1</span> [So<span class="sc">~</span></span>
<span id="cb378-18"><a href="biology-bioinformatics.html#cb378-18" tabindex="-1"></a> <span class="dv">7</span> ENSG00000210100 MT<span class="sc">-</span>TI       mitochondrially encoded tRNA<span class="sc">-</span><span class="fu">Ile</span> (AUU<span class="sc">/</span>C) [Source<span class="sc">:</span>HGNC Symbol;Acc<span class="sc">:</span>HGNC<span class="sc">:</span><span class="dv">748</span><span class="sc">~</span></span>
<span id="cb378-19"><a href="biology-bioinformatics.html#cb378-19" tabindex="-1"></a> <span class="dv">8</span> ENSG00000210107 MT<span class="sc">-</span>TQ       mitochondrially encoded tRNA<span class="sc">-</span><span class="fu">Gln</span> (CAA<span class="sc">/</span>G) [Source<span class="sc">:</span>HGNC Symbol;Acc<span class="sc">:</span>HGNC<span class="sc">:</span><span class="dv">749</span><span class="sc">~</span></span>
<span id="cb378-20"><a href="biology-bioinformatics.html#cb378-20" tabindex="-1"></a> <span class="dv">9</span> ENSG00000210112 MT<span class="sc">-</span>TM       mitochondrially encoded tRNA<span class="sc">-</span><span class="fu">Met</span> (AUA<span class="sc">/</span>G) [Source<span class="sc">:</span>HGNC Symbol;Acc<span class="sc">:</span>HGNC<span class="sc">:</span><span class="dv">749</span><span class="sc">~</span></span>
<span id="cb378-21"><a href="biology-bioinformatics.html#cb378-21" tabindex="-1"></a><span class="dv">10</span> ENSG00000198763 MT<span class="sc">-</span>ND2      mitochondrially encoded NADH<span class="sc">:</span>ubiquinone oxidoreductase core subunit <span class="dv">2</span> [So<span class="sc">~</span></span>
<span id="cb378-22"><a href="biology-bioinformatics.html#cb378-22" tabindex="-1"></a><span class="co"># ... with 68,002 more rows</span></span></code></pre></div>
<p>With this Ensembl Gene ID to HGNC symbol mapping in hand, you can combine the
tibble above with other tibbles containing gene information by <a href="data-wrangling.html#relational-data">joining them
together</a>.</p>
<div class="box note">
<p>BioMart/biomaRt is not the only ways to map different gene identifiers. The
Bioconductor package
<a href="https://bioconductor.org/packages/release/bioc/html/AnnotationDbi.html">AnnotateDbi</a>
also provides this functionality in a flexible format independent of the Ensembl
databases. This package includes not only gene identifier mapping and
information, but also identifiers from technology platforms, e.g. probe set IDs
from <a href="biology-bioinformatics.html#microarrays">microarrays</a>, that can help when working with these types of
data. The package also allows comprehensive and flexible <a href="biology-bioinformatics.html#mapping-homologs">homolog
mapping</a>. As with all Bioconductor packages, the
AnnotationDbi documentation is well written and comprehensive, though knowledge
of <a href="https://en.wikipedia.org/wiki/Relational_database">relational databases</a> is
helpful in understanding how the packages work.</p>
</div>
<div class="box readmore">
<ul>
<li><a href="https://www.ensembl.org/biomart/martview">Ensembl BioMart portal</a></li>
<li><a href="https://useast.ensembl.org/info/data/biomart/index.html">Ensembl BioMart web portal documentation</a></li>
<li><a href="http://www.bioconductor.org/packages/release/bioc/html/biomaRt.html">biomaRt Bioconductor documentation</a></li>
</ul>
</div>
</div>
<div id="mapping-homologs" class="section level3" number="8.6.3">
<h3>
<span class="header-section-number">8.6.3</span> Mapping Homologs<a class="anchor" aria-label="anchor" href="#mapping-homologs"><i class="fas fa-link"></i></a>
</h3>
<p>Sometimes it is necessary to link datasets from different organisms together by
orthology. For example, in an experiment performed in mice, we might be
interested in comparing gene expression patterns observed in the mouse samples
to a publicly available human dataset. In these contexts, we must link gene
identifiers from one organism to their corresponding homologs in the other.
BioMart enables us to extract these linked identifiers by explicitly <a href="https://bioconductor.org/packages/release/bioc/vignettes/biomaRt/inst/doc/accessing_ensembl.html#given-the-human-gene-tp53-retrieve-the-human-chromosomal-location-of-this-gene-and-also-retrieve-the-chromosomal-location-and-refseq-id-of-its-homolog-in-mouse.">connecting
different biomaRt
databases</a>
with the <code>getLDS()</code> function:</p>
<div class="sourceCode" id="cb379"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb379-1"><a href="biology-bioinformatics.html#cb379-1" tabindex="-1"></a>human_db <span class="ot">&lt;-</span> <span class="fu">useEnsembl</span>(<span class="st">"ensembl"</span>, <span class="at">dataset =</span> <span class="st">"hsapiens_gene_ensembl"</span>)</span>
<span id="cb379-2"><a href="biology-bioinformatics.html#cb379-2" tabindex="-1"></a>mouse_db <span class="ot">&lt;-</span> <span class="fu">useEnsembl</span>(<span class="st">"ensembl"</span>, <span class="at">dataset =</span> <span class="st">"mmusculus_gene_ensembl"</span>)</span>
<span id="cb379-3"><a href="biology-bioinformatics.html#cb379-3" tabindex="-1"></a>orth_map <span class="ot">&lt;-</span> <span class="fu">as_tibble</span>(</span>
<span id="cb379-4"><a href="biology-bioinformatics.html#cb379-4" tabindex="-1"></a>  <span class="fu">getLDS</span>(<span class="at">attributes =</span> <span class="fu">c</span>(<span class="st">"ensembl_gene_id"</span>, <span class="st">"hgnc_symbol"</span>),</span>
<span id="cb379-5"><a href="biology-bioinformatics.html#cb379-5" tabindex="-1"></a>       <span class="at">mart =</span> human_db,</span>
<span id="cb379-6"><a href="biology-bioinformatics.html#cb379-6" tabindex="-1"></a>       <span class="at">attributesL =</span> <span class="fu">c</span>(<span class="st">"ensembl_gene_id"</span>, <span class="st">"mgi_symbol"</span>),</span>
<span id="cb379-7"><a href="biology-bioinformatics.html#cb379-7" tabindex="-1"></a>       <span class="at">martL =</span> mouse_db</span>
<span id="cb379-8"><a href="biology-bioinformatics.html#cb379-8" tabindex="-1"></a>  )</span>
<span id="cb379-9"><a href="biology-bioinformatics.html#cb379-9" tabindex="-1"></a>)</span>
<span id="cb379-10"><a href="biology-bioinformatics.html#cb379-10" tabindex="-1"></a>orth_map</span>
<span id="cb379-11"><a href="biology-bioinformatics.html#cb379-11" tabindex="-1"></a><span class="co"># A tibble: 26,390 x 4</span></span>
<span id="cb379-12"><a href="biology-bioinformatics.html#cb379-12" tabindex="-1"></a>   Gene.stable.ID  HGNC.symbol Gene.stable.ID<span class="fl">.1</span>   MGI.symbol</span>
<span id="cb379-13"><a href="biology-bioinformatics.html#cb379-13" tabindex="-1"></a>   <span class="sc">&lt;</span>chr<span class="sc">&gt;</span>           <span class="er">&lt;</span>chr<span class="sc">&gt;</span>       <span class="er">&lt;</span>chr<span class="sc">&gt;</span>              <span class="er">&lt;</span>chr<span class="sc">&gt;</span>     </span>
<span id="cb379-14"><a href="biology-bioinformatics.html#cb379-14" tabindex="-1"></a> <span class="dv">1</span> ENSG00000198695 MT<span class="sc">-</span>ND6      ENSMUSG00000064368 <span class="st">"mt-Nd6"</span>  </span>
<span id="cb379-15"><a href="biology-bioinformatics.html#cb379-15" tabindex="-1"></a> <span class="dv">2</span> ENSG00000212907 MT<span class="sc">-</span>ND4L     ENSMUSG00000065947 <span class="st">"mt-Nd4l"</span></span>
<span id="cb379-16"><a href="biology-bioinformatics.html#cb379-16" tabindex="-1"></a> <span class="dv">3</span> ENSG00000279169 PRAMEF13    ENSMUSG00000094303 <span class="st">""</span>        </span>
<span id="cb379-17"><a href="biology-bioinformatics.html#cb379-17" tabindex="-1"></a> <span class="dv">4</span> ENSG00000279169 PRAMEF13    ENSMUSG00000094722 <span class="st">""</span>        </span>
<span id="cb379-18"><a href="biology-bioinformatics.html#cb379-18" tabindex="-1"></a> <span class="dv">5</span> ENSG00000279169 PRAMEF13    ENSMUSG00000095666 <span class="st">""</span>        </span>
<span id="cb379-19"><a href="biology-bioinformatics.html#cb379-19" tabindex="-1"></a> <span class="dv">6</span> ENSG00000279169 PRAMEF13    ENSMUSG00000094741 <span class="st">""</span>        </span>
<span id="cb379-20"><a href="biology-bioinformatics.html#cb379-20" tabindex="-1"></a> <span class="dv">7</span> ENSG00000279169 PRAMEF13    ENSMUSG00000094836 <span class="st">""</span>        </span>
<span id="cb379-21"><a href="biology-bioinformatics.html#cb379-21" tabindex="-1"></a> <span class="dv">8</span> ENSG00000279169 PRAMEF13    ENSMUSG00000074720 <span class="st">""</span>        </span>
<span id="cb379-22"><a href="biology-bioinformatics.html#cb379-22" tabindex="-1"></a> <span class="dv">9</span> ENSG00000279169 PRAMEF13    ENSMUSG00000096236 <span class="st">""</span>        </span>
<span id="cb379-23"><a href="biology-bioinformatics.html#cb379-23" tabindex="-1"></a><span class="dv">10</span> ENSG00000198763 MT<span class="sc">-</span>ND2      ENSMUSG00000064345 <span class="st">"mt-Nd2"</span>  </span>
<span id="cb379-24"><a href="biology-bioinformatics.html#cb379-24" tabindex="-1"></a><span class="co"># ... with 26,380 more rows</span></span></code></pre></div>
<p>The <code>mgi_symbol</code> field refers to the gene symbol assigned in the <a href="https://www.jax.org/research-and-faculty/resources/mouse-genome-informatics">Mouse Genome
Informatics database</a>
maintained by <a href="https://www.jax.org/">The Jackson Laboratory</a>.</p>

</div>
</div>
<div id="gene-expression" class="section level2" number="8.7">
<h2>
<span class="header-section-number">8.7</span> Gene Expression<a class="anchor" aria-label="anchor" href="#gene-expression"><i class="fas fa-link"></i></a>
</h2>
<p>Gene expression is the process by which information from a gene is used in the
synthesis of functional gene products that affect a phenotype. While gene
expression studies are often focused on protein coding genes, there are many
other functional molecules, e.g. <a href="https://www.genome.gov/genetics-glossary/Transfer-RNA">transfer
RNAs</a>,
<a href="https://en.wikipedia.org/wiki/Long_non-coding_RNA">lncRNAs</a>, etc, that are
produced by this process. The gene expression process has many steps and
intermediate products, as depicted in the following figure:</p>
<div class="float">
<img src="gene_expression_nature.jpg" alt="The Gene Expression Process - An overview of the flow of information from DNA to protein in a eukaryote"><div class="figcaption">The Gene Expression Process - An overview of the flow of information from DNA
to protein in a eukaryote</div>
</div>
<p>The specific parts of the genome that code for genes are copied, or
<em>transcribed</em> into RNA molecules called <em>transcripts</em>. In lower organisms like
bacteria, these RNA molecules are passed on directly to ribosomes, which
translate them into proteins. In most higher organisms like eukaryotes, these
initial transcripts, called <em>pre-messenger RNA (pre-mRNA) transcripts</em> are
further processed such that certain parts of the transcript, called <em>introns</em>,
are spliced out and the flanking sequences, called <em>exons</em>, are concatenated
together. After this splicing process is complete, the pre-mRNA transcripts
become mature <em>messenger RNA (mRNA) transcripts</em> which go on to be exported from
the nucleus and loaded into ribosomes in the cytoplasm to undergo translation
into proteins.</p>
<p>In gene expression studies, the relative abundance, or number of copies, of RNA
or mRNA transcripts in a sample is measured. The measurements are non-negative
numbers that are proportional to the relative abundance of a transcript with
respect to some reference, either another gene, or in the case of high
throughput assays like microarrays or high throughput sequencing, relative to
measurements of all distinct transcripts examined in the experiment.
Conceptually, the larger the magnitude of a transcript abundance measurement,
the more copies of that transcript were in the original sample.</p>
<p>There are many ways to measure the abundance of RNA transcripts; the following
are some of the common methods at the time of writing:</p>
<ul>
<li>Light absorbance - RNA absorbs ultraviolet light with wavelength 260 nm, which
can be used to determine RNA concentration in a sample using specialized equipment</li>
<li>
<a href="https://www.genome.gov/genetics-glossary/Northern-Blot">Northern blot</a> -
measures relative abundance of an RNA with a specific sequence</li>
<li>
<a href="https://www.thermofisher.com/blog/ask-a-scientist/what-is-qpcr/">quantitative polymerase chain reaction
(qPCR)</a> -
measures relative abundance of an RNA with a specific sequence using <a href="https://en.wikipedia.org/wiki/Polymerase_chain_reaction">PCR
amplification</a>
</li>
<li>
<a href="biology-bioinformatics.html#microarrays">Oligonucleotide and microarrays</a> - measures relative abundance
of thousands of genes simultaneously using known DNA probe sequences and
fluorescently tagged RNA molecules</li>
<li>
<a href="biology-bioinformatics.html#rnaseq">High throughput RNA sequencing (RNASeq)</a> - measures thousands to
millions of RNA fragments simultaneously in proportion to their relative
abundance</li>
</ul>
<p>While any of the measurement methods above may be analyzed in R, the high
throughput methods (i.e. microarray, high throughput sequencing) are the primary
concern of this chapter. These methods generate measurements for thousands of
transcripts or genes simultaneously, requiring the power and flexibility of
programmatic analysis to process in a practical amount of time. The remainder of
this chapter is devoted to understanding the specific technologies, data types,
and analytical strategies involved in working with these data.</p>
<div class="box important">
<p>Gene expression measurements are almost always inherently relative, either due
to limitations of the measurement methods (e.g. microarrays, described below) or
because measuring all molecules in a sample will almost always be prohibitively
expensive, have diminishing returns, and it is very difficult if not impossible
to determine if all the molecules have been measured. This means we cannot in
general associate the numbers associated with the measurements with an absolute
molecular copy number. An important implication of the inherent relativity of
these measurements is: <strong>absence of evidence is not evidence of absence</strong>. In
other words, if a transcript has an abundance measurement of zero, this does
<em>not necessarily imply that the gene is not expressed</em>. It may be that the gene
is indeed expressed, but the copy number is sufficiently small that it was not
detected by the assay.</p>
</div>
<div id="gene-expression-data-in-bioconductor" class="section level3" number="8.7.1">
<h3>
<span class="header-section-number">8.7.1</span> Gene Expression Data in Bioconductor<a class="anchor" aria-label="anchor" href="#gene-expression-data-in-bioconductor"><i class="fas fa-link"></i></a>
</h3>
<p>The <a href="https://bioconductor.org/packages/release/bioc/html/SummarizedExperiment.html">SummarizedExperiment
container</a>
is the standard way to load and work with gene expression data in Bioconductor.
This container requires the following information:</p>
<ul>
<li>
<code>assays</code> - one or more measurement assays (e.g. gene expression) in the form of a feature by sample matrix</li>
<li>
<code>colData</code> - metadata associated with the samples (i.e. columns) of the assays</li>
<li>
<code>rowData</code> - metadata associated with the features (i.e. rows) of the assays</li>
<li>
<code>exptData</code> - additional metadata about the experiment itself, like protocol, project name, etc</li>
</ul>
<p>The figure below illustrates how the SummarizedExperiment container is
structured and how the different data elements are accessed:</p>
<div class="float">
<img src="https://media.springernature.com/full/springer-static/image/art%3A10.1038%2Fnmeth.3252/MediaObjects/41592_2015_Article_BFnmeth3252_Fig2_HTML.jpg?as=webp" alt="SummarizedExperiment Schematic - Huber, et al. 2015. “Orchestrating High-Throughput Genomic Analysis with Bioconductor.” Nature Methods 12 (2): 115–21."><div class="figcaption">SummarizedExperiment Schematic - Huber, et al. 2015. “Orchestrating High-Throughput Genomic Analysis with Bioconductor.” Nature Methods 12 (2): 115–21.</div>
</div>
<p>Many Bioconductor packages for specific types of data, e.g.
<a href="#differential-expression--microarrays-limma">limma</a> create these
SummarizedExperiment objects for you, but you may also create your own by
assembling each of these data into data frames manually:</p>
<div class="sourceCode" id="cb380"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># microarray expression dataset intensities</span></span>
<span><span class="va">intensities</span> <span class="op">&lt;-</span> <span class="fu">readr</span><span class="fu">::</span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html">read_delim</a></span><span class="op">(</span><span class="st">"example_intensity_data.csv"</span>,delim<span class="op">=</span><span class="st">" "</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># the first column of intensities tibble is the probesetId, extract to pass as rowData</span></span>
<span><span class="va">rowData</span> <span class="op">&lt;-</span> <span class="va">intensities</span><span class="op">[</span><span class="st">"probeset_id"</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># remove probeset IDs from tibble and turn into a R dataframe so that we can assign rownames</span></span>
<span><span class="co"># since tibbles don't support row names</span></span>
<span><span class="va">intensities</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="va">intensities</span>, <span class="op">-</span><span class="va">probeset_id</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">intensities</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">rowData</span><span class="op">$</span><span class="va">probeset_id</span></span>
<span></span>
<span><span class="co"># these column data are made up, but you would have a sample metadata file to use</span></span>
<span><span class="va">colData</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html">tibble</a></span><span class="op">(</span></span>
<span>  sample_name<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">intensities</span><span class="op">)</span>,</span>
<span>  condition<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"case"</span>,<span class="st">"control"</span><span class="op">)</span>,<span class="fu"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span><span class="op">(</span><span class="va">intensities</span><span class="op">)</span>,replace<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">se</span> <span class="op">&lt;-</span> <span class="fu">SummarizedExperiment</span><span class="op">(</span></span>
<span>   assays<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>intensities<span class="op">=</span><span class="va">intensities</span><span class="op">)</span>,</span>
<span>   colData<span class="op">=</span><span class="va">colData</span>,</span>
<span>   rowData<span class="op">=</span><span class="va">rowData</span></span>
<span><span class="op">)</span></span>
<span><span class="va">se</span></span></code></pre></div>
<pre><code>## class: SummarizedExperiment
## dim: 54675 35
## metadata(0):
## assays(1): intensities
## rownames(54675): 1007_s_at 1053_at ... AFFX-TrpnX-5_at AFFX-TrpnX-M_at
## rowData names(1): probeset_id
## colnames(35): GSM972389 GSM972390 ... GSM972512 GSM972521
## colData names(2): sample_name condition</code></pre>
<p>Detailed documentation of how to create and use the SummarizedExperiment is available in the <a href="https://bioconductor.org/packages/release/bioc/vignettes/SummarizedExperiment/inst/doc/SummarizedExperiment.html#constructing-a-summarizedexperiment">SummarizedExperiment vignette</a>.</p>
<div class="box note">
<p>SummarizedExperiment is the successor to the older
<a href="https://www.bioconductor.org/packages/devel/bioc/vignettes/Biobase/inst/doc/ExpressionSetIntroduction.pdf">ExpressionSet</a>
container. Both are still used by Bioconductor packages, but
SummarizedExperiment is more modern and flexible, so it is suggested for use
whenever possible.</p>
</div>
</div>
<div id="differential-expression-analysis" class="section level3" number="8.7.2">
<h3>
<span class="header-section-number">8.7.2</span> Differential Expression Analysis<a class="anchor" aria-label="anchor" href="#differential-expression-analysis"><i class="fas fa-link"></i></a>
</h3>
<p>Differential expression analysis seeks to identify to what extent gene
expression is associated with one or more variables of interest. For example,
we might be interested in genes that have higher expression in subjects with a
disease, or which genes change in response to a treatment.</p>
<p>Typically, gene expression analysis requires two types of data to run: an
expression matrix and a <em>design matrix</em>. The expression matrix will generally
have features (e.g. genes) as rows and samples as columns. The design matrix is
a numeric matrix that contains the variables we wish to model and any covariates
or confounders we wish to adjust for. The variables in the design matrix then
must be encoded in a way that statistical procedures can understand. The full
details of how design matrices are constructed is beyond the scope of this book.
Fortunately, R makes it very easy to construct these matrices from a tibble with
the <a href="https://www.rdocumentation.org/packages/stats/versions/3.6.2/topics/model.matrix"><code>model.matrix()</code>
function</a>.
Consider for the following imaginary sample information tibble that has 10 AD
and 10 Control subjects with different clinical and protein histology
measurements assayed on their brain tissue:</p>
<div class="sourceCode" id="cb382"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ad_metadata</span></span></code></pre></div>
<pre><code>## # A tibble: 20 x 8
##    ID    age_at_death condition    tau  abeta   iba1   gfap braak_stage
##    &lt;chr&gt;        &lt;dbl&gt; &lt;fct&gt;      &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt; &lt;fct&gt;      
##  1 A1              81 AD         98543 166078  77299  75277 4          
##  2 A2              80 AD         32711  56919  49571 117441 1          
##  3 A3              91 AD        134427 295407 106724  67719 6          
##  4 A4              90 AD         81515 221801 171212  56346 4          
##  5 A5              80 AD         42442  97918   8321 109406 2          
##  6 A6              81 AD         54802  49358  82374  42797 2          
##  7 A7              77 AD        127166   8417  62675  85377 6          
##  8 A8              74 AD        103015  28478 131168 133892 5          
##  9 A9              82 AD        114270 188208 194818  61935 5          
## 10 A10             83 AD         90785  46279 121555 131853 4          
## 11 C1              79 Control      991    237   1062  64752 0          
## 12 C2              76 Control     2804   4423   3441  23384 0          
## 13 C3              73 Control     3120   7073   1402 135049 0          
## 14 C4              73 Control     7167    798   8365  18961 0          
## 15 C5              69 Control     6591   2590   4885  65094 0          
## 16 C6              77 Control    38697 103664  33430  23136 2          
## 17 C7              73 Control    58381  72071  16864  17295 3          
## 18 C8              75 Control     5102  10533   5321  89479 0          
## 19 C9              76 Control    59906  47967  19103 124651 3          
## 20 C10             81 Control    46553  49670  90763  52684 2</code></pre>
<p>We might be interested in identifying genes that are increased or decreased in
people with AD compared with Controls. We can create a model matrix for this as
follows:</p>
<div class="sourceCode" id="cb384"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html">model.matrix</a></span><span class="op">(</span><span class="op">~</span> <span class="va">condition</span>, data<span class="op">=</span><span class="va">ad_metadata</span><span class="op">)</span></span></code></pre></div>
<pre><code>##    (Intercept) conditionAD
## 1            1           1
## 2            1           1
## 3            1           1
## 4            1           1
## 5            1           1
## 6            1           1
## 7            1           1
## 8            1           1
## 9            1           1
## 10           1           1
## 11           1           0
## 12           1           0
## 13           1           0
## 14           1           0
## 15           1           0
## 16           1           0
## 17           1           0
## 18           1           0
## 19           1           0
## 20           1           0
## attr(,"assign")
## [1] 0 1
## attr(,"contrasts")
## attr(,"contrasts")$condition
## [1] "contr.treatment"</code></pre>
<p>The <code>~ condition</code> argument is an an R
<a href="https://rviews.rstudio.com/2017/02/01/the-r-formula-method-the-good-parts/">forumla</a>,
which is a concise description of how variable should be included in the model.
The general format of a formula is as follows:</p>
<pre><code># portions in [] are optional
[&lt;outcome variable&gt;] ~ &lt;predictor variable 1&gt; [+ &lt;predictor variable 2&gt;]...

# examples

# model Gene 3 expression as a function of disease status
`Gene 3` ~ condition

# model the amount of tau pathology as a function of abeta pathology,
# adjusting for age at death
tau ~ age_at_death + abeta

# create a model without an outcome variable that can be used to create the
# model matrix to test for differences with disease status adjusting for age at death
~ age_at_death + condition</code></pre>
<p>For most models, the design matrix will have an intercept term of all ones and
additional colums for the other variables in the model. We might wish to adjust
out the effect of age by including <code>age_at_death</code> as a covariate in the model:</p>
<div class="sourceCode" id="cb387"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html">model.matrix</a></span><span class="op">(</span><span class="op">~</span> <span class="va">age_at_death</span> <span class="op">+</span> <span class="va">condition</span>, data<span class="op">=</span><span class="va">ad_metadata</span><span class="op">)</span></span></code></pre></div>
<pre><code>##    (Intercept) age_at_death conditionAD
## 1            1           81           1
## 2            1           80           1
## 3            1           91           1
## 4            1           90           1
## 5            1           80           1
## 6            1           81           1
## 7            1           77           1
## 8            1           74           1
## 9            1           82           1
## 10           1           83           1
## 11           1           79           0
## 12           1           76           0
## 13           1           73           0
## 14           1           73           0
## 15           1           69           0
## 16           1           77           0
## 17           1           73           0
## 18           1           75           0
## 19           1           76           0
## 20           1           81           0
## attr(,"assign")
## [1] 0 1 2
## attr(,"contrasts")
## attr(,"contrasts")$condition
## [1] "contr.treatment"</code></pre>
<p>The model matrix now includes a column for age at death as well. This model
model matrix is now suitable to pass to differential expression software
packages to look for genes associated with disease status.</p>
<p>As mentioned above, modern gene expression assays measure thousands of genes
simultaneously. This means each gene must be tested for differential expression
individually. In general, each gene is tested using the same statistical model,
so a differential expression analysis package will perform something equivalent
to the following:</p>
<pre><code>Gene 1 ~ age_at_death + condition
Gene 2 ~ age_at_death + condition
Gene 3 ~ age_at_death + condition
...
Gene N ~ age_at_death + condition</code></pre>
<p>Each gene model will have its own statistics associated with it that we must
process and interpret after the analysis is complete.</p>
</div>
<div id="microarray-gene-expression-data" class="section level3" number="8.7.3">
<h3>
<span class="header-section-number">8.7.3</span> Microarray Gene Expression Data<a class="anchor" aria-label="anchor" href="#microarray-gene-expression-data"><i class="fas fa-link"></i></a>
</h3>
<p>On a high level, there are four steps involved in analyzing gene expression
microarray data:</p>
<ol style="list-style-type: decimal">
<li>
<strong>Summarization of probes to probesets.</strong> Each gene is represented by
multiple probes with different sequences. Summarization is a statistical
procedure that computes a single value for each probeset from its corresponding
probes.</li>
<li>
<strong>Normalization.</strong> This includes removing background signal from individual
arrays as well as adjusting probeset intensities so that they are comparable
across multiple sample arrays.</li>
<li>
<strong>Quality control.</strong> Compare all normalized samples within a sample set to
identify, mitigate, or eliminate potential outlier samples.</li>
<li>
<strong>Analysis.</strong> Using the quality controlled expression data, Implement
statistical analysis to answer research questions.</li>
</ol>
<p>The full details of microarray analysis are beyond the scope of this book.
However in the following sections we cover some of the basic entry points to
performing these steps in R and Bioconductor.</p>
<p>The CEL data files from a set of microarrays can be loaded into R for analysis
using the <a href="https://www.bioconductor.org/packages/release/bioc/html/affy.html"><code>affy</code> Bioconductor
package</a>.
This package provides all the functions necessary for loading the data and
performing key preprocessing operations. Typically, two or more samples were
processed in this way, resulting in a set of CEL files that should be processed
together. These CEL files will typically be all stored in the same directory,
and may be loaded using the <code><a href="https://rdrr.io/pkg/affy/man/read.affybatch.html">affy::ReadAffy</a></code> function:</p>
<div class="sourceCode" id="cb390"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># read all CEL files in a single directory</span></span>
<span><span class="va">affy_batch</span> <span class="op">&lt;-</span> <span class="fu">affy</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/affy/man/read.affybatch.html">ReadAffy</a></span><span class="op">(</span>celfile.path<span class="op">=</span><span class="st">"directory/of/CELfiles"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># or individual files in different directories</span></span>
<span><span class="va">cel_filenames</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/list.files.html">list.files</a></span><span class="op">(</span> path<span class="op">=</span><span class="st">"CELfile_dir1"</span>, full.names<span class="op">=</span><span class="cn">TRUE</span>, pattern<span class="op">=</span><span class="st">".CEL$"</span> <span class="op">)</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/list.files.html">list.files</a></span><span class="op">(</span> path<span class="op">=</span><span class="st">"CELfile_dir2"</span>, full.names<span class="op">=</span><span class="cn">TRUE</span>, pattern<span class="op">=</span><span class="st">".CEL$"</span> <span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">affy_batch</span> <span class="op">&lt;-</span> <span class="fu">affy</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/affy/man/read.affybatch.html">ReadAffy</a></span><span class="op">(</span>filenames<span class="op">=</span><span class="va">cel_filenames</span><span class="op">)</span></span></code></pre></div>
<p>The <code>affy_batch</code> variable is a <code>AffyBatch</code> container, which stores information
on the probe definitions based on the type of microarray, probe-level intensity
for each sample, and other information about the experiment contained within the
CEL files.</p>
<p>The <code>affy</code> package provides functions to perform probe summarization and
normalization. The most popular method to accomplish this at the time of writing
is called Robust Multi-array Average or RMA <span class="citation">(<a href="detailed-class-schedule.html#ref-Irizarry2003-hn">Irizarry et al. 2003</a>)</span>, which performs
summarization and normalization of multiple arrays simultaneously. Below, the
RMA algorithm is used to normalize an example dataset provided by the
<a href="https://bioconductor.org/packages/release/data/experiment/html/affydata.html"><code>affydata</code></a>
Bioconductor package.</p>
<div class="box note">
<p>Certain Bioconductor packages provide example datasets for use with companion
analysis packages. For example, the <code>affydata</code> package provides the <code>Dilution</code>
dataset, which was generated using two concentrations of cDNA from human liver
tissue and a central nervous system cell line. To load a data package into R,
first run <code>library(&lt;data package&gt;)</code> and then <code>data(&lt;dataset name&gt;)</code>.</p>
</div>
<div class="sourceCode" id="cb391"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">affy</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">affydata</span><span class="op">)</span></span></code></pre></div>
<pre><code>##      Package    LibPath                                     Item      
## [1,] "affydata" "C:/Users/Adam/Documents/R/win-library/4.1" "Dilution"
##      Title                        
## [1,] "AffyBatch instance Dilution"</code></pre>
<div class="sourceCode" id="cb393"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="va">Dilution</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># normalize the Dilution microarray expression values</span></span>
<span><span class="co"># note Dilution is an AffyBatch object</span></span>
<span><span class="va">eset_rma</span> <span class="op">&lt;-</span> <span class="fu">affy</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/affy/man/rma.html">rma</a></span><span class="op">(</span><span class="va">Dilution</span>,verbose<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># plot distribution of non-normalized probes</span></span>
<span><span class="co"># note rma normalization takes the log2 of the expression values,</span></span>
<span><span class="co"># so we must do so on the raw data to compare</span></span>
<span><span class="va">raw_intensity</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html">as_tibble</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/exprs.html">exprs</a></span><span class="op">(</span><span class="va">Dilution</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>probeset_id<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html">rownames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/exprs.html">exprs</a></span><span class="op">(</span><span class="va">Dilution</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">pivot_longer</span><span class="op">(</span><span class="op">-</span><span class="va">probeset_id</span>, names_to<span class="op">=</span><span class="st">"Sample"</span>, values_to <span class="op">=</span> <span class="st">"Intensity"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span></span>
<span>    `log2 Intensity`<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log2</a></span><span class="op">(</span><span class="va">Intensity</span><span class="op">)</span>,</span>
<span>    Category<span class="op">=</span><span class="st">"Before Normalization"</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span><span class="co"># plot distribution of normalized probes</span></span>
<span><span class="va">rma_intensity</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html">as_tibble</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/exprs.html">exprs</a></span><span class="op">(</span><span class="va">eset_rma</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>probesetid<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/featureNames.html">featureNames</a></span><span class="op">(</span><span class="va">eset_rma</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">pivot_longer</span><span class="op">(</span><span class="op">-</span><span class="va">probesetid</span>, names_to<span class="op">=</span><span class="st">"Sample"</span>, values_to <span class="op">=</span> <span class="st">"log2 Intensity"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>Category<span class="op">=</span><span class="st">"RMA Normalized"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/bind_rows.html">bind_rows</a></span><span class="op">(</span><span class="va">raw_intensity</span>, <span class="va">rma_intensity</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">Sample</span>, y<span class="op">=</span><span class="va">`log2 Intensity`</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html">geom_boxplot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/vars.html">vars</a></span><span class="op">(</span><span class="va">Category</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="main_files/microarray%20norm-1.png" width="672"></div>
<p>Above, we first apply RMA to the <code>Dilution</code> microarray dataset using the
<code><a href="https://rdrr.io/pkg/affy/man/rma.html">affy::rma()</a></code> function. Then we plot the log 2 intensities of the probes in each
array, first using the raw intensities and then afysis seeking to
identify how associated gene expression is with one or more variables of
interest. For example,</p>
</div>
<div id="differential-expression-microarrays-limma" class="section level3" number="8.7.4">
<h3>
<span class="header-section-number">8.7.4</span> Differential Expression: Microarrays (limma)<a class="anchor" aria-label="anchor" href="#differential-expression-microarrays-limma"><i class="fas fa-link"></i></a>
</h3>
<p><a href="https://bioconductor.org/packages/release/bioc/html/limma.html">limma</a>, which
is short for <code>li</code>near <code>m</code>odels of <code>mi</code>croarrays, is one of the <a href="https://bioconductor.org/packages/stats/">top most
downloaded</a> Bioconductor packages.
limma is utilized for analyzing microarray gene expression data, with a focus on
analyses using linear models to integrate <em>all</em> of the data from an experiment.
limma was developed for microarray analysis prior to the development of
sequencing based gene expression methods (i.e. <a href="biology-bioinformatics.html#rnaseq">RNASeq</a>) but has since added
functionality to analyze other types of gene expression data.</p>
<p>limma excels at analyzing these types of data as it can support arbitrarily
complex experimental designs while maintaining strong statistical power. An
experiment with a large number of conditions or predictors can still be analyze
even with small sample sizes. The method accomplishes this by using an
<a href="https://en.wikipedia.org/wiki/Empirical_Bayes_method">empirical Bayes</a> approach
that borrows information across all the genes in the dataset to better control
error in individual gene models. See <span class="citation">Ritchie et al. (<a href="detailed-class-schedule.html#ref-Ritchie2015-bj">2015</a>)</span> for more details on how
limma works.</p>
<p>limma requres an expression matrix like that stored in an ExpressionSet or
SummarizedExperiment container and a design matrix:</p>
<div class="sourceCode" id="cb394"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># intensities is an example gene expression matrix that corresponds to our</span></span>
<span><span class="co"># AD sample metadata of 10 AD and 10 Control individuals</span></span>
<span><span class="va">ad_se</span>  <span class="op">&lt;-</span> <span class="fu">SummarizedExperiment</span><span class="op">(</span></span>
<span>  assays<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>intensities<span class="op">=</span><span class="va">intensities</span><span class="op">)</span>,</span>
<span>  colData<span class="op">=</span><span class="va">ad_metadata</span>,</span>
<span>  rowData<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html">rownames</a></span><span class="op">(</span><span class="va">intensities</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># define our design matrix for AD vs control, adjusting for age at death</span></span>
<span><span class="va">ad_vs_control_model</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html">model.matrix</a></span><span class="op">(</span><span class="op">~</span> <span class="va">age_at_death</span> <span class="op">+</span> <span class="va">condition</span>, data<span class="op">=</span><span class="va">ad_metadata</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># now run limma</span></span>
<span><span class="co"># first fit all genes with the model</span></span>
<span><span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu">limma</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/limma/man/lmFit.html">lmFit</a></span><span class="op">(</span></span>
<span>  <span class="fu">assays</span><span class="op">(</span><span class="va">se</span><span class="op">)</span><span class="op">$</span><span class="va">intensities</span>,</span>
<span>  <span class="va">ad_vs_control_model</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># now better control fit errors with the empirical Bayes method</span></span>
<span><span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu">limma</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/limma/man/ebayes.html">eBayes</a></span><span class="op">(</span><span class="va">fit</span><span class="op">)</span></span></code></pre></div>
<p>We have now conducted a differential expression analysis with limma. We can
extract out the results for our question of interest - which genes are
associated with AD - using the <code><a href="https://rdrr.io/pkg/limma/man/toptable.html">limma::topTable()</a></code> function:</p>
<div class="sourceCode" id="cb395"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># the coefficient name conditionAD is the column name in the design matrix</span></span>
<span><span class="co"># adjust="BH" means perform Benjamini-Hochberg multiple testing adjustment</span></span>
<span><span class="co"># on the nominal p-values from each gene test</span></span>
<span><span class="co"># topTable only returns the top 10 results sorted by ascending p-value by default</span></span>
<span><span class="fu">topTable</span><span class="op">(</span><span class="va">fit</span>, coef<span class="op">=</span><span class="st">"conditionAD"</span>, adjust<span class="op">=</span><span class="st">"BH"</span><span class="op">)</span></span></code></pre></div>
<pre><code>##                   logFC  AveExpr         t      P.Value adj.P.Val         B
## 234942_s_at   1.1231942 8.397425  4.890779 9.613196e-05  0.770605 -2.987573
## 202423_at    -0.7499088 8.646687 -4.785281 1.221646e-04  0.770605 -3.025525
## 1557538_at   -0.8057310 4.348859 -4.697520 1.492072e-04  0.770605 -3.057737
## 1558290_a_at -0.8580750 7.278878 -4.658725 1.630238e-04  0.770605 -3.072162
## 234596_at     0.4925167 3.038448  4.564888 2.020432e-04  0.770605 -3.107523
## 225986_x_at  -0.6030385 6.958092 -4.502976 2.328354e-04  0.770605 -3.131216
## 203124_s_at  -1.2857166 8.565022 -4.488589 2.406451e-04  0.770605 -3.136763
## 209346_s_at   0.6578475 6.010219  4.421869 2.804637e-04  0.770605 -3.162688
## 224825_at     1.0266849 8.147104  4.421597 2.806391e-04  0.770605 -3.162795
## 205647_at    -0.8569475 4.399029 -4.399700 2.951151e-04  0.770605 -3.171376</code></pre>
<p>From the table, none of the probesets in this experiment are significant at FDR
&lt; 0.05.</p>
<div class="box readmore">
<ul>
<li><a href="https://bioinf.wehi.edu.au/limma/">Homepage</a></li>
<li><a href="https://bioconductor.org/packages/release/bioc/html/limma.html">BioConductor</a></li>
<li>
<a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4402510/">Publication</a><br>
</li>
</ul>
</div>
</div>
<div id="rnaseq" class="section level3" number="8.7.5">
<h3>
<span class="header-section-number">8.7.5</span> RNASeq<a class="anchor" aria-label="anchor" href="#rnaseq"><i class="fas fa-link"></i></a>
</h3>
<p>RNA sequencing (RNASeq) is a HTS technology that measures the abundance of RNA
molecules in a sample. Briefly, RNA molecules are extracted from a sample using
a biochemical protocol that isolates RNA from other molecules (i.e. DNA,
proteins, etc) in the sample. Ribosomal RNA is removed from the sample (see Note
box below), and the remaining RNA molecules are size selected to retain only
short molecules (~15-300 nucleotides in length, depending on the protocol). The
size selected molecules are then reverse transcribed into cDNA and converted to
double stranded molecules. Sequencing libraries are prepared with these cDNA
using proprietary biochemical protocols to make them suitable for sequencing on
the sequencing instrument. After sequencing, FASTQ files containing millions of
reads for one or more samples is produced. Except for the initial RNA
extraction, all of these steps are typically performed by specialized
instrumentation facilities called <em>sequencing cores</em> who provide the data in
FASTQ format to the investigators.</p>
<div class="box note">
<p>80%-90% of the RNA mass in a cell is ribosomal RNA (rRNA), the RNA that makes up
ribosomes and is responsible for maintaining protein translation processes in
the cell. If total RNA was sequenced, the large majority of reads would
correspond to rRNAs and would therefore be of little or no use. For this reason,
rRNA molecules are removed from the RNA material that goes on to sequencing
using either a <em>poly-A tail enrichment</em> or <em>ribo-depletion</em> strategy. These two
methods maintain different populations of RNA and therefore sequencing datasets
generated with different strategies require different interpretations. The more
detail on this topic is beyond the scope of this book, but <span class="citation">(<a href="detailed-class-schedule.html#ref-Zhao2018-ao">Zhao et al. 2018</a>)</span>
provide a good introduction and comparison of both approaches.</p>
</div>
<p>The reads produced in an RNASeq experiment represent RNA fragments from
transcripts. The number of reads that correspond to specific transcripts is
assumed to be proportional to the copy number of that transcript in the original
sample. The alignment process described in the <a href="biology-bioinformatics.html#preliminary-hts-data-analysis">Preliminary HTS Data Analysis</a>
section produces alignments of reads against the genome or transcriptome, and
the number of reads that align against each gene or transcript is counted using
a <em>reference annotation</em> that defines the locations of every known gene in the
genome. The resulting counts are therefore estimates of the copy number of the
molecules for each gene or transcript in the original sample. Conceptually, the
more reads that map to a gene, the higher the abundance of transcripts for that
gene, and therefore we infer higher the expression of that gene.</p>
<div class="box important">
<p>Gene expression for different genes can vary by orders of magnitude within a
single cell, where highly expressed genes may have billions of transcripts and
others comparatively few. Therefore, the likelihood of obtaining a read for a
given transcript is proportional to the <em>relative abundance</em> of that transcript
compared with all other transcripts. Relatively rare transcripts have a low
probability of being sampled by the sequencing procedure, and extremely lowly
expressed genes may not be sampled at all. The library size (i.e. the total
number of reads) in the sequencing dataset for a sample determines the range of
gene expression the dataset is likely to detect, where datasets with few reads
are unlikely to sample lowly expressed transcripts even though they are truly
expressed. For this reason <strong>absence of evidence is not evidence of absence</strong> in
RNASeq data. If a read is not observed for any given gene, we cannot say that
the gene is not expressed; we can only say that the relative expression of that
gene is below the detection threshold of the dataset we generated, given the
number of sequenced reads.</p>
</div>
<p>For a single sample, the output of this read counting procedure is a vector of
read counts for each gene in the annotation. Genes with no reads mapping to them
will have a count of zero, and all others will have a read count of one or more;
as mentioned in the <a href="biology-bioinformatics.html#count-data">Count Data</a> section, read counts are typically non-zero
integers. This procedure is typically repeated separately for a set of samples
that make up the experiment, and the counts for each sample are concatenated
together into a counts matrix. This count matrix is the primary result of
interest that is passed on to downstream analysis, most commonly differential
expression analysis.</p>
</div>
<div id="rnaseq-gene-expression-data" class="section level3" number="8.7.6">
<h3>
<span class="header-section-number">8.7.6</span> RNASeq Gene Expression Data<a class="anchor" aria-label="anchor" href="#rnaseq-gene-expression-data"><i class="fas fa-link"></i></a>
</h3>
<p>The read counts for each gene or transcript in the genome are estimates of the
relative abundance of molecules in the original sample. The number of genes
counted depends on the organism being studied and the maturity of the gene
annotation used. Complex multicellular organisms have on the order of
thousands to tens of thousands of genes; for example, humans and mice have
20k-30k genes in their respective genomes. Each RNASeq experiment thus may yield
tens of thousands of measurements for each sample.</p>
<p>We shall consider an example RNASeq dataset generated by <span class="citation">(<a href="detailed-class-schedule.html#ref-OMeara2015-mu">O’Meara et al. 2015</a>)</span>
generated to study how mouse cardiac cells can regenerate up until one week
of age, after which they lose the ability to do so. The dataset has eight
RNASeq samples, two for each of four time points across the developmental window
until adulthood. We will load in the expression data as a tibble:</p>
<div class="sourceCode" id="cb397"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">counts</span> <span class="op">&lt;-</span> <span class="fu">read_tsv</span><span class="op">(</span><span class="st">"verse_counts.tsv"</span><span class="op">)</span></span>
<span><span class="va">counts</span></span></code></pre></div>
<pre><code>## # A tibble: 55,416 x 9
##    gene                  P0_1  P0_2  P4_1  P4_2  P7_1  P7_2  Ad_1  Ad_2
##    &lt;chr&gt;                &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;
##  1 ENSMUSG00000102693.2     0     0     0     0     0     0     0     0
##  2 ENSMUSG00000064842.3     0     0     0     0     0     0     0     0
##  3 ENSMUSG00000051951.6    19    24    17    17    17    12     7     5
##  4 ENSMUSG00000102851.2     0     0     0     0     1     0     0     0
##  5 ENSMUSG00000103377.2     1     0     0     1     0     1     1     0
##  6 ENSMUSG00000104017.2     0     3     0     0     0     1     0     0
##  7 ENSMUSG00000103025.2     0     0     0     0     0     0     0     0
##  8 ENSMUSG00000089699.2     0     0     0     0     0     0     0     0
##  9 ENSMUSG00000103201.2     0     0     0     0     0     0     0     1
## 10 ENSMUSG00000103147.2     0     0     0     0     0     0     0     0
## # ... with 55,406 more rows</code></pre>
<p>The annotation counted reads for 55,416 annnotated Ensembl gene for the mouse
genome, as described in the <a href="biology-bioinformatics.html#gene-identifier-systems">Gene Identifier Systems</a> section.</p>
<p>There are typically three steps when analyzing a RNASeq count matrix:</p>
<ol style="list-style-type: decimal">
<li><strong>Filter genes that are unlikely to be differentially expressed.</strong></li>
<li><strong>Normalize filtered counts to make samples comparable to one another.</strong></li>
<li><strong>Differential expression analysis of filtered, normalized counts.</strong></li>
</ol>
<p>Each of these steps will be described in detail below.</p>
</div>
<div id="filtering-counts" class="section level3" number="8.7.7">
<h3>
<span class="header-section-number">8.7.7</span> Filtering Counts<a class="anchor" aria-label="anchor" href="#filtering-counts"><i class="fas fa-link"></i></a>
</h3>
<p>Filtering genes that are not likely to be differentially expressed is an
important step in differential expression analysis. The primary reason for this
is to reduce the number of genes tested for differential expression, thereby
reducing the multiple testing burden on the results. There are many approaches
and rationales for picking filters, and the choice is generally subjective and
highly dependent upon the specific dataset. In this section we discuss some of
the rationale and considerations for different filtering strategies.</p>
<p>Genes that were not detected in any sample are the easiest to filter, and often
eliminate many genes from consideration. Below we filter out all genes that have
zero counts in all samples:</p>
<div class="sourceCode" id="cb399"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">nonzero_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/colSums.html">rowSums</a></span><span class="op">(</span><span class="va">counts</span><span class="op">[</span><span class="op">-</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span><span class="op">!=</span><span class="fl">0</span></span>
<span><span class="va">filtered_counts</span> <span class="op">&lt;-</span> <span class="va">counts</span><span class="op">[</span><span class="va">nonzero_genes</span>,<span class="op">]</span></span>
<span><span class="va">filtered_counts</span></span></code></pre></div>
<pre><code>## # A tibble: 32,613 x 9
##    gene                   P0_1  P0_2  P4_1  P4_2  P7_1  P7_2  Ad_1  Ad_2
##    &lt;chr&gt;                 &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;
##  1 ENSMUSG00000051951.6     19    24    17    17    17    12     7     5
##  2 ENSMUSG00000102851.2      0     0     0     0     1     0     0     0
##  3 ENSMUSG00000103377.2      1     0     0     1     0     1     1     0
##  4 ENSMUSG00000104017.2      0     3     0     0     0     1     0     0
##  5 ENSMUSG00000103201.2      0     0     0     0     0     0     0     1
##  6 ENSMUSG00000103161.2      0     0     1     0     0     0     0     0
##  7 ENSMUSG00000102331.2      5     8     2     6     6    11     4     3
##  8 ENSMUSG00000102343.2      0     0     0     1     0     0     0     0
##  9 ENSMUSG00000025900.14     4     3     6     7     7     3    36    35
## 10 ENSMUSG00000102948.2      1     0     0     0     0     0     0     1
## # ... with 32,603 more rows</code></pre>
<p>Only 32,613 genes remain after filtering out genes with all zeros, eliminating
nearly 23,000 genes, or more than 40% of the genes in the annotation. Filtering
genes with zeros in all samples as we have done above is somewhat liberal, as
there may be many genes with reads in only one sample. Genes with only a single
sample with a non-zero count also cannot be differentially expressed, so we can
easily decide to eliminate those as well:</p>
<div class="sourceCode" id="cb401"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">nonzero_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/colSums.html">rowSums</a></span><span class="op">(</span><span class="va">counts</span><span class="op">[</span><span class="op">-</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span><span class="op">&gt;=</span><span class="fl">2</span></span>
<span><span class="va">filtered_counts</span> <span class="op">&lt;-</span> <span class="va">counts</span><span class="op">[</span><span class="va">nonzero_genes</span>,<span class="op">]</span></span>
<span><span class="va">filtered_counts</span></span></code></pre></div>
<pre><code>## # A tibble: 28,979 x 9
##    gene                   P0_1  P0_2  P4_1  P4_2  P7_1  P7_2  Ad_1  Ad_2
##    &lt;chr&gt;                 &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;
##  1 ENSMUSG00000051951.6     19    24    17    17    17    12     7     5
##  2 ENSMUSG00000103377.2      1     0     0     1     0     1     1     0
##  3 ENSMUSG00000104017.2      0     3     0     0     0     1     0     0
##  4 ENSMUSG00000102331.2      5     8     2     6     6    11     4     3
##  5 ENSMUSG00000025900.14     4     3     6     7     7     3    36    35
##  6 ENSMUSG00000102948.2      1     0     0     0     0     0     0     1
##  7 ENSMUSG00000025902.14   195   186   204   229   157   173   196   129
##  8 ENSMUSG00000104238.2      0     4     0     2     4     1     3     3
##  9 ENSMUSG00000102269.2      7    16     6     6     4     6     8     3
## 10 ENSMUSG00000118917.1      0     1     0     0     1     1     0     0
## # ... with 28,969 more rows</code></pre>
<p>We have now reduced our number of genes to 28,979 by including only genes with
at least two non-zero counts. We may now wonder how many genes we would include
if we required at least <span class="math inline">\(n\)</span> samples to have non-zero counts. Consider the
following bar chart of the number of genes detected in at least <span class="math inline">\(n\)</span> samples:</p>
<div class="sourceCode" id="cb403"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com">patchwork</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">nonzero_counts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span><span class="va">counts</span>,`Number of samples`<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/colSums.html">rowSums</a></span><span class="op">(</span><span class="va">counts</span><span class="op">[</span><span class="op">-</span><span class="fl">1</span><span class="op">]</span><span class="op">!=</span><span class="fl">0</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">`Number of samples`</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html">summarize</a></span><span class="op">(</span>`Number of genes`<span class="op">=</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/context.html">n</a></span><span class="op">(</span><span class="op">)</span> <span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>`Cumulative number of genes`<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">`Number of genes`</span><span class="op">)</span><span class="op">-</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/lead-lag.html">lag</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/cumsum.html">cumsum</a></span><span class="op">(</span><span class="va">`Number of genes`</span><span class="op">)</span>,<span class="fl">1</span>,default<span class="op">=</span><span class="fl">0</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sum_g</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">nonzero_counts</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html">geom_bar</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">`Number of samples`</span>,y<span class="op">=</span><span class="va">`Cumulative number of genes`</span>,fill<span class="op">=</span><span class="st">"In at least n"</span><span class="op">)</span>,stat<span class="op">=</span><span class="st">"identity"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html">geom_bar</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">`Number of samples`</span>,y<span class="op">=</span><span class="va">`Number of genes`</span>,fill<span class="op">=</span><span class="st">"In exactly n"</span><span class="op">)</span>,stat<span class="op">=</span><span class="st">"identity"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span><span class="op">(</span>title<span class="op">=</span><span class="st">"Number of genes detected in n samples"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_colour_discrete.html">scale_fill_discrete</a></span><span class="op">(</span>name<span class="op">=</span><span class="st">"Nonzero in:"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sum_g</span></span></code></pre></div>
<div class="inline-figure"><img src="main_files/rnaseq%20nonzero%20samples-1.png" width="672"></div>
<p>The largest number of genes is detected in none of the samples, followed by
genes detected in all samples, with many fewer in between. Based on the plot,
there is no obvious choice for how many nonzero samples are sufficient to use as
a filter.</p>
<p>We may gain additional insight into this problem by considering the distribution
of counts for genes present in at least <span class="math inline">\(n\)</span> samples as above, rather than just
the number of nonzero counts. Below is a plot similar to that above but, which
plots the distribution of mean counts across samples within each number of
nonzero samples:</p>
<div class="sourceCode" id="cb404"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># calculate the mean count for non-zero counts in genes with exactly n non-zero counts</span></span>
<span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span><span class="va">counts</span>,`Number of nonzero samples`<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/colSums.html">rowSums</a></span><span class="op">(</span><span class="va">counts</span><span class="op">[</span><span class="op">-</span><span class="fl">1</span><span class="op">]</span><span class="op">!=</span><span class="fl">0</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">pivot_longer</span><span class="op">(</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">gene</span>,<span class="va">`Number of nonzero samples`</span><span class="op">)</span>,names_to<span class="op">=</span><span class="st">"Sample"</span>,values_to<span class="op">=</span><span class="st">"Count"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">`Number of nonzero samples`</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">gene</span>, .add<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html">summarize</a></span><span class="op">(</span>`Nonzero Count Mean`<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/mean.html">mean</a></span><span class="op">(</span><span class="va">Count</span><span class="op">[</span><span class="va">Count</span><span class="op">!=</span><span class="fl">0</span><span class="op">]</span><span class="op">)</span>,.groups<span class="op">=</span><span class="st">"keep"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">ungroup</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">`Number of nonzero samples`</span>,y<span class="op">=</span><span class="va">`Nonzero Count Mean`</span>,fill<span class="op">=</span><span class="va">`Number of nonzero samples`</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_violin.html">geom_violin</a></span><span class="op">(</span>scale<span class="op">=</span><span class="st">"width"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_y_log10</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span><span class="op">(</span>title<span class="op">=</span><span class="st">"Nonzero count mean for genes with exactly n nonzero counts"</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="main_files/rnaseq%20nonzero%20dist-1.png" width="672"></div>
<p>From this plot, we see that the mean count for genes with even a single sample
with a zero is substantially lower than most genes without any zeros. We might
therefore decide to filter out genes that have <em>any</em> zeros, with the rationale
that most genes of appreciable abundance would still be included. However, there
are some genes with a high average count for samples that are not zero that
would be eliminated if we chose to use only genes expressed in all 8 samples.
Again, there is no clear choice of filtering threshold, and a subjective
decision is necessary.</p>
<p>One important property of RNASeq expression data is that genes with higher mean
count also have higher variance. This mean-variance dependence means these count
data are <em>heteroskedastic</em>, a term used to describe data that has inconstant
variance across different values. We can visualize this dependence by
calculating the mean and variance for each of our normalized count genes and
plotting them against each other:</p>
<div class="inline-figure"><img src="main_files/rnaseq%20mean%20variance-1.png" width="672"></div>
<p>Unlike in microarray data, where probesets can be filtered based on variance,
using low variance as a filter for count data would result in filtering out
genes with low mean, which would lead to undesirable bias toward highly
expressed genes.</p>
<p>Depending on the experiment, it may be meaningful if a gene is detected
at all in any of the samples. In these cases, genes with only 1 read in a single
sample might be kept, even though these genes would not result in confident
statistical inference if subjected to differential expression. The scientific
question asked by the experiment is an important consideration when deciding
thresholds.</p>
<p>When the goal of the RNASeq experiment is differential expression analysis, a
good rule of thumb when choosing a filtering threshold is to choose genes that
have sufficiently many non-zero counts such that the statistical method has
enough variance to perform reliable inference. For example, a reasonable
filtering threshold might be to include genes that have non-zero counts in at
least 50% of the samples. This threshold depends only upon the number of samples
being considered. In an experimental design with two or more sample groups, a
reasonable strategy might be to require at least 50% of samples <em>within at least
one group</em> or <em>within all groups</em> to be included. Taking the experimental design
into account when filtering can ensure the most interesting genes, e.g. those
expressed only in one condition, are captured.</p>
<p>Regardless of the filtering strategy used, <strong>there is no biological meaning to
any filtering threshold</strong>. The number of reads detected for each gene is
dependent primarily upon the library size, where larger libraries are more
likely to sample lowly abundant genes. Therefore, we cannot filter genes with
the intent of “removing lowly expressed genes”; instead we can only filter out
genes that are below the detection threshold afforded by the library size of the
dataset.</p>
</div>
<div id="count-distributions" class="section level3" number="8.7.8">
<h3>
<span class="header-section-number">8.7.8</span> Count Distributions<a class="anchor" aria-label="anchor" href="#count-distributions"><i class="fas fa-link"></i></a>
</h3>
<p>The range of gene expression in a cell varies by orders of magnitude. The
following histogram plots the distribution of read count values for one of the
adult samples:</p>
<div class="sourceCode" id="cb405"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="va">filtered_counts</span>, <span class="va">Ad_1</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>`log10(counts+1)`<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="va">Ad_1</span><span class="op">+</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">`log10(counts+1)`</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html">geom_histogram</a></span><span class="op">(</span>bins<span class="op">=</span><span class="fl">50</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span><span class="op">(</span>title<span class="op">=</span><span class="st">'log10(counts) distribution for Adult mouse'</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="main_files/rnaseq%20sample%20count%20dist-1.png" width="672"></div>
<p>We add 1 count, called a <em>pseudocount</em>, to avoid taking the log of zero, therefore
all the genes at <span class="math inline">\(x=0\)</span> are genes that have no reads. From the plot, the largest
count is nearly <span class="math inline">\(10^6\)</span> while the smallest is 1, spanning nearly six orders
of magnitude. There are relatively few of these highly abundant genes, however,
and most genes are betwen the range of 100-10,000 reads. We see that this
distribution is similar for all the samples, when displayed as a <a href="data-visualization.html#ridgeline">ridgeline
plot</a>:</p>
<div class="sourceCode" id="cb406"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://wilkelab.org/ggridges/">ggridges</a></span><span class="op">)</span></span>
<span></span>
<span><span class="fu">pivot_longer</span><span class="op">(</span><span class="va">filtered_counts</span>,<span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">gene</span><span class="op">)</span>,names_to<span class="op">=</span><span class="st">"Sample"</span>,values_to<span class="op">=</span><span class="st">"Count"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>`log10(Count+1)`<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="va">Count</span><span class="op">+</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>y<span class="op">=</span><span class="va">Sample</span>,x<span class="op">=</span><span class="va">`log10(Count+1)`</span>,fill<span class="op">=</span><span class="va">Sample</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://wilkelab.org/ggridges/reference/geom_density_ridges.html">geom_density_ridges</a></span><span class="op">(</span>bandwidth<span class="op">=</span><span class="fl">0.132</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span><span class="op">(</span>title<span class="op">=</span><span class="st">"log10(Count+1) Distributions"</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="main_files/rnaseq%20ridgeline-1.png" width="672"></div>
<p>The general shape of the distributions is similar, with a large mode of counts
at zero, representing genes that had zero counts in that sample, and a wider
mode between 100 and 10,000 where most consistently expressed genes fall.
However, we note that the peak count of these larger modes vary and also recall
that though the difference appears small, the log scale of the x axis means
the differences in counts may be substantial. We will discuss the implications
of this in the next section.</p>
<div id="count-normalization" class="section level4" number="8.7.8.1">
<h4>
<span class="header-section-number">8.7.8.1</span> Count Normalization<a class="anchor" aria-label="anchor" href="#count-normalization"><i class="fas fa-link"></i></a>
</h4>
<p>The number of reads generated for each sample in an RNASeq dataset varies from
sample to sample due to randomness in the biochemical process that generates
the data. The relationship between the number of reads that maps to any given
gene and the relative abundance of the molecule in the sample is dependent upon
the library size of each sample. The direct number of counts, or the <em>raw
counts</em>, is therefore not directly comparable between samples. A simple example
should suffice to make this clear:</p>
<blockquote>
<p>Imagine two RNASeq datasets generated from the same RNA sample. One of the
samples was sequenced to 1000 reads, and the other to 2000 reads. Now imagine
that one gene makes up half of the RNA in the original sample. In the sample
with 1000 reads, we find this gene has 500 reads. In the sample with 2000 reads,
the gene has 1000 of those reads. In both cases the <em>fraction</em> of reads that map
to the gene is 0.5, but the absolute number of reads differs greatly. This is
why the raw counts for a gene are not directly comparable between samples with
different numbers of reads. Since in general every sample will have a unique
number of reads, these raw counts will never be directly comparable. The way we
mitigate this problem is with a statistical procedure called <em>count
normalization</em>.</p>
</blockquote>
<p>Count normalization is the process by which the number of raw counts in each
sample is scaled by a factor to make multiple samples comparable. Many
strategies have been proposed to do this, but the simplest is a family of
methods that divide by some factor of the library size. A common method in this
family is to compute <em>counts per million reads</em> or <em>CPM</em> normalization, which
scales each gene count by the number of millions of reads in the library:</p>
<p><span class="math display">\[
cpm_{s,i} =  \frac{c_{s,i}}{l_s} * 10^6
\]</span></p>
<p>Here, <span class="math inline">\(c_{s,i}\)</span> is the raw count of gene <span class="math inline">\(i\)</span> in sample <span class="math inline">\(s\)</span>, and <span class="math inline">\(l_s\)</span> is the
number of reads in the dataset for sample <span class="math inline">\(s\)</span>. If each sample is scaled by their
own library size, then in principle the proportion of read counts assigned to
each gene out of all reads for that sample will be comparable across samples.
When we have no additional knowledge of the distribution of genes in our system,
library size normalization methods are the most general and make the fewest
assumptions. Below is a ridgeline plot of the counts distribution after CPM
normalization:</p>
<div class="inline-figure"><img src="main_files/rnaseq%20ridgeline%20cpm-1.png" width="672"></div>
<p>The distributions look quite different after CPM normalization, but the modes
now look more consistent than with raw counts.</p>
<p>One drawback of library size normalization is it is sensitive to extreme values;
individual genes with many more reads in one sample than in other samples. These
outliers can lead to pathological effects when comparing gene count values
across samples. However, if we are measuring gene expression in a single
biological system like mouse or human, as is often the case, other methods have
been developed to better use our understanding of the gene expression
distribution to perform more robust and biologically meaningful normalization.
A more robust approach is possible when we make the reasonable assumption that,
for any set of biological conditions in an experiment, <em>most genes are not
differentially expressed</em>. This is the assumption made by the <em>DESeq2
normalization method</em>.</p>
<p>The DESeq2 normalization method is a statistical procedure that uses the median
geometric mean computed across all samples to determine the scale factor for
each sample. The procedure is somewhat complicated and we encourage the reader
to examine it in detail in the DESeq2 publication <span class="citation">(<a href="detailed-class-schedule.html#ref-Love2014-lb">Love, Huber, and Anders 2014</a>)</span>. The method has
proven to be consistent and reliable compared with other strategies
<span class="citation">(<a href="detailed-class-schedule.html#ref-Dillies2013-rq">Dillies et al. 2013</a>)</span> and is the default normalization for the populare DESeq2
differential expression method, which will be described in a later section. The
DESeq2 normalization method may be implemented using the <a href="https://bioconductor.org/packages/release/bioc/html/DESeq2.html">DESeq2 Bioconductor
package</a>:</p>
<div class="sourceCode" id="cb407"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/mikelove/DESeq2">DESeq2</a></span><span class="op">)</span></span>
<span><span class="co"># DESeq2 requires a counts matrix, column data (sample information), and a formula</span></span>
<span><span class="co"># the counts matrix *must be raw counts*</span></span>
<span><span class="va">count_mat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span><span class="op">(</span><span class="va">filtered_counts</span><span class="op">[</span><span class="op">-</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/row.names.html">row.names</a></span><span class="op">(</span><span class="va">count_mat</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">filtered_counts</span><span class="op">$</span><span class="va">gene</span></span>
<span></span>
<span><span class="va">dds</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/DESeqDataSet.html">DESeqDataSetFromMatrix</a></span><span class="op">(</span></span>
<span>  countData<span class="op">=</span><span class="va">count_mat</span>,</span>
<span>  colData<span class="op">=</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html">tibble</a></span><span class="op">(</span>sample_name<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html">colnames</a></span><span class="op">(</span><span class="va">filtered_counts</span><span class="op">[</span><span class="op">-</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  design<span class="op">=</span><span class="op">~</span><span class="fl">1</span> <span class="co"># no formula is needed for normalization, ~1 produces a trivial design matrix</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># compute normalization factors</span></span>
<span><span class="va">dds</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html">estimateSizeFactors</a></span><span class="op">(</span><span class="va">dds</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># extract the normalized counts</span></span>
<span><span class="va">deseq_norm_counts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html">as_tibble</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html">counts</a></span><span class="op">(</span><span class="va">dds</span>,normalized<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>gene<span class="op">=</span><span class="va">filtered_counts</span><span class="op">$</span><span class="va">gene</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/relocate.html">relocate</a></span><span class="op">(</span><span class="va">gene</span><span class="op">)</span></span></code></pre></div>
<p>Using the DESeq2 package, we first construct a DESeq object using the raw counts
matrix, a sample information tibble with only a sample name column, and a
trivial design formula, which will be explained in more depth in the
<a href="biology-bioinformatics.html#deseq2edger">DESeq2/edgeR</a> section. The <code><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html">estimateSizeFactors()</a></code> function performs the
normalization routine on the counts, and the <code>counts(dds,normalized=TRUE)</code> call
extracts the normalized counts matrix. When plotted as a ridgeline plot as
before, we see that the gene expression distribution modes are better behaved
compared to raw counts, and the shape of the distribution is preserved unlike
with CPM normalized counts:</p>
<div class="sourceCode" id="cb408"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">pivot_longer</span><span class="op">(</span><span class="va">deseq_norm_counts</span>,<span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">gene</span><span class="op">)</span>,names_to<span class="op">=</span><span class="st">"Sample"</span>,values_to<span class="op">=</span><span class="st">"Count"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>`log10(Count+1)`<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="va">Count</span><span class="op">+</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>y<span class="op">=</span><span class="va">Sample</span>,x<span class="op">=</span><span class="va">`log10(Count+1)`</span>,fill<span class="op">=</span><span class="va">Sample</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://wilkelab.org/ggridges/reference/geom_density_ridges.html">geom_density_ridges</a></span><span class="op">(</span>bandwidth<span class="op">=</span><span class="fl">0.132</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span><span class="op">(</span>title<span class="op">=</span><span class="st">"log10(Count+1) Distributions"</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="main_files/rnaseq%20ridgeline%20deseq2-1.png" width="672"></div>
<p>The DESeq2 normalization procedure has two important considerations to keep in
mind:</p>
<ul>
<li>
<strong>The procedure borrows information across all samples.</strong> The geometric mean
of counts across all samples is computed as part of the normalization procedure.
This means that the size factors computed <em>depend on the samples in the
dataset</em>. The normalized counts for one sample will likely change if it is
normalized with a different set of samples.</li>
<li>
<strong>The procedure does not use genes with any zero counts.</strong> The geometric mean
calculation across all samples means that any sample with a zero count results
in the entire geometric mean being zero. For this reason <em>only genes with
nonzero counts in every sample are used to calculate the normalization factors</em>.
If one of the samples has a large number of zeros, for example due to a small
library size, this could dramatically influence the normalization factors for
the entire experiment.</li>
</ul>
<p>The CPM normalization procedure does not borrow information across all samples,
and therefore is not subject to these considerations.</p>
<p>These are only two of many possible normalization methods, DESeq2 being the most
popular at the time of writing. For a detailed explanation of many normalization
methods, see <span class="citation">(<a href="detailed-class-schedule.html#ref-Dillies2013-rq">Dillies et al. 2013</a>)</span>.</p>
</div>
<div id="count-transformation" class="section level4" number="8.7.8.2">
<h4>
<span class="header-section-number">8.7.8.2</span> Count Transformation<a class="anchor" aria-label="anchor" href="#count-transformation"><i class="fas fa-link"></i></a>
</h4>
<p>As mentioned in the <a href="biology-bioinformatics.html#count-data">Count Data</a> section, one way to deal with the non-normality
of count data is to perform a data transformation that makes the counts data
follow a normal distribution. Transformation to normality allows common and
powerful statistical methods like linear regression to be used. The most basic
count transformation is to take the logarithm of the counts. However, this can
be problematic for genes with low counts, causing the values to spread further
apart than genes with high counts, increasing the likelihood of false positive
results. To address this issue, the DESeq2 package provides the <code><a href="https://rdrr.io/pkg/DESeq2/man/rlog.html">rlog()</a></code>
function that performs a regularized logarithmic transformation that adjusts the
log transform based on the mean count of each gene.</p>
<div class="box note">
<p>The rlog transformation also has the desirable effect of making the count data
homoskedastic, meaning the variance of a gene is not dependent on the mean. Some
statistical methods assume that a dataset has this property. The relationship of
mean count vs variance will be discussed in more detail in the next section.</p>
</div>
<p>We can clearlysee the effect of rlog on the counts distribution by plotting the
rlog’ed counts using a ridgeline plot as before:</p>
<div class="sourceCode" id="cb409"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># the dds is the DESeq2 object from earlier</span></span>
<span><span class="va">rld</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/rlog.html">rlog</a></span><span class="op">(</span><span class="va">dds</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># extract rlog values as a tibble</span></span>
<span><span class="va">rlog_counts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html">as_tibble</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">assay</a></span><span class="op">(</span><span class="va">rld</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">rlog_counts</span><span class="op">$</span><span class="va">gene</span> <span class="op">&lt;-</span> <span class="va">filtered_counts</span><span class="op">$</span><span class="va">gene</span></span>
<span><span class="fu">pivot_longer</span><span class="op">(</span><span class="va">rlog_counts</span>,<span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">gene</span><span class="op">)</span>,names_to<span class="op">=</span><span class="st">"Sample"</span>,values_to<span class="op">=</span><span class="st">"rlog count"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>y<span class="op">=</span><span class="va">Sample</span>,x<span class="op">=</span><span class="va">`rlog count`</span>,fill<span class="op">=</span><span class="va">Sample</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://wilkelab.org/ggridges/reference/geom_density_ridges.html">geom_density_ridges</a></span><span class="op">(</span>bandwidth<span class="op">=</span><span class="fl">0.132</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span><span class="op">(</span>title<span class="op">=</span><span class="st">"rlog Distributions"</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="main_files/rnaseq%20rlog-1.png" width="672"></div>
<div class="box warning">
<p>It is generally accepted that count transformations have significant drawbacks
compared with statistical methods that model counts and errors explicitly, like
generalized linear models (GLM) such as Poisson and Negative Binomial
regressions <span class="citation">(<a href="detailed-class-schedule.html#ref-St-Pierre2018-tc">St-Pierre, Shikon, and Schneider 2018</a>)</span>. Whenever possible, these GLM based methods
should be used instead of transformations. However, the transformations may
sometimes be necessary depending on the required analysis, so the rlog function
in DESeq2 may be useful for these cases.</p>
</div>
<div class="box readmore">
<ul>
<li><a href="https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#count-data-transformations">DESeq2 package vignette on Count data transformations</a></li>
</ul>
</div>
</div>
</div>
<div id="differential-expression-rnaseq" class="section level3" number="8.7.9">
<h3>
<span class="header-section-number">8.7.9</span> Differential Expression: RNASeq<a class="anchor" aria-label="anchor" href="#differential-expression-rnaseq"><i class="fas fa-link"></i></a>
</h3>
<p>The read counts for each gene or transcript form the primary data used to
identify genes whose expression correlates with some condition of interest. The
counts matrix created by concatenating read counts for all genes across a set of
samples where rows are genes or transcripts and columns are samples is the most
common form of expression data used for this purpose. As mentioned in the
<a href="biology-bioinformatics.html#count-data">Count Data</a> section, these counts are not normally distributed and therefore
require a statistical approach that can accommodate the counts distribution.</p>
<p>The strong relationship between the mean and variance is characteristic of
RNASeq expression data, and motivates the use of a generalized linear model
called <em>negative binomial regression</em>. Negative binomial regression models count
data using the <a href="https://en.wikipedia.org/wiki/Negative_binomial_distribution">negative binomial distribution</a>,
which allows the relationship of mean and variance of a counts distribution to
vary. The statistical details of the negative binomial distribution and negative
binomial regression are beyond the scope of this book, but below we will discuss
the aspects necessary to implement differential expression analysis.</p>
<div id="deseq2edger" class="section level4" number="8.7.9.1">
<h4>
<span class="header-section-number">8.7.9.1</span> DESeq2/EdgeR<a class="anchor" aria-label="anchor" href="#deseq2edger"><i class="fas fa-link"></i></a>
</h4>
<p><a href="https://bioconductor.org/packages/release/bioc/html/DESeq2.html">DESeq2</a> and
<a href="https://bioconductor.org/packages/release/bioc/html/edgeR.html">edgeR</a> are bioconductor packages that both implement negative binomial
regression to perform differential expression on RNASeq data. Both perform
raw counts normalization as part of their function, though they differ in
the normalization method (DESeq2 method is described above, while edgeR uses
trimmed mean of M-values (TMM), see <span class="citation">(<a href="detailed-class-schedule.html#ref-Robinson2010-ik">Robinson and Oshlack 2010</a>)</span>). The interface for using
both packages is also similar, so we will focus on DESeq2 in this section.</p>
<p>As mentioned briefly above, DESeq2 requires three pieces of information to
perform differential expression:</p>
<ol style="list-style-type: decimal">
<li>A <strong>raw</strong> counts matrix with genes as rows and samples as columns</li>
<li>A data frame with metadata associated with the columns</li>
<li>A design formula for the differential expression model</li>
</ol>
<p>DESeq2 requires raw counts due to its normalization procedure, which borrows
information across samples when computing size factors, as described above in
the <a href="biology-bioinformatics.html#count-normalization">Count Normalization</a> section.</p>
<p>With these three pieces of information, we construct a <code>DESeq</code> object:</p>
<div class="sourceCode" id="cb410"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/mikelove/DESeq2">DESeq2</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># filter counts to retain genes with at least 6 out of 8 samples with nonzero counts</span></span>
<span><span class="va">filtered_counts</span> <span class="op">&lt;-</span> <span class="va">counts</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/colSums.html">rowSums</a></span><span class="op">(</span><span class="va">counts</span><span class="op">[</span><span class="op">-</span><span class="fl">1</span><span class="op">]</span><span class="op">!=</span><span class="fl">0</span><span class="op">)</span><span class="op">&gt;=</span><span class="fl">6</span>,<span class="op">]</span></span>
<span></span>
<span><span class="co"># DESeq2 requires a counts matrix, column data (sample information), and a formula</span></span>
<span><span class="co"># the counts matrix *must be raw counts*</span></span>
<span><span class="va">count_mat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span><span class="op">(</span><span class="va">filtered_counts</span><span class="op">[</span><span class="op">-</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/row.names.html">row.names</a></span><span class="op">(</span><span class="va">count_mat</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">filtered_counts</span><span class="op">$</span><span class="va">gene</span></span>
<span></span>
<span><span class="co"># create a sample matrix from sample names</span></span>
<span><span class="va">sample_info</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html">tibble</a></span><span class="op">(</span></span>
<span>  sample_name<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html">colnames</a></span><span class="op">(</span><span class="va">filtered_counts</span><span class="op">[</span><span class="op">-</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">separate</span><span class="op">(</span><span class="va">sample_name</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"timepoint"</span>,<span class="st">"replicate"</span><span class="op">)</span>,sep<span class="op">=</span><span class="st">"_"</span>,remove<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span></span>
<span>    timepoint<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">timepoint</span>,levels<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Ad"</span>,<span class="st">"P0"</span>,<span class="st">"P4"</span>,<span class="st">"P7"</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span><span class="va">design</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/formula.html">formula</a></span><span class="op">(</span><span class="op">~</span> <span class="va">timepoint</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sample_info</span></span></code></pre></div>
<pre><code>## # A tibble: 8 x 3
##   sample_name timepoint replicate
##   &lt;chr&gt;       &lt;fct&gt;     &lt;chr&gt;    
## 1 P0_1        P0        1        
## 2 P0_2        P0        2        
## 3 P4_1        P4        1        
## 4 P4_2        P4        2        
## 5 P7_1        P7        1        
## 6 P7_2        P7        2        
## 7 Ad_1        Ad        1        
## 8 Ad_2        Ad        2</code></pre>
<p>The design formula <code>~ timepoint</code> is similar to that supplied to the
<code><a href="https://rdrr.io/r/stats/model.matrix.html">model.matrix()</a></code> function as described in the <a href="biology-bioinformatics.html#differential-expression-analysis">Differential Expression Analysis</a>
section, where DESeq2 implicitly creates the model matrix by combining the
column data with the formula. In this differential expression analysis, we wish
to identify genes that are different between the timepoints <code>P0</code>, <code>P4</code>, <code>P7</code>,
and <code>Ad</code>. With these three pieces of information prepared, we can create a DESeq
object and perform differential expression analysis:</p>
<div class="sourceCode" id="cb412"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dds</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/DESeqDataSet.html">DESeqDataSetFromMatrix</a></span><span class="op">(</span></span>
<span>  countData<span class="op">=</span><span class="va">count_mat</span>,</span>
<span>  colData<span class="op">=</span><span class="va">sample_info</span>,</span>
<span>  design<span class="op">=</span><span class="va">design</span></span>
<span><span class="op">)</span></span>
<span><span class="va">dds</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/DESeq.html">DESeq</a></span><span class="op">(</span><span class="va">dds</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/results.html">resultsNames</a></span><span class="op">(</span><span class="va">dds</span><span class="op">)</span></span></code></pre></div>
<pre><code>## [1] "Intercept"          "timepoint_P0_vs_Ad" "timepoint_P4_vs_Ad"
## [4] "timepoint_P7_vs_Ad"</code></pre>
<p>DESeq2 performed differential expression on each gene and estimated parameters
for three different comparisons - P0 vs Ad, P4 vs Ad, and P7 vs Ad. Here, Ad is
the <em>reference group</em> so that we may interpret the statistics as relative to the
adult animals. The choice of reference group does not change the magnitude of
the results, but may change the sign of the estimated coefficients, which are
log2 fold changes. To extract out the differential expression statistics for the
P0 vs Ad comparison, we can use the <code><a href="https://rdrr.io/pkg/DESeq2/man/results.html">results()</a></code> function:</p>
<div class="sourceCode" id="cb414"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/results.html">results</a></span><span class="op">(</span><span class="va">dds</span>, name<span class="op">=</span><span class="st">"timepoint_P0_vs_Ad"</span><span class="op">)</span></span>
<span><span class="va">p0_vs_Ad_de</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html">as_tibble</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>gene<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html">rownames</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/relocate.html">relocate</a></span><span class="op">(</span><span class="va">gene</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html">arrange</a></span><span class="op">(</span><span class="va">pvalue</span><span class="op">)</span></span>
<span><span class="va">p0_vs_Ad_de</span></span></code></pre></div>
<pre><code>## # A tibble: 21,213 x 7
##    gene                  baseMean log2FoldChange lfcSE  stat    pvalue      padj
##    &lt;chr&gt;                    &lt;dbl&gt;          &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;
##  1 ENSMUSG00000000031.17   19624.           6.75 0.190  35.6 5.70e-278 1.07e-273
##  2 ENSMUSG00000026418.17    9671.           9.84 0.285  34.5 2.29e-261 2.15e-257
##  3 ENSMUSG00000051855.16    3462.           5.35 0.157  34.1 2.84e-255 1.77e-251
##  4 ENSMUSG00000027750.17    9482.           4.05 0.125  32.4 2.26e-230 1.06e-226
##  5 ENSMUSG00000042828.13    3906.          -5.32 0.167 -31.8 1.67e-221 6.25e-218
##  6 ENSMUSG00000046598.16    1154.          -6.08 0.197 -30.8 4.77e-208 1.49e-204
##  7 ENSMUSG00000019817.19    1952.           5.15 0.171  30.1 1.43e-198 3.83e-195
##  8 ENSMUSG00000021622.4    17146.          -4.57 0.157 -29.0 4.41e-185 1.03e-181
##  9 ENSMUSG00000002500.16    2121.          -6.20 0.217 -28.5 6.69e-179 1.39e-175
## 10 ENSMUSG00000024598.10    1885.           5.90 0.209  28.2 5.56e-175 1.04e-171
## # ... with 21,203 more rows</code></pre>
<p>These are the results of the differential expression comparing P0 with Ad; the
other contrasts (P4 vs Ad and P7 vs Ad) may be extracted by specifying the
appropriate result names to the <code>name</code> argument to the <code><a href="https://rdrr.io/pkg/DESeq2/man/results.html">results()</a></code> function. The columns in the DESeq2 results are defined as follows:</p>
<ul>
<li>
<em>baseMean</em> - the mean normalized count of all samples for the gene</li>
<li>
<em>log2FoldChange</em> - the estimated coefficient (i.e. log2FoldChange) for the comparison of interest</li>
<li>
<em>lfcSE</em> - the standard error for the log2FoldChange estimate</li>
<li>
<em>stat</em> - the <a href="https://en.wikipedia.org/wiki/Wald_test">Wald statistic</a> associated with the log2FoldChange estimate</li>
<li>
<em>pvalue</em> - the nominal p-value of the Wald test (i.e. the signifiance of the association)</li>
<li>
<em>padj</em> - the multiple testing adjusted p-value (i.e. false discovery rate)</li>
</ul>
<p>In the P0 vs Ad comparison, we can check the number of significant results easily:</p>
<div class="sourceCode" id="cb416"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># number of genes significant at FDR &lt; 0.05</span></span>
<span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">p0_vs_Ad_de</span>,<span class="va">padj</span><span class="op">&lt;</span><span class="fl">0.05</span><span class="op">)</span></span></code></pre></div>
<pre><code>## # A tibble: 7,557 x 7
##    gene                  baseMean log2FoldChange lfcSE  stat    pvalue      padj
##    &lt;chr&gt;                    &lt;dbl&gt;          &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;
##  1 ENSMUSG00000000031.17   19624.           6.75 0.190  35.6 5.70e-278 1.07e-273
##  2 ENSMUSG00000026418.17    9671.           9.84 0.285  34.5 2.29e-261 2.15e-257
##  3 ENSMUSG00000051855.16    3462.           5.35 0.157  34.1 2.84e-255 1.77e-251
##  4 ENSMUSG00000027750.17    9482.           4.05 0.125  32.4 2.26e-230 1.06e-226
##  5 ENSMUSG00000042828.13    3906.          -5.32 0.167 -31.8 1.67e-221 6.25e-218
##  6 ENSMUSG00000046598.16    1154.          -6.08 0.197 -30.8 4.77e-208 1.49e-204
##  7 ENSMUSG00000019817.19    1952.           5.15 0.171  30.1 1.43e-198 3.83e-195
##  8 ENSMUSG00000021622.4    17146.          -4.57 0.157 -29.0 4.41e-185 1.03e-181
##  9 ENSMUSG00000002500.16    2121.          -6.20 0.217 -28.5 6.69e-179 1.39e-175
## 10 ENSMUSG00000024598.10    1885.           5.90 0.209  28.2 5.56e-175 1.04e-171
## # ... with 7,547 more rows</code></pre>
<p>There were 7,557 significantly differentially expressed genes associated with
this comparison, which includes both up- and down-regulated genes (examine the
signs of the log2FoldChange column). A common diagnostic plot of differenial
expression results is the so-called <em>volcano plot</em> which plots the
log2FoldChange on the x-axis against -log10 adjusted p-value on the y-axis:</p>
<div class="sourceCode" id="cb418"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span></span>
<span>  <span class="va">p0_vs_Ad_de</span>,</span>
<span>  `-log10(adjusted p)`<span class="op">=</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="va">padj</span><span class="op">)</span>,</span>
<span>  `FDR&lt;0.05`<span class="op">=</span><span class="va">padj</span><span class="op">&lt;</span><span class="fl">0.05</span></span>
<span>  <span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">log2FoldChange</span>,y<span class="op">=</span><span class="va">`-log10(adjusted p)`</span>,color<span class="op">=</span><span class="va">`FDR&lt;0.05`</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="main_files/rnaseq%20deseq2%20volcano-1.png" width="672"></div>
<p>The plot shows that there are significant genes at FDR &lt; 0.05 that are increased
and decreased in the comparison.</p>
<p>DESeq2 has many capabilities beyond this basic differential expression
functionality as described in its <a href="http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html">comprehensive documentation in its vignette</a>.</p>
<p>As mentioned, edgeR implements a very similar model to DESeq2 and also has
<a href="https://bioconductor.org/packages/release/bioc/html/edgeR.html">excellent
documentation</a>.</p>
</div>
<div id="limmavoom" class="section level4" number="8.7.9.2">
<h4>
<span class="header-section-number">8.7.9.2</span> limma/voom<a class="anchor" aria-label="anchor" href="#limmavoom"><i class="fas fa-link"></i></a>
</h4>
<p>While the limma package was initially developed for microarray analysis, the
authors added support for RNASeq soon after the HTS technique became available.
limma accommodates count data by first performing a <a href="biology-bioinformatics.html#count-transformation">Count Transformation</a> using
its own
<a href="https://genomebiology.biomedcentral.com/articles/10.1186/gb-2014-15-2-r29">voom</a>
transformation procedure. voom transforms counts by first performing a counts
per million normalization, taking the logarithm of the CPM values, and finally
estimates the mean-variance relationship across genes for use in its empirical
Bayes statistical framework. The counts data being thus transformed, limma
performs statistical inference using its usual linear model framework to produce
differential expression results with arbitrarily complex statistical models.</p>
<p>This brief example is from the <a href="https://www.bioconductor.org/packages/devel/bioc/vignettes/limma/inst/doc/">limma User
Guide</a>
chapter 15, and covers loading
and processing data from an RNA-seq experiment. We go into more depth while
working with limma in assignment 6.</p>
<p>Without going into too much detail, the <code>design</code> variable is how we inform limma
of our experimental conditions, and where limma draws from to construct its
linear models correctly. This design is relatively simple, just four samples
belonging to two different conditions (the <code>swirls</code> here refer to the swirl of
zebra fish, you can just see them as a phenotypic difference)</p>
<div class="sourceCode" id="cb419"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">design</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>swirl <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"swirl.1"</span>, <span class="st">"swirl.2"</span>, <span class="st">"swirl.3"</span>, <span class="st">"swirl.4"</span><span class="op">)</span>,</span>
<span>                     condition <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="op">-</span><span class="fl">1</span>, <span class="fl">1</span>, <span class="op">-</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">dge</span> <span class="op">&lt;-</span> <span class="fu">DGEList</span><span class="op">(</span>counts<span class="op">=</span><span class="va">counts</span><span class="op">)</span></span>
<span><span class="va">keep</span> <span class="op">&lt;-</span> <span class="fu">filterByExpr</span><span class="op">(</span><span class="va">dge</span>, <span class="va">design</span><span class="op">)</span></span>
<span><span class="va">dge</span> <span class="op">&lt;-</span> <span class="va">dge</span><span class="op">[</span><span class="va">keep</span>,,keep.lib.sizes<span class="op">=</span><span class="cn">FALSE</span><span class="op">]</span></span></code></pre></div>
<p><code>DGEList()</code> is a function from edgeR, of which limma borrows some loading and
filtering functions. This experiment filters by expression level, and uses
square bracket notation (<code>[]</code>) to reduce the number of rows.</p>
<p>Finally, the expression data is transformed into CPM, counts per million, and
a linear model is applied to the data with <code>lmFit()</code>. <code>topTable()</code> is used to view the most
differentially expressed data.</p>
<div class="sourceCode" id="cb420"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># limma trend</span></span>
<span><span class="va">logCPM</span> <span class="op">&lt;-</span> <span class="fu">cpm</span><span class="op">(</span><span class="va">dge</span>, log<span class="op">=</span><span class="cn">TRUE</span>, prior.count<span class="op">=</span><span class="fl">3</span><span class="op">)</span></span>
<span><span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu">lmFit</span><span class="op">(</span><span class="va">logCPM</span>, <span class="va">design</span><span class="op">)</span></span>
<span><span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu">eBayes</span><span class="op">(</span><span class="va">fit</span>, trend<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu">topTable</span><span class="op">(</span><span class="va">fit</span>, coef<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/nrow.html">ncol</a></span><span class="op">(</span><span class="va">design</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>

</div>
</div>
</div>
<div id="gene-set-enrichment-analysis" class="section level2" number="8.8">
<h2>
<span class="header-section-number">8.8</span> Gene Set Enrichment Analysis<a class="anchor" aria-label="anchor" href="#gene-set-enrichment-analysis"><i class="fas fa-link"></i></a>
</h2>
<p>With the constant evolution of high-throughput sequencing (HTS) technologies,
the size and dimensionality of data generated has been ever increasing. The
question of interest has shifted from how do we generate the data to how do we
make meaningful biological interpretations on a genome-wide level. The simplest
and most common output of HTS experiments is a list of “interesting” genes. In
the specific case of differential gene expression analysis, it is possible and
quite common to obtain hundreds or even thousands of differentially expressed
genes in a single experiment that may be directly or indirectly related to the
phenotype of interest.</p>
<p>While it can be helpful and fruitful to research these genes individually, this
form of personal inspection is limited by one’s domain knowledge and by the size
of the results. On a biological level, it is complicated by the fact that most
processes are often coordinated by the actions of many genes/gene products
working in concert. Cellular signaling pathways, phenotypic differences or
responses to various stimuli are typically associated with changes in the
expression pattern of many genes that share common biological functions or
regulation.</p>
<p>Gene set enrichment analysis is an umbrella term for methods designed to analyze
expression data and capture changes in higher-level biological processes and
pathways by organizing genes into biologically relevant groups or gene sets. We
will discuss the background of two common gene set enrichment analyses
(over-representation analysis and rank-based Gene Set Enrichment Analysis),
their advantages and disadvantages, and walk-through an example of how they can
each be implemented in R. Before this, we will briefly touch upon the definition
of a gene set and describe how to construct or obtain gene sets.</p>
<div id="gene-sets" class="section level3" number="8.8.1">
<h3>
<span class="header-section-number">8.8.1</span> Gene Sets<a class="anchor" aria-label="anchor" href="#gene-sets"><i class="fas fa-link"></i></a>
</h3>
<p>Gene sets are curated lists of genes that are grouped together by some logical
association or pre-existing domain knowledge. Their primary use is to facilitate
the biological interpretation of expression data by capturing high-level
interactions between biologically relevant groups of genes. These sets are
highly flexible and may be constructed based on any <em>a priori</em> knowledge or
classification. For example, one could define a gene set that includes all genes
that are members of the same biochemical pathway or a gene set that consists of
all genes that are induced upon treatment with a particular pharmacological
agent. While it is perfectly valid to construct new gene sets, there are many
existing and curated collections of gene sets that are maintained and
contributed to by communities of scientists throughout the world. Below, we will
highlight some of the major collections that are commonly used and cited in
published work:</p>
<p><strong><a href="https://www.genome.jp/kegg/pathway.html">KEGG Pathways</a></strong></p>
<p>The Kyoto Encyclopedia of Genes and Genomes (KEGG) is a repository for
biological pathway information that the authors describe as a means to
“computerize functional interpretations as part of the pathway reconstruction
process based on the hierarchically structured knowledge about the genomic,
chemical, and network spaces”. The KEGG Pathways database consists of maps
displaying the functional and regulatory relationships of genes within various
metabolic and cell signaling pathways. The KEGG offers a web service that allows
for the extraction of the genes and other information belonging to specific
pathways.</p>
<p><strong><a href="http://geneontology.org/docs/go-annotations/">GO Annotations</a></strong></p>
<p>The Gene Ontology provides a network representation of biological systems from
the molecular-level to the organismal level by defining a graph based
representation of connected terms using a controlled vocabulary. At a high
level, GO annotations consist of three broad ontologies termed molecular
function, cellular component, and biological process. The molecular function is
the specific activity a gene product performs on a molecular level (i.e. a
protein kinase would be annotated with <em>protein kinase activity</em>). The
biological process defines the higher-level programs and pathways that’s
accomplished by the activities of the gene product (i.e. if we take our previous
example of a protein kinase, it’s biological process might be assigned as
<em>signal transduction</em>). The cellular component refers to the localization of the
gene product within the cell (i.e. the cellular component of a protein kinase
might be <em>cytosol</em>). A gene product can be annotated with zero or more terms
from each ontology, and these annotations are based on multiple levels of
evidence from published work. The GO annotations may be accessed or downloaded
directly from their own website, and there exist a number of web services that
use GO annotations in the background (<a href="https://david.ncifcrf.gov">DAVID</a>,
<a href="https://maayanlab.cloud/Enrichr/">enrichR</a>, etc.)</p>
<p><strong><a href="http://www.gsea-msigdb.org/gsea/msigdb/">Molecular Signatures Database</a></strong></p>
<p>The Molecular Signatures Database (MSigDB) is a collection of gene sets curated,
maintained and provided by the Broad Institute and UC San Diego. Although
intended and designed for specific use with the Gene Set Enrichment Analysis
Methodology, they are freely available and only require proper attribution for
other uses. The MSigDB consists of 9 major collections of gene sets: H (hallmark
gene sets), C1 (positional gene sets), C2 (curated gene sets), c3 (regulatory
target gene sets), c4 (computational gene sets), c5 (ontology gene sets), c6
(oncogenic signature gene sets), c7 (immunologic signature gene sets), c8 (cell
type signature gene sets). They are available in formats ready for use in the
GSEA methodology and other formats that are easily imported into various
settings for custom use. The gene sets as well as the GSEA methodology are
available directly from their website. There are a number of R-specific packages
that have been developed for working directly with these gene sets such as
<a href="https://bioconductor.org/packages/release/bioc/html/GSEABase.html">GSEABase</a> or
<a href="http://bioconductor.org/packages/release/bioc/html/fgsea.html">fgsea</a>, which we
will discuss later.</p>
</div>
<div id="working-with-gene-sets-in-r" class="section level3" number="8.8.2">
<h3>
<span class="header-section-number">8.8.2</span> Working with gene sets in R<a class="anchor" aria-label="anchor" href="#working-with-gene-sets-in-r"><i class="fas fa-link"></i></a>
</h3>
<p>We will walk through two quick examples of how to read in an example collection
of gene sets. We will utilize the Hallmarks gene set collection, downloaded
directly from the MSigDB website, which consists of 50 gene sets representing
well-defined biological processes and generated by aggregating together many
pre-existing gene sets. The MSigDB provides these gene set collections in the
GMT format. These files are tab-delimited and each row in the format represents
one gene set with the first column being the name of the gene set, and the
second column a short description. The remaining columns each represent a gene
and unequal lengths of columns per row is allowed. As is, the GMT format is
designed to work specifically with the GSEA methodology developed and provided
by the Broad Institute. However, we will also show two ways to manually parse
these gene collections for exploration and further use in R.</p>
<p>The first way to parse these gene sets would be to use various tidyverse
functions that you should be familiar with already to construct a tibble.
Essentially, we read the file, rename the first two columns for convenience, and
use a combination of <code>pivot_longer()</code> and <code><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by()</a></code> to quickly access genes
by pathway. Below, we have the results of these operations and used them to
display a summary of the number of genes in all the pathways contained with the
hallmark pathways gene collection.</p>
<div class="sourceCode" id="cb421"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">read_delim</span><span class="op">(</span><span class="st">'h.all.v7.5.1.symbols.gmt'</span>, delim<span class="op">=</span><span class="st">'\t'</span>, col_names<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/rename.html">rename</a></span><span class="op">(</span>Pathway <span class="op">=</span> <span class="va">X1</span>, Desc<span class="op">=</span><span class="va">X2</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="op">-</span><span class="va">Desc</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">pivot_longer</span><span class="op">(</span><span class="op">!</span><span class="va">Pathway</span>, values_to<span class="op">=</span><span class="st">'genes'</span>, values_drop_na<span class="op">=</span><span class="cn">TRUE</span>, names_to <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">Pathway</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html">summarise</a></span><span class="op">(</span>n<span class="op">=</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/context.html">n</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code>## # A tibble: 50 x 2
##    Pathway                              n
##    &lt;chr&gt;                            &lt;int&gt;
##  1 HALLMARK_ADIPOGENESIS              200
##  2 HALLMARK_ALLOGRAFT_REJECTION       200
##  3 HALLMARK_ANDROGEN_RESPONSE         100
##  4 HALLMARK_ANGIOGENESIS               36
##  5 HALLMARK_APICAL_JUNCTION           200
##  6 HALLMARK_APICAL_SURFACE             44
##  7 HALLMARK_APOPTOSIS                 161
##  8 HALLMARK_BILE_ACID_METABOLISM      112
##  9 HALLMARK_CHOLESTEROL_HOMEOSTASIS    74
## 10 HALLMARK_COAGULATION               138
## # ... with 40 more rows</code></pre>
<p>We can see that there are 50 total pathways in the hallmarks gene collection
that contain varying numbers of genes. We discarded the description column and
if we were to save the results after using <code>pivot_longer()</code> instead of piping
them to <code>group_by</code> and <code><a href="https://dplyr.tidyverse.org/reference/summarise.html">summarise()</a></code>, we would have a tibble in the “long”
format with each row representing a single pathway and single gene.</p>
<p>However, there do exist various packages that have been developed to
specifically handle gene set data in R such as the previously mentioned
GSEABase. This is a collection of functions and class-based objects that
facilitate working with gene sets. The foundation of the GSEABase package is the
GeneSet and GeneSetCollection classes which store gene sets and metadata or a
collection of GeneSet objects, respectively. We will use the GSEABase package to
read in the collection of gene sets we previously downloaded.</p>
<div class="sourceCode" id="cb423"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="st">'GSEABase'</span><span class="op">)</span></span>
<span><span class="va">hallmarks_gmt</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GSEABase/man/getObjects.html">getGmt</a></span><span class="op">(</span>con<span class="op">=</span><span class="st">'h.all.v7.5.1.symbols.gmt'</span><span class="op">)</span></span>
<span><span class="va">hallmarks_gmt</span></span></code></pre></div>
<pre><code>## GeneSetCollection
##   names: HALLMARK_TNFA_SIGNALING_VIA_NFKB, HALLMARK_HYPOXIA, ..., HALLMARK_PANCREAS_BETA_CELLS (50 total)
##   unique identifiers: JUNB, CXCL2, ..., SRP14 (4383 total)
##   types in collection:
##     geneIdType: NullIdentifier (1 total)
##     collectionType: NullCollection (1 total)</code></pre>
<p>If we simply access the hallmarks_gmt variable, we can see that it is a
GeneSetCollection object containing 0 total gene
sets which encompass 4383 unique identifiers or HGNC symbols. Although this
package supports a range of functions, we will focus on the basics. For a more
thorough description of the classes and methods, please read their extended
documentation available
<a href="https://bioconductor.org/packages/release/bioc/manuals/GSEABase/man/GSEABase.pdf">here</a>.</p>
<p>The geneIds method will return a list with each pathway as a named vector of associated gene ids:</p>
<div class="sourceCode" id="cb425"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/GSEABase/man/GeneSet-class.html">geneIds</a></span><span class="op">(</span><span class="va">hallmarks_gmt</span><span class="op">)</span>, <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<pre><code>## $HALLMARK_TNFA_SIGNALING_VIA_NFKB
##   [1] "JUNB"     "CXCL2"    "ATF3"     "NFKBIA"   "TNFAIP3"  "PTGS2"   
##   [7] "CXCL1"    "IER3"     "CD83"     "CCL20"    "CXCL3"    "MAFF"    
##  [13] "NFKB2"    "TNFAIP2"  "HBEGF"    "KLF6"     "BIRC3"    "PLAUR"   
##  [19] "ZFP36"    "ICAM1"    "JUN"      "EGR3"     "IL1B"     "BCL2A1"  
##  [25] "PPP1R15A" "ZC3H12A"  "SOD2"     "NR4A2"    "IL1A"     "RELB"    
##  [31] "TRAF1"    "BTG2"     "DUSP1"    "MAP3K8"   "ETS2"     "F3"      
##  [37] "SDC4"     "EGR1"     "IL6"      "TNF"      "KDM6B"    "NFKB1"   
##  [43] "LIF"      "PTX3"     "FOSL1"    "NR4A1"    "JAG1"     "CCL4"    
##  [49] "GCH1"     "CCL2"     "RCAN1"    "DUSP2"    "EHD1"     "IER2"    
##  [55] "REL"      "CFLAR"    "RIPK2"    "NFKBIE"   "NR4A3"    "PHLDA1"  
##  [61] "IER5"     "TNFSF9"   "GEM"      "GADD45A"  "CXCL10"   "PLK2"    
##  [67] "BHLHE40"  "EGR2"     "SOCS3"    "SLC2A6"   "PTGER4"   "DUSP5"   
##  [73] "SERPINB2" "NFIL3"    "SERPINE1" "TRIB1"    "TIPARP"   "RELA"    
##  [79] "BIRC2"    "CXCL6"    "LITAF"    "TNFAIP6"  "CD44"     "INHBA"   
##  [85] "PLAU"     "MYC"      "TNFRSF9"  "SGK1"     "TNIP1"    "NAMPT"   
##  [91] "FOSL2"    "PNRC1"    "ID2"      "CD69"     "IL7R"     "EFNA1"   
##  [97] "PHLDA2"   "PFKFB3"   "CCL5"     "YRDC"     "IFNGR2"   "SQSTM1"  
## [103] "BTG3"     "GADD45B"  "KYNU"     "G0S2"     "BTG1"     "MCL1"    
## [109] "VEGFA"    "MAP2K3"   "CDKN1A"   "CCN1"     "TANK"     "IFIT2"   
## [115] "IL18"     "TUBB2A"   "IRF1"     "FOS"      "OLR1"     "RHOB"    
## [121] "AREG"     "NINJ1"    "ZBTB10"   "PLPP3"    "KLF4"     "CXCL11"  
## [127] "SAT1"     "CSF1"     "GPR183"   "PMEPA1"   "PTPRE"    "TLR2"    
## [133] "ACKR3"    "KLF10"    "MARCKS"   "LAMB3"    "CEBPB"    "TRIP10"  
## [139] "F2RL1"    "KLF9"     "LDLR"     "TGIF1"    "RNF19B"   "DRAM1"   
## [145] "B4GALT1"  "DNAJB4"   "CSF2"     "PDE4B"    "SNN"      "PLEK"    
## [151] "STAT5A"   "DENND5A"  "CCND1"    "DDX58"    "SPHK1"    "CD80"    
## [157] "TNFAIP8"  "CCNL1"    "FUT4"     "CCRL2"    "SPSB1"    "TSC22D1" 
## [163] "B4GALT5"  "SIK1"     "CLCF1"    "NFE2L2"   "FOSB"     "PER1"    
## [169] "NFAT5"    "ATP2B1"   "IL12B"    "IL6ST"    "SLC16A6"  "ABCA1"   
## [175] "HES1"     "BCL6"     "IRS2"     "SLC2A3"   "CEBPD"    "IL23A"   
## [181] "SMAD3"    "TAP1"     "MSC"      "IFIH1"    "IL15RA"   "TNIP2"   
## [187] "BCL3"     "PANX1"    "FJX1"     "EDN1"     "EIF1"     "BMP2"    
## [193] "DUSP4"    "PDLIM5"   "ICOSLG"   "GFPT2"    "KLF2"     "TNC"     
## [199] "SERPINB8" "MXD1"    
## 
## $HALLMARK_HYPOXIA
##   [1] "PGK1"     "PDK1"     "GBE1"     "PFKL"     "ALDOA"    "ENO2"    
##   [7] "PGM1"     "NDRG1"    "HK2"      "ALDOC"    "GPI"      "MXI1"    
##  [13] "SLC2A1"   "P4HA1"    "ADM"      "P4HA2"    "ENO1"     "PFKP"    
##  [19] "AK4"      "FAM162A"  "PFKFB3"   "VEGFA"    "BNIP3L"   "TPI1"    
##  [25] "ERO1A"    "KDM3A"    "CCNG2"    "LDHA"     "GYS1"     "GAPDH"   
##  [31] "BHLHE40"  "ANGPTL4"  "JUN"      "SERPINE1" "LOX"      "GCK"     
##  [37] "PPFIA4"   "MAFF"     "DDIT4"    "SLC2A3"   "IGFBP3"   "NFIL3"   
##  [43] "FOS"      "RBPJ"     "HK1"      "CITED2"   "ISG20"    "GALK1"   
##  [49] "WSB1"     "PYGM"     "STC1"     "ZNF292"   "BTG1"     "PLIN2"   
##  [55] "CSRP2"    "VLDLR"    "JMJD6"    "EXT1"     "F3"       "PDK3"    
##  [61] "ANKZF1"   "UGP2"     "ALDOB"    "STC2"     "ERRFI1"   "ENO3"    
##  [67] "PNRC1"    "HMOX1"    "PGF"      "GAPDHS"   "CHST2"    "TMEM45A" 
##  [73] "BCAN"     "ATF3"     "CAV1"     "AMPD3"    "GPC3"     "NDST1"   
##  [79] "IRS2"     "SAP30"    "GAA"      "SDC4"     "STBD1"    "IER3"    
##  [85] "PKLR"     "IGFBP1"   "PLAUR"    "CAVIN3"   "CCN5"     "LARGE1"  
##  [91] "NOCT"     "S100A4"   "RRAGD"    "ZFP36"    "EGFR"     "EDN2"    
##  [97] "IDS"      "CDKN1A"   "RORA"     "DUSP1"    "MIF"      "PPP1R3C" 
## [103] "DPYSL4"   "KDELR3"   "DTNA"     "ADORA2B"  "HS3ST1"   "CAVIN1"  
## [109] "NR3C1"    "KLF6"     "GPC4"     "CCN1"     "TNFAIP3"  "CA12"    
## [115] "HEXA"     "BGN"      "PPP1R15A" "PGM2"     "PIM1"     "PRDX5"   
## [121] "NAGK"     "CDKN1B"   "BRS3"     "TKTL1"    "MT1E"     "ATP7A"   
## [127] "MT2A"     "SDC3"     "TIPARP"   "PKP1"     "ANXA2"    "PGAM2"   
## [133] "DDIT3"    "PRKCA"    "SLC37A4"  "CXCR4"    "EFNA3"    "CP"      
## [139] "KLF7"     "CCN2"     "CHST3"    "TPD52"    "LXN"      "B4GALNT2"
## [145] "PPARGC1A" "BCL2"     "GCNT2"    "HAS1"     "KLHL24"   "SCARB1"  
## [151] "SLC25A1"  "SDC2"     "CASP6"    "VHL"      "FOXO3"    "PDGFB"   
## [157] "B3GALT6"  "SLC2A5"   "SRPX"     "EFNA1"    "GLRX"     "ACKR3"   
## [163] "PAM"      "TGFBI"    "DCN"      "SIAH2"    "PLAC8"    "FBP1"    
## [169] "TPST2"    "PHKG1"    "MYH9"     "CDKN1C"   "GRHPR"    "PCK1"    
## [175] "INHA"     "HSPA5"    "NDST2"    "NEDD4L"   "TPBG"     "XPNPEP1" 
## [181] "IL6"      "SLC6A6"   "MAP3K1"   "LDHC"     "AKAP12"   "TES"     
## [187] "KIF5A"    "LALBA"    "COL5A1"   "GPC1"     "HDLBP"    "ILVBL"   
## [193] "NCAN"     "TGM2"     "ETS1"     "HOXB9"    "SELENBP1" "FOSL2"   
## [199] "SULT2B1"  "TGFB3"</code></pre>
<p>The names method will return all of the gene set names contained within a specific collection:</p>
<div class="sourceCode" id="cb427"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">hallmarks_gmt</span><span class="op">)</span>,<span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<pre><code>## [1] "HALLMARK_TNFA_SIGNALING_VIA_NFKB" "HALLMARK_HYPOXIA"</code></pre>
<p>We can access a specific gene set contained within this collection by referring
to its name and using the following notation:</p>
<div class="sourceCode" id="cb429"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hallmarks_gmt</span><span class="op">[[</span><span class="st">'HALLMARK_ANGIOGENESIS'</span><span class="op">]</span><span class="op">]</span></span></code></pre></div>
<pre><code>## setName: HALLMARK_ANGIOGENESIS 
## geneIds: VCAN, POSTN, ..., CXCL6 (total: 36)
## geneIdType: Null
## collectionType: Null 
## details: use 'details(object)'</code></pre>
<p>Simply accessing the object will provide a high-level summary of the information
contained within. To access a specific value of this GeneSet object, we would
call one of the slots (a core concept in object-oriented programming). In our
particular case, we could extract the gene names contained assigned to this
GeneSet by calling the <code>geneIds</code> slot as shown below to return a vector of the
gene names. We can see the first five below and also the length of the returned
vector by using the base R <code><a href="https://rdrr.io/r/base/length.html">length()</a></code> function:</p>
<div class="sourceCode" id="cb431"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">hallmarks_gmt</span><span class="op">[[</span><span class="st">'HALLMARK_ANGIOGENESIS'</span><span class="op">]</span><span class="op">]</span><span class="op">@</span><span class="va">geneIds</span><span class="op">)</span></span></code></pre></div>
<pre><code>## [1] "VCAN"   "POSTN"  "FSTL1"  "LRPAP1" "STC1"   "LPL"</code></pre>
<div class="sourceCode" id="cb433"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">hallmarks_gmt</span><span class="op">[[</span><span class="st">'HALLMARK_ANGIOGENESIS'</span><span class="op">]</span><span class="op">]</span><span class="op">@</span><span class="va">geneIds</span><span class="op">)</span></span></code></pre></div>
<pre><code>## [1] 36</code></pre>
<p>GSEABase includes a number of built-in functions for reading gene sets in from
various sources and performing common operations such as set intersections, set
differences, and ID conversions. We will demonstrate the usage of some of these
in the next section covering Over-representation analysis.</p>
</div>
<div id="over-representation-analysis" class="section level3" number="8.8.3">
<h3>
<span class="header-section-number">8.8.3</span> Over-representation Analysis<a class="anchor" aria-label="anchor" href="#over-representation-analysis"><i class="fas fa-link"></i></a>
</h3>
<p>One of the most common ways to utilize gene sets to evaluate gene expression
results is to perform Over-representation Analysis (ORA). Let us assume that we
have obtained a list of differentially expressed genes from an experiment. We
are curious to know if within this list of differentially expressed genes, do we
see an <em>over-representation</em> or <em>enrichment</em> of genes belonging to gene sets of
interest? In more general terms, the goal of ORA is to determine how likely it
is that there is a non-random association between a gene being differentially
expressed and having membership in a chosen (and ideally relevant) gene set. In
R, we can do a simple ORA by utilizing a Fisher’s exact test and a contingency
table.</p>
<p>For a purely hypothetical example, let us assume that we have performed
differential gene expression analysis between two different cell lines. We
obtain a list of 10,000 total genes (our background) discovered in the
experiment and find that at our chosen statistical threshold, 1,000 of these are
differentially expressed. To keep things simple, we will perform a single ORA
test against the Hallmarks Angiogenesis gene set using a sample list of 1000
differentially expressed genes selected from data generated by <a href="https://pubmed.ncbi.nlm.nih.gov/23700391/">Marisa et al.
2013</a>. The Hallmarks Angiogenesis
gene set consists of 36 genes and we find that 13 of these are also present in
our list of differentially expressed genes.</p>
<div class="box warning">
<p>Please note that we are removing this list of genes from its original meaning
and context found in the publication and simply using it to demonstrate the
basic steps occurring during ORA. All of these numbers and lists were
arbitrarily chosen and this experimental setup is purely hypothetical.</p>
<p>Another issue to note is that typically the “background” list should represent
the entire pool of genes from which any differentially expressed genes could
have been selected. For expression experiments, it is typical to choose all of
the detected genes (regardless of significance) as the background. The number of
genes in the organism’s genome could potentially also be an appropriate
“background”</p>
<p>Also, it is important to keep in mind that in reality, ORA is nearly always
performed on a larger scale against a variety of different gene sets. This
allows for the unbiased discovery of potentially novel and unexpected enrichment
in other biological areas of interest. It also necessitates the need for
multiple-testing correction, which we have discussed in <a href="#multiple-hypothesis-testing">multiple hypothesis
testing</a>.</p>
</div>
<p>To begin, we would want to prepare a contingency table which describes the
various overlaps between our sets of interest. For a 2x2 contingency table,
these four values will be:</p>
<ol style="list-style-type: decimal">
<li>Genes present in our list of differentially expressed genes and present in our gene set</li>
<li>Genes present in our list of differentially expressed genes and not present in our gene set</li>
<li>Genes not present in our list of differentially expressed genes and present in our gene set</li>
<li>Genes not present in our list of differentially expressed genes and not present in our gene set</li>
</ol>
<p>To demonstrate what this would look like, we have manually constructed a
contingency table with labels and totals added below. If you look at the margins
of the table and recall the previously given values above, you can reconstruct
the logic used to generate each of the values in all the cells.</p>
<div class="inline-table"><table class="table table-sm">
<colgroup>
<col width="21%">
<col width="32%">
<col width="38%">
<col width="7%">
</colgroup>
<thead><tr class="header">
<th align="left"></th>
<th align="right">Differentially Expressed</th>
<th align="right">Not Differentially Expressed</th>
<th align="right">Total</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">In Gene Set</td>
<td align="right">13</td>
<td align="right">23</td>
<td align="right">36</td>
</tr>
<tr class="even">
<td align="left">Not in Gene Set</td>
<td align="right">987</td>
<td align="right">8977</td>
<td align="right">9964</td>
</tr>
<tr class="odd">
<td align="left">Total</td>
<td align="right">1000</td>
<td align="right">9000</td>
<td align="right">10000</td>
</tr>
</tbody>
</table></div>
<p>For the purposes of this example, we are reading in our differentially expressed
genes from an external file, but this vector could be generated in any number of
ways depending upon where and how your results are stored. Following good coding
practices, we will write a small function that takes this list of DE genes and a
GeneSet object to programmatically generate a contingency table:</p>
<div class="box note">
<p>There are many ways to construct a contingency table. This is just one way that
was chosen to make calculations of the values contained within the table
transparent and easy to understand.</p>
</div>
<div class="sourceCode" id="cb435"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#load and read our list of DE genes contained within a newline delimited txt file</span></span>
<span><span class="va">de_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/scan.html">scan</a></span><span class="op">(</span><span class="st">'example_de_list.txt'</span>, what<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/character.html">character</a></span><span class="op">(</span><span class="op">)</span>, sep<span class="op">=</span><span class="st">'\n'</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#define a function that takes a list of DE genes, and a specific GeneSet from a GeneSetCollection to generate a contingency table</span></span>
<span><span class="co">#using set operations in GSEABase</span></span>
<span><span class="va">make_contingency</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">de_list</span>, <span class="va">GeneSetobj</span><span class="op">)</span> <span class="op">{</span></span>
<span></span>
<span></span>
<span>  <span class="co">#make our de list into a simple GeneSet object using GSEABase</span></span>
<span>  <span class="va">de_geneset</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GSEABase/man/GeneSet-methods.html">GeneSet</a></span><span class="op">(</span><span class="va">de_list</span>, setName<span class="op">=</span><span class="st">'1000 DE genes'</span><span class="op">)</span></span>
<span></span>
<span>  <span class="co">#If we had the full results, we could determine this value without manually setting it</span></span>
<span>  <span class="va">background_len</span> <span class="op">&lt;-</span> <span class="fl">10000</span></span>
<span></span>
<span>  <span class="co">#Calculate the values inside the contingency table using set operations</span></span>
<span>  <span class="va">de_in</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="op">(</span><span class="va">de_geneset</span> <span class="op">&amp;</span> <span class="va">GeneSetobj</span><span class="op">)</span><span class="op">@</span><span class="va">geneIds</span><span class="op">)</span></span>
<span>  <span class="va">de_notin</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/sets.html">setdiff</a></span><span class="op">(</span><span class="va">de_geneset</span>, <span class="va">GeneSetobj</span><span class="op">)</span><span class="op">@</span><span class="va">geneIds</span><span class="op">)</span></span>
<span>  <span class="va">set_notde</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/sets.html">setdiff</a></span><span class="op">(</span><span class="va">GeneSetobj</span>, <span class="va">de_geneset</span><span class="op">)</span><span class="op">@</span><span class="va">geneIds</span><span class="op">)</span></span>
<span>  <span class="va">notin_notde</span> <span class="op">&lt;-</span> <span class="va">background_len</span> <span class="op">-</span> <span class="op">(</span><span class="va">de_in</span> <span class="op">+</span> <span class="va">de_notin</span> <span class="op">+</span> <span class="va">set_notde</span><span class="op">)</span></span>
<span></span>
<span></span>
<span>  <span class="co">#return a matrix of the contingency values</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/function.html">return</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">de_in</span>, <span class="va">de_notin</span>, <span class="va">set_notde</span>, <span class="va">notin_notde</span><span class="op">)</span>, nrow<span class="op">=</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="va">contingency_table</span> <span class="op">&lt;-</span> <span class="fu">make_contingency</span><span class="op">(</span><span class="va">de_genes</span>, <span class="va">hallmarks_gmt</span><span class="op">[[</span><span class="st">'HALLMARK_ANGIOGENESIS'</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">contingency_table</span></span></code></pre></div>
<pre><code>##      [,1] [,2]
## [1,]   13   23
## [2,]  987 8977</code></pre>
<p>We perform the Fisher’s Exact test using the built-in R function <code><a href="https://rdrr.io/r/stats/fisher.test.html">fisher.test()</a></code>
and view the summarized output by simply calling the variable where we stored
the test results:</p>
<div class="sourceCode" id="cb437"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fisher_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/fisher.test.html">fisher.test</a></span><span class="op">(</span><span class="va">contingency_table</span>, alternative<span class="op">=</span><span class="st">'greater'</span><span class="op">)</span></span>
<span><span class="va">fisher_results</span></span></code></pre></div>
<pre><code>## 
##  Fisher's Exact Test for Count Data
## 
## data:  contingency_table
## p-value = 2.382e-05
## alternative hypothesis: true odds ratio is greater than 1
## 95 percent confidence interval:
##  2.685749      Inf
## sample estimates:
## odds ratio 
##   5.139308</code></pre>
<div class="box note">
<p>Specific values of the results can be accessed by the $notation (i.e.
fisher_results$p.value). The full list of returned values may be found in the R
documentation for <code><a href="https://rdrr.io/r/stats/fisher.test.html">fisher.test()</a></code></p>
</div>
<p>Back to our hypothetical example and focusing on the p-value returned of <code>r fisher_results$p.value</code>, we can interpret this as the probability of randomly
obtaining results as or more extreme than what we observed assuming the null
hypothesis that there is no association between differential expression and gene
set membership is true. Based on these results and if this p-value was below our
pre-determined statistical threshold, we could make the conclusion that there is
an <em>enrichment</em> or <em>over-representation</em> of our differentially expressed genes
from this experiment in the Hallmark Angiogenesis gene set. Relating this back
to the experiment, we might hypothesize that the differences between our two
cell lines might be driving gene expression changes that result in alterations
in genes involved in angiogenesis. This <em>might</em> motivate potential further <em>in
vitro</em> experiments on these cell lines, including migration and proliferation
assays, that could reveal if this enrichment of angiogenesis genes is reflected
at a phenotypic or functional level.</p>
<p>ORA is a quick and useful way to generate further hypotheses to investigate
specific mechanisms of action or regulation. For example, after identifying a
gene set as being <em>enriched</em> or <em>over-represented</em>, one could further test the
specific genes in the set by examining their directionality of change or asking
if dependent pathways/networks are also perturbed.</p>
<p>One major limitation of ORA is that it relies on the choice of arbitrary
statistical thresholds to define “interesting” or “differentially expressed”
genes. To reiterate again, p-value thresholds hold no inherent biological
meaning and are subjectively determined. Changing the p-value threshold may
result in dramatic differences in the outcome of ORA. Additionally, expression
datasets may measure tens of thousands of genes in a single experiment, and
filtering by a p-value threshold may discard potentially useful information.</p>
<p>ORA (though often modified with slightly different statistical methodologies) is
implemented in a number of different R packages such as
<a href="https://bioconductor.org/packages/release/bioc/html/topGO.html">topGO</a> or
various web services including <a href="https://david.ncifcrf.gov">DAVID</a> and
<a href="https://maayanlab.cloud/Enrichr/">enrichR</a></p>
</div>
<div id="rank-based-analysis" class="section level3" number="8.8.4">
<h3>
<span class="header-section-number">8.8.4</span> Rank-based Analysis<a class="anchor" aria-label="anchor" href="#rank-based-analysis"><i class="fas fa-link"></i></a>
</h3>
<p>We will refer to the specific method developed by the Broad Institute and UC San
Diego as Gene Set Enrichment Analysis (note the capitalization). This specific
methodology should not be confused with gene set enrichment analysis, which we
use as an umbrella term covering the many statistical methodologies used to
analyze gene sets.</p>
<p><a href="biology-bioinformatics.html#gene-set-enrichment-analysis">Gene Set Enrichment Analysis</a> (<a href="https://www.gsea-msigdb.org/gsea/index.jsp" class="uri">https://www.gsea-msigdb.org/gsea/index.jsp</a>) was
a method developed to facilitate the biological analysis of genome-wide
expression experiments. It is a rank-based method which utilizes all of the
information from an expression dataset instead of relying on pre-determined
statistical thresholds. In the simplest case, GSEA organizes expression data
into two classes, and ranks all discovered genes by a chosen metric correlating
their expression to these classes. Then, for a pre-defined set of genes, GSEA
tests whether the members of this particular gene set occur more frequently at
the top or bottom of the ranked list or are randomly distributed throughout.
This ranking can be done by a number of different measures, but common ones
include signal-to-noise ratio, signed log p-value or log fold change estimates.
These pre-defined gene sets are flexible and may be manually constructed or
drawn from the many curated databases of gene sets.</p>
<p>Behind the scenes, GSEA functions by descending through the ranked list and
increasing a cumulative score when a gene is encountered that’s contained within
the chosen gene set and decreasing that score when it encounters a gene that is
not within that gene set. This incrementing is weighted to put more emphasis on
genes that are found at the extremes of the ranked list and the final score
taken as the enrichment score (ES) is the maximum of the deviation from zero.
From this score, a p-value is determined by permuting the phenotype labels to
generate a null distribution to compare against. To account for multiple
hypothesis correction, the ES is normalized by the size of the gene set to
create a normalized enrichment score (NES) and the p-values are subjected to
Benjamini-Hochberg correction to generate FDR values for each NES.</p>
<p>To make this explanation more transparent, let us assume that we have performed
a RNAseq experiment to determine which genes are changed when we knock out our
gene of interest in a model cell line. We have obtained our results, which
consists of a list of all genes discovered in the experiment and associated
statistical measures such as p-value and log fold change estimates. To perform a
basic GSEA, we would sort this list of genes, without respect to significance,
by descending fold change to generate a ranked list where genes that are
upregulated (positive fold change) are at the “top” of our list and genes that
are downregulated (negative fold change) are at the “bottom” of our list. We
would use this ranked list and any number or combination of gene sets to perform
GSEA using one of the available means.</p>
<p>GSEA is available as a java application with a graphical user interface as well
as a command-line utility for use in HPC cluster environments. There also exist
several packages in R that also implement the core GSEA algorithm along with
various changes and refinements to the underlying statistical methodology. In
the following section, we will refer to the usage of one such package, fgsea.
Under most circumstances, GSEA can be run with default parameters but please see
the official documentation for a list of customizable parameters and situations
in which to adjust them.</p>
<p>GSEA will return a list of results for gene sets tested consisting of their
associated enrichment scores, and various statistical measures for a given
ranked input With GSEA results, it is typical to set a permissive FDR threshold
to consider which gene sets are significant. Further inspection of individual
gene sets may be performed by analyzing the leading edge subset for each gene
set. This is the group of genes that occur before and contribute to the maximum
deviation from zero score for a given gene set during the analysis. These genes
are typically interpreted as being the more relevant subset of genes that likely
contribute the most to the enrichment signal, and are typically used in further
investigations comparing leading edge subsets from other gene sets or applying
domain knowledge to predict their regulatory effects and functions with respect
to their annotated biological pathway or gene set.</p>
<p>GSEA is a highly flexible method that incorporates information from all of the
data generated in HTS experiments. The choice of ranking methods and gene sets
can all be tailored to the specific experiment or question at hand, and it is a
simple yet fast method that can detect more subtle changes in gene networks than
ORA.</p>
<div class="box warning">
<p>The interpretation of a GSEA result needs to be made with care. Let’s assume
that we have performed GSEA using fold Change as our ranking metric and we find
that the “KEGG Glycolysis” pathway is found to be enriched with a positive
normalized enrichment score and is statistically significant. Alone, this does
not necessarily imply that glycolysis is more active between the conditions of
interest. It simply allows for the conclusion that there is an enrichment of
genes belonging to this gene set among the genes with a positive log fold change
in the experiment. This <em>might</em> imply that your experimental conditions are
affecting some underlying transcriptional regulation that results in these genes
being up-regulated. However, it could also be explained by feedback mechanisms,
direct or indirect, from related biological pathways or gene networks. It <em>does
not</em> allow you to directly conclude that the glycolysis pathway itself is more
active at a functional level. To make that assertion, you would need to perform
further <em>in vitro</em> or <em>in vivo</em> functional assays that directly measure the
activity and outputs of glycolysis given your experimental conditions.</p>
<p>It is important to remember that unless specified, gene sets will often contain
a mix of genes with different regulatory activities including both activators
and inhibitors of that pathway. In addition, for many biological pathways, there
are other factors that influence its regulation that may not be apparent at
a gene expression level. In the specific example of metabolic pathways, it is
common for the enzymatic products of subsequent reactions to feedback and
regulate the activity and flux through the pathway or connected pathways. This
regulation is not easily discerned at the gene expression level alone. All of
this is to say that the results from GSEA must be interpreted with care and
integrated with other analyses and knowledge to gain a holistic understanding
of its meaning.</p>
</div>
</div>
<div id="fgsea" class="section level3" number="8.8.5">
<h3>
<span class="header-section-number">8.8.5</span> fgsea<a class="anchor" aria-label="anchor" href="#fgsea"><i class="fas fa-link"></i></a>
</h3>
<p>To perform GSEA analysis in R, we will be using the <a href="biology-bioinformatics.html#fgsea">fgsea</a>
(<a href="http://bioconductor.org/packages/release/bioc/html/fgsea.html" class="uri">http://bioconductor.org/packages/release/bioc/html/fgsea.html</a>) package. fgsea
uses the core algorithm behind rank-based GSEA but with additional statistical
methodologies meant to provide better estimations of small p-values. For this
example, we have generated simulated and completely artificial data for the
ranked list. We have chosen 1000 genes to represent all of the genes discovered
in our experiment, annotated them with random values for their fold change and
sorted the list in descending order. The top of our list corresponds to genes
with a positive fold change and the bottom those with a negative fold change.
For this hypothetical experiment, let us assume that genes that have a positive
fold change are upregulated in our condition of interest vs. the reference and
genes that have a negative fold change are downregulated in our condition of
interest vs. the reference.</p>
<div class="box note">
<p>The genes that served as our ranked list input were taken from an experiment
looking at differences in cancer subtypes. Although our fold change values
associated with them were generated randomly, there is likely a strong residual
signal / bias towards various cancer pathways as we will see in our results
below</p>
</div>
<div class="sourceCode" id="cb439"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/ctlab/fgsea/">'fgsea'</a></span><span class="op">)</span></span></code></pre></div>
<p>fgsea takes as input, a named vector of gene-level statistics (a ranked list)
and a named list of gene sets. The gene names in the ranked list must exactly
match how they appear in the gene sets (see the note below for additional
considerations to keep in mind). By default, the <code><a href="https://rdrr.io/pkg/fgsea/man/fgsea.html">fgsea()</a></code> function will run a
pre-ranked GSEA meaning that you must generate the ranked list manually
beforehand. Importantly, it will not check your ranked list input and will
assume it is sorted numerically in descending order of your ranking metric.</p>
<p>:::{.box .warning}
Gene sets downloaded from the MSigDB are sets of <strong>human</strong> genes and are
represented by either NCBI Entrez IDs or HGNC symbols. For convenience, we have
been working with gene sets containing HGNC symbols. It is important to remember
to check what identifier system and what species the gene set is provided as.
This also becomes relevant when performing rank-based GSEA on data generated in
non-human species. In these situations, you will need to convert the gene IDs
between species and assuming you’re using a pre-defined gene set, match the
identifiers from your ranked list to the format found in the gene set.</p>
<p>This issue typically becomes relevant when trying to use the MSigDB gene sets to
compare against a ranked list of genes generated from HTS experiments in <em>Mus
musculus</em>. MGI symbols typically begin with an uppercase letter and are followed
by all lowercase letters or numbers. While there are examples of exact
concordance where true orthologous mouse and human genes have matching MGI and
HGNC symbols only differing in case (i.e. Gzmb and GZMB), you should <strong>not</strong>
simply convert MGI symbols to all uppercase. This will work for the subset of
genes where the orthologs have the same symbol but will misidentify many genes
where they do not share the same base name. One way to properly perform this
kind of ‘orthology’ mapping is to use <code>biomaRt</code>, which we have discussed
previously. :::</p>
<p>For this example, we have already made simulated vectors containing the gene
names and their associated fold change values. We will generate the appropriate
format using the <code>setNames</code> function on our two vectors and we can use both
<code>head</code> and <code>tail</code> to quickly check the sorting. For non-simulated data, you
could generate these vectors in a number of ways depending upon the format
of your results data.</p>
<p>These are the top 5 genes in our ranked list:</p>
<div class="sourceCode" id="cb440"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rnk_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/setNames.html">setNames</a></span><span class="op">(</span><span class="va">rnks</span>, <span class="va">de_genes</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">rnk_list</span><span class="op">)</span></span></code></pre></div>
<pre><code>##  RBMS1P1    RBMS1     GAS1    SFRP2   CCDC80      MGP 
## 9.965527 9.907958 9.888917 9.828172 9.826594 9.802787</code></pre>
<p>And these are the bottom 5 genes in the list:</p>
<div class="sourceCode" id="cb442"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">tail</a></span><span class="op">(</span><span class="va">rnk_list</span><span class="op">)</span></span></code></pre></div>
<pre><code>##      KLF7   ANKDD1A    RIMKLB     RGS19    PLPPR4     TENM4 
## -9.865186 -9.922858 -9.930453 -9.931198 -9.959923 -9.984908</code></pre>
<p>For our gene sets, we will use the same Hallmarks gene set provided by MSigDB.
There are two ways we can load these gene sets in the appropriate format
expected. The first is to use one of the built-in functions in fgsea,
<code>gmtPathways</code>, which will read directly from the GMT file. We have displayed
the contents of a randomly selected gene set below:</p>
<div class="sourceCode" id="cb444"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hallmark_pathways_fgsea</span> <span class="op">&lt;-</span> <span class="fu">fgsea</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/fgsea/man/gmtPathways.html">gmtPathways</a></span><span class="op">(</span><span class="st">'h.all.v7.5.1.symbols.gmt'</span><span class="op">)</span></span>
<span><span class="va">hallmark_pathways_fgsea</span><span class="op">$</span><span class="va">HALLMARK_TGF_BETA_SIGNALING</span></span></code></pre></div>
<pre><code>##  [1] "TGFBR1"   "SMAD7"    "TGFB1"    "SMURF2"   "SMURF1"   "BMPR2"   
##  [7] "SKIL"     "SKI"      "ACVR1"    "PMEPA1"   "NCOR2"    "SERPINE1"
## [13] "JUNB"     "SMAD1"    "SMAD6"    "PPP1R15A" "TGIF1"    "FURIN"   
## [19] "SMAD3"    "FKBP1A"   "MAP3K7"   "BMPR1A"   "CTNNB1"   "HIPK2"   
## [25] "KLF10"    "BMP2"     "ENG"      "APC"      "PPM1A"    "XIAP"    
## [31] "CDH1"     "ID1"      "LEFTY2"   "CDKN1C"   "TRIM33"   "RAB31"   
## [37] "TJP1"     "SLC20A1"  "CDK9"     "ID3"      "NOG"      "ARID4B"  
## [43] "IFNGR2"   "ID2"      "PPP1CA"   "SPTBN1"   "WWTR1"    "BCAR3"   
## [49] "THBS1"    "FNTA"     "HDAC1"    "UBE2D3"   "LTBP2"    "RHOA"</code></pre>
<p>If we had previously loaded these gene sets in using GSEABase, we could also
simply do the following:</p>
<div class="sourceCode" id="cb446"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hallmark_pathways_GSEABase</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GSEABase/man/GeneSet-class.html">geneIds</a></span><span class="op">(</span><span class="va">hallmarks_gmt</span><span class="op">)</span></span>
<span><span class="va">hallmark_pathways_GSEABase</span><span class="op">$</span><span class="va">HALLMARK_TGF_BETA_SIGNALING</span></span></code></pre></div>
<pre><code>##  [1] "TGFBR1"   "SMAD7"    "TGFB1"    "SMURF2"   "SMURF1"   "BMPR2"   
##  [7] "SKIL"     "SKI"      "ACVR1"    "PMEPA1"   "NCOR2"    "SERPINE1"
## [13] "JUNB"     "SMAD1"    "SMAD6"    "PPP1R15A" "TGIF1"    "FURIN"   
## [19] "SMAD3"    "FKBP1A"   "MAP3K7"   "BMPR1A"   "CTNNB1"   "HIPK2"   
## [25] "KLF10"    "BMP2"     "ENG"      "APC"      "PPM1A"    "XIAP"    
## [31] "CDH1"     "ID1"      "LEFTY2"   "CDKN1C"   "TRIM33"   "RAB31"   
## [37] "TJP1"     "SLC20A1"  "CDK9"     "ID3"      "NOG"      "ARID4B"  
## [43] "IFNGR2"   "ID2"      "PPP1CA"   "SPTBN1"   "WWTR1"    "BCAR3"   
## [49] "THBS1"    "FNTA"     "HDAC1"    "UBE2D3"   "LTBP2"    "RHOA"</code></pre>
<p>Now to run fgsea, we will use default parameters besides manually setting
thresholds to ignore gene sets in the analysis based on minimum and maximum
values for their size:</p>
<div class="sourceCode" id="cb448"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fgsea_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/fgsea/man/fgsea.html">fgsea</a></span><span class="op">(</span><span class="va">hallmark_pathways_GSEABase</span>, <span class="va">rnk_list</span>, minSize <span class="op">=</span> <span class="fl">15</span>, maxSize<span class="op">=</span><span class="fl">500</span><span class="op">)</span></span>
<span><span class="va">fgsea_results</span> <span class="op">&lt;-</span> <span class="va">fgsea_results</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html">as_tibble</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>As noted in the official documentation for the original GSEA, very small gene
sets or very large gene sets may lead to issues with artificially high scores
or poor normalization, respectively. These values can be adjusted as needed, but
we have used the range suggested by the official documentation.</p>
<p>We will convert the fgsea results into a tibble, and perform some basic
exploration of our results. We can see below the results sorted by ascending
<code>padj</code> and the pathways with the lowest adjusted p-values are largely associated
with cancer pathways including Epithelial Mesenchymal Transition, and Apical
Junctions.</p>
<div class="sourceCode" id="cb449"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fgsea_results</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html">arrange</a></span><span class="op">(</span><span class="va">padj</span><span class="op">)</span></span></code></pre></div>
<pre><code>## # A tibble: 15 x 8
##    pathway                 pval     padj log2err     ES    NES  size leadingEdge
##    &lt;chr&gt;                  &lt;dbl&gt;    &lt;dbl&gt;   &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt; &lt;int&gt; &lt;list&gt;     
##  1 HALLMARK_EPITHELIA~ 4.86e-11 7.28e-10  0.851   0.509  2.60    107 &lt;chr [56]&gt; 
##  2 HALLMARK_APICAL_JU~ 1.50e- 3 1.12e- 2  0.455   0.479  1.87     32 &lt;chr [16]&gt; 
##  3 HALLMARK_MYOGENESIS 3.61e- 3 1.81e- 2  0.432   0.428  1.76     40 &lt;chr [24]&gt; 
##  4 HALLMARK_COAGULATI~ 5.16e- 2 1.55e- 1  0.277   0.383  1.46     30 &lt;chr [13]&gt; 
##  5 HALLMARK_COMPLEMENT 5.15e- 2 1.55e- 1  0.277   0.380  1.49     33 &lt;chr [11]&gt; 
##  6 HALLMARK_KRAS_SIGN~ 1.08e- 1 2.37e- 1  0.188   0.329  1.30     35 &lt;chr [13]&gt; 
##  7 HALLMARK_UV_RESPON~ 1.11e- 1 2.37e- 1  0.186   0.325  1.31     37 &lt;chr [17]&gt; 
##  8 HALLMARK_APOPTOSIS  1.28e- 1 2.39e- 1  0.171   0.378  1.31     21 &lt;chr [10]&gt; 
##  9 HALLMARK_HYPOXIA    2.24e- 1 3.74e- 1  0.124   0.298  1.16     31 &lt;chr [12]&gt; 
## 10 HALLMARK_ADIPOGENE~ 2.73e- 1 4.09e- 1  0.111   0.329  1.13     20 &lt;chr [11]&gt; 
## 11 HALLMARK_IL2_STAT5~ 5.71e- 1 7.14e- 1  0.0747 -0.253 -0.925    24 &lt;chr [6]&gt;  
## 12 HALLMARK_INTERFERO~ 5.63e- 1 7.14e- 1  0.0749 -0.254 -0.924    23 &lt;chr [7]&gt;  
## 13 HALLMARK_ALLOGRAFT~ 6.92e- 1 7.99e- 1  0.0646 -0.236 -0.857    23 &lt;chr [4]&gt;  
## 14 HALLMARK_INFLAMMAT~ 9.10e- 1 9.72e- 1  0.0479  0.184  0.683    26 &lt;chr [7]&gt;  
## 15 HALLMARK_TNFA_SIGN~ 9.72e- 1 9.72e- 1  0.0501 -0.171 -0.591    20 &lt;chr [4]&gt;</code></pre>
<p>To begin to explore these results, we could filter by our FDR threshold and
subset significant gene sets by the direction of their NES (positive indicating
an enrichment at the “top” of our list and negative indicating an enrichment at
the “bottom”).</p>
<div class="sourceCode" id="cb451"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">top_positive_nes</span> <span class="op">&lt;-</span> <span class="va">fgsea_results</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">padj</span> <span class="op">&lt;</span> <span class="fl">.25</span> <span class="op">&amp;</span> <span class="va">NES</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/slice.html">slice_max</a></span><span class="op">(</span><span class="va">NES</span>, n<span class="op">=</span><span class="fl">5</span><span class="op">)</span></span>
<span></span>
<span><span class="va">top_positive_nes</span></span></code></pre></div>
<pre><code>## # A tibble: 5 x 8
##   pathway                    pval     padj log2err    ES   NES  size leadingEdge
##   &lt;chr&gt;                     &lt;dbl&gt;    &lt;dbl&gt;   &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;int&gt; &lt;list&gt;     
## 1 HALLMARK_EPITHELIAL_M~ 4.86e-11 7.28e-10   0.851 0.509  2.60   107 &lt;chr [56]&gt; 
## 2 HALLMARK_APICAL_JUNCT~ 1.50e- 3 1.12e- 2   0.455 0.479  1.87    32 &lt;chr [16]&gt; 
## 3 HALLMARK_MYOGENESIS    3.61e- 3 1.81e- 2   0.432 0.428  1.76    40 &lt;chr [24]&gt; 
## 4 HALLMARK_COMPLEMENT    5.15e- 2 1.55e- 1   0.277 0.380  1.49    33 &lt;chr [11]&gt; 
## 5 HALLMARK_COAGULATION   5.16e- 2 1.55e- 1   0.277 0.383  1.46    30 &lt;chr [13]&gt;</code></pre>
<p>To provide a basic visualization of these results, we can display the normalized
enrichment scores for all of our pathways in a bar chart and <code>fill</code> the bars by
whether or not they meet our <code>padj</code> threshold.</p>
<div class="box note">
<p>Rank-based GSEA is often used as a hypothesis-generating experiment as it can
quickly capture potentially interesting global patterns of regulation among many
biological pathways and processes while considering all the information
generated in a HTS experiment. It is typical to set a permissive FDR for
considering “significant” gene sets. The results of rank-based GSEA need to be
further inspected and most often confirmed through complementary or follow-up
experiments. Thus, we will choose to use a relaxed FDR threshold of &lt; .25.</p>
</div>
<div class="sourceCode" id="cb453"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fgsea_results</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>pathway <span class="op">=</span> <span class="fu">forcats</span><span class="fu">::</span><span class="fu"><a href="https://forcats.tidyverse.org/reference/fct_reorder.html">fct_reorder</a></span><span class="op">(</span><span class="va">pathway</span>, <span class="va">NES</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html">geom_bar</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">pathway</span>, y<span class="op">=</span><span class="va">NES</span>, fill <span class="op">=</span> <span class="va">padj</span> <span class="op">&lt;</span> <span class="fl">.25</span><span class="op">)</span>, stat<span class="op">=</span><span class="st">'identity'</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_fill_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'TRUE'</span> <span class="op">=</span> <span class="st">'red'</span>, <span class="st">'FALSE'</span> <span class="op">=</span> <span class="st">'blue'</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_minimal</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">'fgsea results for Hallmark MSigDB gene sets'</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ylab</a></span><span class="op">(</span><span class="st">'Normalized Enrichment Score (NES)'</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">xlab</a></span><span class="op">(</span><span class="st">''</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html">coord_flip</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="main_files/top%20level%20results%20plot-1.png" alt="GSEA results suggest an enrichment of cancer-related pathways 
amongst upregulated genes. GSEA was performed using a list of genes
generated through RNAseq and ranked by descending fold change. GSEA was run 
using default parameters with minSize = 15 and maxSize = 500. Gene sets with
a Benjamini-Hochberg adjusted p-value &lt; .25 are considered significant." width="672"><p class="caption">
(#fig:top level results plot)GSEA results suggest an enrichment of cancer-related pathways
amongst upregulated genes. GSEA was performed using a list of genes
generated through RNAseq and ranked by descending fold change. GSEA was run
using default parameters with minSize = 15 and maxSize = 500. Gene sets with
a Benjamini-Hochberg adjusted p-value &lt; .25 are considered significant.
</p>
</div>
<p>Given the artificial nature of our data and our choice of a small gene set
collection, we have very few results and are able to display all of them
intelligibly on one plot. In real experiments, there may be several hundred
significant gene sets of interest. In cases like this, you may need to restrict
the number of gene sets by taking the top ten ranked by NES (both positive
and negative) and plotting those in a figure like the one above. This is only a
suggestion and there are many ways to choose ‘interesting’ gene sets to plot
together depending upon the experimental context and future questions of
interest.</p>
<p>If we wished to investigate a single gene set from this list of results, we
could create an enrichment plot displaying the Enrichment Score as it’s
calculated through the ranked list. For convenience, we have wrapped the
pre-built function, <code>plotEnrichment</code> from fgsea, in a user-defined function that
allows us to specify a specific pathway. The <code>plotEnrichment</code> function expects
at minimum the list of genes contained within the pathway to plot and the input
ranked list of gene-level statistics. We have displayed the enrichment plot for
one of the significantly enriched gene sets below:</p>
<div class="sourceCode" id="cb454"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">make_gsea_plot</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">pathway_name</span>, <span class="va">rnks</span><span class="op">)</span> <span class="op">{</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/fgsea/man/plotEnrichment.html">plotEnrichment</a></span><span class="op">(</span><span class="va">hallmark_pathways_fgsea</span><span class="op">[[</span><span class="va">pathway_name</span><span class="op">]</span><span class="op">]</span>, <span class="va">rnks</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span><span class="op">(</span>title<span class="op">=</span><span class="va">pathway_name</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="fu">make_gsea_plot</span><span class="op">(</span><span class="st">'HALLMARK_EPITHELIAL_MESENCHYMAL_TRANSITION'</span>, <span class="va">rnk_list</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="main_files/enrichment%20plot-1.png" alt="The Hallmarks EPITHELIAL_MESENCHYMAL_TRANSITION Pathway
is significantly enriched amongst the upregulated genes. There are 56 genes in the 
leading edge subset including GAS1, POSTN, and LOX. The Normalized Enrichment
Score calculated for this gene set was 2.59 and it has an adjusted p-value of 
7.28e-10" width="672"><p class="caption">
(#fig:enrichment plot)The Hallmarks EPITHELIAL_MESENCHYMAL_TRANSITION Pathway
is significantly enriched amongst the upregulated genes. There are 56 genes in the
leading edge subset including GAS1, POSTN, and LOX. The Normalized Enrichment
Score calculated for this gene set was 2.59 and it has an adjusted p-value of
7.28e-10
</p>
</div>
<p>As mentioned in the prior section, it is often useful for further investigation
to inspect the leading edge subset of genes contained within ‘interesting’ gene
sets. fgsea stores these genes in the form of a list associated with each
pathway in the returned table. We could convert and save these values in a named
list for convenience. This can be accomplished by making use of the <code>deframe()</code>
function. Below we have saved all of the genes contained within the leading edge
subset of significant gene sets, and displayed those belonging to the
HALLMARK_EPITHELIAL_MESENCHYMAL_TRANSITION gene set as determined in this
experiment:</p>
<div class="sourceCode" id="cb455"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gather_leadingedge</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">results</span>, <span class="va">threshold</span><span class="op">)</span> <span class="op">{</span></span>
<span></span>
<span>  <span class="va">genes</span> <span class="op">&lt;-</span> <span class="va">results</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">padj</span> <span class="op">&lt;</span> <span class="va">threshold</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="va">pathway</span>, <span class="va">leadingEdge</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">deframe</span><span class="op">(</span><span class="op">)</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/function.html">return</a></span><span class="op">(</span><span class="va">genes</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="va">leading_edge_genes</span> <span class="op">&lt;-</span> <span class="fu">gather_leadingedge</span><span class="op">(</span><span class="va">fgsea_results</span>, <span class="fl">.25</span><span class="op">)</span></span>
<span><span class="va">leading_edge_genes</span><span class="op">$</span><span class="va">HALLMARK_EPITHELIAL_MESENCHYMAL_TRANSITION</span></span></code></pre></div>
<pre><code>##  [1] "GAS1"   "MGP"    "SPOCK1" "EFEMP2" "FERMT2" "FBN1"   "TAGLN"  "FSTL1" 
##  [9] "TIMP3"  "VIM"    "VCAN"   "THBS2"  "ACTA2"  "DCN"    "BGN"    "NNMT"  
## [17] "CALD1"  "NTM"    "FBLN1"  "SPARC"  "COL5A2" "MYL9"   "SLIT2"  "DPYSL3"
## [25] "VEGFC"  "COL3A1" "LGALS1" "THY1"   "PCOLCE" "ADAM12" "SFRP4"  "WIPF1" 
## [33] "COL6A3" "CTHRC1" "TIMP1"  "TPM2"   "FN1"    "COL5A1" "PMP22"  "MMP2"  
## [41] "HTRA1"  "COL1A2" "POSTN"  "PDGFRB" "EMP3"   "MXRA5"  "FLNA"   "PRRX1" 
## [49] "FAP"    "LOX"    "BASP1"  "GREM1"  "CCN2"   "CDH11"  "LUM"    "LAMC1"</code></pre>
<p>This named list contains all of the leading edge genes for gene sets meeting a
certain <code>padj</code> threshold. We could import these as a GSEABase GeneSetCollection
object for further analyses in R or write them to a file to pass on to
collaborators.</p>
<p>These examples are just some of the ways you can utilize and explore the GSEA
results returned by fgsea. There are also other methods such as the one found in
the built-in function, <code><a href="https://rdrr.io/pkg/fgsea/man/plotGseaTable.html">fgsea::plotGseaTable</a></code>, that provide other ways to
visualize these results for groups of different gene sets. You can also perform
any number of different filtering and sorting options on the table of results
depending upon your interests.</p>

</div>
</div>
<div id="biological-networks" class="section level2" number="8.9">
<h2>
<span class="header-section-number">8.9</span> Biological Networks<a class="anchor" aria-label="anchor" href="#biological-networks"><i class="fas fa-link"></i></a>
</h2>
<p>Biological systems are made up of many components that work together in spatial,
temporal, and chemical ways to maintain biological fidelity. These components
may be molecular components like metabolites, transcripts, or proteins, entire
organisms as in the case of microbial ecologies, or even abstract concepts like
genes and metabolic reactions. While studying individual components is a
necessary part of biological research, understanding how those components
interact and influence each other can help gain a greater understanding of
biological systems as a whole. In biology, mathematics, and other fields, we
can conceptualize a set of these components and their relationships as
<em>network</em>.</p>
<p>In general terms, a <em>network</em> is defined as a set of components (or <em>entities</em>)
and the relationships between them. Some networks are physical, like computer
networks where different hardware devices are connected together with wires.
However, most networks are more abstract, like social networks of friendships
between people, and ecological networks like forests or soil where different
organisms influence each other by predation, synergistic interactions, etc.</p>
<p>Biological networks in molecular biology are similarly abstract, where an
interaction between entities can represent many different types of
relationships. The primary entities of molecular networks are genes,
macromolecules (DNA, RNA transcripts, proteins, etc), metabolites and other
small molecules, cellular organelles and components, and entire cells
themselves. Relationships between these entities may include spatial colocality,
direct physical association, shared molecular functions, genetic and
evolutionary relationships, chemical interactions like biochemical and
structural modifications, modulation of the anabolic or catabolic rates (e.g.
rate of transcription, translation, etc), and many more. In reality, entities
may have all of these types of relationships with other entities, all
potentially occuring simultaneously at any moment in time.</p>
<p>It is a substantial challenge to incorporate all possible types of interactions
between all entities in mind at the same time, so we will typically express only
subsets of any holistic network. These subnetworks reflect different aspects of
the system, e.g. which genes control the transcription of which other genes, or
which proteins physically complex with one another. Fortunately, the concept of
the network is generic, so we may generally use the same tools and techniques to
express and analyze networks for any network type.</p>
<p>This section first describes some of the different types of biological networks
found in molecular biology, and then covers some of the basics of the
mathematical and computational expressions, properties, and algorithms used to
analyze networks.</p>
<div id="biological-pathways" class="section level3" number="8.9.1">
<h3>
<span class="header-section-number">8.9.1</span> Biological Pathways<a class="anchor" aria-label="anchor" href="#biological-pathways"><i class="fas fa-link"></i></a>
</h3>
<p>Biological pathways, or just pathways, are groups of genes that work together to
create some product or cause a change in a cell. Most of the coordinated
activites of a cell are performed by pathways that are induced or repressed
based on conditions within and without the cell. Hundreds of biological pathways
have been characterized and organized into publicly available databases.</p>
</div>
<div id="protein-protein-interaction-network-." class="section level3" number="8.9.2">
<h3>
<span class="header-section-number">8.9.2</span> Protein-Protein Interaction Network .<a class="anchor" aria-label="anchor" href="#protein-protein-interaction-network-."><i class="fas fa-link"></i></a>
</h3>
<p>What are protein-protein interaction (PPI) networks? What information do they
represent (direct association, functional association, etc)? Where does PPI
information come from (datasets, databases, etc)? What are some ways we can use
PPI information when interpreting other biological data (like differential
expression? not sure)?</p>
</div>
<div id="gene-regulatory-networks-." class="section level3" number="8.9.3">
<h3>
<span class="header-section-number">8.9.3</span> Gene Regulatory Networks .<a class="anchor" aria-label="anchor" href="#gene-regulatory-networks-."><i class="fas fa-link"></i></a>
</h3>
<p>What are gene regulatory networks? Why are they important? How do we identify
them (data driven correlation, wetlab experiments, others?)? How are they
useful?</p>
</div>
<div id="wgcna-." class="section level3" number="8.9.4">
<h3>
<span class="header-section-number">8.9.4</span> WGCNA .<a class="anchor" aria-label="anchor" href="#wgcna-."><i class="fas fa-link"></i></a>
</h3>
<p>Weighted correlation network analysis (WGCNA) identifies genes (or other biology features) that are highly correlated with each other, and group them into clusters / modules.<br>
Typical inputs of WGCNA is the expression matrix of genes, proteins or other biology features. Then, WGCNA will construct a undirected and weighted co-expression network. The nodes will be genes (or other features), and the edges connecting them will be the pairwise correlation value of their expression level. Modules are simply the clusters of highly connected genes. After the modules are detected, down stream analyses may include summarizing the module by its “eigengenes”, inferring the functionality of a gene from genes in the same module, or compare different modules.<br>
For more information, you may refer to this <a href="https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-9-559">article</a> or the <a href="https://cran.r-project.org/web/packages/WGCNA/WGCNA.pdf">manual of WGCNA R package</a>.</p>
<div id="how-to-run-wgcna-in-r" class="section level4" number="8.9.4.1">
<h4>
<span class="header-section-number">8.9.4.1</span> How to run WGCNA in R<a class="anchor" aria-label="anchor" href="#how-to-run-wgcna-in-r"><i class="fas fa-link"></i></a>
</h4>
<p>First, install package <code>WGCNA</code>.</p>
<pre><code>BiocManager::install("WGCNA")
library(WGCNA)</code></pre>
<p>note: if you receive error message saying some dependencies are not available, it’s probably because you used to try to install it using standard method <code>install.package("WGCNA")</code>. If that’s the case, try the following to re-install WGCNA:</p>
<pre><code>BiocManager::install("WGCNA",force = T)</code></pre>
<p>Let’s load a toy gene expression matrix. We have 15 samples and 7 genes.</p>
<pre><code>##            CRAB     Whale     Lobst      Octop    Coral BabyShark     Orca
## samp1  6.867731  8.796212  7.820130  6.6081601 15.37307 14.047536 32.96333
## samp2 10.918217  6.775947 16.545788  7.5685123 28.59193  8.923224 19.13865
## samp3  5.821857  8.682239  5.089099  2.8030264 17.99007 14.643098 36.93207
## samp4 17.976404  5.741274 27.596721  8.8660342 34.62918  9.473205 18.68929
## samp5 11.647539 10.866047 14.852970  2.2034360 14.15414 15.949892 42.21472
## samp6  5.897658 11.960800 15.915636 -0.1724085 21.78596 16.336686 44.93125</code></pre>
<p>An optional choice in WGCNA to allow multithreads</p>
<div class="sourceCode" id="cb460"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">allowWGCNAThreads</span><span class="op">(</span><span class="fl">4</span><span class="op">)</span></span></code></pre></div>
<pre><code>## Allowing multi-threading with up to 4 threads.</code></pre>
<p>Now we will choose a “soft-thresholding power” to construct the network. There is no single answer in the power to choose.<br>
If you refer to <a href="https://horvath.genetics.ucla.edu/html/CoexpressionNetwork/Rpackages/WGCNA/faq.html">WGCNA faq</a>, there are more details for power choosing.</p>
<div class="sourceCode" id="cb462"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># A set of soft-thresholding powers to choose from:</span></span>
<span><span class="va">powers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">10</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Call the network topology analysis function</span></span>
<span><span class="va">sft</span> <span class="op">&lt;-</span> <span class="fu">pickSoftThreshold</span><span class="op">(</span></span>
<span>  data <span class="op">=</span> <span class="va">dat</span>,</span>
<span>  powerVector <span class="op">=</span> <span class="va">powers</span>,</span>
<span>  verbose <span class="op">=</span> <span class="fl">5</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code>## pickSoftThreshold: will use block size 7.
##  pickSoftThreshold: calculating connectivity for given powers...
##    ..working on genes 1 through 7 of 7
##    Power SFT.R.sq   slope truncated.R.sq mean.k. median.k. max.k.
## 1      1 0.026500 -12.800        0.38900   1.920     1.820  2.350
## 2      2 0.007510   2.420       -0.18500   1.240     1.180  1.730
## 3      3 0.001090   0.582        0.00340   0.931     0.908  1.430
## 4      4 0.001300  -0.460       -0.14200   0.733     0.708  1.220
## 5      5 0.003130   0.526       -0.03820   0.592     0.556  1.060
## 6      6 0.018100  -1.150        0.04810   0.486     0.440  0.922
## 7      7 0.087400  -2.130       -0.15900   0.404     0.349  0.807
## 8      8 0.063500  -1.570       -0.11100   0.338     0.279  0.708
## 9      9 0.000378  -0.206        0.00319   0.285     0.224  0.621
## 10    10 0.003650  -0.600       -0.00478   0.241     0.181  0.545</code></pre>
<div class="sourceCode" id="cb464"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span><span class="op">(</span>mfrow <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">cex1</span> <span class="op">&lt;-</span> <span class="fl">0.9</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">sft</span><span class="op">$</span><span class="va">fitIndices</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span>,</span>
<span>  <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/sign.html">sign</a></span><span class="op">(</span><span class="va">sft</span><span class="op">$</span><span class="va">fitIndices</span><span class="op">[</span>, <span class="fl">3</span><span class="op">]</span><span class="op">)</span> <span class="op">*</span> <span class="va">sft</span><span class="op">$</span><span class="va">fitIndices</span><span class="op">[</span>, <span class="fl">2</span><span class="op">]</span>,</span>
<span>  xlab <span class="op">=</span> <span class="st">"Soft Threshold (power)"</span>,</span>
<span>  ylab <span class="op">=</span> <span class="st">"Scale Free Topology Model Fit, signed R^2"</span>,</span>
<span>  main <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/paste.html">paste</a></span><span class="op">(</span><span class="st">"Scale independence"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/text.html">text</a></span><span class="op">(</span><span class="va">sft</span><span class="op">$</span><span class="va">fitIndices</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span>,</span>
<span>  <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/sign.html">sign</a></span><span class="op">(</span><span class="va">sft</span><span class="op">$</span><span class="va">fitIndices</span><span class="op">[</span>, <span class="fl">3</span><span class="op">]</span><span class="op">)</span> <span class="op">*</span> <span class="va">sft</span><span class="op">$</span><span class="va">fitIndices</span><span class="op">[</span>, <span class="fl">2</span><span class="op">]</span>,</span>
<span>  labels <span class="op">=</span> <span class="va">powers</span>, cex <span class="op">=</span> <span class="va">cex1</span>, col <span class="op">=</span> <span class="st">"red"</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/abline.html">abline</a></span><span class="op">(</span>h <span class="op">=</span> <span class="fl">0.90</span>, col <span class="op">=</span> <span class="st">"red"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">sft</span><span class="op">$</span><span class="va">fitIndices</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span>,</span>
<span>  <span class="va">sft</span><span class="op">$</span><span class="va">fitIndices</span><span class="op">[</span>, <span class="fl">5</span><span class="op">]</span>,</span>
<span>  xlab <span class="op">=</span> <span class="st">"Soft Threshold (power)"</span>,</span>
<span>  ylab <span class="op">=</span> <span class="st">"Mean Connectivity"</span>,</span>
<span>  type <span class="op">=</span> <span class="st">"n"</span>,</span>
<span>  main <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/paste.html">paste</a></span><span class="op">(</span><span class="st">"Mean connectivity"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/text.html">text</a></span><span class="op">(</span><span class="va">sft</span><span class="op">$</span><span class="va">fitIndices</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span>,</span>
<span>  <span class="va">sft</span><span class="op">$</span><span class="va">fitIndices</span><span class="op">[</span>, <span class="fl">5</span><span class="op">]</span>,</span>
<span>  labels <span class="op">=</span> <span class="va">powers</span>,</span>
<span>  cex <span class="op">=</span> <span class="va">cex1</span>, col <span class="op">=</span> <span class="st">"red"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="main_files/unnamed-chunk-154-1.png" width="672"></div>
<p>Generally, we want to choose a power that give us the “Scale Free Topology Model Fit, signed R^2” around <strong>0.8</strong>. This toy dataset is too small and can’t show it properly.<br>
For illustration, we will pick power 4 in this example.</p>
<div class="sourceCode" id="cb465"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">picked_power</span> <span class="op">&lt;-</span> <span class="fl">4</span></span>
<span></span>
<span><span class="co"># fix a namespace conflict issue, force R to use the cor() function in WGCNA:</span></span>
<span><span class="va">temp_cor</span> <span class="op">&lt;-</span> <span class="va">cor</span></span>
<span><span class="va">cor</span> <span class="op">&lt;-</span> <span class="fu">WGCNA</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/WGCNA/man/cor.html">cor</a></span></span></code></pre></div>
<p>Construct network.<br>
There are a lot of parameters. In most cases, you don’t need to modify them. But, make sure you choose the proper <code>power</code>, <code>networkType</code> (“signed” or “unsigned”), and <code>minModuleSize</code> in your analysis.<br>
- <code>networkType</code> Using “signed” option, the direction of the correlation will be considered, e.g. two genes won’t be clustered together if their correlation value is a perfect -1. On the other hand, use “unsigned” option to consider the absolute value of correlation.<br>
- <code>minModuleSize</code> The minimum umber of features, e.g. genes, in a module. In this example with only 7 genes, so I set this argument to be 2. In real analysis, a typical choice may be 20 or 30 for a matrix with several hundreds or thousands of genes.</p>
<div class="sourceCode" id="cb466"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">netwk</span> <span class="op">&lt;-</span> <span class="fu">blockwiseModules</span><span class="op">(</span></span>
<span>  datExpr <span class="op">=</span> <span class="va">dat</span>,</span>
<span>  power <span class="op">=</span> <span class="va">picked_power</span>,</span>
<span>  networkType <span class="op">=</span> <span class="st">"signed"</span>,</span>
<span>  deepSplit <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  pamRespectsDendro <span class="op">=</span> <span class="cn">F</span>,</span>
<span>  minModuleSize <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  maxBlockSize <span class="op">=</span> <span class="fl">4000</span>,</span>
<span>  reassignThreshold <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  mergeCutHeight <span class="op">=</span> <span class="fl">0.25</span>,</span>
<span>  saveTOMs <span class="op">=</span> <span class="cn">T</span>, <span class="co"># Archive the run results in TOM file (saves time)</span></span>
<span>  saveTOMFileBase <span class="op">=</span> <span class="st">"ER"</span>,</span>
<span>  numericLabels <span class="op">=</span> <span class="cn">T</span>,</span>
<span>  verbose <span class="op">=</span> <span class="fl">3</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code>##  Calculating module eigengenes block-wise from all genes
##    Flagging genes and samples with too many missing values...
##     ..step 1
##  ..Working on block 1 .
##     TOM calculation: adjacency..
##     ..will not use multithreading.
##      Fraction of slow calculations: 0.000000
##     ..connectivity..
##     ..matrix multiplication (system BLAS)..
##     ..normalization..
##     ..done.
##    ..saving TOM for block 1 into file ER-block.1.RData
##  ....clustering..
##  ....detecting modules..
##  ....calculating module eigengenes..
##  ....checking kME in modules..
##  ..merging modules that are too close..
##      mergeCloseModules: Merging modules whose distance is less than 0.25
##        Calculating new MEs...</code></pre>
<p>Now we can see the module membership assignment:</p>
<div class="sourceCode" id="cb468"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">netwk</span><span class="op">$</span><span class="va">colors</span></span></code></pre></div>
<pre><code>##      CRAB     Whale     Lobst     Octop     Coral BabyShark      Orca 
##         1         2         1         1         1         2         2</code></pre>
<p>CRAB, Lobst, Octop and Coral are assigned to module 1. Whale, BabyShark and Orca are assigned to module 2.</p>
<p>We can use function labels2colors() to conveniently convert the module membership to color labels.</p>
<div class="sourceCode" id="cb470"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mergedColors</span> <span class="op">&lt;-</span> <span class="fu">labels2colors</span><span class="op">(</span><span class="va">netwk</span><span class="op">$</span><span class="va">colors</span><span class="op">)</span></span></code></pre></div>
<p>Plot the dendrogram and the module colors:</p>
<div class="sourceCode" id="cb471"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plotDendroAndColors</span><span class="op">(</span></span>
<span>  <span class="va">netwk</span><span class="op">$</span><span class="va">dendrograms</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>,</span>
<span>  <span class="va">mergedColors</span><span class="op">[</span><span class="va">netwk</span><span class="op">$</span><span class="va">blockGenes</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">]</span>,</span>
<span>  <span class="st">"Module colors"</span>,</span>
<span>  dendroLabels <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  hang <span class="op">=</span> <span class="fl">0.03</span>,</span>
<span>  addGuide <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  guideHang <span class="op">=</span> <span class="fl">0.05</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="main_files/unnamed-chunk-159-1.png" width="672"></div>
<p>In a real analysis with a lot of genes, you will see much more modules and colors.<br>
In this dendrogram, you will be able to roughly see the size of each module by looking at the area of different colors.</p>
<p>At this point, the WGCNA analysis is finished.<br>
We can summarize the module information in a data frame:</p>
<div class="sourceCode" id="cb472"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">module_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span></span>
<span>  gene_id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">netwk</span><span class="op">$</span><span class="va">colors</span><span class="op">)</span>,</span>
<span>  modules <span class="op">=</span> <span class="fu">labels2colors</span><span class="op">(</span><span class="va">netwk</span><span class="op">$</span><span class="va">colors</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">module_df</span></span></code></pre></div>
<pre><code>##     gene_id   modules
## 1      CRAB turquoise
## 2     Whale      blue
## 3     Lobst turquoise
## 4     Octop turquoise
## 5     Coral turquoise
## 6 BabyShark      blue
## 7      Orca      blue</code></pre>

</div>
</div>
</div>
</div>
  <div class="chapter-nav">
<div class="prev"><a href="data-visualization.html"><span class="header-section-number">7</span> Data Visualization</a></div>
<div class="next"><a href="engineering.html"><span class="header-section-number">9</span> EngineeRing</a></div>
</div></main><div class="col-md-3 col-lg-2 d-none d-md-block sidebar sidebar-chapter">
    <nav id="toc" data-toggle="toc" aria-label="On this page"><h2>On this page</h2>
      <ul class="nav navbar-nav">
<li><a class="nav-link" href="#biology-bioinformatics"><span class="header-section-number">8</span> Biology &amp; Bioinformatics</a></li>
<li><a class="nav-link" href="#bio-r"><span class="header-section-number">8.1</span> R in Biology</a></li>
<li>
<a class="nav-link" href="#biological-data-overview"><span class="header-section-number">8.2</span> Biological Data Overview</a><ul class="nav navbar-nav">
<li><a class="nav-link" href="#bio-data-types"><span class="header-section-number">8.2.1</span> Types of Biological Data</a></li>
<li><a class="nav-link" href="#csv-files"><span class="header-section-number">8.2.2</span> CSV Files</a></li>
<li><a class="nav-link" href="#biological-data-is-not-tidy"><span class="header-section-number">8.2.3</span> Biological data is NOT Tidy!</a></li>
</ul>
</li>
<li><a class="nav-link" href="#bio-bioconductor"><span class="header-section-number">8.3</span> Bioconductor</a></li>
<li><a class="nav-link" href="#microarrays"><span class="header-section-number">8.4</span> Microarrays</a></li>
<li>
<a class="nav-link" href="#high-throughput-sequencing"><span class="header-section-number">8.5</span> High Throughput Sequencing</a><ul class="nav navbar-nav">
<li><a class="nav-link" href="#raw-hts-data"><span class="header-section-number">8.5.1</span> Raw HTS Data</a></li>
<li><a class="nav-link" href="#preliminary-hts-data-analysis"><span class="header-section-number">8.5.2</span> Preliminary HTS Data Analysis</a></li>
<li><a class="nav-link" href="#count-data"><span class="header-section-number">8.5.3</span> Count Data</a></li>
</ul>
</li>
<li>
<a class="nav-link" href="#gene-identifiers"><span class="header-section-number">8.6</span> Gene Identifiers</a><ul class="nav navbar-nav">
<li><a class="nav-link" href="#gene-identifier-systems"><span class="header-section-number">8.6.1</span> Gene Identifier Systems</a></li>
<li><a class="nav-link" href="#mapping-between-identifier-systems"><span class="header-section-number">8.6.2</span> Mapping Between Identifier Systems</a></li>
<li><a class="nav-link" href="#mapping-homologs"><span class="header-section-number">8.6.3</span> Mapping Homologs</a></li>
</ul>
</li>
<li>
<a class="nav-link" href="#gene-expression"><span class="header-section-number">8.7</span> Gene Expression</a><ul class="nav navbar-nav">
<li><a class="nav-link" href="#gene-expression-data-in-bioconductor"><span class="header-section-number">8.7.1</span> Gene Expression Data in Bioconductor</a></li>
<li><a class="nav-link" href="#differential-expression-analysis"><span class="header-section-number">8.7.2</span> Differential Expression Analysis</a></li>
<li><a class="nav-link" href="#microarray-gene-expression-data"><span class="header-section-number">8.7.3</span> Microarray Gene Expression Data</a></li>
<li><a class="nav-link" href="#differential-expression-microarrays-limma"><span class="header-section-number">8.7.4</span> Differential Expression: Microarrays (limma)</a></li>
<li><a class="nav-link" href="#rnaseq"><span class="header-section-number">8.7.5</span> RNASeq</a></li>
<li><a class="nav-link" href="#rnaseq-gene-expression-data"><span class="header-section-number">8.7.6</span> RNASeq Gene Expression Data</a></li>
<li><a class="nav-link" href="#filtering-counts"><span class="header-section-number">8.7.7</span> Filtering Counts</a></li>
<li><a class="nav-link" href="#count-distributions"><span class="header-section-number">8.7.8</span> Count Distributions</a></li>
<li><a class="nav-link" href="#differential-expression-rnaseq"><span class="header-section-number">8.7.9</span> Differential Expression: RNASeq</a></li>
</ul>
</li>
<li>
<a class="nav-link" href="#gene-set-enrichment-analysis"><span class="header-section-number">8.8</span> Gene Set Enrichment Analysis</a><ul class="nav navbar-nav">
<li><a class="nav-link" href="#gene-sets"><span class="header-section-number">8.8.1</span> Gene Sets</a></li>
<li><a class="nav-link" href="#working-with-gene-sets-in-r"><span class="header-section-number">8.8.2</span> Working with gene sets in R</a></li>
<li><a class="nav-link" href="#over-representation-analysis"><span class="header-section-number">8.8.3</span> Over-representation Analysis</a></li>
<li><a class="nav-link" href="#rank-based-analysis"><span class="header-section-number">8.8.4</span> Rank-based Analysis</a></li>
<li><a class="nav-link" href="#fgsea"><span class="header-section-number">8.8.5</span> fgsea</a></li>
</ul>
</li>
<li>
<a class="nav-link" href="#biological-networks"><span class="header-section-number">8.9</span> Biological Networks</a><ul class="nav navbar-nav">
<li><a class="nav-link" href="#biological-pathways"><span class="header-section-number">8.9.1</span> Biological Pathways</a></li>
<li><a class="nav-link" href="#protein-protein-interaction-network-."><span class="header-section-number">8.9.2</span> Protein-Protein Interaction Network .</a></li>
<li><a class="nav-link" href="#gene-regulatory-networks-."><span class="header-section-number">8.9.3</span> Gene Regulatory Networks .</a></li>
<li><a class="nav-link" href="#wgcna-."><span class="header-section-number">8.9.4</span> WGCNA .</a></li>
</ul>
</li>
</ul>

      <div class="book-extra">
        <ul class="list-unstyled">
          
        </ul>
</div>
    </nav>
</div>

</div>
</div> <!-- .container -->

<footer class="bg-primary text-light mt-5"><div class="container"><div class="row">

  <div class="col-12 col-md-6 mt-3">
    <p>"<strong>BF591 - R for Biological Sciences</strong>" was written by . </p>
  </div>

  <div class="col-12 col-md-6 mt-3">
    <p>This book was built by the <a class="text-light" href="https://bookdown.org">bookdown</a> R package.</p>
  </div>

</div></div>
</footer><!-- dynamically load mathjax for compatibility with self-contained --><script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/latest.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script><script type="text/x-mathjax-config">const popovers = document.querySelectorAll('a.footnote-ref[data-toggle="popover"]');
for (let popover of popovers) {
  const div = document.createElement('div');
  div.setAttribute('style', 'position: absolute; top: 0, left:0; width:0, height:0, overflow: hidden; visibility: hidden;');
  div.innerHTML = popover.getAttribute('data-content');

  var has_math = div.querySelector("span.math");
  if (has_math) {
    document.body.appendChild(div);
    MathJax.Hub.Queue(["Typeset", MathJax.Hub, div]);
    MathJax.Hub.Queue(function() {
      popover.setAttribute('data-content', div.innerHTML);
      document.body.removeChild(div);
    })
  }
}
</script>
</body>
</html>
