<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta name="viewport" content="width=device-width, initial-scale=1" />
<meta property="og:title" content="8.2 Biological Data Overview | BF591 - R for Biological Sciences" />
<meta property="og:type" content="book" />







<script type="text/x-mathjax-config">
MathJax.Hub.Config({
  TeX: { equationNumbers: { autoNumber: "AMS" } }
});
</script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<meta name="description" content="8.2 Biological Data Overview | BF591 - R for Biological Sciences">

<title>8.2 Biological Data Overview | BF591 - R for Biological Sciences</title>

<script src="libs/header-attrs-2.13/header-attrs.js"></script>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="libs/bootstrap-3.3.5/css/darkly.min.css" rel="stylesheet" />
<script src="libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="libs/navigation-1.1/tabsets.js"></script>


<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>


<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />

<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
/* show arrow before summary tag as in bootstrap
TODO: remove if boostrap in updated in html_document (rmarkdown#1485) */
details > summary {
  display: list-item;
  cursor: pointer;
}
</style>
</head>

<body>

<div class="container-fluid main-container">


<div class="row">
<div class="col-sm-12">
<div id="TOC">
<ul>
<li><a href="index.html#syllabus">Syllabus</a>
<ul>
<li><a href="course-schedule.html#course-schedule">Course Schedule</a></li>
<li><a href="instructors.html#instructors">Instructors</a></li>
<li><a href="office-hours.html#office-hours">Office Hours</a></li>
<li><a href="course-values-and-policies.html#course-values-and-policies">Course Values and Policies</a></li>
</ul></li>
<li><a href="introduction.html#introduction"><span class="toc-section-number">1</span> Introduction</a>
<ul>
<li><a href="who-this-book-is-for.html#who-this-book-is-for"><span class="toc-section-number">1.1</span> Who This Book Is For</a></li>
<li><a href="a-note-about-reinventing-the-wheel.html#a-note-about-reinventing-the-wheel"><span class="toc-section-number">1.2</span> A Note About Reinventing the Wheel</a></li>
<li><a href="sources-and-references.html#sources-and-references"><span class="toc-section-number">1.3</span> Sources and References</a></li>
</ul></li>
<li><a href="data-bio.html#data-bio"><span class="toc-section-number">2</span> Data in Biology</a>
<ul>
<li><a href="a-brief-history-of-data-in-molecular-biology.html#a-brief-history-of-data-in-molecular-biology"><span class="toc-section-number">2.1</span> A Brief History of Data in Molecular Biology</a></li>
<li><a href="biology-as-a-mature-data-science.html#biology-as-a-mature-data-science"><span class="toc-section-number">2.2</span> Biology as a Mature Data Science</a></li>
</ul></li>
<li><a href="preliminaries.html#preliminaries"><span class="toc-section-number">3</span> Preliminaries</a>
<ul>
<li><a href="prelim-r.html#prelim-r"><span class="toc-section-number">3.1</span> The R Language</a></li>
<li><a href="prelim-rstudio.html#prelim-rstudio"><span class="toc-section-number">3.2</span> RStudio</a></li>
<li><a href="prog-the-r-script.html#prog-the-r-script"><span class="toc-section-number">3.3</span> The R Script</a></li>
<li><a href="prog-workflow.html#prog-workflow"><span class="toc-section-number">3.4</span> The Scripting Workflow</a></li>
<li><a href="prelim-git.html#prelim-git"><span class="toc-section-number">3.5</span> git + github</a>
<ul>
<li><a href="prelim-git.html#motivation"><span class="toc-section-number">3.5.1</span> Motivation</a></li>
<li><a href="prelim-git.html#git"><span class="toc-section-number">3.5.2</span> git</a></li>
<li><a href="prelim-git.html#git-hosting-platforms-github"><span class="toc-section-number">3.5.3</span> Git hosting platforms (GitHub)</a></li>
</ul></li>
</ul></li>
<li><a href="prog-basics.html#prog-basics"><span class="toc-section-number">4</span> R Programming</a>
<ul>
<li><a href="before-you-begin.html#before-you-begin"><span class="toc-section-number">4.1</span> Before you begin</a></li>
<li><a href="introduction-1.html#introduction-1"><span class="toc-section-number">4.2</span> Introduction</a></li>
<li><a href="prog-r-syntax.html#prog-r-syntax"><span class="toc-section-number">4.3</span> R Syntax Basics</a></li>
<li><a href="prog-types.html#prog-types"><span class="toc-section-number">4.4</span> Basic Types of Values</a>
<ul>
<li><a href="prog-types.html#factors"><span class="toc-section-number">4.4.1</span> Factors</a></li>
</ul></li>
<li><a href="prog-struct.html#prog-struct"><span class="toc-section-number">4.5</span> Data Structures</a>
<ul>
<li><a href="prog-struct.html#vectors"><span class="toc-section-number">4.5.1</span> Vectors</a></li>
<li><a href="prog-struct.html#matrices"><span class="toc-section-number">4.5.2</span> Matrices</a></li>
<li><a href="prog-struct.html#lists-and-data-frames"><span class="toc-section-number">4.5.3</span> Lists and data frames</a></li>
</ul></li>
<li><a href="logical-tests-and-comparators.html#logical-tests-and-comparators"><span class="toc-section-number">4.6</span> Logical Tests and Comparators</a></li>
<li><a href="functions.html#functions"><span class="toc-section-number">4.7</span> Functions</a>
<ul>
<li><a href="functions.html#prog-dry"><span class="toc-section-number">4.7.1</span> DRY: Don’t Repeat Yourself</a></li>
<li><a href="functions.html#writing-your-own-functions"><span class="toc-section-number">4.7.2</span> Writing your own functions</a></li>
<li><a href="functions.html#scope"><span class="toc-section-number">4.7.3</span> Scope</a></li>
</ul></li>
<li><a href="iteration.html#iteration"><span class="toc-section-number">4.8</span> Iteration</a>
<ul>
<li><a href="iteration.html#vectorized-operations"><span class="toc-section-number">4.8.1</span> Vectorized operations</a></li>
<li><a href="iteration.html#functional-programming"><span class="toc-section-number">4.8.2</span> Functional programming</a></li>
<li><a href="iteration.html#apply-and-friends"><span class="toc-section-number">4.8.3</span> <code>apply()</code> and friends</a></li>
</ul></li>
<li><a href="installing-packages.html#installing-packages"><span class="toc-section-number">4.9</span> Installing Packages</a></li>
<li><a href="saving-and-loading-r-data.html#saving-and-loading-r-data"><span class="toc-section-number">4.10</span> Saving and Loading R Data</a></li>
<li><a href="prog-debugging.html#prog-debugging"><span class="toc-section-number">4.11</span> Troubleshooting and Debugging</a></li>
<li><a href="prog-style.html#prog-style"><span class="toc-section-number">4.12</span> Coding Style and Conventions</a>
<ul>
<li><a href="prog-style.html#is-my-code-correct"><span class="toc-section-number">4.12.1</span> Is my code correct?</a></li>
<li><a href="prog-style.html#does-my-code-follow-the-dry-principle"><span class="toc-section-number">4.12.2</span> Does my code follow the <span>DRY principle</span>?</a></li>
<li><a href="prog-style.html#did-i-choose-concise-but-descriptive-variable-and-function-names"><span class="toc-section-number">4.12.3</span> Did I choose concise but descriptive variable and function names?</a></li>
<li><a href="prog-style.html#did-i-use-indentation-and-naming-conventions-consistently-throughout-my-code"><span class="toc-section-number">4.12.4</span> Did I use indentation and naming conventions consistently throughout my code?</a></li>
<li><a href="prog-style.html#did-i-write-comments-especially-when-what-the-code-does-is-not-obvious"><span class="toc-section-number">4.12.5</span> Did I write comments, especially when what the code does is not obvious?</a></li>
<li><a href="prog-style.html#how-easy-would-it-be-for-someone-else-to-understand-my-code"><span class="toc-section-number">4.12.6</span> How easy would it be for someone else to understand my code?</a></li>
<li><a href="prog-style.html#is-my-code-easy-to-maintainchange"><span class="toc-section-number">4.12.7</span> Is my code easy to maintain/change?</a></li>
<li><a href="prog-style.html#the-styler-package"><span class="toc-section-number">4.12.8</span> The <code>styler</code> package</a></li>
</ul></li>
</ul></li>
<li><a href="data-wrangling.html#data-wrangling"><span class="toc-section-number">5</span> Data Wrangling</a>
<ul>
<li><a href="the-tidyverse.html#the-tidyverse"><span class="toc-section-number">5.1</span> The Tidyverse</a></li>
<li><a href="dw-basics.html#dw-basics"><span class="toc-section-number">5.2</span> Tidyverse Basics</a></li>
<li><a href="importing-data.html#importing-data"><span class="toc-section-number">5.3</span> Importing Data</a></li>
<li><a href="data-tibble.html#data-tibble"><span class="toc-section-number">5.4</span> The tibble</a></li>
<li><a href="tidy-data.html#tidy-data"><span class="toc-section-number">5.5</span> Tidy Data</a></li>
<li><a href="pipes.html#pipes"><span class="toc-section-number">5.6</span> pipes</a></li>
<li><a href="arranging-data.html#arranging-data"><span class="toc-section-number">5.7</span> Arranging Data</a>
<ul>
<li><a href="arranging-data.html#dplyrmutate---create-new-columns-using-other-columns"><span class="toc-section-number">5.7.1</span> <code>dplyr::mutate()</code> - Create new columns using other columns</a></li>
<li><a href="arranging-data.html#stringr---working-with-character-values"><span class="toc-section-number">5.7.2</span> <code>stringr</code> - Working with character values</a></li>
<li><a href="arranging-data.html#dplyrselect---subset-columns-by-name"><span class="toc-section-number">5.7.3</span> <code>dplyr::select()</code> - Subset Columns by Name</a></li>
<li><a href="arranging-data.html#dplyrfilter---pick-rows-out-of-a-data-set"><span class="toc-section-number">5.7.4</span> <code>dplyr::filter()</code> - Pick rows out of a data set</a></li>
<li><a href="arranging-data.html#dplyrarrange---order-rows-based-on-their-values"><span class="toc-section-number">5.7.5</span> <code>dplyr::arrange()</code> - Order rows based on their values</a></li>
<li><a href="arranging-data.html#putting-it-all-together"><span class="toc-section-number">5.7.6</span> Putting it all together</a></li>
</ul></li>
<li><a href="grouping-data.html#grouping-data"><span class="toc-section-number">5.8</span> Grouping Data</a></li>
<li><a href="rearranging-data.html#rearranging-data"><span class="toc-section-number">5.9</span> Rearranging Data</a></li>
<li><a href="relational-data.html#relational-data"><span class="toc-section-number">5.10</span> Relational Data</a>
<ul>
<li><a href="relational-data.html#dealing-with-multiple-matches"><span class="toc-section-number">5.10.1</span> Dealing with multiple matches</a></li>
</ul></li>
</ul></li>
<li><a href="data-science.html#data-science"><span class="toc-section-number">6</span> Data Science</a>
<ul>
<li><a href="data-modeling.html#data-modeling"><span class="toc-section-number">6.1</span> Data Modeling</a>
<ul>
<li><a href="data-modeling.html#a-worked-modeling-example"><span class="toc-section-number">6.1.1</span> A Worked Modeling Example</a></li>
<li><a href="data-modeling.html#data-summarization"><span class="toc-section-number">6.1.2</span> Data Summarization</a></li>
<li><a href="data-modeling.html#linear-models"><span class="toc-section-number">6.1.3</span> Linear Models</a></li>
<li><a href="data-modeling.html#flavors-of-linear-models"><span class="toc-section-number">6.1.4</span> Flavors of Linear Models</a></li>
<li><a href="data-modeling.html#assessing-model-accuracy-."><span class="toc-section-number">6.1.5</span> Assessing Model Accuracy .</a></li>
</ul></li>
<li><a href="statistical-distributions.html#statistical-distributions"><span class="toc-section-number">6.2</span> Statistical Distributions</a>
<ul>
<li><a href="statistical-distributions.html#random-variables"><span class="toc-section-number">6.2.1</span> Random Variables</a></li>
<li><a href="statistical-distributions.html#statistical-distribution-basics"><span class="toc-section-number">6.2.2</span> Statistical Distribution Basics</a></li>
<li><a href="statistical-distributions.html#distributions-in-r"><span class="toc-section-number">6.2.3</span> Distributions in R</a></li>
<li><a href="statistical-distributions.html#discrete-distributions"><span class="toc-section-number">6.2.4</span> Discrete Distributions</a></li>
<li><a href="statistical-distributions.html#continuous-distributions"><span class="toc-section-number">6.2.5</span> Continuous Distributions</a></li>
<li><a href="statistical-distributions.html#empirical-distributions"><span class="toc-section-number">6.2.6</span> Empirical Distributions</a></li>
</ul></li>
<li><a href="statistical-tests.html#statistical-tests"><span class="toc-section-number">6.3</span> Statistical Tests</a>
<ul>
<li><a href="statistical-tests.html#what-is-a-statistical-test"><span class="toc-section-number">6.3.1</span> What is a statistical test</a></li>
<li><a href="statistical-tests.html#common-statistical-tests"><span class="toc-section-number">6.3.2</span> Common Statistical Tests</a></li>
<li><a href="statistical-tests.html#choosing-a-test"><span class="toc-section-number">6.3.3</span> Choosing a Test</a></li>
<li><a href="statistical-tests.html#p-values"><span class="toc-section-number">6.3.4</span> P-values</a></li>
<li><a href="statistical-tests.html#multiple-hypothesis-testing-."><span class="toc-section-number">6.3.5</span> Multiple Hypothesis Testing .</a></li>
<li><a href="statistical-tests.html#statistical-power"><span class="toc-section-number">6.3.6</span> Statistical power</a></li>
</ul></li>
<li><a href="exploratory-data-analysis.html#exploratory-data-analysis"><span class="toc-section-number">6.4</span> Exploratory Data Analysis</a>
<ul>
<li><a href="exploratory-data-analysis.html#principal-component-analysis"><span class="toc-section-number">6.4.1</span> Principal Component Analysis</a></li>
<li><a href="exploratory-data-analysis.html#cluster-analysis"><span class="toc-section-number">6.4.2</span> Cluster Analysis</a></li>
<li><a href="exploratory-data-analysis.html#k-means-."><span class="toc-section-number">6.4.3</span> k-means .</a></li>
<li><a href="exploratory-data-analysis.html#others"><span class="toc-section-number">6.4.4</span> Others</a></li>
</ul></li>
<li><a href="network-analysis-..html#network-analysis-."><span class="toc-section-number">6.5</span> Network Analysis .</a>
<ul>
<li><a href="network-analysis-..html#basic-network-analysis"><span class="toc-section-number">6.5.1</span> Basic network analysis</a></li>
<li><a href="network-analysis-..html#create-your-own"><span class="toc-section-number">6.5.2</span> Create your own</a></li>
</ul></li>
</ul></li>
<li><a href="data-visualization.html#data-visualization"><span class="toc-section-number">7</span> Data Visualization</a>
<ul>
<li><a href="responsible-plotting.html#responsible-plotting"><span class="toc-section-number">7.1</span> Responsible Plotting</a>
<ul>
<li><a href="responsible-plotting.html#visualization-principles"><span class="toc-section-number">7.1.1</span> Visualization Principles</a></li>
<li><a href="responsible-plotting.html#human-visual-perception"><span class="toc-section-number">7.1.2</span> Human Visual Perception</a></li>
<li><a href="responsible-plotting.html#visual-encodings"><span class="toc-section-number">7.1.3</span> Visual Encodings</a></li>
<li><a href="responsible-plotting.html#some-opinionated-rules-of-thumb"><span class="toc-section-number">7.1.4</span> Some Opinionated Rules of Thumb</a></li>
</ul></li>
<li><a href="dv-gg.html#dv-gg"><span class="toc-section-number">7.2</span> Grammar of Graphics</a>
<ul>
<li><a href="dv-gg.html#ggplot-mechanics"><span class="toc-section-number">7.2.1</span> <code>ggplot</code> mechanics</a></li>
</ul></li>
<li><a href="plotting-one-dimension.html#plotting-one-dimension"><span class="toc-section-number">7.3</span> Plotting One Dimension</a>
<ul>
<li><a href="plotting-one-dimension.html#dv-bar"><span class="toc-section-number">7.3.1</span> Bar chart</a></li>
<li><a href="plotting-one-dimension.html#lollipop-plots"><span class="toc-section-number">7.3.2</span> Lollipop plots</a></li>
<li><a href="plotting-one-dimension.html#stacked-area-charts"><span class="toc-section-number">7.3.3</span> Stacked Area charts</a></li>
<li><a href="plotting-one-dimension.html#parallel-coordinate-plots"><span class="toc-section-number">7.3.4</span> Parallel Coordinate plots</a></li>
</ul></li>
<li><a href="visualizing-distributions.html#visualizing-distributions"><span class="toc-section-number">7.4</span> Visualizing Distributions</a>
<ul>
<li><a href="visualizing-distributions.html#histogram"><span class="toc-section-number">7.4.1</span> Histogram</a></li>
<li><a href="visualizing-distributions.html#density"><span class="toc-section-number">7.4.2</span> Density</a></li>
<li><a href="visualizing-distributions.html#boxplot"><span class="toc-section-number">7.4.3</span> Boxplot</a></li>
<li><a href="visualizing-distributions.html#violin-plot"><span class="toc-section-number">7.4.4</span> Violin plot</a></li>
<li><a href="visualizing-distributions.html#beeswarm-plot"><span class="toc-section-number">7.4.5</span> Beeswarm plot</a></li>
<li><a href="visualizing-distributions.html#ridgeline"><span class="toc-section-number">7.4.6</span> Ridgeline</a></li>
</ul></li>
<li><a href="plotting-two-or-more-dimensions.html#plotting-two-or-more-dimensions"><span class="toc-section-number">7.5</span> Plotting Two or More Dimensions</a>
<ul>
<li><a href="plotting-two-or-more-dimensions.html#scatter-plots"><span class="toc-section-number">7.5.1</span> Scatter Plots</a></li>
<li><a href="plotting-two-or-more-dimensions.html#bubble-plots"><span class="toc-section-number">7.5.2</span> Bubble Plots</a></li>
<li><a href="plotting-two-or-more-dimensions.html#connected-scatter-plots"><span class="toc-section-number">7.5.3</span> Connected Scatter Plots</a></li>
<li><a href="plotting-two-or-more-dimensions.html#line-plots"><span class="toc-section-number">7.5.4</span> Line Plots</a></li>
<li><a href="plotting-two-or-more-dimensions.html#parallel-coordinate-plots-1"><span class="toc-section-number">7.5.5</span> Parallel Coordinate Plots</a></li>
<li><a href="plotting-two-or-more-dimensions.html#heatmaps"><span class="toc-section-number">7.5.6</span> Heatmaps</a></li>
<li><a href="plotting-two-or-more-dimensions.html#how-to-use-heatmaps-responsibly"><span class="toc-section-number">7.5.7</span> How To Use Heatmaps Responsibly</a></li>
</ul></li>
<li><a href="other-kind-of-plots.html#other-kind-of-plots"><span class="toc-section-number">7.6</span> Other Kind of Plots</a>
<ul>
<li><a href="other-kind-of-plots.html#dendrograms"><span class="toc-section-number">7.6.1</span> Dendrograms</a></li>
<li><a href="other-kind-of-plots.html#chord-diagrams-and-circos-plots"><span class="toc-section-number">7.6.2</span> Chord Diagrams and Circos Plots</a></li>
</ul></li>
<li><a href="multiple-plots.html#multiple-plots"><span class="toc-section-number">7.7</span> Multiple Plots</a>
<ul>
<li><a href="multiple-plots.html#facet-wrapping"><span class="toc-section-number">7.7.1</span> Facet wrapping</a></li>
<li><a href="multiple-plots.html#multipanel-figures"><span class="toc-section-number">7.7.2</span> Multipanel Figures</a></li>
</ul></li>
<li><a href="publication-ready-plots.html#publication-ready-plots"><span class="toc-section-number">7.8</span> Publication Ready Plots</a>
<ul>
<li><a href="publication-ready-plots.html#ggplot-themes"><span class="toc-section-number">7.8.1</span> <code>ggplot</code> Themes</a></li>
<li><a href="publication-ready-plots.html#exporting-to-svg"><span class="toc-section-number">7.8.2</span> Exporting to SVG</a></li>
</ul></li>
<li><a href="network-visualization.html#network-visualization"><span class="toc-section-number">7.9</span> Network visualization</a>
<ul>
<li><a href="network-visualization.html#interactive-network-plots"><span class="toc-section-number">7.9.1</span> Interactive network plots</a></li>
</ul></li>
</ul></li>
<li><a href="biology-bioinformatics.html#biology-bioinformatics"><span class="toc-section-number">8</span> Biology &amp; Bioinformatics</a>
<ul>
<li><a href="bio-r.html#bio-r"><span class="toc-section-number">8.1</span> R in Biology</a></li>
<li><a href="biological-data-overview.html#biological-data-overview"><span class="toc-section-number">8.2</span> Biological Data Overview</a>
<ul>
<li><a href="biological-data-overview.html#bio-data-types"><span class="toc-section-number">8.2.1</span> Types of Biological Data</a></li>
<li><a href="biological-data-overview.html#csv-files"><span class="toc-section-number">8.2.2</span> CSV Files</a></li>
<li><a href="biological-data-overview.html#biological-data-is-not-tidy"><span class="toc-section-number">8.2.3</span> Biological data is NOT Tidy!</a></li>
</ul></li>
<li><a href="bio-bioconductor.html#bio-bioconductor"><span class="toc-section-number">8.3</span> Bioconductor</a></li>
<li><a href="microarrays.html#microarrays"><span class="toc-section-number">8.4</span> Microarrays</a></li>
<li><a href="high-throughput-sequencing.html#high-throughput-sequencing"><span class="toc-section-number">8.5</span> High Throughput Sequencing</a>
<ul>
<li><a href="high-throughput-sequencing.html#raw-hts-data"><span class="toc-section-number">8.5.1</span> Raw HTS Data</a></li>
<li><a href="high-throughput-sequencing.html#preliminary-hts-data-analysis"><span class="toc-section-number">8.5.2</span> Preliminary HTS Data Analysis</a></li>
<li><a href="high-throughput-sequencing.html#count-data"><span class="toc-section-number">8.5.3</span> Count Data</a></li>
</ul></li>
<li><a href="gene-identifiers.html#gene-identifiers"><span class="toc-section-number">8.6</span> Gene Identifiers</a>
<ul>
<li><a href="gene-identifiers.html#gene-identifier-systems"><span class="toc-section-number">8.6.1</span> Gene Identifier Systems</a></li>
<li><a href="gene-identifiers.html#mapping-between-identifier-systems"><span class="toc-section-number">8.6.2</span> Mapping Between Identifier Systems</a></li>
<li><a href="gene-identifiers.html#mapping-homologs"><span class="toc-section-number">8.6.3</span> Mapping Homologs</a></li>
</ul></li>
<li><a href="gene-expression.html#gene-expression"><span class="toc-section-number">8.7</span> Gene Expression</a>
<ul>
<li><a href="gene-expression.html#gene-expression-data-in-bioconductor"><span class="toc-section-number">8.7.1</span> Gene Expression Data in Bioconductor</a></li>
<li><a href="gene-expression.html#differential-expression-analysis"><span class="toc-section-number">8.7.2</span> Differential Expression Analysis</a></li>
<li><a href="gene-expression.html#microarray-gene-expression-data"><span class="toc-section-number">8.7.3</span> Microarray Gene Expression Data</a></li>
<li><a href="gene-expression.html#differential-expression-microarrays-limma"><span class="toc-section-number">8.7.4</span> Differential Expression: Microarrays (limma)</a></li>
<li><a href="gene-expression.html#rnaseq"><span class="toc-section-number">8.7.5</span> RNASeq</a></li>
<li><a href="gene-expression.html#rnaseq-gene-expression-data"><span class="toc-section-number">8.7.6</span> RNASeq Gene Expression Data</a></li>
<li><a href="gene-expression.html#filtering-counts"><span class="toc-section-number">8.7.7</span> Filtering Counts</a></li>
<li><a href="gene-expression.html#count-distributions"><span class="toc-section-number">8.7.8</span> Count Distributions</a></li>
<li><a href="gene-expression.html#differential-expression-rnaseq"><span class="toc-section-number">8.7.9</span> Differential Expression: RNASeq</a></li>
</ul></li>
<li><a href="gene-set-enrichment-analysis.html#gene-set-enrichment-analysis"><span class="toc-section-number">8.8</span> Gene Set Enrichment Analysis</a>
<ul>
<li><a href="gene-set-enrichment-analysis.html#gene-sets"><span class="toc-section-number">8.8.1</span> Gene Sets</a></li>
<li><a href="gene-set-enrichment-analysis.html#working-with-gene-sets-in-r"><span class="toc-section-number">8.8.2</span> Working with gene sets in R</a></li>
<li><a href="gene-set-enrichment-analysis.html#over-representation-analysis"><span class="toc-section-number">8.8.3</span> Over-representation Analysis</a></li>
<li><a href="gene-set-enrichment-analysis.html#rank-based-analysis"><span class="toc-section-number">8.8.4</span> Rank-based Analysis</a></li>
<li><a href="gene-set-enrichment-analysis.html#fgsea"><span class="toc-section-number">8.8.5</span> fgsea</a></li>
</ul></li>
<li><a href="biological-networks-..html#biological-networks-."><span class="toc-section-number">8.9</span> Biological Networks .</a>
<ul>
<li><a href="biological-networks-..html#biological-pathways-."><span class="toc-section-number">8.9.1</span> Biological Pathways .</a></li>
<li><a href="biological-networks-..html#gene-regulatory-networks-."><span class="toc-section-number">8.9.2</span> Gene Regulatory Networks .</a></li>
<li><a href="biological-networks-..html#protein-protein-interaction-network-."><span class="toc-section-number">8.9.3</span> Protein-Protein Interaction Network .</a></li>
<li><a href="biological-networks-..html#wgcna-."><span class="toc-section-number">8.9.4</span> WGCNA .</a></li>
</ul></li>
</ul></li>
<li><a href="engineering.html#engineering"><span class="toc-section-number">9</span> EngineeRing</a>
<ul>
<li><a href="unit-testing.html#unit-testing"><span class="toc-section-number">9.1</span> Unit Testing</a></li>
<li><a href="toolification.html#toolification"><span class="toc-section-number">9.2</span> Toolification</a>
<ul>
<li><a href="toolification.html#the-r-interpreter"><span class="toc-section-number">9.2.1</span> The R Interpreter</a></li>
<li><a href="toolification.html#rscript"><span class="toc-section-number">9.2.2</span> <code>Rscript</code></a></li>
<li><a href="toolification.html#commandargs"><span class="toc-section-number">9.2.3</span> <code>commandArgs()</code></a></li>
</ul></li>
<li><a href="parallel-processing.html#parallel-processing"><span class="toc-section-number">9.3</span> Parallel Processing</a>
<ul>
<li><a href="parallel-processing.html#brief-introduction-to-parallelization"><span class="toc-section-number">9.3.1</span> Brief Introduction to Parallelization</a></li>
<li><a href="parallel-processing.html#apply-and-friends-are-pleasingly-parallel"><span class="toc-section-number">9.3.2</span> <code>apply</code> and Friends Are Pleasingly Parallel</a></li>
<li><a href="parallel-processing.html#the-parallel-package"><span class="toc-section-number">9.3.3</span> The <code>parallel</code> package</a></li>
</ul></li>
<li><a href="object-orientation-in-r.html#object-orientation-in-r"><span class="toc-section-number">9.4</span> Object Orientation in R</a></li>
<li><a href="r-packages.html#r-packages"><span class="toc-section-number">9.5</span> R Packages</a></li>
</ul></li>
<li><a href="rshiny.html#rshiny"><span class="toc-section-number">10</span> RShiny</a>
<ul>
<li><a href="rshiny-introduction.html#rshiny-introduction"><span class="toc-section-number">10.1</span> RShiny Introduction</a>
<ul>
<li><a href="rshiny-introduction.html#how-does-a-web-page-work"><span class="toc-section-number">10.1.1</span> How does a web page work?</a></li>
</ul></li>
<li><a href="application-structure.html#application-structure"><span class="toc-section-number">10.2</span> Application Structure</a>
<ul>
<li><a href="application-structure.html#ui-design"><span class="toc-section-number">10.2.1</span> UI Design</a></li>
<li><a href="application-structure.html#server-functionality"><span class="toc-section-number">10.2.2</span> Server Functionality</a></li>
</ul></li>
<li><a href="reactivity.html#reactivity"><span class="toc-section-number">10.3</span> Reactivity</a></li>
<li><a href="publishing.html#publishing"><span class="toc-section-number">10.4</span> Publishing</a></li>
</ul></li>
<li><a href="communicating-with-r.html#communicating-with-r"><span class="toc-section-number">11</span> Communicating with R</a>
<ul>
<li><a href="rmarkdown-knitr.html#rmarkdown-knitr"><span class="toc-section-number">11.1</span> RMarkdown &amp; knitr</a></li>
<li><a href="bookdown.html#bookdown"><span class="toc-section-number">11.2</span> bookdown</a></li>
</ul></li>
<li><a href="contribution-guide.html#contribution-guide"><span class="toc-section-number">12</span> Contribution Guide</a>
<ul>
<li><a href="custom-blocks.html#custom-blocks"><span class="toc-section-number">12.1</span> Custom blocks</a></li>
</ul></li>
<li><a href="assignments.html#assignments">Assignments</a>
<ul>
<li><a href="assignment-format.html#assignment-format">Assignment Format</a></li>
<li><a href="starting-an-assignment.html#starting-an-assignment">Starting an Assignment</a>
<ul>
<li><a href="starting-an-assignment.html#starting-your-own-git-repository-and-cloning-it">Starting your own <code>git</code> repository and cloning it</a></li>
<li><a href="starting-an-assignment.html#contents">Contents</a></li>
<li><a href="starting-an-assignment.html#github-tutorial">GitHub tutorial</a></li>
<li><a href="starting-an-assignment.html#committing-and-pushing-with-r-and-without">Committing and Pushing with R (and without)</a></li>
<li><a href="starting-an-assignment.html#using-rstudio">Using RStudio</a></li>
<li><a href="starting-an-assignment.html#using-the-command-line">Using the command line</a></li>
</ul></li>
<li><a href="assignment-1.html#assignment-1">Assignment 1</a>
<ul>
<li><a href="assignment-1.html#problem-statement">Problem Statement</a></li>
<li><a href="assignment-1.html#learning-objectives-and-skill-list">Learning Objectives and Skill List</a></li>
<li><a href="assignment-1.html#instructions">Instructions</a></li>
<li><a href="assignment-1.html#hints">Hints</a></li>
</ul></li>
<li><a href="assignment-2.html#assignment-2">Assignment 2</a>
<ul>
<li><a href="assignment-2.html#problem-statement-1">Problem Statement</a></li>
<li><a href="assignment-2.html#learning-objectives">Learning Objectives</a></li>
<li><a href="assignment-2.html#skill-list">Skill List</a></li>
<li><a href="assignment-2.html#instructions-1">Instructions</a></li>
<li><a href="assignment-2.html#function-details">Function Details</a></li>
</ul></li>
<li><a href="assignment-3.html#assignment-3">Assignment 3</a>
<ul>
<li><a href="assignment-3.html#problem-statement-2">Problem Statement</a></li>
<li><a href="assignment-3.html#learning-objectives-1">Learning Objectives</a></li>
<li><a href="assignment-3.html#skill-list-1">Skill List</a></li>
<li><a href="assignment-3.html#background-on-microarrays">Background on Microarrays</a></li>
<li><a href="assignment-3.html#background-on-principal-component-analysis">Background on Principal Component Analysis</a></li>
<li><a href="assignment-3.html#marisa-et-al.-gene-expression-classification-of-colon-cancer-into-molecular-subtypes-characterization-validation-and-prognostic-value.-plos-medicine-may-2013.-pmid-23700391">Marisa et al. Gene Expression Classification of Colon Cancer into Molecular Subtypes: Characterization, Validation, and Prognostic Value. PLoS Medicine, May 2013. PMID: 23700391</a></li>
<li><a href="assignment-3.html#scaling-data-using-r-scale">Scaling data using R <code>scale()</code></a></li>
<li><a href="assignment-3.html#proportion-of-variance-explained">Proportion of variance explained</a></li>
<li><a href="assignment-3.html#plotting-and-visualization-of-pca">Plotting and visualization of PCA</a></li>
<li><a href="assignment-3.html#hierarchical-clustering-and-heatmaps">Hierarchical Clustering and Heatmaps</a></li>
<li><a href="assignment-3.html#references">References</a></li>
</ul></li>
<li><a href="assignment-4.html#assignment-4">Assignment 4</a>
<ul>
<li><a href="assignment-4.html#problem-statement-3">Problem Statement</a></li>
<li><a href="assignment-4.html#learning-objectives-2">Learning Objectives</a></li>
<li><a href="assignment-4.html#skill-list-2">Skill List</a></li>
<li><a href="assignment-4.html#instructions-2">Instructions</a></li>
<li><a href="assignment-4.html#deliverables">Deliverables</a></li>
<li><a href="assignment-4.html#function-details-1">Function Details</a></li>
<li><a href="assignment-4.html#hints-1">Hints</a></li>
</ul></li>
<li><a href="assignment-5.html#assignment-5">Assignment 5</a>
<ul>
<li><a href="assignment-5.html#problem-statement-4">Problem Statement</a></li>
<li><a href="assignment-5.html#learning-objectives-3">Learning Objectives</a></li>
<li><a href="assignment-5.html#skill-list-3">Skill List</a></li>
<li><a href="assignment-5.html#deseq2-background">DESeq2 Background</a></li>
<li><a href="assignment-5.html#generating-a-counts-matrix">Generating a counts matrix</a></li>
<li><a href="assignment-5.html#prefiltering-counts-matrix">Prefiltering Counts matrix</a></li>
<li><a href="assignment-5.html#median-of-ratios-normalization">Median-of-ratios normalization</a></li>
<li><a href="assignment-5.html#deseq2-preparation">DESeq2 preparation</a></li>
<li><a href="assignment-5.html#omeara-et-al.-transcriptional-reversion-of-cardiac-myocyte-fate-during-mammalian-cardiac-regeneration.-circ-res.-feb-2015.-pmid-25477501l">O’Meara et al. Transcriptional Reversion of Cardiac Myocyte Fate During Mammalian Cardiac Regeneration. Circ Res. Feb 2015. PMID: 25477501l</a></li>
<li><a href="assignment-5.html#reading-and-subsetting-the-data-from-verse_counts.tsv-and-sample_metadata.csv">1. Reading and subsetting the data from verse_counts.tsv and sample_metadata.csv</a></li>
<li><a href="assignment-5.html#running-deseq2">2. Running DESeq2</a></li>
<li><a href="assignment-5.html#annotating-results-to-construct-a-labeled-volcano-plot">3. Annotating results to construct a labeled volcano plot</a></li>
<li><a href="assignment-5.html#diagnostic-plot-of-the-raw-p-values-for-all-genes">4. Diagnostic plot of the raw p-values for all genes</a></li>
<li><a href="assignment-5.html#plotting-the-logfoldchanges-for-differentially-expressed-genes">5. Plotting the LogFoldChanges for differentially expressed genes</a></li>
<li><a href="assignment-5.html#the-choice-of-fdr-cutoff-depends-on-cost">The choice of FDR cutoff depends on cost</a></li>
<li><a href="assignment-5.html#plotting-the-normalized-counts-of-differentially-expressed-genes">6. Plotting the normalized counts of differentially expressed genes</a></li>
<li><a href="assignment-5.html#volcano-plot-to-visualize-differential-expression-results">7. Volcano Plot to visualize differential expression results</a></li>
<li><a href="assignment-5.html#running-fgsea-vignette">8. Running fgsea <span>vignette</span></a></li>
<li><a href="assignment-5.html#plotting-the-top-ten-positive-nes-and-top-ten-negative-nes-pathways">9. Plotting the top ten positive NES and top ten negative NES pathways</a></li>
<li><a href="assignment-5.html#references-1">References</a></li>
</ul></li>
<li><a href="assignment-6.html#assignment-6">Assignment 6</a>
<ul>
<li><a href="assignment-6.html#problem-statement-5">Problem Statement</a></li>
<li><a href="assignment-6.html#learning-objectives-4">Learning Objectives</a></li>
<li><a href="assignment-6.html#skill-list-4">Skill List</a></li>
<li><a href="assignment-6.html#instructions-3">Instructions</a></li>
<li><a href="assignment-6.html#function-details-2">Function Details</a></li>
</ul></li>
<li><a href="assignment-7.html#assignment-7">Assignment 7</a>
<ul>
<li><a href="assignment-7.html#problem-statement-6">Problem Statement</a></li>
<li><a href="assignment-7.html#learning-objectives-5">Learning Objectives</a></li>
<li><a href="assignment-7.html#skill-list-5">Skill List</a></li>
<li><a href="assignment-7.html#instructions-4">Instructions</a></li>
<li><a href="assignment-7.html#function-details-3">“Function” Details</a></li>
</ul></li>
</ul></li>
<li class="appendix"><span><b>Appendix</b></span></li>
<li><a href="class-outlines.html#class-outlines"><span class="toc-section-number">A</span> Class Outlines</a>
<ul>
<li><a href="week-1.html#week-1"><span class="toc-section-number">A.1</span> Week 1</a></li>
<li><a href="week-2.html#week-2"><span class="toc-section-number">A.2</span> Week 2</a></li>
<li><a href="week-3.html#week-3"><span class="toc-section-number">A.3</span> Week 3</a></li>
<li><a href="week-4.html#week-4"><span class="toc-section-number">A.4</span> Week 4</a></li>
<li><a href="week-5.html#week-5"><span class="toc-section-number">A.5</span> Week 5</a></li>
<li><a href="week-6.html#week-6"><span class="toc-section-number">A.6</span> Week 6</a></li>
<li><a href="week-7.html#week-7"><span class="toc-section-number">A.7</span> Week 7</a></li>
<li><a href="week-8.html#week-8"><span class="toc-section-number">A.8</span> Week 8</a></li>
<li><a href="week-9.html#week-9"><span class="toc-section-number">A.9</span> Week 9</a></li>
<li><a href="week-10.html#week-10"><span class="toc-section-number">A.10</span> Week 10</a></li>
<li><a href="week-11.html#week-11"><span class="toc-section-number">A.11</span> Week 11</a></li>
<li><a href="week-12.html#week-12"><span class="toc-section-number">A.12</span> Week 12</a></li>
<li><a href="week-13.html#week-13"><span class="toc-section-number">A.13</span> Week 13</a></li>
<li><a href="week-14.html#week-14"><span class="toc-section-number">A.14</span> Week 14</a></li>
</ul></li>
</ul>
</div>
</div>
</div>
<div class="row">
<div class="col-sm-12">
<div id="biological-data-overview" class="section level2" number="8.2">
<h2><span class="header-section-number">8.2</span> Biological Data Overview</h2>
<div id="bio-data-types" class="section level3" number="8.2.1">
<h3><span class="header-section-number">8.2.1</span> Types of Biological Data</h3>
<p>In very general terms, there are five types of data used in biological data
analysis: raw/primary data, processed data, analysis results, metadata, and
annotation data.</p>
<p><strong>Raw/primary data.</strong> These data are the primary observations made by whatever
instruments/techniques we are using for our experiment. These may include
high-throughput sequencing data, mass/charge ratio data from mass spectrometry,
16S rRNA sequencing data from metagenomic studies, SNPs from genotyping assays,
etc. These data can be very large and are often not efficiently processed using
R. Instead, specialized tools built outside of R are used to first process the
primary data into a form that is amenable to analysis. The most common primary
biological data types include <a href="microarrays.html#microarrays">Microarrays</a>, <a href="high-throughput-sequencing.html#high-throughput-sequencing">High Throughput Sequencing</a> data,
and <a href="https://en.wikipedia.org/wiki/Mass_spectrometry">mass spectrometry</a> data.</p>
<p><strong>Processed data.</strong> Processed data is the result of any analysis or
transformation of primary data into an intermediate, more interpretable form.
For example, when performing gene expression analysis with RNASeq, short reads
that form the primary data are typically aligned against a genome and then
counted against annotated genes, resulting in a count for each gene in the
genome. Typically, these counts are then combined for many samples into a single
matrix and subject to downstream analyses like differential expression.</p>
<p><strong>Analysis results.</strong> Analysis results aren’t data <em>per se</em>, but are the results
of analysis of primary data or processed results. For example, in gene
expression studies, a first analysis is often differential expression, where
each gene in the genome is tested for expression associated with some outcome
of interest across many samples. Each gene then has a number of statistics and
related values, like log2 fold change, nominal and adjusted p-value, etc. These
forms of data are usually what we use to form interpretations of our datasets
and therefore we must manipulate them in much the same way as any other dataset.</p>
<p><strong>Metadata.</strong> In most experimental settings, multiple samples have been
processed and had the same primary data type collected. These samples also have
information associated with them, which is here termed <em>metadata</em>, or “data
about data.” In our gene expression of post mortem brain experiments mentioned
above, the information about the individuals themselves, including age at death,
whether the person had a disease, the measurements of tissue quality, etc. is
the metadata. The primary and processed data and metadata are usually stored in
different files, where the metadata (or <em>sample information</em> or <em>sample data</em>,
etc) will have one column indicating the unique identifier (ID) of each sample.
The processed data will typically have columns named for each of the sample IDs.</p>
<p><strong>Annotation data.</strong> A tremendous amount of knowledge has been generated about
biological entities, e.g. genes, especially since the publication of the human
genome. Annotation data is publicly available information about the features we
measure in our experiments. This may come in the form of the coordinates in a
genome where genes exist, any known functions of those genes, the domains found
in proteins and their relative sequence, gene identifier cross references across
different gene naming systems (e.g. symbol vs Ensembl ID), single nucleotide
polymorphism genomic locations and associations with traits or diseases, etc.
This is information that we use to aid in interpretation of our experimental
data but generally do not generate ourselves. Annotation data comes in many
forms, some of which are in CSV format.</p>
<p>The figure below contains a simple diagram of how these different types of data
work together for a typical biological data analysis.</p>
<div class="figure">
<img src="information_flow.png" alt="" />
<p class="caption">Information flow in biological data analysis</p>
</div>
</div>
<div id="csv-files" class="section level3" number="8.2.2">
<h3><span class="header-section-number">8.2.2</span> CSV Files</h3>
<p>The most common, convenient, and flexible data file format in biology and
bioinformatics is the <em>character-delimited</em> or <em>character-separated text file</em>.
These files are plain text files (i.e. not the native file format of any
specific program, like Excel) that generally contain rectangles of data. When
formatted correctly, you can think of these files as containing a grid or matrix
of data values with some number of columns, each of which has the same number of
values. Each line of these files has some number of data values separated by a
consistent character, most commonly the comma which are called <a href="https://en.wikipedia.org/wiki/Comma-separated_values">comma-separated
value, or “CSV,” files</a>
and filenames typically end with the extension <code>.csv</code>. Note that
other characters, especially the <Tab> character, may be used to create valid
files in this format, and all the same general principles apply. This is an
example of a simple CSV file:</p>
<pre><code>id,somevalue,category,genes
1,2.123,A,APOE
4,5.123,B,&quot;HOXA1,HOXB1&quot;
7,8.123,,SNCA</code></pre>
<p>Some properties and principles of CSV files:</p>
<ul>
<li>The first line often but not always contains the column names of each column</li>
<li>Each value is delimited by the same character, in this case <code>,</code></li>
<li>Values can be any value, including numbers and characters</li>
<li>When a value contains the delimiting character (e.g. HOXA1,HOXB1 contains a
<code>,</code>), the value is wrapped in double quotes</li>
<li>Values can be missing, indicated by sequential delimiters (i.e. <code>,,</code> or one
<code>,</code> at the end of the line, if the last column value is missing)</li>
<li>There is no delimiter at the end of the lines</li>
<li>To be well-formatted <em>every line must have the same number of delimited
values</em></li>
</ul>
<p>These same properties and principles apply to all character-separated files,
regardless of the specific delimiter used. If a file follows these principles,
they can be loaded very easily into R or any other data analysis setting.
### Common Biological Data Matrices</p>
<p>Typically the first data set you will work with in R is <em>processed data</em> as
described in the previous section. This data has been transformed from primary
data in some way such that it (usually) forms a numeric matrix with features
as rows and samples as columns. The first column of these files usually contains
a feature identifier, e.g. gene identifier, genomic locus, probe set ID, etc and
the remaining columns have numerical values, one per sample. The first row is
usually column names for all the columns in the file. Below is an example of one
of these files from a microarray gene expression dataset loaded into R:</p>
<div class="sourceCode" id="cb359"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb359-1"><a href="biological-data-overview.html#cb359-1" aria-hidden="true" tabindex="-1"></a>intensities <span class="ot">&lt;-</span> <span class="fu">read_csv</span>(<span class="st">&quot;example_intensity_data.csv&quot;</span>)</span>
<span id="cb359-2"><a href="biological-data-overview.html#cb359-2" aria-hidden="true" tabindex="-1"></a>intensities</span>
<span id="cb359-3"><a href="biological-data-overview.html#cb359-3" aria-hidden="true" tabindex="-1"></a><span class="co"># A tibble: 54,675 x 36</span></span>
<span id="cb359-4"><a href="biological-data-overview.html#cb359-4" aria-hidden="true" tabindex="-1"></a>   probe     GSM972389 GSM972390 GSM972396 GSM972401 GSM972409 GSM972412 GSM972413 GSM972422 GSM972429</span>
<span id="cb359-5"><a href="biological-data-overview.html#cb359-5" aria-hidden="true" tabindex="-1"></a>   <span class="sc">&lt;</span>chr<span class="sc">&gt;</span>         <span class="er">&lt;</span>dbl<span class="sc">&gt;</span>     <span class="er">&lt;</span>dbl<span class="sc">&gt;</span>     <span class="er">&lt;</span>dbl<span class="sc">&gt;</span>     <span class="er">&lt;</span>dbl<span class="sc">&gt;</span>     <span class="er">&lt;</span>dbl<span class="sc">&gt;</span>     <span class="er">&lt;</span>dbl<span class="sc">&gt;</span>     <span class="er">&lt;</span>dbl<span class="sc">&gt;</span>     <span class="er">&lt;</span>dbl<span class="sc">&gt;</span>     <span class="er">&lt;</span>dbl<span class="sc">&gt;</span></span>
<span id="cb359-6"><a href="biological-data-overview.html#cb359-6" aria-hidden="true" tabindex="-1"></a> <span class="dv">1</span> 1007_s_at      <span class="fl">9.54</span>     <span class="fl">10.2</span>       <span class="fl">9.72</span>      <span class="fl">9.68</span>      <span class="fl">9.35</span>      <span class="fl">9.89</span>      <span class="fl">9.70</span>      <span class="fl">9.67</span>      <span class="fl">9.87</span></span>
<span id="cb359-7"><a href="biological-data-overview.html#cb359-7" aria-hidden="true" tabindex="-1"></a> <span class="dv">2</span> 1053_at        <span class="fl">7.62</span>      <span class="fl">7.92</span>      <span class="fl">7.17</span>      <span class="fl">7.24</span>      <span class="fl">8.20</span>      <span class="fl">6.87</span>      <span class="fl">6.62</span>      <span class="fl">7.23</span>      <span class="fl">7.45</span></span>
<span id="cb359-8"><a href="biological-data-overview.html#cb359-8" aria-hidden="true" tabindex="-1"></a> <span class="dv">3</span> 117_at         <span class="fl">5.50</span>      <span class="fl">5.56</span>      <span class="fl">5.06</span>      <span class="fl">7.44</span>      <span class="fl">5.19</span>      <span class="fl">5.72</span>      <span class="fl">5.87</span>      <span class="fl">6.15</span>      <span class="fl">5.46</span></span>
<span id="cb359-9"><a href="biological-data-overview.html#cb359-9" aria-hidden="true" tabindex="-1"></a> <span class="dv">4</span> 121_at         <span class="fl">7.27</span>      <span class="fl">7.96</span>      <span class="fl">7.42</span>      <span class="fl">7.34</span>      <span class="fl">7.49</span>      <span class="fl">7.76</span>      <span class="fl">7.44</span>      <span class="fl">7.66</span>      <span class="fl">8.02</span></span>
<span id="cb359-10"><a href="biological-data-overview.html#cb359-10" aria-hidden="true" tabindex="-1"></a> <span class="dv">5</span> 1255_g_at      <span class="fl">2.79</span>      <span class="fl">3.10</span>      <span class="fl">2.78</span>      <span class="fl">2.91</span>      <span class="fl">3.02</span>      <span class="fl">2.73</span>      <span class="fl">2.78</span>      <span class="fl">3.56</span>      <span class="fl">2.83</span></span>
<span id="cb359-11"><a href="biological-data-overview.html#cb359-11" aria-hidden="true" tabindex="-1"></a> <span class="dv">6</span> 1294_at        <span class="fl">7.51</span>      <span class="fl">7.28</span>      <span class="fl">7.00</span>      <span class="fl">7.18</span>      <span class="fl">7.38</span>      <span class="fl">6.98</span>      <span class="fl">6.90</span>      <span class="fl">7.54</span>      <span class="fl">7.66</span></span>
<span id="cb359-12"><a href="biological-data-overview.html#cb359-12" aria-hidden="true" tabindex="-1"></a> <span class="dv">7</span> 1316_at        <span class="fl">3.89</span>      <span class="fl">4.36</span>      <span class="fl">4.24</span>      <span class="fl">3.94</span>      <span class="fl">4.20</span>      <span class="fl">4.34</span>      <span class="fl">4.06</span>      <span class="fl">4.24</span>      <span class="fl">4.11</span></span>
<span id="cb359-13"><a href="biological-data-overview.html#cb359-13" aria-hidden="true" tabindex="-1"></a> <span class="dv">8</span> 1320_at        <span class="fl">4.65</span>      <span class="fl">4.91</span>      <span class="fl">4.70</span>      <span class="fl">4.78</span>      <span class="fl">5.06</span>      <span class="fl">4.71</span>      <span class="fl">4.55</span>      <span class="fl">4.58</span>      <span class="fl">5.10</span></span>
<span id="cb359-14"><a href="biological-data-overview.html#cb359-14" aria-hidden="true" tabindex="-1"></a> <span class="dv">9</span> 1405_i_at      <span class="fl">8.03</span>      <span class="fl">7.47</span>      <span class="fl">5.42</span>      <span class="fl">7.21</span>      <span class="fl">9.48</span>      <span class="fl">6.79</span>      <span class="fl">6.57</span>      <span class="fl">8.50</span>      <span class="fl">6.36</span></span>
<span id="cb359-15"><a href="biological-data-overview.html#cb359-15" aria-hidden="true" tabindex="-1"></a><span class="dv">10</span> 1431_at        <span class="fl">3.09</span>      <span class="fl">3.78</span>      <span class="fl">3.33</span>      <span class="fl">3.12</span>      <span class="fl">3.21</span>      <span class="fl">3.27</span>      <span class="fl">3.37</span>      <span class="fl">3.84</span>      <span class="fl">3.32</span></span>
<span id="cb359-16"><a href="biological-data-overview.html#cb359-16" aria-hidden="true" tabindex="-1"></a><span class="co"># ... with 54,665 more rows, and 26 more variables: GSM972433 &lt;dbl&gt;, GSM972438 &lt;dbl&gt;, GSM972440 &lt;dbl&gt;,</span></span>
<span id="cb359-17"><a href="biological-data-overview.html#cb359-17" aria-hidden="true" tabindex="-1"></a><span class="co">#   GSM972443 &lt;dbl&gt;, GSM972444 &lt;dbl&gt;, GSM972446 &lt;dbl&gt;, GSM972453 &lt;dbl&gt;, GSM972457 &lt;dbl&gt;,</span></span>
<span id="cb359-18"><a href="biological-data-overview.html#cb359-18" aria-hidden="true" tabindex="-1"></a><span class="co">#   GSM972459 &lt;dbl&gt;, GSM972463 &lt;dbl&gt;, GSM972467 &lt;dbl&gt;, GSM972472 &lt;dbl&gt;, GSM972473 &lt;dbl&gt;,</span></span>
<span id="cb359-19"><a href="biological-data-overview.html#cb359-19" aria-hidden="true" tabindex="-1"></a><span class="co">#   GSM972475 &lt;dbl&gt;, GSM972476 &lt;dbl&gt;, GSM972477 &lt;dbl&gt;, GSM972479 &lt;dbl&gt;, GSM972480 &lt;dbl&gt;,</span></span>
<span id="cb359-20"><a href="biological-data-overview.html#cb359-20" aria-hidden="true" tabindex="-1"></a><span class="co">#   GSM972487 &lt;dbl&gt;, GSM972488 &lt;dbl&gt;, GSM972489 &lt;dbl&gt;, GSM972496 &lt;dbl&gt;, GSM972502 &lt;dbl&gt;,</span></span>
<span id="cb359-21"><a href="biological-data-overview.html#cb359-21" aria-hidden="true" tabindex="-1"></a><span class="co">#   GSM972510 &lt;dbl&gt;, GSM972512 &lt;dbl&gt;, GSM972521 &lt;dbl&gt;</span></span></code></pre></div>
<p>The file has 54,676 rows, consisting of one header row which R loads in as the
column names, and the remaining are probe sets, one per row. There are 36
columns, where the first contains the probe set ID (e.g. <code>1007_s_at</code>) and the
remaining 35 columns correspond to samples.</p>
</div>
<div id="biological-data-is-not-tidy" class="section level3" number="8.2.3">
<h3><span class="header-section-number">8.2.3</span> Biological data is NOT Tidy!</h3>
<p>As mentioned in the <a href="#tidy%20data">tidy data</a> section, the tidyverse packages
assume data to be in so-called “tidy format,” with variables as columns and
observations as rows. Unfortunately, certain forms of biological data are
typically available in the opposite orientation - variables are in rows and
observations are in columns. This is primarily true in feature data matrices,
e.g. gene expression counts matrices, where the number of variables (e.g. genes)
is much larger than the number of samples, which tend to be small very small
compared with the number of features. This format is convenient for humans to
interact with using, e.g. spreadsheet programs like Microsoft Excel, but can
unfortunately make performing certain operations on them challenging in
tidyverse.</p>
<p>For example, consider the microarray expression dataset in the previous section.
Each of the 54,676 rows is a probeset, and each of the 35 numeric columns is a
sample. This is a very large number of probesets to consider, especially if we
plan to conduct a statistical test on each, which would impose a substantial
<a href="#multiple-hypothesis-testing">multiple hypothesis testing</a> burden on our
results. We may therefore wish to eliminate probesets that have very low
variance from the overall dataset, since these probesets are not likely to have
a detectable statistical difference in our tests. However, computing the
variance for each probeset is a computation <em>across all columns</em>, not on columns
themselves, and this is not what tidyverse is designed to do well. Said
differently, R and tidyverse do not operate by default on the rows of a data
frame, tibble, or matrix.</p>
<p>Both base R and tidyverse are optimized to perform computations on columns, not
rows. The reasons for this are buried in the details of how the R program itself
was written to organize data internally and are beyond the scope of this book.
The consequence of this design choice is that, while we can perform operations
on the rows rather than the columns of a data structure, our code may perform
very poorly (i.e. take a very long time to run).</p>
<p>When working with these datasets, we have a couple options to deal with this
issue:</p>
<ul>
<li><p><strong>Pivot into long format.</strong> As described in the <a href="rearranging-data.html#rearranging-data">Rearranging Data</a> section, we
can rearrange our tibble to be more amenable to certain computations. In our
earlier example, we wish to group all of our measurements by probeset and
compute the variance of each, then possibly filter out probesets based on low
variance. We can therefore combine <code>pivot_longer()</code>, <code>group_by()</code>,
<code>summarize()</code>, and finally <code>left_join()</code> to perform this operation. Exactly
how to do this is left as an exercise in <a href="assignment-1.html#assignment-1">Assignment 1</a>.</p></li>
<li><p><strong>Compute row-wise statistics using <code>apply()</code>.</strong> As described in <a href="iteration.html#iteration">Iteration</a>,
R is a functional programming language and implements iteration in a
functional style using the <code>apply()</code> function. The <a href="iteration.html#apply-and-friends"><code>apply()   function</code></a> accepts a <code>MARGIN</code> argument of 1 or 2 if the
provided function is to be applied to the columns or rows, respectively. This
method can be used to compute a summary statistic on each row of a tibble and
the result saved back into the tibble using the column set operator:</p>
<div class="sourceCode" id="cb360"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb360-1"><a href="biological-data-overview.html#cb360-1" aria-hidden="true" tabindex="-1"></a>intensity_variance <span class="ot">&lt;-</span> <span class="fu">apply</span>(intensities, <span class="dv">2</span>, var)</span>
<span id="cb360-2"><a href="biological-data-overview.html#cb360-2" aria-hidden="true" tabindex="-1"></a>intensities<span class="sc">$</span>variance <span class="ot">&lt;-</span> intensity_variance</span></code></pre></div></li>
</ul>

</div>
</div>
<p style="text-align: center;">
<a href="bio-r.html"><button class="btn btn-default">Previous</button></a>
<a href="bio-bioconductor.html"><button class="btn btn-default">Next</button></a>
</p>
</div>
</div>


</div>

<script>

// add bootstrap table styles to pandoc tables
$(document).ready(function () {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
});

</script>

</body>
</html>
