<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>8.9 Biological Networks . | BF591 - R for Biological Sciences</title>
  <meta name="description" content="8.9 Biological Networks . | BF591 - R for Biological Sciences" />
  <meta name="generator" content="bookdown 0.24 and GitBook 2.6.7" />

  <meta property="og:title" content="8.9 Biological Networks . | BF591 - R for Biological Sciences" />
  <meta property="og:type" content="book" />
  
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="8.9 Biological Networks . | BF591 - R for Biological Sciences" />
  
  
  




  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="gene-set-enrichment-analysis.html"/>
<link rel="next" href="engineering.html"/>
<script src="libs/header-attrs-2.13/header-attrs.js"></script>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="index.html#syllabus">Syllabus<span></span></a>
<ul>
<li><a href="course-schedule.html#course-schedule">Course Schedule<span></span></a></li>
<li><a href="instructors.html#instructors">Instructors<span></span></a></li>
<li><a href="office-hours.html#office-hours">Office Hours<span></span></a></li>
<li><a href="course-values-and-policies.html#course-values-and-policies">Course Values and Policies<span></span></a></li>
</ul></li>
<li class="chapter" data-level="1" data-path="introduction.html"><a href="introduction.html"><i class="fa fa-check"></i><b>1</b> Introduction<span></span></a>
<ul>
<li class="chapter" data-level="1.1" data-path="who-this-book-is-for.html"><a href="who-this-book-is-for.html"><i class="fa fa-check"></i><b>1.1</b> Who This Book Is For<span></span></a></li>
<li class="chapter" data-level="1.2" data-path="a-note-about-reinventing-the-wheel.html"><a href="a-note-about-reinventing-the-wheel.html"><i class="fa fa-check"></i><b>1.2</b> A Note About Reinventing the Wheel<span></span></a></li>
<li class="chapter" data-level="1.3" data-path="sources-and-references.html"><a href="sources-and-references.html"><i class="fa fa-check"></i><b>1.3</b> Sources and References<span></span></a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="data-bio.html"><a href="data-bio.html"><i class="fa fa-check"></i><b>2</b> Data in Biology<span></span></a>
<ul>
<li class="chapter" data-level="2.1" data-path="a-brief-history-of-data-in-molecular-biology.html"><a href="a-brief-history-of-data-in-molecular-biology.html"><i class="fa fa-check"></i><b>2.1</b> A Brief History of Data in Molecular Biology<span></span></a></li>
<li class="chapter" data-level="2.2" data-path="biology-as-a-mature-data-science.html"><a href="biology-as-a-mature-data-science.html"><i class="fa fa-check"></i><b>2.2</b> Biology as a Mature Data Science<span></span></a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="preliminaries.html"><a href="preliminaries.html"><i class="fa fa-check"></i><b>3</b> Preliminaries<span></span></a>
<ul>
<li class="chapter" data-level="3.1" data-path="prelim-r.html"><a href="prelim-r.html"><i class="fa fa-check"></i><b>3.1</b> The R Language<span></span></a></li>
<li class="chapter" data-level="3.2" data-path="prelim-rstudio.html"><a href="prelim-rstudio.html"><i class="fa fa-check"></i><b>3.2</b> RStudio<span></span></a></li>
<li class="chapter" data-level="3.3" data-path="prog-the-r-script.html"><a href="prog-the-r-script.html"><i class="fa fa-check"></i><b>3.3</b> The R Script<span></span></a></li>
<li class="chapter" data-level="3.4" data-path="prog-workflow.html"><a href="prog-workflow.html"><i class="fa fa-check"></i><b>3.4</b> The Scripting Workflow<span></span></a></li>
<li class="chapter" data-level="3.5" data-path="prelim-git.html"><a href="prelim-git.html"><i class="fa fa-check"></i><b>3.5</b> git + github<span></span></a>
<ul>
<li class="chapter" data-level="3.5.1" data-path="prelim-git.html"><a href="prelim-git.html#motivation"><i class="fa fa-check"></i><b>3.5.1</b> Motivation<span></span></a></li>
<li class="chapter" data-level="3.5.2" data-path="prelim-git.html"><a href="prelim-git.html#git"><i class="fa fa-check"></i><b>3.5.2</b> git<span></span></a></li>
<li class="chapter" data-level="3.5.3" data-path="prelim-git.html"><a href="prelim-git.html#git-hosting-platforms-github"><i class="fa fa-check"></i><b>3.5.3</b> Git hosting platforms (GitHub)<span></span></a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="prog-basics.html"><a href="prog-basics.html"><i class="fa fa-check"></i><b>4</b> R Programming<span></span></a>
<ul>
<li class="chapter" data-level="4.1" data-path="before-you-begin.html"><a href="before-you-begin.html"><i class="fa fa-check"></i><b>4.1</b> Before you begin<span></span></a></li>
<li class="chapter" data-level="4.2" data-path="introduction-1.html"><a href="introduction-1.html"><i class="fa fa-check"></i><b>4.2</b> Introduction<span></span></a></li>
<li class="chapter" data-level="4.3" data-path="prog-r-syntax.html"><a href="prog-r-syntax.html"><i class="fa fa-check"></i><b>4.3</b> R Syntax Basics<span></span></a></li>
<li class="chapter" data-level="4.4" data-path="prog-types.html"><a href="prog-types.html"><i class="fa fa-check"></i><b>4.4</b> Basic Types of Values<span></span></a>
<ul>
<li class="chapter" data-level="4.4.1" data-path="prog-types.html"><a href="prog-types.html#factors"><i class="fa fa-check"></i><b>4.4.1</b> Factors<span></span></a></li>
</ul></li>
<li class="chapter" data-level="4.5" data-path="prog-struct.html"><a href="prog-struct.html"><i class="fa fa-check"></i><b>4.5</b> Data Structures<span></span></a>
<ul>
<li class="chapter" data-level="4.5.1" data-path="prog-struct.html"><a href="prog-struct.html#vectors"><i class="fa fa-check"></i><b>4.5.1</b> Vectors<span></span></a></li>
<li class="chapter" data-level="4.5.2" data-path="prog-struct.html"><a href="prog-struct.html#matrices"><i class="fa fa-check"></i><b>4.5.2</b> Matrices<span></span></a></li>
<li class="chapter" data-level="4.5.3" data-path="prog-struct.html"><a href="prog-struct.html#lists-and-data-frames"><i class="fa fa-check"></i><b>4.5.3</b> Lists and data frames<span></span></a></li>
</ul></li>
<li class="chapter" data-level="4.6" data-path="logical-tests-and-comparators.html"><a href="logical-tests-and-comparators.html"><i class="fa fa-check"></i><b>4.6</b> Logical Tests and Comparators<span></span></a></li>
<li class="chapter" data-level="4.7" data-path="functions.html"><a href="functions.html"><i class="fa fa-check"></i><b>4.7</b> Functions<span></span></a>
<ul>
<li class="chapter" data-level="4.7.1" data-path="functions.html"><a href="functions.html#prog-dry"><i class="fa fa-check"></i><b>4.7.1</b> DRY: Don’t Repeat Yourself<span></span></a></li>
<li class="chapter" data-level="4.7.2" data-path="functions.html"><a href="functions.html#writing-your-own-functions"><i class="fa fa-check"></i><b>4.7.2</b> Writing your own functions<span></span></a></li>
<li class="chapter" data-level="4.7.3" data-path="functions.html"><a href="functions.html#scope"><i class="fa fa-check"></i><b>4.7.3</b> Scope<span></span></a></li>
</ul></li>
<li class="chapter" data-level="4.8" data-path="iteration.html"><a href="iteration.html"><i class="fa fa-check"></i><b>4.8</b> Iteration<span></span></a>
<ul>
<li class="chapter" data-level="4.8.1" data-path="iteration.html"><a href="iteration.html#vectorized-operations"><i class="fa fa-check"></i><b>4.8.1</b> Vectorized operations<span></span></a></li>
<li class="chapter" data-level="4.8.2" data-path="iteration.html"><a href="iteration.html#functional-programming"><i class="fa fa-check"></i><b>4.8.2</b> Functional programming<span></span></a></li>
<li class="chapter" data-level="4.8.3" data-path="iteration.html"><a href="iteration.html#apply-and-friends"><i class="fa fa-check"></i><b>4.8.3</b> <code>apply()</code> and friends<span></span></a></li>
</ul></li>
<li class="chapter" data-level="4.9" data-path="installing-packages.html"><a href="installing-packages.html"><i class="fa fa-check"></i><b>4.9</b> Installing Packages<span></span></a></li>
<li class="chapter" data-level="4.10" data-path="saving-and-loading-r-data.html"><a href="saving-and-loading-r-data.html"><i class="fa fa-check"></i><b>4.10</b> Saving and Loading R Data<span></span></a></li>
<li class="chapter" data-level="4.11" data-path="prog-debugging.html"><a href="prog-debugging.html"><i class="fa fa-check"></i><b>4.11</b> Troubleshooting and Debugging<span></span></a></li>
<li class="chapter" data-level="4.12" data-path="prog-style.html"><a href="prog-style.html"><i class="fa fa-check"></i><b>4.12</b> Coding Style and Conventions<span></span></a>
<ul>
<li class="chapter" data-level="4.12.1" data-path="prog-style.html"><a href="prog-style.html#is-my-code-correct"><i class="fa fa-check"></i><b>4.12.1</b> Is my code correct?<span></span></a></li>
<li class="chapter" data-level="4.12.2" data-path="prog-style.html"><a href="prog-style.html#does-my-code-follow-the-dry-principle"><i class="fa fa-check"></i><b>4.12.2</b> Does my code follow the <span>DRY principle</span>?<span></span></a></li>
<li class="chapter" data-level="4.12.3" data-path="prog-style.html"><a href="prog-style.html#did-i-choose-concise-but-descriptive-variable-and-function-names"><i class="fa fa-check"></i><b>4.12.3</b> Did I choose concise but descriptive variable and function names?<span></span></a></li>
<li class="chapter" data-level="4.12.4" data-path="prog-style.html"><a href="prog-style.html#did-i-use-indentation-and-naming-conventions-consistently-throughout-my-code"><i class="fa fa-check"></i><b>4.12.4</b> Did I use indentation and naming conventions consistently throughout my code?<span></span></a></li>
<li class="chapter" data-level="4.12.5" data-path="prog-style.html"><a href="prog-style.html#did-i-write-comments-especially-when-what-the-code-does-is-not-obvious"><i class="fa fa-check"></i><b>4.12.5</b> Did I write comments, especially when what the code does is not obvious?<span></span></a></li>
<li class="chapter" data-level="4.12.6" data-path="prog-style.html"><a href="prog-style.html#how-easy-would-it-be-for-someone-else-to-understand-my-code"><i class="fa fa-check"></i><b>4.12.6</b> How easy would it be for someone else to understand my code?<span></span></a></li>
<li class="chapter" data-level="4.12.7" data-path="prog-style.html"><a href="prog-style.html#is-my-code-easy-to-maintainchange"><i class="fa fa-check"></i><b>4.12.7</b> Is my code easy to maintain/change?<span></span></a></li>
<li class="chapter" data-level="4.12.8" data-path="prog-style.html"><a href="prog-style.html#the-styler-package"><i class="fa fa-check"></i><b>4.12.8</b> The <code>styler</code> package<span></span></a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="5" data-path="data-wrangling.html"><a href="data-wrangling.html"><i class="fa fa-check"></i><b>5</b> Data Wrangling<span></span></a>
<ul>
<li class="chapter" data-level="5.1" data-path="the-tidyverse.html"><a href="the-tidyverse.html"><i class="fa fa-check"></i><b>5.1</b> The Tidyverse<span></span></a></li>
<li class="chapter" data-level="5.2" data-path="dw-basics.html"><a href="dw-basics.html"><i class="fa fa-check"></i><b>5.2</b> Tidyverse Basics<span></span></a></li>
<li class="chapter" data-level="5.3" data-path="importing-data.html"><a href="importing-data.html"><i class="fa fa-check"></i><b>5.3</b> Importing Data<span></span></a></li>
<li class="chapter" data-level="5.4" data-path="data-tibble.html"><a href="data-tibble.html"><i class="fa fa-check"></i><b>5.4</b> The tibble<span></span></a></li>
<li class="chapter" data-level="5.5" data-path="tidy-data.html"><a href="tidy-data.html"><i class="fa fa-check"></i><b>5.5</b> Tidy Data<span></span></a></li>
<li class="chapter" data-level="5.6" data-path="pipes.html"><a href="pipes.html"><i class="fa fa-check"></i><b>5.6</b> pipes<span></span></a></li>
<li class="chapter" data-level="5.7" data-path="arranging-data.html"><a href="arranging-data.html"><i class="fa fa-check"></i><b>5.7</b> Arranging Data<span></span></a>
<ul>
<li class="chapter" data-level="5.7.1" data-path="arranging-data.html"><a href="arranging-data.html#dplyrmutate---create-new-columns-using-other-columns"><i class="fa fa-check"></i><b>5.7.1</b> <code>dplyr::mutate()</code> - Create new columns using other columns<span></span></a></li>
<li class="chapter" data-level="5.7.2" data-path="arranging-data.html"><a href="arranging-data.html#stringr---working-with-character-values"><i class="fa fa-check"></i><b>5.7.2</b> <code>stringr</code> - Working with character values<span></span></a></li>
<li class="chapter" data-level="5.7.3" data-path="arranging-data.html"><a href="arranging-data.html#dplyrselect---subset-columns-by-name"><i class="fa fa-check"></i><b>5.7.3</b> <code>dplyr::select()</code> - Subset Columns by Name<span></span></a></li>
<li class="chapter" data-level="5.7.4" data-path="arranging-data.html"><a href="arranging-data.html#dplyrfilter---pick-rows-out-of-a-data-set"><i class="fa fa-check"></i><b>5.7.4</b> <code>dplyr::filter()</code> - Pick rows out of a data set<span></span></a></li>
<li class="chapter" data-level="5.7.5" data-path="arranging-data.html"><a href="arranging-data.html#dplyrarrange---order-rows-based-on-their-values"><i class="fa fa-check"></i><b>5.7.5</b> <code>dplyr::arrange()</code> - Order rows based on their values<span></span></a></li>
<li class="chapter" data-level="5.7.6" data-path="arranging-data.html"><a href="arranging-data.html#putting-it-all-together"><i class="fa fa-check"></i><b>5.7.6</b> Putting it all together<span></span></a></li>
</ul></li>
<li class="chapter" data-level="5.8" data-path="grouping-data.html"><a href="grouping-data.html"><i class="fa fa-check"></i><b>5.8</b> Grouping Data<span></span></a></li>
<li class="chapter" data-level="5.9" data-path="rearranging-data.html"><a href="rearranging-data.html"><i class="fa fa-check"></i><b>5.9</b> Rearranging Data<span></span></a></li>
<li class="chapter" data-level="5.10" data-path="relational-data.html"><a href="relational-data.html"><i class="fa fa-check"></i><b>5.10</b> Relational Data<span></span></a>
<ul>
<li class="chapter" data-level="5.10.1" data-path="relational-data.html"><a href="relational-data.html#dealing-with-multiple-matches"><i class="fa fa-check"></i><b>5.10.1</b> Dealing with multiple matches<span></span></a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="6" data-path="data-science.html"><a href="data-science.html"><i class="fa fa-check"></i><b>6</b> Data Science<span></span></a>
<ul>
<li class="chapter" data-level="6.1" data-path="data-modeling.html"><a href="data-modeling.html"><i class="fa fa-check"></i><b>6.1</b> Data Modeling<span></span></a>
<ul>
<li class="chapter" data-level="6.1.1" data-path="data-modeling.html"><a href="data-modeling.html#a-worked-modeling-example"><i class="fa fa-check"></i><b>6.1.1</b> A Worked Modeling Example<span></span></a></li>
<li class="chapter" data-level="6.1.2" data-path="data-modeling.html"><a href="data-modeling.html#data-summarization"><i class="fa fa-check"></i><b>6.1.2</b> Data Summarization<span></span></a></li>
<li class="chapter" data-level="6.1.3" data-path="data-modeling.html"><a href="data-modeling.html#linear-models"><i class="fa fa-check"></i><b>6.1.3</b> Linear Models<span></span></a></li>
<li class="chapter" data-level="6.1.4" data-path="data-modeling.html"><a href="data-modeling.html#flavors-of-linear-models"><i class="fa fa-check"></i><b>6.1.4</b> Flavors of Linear Models<span></span></a></li>
<li class="chapter" data-level="6.1.5" data-path="data-modeling.html"><a href="data-modeling.html#assessing-model-accuracy-."><i class="fa fa-check"></i><b>6.1.5</b> Assessing Model Accuracy .<span></span></a></li>
</ul></li>
<li class="chapter" data-level="6.2" data-path="statistical-distributions.html"><a href="statistical-distributions.html"><i class="fa fa-check"></i><b>6.2</b> Statistical Distributions<span></span></a>
<ul>
<li class="chapter" data-level="6.2.1" data-path="statistical-distributions.html"><a href="statistical-distributions.html#random-variables"><i class="fa fa-check"></i><b>6.2.1</b> Random Variables<span></span></a></li>
<li class="chapter" data-level="6.2.2" data-path="statistical-distributions.html"><a href="statistical-distributions.html#statistical-distribution-basics"><i class="fa fa-check"></i><b>6.2.2</b> Statistical Distribution Basics<span></span></a></li>
<li class="chapter" data-level="6.2.3" data-path="statistical-distributions.html"><a href="statistical-distributions.html#distributions-in-r"><i class="fa fa-check"></i><b>6.2.3</b> Distributions in R<span></span></a></li>
<li class="chapter" data-level="6.2.4" data-path="statistical-distributions.html"><a href="statistical-distributions.html#discrete-distributions"><i class="fa fa-check"></i><b>6.2.4</b> Discrete Distributions<span></span></a></li>
<li class="chapter" data-level="6.2.5" data-path="statistical-distributions.html"><a href="statistical-distributions.html#continuous-distributions"><i class="fa fa-check"></i><b>6.2.5</b> Continuous Distributions<span></span></a></li>
<li class="chapter" data-level="6.2.6" data-path="statistical-distributions.html"><a href="statistical-distributions.html#empirical-distributions"><i class="fa fa-check"></i><b>6.2.6</b> Empirical Distributions<span></span></a></li>
</ul></li>
<li class="chapter" data-level="6.3" data-path="statistical-tests.html"><a href="statistical-tests.html"><i class="fa fa-check"></i><b>6.3</b> Statistical Tests<span></span></a>
<ul>
<li class="chapter" data-level="6.3.1" data-path="statistical-tests.html"><a href="statistical-tests.html#what-is-a-statistical-test"><i class="fa fa-check"></i><b>6.3.1</b> What is a statistical test<span></span></a></li>
<li class="chapter" data-level="6.3.2" data-path="statistical-tests.html"><a href="statistical-tests.html#common-statistical-tests"><i class="fa fa-check"></i><b>6.3.2</b> Common Statistical Tests<span></span></a></li>
<li class="chapter" data-level="6.3.3" data-path="statistical-tests.html"><a href="statistical-tests.html#choosing-a-test"><i class="fa fa-check"></i><b>6.3.3</b> Choosing a Test<span></span></a></li>
<li class="chapter" data-level="6.3.4" data-path="statistical-tests.html"><a href="statistical-tests.html#p-values"><i class="fa fa-check"></i><b>6.3.4</b> P-values<span></span></a></li>
<li class="chapter" data-level="6.3.5" data-path="statistical-tests.html"><a href="statistical-tests.html#multiple-hypothesis-testing-."><i class="fa fa-check"></i><b>6.3.5</b> Multiple Hypothesis Testing .<span></span></a></li>
<li class="chapter" data-level="6.3.6" data-path="statistical-tests.html"><a href="statistical-tests.html#statistical-power"><i class="fa fa-check"></i><b>6.3.6</b> Statistical power<span></span></a></li>
</ul></li>
<li class="chapter" data-level="6.4" data-path="exploratory-data-analysis.html"><a href="exploratory-data-analysis.html"><i class="fa fa-check"></i><b>6.4</b> Exploratory Data Analysis<span></span></a>
<ul>
<li class="chapter" data-level="6.4.1" data-path="exploratory-data-analysis.html"><a href="exploratory-data-analysis.html#principal-component-analysis"><i class="fa fa-check"></i><b>6.4.1</b> Principal Component Analysis<span></span></a></li>
<li class="chapter" data-level="6.4.2" data-path="exploratory-data-analysis.html"><a href="exploratory-data-analysis.html#cluster-analysis"><i class="fa fa-check"></i><b>6.4.2</b> Cluster Analysis<span></span></a></li>
<li class="chapter" data-level="6.4.3" data-path="exploratory-data-analysis.html"><a href="exploratory-data-analysis.html#k-means-."><i class="fa fa-check"></i><b>6.4.3</b> k-means .<span></span></a></li>
<li class="chapter" data-level="6.4.4" data-path="exploratory-data-analysis.html"><a href="exploratory-data-analysis.html#others"><i class="fa fa-check"></i><b>6.4.4</b> Others<span></span></a></li>
</ul></li>
<li class="chapter" data-level="6.5" data-path="network-analysis-..html"><a href="network-analysis-..html"><i class="fa fa-check"></i><b>6.5</b> Network Analysis .<span></span></a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="data-visualization.html"><a href="data-visualization.html"><i class="fa fa-check"></i><b>7</b> Data Visualization<span></span></a>
<ul>
<li class="chapter" data-level="7.1" data-path="responsible-plotting.html"><a href="responsible-plotting.html"><i class="fa fa-check"></i><b>7.1</b> Responsible Plotting<span></span></a>
<ul>
<li class="chapter" data-level="7.1.1" data-path="responsible-plotting.html"><a href="responsible-plotting.html#visualization-principles"><i class="fa fa-check"></i><b>7.1.1</b> Visualization Principles<span></span></a></li>
<li class="chapter" data-level="7.1.2" data-path="responsible-plotting.html"><a href="responsible-plotting.html#human-visual-perception"><i class="fa fa-check"></i><b>7.1.2</b> Human Visual Perception<span></span></a></li>
<li class="chapter" data-level="7.1.3" data-path="responsible-plotting.html"><a href="responsible-plotting.html#visual-encodings"><i class="fa fa-check"></i><b>7.1.3</b> Visual Encodings<span></span></a></li>
<li class="chapter" data-level="7.1.4" data-path="responsible-plotting.html"><a href="responsible-plotting.html#some-opinionated-rules-of-thumb"><i class="fa fa-check"></i><b>7.1.4</b> Some Opinionated Rules of Thumb<span></span></a></li>
</ul></li>
<li class="chapter" data-level="7.2" data-path="dv-gg.html"><a href="dv-gg.html"><i class="fa fa-check"></i><b>7.2</b> Grammar of Graphics<span></span></a>
<ul>
<li class="chapter" data-level="7.2.1" data-path="dv-gg.html"><a href="dv-gg.html#ggplot-mechanics"><i class="fa fa-check"></i><b>7.2.1</b> <code>ggplot</code> mechanics<span></span></a></li>
</ul></li>
<li class="chapter" data-level="7.3" data-path="plotting-one-dimension.html"><a href="plotting-one-dimension.html"><i class="fa fa-check"></i><b>7.3</b> Plotting One Dimension<span></span></a>
<ul>
<li class="chapter" data-level="7.3.1" data-path="plotting-one-dimension.html"><a href="plotting-one-dimension.html#dv-bar"><i class="fa fa-check"></i><b>7.3.1</b> Bar chart<span></span></a></li>
<li class="chapter" data-level="7.3.2" data-path="plotting-one-dimension.html"><a href="plotting-one-dimension.html#lollipop-plots"><i class="fa fa-check"></i><b>7.3.2</b> Lollipop plots<span></span></a></li>
<li class="chapter" data-level="7.3.3" data-path="plotting-one-dimension.html"><a href="plotting-one-dimension.html#stacked-area-charts"><i class="fa fa-check"></i><b>7.3.3</b> Stacked Area charts<span></span></a></li>
<li class="chapter" data-level="7.3.4" data-path="plotting-one-dimension.html"><a href="plotting-one-dimension.html#parallel-coordinate-plots"><i class="fa fa-check"></i><b>7.3.4</b> Parallel Coordinate plots<span></span></a></li>
</ul></li>
<li class="chapter" data-level="7.4" data-path="visualizing-distributions.html"><a href="visualizing-distributions.html"><i class="fa fa-check"></i><b>7.4</b> Visualizing Distributions<span></span></a>
<ul>
<li class="chapter" data-level="7.4.1" data-path="visualizing-distributions.html"><a href="visualizing-distributions.html#histogram"><i class="fa fa-check"></i><b>7.4.1</b> Histogram<span></span></a></li>
<li class="chapter" data-level="7.4.2" data-path="visualizing-distributions.html"><a href="visualizing-distributions.html#density"><i class="fa fa-check"></i><b>7.4.2</b> Density<span></span></a></li>
<li class="chapter" data-level="7.4.3" data-path="visualizing-distributions.html"><a href="visualizing-distributions.html#boxplot"><i class="fa fa-check"></i><b>7.4.3</b> Boxplot<span></span></a></li>
<li class="chapter" data-level="7.4.4" data-path="visualizing-distributions.html"><a href="visualizing-distributions.html#violin-plot"><i class="fa fa-check"></i><b>7.4.4</b> Violin plot<span></span></a></li>
<li class="chapter" data-level="7.4.5" data-path="visualizing-distributions.html"><a href="visualizing-distributions.html#beeswarm-plot"><i class="fa fa-check"></i><b>7.4.5</b> Beeswarm plot<span></span></a></li>
<li class="chapter" data-level="7.4.6" data-path="visualizing-distributions.html"><a href="visualizing-distributions.html#ridgeline"><i class="fa fa-check"></i><b>7.4.6</b> Ridgeline<span></span></a></li>
</ul></li>
<li class="chapter" data-level="7.5" data-path="plotting-two-or-more-dimensions.html"><a href="plotting-two-or-more-dimensions.html"><i class="fa fa-check"></i><b>7.5</b> Plotting Two or More Dimensions<span></span></a>
<ul>
<li class="chapter" data-level="7.5.1" data-path="plotting-two-or-more-dimensions.html"><a href="plotting-two-or-more-dimensions.html#scatter-plots"><i class="fa fa-check"></i><b>7.5.1</b> Scatter Plots<span></span></a></li>
<li class="chapter" data-level="7.5.2" data-path="plotting-two-or-more-dimensions.html"><a href="plotting-two-or-more-dimensions.html#bubble-plots"><i class="fa fa-check"></i><b>7.5.2</b> Bubble Plots<span></span></a></li>
<li class="chapter" data-level="7.5.3" data-path="plotting-two-or-more-dimensions.html"><a href="plotting-two-or-more-dimensions.html#connected-scatter-plots"><i class="fa fa-check"></i><b>7.5.3</b> Connected Scatter Plots<span></span></a></li>
<li class="chapter" data-level="7.5.4" data-path="plotting-two-or-more-dimensions.html"><a href="plotting-two-or-more-dimensions.html#line-plots"><i class="fa fa-check"></i><b>7.5.4</b> Line Plots<span></span></a></li>
<li class="chapter" data-level="7.5.5" data-path="plotting-two-or-more-dimensions.html"><a href="plotting-two-or-more-dimensions.html#parallel-coordinate-plots-1"><i class="fa fa-check"></i><b>7.5.5</b> Parallel Coordinate Plots<span></span></a></li>
<li class="chapter" data-level="7.5.6" data-path="plotting-two-or-more-dimensions.html"><a href="plotting-two-or-more-dimensions.html#heatmaps"><i class="fa fa-check"></i><b>7.5.6</b> Heatmaps<span></span></a></li>
<li class="chapter" data-level="7.5.7" data-path="plotting-two-or-more-dimensions.html"><a href="plotting-two-or-more-dimensions.html#how-to-use-heatmaps-responsibly"><i class="fa fa-check"></i><b>7.5.7</b> How To Use Heatmaps Responsibly<span></span></a></li>
</ul></li>
<li class="chapter" data-level="7.6" data-path="other-kind-of-plots.html"><a href="other-kind-of-plots.html"><i class="fa fa-check"></i><b>7.6</b> Other Kind of Plots<span></span></a>
<ul>
<li class="chapter" data-level="7.6.1" data-path="other-kind-of-plots.html"><a href="other-kind-of-plots.html#dendrograms"><i class="fa fa-check"></i><b>7.6.1</b> Dendrograms<span></span></a></li>
<li class="chapter" data-level="7.6.2" data-path="other-kind-of-plots.html"><a href="other-kind-of-plots.html#chord-diagrams-and-circos-plots"><i class="fa fa-check"></i><b>7.6.2</b> Chord Diagrams and Circos Plots<span></span></a></li>
</ul></li>
<li class="chapter" data-level="7.7" data-path="multiple-plots.html"><a href="multiple-plots.html"><i class="fa fa-check"></i><b>7.7</b> Multiple Plots<span></span></a>
<ul>
<li class="chapter" data-level="7.7.1" data-path="multiple-plots.html"><a href="multiple-plots.html#facet-wrapping"><i class="fa fa-check"></i><b>7.7.1</b> Facet wrapping<span></span></a></li>
<li class="chapter" data-level="7.7.2" data-path="multiple-plots.html"><a href="multiple-plots.html#multipanel-figures"><i class="fa fa-check"></i><b>7.7.2</b> Multipanel Figures<span></span></a></li>
</ul></li>
<li class="chapter" data-level="7.8" data-path="publication-ready-plots.html"><a href="publication-ready-plots.html"><i class="fa fa-check"></i><b>7.8</b> Publication Ready Plots<span></span></a>
<ul>
<li class="chapter" data-level="7.8.1" data-path="publication-ready-plots.html"><a href="publication-ready-plots.html#ggplot-themes"><i class="fa fa-check"></i><b>7.8.1</b> <code>ggplot</code> Themes<span></span></a></li>
<li class="chapter" data-level="7.8.2" data-path="publication-ready-plots.html"><a href="publication-ready-plots.html#exporting-to-svg"><i class="fa fa-check"></i><b>7.8.2</b> Exporting to SVG<span></span></a></li>
</ul></li>
<li class="chapter" data-level="7.9" data-path="network-visualization.html"><a href="network-visualization.html"><i class="fa fa-check"></i><b>7.9</b> Network visualization<span></span></a>
<ul>
<li class="chapter" data-level="7.9.1" data-path="network-visualization.html"><a href="network-visualization.html#interactive-network-plots"><i class="fa fa-check"></i><b>7.9.1</b> Interactive network plots<span></span></a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="8" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html"><i class="fa fa-check"></i><b>8</b> Biology &amp; Bioinformatics<span></span></a>
<ul>
<li class="chapter" data-level="8.1" data-path="bio-r.html"><a href="bio-r.html"><i class="fa fa-check"></i><b>8.1</b> R in Biology<span></span></a></li>
<li class="chapter" data-level="8.2" data-path="biological-data-overview.html"><a href="biological-data-overview.html"><i class="fa fa-check"></i><b>8.2</b> Biological Data Overview<span></span></a>
<ul>
<li class="chapter" data-level="8.2.1" data-path="biological-data-overview.html"><a href="biological-data-overview.html#bio-data-types"><i class="fa fa-check"></i><b>8.2.1</b> Types of Biological Data<span></span></a></li>
<li class="chapter" data-level="8.2.2" data-path="biological-data-overview.html"><a href="biological-data-overview.html#csv-files"><i class="fa fa-check"></i><b>8.2.2</b> CSV Files<span></span></a></li>
<li class="chapter" data-level="8.2.3" data-path="biological-data-overview.html"><a href="biological-data-overview.html#biological-data-is-not-tidy"><i class="fa fa-check"></i><b>8.2.3</b> Biological data is NOT Tidy!<span></span></a></li>
</ul></li>
<li class="chapter" data-level="8.3" data-path="bio-bioconductor.html"><a href="bio-bioconductor.html"><i class="fa fa-check"></i><b>8.3</b> Bioconductor<span></span></a></li>
<li class="chapter" data-level="8.4" data-path="microarrays.html"><a href="microarrays.html"><i class="fa fa-check"></i><b>8.4</b> Microarrays<span></span></a></li>
<li class="chapter" data-level="8.5" data-path="high-throughput-sequencing.html"><a href="high-throughput-sequencing.html"><i class="fa fa-check"></i><b>8.5</b> High Throughput Sequencing<span></span></a>
<ul>
<li class="chapter" data-level="8.5.1" data-path="high-throughput-sequencing.html"><a href="high-throughput-sequencing.html#raw-hts-data"><i class="fa fa-check"></i><b>8.5.1</b> Raw HTS Data<span></span></a></li>
<li class="chapter" data-level="8.5.2" data-path="high-throughput-sequencing.html"><a href="high-throughput-sequencing.html#preliminary-hts-data-analysis"><i class="fa fa-check"></i><b>8.5.2</b> Preliminary HTS Data Analysis<span></span></a></li>
<li class="chapter" data-level="8.5.3" data-path="high-throughput-sequencing.html"><a href="high-throughput-sequencing.html#count-data"><i class="fa fa-check"></i><b>8.5.3</b> Count Data<span></span></a></li>
</ul></li>
<li class="chapter" data-level="8.6" data-path="gene-identifiers.html"><a href="gene-identifiers.html"><i class="fa fa-check"></i><b>8.6</b> Gene Identifiers<span></span></a>
<ul>
<li class="chapter" data-level="8.6.1" data-path="gene-identifiers.html"><a href="gene-identifiers.html#gene-identifier-systems"><i class="fa fa-check"></i><b>8.6.1</b> Gene Identifier Systems<span></span></a></li>
<li class="chapter" data-level="8.6.2" data-path="gene-identifiers.html"><a href="gene-identifiers.html#mapping-between-identifier-systems"><i class="fa fa-check"></i><b>8.6.2</b> Mapping Between Identifier Systems<span></span></a></li>
<li class="chapter" data-level="8.6.3" data-path="gene-identifiers.html"><a href="gene-identifiers.html#mapping-homologs"><i class="fa fa-check"></i><b>8.6.3</b> Mapping Homologs<span></span></a></li>
</ul></li>
<li class="chapter" data-level="8.7" data-path="gene-expression.html"><a href="gene-expression.html"><i class="fa fa-check"></i><b>8.7</b> Gene Expression<span></span></a>
<ul>
<li class="chapter" data-level="8.7.1" data-path="gene-expression.html"><a href="gene-expression.html#gene-expression-data-in-bioconductor"><i class="fa fa-check"></i><b>8.7.1</b> Gene Expression Data in Bioconductor<span></span></a></li>
<li class="chapter" data-level="8.7.2" data-path="gene-expression.html"><a href="gene-expression.html#differential-expression-analysis"><i class="fa fa-check"></i><b>8.7.2</b> Differential Expression Analysis<span></span></a></li>
<li class="chapter" data-level="8.7.3" data-path="gene-expression.html"><a href="gene-expression.html#microarray-gene-expression-data"><i class="fa fa-check"></i><b>8.7.3</b> Microarray Gene Expression Data<span></span></a></li>
<li class="chapter" data-level="8.7.4" data-path="gene-expression.html"><a href="gene-expression.html#differential-expression-microarrays-limma"><i class="fa fa-check"></i><b>8.7.4</b> Differential Expression: Microarrays (limma)<span></span></a></li>
<li class="chapter" data-level="8.7.5" data-path="gene-expression.html"><a href="gene-expression.html#rnaseq"><i class="fa fa-check"></i><b>8.7.5</b> RNASeq<span></span></a></li>
<li class="chapter" data-level="8.7.6" data-path="gene-expression.html"><a href="gene-expression.html#rnaseq-gene-expression-data"><i class="fa fa-check"></i><b>8.7.6</b> RNASeq Gene Expression Data<span></span></a></li>
<li class="chapter" data-level="8.7.7" data-path="gene-expression.html"><a href="gene-expression.html#filtering-counts"><i class="fa fa-check"></i><b>8.7.7</b> Filtering Counts<span></span></a></li>
<li class="chapter" data-level="8.7.8" data-path="gene-expression.html"><a href="gene-expression.html#count-distributions"><i class="fa fa-check"></i><b>8.7.8</b> Count Distributions<span></span></a></li>
<li class="chapter" data-level="8.7.9" data-path="gene-expression.html"><a href="gene-expression.html#differential-expression-rnaseq"><i class="fa fa-check"></i><b>8.7.9</b> Differential Expression: RNASeq<span></span></a></li>
</ul></li>
<li class="chapter" data-level="8.8" data-path="gene-set-enrichment-analysis.html"><a href="gene-set-enrichment-analysis.html"><i class="fa fa-check"></i><b>8.8</b> Gene Set Enrichment Analysis<span></span></a>
<ul>
<li class="chapter" data-level="8.8.1" data-path="gene-set-enrichment-analysis.html"><a href="gene-set-enrichment-analysis.html#gene-sets"><i class="fa fa-check"></i><b>8.8.1</b> Gene Sets<span></span></a></li>
<li class="chapter" data-level="8.8.2" data-path="gene-set-enrichment-analysis.html"><a href="gene-set-enrichment-analysis.html#working-with-gene-sets-in-r"><i class="fa fa-check"></i><b>8.8.2</b> Working with gene sets in R<span></span></a></li>
<li class="chapter" data-level="8.8.3" data-path="gene-set-enrichment-analysis.html"><a href="gene-set-enrichment-analysis.html#over-representation-analysis"><i class="fa fa-check"></i><b>8.8.3</b> Over-representation Analysis<span></span></a></li>
<li class="chapter" data-level="8.8.4" data-path="gene-set-enrichment-analysis.html"><a href="gene-set-enrichment-analysis.html#rank-based-analysis"><i class="fa fa-check"></i><b>8.8.4</b> Rank-based Analysis<span></span></a></li>
<li class="chapter" data-level="8.8.5" data-path="gene-set-enrichment-analysis.html"><a href="gene-set-enrichment-analysis.html#fgsea"><i class="fa fa-check"></i><b>8.8.5</b> fgsea<span></span></a></li>
</ul></li>
<li class="chapter" data-level="8.9" data-path="biological-networks-..html"><a href="biological-networks-..html"><i class="fa fa-check"></i><b>8.9</b> Biological Networks .<span></span></a>
<ul>
<li class="chapter" data-level="8.9.1" data-path="biological-networks-..html"><a href="biological-networks-..html#biological-pathways-."><i class="fa fa-check"></i><b>8.9.1</b> Biological Pathways .<span></span></a></li>
<li class="chapter" data-level="8.9.2" data-path="biological-networks-..html"><a href="biological-networks-..html#gene-regulatory-networks-."><i class="fa fa-check"></i><b>8.9.2</b> Gene Regulatory Networks .<span></span></a></li>
<li class="chapter" data-level="8.9.3" data-path="biological-networks-..html"><a href="biological-networks-..html#protein-protein-interaction-network-."><i class="fa fa-check"></i><b>8.9.3</b> Protein-Protein Interaction Network .<span></span></a></li>
<li class="chapter" data-level="8.9.4" data-path="biological-networks-..html"><a href="biological-networks-..html#wgcna-."><i class="fa fa-check"></i><b>8.9.4</b> WGCNA .<span></span></a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="9" data-path="engineering.html"><a href="engineering.html"><i class="fa fa-check"></i><b>9</b> EngineeRing<span></span></a>
<ul>
<li class="chapter" data-level="9.1" data-path="unit-testing.html"><a href="unit-testing.html"><i class="fa fa-check"></i><b>9.1</b> Unit Testing<span></span></a></li>
<li class="chapter" data-level="9.2" data-path="toolification.html"><a href="toolification.html"><i class="fa fa-check"></i><b>9.2</b> Toolification<span></span></a>
<ul>
<li class="chapter" data-level="9.2.1" data-path="toolification.html"><a href="toolification.html#the-r-interpreter"><i class="fa fa-check"></i><b>9.2.1</b> The R Interpreter<span></span></a></li>
<li class="chapter" data-level="9.2.2" data-path="toolification.html"><a href="toolification.html#rscript"><i class="fa fa-check"></i><b>9.2.2</b> <code>Rscript</code><span></span></a></li>
<li class="chapter" data-level="9.2.3" data-path="toolification.html"><a href="toolification.html#commandargs"><i class="fa fa-check"></i><b>9.2.3</b> <code>commandArgs()</code><span></span></a></li>
</ul></li>
<li class="chapter" data-level="9.3" data-path="parallel-processing.html"><a href="parallel-processing.html"><i class="fa fa-check"></i><b>9.3</b> Parallel Processing<span></span></a>
<ul>
<li class="chapter" data-level="9.3.1" data-path="parallel-processing.html"><a href="parallel-processing.html#brief-introduction-to-parallelization"><i class="fa fa-check"></i><b>9.3.1</b> Brief Introduction to Parallelization<span></span></a></li>
<li class="chapter" data-level="9.3.2" data-path="parallel-processing.html"><a href="parallel-processing.html#apply-and-friends-are-pleasingly-parallel"><i class="fa fa-check"></i><b>9.3.2</b> <code>apply</code> and Friends Are Pleasingly Parallel<span></span></a></li>
<li class="chapter" data-level="9.3.3" data-path="parallel-processing.html"><a href="parallel-processing.html#the-parallel-package"><i class="fa fa-check"></i><b>9.3.3</b> The <code>parallel</code> package<span></span></a></li>
</ul></li>
<li class="chapter" data-level="9.4" data-path="object-orientation-in-r-..html"><a href="object-orientation-in-r-..html"><i class="fa fa-check"></i><b>9.4</b> Object Orientation in R .<span></span></a></li>
<li class="chapter" data-level="9.5" data-path="r-packages-..html"><a href="r-packages-..html"><i class="fa fa-check"></i><b>9.5</b> R Packages .<span></span></a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="rshiny.html"><a href="rshiny.html"><i class="fa fa-check"></i><b>10</b> RShiny<span></span></a>
<ul>
<li class="chapter" data-level="10.1" data-path="rshiny-introduction.html"><a href="rshiny-introduction.html"><i class="fa fa-check"></i><b>10.1</b> RShiny Introduction<span></span></a>
<ul>
<li class="chapter" data-level="10.1.1" data-path="rshiny-introduction.html"><a href="rshiny-introduction.html#how-does-a-web-page-work"><i class="fa fa-check"></i><b>10.1.1</b> How does a web page work?<span></span></a></li>
</ul></li>
<li class="chapter" data-level="10.2" data-path="application-structure.html"><a href="application-structure.html"><i class="fa fa-check"></i><b>10.2</b> Application Structure<span></span></a>
<ul>
<li class="chapter" data-level="10.2.1" data-path="application-structure.html"><a href="application-structure.html#ui-design"><i class="fa fa-check"></i><b>10.2.1</b> UI Design<span></span></a></li>
<li class="chapter" data-level="10.2.2" data-path="application-structure.html"><a href="application-structure.html#server-functionality"><i class="fa fa-check"></i><b>10.2.2</b> Server Functionality<span></span></a></li>
</ul></li>
<li class="chapter" data-level="10.3" data-path="reactivity.html"><a href="reactivity.html"><i class="fa fa-check"></i><b>10.3</b> Reactivity<span></span></a></li>
<li class="chapter" data-level="10.4" data-path="publishing.html"><a href="publishing.html"><i class="fa fa-check"></i><b>10.4</b> Publishing<span></span></a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="communicating-with-r.html"><a href="communicating-with-r.html"><i class="fa fa-check"></i><b>11</b> Communicating with R<span></span></a>
<ul>
<li class="chapter" data-level="11.1" data-path="rmarkdown-knitr.html"><a href="rmarkdown-knitr.html"><i class="fa fa-check"></i><b>11.1</b> RMarkdown &amp; knitr<span></span></a></li>
<li class="chapter" data-level="11.2" data-path="bookdown.html"><a href="bookdown.html"><i class="fa fa-check"></i><b>11.2</b> bookdown<span></span></a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="contribution-guide.html"><a href="contribution-guide.html"><i class="fa fa-check"></i><b>12</b> Contribution Guide<span></span></a>
<ul>
<li class="chapter" data-level="12.1" data-path="custom-blocks.html"><a href="custom-blocks.html"><i class="fa fa-check"></i><b>12.1</b> Custom blocks<span></span></a></li>
</ul></li>
<li><a href="assignments.html#assignments">Assignments<span></span></a>
<ul>
<li><a href="assignment-format.html#assignment-format">Assignment Format<span></span></a></li>
<li><a href="starting-an-assignment.html#starting-an-assignment">Starting an Assignment<span></span></a>
<ul>
<li><a href="starting-an-assignment.html#starting-your-own-git-repository-and-cloning-it">Starting your own <code>git</code> repository and cloning it<span></span></a></li>
<li><a href="starting-an-assignment.html#contents">Contents<span></span></a></li>
<li><a href="starting-an-assignment.html#github-tutorial">GitHub tutorial<span></span></a></li>
<li><a href="starting-an-assignment.html#committing-and-pushing-with-r-and-without">Committing and Pushing with R (and without)<span></span></a></li>
<li><a href="starting-an-assignment.html#using-rstudio">Using RStudio<span></span></a></li>
<li><a href="starting-an-assignment.html#using-the-command-line">Using the command line<span></span></a></li>
</ul></li>
<li><a href="assignment-1.html#assignment-1">Assignment 1<span></span></a>
<ul>
<li><a href="assignment-1.html#problem-statement">Problem Statement<span></span></a></li>
<li><a href="assignment-1.html#learning-objectives-and-skill-list">Learning Objectives and Skill List<span></span></a></li>
<li><a href="assignment-1.html#instructions">Instructions<span></span></a></li>
<li><a href="assignment-1.html#hints">Hints<span></span></a></li>
</ul></li>
<li><a href="assignment-2.html#assignment-2">Assignment 2<span></span></a>
<ul>
<li><a href="assignment-2.html#problem-statement-1">Problem Statement<span></span></a></li>
<li><a href="assignment-2.html#learning-objectives">Learning Objectives<span></span></a></li>
<li><a href="assignment-2.html#skill-list">Skill List<span></span></a></li>
<li><a href="assignment-2.html#instructions-1">Instructions<span></span></a></li>
<li><a href="assignment-2.html#function-details">Function Details<span></span></a></li>
</ul></li>
<li><a href="assignment-3.html#assignment-3">Assignment 3<span></span></a>
<ul>
<li><a href="assignment-3.html#problem-statement-2">Problem Statement<span></span></a></li>
<li><a href="assignment-3.html#learning-objectives-1">Learning Objectives<span></span></a></li>
<li><a href="assignment-3.html#skill-list-1">Skill List<span></span></a></li>
<li><a href="assignment-3.html#background-on-microarrays">Background on Microarrays<span></span></a></li>
<li><a href="assignment-3.html#background-on-principal-component-analysis">Background on Principal Component Analysis<span></span></a></li>
<li><a href="assignment-3.html#marisa-et-al.-gene-expression-classification-of-colon-cancer-into-molecular-subtypes-characterization-validation-and-prognostic-value.-plos-medicine-may-2013.-pmid-23700391">Marisa et al. Gene Expression Classification of Colon Cancer into Molecular Subtypes: Characterization, Validation, and Prognostic Value. PLoS Medicine, May 2013. PMID: 23700391<span></span></a></li>
<li><a href="assignment-3.html#scaling-data-using-r-scale">Scaling data using R <code>scale()</code><span></span></a></li>
<li><a href="assignment-3.html#proportion-of-variance-explained">Proportion of variance explained<span></span></a></li>
<li><a href="assignment-3.html#plotting-and-visualization-of-pca">Plotting and visualization of PCA<span></span></a></li>
<li><a href="assignment-3.html#hierarchical-clustering-and-heatmaps">Hierarchical Clustering and Heatmaps<span></span></a></li>
<li><a href="assignment-3.html#references">References<span></span></a></li>
</ul></li>
<li><a href="assignment-4.html#assignment-4">Assignment 4<span></span></a>
<ul>
<li><a href="assignment-4.html#problem-statement-3">Problem Statement<span></span></a></li>
<li><a href="assignment-4.html#learning-objectives-2">Learning Objectives<span></span></a></li>
<li><a href="assignment-4.html#skill-list-2">Skill List<span></span></a></li>
<li><a href="assignment-4.html#instructions-2">Instructions<span></span></a></li>
<li><a href="assignment-4.html#deliverables">Deliverables<span></span></a></li>
<li><a href="assignment-4.html#function-details-1">Function Details<span></span></a></li>
<li><a href="assignment-4.html#hints-1">Hints<span></span></a></li>
</ul></li>
<li><a href="assignment-5.html#assignment-5">Assignment 5<span></span></a>
<ul>
<li><a href="assignment-5.html#problem-statement-4">Problem Statement<span></span></a></li>
<li><a href="assignment-5.html#learning-objectives-3">Learning Objectives<span></span></a></li>
<li><a href="assignment-5.html#skill-list-3">Skill List<span></span></a></li>
<li><a href="assignment-5.html#deseq2-background">DESeq2 Background<span></span></a></li>
<li><a href="assignment-5.html#generating-a-counts-matrix">Generating a counts matrix<span></span></a></li>
<li><a href="assignment-5.html#prefiltering-counts-matrix">Prefiltering Counts matrix<span></span></a></li>
<li><a href="assignment-5.html#median-of-ratios-normalization">Median-of-ratios normalization<span></span></a></li>
<li><a href="assignment-5.html#deseq2-preparation">DESeq2 preparation<span></span></a></li>
<li><a href="assignment-5.html#omeara-et-al.-transcriptional-reversion-of-cardiac-myocyte-fate-during-mammalian-cardiac-regeneration.-circ-res.-feb-2015.-pmid-25477501l">O’Meara et al. Transcriptional Reversion of Cardiac Myocyte Fate During Mammalian Cardiac Regeneration. Circ Res. Feb 2015. PMID: 25477501l<span></span></a></li>
<li><a href="assignment-5.html#reading-and-subsetting-the-data-from-verse_counts.tsv-and-sample_metadata.csv">1. Reading and subsetting the data from verse_counts.tsv and sample_metadata.csv<span></span></a></li>
<li><a href="assignment-5.html#running-deseq2">2. Running DESeq2<span></span></a></li>
<li><a href="assignment-5.html#annotating-results-to-construct-a-labeled-volcano-plot">3. Annotating results to construct a labeled volcano plot<span></span></a></li>
<li><a href="assignment-5.html#diagnostic-plot-of-the-raw-p-values-for-all-genes">4. Diagnostic plot of the raw p-values for all genes<span></span></a></li>
<li><a href="assignment-5.html#plotting-the-logfoldchanges-for-differentially-expressed-genes">5. Plotting the LogFoldChanges for differentially expressed genes<span></span></a></li>
<li><a href="assignment-5.html#the-choice-of-fdr-cutoff-depends-on-cost">The choice of FDR cutoff depends on cost<span></span></a></li>
<li><a href="assignment-5.html#plotting-the-normalized-counts-of-differentially-expressed-genes">6. Plotting the normalized counts of differentially expressed genes<span></span></a></li>
<li><a href="assignment-5.html#volcano-plot-to-visualize-differential-expression-results">7. Volcano Plot to visualize differential expression results<span></span></a></li>
<li><a href="assignment-5.html#running-fgsea-vignette">8. Running fgsea <span>vignette</span><span></span></a></li>
<li><a href="assignment-5.html#plotting-the-top-ten-positive-nes-and-top-ten-negative-nes-pathways">9. Plotting the top ten positive NES and top ten negative NES pathways<span></span></a></li>
<li><a href="assignment-5.html#references-1">References<span></span></a></li>
</ul></li>
<li><a href="assignment-6.html#assignment-6">Assignment 6<span></span></a>
<ul>
<li><a href="assignment-6.html#problem-statement-5">Problem Statement<span></span></a></li>
<li><a href="assignment-6.html#learning-objectives-4">Learning Objectives<span></span></a></li>
<li><a href="assignment-6.html#skill-list-4">Skill List<span></span></a></li>
<li><a href="assignment-6.html#instructions-3">Instructions<span></span></a></li>
<li><a href="assignment-6.html#function-details-2">Function Details<span></span></a></li>
</ul></li>
<li><a href="assignment-7.html#assignment-7">Assignment 7<span></span></a>
<ul>
<li><a href="assignment-7.html#problem-statement-6">Problem Statement<span></span></a></li>
<li><a href="assignment-7.html#learning-objectives-5">Learning Objectives<span></span></a></li>
<li><a href="assignment-7.html#skill-list-5">Skill List<span></span></a></li>
<li><a href="assignment-7.html#instructions-4">Instructions<span></span></a></li>
<li><a href="assignment-7.html#function-details-3">“Function” Details<span></span></a></li>
</ul></li>
</ul></li>
<li class="appendix"><span><b>Appendix<span></span></b></span></li>
<li class="chapter" data-level="A" data-path="class-outlines.html"><a href="class-outlines.html"><i class="fa fa-check"></i><b>A</b> Class Outlines<span></span></a>
<ul>
<li class="chapter" data-level="A.1" data-path="week-1.html"><a href="week-1.html"><i class="fa fa-check"></i><b>A.1</b> Week 1<span></span></a></li>
<li class="chapter" data-level="A.2" data-path="week-2.html"><a href="week-2.html"><i class="fa fa-check"></i><b>A.2</b> Week 2<span></span></a></li>
<li class="chapter" data-level="A.3" data-path="week-3.html"><a href="week-3.html"><i class="fa fa-check"></i><b>A.3</b> Week 3<span></span></a></li>
<li class="chapter" data-level="A.4" data-path="week-4.html"><a href="week-4.html"><i class="fa fa-check"></i><b>A.4</b> Week 4<span></span></a></li>
<li class="chapter" data-level="A.5" data-path="week-5.html"><a href="week-5.html"><i class="fa fa-check"></i><b>A.5</b> Week 5<span></span></a></li>
<li class="chapter" data-level="A.6" data-path="week-6.html"><a href="week-6.html"><i class="fa fa-check"></i><b>A.6</b> Week 6<span></span></a></li>
<li class="chapter" data-level="A.7" data-path="week-7.html"><a href="week-7.html"><i class="fa fa-check"></i><b>A.7</b> Week 7<span></span></a></li>
<li class="chapter" data-level="A.8" data-path="week-8.html"><a href="week-8.html"><i class="fa fa-check"></i><b>A.8</b> Week 8<span></span></a></li>
<li class="chapter" data-level="A.9" data-path="week-9.html"><a href="week-9.html"><i class="fa fa-check"></i><b>A.9</b> Week 9<span></span></a></li>
<li class="chapter" data-level="A.10" data-path="week-10.html"><a href="week-10.html"><i class="fa fa-check"></i><b>A.10</b> Week 10<span></span></a></li>
<li class="chapter" data-level="A.11" data-path="week-11.html"><a href="week-11.html"><i class="fa fa-check"></i><b>A.11</b> Week 11<span></span></a></li>
<li class="chapter" data-level="A.12" data-path="week-12.html"><a href="week-12.html"><i class="fa fa-check"></i><b>A.12</b> Week 12<span></span></a></li>
<li class="chapter" data-level="A.13" data-path="week-13.html"><a href="week-13.html"><i class="fa fa-check"></i><b>A.13</b> Week 13<span></span></a></li>
<li class="chapter" data-level="A.14" data-path="week-14.html"><a href="week-14.html"><i class="fa fa-check"></i><b>A.14</b> Week 14<span></span></a></li>
</ul></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">BF591 - R for Biological Sciences</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="biological-networks-." class="section level2 hasAnchor" number="8.9">
<h2><span class="header-section-number">8.9</span> Biological Networks .<a href="biological-networks-..html#biological-networks-." class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>What are biological networks? What do they represent (networks encode
relationships between entities)? Why are biological networks useful and what can
we learn from them?</p>
<div id="biological-pathways-." class="section level3 hasAnchor" number="8.9.1">
<h3><span class="header-section-number">8.9.1</span> Biological Pathways .<a href="biological-networks-..html#biological-pathways-." class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>What is a biological pathway? Why are they important, and how are they useful?</p>
</div>
<div id="gene-regulatory-networks-." class="section level3 hasAnchor" number="8.9.2">
<h3><span class="header-section-number">8.9.2</span> Gene Regulatory Networks .<a href="biological-networks-..html#gene-regulatory-networks-." class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>What are gene regulatory networks? Why are they important? How do we identify
them (data driven correlation, wetlab experiments, others?)? How are they
useful?</p>
</div>
<div id="protein-protein-interaction-network-." class="section level3 hasAnchor" number="8.9.3">
<h3><span class="header-section-number">8.9.3</span> Protein-Protein Interaction Network .<a href="biological-networks-..html#protein-protein-interaction-network-." class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>What are protein-protein interaction (PPI) networks? What information do they
represent (direct association, functional association, etc)? Where does PPI
information come from (datasets, databases, etc)? What are some ways we can use
PPI information when interpreting other biological data (like differential
expression? not sure)?</p>
</div>
<div id="wgcna-." class="section level3 hasAnchor" number="8.9.4">
<h3><span class="header-section-number">8.9.4</span> WGCNA .<a href="biological-networks-..html#wgcna-." class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Weighted correlation network analysis (WGCNA) identifies genes (or other biology features) that are highly correlated with each other, and group them into clusters / modules.<br />
Typical inputs of WGCNA is the expression matrix of genes, proteins or other biology features. Then, WGCNA will construct a undirected and weighted co-expression network. The nodes will be genes (or other features), and the edges connecting them will be the pairwise correlation value of their expression level. Modules are simply the clusters of highly connected genes. After the modules are detected, down stream analyses may include summarizing the module by its “eigengenes,” inferring the functionality of a gene from genes in the same module, or compare different modules.<br />
For more information, you may refer to this <a href="https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-9-559">article</a> or the <a href="https://cran.r-project.org/web/packages/WGCNA/WGCNA.pdf">manual of WGCNA R package</a>.</p>
<div id="how-to-run-wgcna-in-r" class="section level4 hasAnchor" number="8.9.4.1">
<h4><span class="header-section-number">8.9.4.1</span> How to run WGCNA in R<a href="biological-networks-..html#how-to-run-wgcna-in-r" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>First, install package <code>WGCNA</code>.</p>
<pre><code>BiocManager::install(&quot;WGCNA&quot;)
library(WGCNA)</code></pre>
<p>note: if you receive error message saying some dependencies are not available, it’s probably because you used to try to install it using standard method <code>install.package("WGCNA")</code>. If that’s the case, try the following to re-install WGCNA:</p>
<pre><code>BiocManager::install(&quot;WGCNA&quot;,force = T)</code></pre>
<p>Let’s load a toy gene expression matrix. We have 15 samples and 7 genes.</p>
<pre><code>##            CRAB     Whale     Lobst      Octop    Coral BabyShark     Orca
## samp1  6.867731  8.796212  7.820130  6.6081601 15.37307 14.047536 32.96333
## samp2 10.918217  6.775947 16.545788  7.5685123 28.59193  8.923224 19.13865
## samp3  5.821857  8.682239  5.089099  2.8030264 17.99007 14.643098 36.93207
## samp4 17.976404  5.741274 27.596721  8.8660342 34.62918  9.473205 18.68929
## samp5 11.647539 10.866047 14.852970  2.2034360 14.15414 15.949892 42.21472
## samp6  5.897658 11.960800 15.915636 -0.1724085 21.78596 16.336686 44.93125</code></pre>
<p>An optional choice in WGCNA to allow multithreads</p>
<div class="sourceCode" id="cb427"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb427-1"><a href="biological-networks-..html#cb427-1" aria-hidden="true" tabindex="-1"></a><span class="fu">allowWGCNAThreads</span>(<span class="dv">4</span>)</span></code></pre></div>
<pre><code>## Allowing multi-threading with up to 4 threads.</code></pre>
<p>Now we will choose a “soft-thresholding power” to construct the network. There is no single answer in the power to choose.<br />
If you refer to <a href="https://horvath.genetics.ucla.edu/html/CoexpressionNetwork/Rpackages/WGCNA/faq.html">WGCNA faq</a>, there are more details for power choosing.</p>
<div class="sourceCode" id="cb429"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb429-1"><a href="biological-networks-..html#cb429-1" aria-hidden="true" tabindex="-1"></a><span class="co"># A set of soft-thresholding powers to choose from:</span></span>
<span id="cb429-2"><a href="biological-networks-..html#cb429-2" aria-hidden="true" tabindex="-1"></a>powers <span class="ot">&lt;-</span> <span class="fu">seq</span>(<span class="dv">1</span>, <span class="dv">10</span>)</span>
<span id="cb429-3"><a href="biological-networks-..html#cb429-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb429-4"><a href="biological-networks-..html#cb429-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Call the network topology analysis function</span></span>
<span id="cb429-5"><a href="biological-networks-..html#cb429-5" aria-hidden="true" tabindex="-1"></a>sft <span class="ot">&lt;-</span> <span class="fu">pickSoftThreshold</span>(</span>
<span id="cb429-6"><a href="biological-networks-..html#cb429-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">data =</span> dat,</span>
<span id="cb429-7"><a href="biological-networks-..html#cb429-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">powerVector =</span> powers,</span>
<span id="cb429-8"><a href="biological-networks-..html#cb429-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">verbose =</span> <span class="dv">5</span></span>
<span id="cb429-9"><a href="biological-networks-..html#cb429-9" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div>
<pre><code>## pickSoftThreshold: will use block size 7.
##  pickSoftThreshold: calculating connectivity for given powers...
##    ..working on genes 1 through 7 of 7
##    Power SFT.R.sq   slope truncated.R.sq mean.k. median.k. max.k.
## 1      1 0.026500 -12.800        0.38900   1.920     1.820  2.350
## 2      2 0.007510   2.420       -0.18500   1.240     1.180  1.730
## 3      3 0.001090   0.582        0.00340   0.931     0.908  1.430
## 4      4 0.001300  -0.460       -0.14200   0.733     0.708  1.220
## 5      5 0.003130   0.526       -0.03820   0.592     0.556  1.060
## 6      6 0.018100  -1.150        0.04810   0.486     0.440  0.922
## 7      7 0.087400  -2.130       -0.15900   0.404     0.349  0.807
## 8      8 0.063500  -1.570       -0.11100   0.338     0.279  0.708
## 9      9 0.000378  -0.206        0.00319   0.285     0.224  0.621
## 10    10 0.003650  -0.600       -0.00478   0.241     0.181  0.545</code></pre>
<div class="sourceCode" id="cb431"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb431-1"><a href="biological-networks-..html#cb431-1" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>))</span>
<span id="cb431-2"><a href="biological-networks-..html#cb431-2" aria-hidden="true" tabindex="-1"></a>cex1 <span class="ot">&lt;-</span> <span class="fl">0.9</span></span>
<span id="cb431-3"><a href="biological-networks-..html#cb431-3" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(sft<span class="sc">$</span>fitIndices[, <span class="dv">1</span>],</span>
<span id="cb431-4"><a href="biological-networks-..html#cb431-4" aria-hidden="true" tabindex="-1"></a>  <span class="sc">-</span><span class="fu">sign</span>(sft<span class="sc">$</span>fitIndices[, <span class="dv">3</span>]) <span class="sc">*</span> sft<span class="sc">$</span>fitIndices[, <span class="dv">2</span>],</span>
<span id="cb431-5"><a href="biological-networks-..html#cb431-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">xlab =</span> <span class="st">&quot;Soft Threshold (power)&quot;</span>,</span>
<span id="cb431-6"><a href="biological-networks-..html#cb431-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">ylab =</span> <span class="st">&quot;Scale Free Topology Model Fit, signed R^2&quot;</span>,</span>
<span id="cb431-7"><a href="biological-networks-..html#cb431-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">main =</span> <span class="fu">paste</span>(<span class="st">&quot;Scale independence&quot;</span>)</span>
<span id="cb431-8"><a href="biological-networks-..html#cb431-8" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb431-9"><a href="biological-networks-..html#cb431-9" aria-hidden="true" tabindex="-1"></a><span class="fu">text</span>(sft<span class="sc">$</span>fitIndices[, <span class="dv">1</span>],</span>
<span id="cb431-10"><a href="biological-networks-..html#cb431-10" aria-hidden="true" tabindex="-1"></a>  <span class="sc">-</span><span class="fu">sign</span>(sft<span class="sc">$</span>fitIndices[, <span class="dv">3</span>]) <span class="sc">*</span> sft<span class="sc">$</span>fitIndices[, <span class="dv">2</span>],</span>
<span id="cb431-11"><a href="biological-networks-..html#cb431-11" aria-hidden="true" tabindex="-1"></a>  <span class="at">labels =</span> powers, <span class="at">cex =</span> cex1, <span class="at">col =</span> <span class="st">&quot;red&quot;</span></span>
<span id="cb431-12"><a href="biological-networks-..html#cb431-12" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb431-13"><a href="biological-networks-..html#cb431-13" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="at">h =</span> <span class="fl">0.90</span>, <span class="at">col =</span> <span class="st">&quot;red&quot;</span>)</span>
<span id="cb431-14"><a href="biological-networks-..html#cb431-14" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(sft<span class="sc">$</span>fitIndices[, <span class="dv">1</span>],</span>
<span id="cb431-15"><a href="biological-networks-..html#cb431-15" aria-hidden="true" tabindex="-1"></a>  sft<span class="sc">$</span>fitIndices[, <span class="dv">5</span>],</span>
<span id="cb431-16"><a href="biological-networks-..html#cb431-16" aria-hidden="true" tabindex="-1"></a>  <span class="at">xlab =</span> <span class="st">&quot;Soft Threshold (power)&quot;</span>,</span>
<span id="cb431-17"><a href="biological-networks-..html#cb431-17" aria-hidden="true" tabindex="-1"></a>  <span class="at">ylab =</span> <span class="st">&quot;Mean Connectivity&quot;</span>,</span>
<span id="cb431-18"><a href="biological-networks-..html#cb431-18" aria-hidden="true" tabindex="-1"></a>  <span class="at">type =</span> <span class="st">&quot;n&quot;</span>,</span>
<span id="cb431-19"><a href="biological-networks-..html#cb431-19" aria-hidden="true" tabindex="-1"></a>  <span class="at">main =</span> <span class="fu">paste</span>(<span class="st">&quot;Mean connectivity&quot;</span>)</span>
<span id="cb431-20"><a href="biological-networks-..html#cb431-20" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb431-21"><a href="biological-networks-..html#cb431-21" aria-hidden="true" tabindex="-1"></a><span class="fu">text</span>(sft<span class="sc">$</span>fitIndices[, <span class="dv">1</span>],</span>
<span id="cb431-22"><a href="biological-networks-..html#cb431-22" aria-hidden="true" tabindex="-1"></a>  sft<span class="sc">$</span>fitIndices[, <span class="dv">5</span>],</span>
<span id="cb431-23"><a href="biological-networks-..html#cb431-23" aria-hidden="true" tabindex="-1"></a>  <span class="at">labels =</span> powers,</span>
<span id="cb431-24"><a href="biological-networks-..html#cb431-24" aria-hidden="true" tabindex="-1"></a>  <span class="at">cex =</span> cex1, <span class="at">col =</span> <span class="st">&quot;red&quot;</span></span>
<span id="cb431-25"><a href="biological-networks-..html#cb431-25" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div>
<p><img src="main_files/unnamed-chunk-139-1.png" width="672" /></p>
<p>Generally, we want to choose a power that give us the “Scale Free Topology Model Fit, signed R^2” around <strong>0.8</strong>. This toy dataset is too small and can’t show it properly.<br />
For illustration, we will pick power 4 in this example.</p>
<div class="sourceCode" id="cb432"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb432-1"><a href="biological-networks-..html#cb432-1" aria-hidden="true" tabindex="-1"></a>picked_power <span class="ot">&lt;-</span> <span class="dv">4</span></span>
<span id="cb432-2"><a href="biological-networks-..html#cb432-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb432-3"><a href="biological-networks-..html#cb432-3" aria-hidden="true" tabindex="-1"></a><span class="co"># fix a namespace conflict issue, force R to use the cor() function in WGCNA:</span></span>
<span id="cb432-4"><a href="biological-networks-..html#cb432-4" aria-hidden="true" tabindex="-1"></a>temp_cor <span class="ot">&lt;-</span> cor</span>
<span id="cb432-5"><a href="biological-networks-..html#cb432-5" aria-hidden="true" tabindex="-1"></a>cor <span class="ot">&lt;-</span> WGCNA<span class="sc">::</span>cor</span></code></pre></div>
<p>Construct network.<br />
There are a lot of parameters. In most cases, you don’t need to modify them. But, make sure you choose the proper <code>power</code>, <code>networkType</code> (“signed” or “unsigned”), and <code>minModuleSize</code> in your analysis.<br />
- <code>networkType</code> Using “signed” option, the direction of the correlation will be considered, e.g. two genes won’t be clustered together if their correlation value is a perfect -1. On the other hand, use “unsigned” option to consider the absolute value of correlation.<br />
- <code>minModuleSize</code> The minimum umber of features, e.g. genes, in a module. In this example with only 7 genes, so I set this argument to be 2. In real analysis, a typical choice may be 20 or 30 for a matrix with several hundreds or thousands of genes.</p>
<div class="sourceCode" id="cb433"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb433-1"><a href="biological-networks-..html#cb433-1" aria-hidden="true" tabindex="-1"></a>netwk <span class="ot">&lt;-</span> <span class="fu">blockwiseModules</span>(</span>
<span id="cb433-2"><a href="biological-networks-..html#cb433-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">datExpr =</span> dat,</span>
<span id="cb433-3"><a href="biological-networks-..html#cb433-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">power =</span> picked_power,</span>
<span id="cb433-4"><a href="biological-networks-..html#cb433-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">networkType =</span> <span class="st">&quot;signed&quot;</span>,</span>
<span id="cb433-5"><a href="biological-networks-..html#cb433-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">deepSplit =</span> <span class="dv">2</span>,</span>
<span id="cb433-6"><a href="biological-networks-..html#cb433-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">pamRespectsDendro =</span> F,</span>
<span id="cb433-7"><a href="biological-networks-..html#cb433-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">minModuleSize =</span> <span class="dv">2</span>,</span>
<span id="cb433-8"><a href="biological-networks-..html#cb433-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">maxBlockSize =</span> <span class="dv">4000</span>,</span>
<span id="cb433-9"><a href="biological-networks-..html#cb433-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">reassignThreshold =</span> <span class="dv">0</span>,</span>
<span id="cb433-10"><a href="biological-networks-..html#cb433-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">mergeCutHeight =</span> <span class="fl">0.25</span>,</span>
<span id="cb433-11"><a href="biological-networks-..html#cb433-11" aria-hidden="true" tabindex="-1"></a>  <span class="at">saveTOMs =</span> T, <span class="co"># Archive the run results in TOM file (saves time)</span></span>
<span id="cb433-12"><a href="biological-networks-..html#cb433-12" aria-hidden="true" tabindex="-1"></a>  <span class="at">saveTOMFileBase =</span> <span class="st">&quot;ER&quot;</span>,</span>
<span id="cb433-13"><a href="biological-networks-..html#cb433-13" aria-hidden="true" tabindex="-1"></a>  <span class="at">numericLabels =</span> T,</span>
<span id="cb433-14"><a href="biological-networks-..html#cb433-14" aria-hidden="true" tabindex="-1"></a>  <span class="at">verbose =</span> <span class="dv">3</span></span>
<span id="cb433-15"><a href="biological-networks-..html#cb433-15" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div>
<pre><code>##  Calculating module eigengenes block-wise from all genes
##    Flagging genes and samples with too many missing values...
##     ..step 1
##  ..Working on block 1 .
##     TOM calculation: adjacency..
##     ..will not use multithreading.
##      Fraction of slow calculations: 0.000000
##     ..connectivity..
##     ..matrix multiplication (system BLAS)..
##     ..normalization..
##     ..done.
##    ..saving TOM for block 1 into file ER-block.1.RData
##  ....clustering..
##  ....detecting modules..
##  ....calculating module eigengenes..
##  ....checking kME in modules..
##  ..merging modules that are too close..
##      mergeCloseModules: Merging modules whose distance is less than 0.25
##        Calculating new MEs...</code></pre>
<p>Now we can see the module membership assignment:</p>
<div class="sourceCode" id="cb435"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb435-1"><a href="biological-networks-..html#cb435-1" aria-hidden="true" tabindex="-1"></a>netwk<span class="sc">$</span>colors</span></code></pre></div>
<pre><code>##      CRAB     Whale     Lobst     Octop     Coral BabyShark      Orca 
##         1         2         1         1         1         2         2</code></pre>
<p>CRAB, Lobst, Octop and Coral are assigned to module 1. Whale, BabyShark and Orca are assigned to module 2.</p>
<p>We can use function labels2colors() to conveniently convert the module membership to color labels.</p>
<div class="sourceCode" id="cb437"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb437-1"><a href="biological-networks-..html#cb437-1" aria-hidden="true" tabindex="-1"></a>mergedColors <span class="ot">&lt;-</span> <span class="fu">labels2colors</span>(netwk<span class="sc">$</span>colors)</span></code></pre></div>
<p>Plot the dendrogram and the module colors:</p>
<div class="sourceCode" id="cb438"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb438-1"><a href="biological-networks-..html#cb438-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotDendroAndColors</span>(</span>
<span id="cb438-2"><a href="biological-networks-..html#cb438-2" aria-hidden="true" tabindex="-1"></a>  netwk<span class="sc">$</span>dendrograms[[<span class="dv">1</span>]],</span>
<span id="cb438-3"><a href="biological-networks-..html#cb438-3" aria-hidden="true" tabindex="-1"></a>  mergedColors[netwk<span class="sc">$</span>blockGenes[[<span class="dv">1</span>]]],</span>
<span id="cb438-4"><a href="biological-networks-..html#cb438-4" aria-hidden="true" tabindex="-1"></a>  <span class="st">&quot;Module colors&quot;</span>,</span>
<span id="cb438-5"><a href="biological-networks-..html#cb438-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">dendroLabels =</span> <span class="cn">FALSE</span>,</span>
<span id="cb438-6"><a href="biological-networks-..html#cb438-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">hang =</span> <span class="fl">0.03</span>,</span>
<span id="cb438-7"><a href="biological-networks-..html#cb438-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">addGuide =</span> <span class="cn">TRUE</span>,</span>
<span id="cb438-8"><a href="biological-networks-..html#cb438-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">guideHang =</span> <span class="fl">0.05</span></span>
<span id="cb438-9"><a href="biological-networks-..html#cb438-9" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div>
<p><img src="main_files/unnamed-chunk-144-1.png" width="672" /></p>
<p>In a real analysis with a lot of genes, you will see much more modules and colors.<br />
In this dendrogram, you will be able to roughly see the size of each module by looking at the area of different colors.</p>
<p>At this point, the WGCNA analysis is finished.<br />
We can summarize the module information in a data frame:</p>
<div class="sourceCode" id="cb439"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb439-1"><a href="biological-networks-..html#cb439-1" aria-hidden="true" tabindex="-1"></a>module_df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb439-2"><a href="biological-networks-..html#cb439-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">gene_id =</span> <span class="fu">names</span>(netwk<span class="sc">$</span>colors),</span>
<span id="cb439-3"><a href="biological-networks-..html#cb439-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">modules =</span> <span class="fu">labels2colors</span>(netwk<span class="sc">$</span>colors)</span>
<span id="cb439-4"><a href="biological-networks-..html#cb439-4" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb439-5"><a href="biological-networks-..html#cb439-5" aria-hidden="true" tabindex="-1"></a>module_df</span></code></pre></div>
<pre><code>##     gene_id   modules
## 1      CRAB turquoise
## 2     Whale      blue
## 3     Lobst turquoise
## 4     Octop turquoise
## 5     Coral turquoise
## 6 BabyShark      blue
## 7      Orca      blue</code></pre>

</div>
</div>
</div>
<!-- </div> -->
            </section>

          </div>
        </div>
      </div>
<a href="gene-set-enrichment-analysis.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="engineering.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
