<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>6 Data Science | BF591 - R for Biological Sciences</title>
  <meta name="description" content="6 Data Science | BF591 - R for Biological Sciences" />
  <meta name="generator" content="bookdown 0.24 and GitBook 2.6.7" />

  <meta property="og:title" content="6 Data Science | BF591 - R for Biological Sciences" />
  <meta property="og:type" content="book" />
  
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="6 Data Science | BF591 - R for Biological Sciences" />
  
  
  




  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="data-wrangling.html"/>
<link rel="next" href="data-visualization.html"/>
<script src="libs/header-attrs-2.11/header-attrs.js"></script>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.0.1/anchor-sections.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.0.1/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Syllabus</a>
<ul>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#course-schedule"><i class="fa fa-check"></i>Course Schedule</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#instructors"><i class="fa fa-check"></i>Instructors</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#office-hours"><i class="fa fa-check"></i>Office Hours</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#course-values-and-policies"><i class="fa fa-check"></i>Course Values and Policies</a></li>
</ul></li>
<li class="chapter" data-level="1" data-path="introduction.html"><a href="introduction.html"><i class="fa fa-check"></i><b>1</b> Introduction</a>
<ul>
<li class="chapter" data-level="1.1" data-path="introduction.html"><a href="introduction.html#who-this-book-is-for"><i class="fa fa-check"></i><b>1.1</b> Who This Book Is For</a></li>
<li class="chapter" data-level="1.2" data-path="introduction.html"><a href="introduction.html#a-note-about-reinventing-the-wheel"><i class="fa fa-check"></i><b>1.2</b> A Note About Reinventing the Wheel</a></li>
<li class="chapter" data-level="1.3" data-path="introduction.html"><a href="introduction.html#sources-and-references"><i class="fa fa-check"></i><b>1.3</b> Sources and References</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="data-bio.html"><a href="data-bio.html"><i class="fa fa-check"></i><b>2</b> Data in Biology</a>
<ul>
<li class="chapter" data-level="2.1" data-path="data-bio.html"><a href="data-bio.html#a-brief-history-of-data-in-molecular-biology"><i class="fa fa-check"></i><b>2.1</b> A Brief History of Data in Molecular Biology</a></li>
<li class="chapter" data-level="2.2" data-path="data-bio.html"><a href="data-bio.html#biology-as-a-mature-data-science"><i class="fa fa-check"></i><b>2.2</b> Biology as a Mature Data Science</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="preliminaries.html"><a href="preliminaries.html"><i class="fa fa-check"></i><b>3</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="3.1" data-path="preliminaries.html"><a href="preliminaries.html#prelim-r"><i class="fa fa-check"></i><b>3.1</b> The R Language</a></li>
<li class="chapter" data-level="3.2" data-path="preliminaries.html"><a href="preliminaries.html#prelim-rstudio"><i class="fa fa-check"></i><b>3.2</b> RStudio</a></li>
<li class="chapter" data-level="3.3" data-path="preliminaries.html"><a href="preliminaries.html#prog-the-r-script"><i class="fa fa-check"></i><b>3.3</b> The R Script</a></li>
<li class="chapter" data-level="3.4" data-path="preliminaries.html"><a href="preliminaries.html#prog-workflow"><i class="fa fa-check"></i><b>3.4</b> The Scripting Workflow</a></li>
<li class="chapter" data-level="3.5" data-path="preliminaries.html"><a href="preliminaries.html#prelim-git"><i class="fa fa-check"></i><b>3.5</b> git + github</a>
<ul>
<li class="chapter" data-level="3.5.1" data-path="preliminaries.html"><a href="preliminaries.html#motivation"><i class="fa fa-check"></i><b>3.5.1</b> Motivation</a></li>
<li class="chapter" data-level="3.5.2" data-path="preliminaries.html"><a href="preliminaries.html#git"><i class="fa fa-check"></i><b>3.5.2</b> git</a></li>
<li class="chapter" data-level="3.5.3" data-path="preliminaries.html"><a href="preliminaries.html#git-hosting-platforms-github"><i class="fa fa-check"></i><b>3.5.3</b> Git hosting platforms (GitHub)</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="prog-basics.html"><a href="prog-basics.html"><i class="fa fa-check"></i><b>4</b> R Programming</a>
<ul>
<li class="chapter" data-level="4.1" data-path="prog-basics.html"><a href="prog-basics.html#before-you-begin"><i class="fa fa-check"></i><b>4.1</b> Before you begin</a></li>
<li class="chapter" data-level="4.2" data-path="prog-basics.html"><a href="prog-basics.html#introduction-1"><i class="fa fa-check"></i><b>4.2</b> Introduction</a></li>
<li class="chapter" data-level="4.3" data-path="prog-basics.html"><a href="prog-basics.html#prog-r-syntax"><i class="fa fa-check"></i><b>4.3</b> R Syntax Basics</a></li>
<li class="chapter" data-level="4.4" data-path="prog-basics.html"><a href="prog-basics.html#prog-types"><i class="fa fa-check"></i><b>4.4</b> Basic Types of Values</a>
<ul>
<li class="chapter" data-level="4.4.1" data-path="prog-basics.html"><a href="prog-basics.html#factors"><i class="fa fa-check"></i><b>4.4.1</b> Factors</a></li>
</ul></li>
<li class="chapter" data-level="4.5" data-path="prog-basics.html"><a href="prog-basics.html#prog-struct"><i class="fa fa-check"></i><b>4.5</b> Data Structures</a>
<ul>
<li class="chapter" data-level="4.5.1" data-path="prog-basics.html"><a href="prog-basics.html#vectors"><i class="fa fa-check"></i><b>4.5.1</b> Vectors</a></li>
<li class="chapter" data-level="4.5.2" data-path="prog-basics.html"><a href="prog-basics.html#matrices"><i class="fa fa-check"></i><b>4.5.2</b> Matrices</a></li>
<li class="chapter" data-level="4.5.3" data-path="prog-basics.html"><a href="prog-basics.html#lists-and-data-frames"><i class="fa fa-check"></i><b>4.5.3</b> Lists and data frames</a></li>
</ul></li>
<li class="chapter" data-level="4.6" data-path="prog-basics.html"><a href="prog-basics.html#logical-tests-and-comparators"><i class="fa fa-check"></i><b>4.6</b> Logical Tests and Comparators</a></li>
<li class="chapter" data-level="4.7" data-path="prog-basics.html"><a href="prog-basics.html#functions"><i class="fa fa-check"></i><b>4.7</b> Functions</a>
<ul>
<li class="chapter" data-level="4.7.1" data-path="prog-basics.html"><a href="prog-basics.html#prog-dry"><i class="fa fa-check"></i><b>4.7.1</b> DRY: Don’t Repeat Yourself</a></li>
<li class="chapter" data-level="4.7.2" data-path="prog-basics.html"><a href="prog-basics.html#writing-your-own-functions"><i class="fa fa-check"></i><b>4.7.2</b> Writing your own functions</a></li>
<li class="chapter" data-level="4.7.3" data-path="prog-basics.html"><a href="prog-basics.html#scope"><i class="fa fa-check"></i><b>4.7.3</b> Scope</a></li>
</ul></li>
<li class="chapter" data-level="4.8" data-path="prog-basics.html"><a href="prog-basics.html#iteration"><i class="fa fa-check"></i><b>4.8</b> Iteration</a>
<ul>
<li class="chapter" data-level="4.8.1" data-path="prog-basics.html"><a href="prog-basics.html#vectorized-operations"><i class="fa fa-check"></i><b>4.8.1</b> Vectorized operations</a></li>
<li class="chapter" data-level="4.8.2" data-path="prog-basics.html"><a href="prog-basics.html#functional-programming"><i class="fa fa-check"></i><b>4.8.2</b> Functional programming</a></li>
<li class="chapter" data-level="4.8.3" data-path="prog-basics.html"><a href="prog-basics.html#apply-and-friends"><i class="fa fa-check"></i><b>4.8.3</b> <code>apply()</code> and friends</a></li>
</ul></li>
<li class="chapter" data-level="4.9" data-path="prog-basics.html"><a href="prog-basics.html#installing-packages"><i class="fa fa-check"></i><b>4.9</b> Installing Packages</a></li>
<li class="chapter" data-level="4.10" data-path="prog-basics.html"><a href="prog-basics.html#saving-and-loading-r-data"><i class="fa fa-check"></i><b>4.10</b> Saving and Loading R Data</a></li>
<li class="chapter" data-level="4.11" data-path="prog-basics.html"><a href="prog-basics.html#prog-debugging"><i class="fa fa-check"></i><b>4.11</b> Troubleshooting and Debugging</a></li>
<li class="chapter" data-level="4.12" data-path="prog-basics.html"><a href="prog-basics.html#prog-style"><i class="fa fa-check"></i><b>4.12</b> Coding Style and Conventions</a>
<ul>
<li class="chapter" data-level="4.12.1" data-path="prog-basics.html"><a href="prog-basics.html#is-my-code-correct"><i class="fa fa-check"></i><b>4.12.1</b> Is my code correct?</a></li>
<li class="chapter" data-level="4.12.2" data-path="prog-basics.html"><a href="prog-basics.html#does-my-code-follow-the-dry-principle"><i class="fa fa-check"></i><b>4.12.2</b> Does my code follow the <span>DRY principle</span>?</a></li>
<li class="chapter" data-level="4.12.3" data-path="prog-basics.html"><a href="prog-basics.html#did-i-choose-concise-but-descriptive-variable-and-function-names"><i class="fa fa-check"></i><b>4.12.3</b> Did I choose concise but descriptive variable and function names?</a></li>
<li class="chapter" data-level="4.12.4" data-path="prog-basics.html"><a href="prog-basics.html#did-i-use-indentation-and-naming-conventions-consistently-throughout-my-code"><i class="fa fa-check"></i><b>4.12.4</b> Did I use indentation and naming conventions consistently throughout my code?</a></li>
<li class="chapter" data-level="4.12.5" data-path="prog-basics.html"><a href="prog-basics.html#did-i-write-comments-especially-when-what-the-code-does-is-not-obvious"><i class="fa fa-check"></i><b>4.12.5</b> Did I write comments, especially when what the code does is not obvious?</a></li>
<li class="chapter" data-level="4.12.6" data-path="prog-basics.html"><a href="prog-basics.html#how-easy-would-it-be-for-someone-else-to-understand-my-code"><i class="fa fa-check"></i><b>4.12.6</b> How easy would it be for someone else to understand my code?</a></li>
<li class="chapter" data-level="4.12.7" data-path="prog-basics.html"><a href="prog-basics.html#is-my-code-easy-to-maintainchange"><i class="fa fa-check"></i><b>4.12.7</b> Is my code easy to maintain/change?</a></li>
<li class="chapter" data-level="4.12.8" data-path="prog-basics.html"><a href="prog-basics.html#the-styler-package-."><i class="fa fa-check"></i><b>4.12.8</b> The <code>styler</code> package .</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="5" data-path="data-wrangling.html"><a href="data-wrangling.html"><i class="fa fa-check"></i><b>5</b> Data Wrangling</a>
<ul>
<li class="chapter" data-level="5.1" data-path="data-wrangling.html"><a href="data-wrangling.html#the-tidyverse"><i class="fa fa-check"></i><b>5.1</b> The Tidyverse</a></li>
<li class="chapter" data-level="5.2" data-path="data-wrangling.html"><a href="data-wrangling.html#dw-basics"><i class="fa fa-check"></i><b>5.2</b> Tidyverse Basics</a></li>
<li class="chapter" data-level="5.3" data-path="data-wrangling.html"><a href="data-wrangling.html#importing-data"><i class="fa fa-check"></i><b>5.3</b> Importing Data</a></li>
<li class="chapter" data-level="5.4" data-path="data-wrangling.html"><a href="data-wrangling.html#data-tibble"><i class="fa fa-check"></i><b>5.4</b> The tibble</a></li>
<li class="chapter" data-level="5.5" data-path="data-wrangling.html"><a href="data-wrangling.html#tidy-data"><i class="fa fa-check"></i><b>5.5</b> Tidy Data</a></li>
<li class="chapter" data-level="5.6" data-path="data-wrangling.html"><a href="data-wrangling.html#pipes"><i class="fa fa-check"></i><b>5.6</b> pipes</a></li>
<li class="chapter" data-level="5.7" data-path="data-wrangling.html"><a href="data-wrangling.html#arranging-data"><i class="fa fa-check"></i><b>5.7</b> Arranging Data</a>
<ul>
<li class="chapter" data-level="5.7.1" data-path="data-wrangling.html"><a href="data-wrangling.html#dplyrmutate---create-new-columns-using-other-columns"><i class="fa fa-check"></i><b>5.7.1</b> <code>dplyr::mutate()</code> - Create new columns using other columns</a></li>
<li class="chapter" data-level="5.7.2" data-path="data-wrangling.html"><a href="data-wrangling.html#stringr---working-with-character-values"><i class="fa fa-check"></i><b>5.7.2</b> <code>stringr</code> - Working with character values</a></li>
<li class="chapter" data-level="5.7.3" data-path="data-wrangling.html"><a href="data-wrangling.html#dplyrselect---subset-columns-by-name"><i class="fa fa-check"></i><b>5.7.3</b> <code>dplyr::select()</code> - Subset Columns by Name</a></li>
<li class="chapter" data-level="5.7.4" data-path="data-wrangling.html"><a href="data-wrangling.html#dplyrfilter---pick-rows-out-of-a-data-set"><i class="fa fa-check"></i><b>5.7.4</b> <code>dplyr::filter()</code> - Pick rows out of a data set</a></li>
<li class="chapter" data-level="5.7.5" data-path="data-wrangling.html"><a href="data-wrangling.html#dplyrarrange---order-rows-based-on-their-values"><i class="fa fa-check"></i><b>5.7.5</b> <code>dplyr::arrange()</code> - Order rows based on their values</a></li>
<li class="chapter" data-level="5.7.6" data-path="data-wrangling.html"><a href="data-wrangling.html#putting-it-all-together"><i class="fa fa-check"></i><b>5.7.6</b> Putting it all together</a></li>
</ul></li>
<li class="chapter" data-level="5.8" data-path="data-wrangling.html"><a href="data-wrangling.html#grouping-data"><i class="fa fa-check"></i><b>5.8</b> Grouping Data</a></li>
<li class="chapter" data-level="5.9" data-path="data-wrangling.html"><a href="data-wrangling.html#rearranging-data"><i class="fa fa-check"></i><b>5.9</b> Rearranging Data</a></li>
<li class="chapter" data-level="5.10" data-path="data-wrangling.html"><a href="data-wrangling.html#relational-data"><i class="fa fa-check"></i><b>5.10</b> Relational Data</a>
<ul>
<li class="chapter" data-level="5.10.1" data-path="data-wrangling.html"><a href="data-wrangling.html#dealing-with-multiple-matches"><i class="fa fa-check"></i><b>5.10.1</b> Dealing with multiple matches</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="6" data-path="data-science.html"><a href="data-science.html"><i class="fa fa-check"></i><b>6</b> Data Science</a>
<ul>
<li class="chapter" data-level="6.1" data-path="data-science.html"><a href="data-science.html#data-modeling"><i class="fa fa-check"></i><b>6.1</b> Data Modeling</a>
<ul>
<li class="chapter" data-level="6.1.1" data-path="data-science.html"><a href="data-science.html#a-worked-modeling-example"><i class="fa fa-check"></i><b>6.1.1</b> A Worked Modeling Example</a></li>
<li class="chapter" data-level="6.1.2" data-path="data-science.html"><a href="data-science.html#data-summarization"><i class="fa fa-check"></i><b>6.1.2</b> Data Summarization</a></li>
<li class="chapter" data-level="6.1.3" data-path="data-science.html"><a href="data-science.html#linear-models"><i class="fa fa-check"></i><b>6.1.3</b> Linear Models</a></li>
<li class="chapter" data-level="6.1.4" data-path="data-science.html"><a href="data-science.html#flavors-of-linear-models"><i class="fa fa-check"></i><b>6.1.4</b> Flavors of Linear Models</a></li>
<li class="chapter" data-level="6.1.5" data-path="data-science.html"><a href="data-science.html#assessing-model-accuracy-."><i class="fa fa-check"></i><b>6.1.5</b> Assessing Model Accuracy .</a></li>
</ul></li>
<li class="chapter" data-level="6.2" data-path="data-science.html"><a href="data-science.html#statistical-distributions-tests-."><i class="fa fa-check"></i><b>6.2</b> Statistical Distributions &amp; Tests .</a>
<ul>
<li class="chapter" data-level="6.2.1" data-path="data-science.html"><a href="data-science.html#statistical-distributions"><i class="fa fa-check"></i><b>6.2.1</b> Statistical Distributions</a></li>
<li class="chapter" data-level="6.2.2" data-path="data-science.html"><a href="data-science.html#statistical-tests-."><i class="fa fa-check"></i><b>6.2.2</b> Statistical Tests .</a></li>
<li class="chapter" data-level="6.2.3" data-path="data-science.html"><a href="data-science.html#p-values-."><i class="fa fa-check"></i><b>6.2.3</b> p-values .</a></li>
<li class="chapter" data-level="6.2.4" data-path="data-science.html"><a href="data-science.html#multiple-hypothesis-testing-."><i class="fa fa-check"></i><b>6.2.4</b> Multiple Hypothesis Testing .</a></li>
<li class="chapter" data-level="6.2.5" data-path="data-science.html"><a href="data-science.html#statistical-power-."><i class="fa fa-check"></i><b>6.2.5</b> Statistical power .</a></li>
</ul></li>
<li class="chapter" data-level="6.3" data-path="data-science.html"><a href="data-science.html#exploratory-analysis"><i class="fa fa-check"></i><b>6.3</b> Exploratory Analysis</a>
<ul>
<li class="chapter" data-level="6.3.1" data-path="data-science.html"><a href="data-science.html#principal-component-analysis"><i class="fa fa-check"></i><b>6.3.1</b> Principal Component Analysis</a></li>
<li class="chapter" data-level="6.3.2" data-path="data-science.html"><a href="data-science.html#cluster-analysis"><i class="fa fa-check"></i><b>6.3.2</b> Cluster Analysis</a></li>
<li class="chapter" data-level="6.3.3" data-path="data-science.html"><a href="data-science.html#k-means-."><i class="fa fa-check"></i><b>6.3.3</b> k-means .</a></li>
<li class="chapter" data-level="6.3.4" data-path="data-science.html"><a href="data-science.html#others"><i class="fa fa-check"></i><b>6.3.4</b> Others</a></li>
</ul></li>
<li class="chapter" data-level="6.4" data-path="data-science.html"><a href="data-science.html#network-analysis-."><i class="fa fa-check"></i><b>6.4</b> Network Analysis .</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="data-visualization.html"><a href="data-visualization.html"><i class="fa fa-check"></i><b>7</b> Data Visualization</a>
<ul>
<li class="chapter" data-level="7.1" data-path="data-visualization.html"><a href="data-visualization.html#dv-gg"><i class="fa fa-check"></i><b>7.1</b> Grammar of Graphics</a></li>
<li class="chapter" data-level="7.2" data-path="data-visualization.html"><a href="data-visualization.html#plotting-one-dimension"><i class="fa fa-check"></i><b>7.2</b> Plotting One Dimension</a>
<ul>
<li class="chapter" data-level="7.2.1" data-path="data-visualization.html"><a href="data-visualization.html#dv-bar"><i class="fa fa-check"></i><b>7.2.1</b> Bar chart</a></li>
<li class="chapter" data-level="7.2.2" data-path="data-visualization.html"><a href="data-visualization.html#lollipop-plots"><i class="fa fa-check"></i><b>7.2.2</b> Lollipop plots</a></li>
<li class="chapter" data-level="7.2.3" data-path="data-visualization.html"><a href="data-visualization.html#stacked-area-charts"><i class="fa fa-check"></i><b>7.2.3</b> Stacked Area charts</a></li>
<li class="chapter" data-level="7.2.4" data-path="data-visualization.html"><a href="data-visualization.html#parallel-coordinate-plots"><i class="fa fa-check"></i><b>7.2.4</b> Parallel Coordinate plots</a></li>
</ul></li>
<li class="chapter" data-level="7.3" data-path="data-visualization.html"><a href="data-visualization.html#visualizing-distributions"><i class="fa fa-check"></i><b>7.3</b> Visualizing Distributions</a>
<ul>
<li class="chapter" data-level="7.3.1" data-path="data-visualization.html"><a href="data-visualization.html#histogram"><i class="fa fa-check"></i><b>7.3.1</b> Histogram</a></li>
<li class="chapter" data-level="7.3.2" data-path="data-visualization.html"><a href="data-visualization.html#density"><i class="fa fa-check"></i><b>7.3.2</b> Density</a></li>
<li class="chapter" data-level="7.3.3" data-path="data-visualization.html"><a href="data-visualization.html#boxplot"><i class="fa fa-check"></i><b>7.3.3</b> Boxplot</a></li>
<li class="chapter" data-level="7.3.4" data-path="data-visualization.html"><a href="data-visualization.html#violin-plot"><i class="fa fa-check"></i><b>7.3.4</b> Violin plot</a></li>
<li class="chapter" data-level="7.3.5" data-path="data-visualization.html"><a href="data-visualization.html#beeswarm-plot"><i class="fa fa-check"></i><b>7.3.5</b> Beeswarm plot</a></li>
<li class="chapter" data-level="7.3.6" data-path="data-visualization.html"><a href="data-visualization.html#ridgeline"><i class="fa fa-check"></i><b>7.3.6</b> Ridgeline</a></li>
</ul></li>
<li class="chapter" data-level="7.4" data-path="data-visualization.html"><a href="data-visualization.html#plotting-two-or-more-dimensions-."><i class="fa fa-check"></i><b>7.4</b> Plotting Two or More Dimensions .</a>
<ul>
<li class="chapter" data-level="7.4.1" data-path="data-visualization.html"><a href="data-visualization.html#scatter-plot-."><i class="fa fa-check"></i><b>7.4.1</b> Scatter plot .</a></li>
<li class="chapter" data-level="7.4.2" data-path="data-visualization.html"><a href="data-visualization.html#line-plot-."><i class="fa fa-check"></i><b>7.4.2</b> Line plot .</a></li>
<li class="chapter" data-level="7.4.3" data-path="data-visualization.html"><a href="data-visualization.html#bubble-plot-."><i class="fa fa-check"></i><b>7.4.3</b> Bubble plot .</a></li>
<li class="chapter" data-level="7.4.4" data-path="data-visualization.html"><a href="data-visualization.html#heatmaps"><i class="fa fa-check"></i><b>7.4.4</b> Heatmaps</a></li>
</ul></li>
<li class="chapter" data-level="7.5" data-path="data-visualization.html"><a href="data-visualization.html#other-kind-of-plots-."><i class="fa fa-check"></i><b>7.5</b> Other Kind of Plots .</a>
<ul>
<li class="chapter" data-level="7.5.1" data-path="data-visualization.html"><a href="data-visualization.html#chord-."><i class="fa fa-check"></i><b>7.5.1</b> Chord .</a></li>
<li class="chapter" data-level="7.5.2" data-path="data-visualization.html"><a href="data-visualization.html#sankey-."><i class="fa fa-check"></i><b>7.5.2</b> Sankey .</a></li>
<li class="chapter" data-level="7.5.3" data-path="data-visualization.html"><a href="data-visualization.html#dendrograms"><i class="fa fa-check"></i><b>7.5.3</b> Dendrograms</a></li>
</ul></li>
<li class="chapter" data-level="7.6" data-path="data-visualization.html"><a href="data-visualization.html#tips-and-tricks-."><i class="fa fa-check"></i><b>7.6</b> Tips and Tricks .</a>
<ul>
<li class="chapter" data-level="7.6.1" data-path="data-visualization.html"><a href="data-visualization.html#reordering-1-d-data-elements-."><i class="fa fa-check"></i><b>7.6.1</b> Reordering 1-D Data Elements .</a></li>
<li class="chapter" data-level="7.6.2" data-path="data-visualization.html"><a href="data-visualization.html#confidence-intervals-."><i class="fa fa-check"></i><b>7.6.2</b> Confidence Intervals .</a></li>
<li class="chapter" data-level="7.6.3" data-path="data-visualization.html"><a href="data-visualization.html#annotations-."><i class="fa fa-check"></i><b>7.6.3</b> Annotations .</a></li>
</ul></li>
<li class="chapter" data-level="7.7" data-path="data-visualization.html"><a href="data-visualization.html#multiple-plots-."><i class="fa fa-check"></i><b>7.7</b> Multiple Plots .</a>
<ul>
<li class="chapter" data-level="7.7.1" data-path="data-visualization.html"><a href="data-visualization.html#secondary-axes-."><i class="fa fa-check"></i><b>7.7.1</b> Secondary Axes .</a></li>
<li class="chapter" data-level="7.7.2" data-path="data-visualization.html"><a href="data-visualization.html#facet-wrapping-."><i class="fa fa-check"></i><b>7.7.2</b> Facet wrapping .</a></li>
<li class="chapter" data-level="7.7.3" data-path="data-visualization.html"><a href="data-visualization.html#multipanel-figures-."><i class="fa fa-check"></i><b>7.7.3</b> Multipanel Figures .</a></li>
</ul></li>
<li class="chapter" data-level="7.8" data-path="data-visualization.html"><a href="data-visualization.html#responsible-plotting-."><i class="fa fa-check"></i><b>7.8</b> Responsible Plotting .</a></li>
<li class="chapter" data-level="7.9" data-path="data-visualization.html"><a href="data-visualization.html#publication-ready-plots-."><i class="fa fa-check"></i><b>7.9</b> Publication Ready Plots .</a></li>
<li class="chapter" data-level="7.10" data-path="data-visualization.html"><a href="data-visualization.html#network-visualization-."><i class="fa fa-check"></i><b>7.10</b> Network visualization .</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html"><i class="fa fa-check"></i><b>8</b> Biology &amp; Bioinformatics</a>
<ul>
<li class="chapter" data-level="8.1" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#bio-r"><i class="fa fa-check"></i><b>8.1</b> R in Biology</a></li>
<li class="chapter" data-level="8.2" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#biological-data-overview"><i class="fa fa-check"></i><b>8.2</b> Biological Data Overview</a>
<ul>
<li class="chapter" data-level="8.2.1" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#csv-files"><i class="fa fa-check"></i><b>8.2.1</b> CSV Files</a></li>
<li class="chapter" data-level="8.2.2" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#bio-data-types"><i class="fa fa-check"></i><b>8.2.2</b> Types of Biological Data</a></li>
<li class="chapter" data-level="8.2.3" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#common-biological-data-matrices"><i class="fa fa-check"></i><b>8.2.3</b> Common Biological Data Matrices</a></li>
<li class="chapter" data-level="8.2.4" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#biological-data-is-not-tidy"><i class="fa fa-check"></i><b>8.2.4</b> Biological data is NOT Tidy!</a></li>
</ul></li>
<li class="chapter" data-level="8.3" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#bio-bioconductor"><i class="fa fa-check"></i><b>8.3</b> Bioconductor</a></li>
<li class="chapter" data-level="8.4" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#gene-identifiers"><i class="fa fa-check"></i><b>8.4</b> Gene Identifiers</a>
<ul>
<li class="chapter" data-level="8.4.1" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#gene-identifier-systems"><i class="fa fa-check"></i><b>8.4.1</b> Gene Identifier Systems</a></li>
<li class="chapter" data-level="8.4.2" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#mapping-between-identifier-systems"><i class="fa fa-check"></i><b>8.4.2</b> Mapping Between Identifier Systems</a></li>
<li class="chapter" data-level="8.4.3" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#mapping-homologs"><i class="fa fa-check"></i><b>8.4.3</b> Mapping Homologs</a></li>
</ul></li>
<li class="chapter" data-level="8.5" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#gene-expression"><i class="fa fa-check"></i><b>8.5</b> Gene Expression</a>
<ul>
<li class="chapter" data-level="8.5.1" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#gene-expression-data-in-bioconductor"><i class="fa fa-check"></i><b>8.5.1</b> Gene Expression Data in Bioconductor</a></li>
<li class="chapter" data-level="8.5.2" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#microarrays"><i class="fa fa-check"></i><b>8.5.2</b> Microarrays</a></li>
<li class="chapter" data-level="8.5.3" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#microarray-data-in-bioconductor"><i class="fa fa-check"></i><b>8.5.3</b> Microarray Data in Bioconductor</a></li>
<li class="chapter" data-level="8.5.4" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#differential-expression-analysis"><i class="fa fa-check"></i><b>8.5.4</b> Differential Expression Analysis</a></li>
<li class="chapter" data-level="8.5.5" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#differential-expression-microarrays-limma"><i class="fa fa-check"></i><b>8.5.5</b> Differential Expression: Microarrays (limma)</a></li>
<li class="chapter" data-level="8.5.6" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#high-throughput-sequencing-."><i class="fa fa-check"></i><b>8.5.6</b> High Throughput Sequencing .</a></li>
<li class="chapter" data-level="8.5.7" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#count-data-."><i class="fa fa-check"></i><b>8.5.7</b> Count Data .</a></li>
<li class="chapter" data-level="8.5.8" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#rnaseq-."><i class="fa fa-check"></i><b>8.5.8</b> RNASeq .</a></li>
<li class="chapter" data-level="8.5.9" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#differential-expression-rnaseq-."><i class="fa fa-check"></i><b>8.5.9</b> Differential Expression: RNASeq .</a></li>
</ul></li>
<li class="chapter" data-level="8.6" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#gene-set-enrichment-analysis"><i class="fa fa-check"></i><b>8.6</b> Gene Set Enrichment Analysis</a>
<ul>
<li class="chapter" data-level="8.6.1" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#gene-sets"><i class="fa fa-check"></i><b>8.6.1</b> Gene Sets</a></li>
<li class="chapter" data-level="8.6.2" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#working-with-gene-sets-in-r"><i class="fa fa-check"></i><b>8.6.2</b> Working with gene sets in R</a></li>
<li class="chapter" data-level="8.6.3" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#over-representation-analysis"><i class="fa fa-check"></i><b>8.6.3</b> Over-representation Analysis</a></li>
<li class="chapter" data-level="8.6.4" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#rank-based-analysis-."><i class="fa fa-check"></i><b>8.6.4</b> Rank-based Analysis .</a></li>
</ul></li>
<li class="chapter" data-level="8.7" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#biological-networks-."><i class="fa fa-check"></i><b>8.7</b> Biological Networks .</a>
<ul>
<li class="chapter" data-level="8.7.1" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#biological-pathways-."><i class="fa fa-check"></i><b>8.7.1</b> Biological Pathways .</a></li>
<li class="chapter" data-level="8.7.2" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#gene-regulatory-networks-."><i class="fa fa-check"></i><b>8.7.2</b> Gene Regulatory Networks .</a></li>
<li class="chapter" data-level="8.7.3" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#protein-protein-interaction-network-."><i class="fa fa-check"></i><b>8.7.3</b> Protein-Protein Interaction Network .</a></li>
<li class="chapter" data-level="8.7.4" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#wgcna-."><i class="fa fa-check"></i><b>8.7.4</b> WGCNA .</a></li>
</ul></li>
<li class="chapter" data-level="8.8" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#single-cell-sequencing-analysis-."><i class="fa fa-check"></i><b>8.8</b> Single Cell Sequencing Analysis .</a>
<ul>
<li class="chapter" data-level="8.8.1" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#single-cell-sequencing-."><i class="fa fa-check"></i><b>8.8.1</b> Single Cell Sequencing .</a></li>
<li class="chapter" data-level="8.8.2" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#single-cell-analysis-."><i class="fa fa-check"></i><b>8.8.2</b> Single Cell Analysis .</a></li>
<li class="chapter" data-level="8.8.3" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#single-cell-clustering-."><i class="fa fa-check"></i><b>8.8.3</b> Single Cell Clustering .</a></li>
<li class="chapter" data-level="8.8.4" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#dimensionality-reduction-projection-."><i class="fa fa-check"></i><b>8.8.4</b> Dimensionality Reduction &amp; Projection .</a></li>
<li class="chapter" data-level="8.8.5" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#single-cell-data-visualization-."><i class="fa fa-check"></i><b>8.8.5</b> Single Cell Data Visualization .</a></li>
<li class="chapter" data-level="8.8.6" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#single-cell-marker-analysis-."><i class="fa fa-check"></i><b>8.8.6</b> Single Cell Marker Analysis .</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="9" data-path="engineering.html"><a href="engineering.html"><i class="fa fa-check"></i><b>9</b> EngineeRing</a>
<ul>
<li class="chapter" data-level="9.1" data-path="engineering.html"><a href="engineering.html#unit-testing"><i class="fa fa-check"></i><b>9.1</b> Unit Testing</a></li>
<li class="chapter" data-level="9.2" data-path="engineering.html"><a href="engineering.html#object-orientation-in-r-."><i class="fa fa-check"></i><b>9.2</b> Object Orientation in R .</a></li>
<li class="chapter" data-level="9.3" data-path="engineering.html"><a href="engineering.html#toolification-."><i class="fa fa-check"></i><b>9.3</b> Toolification .</a></li>
<li class="chapter" data-level="9.4" data-path="engineering.html"><a href="engineering.html#pipelines-workflows-."><i class="fa fa-check"></i><b>9.4</b> Pipelines &amp; Workflows .</a></li>
<li class="chapter" data-level="9.5" data-path="engineering.html"><a href="engineering.html#parallel-processing-."><i class="fa fa-check"></i><b>9.5</b> Parallel Processing .</a></li>
<li class="chapter" data-level="9.6" data-path="engineering.html"><a href="engineering.html#r-packages-."><i class="fa fa-check"></i><b>9.6</b> R Packages .</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="rshiny.html"><a href="rshiny.html"><i class="fa fa-check"></i><b>10</b> RShiny</a>
<ul>
<li class="chapter" data-level="10.1" data-path="rshiny.html"><a href="rshiny.html#overview"><i class="fa fa-check"></i><b>10.1</b> Overview</a></li>
<li class="chapter" data-level="10.2" data-path="rshiny.html"><a href="rshiny.html#learning-objectives"><i class="fa fa-check"></i><b>10.2</b> Learning Objectives</a></li>
<li class="chapter" data-level="10.3" data-path="rshiny.html"><a href="rshiny.html#skill-list"><i class="fa fa-check"></i><b>10.3</b> Skill List</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="communicating-with-r.html"><a href="communicating-with-r.html"><i class="fa fa-check"></i><b>11</b> Communicating with R</a>
<ul>
<li class="chapter" data-level="11.1" data-path="communicating-with-r.html"><a href="communicating-with-r.html#rmarkdown-knitr"><i class="fa fa-check"></i><b>11.1</b> RMarkdown &amp; knitr</a></li>
<li class="chapter" data-level="11.2" data-path="communicating-with-r.html"><a href="communicating-with-r.html#bookdown"><i class="fa fa-check"></i><b>11.2</b> bookdown</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="contribution-guide.html"><a href="contribution-guide.html"><i class="fa fa-check"></i><b>12</b> Contribution Guide</a>
<ul>
<li class="chapter" data-level="12.1" data-path="contribution-guide.html"><a href="contribution-guide.html#custom-blocks"><i class="fa fa-check"></i><b>12.1</b> Custom blocks</a></li>
</ul></li>
<li class="part"><span><b>I Assignments</b></span></li>
<li class="chapter" data-level="" data-path="assignment-format.html"><a href="assignment-format.html"><i class="fa fa-check"></i>Assignment Format</a></li>
<li class="chapter" data-level="" data-path="starting-an-assignment.html"><a href="starting-an-assignment.html"><i class="fa fa-check"></i>Starting an Assignment</a>
<ul>
<li><a href="starting-an-assignment.html#starting-your-own-git-repository-and-cloning-it">Starting your own <code>git</code> repository and cloning it</a>
<ul>
<li class="chapter" data-level="" data-path="starting-an-assignment.html"><a href="starting-an-assignment.html#contents"><i class="fa fa-check"></i>Contents</a></li>
<li class="chapter" data-level="" data-path="starting-an-assignment.html"><a href="starting-an-assignment.html#github-tutorial"><i class="fa fa-check"></i>GitHub tutorial</a></li>
<li class="chapter" data-level="" data-path="starting-an-assignment.html"><a href="starting-an-assignment.html#committing-and-pushing-with-r-and-without"><i class="fa fa-check"></i>Committing and Pushing with R (and without)</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="assignment-1.html"><a href="assignment-1.html"><i class="fa fa-check"></i>Assignment 1</a>
<ul>
<li class="chapter" data-level="" data-path="assignment-1.html"><a href="assignment-1.html#problem-statement"><i class="fa fa-check"></i>Problem Statement</a></li>
<li class="chapter" data-level="" data-path="assignment-1.html"><a href="assignment-1.html#learning-objectives-and-skill-list"><i class="fa fa-check"></i>Learning Objectives and Skill List</a></li>
<li class="chapter" data-level="" data-path="assignment-1.html"><a href="assignment-1.html#instructions"><i class="fa fa-check"></i>Instructions</a></li>
<li class="chapter" data-level="" data-path="assignment-1.html"><a href="assignment-1.html#hints"><i class="fa fa-check"></i>Hints</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="assignment-2.html"><a href="assignment-2.html"><i class="fa fa-check"></i>Assignment 2</a>
<ul>
<li class="chapter" data-level="" data-path="assignment-2.html"><a href="assignment-2.html#problem-statement-1"><i class="fa fa-check"></i>Problem Statement</a></li>
<li class="chapter" data-level="" data-path="assignment-2.html"><a href="assignment-2.html#learning-objectives-1"><i class="fa fa-check"></i>Learning Objectives</a></li>
<li class="chapter" data-level="" data-path="assignment-2.html"><a href="assignment-2.html#skill-list-1"><i class="fa fa-check"></i>Skill List</a></li>
<li class="chapter" data-level="" data-path="assignment-2.html"><a href="assignment-2.html#instructions-1"><i class="fa fa-check"></i>Instructions</a></li>
<li class="chapter" data-level="" data-path="assignment-2.html"><a href="assignment-2.html#function-details"><i class="fa fa-check"></i>Function Details</a>
<ul>
<li class="chapter" data-level="" data-path="assignment-2.html"><a href="assignment-2.html#bioconductor"><i class="fa fa-check"></i>1. Bioconductor</a></li>
<li><a href="assignment-2.html#load_expression">2. <code>load_expression()</code></a></li>
<li><a href="assignment-2.html#filter_15">3. <code>filter_15()</code></a></li>
<li><a href="assignment-2.html#affy_to_hgnc">4. <code>affy_to_hgnc()</code></a></li>
<li><a href="assignment-2.html#reduce_data">5. <code>reduce_data()</code></a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="assignment-3.html"><a href="assignment-3.html"><i class="fa fa-check"></i>Assignment 3</a>
<ul>
<li class="chapter" data-level="" data-path="assignment-3.html"><a href="assignment-3.html#problem-statement-2"><i class="fa fa-check"></i>Problem Statement</a></li>
<li class="chapter" data-level="" data-path="assignment-3.html"><a href="assignment-3.html#learning-objectives-2"><i class="fa fa-check"></i>Learning Objectives</a></li>
<li class="chapter" data-level="" data-path="assignment-3.html"><a href="assignment-3.html#skill-list-2"><i class="fa fa-check"></i>Skill List</a></li>
<li class="chapter" data-level="" data-path="assignment-3.html"><a href="assignment-3.html#background-on-microarrays"><i class="fa fa-check"></i>Background on Microarrays</a></li>
<li class="chapter" data-level="" data-path="assignment-3.html"><a href="assignment-3.html#background-on-principal-component-analysis"><i class="fa fa-check"></i>Background on Principal Component Analysis</a></li>
<li class="chapter" data-level="" data-path="assignment-3.html"><a href="assignment-3.html#marisa-et-al.-gene-expression-classification-of-colon-cancer-into-molecular-subtypes-characterization-validation-and-prognostic-value.-plos-medicine-may-2013.-pmid-23700391"><i class="fa fa-check"></i>Marisa et al. Gene Expression Classification of Colon Cancer into Molecular Subtypes: Characterization, Validation, and Prognostic Value. PLoS Medicine, May 2013. PMID: 23700391</a></li>
<li><a href="assignment-3.html#scaling-data-using-r-scale">Scaling data using R <code>scale()</code></a></li>
<li class="chapter" data-level="" data-path="assignment-3.html"><a href="assignment-3.html#proportion-of-variance-explained"><i class="fa fa-check"></i>Proportion of variance explained</a></li>
<li class="chapter" data-level="" data-path="assignment-3.html"><a href="assignment-3.html#plotting-and-visualization-of-pca"><i class="fa fa-check"></i>Plotting and visualization of PCA</a></li>
<li class="chapter" data-level="" data-path="assignment-3.html"><a href="assignment-3.html#hierarchical-clustering-and-heatmaps"><i class="fa fa-check"></i>Hierarchical Clustering and Heatmaps</a></li>
<li class="chapter" data-level="" data-path="assignment-3.html"><a href="assignment-3.html#references"><i class="fa fa-check"></i>References</a></li>
</ul></li>
<li class="appendix"><span><b>Appendix</b></span></li>
<li class="chapter" data-level="A" data-path="class-outlines.html"><a href="class-outlines.html"><i class="fa fa-check"></i><b>A</b> Class Outlines</a>
<ul>
<li class="chapter" data-level="A.1" data-path="class-outlines.html"><a href="class-outlines.html#week-1"><i class="fa fa-check"></i><b>A.1</b> Week 1</a></li>
<li class="chapter" data-level="A.2" data-path="class-outlines.html"><a href="class-outlines.html#week-2"><i class="fa fa-check"></i><b>A.2</b> Week 2</a></li>
<li class="chapter" data-level="A.3" data-path="class-outlines.html"><a href="class-outlines.html#week-3"><i class="fa fa-check"></i><b>A.3</b> Week 3</a></li>
<li class="chapter" data-level="A.4" data-path="class-outlines.html"><a href="class-outlines.html#week-4"><i class="fa fa-check"></i><b>A.4</b> Week 4</a></li>
<li class="chapter" data-level="A.5" data-path="class-outlines.html"><a href="class-outlines.html#week-5"><i class="fa fa-check"></i><b>A.5</b> Week 5</a></li>
<li class="chapter" data-level="A.6" data-path="class-outlines.html"><a href="class-outlines.html#week-6"><i class="fa fa-check"></i><b>A.6</b> Week 6</a></li>
<li class="chapter" data-level="A.7" data-path="class-outlines.html"><a href="class-outlines.html#week-7"><i class="fa fa-check"></i><b>A.7</b> Week 7</a></li>
<li class="chapter" data-level="A.8" data-path="class-outlines.html"><a href="class-outlines.html#week-8"><i class="fa fa-check"></i><b>A.8</b> Week 8</a></li>
<li class="chapter" data-level="A.9" data-path="class-outlines.html"><a href="class-outlines.html#week-9"><i class="fa fa-check"></i><b>A.9</b> Week 9</a></li>
<li class="chapter" data-level="A.10" data-path="class-outlines.html"><a href="class-outlines.html#week-10"><i class="fa fa-check"></i><b>A.10</b> Week 10</a></li>
<li class="chapter" data-level="A.11" data-path="class-outlines.html"><a href="class-outlines.html#week-11"><i class="fa fa-check"></i><b>A.11</b> Week 11</a></li>
<li class="chapter" data-level="A.12" data-path="class-outlines.html"><a href="class-outlines.html#week-12"><i class="fa fa-check"></i><b>A.12</b> Week 12</a></li>
<li class="chapter" data-level="A.13" data-path="class-outlines.html"><a href="class-outlines.html#week-13---no-class-for-project-work"><i class="fa fa-check"></i><b>A.13</b> Week 13 - no class for project work</a></li>
<li class="chapter" data-level="A.14" data-path="class-outlines.html"><a href="class-outlines.html#week-14---no-class-for-project-work"><i class="fa fa-check"></i><b>A.14</b> Week 14 - no class for project work</a></li>
</ul></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">BF591 - R for Biological Sciences</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="data-science" class="section level1" number="6">
<h1><span class="header-section-number">6</span> Data Science</h1>
<p><a href="https://en.wikipedia.org/wiki/Data_science">Data science</a> is an enormous and
rapidly growing field that incorporates elements of statistics, computer
science, software engineering, high performance and cloud computing, and big
data management, as well as syntheses with knowledge from other social and
physical science fields to model and predict phenomena captured by data
collected from the “real world.” Many books have and will be written on this
subject, and this one does not pretend to even attempt to give this area an
adequate treatment. Like the rest of this book, the topics covered here are
opinionated and presented through the lens of a biological analyst practitioner
with enough high level conceptual details and a few general principles to
hopefully be useful in that context.</p>

<div id="data-modeling" class="section level2" number="6.1">
<h2><span class="header-section-number">6.1</span> Data Modeling</h2>
<p>The goal of data modeling is to describe a dataset using a relatively small
number of mathematical relationships. Said differently, a model uses some parts
of a dataset to try to accurately predict other parts of the dataset in a way
that is useful to us.</p>
<p>Models are human inventions; they reflect our beliefs about the way the universe
works. The successful model identifies patterns within a dataset that are the
result of causal relationships in the universe that led to the phenomena that
were measured while accounting for noise in the data. However, <em>the model itself
does not identify or even accurately reflect those causal effects</em>. The model
merely summarizes patterns and we as scientists are left to interpret those
patterns and design follow up experiments to investigate the nature of those
causal relationships using our prior knowledge of the world.</p>
<p>There are several principles to keep in mind when modeling data:</p>
<ol style="list-style-type: decimal">
<li><p><strong>Data are never wrong.</strong> All data are collected using processes and
devices designed and implemented by humans, who always have biases and make
assumptions. All data measure something about the universe, and so are “true” in
some sense of the word. If what we intended to measure and what we actually
measured were not the same thing, that is due to our errors in collection or
interpretation, not due to the data being wrong. If we approach our dataset with
a particular set of hypotheses and the data don’t support those hypotheses, it
is our beliefs of the world and our understanding of the dataset that are wrong,
not the data itself.</p></li>
<li><p><strong>Not all data are useful.</strong> Just because data isn’t wrong, it doesn’t mean
it is useful. There may have been systematic errors in the collection of the
data that makes interpreting them difficult. Data collected for one purpose may
not be useful for any other purposes. And sometimes, a dataset collected for a
particular purpose may simply not have the information needed to answer our
questions; if what we measure has no relationship to what we wish to predict,
the data itself is not useful - though the knowledge that what we measured has
no detectable effect on the thing we wish to predict may be very useful!</p></li>
<li><p><strong>“All models are wrong, but some are useful.”</strong> George Box, the renowned
British statistician, famously asserted this in a 1976 paper to the Journal of
the American Statistical Association. <span class="citation">(<a href="#ref-Box1976-rd" role="doc-biblioref">Box 1976</a>)</span>. By this he meant that every
model we create is a simplification of the system we are seeking to model, which
is by definition not identical to that system. To perfectly model a system, our
model would need to be precisely the system we are modeling, which is no longer
a model but the system itself. Fortunately, even though we know our models are
always wrong to some degree, they may nonetheless be useful because they are not
<strong>too</strong> wrong. Some models may indeed be too wrong, though.</p></li>
<li><p><strong>Data do not contain causal information.</strong> “Correlation does not mean
causation.” Data are measurements of the results of a process in the universe
that we wish to understand; the data are possibly reflective of that process,
but do not contain any information about the process itself. We cannot infer
causal relationships from a dataset alone. We must construct possible causal
models using our knowledge of the world first, then apply our data to our model
and other alternative models to compare their relative plausibility.</p></li>
<li><p><strong>All data have noise.</strong> The usefulness of a model to describe a dataset is
related to the relative strength of the patterns and noise in the dataset when
viewed through the lens of the model; conceptually, the so-called “signal to
noise ratio” of the data. The fundamental concern of statistics is quantifying
uncertainty (i.e. noise), and separating it from real signal, though different
statistical approaches (e.g. frequentist vs Bayesian) reason about uncertainty
in different ways.</p></li>
</ol>
<p>Modeling begins (or should begin) with posing one or more <strong>scientific models</strong>
of the process or phenomenon we wish to understand. The scientific model is
conceptual; it reflects our belief of the universe and proposes a causal
explanation for the phenomenon. We then decide how to map that scientific model
onto a <strong>statistical model</strong>, which is a mechanical procedure that quantifies
how well our scientific model explains a dataset. <em>The scientific model and
statistical model are related but independent choices we make.</em> There may be
many valid statistical models that represent a given scientific model. However,
sometimes in practice we lack sufficient knowledge about the process to propose
scientific models first, requiring data exploration and summarization first to
suggest reasonable starting points.</p>
<div class="box important">
<p>This section pertains primarily to models specified explicitly by humans. There
is another class of models, namely those created by certain machine learning
algorithms like neural networks and deep learning, that <em>discover models from
data</em>. These models are fundamentally different than those designed by human
minds, in that they are often accurate and therefore useful, but it can be very
difficult if not impossible to understand how they work. While these are
important types of models that fall under the umbrella of data science, we limit
the content of this chapter to human designed statistical models.</p>
</div>
<div id="a-worked-modeling-example" class="section level3" number="6.1.1">
<h3><span class="header-section-number">6.1.1</span> A Worked Modeling Example</h3>
<p>As an example, let’s consider a scenario where we wish to assess whether any of
three genes can help us distinguish between patients who have Parkinson’s
Disease and those who don’t by measuring the relative activity of those genes in
blood samples. We have the following made-up dataset:</p>
<div class="sourceCode" id="cb161"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb161-1"><a href="data-science.html#cb161-1" aria-hidden="true" tabindex="-1"></a>gene_exp</span></code></pre></div>
<pre><code>## # A tibble: 200 x 5
##    sample_name `Disease Status` `Gene 1` `Gene 2` `Gene 3`
##    &lt;chr&gt;       &lt;fct&gt;               &lt;dbl&gt;    &lt;dbl&gt;    &lt;dbl&gt;
##  1 P1          Parkinson&#39;s          257.     636.     504.
##  2 P2          Parkinson&#39;s          241.     556.     484.
##  3 P3          Parkinson&#39;s          252.     426.     476.
##  4 P4          Parkinson&#39;s          271.     405.     458.
##  5 P5          Parkinson&#39;s          248.     482.     520.
##  6 P6          Parkinson&#39;s          275.     521.     460.
##  7 P7          Parkinson&#39;s          264.     415.     568.
##  8 P8          Parkinson&#39;s          276.     450.     495.
##  9 P9          Parkinson&#39;s          274.     490.     496.
## 10 P10         Parkinson&#39;s          251.     584.     573.
## # ... with 190 more rows</code></pre>
<p>Our imaginary dataset has 100 Parkinson’s and 100 control subjects. For each of
our samples, we have a sample ID, Disease Status of <code>Parkinson's</code> or <code>Control</code>,
and numeric measurements of each of three genes. Below are <a href="data-visualization.html#violin-plot">violin
plots</a> of our (made-up) data set for these three genes:</p>
<p><img src="main_files/data%20modeling%20violin-1.png" width="672" /></p>
<p>By inspection, it appears that Gene 1 has no relationship with disease; we may
safely eliminate this gene from further consideration. Gene 2 appears to have a
different profile depending on disease status, where control individuals have
a higher average expression and a lower variance. Unfortunately, despite this
qualitative difference, this gene may not be useful for telling whether someone
has disease or not - the ranges completely overlap. Gene 3 appears to
discriminate between disease and control. There is some overlap in the two
expression distributions, but above a certain expression value these data
suggest a high degree of predictive accuracy may be obtained with this gene.
Measuring this gene may therefore be useful, if the results from this dataset
generalize to all people with Parkinson’s Disease.</p>
<p>So far, we have not done any modeling, but instead relied on plotting and our
eyes. A more quantitative question might be: how much higher is Gene 3
expression in Parkinson’s Disease than control? Another way of posing this
question is: if I know a patient has Parkinson’s Disease, what Gene 3 expression
value do I expect them to have? Written this way, we have turned our question
into a prediction problem: if we only had information that a patient had
Parkinson’s Disease, what is the predicted expression value of their Gene 3?</p>
<p>Another way to pose this prediction question is in the opposite (and arguably
more useful) direction: if all we knew about a person was their Gene 3 gene
expression, how likely is it that the person has Parkinson’s Disease? If this
gene expression is predictive enough of a person’s disease status, it may be a
viable <a href="https://en.wikipedia.org/wiki/Biomarker">biomarker</a> of disease and thus
might be useful in a clinical setting, for example when identifying
presymptomatic individuals or assessing the efficacy of a pharmacological
treatment.</p>
<p>Although it may seems obvious, before beginning to model a dataset, we must
start by posing the <strong>scientific question</strong> as concisely as possible, as we have
done above. These questions will help us identify which modeling techniques are
appropriate and help us ensure we interpret our results correctly.</p>
<p>We will use this example dataset throughout this chapter to illustrate some key
concepts.</p>
</div>
<div id="data-summarization" class="section level3" number="6.1.2">
<h3><span class="header-section-number">6.1.2</span> Data Summarization</h3>
<p>Broadly speaking, <em>data summarization</em> is the process of finding a
lower-dimensional representation of a larger dataset. There are many ways to
summarize a set of data; each approach will emphasize different aspects of the
dataset, and have varying degrees of accuracy. Consider the gene expression of
Gene 1 for all individuals in our example above, plotted as a distribution with
a <a href="data-visualization.html#histogram">histogram</a>:</p>
<div class="sourceCode" id="cb163"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb163-1"><a href="data-science.html#cb163-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(gene_exp, <span class="fu">aes</span>(<span class="at">x=</span><span class="st">`</span><span class="at">Gene 1</span><span class="st">`</span>)) <span class="sc">+</span></span>
<span id="cb163-2"><a href="data-science.html#cb163-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">bins=</span><span class="dv">30</span>,<span class="at">fill=</span><span class="st">&quot;#a93c13&quot;</span>)</span></code></pre></div>
<p><img src="main_files/data%20sci%20summ%20hist-1.png" width="672" /></p>
<div id="point-estimates" class="section level4" number="6.1.2.1">
<h4><span class="header-section-number">6.1.2.1</span> Point Estimates</h4>
<p>The data are concentrated around the value 250, and become less common for
larger and smaller values. Since the extents to the left and right of the middle
of the distribution appear to be equally distant, perhaps the arithmetic mean
is a good way to identify the middle of the distribution:</p>
<div class="sourceCode" id="cb164"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb164-1"><a href="data-science.html#cb164-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(gene_exp, <span class="fu">aes</span>(<span class="at">x=</span><span class="st">`</span><span class="at">Gene 1</span><span class="st">`</span>)) <span class="sc">+</span></span>
<span id="cb164-2"><a href="data-science.html#cb164-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">bins=</span><span class="dv">30</span>,<span class="at">fill=</span><span class="st">&quot;#a93c13&quot;</span>) <span class="sc">+</span></span>
<span id="cb164-3"><a href="data-science.html#cb164-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept=</span><span class="fu">mean</span>(gene_exp<span class="sc">$</span><span class="st">`</span><span class="at">Gene 1</span><span class="st">`</span>))</span></code></pre></div>
<p><img src="main_files/data%20sci%20summ%20hist2-1.png" width="672" /></p>
<p>By eye, the mean does seem to correspond well to the value that is among the
most frequent, and successfully captures an important aspect of the data: its
<em>central tendency</em>. Summaries that compute a single number are called <em>point
estimates</em>. Point estimates collapse the data into one singular point, one
value.</p>
<div class="box">
<p>The arithmetic mean is just one measure of central tendency, computed by taking
the sum of all the values and dividing by the number of values. The mean may be
a good point estimate of the central tendency of a dataset, but it is sensitive
to outlier samples. Consider the following examples:</p>
<div class="sourceCode" id="cb165"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb165-1"><a href="data-science.html#cb165-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(patchwork)</span>
<span id="cb165-2"><a href="data-science.html#cb165-2" aria-hidden="true" tabindex="-1"></a>well_behaved_data <span class="ot">&lt;-</span> <span class="fu">tibble</span>(<span class="at">data =</span> <span class="fu">rnorm</span>(<span class="dv">1000</span>))</span>
<span id="cb165-3"><a href="data-science.html#cb165-3" aria-hidden="true" tabindex="-1"></a>data_w_outliers <span class="ot">&lt;-</span> <span class="fu">tibble</span>(<span class="at">data =</span> <span class="fu">c</span>(<span class="fu">rnorm</span>(<span class="dv">800</span>), <span class="fu">rep</span>(<span class="dv">5</span>, <span class="dv">200</span>))) <span class="co"># oops I add some outliers :^)</span></span>
<span id="cb165-4"><a href="data-science.html#cb165-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb165-5"><a href="data-science.html#cb165-5" aria-hidden="true" tabindex="-1"></a>g_no_outlier <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(well_behaved_data, <span class="fu">aes</span>(<span class="at">x =</span> data)) <span class="sc">+</span></span>
<span id="cb165-6"><a href="data-science.html#cb165-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">fill =</span> <span class="st">&quot;#56CBF9&quot;</span>, <span class="at">color =</span> <span class="st">&quot;grey&quot;</span>, <span class="at">bins =</span> <span class="dv">30</span>) <span class="sc">+</span></span>
<span id="cb165-7"><a href="data-science.html#cb165-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="fu">mean</span>(well_behaved_data<span class="sc">$</span>data)) <span class="sc">+</span></span>
<span id="cb165-8"><a href="data-science.html#cb165-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">&quot;Mean example, no outliers&quot;</span>)</span>
<span id="cb165-9"><a href="data-science.html#cb165-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb165-10"><a href="data-science.html#cb165-10" aria-hidden="true" tabindex="-1"></a>g_outlier <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(data_w_outliers, <span class="fu">aes</span>(<span class="at">x =</span> data)) <span class="sc">+</span></span>
<span id="cb165-11"><a href="data-science.html#cb165-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">fill =</span> <span class="st">&quot;#7FBEEB&quot;</span>, <span class="at">color =</span> <span class="st">&quot;grey&quot;</span>, <span class="at">bins =</span> <span class="dv">30</span>) <span class="sc">+</span></span>
<span id="cb165-12"><a href="data-science.html#cb165-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="fu">mean</span>(data_w_outliers<span class="sc">$</span>data)) <span class="sc">+</span></span>
<span id="cb165-13"><a href="data-science.html#cb165-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">&quot;Mean example, big outliers&quot;</span>)</span>
<span id="cb165-14"><a href="data-science.html#cb165-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb165-15"><a href="data-science.html#cb165-15" aria-hidden="true" tabindex="-1"></a>g_no_outlier <span class="sc">|</span> g_outlier</span></code></pre></div>
<p><img src="main_files/unnamed-chunk-43-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>The median is another measure of central tendency, which is found by identifying
the value that divides the samples into equal halves when sorted from smallest
to largest. The median is more robust in the presence of outliers.</p>
<div class="sourceCode" id="cb166"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb166-1"><a href="data-science.html#cb166-1" aria-hidden="true" tabindex="-1"></a>g_no_outlier <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(well_behaved_data, <span class="fu">aes</span>(<span class="at">x =</span> data)) <span class="sc">+</span></span>
<span id="cb166-2"><a href="data-science.html#cb166-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">fill =</span> <span class="st">&quot;#AFBED1&quot;</span>, <span class="at">color =</span> <span class="st">&quot;grey&quot;</span>, <span class="at">bins =</span> <span class="dv">30</span>) <span class="sc">+</span></span>
<span id="cb166-3"><a href="data-science.html#cb166-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="fu">median</span>(well_behaved_data<span class="sc">$</span>data)) <span class="sc">+</span></span>
<span id="cb166-4"><a href="data-science.html#cb166-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">&quot;Median example&quot;</span>)</span>
<span id="cb166-5"><a href="data-science.html#cb166-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb166-6"><a href="data-science.html#cb166-6" aria-hidden="true" tabindex="-1"></a>g_outlier <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(data_w_outliers, <span class="fu">aes</span>(<span class="at">x =</span> data)) <span class="sc">+</span></span>
<span id="cb166-7"><a href="data-science.html#cb166-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">fill =</span> <span class="st">&quot;#7FBEEB&quot;</span>, <span class="at">color =</span> <span class="st">&quot;grey&quot;</span>, <span class="at">bins =</span> <span class="dv">30</span>) <span class="sc">+</span></span>
<span id="cb166-8"><a href="data-science.html#cb166-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="fu">median</span>(data_w_outliers<span class="sc">$</span>data)) <span class="sc">+</span></span>
<span id="cb166-9"><a href="data-science.html#cb166-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">&quot;Median example, big outliers&quot;</span>)</span>
<span id="cb166-10"><a href="data-science.html#cb166-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb166-11"><a href="data-science.html#cb166-11" aria-hidden="true" tabindex="-1"></a>g_no_outlier <span class="sc">|</span> g_outlier</span></code></pre></div>
<p><img src="main_files/unnamed-chunk-44-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
</div>
<div id="dispersion" class="section level4" number="6.1.2.2">
<h4><span class="header-section-number">6.1.2.2</span> Dispersion</h4>
<p>Central tendencies are important aspects of the data but don’t describe what the
data do for values outside this point estimate of central tendency; in other
words, we have not expressed the spread, or <em>dispersion</em> of the data.</p>
<p>We decide that perhaps computing the <a href="https://en.wikipedia.org/wiki/Standard_deviation">standard
deviation</a> of the data may
characterize the spread well, since it appears to be symmetric around the mean.
We can layer this information on the plot as well to inspect it:</p>
<div class="sourceCode" id="cb167"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb167-1"><a href="data-science.html#cb167-1" aria-hidden="true" tabindex="-1"></a>g1_mean <span class="ot">&lt;-</span> <span class="fu">mean</span>(gene_exp<span class="sc">$</span><span class="st">`</span><span class="at">Gene 1</span><span class="st">`</span>)</span>
<span id="cb167-2"><a href="data-science.html#cb167-2" aria-hidden="true" tabindex="-1"></a>g1_sd <span class="ot">&lt;-</span> <span class="fu">sd</span>(gene_exp<span class="sc">$</span><span class="st">`</span><span class="at">Gene 1</span><span class="st">`</span>)</span>
<span id="cb167-3"><a href="data-science.html#cb167-3" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(gene_exp, <span class="fu">aes</span>(<span class="at">x=</span><span class="st">`</span><span class="at">Gene 1</span><span class="st">`</span>)) <span class="sc">+</span></span>
<span id="cb167-4"><a href="data-science.html#cb167-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">bins=</span><span class="dv">30</span>,<span class="at">fill=</span><span class="st">&quot;#a93c13&quot;</span>) <span class="sc">+</span></span>
<span id="cb167-5"><a href="data-science.html#cb167-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept=</span>g1_mean) <span class="sc">+</span></span>
<span id="cb167-6"><a href="data-science.html#cb167-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_segment</span>(<span class="at">x=</span>g1_mean<span class="sc">-</span>g1_sd, <span class="at">y=</span><span class="dv">10</span>, <span class="at">xend=</span>g1_mean<span class="sc">+</span>g1_sd, <span class="at">yend=</span><span class="dv">10</span>)</span></code></pre></div>
<p><img src="main_files/data%20sci%20summ%20sd-1.png" width="672" /></p>
<p>The width of the horizontal line is proportional to the mean +/- one standard
deviation around the mean, and has been placed arbitrarily on the y axis at <code>y = 10</code> to show how this range covers the data in the histogram. The +/- 1 standard
deviation around the mean visually describes the spread of the data reasonably
well.</p>
<div class="box">
<p>Measures the spread of the data, typically around its perceived center (a mean).
Often related to the distribution of the data.</p>
<p><strong>Standard deviation</strong>: A measure of how close values are to the mean. Bigger
standard deviations mean data is more spread out.</p>
<div class="sourceCode" id="cb168"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb168-1"><a href="data-science.html#cb168-1" aria-hidden="true" tabindex="-1"></a>data <span class="ot">&lt;-</span> <span class="fu">tibble</span>(<span class="at">data =</span> <span class="fu">c</span>(<span class="fu">rnorm</span>(<span class="dv">1000</span>, <span class="at">sd=</span><span class="fl">1.75</span>)))</span>
<span id="cb168-2"><a href="data-science.html#cb168-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(data, <span class="fu">aes</span>(<span class="at">x =</span> data)) <span class="sc">+</span></span>
<span id="cb168-3"><a href="data-science.html#cb168-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">fill =</span> <span class="st">&quot;#EAC5D8&quot;</span>, <span class="at">color =</span> <span class="st">&quot;white&quot;</span>, <span class="at">bins =</span> <span class="dv">30</span>) <span class="sc">+</span></span>
<span id="cb168-4"><a href="data-science.html#cb168-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="fu">c</span>(<span class="sc">-</span><span class="dv">2</span>, <span class="sc">-</span><span class="dv">1</span>, <span class="dv">0</span>, <span class="dv">1</span>, <span class="dv">2</span>) <span class="sc">*</span> <span class="fu">sd</span>(data<span class="sc">$</span>data)) <span class="sc">+</span></span>
<span id="cb168-5"><a href="data-science.html#cb168-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlim</span>(<span class="fu">c</span>(<span class="sc">-</span><span class="dv">6</span>, <span class="dv">6</span>)) <span class="sc">+</span></span>
<span id="cb168-6"><a href="data-science.html#cb168-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">&quot;Standard deviations aplenty&quot;</span>, <span class="fu">paste</span>(<span class="st">&quot;SD:&quot;</span>, <span class="fu">sd</span>(data<span class="sc">$</span>data)))</span></code></pre></div>
<p><img src="main_files/taylor%20sd-1.png" width="672" style="display: block; margin: auto;" /></p>
<p><strong>Variance</strong>:
Similar to SD (it’s the square of SD), variance measures how far a random value
is from the mean.</p>
<div class="sourceCode" id="cb169"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb169-1"><a href="data-science.html#cb169-1" aria-hidden="true" tabindex="-1"></a>data <span class="ot">&lt;-</span> <span class="fu">tibble</span>(<span class="at">data =</span> <span class="fu">c</span>(<span class="fu">rnorm</span>(<span class="dv">1000</span>, <span class="at">sd=</span><span class="fl">0.5</span>)))</span>
<span id="cb169-2"><a href="data-science.html#cb169-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(data, <span class="fu">aes</span>(<span class="at">x =</span> data)) <span class="sc">+</span></span>
<span id="cb169-3"><a href="data-science.html#cb169-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">fill =</span> <span class="st">&quot;#DBD8F0&quot;</span>, <span class="at">color =</span> <span class="st">&quot;white&quot;</span>, <span class="at">bins =</span> <span class="dv">30</span>) <span class="sc">+</span></span>
<span id="cb169-4"><a href="data-science.html#cb169-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="fu">mean</span>(data<span class="sc">$</span>data)) <span class="sc">+</span></span>
<span id="cb169-5"><a href="data-science.html#cb169-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlim</span>(<span class="fu">c</span>(<span class="sc">-</span><span class="dv">6</span>, <span class="dv">6</span>)) <span class="sc">+</span></span>
<span id="cb169-6"><a href="data-science.html#cb169-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">&quot;Same mean as SD plot, but different variance&quot;</span>,</span>
<span id="cb169-7"><a href="data-science.html#cb169-7" aria-hidden="true" tabindex="-1"></a>          <span class="fu">paste</span>(<span class="st">&quot;Variance:&quot;</span>, <span class="fu">sd</span>(data<span class="sc">$</span>data)))</span></code></pre></div>
<p><img src="main_files/taylor%20var-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
</div>
<div id="distributions" class="section level4" number="6.1.2.3">
<h4><span class="header-section-number">6.1.2.3</span> Distributions</h4>
<p>With these two pieces of knowledge - the mean accurately describes the center of
the data and the standard deviation describes the spread - we now recognize that
these data may be <a href="https://en.wikipedia.org/wiki/Normal_distribution">normally distributed</a>, and therefore
we can potentially describe the dataset mathematically. We decide to visually
inspect this possibility by layering a normal distribution on top of our data
using
<a href="https://ggplot2.tidyverse.org/reference/geom_function.html"><code>stat_function</code></a>:</p>
<div class="sourceCode" id="cb170"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb170-1"><a href="data-science.html#cb170-1" aria-hidden="true" tabindex="-1"></a>g1_mean <span class="ot">&lt;-</span> <span class="fu">mean</span>(gene_exp<span class="sc">$</span><span class="st">`</span><span class="at">Gene 1</span><span class="st">`</span>)</span>
<span id="cb170-2"><a href="data-science.html#cb170-2" aria-hidden="true" tabindex="-1"></a>g1_sd <span class="ot">&lt;-</span> <span class="fu">sd</span>(gene_exp<span class="sc">$</span><span class="st">`</span><span class="at">Gene 1</span><span class="st">`</span>)</span>
<span id="cb170-3"><a href="data-science.html#cb170-3" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(gene_exp, <span class="fu">aes</span>(<span class="at">x=</span><span class="st">`</span><span class="at">Gene 1</span><span class="st">`</span>)) <span class="sc">+</span></span>
<span id="cb170-4"><a href="data-science.html#cb170-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(</span>
<span id="cb170-5"><a href="data-science.html#cb170-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">aes</span>(<span class="at">y=</span><span class="fu">after_stat</span>(density)),</span>
<span id="cb170-6"><a href="data-science.html#cb170-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">bins=</span><span class="dv">30</span>,</span>
<span id="cb170-7"><a href="data-science.html#cb170-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">fill=</span><span class="st">&quot;#a93c13&quot;</span></span>
<span id="cb170-8"><a href="data-science.html#cb170-8" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">+</span></span>
<span id="cb170-9"><a href="data-science.html#cb170-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">stat_function</span>(<span class="at">fun=</span>dnorm, <span class="at">args =</span> <span class="fu">list</span>(<span class="at">mean=</span>g1_mean, <span class="at">sd=</span>g1_sd), <span class="at">size=</span><span class="dv">2</span>)</span></code></pre></div>
<p><img src="main_files/unnamed-chunk-45-1.png" width="672" /></p>
<p>Note the histogram bars are scaled with
<code>aes(y=[after_stat](https://ggplot2.tidyverse.org/reference/aes_eval.html)(density))</code>
to the density of values in each bin to make all the bar heights sum to 1 so
that the y scale matches that of a normal distribution.</p>
<p>We have now created our first model: we chose to express the dataset as a normal
distribution parameterized by the mean and standard deviation and standard
deviation of the data. Using the values of 254 and 11
as our mean and standard deviation, respectively, we can express our model
mathematically as follows:</p>
<p><span class="math display">\[
Gene\;1 \sim \mathcal{N}(254, 11)
\]</span></p>
<p>Here the <span class="math inline">\(\sim\)</span> symbol means “distributed as” and the <span class="math inline">\(\mathcal{N}(\mu,\sigma)\)</span>
represents a normal distribution with mean of <span class="math inline">\(\mu\)</span> and standard deviation of
<span class="math inline">\(\sigma\)</span>. This is mathematical formulation means the same thing as saying we are
modeling Gene 1 expression as a normal distribution with mean of 254 and
standard deviation of 11. Without any additional information about a new sample,
we would expect the expression of that gene to be 254, although it may vary from
this value.</p>
<div class="box">
<p>The normal distribution is the most common distribution observed in nature, but
it is hardly the only one. We could have proposed other distributions to instead
summarize our data:</p>
<div class="sourceCode" id="cb171"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb171-1"><a href="data-science.html#cb171-1" aria-hidden="true" tabindex="-1"></a>g_norm <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(<span class="fu">tibble</span>(<span class="at">data =</span> <span class="fu">rnorm</span>(<span class="dv">5000</span>)), <span class="fu">aes</span>(<span class="at">x =</span> data)) <span class="sc">+</span></span>
<span id="cb171-2"><a href="data-science.html#cb171-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">fill =</span> <span class="st">&quot;#D0FCB3&quot;</span>, <span class="at">bins =</span> <span class="dv">50</span>, <span class="at">color =</span> <span class="st">&quot;gray&quot;</span>) <span class="sc">+</span></span>
<span id="cb171-3"><a href="data-science.html#cb171-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">&quot;Normal distribution&quot;</span>, <span class="st">&quot;rnorm(n = 1000)&quot;</span>)</span>
<span id="cb171-4"><a href="data-science.html#cb171-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb171-5"><a href="data-science.html#cb171-5" aria-hidden="true" tabindex="-1"></a>g_unif <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(<span class="fu">tibble</span>(<span class="at">data =</span> <span class="fu">runif</span>(<span class="dv">5000</span>)), <span class="fu">aes</span>(<span class="at">x =</span> data)) <span class="sc">+</span></span>
<span id="cb171-6"><a href="data-science.html#cb171-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">fill =</span> <span class="st">&quot;#271F30&quot;</span>, <span class="at">bins =</span> <span class="dv">50</span>, <span class="at">color =</span> <span class="st">&quot;white&quot;</span>) <span class="sc">+</span></span>
<span id="cb171-7"><a href="data-science.html#cb171-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">&quot;Uniform distribution&quot;</span>, <span class="st">&quot;runif(n = 1000)&quot;</span>)</span>
<span id="cb171-8"><a href="data-science.html#cb171-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb171-9"><a href="data-science.html#cb171-9" aria-hidden="true" tabindex="-1"></a>g_logistic <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(<span class="fu">tibble</span>(<span class="at">data =</span> <span class="fu">rlogis</span>(<span class="dv">5000</span>)), <span class="fu">aes</span>(<span class="at">x =</span> data)) <span class="sc">+</span></span>
<span id="cb171-10"><a href="data-science.html#cb171-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">fill =</span> <span class="st">&quot;#9BC59D&quot;</span>, <span class="at">bins =</span> <span class="dv">50</span>, <span class="at">color =</span> <span class="st">&quot;black&quot;</span>) <span class="sc">+</span></span>
<span id="cb171-11"><a href="data-science.html#cb171-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">&quot;Logistic distribution&quot;</span>, <span class="st">&quot;rlogis(n = 1000)&quot;</span>)</span>
<span id="cb171-12"><a href="data-science.html#cb171-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb171-13"><a href="data-science.html#cb171-13" aria-hidden="true" tabindex="-1"></a>g_exp <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(<span class="fu">tibble</span>(<span class="at">data =</span> <span class="fu">rexp</span>(<span class="dv">5000</span>, <span class="at">rate =</span> <span class="dv">1</span>)), <span class="fu">aes</span>(<span class="at">x =</span> data)) <span class="sc">+</span></span>
<span id="cb171-14"><a href="data-science.html#cb171-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">fill =</span> <span class="st">&quot;#6C5A49&quot;</span>, <span class="at">bins =</span> <span class="dv">50</span>, <span class="at">color =</span> <span class="st">&quot;white&quot;</span>) <span class="sc">+</span></span>
<span id="cb171-15"><a href="data-science.html#cb171-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">&quot;Exponential distribution&quot;</span>, <span class="st">&quot;rexp(n = 1000, rate = 1)&quot;</span>)</span>
<span id="cb171-16"><a href="data-science.html#cb171-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb171-17"><a href="data-science.html#cb171-17" aria-hidden="true" tabindex="-1"></a>(g_norm <span class="sc">|</span> g_unif) <span class="sc">/</span> (g_logistic <span class="sc">|</span> g_exp)</span></code></pre></div>
<p><img src="main_files/unnamed-chunk-46-1.png" width="672" style="display: block; margin: auto;" />
In addition to the normal distribution, we have also plotted samples drawn from
a <a href="ihttps://en.wikipedia.org/wiki/Continuous_uniform_distribution">continuous uniform
distribution</a>
between 0 and 1, a <a href="https://en.wikipedia.org/wiki/Logistic_distribution">logistic
distribution</a> which is
similar to the normal distribution but has heavier “tails,” and an <a href="https://en.wikipedia.org/wiki/Exponential_distribution">exponential
distribution</a>.
There are many more distributions than these, and many of them were discovered
to arise in nature and encode different types of processes and relationships.</p>
</div>
<p>A few notes on our data modeling example before we move on:</p>
<ol style="list-style-type: decimal">
<li><p><strong>Our model choice was totally subjective.</strong> We looked at the data and
decided that a normal distribution was a reasonable choice. There were many
other choices we could have made, and all of them would be equally valid, though
they may not all describe the data equally well.</p></li>
<li><p><strong>We can’t know if this is the “correct” model for the data.</strong> By eye, it
appears to be a reasonably accurate summary. However, there is no such thing as
a correct model; some models are simply better at describing the data than
others. Recall that all models are wrong, and some models are useful. Our model
may be useful, but it is definitely wrong to some extent.</p></li>
<li><p><strong>We don’t know how well our model describes the data yet.</strong> So far we’ve
only used our eyes to choose our model which might be a good starting point
considering our data are so simple, but we have not yet quantified how well our
model describes the data, or compared it to alternative models to see which is
better. This will be discussed briefly in a later section.</p></li>
</ol>
</div>
</div>
<div id="linear-models" class="section level3" number="6.1.3">
<h3><span class="header-section-number">6.1.3</span> Linear Models</h3>
<p>Our choice of a normal distribution to model our Gene 1 gene expression was only
<em>descriptive</em>; it was a low-dimensional summary of our dataset. However, it was
not very <em>informative</em>; it doesn’t tell us anything useful about Gene 1
expression with respect to our scientific question of distinguishing between
Parkinson’s Disease and Control individuals. To do that, we will need to find a
model that can make <em>predictions</em> that we may find useful if we receive new
data. To do that, we will introduce a new type of model: the linear model.</p>
<p>A linear model is any statistical model that relates one outcome variable as a
linear combination (i.e. sum) of one or more explanatory variables. This may be
expressed mathematically as follows:</p>
<p><span class="math display">\[
Y_i = \beta_0 + \beta_1 \phi_1 ( X_{i1} ) + \beta_2 \phi_2 ( X_{i2} ) + \ldots + \beta_p \phi_p ( X_{ip} ) + \epsilon_i
\]</span></p>
<p>Above, <span class="math inline">\(Y_i\)</span> is some outcome or response variable we wish to model, <span class="math inline">\(X_{ij}\)</span> is
our explanatory or predictor variable <span class="math inline">\(j\)</span> for observatoin <span class="math inline">\(i\)</span>, and <span class="math inline">\(\beta_j\)</span> are
coefficients estimated to minimize the difference between the predicted outcome
<span class="math inline">\(\hat{Y_i}\)</span> and the observed <span class="math inline">\(Y_i\)</span> over all observations. <span class="math inline">\(\phi_j\)</span> is a possibly
non-linear transformation of the explanatory variable <span class="math inline">\(X_ij\)</span>; note these
functions may be non-linear so long as the predicted outcome is modeled as a
linear combination of the transformed variables. The rest of this section is
dedicated to a worked example of a linear model for gene expression data.</p>
<p>Let us begin with a <a href="data-visualization.html#beeswarm-plot">beeswarm plot</a> plot of Gene 3:</p>
<div class="sourceCode" id="cb172"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb172-1"><a href="data-science.html#cb172-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggbeeswarm)</span>
<span id="cb172-2"><a href="data-science.html#cb172-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(gene_exp, <span class="fu">aes</span>(<span class="at">x=</span><span class="st">`</span><span class="at">Disease Status</span><span class="st">`</span>, <span class="at">y=</span><span class="st">`</span><span class="at">Gene 3</span><span class="st">`</span>, <span class="at">color=</span><span class="st">`</span><span class="at">Disease Status</span><span class="st">`</span>)) <span class="sc">+</span></span>
<span id="cb172-3"><a href="data-science.html#cb172-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_beeswarm</span>()</span></code></pre></div>
<p><img src="main_files/lin%20model%20gene%203%20violin-1.png" width="672" /></p>
<p>The expression values within each disease status look like they might be
normally distributed just like Gene 1, so let’s summarize each group with the
arithmetic mean and standard deviation as before, and instead plot both
distributions as histograms:</p>
<div class="sourceCode" id="cb173"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb173-1"><a href="data-science.html#cb173-1" aria-hidden="true" tabindex="-1"></a>exp_summ <span class="ot">&lt;-</span> <span class="fu">pivot_longer</span>(</span>
<span id="cb173-2"><a href="data-science.html#cb173-2" aria-hidden="true" tabindex="-1"></a>  gene_exp,</span>
<span id="cb173-3"><a href="data-science.html#cb173-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">c</span>(<span class="st">`</span><span class="at">Gene 3</span><span class="st">`</span>)</span>
<span id="cb173-4"><a href="data-science.html#cb173-4" aria-hidden="true" tabindex="-1"></a>) <span class="sc">%&gt;%</span></span>
<span id="cb173-5"><a href="data-science.html#cb173-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(<span class="st">`</span><span class="at">Disease Status</span><span class="st">`</span>) <span class="sc">%&gt;%</span></span>
<span id="cb173-6"><a href="data-science.html#cb173-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarize</span>(<span class="at">mean=</span><span class="fu">mean</span>(value),<span class="at">sd=</span><span class="fu">sd</span>(value))</span>
<span id="cb173-7"><a href="data-science.html#cb173-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb173-8"><a href="data-science.html#cb173-8" aria-hidden="true" tabindex="-1"></a>pd_mean <span class="ot">&lt;-</span> <span class="fu">filter</span>(exp_summ, <span class="st">`</span><span class="at">Disease Status</span><span class="st">`</span> <span class="sc">==</span> <span class="st">&quot;Parkinson&#39;s&quot;</span>)<span class="sc">$</span>mean</span>
<span id="cb173-9"><a href="data-science.html#cb173-9" aria-hidden="true" tabindex="-1"></a>c_mean <span class="ot">&lt;-</span> <span class="fu">filter</span>(exp_summ, <span class="st">`</span><span class="at">Disease Status</span><span class="st">`</span> <span class="sc">==</span> <span class="st">&quot;Control&quot;</span>)<span class="sc">$</span>mean</span>
<span id="cb173-10"><a href="data-science.html#cb173-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb173-11"><a href="data-science.html#cb173-11" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(gene_exp, <span class="fu">aes</span>(<span class="at">x=</span><span class="st">`</span><span class="at">Gene 3</span><span class="st">`</span>, <span class="at">fill=</span><span class="st">`</span><span class="at">Disease Status</span><span class="st">`</span>)) <span class="sc">+</span></span>
<span id="cb173-12"><a href="data-science.html#cb173-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">bins=</span><span class="dv">20</span>, <span class="at">alpha=</span><span class="fl">0.6</span>,<span class="at">position=</span><span class="st">&quot;identity&quot;</span>) <span class="sc">+</span></span>
<span id="cb173-13"><a href="data-science.html#cb173-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">annotate</span>(<span class="st">&quot;segment&quot;</span>, <span class="at">x=</span>c_mean, <span class="at">xend=</span>pd_mean, <span class="at">y=</span><span class="dv">20</span>, <span class="at">yend=</span><span class="dv">20</span>, <span class="at">arrow=</span><span class="fu">arrow</span>(<span class="at">ends=</span><span class="st">&quot;both&quot;</span>, <span class="at">angle=</span><span class="dv">90</span>)) <span class="sc">+</span></span>
<span id="cb173-14"><a href="data-science.html#cb173-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">annotate</span>(<span class="st">&quot;text&quot;</span>, <span class="at">x=</span><span class="fu">mean</span>(<span class="fu">c</span>(c_mean,pd_mean)), <span class="at">y=</span><span class="dv">21</span>, <span class="at">hjust=</span><span class="fl">0.5</span>, <span class="at">label=</span><span class="st">&quot;How different?&quot;</span>)</span></code></pre></div>
<p><img src="main_files/unnamed-chunk-47-1.png" width="672" /></p>
<p>We can make a point estimate of this difference by simply subtracting the means:</p>
<div class="sourceCode" id="cb174"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb174-1"><a href="data-science.html#cb174-1" aria-hidden="true" tabindex="-1"></a>pd_mean <span class="sc">-</span> c_mean</span></code></pre></div>
<pre><code>## [1] 164.0942</code></pre>
<p>In other words, this point estimate suggests that on average Parkinson’s
patients have 164.1 greater expression than Controls. We can
plot this relationship relatively simply:</p>
<div class="sourceCode" id="cb176"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb176-1"><a href="data-science.html#cb176-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(gene_exp, <span class="fu">aes</span>(<span class="at">x=</span><span class="st">`</span><span class="at">Disease Status</span><span class="st">`</span>, <span class="at">y=</span><span class="st">`</span><span class="at">Gene 3</span><span class="st">`</span>, <span class="at">color=</span><span class="st">`</span><span class="at">Disease Status</span><span class="st">`</span>)) <span class="sc">+</span></span>
<span id="cb176-2"><a href="data-science.html#cb176-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_beeswarm</span>() <span class="sc">+</span></span>
<span id="cb176-3"><a href="data-science.html#cb176-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">annotate</span>(<span class="st">&quot;segment&quot;</span>, <span class="at">x=</span><span class="dv">0</span>, <span class="at">xend=</span><span class="dv">3</span>, <span class="at">y=</span><span class="dv">2</span><span class="sc">*</span>c_mean<span class="sc">-</span>pd_mean, <span class="at">yend=</span><span class="dv">2</span><span class="sc">*</span>pd_mean<span class="sc">-</span>c_mean)</span></code></pre></div>
<p><img src="main_files/lin%20model%20gene%203%20fit-1.png" width="672" /></p>
<p>However, this point estimate tells us nothing about how confident we are about
the difference. We can do this by using an <a href="https://en.wikipedia.org/wiki/Least_squares">linear
regression</a> by modeling Gene 3 as a
function of disease status:</p>
<div class="sourceCode" id="cb177"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb177-1"><a href="data-science.html#cb177-1" aria-hidden="true" tabindex="-1"></a>fit <span class="ot">&lt;-</span> <span class="fu">lm</span>(<span class="st">`</span><span class="at">Gene 3</span><span class="st">`</span> <span class="sc">~</span> <span class="st">`</span><span class="at">Disease Status</span><span class="st">`</span>, <span class="at">data=</span>gene_exp)</span>
<span id="cb177-2"><a href="data-science.html#cb177-2" aria-hidden="true" tabindex="-1"></a>fit</span></code></pre></div>
<pre><code>## 
## Call:
## lm(formula = `Gene 3` ~ `Disease Status`, data = gene_exp)
## 
## Coefficients:
##                 (Intercept)  `Disease Status`Parkinson&#39;s  
##                       334.6                        164.1</code></pre>
<p>The coefficient associated with having the disease status of Parkinson’s disease
is almost exactly equal to our difference in means. We also note that the
coefficient labeled <code>(Intercept)</code> is nearly equal to the mean of our control
samples (334.6). Under the hood, this simple linear model did the
same calculation we did by subtracting the means of each group but estimated the
means using all the data at once, instead of point estimates. Another advantage
of using <code>lm()</code> over the point estimate method is the model can estimate how
confident the model was that the difference in mean between Parkinson’s and
controls subjects. Let’s print some more information about the model than
before:</p>
<div class="sourceCode" id="cb179"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb179-1"><a href="data-science.html#cb179-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(fit)</span></code></pre></div>
<pre><code>## 
## Call:
## lm(formula = `Gene 3` ~ `Disease Status`, data = gene_exp)
## 
## Residuals:
##      Min       1Q   Median       3Q      Max 
## -124.303  -25.758   -2.434   30.518  119.348 
## 
## Coefficients:
##                             Estimate Std. Error t value Pr(&gt;|t|)    
## (Intercept)                  334.578      4.414   75.80   &lt;2e-16 ***
## `Disease Status`Parkinson&#39;s  164.094      6.243   26.29   &lt;2e-16 ***
## ---
## Signif. codes:  0 &#39;***&#39; 0.001 &#39;**&#39; 0.01 &#39;*&#39; 0.05 &#39;.&#39; 0.1 &#39; &#39; 1
## 
## Residual standard error: 44.14 on 198 degrees of freedom
## Multiple R-squared:  0.7773, Adjusted R-squared:  0.7761 
## F-statistic:   691 on 1 and 198 DF,  p-value: &lt; 2.2e-16</code></pre>
<p>We again see our coefficient estimates for the intercept (i.e. mean control
expression) and our increase in Parkinson’s, but also a number of other terms,
in particular <code>Pr(&gt;|t|)</code> and <code>Multiple R-squared</code>. The former is the
<a href="#p-values">p-value</a> associated with each of the coefficient estimates, both of
which are very small, indicating to us that the model was very confident of the
estimated values. The latter, multiple R-squared or <span class="math inline">\(R^2\)</span>, describes how much of
the variance in the data was explained by the model it found as a fraction
between 0 and 1. This model explains 77.7% of the variance of the data, which is
substantial. The <span class="math inline">\(R^2\)</span> value also has an associated p-value, which is also very
small. Overall, these statistics suggest this model fits the data very well.</p>
<p>We can plot the results of a linear model for each of our genes relatively
easily:</p>
<div class="sourceCode" id="cb181"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb181-1"><a href="data-science.html#cb181-1" aria-hidden="true" tabindex="-1"></a>pd_mean <span class="ot">&lt;-</span> <span class="fu">mean</span>(<span class="fu">filter</span>(gene_exp,<span class="st">`</span><span class="at">Disease Status</span><span class="st">`</span><span class="sc">==</span><span class="st">&quot;Parkinson&#39;s&quot;</span>)<span class="sc">$</span><span class="st">`</span><span class="at">Gene 1</span><span class="st">`</span>)</span>
<span id="cb181-2"><a href="data-science.html#cb181-2" aria-hidden="true" tabindex="-1"></a>c_mean <span class="ot">&lt;-</span> <span class="fu">mean</span>(<span class="fu">filter</span>(gene_exp,<span class="st">`</span><span class="at">Disease Status</span><span class="st">`</span><span class="sc">==</span><span class="st">&quot;Control&quot;</span>)<span class="sc">$</span><span class="st">`</span><span class="at">Gene 1</span><span class="st">`</span>)</span>
<span id="cb181-3"><a href="data-science.html#cb181-3" aria-hidden="true" tabindex="-1"></a>g1 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(gene_exp, <span class="fu">aes</span>(<span class="at">x=</span><span class="st">`</span><span class="at">Disease Status</span><span class="st">`</span>, <span class="at">y=</span><span class="st">`</span><span class="at">Gene 1</span><span class="st">`</span>, <span class="at">color=</span><span class="st">`</span><span class="at">Disease Status</span><span class="st">`</span>)) <span class="sc">+</span></span>
<span id="cb181-4"><a href="data-science.html#cb181-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_beeswarm</span>() <span class="sc">+</span></span>
<span id="cb181-5"><a href="data-science.html#cb181-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">annotate</span>(<span class="st">&quot;segment&quot;</span>, <span class="at">x=</span><span class="dv">0</span>, <span class="at">xend=</span><span class="dv">3</span>, <span class="at">y=</span><span class="dv">2</span><span class="sc">*</span>c_mean<span class="sc">-</span>pd_mean, <span class="at">yend=</span><span class="dv">2</span><span class="sc">*</span>pd_mean<span class="sc">-</span>c_mean) <span class="sc">+</span></span>
<span id="cb181-6"><a href="data-science.html#cb181-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position=</span><span class="st">&quot;none&quot;</span>)</span>
<span id="cb181-7"><a href="data-science.html#cb181-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb181-8"><a href="data-science.html#cb181-8" aria-hidden="true" tabindex="-1"></a>pd_mean <span class="ot">&lt;-</span> <span class="fu">mean</span>(<span class="fu">filter</span>(gene_exp,<span class="st">`</span><span class="at">Disease Status</span><span class="st">`</span><span class="sc">==</span><span class="st">&quot;Parkinson&#39;s&quot;</span>)<span class="sc">$</span><span class="st">`</span><span class="at">Gene 2</span><span class="st">`</span>)</span>
<span id="cb181-9"><a href="data-science.html#cb181-9" aria-hidden="true" tabindex="-1"></a>c_mean <span class="ot">&lt;-</span> <span class="fu">mean</span>(<span class="fu">filter</span>(gene_exp,<span class="st">`</span><span class="at">Disease Status</span><span class="st">`</span><span class="sc">==</span><span class="st">&quot;Control&quot;</span>)<span class="sc">$</span><span class="st">`</span><span class="at">Gene 2</span><span class="st">`</span>)</span>
<span id="cb181-10"><a href="data-science.html#cb181-10" aria-hidden="true" tabindex="-1"></a>g2 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(gene_exp, <span class="fu">aes</span>(<span class="at">x=</span><span class="st">`</span><span class="at">Disease Status</span><span class="st">`</span>, <span class="at">y=</span><span class="st">`</span><span class="at">Gene 2</span><span class="st">`</span>, <span class="at">color=</span><span class="st">`</span><span class="at">Disease Status</span><span class="st">`</span>)) <span class="sc">+</span></span>
<span id="cb181-11"><a href="data-science.html#cb181-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_beeswarm</span>() <span class="sc">+</span></span>
<span id="cb181-12"><a href="data-science.html#cb181-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">annotate</span>(<span class="st">&quot;segment&quot;</span>, <span class="at">x=</span><span class="dv">0</span>, <span class="at">xend=</span><span class="dv">3</span>, <span class="at">y=</span><span class="dv">2</span><span class="sc">*</span>c_mean<span class="sc">-</span>pd_mean, <span class="at">yend=</span><span class="dv">2</span><span class="sc">*</span>pd_mean<span class="sc">-</span>c_mean) <span class="sc">+</span></span>
<span id="cb181-13"><a href="data-science.html#cb181-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position=</span><span class="st">&quot;none&quot;</span>)</span>
<span id="cb181-14"><a href="data-science.html#cb181-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb181-15"><a href="data-science.html#cb181-15" aria-hidden="true" tabindex="-1"></a>pd_mean <span class="ot">&lt;-</span> <span class="fu">mean</span>(<span class="fu">filter</span>(gene_exp,<span class="st">`</span><span class="at">Disease Status</span><span class="st">`</span><span class="sc">==</span><span class="st">&quot;Parkinson&#39;s&quot;</span>)<span class="sc">$</span><span class="st">`</span><span class="at">Gene 3</span><span class="st">`</span>)</span>
<span id="cb181-16"><a href="data-science.html#cb181-16" aria-hidden="true" tabindex="-1"></a>c_mean <span class="ot">&lt;-</span> <span class="fu">mean</span>(<span class="fu">filter</span>(gene_exp,<span class="st">`</span><span class="at">Disease Status</span><span class="st">`</span><span class="sc">==</span><span class="st">&quot;Control&quot;</span>)<span class="sc">$</span><span class="st">`</span><span class="at">Gene 3</span><span class="st">`</span>)</span>
<span id="cb181-17"><a href="data-science.html#cb181-17" aria-hidden="true" tabindex="-1"></a>g3 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(gene_exp, <span class="fu">aes</span>(<span class="at">x=</span><span class="st">`</span><span class="at">Disease Status</span><span class="st">`</span>, <span class="at">y=</span><span class="st">`</span><span class="at">Gene 3</span><span class="st">`</span>, <span class="at">color=</span><span class="st">`</span><span class="at">Disease Status</span><span class="st">`</span>)) <span class="sc">+</span></span>
<span id="cb181-18"><a href="data-science.html#cb181-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_beeswarm</span>() <span class="sc">+</span></span>
<span id="cb181-19"><a href="data-science.html#cb181-19" aria-hidden="true" tabindex="-1"></a>  <span class="fu">annotate</span>(<span class="st">&quot;segment&quot;</span>, <span class="at">x=</span><span class="dv">0</span>, <span class="at">xend=</span><span class="dv">3</span>, <span class="at">y=</span><span class="dv">2</span><span class="sc">*</span>c_mean<span class="sc">-</span>pd_mean, <span class="at">yend=</span><span class="dv">2</span><span class="sc">*</span>pd_mean<span class="sc">-</span>c_mean) <span class="sc">+</span></span>
<span id="cb181-20"><a href="data-science.html#cb181-20" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position=</span><span class="st">&quot;none&quot;</span>)</span>
<span id="cb181-21"><a href="data-science.html#cb181-21" aria-hidden="true" tabindex="-1"></a>g1 <span class="sc">|</span> g2 <span class="sc">|</span> g3</span></code></pre></div>
<p><img src="main_files/lin%20model%203%20genes-1.png" width="672" /></p>
<p>We can also compute the corresponding linear model fits, and confirm that the
coefficients agree with the directions observed in the plot, as well as that all
the associations are significant FDR &lt; 0.05:</p>
<div class="sourceCode" id="cb182"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb182-1"><a href="data-science.html#cb182-1" aria-hidden="true" tabindex="-1"></a>fit1 <span class="ot">&lt;-</span> <span class="fu">lm</span>(<span class="st">`</span><span class="at">Gene 1</span><span class="st">`</span> <span class="sc">~</span> <span class="st">`</span><span class="at">Disease Status</span><span class="st">`</span>, <span class="at">data=</span>gene_exp)</span>
<span id="cb182-2"><a href="data-science.html#cb182-2" aria-hidden="true" tabindex="-1"></a>fit2 <span class="ot">&lt;-</span> <span class="fu">lm</span>(<span class="st">`</span><span class="at">Gene 2</span><span class="st">`</span> <span class="sc">~</span> <span class="st">`</span><span class="at">Disease Status</span><span class="st">`</span>, <span class="at">data=</span>gene_exp)</span>
<span id="cb182-3"><a href="data-science.html#cb182-3" aria-hidden="true" tabindex="-1"></a>fit3 <span class="ot">&lt;-</span> <span class="fu">lm</span>(<span class="st">`</span><span class="at">Gene 3</span><span class="st">`</span> <span class="sc">~</span> <span class="st">`</span><span class="at">Disease Status</span><span class="st">`</span>, <span class="at">data=</span>gene_exp)</span>
<span id="cb182-4"><a href="data-science.html#cb182-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb182-5"><a href="data-science.html#cb182-5" aria-hidden="true" tabindex="-1"></a>gene_stats <span class="ot">&lt;-</span> <span class="fu">bind_rows</span>(</span>
<span id="cb182-6"><a href="data-science.html#cb182-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">c</span>(<span class="st">&quot;Gene 1&quot;</span>,<span class="fu">coefficients</span>(fit1),<span class="fu">summary</span>(fit1)<span class="sc">$</span>coefficients[<span class="dv">2</span>,<span class="dv">4</span>]),</span>
<span id="cb182-7"><a href="data-science.html#cb182-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">c</span>(<span class="st">&quot;Gene 2&quot;</span>,<span class="fu">coefficients</span>(fit2),<span class="fu">summary</span>(fit2)<span class="sc">$</span>coefficients[<span class="dv">2</span>,<span class="dv">4</span>]),</span>
<span id="cb182-8"><a href="data-science.html#cb182-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">c</span>(<span class="st">&quot;Gene 3&quot;</span>,<span class="fu">coefficients</span>(fit3),<span class="fu">summary</span>(fit3)<span class="sc">$</span>coefficients[<span class="dv">2</span>,<span class="dv">4</span>])</span>
<span id="cb182-9"><a href="data-science.html#cb182-9" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb182-10"><a href="data-science.html#cb182-10" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(gene_stats) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;Gene&quot;</span>,<span class="st">&quot;Intercept&quot;</span>,<span class="st">&quot;Parkinson&#39;s&quot;</span>,<span class="st">&quot;pvalue&quot;</span>)</span>
<span id="cb182-11"><a href="data-science.html#cb182-11" aria-hidden="true" tabindex="-1"></a>gene_stats<span class="sc">$</span>padj <span class="ot">&lt;-</span> <span class="fu">p.adjust</span>(gene_stats<span class="sc">$</span>pvalue,<span class="at">method=</span><span class="st">&quot;fdr&quot;</span>)</span>
<span id="cb182-12"><a href="data-science.html#cb182-12" aria-hidden="true" tabindex="-1"></a>gene_stats</span></code></pre></div>
<pre><code>## # A tibble: 3 x 5
##   Gene   Intercept        `Parkinson&#39;s`     pvalue                   padj
##   &lt;chr&gt;  &lt;chr&gt;            &lt;chr&gt;             &lt;chr&gt;                   &lt;dbl&gt;
## 1 Gene 1 250.410735540151 6.95982194659045  3.92131170913765e-06 3.92e- 6
## 2 Gene 2 597.763814010886 -93.6291659250254 2.37368730897756e-13 3.56e-13
## 3 Gene 3 334.57788193953  164.094204856583  1.72774902843408e-66 5.18e-66</code></pre>
<p>We have just performed our first elementary differential expression analysis
using a linear model. Specifically, we examined each of our genes for a
statistical relationship with having Parkinson’s disease. The mechanics of this
analysis are beyond the scope of this book, but later when we consider
differential expression analysis packages in <a href="biology-bioinformatics.html#biology-bioinformatics">chapter
7</a> this pattern should be familiar after this example.</p>
<div class="box note">
<p>If you are familiar with <a href="https://en.wikipedia.org/wiki/Logistic_regression">logistic
regression</a>, you might have
wondered why we didn’t model disease status, which is a binary variable, as a
function of gene expression like <code>Disease Status ~ Gene 3</code>. There are several
reasons for this, <a href="https://en.wikipedia.org/wiki/Separation_(statistics)">complete
separation</a> principally
among them. In logistic regression, complete separation occurs when all the
predictor values (e.g. gene expression) for one outcome group are greater or
smaller than the other; i.e. there is no overlap in the values between groups.
This causes logistic regression to fail to converge, leaving these genes with no
statistics even though these genes are potentially the most interesting! There
are methods (e.g. <a href="https://cran.r-project.org/web/packages/logistf/index.html">Firth’s Logistic
Regression</a>) that
overcome this problem, but methods that model gene expression as a function of
other outcome variables were developed first and remain the most popular.</p>
</div>
<div class="box readmore">
<p><a href="https://r4ds.had.co.nz/model-basics.html">Modeling Basics - R for Data Science</a>
<a href="https://en.wikipedia.org/wiki/Least_squares">Least squares regression</a></p>
</div>
</div>
<div id="flavors-of-linear-models" class="section level3" number="6.1.4">
<h3><span class="header-section-number">6.1.4</span> Flavors of Linear Models</h3>
<p>The linear model implemented above is termed <a href="https://en.wikipedia.org/wiki/Linear_regression">linear
regression</a> due to the way it
models the relationship between the predictor variables and the outcome.
Specifically, linear regression makes some strong assumptions about that
relationship that may not always hold for all datasets. To address this
limitation, a more flexible class of linear models called <a href="https://en.wikipedia.org/wiki/Generalized_linear_model">generalized linear
models</a> that allow these
assumptions to be relaxed by changing the mathematical relationship between the
predictors and outcome using a <a href="https://en.wikipedia.org/wiki/Generalized_linear_model#Link_function">link
function</a>,
and/or by mathematically transforming the predictors themselves. Some of the
more common generalized linear models are listed below:</p>
<ul>
<li><strong>Logistic regression</strong> - models a binary outcome (e.g. disease vs control) as
a linear combination of predictor variables by using the <a href="https://en.wikipedia.org/wiki/Logistic_function">logistic
function</a> as the link function.</li>
<li><strong>Multinomial regression</strong> - models a multinomial (i.e. categorical variable
with more than two categories) using a <a href="https://en.wikipedia.org/wiki/Multinomial_logistic_regression">multinomial logit function</a> link
function</li>
<li><strong>Poisson regression</strong> - models an outcome variable that are count data as
a linear combination of predictors using the logarithm as the link function;
this model is commonly used when modeling certain types of high throughput
sequencing data</li>
<li><strong>Negative binomial regression</strong> - also models an outcome variable that are
count data but relaxes some of the assumptions of Poisson regression, namely the
mean-variance equality constraint; negative binomial regression is commonly used
to model gene expression estimated from RNASeq data.</li>
</ul>
<p>Generalized linear models are very flexible, and there are many other types of
these models used in biology and data science. It is important to be aware of
the characteristics of the outcome you wish to model and choose the modeling
method that is most suitable.</p>
<div class="box readmore">
<p><a href="https://en.wikipedia.org/wiki/Generalized_linear_model">Generalized linear models</a></p>
</div>
</div>
<div id="assessing-model-accuracy-." class="section level3" number="6.1.5">
<h3><span class="header-section-number">6.1.5</span> Assessing Model Accuracy .</h3>

</div>
</div>
<div id="statistical-distributions-tests-." class="section level2" number="6.2">
<h2><span class="header-section-number">6.2</span> Statistical Distributions &amp; Tests .</h2>
<div id="statistical-distributions" class="section level3" number="6.2.1">
<h3><span class="header-section-number">6.2.1</span> Statistical Distributions</h3>
<p>By definition, a distribution is a function that shows
<strong>the possible values for a variable and how often they occur</strong>.
Generally, they are divided into two categories: discrete distributions and
continuous distributions.</p>
<div id="discrete-distributions" class="section level4" number="6.2.1.1">
<h4><span class="header-section-number">6.2.1.1</span> Discrete distributions</h4>
<div id="bernoulli-random-trail-and-more" class="section level5" number="6.2.1.1.1">
<h5><span class="header-section-number">6.2.1.1.1</span> Bernoulli random trail and more</h5>
<p>One of the examples for discrete random variable distribution is the <strong>Bernoulli</strong>
function. A Bernoulli trail has only 2 outcomes with probability <span class="math inline">\(p\)</span> and <span class="math inline">\((1-p)\)</span>.
Consider flipping a fair coin, and the random variable X can take value 0 or 1
indicating you get a head or a tail. If it’s a fair coin, we would expect the
Pr{ X = 0 } = Pr{ X = 1 } = 0.5. Or, if we throw a die and we record X = 1 when
we get a six, and X = 0 otherwise, then Pr{ X = 0 } = 5/6 and Pr{ X = 1 } = 1/6.</p>
<p>Now consider a slightly complicated situation: what if we are throwing the die
<span class="math inline">\(n\)</span> times and we would like to analyze the total number of six, say <span class="math inline">\(x\)</span>, we get
during those <span class="math inline">\(n\)</span> throws? Now, this leads us to the <strong>binomial distribution</strong>.
If it’s a fair die, we would say our proportion parameter p = 1/6, which means
the probability we are getting a six is 1/6 for each throw.</p>
<p><span class="math display">\[\begin{equation}
f(x) = \frac {n!} {x!(n-x)!}p^x (1-p) ^{(n-x)}
\end{equation}\]</span></p>
<p>The <strong>geometric</strong> random variable, similar to the binomial, is also from a
sequence of random Bernoulli trials with a constant probability parameter p.
But this time, we define the random variable X as the number of consecutive
failures before the first success. In this case, the probability of x
consecutive failures followed by success on trial x+1 is:</p>
<p><span class="math display">\[\begin{equation}
f(x) = p * (1-p)^x
\end{equation}\]</span></p>
<p>The <strong>negative binomial</strong> distribution goes one step forward. This time, we are
still performing a sequence of independent Bernoulli random trials with a
constant probability of success equal to p. But now we would like to record the
random variable Z to be the total number of failures before we finally get to
the <span class="math inline">\(r^{th}\)</span> success. In other words, when we get to the <span class="math inline">\(r^{th}\)</span> success, we
had x+r Bernoulli random trails, in which x times failed and r times succeeded.</p>
<p><span class="math display">\[\begin{equation}
f(x) = \frac {x+r-1} {r-1} p^r {(1-p)}^x
\end{equation}\]</span></p>
</div>
<div id="poisson" class="section level5" number="6.2.1.1.2">
<h5><span class="header-section-number">6.2.1.1.2</span> Poisson</h5>
<p>The Poisson distribution is used to express the probability of a given number
of events occurring in a fixed interval of time or space, and these events
occur with a known constant mean rate and independently of the time since
the last event. <del>But no one understands this definition.</del></p>
<p>The formula for Poisson distribution is:<br />
<span class="math display">\[\begin{equation}
f(k; \lambda) = Pr(X=k) = \frac {\lambda^k e^{-\lambda}} {k!}
\end{equation}\]</span></p>
<ul>
<li>lambda is the expected value of the random variable X<br />
</li>
<li>k is the number of occurrences<br />
</li>
<li>e is Euler’s number (e=2.71828)<br />
<del>okay, the formula makes it even more confusing.</del></li>
</ul>
<p>Imagine you are working at a mail reception center, and your responsibility
is to receive incoming letters. Assume the number of incoming letters is not
affected by the day of the week or season of the year. You are expected to get
20 letters on average in a day. But, the actual number of letters you receive
each day will not be perfectly 20.<br />
You recorded the number of letters you receive each day in a month (30 days).
In the following plot, each dot represents a day. The x-axis is calender day,
and y-axis is the number of letters you receive on that day. Although on average
you are receiving 20 letters each day, the actual number of letters each day vary a lot.</p>
<div class="sourceCode" id="cb184"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb184-1"><a href="data-science.html#cb184-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">2</span>)</span>
<span id="cb184-2"><a href="data-science.html#cb184-2" aria-hidden="true" tabindex="-1"></a>my_letter <span class="ot">&lt;-</span> <span class="fu">rpois</span>(<span class="at">n =</span> <span class="dv">30</span>, <span class="at">lambda =</span> <span class="dv">20</span>)</span>
<span id="cb184-3"><a href="data-science.html#cb184-3" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(my_letter,</span>
<span id="cb184-4"><a href="data-science.html#cb184-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">main =</span> <span class="st">&quot;Letters received each day&quot;</span>,</span>
<span id="cb184-5"><a href="data-science.html#cb184-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">xlab =</span> <span class="st">&quot;day of the month&quot;</span>, <span class="at">ylab =</span> <span class="st">&quot;number of letters&quot;</span>,</span>
<span id="cb184-6"><a href="data-science.html#cb184-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">pch =</span> <span class="dv">19</span>, <span class="at">col =</span> <span class="st">&quot;royalblue&quot;</span></span>
<span id="cb184-7"><a href="data-science.html#cb184-7" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb184-8"><a href="data-science.html#cb184-8" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="at">a =</span> <span class="dv">20</span>, <span class="at">b =</span> <span class="dv">0</span>, <span class="at">lwd =</span> <span class="dv">2</span>, <span class="at">lty =</span> <span class="dv">3</span>, <span class="at">col =</span> <span class="st">&quot;salmon&quot;</span>)</span></code></pre></div>
<p><img src="main_files/unnamed-chunk-48-1.png" width="672" /></p>
<p>Now, let’s plot the density plot of our data. The x-axis is the number of
letters on a single day, and the y-axis is the probability.</p>
<div class="sourceCode" id="cb185"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb185-1"><a href="data-science.html#cb185-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(<span class="fu">density</span>(my_letter),</span>
<span id="cb185-2"><a href="data-science.html#cb185-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">lwd =</span> <span class="dv">2</span>, <span class="at">col =</span> <span class="st">&quot;royalblue&quot;</span>,</span>
<span id="cb185-3"><a href="data-science.html#cb185-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">main =</span> <span class="st">&quot;Probability of number of letters each day&quot;</span>,</span>
<span id="cb185-4"><a href="data-science.html#cb185-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">xlab =</span> <span class="st">&quot;number of letters&quot;</span></span>
<span id="cb185-5"><a href="data-science.html#cb185-5" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div>
<p><img src="main_files/unnamed-chunk-49-1.png" width="672" /></p>
<p>Since we only have 30 data points, it doesn’t look like a good curve. But,
after we worked at the mail reception for 5000 days, it becomes much closer
to the theoretical Poisson distribution with lambda = 20.</p>
<div class="sourceCode" id="cb186"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb186-1"><a href="data-science.html#cb186-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">3</span>)</span>
<span id="cb186-2"><a href="data-science.html#cb186-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(<span class="fu">density</span>(<span class="fu">rpois</span>(<span class="at">n =</span> <span class="dv">5000</span>, <span class="at">lambda =</span> <span class="dv">20</span>)),</span>
<span id="cb186-3"><a href="data-science.html#cb186-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">lwd =</span> <span class="dv">2</span>, <span class="at">col =</span> <span class="st">&quot;royalblue&quot;</span>,</span>
<span id="cb186-4"><a href="data-science.html#cb186-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">main =</span> <span class="st">&quot;Probability of number of letters each day&quot;</span>,</span>
<span id="cb186-5"><a href="data-science.html#cb186-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">xlab =</span> <span class="st">&quot;number of letters&quot;</span></span>
<span id="cb186-6"><a href="data-science.html#cb186-6" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div>
<p><img src="main_files/unnamed-chunk-50-1.png" width="672" /></p>
<p>Here is the theoretical Poisson distribution with lambda = 20.</p>
<div class="sourceCode" id="cb187"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb187-1"><a href="data-science.html#cb187-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(<span class="fu">dpois</span>(<span class="fu">c</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">40</span>), <span class="at">lambda =</span> <span class="dv">20</span>),</span>
<span id="cb187-2"><a href="data-science.html#cb187-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">lwd =</span> <span class="dv">2</span>, <span class="at">type =</span> <span class="st">&quot;l&quot;</span>, <span class="at">col =</span> <span class="st">&quot;royalblue&quot;</span>,</span>
<span id="cb187-3"><a href="data-science.html#cb187-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">ylab =</span> <span class="st">&quot;probability&quot;</span>, <span class="at">main =</span> <span class="st">&quot;Poisson lambda=20&quot;</span></span>
<span id="cb187-4"><a href="data-science.html#cb187-4" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div>
<p><img src="main_files/unnamed-chunk-51-1.png" width="672" /></p>
<p>If we want to know what’s the probability to receive, for example, 18 letters,<br />
we can use <code>dpois()</code> function.</p>
<div class="sourceCode" id="cb188"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb188-1"><a href="data-science.html#cb188-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dpois</span>(<span class="at">x =</span> <span class="dv">18</span>, <span class="at">lambda =</span> <span class="dv">20</span>)</span></code></pre></div>
<pre><code>## [1] 0.08439355</code></pre>
<p>If we want to know the probability of receiving 18 or less letters,
use <code>ppois()</code> function.</p>
<div class="sourceCode" id="cb190"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb190-1"><a href="data-science.html#cb190-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ppois</span>(<span class="at">q =</span> <span class="dv">18</span>, <span class="at">lambda =</span> <span class="dv">20</span>, <span class="at">lower.tail =</span> T)</span></code></pre></div>
<pre><code>## [1] 0.3814219</code></pre>
<p>It is the cumulative area colored in the following plot:</p>
<div class="sourceCode" id="cb192"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb192-1"><a href="data-science.html#cb192-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(<span class="fu">dpois</span>(<span class="fu">c</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">40</span>), <span class="at">lambda =</span> <span class="dv">20</span>),</span>
<span id="cb192-2"><a href="data-science.html#cb192-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">lwd =</span> <span class="dv">2</span>, <span class="at">type =</span> <span class="st">&quot;l&quot;</span>, <span class="at">col =</span> <span class="st">&quot;royalblue&quot;</span>,</span>
<span id="cb192-3"><a href="data-science.html#cb192-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">ylab =</span> <span class="st">&quot;probability&quot;</span>, <span class="at">main =</span> <span class="st">&quot;Poisson lambda=20&quot;</span></span>
<span id="cb192-4"><a href="data-science.html#cb192-4" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb192-5"><a href="data-science.html#cb192-5" aria-hidden="true" tabindex="-1"></a><span class="fu">polygon</span>(</span>
<span id="cb192-6"><a href="data-science.html#cb192-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">x =</span> <span class="fu">c</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">18</span>, <span class="dv">18</span>),</span>
<span id="cb192-7"><a href="data-science.html#cb192-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">y =</span> <span class="fu">c</span>(<span class="fu">dpois</span>(<span class="fu">c</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">18</span>), <span class="at">lambda =</span> <span class="dv">20</span>), <span class="dv">0</span>),</span>
<span id="cb192-8"><a href="data-science.html#cb192-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">border =</span> <span class="st">&quot;royalblue&quot;</span>, <span class="at">col =</span> <span class="st">&quot;lightblue1&quot;</span></span>
<span id="cb192-9"><a href="data-science.html#cb192-9" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb192-10"><a href="data-science.html#cb192-10" aria-hidden="true" tabindex="-1"></a><span class="fu">segments</span>(<span class="at">x0 =</span> <span class="dv">18</span>, <span class="at">y0 =</span> <span class="dv">0</span>, <span class="at">y1 =</span> <span class="fl">0.08439355</span>, <span class="at">lwd =</span> <span class="dv">2</span>, <span class="at">lty =</span> <span class="dv">2</span>, <span class="at">col =</span> <span class="st">&quot;salmon&quot;</span>)</span></code></pre></div>
<p><img src="main_files/unnamed-chunk-54-1.png" width="672" /></p>
<p><code>qpois()</code> is like the “reverse” of <code>ppois()</code>.</p>
<div class="sourceCode" id="cb193"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb193-1"><a href="data-science.html#cb193-1" aria-hidden="true" tabindex="-1"></a><span class="fu">qpois</span>(<span class="at">p =</span> <span class="fl">0.3814219</span>, <span class="at">lambda =</span> <span class="dv">20</span>)</span></code></pre></div>
<pre><code>## [1] 18</code></pre>
<p>Let’s review the definition of Poisson distribution.<br />
- “these events occur in a fixed interval of time or space,” which is a day;<br />
- “these events occur with a known constant mean rate,” which is 20 letters;<br />
- “independently of the time since the last event,” which means the number of
letters you receive today is independent of the letter you receive tomorrow.
<del>more work today don’t guarantee less work tomorrow, just like in real life</del></p>
<p>Now let’s re-visit the formula.<br />
<span class="math display">\[\begin{equation}
f(k; \lambda) = Pr(X=k) = \frac {\lambda^k e^{-\lambda}} {k!}
\end{equation}\]</span></p>
<ul>
<li>lambda is the expected value of X, which is 20 letters in this example.<br />
</li>
<li>k is the number of occurrences, which is the number of letters you get on a specific day.<br />
</li>
<li>e is Euler’s number (e=2.71828)</li>
</ul>
</div>
</div>
<div id="continuous-distributions" class="section level4" number="6.2.1.2">
<h4><span class="header-section-number">6.2.1.2</span> Continuous distributions</h4>
<p>There are a number of different continuous distributions. Here we will focus
on the <strong>Normal distribution (Gaussian distribution)</strong>. A lot of variables in
real life are normally distributed, common examples include people’s height,
blood pressure, and students’ exam score.</p>
<p>This is what a normal distribution with mean = 0 and standard deviation = 1 looks like:</p>
<div class="sourceCode" id="cb195"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb195-1"><a href="data-science.html#cb195-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">2</span>)</span>
<span id="cb195-2"><a href="data-science.html#cb195-2" aria-hidden="true" tabindex="-1"></a>norm <span class="ot">&lt;-</span> <span class="fu">rnorm</span>(<span class="at">n =</span> <span class="dv">50000</span>, <span class="at">mean =</span> <span class="dv">0</span>, <span class="at">sd =</span> <span class="dv">1</span>)</span>
<span id="cb195-3"><a href="data-science.html#cb195-3" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(<span class="fu">density</span>(norm),</span>
<span id="cb195-4"><a href="data-science.html#cb195-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">main =</span> <span class="st">&quot;A Normal distribution&quot;</span>, <span class="at">xlab =</span> <span class="st">&quot;x&quot;</span>,</span>
<span id="cb195-5"><a href="data-science.html#cb195-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">lwd =</span> <span class="dv">2</span>, <span class="at">col =</span> <span class="st">&quot;royalblue&quot;</span></span>
<span id="cb195-6"><a href="data-science.html#cb195-6" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div>
<p><img src="main_files/unnamed-chunk-56-1.png" width="672" /></p>
<p>Similar as the Poisson distribution above, there are several functions to work
with normal distribution, including <code>rnorm()</code>, <code>dnorm()</code>, <code>pnorm()</code>, and <code>qnorm()</code>.</p>
<p><code>rnorm()</code> is used to draw random data points from a normal distribution with
a given mean and standard deviation.</p>
<div class="sourceCode" id="cb196"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb196-1"><a href="data-science.html#cb196-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">2</span>)</span>
<span id="cb196-2"><a href="data-science.html#cb196-2" aria-hidden="true" tabindex="-1"></a><span class="fu">rnorm</span>(</span>
<span id="cb196-3"><a href="data-science.html#cb196-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">n =</span> <span class="dv">6</span>, <span class="co"># number of data points to draw</span></span>
<span id="cb196-4"><a href="data-science.html#cb196-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">mean =</span> <span class="dv">0</span>, <span class="co"># mean</span></span>
<span id="cb196-5"><a href="data-science.html#cb196-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">sd =</span> <span class="dv">1</span></span>
<span id="cb196-6"><a href="data-science.html#cb196-6" aria-hidden="true" tabindex="-1"></a>) <span class="co"># standard deviation</span></span></code></pre></div>
<pre><code>## [1] -0.89691455  0.18484918  1.58784533 -1.13037567 -0.08025176  0.13242028</code></pre>
<p><code>dnorm()</code> is the density at a given quantile. For instance, in the normal
distribution (mean=0, sd=1) above, the probability density at 0.5 is roughly 0.35.</p>
<div class="sourceCode" id="cb198"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb198-1"><a href="data-science.html#cb198-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dnorm</span>(<span class="at">x =</span> <span class="fl">0.5</span>, <span class="at">mean =</span> <span class="dv">0</span>, <span class="at">sd =</span> <span class="dv">1</span>)</span></code></pre></div>
<pre><code>## [1] 0.3520653</code></pre>
<div class="sourceCode" id="cb200"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb200-1"><a href="data-science.html#cb200-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(<span class="fu">density</span>(norm),</span>
<span id="cb200-2"><a href="data-science.html#cb200-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">main =</span> <span class="st">&quot;A Normal distribution&quot;</span>, <span class="at">xlab =</span> <span class="st">&quot;x&quot;</span>,</span>
<span id="cb200-3"><a href="data-science.html#cb200-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">lwd =</span> <span class="dv">2</span>, <span class="at">col =</span> <span class="st">&quot;royalblue&quot;</span></span>
<span id="cb200-4"><a href="data-science.html#cb200-4" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb200-5"><a href="data-science.html#cb200-5" aria-hidden="true" tabindex="-1"></a><span class="fu">segments</span>(<span class="at">x0 =</span> <span class="fl">0.5</span>, <span class="at">y0 =</span> <span class="dv">0</span>, <span class="at">y1 =</span> <span class="fl">0.3520653</span>, <span class="at">lwd =</span> <span class="dv">2</span>, <span class="at">lty =</span> <span class="dv">3</span>, <span class="at">col =</span> <span class="st">&quot;salmon&quot;</span>)</span>
<span id="cb200-6"><a href="data-science.html#cb200-6" aria-hidden="true" tabindex="-1"></a><span class="fu">segments</span>(<span class="at">x0 =</span> <span class="sc">-</span><span class="dv">5</span>, <span class="at">y0 =</span> <span class="fl">0.3520653</span>, <span class="at">x1 =</span> <span class="fl">0.5</span>, <span class="at">lwd =</span> <span class="dv">2</span>, <span class="at">lty =</span> <span class="dv">3</span>, <span class="at">col =</span> <span class="st">&quot;salmon&quot;</span>)</span>
<span id="cb200-7"><a href="data-science.html#cb200-7" aria-hidden="true" tabindex="-1"></a><span class="fu">points</span>(<span class="at">x =</span> <span class="fl">0.5</span>, <span class="at">y =</span> <span class="fl">0.3520653</span>, <span class="at">cex =</span> <span class="fl">1.5</span>, <span class="at">lwd =</span> <span class="dv">2</span>, <span class="at">col =</span> <span class="st">&quot;red&quot;</span>)</span>
<span id="cb200-8"><a href="data-science.html#cb200-8" aria-hidden="true" tabindex="-1"></a><span class="fu">text</span>(<span class="at">x =</span> <span class="fl">1.1</span>, <span class="at">y =</span> <span class="fl">0.36</span>, <span class="at">labels =</span> <span class="st">&quot;0.3521&quot;</span>, <span class="at">col =</span> <span class="st">&quot;red2&quot;</span>)</span></code></pre></div>
<p><img src="main_files/unnamed-chunk-59-1.png" width="672" /></p>
<p><code>pnorm()</code> gives the distribution function. Or, you can think of it as the
cumulative of the left side of the density function until a given value,
which is the area colored in light blue in the following plot.</p>
<div class="sourceCode" id="cb201"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb201-1"><a href="data-science.html#cb201-1" aria-hidden="true" tabindex="-1"></a><span class="fu">pnorm</span>(<span class="at">q =</span> <span class="fl">0.5</span>, <span class="at">mean =</span> <span class="dv">0</span>, <span class="at">sd =</span> <span class="dv">1</span>)</span></code></pre></div>
<pre><code>## [1] 0.6914625</code></pre>
<div class="sourceCode" id="cb203"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb203-1"><a href="data-science.html#cb203-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(<span class="fu">density</span>(norm),</span>
<span id="cb203-2"><a href="data-science.html#cb203-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">main =</span> <span class="st">&quot;A Normal distribution&quot;</span>, <span class="at">xlab =</span> <span class="st">&quot;x&quot;</span>,</span>
<span id="cb203-3"><a href="data-science.html#cb203-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">lwd =</span> <span class="dv">2</span>, <span class="at">col =</span> <span class="st">&quot;royalblue&quot;</span></span>
<span id="cb203-4"><a href="data-science.html#cb203-4" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb203-5"><a href="data-science.html#cb203-5" aria-hidden="true" tabindex="-1"></a><span class="fu">polygon</span>(</span>
<span id="cb203-6"><a href="data-science.html#cb203-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">x =</span> <span class="fu">c</span>(<span class="fu">density</span>(norm)<span class="sc">$</span>x[<span class="fu">density</span>(norm)<span class="sc">$</span>x <span class="sc">&lt;=</span> <span class="fl">0.5</span>], <span class="fl">0.5</span>),</span>
<span id="cb203-7"><a href="data-science.html#cb203-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">y =</span> <span class="fu">c</span>(<span class="fu">density</span>(norm)<span class="sc">$</span>y[<span class="fu">density</span>(norm)<span class="sc">$</span>x <span class="sc">&lt;=</span> <span class="fl">0.5</span>], <span class="dv">0</span>),</span>
<span id="cb203-8"><a href="data-science.html#cb203-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">border =</span> <span class="st">&quot;royalblue&quot;</span>, <span class="at">col =</span> <span class="st">&quot;lightblue1&quot;</span></span>
<span id="cb203-9"><a href="data-science.html#cb203-9" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb203-10"><a href="data-science.html#cb203-10" aria-hidden="true" tabindex="-1"></a><span class="fu">segments</span>(<span class="at">x0 =</span> <span class="fl">0.5</span>, <span class="at">y0 =</span> <span class="dv">0</span>, <span class="at">y1 =</span> <span class="fl">0.3520653</span>, <span class="at">lwd =</span> <span class="dv">2</span>, <span class="at">lty =</span> <span class="dv">2</span>, <span class="at">col =</span> <span class="st">&quot;salmon&quot;</span>)</span></code></pre></div>
<p><img src="main_files/unnamed-chunk-61-1.png" width="672" /></p>
<p><code>qnorm()</code> gives the quantile function.
You can think of it as the “reverse” of <code>pnorm()</code>. For instance:</p>
<div class="sourceCode" id="cb204"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb204-1"><a href="data-science.html#cb204-1" aria-hidden="true" tabindex="-1"></a><span class="fu">pnorm</span>(<span class="at">q =</span> <span class="fl">0.5</span>, <span class="at">mean =</span> <span class="dv">0</span>, <span class="at">sd =</span> <span class="dv">1</span>)</span></code></pre></div>
<pre><code>## [1] 0.6914625</code></pre>
<div class="sourceCode" id="cb206"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb206-1"><a href="data-science.html#cb206-1" aria-hidden="true" tabindex="-1"></a><span class="fu">qnorm</span>(<span class="at">p =</span> <span class="fl">0.6914625</span>, <span class="at">mean =</span> <span class="dv">0</span>, <span class="at">sd =</span> <span class="dv">1</span>)</span></code></pre></div>
<pre><code>## [1] 0.5000001</code></pre>
<p>Similarly, other distributions such as chi-square distribution, they all have
the same set of functions: <code>rchisq()</code>, <code>dchisq()</code>, <code>pchisq()</code> and <code>qchisq()</code> etc.</p>
</div>
</div>
<div id="statistical-tests-." class="section level3" number="6.2.2">
<h3><span class="header-section-number">6.2.2</span> Statistical Tests .</h3>
<p>What is a statistical test, and when is it appropriate to run one? How do we
know which tests are appropriate in a given situation? What is the difference
between a parametric and non-parametric test? What are some common tests (e.g.
t-test, Chi-sq, hypergeometric, etc), and how do we run them in R?</p>
</div>
<div id="p-values-." class="section level3" number="6.2.3">
<h3><span class="header-section-number">6.2.3</span> p-values .</h3>
<p>What is a p-value? What is the relationship between a p-value and a distribution?
What does a p-value look like when plotted with a distribution? How do we
interpret p-values? How do we compute a p-value using a distribution and a
statistic using R? (Should we mention critical values?)</p>
</div>
<div id="multiple-hypothesis-testing-." class="section level3" number="6.2.4">
<h3><span class="header-section-number">6.2.4</span> Multiple Hypothesis Testing .</h3>
<p>What is multiply hypothesis testing adjustment? Why is it important? What are
some common adjustment methods (Bonferroni (FWER) and Benjamini-Hochberg (FDR))?
How do we interpret adjusted p-values (depends on the adjustment method)? How
do we compute adjusted p-values in R?</p>
</div>
<div id="statistical-power-." class="section level3" number="6.2.5">
<h3><span class="header-section-number">6.2.5</span> Statistical power .</h3>
<p>Conceptually, what is statistical power, and what does it mean? Why is it
important? What is the relationship between a dataset, an analysis, and power?
What aspects of an analysis influence the statistical power of a test? (I don’t
think it’s reasonable to ask students to perform power calculations, just to be
aware that power is a thing and generally what it is).</p>

</div>
</div>
<div id="exploratory-analysis" class="section level2" number="6.3">
<h2><span class="header-section-number">6.3</span> Exploratory Analysis</h2>
<p>So far, we have discussed methods where we chose an explicit model to summarize
our data. However, sometimes we don’t have enough information about our data to
propose reasonable models, as we did earlier when exploring the distribution of
our imaginary gene expression dataset. There may be patterns in the data that
emerge when we compute different summaries and ask whether there is a non-random
structure to how the individual samples or features compare with one another.
The types of methods we use to take a dataset and examine it for structure
without a prefigured hypothesis is called <em>exploratory analysis</em> and is a key
approach to working with biological data.</p>
<div id="principal-component-analysis" class="section level3" number="6.3.1">
<h3><span class="header-section-number">6.3.1</span> Principal Component Analysis</h3>
<p>A very common method of exploratory analysis is <a href="https://en.wikipedia.org/wiki/Principal_component_analysis">principal component
analysis</a> or PCA.
PCA is a statistical procedure that identifies so called <em>directions of
orthogonal variance</em> that capture covariance between different dimensions in a
dataset. Because the approach captures this covariance between features of
arbitrary dimension, it is often used for so-called <a href="https://en.wikipedia.org/wiki/Dimensionality_reduction">dimensionality
reduction</a>, where a
large amount of variance in a dataset with a potentially large number of
dimensions may be expressed in terms of a set of <a href="https://en.wikipedia.org/wiki/Change_of_basis">basis
vectors</a> of smaller dimension.
The mathematical details of this approach are beyond the scope of this book, but
below we explain in general terms the intuition behind what PCA does, and
present an example of how it is used in a biological context.</p>
<p>PCA decomposes a dataset into a set of <a href="https://en.wikipedia.org/wiki/Orthonormal_basis">orthonormal basis
vectors</a> that collectively
capture all the variance in the dataset, where the first basis vector, called
the <em>first principal component</em> explains the largest fraction of the variance,
the second principal component explains the second largest fraction, and so on.
There are always an equal number of principal components as there are dimensions
in the dataset or the number of samples, whichever is smaller. Typically only a
small number of these components are needed to explain most of the variance.</p>
<p>Each principal component is a <span class="math inline">\(p\)</span>-dimensional <a href="https://en.wikipedia.org/wiki/Unit_vector">unit
vector</a> (i.e. a vector of magnitude
1), where <span class="math inline">\(p\)</span> is the number of features in the dataset, and the values are
weights that describe the component’s direction of variance. By multiplying this
component with the values in each sample, we obtain the <em>projection</em> of each
sample with respect to the basis of the component. The projections of each
sample made with each principal component produces a rotation of the dataset in
<span class="math inline">\(p\)</span> dimensional space. The figure below presents a geometric intuition of PCA.</p>
<div class="figure">
<img src="PCA.png" alt="" />
<p class="caption">Principal Component Analysis - Geometric Intuition Illustration</p>
</div>
<p>Many biological datasets, especially those that make genome-wide measurements
like with gene expression assays, have many thousands of features (e.g. genes)
and comparatively few samples. Since PCA can only determine a maximum number of
principal components as the smaller of the number of features or samples, we
will almost always only have as many components as samples. To demonstrate this,
we perform PCA using the
<a href="https://www.rdocumentation.org/packages/stats/versions/3.6.2/topics/prcomp"><code>stats::prcomp()</code></a>
function on an example microarray gene expression intensity dataset:</p>
<div class="sourceCode" id="cb208"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb208-1"><a href="data-science.html#cb208-1" aria-hidden="true" tabindex="-1"></a><span class="co"># intensities contains microarray expression data for ~54k probesets x 20 samples</span></span>
<span id="cb208-2"><a href="data-science.html#cb208-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb208-3"><a href="data-science.html#cb208-3" aria-hidden="true" tabindex="-1"></a><span class="co"># transpose expression values so samples are rows</span></span>
<span id="cb208-4"><a href="data-science.html#cb208-4" aria-hidden="true" tabindex="-1"></a>expr_mat <span class="ot">&lt;-</span> intensities <span class="sc">%&gt;%</span></span>
<span id="cb208-5"><a href="data-science.html#cb208-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(<span class="sc">-</span><span class="fu">c</span>(probeset_id),<span class="at">names_to=</span><span class="st">&quot;sample&quot;</span>) <span class="sc">%&gt;%</span></span>
<span id="cb208-6"><a href="data-science.html#cb208-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_wider</span>(<span class="at">names_from=</span>probeset_id)</span>
<span id="cb208-7"><a href="data-science.html#cb208-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb208-8"><a href="data-science.html#cb208-8" aria-hidden="true" tabindex="-1"></a><span class="co"># PCA expects all features (i.e. probesets) to be mean-centered,</span></span>
<span id="cb208-9"><a href="data-science.html#cb208-9" aria-hidden="true" tabindex="-1"></a><span class="co"># convert to dataframe so we can use rownames</span></span>
<span id="cb208-10"><a href="data-science.html#cb208-10" aria-hidden="true" tabindex="-1"></a>expr_mat_centered <span class="ot">&lt;-</span>  <span class="fu">as.data.frame</span>(</span>
<span id="cb208-11"><a href="data-science.html#cb208-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">lapply</span>(dplyr<span class="sc">::</span><span class="fu">select</span>(expr_mat,<span class="sc">-</span><span class="fu">c</span>(sample)),<span class="cf">function</span>(x) x<span class="sc">-</span><span class="fu">mean</span>(x))</span>
<span id="cb208-12"><a href="data-science.html#cb208-12" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb208-13"><a href="data-science.html#cb208-13" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(expr_mat_centered) <span class="ot">&lt;-</span> expr_mat<span class="sc">$</span>sample</span>
<span id="cb208-14"><a href="data-science.html#cb208-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb208-15"><a href="data-science.html#cb208-15" aria-hidden="true" tabindex="-1"></a><span class="co"># prcomp performs PCA</span></span>
<span id="cb208-16"><a href="data-science.html#cb208-16" aria-hidden="true" tabindex="-1"></a>pca <span class="ot">&lt;-</span> <span class="fu">prcomp</span>(</span>
<span id="cb208-17"><a href="data-science.html#cb208-17" aria-hidden="true" tabindex="-1"></a>  expr_mat_centered,</span>
<span id="cb208-18"><a href="data-science.html#cb208-18" aria-hidden="true" tabindex="-1"></a>  <span class="at">center=</span><span class="cn">FALSE</span>, <span class="co"># prcomp centers features to have mean zero by default, but we already did it</span></span>
<span id="cb208-19"><a href="data-science.html#cb208-19" aria-hidden="true" tabindex="-1"></a>  <span class="at">scale=</span><span class="cn">TRUE</span> <span class="co"># prcomp scales each feature to have unit variance</span></span>
<span id="cb208-20"><a href="data-science.html#cb208-20" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb208-21"><a href="data-science.html#cb208-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb208-22"><a href="data-science.html#cb208-22" aria-hidden="true" tabindex="-1"></a><span class="co"># the str() function prints the structure of its argument</span></span>
<span id="cb208-23"><a href="data-science.html#cb208-23" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(pca)</span></code></pre></div>
<pre><code>## List of 5
##  $ sdev    : num [1:20] 101.5 81 71.2 61.9 57.9 ...
##  $ rotation: num [1:54675, 1:20] 0.00219 0.00173 -0.00313 0.00465 0.0034 ...
##   ..- attr(*, &quot;dimnames&quot;)=List of 2
##   .. ..$ : chr [1:54675] &quot;X1007_s_at&quot; &quot;X1053_at&quot; &quot;X117_at&quot; &quot;X121_at&quot; ...
##   .. ..$ : chr [1:20] &quot;PC1&quot; &quot;PC2&quot; &quot;PC3&quot; &quot;PC4&quot; ...
##  $ center  : logi FALSE
##  $ scale   : Named num [1:54675] 0.379 0.46 0.628 0.272 0.196 ...
##   ..- attr(*, &quot;names&quot;)= chr [1:54675] &quot;X1007_s_at&quot; &quot;X1053_at&quot; &quot;X117_at&quot; &quot;X121_at&quot; ...
##  $ x       : num [1:20, 1:20] -67.94 186.14 6.07 -70.72 9.58 ...
##   ..- attr(*, &quot;dimnames&quot;)=List of 2
##   .. ..$ : chr [1:20] &quot;A1&quot; &quot;A2&quot; &quot;A3&quot; &quot;A4&quot; ...
##   .. ..$ : chr [1:20] &quot;PC1&quot; &quot;PC2&quot; &quot;PC3&quot; &quot;PC4&quot; ...
##  - attr(*, &quot;class&quot;)= chr &quot;prcomp&quot;</code></pre>
<p>The result of <code>prcomp()</code> is a list with five members:</p>
<ul>
<li><code>sdev</code> - the standard deviation (i.e. the square root of the variance) for each component</li>
<li><code>rotation</code> - a matrix where the principal components are in columns</li>
<li><code>x</code> - the projections of the original data, aka the rotated data matrix</li>
<li><code>center</code> - if <code>center=TRUE</code> was passed, a vector of feature means</li>
<li><code>scale</code> - if <code>scale=TRUE</code> was passed, a vector of the feature variances</li>
</ul>
<p>Recall that each principal component explains a fraction of the overall variance
in the dataset. The <code>sdev</code> variable returned by <code>prcomp()</code> may be used to first
calculate the variance explained by each component by squaring it, then dividing
by the sum:</p>
<div class="sourceCode" id="cb210"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb210-1"><a href="data-science.html#cb210-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(patchwork)</span>
<span id="cb210-2"><a href="data-science.html#cb210-2" aria-hidden="true" tabindex="-1"></a>pca_var <span class="ot">&lt;-</span> <span class="fu">tibble</span>(</span>
<span id="cb210-3"><a href="data-science.html#cb210-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">PC=</span><span class="fu">factor</span>(<span class="fu">str_c</span>(<span class="st">&quot;PC&quot;</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">20</span>),<span class="fu">str_c</span>(<span class="st">&quot;PC&quot;</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">20</span>)),</span>
<span id="cb210-4"><a href="data-science.html#cb210-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">Variance=</span>pca<span class="sc">$</span>sdev<span class="sc">**</span><span class="dv">2</span>,</span>
<span id="cb210-5"><a href="data-science.html#cb210-5" aria-hidden="true" tabindex="-1"></a>  <span class="st">`</span><span class="at">% Explained Variance</span><span class="st">`</span><span class="ot">=</span>Variance<span class="sc">/</span><span class="fu">sum</span>(Variance)<span class="sc">*</span><span class="dv">100</span>,</span>
<span id="cb210-6"><a href="data-science.html#cb210-6" aria-hidden="true" tabindex="-1"></a>  <span class="st">`</span><span class="at">Cumulative % Explained Variance</span><span class="st">`</span><span class="ot">=</span><span class="fu">cumsum</span>(<span class="st">`</span><span class="at">% Explained Variance</span><span class="st">`</span>)</span>
<span id="cb210-7"><a href="data-science.html#cb210-7" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb210-8"><a href="data-science.html#cb210-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb210-9"><a href="data-science.html#cb210-9" aria-hidden="true" tabindex="-1"></a>exp_g <span class="ot">&lt;-</span> pca_var <span class="sc">%&gt;%</span></span>
<span id="cb210-10"><a href="data-science.html#cb210-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x=</span>PC,<span class="at">y=</span><span class="st">`</span><span class="at">% Explained Variance</span><span class="st">`</span>,<span class="at">group=</span><span class="dv">1</span>)) <span class="sc">+</span></span>
<span id="cb210-11"><a href="data-science.html#cb210-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb210-12"><a href="data-science.html#cb210-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_line</span>() <span class="sc">+</span></span>
<span id="cb210-13"><a href="data-science.html#cb210-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.text.x=</span><span class="fu">element_text</span>(<span class="at">angle=</span><span class="dv">90</span>,<span class="at">hjust=</span><span class="dv">0</span>,<span class="at">vjust=</span><span class="fl">0.5</span>)) <span class="co"># make x labels rotate 90 degrees</span></span>
<span id="cb210-14"><a href="data-science.html#cb210-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb210-15"><a href="data-science.html#cb210-15" aria-hidden="true" tabindex="-1"></a>cum_g <span class="ot">&lt;-</span> pca_var <span class="sc">%&gt;%</span></span>
<span id="cb210-16"><a href="data-science.html#cb210-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x=</span>PC,<span class="at">y=</span><span class="st">`</span><span class="at">Cumulative % Explained Variance</span><span class="st">`</span>,<span class="at">group=</span><span class="dv">1</span>)) <span class="sc">+</span></span>
<span id="cb210-17"><a href="data-science.html#cb210-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb210-18"><a href="data-science.html#cb210-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_line</span>() <span class="sc">+</span></span>
<span id="cb210-19"><a href="data-science.html#cb210-19" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylim</span>(<span class="dv">0</span>,<span class="dv">100</span>) <span class="sc">+</span> <span class="co"># set y limits to [0,100]</span></span>
<span id="cb210-20"><a href="data-science.html#cb210-20" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.text.x=</span><span class="fu">element_text</span>(<span class="at">angle=</span><span class="dv">90</span>,<span class="at">hjust=</span><span class="dv">0</span>,<span class="at">vjust=</span><span class="fl">0.5</span>))</span>
<span id="cb210-21"><a href="data-science.html#cb210-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb210-22"><a href="data-science.html#cb210-22" aria-hidden="true" tabindex="-1"></a>exp_g <span class="sc">|</span> cum_g</span></code></pre></div>
<p><img src="main_files/pca%20var-1.png" width="672" /></p>
<p>The first component explains nearly 20% of all variance in the dataset, followed
by the second component with about 12%, the third component 9%, and so on. The
cumulative variance plot on the right shows that the top 15 components are
required to capture 90% of the variance in the dataset. This suggests that each
sample contributes a significant amount of variance to the overall dataset, but
that there are still some features that covary among them.</p>
<p>An important use of PCA results is the identification of outlier samples. This
is accomplished by plotting the projections of each sample and examining the
result by eye to identify samples that are “far away” from the other samples.
This is usually done by inspection and outlier samples chosen subjectively;
there are no general rules to decide when a sample is an outlier by this method.
Below the projections for components 1 and 2 are plotted against each other as a
scatter plot:</p>
<div class="sourceCode" id="cb211"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb211-1"><a href="data-science.html#cb211-1" aria-hidden="true" tabindex="-1"></a><span class="fu">as_tibble</span>(pca<span class="sc">$</span>x) <span class="sc">%&gt;%</span></span>
<span id="cb211-2"><a href="data-science.html#cb211-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x=</span>PC1,<span class="at">y=</span>PC2)) <span class="sc">+</span></span>
<span id="cb211-3"><a href="data-science.html#cb211-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>()</span></code></pre></div>
<p><img src="main_files/pca%20pairwise-1.png" width="672" /></p>
<p>By eye, no samples appear to be obvious outliers. However, this plot is just one
of many pairs of projections. We could plot all pairs of the first six
components using the <a href="https://ggobi.github.io/ggally/reference/ggpairs.html"><code>ggpairs() function</code></a> in the
<a href="https://ggobi.github.io/ggally/">GGally</a> package:</p>
<div class="sourceCode" id="cb212"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb212-1"><a href="data-science.html#cb212-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(GGally)</span>
<span id="cb212-2"><a href="data-science.html#cb212-2" aria-hidden="true" tabindex="-1"></a><span class="fu">as_tibble</span>(pca<span class="sc">$</span>x) <span class="sc">%&gt;%</span></span>
<span id="cb212-3"><a href="data-science.html#cb212-3" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">select</span>(PC1<span class="sc">:</span>PC6) <span class="sc">%&gt;%</span></span>
<span id="cb212-4"><a href="data-science.html#cb212-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggpairs</span>()</span></code></pre></div>
<p><img src="main_files/pca%20pairs-1.png" width="672" /></p>
<p>This is already nearly unintelligible and mostly uninformative. While it is very
common for projections for pairs of components to be plotted, an alternative way
to visualize projections across all samples is by plotting the distribution of
projections for each component using a <a href="#beeswarm-plots">beeswarm plot</a>:</p>
<div class="sourceCode" id="cb213"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb213-1"><a href="data-science.html#cb213-1" aria-hidden="true" tabindex="-1"></a><span class="fu">as_tibble</span>(pca<span class="sc">$</span>x) <span class="sc">%&gt;%</span></span>
<span id="cb213-2"><a href="data-science.html#cb213-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(<span class="fu">everything</span>(),<span class="at">names_to=</span><span class="st">&quot;PC&quot;</span>,<span class="at">values_to=</span><span class="st">&quot;projection&quot;</span>) <span class="sc">%&gt;%</span></span>
<span id="cb213-3"><a href="data-science.html#cb213-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">PC=</span><span class="fu">fct_relevel</span>(PC,<span class="fu">str_c</span>(<span class="st">&quot;PC&quot;</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">20</span>))) <span class="sc">%&gt;%</span></span>
<span id="cb213-4"><a href="data-science.html#cb213-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x=</span>PC,<span class="at">y=</span>projection)) <span class="sc">+</span></span>
<span id="cb213-5"><a href="data-science.html#cb213-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_beeswarm</span>() <span class="sc">+</span> <span class="fu">labs</span>(<span class="at">title=</span><span class="st">&quot;PCA Projection Plot&quot;</span>) <span class="sc">+</span></span>
<span id="cb213-6"><a href="data-science.html#cb213-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.text.x=</span><span class="fu">element_text</span>(<span class="at">angle=</span><span class="dv">90</span>,<span class="at">hjust=</span><span class="dv">0</span>,<span class="at">vjust=</span><span class="fl">0.5</span>))</span></code></pre></div>
<p><img src="main_files/pca%20proj%20beeswarm-1.png" width="672" /></p>
<p>Now we can see all the projections for all components in the same plot, although
we cannot see how they relate to one another.</p>
<p>These projection plots may become more useful if we layer on additional
information about our samples. There are two types of samples in our dataset (
labelled <code>A</code> and <code>C</code>. We can color our pairwise scatter plot by type like so:</p>
<div class="sourceCode" id="cb214"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb214-1"><a href="data-science.html#cb214-1" aria-hidden="true" tabindex="-1"></a><span class="fu">as_tibble</span>(pca<span class="sc">$</span>x) <span class="sc">%&gt;%</span></span>
<span id="cb214-2"><a href="data-science.html#cb214-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">type=</span>stringr<span class="sc">::</span><span class="fu">str_sub</span>(<span class="fu">rownames</span>(pca<span class="sc">$</span>x),<span class="dv">1</span>,<span class="dv">1</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb214-3"><a href="data-science.html#cb214-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x=</span>PC1,<span class="at">y=</span>PC2,<span class="at">color=</span>type)) <span class="sc">+</span></span>
<span id="cb214-4"><a href="data-science.html#cb214-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>()</span></code></pre></div>
<p><img src="main_files/pca%20color%20scatter-1.png" width="672" /></p>
<p>Little pattern is obvious from the plot, but we can plot pairs of components as
before but now with type information:</p>
<div class="sourceCode" id="cb215"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb215-1"><a href="data-science.html#cb215-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(GGally)</span>
<span id="cb215-2"><a href="data-science.html#cb215-2" aria-hidden="true" tabindex="-1"></a><span class="fu">as_tibble</span>(pca<span class="sc">$</span>x) <span class="sc">%&gt;%</span></span>
<span id="cb215-3"><a href="data-science.html#cb215-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">type=</span>stringr<span class="sc">::</span><span class="fu">str_sub</span>(<span class="fu">rownames</span>(pca<span class="sc">$</span>x),<span class="dv">1</span>,<span class="dv">1</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb215-4"><a href="data-science.html#cb215-4" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">select</span>(<span class="fu">c</span>(type,PC1<span class="sc">:</span>PC6)) <span class="sc">%&gt;%</span></span>
<span id="cb215-5"><a href="data-science.html#cb215-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggpairs</span>(<span class="at">columns=</span><span class="dv">1</span><span class="sc">:</span><span class="dv">6</span>,<span class="at">mapping=</span><span class="fu">aes</span>(<span class="at">fill=</span>type))</span></code></pre></div>
<p><img src="main_files/pca%20pairs%20type-1.png" width="672" /></p>
<p>Examining PC3 and PC4, we now observe that there may indeed be some genes that
distinguish between types based on the separation of projection scores for the
two types. Finally, we can also color our beeswarm plot by type:</p>
<div class="sourceCode" id="cb216"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb216-1"><a href="data-science.html#cb216-1" aria-hidden="true" tabindex="-1"></a><span class="fu">as_tibble</span>(pca<span class="sc">$</span>x) <span class="sc">%&gt;%</span></span>
<span id="cb216-2"><a href="data-science.html#cb216-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(</span>
<span id="cb216-3"><a href="data-science.html#cb216-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">sample=</span><span class="fu">rownames</span>(pca<span class="sc">$</span>x),</span>
<span id="cb216-4"><a href="data-science.html#cb216-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">type=</span>stringr<span class="sc">::</span><span class="fu">str_sub</span>(sample,<span class="dv">1</span>,<span class="dv">1</span>)</span>
<span id="cb216-5"><a href="data-science.html#cb216-5" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">%&gt;%</span></span>
<span id="cb216-6"><a href="data-science.html#cb216-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(PC1<span class="sc">:</span>PC20,<span class="at">names_to=</span><span class="st">&quot;PC&quot;</span>,<span class="at">values_to=</span><span class="st">&quot;projection&quot;</span>) <span class="sc">%&gt;%</span></span>
<span id="cb216-7"><a href="data-science.html#cb216-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">PC=</span><span class="fu">fct_relevel</span>(PC,<span class="fu">str_c</span>(<span class="st">&quot;PC&quot;</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">20</span>))) <span class="sc">%&gt;%</span></span>
<span id="cb216-8"><a href="data-science.html#cb216-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x=</span>PC,<span class="at">y=</span>projection,<span class="at">color=</span>type)) <span class="sc">+</span></span>
<span id="cb216-9"><a href="data-science.html#cb216-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_beeswarm</span>() <span class="sc">+</span> <span class="fu">labs</span>(<span class="at">title=</span><span class="st">&quot;PCA Projection Plot&quot;</span>) <span class="sc">+</span></span>
<span id="cb216-10"><a href="data-science.html#cb216-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.text.x=</span><span class="fu">element_text</span>(<span class="at">angle=</span><span class="dv">90</span>,<span class="at">hjust=</span><span class="dv">0</span>,<span class="at">vjust=</span><span class="fl">0.5</span>))</span></code></pre></div>
<p><img src="main_files/pca%20proj%20beeswarm%20type-1.png" width="672" /></p>
<p>These approaches to plotting the results of a PCA are complementary, and all
may be useful in understanding the structure of a dataset.</p>
</div>
<div id="cluster-analysis" class="section level3" number="6.3.2">
<h3><span class="header-section-number">6.3.2</span> Cluster Analysis</h3>
<p>Cluster analysis or simple clustering is the task of grouping objects together
such that objects within the same group, or <em>cluster</em>, are more similar to each
other than to objects in other clusters. It is a type of <a href="https://en.wikipedia.org/wiki/Exploratory_data_analysis">exploratory data
analysis</a> that seeks to
identify structure or organization in a dataset without making strong
assumptions about the data or testing a specific hypothesis. Many different
<a href="https://en.wikipedia.org/wiki/Category:Cluster_analysis_algorithms">clustering algorithms</a>
have been developed that employ different computational strategies and are
designed for data with different types and properties. The choice of algorithm
is often dependent upon not only the type of data being clustered but also the
comparative performance of different clustering algorithms applied to the same
data. Different algorithms may identify different types of patterns, and so no
one clustering method is better than any other in general.</p>
<p>The goal of clustering may be easily illustrated with the following plot:</p>
<div class="sourceCode" id="cb217"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb217-1"><a href="data-science.html#cb217-1" aria-hidden="true" tabindex="-1"></a>unclustered <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(well_clustered_data, <span class="fu">aes</span>(<span class="at">x=</span>f1,<span class="at">y=</span>f2)) <span class="sc">+</span></span>
<span id="cb217-2"><a href="data-science.html#cb217-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb217-3"><a href="data-science.html#cb217-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title=</span><span class="st">&quot;Unclustered&quot;</span>)</span>
<span id="cb217-4"><a href="data-science.html#cb217-4" aria-hidden="true" tabindex="-1"></a>clustered <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(well_clustered_data, <span class="fu">aes</span>(<span class="at">x=</span>f1,<span class="at">y=</span>f2,<span class="at">color=</span>cluster,<span class="at">shape=</span>cluster)) <span class="sc">+</span></span>
<span id="cb217-5"><a href="data-science.html#cb217-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb217-6"><a href="data-science.html#cb217-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title=</span><span class="st">&quot;Clustered&quot;</span>)</span>
<span id="cb217-7"><a href="data-science.html#cb217-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb217-8"><a href="data-science.html#cb217-8" aria-hidden="true" tabindex="-1"></a>unclustered <span class="sc">|</span> clustered</span></code></pre></div>
<p><img src="main_files/clustering%20goal-1.png" width="672" /></p>
<p>That is, the goal of clustering is to take unlabeled data that has some
structure like on the left and label data points that are similar to one another
to be in the same cluster. A full treatment of cluster analysis is beyond the
scope of this book, but below we discuss a few of the most common and general
clustering algorithms used in biology and bioinformatics.</p>
<div id="hierarchical-clustering" class="section level4" number="6.3.2.1">
<h4><span class="header-section-number">6.3.2.1</span> Hierarchical Clustering</h4>
<p>Hierarchical clustering is a form of clustering that clusters data points
together in nested, or <em>hierarchical</em>, groups based on their dissimilarity from
one another. There are two broad strategies to accomplish this:</p>
<ul>
<li><strong>Agglomerative</strong>: all data points start in their own groups, and groups are
iteratively merged hierarchically into larger groups based on their similarity
until all data points have been added to a group</li>
<li><strong>Divisive</strong>: all data points start in the same group, and are recursively
split into smaller groups based on their dissimilarity</li>
</ul>
<p>Whichever approach is used, the critical step in clustering is choosing the
<a href="https://en.wikipedia.org/wiki/Metric_(mathematics)">distance function</a> that
quantifies how dissimilar two data points are. Distance functions, or <em>metrics</em>
are non-negative real numbers whose magnitude is proportional to some notion of
distance between a pair of points. There are <a href="https://en.wikipedia.org/wiki/Metric_(mathematics)#Generalized_metrics">many different distance
functions</a>
to choose from, and the choice is made based on the type of data being
clustered. For numeric data, <a href="https://en.wikipedia.org/wiki/Euclidean_distance">euclidean
distance</a>, which corresponds
to the length of a line segment drawn between two points in any <span class="math inline">\(n\)</span>-dimensional
<a href="https://en.wikipedia.org/wiki/Euclidean_space">Euclidean space</a> is often a
reasonable choice.</p>
<p>Because the divisive strategy is conceptually the inverse of
agglomerative clustering, we will limit our description to agglomerative
clustering. Once the distance function has been chosen, distances between all
pairs of points are computed and the two nearest points are merged into a group.
The new group of points is used for computing distances to all other points as a
set using a <em>linkage function</em>
which defines how the group is summarized into a single point for the purposes
of distance calculation. The linkage function is what distinguishes different
clustering algorithms, which include:</p>
<ul>
<li><a href="https://en.wikipedia.org/wiki/Single-linkage_clustering">Single-linkage</a> - distance between two groups is computed as the distance between the two nearest members of the groups</li>
<li><a href="https://en.wikipedia.org/wiki/Complete-linkage_clustering">Complete-linkage</a> - distance between two groups is computed as the distance between the two farthest members of the groups</li>
<li><a href="https://en.wikipedia.org/wiki/UPGMA">Unweighted Pair Group Method with Arithmetic mean (UPGMA)</a> - distance between two groups is the average distance of all pairs of points between groups</li>
<li><a href="https://en.wikipedia.org/wiki/WPGMA">Weighted Pair Group Method with Arithmetic mean (WPGMA)</a> - similar to UPGMA, but weights distances from pairs of groups evenly when merging</li>
</ul>
<p>The choice of linkage function (and therefore clustering algorithm) should be
determined based on knowledge of the dataset and assessment of clustering
performance. In general, there is no one linkage function that will always
perform well.</p>
<p>The following figure illustrates a simple example of clustering a 1 dimensional
set of points using WPGMA:</p>
<div class="figure">
<img src="hierarchical_clustering.png" alt="" />
<p class="caption">Conceptual illustration of agglomerative hierarchical clustering</p>
</div>
<p>Below we will cluster the synthetic dataset introduced above in R:</p>
<div class="sourceCode" id="cb218"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb218-1"><a href="data-science.html#cb218-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(well_clustered_data, <span class="fu">aes</span>(<span class="at">x=</span>f1,<span class="at">y=</span>f2)) <span class="sc">+</span></span>
<span id="cb218-2"><a href="data-science.html#cb218-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb218-3"><a href="data-science.html#cb218-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title=</span><span class="st">&quot;Unclustered&quot;</span>)</span></code></pre></div>
<p><img src="main_files/hclust%20data-1.png" width="672" /></p>
<p>This synthetic dataset has two distinct groups of samples drawn from
multivariate normal samples. To hierarchically cluster these samples, we use the
<a href="https://www.rdocumentation.org/packages/stats/versions/3.6.2/topics/dist"><code>dist()</code></a>
and <a href="https://www.rdocumentation.org/packages/stats/versions/3.6.2/topics/hclust"><code>hclust()</code></a>
functions:</p>
<div class="sourceCode" id="cb219"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb219-1"><a href="data-science.html#cb219-1" aria-hidden="true" tabindex="-1"></a><span class="co"># compute all pairwise distances using euclidean distance as the distance metric</span></span>
<span id="cb219-2"><a href="data-science.html#cb219-2" aria-hidden="true" tabindex="-1"></a>euc_dist <span class="ot">&lt;-</span> <span class="fu">dist</span>(dplyr<span class="sc">::</span><span class="fu">select</span>(well_clustered_data,<span class="sc">-</span>ID))</span>
<span id="cb219-3"><a href="data-science.html#cb219-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb219-4"><a href="data-science.html#cb219-4" aria-hidden="true" tabindex="-1"></a><span class="co"># produce a clustering of the data using the hclust for hierarchical clustering</span></span>
<span id="cb219-5"><a href="data-science.html#cb219-5" aria-hidden="true" tabindex="-1"></a>hc <span class="ot">&lt;-</span> <span class="fu">hclust</span>(euc_dist, <span class="at">method=</span><span class="st">&quot;ave&quot;</span>)</span>
<span id="cb219-6"><a href="data-science.html#cb219-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb219-7"><a href="data-science.html#cb219-7" aria-hidden="true" tabindex="-1"></a><span class="co"># add ID as labels to the clustering object</span></span>
<span id="cb219-8"><a href="data-science.html#cb219-8" aria-hidden="true" tabindex="-1"></a>hc<span class="sc">$</span>labels <span class="ot">&lt;-</span> well_clustered_data<span class="sc">$</span>ID</span>
<span id="cb219-9"><a href="data-science.html#cb219-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb219-10"><a href="data-science.html#cb219-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb219-11"><a href="data-science.html#cb219-11" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(hc)</span></code></pre></div>
<pre><code>## List of 7
##  $ merge      : int [1:59, 1:2] -50 -33 -48 -46 -44 -21 -11 -4 -36 -3 ...
##  $ height     : num [1:59] 0.00429 0.08695 0.23536 0.24789 0.25588 ...
##  $ order      : int [1:60] 8 9 7 6 19 18 11 16 4 14 ...
##  $ labels     : chr [1:60] &quot;A1&quot; &quot;A2&quot; &quot;A3&quot; &quot;A4&quot; ...
##  $ method     : chr &quot;average&quot;
##  $ call       : language hclust(d = euc_dist, method = &quot;ave&quot;)
##  $ dist.method: chr &quot;euclidean&quot;
##  - attr(*, &quot;class&quot;)= chr &quot;hclust&quot;</code></pre>
<p>The <code>hclust()</code> return object describes the clustering as a tree that can be
visualized using a <a href="data-visualization.html#dendrograms">dendrogram</a>:</p>
<div class="sourceCode" id="cb221"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb221-1"><a href="data-science.html#cb221-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggdendro)</span>
<span id="cb221-2"><a href="data-science.html#cb221-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ggdendrogram</span>(hc)</span></code></pre></div>
<p><img src="main_files/hclust%20dendro-1.png" width="672" />
Our data do indeed cluster well, where all samples from the same group cluster
together perfectly. See the <a href="data-visualization.html#dendrograms">dendrogram</a> section in the <a href="#data-viz">data
vizualization chapter</a> for more detail on how to plot clustering
results.</p>
<p>It is sometimes desirable to use split hierarchical clustering into groups based
on their pattern. In the above clustering, three discrete clusters corresponding
sample groups are clearly visible. If we wished to separate these three groups,
we can use the
<a href="https://www.rdocumentation.org/packages/stats/versions/3.6.2/topics/cutree"><code>cutree</code></a>
to divide the tree into three groups using <code>k=3</code>:</p>
<div class="sourceCode" id="cb222"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb222-1"><a href="data-science.html#cb222-1" aria-hidden="true" tabindex="-1"></a>labels <span class="ot">&lt;-</span> <span class="fu">cutree</span>(hc,<span class="at">k=</span><span class="dv">3</span>)</span>
<span id="cb222-2"><a href="data-science.html#cb222-2" aria-hidden="true" tabindex="-1"></a>labels</span></code></pre></div>
<pre><code>##  A1  A2  A3  A4  A5  A6  A7  A8  A9 A10 A11 A12 A13 A14 A15 A16 A17 A18 A19 A20 
##   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1 
##  B1  B2  B3  B4  B5  B6  B7  B8  B9 B10 B11 B12 B13 B14 B15 B16 B17 B18 B19 B20 
##   2   2   2   2   2   2   2   2   2   2   2   2   2   2   2   2   2   2   2   2 
##  C1  C2  C3  C4  C5  C6  C7  C8  C9 C10 C11 C12 C13 C14 C15 C16 C17 C18 C19 C20 
##   3   3   3   3   3   3   3   3   3   3   3   3   3   3   3   3   3   3   3   3</code></pre>
<p>We can then use these samples and labels to color our original plot as desired:</p>
<div class="sourceCode" id="cb224"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb224-1"><a href="data-science.html#cb224-1" aria-hidden="true" tabindex="-1"></a><span class="co"># we turn our labels into a tibble so we can join them with the original tibble</span></span>
<span id="cb224-2"><a href="data-science.html#cb224-2" aria-hidden="true" tabindex="-1"></a>well_clustered_data <span class="sc">%&gt;%</span></span>
<span id="cb224-3"><a href="data-science.html#cb224-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">left_join</span>(</span>
<span id="cb224-4"><a href="data-science.html#cb224-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">tibble</span>(</span>
<span id="cb224-5"><a href="data-science.html#cb224-5" aria-hidden="true" tabindex="-1"></a>      <span class="at">ID=</span><span class="fu">names</span>(labels),</span>
<span id="cb224-6"><a href="data-science.html#cb224-6" aria-hidden="true" tabindex="-1"></a>      <span class="at">label=</span><span class="fu">as.factor</span>(labels)</span>
<span id="cb224-7"><a href="data-science.html#cb224-7" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb224-8"><a href="data-science.html#cb224-8" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">%&gt;%</span></span>
<span id="cb224-9"><a href="data-science.html#cb224-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x=</span>f1,<span class="at">y=</span>f2,<span class="at">color=</span>label)) <span class="sc">+</span></span>
<span id="cb224-10"><a href="data-science.html#cb224-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb224-11"><a href="data-science.html#cb224-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title=</span><span class="st">&quot;Clustered&quot;</span>)</span></code></pre></div>
<p><img src="main_files/hclust%20cutree%20color-1.png" width="672" /></p>
<p>We were able to recover the correct clustering because this dataset was easy to
cluster by construction. Real data are seldom so well behaved.</p>
</div>
</div>
<div id="k-means-." class="section level3" number="6.3.3">
<h3><span class="header-section-number">6.3.3</span> k-means .</h3>
</div>
<div id="others" class="section level3" number="6.3.4">
<h3><span class="header-section-number">6.3.4</span> Others</h3>

</div>
</div>
<div id="network-analysis-." class="section level2" number="6.4">
<h2><span class="header-section-number">6.4</span> Network Analysis .</h2>
<p>What is a network? What is network analysis? What are some example network
analysis applications in biology and bioinformatics (there is a full ksection on
[Biological Networks] in the bio chapter, these are just illustrative examples)?
How do we represent networks in R, and how do we analyze them? ([Network
visualization] is covered in the data viz chapter, might be helpful to write
these two sections together)</p>

</div>
</div>
<h3>References</h3>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-Box1976-rd" class="csl-entry">
Box, George E P. 1976. <span>“Science and Statistics.”</span> <em>J. Am. Stat. Assoc.</em> 71 (356): 791–99.
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="data-wrangling.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="data-visualization.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
