<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>6 Data Science | BF591 - R for Biological Sciences</title>
  <meta name="description" content="6 Data Science | BF591 - R for Biological Sciences" />
  <meta name="generator" content="bookdown 0.24 and GitBook 2.6.7" />

  <meta property="og:title" content="6 Data Science | BF591 - R for Biological Sciences" />
  <meta property="og:type" content="book" />
  
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="6 Data Science | BF591 - R for Biological Sciences" />
  
  
  




  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="data-wrangling.html"/>
<link rel="next" href="biology-bioinformatics.html"/>
<script src="libs/header-attrs-2.11/header-attrs.js"></script>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.0.1/anchor-sections.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.0.1/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Syllabus</a>
<ul>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#course-schedule"><i class="fa fa-check"></i>Course Schedule</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#instructors"><i class="fa fa-check"></i>Instructors</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#office-hours"><i class="fa fa-check"></i>Office Hours</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#course-values-and-policies"><i class="fa fa-check"></i>Course Values and Policies</a></li>
</ul></li>
<li class="chapter" data-level="1" data-path="introduction.html"><a href="introduction.html"><i class="fa fa-check"></i><b>1</b> Introduction</a>
<ul>
<li class="chapter" data-level="1.1" data-path="introduction.html"><a href="introduction.html#who-this-book-is-for"><i class="fa fa-check"></i><b>1.1</b> Who This Book Is For</a></li>
<li class="chapter" data-level="1.2" data-path="introduction.html"><a href="introduction.html#a-note-about-reinventing-the-wheel"><i class="fa fa-check"></i><b>1.2</b> A Note About Reinventing the Wheel</a></li>
<li class="chapter" data-level="1.3" data-path="introduction.html"><a href="introduction.html#sources-and-references"><i class="fa fa-check"></i><b>1.3</b> Sources and References</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="data-bio.html"><a href="data-bio.html"><i class="fa fa-check"></i><b>2</b> Data in Biology</a>
<ul>
<li class="chapter" data-level="2.1" data-path="data-bio.html"><a href="data-bio.html#a-brief-history-of-data-in-molecular-biology"><i class="fa fa-check"></i><b>2.1</b> A Brief History of Data in Molecular Biology</a></li>
<li class="chapter" data-level="2.2" data-path="data-bio.html"><a href="data-bio.html#biology-as-a-mature-data-science"><i class="fa fa-check"></i><b>2.2</b> Biology as a Mature Data Science</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="preliminaries.html"><a href="preliminaries.html"><i class="fa fa-check"></i><b>3</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="3.1" data-path="preliminaries.html"><a href="preliminaries.html#prelim-r"><i class="fa fa-check"></i><b>3.1</b> The R Language</a></li>
<li class="chapter" data-level="3.2" data-path="preliminaries.html"><a href="preliminaries.html#prelim-rstudio"><i class="fa fa-check"></i><b>3.2</b> RStudio</a></li>
<li class="chapter" data-level="3.3" data-path="preliminaries.html"><a href="preliminaries.html#prog-the-r-script"><i class="fa fa-check"></i><b>3.3</b> The R Script</a></li>
<li class="chapter" data-level="3.4" data-path="preliminaries.html"><a href="preliminaries.html#prog-workflow"><i class="fa fa-check"></i><b>3.4</b> The Scripting Workflow</a></li>
<li class="chapter" data-level="3.5" data-path="preliminaries.html"><a href="preliminaries.html#prelim-git"><i class="fa fa-check"></i><b>3.5</b> git + github</a>
<ul>
<li class="chapter" data-level="3.5.1" data-path="preliminaries.html"><a href="preliminaries.html#motivation"><i class="fa fa-check"></i><b>3.5.1</b> Motivation</a></li>
<li class="chapter" data-level="3.5.2" data-path="preliminaries.html"><a href="preliminaries.html#git"><i class="fa fa-check"></i><b>3.5.2</b> git</a></li>
<li class="chapter" data-level="3.5.3" data-path="preliminaries.html"><a href="preliminaries.html#git-hosting-platforms-github"><i class="fa fa-check"></i><b>3.5.3</b> Git hosting platforms (GitHub)</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="prog-basics.html"><a href="prog-basics.html"><i class="fa fa-check"></i><b>4</b> R Programming</a>
<ul>
<li class="chapter" data-level="4.1" data-path="prog-basics.html"><a href="prog-basics.html#before-you-begin"><i class="fa fa-check"></i><b>4.1</b> Before you begin</a></li>
<li class="chapter" data-level="4.2" data-path="prog-basics.html"><a href="prog-basics.html#introduction-1"><i class="fa fa-check"></i><b>4.2</b> Introduction</a></li>
<li class="chapter" data-level="4.3" data-path="prog-basics.html"><a href="prog-basics.html#prog-r-syntax"><i class="fa fa-check"></i><b>4.3</b> R Syntax Basics</a></li>
<li class="chapter" data-level="4.4" data-path="prog-basics.html"><a href="prog-basics.html#prog-types"><i class="fa fa-check"></i><b>4.4</b> Basic Types of Values</a>
<ul>
<li class="chapter" data-level="4.4.1" data-path="prog-basics.html"><a href="prog-basics.html#factors-."><i class="fa fa-check"></i><b>4.4.1</b> Factors .</a></li>
</ul></li>
<li class="chapter" data-level="4.5" data-path="prog-basics.html"><a href="prog-basics.html#prog-struct"><i class="fa fa-check"></i><b>4.5</b> Data Structures</a>
<ul>
<li class="chapter" data-level="4.5.1" data-path="prog-basics.html"><a href="prog-basics.html#vectors"><i class="fa fa-check"></i><b>4.5.1</b> Vectors</a></li>
<li class="chapter" data-level="4.5.2" data-path="prog-basics.html"><a href="prog-basics.html#matrices"><i class="fa fa-check"></i><b>4.5.2</b> Matrices</a></li>
<li class="chapter" data-level="4.5.3" data-path="prog-basics.html"><a href="prog-basics.html#lists-and-data-frames"><i class="fa fa-check"></i><b>4.5.3</b> Lists and data frames</a></li>
</ul></li>
<li class="chapter" data-level="4.6" data-path="prog-basics.html"><a href="prog-basics.html#logical-tests-and-comparators"><i class="fa fa-check"></i><b>4.6</b> Logical Tests and Comparators</a></li>
<li class="chapter" data-level="4.7" data-path="prog-basics.html"><a href="prog-basics.html#functions"><i class="fa fa-check"></i><b>4.7</b> Functions</a>
<ul>
<li class="chapter" data-level="4.7.1" data-path="prog-basics.html"><a href="prog-basics.html#prog-dry"><i class="fa fa-check"></i><b>4.7.1</b> DRY: Don’t Repeat Yourself</a></li>
<li class="chapter" data-level="4.7.2" data-path="prog-basics.html"><a href="prog-basics.html#writing-your-own-functions"><i class="fa fa-check"></i><b>4.7.2</b> Writing your own functions</a></li>
<li class="chapter" data-level="4.7.3" data-path="prog-basics.html"><a href="prog-basics.html#scope"><i class="fa fa-check"></i><b>4.7.3</b> Scope</a></li>
</ul></li>
<li class="chapter" data-level="4.8" data-path="prog-basics.html"><a href="prog-basics.html#iteration"><i class="fa fa-check"></i><b>4.8</b> Iteration</a>
<ul>
<li class="chapter" data-level="4.8.1" data-path="prog-basics.html"><a href="prog-basics.html#vectorized-operations"><i class="fa fa-check"></i><b>4.8.1</b> Vectorized operations</a></li>
<li class="chapter" data-level="4.8.2" data-path="prog-basics.html"><a href="prog-basics.html#functional-programming"><i class="fa fa-check"></i><b>4.8.2</b> Functional programming</a></li>
<li class="chapter" data-level="4.8.3" data-path="prog-basics.html"><a href="prog-basics.html#apply-and-friends"><i class="fa fa-check"></i><b>4.8.3</b> <code>apply()</code> and friends</a></li>
</ul></li>
<li class="chapter" data-level="4.9" data-path="prog-basics.html"><a href="prog-basics.html#installing-packages"><i class="fa fa-check"></i><b>4.9</b> Installing Packages</a></li>
<li class="chapter" data-level="4.10" data-path="prog-basics.html"><a href="prog-basics.html#saving-and-loading-r-data"><i class="fa fa-check"></i><b>4.10</b> Saving and Loading R Data</a></li>
<li class="chapter" data-level="4.11" data-path="prog-basics.html"><a href="prog-basics.html#prog-debugging"><i class="fa fa-check"></i><b>4.11</b> Troubleshooting and Debugging</a></li>
<li class="chapter" data-level="4.12" data-path="prog-basics.html"><a href="prog-basics.html#prog-style"><i class="fa fa-check"></i><b>4.12</b> Coding Style and Conventions</a>
<ul>
<li class="chapter" data-level="4.12.1" data-path="prog-basics.html"><a href="prog-basics.html#is-my-code-correct"><i class="fa fa-check"></i><b>4.12.1</b> Is my code correct?</a></li>
<li class="chapter" data-level="4.12.2" data-path="prog-basics.html"><a href="prog-basics.html#does-my-code-follow-the-dry-principle"><i class="fa fa-check"></i><b>4.12.2</b> Does my code follow the <span>DRY principle</span>?</a></li>
<li class="chapter" data-level="4.12.3" data-path="prog-basics.html"><a href="prog-basics.html#did-i-choose-concise-but-descriptive-variable-and-function-names"><i class="fa fa-check"></i><b>4.12.3</b> Did I choose concise but descriptive variable and function names?</a></li>
<li class="chapter" data-level="4.12.4" data-path="prog-basics.html"><a href="prog-basics.html#did-i-use-indentation-and-naming-conventions-consistently-throughout-my-code"><i class="fa fa-check"></i><b>4.12.4</b> Did I use indentation and naming conventions consistently throughout my code?</a></li>
<li class="chapter" data-level="4.12.5" data-path="prog-basics.html"><a href="prog-basics.html#did-i-write-comments-especially-when-what-the-code-does-is-not-obvious"><i class="fa fa-check"></i><b>4.12.5</b> Did I write comments, especially when what the code does is not obvious?</a></li>
<li class="chapter" data-level="4.12.6" data-path="prog-basics.html"><a href="prog-basics.html#how-easy-would-it-be-for-someone-else-to-understand-my-code"><i class="fa fa-check"></i><b>4.12.6</b> How easy would it be for someone else to understand my code?</a></li>
<li class="chapter" data-level="4.12.7" data-path="prog-basics.html"><a href="prog-basics.html#is-my-code-easy-to-maintainchange"><i class="fa fa-check"></i><b>4.12.7</b> Is my code easy to maintain/change?</a></li>
<li class="chapter" data-level="4.12.8" data-path="prog-basics.html"><a href="prog-basics.html#the-styler-package-."><i class="fa fa-check"></i><b>4.12.8</b> The <code>styler</code> package .</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="5" data-path="data-wrangling.html"><a href="data-wrangling.html"><i class="fa fa-check"></i><b>5</b> Data Wrangling</a>
<ul>
<li class="chapter" data-level="5.1" data-path="data-wrangling.html"><a href="data-wrangling.html#the-tidyverse"><i class="fa fa-check"></i><b>5.1</b> The Tidyverse</a></li>
<li class="chapter" data-level="5.2" data-path="data-wrangling.html"><a href="data-wrangling.html#dw-basics"><i class="fa fa-check"></i><b>5.2</b> Tidyverse Basics</a></li>
<li class="chapter" data-level="5.3" data-path="data-wrangling.html"><a href="data-wrangling.html#importing-data"><i class="fa fa-check"></i><b>5.3</b> Importing Data</a></li>
<li class="chapter" data-level="5.4" data-path="data-wrangling.html"><a href="data-wrangling.html#data-tibble"><i class="fa fa-check"></i><b>5.4</b> The tibble</a></li>
<li class="chapter" data-level="5.5" data-path="data-wrangling.html"><a href="data-wrangling.html#tidy-data"><i class="fa fa-check"></i><b>5.5</b> Tidy Data</a></li>
<li class="chapter" data-level="5.6" data-path="data-wrangling.html"><a href="data-wrangling.html#pipes"><i class="fa fa-check"></i><b>5.6</b> pipes</a></li>
<li class="chapter" data-level="5.7" data-path="data-wrangling.html"><a href="data-wrangling.html#arranging-data"><i class="fa fa-check"></i><b>5.7</b> Arranging Data</a>
<ul>
<li class="chapter" data-level="5.7.1" data-path="data-wrangling.html"><a href="data-wrangling.html#dplyrmutate---create-new-columns-using-other-columns"><i class="fa fa-check"></i><b>5.7.1</b> <code>dplyr::mutate()</code> - Create new columns using other columns</a></li>
<li class="chapter" data-level="5.7.2" data-path="data-wrangling.html"><a href="data-wrangling.html#stringr---working-with-character-values"><i class="fa fa-check"></i><b>5.7.2</b> <code>stringr</code> - Working with character values</a></li>
<li class="chapter" data-level="5.7.3" data-path="data-wrangling.html"><a href="data-wrangling.html#dplyrselect---subset-columns-by-name"><i class="fa fa-check"></i><b>5.7.3</b> <code>dplyr::select()</code> - Subset Columns by Name</a></li>
<li class="chapter" data-level="5.7.4" data-path="data-wrangling.html"><a href="data-wrangling.html#dplyrfilter---pick-rows-out-of-a-data-set"><i class="fa fa-check"></i><b>5.7.4</b> <code>dplyr::filter()</code> - Pick rows out of a data set</a></li>
<li class="chapter" data-level="5.7.5" data-path="data-wrangling.html"><a href="data-wrangling.html#dplyrarrange---order-rows-based-on-their-values"><i class="fa fa-check"></i><b>5.7.5</b> <code>dplyr::arrange()</code> - Order rows based on their values</a></li>
<li class="chapter" data-level="5.7.6" data-path="data-wrangling.html"><a href="data-wrangling.html#putting-it-all-together"><i class="fa fa-check"></i><b>5.7.6</b> Putting it all together</a></li>
</ul></li>
<li class="chapter" data-level="5.8" data-path="data-wrangling.html"><a href="data-wrangling.html#grouping-data"><i class="fa fa-check"></i><b>5.8</b> Grouping Data</a></li>
<li class="chapter" data-level="5.9" data-path="data-wrangling.html"><a href="data-wrangling.html#rearranging-data"><i class="fa fa-check"></i><b>5.9</b> Rearranging Data</a></li>
<li class="chapter" data-level="5.10" data-path="data-wrangling.html"><a href="data-wrangling.html#relational-data"><i class="fa fa-check"></i><b>5.10</b> Relational Data</a>
<ul>
<li class="chapter" data-level="5.10.1" data-path="data-wrangling.html"><a href="data-wrangling.html#dealing-with-multiple-matches"><i class="fa fa-check"></i><b>5.10.1</b> Dealing with multiple matches</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="6" data-path="data-science.html"><a href="data-science.html"><i class="fa fa-check"></i><b>6</b> Data Science</a>
<ul>
<li class="chapter" data-level="6.1" data-path="data-science.html"><a href="data-science.html#data-modeling"><i class="fa fa-check"></i><b>6.1</b> Data Modeling</a>
<ul>
<li class="chapter" data-level="6.1.1" data-path="data-science.html"><a href="data-science.html#a-worked-modeling-example"><i class="fa fa-check"></i><b>6.1.1</b> A Worked Modeling Example</a></li>
<li class="chapter" data-level="6.1.2" data-path="data-science.html"><a href="data-science.html#data-summarization"><i class="fa fa-check"></i><b>6.1.2</b> Data Summarization</a></li>
<li class="chapter" data-level="6.1.3" data-path="data-science.html"><a href="data-science.html#assessing-model-accuracy"><i class="fa fa-check"></i><b>6.1.3</b> Assessing Model Accuracy</a></li>
<li class="chapter" data-level="6.1.4" data-path="data-science.html"><a href="data-science.html#linear-models-."><i class="fa fa-check"></i><b>6.1.4</b> Linear Models .</a></li>
</ul></li>
<li class="chapter" data-level="6.2" data-path="data-science.html"><a href="data-science.html#statistical-distributions-tests-."><i class="fa fa-check"></i><b>6.2</b> Statistical Distributions &amp; Tests .</a>
<ul>
<li class="chapter" data-level="6.2.1" data-path="data-science.html"><a href="data-science.html#statistical-distributions-."><i class="fa fa-check"></i><b>6.2.1</b> Statistical Distributions .</a></li>
<li class="chapter" data-level="6.2.2" data-path="data-science.html"><a href="data-science.html#statistical-tests-."><i class="fa fa-check"></i><b>6.2.2</b> Statistical Tests .</a></li>
<li class="chapter" data-level="6.2.3" data-path="data-science.html"><a href="data-science.html#p-values-."><i class="fa fa-check"></i><b>6.2.3</b> p-values .</a></li>
<li class="chapter" data-level="6.2.4" data-path="data-science.html"><a href="data-science.html#multiple-hypothesis-testing-."><i class="fa fa-check"></i><b>6.2.4</b> Multiple Hypothesis Testing .</a></li>
<li class="chapter" data-level="6.2.5" data-path="data-science.html"><a href="data-science.html#statistical-power-."><i class="fa fa-check"></i><b>6.2.5</b> Statistical power .</a></li>
</ul></li>
<li class="chapter" data-level="6.3" data-path="data-science.html"><a href="data-science.html#clustering-."><i class="fa fa-check"></i><b>6.3</b> Clustering .</a></li>
<li class="chapter" data-level="6.4" data-path="data-science.html"><a href="data-science.html#network-analysis-."><i class="fa fa-check"></i><b>6.4</b> Network Analysis .</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html"><i class="fa fa-check"></i><b>7</b> Biology &amp; Bioinformatics</a>
<ul>
<li class="chapter" data-level="7.1" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#bio-r"><i class="fa fa-check"></i><b>7.1</b> R in Biology</a></li>
<li class="chapter" data-level="7.2" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#csv-files"><i class="fa fa-check"></i><b>7.2</b> CSV Files</a></li>
<li class="chapter" data-level="7.3" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#bio-data-types"><i class="fa fa-check"></i><b>7.3</b> Types of Biological Data</a></li>
<li class="chapter" data-level="7.4" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#common-biological-data-matrices"><i class="fa fa-check"></i><b>7.4</b> Common Biological Data Matrices</a>
<ul>
<li class="chapter" data-level="7.4.1" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#biological-data-is-not-tidy"><i class="fa fa-check"></i><b>7.4.1</b> Biological data is NOT Tidy!</a></li>
</ul></li>
<li class="chapter" data-level="7.5" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#bio-bioconductor"><i class="fa fa-check"></i><b>7.5</b> Bioconductor</a></li>
<li class="chapter" data-level="7.6" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#gene-identifiers"><i class="fa fa-check"></i><b>7.6</b> Gene Identifiers</a>
<ul>
<li class="chapter" data-level="7.6.1" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#gene-identifier-systems"><i class="fa fa-check"></i><b>7.6.1</b> Gene Identifier Systems</a></li>
<li class="chapter" data-level="7.6.2" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#mapping-between-identifier-systems"><i class="fa fa-check"></i><b>7.6.2</b> Mapping Between Identifier Systems</a></li>
<li class="chapter" data-level="7.6.3" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#mapping-homologs"><i class="fa fa-check"></i><b>7.6.3</b> Mapping Homologs</a></li>
</ul></li>
<li class="chapter" data-level="7.7" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#gene-expression"><i class="fa fa-check"></i><b>7.7</b> Gene Expression</a>
<ul>
<li class="chapter" data-level="7.7.1" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#gene-expression-data-in-bioconductor"><i class="fa fa-check"></i><b>7.7.1</b> Gene Expression Data in Bioconductor</a></li>
<li class="chapter" data-level="7.7.2" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#microarrays-."><i class="fa fa-check"></i><b>7.7.2</b> Microarrays .</a></li>
<li class="chapter" data-level="7.7.3" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#differential-expression-microarrays-limma-."><i class="fa fa-check"></i><b>7.7.3</b> Differential Expression: Microarrays (limma) .</a></li>
<li class="chapter" data-level="7.7.4" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#high-throughput-sequencing-."><i class="fa fa-check"></i><b>7.7.4</b> High Throughput Sequencing .</a></li>
<li class="chapter" data-level="7.7.5" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#count-data-."><i class="fa fa-check"></i><b>7.7.5</b> Count Data .</a></li>
<li class="chapter" data-level="7.7.6" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#differential-expression-rnaseq-."><i class="fa fa-check"></i><b>7.7.6</b> Differential Expression: RNASeq .</a></li>
</ul></li>
<li class="chapter" data-level="7.8" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#genomic-intervals-."><i class="fa fa-check"></i><b>7.8</b> Genomic Intervals .</a></li>
<li class="chapter" data-level="7.9" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#gene-set-enrichment-analysis-."><i class="fa fa-check"></i><b>7.9</b> Gene Set Enrichment Analysis .</a>
<ul>
<li class="chapter" data-level="7.9.1" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#gene-sets-."><i class="fa fa-check"></i><b>7.9.1</b> Gene Sets .</a></li>
<li class="chapter" data-level="7.9.2" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#over-representation-analysis-."><i class="fa fa-check"></i><b>7.9.2</b> Over-representation Analysis .</a></li>
<li class="chapter" data-level="7.9.3" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#rank-based-analysis-."><i class="fa fa-check"></i><b>7.9.3</b> Rank-based Analysis .</a></li>
</ul></li>
<li class="chapter" data-level="7.10" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#biological-networks-."><i class="fa fa-check"></i><b>7.10</b> Biological Networks .</a>
<ul>
<li class="chapter" data-level="7.10.1" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#biological-pathways-."><i class="fa fa-check"></i><b>7.10.1</b> Biological Pathways .</a></li>
<li class="chapter" data-level="7.10.2" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#gene-regulatory-networks-."><i class="fa fa-check"></i><b>7.10.2</b> Gene Regulatory Networks .</a></li>
<li class="chapter" data-level="7.10.3" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#protein-protein-interaction-network-."><i class="fa fa-check"></i><b>7.10.3</b> Protein-Protein Interaction Network .</a></li>
<li class="chapter" data-level="7.10.4" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#wgcna-."><i class="fa fa-check"></i><b>7.10.4</b> WGCNA .</a></li>
</ul></li>
<li class="chapter" data-level="7.11" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#single-cell-sequencing-analysis-."><i class="fa fa-check"></i><b>7.11</b> Single Cell Sequencing Analysis .</a>
<ul>
<li class="chapter" data-level="7.11.1" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#single-cell-sequencing-."><i class="fa fa-check"></i><b>7.11.1</b> Single Cell Sequencing .</a></li>
<li class="chapter" data-level="7.11.2" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#single-cell-analysis-."><i class="fa fa-check"></i><b>7.11.2</b> Single Cell Analysis .</a></li>
<li class="chapter" data-level="7.11.3" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#single-cell-clustering-."><i class="fa fa-check"></i><b>7.11.3</b> Single Cell Clustering .</a></li>
<li class="chapter" data-level="7.11.4" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#dimensionality-reduction-projection-."><i class="fa fa-check"></i><b>7.11.4</b> Dimensionality Reduction &amp; Projection .</a></li>
<li class="chapter" data-level="7.11.5" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#single-cell-data-visualization-."><i class="fa fa-check"></i><b>7.11.5</b> Single Cell Data Visualization .</a></li>
<li class="chapter" data-level="7.11.6" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#single-cell-marker-analysis-."><i class="fa fa-check"></i><b>7.11.6</b> Single Cell Marker Analysis .</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="8" data-path="data-visualization.html"><a href="data-visualization.html"><i class="fa fa-check"></i><b>8</b> Data Visualization</a>
<ul>
<li class="chapter" data-level="8.1" data-path="data-visualization.html"><a href="data-visualization.html#dv-gg"><i class="fa fa-check"></i><b>8.1</b> Grammar of Graphics</a></li>
<li class="chapter" data-level="8.2" data-path="data-visualization.html"><a href="data-visualization.html#plotting-one-dimension"><i class="fa fa-check"></i><b>8.2</b> Plotting One Dimension</a>
<ul>
<li class="chapter" data-level="8.2.1" data-path="data-visualization.html"><a href="data-visualization.html#dv-bar"><i class="fa fa-check"></i><b>8.2.1</b> Bar chart</a></li>
<li class="chapter" data-level="8.2.2" data-path="data-visualization.html"><a href="data-visualization.html#lollipop-plots"><i class="fa fa-check"></i><b>8.2.2</b> Lollipop plots</a></li>
<li class="chapter" data-level="8.2.3" data-path="data-visualization.html"><a href="data-visualization.html#parallel-coordinate-plots"><i class="fa fa-check"></i><b>8.2.3</b> Parallel Coordinate plots</a></li>
</ul></li>
<li class="chapter" data-level="8.3" data-path="data-visualization.html"><a href="data-visualization.html#visualizing-distributions"><i class="fa fa-check"></i><b>8.3</b> Visualizing Distributions</a>
<ul>
<li class="chapter" data-level="8.3.1" data-path="data-visualization.html"><a href="data-visualization.html#histogram"><i class="fa fa-check"></i><b>8.3.1</b> Histogram</a></li>
<li class="chapter" data-level="8.3.2" data-path="data-visualization.html"><a href="data-visualization.html#density"><i class="fa fa-check"></i><b>8.3.2</b> Density</a></li>
<li class="chapter" data-level="8.3.3" data-path="data-visualization.html"><a href="data-visualization.html#boxplot"><i class="fa fa-check"></i><b>8.3.3</b> Boxplot</a></li>
<li class="chapter" data-level="8.3.4" data-path="data-visualization.html"><a href="data-visualization.html#violin-plot"><i class="fa fa-check"></i><b>8.3.4</b> Violin plot</a></li>
<li class="chapter" data-level="8.3.5" data-path="data-visualization.html"><a href="data-visualization.html#beeswarm-plot"><i class="fa fa-check"></i><b>8.3.5</b> Beeswarm plot</a></li>
<li class="chapter" data-level="8.3.6" data-path="data-visualization.html"><a href="data-visualization.html#ridgeline"><i class="fa fa-check"></i><b>8.3.6</b> Ridgeline</a></li>
</ul></li>
<li class="chapter" data-level="8.4" data-path="data-visualization.html"><a href="data-visualization.html#plotting-two-or-more-dimensions-."><i class="fa fa-check"></i><b>8.4</b> Plotting Two or More Dimensions .</a>
<ul>
<li class="chapter" data-level="8.4.1" data-path="data-visualization.html"><a href="data-visualization.html#scatter-plot-."><i class="fa fa-check"></i><b>8.4.1</b> Scatter plot .</a></li>
<li class="chapter" data-level="8.4.2" data-path="data-visualization.html"><a href="data-visualization.html#line-plot-."><i class="fa fa-check"></i><b>8.4.2</b> Line plot .</a></li>
<li class="chapter" data-level="8.4.3" data-path="data-visualization.html"><a href="data-visualization.html#bubble-plot-."><i class="fa fa-check"></i><b>8.4.3</b> Bubble plot .</a></li>
<li class="chapter" data-level="8.4.4" data-path="data-visualization.html"><a href="data-visualization.html#heatmaps-."><i class="fa fa-check"></i><b>8.4.4</b> Heatmaps .</a></li>
</ul></li>
<li class="chapter" data-level="8.5" data-path="data-visualization.html"><a href="data-visualization.html#other-kind-of-plots-."><i class="fa fa-check"></i><b>8.5</b> Other Kind of Plots .</a>
<ul>
<li class="chapter" data-level="8.5.1" data-path="data-visualization.html"><a href="data-visualization.html#chord-."><i class="fa fa-check"></i><b>8.5.1</b> Chord .</a></li>
<li class="chapter" data-level="8.5.2" data-path="data-visualization.html"><a href="data-visualization.html#sankey-."><i class="fa fa-check"></i><b>8.5.2</b> Sankey .</a></li>
<li class="chapter" data-level="8.5.3" data-path="data-visualization.html"><a href="data-visualization.html#dendrogram-."><i class="fa fa-check"></i><b>8.5.3</b> Dendrogram .</a></li>
</ul></li>
<li class="chapter" data-level="8.6" data-path="data-visualization.html"><a href="data-visualization.html#network-visualization-."><i class="fa fa-check"></i><b>8.6</b> Network visualization .</a></li>
<li class="chapter" data-level="8.7" data-path="data-visualization.html"><a href="data-visualization.html#tips-and-tricks-."><i class="fa fa-check"></i><b>8.7</b> Tips and Tricks .</a>
<ul>
<li class="chapter" data-level="8.7.1" data-path="data-visualization.html"><a href="data-visualization.html#reordering-1-d-data-elements-."><i class="fa fa-check"></i><b>8.7.1</b> Reordering 1-D Data Elements .</a></li>
<li class="chapter" data-level="8.7.2" data-path="data-visualization.html"><a href="data-visualization.html#confidence-intervals-."><i class="fa fa-check"></i><b>8.7.2</b> Confidence Intervals .</a></li>
<li class="chapter" data-level="8.7.3" data-path="data-visualization.html"><a href="data-visualization.html#annotations-."><i class="fa fa-check"></i><b>8.7.3</b> Annotations .</a></li>
</ul></li>
<li class="chapter" data-level="8.8" data-path="data-visualization.html"><a href="data-visualization.html#multiple-plots-."><i class="fa fa-check"></i><b>8.8</b> Multiple Plots .</a>
<ul>
<li class="chapter" data-level="8.8.1" data-path="data-visualization.html"><a href="data-visualization.html#secondary-axes-."><i class="fa fa-check"></i><b>8.8.1</b> Secondary Axes .</a></li>
<li class="chapter" data-level="8.8.2" data-path="data-visualization.html"><a href="data-visualization.html#facet-wrapping-."><i class="fa fa-check"></i><b>8.8.2</b> Facet wrapping .</a></li>
<li class="chapter" data-level="8.8.3" data-path="data-visualization.html"><a href="data-visualization.html#multipanel-figures-."><i class="fa fa-check"></i><b>8.8.3</b> Multipanel Figures .</a></li>
</ul></li>
<li class="chapter" data-level="8.9" data-path="data-visualization.html"><a href="data-visualization.html#responsible-plotting-."><i class="fa fa-check"></i><b>8.9</b> Responsible Plotting .</a></li>
<li class="chapter" data-level="8.10" data-path="data-visualization.html"><a href="data-visualization.html#publication-ready-plots-."><i class="fa fa-check"></i><b>8.10</b> Publication Ready Plots .</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="engineering.html"><a href="engineering.html"><i class="fa fa-check"></i><b>9</b> EngineeRing</a>
<ul>
<li class="chapter" data-level="9.1" data-path="engineering.html"><a href="engineering.html#unit-testing"><i class="fa fa-check"></i><b>9.1</b> Unit Testing</a></li>
<li class="chapter" data-level="9.2" data-path="engineering.html"><a href="engineering.html#toolification-."><i class="fa fa-check"></i><b>9.2</b> Toolification .</a></li>
<li class="chapter" data-level="9.3" data-path="engineering.html"><a href="engineering.html#pipelines-workflows-."><i class="fa fa-check"></i><b>9.3</b> Pipelines &amp; Workflows .</a></li>
<li class="chapter" data-level="9.4" data-path="engineering.html"><a href="engineering.html#parallel-processing-."><i class="fa fa-check"></i><b>9.4</b> Parallel Processing .</a></li>
<li class="chapter" data-level="9.5" data-path="engineering.html"><a href="engineering.html#r-packages-."><i class="fa fa-check"></i><b>9.5</b> R Packages .</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="rshiny.html"><a href="rshiny.html"><i class="fa fa-check"></i><b>10</b> RShiny</a>
<ul>
<li class="chapter" data-level="10.1" data-path="rshiny.html"><a href="rshiny.html#overview"><i class="fa fa-check"></i><b>10.1</b> Overview</a></li>
<li class="chapter" data-level="10.2" data-path="rshiny.html"><a href="rshiny.html#learning-objectives"><i class="fa fa-check"></i><b>10.2</b> Learning Objectives</a></li>
<li class="chapter" data-level="10.3" data-path="rshiny.html"><a href="rshiny.html#skill-list"><i class="fa fa-check"></i><b>10.3</b> Skill List</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="communicating-with-r.html"><a href="communicating-with-r.html"><i class="fa fa-check"></i><b>11</b> Communicating with R</a>
<ul>
<li class="chapter" data-level="11.1" data-path="communicating-with-r.html"><a href="communicating-with-r.html#rmarkdown-knitr"><i class="fa fa-check"></i><b>11.1</b> RMarkdown &amp; knitr</a></li>
<li class="chapter" data-level="11.2" data-path="communicating-with-r.html"><a href="communicating-with-r.html#bookdown"><i class="fa fa-check"></i><b>11.2</b> bookdown</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="contribution-guide.html"><a href="contribution-guide.html"><i class="fa fa-check"></i><b>12</b> Contribution Guide</a>
<ul>
<li class="chapter" data-level="12.1" data-path="contribution-guide.html"><a href="contribution-guide.html#custom-blocks"><i class="fa fa-check"></i><b>12.1</b> Custom blocks</a></li>
</ul></li>
<li class="part"><span><b>I Assignments</b></span></li>
<li class="chapter" data-level="" data-path="assignment-format.html"><a href="assignment-format.html"><i class="fa fa-check"></i>Assignment Format</a></li>
<li class="chapter" data-level="" data-path="starting-an-assignment.html"><a href="starting-an-assignment.html"><i class="fa fa-check"></i>Starting an Assignment</a>
<ul>
<li><a href="starting-an-assignment.html#starting-your-own-git-repository-and-cloning-it">Starting your own <code>git</code> repository and cloning it</a>
<ul>
<li class="chapter" data-level="" data-path="starting-an-assignment.html"><a href="starting-an-assignment.html#contents"><i class="fa fa-check"></i>Contents</a></li>
<li class="chapter" data-level="" data-path="starting-an-assignment.html"><a href="starting-an-assignment.html#github-tutorial"><i class="fa fa-check"></i>GitHub tutorial</a></li>
<li class="chapter" data-level="" data-path="starting-an-assignment.html"><a href="starting-an-assignment.html#committing-and-pushing-with-r-and-without"><i class="fa fa-check"></i>Committing and Pushing with R (and without)</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="assignment-1.html"><a href="assignment-1.html"><i class="fa fa-check"></i>Assignment 1</a>
<ul>
<li class="chapter" data-level="" data-path="assignment-1.html"><a href="assignment-1.html#problem-statement"><i class="fa fa-check"></i>Problem Statement</a></li>
<li class="chapter" data-level="" data-path="assignment-1.html"><a href="assignment-1.html#learning-objectives-and-skill-list"><i class="fa fa-check"></i>Learning Objectives and Skill List</a></li>
<li class="chapter" data-level="" data-path="assignment-1.html"><a href="assignment-1.html#instructions"><i class="fa fa-check"></i>Instructions</a></li>
<li class="chapter" data-level="" data-path="assignment-1.html"><a href="assignment-1.html#hints"><i class="fa fa-check"></i>Hints</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="assignment-2.html"><a href="assignment-2.html"><i class="fa fa-check"></i>Assignment 2</a>
<ul>
<li class="chapter" data-level="" data-path="assignment-2.html"><a href="assignment-2.html#problem-statement-1"><i class="fa fa-check"></i>Problem Statement</a></li>
<li class="chapter" data-level="" data-path="assignment-2.html"><a href="assignment-2.html#learning-objectives-1"><i class="fa fa-check"></i>Learning Objectives</a></li>
<li class="chapter" data-level="" data-path="assignment-2.html"><a href="assignment-2.html#skill-list-1"><i class="fa fa-check"></i>Skill List</a></li>
<li class="chapter" data-level="" data-path="assignment-2.html"><a href="assignment-2.html#instructions-1"><i class="fa fa-check"></i>Instructions</a></li>
<li class="chapter" data-level="" data-path="assignment-2.html"><a href="assignment-2.html#function-details"><i class="fa fa-check"></i>Function Details</a>
<ul>
<li class="chapter" data-level="" data-path="assignment-2.html"><a href="assignment-2.html#bioconductor"><i class="fa fa-check"></i>1. Bioconductor</a></li>
<li><a href="assignment-2.html#load_expression">2. <code>load_expression()</code></a></li>
<li><a href="assignment-2.html#filter_15">3. <code>filter_15()</code></a></li>
<li><a href="assignment-2.html#affy_to_hgnc">4. <code>affy_to_hgnc()</code></a></li>
<li><a href="assignment-2.html#reduce_data">5. <code>reduce_data()</code></a></li>
</ul></li>
</ul></li>
<li class="appendix"><span><b>Appendix</b></span></li>
<li class="chapter" data-level="A" data-path="class-outlines.html"><a href="class-outlines.html"><i class="fa fa-check"></i><b>A</b> Class Outlines</a>
<ul>
<li class="chapter" data-level="A.1" data-path="class-outlines.html"><a href="class-outlines.html#week-1"><i class="fa fa-check"></i><b>A.1</b> Week 1</a></li>
<li class="chapter" data-level="A.2" data-path="class-outlines.html"><a href="class-outlines.html#week-2"><i class="fa fa-check"></i><b>A.2</b> Week 2</a></li>
<li class="chapter" data-level="A.3" data-path="class-outlines.html"><a href="class-outlines.html#week-3"><i class="fa fa-check"></i><b>A.3</b> Week 3</a></li>
<li class="chapter" data-level="A.4" data-path="class-outlines.html"><a href="class-outlines.html#week-4"><i class="fa fa-check"></i><b>A.4</b> Week 4</a></li>
<li class="chapter" data-level="A.5" data-path="class-outlines.html"><a href="class-outlines.html#week-5"><i class="fa fa-check"></i><b>A.5</b> Week 5</a></li>
<li class="chapter" data-level="A.6" data-path="class-outlines.html"><a href="class-outlines.html#week-6"><i class="fa fa-check"></i><b>A.6</b> Week 6</a></li>
<li class="chapter" data-level="A.7" data-path="class-outlines.html"><a href="class-outlines.html#week-7"><i class="fa fa-check"></i><b>A.7</b> Week 7</a></li>
<li class="chapter" data-level="A.8" data-path="class-outlines.html"><a href="class-outlines.html#week-8"><i class="fa fa-check"></i><b>A.8</b> Week 8</a></li>
<li class="chapter" data-level="A.9" data-path="class-outlines.html"><a href="class-outlines.html#week-9"><i class="fa fa-check"></i><b>A.9</b> Week 9</a></li>
<li class="chapter" data-level="A.10" data-path="class-outlines.html"><a href="class-outlines.html#week-10"><i class="fa fa-check"></i><b>A.10</b> Week 10</a></li>
<li class="chapter" data-level="A.11" data-path="class-outlines.html"><a href="class-outlines.html#week-11"><i class="fa fa-check"></i><b>A.11</b> Week 11</a></li>
<li class="chapter" data-level="A.12" data-path="class-outlines.html"><a href="class-outlines.html#week-12"><i class="fa fa-check"></i><b>A.12</b> Week 12</a></li>
<li class="chapter" data-level="A.13" data-path="class-outlines.html"><a href="class-outlines.html#week-13---no-class-for-project-work"><i class="fa fa-check"></i><b>A.13</b> Week 13 - no class for project work</a></li>
<li class="chapter" data-level="A.14" data-path="class-outlines.html"><a href="class-outlines.html#week-14---no-class-for-project-work"><i class="fa fa-check"></i><b>A.14</b> Week 14 - no class for project work</a></li>
</ul></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">BF591 - R for Biological Sciences</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="data-science" class="section level1" number="6">
<h1><span class="header-section-number">6</span> Data Science</h1>
<p><a href="https://en.wikipedia.org/wiki/Data_science">Data science</a> is an enormous and
rapidly growing field that incorporates elements of statistics, computer
science, software engineering, high performance and cloud computing, and big
data management, as well as syntheses with knowledge from other social and
physical science fields to model and predict phenomena captured by data
collected from the “real world.” Many books have and will be written on this
subject, and this one does not pretend to even attempt to give this area an
adequate treatment. Like the rest of this book, the topics covered here are
opinionated and presented through the lens of a biological analyst practitioner
with enough high level conceptual details and a few general principles to
hopefully be useful in that context.</p>
<div id="data-modeling" class="section level2" number="6.1">
<h2><span class="header-section-number">6.1</span> Data Modeling</h2>
<p>The goal of data modeling is to describe a dataset using a relatively small
number of mathematical relationships. Said differently, a model uses some parts
of a dataset to try to accurately predict other parts of the dataset in a way
that is useful to us.</p>
<p>Models are human inventions; they reflect our beliefs about the way the universe
works. The successful model identifies patterns within a dataset that are the
result of causal relationships in the universe that led to the phenomena that
were measured while accounting for noise in the data. However, <em>the model itself
does not identify or even accurately reflect those causal effects</em>. The model
merely summarizes patterns and we as scientists are left to interpret those
patterns and design follow up experiments to investigate the nature of those
causal relationships using our prior knowledge of the world.</p>
<p>There are several principles to keep in mind when modeling data:</p>
<ol style="list-style-type: decimal">
<li><p><strong>Data are never wrong.</strong> All data are collected using processes and
devices designed and implemented by humans, who always have biases and make
assumptions. All data measure something about the universe, and so are “true” in
some sense of the word. If what we intended to measure and what we actually
measured were not the same thing, that is due to our errors in collection or
interpretation, not due to the data being wrong. If we approach our dataset with
a particular set of hypotheses and the data don’t support those hypotheses, it
is our beliefs of the world and our understanding of the dataset that are wrong,
not the data itself.</p></li>
<li><p><strong>Not all data are useful.</strong> Just because data isn’t wrong, it doesn’t mean
it is useful. There may have been systematic errors in the collection of the
data that makes interpreting them difficult. Data collected for one purpose may
not be useful for any other purposes. And sometimes, a dataset collected for a
particular purpose may simply not have the information needed to answer our
questions; if what we measure has no relationship to what we wish to predict,
the data itself is not useful - though the knowledge that what we measured has
no detectable effect on the thing we wish to predict may be very useful!</p></li>
<li><p><strong>“All models are wrong, but some are useful.”</strong> George Box, the renowned
British statistician, famously asserted this in a 1976 paper to the Journal of
the American Statistical Association. <span class="citation">(<a href="#ref-Box1976-rd" role="doc-biblioref">Box 1976</a>)</span>. By this he meant that every
model we create is a simplification of the system we are seeking to model, which
is by definition not identical to that system. To perfectly model a system, our
model would need to be precisely the system we are modeling, which is no longer
a model but the system itself. Fortunately, even though we know our models are
always wrong to some degree, they may nonetheless be useful because they are not
<strong>too</strong> wrong. Some models may indeed be too wrong, though.</p></li>
<li><p><strong>Data do not contain causal information.</strong> “Correlation does not mean
causation.” Data are measurements of the results of a process in the universe
that we wish to understand; the data are possibly reflective of that process,
but do not contain any information about the process itself. We cannot infer
causal relationships from a dataset alone. We must construct possible causal
models using our knowledge of the world first, then apply our data to our model
and other alternative models to compare their relative plausibility.</p></li>
<li><p><strong>All data have noise.</strong> The usefulness of a model to describe a dataset is
related to the relative strength of the patterns and noise in the dataset when
viewed through the lens of the model; conceptually, the so-called “signal to
noise ratio” of the data. The fundamental concern of statistics is quantifying
uncertainty (i.e. noise), and separating it from real signal, though different
statistical approaches (e.g. frequentist vs Bayesian) reason about uncertainty
in different ways.</p></li>
</ol>
<p>Modeling begins (or should begin) with posing one or more <strong>scientific models</strong>
of the process or phenomenon we wish to understand. The scientific model is
conceptual; it reflects our belief of the universe and proposes a causal
explanation for the phenomenon. We then decide how to map that scientific model
onto a <strong>statistical model</strong>, which is a mechanical procedure that quantifies
how well our scientific model explains a dataset. <em>The scientific model and
statistical model are related but independent choices we make.</em> There may be
many valid statistical models that represent a given scientific model. However,
sometimes in practice we lack sufficient knowledge about the process to propose
scientific models first, requiring data exploration and summarization first to
suggest reasonable starting points.</p>
<div class="box important">
<p>This section pertains primarily to models specified explicitly by humans. There
is another class of models, namely those created by certain machine learning
algorithms like neural networks and deep learning, that <em>discover models from
data</em>. These models are fundamentally different than those designed by human
minds, in that they are often accurate and therefore useful, but it can be very
difficult if not impossible to understand how they work. While these are
important types of models that fall under the umbrella of data science, we limit
the content of this chapter to human designed statistical models.</p>
</div>
<div id="a-worked-modeling-example" class="section level3" number="6.1.1">
<h3><span class="header-section-number">6.1.1</span> A Worked Modeling Example</h3>
<p>As an example, let’s consider a scenario where we wish to assess whether any of
three genes can help us distinguish between patients who have Parkinson’s
Disease and those who don’t by measuring the relative activity of those genes in
blood samples. We have the following made-up dataset:</p>
<div class="sourceCode" id="cb161"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb161-1"><a href="data-science.html#cb161-1" aria-hidden="true" tabindex="-1"></a>gene_exp</span></code></pre></div>
<pre><code>## # A tibble: 200 x 5
##    sample_name `Disease Status` `Gene 1` `Gene 2` `Gene 3`
##    &lt;chr&gt;       &lt;chr&gt;               &lt;dbl&gt;    &lt;dbl&gt;    &lt;dbl&gt;
##  1 P1          Parkinson&#39;s          257.     636.     504.
##  2 P2          Parkinson&#39;s          241.     556.     484.
##  3 P3          Parkinson&#39;s          252.     426.     476.
##  4 P4          Parkinson&#39;s          271.     405.     458.
##  5 P5          Parkinson&#39;s          248.     482.     520.
##  6 P6          Parkinson&#39;s          275.     521.     460.
##  7 P7          Parkinson&#39;s          264.     415.     568.
##  8 P8          Parkinson&#39;s          276.     450.     495.
##  9 P9          Parkinson&#39;s          274.     490.     496.
## 10 P10         Parkinson&#39;s          251.     584.     573.
## # ... with 190 more rows</code></pre>
<p>Our imaginary dataset has 100 Parkinson’s and 100 control subjects. For each of
our samples, we have a sample ID, Disease Status of <code>Parkinson's</code> or <code>Control</code>,
and numeric measurements of each of three genes. Below are <a href="data-visualization.html#violin-plot">violin
plots</a> of our (made-up) data set for these three genes:</p>
<p><img src="main_files/data%20modeling%20violin-1.png" width="672" /></p>
<p>By inspection, it appears that Gene 1 has no relationship with disease; we may
safely eliminate this gene from further consideration. Gene 2 appears to have a
different profile depending on disease status, where control individuals have
a higher average expression and a lower variance. Unfortunately, despite this
qualitative difference, this gene may not be useful for telling whether someone
has disease or not - the ranges completely overlap. Gene 3 appears to
discriminate between disease and control. There is some overlap in the two
expression distributions, but above a certain expression value these data
suggest a high degree of predictive accuracy may be obtained with this gene.
Measuring this gene may therefore be useful, if the results from this dataset
generalize to all people with Parkinson’s Disease.</p>
<p>So far, we have not done any modeling, but instead relied on plotting and our
eyes. A more quantitative question might be: how much higher is Gene 3
expression in Parkinson’s Disease than control? Another way of posing this
question is: if I know a patient has Parkinson’s Disease, what Gene 3 expression
value do I expect them to have? Written this way, we have turned our question
into a prediction problem: if we only had information that a patient had
Parkinson’s Disease, what is the predicted expression value of their Gene 3?</p>
<p>Another way to pose this prediction question is in the opposite (and arguably
more useful) direction: if all we knew about a person was their Gene 3 gene
expression, how likely is it that the person has Parkinson’s Disease? If this
gene expression is predictive enough of a person’s disease status, it may be a
viable <a href="https://en.wikipedia.org/wiki/Biomarker">biomarker</a> of disease and thus
might be useful in a clinical setting, for example when identifying
presymptomatic individuals or assessing the efficacy of a pharmacological
treatment.</p>
<p>Although it may seems obvious, before beginning to model a dataset, we must
start by posing the <strong>scientific question</strong> as concisely as possible, as we have
done above. These questions will help us identify which modeling techniques are
appropriate and help us ensure we interpret our results correctly.</p>
</div>
<div id="data-summarization" class="section level3" number="6.1.2">
<h3><span class="header-section-number">6.1.2</span> Data Summarization</h3>
<p>Broadly speaking, <em>data summarization</em> is the process of finding a
lower-dimensional representation of a larger dataset. There are many ways to
summarize a set of data; each approach will emphasize different aspects of the
dataset, and have varying degrees of accuracy. Consider the gene expression of
Gene 1 for all individuals in our example above, plotted as a distribution with
a <a href="data-visualization.html#histogram">histogram</a>:</p>
<div class="sourceCode" id="cb163"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb163-1"><a href="data-science.html#cb163-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(gene_exp, <span class="fu">aes</span>(<span class="at">x=</span><span class="st">`</span><span class="at">Gene 1</span><span class="st">`</span>)) <span class="sc">+</span></span>
<span id="cb163-2"><a href="data-science.html#cb163-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">bins=</span><span class="dv">30</span>,<span class="at">fill=</span><span class="st">&quot;#a93c13&quot;</span>)</span></code></pre></div>
<p><img src="main_files/data%20sci%20summ%20hist-1.png" width="672" /></p>
<div id="point-estimates" class="section level4" number="6.1.2.1">
<h4><span class="header-section-number">6.1.2.1</span> Point Estimates</h4>
<p>The data are concentrated around the value 250, and become less common for
larger and smaller values. Since the extents to the left and right of the middle
of the distribution appear to be equally distant, perhaps the arithmetic mean
is a good way to identify the middle of the distribution:</p>
<div class="sourceCode" id="cb164"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb164-1"><a href="data-science.html#cb164-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(gene_exp, <span class="fu">aes</span>(<span class="at">x=</span><span class="st">`</span><span class="at">Gene 1</span><span class="st">`</span>)) <span class="sc">+</span></span>
<span id="cb164-2"><a href="data-science.html#cb164-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">bins=</span><span class="dv">30</span>,<span class="at">fill=</span><span class="st">&quot;#a93c13&quot;</span>) <span class="sc">+</span></span>
<span id="cb164-3"><a href="data-science.html#cb164-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept=</span><span class="fu">mean</span>(gene_exp<span class="sc">$</span><span class="st">`</span><span class="at">Gene 1</span><span class="st">`</span>))</span></code></pre></div>
<p><img src="main_files/data%20sci%20summ%20hist2-1.png" width="672" /></p>
<p>By eye, the mean does seem to correspond well to the value that is among the
most frequent, and successfully captures an important aspect of the data: its
<em>central tendency</em>. Summaries that compute a single number are called <em>point
estimates</em>. Point estimates collapse the data into one singular point, one
value. The arithmetic mean is just one of these measures of central tendency,
computed by taking the sum of all the values and dividing by the number of
values. The mean may be a good point estimate of the central tendency of a
dataset, but it is sensitive to outlier samples. Consider the following
examples:</p>
<div class="sourceCode" id="cb165"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb165-1"><a href="data-science.html#cb165-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(patchwork)</span>
<span id="cb165-2"><a href="data-science.html#cb165-2" aria-hidden="true" tabindex="-1"></a>well_behaved_data <span class="ot">&lt;-</span> <span class="fu">tibble</span>(<span class="at">data =</span> <span class="fu">rnorm</span>(<span class="dv">1000</span>))</span>
<span id="cb165-3"><a href="data-science.html#cb165-3" aria-hidden="true" tabindex="-1"></a>data_w_outliers <span class="ot">&lt;-</span> <span class="fu">tibble</span>(<span class="at">data =</span> <span class="fu">c</span>(<span class="fu">rnorm</span>(<span class="dv">800</span>), <span class="fu">rep</span>(<span class="dv">5</span>, <span class="dv">200</span>))) <span class="co"># oops I add some outliers :^)</span></span>
<span id="cb165-4"><a href="data-science.html#cb165-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb165-5"><a href="data-science.html#cb165-5" aria-hidden="true" tabindex="-1"></a>g_no_outlier <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(well_behaved_data, <span class="fu">aes</span>(<span class="at">x =</span> data)) <span class="sc">+</span></span>
<span id="cb165-6"><a href="data-science.html#cb165-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">fill =</span> <span class="st">&quot;#56CBF9&quot;</span>, <span class="at">color =</span> <span class="st">&quot;grey&quot;</span>, <span class="at">bins =</span> <span class="dv">30</span>) <span class="sc">+</span></span>
<span id="cb165-7"><a href="data-science.html#cb165-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="fu">mean</span>(well_behaved_data<span class="sc">$</span>data)) <span class="sc">+</span></span>
<span id="cb165-8"><a href="data-science.html#cb165-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">&quot;Mean example, no outliers&quot;</span>)</span>
<span id="cb165-9"><a href="data-science.html#cb165-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb165-10"><a href="data-science.html#cb165-10" aria-hidden="true" tabindex="-1"></a>g_outlier <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(data_w_outliers, <span class="fu">aes</span>(<span class="at">x =</span> data)) <span class="sc">+</span></span>
<span id="cb165-11"><a href="data-science.html#cb165-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">fill =</span> <span class="st">&quot;#7FBEEB&quot;</span>, <span class="at">color =</span> <span class="st">&quot;grey&quot;</span>, <span class="at">bins =</span> <span class="dv">30</span>) <span class="sc">+</span></span>
<span id="cb165-12"><a href="data-science.html#cb165-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="fu">mean</span>(data_w_outliers<span class="sc">$</span>data)) <span class="sc">+</span></span>
<span id="cb165-13"><a href="data-science.html#cb165-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">&quot;Mean example, big outliers&quot;</span>)</span>
<span id="cb165-14"><a href="data-science.html#cb165-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb165-15"><a href="data-science.html#cb165-15" aria-hidden="true" tabindex="-1"></a>g_no_outlier <span class="sc">|</span> g_outlier</span></code></pre></div>
<p><img src="main_files/unnamed-chunk-43-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>The median is another measure of central tendency, which is found by identifying
the value that divides the samples into equal halves when sorted from smallest
to largest. The median is more robust in the presence of outliers.</p>
<div class="sourceCode" id="cb166"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb166-1"><a href="data-science.html#cb166-1" aria-hidden="true" tabindex="-1"></a>g_no_outlier <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(well_behaved_data, <span class="fu">aes</span>(<span class="at">x =</span> data)) <span class="sc">+</span></span>
<span id="cb166-2"><a href="data-science.html#cb166-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">fill =</span> <span class="st">&quot;#AFBED1&quot;</span>, <span class="at">color =</span> <span class="st">&quot;white&quot;</span>, <span class="at">bins =</span> <span class="dv">30</span>) <span class="sc">+</span></span>
<span id="cb166-3"><a href="data-science.html#cb166-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="fu">median</span>(well_behaved_data<span class="sc">$</span>data)) <span class="sc">+</span></span>
<span id="cb166-4"><a href="data-science.html#cb166-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">&quot;Median example&quot;</span>) <span class="sc">+</span> <span class="fu">xlab</span>(<span class="st">&quot;&quot;</span>) <span class="sc">+</span> <span class="fu">ylab</span>(<span class="st">&quot;&quot;</span>)</span>
<span id="cb166-5"><a href="data-science.html#cb166-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb166-6"><a href="data-science.html#cb166-6" aria-hidden="true" tabindex="-1"></a>g_outlier <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(data_w_outliers, <span class="fu">aes</span>(<span class="at">x =</span> data)) <span class="sc">+</span></span>
<span id="cb166-7"><a href="data-science.html#cb166-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">fill =</span> <span class="st">&quot;#7FBEEB&quot;</span>, <span class="at">color =</span> <span class="st">&quot;grey&quot;</span>, <span class="at">bins =</span> <span class="dv">30</span>) <span class="sc">+</span></span>
<span id="cb166-8"><a href="data-science.html#cb166-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="fu">median</span>(data_w_outliers<span class="sc">$</span>data)) <span class="sc">+</span></span>
<span id="cb166-9"><a href="data-science.html#cb166-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">&quot;Median example, big outliers&quot;</span>) <span class="sc">+</span> <span class="fu">xlab</span>(<span class="st">&quot;&quot;</span>) <span class="sc">+</span> <span class="fu">ylab</span>(<span class="st">&quot;&quot;</span>)</span>
<span id="cb166-10"><a href="data-science.html#cb166-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb166-11"><a href="data-science.html#cb166-11" aria-hidden="true" tabindex="-1"></a>g_no_outlier <span class="sc">|</span> g_outlier</span></code></pre></div>
<p><img src="main_files/unnamed-chunk-44-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="dispersion" class="section level4" number="6.1.2.2">
<h4><span class="header-section-number">6.1.2.2</span> Dispersion</h4>
<p>Central tendencies are important aspects of the data but don’t describe what the
data do for values outside this point estimate of central tendency; in other
words, we have not expressed the spread, or <em>dispersion</em> of the data.</p>
<p>We decide that perhaps computing the <a href="https://en.wikipedia.org/wiki/Standard_deviation">standard
deviation</a> of the data may
characterize the spread well, since it appears to be symmetric around the mean.
We can layer this information on the plot as well to inspect it:</p>
<div class="sourceCode" id="cb167"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb167-1"><a href="data-science.html#cb167-1" aria-hidden="true" tabindex="-1"></a>g1_mean <span class="ot">&lt;-</span> <span class="fu">mean</span>(gene_exp<span class="sc">$</span><span class="st">`</span><span class="at">Gene 1</span><span class="st">`</span>)</span>
<span id="cb167-2"><a href="data-science.html#cb167-2" aria-hidden="true" tabindex="-1"></a>g1_sd <span class="ot">&lt;-</span> <span class="fu">sd</span>(gene_exp<span class="sc">$</span><span class="st">`</span><span class="at">Gene 1</span><span class="st">`</span>)</span>
<span id="cb167-3"><a href="data-science.html#cb167-3" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(gene_exp, <span class="fu">aes</span>(<span class="at">x=</span><span class="st">`</span><span class="at">Gene 1</span><span class="st">`</span>)) <span class="sc">+</span></span>
<span id="cb167-4"><a href="data-science.html#cb167-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">bins=</span><span class="dv">30</span>,<span class="at">fill=</span><span class="st">&quot;#a93c13&quot;</span>) <span class="sc">+</span></span>
<span id="cb167-5"><a href="data-science.html#cb167-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept=</span>g1_mean) <span class="sc">+</span></span>
<span id="cb167-6"><a href="data-science.html#cb167-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_segment</span>(<span class="at">x=</span>g1_mean<span class="sc">-</span>g1_sd, <span class="at">y=</span><span class="dv">10</span>, <span class="at">xend=</span>g1_mean<span class="sc">+</span>g1_sd, <span class="at">yend=</span><span class="dv">10</span>)</span></code></pre></div>
<p><img src="main_files/data%20sci%20summ%20sd-1.png" width="672" /></p>
<p>The width of the horizontal line is proportional to the mean +/- one standard
deviation around the mean, and has been placed arbitrarily on the y axis at <code>y = 10</code> to show how this range covers the data in the histogram. The +/- 1 standard
deviation around the mean visually describes the spread of the data reasonably
well.</p>
<p>Measures the spread of the data, typically around its perceived center (a mean).
Often related to the distribution of the data.</p>
<p><strong>Standard deviation</strong>: A measure of how close values are to the mean. Bigger
standard deviations mean data is more spread out.</p>
<div class="sourceCode" id="cb168"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb168-1"><a href="data-science.html#cb168-1" aria-hidden="true" tabindex="-1"></a>data <span class="ot">&lt;-</span> <span class="fu">tibble</span>(<span class="at">data =</span> <span class="fu">c</span>(<span class="fu">rnorm</span>(<span class="dv">1000</span>, <span class="at">sd=</span><span class="fl">1.75</span>)))</span>
<span id="cb168-2"><a href="data-science.html#cb168-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(data, <span class="fu">aes</span>(<span class="at">x =</span> data)) <span class="sc">+</span></span>
<span id="cb168-3"><a href="data-science.html#cb168-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">fill =</span> <span class="st">&quot;#EAC5D8&quot;</span>, <span class="at">color =</span> <span class="st">&quot;white&quot;</span>, <span class="at">bins =</span> <span class="dv">30</span>) <span class="sc">+</span></span>
<span id="cb168-4"><a href="data-science.html#cb168-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="fu">c</span>(<span class="sc">-</span><span class="dv">2</span>, <span class="sc">-</span><span class="dv">1</span>, <span class="dv">0</span>, <span class="dv">1</span>, <span class="dv">2</span>) <span class="sc">*</span> <span class="fu">sd</span>(data<span class="sc">$</span>data)) <span class="sc">+</span></span>
<span id="cb168-5"><a href="data-science.html#cb168-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span> <span class="fu">xlab</span>(<span class="st">&quot;&quot;</span>) <span class="sc">+</span> <span class="fu">ylab</span>(<span class="st">&quot;&quot;</span>) <span class="sc">+</span> <span class="fu">xlim</span>(<span class="fu">c</span>(<span class="sc">-</span><span class="dv">6</span>, <span class="dv">6</span>)) <span class="sc">+</span></span>
<span id="cb168-6"><a href="data-science.html#cb168-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">&quot;Standard deviations aplenty&quot;</span>, <span class="fu">paste</span>(<span class="st">&quot;SD:&quot;</span>, <span class="fu">sd</span>(data<span class="sc">$</span>data)))</span></code></pre></div>
<p><img src="main_files/taylor%20sd-1.png" width="672" style="display: block; margin: auto;" /></p>
<p><strong>Variance</strong>:
Similar to SD (it’s the square of SD), variance measures how far a random value
is from the mean.</p>
<div class="sourceCode" id="cb169"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb169-1"><a href="data-science.html#cb169-1" aria-hidden="true" tabindex="-1"></a>data <span class="ot">&lt;-</span> <span class="fu">tibble</span>(<span class="at">data =</span> <span class="fu">c</span>(<span class="fu">rnorm</span>(<span class="dv">1000</span>, <span class="at">sd=</span><span class="fl">0.5</span>)))</span>
<span id="cb169-2"><a href="data-science.html#cb169-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(data, <span class="fu">aes</span>(<span class="at">x =</span> data)) <span class="sc">+</span></span>
<span id="cb169-3"><a href="data-science.html#cb169-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">fill =</span> <span class="st">&quot;#DBD8F0&quot;</span>, <span class="at">color =</span> <span class="st">&quot;white&quot;</span>, <span class="at">bins =</span> <span class="dv">30</span>) <span class="sc">+</span></span>
<span id="cb169-4"><a href="data-science.html#cb169-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="fu">mean</span>(data<span class="sc">$</span>data)) <span class="sc">+</span></span>
<span id="cb169-5"><a href="data-science.html#cb169-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span> <span class="fu">xlab</span>(<span class="st">&quot;&quot;</span>) <span class="sc">+</span> <span class="fu">ylab</span>(<span class="st">&quot;&quot;</span>) <span class="sc">+</span> <span class="fu">xlim</span>(<span class="fu">c</span>(<span class="sc">-</span><span class="dv">6</span>, <span class="dv">6</span>)) <span class="sc">+</span></span>
<span id="cb169-6"><a href="data-science.html#cb169-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">&quot;Same mean as SD plot, but different variance&quot;</span>,</span>
<span id="cb169-7"><a href="data-science.html#cb169-7" aria-hidden="true" tabindex="-1"></a>          <span class="fu">paste</span>(<span class="st">&quot;Variance:&quot;</span>, <span class="fu">sd</span>(data<span class="sc">$</span>data)))</span></code></pre></div>
<p><img src="main_files/taylor%20var-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="distributions" class="section level4" number="6.1.2.3">
<h4><span class="header-section-number">6.1.2.3</span> Distributions</h4>
<p>With these two pieces of knowledge - the mean accurately describes the center of
the data and the standard deviation describes the spread - we now recognize that
these data may be <a href="https://en.wikipedia.org/wiki/Normal_distribution">normally distributed</a>, and therefore
we can potentially describe the dataset mathematically. We decide to visually
inspect this possibility by layering a normal distribution on top of our data
using
<a href="https://ggplot2.tidyverse.org/reference/geom_function.html"><code>stat_function</code></a>:</p>
<div class="sourceCode" id="cb170"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb170-1"><a href="data-science.html#cb170-1" aria-hidden="true" tabindex="-1"></a>g1_mean <span class="ot">&lt;-</span> <span class="fu">mean</span>(gene_exp<span class="sc">$</span><span class="st">`</span><span class="at">Gene 1</span><span class="st">`</span>)</span>
<span id="cb170-2"><a href="data-science.html#cb170-2" aria-hidden="true" tabindex="-1"></a>g1_sd <span class="ot">&lt;-</span> <span class="fu">sd</span>(gene_exp<span class="sc">$</span><span class="st">`</span><span class="at">Gene 1</span><span class="st">`</span>)</span>
<span id="cb170-3"><a href="data-science.html#cb170-3" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(gene_exp, <span class="fu">aes</span>(<span class="at">x=</span><span class="st">`</span><span class="at">Gene 1</span><span class="st">`</span>)) <span class="sc">+</span></span>
<span id="cb170-4"><a href="data-science.html#cb170-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(</span>
<span id="cb170-5"><a href="data-science.html#cb170-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">aes</span>(<span class="at">y=</span><span class="fu">after_stat</span>(density)),</span>
<span id="cb170-6"><a href="data-science.html#cb170-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">bins=</span><span class="dv">30</span>,</span>
<span id="cb170-7"><a href="data-science.html#cb170-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">fill=</span><span class="st">&quot;#a93c13&quot;</span></span>
<span id="cb170-8"><a href="data-science.html#cb170-8" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">+</span></span>
<span id="cb170-9"><a href="data-science.html#cb170-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">stat_function</span>(<span class="at">fun=</span>dnorm, <span class="at">args =</span> <span class="fu">list</span>(<span class="at">mean=</span>g1_mean, <span class="at">sd=</span>g1_sd), <span class="at">size=</span><span class="dv">2</span>)</span></code></pre></div>
<p><img src="main_files/unnamed-chunk-45-1.png" width="672" /></p>
<p>Note the histogram bars are scaled to the density of values in each bin to make
all the bar heights sum to 1 so that the y scale matches that of a normal
distribution.</p>
<p>We have now created our first model: we chose to express the dataset as a normal
distribution parameterized by the mean and standard deviation and standard
deviation of the data. Using the values of 254 and 11
as our mean and standard deviation, respectively, we can express our model mathematically as follows:</p>
<p><span class="math display">\[
Gene\;1 \sim \mathcal{N}(254, 11)
\]</span></p>
<p>Here the <span class="math inline">\(\sim\)</span> symbol means “distributed as” and the <span class="math inline">\(\mathcal{N}(\mu,\sigma)\)</span>
represents a normal distribution with mean of <span class="math inline">\(\mu\)</span> and standard deviation of
<span class="math inline">\(\sigma\)</span>. This is mathematical formulation means the same thing as saying we are
modeling Gene 1 expression as a normal distribution with mean of 254 and
standard deviation of 11. Without any additional information about a new sample,
we would expect the expression of that gene to be 254, although it may vary from
this value.</p>
<p>The normal distribution is the most common distribution observed in nature, but
it is hardly the only one. We could have proposed other distributions to instead
summarize our data:</p>
<div class="sourceCode" id="cb171"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb171-1"><a href="data-science.html#cb171-1" aria-hidden="true" tabindex="-1"></a>g_norm <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(<span class="fu">tibble</span>(<span class="at">data =</span> <span class="fu">rnorm</span>(<span class="dv">5000</span>)), <span class="fu">aes</span>(<span class="at">x =</span> data)) <span class="sc">+</span></span>
<span id="cb171-2"><a href="data-science.html#cb171-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">fill =</span> <span class="st">&quot;#D0FCB3&quot;</span>, <span class="at">bins =</span> <span class="dv">50</span>, <span class="at">color =</span> <span class="st">&quot;gray&quot;</span>) <span class="sc">+</span></span>
<span id="cb171-3"><a href="data-science.html#cb171-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">&quot;Normal distribution&quot;</span>, <span class="st">&quot;rnorm(n = 1000)&quot;</span>)</span>
<span id="cb171-4"><a href="data-science.html#cb171-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb171-5"><a href="data-science.html#cb171-5" aria-hidden="true" tabindex="-1"></a>g_unif <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(<span class="fu">tibble</span>(<span class="at">data =</span> <span class="fu">runif</span>(<span class="dv">5000</span>)), <span class="fu">aes</span>(<span class="at">x =</span> data)) <span class="sc">+</span></span>
<span id="cb171-6"><a href="data-science.html#cb171-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">fill =</span> <span class="st">&quot;#271F30&quot;</span>, <span class="at">bins =</span> <span class="dv">50</span>, <span class="at">color =</span> <span class="st">&quot;white&quot;</span>) <span class="sc">+</span></span>
<span id="cb171-7"><a href="data-science.html#cb171-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">&quot;Uniform distribution&quot;</span>, <span class="st">&quot;runif(n = 1000)&quot;</span>)</span>
<span id="cb171-8"><a href="data-science.html#cb171-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb171-9"><a href="data-science.html#cb171-9" aria-hidden="true" tabindex="-1"></a>g_logistic <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(<span class="fu">tibble</span>(<span class="at">data =</span> <span class="fu">rlogis</span>(<span class="dv">5000</span>)), <span class="fu">aes</span>(<span class="at">x =</span> data)) <span class="sc">+</span></span>
<span id="cb171-10"><a href="data-science.html#cb171-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">fill =</span> <span class="st">&quot;#9BC59D&quot;</span>, <span class="at">bins =</span> <span class="dv">50</span>, <span class="at">color =</span> <span class="st">&quot;black&quot;</span>) <span class="sc">+</span></span>
<span id="cb171-11"><a href="data-science.html#cb171-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">&quot;Logistic distribution&quot;</span>, <span class="st">&quot;rlogis(n = 1000)&quot;</span>)</span>
<span id="cb171-12"><a href="data-science.html#cb171-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb171-13"><a href="data-science.html#cb171-13" aria-hidden="true" tabindex="-1"></a>g_exp <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(<span class="fu">tibble</span>(<span class="at">data =</span> <span class="fu">rexp</span>(<span class="dv">5000</span>, <span class="at">rate =</span> <span class="dv">1</span>)), <span class="fu">aes</span>(<span class="at">x =</span> data)) <span class="sc">+</span></span>
<span id="cb171-14"><a href="data-science.html#cb171-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">fill =</span> <span class="st">&quot;#6C5A49&quot;</span>, <span class="at">bins =</span> <span class="dv">50</span>, <span class="at">color =</span> <span class="st">&quot;white&quot;</span>) <span class="sc">+</span></span>
<span id="cb171-15"><a href="data-science.html#cb171-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">&quot;Exponential distribution&quot;</span>, <span class="st">&quot;rexp(n = 1000, rate = 1)&quot;</span>)</span>
<span id="cb171-16"><a href="data-science.html#cb171-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb171-17"><a href="data-science.html#cb171-17" aria-hidden="true" tabindex="-1"></a>(g_norm <span class="sc">|</span> g_unif) <span class="sc">/</span> (g_logistic <span class="sc">|</span> g_exp)</span></code></pre></div>
<p><img src="main_files/unnamed-chunk-46-1.png" width="672" style="display: block; margin: auto;" />
In addition to the normal distribution, we have also plotted samples drawn from
a <a href="ihttps://en.wikipedia.org/wiki/Continuous_uniform_distribution">continuous uniform
distribution</a>
between 0 and 1, a <a href="https://en.wikipedia.org/wiki/Logistic_distribution">logistic
distribution</a> which is
similar to the normal distribution but has heavier “tails,” and an <a href="https://en.wikipedia.org/wiki/Exponential_distribution">exponential
distribution</a>.
There are many more distributions than these, and many of them were discovered
to arise in nature and encode different types of processes and relationships.</p>
</div>
<div id="final-modeling-thoughts" class="section level4" number="6.1.2.4">
<h4><span class="header-section-number">6.1.2.4</span> Final Modeling Thoughts</h4>
<p>A few notes on our data modeling example before we move on to modeling
biological data:</p>
<ol style="list-style-type: decimal">
<li><p><strong>Our model choice was totally subjective.</strong> We looked at the data and
decided that a normal distribution was a reasonable choice. There were many
other choices we could have made, and all of them would be equally valid, though
they may not all describe the data well.</p></li>
<li><p><strong>We can’t know if this is the “correct” model for the data.</strong> By eye, it
appears to be a reasonably accurate summary. However, there is no such thing as
a correct model; some models are simply better at describing the data than
others. Recall that all models are wrong, and some models are useful. Our model
may be useful, but it is definitely wrong to some extent.</p></li>
<li><p><strong>We don’t know how well our model describes the data yet.</strong> So far we’ve
only used our eyes to choose our model which might be a good starting point
considering our data are so simple, but we have not yet quantified how well our
model describes the data, or compared it to alternative models to see which is
better. This will be discussed briefly in the next section.</p></li>
</ol>
</div>
</div>
<div id="assessing-model-accuracy" class="section level3" number="6.1.3">
<h3><span class="header-section-number">6.1.3</span> Assessing Model Accuracy</h3>
</div>
<div id="linear-models-." class="section level3" number="6.1.4">
<h3><span class="header-section-number">6.1.4</span> Linear Models .</h3>
<p>What is a linear model? Conceptually, how do they work? How do you interpret the
results of a linear model? How do you implement a linear model in R?</p>
</div>
</div>
<div id="statistical-distributions-tests-." class="section level2" number="6.2">
<h2><span class="header-section-number">6.2</span> Statistical Distributions &amp; Tests .</h2>
<div id="statistical-distributions-." class="section level3" number="6.2.1">
<h3><span class="header-section-number">6.2.1</span> Statistical Distributions .</h3>
<p>What is a distribution? Why are they important? In brief terms, how are
distributions defined mathematically? What are some common distributions, and
when are they used? What do different distributions look like when plotted? How
do we work with distributions in R (e.g. dnorm, pnorm, qnorm, rnorm)?</p>
</div>
<div id="statistical-tests-." class="section level3" number="6.2.2">
<h3><span class="header-section-number">6.2.2</span> Statistical Tests .</h3>
<p>What is a statistical test, and when is it appropriate to run one? How do we
know which tests are appropriate in a given situation? What is the difference
between a parametric and non-parametric test? What are some common tests (e.g.
t-test, Chi-sq, hypergeometric, etc), and how do we run them in R?</p>
</div>
<div id="p-values-." class="section level3" number="6.2.3">
<h3><span class="header-section-number">6.2.3</span> p-values .</h3>
<p>What is a p-value? What is the relationship between a p-value and a distribution?
What does a p-value look like when plotted with a distribution? How do we
interpret p-values? How do we compute a p-value using a distribution and a
statistic using R? (Should we mention critical values?)</p>
</div>
<div id="multiple-hypothesis-testing-." class="section level3" number="6.2.4">
<h3><span class="header-section-number">6.2.4</span> Multiple Hypothesis Testing .</h3>
<p>What is multiply hypothesis testing adjustment? Why is it important? What are
some common adjustment methods (Bonferroni (FWER) and Benjamini-Hochberg (FDR))?
How do we interpret adjusted p-values (depends on the adjustment method)? How
do we compute adjusted p-values in R?</p>
</div>
<div id="statistical-power-." class="section level3" number="6.2.5">
<h3><span class="header-section-number">6.2.5</span> Statistical power .</h3>
<p>Conceptually, what is statistical power, and what does it mean? Why is it
important? What is the relationship between a dataset, an analysis, and power?
What aspects of an analysis influence the statistical power of a test? (I don’t
think it’s reasonable to ask students to perform power calculations, just to be
aware that power is a thing and generally what it is).</p>
</div>
</div>
<div id="clustering-." class="section level2" number="6.3">
<h2><span class="header-section-number">6.3</span> Clustering .</h2>
<p>What is clustering? What is the goal of clustering? How does clustering relate
to data modeling (e.g. clustering is usually hypothesis-free)? How is clustering
useful (i.e. what can we do with different clusters)? What are some common
clustering methods, and how do we implement and visualize the results in R?</p>
</div>
<div id="network-analysis-." class="section level2" number="6.4">
<h2><span class="header-section-number">6.4</span> Network Analysis .</h2>
<p>What is a network? What is network analysis? What are some example network
analysis applications in biology and bioinformatics (there is a full ksection on
[Biological Networks] in the bio chapter, these are just illustrative examples)?
How do we represent networks in R, and how do we analyze them? ([Network
visualization] is covered in the data viz chapter, might be helpful to write
these two sections together)</p>

</div>
</div>
<h3>References</h3>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-Box1976-rd" class="csl-entry">
Box, George E P. 1976. <span>“Science and Statistics.”</span> <em>J. Am. Stat. Assoc.</em> 71 (356): 791–99.
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="data-wrangling.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="biology-bioinformatics.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
