<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>8 Data Visualization | BF591 - R for Biological Sciences</title>
  <meta name="description" content="8 Data Visualization | BF591 - R for Biological Sciences" />
  <meta name="generator" content="bookdown 0.24 and GitBook 2.6.7" />

  <meta property="og:title" content="8 Data Visualization | BF591 - R for Biological Sciences" />
  <meta property="og:type" content="book" />
  
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="8 Data Visualization | BF591 - R for Biological Sciences" />
  
  
  




  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="biology-bioinformatics.html"/>
<link rel="next" href="engineering.html"/>
<script src="libs/header-attrs-2.11/header-attrs.js"></script>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.0.1/anchor-sections.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.0.1/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Syllabus</a>
<ul>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#course-schedule"><i class="fa fa-check"></i>Course Schedule</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#course-values-and-policies"><i class="fa fa-check"></i>Course Values and Policies</a></li>
</ul></li>
<li class="chapter" data-level="1" data-path="introduction.html"><a href="introduction.html"><i class="fa fa-check"></i><b>1</b> Introduction</a>
<ul>
<li class="chapter" data-level="1.1" data-path="introduction.html"><a href="introduction.html#who-this-book-is-for"><i class="fa fa-check"></i><b>1.1</b> Who This Book Is For</a></li>
<li class="chapter" data-level="1.2" data-path="introduction.html"><a href="introduction.html#a-note-about-reinventing-the-wheel"><i class="fa fa-check"></i><b>1.2</b> A Note About Reinventing the Wheel</a></li>
<li class="chapter" data-level="1.3" data-path="introduction.html"><a href="introduction.html#sources-and-references"><i class="fa fa-check"></i><b>1.3</b> Sources and References</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="data-bio.html"><a href="data-bio.html"><i class="fa fa-check"></i><b>2</b> Data in Biology</a>
<ul>
<li class="chapter" data-level="2.1" data-path="data-bio.html"><a href="data-bio.html#a-brief-history-of-data-in-molecular-biology"><i class="fa fa-check"></i><b>2.1</b> A Brief History of Data in Molecular Biology</a></li>
<li class="chapter" data-level="2.2" data-path="data-bio.html"><a href="data-bio.html#biology-as-a-mature-data-science"><i class="fa fa-check"></i><b>2.2</b> Biology as a Mature Data Science</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="preliminaries.html"><a href="preliminaries.html"><i class="fa fa-check"></i><b>3</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="3.1" data-path="preliminaries.html"><a href="preliminaries.html#prelim-r"><i class="fa fa-check"></i><b>3.1</b> The R Language</a></li>
<li class="chapter" data-level="3.2" data-path="preliminaries.html"><a href="preliminaries.html#prelim-rstudio"><i class="fa fa-check"></i><b>3.2</b> RStudio</a></li>
<li class="chapter" data-level="3.3" data-path="preliminaries.html"><a href="preliminaries.html#prog-the-r-script"><i class="fa fa-check"></i><b>3.3</b> The R Script</a></li>
<li class="chapter" data-level="3.4" data-path="preliminaries.html"><a href="preliminaries.html#prog-workflow"><i class="fa fa-check"></i><b>3.4</b> The Scripting Workflow</a></li>
<li class="chapter" data-level="3.5" data-path="preliminaries.html"><a href="preliminaries.html#prelim-git"><i class="fa fa-check"></i><b>3.5</b> git + github</a>
<ul>
<li class="chapter" data-level="3.5.1" data-path="preliminaries.html"><a href="preliminaries.html#motivation"><i class="fa fa-check"></i><b>3.5.1</b> Motivation</a></li>
<li class="chapter" data-level="3.5.2" data-path="preliminaries.html"><a href="preliminaries.html#git"><i class="fa fa-check"></i><b>3.5.2</b> git</a></li>
<li class="chapter" data-level="3.5.3" data-path="preliminaries.html"><a href="preliminaries.html#git-hosting-platforms-github"><i class="fa fa-check"></i><b>3.5.3</b> Git hosting platforms (GitHub)</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="prog-basics.html"><a href="prog-basics.html"><i class="fa fa-check"></i><b>4</b> R Programming</a>
<ul>
<li class="chapter" data-level="4.1" data-path="prog-basics.html"><a href="prog-basics.html#before-you-begin"><i class="fa fa-check"></i><b>4.1</b> Before you begin</a></li>
<li class="chapter" data-level="4.2" data-path="prog-basics.html"><a href="prog-basics.html#introduction-1"><i class="fa fa-check"></i><b>4.2</b> Introduction</a></li>
<li class="chapter" data-level="4.3" data-path="prog-basics.html"><a href="prog-basics.html#prog-r-syntax"><i class="fa fa-check"></i><b>4.3</b> R Syntax Basics</a></li>
<li class="chapter" data-level="4.4" data-path="prog-basics.html"><a href="prog-basics.html#prog-types"><i class="fa fa-check"></i><b>4.4</b> Basic Types of Values</a></li>
<li class="chapter" data-level="4.5" data-path="prog-basics.html"><a href="prog-basics.html#prog-struct"><i class="fa fa-check"></i><b>4.5</b> Data Structures</a>
<ul>
<li class="chapter" data-level="4.5.1" data-path="prog-basics.html"><a href="prog-basics.html#vectors"><i class="fa fa-check"></i><b>4.5.1</b> Vectors</a></li>
<li class="chapter" data-level="4.5.2" data-path="prog-basics.html"><a href="prog-basics.html#matrices"><i class="fa fa-check"></i><b>4.5.2</b> Matrices</a></li>
<li class="chapter" data-level="4.5.3" data-path="prog-basics.html"><a href="prog-basics.html#lists-and-data-frames"><i class="fa fa-check"></i><b>4.5.3</b> Lists and data frames</a></li>
</ul></li>
<li class="chapter" data-level="4.6" data-path="prog-basics.html"><a href="prog-basics.html#logical-tests-and-comparators"><i class="fa fa-check"></i><b>4.6</b> Logical Tests and Comparators</a></li>
<li class="chapter" data-level="4.7" data-path="prog-basics.html"><a href="prog-basics.html#functions"><i class="fa fa-check"></i><b>4.7</b> Functions</a>
<ul>
<li class="chapter" data-level="4.7.1" data-path="prog-basics.html"><a href="prog-basics.html#prog-dry"><i class="fa fa-check"></i><b>4.7.1</b> DRY: Don’t Repeat Yourself</a></li>
<li class="chapter" data-level="4.7.2" data-path="prog-basics.html"><a href="prog-basics.html#writing-your-own-functions"><i class="fa fa-check"></i><b>4.7.2</b> Writing your own functions</a></li>
<li class="chapter" data-level="4.7.3" data-path="prog-basics.html"><a href="prog-basics.html#scope"><i class="fa fa-check"></i><b>4.7.3</b> Scope</a></li>
</ul></li>
<li class="chapter" data-level="4.8" data-path="prog-basics.html"><a href="prog-basics.html#iteration"><i class="fa fa-check"></i><b>4.8</b> Iteration</a>
<ul>
<li class="chapter" data-level="4.8.1" data-path="prog-basics.html"><a href="prog-basics.html#vectorized-operations"><i class="fa fa-check"></i><b>4.8.1</b> Vectorized operations</a></li>
<li class="chapter" data-level="4.8.2" data-path="prog-basics.html"><a href="prog-basics.html#functional-programming"><i class="fa fa-check"></i><b>4.8.2</b> Functional programming</a></li>
<li class="chapter" data-level="4.8.3" data-path="prog-basics.html"><a href="prog-basics.html#apply-and-friends"><i class="fa fa-check"></i><b>4.8.3</b> <code>apply()</code> and friends</a></li>
</ul></li>
<li class="chapter" data-level="4.9" data-path="prog-basics.html"><a href="prog-basics.html#installing-packages"><i class="fa fa-check"></i><b>4.9</b> Installing Packages</a></li>
<li class="chapter" data-level="4.10" data-path="prog-basics.html"><a href="prog-basics.html#prog-debugging"><i class="fa fa-check"></i><b>4.10</b> Troubleshooting and Debugging</a></li>
<li class="chapter" data-level="4.11" data-path="prog-basics.html"><a href="prog-basics.html#prog-style"><i class="fa fa-check"></i><b>4.11</b> Coding Style and Conventions</a>
<ul>
<li class="chapter" data-level="4.11.1" data-path="prog-basics.html"><a href="prog-basics.html#is-my-code-correct"><i class="fa fa-check"></i><b>4.11.1</b> Is my code correct?</a></li>
<li class="chapter" data-level="4.11.2" data-path="prog-basics.html"><a href="prog-basics.html#does-my-code-follow-the-dry-principle"><i class="fa fa-check"></i><b>4.11.2</b> Does my code follow the <span>DRY principle</span>?</a></li>
<li class="chapter" data-level="4.11.3" data-path="prog-basics.html"><a href="prog-basics.html#did-i-choose-concise-but-descriptive-variable-and-function-names"><i class="fa fa-check"></i><b>4.11.3</b> Did I choose concise but descriptive variable and function names?</a></li>
<li class="chapter" data-level="4.11.4" data-path="prog-basics.html"><a href="prog-basics.html#did-i-use-indentation-and-naming-conventions-consistently-throughout-my-code"><i class="fa fa-check"></i><b>4.11.4</b> Did I use indentation and naming conventions consistently throughout my code?</a></li>
<li class="chapter" data-level="4.11.5" data-path="prog-basics.html"><a href="prog-basics.html#did-i-write-comments-especially-when-what-the-code-does-is-not-obvious"><i class="fa fa-check"></i><b>4.11.5</b> Did I write comments, especially when what the code does is not obvious?</a></li>
<li class="chapter" data-level="4.11.6" data-path="prog-basics.html"><a href="prog-basics.html#how-easy-would-it-be-for-someone-else-to-understand-my-code"><i class="fa fa-check"></i><b>4.11.6</b> How easy would it be for someone else to understand my code?</a></li>
<li class="chapter" data-level="4.11.7" data-path="prog-basics.html"><a href="prog-basics.html#is-my-code-easy-to-maintainchange"><i class="fa fa-check"></i><b>4.11.7</b> Is my code easy to maintain/change?</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="5" data-path="data-wrangling.html"><a href="data-wrangling.html"><i class="fa fa-check"></i><b>5</b> Data Wrangling</a>
<ul>
<li class="chapter" data-level="5.1" data-path="data-wrangling.html"><a href="data-wrangling.html#the-tidyverse"><i class="fa fa-check"></i><b>5.1</b> The Tidyverse</a></li>
<li class="chapter" data-level="5.2" data-path="data-wrangling.html"><a href="data-wrangling.html#dw-basics"><i class="fa fa-check"></i><b>5.2</b> Tidyverse Basics</a></li>
<li class="chapter" data-level="5.3" data-path="data-wrangling.html"><a href="data-wrangling.html#importing-data"><i class="fa fa-check"></i><b>5.3</b> Importing Data</a></li>
<li class="chapter" data-level="5.4" data-path="data-wrangling.html"><a href="data-wrangling.html#data-tibble"><i class="fa fa-check"></i><b>5.4</b> The tibble</a></li>
<li class="chapter" data-level="5.5" data-path="data-wrangling.html"><a href="data-wrangling.html#pipes"><i class="fa fa-check"></i><b>5.5</b> pipes</a></li>
<li class="chapter" data-level="5.6" data-path="data-wrangling.html"><a href="data-wrangling.html#tidying-data"><i class="fa fa-check"></i><b>5.6</b> Tidying Data</a></li>
<li class="chapter" data-level="5.7" data-path="data-wrangling.html"><a href="data-wrangling.html#arranging-data"><i class="fa fa-check"></i><b>5.7</b> Arranging Data</a>
<ul>
<li class="chapter" data-level="5.7.1" data-path="data-wrangling.html"><a href="data-wrangling.html#dplyrmutate---create-new-columns-using-other-columns"><i class="fa fa-check"></i><b>5.7.1</b> <code>dplyr::mutate()</code> - Create new columns using other columns</a></li>
<li class="chapter" data-level="5.7.2" data-path="data-wrangling.html"><a href="data-wrangling.html#stringr---working-with-character-values"><i class="fa fa-check"></i><b>5.7.2</b> <code>stringr</code> - Working with character values</a></li>
<li class="chapter" data-level="5.7.3" data-path="data-wrangling.html"><a href="data-wrangling.html#dplyrselect---subset-columns-by-name"><i class="fa fa-check"></i><b>5.7.3</b> <code>dplyr::select()</code> - Subset Columns by Name</a></li>
<li class="chapter" data-level="5.7.4" data-path="data-wrangling.html"><a href="data-wrangling.html#dplyrfilter---pick-rows-out-of-a-data-set"><i class="fa fa-check"></i><b>5.7.4</b> <code>dplyr::filter()</code> - Pick rows out of a data set</a></li>
<li class="chapter" data-level="5.7.5" data-path="data-wrangling.html"><a href="data-wrangling.html#dplyrarrange---order-rows-based-on-their-values"><i class="fa fa-check"></i><b>5.7.5</b> <code>dplyr::arrange()</code> - Order rows based on their values</a></li>
<li class="chapter" data-level="5.7.6" data-path="data-wrangling.html"><a href="data-wrangling.html#putting-it-all-together"><i class="fa fa-check"></i><b>5.7.6</b> Putting it all together</a></li>
</ul></li>
<li class="chapter" data-level="5.8" data-path="data-wrangling.html"><a href="data-wrangling.html#grouping-data"><i class="fa fa-check"></i><b>5.8</b> Grouping Data</a></li>
<li class="chapter" data-level="5.9" data-path="data-wrangling.html"><a href="data-wrangling.html#relational-data"><i class="fa fa-check"></i><b>5.9</b> Relational Data</a>
<ul>
<li class="chapter" data-level="5.9.1" data-path="data-wrangling.html"><a href="data-wrangling.html#dealing-with-multiple-matches"><i class="fa fa-check"></i><b>5.9.1</b> Dealing with multiple matches</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="6" data-path="data-science.html"><a href="data-science.html"><i class="fa fa-check"></i><b>6</b> Data Science</a>
<ul>
<li class="chapter" data-level="6.1" data-path="data-science.html"><a href="data-science.html#ds-dists"><i class="fa fa-check"></i><b>6.1</b> Distributions &amp; Tests</a></li>
<li class="chapter" data-level="6.2" data-path="data-science.html"><a href="data-science.html#ds-summ"><i class="fa fa-check"></i><b>6.2</b> Data summarization &amp; cleaning</a></li>
<li class="chapter" data-level="6.3" data-path="data-science.html"><a href="data-science.html#ds-clust"><i class="fa fa-check"></i><b>6.3</b> Clustering</a></li>
<li class="chapter" data-level="6.4" data-path="data-science.html"><a href="data-science.html#ds-model"><i class="fa fa-check"></i><b>6.4</b> Data Modeling</a></li>
<li class="chapter" data-level="6.5" data-path="data-science.html"><a href="data-science.html#ds-networks"><i class="fa fa-check"></i><b>6.5</b> Network Analysis</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html"><i class="fa fa-check"></i><b>7</b> Biology &amp; Bioinformatics</a>
<ul>
<li class="chapter" data-level="7.1" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#bio-r"><i class="fa fa-check"></i><b>7.1</b> R in Biology</a></li>
<li class="chapter" data-level="7.2" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#csv-files"><i class="fa fa-check"></i><b>7.2</b> CSV Files</a></li>
<li class="chapter" data-level="7.3" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#bio-data-types"><i class="fa fa-check"></i><b>7.3</b> Types of Biological Data</a></li>
<li class="chapter" data-level="7.4" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#bio-bioconductor"><i class="fa fa-check"></i><b>7.4</b> Bioconductor</a></li>
<li class="chapter" data-level="7.5" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#bio-geneexp"><i class="fa fa-check"></i><b>7.5</b> Gene Expression</a></li>
<li class="chapter" data-level="7.6" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#bio-micro"><i class="fa fa-check"></i><b>7.6</b> Differential Expression: Microarrays</a></li>
<li class="chapter" data-level="7.7" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#bio-gsea"><i class="fa fa-check"></i><b>7.7</b> Gene Set Enrichment Analysis</a></li>
<li class="chapter" data-level="7.8" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#bio-hts"><i class="fa fa-check"></i><b>7.8</b> High Throughput Sequencing</a></li>
<li class="chapter" data-level="7.9" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#bio-rnaseq"><i class="fa fa-check"></i><b>7.9</b> Differential Expression: RNASeq</a></li>
<li class="chapter" data-level="7.10" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#bio-networks"><i class="fa fa-check"></i><b>7.10</b> Biological Network Analysis</a></li>
<li class="chapter" data-level="7.11" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#bio-scseq"><i class="fa fa-check"></i><b>7.11</b> Single Cell Sequencing Analysis</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="data-visualization.html"><a href="data-visualization.html"><i class="fa fa-check"></i><b>8</b> Data Visualization</a>
<ul>
<li class="chapter" data-level="8.1" data-path="data-visualization.html"><a href="data-visualization.html#dv-gg"><i class="fa fa-check"></i><b>8.1</b> Grammar of Graphics</a></li>
<li class="chapter" data-level="8.2" data-path="data-visualization.html"><a href="data-visualization.html#plotting-one-dimension"><i class="fa fa-check"></i><b>8.2</b> Plotting One Dimension</a>
<ul>
<li class="chapter" data-level="8.2.1" data-path="data-visualization.html"><a href="data-visualization.html#bar-chart"><i class="fa fa-check"></i><b>8.2.1</b> Bar chart</a></li>
</ul></li>
<li class="chapter" data-level="8.3" data-path="data-visualization.html"><a href="data-visualization.html#plotting-two-or-more-dimensions"><i class="fa fa-check"></i><b>8.3</b> Plotting Two or More Dimensions</a></li>
<li class="chapter" data-level="8.4" data-path="data-visualization.html"><a href="data-visualization.html#visualizing-distributions"><i class="fa fa-check"></i><b>8.4</b> Visualizing Distributions</a></li>
<li class="chapter" data-level="8.5" data-path="data-visualization.html"><a href="data-visualization.html#other-kind-of-plots"><i class="fa fa-check"></i><b>8.5</b> Other Kind of Plots</a></li>
<li class="chapter" data-level="8.6" data-path="data-visualization.html"><a href="data-visualization.html#ds-network-viz"><i class="fa fa-check"></i><b>8.6</b> Network Viz</a></li>
<li class="chapter" data-level="8.7" data-path="data-visualization.html"><a href="data-visualization.html#tips-and-tricks"><i class="fa fa-check"></i><b>8.7</b> Tips and Tricks</a></li>
<li class="chapter" data-level="8.8" data-path="data-visualization.html"><a href="data-visualization.html#multiple-plots"><i class="fa fa-check"></i><b>8.8</b> Multiple Plots</a></li>
<li class="chapter" data-level="8.9" data-path="data-visualization.html"><a href="data-visualization.html#dv-resp"><i class="fa fa-check"></i><b>8.9</b> Responsible Plotting</a></li>
<li class="chapter" data-level="8.10" data-path="data-visualization.html"><a href="data-visualization.html#publication-ready-plots"><i class="fa fa-check"></i><b>8.10</b> Publication Ready Plots</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="engineering.html"><a href="engineering.html"><i class="fa fa-check"></i><b>9</b> EngineeRing</a>
<ul>
<li class="chapter" data-level="9.1" data-path="engineering.html"><a href="engineering.html#unit-testing"><i class="fa fa-check"></i><b>9.1</b> Unit Testing</a></li>
<li class="chapter" data-level="9.2" data-path="engineering.html"><a href="engineering.html#toolification"><i class="fa fa-check"></i><b>9.2</b> Toolification</a></li>
<li class="chapter" data-level="9.3" data-path="engineering.html"><a href="engineering.html#pipelines-workflows"><i class="fa fa-check"></i><b>9.3</b> Pipelines &amp; Workflows</a></li>
<li class="chapter" data-level="9.4" data-path="engineering.html"><a href="engineering.html#parallel-processing"><i class="fa fa-check"></i><b>9.4</b> Parallel Processing</a></li>
<li class="chapter" data-level="9.5" data-path="engineering.html"><a href="engineering.html#r-packages"><i class="fa fa-check"></i><b>9.5</b> R Packages</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="rshiny.html"><a href="rshiny.html"><i class="fa fa-check"></i><b>10</b> RShiny</a>
<ul>
<li class="chapter" data-level="10.1" data-path="rshiny.html"><a href="rshiny.html#overview"><i class="fa fa-check"></i><b>10.1</b> Overview</a></li>
<li class="chapter" data-level="10.2" data-path="rshiny.html"><a href="rshiny.html#learning-objectives"><i class="fa fa-check"></i><b>10.2</b> Learning Objectives</a></li>
<li class="chapter" data-level="10.3" data-path="rshiny.html"><a href="rshiny.html#skill-list"><i class="fa fa-check"></i><b>10.3</b> Skill List</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="communicating-with-r.html"><a href="communicating-with-r.html"><i class="fa fa-check"></i><b>11</b> Communicating with R</a>
<ul>
<li class="chapter" data-level="11.1" data-path="communicating-with-r.html"><a href="communicating-with-r.html#rmarkdown-knitr"><i class="fa fa-check"></i><b>11.1</b> RMarkdown &amp; knitr</a></li>
<li class="chapter" data-level="11.2" data-path="communicating-with-r.html"><a href="communicating-with-r.html#bookdown"><i class="fa fa-check"></i><b>11.2</b> bookdown</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="contribution-guide.html"><a href="contribution-guide.html"><i class="fa fa-check"></i><b>12</b> Contribution Guide</a>
<ul>
<li class="chapter" data-level="12.1" data-path="contribution-guide.html"><a href="contribution-guide.html#custom-blocks"><i class="fa fa-check"></i><b>12.1</b> Custom blocks</a></li>
</ul></li>
<li class="part"><span><b>I Assignments</b></span></li>
<li class="chapter" data-level="" data-path="assignment-format.html"><a href="assignment-format.html"><i class="fa fa-check"></i>Assignment Format</a></li>
<li class="chapter" data-level="" data-path="starting-an-assignment.html"><a href="starting-an-assignment.html"><i class="fa fa-check"></i>Starting an Assignment</a>
<ul>
<li><a href="starting-an-assignment.html#starting-your-own-git-repository-and-cloning-it">Starting your own <code>git</code> repository and cloning it</a>
<ul>
<li class="chapter" data-level="" data-path="starting-an-assignment.html"><a href="starting-an-assignment.html#contents"><i class="fa fa-check"></i>Contents</a></li>
<li class="chapter" data-level="" data-path="starting-an-assignment.html"><a href="starting-an-assignment.html#uhhh-various-github-classroom-details-idk.-todo-or-whatever."><i class="fa fa-check"></i>Uhhh various github classroom details, idk. TODO: or whatever.</a></li>
<li class="chapter" data-level="12.1.1" data-path="starting-an-assignment.html"><a href="starting-an-assignment.html#committing-and-pushing-with-r-and-without"><i class="fa fa-check"></i><b>12.1.1</b> Committing and Pushing with R (and without)</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="assignment-2.html"><a href="assignment-2.html"><i class="fa fa-check"></i>Assignment 2</a>
<ul>
<li class="chapter" data-level="" data-path="assignment-2.html"><a href="assignment-2.html#problem-statement"><i class="fa fa-check"></i>Problem Statement</a></li>
<li class="chapter" data-level="" data-path="assignment-2.html"><a href="assignment-2.html#learning-objectives-1"><i class="fa fa-check"></i>Learning Objectives</a></li>
<li class="chapter" data-level="" data-path="assignment-2.html"><a href="assignment-2.html#skill-list-1"><i class="fa fa-check"></i>Skill List</a></li>
<li class="chapter" data-level="" data-path="assignment-2.html"><a href="assignment-2.html#instructions"><i class="fa fa-check"></i>Instructions</a></li>
<li class="chapter" data-level="" data-path="assignment-2.html"><a href="assignment-2.html#function-details"><i class="fa fa-check"></i>Function Details</a>
<ul>
<li class="chapter" data-level="" data-path="assignment-2.html"><a href="assignment-2.html#bioconductor"><i class="fa fa-check"></i>1. Bioconductor</a></li>
<li><a href="assignment-2.html#load_expression">2. <code>load_expression()</code></a></li>
<li><a href="assignment-2.html#filter_15">3. <code>filter_15()</code></a></li>
<li><a href="assignment-2.html#affy_to_hgnc">4. <code>affy_to_hgnc()</code></a></li>
<li><a href="assignment-2.html#reduce_data">5. <code>reduce_data()</code></a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="assignment-5---deseq2.html"><a href="assignment-5---deseq2.html"><i class="fa fa-check"></i>Assignment 5 - DESeq2</a>
<ul>
<li class="chapter" data-level="" data-path="assignment-5---deseq2.html"><a href="assignment-5---deseq2.html#problem-statement-1"><i class="fa fa-check"></i>Problem Statement</a></li>
<li class="chapter" data-level="" data-path="assignment-5---deseq2.html"><a href="assignment-5---deseq2.html#learning-objectives-2"><i class="fa fa-check"></i>Learning Objectives</a></li>
<li class="chapter" data-level="" data-path="assignment-5---deseq2.html"><a href="assignment-5---deseq2.html#skill-list-2"><i class="fa fa-check"></i>Skill List</a></li>
<li class="chapter" data-level="" data-path="assignment-5---deseq2.html"><a href="assignment-5---deseq2.html#deseq2-background"><i class="fa fa-check"></i>DESeq2 Background</a></li>
<li class="chapter" data-level="" data-path="assignment-5---deseq2.html"><a href="assignment-5---deseq2.html#generating-a-counts-matrix"><i class="fa fa-check"></i>Generating a counts matrix</a></li>
<li class="chapter" data-level="" data-path="assignment-5---deseq2.html"><a href="assignment-5---deseq2.html#prefiltering-counts-matrix"><i class="fa fa-check"></i>Prefiltering Counts matrix</a></li>
<li class="chapter" data-level="" data-path="assignment-5---deseq2.html"><a href="assignment-5---deseq2.html#median-of-ratios-normalization"><i class="fa fa-check"></i>Median-of-ratios normalization</a></li>
<li class="chapter" data-level="" data-path="assignment-5---deseq2.html"><a href="assignment-5---deseq2.html#deseq2-preparation"><i class="fa fa-check"></i>DESeq2 preparation</a></li>
<li class="chapter" data-level="" data-path="assignment-5---deseq2.html"><a href="assignment-5---deseq2.html#omeara-et-al.-transcriptional-reversion-of-cardiac-myocyte-fate-during-mammalian-cardiac-regeneration.-circ-res.-feb-2015.-pmid-25477501l"><i class="fa fa-check"></i>O’Meara et al. Transcriptional Reversion of Cardiac Myocyte Fate During Mammalian Cardiac Regeneration. Circ Res. Feb 2015. PMID: 25477501l</a></li>
<li class="chapter" data-level="" data-path="assignment-5---deseq2.html"><a href="assignment-5---deseq2.html#understanding-factor-levels"><i class="fa fa-check"></i>Understanding Factor Levels</a></li>
<li class="chapter" data-level="" data-path="assignment-5---deseq2.html"><a href="assignment-5---deseq2.html#performing-differential-expression-analysis"><i class="fa fa-check"></i>Performing Differential Expression Analysis</a></li>
<li class="chapter" data-level="" data-path="assignment-5---deseq2.html"><a href="assignment-5---deseq2.html#p-value-distribution-and-multiple-test-correction"><i class="fa fa-check"></i>P-value distribution and multiple test correction</a></li>
<li class="chapter" data-level="" data-path="assignment-5---deseq2.html"><a href="assignment-5---deseq2.html#histogram-of-log-fold-changes"><i class="fa fa-check"></i>Histogram of Log Fold Changes</a></li>
<li class="chapter" data-level="" data-path="assignment-5---deseq2.html"><a href="assignment-5---deseq2.html#make-a-volcano-plot"><i class="fa fa-check"></i>Make a volcano plot</a></li>
<li class="chapter" data-level="" data-path="assignment-5---deseq2.html"><a href="assignment-5---deseq2.html#visualize-normalized-counts-for-the-ten-differentially-expressed-genes-with-the-lowest-adjusted-p-values"><i class="fa fa-check"></i>Visualize normalized counts for the ten differentially expressed genes with the lowest adjusted p-values</a></li>
<li class="chapter" data-level="" data-path="assignment-5---deseq2.html"><a href="assignment-5---deseq2.html#time-series-with-deseq2-likelihood-ratio-test-lrt"><i class="fa fa-check"></i>Time series with DESeq2 Likelihood Ratio Test (LRT)</a></li>
<li class="chapter" data-level="" data-path="assignment-5---deseq2.html"><a href="assignment-5---deseq2.html#plotting-the-normalized-counts-over-time-for-the-top-differentially-expressed-gene-from-the-lrt-analysis"><i class="fa fa-check"></i>Plotting the normalized counts over time for the top differentially expressed gene from the LRT analysis</a></li>
<li class="chapter" data-level="" data-path="assignment-5---deseq2.html"><a href="assignment-5---deseq2.html#optional-exercises"><i class="fa fa-check"></i>Optional exercises</a></li>
<li class="chapter" data-level="" data-path="assignment-5---deseq2.html"><a href="assignment-5---deseq2.html#references"><i class="fa fa-check"></i>References</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="assignment-6.html"><a href="assignment-6.html"><i class="fa fa-check"></i>Assignment 6</a>
<ul>
<li class="chapter" data-level="" data-path="assignment-6.html"><a href="assignment-6.html#problem-statement-2"><i class="fa fa-check"></i>Problem Statement</a></li>
<li class="chapter" data-level="" data-path="assignment-6.html"><a href="assignment-6.html#learning-objectives-3"><i class="fa fa-check"></i>Learning Objectives</a></li>
<li class="chapter" data-level="" data-path="assignment-6.html"><a href="assignment-6.html#skill-list-3"><i class="fa fa-check"></i>Skill List</a></li>
<li class="chapter" data-level="" data-path="assignment-6.html"><a href="assignment-6.html#instructions-1"><i class="fa fa-check"></i>Instructions</a></li>
</ul></li>
<li class="appendix"><span><b>Appendix</b></span></li>
<li class="chapter" data-level="A" data-path="class-outlines.html"><a href="class-outlines.html"><i class="fa fa-check"></i><b>A</b> Class Outlines</a>
<ul>
<li class="chapter" data-level="A.1" data-path="class-outlines.html"><a href="class-outlines.html#week-1"><i class="fa fa-check"></i><b>A.1</b> Week 1</a></li>
<li class="chapter" data-level="A.2" data-path="class-outlines.html"><a href="class-outlines.html#week-2"><i class="fa fa-check"></i><b>A.2</b> Week 2</a></li>
</ul></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">BF591 - R for Biological Sciences</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="data-visualization" class="section level1" number="8">
<h1><span class="header-section-number">8</span> Data Visualization</h1>
<p>Data visualization is a core component of both exploring data and communicating
results to others. The goal of data visualization is to present data in a
graphical way that shows the reader patterns that would not otherwise be
visible. Despite its ubiquity and importance, effective data visualization is
challenging, and while many tools and approaches exist there is no “gold
standard” to follow in any meaningful sense. Rather, an effective visualization
has the following properties:</p>
<ol style="list-style-type: decimal">
<li>Depicts accurate data</li>
<li>Depicts data accurately</li>
<li>Shows enough, but not too much, of the data for the viewer to gain insight</li>
<li>Is self contained - no additional information (except a caption) is required
to understand the contents of the figure</li>
</ol>
<p>Beyond these, a <em>great</em> visualization has some additional properties:</p>
<ol start="5" style="list-style-type: decimal">
<li>Exposes patterns in the data not easily observable by other methods</li>
<li>Invites the viewer to ask more questions about the data</li>
</ol>
<div id="dv-gg" class="section level2" number="8.1">
<h2><span class="header-section-number">8.1</span> Grammar of Graphics</h2>
<p>The grammar of graphics is a system of rules that describes how data and
graphical aesthetics (e.g. color, size, shape, etc) are combined to form
graphics and plots. First popularized in the book <a href="https://www.google.com/books/edition/_/ZiwLCAAAQBAJ?hl=en">The Grammar of
Graphics</a> by Leland
Wilkinson and co-authors in 1999, this grammar is a major contribution to the
structural theory of statistical graphics. In 2005, Hadley Wickam wrote an
implementation of the grammar of graphics in R called
<a href="https://ggplot2.tidyverse.org/">ggplot2</a> (gg stands for grammar of graphics).</p>
<p>Under the grammar of graphics, every plot is the combination of three types of
information: <strong>data</strong>, <strong>geometry</strong>, and <strong>aesthetics</strong>. Data is the data we
wish to plot. Geometry is the type of geometry we wish to use to depict the data
(e.g. circles, squares, lines, etc). Aesthetics connect the data to the geometry
and defines how the data controls the way the selected geometry looks.</p>
<p>A simple example will help to explain. Consider the following made up sample
metadata tibble for a study of subjects who died with Alzheimer’s Disease (AD)
and neuropathologically normal controls:</p>
<div class="sourceCode" id="cb136"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb136-1"><a href="data-visualization.html#cb136-1" aria-hidden="true" tabindex="-1"></a>ad_metadata</span></code></pre></div>
<pre><code>## # A tibble: 20 x 8
##    ID    age_at_death condition    tau  abeta   iba1   gfap braak_stage
##    &lt;chr&gt;        &lt;dbl&gt; &lt;chr&gt;      &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;       &lt;dbl&gt;
##  1 A1              73 AD         96548 176324 157501  78139           4
##  2 A2              82 AD         95251      0 147637  79348           4
##  3 A3              82 AD         40287  45365  17655 127131           2
##  4 A4              81 AD         82833 181513 225558 122615           4
##  5 A5              84 AD         73807  69332  47247  46968           3
##  6 A6              81 AD         69108 139420  57814  24885           3
##  7 A7              91 AD        138135 209045 101545  32704           6
##  8 A8              88 AD         63485 104154  80123  31620           3
##  9 A9              74 AD         59485 148125  27375  50549           3
## 10 A10             69 AD         48357  27260  47024  78507           2
## 11 C1              80 Control    62684  93739 131595 124075           3
## 12 C2              77 Control    63598  69838   7189  35597           3
## 13 C3              72 Control    52951  19444  54673  17221           2
## 14 C4              75 Control    12415   7451  16481  38510           0
## 15 C5              77 Control    62805      0  40419 108479           3
## 16 C6              77 Control     1615   3921   2439  35430           0
## 17 C7              73 Control    62233  77113  81887  96024           3
## 18 C8              77 Control    59123  94715  72350  76833           3
## 19 C9              80 Control    21705   3046  26768 113093           1
## 20 C10             73 Control    15781  16592  10271 100858           1</code></pre>
<p>For context, <a href="https://en.wikipedia.org/wiki/Tau_protein">tau protein</a> and
<a href="https://en.wikipedia.org/wiki/Amyloid_beta">amyloid beta peptides from the amyloid precursor
protein</a>
aggregate into neurofibrillary tangles and A-beta plaques, respectively, the
brains of people with AD. Generally, the amount of both of these pathologies is
associated with more severe disease. <a href="https://en.wikipedia.org/wiki/Braak_staging">Braak
stage</a> is a neuropathological
assessment of the amount of pathology in a brain that is associated with the
severity of disease, where 0 indicates absence of pathology and 6 with
widespread involvement in multiple brain regions. Aggregation of tau is also a
<a href="https://pubmed.ncbi.nlm.nih.gov/8866683/">consequence of normal aging</a>, so must
accompany neurological symptoms such as dementia to indicate an AD diagnosis
post mortem. Note we have control samples as well as AD.</p>
<div class="figure">
<img src="https://upload.wikimedia.org/wikipedia/commons/f/f7/Tauopathy_in_Alzheimer%27s_disease.jpg" alt="" />
<p class="caption">Tauopathy: tau protein accumulates in the cell bodies of affected neurons -
Wikipedia</p>
</div>
<p>The histology measures <code>tau</code>, <code>abeta</code>, <code>iba1</code>, and <code>gfap</code> have been quantified
using digital microscopy, where brain sections are stained with
immunohistochemistry to identify the location and degree of pathology; the
measures in the table are the number of pixels of a 400 x 400 pixel image of a
piece of brain tissue that fluoresce when stained with the corresponding
antibody. Tau and A-beta antibodies are specialized to the types of aggregated
proteins mentioned above and provide a quantification of the level of overall
AD pathology. Ionized calcium binding adaptor molecule 1 (IBA1) is a marker of
activated <a href="https://en.wikipedia.org/wiki/Microglia">microglia</a>, the resident
macrophages of the brain, which is an
indication of neuroinflammation. Glial fibrillary acidic protein (GFAP) is a
marker for activated <a href="https://en.wikipedia.org/wiki/Astrocyte">astrocytes</a>,
specialized cells that derive from the neuron lineage, are critical for
maintaining the <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4292164/">blood brain
barrier</a>, and are also
involved in the neuroinflammatory response.</p>
<p>Let’s say we wished to visualize the relationship between age at death and the
amount of tau pathology. A scatter plot where each marker is a subject with <span class="math inline">\(x\)</span>
and <span class="math inline">\(y\)</span> position corresponding to <code>age_at_death</code> and <code>tau</code> respectively. The
following R code creates such a plot with ggplot2:</p>
<div class="sourceCode" id="cb138"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb138-1"><a href="data-visualization.html#cb138-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="at">data=</span>ad_metadata, <span class="at">mapping =</span> <span class="fu">aes</span>(<span class="at">x =</span> age_at_death, <span class="at">y=</span>tau)) <span class="sc">+</span></span>
<span id="cb138-2"><a href="data-visualization.html#cb138-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>()</span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-42-1.png" width="672" /></p>
<p>All ggplot2 plots begin with the <code>ggplot()</code> function call, which is passed a
tibble with the <strong>data</strong> to be plotted. We then define the <strong>aesthetics</strong> are
defined by mapping the x coordinate to the <code>age_at_death</code> column and the y
coordinate to the <code>tau</code> column with <code>aes(x = age_at_death, y=tau)</code>. Finally, the
<strong>geometry</strong> as ‘point’ with <code>geom_point()</code>, meaning marks will be made at pairs
of x,y coordinates. The plot shows what we expect given our knowledge of the
relationship between age and amount of tau; the two look to be positively
correlated.</p>
<p>However, we are not capturing the whole story: we know that there are both AD
and Control subjects in this dataset. How does <code>condition</code> relate to this
relationship we see? We can layer on an additional aesthetic of color to add
this information to the plot:</p>
<div class="sourceCode" id="cb139"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb139-1"><a href="data-visualization.html#cb139-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="at">data=</span>ad_metadata, <span class="at">mapping =</span> <span class="fu">aes</span>(<span class="at">x =</span> age_at_death, <span class="at">y=</span>tau, <span class="at">color=</span>condition)) <span class="sc">+</span></span>
<span id="cb139-2"><a href="data-visualization.html#cb139-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>()</span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-43-1.png" width="672" /></p>
<p>This looks a little clearer, showing that Control subjects generally have both
an earlier age at death and a lower amount of tau pathology. This might be a
problem, however, since if the age distributions of AD and Control groups are
different that might pose a problem with
<a href="https://en.wikipedia.org/wiki/Confounding">confounding</a>. We should investigate
this.</p>
<p>Instead of plotting age at death and tau against each other, we will examine the
distributions of each of these variables for AD and Control samples separately.
We will use the <a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html">box and whisker
geometry</a> with
<code>geom_boxplot()</code> to look at the distributions of <code>age_at_death</code>:</p>
<div class="sourceCode" id="cb140"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb140-1"><a href="data-visualization.html#cb140-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="at">data=</span>ad_metadata, <span class="at">mapping =</span> <span class="fu">aes</span>(<span class="at">x=</span>condition, <span class="at">y=</span>age_at_death)) <span class="sc">+</span></span>
<span id="cb140-2"><a href="data-visualization.html#cb140-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>()</span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-44-1.png" width="672" /></p>
<p>We can see immediately that there are big differences between the age
distributions of the two groups. This is not ideal, but perhaps we can adjust
for these effects in downstream analyses. We’d like to look at the tau
distributions as well, but it would be nice to have these two plots side by side
in the same plot. To do that, we will use another library called
<a href="https://patchwork.data-imaginist.com/">patchwork</a>, which allows independent
ggplot2 plots to be arranged together with a simple expressive syntax:</p>
<div class="sourceCode" id="cb141"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb141-1"><a href="data-visualization.html#cb141-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(patchwork)</span>
<span id="cb141-2"><a href="data-visualization.html#cb141-2" aria-hidden="true" tabindex="-1"></a>age_boxplot <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(<span class="at">data=</span>ad_metadata, <span class="at">mapping =</span> <span class="fu">aes</span>(<span class="at">x=</span>condition, <span class="at">y=</span>age_at_death)) <span class="sc">+</span></span>
<span id="cb141-3"><a href="data-visualization.html#cb141-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>()</span>
<span id="cb141-4"><a href="data-visualization.html#cb141-4" aria-hidden="true" tabindex="-1"></a>tau_boxplot <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(<span class="at">data=</span>ad_metadata, <span class="at">mapping=</span><span class="fu">aes</span>(<span class="at">x=</span>condition, <span class="at">y=</span>tau)) <span class="sc">+</span></span>
<span id="cb141-5"><a href="data-visualization.html#cb141-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>()</span>
<span id="cb141-6"><a href="data-visualization.html#cb141-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb141-7"><a href="data-visualization.html#cb141-7" aria-hidden="true" tabindex="-1"></a>age_boxplot <span class="sc">|</span> tau_boxplot <span class="co"># this puts the plots side by side</span></span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-45-1.png" width="672" /></p>
<p>This confirms our suspicion, and also reveals a serious problem with our
samples: we have strong confounding of tau and age at death between AD and
Control samples. This means that if we look for differences between AD and
Control, we won’t know if the difference is due to the amount of tau pathology
or due to age of the subjects. With this sample set, we simply cannot
confidently answer that question. Just a few simple plots alerted us to this
problem; hopefully more expensive datasets have not already been generated for
these samples, so that hopefully different subjects are available that could
avoid this confounding.</p>
<p>This has been a biological data analysis oriented tutorial on plotting meant to
illustrate the principles of the grammar of graphics. Namely, every plot has
<strong>data</strong>, <strong>geometry</strong>, and <strong>aesthetics</strong> that can be independently controlled
to produce many types of plots. Many of these plots have names, like scatter
plots and boxplots, but as you compose different types of geometries and
aesthetics together you may find yourself generating plots that aren’t so easily
named.</p>
<p>The next sections of this chapter are a kind of “cook book” of different kinds
plots you can generate with data of different shapes. It is not intended to be
comprehensive, but a helpful guide when you are trying to decide how to
visualize your own datasets.</p>
<div class="box note">
<p>If you want to go directly to the comprehensive documentation of the many types
of ggplot2 plots, peruse the <a href="https://www.r-graph-gallery.com/">R Graph Gallery</a>
site.</p>
</div>
<div class="box readmore">
<ul>
<li><a href="https://ggplot2-book.org/index.html">ggplot2 - Elegant Graphics for Data Analysis</a>, by Hadley Wickam</li>
<li><a href="https://r4ds.had.co.nz/data-visualisation.html">R for Data Science - Data Vizualization</a></li>
<li><a href="https://patchwork.data-imaginist.com/">patchwork package</a></li>
</ul>
</div>
</div>
<div id="plotting-one-dimension" class="section level2" number="8.2">
<h2><span class="header-section-number">8.2</span> Plotting One Dimension</h2>
<p>The simplest plots involve plotting a single vector of numbers, or several such
vectors (e.g. for different samples). Each value in the vector typically
corresponds to a category or fixed value, for example the <code>tau</code> column from the
example above has pairs of <em>(ID, tau value)</em>. The order of these numbers can be
changed, but the vector remains one dimensional.</p>
<div id="bar-chart" class="section level3" number="8.2.1">
<h3><span class="header-section-number">8.2.1</span> Bar chart</h3>
<p>Bar charts map length (i.e. the height or width of a box) to the scalar value of
a number. The difference in visual length can help the viewer notice consistent
patterns in groups of bars, depending on how they are arranged:</p>
<div class="sourceCode" id="cb142"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb142-1"><a href="data-visualization.html#cb142-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(ad_metadata, <span class="at">mapping =</span> <span class="fu">aes</span>(<span class="at">x=</span>ID,<span class="at">y=</span>tau)) <span class="sc">+</span></span>
<span id="cb142-2"><a href="data-visualization.html#cb142-2" aria-hidden="true" tabindex="-1"></a><span class="fu">geom_bar</span>(<span class="at">stat=</span><span class="st">&quot;identity&quot;</span>)</span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-46-1.png" width="672" /></p>
<p>Note the <code>stat="identity"</code> argument is required because by default <code>geom_bar</code>
counts the number of values for each value of x, which in our case is only ever
one. This plot is not particularly helpful, so let’s change the fill color of
the bars based on condition:</p>
<div class="sourceCode" id="cb143"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb143-1"><a href="data-visualization.html#cb143-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(ad_metadata, <span class="at">mapping =</span> <span class="fu">aes</span>(<span class="at">x=</span>ID,<span class="at">y=</span>tau,<span class="at">fill=</span>condition)) <span class="sc">+</span></span>
<span id="cb143-2"><a href="data-visualization.html#cb143-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_bar</span>(<span class="at">stat=</span><span class="st">&quot;identity&quot;</span>)</span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-47-1.png" width="672" /></p>
<p>Slightly better, but maybe we can see even more clearly if we sort our tibble
by tau first:</p>
<div class="sourceCode" id="cb144"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb144-1"><a href="data-visualization.html#cb144-1" aria-hidden="true" tabindex="-1"></a>ad_metadata <span class="sc">%&gt;%</span></span>
<span id="cb144-2"><a href="data-visualization.html#cb144-2" aria-hidden="true" tabindex="-1"></a><span class="fu">arrange</span>(tau) <span class="sc">%&gt;%</span></span>
<span id="cb144-3"><a href="data-visualization.html#cb144-3" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="at">mapping =</span> <span class="fu">aes</span>(<span class="at">x=</span>ID,<span class="at">y=</span>tau,<span class="at">fill=</span>condition)) <span class="sc">+</span></span>
<span id="cb144-4"><a href="data-visualization.html#cb144-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_bar</span>(<span class="at">stat=</span><span class="st">&quot;identity&quot;</span>)</span></code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-48-1.png" width="672" /></p>
<p>But wait, our bars are in the same order as in the previous plot. Sorting our
data prior to plotting did not have the desired effect. The reason is that
although the <code>ID</code> column is a character string in our original tibble, ggplot
converts it into a <a href="#prog-factor">factor</a> prior to plotting. Recall that a
factor is a numeric type composed of sequential integers, each of which having a
character label associated with it.</p>
<p>Lollipop plots</p>
<p>Parallel Coordinate plots</p>
<p>Area/Stacked Area charts</p>
</div>
</div>
<div id="plotting-two-or-more-dimensions" class="section level2" number="8.3">
<h2><span class="header-section-number">8.3</span> Plotting Two or More Dimensions</h2>
<p>Scatter plot</p>
<p>Line plot</p>
<p>Bubble plot</p>
<p>Heatmaps</p>
</div>
<div id="visualizing-distributions" class="section level2" number="8.4">
<h2><span class="header-section-number">8.4</span> Visualizing Distributions</h2>
<p>Box plot</p>
<p>Violin plot</p>
<p>Beeswarm plot</p>
<p>Histogram</p>
<p>Density</p>
<p>Ridgeline</p>
</div>
<div id="other-kind-of-plots" class="section level2" number="8.5">
<h2><span class="header-section-number">8.5</span> Other Kind of Plots</h2>
<p>Chord</p>
<p>Sankey</p>
<p>Dendrogram</p>
</div>
<div id="ds-network-viz" class="section level2" number="8.6">
<h2><span class="header-section-number">8.6</span> Network Viz</h2>
</div>
<div id="tips-and-tricks" class="section level2" number="8.7">
<h2><span class="header-section-number">8.7</span> Tips and Tricks</h2>
<p>Confidence Intervals</p>
<p>Annotations</p>
</div>
<div id="multiple-plots" class="section level2" number="8.8">
<h2><span class="header-section-number">8.8</span> Multiple Plots</h2>
</div>
<div id="dv-resp" class="section level2" number="8.9">
<h2><span class="header-section-number">8.9</span> Responsible Plotting</h2>
<p>“Good plots empower us to ask good questions.” - Alberto Cairo, How Charts Lie</p>
</div>
<div id="publication-ready-plots" class="section level2" number="8.10">
<h2><span class="header-section-number">8.10</span> Publication Ready Plots</h2>
<div class="box readmore">
<ul>
<li><a href="https://rpkgs.datanovia.com/ggpubr/">ggpubr - publication ready plotting with ggplot</a></li>
</ul>
</div>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="biology-bioinformatics.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="engineering.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
