<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>7 Data Visualization | BF591 - R for Biological Sciences</title>
  <meta name="description" content="7 Data Visualization | BF591 - R for Biological Sciences" />
  <meta name="generator" content="bookdown 0.24 and GitBook 2.6.7" />

  <meta property="og:title" content="7 Data Visualization | BF591 - R for Biological Sciences" />
  <meta property="og:type" content="book" />
  
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="7 Data Visualization | BF591 - R for Biological Sciences" />
  
  
  




  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="data-science.html"/>
<link rel="next" href="biology-bioinformatics.html"/>
<script src="libs/header-attrs-2.11/header-attrs.js"></script>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.0.1/anchor-sections.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.0.1/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Syllabus</a>
<ul>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#course-schedule"><i class="fa fa-check"></i>Course Schedule</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#instructors"><i class="fa fa-check"></i>Instructors</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#office-hours"><i class="fa fa-check"></i>Office Hours</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#course-values-and-policies"><i class="fa fa-check"></i>Course Values and Policies</a></li>
</ul></li>
<li class="chapter" data-level="1" data-path="introduction.html"><a href="introduction.html"><i class="fa fa-check"></i><b>1</b> Introduction</a>
<ul>
<li class="chapter" data-level="1.1" data-path="introduction.html"><a href="introduction.html#who-this-book-is-for"><i class="fa fa-check"></i><b>1.1</b> Who This Book Is For</a></li>
<li class="chapter" data-level="1.2" data-path="introduction.html"><a href="introduction.html#a-note-about-reinventing-the-wheel"><i class="fa fa-check"></i><b>1.2</b> A Note About Reinventing the Wheel</a></li>
<li class="chapter" data-level="1.3" data-path="introduction.html"><a href="introduction.html#sources-and-references"><i class="fa fa-check"></i><b>1.3</b> Sources and References</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="data-bio.html"><a href="data-bio.html"><i class="fa fa-check"></i><b>2</b> Data in Biology</a>
<ul>
<li class="chapter" data-level="2.1" data-path="data-bio.html"><a href="data-bio.html#a-brief-history-of-data-in-molecular-biology"><i class="fa fa-check"></i><b>2.1</b> A Brief History of Data in Molecular Biology</a></li>
<li class="chapter" data-level="2.2" data-path="data-bio.html"><a href="data-bio.html#biology-as-a-mature-data-science"><i class="fa fa-check"></i><b>2.2</b> Biology as a Mature Data Science</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="preliminaries.html"><a href="preliminaries.html"><i class="fa fa-check"></i><b>3</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="3.1" data-path="preliminaries.html"><a href="preliminaries.html#prelim-r"><i class="fa fa-check"></i><b>3.1</b> The R Language</a></li>
<li class="chapter" data-level="3.2" data-path="preliminaries.html"><a href="preliminaries.html#prelim-rstudio"><i class="fa fa-check"></i><b>3.2</b> RStudio</a></li>
<li class="chapter" data-level="3.3" data-path="preliminaries.html"><a href="preliminaries.html#prog-the-r-script"><i class="fa fa-check"></i><b>3.3</b> The R Script</a></li>
<li class="chapter" data-level="3.4" data-path="preliminaries.html"><a href="preliminaries.html#prog-workflow"><i class="fa fa-check"></i><b>3.4</b> The Scripting Workflow</a></li>
<li class="chapter" data-level="3.5" data-path="preliminaries.html"><a href="preliminaries.html#prelim-git"><i class="fa fa-check"></i><b>3.5</b> git + github</a>
<ul>
<li class="chapter" data-level="3.5.1" data-path="preliminaries.html"><a href="preliminaries.html#motivation"><i class="fa fa-check"></i><b>3.5.1</b> Motivation</a></li>
<li class="chapter" data-level="3.5.2" data-path="preliminaries.html"><a href="preliminaries.html#git"><i class="fa fa-check"></i><b>3.5.2</b> git</a></li>
<li class="chapter" data-level="3.5.3" data-path="preliminaries.html"><a href="preliminaries.html#git-hosting-platforms-github"><i class="fa fa-check"></i><b>3.5.3</b> Git hosting platforms (GitHub)</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="prog-basics.html"><a href="prog-basics.html"><i class="fa fa-check"></i><b>4</b> R Programming</a>
<ul>
<li class="chapter" data-level="4.1" data-path="prog-basics.html"><a href="prog-basics.html#before-you-begin"><i class="fa fa-check"></i><b>4.1</b> Before you begin</a></li>
<li class="chapter" data-level="4.2" data-path="prog-basics.html"><a href="prog-basics.html#introduction-1"><i class="fa fa-check"></i><b>4.2</b> Introduction</a></li>
<li class="chapter" data-level="4.3" data-path="prog-basics.html"><a href="prog-basics.html#prog-r-syntax"><i class="fa fa-check"></i><b>4.3</b> R Syntax Basics</a></li>
<li class="chapter" data-level="4.4" data-path="prog-basics.html"><a href="prog-basics.html#prog-types"><i class="fa fa-check"></i><b>4.4</b> Basic Types of Values</a>
<ul>
<li class="chapter" data-level="4.4.1" data-path="prog-basics.html"><a href="prog-basics.html#factors-."><i class="fa fa-check"></i><b>4.4.1</b> Factors .</a></li>
</ul></li>
<li class="chapter" data-level="4.5" data-path="prog-basics.html"><a href="prog-basics.html#prog-struct"><i class="fa fa-check"></i><b>4.5</b> Data Structures</a>
<ul>
<li class="chapter" data-level="4.5.1" data-path="prog-basics.html"><a href="prog-basics.html#vectors"><i class="fa fa-check"></i><b>4.5.1</b> Vectors</a></li>
<li class="chapter" data-level="4.5.2" data-path="prog-basics.html"><a href="prog-basics.html#matrices"><i class="fa fa-check"></i><b>4.5.2</b> Matrices</a></li>
<li class="chapter" data-level="4.5.3" data-path="prog-basics.html"><a href="prog-basics.html#lists-and-data-frames"><i class="fa fa-check"></i><b>4.5.3</b> Lists and data frames</a></li>
</ul></li>
<li class="chapter" data-level="4.6" data-path="prog-basics.html"><a href="prog-basics.html#logical-tests-and-comparators"><i class="fa fa-check"></i><b>4.6</b> Logical Tests and Comparators</a></li>
<li class="chapter" data-level="4.7" data-path="prog-basics.html"><a href="prog-basics.html#functions"><i class="fa fa-check"></i><b>4.7</b> Functions</a>
<ul>
<li class="chapter" data-level="4.7.1" data-path="prog-basics.html"><a href="prog-basics.html#prog-dry"><i class="fa fa-check"></i><b>4.7.1</b> DRY: Don’t Repeat Yourself</a></li>
<li class="chapter" data-level="4.7.2" data-path="prog-basics.html"><a href="prog-basics.html#writing-your-own-functions"><i class="fa fa-check"></i><b>4.7.2</b> Writing your own functions</a></li>
<li class="chapter" data-level="4.7.3" data-path="prog-basics.html"><a href="prog-basics.html#scope"><i class="fa fa-check"></i><b>4.7.3</b> Scope</a></li>
</ul></li>
<li class="chapter" data-level="4.8" data-path="prog-basics.html"><a href="prog-basics.html#iteration"><i class="fa fa-check"></i><b>4.8</b> Iteration</a>
<ul>
<li class="chapter" data-level="4.8.1" data-path="prog-basics.html"><a href="prog-basics.html#vectorized-operations"><i class="fa fa-check"></i><b>4.8.1</b> Vectorized operations</a></li>
<li class="chapter" data-level="4.8.2" data-path="prog-basics.html"><a href="prog-basics.html#functional-programming"><i class="fa fa-check"></i><b>4.8.2</b> Functional programming</a></li>
<li class="chapter" data-level="4.8.3" data-path="prog-basics.html"><a href="prog-basics.html#apply-and-friends"><i class="fa fa-check"></i><b>4.8.3</b> <code>apply()</code> and friends</a></li>
</ul></li>
<li class="chapter" data-level="4.9" data-path="prog-basics.html"><a href="prog-basics.html#installing-packages"><i class="fa fa-check"></i><b>4.9</b> Installing Packages</a></li>
<li class="chapter" data-level="4.10" data-path="prog-basics.html"><a href="prog-basics.html#saving-and-loading-r-data"><i class="fa fa-check"></i><b>4.10</b> Saving and Loading R Data</a></li>
<li class="chapter" data-level="4.11" data-path="prog-basics.html"><a href="prog-basics.html#prog-debugging"><i class="fa fa-check"></i><b>4.11</b> Troubleshooting and Debugging</a></li>
<li class="chapter" data-level="4.12" data-path="prog-basics.html"><a href="prog-basics.html#prog-style"><i class="fa fa-check"></i><b>4.12</b> Coding Style and Conventions</a>
<ul>
<li class="chapter" data-level="4.12.1" data-path="prog-basics.html"><a href="prog-basics.html#is-my-code-correct"><i class="fa fa-check"></i><b>4.12.1</b> Is my code correct?</a></li>
<li class="chapter" data-level="4.12.2" data-path="prog-basics.html"><a href="prog-basics.html#does-my-code-follow-the-dry-principle"><i class="fa fa-check"></i><b>4.12.2</b> Does my code follow the <span>DRY principle</span>?</a></li>
<li class="chapter" data-level="4.12.3" data-path="prog-basics.html"><a href="prog-basics.html#did-i-choose-concise-but-descriptive-variable-and-function-names"><i class="fa fa-check"></i><b>4.12.3</b> Did I choose concise but descriptive variable and function names?</a></li>
<li class="chapter" data-level="4.12.4" data-path="prog-basics.html"><a href="prog-basics.html#did-i-use-indentation-and-naming-conventions-consistently-throughout-my-code"><i class="fa fa-check"></i><b>4.12.4</b> Did I use indentation and naming conventions consistently throughout my code?</a></li>
<li class="chapter" data-level="4.12.5" data-path="prog-basics.html"><a href="prog-basics.html#did-i-write-comments-especially-when-what-the-code-does-is-not-obvious"><i class="fa fa-check"></i><b>4.12.5</b> Did I write comments, especially when what the code does is not obvious?</a></li>
<li class="chapter" data-level="4.12.6" data-path="prog-basics.html"><a href="prog-basics.html#how-easy-would-it-be-for-someone-else-to-understand-my-code"><i class="fa fa-check"></i><b>4.12.6</b> How easy would it be for someone else to understand my code?</a></li>
<li class="chapter" data-level="4.12.7" data-path="prog-basics.html"><a href="prog-basics.html#is-my-code-easy-to-maintainchange"><i class="fa fa-check"></i><b>4.12.7</b> Is my code easy to maintain/change?</a></li>
<li class="chapter" data-level="4.12.8" data-path="prog-basics.html"><a href="prog-basics.html#the-styler-package-."><i class="fa fa-check"></i><b>4.12.8</b> The <code>styler</code> package .</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="5" data-path="data-wrangling.html"><a href="data-wrangling.html"><i class="fa fa-check"></i><b>5</b> Data Wrangling</a>
<ul>
<li class="chapter" data-level="5.1" data-path="data-wrangling.html"><a href="data-wrangling.html#the-tidyverse"><i class="fa fa-check"></i><b>5.1</b> The Tidyverse</a></li>
<li class="chapter" data-level="5.2" data-path="data-wrangling.html"><a href="data-wrangling.html#dw-basics"><i class="fa fa-check"></i><b>5.2</b> Tidyverse Basics</a></li>
<li class="chapter" data-level="5.3" data-path="data-wrangling.html"><a href="data-wrangling.html#importing-data"><i class="fa fa-check"></i><b>5.3</b> Importing Data</a></li>
<li class="chapter" data-level="5.4" data-path="data-wrangling.html"><a href="data-wrangling.html#data-tibble"><i class="fa fa-check"></i><b>5.4</b> The tibble</a></li>
<li class="chapter" data-level="5.5" data-path="data-wrangling.html"><a href="data-wrangling.html#tidy-data"><i class="fa fa-check"></i><b>5.5</b> Tidy Data</a></li>
<li class="chapter" data-level="5.6" data-path="data-wrangling.html"><a href="data-wrangling.html#pipes"><i class="fa fa-check"></i><b>5.6</b> pipes</a></li>
<li class="chapter" data-level="5.7" data-path="data-wrangling.html"><a href="data-wrangling.html#arranging-data"><i class="fa fa-check"></i><b>5.7</b> Arranging Data</a>
<ul>
<li class="chapter" data-level="5.7.1" data-path="data-wrangling.html"><a href="data-wrangling.html#dplyrmutate---create-new-columns-using-other-columns"><i class="fa fa-check"></i><b>5.7.1</b> <code>dplyr::mutate()</code> - Create new columns using other columns</a></li>
<li class="chapter" data-level="5.7.2" data-path="data-wrangling.html"><a href="data-wrangling.html#stringr---working-with-character-values"><i class="fa fa-check"></i><b>5.7.2</b> <code>stringr</code> - Working with character values</a></li>
<li class="chapter" data-level="5.7.3" data-path="data-wrangling.html"><a href="data-wrangling.html#dplyrselect---subset-columns-by-name"><i class="fa fa-check"></i><b>5.7.3</b> <code>dplyr::select()</code> - Subset Columns by Name</a></li>
<li class="chapter" data-level="5.7.4" data-path="data-wrangling.html"><a href="data-wrangling.html#dplyrfilter---pick-rows-out-of-a-data-set"><i class="fa fa-check"></i><b>5.7.4</b> <code>dplyr::filter()</code> - Pick rows out of a data set</a></li>
<li class="chapter" data-level="5.7.5" data-path="data-wrangling.html"><a href="data-wrangling.html#dplyrarrange---order-rows-based-on-their-values"><i class="fa fa-check"></i><b>5.7.5</b> <code>dplyr::arrange()</code> - Order rows based on their values</a></li>
<li class="chapter" data-level="5.7.6" data-path="data-wrangling.html"><a href="data-wrangling.html#putting-it-all-together"><i class="fa fa-check"></i><b>5.7.6</b> Putting it all together</a></li>
</ul></li>
<li class="chapter" data-level="5.8" data-path="data-wrangling.html"><a href="data-wrangling.html#grouping-data"><i class="fa fa-check"></i><b>5.8</b> Grouping Data</a></li>
<li class="chapter" data-level="5.9" data-path="data-wrangling.html"><a href="data-wrangling.html#rearranging-data"><i class="fa fa-check"></i><b>5.9</b> Rearranging Data</a></li>
<li class="chapter" data-level="5.10" data-path="data-wrangling.html"><a href="data-wrangling.html#relational-data"><i class="fa fa-check"></i><b>5.10</b> Relational Data</a>
<ul>
<li class="chapter" data-level="5.10.1" data-path="data-wrangling.html"><a href="data-wrangling.html#dealing-with-multiple-matches"><i class="fa fa-check"></i><b>5.10.1</b> Dealing with multiple matches</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="6" data-path="data-science.html"><a href="data-science.html"><i class="fa fa-check"></i><b>6</b> Data Science</a>
<ul>
<li class="chapter" data-level="6.1" data-path="data-science.html"><a href="data-science.html#data-modeling"><i class="fa fa-check"></i><b>6.1</b> Data Modeling</a>
<ul>
<li class="chapter" data-level="6.1.1" data-path="data-science.html"><a href="data-science.html#a-worked-modeling-example"><i class="fa fa-check"></i><b>6.1.1</b> A Worked Modeling Example</a></li>
<li class="chapter" data-level="6.1.2" data-path="data-science.html"><a href="data-science.html#data-summarization"><i class="fa fa-check"></i><b>6.1.2</b> Data Summarization</a></li>
<li class="chapter" data-level="6.1.3" data-path="data-science.html"><a href="data-science.html#linear-models"><i class="fa fa-check"></i><b>6.1.3</b> Linear Models</a></li>
<li class="chapter" data-level="6.1.4" data-path="data-science.html"><a href="data-science.html#flavors-of-linear-models"><i class="fa fa-check"></i><b>6.1.4</b> Flavors of Linear Models</a></li>
<li class="chapter" data-level="6.1.5" data-path="data-science.html"><a href="data-science.html#assessing-model-accuracy-."><i class="fa fa-check"></i><b>6.1.5</b> Assessing Model Accuracy .</a></li>
</ul></li>
<li class="chapter" data-level="6.2" data-path="data-science.html"><a href="data-science.html#statistical-distributions-tests-."><i class="fa fa-check"></i><b>6.2</b> Statistical Distributions &amp; Tests .</a>
<ul>
<li class="chapter" data-level="6.2.1" data-path="data-science.html"><a href="data-science.html#statistical-distributions"><i class="fa fa-check"></i><b>6.2.1</b> Statistical Distributions</a></li>
<li class="chapter" data-level="6.2.2" data-path="data-science.html"><a href="data-science.html#statistical-tests-."><i class="fa fa-check"></i><b>6.2.2</b> Statistical Tests .</a></li>
<li class="chapter" data-level="6.2.3" data-path="data-science.html"><a href="data-science.html#p-values-."><i class="fa fa-check"></i><b>6.2.3</b> p-values .</a></li>
<li class="chapter" data-level="6.2.4" data-path="data-science.html"><a href="data-science.html#multiple-hypothesis-testing-."><i class="fa fa-check"></i><b>6.2.4</b> Multiple Hypothesis Testing .</a></li>
<li class="chapter" data-level="6.2.5" data-path="data-science.html"><a href="data-science.html#statistical-power-."><i class="fa fa-check"></i><b>6.2.5</b> Statistical power .</a></li>
</ul></li>
<li class="chapter" data-level="6.3" data-path="data-science.html"><a href="data-science.html#exploratory-analysis"><i class="fa fa-check"></i><b>6.3</b> Exploratory Analysis</a>
<ul>
<li class="chapter" data-level="6.3.1" data-path="data-science.html"><a href="data-science.html#principal-component-analysis"><i class="fa fa-check"></i><b>6.3.1</b> Principal Component Analysis</a></li>
<li class="chapter" data-level="6.3.2" data-path="data-science.html"><a href="data-science.html#cluster-analysis"><i class="fa fa-check"></i><b>6.3.2</b> Cluster Analysis</a></li>
<li class="chapter" data-level="6.3.3" data-path="data-science.html"><a href="data-science.html#k-means-."><i class="fa fa-check"></i><b>6.3.3</b> k-means .</a></li>
<li class="chapter" data-level="6.3.4" data-path="data-science.html"><a href="data-science.html#others"><i class="fa fa-check"></i><b>6.3.4</b> Others</a></li>
</ul></li>
<li class="chapter" data-level="6.4" data-path="data-science.html"><a href="data-science.html#network-analysis-."><i class="fa fa-check"></i><b>6.4</b> Network Analysis .</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="data-visualization.html"><a href="data-visualization.html"><i class="fa fa-check"></i><b>7</b> Data Visualization</a>
<ul>
<li class="chapter" data-level="7.1" data-path="data-visualization.html"><a href="data-visualization.html#dv-gg"><i class="fa fa-check"></i><b>7.1</b> Grammar of Graphics</a></li>
<li class="chapter" data-level="7.2" data-path="data-visualization.html"><a href="data-visualization.html#plotting-one-dimension"><i class="fa fa-check"></i><b>7.2</b> Plotting One Dimension</a>
<ul>
<li class="chapter" data-level="7.2.1" data-path="data-visualization.html"><a href="data-visualization.html#dv-bar"><i class="fa fa-check"></i><b>7.2.1</b> Bar chart</a></li>
<li class="chapter" data-level="7.2.2" data-path="data-visualization.html"><a href="data-visualization.html#lollipop-plots"><i class="fa fa-check"></i><b>7.2.2</b> Lollipop plots</a></li>
<li class="chapter" data-level="7.2.3" data-path="data-visualization.html"><a href="data-visualization.html#stacked-area-charts"><i class="fa fa-check"></i><b>7.2.3</b> Stacked Area charts</a></li>
<li class="chapter" data-level="7.2.4" data-path="data-visualization.html"><a href="data-visualization.html#parallel-coordinate-plots"><i class="fa fa-check"></i><b>7.2.4</b> Parallel Coordinate plots</a></li>
</ul></li>
<li class="chapter" data-level="7.3" data-path="data-visualization.html"><a href="data-visualization.html#visualizing-distributions"><i class="fa fa-check"></i><b>7.3</b> Visualizing Distributions</a>
<ul>
<li class="chapter" data-level="7.3.1" data-path="data-visualization.html"><a href="data-visualization.html#histogram"><i class="fa fa-check"></i><b>7.3.1</b> Histogram</a></li>
<li class="chapter" data-level="7.3.2" data-path="data-visualization.html"><a href="data-visualization.html#density"><i class="fa fa-check"></i><b>7.3.2</b> Density</a></li>
<li class="chapter" data-level="7.3.3" data-path="data-visualization.html"><a href="data-visualization.html#boxplot"><i class="fa fa-check"></i><b>7.3.3</b> Boxplot</a></li>
<li class="chapter" data-level="7.3.4" data-path="data-visualization.html"><a href="data-visualization.html#violin-plot"><i class="fa fa-check"></i><b>7.3.4</b> Violin plot</a></li>
<li class="chapter" data-level="7.3.5" data-path="data-visualization.html"><a href="data-visualization.html#beeswarm-plot"><i class="fa fa-check"></i><b>7.3.5</b> Beeswarm plot</a></li>
<li class="chapter" data-level="7.3.6" data-path="data-visualization.html"><a href="data-visualization.html#ridgeline"><i class="fa fa-check"></i><b>7.3.6</b> Ridgeline</a></li>
</ul></li>
<li class="chapter" data-level="7.4" data-path="data-visualization.html"><a href="data-visualization.html#plotting-two-or-more-dimensions-."><i class="fa fa-check"></i><b>7.4</b> Plotting Two or More Dimensions .</a>
<ul>
<li class="chapter" data-level="7.4.1" data-path="data-visualization.html"><a href="data-visualization.html#scatter-plot-."><i class="fa fa-check"></i><b>7.4.1</b> Scatter plot .</a></li>
<li class="chapter" data-level="7.4.2" data-path="data-visualization.html"><a href="data-visualization.html#line-plot-."><i class="fa fa-check"></i><b>7.4.2</b> Line plot .</a></li>
<li class="chapter" data-level="7.4.3" data-path="data-visualization.html"><a href="data-visualization.html#bubble-plot-."><i class="fa fa-check"></i><b>7.4.3</b> Bubble plot .</a></li>
<li class="chapter" data-level="7.4.4" data-path="data-visualization.html"><a href="data-visualization.html#heatmaps"><i class="fa fa-check"></i><b>7.4.4</b> Heatmaps</a></li>
</ul></li>
<li class="chapter" data-level="7.5" data-path="data-visualization.html"><a href="data-visualization.html#other-kind-of-plots-."><i class="fa fa-check"></i><b>7.5</b> Other Kind of Plots .</a>
<ul>
<li class="chapter" data-level="7.5.1" data-path="data-visualization.html"><a href="data-visualization.html#chord-."><i class="fa fa-check"></i><b>7.5.1</b> Chord .</a></li>
<li class="chapter" data-level="7.5.2" data-path="data-visualization.html"><a href="data-visualization.html#sankey-."><i class="fa fa-check"></i><b>7.5.2</b> Sankey .</a></li>
<li class="chapter" data-level="7.5.3" data-path="data-visualization.html"><a href="data-visualization.html#dendrograms"><i class="fa fa-check"></i><b>7.5.3</b> Dendrograms</a></li>
</ul></li>
<li class="chapter" data-level="7.6" data-path="data-visualization.html"><a href="data-visualization.html#tips-and-tricks-."><i class="fa fa-check"></i><b>7.6</b> Tips and Tricks .</a>
<ul>
<li class="chapter" data-level="7.6.1" data-path="data-visualization.html"><a href="data-visualization.html#reordering-1-d-data-elements-."><i class="fa fa-check"></i><b>7.6.1</b> Reordering 1-D Data Elements .</a></li>
<li class="chapter" data-level="7.6.2" data-path="data-visualization.html"><a href="data-visualization.html#confidence-intervals-."><i class="fa fa-check"></i><b>7.6.2</b> Confidence Intervals .</a></li>
<li class="chapter" data-level="7.6.3" data-path="data-visualization.html"><a href="data-visualization.html#annotations-."><i class="fa fa-check"></i><b>7.6.3</b> Annotations .</a></li>
</ul></li>
<li class="chapter" data-level="7.7" data-path="data-visualization.html"><a href="data-visualization.html#multiple-plots-."><i class="fa fa-check"></i><b>7.7</b> Multiple Plots .</a>
<ul>
<li class="chapter" data-level="7.7.1" data-path="data-visualization.html"><a href="data-visualization.html#secondary-axes-."><i class="fa fa-check"></i><b>7.7.1</b> Secondary Axes .</a></li>
<li class="chapter" data-level="7.7.2" data-path="data-visualization.html"><a href="data-visualization.html#facet-wrapping-."><i class="fa fa-check"></i><b>7.7.2</b> Facet wrapping .</a></li>
<li class="chapter" data-level="7.7.3" data-path="data-visualization.html"><a href="data-visualization.html#multipanel-figures-."><i class="fa fa-check"></i><b>7.7.3</b> Multipanel Figures .</a></li>
</ul></li>
<li class="chapter" data-level="7.8" data-path="data-visualization.html"><a href="data-visualization.html#responsible-plotting-."><i class="fa fa-check"></i><b>7.8</b> Responsible Plotting .</a></li>
<li class="chapter" data-level="7.9" data-path="data-visualization.html"><a href="data-visualization.html#publication-ready-plots-."><i class="fa fa-check"></i><b>7.9</b> Publication Ready Plots .</a></li>
<li class="chapter" data-level="7.10" data-path="data-visualization.html"><a href="data-visualization.html#network-visualization-."><i class="fa fa-check"></i><b>7.10</b> Network visualization .</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html"><i class="fa fa-check"></i><b>8</b> Biology &amp; Bioinformatics</a>
<ul>
<li class="chapter" data-level="8.1" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#bio-r"><i class="fa fa-check"></i><b>8.1</b> R in Biology</a></li>
<li class="chapter" data-level="8.2" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#biological-data-overview"><i class="fa fa-check"></i><b>8.2</b> Biological Data Overview</a>
<ul>
<li class="chapter" data-level="8.2.1" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#csv-files"><i class="fa fa-check"></i><b>8.2.1</b> CSV Files</a></li>
<li class="chapter" data-level="8.2.2" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#bio-data-types"><i class="fa fa-check"></i><b>8.2.2</b> Types of Biological Data</a></li>
<li class="chapter" data-level="8.2.3" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#common-biological-data-matrices"><i class="fa fa-check"></i><b>8.2.3</b> Common Biological Data Matrices</a></li>
<li class="chapter" data-level="8.2.4" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#biological-data-is-not-tidy"><i class="fa fa-check"></i><b>8.2.4</b> Biological data is NOT Tidy!</a></li>
</ul></li>
<li class="chapter" data-level="8.3" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#bio-bioconductor"><i class="fa fa-check"></i><b>8.3</b> Bioconductor</a></li>
<li class="chapter" data-level="8.4" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#gene-identifiers"><i class="fa fa-check"></i><b>8.4</b> Gene Identifiers</a>
<ul>
<li class="chapter" data-level="8.4.1" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#gene-identifier-systems"><i class="fa fa-check"></i><b>8.4.1</b> Gene Identifier Systems</a></li>
<li class="chapter" data-level="8.4.2" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#mapping-between-identifier-systems"><i class="fa fa-check"></i><b>8.4.2</b> Mapping Between Identifier Systems</a></li>
<li class="chapter" data-level="8.4.3" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#mapping-homologs"><i class="fa fa-check"></i><b>8.4.3</b> Mapping Homologs</a></li>
</ul></li>
<li class="chapter" data-level="8.5" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#gene-expression"><i class="fa fa-check"></i><b>8.5</b> Gene Expression</a>
<ul>
<li class="chapter" data-level="8.5.1" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#gene-expression-data-in-bioconductor"><i class="fa fa-check"></i><b>8.5.1</b> Gene Expression Data in Bioconductor</a></li>
<li class="chapter" data-level="8.5.2" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#microarrays"><i class="fa fa-check"></i><b>8.5.2</b> Microarrays</a></li>
<li class="chapter" data-level="8.5.3" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#microarray-data-in-bioconductor"><i class="fa fa-check"></i><b>8.5.3</b> Microarray Data in Bioconductor</a></li>
<li class="chapter" data-level="8.5.4" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#differential-expression-analysis"><i class="fa fa-check"></i><b>8.5.4</b> Differential Expression Analysis</a></li>
<li class="chapter" data-level="8.5.5" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#differential-expression-microarrays-limma"><i class="fa fa-check"></i><b>8.5.5</b> Differential Expression: Microarrays (limma)</a></li>
<li class="chapter" data-level="8.5.6" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#high-throughput-sequencing-."><i class="fa fa-check"></i><b>8.5.6</b> High Throughput Sequencing .</a></li>
<li class="chapter" data-level="8.5.7" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#count-data-."><i class="fa fa-check"></i><b>8.5.7</b> Count Data .</a></li>
<li class="chapter" data-level="8.5.8" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#rnaseq-."><i class="fa fa-check"></i><b>8.5.8</b> RNASeq .</a></li>
<li class="chapter" data-level="8.5.9" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#differential-expression-rnaseq-."><i class="fa fa-check"></i><b>8.5.9</b> Differential Expression: RNASeq .</a></li>
</ul></li>
<li class="chapter" data-level="8.6" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#gene-set-enrichment-analysis"><i class="fa fa-check"></i><b>8.6</b> Gene Set Enrichment Analysis</a>
<ul>
<li class="chapter" data-level="8.6.1" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#gene-sets"><i class="fa fa-check"></i><b>8.6.1</b> Gene Sets</a></li>
<li class="chapter" data-level="8.6.2" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#working-with-gene-sets-in-r"><i class="fa fa-check"></i><b>8.6.2</b> Working with gene sets in R</a></li>
<li class="chapter" data-level="8.6.3" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#over-representation-analysis"><i class="fa fa-check"></i><b>8.6.3</b> Over-representation Analysis</a></li>
<li class="chapter" data-level="8.6.4" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#rank-based-analysis-."><i class="fa fa-check"></i><b>8.6.4</b> Rank-based Analysis .</a></li>
</ul></li>
<li class="chapter" data-level="8.7" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#genomic-intervals-."><i class="fa fa-check"></i><b>8.7</b> Genomic Intervals .</a></li>
<li class="chapter" data-level="8.8" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#biological-networks-."><i class="fa fa-check"></i><b>8.8</b> Biological Networks .</a>
<ul>
<li class="chapter" data-level="8.8.1" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#biological-pathways-."><i class="fa fa-check"></i><b>8.8.1</b> Biological Pathways .</a></li>
<li class="chapter" data-level="8.8.2" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#gene-regulatory-networks-."><i class="fa fa-check"></i><b>8.8.2</b> Gene Regulatory Networks .</a></li>
<li class="chapter" data-level="8.8.3" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#protein-protein-interaction-network-."><i class="fa fa-check"></i><b>8.8.3</b> Protein-Protein Interaction Network .</a></li>
<li class="chapter" data-level="8.8.4" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#wgcna-."><i class="fa fa-check"></i><b>8.8.4</b> WGCNA .</a></li>
</ul></li>
<li class="chapter" data-level="8.9" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#single-cell-sequencing-analysis-."><i class="fa fa-check"></i><b>8.9</b> Single Cell Sequencing Analysis .</a>
<ul>
<li class="chapter" data-level="8.9.1" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#single-cell-sequencing-."><i class="fa fa-check"></i><b>8.9.1</b> Single Cell Sequencing .</a></li>
<li class="chapter" data-level="8.9.2" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#single-cell-analysis-."><i class="fa fa-check"></i><b>8.9.2</b> Single Cell Analysis .</a></li>
<li class="chapter" data-level="8.9.3" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#single-cell-clustering-."><i class="fa fa-check"></i><b>8.9.3</b> Single Cell Clustering .</a></li>
<li class="chapter" data-level="8.9.4" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#dimensionality-reduction-projection-."><i class="fa fa-check"></i><b>8.9.4</b> Dimensionality Reduction &amp; Projection .</a></li>
<li class="chapter" data-level="8.9.5" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#single-cell-data-visualization-."><i class="fa fa-check"></i><b>8.9.5</b> Single Cell Data Visualization .</a></li>
<li class="chapter" data-level="8.9.6" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#single-cell-marker-analysis-."><i class="fa fa-check"></i><b>8.9.6</b> Single Cell Marker Analysis .</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="9" data-path="engineering.html"><a href="engineering.html"><i class="fa fa-check"></i><b>9</b> EngineeRing</a>
<ul>
<li class="chapter" data-level="9.1" data-path="engineering.html"><a href="engineering.html#unit-testing"><i class="fa fa-check"></i><b>9.1</b> Unit Testing</a></li>
<li class="chapter" data-level="9.2" data-path="engineering.html"><a href="engineering.html#toolification-."><i class="fa fa-check"></i><b>9.2</b> Toolification .</a></li>
<li class="chapter" data-level="9.3" data-path="engineering.html"><a href="engineering.html#pipelines-workflows-."><i class="fa fa-check"></i><b>9.3</b> Pipelines &amp; Workflows .</a></li>
<li class="chapter" data-level="9.4" data-path="engineering.html"><a href="engineering.html#parallel-processing-."><i class="fa fa-check"></i><b>9.4</b> Parallel Processing .</a></li>
<li class="chapter" data-level="9.5" data-path="engineering.html"><a href="engineering.html#r-packages-."><i class="fa fa-check"></i><b>9.5</b> R Packages .</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="rshiny.html"><a href="rshiny.html"><i class="fa fa-check"></i><b>10</b> RShiny</a>
<ul>
<li class="chapter" data-level="10.1" data-path="rshiny.html"><a href="rshiny.html#overview"><i class="fa fa-check"></i><b>10.1</b> Overview</a></li>
<li class="chapter" data-level="10.2" data-path="rshiny.html"><a href="rshiny.html#learning-objectives"><i class="fa fa-check"></i><b>10.2</b> Learning Objectives</a></li>
<li class="chapter" data-level="10.3" data-path="rshiny.html"><a href="rshiny.html#skill-list"><i class="fa fa-check"></i><b>10.3</b> Skill List</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="communicating-with-r.html"><a href="communicating-with-r.html"><i class="fa fa-check"></i><b>11</b> Communicating with R</a>
<ul>
<li class="chapter" data-level="11.1" data-path="communicating-with-r.html"><a href="communicating-with-r.html#rmarkdown-knitr"><i class="fa fa-check"></i><b>11.1</b> RMarkdown &amp; knitr</a></li>
<li class="chapter" data-level="11.2" data-path="communicating-with-r.html"><a href="communicating-with-r.html#bookdown"><i class="fa fa-check"></i><b>11.2</b> bookdown</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="contribution-guide.html"><a href="contribution-guide.html"><i class="fa fa-check"></i><b>12</b> Contribution Guide</a>
<ul>
<li class="chapter" data-level="12.1" data-path="contribution-guide.html"><a href="contribution-guide.html#custom-blocks"><i class="fa fa-check"></i><b>12.1</b> Custom blocks</a></li>
</ul></li>
<li class="part"><span><b>I Assignments</b></span></li>
<li class="chapter" data-level="" data-path="assignment-format.html"><a href="assignment-format.html"><i class="fa fa-check"></i>Assignment Format</a></li>
<li class="chapter" data-level="" data-path="starting-an-assignment.html"><a href="starting-an-assignment.html"><i class="fa fa-check"></i>Starting an Assignment</a>
<ul>
<li><a href="starting-an-assignment.html#starting-your-own-git-repository-and-cloning-it">Starting your own <code>git</code> repository and cloning it</a>
<ul>
<li class="chapter" data-level="" data-path="starting-an-assignment.html"><a href="starting-an-assignment.html#contents"><i class="fa fa-check"></i>Contents</a></li>
<li class="chapter" data-level="" data-path="starting-an-assignment.html"><a href="starting-an-assignment.html#github-tutorial"><i class="fa fa-check"></i>GitHub tutorial</a></li>
<li class="chapter" data-level="" data-path="starting-an-assignment.html"><a href="starting-an-assignment.html#committing-and-pushing-with-r-and-without"><i class="fa fa-check"></i>Committing and Pushing with R (and without)</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="assignment-1.html"><a href="assignment-1.html"><i class="fa fa-check"></i>Assignment 1</a>
<ul>
<li class="chapter" data-level="" data-path="assignment-1.html"><a href="assignment-1.html#problem-statement"><i class="fa fa-check"></i>Problem Statement</a></li>
<li class="chapter" data-level="" data-path="assignment-1.html"><a href="assignment-1.html#learning-objectives-and-skill-list"><i class="fa fa-check"></i>Learning Objectives and Skill List</a></li>
<li class="chapter" data-level="" data-path="assignment-1.html"><a href="assignment-1.html#instructions"><i class="fa fa-check"></i>Instructions</a></li>
<li class="chapter" data-level="" data-path="assignment-1.html"><a href="assignment-1.html#hints"><i class="fa fa-check"></i>Hints</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="assignment-2.html"><a href="assignment-2.html"><i class="fa fa-check"></i>Assignment 2</a>
<ul>
<li class="chapter" data-level="" data-path="assignment-2.html"><a href="assignment-2.html#problem-statement-1"><i class="fa fa-check"></i>Problem Statement</a></li>
<li class="chapter" data-level="" data-path="assignment-2.html"><a href="assignment-2.html#learning-objectives-1"><i class="fa fa-check"></i>Learning Objectives</a></li>
<li class="chapter" data-level="" data-path="assignment-2.html"><a href="assignment-2.html#skill-list-1"><i class="fa fa-check"></i>Skill List</a></li>
<li class="chapter" data-level="" data-path="assignment-2.html"><a href="assignment-2.html#instructions-1"><i class="fa fa-check"></i>Instructions</a></li>
<li class="chapter" data-level="" data-path="assignment-2.html"><a href="assignment-2.html#function-details"><i class="fa fa-check"></i>Function Details</a>
<ul>
<li class="chapter" data-level="" data-path="assignment-2.html"><a href="assignment-2.html#bioconductor"><i class="fa fa-check"></i>1. Bioconductor</a></li>
<li><a href="assignment-2.html#load_expression">2. <code>load_expression()</code></a></li>
<li><a href="assignment-2.html#filter_15">3. <code>filter_15()</code></a></li>
<li><a href="assignment-2.html#affy_to_hgnc">4. <code>affy_to_hgnc()</code></a></li>
<li><a href="assignment-2.html#reduce_data">5. <code>reduce_data()</code></a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="assignment-3.html"><a href="assignment-3.html"><i class="fa fa-check"></i>Assignment 3</a>
<ul>
<li class="chapter" data-level="" data-path="assignment-3.html"><a href="assignment-3.html#problem-statement-2"><i class="fa fa-check"></i>Problem Statement</a></li>
<li class="chapter" data-level="" data-path="assignment-3.html"><a href="assignment-3.html#learning-objectives-2"><i class="fa fa-check"></i>Learning Objectives</a></li>
<li class="chapter" data-level="" data-path="assignment-3.html"><a href="assignment-3.html#skill-list-2"><i class="fa fa-check"></i>Skill List</a></li>
<li class="chapter" data-level="" data-path="assignment-3.html"><a href="assignment-3.html#background-on-microarrays"><i class="fa fa-check"></i>Background on Microarrays</a></li>
<li class="chapter" data-level="" data-path="assignment-3.html"><a href="assignment-3.html#background-on-principal-component-analysis"><i class="fa fa-check"></i>Background on Principal Component Analysis</a></li>
<li class="chapter" data-level="" data-path="assignment-3.html"><a href="assignment-3.html#marisa-et-al.-gene-expression-classification-of-colon-cancer-into-molecular-subtypes-characterization-validation-and-prognostic-value.-plos-medicine-may-2013.-pmid-23700391"><i class="fa fa-check"></i>Marisa et al. Gene Expression Classification of Colon Cancer into Molecular Subtypes: Characterization, Validation, and Prognostic Value. PLoS Medicine, May 2013. PMID: 23700391</a></li>
<li><a href="assignment-3.html#scaling-data-using-r-scale">Scaling data using R <code>scale()</code></a></li>
<li class="chapter" data-level="" data-path="assignment-3.html"><a href="assignment-3.html#proportion-of-variance-explained"><i class="fa fa-check"></i>Proportion of variance explained</a></li>
<li class="chapter" data-level="" data-path="assignment-3.html"><a href="assignment-3.html#plotting-and-visualization-of-pca"><i class="fa fa-check"></i>Plotting and visualization of PCA</a></li>
<li class="chapter" data-level="" data-path="assignment-3.html"><a href="assignment-3.html#hierarchical-clustering-and-heatmaps"><i class="fa fa-check"></i>Hierarchical Clustering and Heatmaps</a></li>
<li class="chapter" data-level="" data-path="assignment-3.html"><a href="assignment-3.html#references"><i class="fa fa-check"></i>References</a></li>
</ul></li>
<li class="appendix"><span><b>Appendix</b></span></li>
<li class="chapter" data-level="A" data-path="class-outlines.html"><a href="class-outlines.html"><i class="fa fa-check"></i><b>A</b> Class Outlines</a>
<ul>
<li class="chapter" data-level="A.1" data-path="class-outlines.html"><a href="class-outlines.html#week-1"><i class="fa fa-check"></i><b>A.1</b> Week 1</a></li>
<li class="chapter" data-level="A.2" data-path="class-outlines.html"><a href="class-outlines.html#week-2"><i class="fa fa-check"></i><b>A.2</b> Week 2</a></li>
<li class="chapter" data-level="A.3" data-path="class-outlines.html"><a href="class-outlines.html#week-3"><i class="fa fa-check"></i><b>A.3</b> Week 3</a></li>
<li class="chapter" data-level="A.4" data-path="class-outlines.html"><a href="class-outlines.html#week-4"><i class="fa fa-check"></i><b>A.4</b> Week 4</a></li>
<li class="chapter" data-level="A.5" data-path="class-outlines.html"><a href="class-outlines.html#week-5"><i class="fa fa-check"></i><b>A.5</b> Week 5</a></li>
<li class="chapter" data-level="A.6" data-path="class-outlines.html"><a href="class-outlines.html#week-6"><i class="fa fa-check"></i><b>A.6</b> Week 6</a></li>
<li class="chapter" data-level="A.7" data-path="class-outlines.html"><a href="class-outlines.html#week-7"><i class="fa fa-check"></i><b>A.7</b> Week 7</a></li>
<li class="chapter" data-level="A.8" data-path="class-outlines.html"><a href="class-outlines.html#week-8"><i class="fa fa-check"></i><b>A.8</b> Week 8</a></li>
<li class="chapter" data-level="A.9" data-path="class-outlines.html"><a href="class-outlines.html#week-9"><i class="fa fa-check"></i><b>A.9</b> Week 9</a></li>
<li class="chapter" data-level="A.10" data-path="class-outlines.html"><a href="class-outlines.html#week-10"><i class="fa fa-check"></i><b>A.10</b> Week 10</a></li>
<li class="chapter" data-level="A.11" data-path="class-outlines.html"><a href="class-outlines.html#week-11"><i class="fa fa-check"></i><b>A.11</b> Week 11</a></li>
<li class="chapter" data-level="A.12" data-path="class-outlines.html"><a href="class-outlines.html#week-12"><i class="fa fa-check"></i><b>A.12</b> Week 12</a></li>
<li class="chapter" data-level="A.13" data-path="class-outlines.html"><a href="class-outlines.html#week-13---no-class-for-project-work"><i class="fa fa-check"></i><b>A.13</b> Week 13 - no class for project work</a></li>
<li class="chapter" data-level="A.14" data-path="class-outlines.html"><a href="class-outlines.html#week-14---no-class-for-project-work"><i class="fa fa-check"></i><b>A.14</b> Week 14 - no class for project work</a></li>
</ul></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">BF591 - R for Biological Sciences</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="data-visualization" class="section level1" number="7">
<h1><span class="header-section-number">7</span> Data Visualization</h1>
<p>Data visualization is a core component of both exploring data and communicating
results to others. The goal of data visualization is to present data in a
graphical way that shows the reader patterns that would not otherwise be
visible. Despite its ubiquity and importance, effective data visualization is
challenging, and while many tools and approaches exist there is no “gold
standard” to follow in any meaningful sense. Rather, an effective visualization
has the following properties:</p>
<ol style="list-style-type: decimal">
<li>Depicts accurate data</li>
<li>Depicts data accurately</li>
<li>Shows enough, but not too much, of the data for the viewer to gain insight</li>
<li>Is self contained - no additional information (except a caption) is required
to understand the contents of the figure</li>
</ol>
<p>Beyond these, a <em>great</em> visualization has some additional properties:</p>
<ol start="5" style="list-style-type: decimal">
<li>Exposes patterns in the data not easily observable by other methods</li>
<li>Invites the viewer to ask more questions about the data</li>
</ol>
<div id="dv-gg" class="section level2" number="7.1">
<h2><span class="header-section-number">7.1</span> Grammar of Graphics</h2>
<p>The grammar of graphics is a system of rules that describes how data and
graphical aesthetics (e.g. color, size, shape, etc) are combined to form
graphics and plots. First popularized in the book <a href="https://www.google.com/books/edition/_/ZiwLCAAAQBAJ?hl=en">The Grammar of
Graphics</a> by Leland
Wilkinson and co-authors in 1999, this grammar is a major contribution to the
structural theory of statistical graphics. In 2005, Hadley Wickam wrote an
implementation of the grammar of graphics in R called
<a href="https://ggplot2.tidyverse.org/">ggplot2</a> (gg stands for grammar of graphics).</p>
<p>Under the grammar of graphics, every plot is the combination of three types of
information: <strong>data</strong>, <strong>geometry</strong>, and <strong>aesthetics</strong>. Data is the data we
wish to plot. Geometry is the type of geometry we wish to use to depict the data
(e.g. circles, squares, lines, etc). Aesthetics connect the data to the geometry
and defines how the data controls the way the selected geometry looks.</p>
<p>A simple example will help to explain. Consider the following made up sample
metadata tibble for a study of subjects who died with Alzheimer’s Disease (AD)
and neuropathologically normal controls:</p>
<div class="sourceCode" id="cb220"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb220-1"><a href="data-visualization.html#cb220-1" aria-hidden="true" tabindex="-1"></a>ad_metadata</span></code></pre></div>
<pre><code>## # A tibble: 20 x 8
##    ID    age_at_death condition    tau  abeta   iba1   gfap braak_stage
##    &lt;chr&gt;        &lt;dbl&gt; &lt;fct&gt;      &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt; &lt;fct&gt;      
##  1 A1              81 AD         74165  35842  11761 129300 3          
##  2 A2              82 AD        138292 171416  83640  42844 6          
##  3 A3              80 AD         92759  60277 120280  41222 4          
##  4 A4              66 AD         71894  47856  67188  32846 3          
##  5 A5              75 AD         64910 153513 111125 120407 3          
##  6 A6              70 AD         39788  52851  33094  43674 2          
##  7 A7              84 AD         39993  81023  36641  79634 2          
##  8 A8              85 AD        134645 290511  41510  25664 6          
##  9 A9              73 AD         82030 285705  70156  42536 4          
## 10 A10             80 AD         57169  83868  28670  31079 2          
## 11 C1              76 Control    57708  45956  46536 100924 2          
## 12 C2              78 Control    46347  12586  27588  76055 2          
## 13 C3              67 Control    20955  22866  11881 107339 1          
## 14 C4              86 Control    48941   7898  66129  92690 2          
## 15 C5              73 Control    16914  20748  15493 131424 1          
## 16 C6              83 Control     3878   4489   2416  70401 0          
## 17 C7              76 Control    17960  14331  17473  54470 1          
## 18 C8              76 Control    25712  46254  17499 142991 1          
## 19 C9              68 Control      651    595    601 135552 0          
## 20 C10             90 Control    21876  60531  39058  76933 1</code></pre>
<p>For context, <a href="https://en.wikipedia.org/wiki/Tau_protein">tau protein</a> and
<a href="https://en.wikipedia.org/wiki/Amyloid_beta">amyloid beta peptides from the amyloid precursor
protein</a>
aggregate into neurofibrillary tangles and A-beta plaques, respectively, the
brains of people with AD. Generally, the amount of both of these pathologies is
associated with more severe disease. <a href="https://en.wikipedia.org/wiki/Braak_staging">Braak
stage</a> is a neuropathological
assessment of the amount of pathology in a brain that is associated with the
severity of disease, where 0 indicates absence of pathology and 6 with
widespread involvement in multiple brain regions. Aggregation of tau is also a
<a href="https://pubmed.ncbi.nlm.nih.gov/8866683/">consequence of normal aging</a>, so must
accompany neurological symptoms such as dementia to indicate an AD diagnosis
post mortem. Note we have control samples as well as AD.</p>
<div class="figure">
<img src="https://upload.wikimedia.org/wikipedia/commons/f/f7/Tauopathy_in_Alzheimer%27s_disease.jpg" alt="" />
<p class="caption">Tauopathy: tau protein accumulates in the cell bodies of affected neurons -
Wikipedia</p>
</div>
<p>The histology measures <code>tau</code>, <code>abeta</code>, <code>iba1</code>, and <code>gfap</code> have been quantified
using digital microscopy, where brain sections are stained with
immunohistochemistry to identify the location and degree of pathology; the
measures in the table are the number of pixels of a 400 x 400 pixel image of a
piece of brain tissue that fluoresce when stained with the corresponding
antibody. Tau and A-beta antibodies are specialized to the types of aggregated
proteins mentioned above and provide a quantification of the level of overall
AD pathology. Ionized calcium binding adaptor molecule 1 (IBA1) is a marker of
activated <a href="https://en.wikipedia.org/wiki/Microglia">microglia</a>, the resident
macrophages of the brain, which is an
indication of neuroinflammation. Glial fibrillary acidic protein (GFAP) is a
marker for activated <a href="https://en.wikipedia.org/wiki/Astrocyte">astrocytes</a>,
specialized cells that derive from the neuron lineage, are critical for
maintaining the <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4292164/">blood brain
barrier</a>, and are also
involved in the neuroinflammatory response.</p>
<p>Let’s say we wished to visualize the relationship between age at death and the
amount of tau pathology. A scatter plot where each marker is a subject with <span class="math inline">\(x\)</span>
and <span class="math inline">\(y\)</span> position corresponding to <code>age_at_death</code> and <code>tau</code> respectively. The
following R code creates such a plot with ggplot2:</p>
<div class="sourceCode" id="cb222"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb222-1"><a href="data-visualization.html#cb222-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="at">data=</span>ad_metadata, <span class="at">mapping =</span> <span class="fu">aes</span>(<span class="at">x =</span> age_at_death, <span class="at">y=</span>tau)) <span class="sc">+</span></span>
<span id="cb222-2"><a href="data-visualization.html#cb222-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>()</span></code></pre></div>
<p><img src="main_files/unnamed-chunk-66-1.png" width="672" /></p>
<p>All ggplot2 plots begin with the <code>ggplot()</code> function call, which is passed a
tibble with the <strong>data</strong> to be plotted. We then define the <strong>aesthetics</strong> are
defined by mapping the x coordinate to the <code>age_at_death</code> column and the y
coordinate to the <code>tau</code> column with <code>aes(x = age_at_death, y=tau)</code>. Finally, the
<strong>geometry</strong> as ‘point’ with <code>geom_point()</code>, meaning marks will be made at pairs
of x,y coordinates. The plot shows what we expect given our knowledge of the
relationship between age and amount of tau; the two look to be positively
correlated.</p>
<p>However, we are not capturing the whole story: we know that there are both AD
and Control subjects in this dataset. How does <code>condition</code> relate to this
relationship we see? We can layer on an additional aesthetic of color to add
this information to the plot:</p>
<div class="sourceCode" id="cb223"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb223-1"><a href="data-visualization.html#cb223-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="at">data=</span>ad_metadata, <span class="at">mapping =</span> <span class="fu">aes</span>(<span class="at">x =</span> age_at_death, <span class="at">y=</span>tau, <span class="at">color=</span>condition)) <span class="sc">+</span></span>
<span id="cb223-2"><a href="data-visualization.html#cb223-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>()</span></code></pre></div>
<p><img src="main_files/unnamed-chunk-67-1.png" width="672" /></p>
<p>This looks a little clearer, showing that Control subjects generally have both
an earlier age at death and a lower amount of tau pathology. This might be a
problem, however, since if the age distributions of AD and Control groups are
different that might pose a problem with
<a href="https://en.wikipedia.org/wiki/Confounding">confounding</a>. We should investigate
this.</p>
<p>Instead of plotting age at death and tau against each other, we will examine the
distributions of each of these variables for AD and Control samples separately.
We will use the <a href="https://ggplot2.tidyverse.org/reference/geom_violin.html">violin
geometry</a> with
<code>geom_violin()</code> to look at the distributions of <code>age_at_death</code>:</p>
<div class="sourceCode" id="cb224"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb224-1"><a href="data-visualization.html#cb224-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="at">data=</span>ad_metadata, <span class="at">mapping =</span> <span class="fu">aes</span>(<span class="at">x=</span>condition, <span class="at">y=</span>age_at_death)) <span class="sc">+</span></span>
<span id="cb224-2"><a href="data-visualization.html#cb224-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_violin</span>()</span></code></pre></div>
<p><img src="main_files/unnamed-chunk-68-1.png" width="672" /></p>
<p>We can see immediately that there are big differences between the age
distributions of the two groups. This is not ideal, but perhaps we can adjust
for these effects in downstream analyses. We’d like to look at the tau
distributions as well, but it would be nice to have these two plots side by side
in the same plot. To do that, we will use another library called
<a href="https://patchwork.data-imaginist.com/">patchwork</a>, which allows independent
ggplot2 plots to be arranged together with a simple expressive syntax:</p>
<div class="sourceCode" id="cb225"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb225-1"><a href="data-visualization.html#cb225-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(patchwork)</span>
<span id="cb225-2"><a href="data-visualization.html#cb225-2" aria-hidden="true" tabindex="-1"></a>age_boxplot <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(<span class="at">data=</span>ad_metadata, <span class="at">mapping =</span> <span class="fu">aes</span>(<span class="at">x=</span>condition, <span class="at">y=</span>age_at_death)) <span class="sc">+</span></span>
<span id="cb225-3"><a href="data-visualization.html#cb225-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>()</span>
<span id="cb225-4"><a href="data-visualization.html#cb225-4" aria-hidden="true" tabindex="-1"></a>tau_boxplot <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(<span class="at">data=</span>ad_metadata, <span class="at">mapping=</span><span class="fu">aes</span>(<span class="at">x=</span>condition, <span class="at">y=</span>tau)) <span class="sc">+</span></span>
<span id="cb225-5"><a href="data-visualization.html#cb225-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>()</span>
<span id="cb225-6"><a href="data-visualization.html#cb225-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb225-7"><a href="data-visualization.html#cb225-7" aria-hidden="true" tabindex="-1"></a>age_boxplot <span class="sc">|</span> tau_boxplot <span class="co"># this puts the plots side by side</span></span></code></pre></div>
<p><img src="main_files/unnamed-chunk-69-1.png" width="672" /></p>
<p>This confirms our suspicion, and also reveals a serious problem with our
samples: we have strong confounding of tau and age at death between AD and
Control samples. This means that if we look for differences between AD and
Control, we won’t know if the difference is due to the amount of tau pathology
or due to age of the subjects. With this sample set, we simply cannot
confidently answer that question. Just a few simple plots alerted us to this
problem; hopefully more expensive datasets have not already been generated for
these samples, so that hopefully different subjects are available that could
avoid this confounding.</p>
<p>This has been a biological data analysis oriented tutorial on plotting meant to
illustrate the principles of the grammar of graphics. Namely, every plot has
<strong>data</strong>, <strong>geometry</strong>, and <strong>aesthetics</strong> that can be independently controlled
to produce many types of plots. Many of these plots have names, like scatter
plots and boxplots, but as you compose different types of geometries and
aesthetics together you may find yourself generating plots that aren’t so easily
named.</p>
<p>The next sections of this chapter are a kind of “cook book” of different kinds
plots you can generate with data of different shapes. It is not intended to be
comprehensive, but a helpful guide when you are trying to decide how to
visualize your own datasets.</p>
<div class="box note">
<p>If you want to go directly to the comprehensive documentation of the many types
of ggplot2 plots, peruse the <a href="https://www.r-graph-gallery.com/">R Graph Gallery</a>
site.</p>
</div>
<div class="box readmore">
<ul>
<li><a href="https://ggplot2-book.org/index.html">ggplot2 - Elegant Graphics for Data Analysis</a>, by Hadley Wickam</li>
<li><a href="https://r4ds.had.co.nz/data-visualisation.html">R for Data Science - Data Vizualization</a></li>
<li><a href="https://www.r-graph-gallery.com/">R Graph Gallery</a></li>
<li><a href="https://patchwork.data-imaginist.com/">patchwork package</a></li>
</ul>
</div>

</div>
<div id="plotting-one-dimension" class="section level2" number="7.2">
<h2><span class="header-section-number">7.2</span> Plotting One Dimension</h2>
<p>The simplest plots involve plotting a single vector of numbers, or several such
vectors (e.g. for different samples). Each value in the vector typically
corresponds to a category or fixed value, for example the <code>tau</code> column from the
example above has pairs of <em>(ID, tau value)</em>. The order of these numbers can be
changed, but the vector remains one dimensional or 1-D.</p>
<div id="dv-bar" class="section level3" number="7.2.1">
<h3><span class="header-section-number">7.2.1</span> Bar chart</h3>
<p>Bar charts map length (i.e. the height or width of a box) to the scalar value of
a number. The difference in visual length can help the viewer notice consistent
patterns in groups of bars, depending on how they are arranged:</p>
<div class="sourceCode" id="cb226"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb226-1"><a href="data-visualization.html#cb226-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(ad_metadata, <span class="at">mapping =</span> <span class="fu">aes</span>(<span class="at">x=</span>ID,<span class="at">y=</span>tau)) <span class="sc">+</span></span>
<span id="cb226-2"><a href="data-visualization.html#cb226-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_bar</span>(<span class="at">stat=</span><span class="st">&quot;identity&quot;</span>)</span></code></pre></div>
<p><img src="main_files/unnamed-chunk-71-1.png" width="672" /></p>
<p>Note the <code>stat="identity"</code> argument is required because by default <code>geom_bar</code>
counts the number of values for each value of x, which in our case is only ever
one. This plot is not particularly helpful, so let’s change the fill color of
the bars based on condition:</p>
<div class="sourceCode" id="cb227"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb227-1"><a href="data-visualization.html#cb227-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(ad_metadata, <span class="at">mapping =</span> <span class="fu">aes</span>(<span class="at">x=</span>ID,<span class="at">y=</span>tau,<span class="at">fill=</span>condition)) <span class="sc">+</span></span>
<span id="cb227-2"><a href="data-visualization.html#cb227-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_bar</span>(<span class="at">stat=</span><span class="st">&quot;identity&quot;</span>)</span></code></pre></div>
<p><img src="main_files/unnamed-chunk-72-1.png" width="672" /></p>
<p>Slightly better, but maybe we can see even more clearly if we sort our tibble
by tau first. Sorting elements in these 1-D charts is somewhat complicated, and
is explained in the [Reordering 1-D Data Elements] section below.</p>
<p>Bar charts can also plot negative numbers. In the following example, we center
the tau measurements by subtracting the mean from each value before plotting:</p>
<div class="sourceCode" id="cb228"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb228-1"><a href="data-visualization.html#cb228-1" aria-hidden="true" tabindex="-1"></a><span class="fu">mutate</span>(ad_metadata, <span class="at">tau_centered=</span>(tau <span class="sc">-</span> <span class="fu">mean</span>(tau))) <span class="sc">%&gt;%</span></span>
<span id="cb228-2"><a href="data-visualization.html#cb228-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="at">mapping =</span> <span class="fu">aes</span>(<span class="at">x=</span>ID, <span class="at">y=</span>tau_centered, <span class="at">fill=</span>condition)) <span class="sc">+</span></span>
<span id="cb228-3"><a href="data-visualization.html#cb228-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_bar</span>(<span class="at">stat=</span><span class="st">&quot;identity&quot;</span>)</span></code></pre></div>
<p><img src="main_files/unnamed-chunk-73-1.png" width="672" /></p>
</div>
<div id="lollipop-plots" class="section level3" number="7.2.2">
<h3><span class="header-section-number">7.2.2</span> Lollipop plots</h3>
<p>Similar to <a href="data-visualization.html#dv-bar">bar charts</a>, so-called “lollipop plots” replace the bar
with a line segment and a circle. The length of the line segment is proportional
to the magnitude of the number, and the point marks the length of the segment as
a height on the y or length on the x axis, depending on orientation.</p>
<div class="sourceCode" id="cb229"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb229-1"><a href="data-visualization.html#cb229-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(ad_metadata) <span class="sc">+</span></span>
<span id="cb229-2"><a href="data-visualization.html#cb229-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">mapping=</span><span class="fu">aes</span>(<span class="at">x=</span>ID, <span class="at">y=</span>tau)) <span class="sc">+</span></span>
<span id="cb229-3"><a href="data-visualization.html#cb229-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_segment</span>(<span class="at">mapping=</span><span class="fu">aes</span>(<span class="at">x=</span>ID, <span class="at">xend=</span>ID, <span class="at">y=</span><span class="dv">0</span>, <span class="at">yend=</span>tau))</span></code></pre></div>
<p><img src="main_files/unnamed-chunk-74-1.png" width="672" /></p>
<p>Note that <code>aes()</code> mappings can be made on the <code>ggplot()</code> object or on each
individual geometry function call, to specify different mappings based on
geometry.</p>
</div>
<div id="stacked-area-charts" class="section level3" number="7.2.3">
<h3><span class="header-section-number">7.2.3</span> Stacked Area charts</h3>
<p>Stacked area charts can visualize multiple 1D data that share a common
categorical axis. The charts consist of one line per variable with vertices that
correspond to x and y values similar to a <a href="#bar-plots">bar</a> or
<a href="data-visualization.html#lollipop-plots">lollipop</a> plots. Each variable is plotted using the previous
one as a baseline, so that the height of the data points for each category is
proportional to their sum. The space between the lines for each variable and the
previous one are filled with a color. The following plot visualizes the amount
of marker stain for each of the four genes for each individaul:</p>
<div class="sourceCode" id="cb230"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb230-1"><a href="data-visualization.html#cb230-1" aria-hidden="true" tabindex="-1"></a><span class="fu">pivot_longer</span>(</span>
<span id="cb230-2"><a href="data-visualization.html#cb230-2" aria-hidden="true" tabindex="-1"></a>  ad_metadata,</span>
<span id="cb230-3"><a href="data-visualization.html#cb230-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">c</span>(tau,abeta,iba1,gfap),</span>
<span id="cb230-4"><a href="data-visualization.html#cb230-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">names_to=</span><span class="st">&#39;Marker&#39;</span>,</span>
<span id="cb230-5"><a href="data-visualization.html#cb230-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">values_to=</span><span class="st">&#39;Intensity&#39;</span></span>
<span id="cb230-6"><a href="data-visualization.html#cb230-6" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">%&gt;%</span></span>
<span id="cb230-7"><a href="data-visualization.html#cb230-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x=</span>ID,<span class="at">y=</span>Intensity,<span class="at">group=</span>Marker,<span class="at">fill=</span>Marker)) <span class="sc">+</span></span>
<span id="cb230-8"><a href="data-visualization.html#cb230-8" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_area</span>()</span></code></pre></div>
<p><img src="main_files/unnamed-chunk-75-1.png" width="672" /></p>
<p>We notice that subject A4 has the highest overall level of marker intensity,
followed by A1, A7, etc. The control samples overall have less intensity across
all markers. Certain samples, A2 and C5, have little to no <code>abeta</code> aggregation,
and C6 has little to no <code>tau</code>.</p>
<p>Stacked area plots require three pieces of data:</p>
<ul>
<li><strong>x</strong> - a numeric or categorical axis for vertical alignment</li>
<li><strong>y</strong> - a numeric axis to draw vertical proportions</li>
<li><strong>group</strong> - a categorical variable that indicates which (x,y) pairs correspond
to the same line</li>
</ul>
<p>In the example above, we needed to pivot our tibble so that the different
markers and their values were placed into columns <code>Marker</code> and <code>Intensity</code>,
respectively. Data for stacked bar charts will usually need to be in this ‘long’
format, as described in <a href="data-wrangling.html#rearranging-data">Rearranging Data</a>.</p>
<p>Sometimes it is more helpful to view the relative proportion of values in each
category rather than the actual values. The result is called a <em>proportional</em>
stacked area plots. While not a distinct plot type, we can create one by
preprocessing our data by dividing each value by the column sum:</p>
<div class="sourceCode" id="cb231"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb231-1"><a href="data-visualization.html#cb231-1" aria-hidden="true" tabindex="-1"></a><span class="fu">pivot_longer</span>(</span>
<span id="cb231-2"><a href="data-visualization.html#cb231-2" aria-hidden="true" tabindex="-1"></a>  ad_metadata,</span>
<span id="cb231-3"><a href="data-visualization.html#cb231-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">c</span>(tau,abeta,iba1,gfap),</span>
<span id="cb231-4"><a href="data-visualization.html#cb231-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">names_to=</span><span class="st">&#39;Marker&#39;</span>,</span>
<span id="cb231-5"><a href="data-visualization.html#cb231-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">values_to=</span><span class="st">&#39;Intensity&#39;</span></span>
<span id="cb231-6"><a href="data-visualization.html#cb231-6" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">%&gt;%</span></span>
<span id="cb231-7"><a href="data-visualization.html#cb231-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(ID) <span class="sc">%&gt;%</span> <span class="co"># we want to divide each subjects intensity values by the sum of all four markers</span></span>
<span id="cb231-8"><a href="data-visualization.html#cb231-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(</span>
<span id="cb231-9"><a href="data-visualization.html#cb231-9" aria-hidden="true" tabindex="-1"></a>    <span class="st">`</span><span class="at">Relative Intensity</span><span class="st">`</span><span class="ot">=</span>Intensity<span class="sc">/</span><span class="fu">sum</span>(Intensity)</span>
<span id="cb231-10"><a href="data-visualization.html#cb231-10" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">%&gt;%</span></span>
<span id="cb231-11"><a href="data-visualization.html#cb231-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ungroup</span>() <span class="sc">%&gt;%</span> <span class="co"># ungroup restores the tibble to its original number of rows after the transformation</span></span>
<span id="cb231-12"><a href="data-visualization.html#cb231-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x=</span>ID,<span class="at">y=</span><span class="st">`</span><span class="at">Relative Intensity</span><span class="st">`</span>,<span class="at">group=</span>Marker,<span class="at">fill=</span>Marker)) <span class="sc">+</span></span>
<span id="cb231-13"><a href="data-visualization.html#cb231-13" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_area</span>()</span></code></pre></div>
<p><img src="main_files/unnamed-chunk-76-1.png" width="672" />
Now the values for each subject have been normalized to each sum to 1. In this
way, we might note that the relative proportion of <code>abeta</code> seems to be greater
in AD samples than Controls, but that may not be true of <code>tau</code>. These
observations may inspire us to ask these questions more rigorously than we have
done so far by inspection.</p>
</div>
<div id="parallel-coordinate-plots" class="section level3" number="7.2.4">
<h3><span class="header-section-number">7.2.4</span> Parallel Coordinate plots</h3>

</div>
</div>
<div id="visualizing-distributions" class="section level2" number="7.3">
<h2><span class="header-section-number">7.3</span> Visualizing Distributions</h2>
<p>The distribution is one of the most important properties of a set of numbers.
A distribution describes the general “shape” of the numbers, i.e. what is the
relative frequency of the values, or ranges of values, within the data.
Understanding the distribution of a data set is critical when choosing methods
to apply, since many methods are only appropriate when data is distributed in
certain ways, e.g. <a href="https://en.wikipedia.org/wiki/Linear_regression">linear
regression</a> assumes the
response variable is <a href="https://en.wikipedia.org/wiki/Normal_distribution">normally
distributed</a>, otherwise the
result of the model cannot be interpreted properly. Often, we don’t know how our
data are distributed when we obtain it and so we must examine the distribution
empirically. The visualizations in this section are all used for the purpose of
depicting the distribution of a set of numbers.</p>
<div id="histogram" class="section level3" number="7.3.1">
<h3><span class="header-section-number">7.3.1</span> Histogram</h3>
<p>The most common way to plot the distribution of a 1-D set of data is the
<a href="https://en.wikipedia.org/wiki/Histogram">histogram</a>. The histogram divides up
the range of a dataset from minimum to maximum into bins usually of the same
width and tabulates the number of values that fall within each bin. Below is a
histogram of our <code>age_at_death</code> measurement for all samples:</p>
<div class="sourceCode" id="cb232"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb232-1"><a href="data-visualization.html#cb232-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(ad_metadata) <span class="sc">+</span></span>
<span id="cb232-2"><a href="data-visualization.html#cb232-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">mapping=</span><span class="fu">aes</span>(<span class="at">x=</span>age_at_death))</span></code></pre></div>
<p><img src="main_files/unnamed-chunk-77-1.png" width="672" /></p>
<p>Note that the histogram does not look very complete, because there are only 20
values in our data. We can mitigate this somewhat by increasing the number of
bins the data range is divided into:</p>
<div class="sourceCode" id="cb233"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb233-1"><a href="data-visualization.html#cb233-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(ad_metadata) <span class="sc">+</span></span>
<span id="cb233-2"><a href="data-visualization.html#cb233-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">mapping=</span><span class="fu">aes</span>(<span class="at">x=</span>age_at_death),<span class="at">bins=</span><span class="dv">10</span>)</span></code></pre></div>
<p><img src="main_files/unnamed-chunk-78-1.png" width="672" /></p>
<p>This is a little bit better, but there are still some bins (76-79, 84-87) that
have no values. Compare this to the following synthetic dataset of 1000
normally distributed values:</p>
<div class="sourceCode" id="cb234"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb234-1"><a href="data-visualization.html#cb234-1" aria-hidden="true" tabindex="-1"></a><span class="fu">tibble</span>(</span>
<span id="cb234-2"><a href="data-visualization.html#cb234-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">x=</span><span class="fu">rnorm</span>(<span class="dv">1000</span>)</span>
<span id="cb234-3"><a href="data-visualization.html#cb234-3" aria-hidden="true" tabindex="-1"></a>) <span class="sc">%&gt;%</span></span>
<span id="cb234-4"><a href="data-visualization.html#cb234-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb234-5"><a href="data-visualization.html#cb234-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="fu">aes</span>(<span class="at">x=</span>x))</span></code></pre></div>
<p><img src="main_files/unnamed-chunk-79-1.png" width="672" /></p>
<p>For distributions with a small number of samples, histograms might not be the
best visualization. We will continue with synthetic normally distributed dataset
for the remaining examples.</p>
<p>ggplot allows you to easily plot multiple distributions on the same plot:</p>
<div class="sourceCode" id="cb235"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb235-1"><a href="data-visualization.html#cb235-1" aria-hidden="true" tabindex="-1"></a><span class="fu">tibble</span>(</span>
<span id="cb235-2"><a href="data-visualization.html#cb235-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">x=</span><span class="fu">c</span>(<span class="fu">rnorm</span>(<span class="dv">1000</span>),<span class="fu">rnorm</span>(<span class="dv">1000</span>,<span class="at">mean=</span><span class="dv">4</span>)),</span>
<span id="cb235-3"><a href="data-visualization.html#cb235-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">type=</span><span class="fu">c</span>(<span class="fu">rep</span>(<span class="st">&#39;A&#39;</span>,<span class="dv">1000</span>),<span class="fu">rep</span>(<span class="st">&#39;B&#39;</span>,<span class="dv">1000</span>))</span>
<span id="cb235-4"><a href="data-visualization.html#cb235-4" aria-hidden="true" tabindex="-1"></a>) <span class="sc">%&gt;%</span></span>
<span id="cb235-5"><a href="data-visualization.html#cb235-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x=</span>x,<span class="at">fill=</span>type)) <span class="sc">+</span></span>
<span id="cb235-6"><a href="data-visualization.html#cb235-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">bins=</span><span class="dv">30</span>, <span class="at">alpha=</span><span class="fl">0.6</span>, <span class="at">position=</span><span class="st">&quot;identity&quot;</span>)</span></code></pre></div>
<p><img src="main_files/unnamed-chunk-80-1.png" width="672" /></p>
<p>The <code>alpha=0.6, position="identity"</code> arguments makes the bars partially
transparent so you can see the overlap more clearly.</p>
<div class="box readmore">
<ul>
<li><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html?q=histogram">ggplot2 geom_histogram reference</a></li>
<li><a href="https://www.r-graph-gallery.com/histogram.html">R Graph Gallery - histograms</a></li>
</ul>
</div>
</div>
<div id="density" class="section level3" number="7.3.2">
<h3><span class="header-section-number">7.3.2</span> Density</h3>
<p>Another way to describe a distribution is with a <a href="https://www.r-graph-gallery.com/density-plot.html">density
plot</a>. Instead of binning the
values into intervals and drawing bars with height proportional to the number of
values in each bin, a density plot draws a smoothly interpolated line that
approximates the distribution instead. A key difference between a histogram and
a density plot is the density plot is always normalized so the integral under
the curve is approximately 1, whereas a histogram may be either counts or, if
the counts in each bin are divided by the total number of data points, a
proportion.</p>
<p>Compare the histogram and density plots
of the <code>age_at_death</code> variable from our example tibble:</p>
<div class="sourceCode" id="cb236"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb236-1"><a href="data-visualization.html#cb236-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(patchwork)</span>
<span id="cb236-2"><a href="data-visualization.html#cb236-2" aria-hidden="true" tabindex="-1"></a>hist_g <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(ad_metadata) <span class="sc">+</span></span>
<span id="cb236-3"><a href="data-visualization.html#cb236-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">mapping=</span><span class="fu">aes</span>(<span class="at">x=</span>age_at_death),<span class="at">bins=</span><span class="dv">30</span>)</span>
<span id="cb236-4"><a href="data-visualization.html#cb236-4" aria-hidden="true" tabindex="-1"></a>density_g <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(ad_metadata) <span class="sc">+</span></span>
<span id="cb236-5"><a href="data-visualization.html#cb236-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_density</span>(<span class="at">mapping=</span><span class="fu">aes</span>(<span class="at">x=</span>age_at_death),<span class="at">fill=</span><span class="st">&quot;#c9a13daa&quot;</span>)</span>
<span id="cb236-6"><a href="data-visualization.html#cb236-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb236-7"><a href="data-visualization.html#cb236-7" aria-hidden="true" tabindex="-1"></a>hist_g <span class="sc">|</span> density_g</span></code></pre></div>
<p><img src="main_files/unnamed-chunk-81-1.png" width="672" /></p>
<p>Notice the overall shape of the two distributions is similar, with the highest
values in both around age 77. The density plot is a smoother representation of
a histogram, but its accuracy is still highly sensitive to the number of
measurements used to construct it. Compare the histogram and density plots of
two sets of 1000 normally distributed samples with different means:</p>
<div class="sourceCode" id="cb237"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb237-1"><a href="data-visualization.html#cb237-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(patchwork)</span>
<span id="cb237-2"><a href="data-visualization.html#cb237-2" aria-hidden="true" tabindex="-1"></a>normal_samples <span class="ot">&lt;-</span> <span class="fu">tibble</span>(</span>
<span id="cb237-3"><a href="data-visualization.html#cb237-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">x=</span><span class="fu">c</span>(<span class="fu">rnorm</span>(<span class="dv">1000</span>),<span class="fu">rnorm</span>(<span class="dv">1000</span>,<span class="at">mean=</span><span class="dv">4</span>)),</span>
<span id="cb237-4"><a href="data-visualization.html#cb237-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">type=</span><span class="fu">c</span>(<span class="fu">rep</span>(<span class="st">&#39;A&#39;</span>,<span class="dv">1000</span>),<span class="fu">rep</span>(<span class="st">&#39;B&#39;</span>,<span class="dv">1000</span>))</span>
<span id="cb237-5"><a href="data-visualization.html#cb237-5" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb237-6"><a href="data-visualization.html#cb237-6" aria-hidden="true" tabindex="-1"></a>hist_g <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(normal_samples) <span class="sc">+</span></span>
<span id="cb237-7"><a href="data-visualization.html#cb237-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(</span>
<span id="cb237-8"><a href="data-visualization.html#cb237-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">mapping=</span><span class="fu">aes</span>(<span class="at">x=</span>x,<span class="at">fill=</span>type),</span>
<span id="cb237-9"><a href="data-visualization.html#cb237-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">alpha=</span><span class="fl">0.6</span>,</span>
<span id="cb237-10"><a href="data-visualization.html#cb237-10" aria-hidden="true" tabindex="-1"></a>    <span class="at">position=</span><span class="st">&quot;identity&quot;</span>,</span>
<span id="cb237-11"><a href="data-visualization.html#cb237-11" aria-hidden="true" tabindex="-1"></a>    <span class="at">bins=</span><span class="dv">30</span></span>
<span id="cb237-12"><a href="data-visualization.html#cb237-12" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb237-13"><a href="data-visualization.html#cb237-13" aria-hidden="true" tabindex="-1"></a>density_g <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(normal_samples) <span class="sc">+</span></span>
<span id="cb237-14"><a href="data-visualization.html#cb237-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_density</span>(</span>
<span id="cb237-15"><a href="data-visualization.html#cb237-15" aria-hidden="true" tabindex="-1"></a>    <span class="at">mapping=</span><span class="fu">aes</span>(<span class="at">x=</span>x,<span class="at">fill=</span>type),</span>
<span id="cb237-16"><a href="data-visualization.html#cb237-16" aria-hidden="true" tabindex="-1"></a>    <span class="at">alpha=</span><span class="fl">0.6</span>,</span>
<span id="cb237-17"><a href="data-visualization.html#cb237-17" aria-hidden="true" tabindex="-1"></a>    <span class="at">position=</span><span class="st">&quot;identity&quot;</span></span>
<span id="cb237-18"><a href="data-visualization.html#cb237-18" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb237-19"><a href="data-visualization.html#cb237-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb237-20"><a href="data-visualization.html#cb237-20" aria-hidden="true" tabindex="-1"></a>hist_g <span class="sc">|</span> density_g</span></code></pre></div>
<p><img src="main_files/unnamed-chunk-82-1.png" width="672" /></p>
<p>Again the two types of plots depict similar distributions, although they are
different enough to possibly suggest different interpretations. In general,
density plots might be preferable over histograms if the data are noisy or
sparse in that they produce cleaner plots, but potentially at the expense of
accuracy when the number of samples is low.</p>
<div class="box readmore">
<ul>
<li><a href="https://ggplot2.tidyverse.org/reference/geom_density.html?q=density">ggplot2 geom_density reference</a></li>
<li><a href="https://www.r-graph-gallery.com/density-plot.html">R Graph Gallery - density plots</a></li>
</ul>
</div>
</div>
<div id="boxplot" class="section level3" number="7.3.3">
<h3><span class="header-section-number">7.3.3</span> Boxplot</h3>
<p>Box plots, or box and whisker plots, are extremely common when used to describe
distributions. Below is a boxplot of age at death divided by condition:</p>
<div class="sourceCode" id="cb238"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb238-1"><a href="data-visualization.html#cb238-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(ad_metadata) <span class="sc">+</span></span>
<span id="cb238-2"><a href="data-visualization.html#cb238-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>(<span class="at">mapping=</span><span class="fu">aes</span>(<span class="at">x=</span>condition,<span class="at">y=</span>age_at_death))</span></code></pre></div>
<p><img src="main_files/unnamed-chunk-83-1.png" width="672" /></p>
<p>Boxplots are drawn assuming the data are
<a href="https://en.wikipedia.org/wiki/Unimodality">unimodal</a> (i.e. shaped like a hill,
possibly slanted to one side or the other), where the extents of the box
represent the 1st and 3rd quartile of the data, the central line is the median,
the whiskers are drawn as 1.5 times the value outside the 1st and 3rd quartiles.
Sometimes individual values more extreme than the whiskers are drawn
individually to identify them as outliers.</p>
<div class="figure">
<img src="https://upload.wikimedia.org/wikipedia/commons/1/1a/Boxplot_vs_PDF.svg" alt="" />
<p class="caption">Boxplot anatomy. IQR stands for “inner quartile range,” the distance between
the 1st and 3rd quartile - Wikipedia</p>
</div>
<p>However, boxplots have some significant shortcomings. Primarily, the rectangle
of the inner quartile range does not describe the actual distribution of the
samples within it. Although the median can give a sense of skewness, if the
data are not unimodal this may be misleading. Consider the following
distributions plotted as boxplots or as violin plots (described in the next
section):</p>
<div class="sourceCode" id="cb239"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb239-1"><a href="data-visualization.html#cb239-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(patchwork)</span>
<span id="cb239-2"><a href="data-visualization.html#cb239-2" aria-hidden="true" tabindex="-1"></a>normal_samples <span class="ot">&lt;-</span> <span class="fu">tibble</span>(</span>
<span id="cb239-3"><a href="data-visualization.html#cb239-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">x=</span><span class="fu">c</span>(<span class="fu">rnorm</span>(<span class="dv">1000</span>),<span class="fu">rnorm</span>(<span class="dv">1000</span>,<span class="dv">4</span>),<span class="fu">rnorm</span>(<span class="dv">1000</span>,<span class="dv">2</span>,<span class="dv">3</span>)),</span>
<span id="cb239-4"><a href="data-visualization.html#cb239-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">type=</span><span class="fu">c</span>(<span class="fu">rep</span>(<span class="st">&#39;A&#39;</span>,<span class="dv">2000</span>),<span class="fu">rep</span>(<span class="st">&#39;B&#39;</span>,<span class="dv">1000</span>))</span>
<span id="cb239-5"><a href="data-visualization.html#cb239-5" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb239-6"><a href="data-visualization.html#cb239-6" aria-hidden="true" tabindex="-1"></a>g <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(normal_samples, <span class="fu">aes</span>(<span class="at">x=</span>type,<span class="at">y=</span>x,<span class="at">fill=</span>type))</span>
<span id="cb239-7"><a href="data-visualization.html#cb239-7" aria-hidden="true" tabindex="-1"></a>boxplot_g <span class="ot">&lt;-</span> g <span class="sc">+</span>  <span class="fu">geom_boxplot</span>()</span>
<span id="cb239-8"><a href="data-visualization.html#cb239-8" aria-hidden="true" tabindex="-1"></a>violin_g <span class="ot">&lt;-</span> g <span class="sc">+</span> <span class="fu">geom_violin</span>()</span>
<span id="cb239-9"><a href="data-visualization.html#cb239-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb239-10"><a href="data-visualization.html#cb239-10" aria-hidden="true" tabindex="-1"></a>boxplot_g <span class="sc">|</span> violin_g</span></code></pre></div>
<p><img src="main_files/unnamed-chunk-84-1.png" width="672" /></p>
<p>The two distributions look almost identical in the boxplot figure; however they
are dramatically different when visualized using a method like a violin plot
where the contours of the entire distribution are depicted. Unless you are
certain that your data are unimodal, one of the other distribution visualization
methods in this section will likely more accurately depict your data than a
boxplot.</p>
<div class="box readmore">
<ul>
<li><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html">ggplot2 geom_boxplot reference</a></li>
<li><a href="https://www.r-graph-gallery.com/boxplot.html">R Graph Gallery - boxplots</a></li>
</ul>
</div>
</div>
<div id="violin-plot" class="section level3" number="7.3.4">
<h3><span class="header-section-number">7.3.4</span> Violin plot</h3>
<p>As seen in the last section, a <a href="https://ggplot2.tidyverse.org/reference/geom_violin.html">violin
plot</a> is another way
to depict a distribution by producing a shape where the width is proportional to
the value along the x or y axis, depending on orientation. The “violin” shape is
similar in principle to a histogram or a density plot, in that it describes the
contour of all the data in the distribution, not just the quantiles and extents,
as in a box plot. Below is a violin plot of the <code>tau</code> measures from our example
tibble:</p>
<div class="sourceCode" id="cb240"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb240-1"><a href="data-visualization.html#cb240-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(ad_metadata) <span class="sc">+</span></span>
<span id="cb240-2"><a href="data-visualization.html#cb240-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_violin</span>(<span class="fu">aes</span>(<span class="at">x=</span>condition,<span class="at">y=</span>tau,<span class="at">fill=</span>condition))</span></code></pre></div>
<p><img src="main_files/unnamed-chunk-85-1.png" width="672" /></p>
<p>The violin plot is both more and less descriptive than a boxplot; it does depict
the entire distribution of the data, but also doesn’t include features like
median by default.</p>
<div class="box readmore">
<ul>
<li><a href="https://ggplot2.tidyverse.org/reference/geom_violin.html">ggplot2 geom_violin reference</a></li>
<li><a href="https://www.r-graph-gallery.com/violin.html">R Graph Gallery - violin plots</a></li>
</ul>
</div>
</div>
<div id="beeswarm-plot" class="section level3" number="7.3.5">
<h3><span class="header-section-number">7.3.5</span> Beeswarm plot</h3>
<p>The <a href="https://r-charts.com/distribution/ggbeeswarm/">beeswarm plot</a> is similar to
a violin plot, but instead of plotting the contours of the data, it plots the
data itself as points like in a scatter plot. The individual values of the
distribution are organized vertically and spaced such that the points don’t
overlap. In this plot, the distribution of age at death is plotted for each kind
of sample and the markers are colored by the amount of tau:</p>
<div class="sourceCode" id="cb241"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb241-1"><a href="data-visualization.html#cb241-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggbeeswarm)</span>
<span id="cb241-2"><a href="data-visualization.html#cb241-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(ad_metadata) <span class="sc">+</span></span>
<span id="cb241-3"><a href="data-visualization.html#cb241-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_beeswarm</span>(<span class="fu">aes</span>(<span class="at">x=</span>condition,<span class="at">y=</span>age_at_death,<span class="at">color=</span>condition),<span class="at">cex=</span><span class="dv">2</span>,<span class="at">size=</span><span class="dv">2</span>)</span></code></pre></div>
<p><img src="main_files/unnamed-chunk-86-1.png" width="672" /></p>
<p>We may not have noticed before that our AD samples have a big gap in ages
between 74 and 81; since the beeswarm plot displays all the data, we can see it
easily here.</p>
<p>Beeswarm plots are typically only useful when the number of values is within a
range; not too many and not too few. The example above is close to having too
few values per group for this plot to be useful, but consider the following with
too many samples:</p>
<div class="sourceCode" id="cb242"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb242-1"><a href="data-visualization.html#cb242-1" aria-hidden="true" tabindex="-1"></a>normal_samples <span class="ot">&lt;-</span> <span class="fu">tibble</span>(</span>
<span id="cb242-2"><a href="data-visualization.html#cb242-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">x=</span><span class="fu">c</span>(<span class="fu">rnorm</span>(<span class="dv">1000</span>),<span class="fu">rnorm</span>(<span class="dv">1000</span>,<span class="dv">4</span>),<span class="fu">rnorm</span>(<span class="dv">1000</span>,<span class="dv">2</span>,<span class="dv">3</span>)),</span>
<span id="cb242-3"><a href="data-visualization.html#cb242-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">type=</span><span class="fu">c</span>(<span class="fu">rep</span>(<span class="st">&#39;A&#39;</span>,<span class="dv">2000</span>),<span class="fu">rep</span>(<span class="st">&#39;B&#39;</span>,<span class="dv">1000</span>))</span>
<span id="cb242-4"><a href="data-visualization.html#cb242-4" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb242-5"><a href="data-visualization.html#cb242-5" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(normal_samples, <span class="fu">aes</span>(<span class="at">x=</span>type,<span class="at">y=</span>x,<span class="at">color=</span>type)) <span class="sc">+</span></span>
<span id="cb242-6"><a href="data-visualization.html#cb242-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_beeswarm</span>()</span></code></pre></div>
<p><img src="main_files/unnamed-chunk-87-1.png" width="672" /></p>
<p>This plot likely has too many samples to be the right choice (it’s also ugly),
but it does give an idea of the distribution of the data.</p>
<p>In the previous examples the markers for each group also determined the color of
the group. This makes the chart a bit easier to read and more pleasing to the
eye, but is technically redundant. You can use however profitably however to
color markers by some other value that might be of interest. Consider this final
example where markers are colored by another randomly generated variable:</p>
<div class="sourceCode" id="cb243"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb243-1"><a href="data-visualization.html#cb243-1" aria-hidden="true" tabindex="-1"></a>normal_samples <span class="ot">&lt;-</span> <span class="fu">tibble</span>(</span>
<span id="cb243-2"><a href="data-visualization.html#cb243-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">x=</span><span class="fu">c</span>(<span class="fu">rnorm</span>(<span class="dv">100</span>),<span class="fu">rnorm</span>(<span class="dv">100</span>,<span class="dv">4</span>),<span class="fu">rnorm</span>(<span class="dv">100</span>,<span class="dv">2</span>,<span class="dv">3</span>)),</span>
<span id="cb243-3"><a href="data-visualization.html#cb243-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">type=</span><span class="fu">c</span>(<span class="fu">rep</span>(<span class="st">&#39;A&#39;</span>,<span class="dv">200</span>),<span class="fu">rep</span>(<span class="st">&#39;B&#39;</span>,<span class="dv">100</span>)),</span>
<span id="cb243-4"><a href="data-visualization.html#cb243-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">category=</span><span class="fu">sample</span>(<span class="fu">c</span>(<span class="st">&#39;healthy&#39;</span>,<span class="st">&#39;disease&#39;</span>),<span class="dv">300</span>,<span class="at">replace=</span><span class="cn">TRUE</span>)</span>
<span id="cb243-5"><a href="data-visualization.html#cb243-5" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb243-6"><a href="data-visualization.html#cb243-6" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(normal_samples, <span class="fu">aes</span>(<span class="at">x=</span>type,<span class="at">y=</span>x,<span class="at">color=</span>category)) <span class="sc">+</span></span>
<span id="cb243-7"><a href="data-visualization.html#cb243-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_beeswarm</span>()</span></code></pre></div>
<p><img src="main_files/unnamed-chunk-88-1.png" width="672" /></p>
<p>We are now effectively visualizing three dimensions which may provide insight
into the data.</p>
<div class="box readmore">
<ul>
<li><a href="https://r-charts.com/distribution/ggbeeswarm/">beeswarm package reference</a></li>
</ul>
</div>
</div>
<div id="ridgeline" class="section level3" number="7.3.6">
<h3><span class="header-section-number">7.3.6</span> Ridgeline</h3>
<p>If you have many non-trivial distributions that you would like the user to
compare, a good option is a <a href="https://www.r-graph-gallery.com/294-basic-ridgeline-plot.html">ridgeline
chart</a>. The
ridgeline plot is simply multiple density plots drawn for different variables
within the same plot. Like the beeswarm plot, ridgeline plots are provided by
another package outside ggplot2.</p>
<div class="sourceCode" id="cb244"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb244-1"><a href="data-visualization.html#cb244-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggridges)</span>
<span id="cb244-2"><a href="data-visualization.html#cb244-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb244-3"><a href="data-visualization.html#cb244-3" aria-hidden="true" tabindex="-1"></a><span class="fu">tibble</span>(</span>
<span id="cb244-4"><a href="data-visualization.html#cb244-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">x=</span><span class="fu">c</span>(<span class="fu">rnorm</span>(<span class="dv">100</span>),<span class="fu">rnorm</span>(<span class="dv">100</span>,<span class="dv">4</span>),<span class="fu">rnorm</span>(<span class="dv">100</span>,<span class="dv">2</span>,<span class="dv">3</span>)),</span>
<span id="cb244-5"><a href="data-visualization.html#cb244-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">type=</span><span class="fu">c</span>(<span class="fu">rep</span>(<span class="st">&#39;A&#39;</span>,<span class="dv">200</span>),<span class="fu">rep</span>(<span class="st">&#39;B&#39;</span>,<span class="dv">100</span>)),</span>
<span id="cb244-6"><a href="data-visualization.html#cb244-6" aria-hidden="true" tabindex="-1"></a>) <span class="sc">%&gt;%</span></span>
<span id="cb244-7"><a href="data-visualization.html#cb244-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">y=</span>type,<span class="at">x=</span>x,<span class="at">fill=</span>type)) <span class="sc">+</span></span>
<span id="cb244-8"><a href="data-visualization.html#cb244-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_density_ridges</span>()</span></code></pre></div>
<p><img src="main_files/unnamed-chunk-89-1.png" width="672" /></p>
<p>Many distributions may be plotted:</p>
<div class="sourceCode" id="cb245"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb245-1"><a href="data-visualization.html#cb245-1" aria-hidden="true" tabindex="-1"></a><span class="fu">tibble</span>(</span>
<span id="cb245-2"><a href="data-visualization.html#cb245-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">x=</span><span class="fu">rnorm</span>(<span class="dv">10000</span>,<span class="at">mean=</span><span class="fu">runif</span>(<span class="dv">10</span>,<span class="dv">1</span>,<span class="dv">10</span>),<span class="at">sd=</span><span class="fu">runif</span>(<span class="dv">2</span>,<span class="dv">1</span>,<span class="dv">4</span>)),</span>
<span id="cb245-3"><a href="data-visualization.html#cb245-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">type=</span><span class="fu">rep</span>(<span class="fu">c</span>(<span class="st">&quot;A&quot;</span>,<span class="st">&quot;B&quot;</span>,<span class="st">&quot;C&quot;</span>,<span class="st">&quot;D&quot;</span>,<span class="st">&quot;E&quot;</span>,<span class="st">&quot;F&quot;</span>,<span class="st">&quot;G&quot;</span>,<span class="st">&quot;H&quot;</span>,<span class="st">&quot;I&quot;</span>,<span class="st">&quot;J&quot;</span>),<span class="dv">1000</span>)</span>
<span id="cb245-4"><a href="data-visualization.html#cb245-4" aria-hidden="true" tabindex="-1"></a>) <span class="sc">%&gt;%</span></span>
<span id="cb245-5"><a href="data-visualization.html#cb245-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">y=</span>type,<span class="at">x=</span>x,<span class="at">fill=</span>type)) <span class="sc">+</span></span>
<span id="cb245-6"><a href="data-visualization.html#cb245-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_density_ridges</span>(<span class="at">alpha=</span><span class="fl">0.6</span>,<span class="at">position=</span><span class="st">&quot;identity&quot;</span>)</span></code></pre></div>
<p><img src="main_files/unnamed-chunk-90-1.png" width="672" /></p>
<div class="box readmore">
<ul>
<li><a href="https://www.r-graph-gallery.com/294-basic-ridgeline-plot.html">R Graph Gallery - ridgeline plots</a></li>
<li><a href="https://cran.r-project.org/web/packages/ggridges/vignettes/introduction.html">ggridges package vignette</a></li>
<li><a href="https://cran.r-project.org/web/packages/ggridges/index.html">ggridges package index on CRAN</a></li>
</ul>
</div>

</div>
</div>
<div id="plotting-two-or-more-dimensions-." class="section level2" number="7.4">
<h2><span class="header-section-number">7.4</span> Plotting Two or More Dimensions .</h2>
<div id="scatter-plot-." class="section level3" number="7.4.1">
<h3><span class="header-section-number">7.4.1</span> Scatter plot .</h3>
</div>
<div id="line-plot-." class="section level3" number="7.4.2">
<h3><span class="header-section-number">7.4.2</span> Line plot .</h3>
<div class="sourceCode" id="cb246"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb246-1"><a href="data-visualization.html#cb246-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(ad_metadata,<span class="at">mapping=</span><span class="fu">aes</span>(<span class="at">x=</span>abeta, <span class="at">y=</span>tau)) <span class="sc">+</span></span>
<span id="cb246-2"><a href="data-visualization.html#cb246-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_line</span>()</span></code></pre></div>
<p><img src="main_files/unnamed-chunk-92-1.png" width="672" /></p>
<div class="sourceCode" id="cb247"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb247-1"><a href="data-visualization.html#cb247-1" aria-hidden="true" tabindex="-1"></a><span class="fu">pivot_longer</span>(ad_metadata,</span>
<span id="cb247-2"><a href="data-visualization.html#cb247-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">c</span>(tau,abeta,iba1,gfap),</span>
<span id="cb247-3"><a href="data-visualization.html#cb247-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">names_to=</span><span class="st">&#39;Marker&#39;</span>,</span>
<span id="cb247-4"><a href="data-visualization.html#cb247-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">values_to=</span><span class="st">&#39;Intensity&#39;</span></span>
<span id="cb247-5"><a href="data-visualization.html#cb247-5" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">%&gt;%</span></span>
<span id="cb247-6"><a href="data-visualization.html#cb247-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(ad_metadata,<span class="at">mapping=</span><span class="fu">aes</span>(<span class="at">x=</span>ID, <span class="at">y=</span>Intensity, <span class="at">group=</span>Marker, <span class="at">color=</span>Marker)) <span class="sc">+</span></span>
<span id="cb247-7"><a href="data-visualization.html#cb247-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_line</span>()</span></code></pre></div>
<p><img src="main_files/unnamed-chunk-93-1.png" width="672" /></p>
</div>
<div id="bubble-plot-." class="section level3" number="7.4.3">
<h3><span class="header-section-number">7.4.3</span> Bubble plot .</h3>
</div>
<div id="heatmaps" class="section level3" number="7.4.4">
<h3><span class="header-section-number">7.4.4</span> Heatmaps</h3>
<p>Heatmaps visualize values associated with a grid of points <span class="math inline">\((x,y,z)\)</span> as a grid
of colored rectangles, where <span class="math inline">\(x\)</span> and <span class="math inline">\(y\)</span> define the grid point coordinates and
<span class="math inline">\(z\)</span> is a continuous value. A common heatmap you might have seen is the <a href="https://en.wikipedia.org/wiki/Weather_map">weather
map</a>, which plots current or
predicted weather patterns on top of a geographic map:</p>
<div class="figure">
<img src="weather_map.png" alt="" />
<p class="caption">Weather maps are heatmaps</p>
</div>
<p>The color at each grid coordinate, which in this case are spaced closely enough
so we cannot distinguish the boundaries between the rectangles without zooming
in, is proportional to the type and intensity of weather at that location.
In the weather map, the weather values are mapped to a <em>color scales</em> that
describe the amount of precipitation as shown in the map legend.</p>
<p>In biology, heatmaps are more typically used to visualize matrices. The concept
is the same as with the weather map, except instead of geographic coordinates
the grid corresponds to the rows and columns of a data matrix, and the color of
each grid point is mapped to a color scale chosen by the designer. The grid of
rectangles is typically not overlayed over anything, as in the weather map, but
is instead the entire visualization.</p>
<p>We can create heatmap in R using the base R
<a href="https://www.rdocumentation.org/packages/stats/versions/3.6.2/topics/heatmap"><code>heatmap()</code></a>
function. Below is a heatmap visualization of our histological markers from the
AD example:</p>
<div class="sourceCode" id="cb248"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb248-1"><a href="data-visualization.html#cb248-1" aria-hidden="true" tabindex="-1"></a><span class="co"># heatmap() requires a R matrix, and cannot accept a tibble or a dataframe</span></span>
<span id="cb248-2"><a href="data-visualization.html#cb248-2" aria-hidden="true" tabindex="-1"></a>marker_matrix <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(</span>
<span id="cb248-3"><a href="data-visualization.html#cb248-3" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">select</span>(ad_metadata,<span class="fu">c</span>(tau,abeta,iba1,gfap))</span>
<span id="cb248-4"><a href="data-visualization.html#cb248-4" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb248-5"><a href="data-visualization.html#cb248-5" aria-hidden="true" tabindex="-1"></a><span class="co"># rownames of the matrix become y labels</span></span>
<span id="cb248-6"><a href="data-visualization.html#cb248-6" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(marker_matrix) <span class="ot">&lt;-</span> ad_metadata<span class="sc">$</span>ID</span>
<span id="cb248-7"><a href="data-visualization.html#cb248-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb248-8"><a href="data-visualization.html#cb248-8" aria-hidden="true" tabindex="-1"></a><span class="fu">heatmap</span>(marker_matrix)</span></code></pre></div>
<p><img src="main_files/data%20viz%20marker%20heatmap-1.png" width="672" /></p>
<p>The lower right portion of the plot is the heatmap, where the matrix values are
visualized on a color gradient from light yellow to dark red.</p>
<p>More precisely, the <code>heatmap()</code> function creates a <em>clustered heatmap</em>, where
the rows and columns have been <a href="data-science.html#hierarchical-clustering">hierarchically
clustered</a> separately and ordered according to how
similar they are to one another. The left and top line diagrams are
<a href="data-visualization.html#dendrograms">dendrograms</a>, which depict the similarity between rows and
columns as a tree, where the total branch length summed from one row/column to
another is proportional to the dissimilarity (i.e. numeric distance) of the two.</p>
<div class="box note">
<p>The base R <code>heatmap()</code> function performs many different operations on the input
matrix than just draw a grid of rectangles in proportion to the values. By
default, it also:</p>
<ul>
<li>performs <a href="data-science.html#hierarchical-clustering">hierarchical clustering</a> of the rows and
columns using a Euclidean distance function and orders them accordingly</li>
<li>draws <a href="#dendrogram">dendrograms</a> on the rows and columns according to the
clustering</li>
<li>scales the data in the rows to have mean zero and standard deviation 1</li>
</ul>
<p>Each of these defaults can be changed by passing arguments into the function
call. The following turns off all of the extra functionality and produces only a
heatmap of the matrix:</p>
<div class="sourceCode" id="cb249"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb249-1"><a href="data-visualization.html#cb249-1" aria-hidden="true" tabindex="-1"></a><span class="fu">heatmap</span>(</span>
<span id="cb249-2"><a href="data-visualization.html#cb249-2" aria-hidden="true" tabindex="-1"></a>  marker_matrix,</span>
<span id="cb249-3"><a href="data-visualization.html#cb249-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">Rowv=</span><span class="cn">NA</span>,</span>
<span id="cb249-4"><a href="data-visualization.html#cb249-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">Colv=</span><span class="cn">NA</span>,</span>
<span id="cb249-5"><a href="data-visualization.html#cb249-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">scale=</span><span class="st">&quot;none&quot;</span>,</span>
<span id="cb249-6"><a href="data-visualization.html#cb249-6" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div>
<p><img src="main_files/data%20viz%20pure%20heatmap-1.png" width="672" />
Note how the colors in this heatmap are very different for many of the
marker/sample pairs. This illustrates some of the dangers of using heatmaps,
which are described more fully in the [How To Use Heatmaps Responsibly] section
below.</p>
</div>
<p>The base R heatmap function has the major drawback that no color key is provided
to indicate how values map to colors. Another heatmap
<a href="https://www.rdocumentation.org/packages/gplots/versions/3.1.1/topics/heatmap.2"><code>heatmap.2</code></a>
function in the <a href="https://cran.r-project.org/web/packages/gplots/index.html"><code>gplots</code> package</a>
which has a similar interface to <code>heatmap()</code>, but allows provides more
parameters to control the behavior of the plot and includes a color key:</p>
<div class="sourceCode" id="cb250"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb250-1"><a href="data-visualization.html#cb250-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(gplots)</span>
<span id="cb250-2"><a href="data-visualization.html#cb250-2" aria-hidden="true" tabindex="-1"></a><span class="fu">heatmap.2</span>(marker_matrix)</span></code></pre></div>
<p><img src="main_files/heatmap.2-1.png" width="672" /></p>
<p>The extra decoration (dashed and solid vertical lines called <em>traces</em> by the
package) provide another way to understand the magnitude of the value in each
cell. Refer to the <a href="https://www.rdocumentation.org/packages/gplots/versions/3.1.1/topics/heatmap.2"><code>gplots</code> documentation of
<code>heatmap2</code></a>
for more information on how to interpret the trace (hint: you may turn it off
with by passing the argument <code>trace = "none"</code> to the function call).</p>
<p>Both <code>heatmap()</code> and <code>heatmap.2()</code> provide a useful method to annotate rows and
columns with a categorical variable. Each of our subjects either has AD or is a
control, and we can add a color bar along the margin of the plot to indicate the
status of each subject:</p>
<div class="sourceCode" id="cb251"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb251-1"><a href="data-visualization.html#cb251-1" aria-hidden="true" tabindex="-1"></a>condition_colors <span class="ot">&lt;-</span></span>
<span id="cb251-2"><a href="data-visualization.html#cb251-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">transmute</span>(</span>
<span id="cb251-3"><a href="data-visualization.html#cb251-3" aria-hidden="true" tabindex="-1"></a>    ad_metadata,</span>
<span id="cb251-4"><a href="data-visualization.html#cb251-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">color=</span><span class="fu">if_else</span>(condition <span class="sc">==</span> <span class="st">&quot;AD&quot;</span>,<span class="st">&quot;red&quot;</span>,<span class="st">&quot;blue&quot;</span>)</span>
<span id="cb251-5"><a href="data-visualization.html#cb251-5" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb251-6"><a href="data-visualization.html#cb251-6" aria-hidden="true" tabindex="-1"></a><span class="fu">heatmap</span>(</span>
<span id="cb251-7"><a href="data-visualization.html#cb251-7" aria-hidden="true" tabindex="-1"></a>  marker_matrix,</span>
<span id="cb251-8"><a href="data-visualization.html#cb251-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">RowSideColors=</span>condition_colors<span class="sc">$</span>color</span>
<span id="cb251-9"><a href="data-visualization.html#cb251-9" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div>
<p><img src="main_files/heatmap%20margin-1.png" width="672" /></p>
<p>And with <code>heatmap.2</code>:</p>
<div class="sourceCode" id="cb252"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb252-1"><a href="data-visualization.html#cb252-1" aria-hidden="true" tabindex="-1"></a><span class="fu">heatmap.2</span>(</span>
<span id="cb252-2"><a href="data-visualization.html#cb252-2" aria-hidden="true" tabindex="-1"></a>  marker_matrix,</span>
<span id="cb252-3"><a href="data-visualization.html#cb252-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">RowSideColors=</span>condition_colors<span class="sc">$</span>color</span>
<span id="cb252-4"><a href="data-visualization.html#cb252-4" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div>
<p><img src="main_files/heatmap.2%20margin-1.png" width="672" /></p>
<p>Heatmaps may also be created using ggplot and the
<a href="https://ggplot2.tidyverse.org/reference/geom_tile.html"><code>geom_tile</code></a> geometry.
This method of generating heatmaps is more “manual” than the other functions
covered so far, and so are both flexible and require more work to obtain certain
functionality (e.g. clustering, color bars on side margins, etc). This geometry
requires the data to be in long format, with one column for <code>x</code>, <code>y</code>, and <code>z</code>
values:</p>
<div class="sourceCode" id="cb253"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb253-1"><a href="data-visualization.html#cb253-1" aria-hidden="true" tabindex="-1"></a><span class="fu">pivot_longer</span>(</span>
<span id="cb253-2"><a href="data-visualization.html#cb253-2" aria-hidden="true" tabindex="-1"></a>  ad_metadata,</span>
<span id="cb253-3"><a href="data-visualization.html#cb253-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">c</span>(tau,abeta,iba1,gfap),</span>
<span id="cb253-4"><a href="data-visualization.html#cb253-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">names_to=</span><span class="st">&quot;Marker&quot;</span>,</span>
<span id="cb253-5"><a href="data-visualization.html#cb253-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">values_to=</span><span class="st">&quot;Intensity&quot;</span></span>
<span id="cb253-6"><a href="data-visualization.html#cb253-6" aria-hidden="true" tabindex="-1"></a>) <span class="sc">%&gt;%</span></span>
<span id="cb253-7"><a href="data-visualization.html#cb253-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x=</span>Marker,<span class="at">y=</span>ID,<span class="at">fill=</span>Intensity)) <span class="sc">+</span></span>
<span id="cb253-8"><a href="data-visualization.html#cb253-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_tile</span>()</span></code></pre></div>
<p><img src="main_files/heatmap%20geom%20tile-1.png" width="672" /></p>
<p>Note this ggplot method does not scale or reorder rows or columns.</p>
<div id="specifying-heatmap-colors" class="section level4" number="7.4.4.1">
<h4><span class="header-section-number">7.4.4.1</span> Specifying Heatmap Colors</h4>
<p>The colors of the heatmap may be changed by passing one of the <a href="https://stat.ethz.ch/R-manual/R-patched/library/grDevices/html/palettes.html">native color
palettes</a>
to the heatmap function with the <code>col</code> argument:</p>
<div class="sourceCode" id="cb254"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb254-1"><a href="data-visualization.html#cb254-1" aria-hidden="true" tabindex="-1"></a><span class="co"># native R colors are:</span></span>
<span id="cb254-2"><a href="data-visualization.html#cb254-2" aria-hidden="true" tabindex="-1"></a><span class="co"># - rainbow(n, start=.7, end=.1)</span></span>
<span id="cb254-3"><a href="data-visualization.html#cb254-3" aria-hidden="true" tabindex="-1"></a><span class="co"># - heat.colors(n)</span></span>
<span id="cb254-4"><a href="data-visualization.html#cb254-4" aria-hidden="true" tabindex="-1"></a><span class="co"># - terrain.colors(n)</span></span>
<span id="cb254-5"><a href="data-visualization.html#cb254-5" aria-hidden="true" tabindex="-1"></a><span class="co"># - topo.colors(n)</span></span>
<span id="cb254-6"><a href="data-visualization.html#cb254-6" aria-hidden="true" tabindex="-1"></a><span class="co"># - cm.colors(n)</span></span>
<span id="cb254-7"><a href="data-visualization.html#cb254-7" aria-hidden="true" tabindex="-1"></a><span class="co"># the n argument specifies the number of colors (i.e. resolution) of the colormap to return</span></span>
<span id="cb254-8"><a href="data-visualization.html#cb254-8" aria-hidden="true" tabindex="-1"></a><span class="fu">heatmap</span>(marker_matrix,<span class="at">col=</span><span class="fu">cm.colors</span>(<span class="dv">256</span>))</span></code></pre></div>
<p><img src="main_files/heatmap%20cm%20colors-1.png" width="672" /></p>
<p>To change the color using ggplot and <code>geom_tile()</code>, use the
<a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html"><code>scale_fill_gradientn</code></a>
function to specify a different color palette:</p>
<div class="sourceCode" id="cb255"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb255-1"><a href="data-visualization.html#cb255-1" aria-hidden="true" tabindex="-1"></a><span class="fu">pivot_longer</span>(</span>
<span id="cb255-2"><a href="data-visualization.html#cb255-2" aria-hidden="true" tabindex="-1"></a>  ad_metadata,</span>
<span id="cb255-3"><a href="data-visualization.html#cb255-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">c</span>(tau,abeta,iba1,gfap),</span>
<span id="cb255-4"><a href="data-visualization.html#cb255-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">names_to=</span><span class="st">&quot;Marker&quot;</span>,</span>
<span id="cb255-5"><a href="data-visualization.html#cb255-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">values_to=</span><span class="st">&quot;Intensity&quot;</span></span>
<span id="cb255-6"><a href="data-visualization.html#cb255-6" aria-hidden="true" tabindex="-1"></a>) <span class="sc">%&gt;%</span></span>
<span id="cb255-7"><a href="data-visualization.html#cb255-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x=</span>Marker,<span class="at">y=</span>ID,<span class="at">fill=</span>Intensity)) <span class="sc">+</span></span>
<span id="cb255-8"><a href="data-visualization.html#cb255-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_tile</span>() <span class="sc">+</span></span>
<span id="cb255-9"><a href="data-visualization.html#cb255-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_gradientn</span>(<span class="at">colors=</span><span class="fu">cm.colors</span>(<span class="dv">256</span>))</span></code></pre></div>
<p><img src="main_files/heatmap%20geom%20tile%20color%20change-1.png" width="672" />
The native color palettes in R look as follows:</p>
<p><img src="main_files/data%20viz%20heatmap%20native%20palettes-1.png" width="672" /></p>
<p>Instead of these color palettes, you may use one of the
<a href="https://colorbrewer2.org/#type=sequential&amp;scheme=BuGn&amp;n=3">ColorBrewer</a>
palettes in the <a href="https://cran.r-project.org/web/packages/RColorBrewer/index.html">RColorBrewer
package</a></p>
<div class="sourceCode" id="cb256"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb256-1"><a href="data-visualization.html#cb256-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(RColorBrewer)</span>
<span id="cb256-2"><a href="data-visualization.html#cb256-2" aria-hidden="true" tabindex="-1"></a><span class="fu">display.brewer.all</span>()</span></code></pre></div>
<p><img src="main_files/heatmap%20colorbrewer%20palettes-1.png" width="672" /></p>
<p>There are very many color palettes to choose from, and you may also specify your
own palettes with as much detail and complexity as you desire. More discussion
on how to choose colors is described in the next section.</p>
</div>
<div id="how-to-use-heatmaps-resonsibly-." class="section level4" number="7.4.4.2">
<h4><span class="header-section-number">7.4.4.2</span> How To Use Heatmaps Resonsibly .</h4>
<p>While heatmaps may seem intuitive, they are actually quite complicated and can
be difficult to create in a way that accurately depicts the underlying matrix.
More to come…</p>

</div>
</div>
</div>
<div id="other-kind-of-plots-." class="section level2" number="7.5">
<h2><span class="header-section-number">7.5</span> Other Kind of Plots .</h2>
<div id="chord-." class="section level3" number="7.5.1">
<h3><span class="header-section-number">7.5.1</span> Chord .</h3>
</div>
<div id="sankey-." class="section level3" number="7.5.2">
<h3><span class="header-section-number">7.5.2</span> Sankey .</h3>
</div>
<div id="dendrograms" class="section level3" number="7.5.3">
<h3><span class="header-section-number">7.5.3</span> Dendrograms</h3>
<p>Dendrograms are visualizations of tree structures (<em>dendron</em> means tree in
Greek). In the <a href="https://en.wikipedia.org/wiki/Graph_theory">graph theory</a> branch
of mathematics, a <a href="https://en.wikipedia.org/wiki/Tree_(graph_theory)">tree</a> is
defined as a graph composed of nodes and edges in which any two nodes are
connected by exactly one path. Such a graph is also called an <em>acyclic graph</em>,
where no paths exist that can be followed from a node back to iself. A full
treatment of trees and graph theory is beyond the scope of this book, but we
will discuss trees to the extent they are necessary to understand how to read a
dendrogram.</p>
<p>The concepts of a dendrogram are illustrated in the following figure:</p>
<div class="figure">
<img src="dendrograms.png" alt="" />
<p class="caption">Dendrogram Illustration</p>
</div>
<p>The tree drawn in a typical dendrogram is created using the output of a
clustering algorithm, e.g. <a href="data-science.html#hierarchical-clustering">hierarchical clustering</a>
that groups data points together based on their dissimilarity. Briefly, in the
figure, the data are one dimensional scalars, so the distance between all points
is simply the absolute value of the difference between them. The two closest
points are then merged into a single group, and the pairwise distances are
recomputed for all remaining data points with a summary of this new group (the
group summarization method is specified by the user and is called the <a href="https://en.wikipedia.org/wiki/Hierarchical_clustering#Linkage_criteria">linkage
critera</a>).
Then the process repeats, where the two nearest data points or summarized groups
are merged into a new group, the new group is summarized, pairwise distances
recomputed, etc. until all the data points have been included in a group. In
this way, all data points are assigned to a <em>hierarchy</em> of groups. In general,
any hierarchical grouping of data can be used to generate a dendrogram.</p>
<p>Given a hierarchical clustering result, a dendrogram can be drawn in several
ways in R. Using our AD marker example data, we hierarchically cluster the
subjects based on the content of all their markers and visualize the result as
follows:</p>
<div class="sourceCode" id="cb257"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb257-1"><a href="data-visualization.html#cb257-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggdendro)</span>
<span id="cb257-2"><a href="data-visualization.html#cb257-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb257-3"><a href="data-visualization.html#cb257-3" aria-hidden="true" tabindex="-1"></a><span class="co"># produce a clustering of the data using the hclust for hierarchical clustering</span></span>
<span id="cb257-4"><a href="data-visualization.html#cb257-4" aria-hidden="true" tabindex="-1"></a><span class="co"># and euclidean distance as the distance metric</span></span>
<span id="cb257-5"><a href="data-visualization.html#cb257-5" aria-hidden="true" tabindex="-1"></a>euc_dist <span class="ot">&lt;-</span> <span class="fu">dist</span>(dplyr<span class="sc">::</span><span class="fu">select</span>(ad_metadata,<span class="fu">c</span>(tau,abeta,iba1,gfap)))</span>
<span id="cb257-6"><a href="data-visualization.html#cb257-6" aria-hidden="true" tabindex="-1"></a>hc <span class="ot">&lt;-</span> <span class="fu">hclust</span>(euc_dist, <span class="at">method=</span><span class="st">&quot;ave&quot;</span>)</span>
<span id="cb257-7"><a href="data-visualization.html#cb257-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb257-8"><a href="data-visualization.html#cb257-8" aria-hidden="true" tabindex="-1"></a><span class="co"># add ID as labels to the clustering object</span></span>
<span id="cb257-9"><a href="data-visualization.html#cb257-9" aria-hidden="true" tabindex="-1"></a>hc<span class="sc">$</span>labels <span class="ot">&lt;-</span> ad_metadata<span class="sc">$</span>ID</span>
<span id="cb257-10"><a href="data-visualization.html#cb257-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb257-11"><a href="data-visualization.html#cb257-11" aria-hidden="true" tabindex="-1"></a><span class="fu">ggdendrogram</span>(hc)</span></code></pre></div>
<p><img src="main_files/dendro%20full-1.png" width="672" /></p>
<p>This dendrogram does not produce highly distinct clusters of AD and control
samples with all the data.</p>
<p>To illustrate what a dataset with strong clustering looks like, we draw
multivariate samples from two normal distributions and cluster the results:</p>
<div class="sourceCode" id="cb258"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb258-1"><a href="data-visualization.html#cb258-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(patchwork)</span>
<span id="cb258-2"><a href="data-visualization.html#cb258-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggdendro)</span>
<span id="cb258-3"><a href="data-visualization.html#cb258-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb258-4"><a href="data-visualization.html#cb258-4" aria-hidden="true" tabindex="-1"></a>well_clustered_data <span class="ot">&lt;-</span> <span class="fu">tibble</span>(</span>
<span id="cb258-5"><a href="data-visualization.html#cb258-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">ID=</span><span class="fu">c</span>(stringr<span class="sc">::</span><span class="fu">str_c</span>(<span class="st">&quot;A&quot;</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>),stringr<span class="sc">::</span><span class="fu">str_c</span>(<span class="st">&quot;B&quot;</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>)),</span>
<span id="cb258-6"><a href="data-visualization.html#cb258-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">f1=</span><span class="fu">c</span>(<span class="fu">rnorm</span>(<span class="dv">10</span>,<span class="dv">0</span>,<span class="dv">1</span>),<span class="fu">rnorm</span>(<span class="dv">10</span>,<span class="dv">10</span>,<span class="dv">1</span>)),</span>
<span id="cb258-7"><a href="data-visualization.html#cb258-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">f2=</span><span class="fu">c</span>(<span class="fu">rnorm</span>(<span class="dv">10</span>,<span class="dv">0</span>,<span class="dv">1</span>),<span class="fu">rnorm</span>(<span class="dv">10</span>,<span class="dv">10</span>,<span class="dv">1</span>))</span>
<span id="cb258-8"><a href="data-visualization.html#cb258-8" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb258-9"><a href="data-visualization.html#cb258-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb258-10"><a href="data-visualization.html#cb258-10" aria-hidden="true" tabindex="-1"></a>scatter_g <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(well_clustered_data, <span class="fu">aes</span>(<span class="at">x=</span>f1,<span class="at">y=</span>f2)) <span class="sc">+</span> <span class="fu">geom_point</span>()</span>
<span id="cb258-11"><a href="data-visualization.html#cb258-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb258-12"><a href="data-visualization.html#cb258-12" aria-hidden="true" tabindex="-1"></a><span class="co"># produce a clustering of the data using the hclust for hierarchical clustering</span></span>
<span id="cb258-13"><a href="data-visualization.html#cb258-13" aria-hidden="true" tabindex="-1"></a><span class="co"># and euclidean distance as the distance metric</span></span>
<span id="cb258-14"><a href="data-visualization.html#cb258-14" aria-hidden="true" tabindex="-1"></a>euc_dist <span class="ot">&lt;-</span> <span class="fu">dist</span>(dplyr<span class="sc">::</span><span class="fu">select</span>(well_clustered_data,<span class="sc">-</span>ID))</span>
<span id="cb258-15"><a href="data-visualization.html#cb258-15" aria-hidden="true" tabindex="-1"></a>hc <span class="ot">&lt;-</span> <span class="fu">hclust</span>(euc_dist, <span class="at">method=</span><span class="st">&quot;ave&quot;</span>)</span>
<span id="cb258-16"><a href="data-visualization.html#cb258-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb258-17"><a href="data-visualization.html#cb258-17" aria-hidden="true" tabindex="-1"></a><span class="co"># add ID as labels to the clustering object</span></span>
<span id="cb258-18"><a href="data-visualization.html#cb258-18" aria-hidden="true" tabindex="-1"></a>hc<span class="sc">$</span>labels <span class="ot">&lt;-</span> well_clustered_data<span class="sc">$</span>ID</span>
<span id="cb258-19"><a href="data-visualization.html#cb258-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb258-20"><a href="data-visualization.html#cb258-20" aria-hidden="true" tabindex="-1"></a>dendro_g <span class="ot">&lt;-</span> <span class="fu">ggdendrogram</span>(hc)</span>
<span id="cb258-21"><a href="data-visualization.html#cb258-21" aria-hidden="true" tabindex="-1"></a>scatter_g <span class="sc">|</span> dendro_g</span></code></pre></div>
<p><img src="main_files/dendro%20strong-1.png" width="672" /></p>
<p>Here, all the A and B samples strongly cluster together, with a large distance
between clusters.</p>
<div class="box readmore">
<ul>
<li><a href="https://en.wikipedia.org/wiki/Dendrogram">Dendrograms</a> Wikipedia page</li>
<li><a href="https://cran.r-project.org/web/packages/ggdendro/vignettes/ggdendro.html"><code>ggdendro</code> package vignette</a></li>
</ul>
</div>

</div>
</div>
<div id="tips-and-tricks-." class="section level2" number="7.6">
<h2><span class="header-section-number">7.6</span> Tips and Tricks .</h2>
<div id="reordering-1-d-data-elements-." class="section level3" number="7.6.1">
<h3><span class="header-section-number">7.6.1</span> Reordering 1-D Data Elements .</h3>
<p>How do we enforce an order on elements in a plot? Why might we want to do that?</p>
<div class="sourceCode" id="cb259"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb259-1"><a href="data-visualization.html#cb259-1" aria-hidden="true" tabindex="-1"></a>ad_metadata <span class="sc">%&gt;%</span></span>
<span id="cb259-2"><a href="data-visualization.html#cb259-2" aria-hidden="true" tabindex="-1"></a><span class="fu">arrange</span>(tau) <span class="sc">%&gt;%</span></span>
<span id="cb259-3"><a href="data-visualization.html#cb259-3" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="at">mapping =</span> <span class="fu">aes</span>(<span class="at">x=</span>ID,<span class="at">y=</span>tau,<span class="at">fill=</span>condition)) <span class="sc">+</span></span>
<span id="cb259-4"><a href="data-visualization.html#cb259-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_bar</span>(<span class="at">stat=</span><span class="st">&quot;identity&quot;</span>)</span></code></pre></div>
<p><img src="main_files/unnamed-chunk-95-1.png" width="672" /></p>
<p>But wait, our bars are in the same order as in the previous plot. Sorting our
data prior to plotting did not have the desired effect. The reason is that
although the <code>ID</code> column is a character string in our original tibble, ggplot
converts it into a <a href="#prog-factor">factor</a> prior to plotting. Recall that a
factor is a numeric type composed of sequential integers, each of which having a
character label associated with it.</p>
</div>
<div id="confidence-intervals-." class="section level3" number="7.6.2">
<h3><span class="header-section-number">7.6.2</span> Confidence Intervals .</h3>
<p>What are some ways to depict uncertainty in a plot?</p>
</div>
<div id="annotations-." class="section level3" number="7.6.3">
<h3><span class="header-section-number">7.6.3</span> Annotations .</h3>
<p>How do we add text annotations to plots? Manually vs data-driven.</p>

</div>
</div>
<div id="multiple-plots-." class="section level2" number="7.7">
<h2><span class="header-section-number">7.7</span> Multiple Plots .</h2>
<p>How do we add multiple datasets to plots?</p>
<div id="secondary-axes-." class="section level3" number="7.7.1">
<h3><span class="header-section-number">7.7.1</span> Secondary Axes .</h3>
<p>How and when do we add series with different scales to the same plot?</p>
</div>
<div id="facet-wrapping-." class="section level3" number="7.7.2">
<h3><span class="header-section-number">7.7.2</span> Facet wrapping .</h3>
<p>What is facet wrapping and in which contexts is it useful (i.e. group_by)?</p>
</div>
<div id="multipanel-figures-." class="section level3" number="7.7.3">
<h3><span class="header-section-number">7.7.3</span> Multipanel Figures .</h3>
<p>How do we add multiple, unrelated subplots to the same larger plot?</p>

</div>
</div>
<div id="responsible-plotting-." class="section level2" number="7.8">
<h2><span class="header-section-number">7.8</span> Responsible Plotting .</h2>
<p>“Good plots empower us to ask good questions.” - Alberto Cairo, How Charts Lie</p>

</div>
<div id="publication-ready-plots-." class="section level2" number="7.9">
<h2><span class="header-section-number">7.9</span> Publication Ready Plots .</h2>
<div class="box readmore">
<ul>
<li><a href="https://rpkgs.datanovia.com/ggpubr/">ggpubr - publication ready plotting with ggplot</a></li>
</ul>
</div>

</div>
<div id="network-visualization-." class="section level2" number="7.10">
<h2><span class="header-section-number">7.10</span> Network visualization .</h2>
<p>How do we visualize networks?</p>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="data-science.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="biology-bioinformatics.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
