<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta name="viewport" content="width=device-width, initial-scale=1" />
<meta property="og:title" content="6.1 Data Modeling | BF591 - R for Biological Sciences" />
<meta property="og:type" content="book" />







<script type="text/x-mathjax-config">
MathJax.Hub.Config({
  TeX: { equationNumbers: { autoNumber: "AMS" } }
});
</script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<meta name="description" content="6.1 Data Modeling | BF591 - R for Biological Sciences">

<title>6.1 Data Modeling | BF591 - R for Biological Sciences</title>

<script src="libs/header-attrs-2.13/header-attrs.js"></script>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="libs/bootstrap-3.3.5/css/darkly.min.css" rel="stylesheet" />
<script src="libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="libs/navigation-1.1/tabsets.js"></script>


<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>


<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />

<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
/* show arrow before summary tag as in bootstrap
TODO: remove if boostrap in updated in html_document (rmarkdown#1485) */
details > summary {
  display: list-item;
  cursor: pointer;
}
</style>
</head>

<body>

<div class="container-fluid main-container">


<div class="row">
<div class="col-sm-12">
<div id="TOC">
<ul>
<li><a href="index.html#syllabus">Syllabus</a>
<ul>
<li><a href="course-schedule.html#course-schedule">Course Schedule</a></li>
<li><a href="instructors.html#instructors">Instructors</a></li>
<li><a href="office-hours.html#office-hours">Office Hours</a></li>
<li><a href="course-values-and-policies.html#course-values-and-policies">Course Values and Policies</a></li>
</ul></li>
<li><a href="introduction.html#introduction"><span class="toc-section-number">1</span> Introduction</a>
<ul>
<li><a href="who-this-book-is-for.html#who-this-book-is-for"><span class="toc-section-number">1.1</span> Who This Book Is For</a></li>
<li><a href="a-note-about-reinventing-the-wheel.html#a-note-about-reinventing-the-wheel"><span class="toc-section-number">1.2</span> A Note About Reinventing the Wheel</a></li>
<li><a href="sources-and-references.html#sources-and-references"><span class="toc-section-number">1.3</span> Sources and References</a></li>
</ul></li>
<li><a href="data-bio.html#data-bio"><span class="toc-section-number">2</span> Data in Biology</a>
<ul>
<li><a href="a-brief-history-of-data-in-molecular-biology.html#a-brief-history-of-data-in-molecular-biology"><span class="toc-section-number">2.1</span> A Brief History of Data in Molecular Biology</a></li>
<li><a href="biology-as-a-mature-data-science.html#biology-as-a-mature-data-science"><span class="toc-section-number">2.2</span> Biology as a Mature Data Science</a></li>
</ul></li>
<li><a href="preliminaries.html#preliminaries"><span class="toc-section-number">3</span> Preliminaries</a>
<ul>
<li><a href="prelim-r.html#prelim-r"><span class="toc-section-number">3.1</span> The R Language</a></li>
<li><a href="prelim-rstudio.html#prelim-rstudio"><span class="toc-section-number">3.2</span> RStudio</a></li>
<li><a href="prog-the-r-script.html#prog-the-r-script"><span class="toc-section-number">3.3</span> The R Script</a></li>
<li><a href="prog-workflow.html#prog-workflow"><span class="toc-section-number">3.4</span> The Scripting Workflow</a></li>
<li><a href="prelim-git.html#prelim-git"><span class="toc-section-number">3.5</span> git + github</a>
<ul>
<li><a href="prelim-git.html#motivation"><span class="toc-section-number">3.5.1</span> Motivation</a></li>
<li><a href="prelim-git.html#git"><span class="toc-section-number">3.5.2</span> git</a></li>
<li><a href="prelim-git.html#git-hosting-platforms-github"><span class="toc-section-number">3.5.3</span> Git hosting platforms (GitHub)</a></li>
</ul></li>
</ul></li>
<li><a href="prog-basics.html#prog-basics"><span class="toc-section-number">4</span> R Programming</a>
<ul>
<li><a href="before-you-begin.html#before-you-begin"><span class="toc-section-number">4.1</span> Before you begin</a></li>
<li><a href="introduction-1.html#introduction-1"><span class="toc-section-number">4.2</span> Introduction</a></li>
<li><a href="prog-r-syntax.html#prog-r-syntax"><span class="toc-section-number">4.3</span> R Syntax Basics</a></li>
<li><a href="prog-types.html#prog-types"><span class="toc-section-number">4.4</span> Basic Types of Values</a>
<ul>
<li><a href="prog-types.html#factors"><span class="toc-section-number">4.4.1</span> Factors</a></li>
</ul></li>
<li><a href="prog-struct.html#prog-struct"><span class="toc-section-number">4.5</span> Data Structures</a>
<ul>
<li><a href="prog-struct.html#vectors"><span class="toc-section-number">4.5.1</span> Vectors</a></li>
<li><a href="prog-struct.html#matrices"><span class="toc-section-number">4.5.2</span> Matrices</a></li>
<li><a href="prog-struct.html#lists-and-data-frames"><span class="toc-section-number">4.5.3</span> Lists and data frames</a></li>
</ul></li>
<li><a href="logical-tests-and-comparators.html#logical-tests-and-comparators"><span class="toc-section-number">4.6</span> Logical Tests and Comparators</a></li>
<li><a href="functions.html#functions"><span class="toc-section-number">4.7</span> Functions</a>
<ul>
<li><a href="functions.html#prog-dry"><span class="toc-section-number">4.7.1</span> DRY: Don’t Repeat Yourself</a></li>
<li><a href="functions.html#writing-your-own-functions"><span class="toc-section-number">4.7.2</span> Writing your own functions</a></li>
<li><a href="functions.html#scope"><span class="toc-section-number">4.7.3</span> Scope</a></li>
</ul></li>
<li><a href="iteration.html#iteration"><span class="toc-section-number">4.8</span> Iteration</a>
<ul>
<li><a href="iteration.html#vectorized-operations"><span class="toc-section-number">4.8.1</span> Vectorized operations</a></li>
<li><a href="iteration.html#functional-programming"><span class="toc-section-number">4.8.2</span> Functional programming</a></li>
<li><a href="iteration.html#apply-and-friends"><span class="toc-section-number">4.8.3</span> <code>apply()</code> and friends</a></li>
</ul></li>
<li><a href="installing-packages.html#installing-packages"><span class="toc-section-number">4.9</span> Installing Packages</a></li>
<li><a href="saving-and-loading-r-data.html#saving-and-loading-r-data"><span class="toc-section-number">4.10</span> Saving and Loading R Data</a></li>
<li><a href="prog-debugging.html#prog-debugging"><span class="toc-section-number">4.11</span> Troubleshooting and Debugging</a></li>
<li><a href="prog-style.html#prog-style"><span class="toc-section-number">4.12</span> Coding Style and Conventions</a>
<ul>
<li><a href="prog-style.html#is-my-code-correct"><span class="toc-section-number">4.12.1</span> Is my code correct?</a></li>
<li><a href="prog-style.html#does-my-code-follow-the-dry-principle"><span class="toc-section-number">4.12.2</span> Does my code follow the <span>DRY principle</span>?</a></li>
<li><a href="prog-style.html#did-i-choose-concise-but-descriptive-variable-and-function-names"><span class="toc-section-number">4.12.3</span> Did I choose concise but descriptive variable and function names?</a></li>
<li><a href="prog-style.html#did-i-use-indentation-and-naming-conventions-consistently-throughout-my-code"><span class="toc-section-number">4.12.4</span> Did I use indentation and naming conventions consistently throughout my code?</a></li>
<li><a href="prog-style.html#did-i-write-comments-especially-when-what-the-code-does-is-not-obvious"><span class="toc-section-number">4.12.5</span> Did I write comments, especially when what the code does is not obvious?</a></li>
<li><a href="prog-style.html#how-easy-would-it-be-for-someone-else-to-understand-my-code"><span class="toc-section-number">4.12.6</span> How easy would it be for someone else to understand my code?</a></li>
<li><a href="prog-style.html#is-my-code-easy-to-maintainchange"><span class="toc-section-number">4.12.7</span> Is my code easy to maintain/change?</a></li>
<li><a href="prog-style.html#the-styler-package"><span class="toc-section-number">4.12.8</span> The <code>styler</code> package</a></li>
</ul></li>
</ul></li>
<li><a href="data-wrangling.html#data-wrangling"><span class="toc-section-number">5</span> Data Wrangling</a>
<ul>
<li><a href="the-tidyverse.html#the-tidyverse"><span class="toc-section-number">5.1</span> The Tidyverse</a></li>
<li><a href="dw-basics.html#dw-basics"><span class="toc-section-number">5.2</span> Tidyverse Basics</a></li>
<li><a href="importing-data.html#importing-data"><span class="toc-section-number">5.3</span> Importing Data</a></li>
<li><a href="data-tibble.html#data-tibble"><span class="toc-section-number">5.4</span> The tibble</a></li>
<li><a href="tidy-data.html#tidy-data"><span class="toc-section-number">5.5</span> Tidy Data</a></li>
<li><a href="pipes.html#pipes"><span class="toc-section-number">5.6</span> pipes</a></li>
<li><a href="arranging-data.html#arranging-data"><span class="toc-section-number">5.7</span> Arranging Data</a>
<ul>
<li><a href="arranging-data.html#dplyrmutate---create-new-columns-using-other-columns"><span class="toc-section-number">5.7.1</span> <code>dplyr::mutate()</code> - Create new columns using other columns</a></li>
<li><a href="arranging-data.html#stringr---working-with-character-values"><span class="toc-section-number">5.7.2</span> <code>stringr</code> - Working with character values</a></li>
<li><a href="arranging-data.html#dplyrselect---subset-columns-by-name"><span class="toc-section-number">5.7.3</span> <code>dplyr::select()</code> - Subset Columns by Name</a></li>
<li><a href="arranging-data.html#dplyrfilter---pick-rows-out-of-a-data-set"><span class="toc-section-number">5.7.4</span> <code>dplyr::filter()</code> - Pick rows out of a data set</a></li>
<li><a href="arranging-data.html#dplyrarrange---order-rows-based-on-their-values"><span class="toc-section-number">5.7.5</span> <code>dplyr::arrange()</code> - Order rows based on their values</a></li>
<li><a href="arranging-data.html#putting-it-all-together"><span class="toc-section-number">5.7.6</span> Putting it all together</a></li>
</ul></li>
<li><a href="grouping-data.html#grouping-data"><span class="toc-section-number">5.8</span> Grouping Data</a></li>
<li><a href="rearranging-data.html#rearranging-data"><span class="toc-section-number">5.9</span> Rearranging Data</a></li>
<li><a href="relational-data.html#relational-data"><span class="toc-section-number">5.10</span> Relational Data</a>
<ul>
<li><a href="relational-data.html#dealing-with-multiple-matches"><span class="toc-section-number">5.10.1</span> Dealing with multiple matches</a></li>
</ul></li>
</ul></li>
<li><a href="data-science.html#data-science"><span class="toc-section-number">6</span> Data Science</a>
<ul>
<li><a href="data-modeling.html#data-modeling"><span class="toc-section-number">6.1</span> Data Modeling</a>
<ul>
<li><a href="data-modeling.html#a-worked-modeling-example"><span class="toc-section-number">6.1.1</span> A Worked Modeling Example</a></li>
<li><a href="data-modeling.html#data-summarization"><span class="toc-section-number">6.1.2</span> Data Summarization</a></li>
<li><a href="data-modeling.html#linear-models"><span class="toc-section-number">6.1.3</span> Linear Models</a></li>
<li><a href="data-modeling.html#flavors-of-linear-models"><span class="toc-section-number">6.1.4</span> Flavors of Linear Models</a></li>
<li><a href="data-modeling.html#assessing-model-accuracy-."><span class="toc-section-number">6.1.5</span> Assessing Model Accuracy .</a></li>
</ul></li>
<li><a href="statistical-distributions.html#statistical-distributions"><span class="toc-section-number">6.2</span> Statistical Distributions</a>
<ul>
<li><a href="statistical-distributions.html#random-variables"><span class="toc-section-number">6.2.1</span> Random Variables</a></li>
<li><a href="statistical-distributions.html#statistical-distribution-basics"><span class="toc-section-number">6.2.2</span> Statistical Distribution Basics</a></li>
<li><a href="statistical-distributions.html#distributions-in-r"><span class="toc-section-number">6.2.3</span> Distributions in R</a></li>
<li><a href="statistical-distributions.html#discrete-distributions"><span class="toc-section-number">6.2.4</span> Discrete Distributions</a></li>
<li><a href="statistical-distributions.html#continuous-distributions"><span class="toc-section-number">6.2.5</span> Continuous Distributions</a></li>
<li><a href="statistical-distributions.html#empirical-distributions"><span class="toc-section-number">6.2.6</span> Empirical Distributions</a></li>
</ul></li>
<li><a href="statistical-tests.html#statistical-tests"><span class="toc-section-number">6.3</span> Statistical Tests</a>
<ul>
<li><a href="statistical-tests.html#what-is-a-statistical-test"><span class="toc-section-number">6.3.1</span> What is a statistical test</a></li>
<li><a href="statistical-tests.html#common-statistical-tests"><span class="toc-section-number">6.3.2</span> Common Statistical Tests</a></li>
<li><a href="statistical-tests.html#choosing-a-test"><span class="toc-section-number">6.3.3</span> Choosing a Test</a></li>
<li><a href="statistical-tests.html#p-values"><span class="toc-section-number">6.3.4</span> P-values</a></li>
<li><a href="statistical-tests.html#multiple-hypothesis-testing-."><span class="toc-section-number">6.3.5</span> Multiple Hypothesis Testing .</a></li>
<li><a href="statistical-tests.html#statistical-power"><span class="toc-section-number">6.3.6</span> Statistical power</a></li>
</ul></li>
<li><a href="exploratory-data-analysis.html#exploratory-data-analysis"><span class="toc-section-number">6.4</span> Exploratory Data Analysis</a>
<ul>
<li><a href="exploratory-data-analysis.html#principal-component-analysis"><span class="toc-section-number">6.4.1</span> Principal Component Analysis</a></li>
<li><a href="exploratory-data-analysis.html#cluster-analysis"><span class="toc-section-number">6.4.2</span> Cluster Analysis</a></li>
<li><a href="exploratory-data-analysis.html#k-means-."><span class="toc-section-number">6.4.3</span> k-means .</a></li>
<li><a href="exploratory-data-analysis.html#others"><span class="toc-section-number">6.4.4</span> Others</a></li>
</ul></li>
<li><a href="network-analysis-..html#network-analysis-."><span class="toc-section-number">6.5</span> Network Analysis .</a>
<ul>
<li><a href="network-analysis-..html#basic-network-analysis"><span class="toc-section-number">6.5.1</span> Basic network analysis</a></li>
<li><a href="network-analysis-..html#create-your-own"><span class="toc-section-number">6.5.2</span> Create your own</a></li>
</ul></li>
</ul></li>
<li><a href="data-visualization.html#data-visualization"><span class="toc-section-number">7</span> Data Visualization</a>
<ul>
<li><a href="responsible-plotting.html#responsible-plotting"><span class="toc-section-number">7.1</span> Responsible Plotting</a>
<ul>
<li><a href="responsible-plotting.html#visualization-principles"><span class="toc-section-number">7.1.1</span> Visualization Principles</a></li>
<li><a href="responsible-plotting.html#human-visual-perception"><span class="toc-section-number">7.1.2</span> Human Visual Perception</a></li>
<li><a href="responsible-plotting.html#visual-encodings"><span class="toc-section-number">7.1.3</span> Visual Encodings</a></li>
<li><a href="responsible-plotting.html#some-opinionated-rules-of-thumb"><span class="toc-section-number">7.1.4</span> Some Opinionated Rules of Thumb</a></li>
</ul></li>
<li><a href="dv-gg.html#dv-gg"><span class="toc-section-number">7.2</span> Grammar of Graphics</a>
<ul>
<li><a href="dv-gg.html#ggplot-mechanics"><span class="toc-section-number">7.2.1</span> <code>ggplot</code> mechanics</a></li>
</ul></li>
<li><a href="plotting-one-dimension.html#plotting-one-dimension"><span class="toc-section-number">7.3</span> Plotting One Dimension</a>
<ul>
<li><a href="plotting-one-dimension.html#dv-bar"><span class="toc-section-number">7.3.1</span> Bar chart</a></li>
<li><a href="plotting-one-dimension.html#lollipop-plots"><span class="toc-section-number">7.3.2</span> Lollipop plots</a></li>
<li><a href="plotting-one-dimension.html#stacked-area-charts"><span class="toc-section-number">7.3.3</span> Stacked Area charts</a></li>
<li><a href="plotting-one-dimension.html#parallel-coordinate-plots"><span class="toc-section-number">7.3.4</span> Parallel Coordinate plots</a></li>
</ul></li>
<li><a href="visualizing-distributions.html#visualizing-distributions"><span class="toc-section-number">7.4</span> Visualizing Distributions</a>
<ul>
<li><a href="visualizing-distributions.html#histogram"><span class="toc-section-number">7.4.1</span> Histogram</a></li>
<li><a href="visualizing-distributions.html#density"><span class="toc-section-number">7.4.2</span> Density</a></li>
<li><a href="visualizing-distributions.html#boxplot"><span class="toc-section-number">7.4.3</span> Boxplot</a></li>
<li><a href="visualizing-distributions.html#violin-plot"><span class="toc-section-number">7.4.4</span> Violin plot</a></li>
<li><a href="visualizing-distributions.html#beeswarm-plot"><span class="toc-section-number">7.4.5</span> Beeswarm plot</a></li>
<li><a href="visualizing-distributions.html#ridgeline"><span class="toc-section-number">7.4.6</span> Ridgeline</a></li>
</ul></li>
<li><a href="plotting-two-or-more-dimensions.html#plotting-two-or-more-dimensions"><span class="toc-section-number">7.5</span> Plotting Two or More Dimensions</a>
<ul>
<li><a href="plotting-two-or-more-dimensions.html#scatter-plots"><span class="toc-section-number">7.5.1</span> Scatter Plots</a></li>
<li><a href="plotting-two-or-more-dimensions.html#bubble-plots"><span class="toc-section-number">7.5.2</span> Bubble Plots</a></li>
<li><a href="plotting-two-or-more-dimensions.html#connected-scatter-plots"><span class="toc-section-number">7.5.3</span> Connected Scatter Plots</a></li>
<li><a href="plotting-two-or-more-dimensions.html#line-plots"><span class="toc-section-number">7.5.4</span> Line Plots</a></li>
<li><a href="plotting-two-or-more-dimensions.html#parallel-coordinate-plots-1"><span class="toc-section-number">7.5.5</span> Parallel Coordinate Plots</a></li>
<li><a href="plotting-two-or-more-dimensions.html#heatmaps"><span class="toc-section-number">7.5.6</span> Heatmaps</a></li>
<li><a href="plotting-two-or-more-dimensions.html#how-to-use-heatmaps-responsibly"><span class="toc-section-number">7.5.7</span> How To Use Heatmaps Responsibly</a></li>
</ul></li>
<li><a href="other-kind-of-plots.html#other-kind-of-plots"><span class="toc-section-number">7.6</span> Other Kind of Plots</a>
<ul>
<li><a href="other-kind-of-plots.html#dendrograms"><span class="toc-section-number">7.6.1</span> Dendrograms</a></li>
<li><a href="other-kind-of-plots.html#chord-diagrams-and-circos-plots"><span class="toc-section-number">7.6.2</span> Chord Diagrams and Circos Plots</a></li>
</ul></li>
<li><a href="multiple-plots.html#multiple-plots"><span class="toc-section-number">7.7</span> Multiple Plots</a>
<ul>
<li><a href="multiple-plots.html#facet-wrapping"><span class="toc-section-number">7.7.1</span> Facet wrapping</a></li>
<li><a href="multiple-plots.html#multipanel-figures"><span class="toc-section-number">7.7.2</span> Multipanel Figures</a></li>
</ul></li>
<li><a href="publication-ready-plots.html#publication-ready-plots"><span class="toc-section-number">7.8</span> Publication Ready Plots</a>
<ul>
<li><a href="publication-ready-plots.html#ggplot-themes"><span class="toc-section-number">7.8.1</span> <code>ggplot</code> Themes</a></li>
<li><a href="publication-ready-plots.html#exporting-to-svg"><span class="toc-section-number">7.8.2</span> Exporting to SVG</a></li>
</ul></li>
<li><a href="network-visualization.html#network-visualization"><span class="toc-section-number">7.9</span> Network visualization</a>
<ul>
<li><a href="network-visualization.html#interactive-network-plots"><span class="toc-section-number">7.9.1</span> Interactive network plots</a></li>
</ul></li>
</ul></li>
<li><a href="biology-bioinformatics.html#biology-bioinformatics"><span class="toc-section-number">8</span> Biology &amp; Bioinformatics</a>
<ul>
<li><a href="bio-r.html#bio-r"><span class="toc-section-number">8.1</span> R in Biology</a></li>
<li><a href="biological-data-overview.html#biological-data-overview"><span class="toc-section-number">8.2</span> Biological Data Overview</a>
<ul>
<li><a href="biological-data-overview.html#bio-data-types"><span class="toc-section-number">8.2.1</span> Types of Biological Data</a></li>
<li><a href="biological-data-overview.html#csv-files"><span class="toc-section-number">8.2.2</span> CSV Files</a></li>
<li><a href="biological-data-overview.html#biological-data-is-not-tidy"><span class="toc-section-number">8.2.3</span> Biological data is NOT Tidy!</a></li>
</ul></li>
<li><a href="bio-bioconductor.html#bio-bioconductor"><span class="toc-section-number">8.3</span> Bioconductor</a></li>
<li><a href="microarrays.html#microarrays"><span class="toc-section-number">8.4</span> Microarrays</a></li>
<li><a href="high-throughput-sequencing.html#high-throughput-sequencing"><span class="toc-section-number">8.5</span> High Throughput Sequencing</a>
<ul>
<li><a href="high-throughput-sequencing.html#raw-hts-data"><span class="toc-section-number">8.5.1</span> Raw HTS Data</a></li>
<li><a href="high-throughput-sequencing.html#preliminary-hts-data-analysis"><span class="toc-section-number">8.5.2</span> Preliminary HTS Data Analysis</a></li>
<li><a href="high-throughput-sequencing.html#count-data"><span class="toc-section-number">8.5.3</span> Count Data</a></li>
</ul></li>
<li><a href="gene-identifiers.html#gene-identifiers"><span class="toc-section-number">8.6</span> Gene Identifiers</a>
<ul>
<li><a href="gene-identifiers.html#gene-identifier-systems"><span class="toc-section-number">8.6.1</span> Gene Identifier Systems</a></li>
<li><a href="gene-identifiers.html#mapping-between-identifier-systems"><span class="toc-section-number">8.6.2</span> Mapping Between Identifier Systems</a></li>
<li><a href="gene-identifiers.html#mapping-homologs"><span class="toc-section-number">8.6.3</span> Mapping Homologs</a></li>
</ul></li>
<li><a href="gene-expression.html#gene-expression"><span class="toc-section-number">8.7</span> Gene Expression</a>
<ul>
<li><a href="gene-expression.html#gene-expression-data-in-bioconductor"><span class="toc-section-number">8.7.1</span> Gene Expression Data in Bioconductor</a></li>
<li><a href="gene-expression.html#differential-expression-analysis"><span class="toc-section-number">8.7.2</span> Differential Expression Analysis</a></li>
<li><a href="gene-expression.html#microarray-gene-expression-data"><span class="toc-section-number">8.7.3</span> Microarray Gene Expression Data</a></li>
<li><a href="gene-expression.html#differential-expression-microarrays-limma"><span class="toc-section-number">8.7.4</span> Differential Expression: Microarrays (limma)</a></li>
<li><a href="gene-expression.html#rnaseq"><span class="toc-section-number">8.7.5</span> RNASeq</a></li>
<li><a href="gene-expression.html#rnaseq-gene-expression-data"><span class="toc-section-number">8.7.6</span> RNASeq Gene Expression Data</a></li>
<li><a href="gene-expression.html#filtering-counts"><span class="toc-section-number">8.7.7</span> Filtering Counts</a></li>
<li><a href="gene-expression.html#count-distributions"><span class="toc-section-number">8.7.8</span> Count Distributions</a></li>
<li><a href="gene-expression.html#differential-expression-rnaseq"><span class="toc-section-number">8.7.9</span> Differential Expression: RNASeq</a></li>
</ul></li>
<li><a href="gene-set-enrichment-analysis.html#gene-set-enrichment-analysis"><span class="toc-section-number">8.8</span> Gene Set Enrichment Analysis</a>
<ul>
<li><a href="gene-set-enrichment-analysis.html#gene-sets"><span class="toc-section-number">8.8.1</span> Gene Sets</a></li>
<li><a href="gene-set-enrichment-analysis.html#working-with-gene-sets-in-r"><span class="toc-section-number">8.8.2</span> Working with gene sets in R</a></li>
<li><a href="gene-set-enrichment-analysis.html#over-representation-analysis"><span class="toc-section-number">8.8.3</span> Over-representation Analysis</a></li>
<li><a href="gene-set-enrichment-analysis.html#rank-based-analysis"><span class="toc-section-number">8.8.4</span> Rank-based Analysis</a></li>
<li><a href="gene-set-enrichment-analysis.html#fgsea"><span class="toc-section-number">8.8.5</span> fgsea</a></li>
</ul></li>
<li><a href="biological-networks-..html#biological-networks-."><span class="toc-section-number">8.9</span> Biological Networks .</a>
<ul>
<li><a href="biological-networks-..html#biological-pathways-."><span class="toc-section-number">8.9.1</span> Biological Pathways .</a></li>
<li><a href="biological-networks-..html#gene-regulatory-networks-."><span class="toc-section-number">8.9.2</span> Gene Regulatory Networks .</a></li>
<li><a href="biological-networks-..html#protein-protein-interaction-network-."><span class="toc-section-number">8.9.3</span> Protein-Protein Interaction Network .</a></li>
<li><a href="biological-networks-..html#wgcna-."><span class="toc-section-number">8.9.4</span> WGCNA .</a></li>
</ul></li>
</ul></li>
<li><a href="engineering.html#engineering"><span class="toc-section-number">9</span> EngineeRing</a>
<ul>
<li><a href="unit-testing.html#unit-testing"><span class="toc-section-number">9.1</span> Unit Testing</a></li>
<li><a href="toolification.html#toolification"><span class="toc-section-number">9.2</span> Toolification</a>
<ul>
<li><a href="toolification.html#the-r-interpreter"><span class="toc-section-number">9.2.1</span> The R Interpreter</a></li>
<li><a href="toolification.html#rscript"><span class="toc-section-number">9.2.2</span> <code>Rscript</code></a></li>
<li><a href="toolification.html#commandargs"><span class="toc-section-number">9.2.3</span> <code>commandArgs()</code></a></li>
</ul></li>
<li><a href="parallel-processing.html#parallel-processing"><span class="toc-section-number">9.3</span> Parallel Processing</a>
<ul>
<li><a href="parallel-processing.html#brief-introduction-to-parallelization"><span class="toc-section-number">9.3.1</span> Brief Introduction to Parallelization</a></li>
<li><a href="parallel-processing.html#apply-and-friends-are-pleasingly-parallel"><span class="toc-section-number">9.3.2</span> <code>apply</code> and Friends Are Pleasingly Parallel</a></li>
<li><a href="parallel-processing.html#the-parallel-package"><span class="toc-section-number">9.3.3</span> The <code>parallel</code> package</a></li>
</ul></li>
<li><a href="object-orientation-in-r.html#object-orientation-in-r"><span class="toc-section-number">9.4</span> Object Orientation in R</a></li>
<li><a href="r-packages.html#r-packages"><span class="toc-section-number">9.5</span> R Packages</a></li>
</ul></li>
<li><a href="rshiny.html#rshiny"><span class="toc-section-number">10</span> RShiny</a>
<ul>
<li><a href="rshiny-introduction.html#rshiny-introduction"><span class="toc-section-number">10.1</span> RShiny Introduction</a>
<ul>
<li><a href="rshiny-introduction.html#how-does-a-web-page-work"><span class="toc-section-number">10.1.1</span> How does a web page work?</a></li>
</ul></li>
<li><a href="application-structure.html#application-structure"><span class="toc-section-number">10.2</span> Application Structure</a>
<ul>
<li><a href="application-structure.html#ui-design"><span class="toc-section-number">10.2.1</span> UI Design</a></li>
<li><a href="application-structure.html#server-functionality"><span class="toc-section-number">10.2.2</span> Server Functionality</a></li>
</ul></li>
<li><a href="reactivity.html#reactivity"><span class="toc-section-number">10.3</span> Reactivity</a></li>
<li><a href="publishing.html#publishing"><span class="toc-section-number">10.4</span> Publishing</a></li>
</ul></li>
<li><a href="communicating-with-r.html#communicating-with-r"><span class="toc-section-number">11</span> Communicating with R</a>
<ul>
<li><a href="rmarkdown-knitr.html#rmarkdown-knitr"><span class="toc-section-number">11.1</span> RMarkdown &amp; knitr</a></li>
<li><a href="bookdown.html#bookdown"><span class="toc-section-number">11.2</span> bookdown</a></li>
</ul></li>
<li><a href="contribution-guide.html#contribution-guide"><span class="toc-section-number">12</span> Contribution Guide</a>
<ul>
<li><a href="custom-blocks.html#custom-blocks"><span class="toc-section-number">12.1</span> Custom blocks</a></li>
</ul></li>
<li><a href="assignments.html#assignments">Assignments</a>
<ul>
<li><a href="assignment-format.html#assignment-format">Assignment Format</a></li>
<li><a href="starting-an-assignment.html#starting-an-assignment">Starting an Assignment</a>
<ul>
<li><a href="starting-an-assignment.html#starting-your-own-git-repository-and-cloning-it">Starting your own <code>git</code> repository and cloning it</a></li>
<li><a href="starting-an-assignment.html#contents">Contents</a></li>
<li><a href="starting-an-assignment.html#github-tutorial">GitHub tutorial</a></li>
<li><a href="starting-an-assignment.html#committing-and-pushing-with-r-and-without">Committing and Pushing with R (and without)</a></li>
<li><a href="starting-an-assignment.html#using-rstudio">Using RStudio</a></li>
<li><a href="starting-an-assignment.html#using-the-command-line">Using the command line</a></li>
</ul></li>
<li><a href="assignment-1.html#assignment-1">Assignment 1</a>
<ul>
<li><a href="assignment-1.html#problem-statement">Problem Statement</a></li>
<li><a href="assignment-1.html#learning-objectives-and-skill-list">Learning Objectives and Skill List</a></li>
<li><a href="assignment-1.html#instructions">Instructions</a></li>
<li><a href="assignment-1.html#hints">Hints</a></li>
</ul></li>
<li><a href="assignment-2.html#assignment-2">Assignment 2</a>
<ul>
<li><a href="assignment-2.html#problem-statement-1">Problem Statement</a></li>
<li><a href="assignment-2.html#learning-objectives">Learning Objectives</a></li>
<li><a href="assignment-2.html#skill-list">Skill List</a></li>
<li><a href="assignment-2.html#instructions-1">Instructions</a></li>
<li><a href="assignment-2.html#function-details">Function Details</a></li>
</ul></li>
<li><a href="assignment-3.html#assignment-3">Assignment 3</a>
<ul>
<li><a href="assignment-3.html#problem-statement-2">Problem Statement</a></li>
<li><a href="assignment-3.html#learning-objectives-1">Learning Objectives</a></li>
<li><a href="assignment-3.html#skill-list-1">Skill List</a></li>
<li><a href="assignment-3.html#background-on-microarrays">Background on Microarrays</a></li>
<li><a href="assignment-3.html#background-on-principal-component-analysis">Background on Principal Component Analysis</a></li>
<li><a href="assignment-3.html#marisa-et-al.-gene-expression-classification-of-colon-cancer-into-molecular-subtypes-characterization-validation-and-prognostic-value.-plos-medicine-may-2013.-pmid-23700391">Marisa et al. Gene Expression Classification of Colon Cancer into Molecular Subtypes: Characterization, Validation, and Prognostic Value. PLoS Medicine, May 2013. PMID: 23700391</a></li>
<li><a href="assignment-3.html#scaling-data-using-r-scale">Scaling data using R <code>scale()</code></a></li>
<li><a href="assignment-3.html#proportion-of-variance-explained">Proportion of variance explained</a></li>
<li><a href="assignment-3.html#plotting-and-visualization-of-pca">Plotting and visualization of PCA</a></li>
<li><a href="assignment-3.html#hierarchical-clustering-and-heatmaps">Hierarchical Clustering and Heatmaps</a></li>
<li><a href="assignment-3.html#references">References</a></li>
</ul></li>
<li><a href="assignment-4.html#assignment-4">Assignment 4</a>
<ul>
<li><a href="assignment-4.html#problem-statement-3">Problem Statement</a></li>
<li><a href="assignment-4.html#learning-objectives-2">Learning Objectives</a></li>
<li><a href="assignment-4.html#skill-list-2">Skill List</a></li>
<li><a href="assignment-4.html#instructions-2">Instructions</a></li>
<li><a href="assignment-4.html#deliverables">Deliverables</a></li>
<li><a href="assignment-4.html#function-details-1">Function Details</a></li>
<li><a href="assignment-4.html#hints-1">Hints</a></li>
</ul></li>
<li><a href="assignment-5.html#assignment-5">Assignment 5</a>
<ul>
<li><a href="assignment-5.html#problem-statement-4">Problem Statement</a></li>
<li><a href="assignment-5.html#learning-objectives-3">Learning Objectives</a></li>
<li><a href="assignment-5.html#skill-list-3">Skill List</a></li>
<li><a href="assignment-5.html#deseq2-background">DESeq2 Background</a></li>
<li><a href="assignment-5.html#generating-a-counts-matrix">Generating a counts matrix</a></li>
<li><a href="assignment-5.html#prefiltering-counts-matrix">Prefiltering Counts matrix</a></li>
<li><a href="assignment-5.html#median-of-ratios-normalization">Median-of-ratios normalization</a></li>
<li><a href="assignment-5.html#deseq2-preparation">DESeq2 preparation</a></li>
<li><a href="assignment-5.html#omeara-et-al.-transcriptional-reversion-of-cardiac-myocyte-fate-during-mammalian-cardiac-regeneration.-circ-res.-feb-2015.-pmid-25477501l">O’Meara et al. Transcriptional Reversion of Cardiac Myocyte Fate During Mammalian Cardiac Regeneration. Circ Res. Feb 2015. PMID: 25477501l</a></li>
<li><a href="assignment-5.html#reading-and-subsetting-the-data-from-verse_counts.tsv-and-sample_metadata.csv">1. Reading and subsetting the data from verse_counts.tsv and sample_metadata.csv</a></li>
<li><a href="assignment-5.html#running-deseq2">2. Running DESeq2</a></li>
<li><a href="assignment-5.html#annotating-results-to-construct-a-labeled-volcano-plot">3. Annotating results to construct a labeled volcano plot</a></li>
<li><a href="assignment-5.html#diagnostic-plot-of-the-raw-p-values-for-all-genes">4. Diagnostic plot of the raw p-values for all genes</a></li>
<li><a href="assignment-5.html#plotting-the-logfoldchanges-for-differentially-expressed-genes">5. Plotting the LogFoldChanges for differentially expressed genes</a></li>
<li><a href="assignment-5.html#the-choice-of-fdr-cutoff-depends-on-cost">The choice of FDR cutoff depends on cost</a></li>
<li><a href="assignment-5.html#plotting-the-normalized-counts-of-differentially-expressed-genes">6. Plotting the normalized counts of differentially expressed genes</a></li>
<li><a href="assignment-5.html#volcano-plot-to-visualize-differential-expression-results">7. Volcano Plot to visualize differential expression results</a></li>
<li><a href="assignment-5.html#running-fgsea-vignette">8. Running fgsea <span>vignette</span></a></li>
<li><a href="assignment-5.html#plotting-the-top-ten-positive-nes-and-top-ten-negative-nes-pathways">9. Plotting the top ten positive NES and top ten negative NES pathways</a></li>
<li><a href="assignment-5.html#references-1">References</a></li>
</ul></li>
<li><a href="assignment-6.html#assignment-6">Assignment 6</a>
<ul>
<li><a href="assignment-6.html#problem-statement-5">Problem Statement</a></li>
<li><a href="assignment-6.html#learning-objectives-4">Learning Objectives</a></li>
<li><a href="assignment-6.html#skill-list-4">Skill List</a></li>
<li><a href="assignment-6.html#instructions-3">Instructions</a></li>
<li><a href="assignment-6.html#function-details-2">Function Details</a></li>
</ul></li>
<li><a href="assignment-7.html#assignment-7">Assignment 7</a>
<ul>
<li><a href="assignment-7.html#problem-statement-6">Problem Statement</a></li>
<li><a href="assignment-7.html#learning-objectives-5">Learning Objectives</a></li>
<li><a href="assignment-7.html#skill-list-5">Skill List</a></li>
<li><a href="assignment-7.html#instructions-4">Instructions</a></li>
<li><a href="assignment-7.html#function-details-3">“Function” Details</a></li>
</ul></li>
</ul></li>
<li class="appendix"><span><b>Appendix</b></span></li>
<li><a href="class-outlines.html#class-outlines"><span class="toc-section-number">A</span> Class Outlines</a>
<ul>
<li><a href="week-1.html#week-1"><span class="toc-section-number">A.1</span> Week 1</a></li>
<li><a href="week-2.html#week-2"><span class="toc-section-number">A.2</span> Week 2</a></li>
<li><a href="week-3.html#week-3"><span class="toc-section-number">A.3</span> Week 3</a></li>
<li><a href="week-4.html#week-4"><span class="toc-section-number">A.4</span> Week 4</a></li>
<li><a href="week-5.html#week-5"><span class="toc-section-number">A.5</span> Week 5</a></li>
<li><a href="week-6.html#week-6"><span class="toc-section-number">A.6</span> Week 6</a></li>
<li><a href="week-7.html#week-7"><span class="toc-section-number">A.7</span> Week 7</a></li>
<li><a href="week-8.html#week-8"><span class="toc-section-number">A.8</span> Week 8</a></li>
<li><a href="week-9.html#week-9"><span class="toc-section-number">A.9</span> Week 9</a></li>
<li><a href="week-10.html#week-10"><span class="toc-section-number">A.10</span> Week 10</a></li>
<li><a href="week-11.html#week-11"><span class="toc-section-number">A.11</span> Week 11</a></li>
<li><a href="week-12.html#week-12"><span class="toc-section-number">A.12</span> Week 12</a></li>
<li><a href="week-13.html#week-13"><span class="toc-section-number">A.13</span> Week 13</a></li>
<li><a href="week-14.html#week-14"><span class="toc-section-number">A.14</span> Week 14</a></li>
</ul></li>
</ul>
</div>
</div>
</div>
<div class="row">
<div class="col-sm-12">
<div id="data-modeling" class="section level2" number="6.1">
<h2><span class="header-section-number">6.1</span> Data Modeling</h2>
<p>The goal of data modeling is to describe a dataset using a relatively small
number of mathematical relationships. Said differently, a model uses some parts
of a dataset to try to accurately predict other parts of the dataset in a way
that is useful to us.</p>
<p>Models are human inventions; they reflect our beliefs about the way the universe
works. The successful model identifies patterns within a dataset that are the
result of causal relationships in the universe that led to the phenomena that
were measured while accounting for noise in the data. However, <em>the model itself
does not identify or even accurately reflect those causal effects</em>. The model
merely summarizes patterns and we as scientists are left to interpret those
patterns and design follow up experiments to investigate the nature of those
causal relationships using our prior knowledge of the world.</p>
<p>There are several principles to keep in mind when modeling data:</p>
<ol style="list-style-type: decimal">
<li><p><strong>Data are never wrong.</strong> All data are collected using processes and
devices designed and implemented by humans, who always have biases and make
assumptions. All data measure something about the universe, and so are “true” in
some sense of the word. If what we intended to measure and what we actually
measured were not the same thing, that is due to our errors in collection or
interpretation, not due to the data being wrong. If we approach our dataset with
a particular set of hypotheses and the data don’t support those hypotheses, it
is our beliefs of the world and our understanding of the dataset that are wrong,
not the data itself.</p></li>
<li><p><strong>Not all data are useful.</strong> Just because data isn’t wrong, it doesn’t mean
it is useful. There may have been systematic errors in the collection of the
data that makes interpreting them difficult. Data collected for one purpose may
not be useful for any other purposes. And sometimes, a dataset collected for a
particular purpose may simply not have the information needed to answer our
questions; if what we measure has no relationship to what we wish to predict,
the data itself is not useful - though the knowledge that what we measured has
no detectable effect on the thing we wish to predict may be very useful!</p></li>
<li><p><strong>“All models are wrong, but some are useful.”</strong> George Box, the renowned
British statistician, famously asserted this in a 1976 paper to the Journal of
the American Statistical Association. <span class="citation">(<a href="#ref-Box1976-rd" role="doc-biblioref">Box 1976</a>)</span>. By this he meant that every
model we create is a simplification of the system we are seeking to model, which
is by definition not identical to that system. To perfectly model a system, our
model would need to be precisely the system we are modeling, which is no longer
a model but the system itself. Fortunately, even though we know our models are
always wrong to some degree, they may nonetheless be useful because they are not
<strong>too</strong> wrong. Some models may indeed be too wrong, though.</p></li>
<li><p><strong>Data do not contain causal information.</strong> “Correlation does not mean
causation.” Data are measurements of the results of a process in the universe
that we wish to understand; the data are possibly reflective of that process,
but do not contain any information about the process itself. We cannot infer
causal relationships from a dataset alone. We must construct possible causal
models using our knowledge of the world first, then apply our data to our model
and other alternative models to compare their relative plausibility.</p></li>
<li><p><strong>All data have noise.</strong> The usefulness of a model to describe a dataset is
related to the relative strength of the patterns and noise in the dataset when
viewed through the lens of the model; conceptually, the so-called “signal to
noise ratio” of the data. The fundamental concern of statistics is quantifying
uncertainty (i.e. noise), and separating it from real signal, though different
statistical approaches (e.g. frequentist vs Bayesian) reason about uncertainty
in different ways.</p></li>
</ol>
<p>Modeling begins (or should begin) with posing one or more <strong>scientific models</strong>
of the process or phenomenon we wish to understand. The scientific model is
conceptual; it reflects our belief of the universe and proposes a causal
explanation for the phenomenon. We then decide how to map that scientific model
onto a <strong>statistical model</strong>, which is a mechanical procedure that quantifies
how well our scientific model explains a dataset. <em>The scientific model and
statistical model are related but independent choices we make.</em> There may be
many valid statistical models that represent a given scientific model. However,
sometimes in practice we lack sufficient knowledge about the process to propose
scientific models first, requiring data exploration and summarization first to
suggest reasonable starting points.</p>
<div class="box important">
<p>This section pertains primarily to models specified explicitly by humans. There
is another class of models, namely those created by certain machine learning
algorithms like neural networks and deep learning, that <em>discover models from
data</em>. These models are fundamentally different than those designed by human
minds, in that they are often accurate and therefore useful, but it can be very
difficult if not impossible to understand how they work. While these are
important types of models that fall under the umbrella of data science, we limit
the content of this chapter to human designed statistical models.</p>
</div>
<div id="a-worked-modeling-example" class="section level3" number="6.1.1">
<h3><span class="header-section-number">6.1.1</span> A Worked Modeling Example</h3>
<p>As an example, let’s consider a scenario where we wish to assess whether any of
three genes can help us distinguish between patients who have Parkinson’s
Disease and those who don’t by measuring the relative activity of those genes in
blood samples. We have the following made-up dataset:</p>
<div class="sourceCode" id="cb161"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb161-1"><a href="data-modeling.html#cb161-1" aria-hidden="true" tabindex="-1"></a>gene_exp</span></code></pre></div>
<pre><code>## # A tibble: 200 x 5
##    sample_name `Disease Status` `Gene 1` `Gene 2` `Gene 3`
##    &lt;chr&gt;       &lt;fct&gt;               &lt;dbl&gt;    &lt;dbl&gt;    &lt;dbl&gt;
##  1 P1          Parkinson&#39;s          257.     636.     504.
##  2 P2          Parkinson&#39;s          241.     556.     484.
##  3 P3          Parkinson&#39;s          252.     426.     476.
##  4 P4          Parkinson&#39;s          271.     405.     458.
##  5 P5          Parkinson&#39;s          248.     482.     520.
##  6 P6          Parkinson&#39;s          275.     521.     460.
##  7 P7          Parkinson&#39;s          264.     415.     568.
##  8 P8          Parkinson&#39;s          276.     450.     495.
##  9 P9          Parkinson&#39;s          274.     490.     496.
## 10 P10         Parkinson&#39;s          251.     584.     573.
## # ... with 190 more rows</code></pre>
<p>Our imaginary dataset has 100 Parkinson’s and 100 control subjects. For each of
our samples, we have a sample ID, Disease Status of <code>Parkinson's</code> or <code>Control</code>,
and numeric measurements of each of three genes. Below are <a href="visualizing-distributions.html#violin-plot">violin
plots</a> of our (made-up) data set for these three genes:</p>
<p><img src="main_files/data%20modeling%20violin-1.png" width="672" /></p>
<p>By inspection, it appears that Gene 1 has no relationship with disease; we may
safely eliminate this gene from further consideration. Gene 2 appears to have a
different profile depending on disease status, where control individuals have
a higher average expression and a lower variance. Unfortunately, despite this
qualitative difference, this gene may not be useful for telling whether someone
has disease or not - the ranges completely overlap. Gene 3 appears to
discriminate between disease and control. There is some overlap in the two
expression distributions, but above a certain expression value these data
suggest a high degree of predictive accuracy may be obtained with this gene.
Measuring this gene may therefore be useful, if the results from this dataset
generalize to all people with Parkinson’s Disease.</p>
<p>So far, we have not done any modeling, but instead relied on plotting and our
eyes. A more quantitative question might be: how much higher is Gene 3
expression in Parkinson’s Disease than control? Another way of posing this
question is: if I know a patient has Parkinson’s Disease, what Gene 3 expression
value do I expect them to have? Written this way, we have turned our question
into a prediction problem: if we only had information that a patient had
Parkinson’s Disease, what is the predicted expression value of their Gene 3?</p>
<p>Another way to pose this prediction question is in the opposite (and arguably
more useful) direction: if all we knew about a person was their Gene 3 gene
expression, how likely is it that the person has Parkinson’s Disease? If this
gene expression is predictive enough of a person’s disease status, it may be a
viable <a href="https://en.wikipedia.org/wiki/Biomarker">biomarker</a> of disease and thus
might be useful in a clinical setting, for example when identifying
presymptomatic individuals or assessing the efficacy of a pharmacological
treatment.</p>
<p>Although it may seems obvious, before beginning to model a dataset, we must
start by posing the <strong>scientific question</strong> as concisely as possible, as we have
done above. These questions will help us identify which modeling techniques are
appropriate and help us ensure we interpret our results correctly.</p>
<p>We will use this example dataset throughout this chapter to illustrate some key
concepts.</p>
</div>
<div id="data-summarization" class="section level3" number="6.1.2">
<h3><span class="header-section-number">6.1.2</span> Data Summarization</h3>
<p>Broadly speaking, <em>data summarization</em> is the process of finding a
lower-dimensional representation of a larger dataset. There are many ways to
summarize a set of data; each approach will emphasize different aspects of the
dataset, and have varying degrees of accuracy. Consider the gene expression of
Gene 1 for all individuals in our example above, plotted as a distribution with
a <a href="visualizing-distributions.html#histogram">histogram</a>:</p>
<div class="sourceCode" id="cb163"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb163-1"><a href="data-modeling.html#cb163-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(gene_exp, <span class="fu">aes</span>(<span class="at">x=</span><span class="st">`</span><span class="at">Gene 1</span><span class="st">`</span>)) <span class="sc">+</span></span>
<span id="cb163-2"><a href="data-modeling.html#cb163-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">bins=</span><span class="dv">30</span>,<span class="at">fill=</span><span class="st">&quot;#a93c13&quot;</span>)</span></code></pre></div>
<p><img src="main_files/data%20sci%20summ%20hist-1.png" width="672" /></p>
<div id="point-estimates" class="section level4" number="6.1.2.1">
<h4><span class="header-section-number">6.1.2.1</span> Point Estimates</h4>
<p>The data are concentrated around the value 250, and become less common for
larger and smaller values. Since the extents to the left and right of the middle
of the distribution appear to be equally distant, perhaps the arithmetic mean
is a good way to identify the middle of the distribution:</p>
<div class="sourceCode" id="cb164"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb164-1"><a href="data-modeling.html#cb164-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(gene_exp, <span class="fu">aes</span>(<span class="at">x=</span><span class="st">`</span><span class="at">Gene 1</span><span class="st">`</span>)) <span class="sc">+</span></span>
<span id="cb164-2"><a href="data-modeling.html#cb164-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">bins=</span><span class="dv">30</span>,<span class="at">fill=</span><span class="st">&quot;#a93c13&quot;</span>) <span class="sc">+</span></span>
<span id="cb164-3"><a href="data-modeling.html#cb164-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept=</span><span class="fu">mean</span>(gene_exp<span class="sc">$</span><span class="st">`</span><span class="at">Gene 1</span><span class="st">`</span>))</span></code></pre></div>
<p><img src="main_files/data%20sci%20summ%20hist2-1.png" width="672" /></p>
<p>By eye, the mean does seem to correspond well to the value that is among the
most frequent, and successfully captures an important aspect of the data: its
<em>central tendency</em>. Summaries that compute a single number are called <em>point
estimates</em>. Point estimates collapse the data into one singular point, one
value.</p>
<div class="box">
<p>The arithmetic mean is just one measure of central tendency, computed by taking
the sum of all the values and dividing by the number of values. The mean may be
a good point estimate of the central tendency of a dataset, but it is sensitive
to outlier samples. Consider the following examples:</p>
<div class="sourceCode" id="cb165"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb165-1"><a href="data-modeling.html#cb165-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(patchwork)</span>
<span id="cb165-2"><a href="data-modeling.html#cb165-2" aria-hidden="true" tabindex="-1"></a>well_behaved_data <span class="ot">&lt;-</span> <span class="fu">tibble</span>(<span class="at">data =</span> <span class="fu">rnorm</span>(<span class="dv">1000</span>))</span>
<span id="cb165-3"><a href="data-modeling.html#cb165-3" aria-hidden="true" tabindex="-1"></a>data_w_outliers <span class="ot">&lt;-</span> <span class="fu">tibble</span>(<span class="at">data =</span> <span class="fu">c</span>(<span class="fu">rnorm</span>(<span class="dv">800</span>), <span class="fu">rep</span>(<span class="dv">5</span>, <span class="dv">200</span>))) <span class="co"># oops I add some outliers :^)</span></span>
<span id="cb165-4"><a href="data-modeling.html#cb165-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb165-5"><a href="data-modeling.html#cb165-5" aria-hidden="true" tabindex="-1"></a>g_no_outlier <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(well_behaved_data, <span class="fu">aes</span>(<span class="at">x =</span> data)) <span class="sc">+</span></span>
<span id="cb165-6"><a href="data-modeling.html#cb165-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">fill =</span> <span class="st">&quot;#56CBF9&quot;</span>, <span class="at">color =</span> <span class="st">&quot;grey&quot;</span>, <span class="at">bins =</span> <span class="dv">30</span>) <span class="sc">+</span></span>
<span id="cb165-7"><a href="data-modeling.html#cb165-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="fu">mean</span>(well_behaved_data<span class="sc">$</span>data)) <span class="sc">+</span></span>
<span id="cb165-8"><a href="data-modeling.html#cb165-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">&quot;Mean example, no outliers&quot;</span>)</span>
<span id="cb165-9"><a href="data-modeling.html#cb165-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb165-10"><a href="data-modeling.html#cb165-10" aria-hidden="true" tabindex="-1"></a>g_outlier <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(data_w_outliers, <span class="fu">aes</span>(<span class="at">x =</span> data)) <span class="sc">+</span></span>
<span id="cb165-11"><a href="data-modeling.html#cb165-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">fill =</span> <span class="st">&quot;#7FBEEB&quot;</span>, <span class="at">color =</span> <span class="st">&quot;grey&quot;</span>, <span class="at">bins =</span> <span class="dv">30</span>) <span class="sc">+</span></span>
<span id="cb165-12"><a href="data-modeling.html#cb165-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="fu">mean</span>(data_w_outliers<span class="sc">$</span>data)) <span class="sc">+</span></span>
<span id="cb165-13"><a href="data-modeling.html#cb165-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">&quot;Mean example, big outliers&quot;</span>)</span>
<span id="cb165-14"><a href="data-modeling.html#cb165-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb165-15"><a href="data-modeling.html#cb165-15" aria-hidden="true" tabindex="-1"></a>g_no_outlier <span class="sc">|</span> g_outlier</span></code></pre></div>
<p><img src="main_files/unnamed-chunk-43-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>The median is another measure of central tendency, which is found by identifying
the value that divides the samples into equal halves when sorted from smallest
to largest. The median is more robust in the presence of outliers.</p>
<div class="sourceCode" id="cb166"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb166-1"><a href="data-modeling.html#cb166-1" aria-hidden="true" tabindex="-1"></a>g_no_outlier <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(well_behaved_data, <span class="fu">aes</span>(<span class="at">x =</span> data)) <span class="sc">+</span></span>
<span id="cb166-2"><a href="data-modeling.html#cb166-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">fill =</span> <span class="st">&quot;#AFBED1&quot;</span>, <span class="at">color =</span> <span class="st">&quot;grey&quot;</span>, <span class="at">bins =</span> <span class="dv">30</span>) <span class="sc">+</span></span>
<span id="cb166-3"><a href="data-modeling.html#cb166-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="fu">median</span>(well_behaved_data<span class="sc">$</span>data)) <span class="sc">+</span></span>
<span id="cb166-4"><a href="data-modeling.html#cb166-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">&quot;Median example&quot;</span>)</span>
<span id="cb166-5"><a href="data-modeling.html#cb166-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb166-6"><a href="data-modeling.html#cb166-6" aria-hidden="true" tabindex="-1"></a>g_outlier <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(data_w_outliers, <span class="fu">aes</span>(<span class="at">x =</span> data)) <span class="sc">+</span></span>
<span id="cb166-7"><a href="data-modeling.html#cb166-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">fill =</span> <span class="st">&quot;#7FBEEB&quot;</span>, <span class="at">color =</span> <span class="st">&quot;grey&quot;</span>, <span class="at">bins =</span> <span class="dv">30</span>) <span class="sc">+</span></span>
<span id="cb166-8"><a href="data-modeling.html#cb166-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="fu">median</span>(data_w_outliers<span class="sc">$</span>data)) <span class="sc">+</span></span>
<span id="cb166-9"><a href="data-modeling.html#cb166-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">&quot;Median example, big outliers&quot;</span>)</span>
<span id="cb166-10"><a href="data-modeling.html#cb166-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb166-11"><a href="data-modeling.html#cb166-11" aria-hidden="true" tabindex="-1"></a>g_no_outlier <span class="sc">|</span> g_outlier</span></code></pre></div>
<p><img src="main_files/unnamed-chunk-44-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
</div>
<div id="dispersion" class="section level4" number="6.1.2.2">
<h4><span class="header-section-number">6.1.2.2</span> Dispersion</h4>
<p>Central tendencies are important aspects of the data but don’t describe what the
data do for values outside this point estimate of central tendency; in other
words, we have not expressed the spread, or <em>dispersion</em> of the data.</p>
<p>We decide that perhaps computing the <a href="https://en.wikipedia.org/wiki/Standard_deviation">standard
deviation</a> of the data may
characterize the spread well, since it appears to be symmetric around the mean.
We can layer this information on the plot as well to inspect it:</p>
<div class="sourceCode" id="cb167"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb167-1"><a href="data-modeling.html#cb167-1" aria-hidden="true" tabindex="-1"></a>g1_mean <span class="ot">&lt;-</span> <span class="fu">mean</span>(gene_exp<span class="sc">$</span><span class="st">`</span><span class="at">Gene 1</span><span class="st">`</span>)</span>
<span id="cb167-2"><a href="data-modeling.html#cb167-2" aria-hidden="true" tabindex="-1"></a>g1_sd <span class="ot">&lt;-</span> <span class="fu">sd</span>(gene_exp<span class="sc">$</span><span class="st">`</span><span class="at">Gene 1</span><span class="st">`</span>)</span>
<span id="cb167-3"><a href="data-modeling.html#cb167-3" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(gene_exp, <span class="fu">aes</span>(<span class="at">x=</span><span class="st">`</span><span class="at">Gene 1</span><span class="st">`</span>)) <span class="sc">+</span></span>
<span id="cb167-4"><a href="data-modeling.html#cb167-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">bins=</span><span class="dv">30</span>,<span class="at">fill=</span><span class="st">&quot;#a93c13&quot;</span>) <span class="sc">+</span></span>
<span id="cb167-5"><a href="data-modeling.html#cb167-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept=</span>g1_mean) <span class="sc">+</span></span>
<span id="cb167-6"><a href="data-modeling.html#cb167-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_segment</span>(<span class="at">x=</span>g1_mean<span class="sc">-</span>g1_sd, <span class="at">y=</span><span class="dv">10</span>, <span class="at">xend=</span>g1_mean<span class="sc">+</span>g1_sd, <span class="at">yend=</span><span class="dv">10</span>)</span></code></pre></div>
<p><img src="main_files/data%20sci%20summ%20sd-1.png" width="672" /></p>
<p>The width of the horizontal line is proportional to the mean +/- one standard
deviation around the mean, and has been placed arbitrarily on the y axis at <code>y = 10</code> to show how this range covers the data in the histogram. The +/- 1 standard
deviation around the mean visually describes the spread of the data reasonably
well.</p>
<div class="box">
<p>Measures the spread of the data, typically around its perceived center (a mean).
Often related to the distribution of the data.</p>
<p><strong>Standard deviation</strong>: A measure of how close values are to the mean. Bigger
standard deviations mean data is more spread out.</p>
<div class="sourceCode" id="cb168"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb168-1"><a href="data-modeling.html#cb168-1" aria-hidden="true" tabindex="-1"></a>data <span class="ot">&lt;-</span> <span class="fu">tibble</span>(<span class="at">data =</span> <span class="fu">c</span>(<span class="fu">rnorm</span>(<span class="dv">1000</span>, <span class="at">sd=</span><span class="fl">1.75</span>)))</span>
<span id="cb168-2"><a href="data-modeling.html#cb168-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(data, <span class="fu">aes</span>(<span class="at">x =</span> data)) <span class="sc">+</span></span>
<span id="cb168-3"><a href="data-modeling.html#cb168-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">fill =</span> <span class="st">&quot;#EAC5D8&quot;</span>, <span class="at">color =</span> <span class="st">&quot;white&quot;</span>, <span class="at">bins =</span> <span class="dv">30</span>) <span class="sc">+</span></span>
<span id="cb168-4"><a href="data-modeling.html#cb168-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="fu">c</span>(<span class="sc">-</span><span class="dv">2</span>, <span class="sc">-</span><span class="dv">1</span>, <span class="dv">0</span>, <span class="dv">1</span>, <span class="dv">2</span>) <span class="sc">*</span> <span class="fu">sd</span>(data<span class="sc">$</span>data)) <span class="sc">+</span></span>
<span id="cb168-5"><a href="data-modeling.html#cb168-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlim</span>(<span class="fu">c</span>(<span class="sc">-</span><span class="dv">6</span>, <span class="dv">6</span>)) <span class="sc">+</span></span>
<span id="cb168-6"><a href="data-modeling.html#cb168-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">&quot;Standard deviations aplenty&quot;</span>, <span class="fu">paste</span>(<span class="st">&quot;SD:&quot;</span>, <span class="fu">sd</span>(data<span class="sc">$</span>data)))</span></code></pre></div>
<p><img src="main_files/taylor%20sd-1.png" width="672" style="display: block; margin: auto;" /></p>
<p><strong>Variance</strong>:
Similar to SD (it’s the square of SD), variance measures how far a random value
is from the mean.</p>
<div class="sourceCode" id="cb169"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb169-1"><a href="data-modeling.html#cb169-1" aria-hidden="true" tabindex="-1"></a>data <span class="ot">&lt;-</span> <span class="fu">tibble</span>(<span class="at">data =</span> <span class="fu">c</span>(<span class="fu">rnorm</span>(<span class="dv">1000</span>, <span class="at">sd=</span><span class="fl">0.5</span>)))</span>
<span id="cb169-2"><a href="data-modeling.html#cb169-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(data, <span class="fu">aes</span>(<span class="at">x =</span> data)) <span class="sc">+</span></span>
<span id="cb169-3"><a href="data-modeling.html#cb169-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">fill =</span> <span class="st">&quot;#DBD8F0&quot;</span>, <span class="at">color =</span> <span class="st">&quot;white&quot;</span>, <span class="at">bins =</span> <span class="dv">30</span>) <span class="sc">+</span></span>
<span id="cb169-4"><a href="data-modeling.html#cb169-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="fu">mean</span>(data<span class="sc">$</span>data)) <span class="sc">+</span></span>
<span id="cb169-5"><a href="data-modeling.html#cb169-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlim</span>(<span class="fu">c</span>(<span class="sc">-</span><span class="dv">6</span>, <span class="dv">6</span>)) <span class="sc">+</span></span>
<span id="cb169-6"><a href="data-modeling.html#cb169-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">&quot;Same mean as SD plot, but different variance&quot;</span>,</span>
<span id="cb169-7"><a href="data-modeling.html#cb169-7" aria-hidden="true" tabindex="-1"></a>          <span class="fu">paste</span>(<span class="st">&quot;Variance:&quot;</span>, <span class="fu">sd</span>(data<span class="sc">$</span>data)))</span></code></pre></div>
<p><img src="main_files/taylor%20var-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
</div>
<div id="distributions" class="section level4" number="6.1.2.3">
<h4><span class="header-section-number">6.1.2.3</span> Distributions</h4>
<p>With these two pieces of knowledge - the mean accurately describes the center of
the data and the standard deviation describes the spread - we now recognize that
these data may be <a href="https://en.wikipedia.org/wiki/Normal_distribution">normally distributed</a>, and therefore
we can potentially describe the dataset mathematically. We decide to visually
inspect this possibility by layering a normal distribution on top of our data
using
<a href="https://ggplot2.tidyverse.org/reference/geom_function.html"><code>stat_function</code></a>:</p>
<div class="sourceCode" id="cb170"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb170-1"><a href="data-modeling.html#cb170-1" aria-hidden="true" tabindex="-1"></a>g1_mean <span class="ot">&lt;-</span> <span class="fu">mean</span>(gene_exp<span class="sc">$</span><span class="st">`</span><span class="at">Gene 1</span><span class="st">`</span>)</span>
<span id="cb170-2"><a href="data-modeling.html#cb170-2" aria-hidden="true" tabindex="-1"></a>g1_sd <span class="ot">&lt;-</span> <span class="fu">sd</span>(gene_exp<span class="sc">$</span><span class="st">`</span><span class="at">Gene 1</span><span class="st">`</span>)</span>
<span id="cb170-3"><a href="data-modeling.html#cb170-3" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(gene_exp, <span class="fu">aes</span>(<span class="at">x=</span><span class="st">`</span><span class="at">Gene 1</span><span class="st">`</span>)) <span class="sc">+</span></span>
<span id="cb170-4"><a href="data-modeling.html#cb170-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(</span>
<span id="cb170-5"><a href="data-modeling.html#cb170-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">aes</span>(<span class="at">y=</span><span class="fu">after_stat</span>(density)),</span>
<span id="cb170-6"><a href="data-modeling.html#cb170-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">bins=</span><span class="dv">30</span>,</span>
<span id="cb170-7"><a href="data-modeling.html#cb170-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">fill=</span><span class="st">&quot;#a93c13&quot;</span></span>
<span id="cb170-8"><a href="data-modeling.html#cb170-8" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">+</span></span>
<span id="cb170-9"><a href="data-modeling.html#cb170-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">stat_function</span>(<span class="at">fun=</span>dnorm, <span class="at">args =</span> <span class="fu">list</span>(<span class="at">mean=</span>g1_mean, <span class="at">sd=</span>g1_sd), <span class="at">size=</span><span class="dv">2</span>)</span></code></pre></div>
<p><img src="main_files/unnamed-chunk-45-1.png" width="672" /></p>
<p>Note the histogram bars are scaled with
<code>aes(y=[after_stat](https://ggplot2.tidyverse.org/reference/aes_eval.html)(density))</code>
to the density of values in each bin to make all the bar heights sum to 1 so
that the y scale matches that of a normal distribution.</p>
<p>We have now created our first model: we chose to express the dataset as a normal
distribution parameterized by the mean and standard deviation and standard
deviation of the data. Using the values of 254 and 11
as our mean and standard deviation, respectively, we can express our model
mathematically as follows:</p>
<p><span class="math display">\[
Gene\;1 \sim \mathcal{N}(254, 11)
\]</span></p>
<p>Here the <span class="math inline">\(\sim\)</span> symbol means “distributed as” and the <span class="math inline">\(\mathcal{N}(\mu,\sigma)\)</span>
represents a normal distribution with mean of <span class="math inline">\(\mu\)</span> and standard deviation of
<span class="math inline">\(\sigma\)</span>. This is mathematical formulation means the same thing as saying we are
modeling Gene 1 expression as a normal distribution with mean of 254 and
standard deviation of 11. Without any additional information about a new sample,
we would expect the expression of that gene to be 254, although it may vary from
this value.</p>
<div class="box">
<p>The normal distribution is the most common distribution observed in nature, but
it is hardly the only one. We could have proposed other distributions to instead
summarize our data:</p>
<div class="sourceCode" id="cb171"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb171-1"><a href="data-modeling.html#cb171-1" aria-hidden="true" tabindex="-1"></a>g_norm <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(<span class="fu">tibble</span>(<span class="at">data =</span> <span class="fu">rnorm</span>(<span class="dv">5000</span>)), <span class="fu">aes</span>(<span class="at">x =</span> data)) <span class="sc">+</span></span>
<span id="cb171-2"><a href="data-modeling.html#cb171-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">fill =</span> <span class="st">&quot;#D0FCB3&quot;</span>, <span class="at">bins =</span> <span class="dv">50</span>, <span class="at">color =</span> <span class="st">&quot;gray&quot;</span>) <span class="sc">+</span></span>
<span id="cb171-3"><a href="data-modeling.html#cb171-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">&quot;Normal distribution&quot;</span>, <span class="st">&quot;rnorm(n = 1000)&quot;</span>)</span>
<span id="cb171-4"><a href="data-modeling.html#cb171-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb171-5"><a href="data-modeling.html#cb171-5" aria-hidden="true" tabindex="-1"></a>g_unif <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(<span class="fu">tibble</span>(<span class="at">data =</span> <span class="fu">runif</span>(<span class="dv">5000</span>)), <span class="fu">aes</span>(<span class="at">x =</span> data)) <span class="sc">+</span></span>
<span id="cb171-6"><a href="data-modeling.html#cb171-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">fill =</span> <span class="st">&quot;#271F30&quot;</span>, <span class="at">bins =</span> <span class="dv">50</span>, <span class="at">color =</span> <span class="st">&quot;white&quot;</span>) <span class="sc">+</span></span>
<span id="cb171-7"><a href="data-modeling.html#cb171-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">&quot;Uniform distribution&quot;</span>, <span class="st">&quot;runif(n = 1000)&quot;</span>)</span>
<span id="cb171-8"><a href="data-modeling.html#cb171-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb171-9"><a href="data-modeling.html#cb171-9" aria-hidden="true" tabindex="-1"></a>g_logistic <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(<span class="fu">tibble</span>(<span class="at">data =</span> <span class="fu">rlogis</span>(<span class="dv">5000</span>)), <span class="fu">aes</span>(<span class="at">x =</span> data)) <span class="sc">+</span></span>
<span id="cb171-10"><a href="data-modeling.html#cb171-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">fill =</span> <span class="st">&quot;#9BC59D&quot;</span>, <span class="at">bins =</span> <span class="dv">50</span>, <span class="at">color =</span> <span class="st">&quot;black&quot;</span>) <span class="sc">+</span></span>
<span id="cb171-11"><a href="data-modeling.html#cb171-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">&quot;Logistic distribution&quot;</span>, <span class="st">&quot;rlogis(n = 1000)&quot;</span>)</span>
<span id="cb171-12"><a href="data-modeling.html#cb171-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb171-13"><a href="data-modeling.html#cb171-13" aria-hidden="true" tabindex="-1"></a>g_exp <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(<span class="fu">tibble</span>(<span class="at">data =</span> <span class="fu">rexp</span>(<span class="dv">5000</span>, <span class="at">rate =</span> <span class="dv">1</span>)), <span class="fu">aes</span>(<span class="at">x =</span> data)) <span class="sc">+</span></span>
<span id="cb171-14"><a href="data-modeling.html#cb171-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">fill =</span> <span class="st">&quot;#6C5A49&quot;</span>, <span class="at">bins =</span> <span class="dv">50</span>, <span class="at">color =</span> <span class="st">&quot;white&quot;</span>) <span class="sc">+</span></span>
<span id="cb171-15"><a href="data-modeling.html#cb171-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">&quot;Exponential distribution&quot;</span>, <span class="st">&quot;rexp(n = 1000, rate = 1)&quot;</span>)</span>
<span id="cb171-16"><a href="data-modeling.html#cb171-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb171-17"><a href="data-modeling.html#cb171-17" aria-hidden="true" tabindex="-1"></a>(g_norm <span class="sc">|</span> g_unif) <span class="sc">/</span> (g_logistic <span class="sc">|</span> g_exp)</span></code></pre></div>
<p><img src="main_files/unnamed-chunk-46-1.png" width="672" style="display: block; margin: auto;" />
In addition to the normal distribution, we have also plotted samples drawn from
a <a href="ihttps://en.wikipedia.org/wiki/Continuous_uniform_distribution">continuous uniform
distribution</a>
between 0 and 1, a <a href="https://en.wikipedia.org/wiki/Logistic_distribution">logistic
distribution</a> which is
similar to the normal distribution but has heavier “tails,” and an <a href="https://en.wikipedia.org/wiki/Exponential_distribution">exponential
distribution</a>.
There are many more distributions than these, and many of them were discovered
to arise in nature and encode different types of processes and relationships.</p>
</div>
<p>A few notes on our data modeling example before we move on:</p>
<ol style="list-style-type: decimal">
<li><p><strong>Our model choice was totally subjective.</strong> We looked at the data and
decided that a normal distribution was a reasonable choice. There were many
other choices we could have made, and all of them would be equally valid, though
they may not all describe the data equally well.</p></li>
<li><p><strong>We can’t know if this is the “correct” model for the data.</strong> By eye, it
appears to be a reasonably accurate summary. However, there is no such thing as
a correct model; some models are simply better at describing the data than
others. Recall that all models are wrong, and some models are useful. Our model
may be useful, but it is definitely wrong to some extent.</p></li>
<li><p><strong>We don’t know how well our model describes the data yet.</strong> So far we’ve
only used our eyes to choose our model which might be a good starting point
considering our data are so simple, but we have not yet quantified how well our
model describes the data, or compared it to alternative models to see which is
better. This will be discussed briefly in a later section.</p></li>
</ol>
</div>
</div>
<div id="linear-models" class="section level3" number="6.1.3">
<h3><span class="header-section-number">6.1.3</span> Linear Models</h3>
<p>Our choice of a normal distribution to model our Gene 1 gene expression was only
<em>descriptive</em>; it was a low-dimensional summary of our dataset. However, it was
not very <em>informative</em>; it doesn’t tell us anything useful about Gene 1
expression with respect to our scientific question of distinguishing between
Parkinson’s Disease and Control individuals. To do that, we will need to find a
model that can make <em>predictions</em> that we may find useful if we receive new
data. To do that, we will introduce a new type of model: the linear model.</p>
<p>A linear model is any statistical model that relates one outcome variable as a
linear combination (i.e. sum) of one or more explanatory variables. This may be
expressed mathematically as follows:</p>
<p><span class="math display">\[
Y_i = \beta_0 + \beta_1 \phi_1 ( X_{i1} ) + \beta_2 \phi_2 ( X_{i2} ) + \ldots + \beta_p \phi_p ( X_{ip} ) + \epsilon_i
\]</span></p>
<p>Above, <span class="math inline">\(Y_i\)</span> is some outcome or response variable we wish to model, <span class="math inline">\(X_{ij}\)</span> is
our explanatory or predictor variable <span class="math inline">\(j\)</span> for observatoin <span class="math inline">\(i\)</span>, and <span class="math inline">\(\beta_j\)</span> are
coefficients estimated to minimize the difference between the predicted outcome
<span class="math inline">\(\hat{Y_i}\)</span> and the observed <span class="math inline">\(Y_i\)</span> over all observations. <span class="math inline">\(\phi_j\)</span> is a possibly
non-linear transformation of the explanatory variable <span class="math inline">\(X_ij\)</span>; note these
functions may be non-linear so long as the predicted outcome is modeled as a
linear combination of the transformed variables. The rest of this section is
dedicated to a worked example of a linear model for gene expression data.</p>
<p>Let us begin with a <a href="visualizing-distributions.html#beeswarm-plot">beeswarm plot</a> plot of Gene 3:</p>
<div class="sourceCode" id="cb172"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb172-1"><a href="data-modeling.html#cb172-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggbeeswarm)</span>
<span id="cb172-2"><a href="data-modeling.html#cb172-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(gene_exp, <span class="fu">aes</span>(<span class="at">x=</span><span class="st">`</span><span class="at">Disease Status</span><span class="st">`</span>, <span class="at">y=</span><span class="st">`</span><span class="at">Gene 3</span><span class="st">`</span>, <span class="at">color=</span><span class="st">`</span><span class="at">Disease Status</span><span class="st">`</span>)) <span class="sc">+</span></span>
<span id="cb172-3"><a href="data-modeling.html#cb172-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_beeswarm</span>()</span></code></pre></div>
<p><img src="main_files/lin%20model%20gene%203%20violin-1.png" width="672" /></p>
<p>The expression values within each disease status look like they might be
normally distributed just like Gene 1, so let’s summarize each group with the
arithmetic mean and standard deviation as before, and instead plot both
distributions as histograms:</p>
<div class="sourceCode" id="cb173"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb173-1"><a href="data-modeling.html#cb173-1" aria-hidden="true" tabindex="-1"></a>exp_summ <span class="ot">&lt;-</span> <span class="fu">pivot_longer</span>(</span>
<span id="cb173-2"><a href="data-modeling.html#cb173-2" aria-hidden="true" tabindex="-1"></a>  gene_exp,</span>
<span id="cb173-3"><a href="data-modeling.html#cb173-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">c</span>(<span class="st">`</span><span class="at">Gene 3</span><span class="st">`</span>)</span>
<span id="cb173-4"><a href="data-modeling.html#cb173-4" aria-hidden="true" tabindex="-1"></a>) <span class="sc">%&gt;%</span></span>
<span id="cb173-5"><a href="data-modeling.html#cb173-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(<span class="st">`</span><span class="at">Disease Status</span><span class="st">`</span>) <span class="sc">%&gt;%</span></span>
<span id="cb173-6"><a href="data-modeling.html#cb173-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarize</span>(<span class="at">mean=</span><span class="fu">mean</span>(value),<span class="at">sd=</span><span class="fu">sd</span>(value))</span>
<span id="cb173-7"><a href="data-modeling.html#cb173-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb173-8"><a href="data-modeling.html#cb173-8" aria-hidden="true" tabindex="-1"></a>pd_mean <span class="ot">&lt;-</span> <span class="fu">filter</span>(exp_summ, <span class="st">`</span><span class="at">Disease Status</span><span class="st">`</span> <span class="sc">==</span> <span class="st">&quot;Parkinson&#39;s&quot;</span>)<span class="sc">$</span>mean</span>
<span id="cb173-9"><a href="data-modeling.html#cb173-9" aria-hidden="true" tabindex="-1"></a>c_mean <span class="ot">&lt;-</span> <span class="fu">filter</span>(exp_summ, <span class="st">`</span><span class="at">Disease Status</span><span class="st">`</span> <span class="sc">==</span> <span class="st">&quot;Control&quot;</span>)<span class="sc">$</span>mean</span>
<span id="cb173-10"><a href="data-modeling.html#cb173-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb173-11"><a href="data-modeling.html#cb173-11" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(gene_exp, <span class="fu">aes</span>(<span class="at">x=</span><span class="st">`</span><span class="at">Gene 3</span><span class="st">`</span>, <span class="at">fill=</span><span class="st">`</span><span class="at">Disease Status</span><span class="st">`</span>)) <span class="sc">+</span></span>
<span id="cb173-12"><a href="data-modeling.html#cb173-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">bins=</span><span class="dv">20</span>, <span class="at">alpha=</span><span class="fl">0.6</span>,<span class="at">position=</span><span class="st">&quot;identity&quot;</span>) <span class="sc">+</span></span>
<span id="cb173-13"><a href="data-modeling.html#cb173-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">annotate</span>(<span class="st">&quot;segment&quot;</span>, <span class="at">x=</span>c_mean, <span class="at">xend=</span>pd_mean, <span class="at">y=</span><span class="dv">20</span>, <span class="at">yend=</span><span class="dv">20</span>, <span class="at">arrow=</span><span class="fu">arrow</span>(<span class="at">ends=</span><span class="st">&quot;both&quot;</span>, <span class="at">angle=</span><span class="dv">90</span>)) <span class="sc">+</span></span>
<span id="cb173-14"><a href="data-modeling.html#cb173-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">annotate</span>(<span class="st">&quot;text&quot;</span>, <span class="at">x=</span><span class="fu">mean</span>(<span class="fu">c</span>(c_mean,pd_mean)), <span class="at">y=</span><span class="dv">21</span>, <span class="at">hjust=</span><span class="fl">0.5</span>, <span class="at">label=</span><span class="st">&quot;How different?&quot;</span>)</span></code></pre></div>
<p><img src="main_files/unnamed-chunk-47-1.png" width="672" /></p>
<p>We can make a point estimate of this difference by simply subtracting the means:</p>
<div class="sourceCode" id="cb174"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb174-1"><a href="data-modeling.html#cb174-1" aria-hidden="true" tabindex="-1"></a>pd_mean <span class="sc">-</span> c_mean</span></code></pre></div>
<pre><code>## [1] 164.0942</code></pre>
<p>In other words, this point estimate suggests that on average Parkinson’s
patients have 164.1 greater expression than Controls. We can
plot this relationship relatively simply:</p>
<div class="sourceCode" id="cb176"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb176-1"><a href="data-modeling.html#cb176-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(gene_exp, <span class="fu">aes</span>(<span class="at">x=</span><span class="st">`</span><span class="at">Disease Status</span><span class="st">`</span>, <span class="at">y=</span><span class="st">`</span><span class="at">Gene 3</span><span class="st">`</span>, <span class="at">color=</span><span class="st">`</span><span class="at">Disease Status</span><span class="st">`</span>)) <span class="sc">+</span></span>
<span id="cb176-2"><a href="data-modeling.html#cb176-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_beeswarm</span>() <span class="sc">+</span></span>
<span id="cb176-3"><a href="data-modeling.html#cb176-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">annotate</span>(<span class="st">&quot;segment&quot;</span>, <span class="at">x=</span><span class="dv">0</span>, <span class="at">xend=</span><span class="dv">3</span>, <span class="at">y=</span><span class="dv">2</span><span class="sc">*</span>c_mean<span class="sc">-</span>pd_mean, <span class="at">yend=</span><span class="dv">2</span><span class="sc">*</span>pd_mean<span class="sc">-</span>c_mean)</span></code></pre></div>
<p><img src="main_files/lin%20model%20gene%203%20fit-1.png" width="672" /></p>
<p>However, this point estimate tells us nothing about how confident we are about
the difference. We can do this by using an <a href="https://en.wikipedia.org/wiki/Least_squares">linear
regression</a> by modeling Gene 3 as a
function of disease status:</p>
<div class="sourceCode" id="cb177"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb177-1"><a href="data-modeling.html#cb177-1" aria-hidden="true" tabindex="-1"></a>fit <span class="ot">&lt;-</span> <span class="fu">lm</span>(<span class="st">`</span><span class="at">Gene 3</span><span class="st">`</span> <span class="sc">~</span> <span class="st">`</span><span class="at">Disease Status</span><span class="st">`</span>, <span class="at">data=</span>gene_exp)</span>
<span id="cb177-2"><a href="data-modeling.html#cb177-2" aria-hidden="true" tabindex="-1"></a>fit</span></code></pre></div>
<pre><code>## 
## Call:
## lm(formula = `Gene 3` ~ `Disease Status`, data = gene_exp)
## 
## Coefficients:
##                 (Intercept)  `Disease Status`Parkinson&#39;s  
##                       334.6                        164.1</code></pre>
<p>The coefficient associated with having the disease status of Parkinson’s disease
is almost exactly equal to our difference in means. We also note that the
coefficient labeled <code>(Intercept)</code> is nearly equal to the mean of our control
samples (334.6). Under the hood, this simple linear model did the
same calculation we did by subtracting the means of each group but estimated the
means using all the data at once, instead of point estimates. Another advantage
of using <code>lm()</code> over the point estimate method is the model can estimate how
confident the model was that the difference in mean between Parkinson’s and
controls subjects. Let’s print some more information about the model than
before:</p>
<div class="sourceCode" id="cb179"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb179-1"><a href="data-modeling.html#cb179-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(fit)</span></code></pre></div>
<pre><code>## 
## Call:
## lm(formula = `Gene 3` ~ `Disease Status`, data = gene_exp)
## 
## Residuals:
##      Min       1Q   Median       3Q      Max 
## -124.303  -25.758   -2.434   30.518  119.348 
## 
## Coefficients:
##                             Estimate Std. Error t value Pr(&gt;|t|)    
## (Intercept)                  334.578      4.414   75.80   &lt;2e-16 ***
## `Disease Status`Parkinson&#39;s  164.094      6.243   26.29   &lt;2e-16 ***
## ---
## Signif. codes:  0 &#39;***&#39; 0.001 &#39;**&#39; 0.01 &#39;*&#39; 0.05 &#39;.&#39; 0.1 &#39; &#39; 1
## 
## Residual standard error: 44.14 on 198 degrees of freedom
## Multiple R-squared:  0.7773, Adjusted R-squared:  0.7761 
## F-statistic:   691 on 1 and 198 DF,  p-value: &lt; 2.2e-16</code></pre>
<p>We again see our coefficient estimates for the intercept (i.e. mean control
expression) and our increase in Parkinson’s, but also a number of other terms,
in particular <code>Pr(&gt;|t|)</code> and <code>Multiple R-squared</code>. The former is the
<a href="statistical-tests.html#p-values">p-value</a> associated with each of the coefficient estimates, both of
which are very small, indicating to us that the model was very confident of the
estimated values. The latter, multiple R-squared or <span class="math inline">\(R^2\)</span>, describes how much of
the variance in the data was explained by the model it found as a fraction
between 0 and 1. This model explains 77.7% of the variance of the data, which is
substantial. The <span class="math inline">\(R^2\)</span> value also has an associated p-value, which is also very
small. Overall, these statistics suggest this model fits the data very well.</p>
<p>We can plot the results of a linear model for each of our genes relatively
easily:</p>
<div class="sourceCode" id="cb181"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb181-1"><a href="data-modeling.html#cb181-1" aria-hidden="true" tabindex="-1"></a>pd_mean <span class="ot">&lt;-</span> <span class="fu">mean</span>(<span class="fu">filter</span>(gene_exp,<span class="st">`</span><span class="at">Disease Status</span><span class="st">`</span><span class="sc">==</span><span class="st">&quot;Parkinson&#39;s&quot;</span>)<span class="sc">$</span><span class="st">`</span><span class="at">Gene 1</span><span class="st">`</span>)</span>
<span id="cb181-2"><a href="data-modeling.html#cb181-2" aria-hidden="true" tabindex="-1"></a>c_mean <span class="ot">&lt;-</span> <span class="fu">mean</span>(<span class="fu">filter</span>(gene_exp,<span class="st">`</span><span class="at">Disease Status</span><span class="st">`</span><span class="sc">==</span><span class="st">&quot;Control&quot;</span>)<span class="sc">$</span><span class="st">`</span><span class="at">Gene 1</span><span class="st">`</span>)</span>
<span id="cb181-3"><a href="data-modeling.html#cb181-3" aria-hidden="true" tabindex="-1"></a>g1 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(gene_exp, <span class="fu">aes</span>(<span class="at">x=</span><span class="st">`</span><span class="at">Disease Status</span><span class="st">`</span>, <span class="at">y=</span><span class="st">`</span><span class="at">Gene 1</span><span class="st">`</span>, <span class="at">color=</span><span class="st">`</span><span class="at">Disease Status</span><span class="st">`</span>)) <span class="sc">+</span></span>
<span id="cb181-4"><a href="data-modeling.html#cb181-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_beeswarm</span>() <span class="sc">+</span></span>
<span id="cb181-5"><a href="data-modeling.html#cb181-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">annotate</span>(<span class="st">&quot;segment&quot;</span>, <span class="at">x=</span><span class="dv">0</span>, <span class="at">xend=</span><span class="dv">3</span>, <span class="at">y=</span><span class="dv">2</span><span class="sc">*</span>c_mean<span class="sc">-</span>pd_mean, <span class="at">yend=</span><span class="dv">2</span><span class="sc">*</span>pd_mean<span class="sc">-</span>c_mean) <span class="sc">+</span></span>
<span id="cb181-6"><a href="data-modeling.html#cb181-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position=</span><span class="st">&quot;none&quot;</span>)</span>
<span id="cb181-7"><a href="data-modeling.html#cb181-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb181-8"><a href="data-modeling.html#cb181-8" aria-hidden="true" tabindex="-1"></a>pd_mean <span class="ot">&lt;-</span> <span class="fu">mean</span>(<span class="fu">filter</span>(gene_exp,<span class="st">`</span><span class="at">Disease Status</span><span class="st">`</span><span class="sc">==</span><span class="st">&quot;Parkinson&#39;s&quot;</span>)<span class="sc">$</span><span class="st">`</span><span class="at">Gene 2</span><span class="st">`</span>)</span>
<span id="cb181-9"><a href="data-modeling.html#cb181-9" aria-hidden="true" tabindex="-1"></a>c_mean <span class="ot">&lt;-</span> <span class="fu">mean</span>(<span class="fu">filter</span>(gene_exp,<span class="st">`</span><span class="at">Disease Status</span><span class="st">`</span><span class="sc">==</span><span class="st">&quot;Control&quot;</span>)<span class="sc">$</span><span class="st">`</span><span class="at">Gene 2</span><span class="st">`</span>)</span>
<span id="cb181-10"><a href="data-modeling.html#cb181-10" aria-hidden="true" tabindex="-1"></a>g2 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(gene_exp, <span class="fu">aes</span>(<span class="at">x=</span><span class="st">`</span><span class="at">Disease Status</span><span class="st">`</span>, <span class="at">y=</span><span class="st">`</span><span class="at">Gene 2</span><span class="st">`</span>, <span class="at">color=</span><span class="st">`</span><span class="at">Disease Status</span><span class="st">`</span>)) <span class="sc">+</span></span>
<span id="cb181-11"><a href="data-modeling.html#cb181-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_beeswarm</span>() <span class="sc">+</span></span>
<span id="cb181-12"><a href="data-modeling.html#cb181-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">annotate</span>(<span class="st">&quot;segment&quot;</span>, <span class="at">x=</span><span class="dv">0</span>, <span class="at">xend=</span><span class="dv">3</span>, <span class="at">y=</span><span class="dv">2</span><span class="sc">*</span>c_mean<span class="sc">-</span>pd_mean, <span class="at">yend=</span><span class="dv">2</span><span class="sc">*</span>pd_mean<span class="sc">-</span>c_mean) <span class="sc">+</span></span>
<span id="cb181-13"><a href="data-modeling.html#cb181-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position=</span><span class="st">&quot;none&quot;</span>)</span>
<span id="cb181-14"><a href="data-modeling.html#cb181-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb181-15"><a href="data-modeling.html#cb181-15" aria-hidden="true" tabindex="-1"></a>pd_mean <span class="ot">&lt;-</span> <span class="fu">mean</span>(<span class="fu">filter</span>(gene_exp,<span class="st">`</span><span class="at">Disease Status</span><span class="st">`</span><span class="sc">==</span><span class="st">&quot;Parkinson&#39;s&quot;</span>)<span class="sc">$</span><span class="st">`</span><span class="at">Gene 3</span><span class="st">`</span>)</span>
<span id="cb181-16"><a href="data-modeling.html#cb181-16" aria-hidden="true" tabindex="-1"></a>c_mean <span class="ot">&lt;-</span> <span class="fu">mean</span>(<span class="fu">filter</span>(gene_exp,<span class="st">`</span><span class="at">Disease Status</span><span class="st">`</span><span class="sc">==</span><span class="st">&quot;Control&quot;</span>)<span class="sc">$</span><span class="st">`</span><span class="at">Gene 3</span><span class="st">`</span>)</span>
<span id="cb181-17"><a href="data-modeling.html#cb181-17" aria-hidden="true" tabindex="-1"></a>g3 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(gene_exp, <span class="fu">aes</span>(<span class="at">x=</span><span class="st">`</span><span class="at">Disease Status</span><span class="st">`</span>, <span class="at">y=</span><span class="st">`</span><span class="at">Gene 3</span><span class="st">`</span>, <span class="at">color=</span><span class="st">`</span><span class="at">Disease Status</span><span class="st">`</span>)) <span class="sc">+</span></span>
<span id="cb181-18"><a href="data-modeling.html#cb181-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_beeswarm</span>() <span class="sc">+</span></span>
<span id="cb181-19"><a href="data-modeling.html#cb181-19" aria-hidden="true" tabindex="-1"></a>  <span class="fu">annotate</span>(<span class="st">&quot;segment&quot;</span>, <span class="at">x=</span><span class="dv">0</span>, <span class="at">xend=</span><span class="dv">3</span>, <span class="at">y=</span><span class="dv">2</span><span class="sc">*</span>c_mean<span class="sc">-</span>pd_mean, <span class="at">yend=</span><span class="dv">2</span><span class="sc">*</span>pd_mean<span class="sc">-</span>c_mean) <span class="sc">+</span></span>
<span id="cb181-20"><a href="data-modeling.html#cb181-20" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position=</span><span class="st">&quot;none&quot;</span>)</span>
<span id="cb181-21"><a href="data-modeling.html#cb181-21" aria-hidden="true" tabindex="-1"></a>g1 <span class="sc">|</span> g2 <span class="sc">|</span> g3</span></code></pre></div>
<p><img src="main_files/lin%20model%203%20genes-1.png" width="672" /></p>
<p>We can also compute the corresponding linear model fits, and confirm that the
coefficients agree with the directions observed in the plot, as well as that all
the associations are significant FDR &lt; 0.05:</p>
<div class="sourceCode" id="cb182"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb182-1"><a href="data-modeling.html#cb182-1" aria-hidden="true" tabindex="-1"></a>fit1 <span class="ot">&lt;-</span> <span class="fu">lm</span>(<span class="st">`</span><span class="at">Gene 1</span><span class="st">`</span> <span class="sc">~</span> <span class="st">`</span><span class="at">Disease Status</span><span class="st">`</span>, <span class="at">data=</span>gene_exp)</span>
<span id="cb182-2"><a href="data-modeling.html#cb182-2" aria-hidden="true" tabindex="-1"></a>fit2 <span class="ot">&lt;-</span> <span class="fu">lm</span>(<span class="st">`</span><span class="at">Gene 2</span><span class="st">`</span> <span class="sc">~</span> <span class="st">`</span><span class="at">Disease Status</span><span class="st">`</span>, <span class="at">data=</span>gene_exp)</span>
<span id="cb182-3"><a href="data-modeling.html#cb182-3" aria-hidden="true" tabindex="-1"></a>fit3 <span class="ot">&lt;-</span> <span class="fu">lm</span>(<span class="st">`</span><span class="at">Gene 3</span><span class="st">`</span> <span class="sc">~</span> <span class="st">`</span><span class="at">Disease Status</span><span class="st">`</span>, <span class="at">data=</span>gene_exp)</span>
<span id="cb182-4"><a href="data-modeling.html#cb182-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb182-5"><a href="data-modeling.html#cb182-5" aria-hidden="true" tabindex="-1"></a>gene_stats <span class="ot">&lt;-</span> <span class="fu">bind_rows</span>(</span>
<span id="cb182-6"><a href="data-modeling.html#cb182-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">c</span>(<span class="st">&quot;Gene 1&quot;</span>,<span class="fu">coefficients</span>(fit1),<span class="fu">summary</span>(fit1)<span class="sc">$</span>coefficients[<span class="dv">2</span>,<span class="dv">4</span>]),</span>
<span id="cb182-7"><a href="data-modeling.html#cb182-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">c</span>(<span class="st">&quot;Gene 2&quot;</span>,<span class="fu">coefficients</span>(fit2),<span class="fu">summary</span>(fit2)<span class="sc">$</span>coefficients[<span class="dv">2</span>,<span class="dv">4</span>]),</span>
<span id="cb182-8"><a href="data-modeling.html#cb182-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">c</span>(<span class="st">&quot;Gene 3&quot;</span>,<span class="fu">coefficients</span>(fit3),<span class="fu">summary</span>(fit3)<span class="sc">$</span>coefficients[<span class="dv">2</span>,<span class="dv">4</span>])</span>
<span id="cb182-9"><a href="data-modeling.html#cb182-9" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb182-10"><a href="data-modeling.html#cb182-10" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(gene_stats) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;Gene&quot;</span>,<span class="st">&quot;Intercept&quot;</span>,<span class="st">&quot;Parkinson&#39;s&quot;</span>,<span class="st">&quot;pvalue&quot;</span>)</span>
<span id="cb182-11"><a href="data-modeling.html#cb182-11" aria-hidden="true" tabindex="-1"></a>gene_stats<span class="sc">$</span>padj <span class="ot">&lt;-</span> <span class="fu">p.adjust</span>(gene_stats<span class="sc">$</span>pvalue,<span class="at">method=</span><span class="st">&quot;fdr&quot;</span>)</span>
<span id="cb182-12"><a href="data-modeling.html#cb182-12" aria-hidden="true" tabindex="-1"></a>gene_stats</span></code></pre></div>
<pre><code>## # A tibble: 3 x 5
##   Gene   Intercept        `Parkinson&#39;s`     pvalue                   padj
##   &lt;chr&gt;  &lt;chr&gt;            &lt;chr&gt;             &lt;chr&gt;                   &lt;dbl&gt;
## 1 Gene 1 250.410735540151 6.95982194659045  3.92131170913765e-06 3.92e- 6
## 2 Gene 2 597.763814010886 -93.6291659250254 2.37368730897756e-13 3.56e-13
## 3 Gene 3 334.57788193953  164.094204856583  1.72774902843408e-66 5.18e-66</code></pre>
<p>We have just performed our first elementary differential expression analysis
using a linear model. Specifically, we examined each of our genes for a
statistical relationship with having Parkinson’s disease. The mechanics of this
analysis are beyond the scope of this book, but later when we consider
differential expression analysis packages in <a href="biology-bioinformatics.html#biology-bioinformatics">chapter
7</a> this pattern should be familiar after this example.</p>
<div class="box note">
<p>If you are familiar with <a href="https://en.wikipedia.org/wiki/Logistic_regression">logistic
regression</a>, you might have
wondered why we didn’t model disease status, which is a binary variable, as a
function of gene expression like <code>Disease Status ~ Gene 3</code>. There are several
reasons for this, <a href="https://en.wikipedia.org/wiki/Separation_(statistics)">complete
separation</a> principally
among them. In logistic regression, complete separation occurs when all the
predictor values (e.g. gene expression) for one outcome group are greater or
smaller than the other; i.e. there is no overlap in the values between groups.
This causes logistic regression to fail to converge, leaving these genes with no
statistics even though these genes are potentially the most interesting! There
are methods (e.g. <a href="https://cran.r-project.org/web/packages/logistf/index.html">Firth’s Logistic
Regression</a>) that
overcome this problem, but methods that model gene expression as a function of
other outcome variables were developed first and remain the most popular.</p>
</div>
<div class="box readmore">
<p><a href="https://r4ds.had.co.nz/model-basics.html">Modeling Basics - R for Data Science</a>
<a href="https://en.wikipedia.org/wiki/Least_squares">Least squares regression</a></p>
</div>
</div>
<div id="flavors-of-linear-models" class="section level3" number="6.1.4">
<h3><span class="header-section-number">6.1.4</span> Flavors of Linear Models</h3>
<p>The linear model implemented above is termed <a href="https://en.wikipedia.org/wiki/Linear_regression">linear
regression</a> due to the way it
models the relationship between the predictor variables and the outcome.
Specifically, linear regression makes some strong assumptions about that
relationship that may not always hold for all datasets. To address this
limitation, a more flexible class of linear models called <a href="https://en.wikipedia.org/wiki/Generalized_linear_model">generalized linear
models</a> that allow these
assumptions to be relaxed by changing the mathematical relationship between the
predictors and outcome using a <a href="https://en.wikipedia.org/wiki/Generalized_linear_model#Link_function">link
function</a>,
and/or by mathematically transforming the predictors themselves. Some of the
more common generalized linear models are listed below:</p>
<ul>
<li><strong>Logistic regression</strong> - models a binary outcome (e.g. disease vs control) as
a linear combination of predictor variables by using the <a href="https://en.wikipedia.org/wiki/Logistic_function">logistic
function</a> as the link function.</li>
<li><strong>Multinomial regression</strong> - models a multinomial (i.e. categorical variable
with more than two categories) using a <a href="https://en.wikipedia.org/wiki/Multinomial_logistic_regression">multinomial logit function</a> link
function</li>
<li><strong>Poisson regression</strong> - models an outcome variable that are count data as
a linear combination of predictors using the logarithm as the link function;
this model is commonly used when modeling certain types of high throughput
sequencing data</li>
<li><strong>Negative binomial regression</strong> - also models an outcome variable that are
count data but relaxes some of the assumptions of Poisson regression, namely the
mean-variance equality constraint; negative binomial regression is commonly used
to model gene expression estimated from RNASeq data.</li>
</ul>
<p>Generalized linear models are very flexible, and there are many other types of
these models used in biology and data science. It is important to be aware of
the characteristics of the outcome you wish to model and choose the modeling
method that is most suitable.</p>
<div class="box readmore">
<p><a href="https://en.wikipedia.org/wiki/Generalized_linear_model">Generalized linear models</a></p>
</div>
</div>
<div id="assessing-model-accuracy-." class="section level3" number="6.1.5">
<h3><span class="header-section-number">6.1.5</span> Assessing Model Accuracy .</h3>

</div>
</div>
<h3>References</h3>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-Box1976-rd" class="csl-entry">
Box, George E P. 1976. <span>“Science and Statistics.”</span> <em>J. Am. Stat. Assoc.</em> 71 (356): 791–99.
</div>
</div>
<p style="text-align: center;">
<a href="data-science.html"><button class="btn btn-default">Previous</button></a>
<a href="statistical-distributions.html"><button class="btn btn-default">Next</button></a>
</p>
</div>
</div>


</div>

<script>

// add bootstrap table styles to pandoc tables
$(document).ready(function () {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
});

</script>

</body>
</html>
