<!DOCTYPE html>
<html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>7 Data Visualization | BF591 - R for Biological Sciences</title>
<meta name="generator" content="bookdown 0.32 with bs4_book()">
<meta property="og:title" content="7 Data Visualization | BF591 - R for Biological Sciences">
<meta property="og:type" content="book">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="7 Data Visualization | BF591 - R for Biological Sciences">
<!-- JS --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://kit.fontawesome.com/6ecbd6c532.js" crossorigin="anonymous"></script><script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="libs/bootstrap-4.6.0/bootstrap.min.css" rel="stylesheet">
<script src="libs/bootstrap-4.6.0/bootstrap.bundle.min.js"></script><script src="libs/bs3compat-0.4.2/transition.js"></script><script src="libs/bs3compat-0.4.2/tabs.js"></script><script src="libs/bs3compat-0.4.2/bs3compat.js"></script><link href="libs/bs4_book-1.0.0/bs4_book.css" rel="stylesheet">
<script src="libs/bs4_book-1.0.0/bs4_book.js"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- CSS --><style type="text/css">
    
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  </style>
<style type="text/css">
    /* Used with Pandoc 2.11+ new --citeproc when CSL is used */
    div.csl-bib-body { }
    div.csl-entry {
      clear: both;
        }
    .hanging div.csl-entry {
      margin-left:2em;
      text-indent:-2em;
    }
    div.csl-left-margin {
      min-width:2em;
      float:left;
    }
    div.csl-right-inline {
      margin-left:2em;
      padding-left:1em;
    }
    div.csl-indent {
      margin-left: 2em;
    }
  </style>
<link rel="stylesheet" href="style.css">
<meta name="description" content="Data visualization is a core component of both exploring data and communicating results to others. The goal of data visualization is to present data in a graphical way that shows the reader...">
<meta property="og:description" content="Data visualization is a core component of both exploring data and communicating results to others. The goal of data visualization is to present data in a graphical way that shows the reader...">
<meta name="twitter:description" content="Data visualization is a core component of both exploring data and communicating results to others. The goal of data visualization is to present data in a graphical way that shows the reader...">
</head>
<body data-spy="scroll" data-target="#toc">

<div class="container-fluid">
<div class="row">
  <header class="col-sm-12 col-lg-3 sidebar sidebar-book"><a class="sr-only sr-only-focusable" href="#content">Skip to main content</a>

    <div class="d-flex align-items-start justify-content-between">
      <h1>
        <a href="index.html" title="">BF591 - R for Biological Sciences</a>
      </h1>
      <button class="btn btn-outline-primary d-lg-none ml-2 mt-1" type="button" data-toggle="collapse" data-target="#main-nav" aria-expanded="true" aria-controls="main-nav"><i class="fas fa-bars"></i><span class="sr-only">Show table of contents</span></button>
    </div>

    <div id="main-nav" class="collapse-lg">
      <form role="search">
        <input id="search" class="form-control" type="search" placeholder="Search" aria-label="Search">
</form>

      <nav aria-label="Table of contents"><h2>Table of contents</h2>
        <ul class="book-toc list-unstyled">
<li><a class="" href="index.html">R for Biological Sciences</a></li>
<li><a class="" href="introduction.html"><span class="header-section-number">1</span> Introduction</a></li>
<li><a class="" href="data-bio.html"><span class="header-section-number">2</span> Data in Biology</a></li>
<li><a class="" href="preliminaries.html"><span class="header-section-number">3</span> Preliminaries</a></li>
<li><a class="" href="prog-basics.html"><span class="header-section-number">4</span> R Programming</a></li>
<li><a class="" href="data-wrangling.html"><span class="header-section-number">5</span> Data Wrangling</a></li>
<li><a class="" href="data-science.html"><span class="header-section-number">6</span> Data Science</a></li>
<li><a class="active" href="data-visualization.html"><span class="header-section-number">7</span> Data Visualization</a></li>
<li><a class="" href="biology-bioinformatics.html"><span class="header-section-number">8</span> Biology &amp; Bioinformatics</a></li>
<li><a class="" href="engineering.html"><span class="header-section-number">9</span> EngineeRing</a></li>
<li><a class="" href="rshiny.html"><span class="header-section-number">10</span> RShiny</a></li>
<li><a class="" href="communicating-with-r.html"><span class="header-section-number">11</span> Communicating with R</a></li>
<li><a class="" href="contribution-guide.html"><span class="header-section-number">12</span> Contribution Guide</a></li>
<li class="book-part">Assignments</li>
<li><a class="" href="starting-an-assignment.html">Starting an Assignment</a></li>
<li><a class="" href="assignment-overview.html">Assignment Overview</a></li>
<li><a class="" href="assignment-0.html">Assignment 0</a></li>
<li><a class="" href="assignment-1.html">Assignment 1</a></li>
<li><a class="" href="assignment-2.html">Assignment 2</a></li>
<li><a class="" href="assignment-3.html">Assignment 3</a></li>
<li><a class="" href="assignment-4.html">Assignment 4</a></li>
<li><a class="" href="assignment-5.html">Assignment 5</a></li>
<li><a class="" href="assignment-6.html">Assignment 6</a></li>
<li><a class="" href="assignment-7.html">Assignment 7</a></li>
<li><a class="" href="final-project.html">Final Project</a></li>
<li class="book-part">Appendix</li>
<li><a class="" href="detailed-class-schedule.html"><span class="header-section-number">A</span> Detailed Class Schedule</a></li>
</ul>

        <div class="book-extra">
          
        </div>
      </nav>
</div>
  </header><main class="col-sm-12 col-md-9 col-lg-7" id="content"><div id="data-visualization" class="section level1" number="7">
<h1>
<span class="header-section-number">7</span> Data Visualization<a class="anchor" aria-label="anchor" href="#data-visualization"><i class="fas fa-link"></i></a>
</h1>
<p>Data visualization is a core component of both exploring data and communicating
results to others. The goal of data visualization is to present data in a
graphical way that shows the reader patterns that would not otherwise be
visible. Despite its ubiquity and importance, effective data visualization is
challenging, and while many tools and approaches exist there is no “gold
standard” to follow in any meaningful sense. Rather, an effective visualization
has the following properties:</p>
<ol style="list-style-type: decimal">
<li>Depicts accurate data</li>
<li>Depicts data accurately</li>
<li>Shows enough, but not too much, of the data for the viewer to gain insight</li>
<li>Is self contained - no additional information (except a caption) is required
to understand the contents of the figure</li>
</ol>
<p>Beyond these, a <em>great</em> visualization has some additional properties:</p>
<ol start="5" style="list-style-type: decimal">
<li>Exposes patterns in the data not easily observable by other methods</li>
<li>Invites the viewer to ask more questions about the data</li>
</ol>
<p>In this chapter, we explore some of the principles of visualizing data and
present many examples of plots and data visualization strategies commonly used
in biological data analysis and bioinformatics.</p>

<div id="responsible-plotting" class="section level2" number="7.1">
<h2>
<span class="header-section-number">7.1</span> Responsible Plotting<a class="anchor" aria-label="anchor" href="#responsible-plotting"><i class="fas fa-link"></i></a>
</h2>
<p>“Good plots empower us to ask good questions.” - Alberto Cairo, How Charts Lie</p>
<p>Plots are the primary vehicles of communicating scientific results. A very
common and useful approach to writing scientific papers is to decide how many
figures are needed to tell the story, and then create plots as necessary to
convey that story. The text plays a supporting role for the figures. Effective
data visualization has immense power to communicate the ideas (and beliefs) of
the chart creator. Consider the famous “hockey stick graph” depicting average
global temperatures created by Jerry Mahlman to describe the pattern of global
warming shown by Mann, Bradkely &amp; Hughes in 1999:</p>
<div class="float">
<img src="hockey_stick_chart.jpg" alt="“Hockey stick graph” by Jerry Mahlman, showing the trend of temperature rise over the past 1000 years in the northern hemisphere"><div class="figcaption">“Hockey stick graph” by Jerry Mahlman, showing the trend of temperature rise over the past 1000 years in the northern hemisphere</div>
</div>
<p>The data and models underlying the graph are very complex, but the story is
clear: the past 100 years has seen a dramatic increase in average temperature in
the northern hemisphere. This chart was one of the main drivers behind renewed
main stream awareness of climate change, in no small part because it is easy to
understand.</p>
<p>Given the power visualization has, and the increasing amount of data created by
and required for daily life, it is ever more important that data visualization
designers understand the principles and pitfalls of effective plotting. There is
no easy solution to accomplishing this, but Alberto Cairo and Enrico Bertini,
two journalists who specialize in data visualization and infographics, lay out
five qualities of an effective visualization, in Cairo’s book <a href="http://www.thefunctionalart.com/p/the-truthful-art-book.html">the truthful
art</a>:</p>
<ol style="list-style-type: decimal">
<li>
<strong>It is truthful</strong> - the message conveyed by the visualization expresses some
truth of the underlying data (to the extent that anything is “true”)</li>
<li>
<strong>It is functional</strong> - the visualization is accurate, and allows the reader
to do something useful with the insights they gain</li>
<li>
<strong>It is insightful</strong> - the visualization should reveal patterns in the data
that would otherwise be difficult or impossible to see and understand</li>
<li>
<strong>It is enlightening</strong> - the reader should leave the experience of engaging
with the visualization convinced of something, ideally what the designer intends</li>
<li>
<strong>It is beautiful</strong> - it is visually appealing to its target audience, which
helps the user more easily spend time and engage with the visualization</li>
</ol>
<p>A full treatment of these ideas are included in, among others, three excellent
books by Alberto Cairo, linked at the end of this section. The sections that
follow describe some of the ideas and concepts that make visualization
challenging and offer some guidance from the books mentioned on how to produce
effective visualizations.</p>
<div class="box readmore">
<p>All three of these books were written by Alberto Cairo:</p>
<ul>
<li><a href="http://www.thefunctionalart.com/p/reviews.html">How Charts Lie</a></li>
<li><a href="http://www.thefunctionalart.com/p/about-book.html">the funcitonal art</a></li>
<li><a href="http://www.thefunctionalart.com/p/the-truthful-art-book.html">the truthful art</a></li>
</ul>
</div>
<div id="visualization-principles" class="section level3" number="7.1.1">
<h3>
<span class="header-section-number">7.1.1</span> Visualization Principles<a class="anchor" aria-label="anchor" href="#visualization-principles"><i class="fas fa-link"></i></a>
</h3>
<p>The purpose of data visualization is to illustrate (from the latin <em>illustrare</em>,
meaning to light up or illuminate) relationships between quantities. Effective
data visualization takes advantage of our highly evolved <a href="https://en.wikipedia.org/wiki/Visual_perception">visual perception
system</a> to communicate these
relationships to our cognitive and decision making systems. Therefore, our
visualization strategies should take knowledge of our perceptual biases into
account when deciding how to present data. Poor data visualization choices may
make interpretation difficult, or worse lead us to incorrect inferences and
conclusions.</p>
</div>
<div id="human-visual-perception" class="section level3" number="7.1.2">
<h3>
<span class="header-section-number">7.1.2</span> Human Visual Perception<a class="anchor" aria-label="anchor" href="#human-visual-perception"><i class="fas fa-link"></i></a>
</h3>
<p>The human visual perceptual system is intrinsically integrated with our
cognitive system. This makes it a <em>predictive</em> or <em>interpretive</em> system, where
patterns of light received by the eye are pre-processed before the relatively
slower executive function parts of our brains can make judgements.
If we compare our perception of the world around us to a cinematic film, this
makes our visual perception less raw video footage and more like a director’s
commentary, where we add our interpretations, motivations, experiences, and
anecdotal annotations to the images before we cognitively interpret them. Said
differently, our visuo-cognitive system performs <em>pattern recognition</em> on the
information we receive, and annotates the information with the patterns it
thinks it sees. Very often, our cognition predicts patterns that are accurate,
which therefore may be useful for us in making decisions, e.g. how to move out
of the trajectory of a falling tree branch. However, sometimes our pattern
recognition system, which is always <em>trying</em> to match patterns, sees patterns
that don’t exist. Some of the most well known and well studied examples of these
perceptual errors are <em>optical illusions.</em></p>
<p>Optical illusions are images or animations that reveal the errors and bias of
our perceptual system. Some illusions that are static images appear to move.
Others distort the true relations between lines or shapes, or lead to incorrect
or inconsistent perception of color. Consider the famous <a href="https://en.wikipedia.org/wiki/Checker_shadow_illusion">checkerboard shadow
illusion</a> created by
Edward H. Adelson in 1995:</p>
<div class="float">
<img src="checker_illusion.png" alt="Checkerboard shadow illusion, the squares marked A and B are the same color"><div class="figcaption">Checkerboard shadow illusion, the squares marked A and B are the same color</div>
</div>
<p>The image on the right clearly shows the two squares A and B are the same shade
of grey, though A appears darker than B in the left image. This is because our
perceptual system identifies color by <em>contrasting</em> adjacent colors. Since the
colors adjacent to the A and B squares differ, our perception of the colors also
differ. This example illustrates the contrast bias according to the <em>value</em>, or
lightness, of the color, but the principle also applies to different hues:</p>
<div class="float">
<img src="gradient_illusion.png" alt="The three squares are drawn on top of and below the image and are the same color"><div class="figcaption">The three squares are drawn on top of and below the image and are the same color</div>
</div>
<p>An illusion that shows how our perceptual system distorts relationships between
geometry is as follows:</p>
<div class="float">
<img src="wavy_grid.png" alt="The horizontal and vertical pattern appears to be “wavy” but in fact form straight lines"><div class="figcaption">The horizontal and vertical pattern appears to be “wavy” but in fact form straight lines</div>
</div>
<p>The pattern of light and dark squares intersecting with the grid makes the lines
appear “wavy” when all the horizontal and vertical lines are in fact straight,
parallel, and perpendicular.</p>
<p>A final example illusion that is relevant for data visualization demonstrates
a similar contextual bias of relative area:</p>
<div class="float">
<img src="area_illusion.png" alt="The blue dots have the same diameter"><div class="figcaption">The blue dots have the same diameter</div>
</div>
<p>The two blue dots in the image have the same radius, although the dot within
the large black circles appears smaller.</p>
<p>Not only are there illusions that distort our perception of shapes that are
there, but there are also illusions that convince us of the existence of shapes
that are <em>not</em> there. The following illusion illustrates the point:</p>
<div class="float">
<img src="negative_shape.png" alt="The shapes are drawn to suggest a triangle in the middle, though no triangle is actually drawn"><div class="figcaption">The shapes are drawn to suggest a triangle in the middle, though no triangle is actually drawn</div>
</div>
<p>Seeing shapes when they aren’t there is nothing new to humans, of course.</p>
<div class="float">
<img src="cloud_dog.png" alt="Clouds that resemble shapes - this author sees a running dog"><div class="figcaption">Clouds that resemble shapes - this author sees a running dog</div>
</div>
<p>Each of these perceptual biases relate to how we interpret many common data
visualization strategies. We cannot avoid making these interpretive errors via
our cognition alone, but we can use this knowledge in the design of our plots
to mitigate their effects. To quickly summarize the perceptual biases we have
demonstrated:</p>
<ol style="list-style-type: decimal">
<li>We perceive the value and hue of colors in contrast to adjacent colors</li>
<li>Certain types of geometry interfere with our ability to assess accurate
relationships between geometric shapes</li>
<li>We perceive the size (area) of shapes relative to nearby shapes</li>
<li>We may perceive shapes where there are none drawn, based on nearby shapes</li>
<li>We may perceive shapes that are similar to other familiar shapes, but have no
true relationship</li>
</ol>
<p>Each of these biases influence how we perceive and interpret plots, as we will
see in the rest of this chapter.</p>
</div>
<div id="visual-encodings" class="section level3" number="7.1.3">
<h3>
<span class="header-section-number">7.1.3</span> Visual Encodings<a class="anchor" aria-label="anchor" href="#visual-encodings"><i class="fas fa-link"></i></a>
</h3>
<p>Data visualization is the process of encoding numbers as positions, shapes, and
colors in a two- or sometimes three-dimensional space. There are many different
geometric properties we can use to create an encoding, including:</p>
<ol style="list-style-type: decimal">
<li>Length, width, or height</li>
<li>Position</li>
<li>Area</li>
<li>Angle/proportional area</li>
<li>Shape</li>
<li>Hue and value of colors</li>
</ol>
<p>Examples of each of these encodings is depicted for a fictitious dataset of
values between zero and one in the following figure:</p>
<div class="float">
<img src="viz_encodings.png" alt="Different encodings for the same data. This figure was inspired by Alberto Cairo in his book the truthful art."><div class="figcaption">Different encodings for the same data. This figure was inspired by Alberto Cairo in his book the truthful art.</div>
</div>
<p>The example encodings in the figure are among the most common and recognizable
elements used in scientific plotting. Every time a reader examines a plot, they
must <em>decode</em> it to a (hopefully) accurate mental model of the quantities that
underlie it. Using familiar encodings that a reader will already know how to
interpret makes this decoding cognitively easier, thereby enabling the plot to
be more quickly understood.</p>
<p>Every quantity we wish to plot must be mapped to a visualization using one or
more of these geometric encodings. By combining different variables with
different encodings, we construct more complex plots. Consider the scatter plot,
which contains two continuous values per datum, each of which is mapped to a
visualization with a position encoding:</p>
<div class="sourceCode" id="cb261"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html">tibble</a></span><span class="op">(</span></span>
<span>  x<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span><span class="op">(</span><span class="fl">100</span><span class="op">)</span>,</span>
<span>  y<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span><span class="op">(</span><span class="fl">100</span>,<span class="fl">0</span>,<span class="fl">5</span><span class="op">)</span></span>
<span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">x</span>,y<span class="op">=</span><span class="va">y</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="main_files/viz%20scatter-1.png" width="672"></div>
<p>We might layer on another encoding by introducing a new variable that is the
absolute value of the product of <span class="math inline">\(x\)</span> and <span class="math inline">\(y\)</span> and make the area of each datum
proportional to that value:</p>
<div class="sourceCode" id="cb262"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data</span> <span class="op">&lt;-</span>  <span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html">tibble</a></span><span class="op">(</span></span>
<span>  x<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span><span class="op">(</span><span class="fl">100</span><span class="op">)</span>,</span>
<span>  y<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span><span class="op">(</span><span class="fl">100</span>,<span class="fl">0</span>,<span class="fl">5</span><span class="op">)</span>,</span>
<span>  `x times y`<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html">abs</a></span><span class="op">(</span><span class="va">x</span><span class="op">*</span><span class="va">y</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">data</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">x</span>,y<span class="op">=</span><span class="va">y</span>,size<span class="op">=</span><span class="va">`x times y`</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="main_files/viz%20scatter%202-1.png" width="672"></div>
<p>We could layer yet more encodings to color each marker proportional to a value
between 0 and 1, and also change the shape based on a category:</p>
<div class="sourceCode" id="cb263"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span><span class="va">data</span>,</span>
<span>    z<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html">runif</a></span><span class="op">(</span><span class="fl">100</span><span class="op">)</span>,</span>
<span>    category<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'A'</span>,<span class="st">'B'</span>,<span class="st">'C'</span><span class="op">)</span>, <span class="fl">100</span>, replace<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span><span class="va">data</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">x</span>, y<span class="op">=</span><span class="va">y</span>, size<span class="op">=</span><span class="va">`x times y`</span>, color<span class="op">=</span><span class="va">z</span>, shape<span class="op">=</span><span class="va">category</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="main_files/viz%20scatter%203-1.png" width="672"></div>
<p>Let’s say the sum of <span class="math inline">\(x\)</span> and <span class="math inline">\(y\)</span> is meaningful as well, and we are interested in
the order of the data along this axis. We can connect all the points with lines
by using a length encoding between adjacent data points after sorting by <span class="math inline">\(x+y\)</span>:</p>
<div class="sourceCode" id="cb264"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span><span class="va">data</span>,</span>
<span>  `x + y`<span class="op">=</span><span class="va">x</span><span class="op">+</span><span class="va">y</span>,</span>
<span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html">arrange</a></span><span class="op">(</span><span class="va">`x + y`</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span></span>
<span>    xend<span class="op">=</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/lead-lag.html">lag</a></span><span class="op">(</span><span class="va">x</span>,<span class="fl">1</span><span class="op">)</span>,</span>
<span>    yend<span class="op">=</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/lead-lag.html">lag</a></span><span class="op">(</span><span class="va">y</span>,<span class="fl">1</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span><span class="va">data</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_segment.html">geom_segment</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">x</span>,xend<span class="op">=</span><span class="va">xend</span>,y<span class="op">=</span><span class="va">y</span>,yend<span class="op">=</span><span class="va">yend</span>,alpha<span class="op">=</span><span class="fl">0.5</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">x</span>, y<span class="op">=</span><span class="va">y</span>, size<span class="op">=</span><span class="va">`x times y`</span>, color<span class="op">=</span><span class="va">z</span>, shape<span class="op">=</span><span class="va">category</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="main_files/viz%20scatter%204-1.png" width="672"></div>
<p>The plot is becoming very busy and difficult to interpret because we have
encoded six different dimensions in the same plot:</p>
<ol style="list-style-type: decimal">
<li>
<span class="math inline">\(x\)</span> value has a positional encoding</li>
<li>
<span class="math inline">\(y\)</span> value has another positional encoding</li>
<li>The product <span class="math inline">\(xy\)</span> has an area encoding</li>
<li>
<span class="math inline">\(z\)</span> has a quantitative color value encoding</li>
<li>category has a categorical encoding to shape</li>
<li>The adjacency of coordinates along the <span class="math inline">\(x + y\)</span> axis with a length encoding</li>
</ol>
<p>Very complex plots may be built by thinking explicitly about how each dimension
of the data are encoded, but the designer (i.e. you) must carefully consider
whether the chosen encodings faithfully depict the relationships in the data.</p>
<p>These common encodings depicted in the earlier figure are not the only
possibilities for mapping visual properties to a set of values. Consider the
violin plot below:</p>
<div class="sourceCode" id="cb265"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">1337</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html">tibble</a></span><span class="op">(</span></span>
<span>  value<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span><span class="op">(</span><span class="fl">100</span><span class="op">)</span>,</span>
<span>  category<span class="op">=</span><span class="st">'A'</span></span>
<span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">category</span>,y<span class="op">=</span><span class="va">value</span>,fill<span class="op">=</span><span class="va">category</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_violin.html">geom_violin</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="main_files/viz%20violin-1.png" width="672"></div>
<p>Here, <code><a href="https://ggplot2.tidyverse.org/reference/geom_violin.html">geom_violin()</a></code> has preprocessed the data to map the number of values
across the full range of values into a density. The density is then encoded with
a width encoding, where the width of the violin across the range of values,
which is encoded with a position encoding.</p>
<p>Most common chart types are simply combinations of variables with different
encodings. The following table contains the encodings employed for different
types of plots:</p>
<div class="inline-table"><table style="width:99%;" class="table table-sm">
<colgroup>
<col width="20%">
<col width="10%">
<col width="27%">
<col width="39%">
</colgroup>
<thead><tr class="header">
<th>Plot</th>
<th>
<span class="math inline">\(x\)</span> encoding</th>
<th>
<span class="math inline">\(y\)</span> encoding</th>
<th>note</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><a href="scatter-plot">scatter</a></td>
<td>position</td>
<td>position</td>
<td></td>
</tr>
<tr class="even">
<td><a href="bar-chart">vertical bar</a></td>
<td>position</td>
<td>height</td>
<td></td>
</tr>
<tr class="odd">
<td><a href="bar-chart">horizontal bar</a></td>
<td>width</td>
<td>position</td>
<td></td>
</tr>
<tr class="even">
<td><a href="lollipop-plots">lollipop</a></td>
<td>position</td>
<td>height for line + position for “head”</td>
<td></td>
</tr>
<tr class="odd">
<td><a href="violin-plots">violin</a></td>
<td>position</td>
<td>width</td>
<td>
<span class="math inline">\(x\)</span> transformed to range, <span class="math inline">\(y\)</span> transformed to densities</td>
</tr>
</tbody>
</table></div>
<div id="elementary-perceptual-tasks" class="section level4" number="7.1.3.1">
<h4>
<span class="header-section-number">7.1.3.1</span> Elementary Perceptual Tasks<a class="anchor" aria-label="anchor" href="#elementary-perceptual-tasks"><i class="fas fa-link"></i></a>
</h4>
<p>The purpose of visualizations is to enable the reader to construct an accurate
mental model of the relationships or ideas contained within the data. As
discussed in the <a href="data-visualization.html#human-visual-perception">Human Visual Perception</a> section, not all visualizations allow
for easy and accurate judgments of this kind due to the biases of our biological
visuo-cognitive biases. However, some encodings of data allow <em>more accurate</em>
judgments than others. William Cleveland and Robert McGill, two statisticians,
proposed a hierarchy of visual encodings that span the spectrum of precision of
our ability to accurately interpret visual information <span class="citation">(<a href="detailed-class-schedule.html#ref-Cleveland1984-ay">Cleveland and McGill 1984</a>)</span>. They
defined the act of mapping a specific type of encoding to a cognitive model an
<em>elementary perceptual task.</em> The hierarchy is depicted in the following figure:</p>
<div class="float">
<img src="elementary_perceptual_tasks.png" alt="Elementary Perceptual Task Hierarchy - inspired by Alberto Cairo in the truthful art, which was inspired by Willian Cleveland and Robert McGill"><div class="figcaption">Elementary Perceptual Task Hierarchy - inspired by Alberto Cairo in the truthful art, which was inspired by Willian Cleveland and Robert McGill</div>
</div>
<p>The hierarchy is oriented from least to most cognitively difficult decoding
tasks, where the easier tasks lend themselves to easier comparisons and
judgements. For example, the easiest perceptual task is to compare visual
elements with a position encoding that are all aligned across all charts; the
positional decoding is the same for each chart, even if they use different
stylistic elements. Further down the hierarchy, more cognitive effort is
required to assess the precise relationships between elements of the
plot, therefore possibly reducing the accuracy of our judgements.</p>
<p>This is not to say that it is always better for the reader to make highly
accurate estimates. General trends and patterns may be more readily (and
beautifully) apparent with less accurate encodings. Consider the following
clustered heatmap:</p>
<div class="sourceCode" id="cb266"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># these data have two groups of samples with similar random data profiles</span></span>
<span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html">tibble</a></span><span class="op">(</span></span>
<span>    A<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span><span class="op">(</span><span class="fl">10</span><span class="op">)</span>,<span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span><span class="op">(</span><span class="fl">10</span>,<span class="fl">2</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    B<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span><span class="op">(</span><span class="fl">10</span><span class="op">)</span>,<span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span><span class="op">(</span><span class="fl">10</span>,<span class="fl">2</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    C<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span><span class="op">(</span><span class="fl">10</span><span class="op">)</span>,<span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span><span class="op">(</span><span class="fl">10</span>,<span class="fl">2</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    D<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span><span class="op">(</span><span class="fl">10</span>,<span class="fl">4</span><span class="op">)</span>,<span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span><span class="op">(</span><span class="fl">10</span>,<span class="op">-</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    E<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span><span class="op">(</span><span class="fl">10</span>,<span class="fl">4</span><span class="op">)</span>,<span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span><span class="op">(</span><span class="fl">10</span>,<span class="op">-</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    F<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span><span class="op">(</span><span class="fl">10</span>,<span class="fl">4</span><span class="op">)</span>,<span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span><span class="op">(</span><span class="fl">10</span>,<span class="op">-</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">'G'</span>,<span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/stats/heatmap.html">heatmap</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="main_files/heatmap%20perceptual%20task-1.png" width="672"></div>
<p>In this toy example, it may be sufficient for the reader’s comprehension to show
that samples A, B, and C cluster together, as do D, E, and F, without requiring
them to know <em>precisely</em> how different they are. In this case, the clustering
suggested by the visualization should be presented in some other way that is
more precise as well, for example a statistical comparison with results reported
in a table, to confirm what our eyes tell us.</p>
<p>It is usually a good idea to plot data multiple ways, to ensure the patterns we
see with one encoding are also seen with another. We might consider plotting
the data again using a <a href="parallel-coordinate-plots">parallel coordinates plot</a>:</p>
<div class="sourceCode" id="cb267"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ggobi.github.io/ggally/">GGally</a></span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html">as_tibble</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span></span>
<span>  group<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="st">'Group 1'</span>,<span class="fl">10</span><span class="op">)</span>,<span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="st">'Group 2'</span>,<span class="fl">10</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://ggobi.github.io/ggally/reference/ggparcoord.html">ggparcoord</a></span><span class="op">(</span>columns<span class="op">=</span><span class="fl">1</span><span class="op">:</span><span class="fl">6</span>,groupColumn<span class="op">=</span><span class="fl">7</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="main_files/viz%20element%20pcp-1.png" width="672"></div>
<p>For simplicity, we added a new variable <code>group</code> that is consistent with the
clustering we observed in the heatmap, which we might have extracted
computationally by cutting the row dendrogram into two flat clusters. The
parallel coordinate plot also visually confirms our interpretation of two
distinct clusters. We may now feel more comfortable including one or the other
plot in a figure, though this author finds the parallel coordinate plot more
accurate and enlightening. Recall the elementary perceptual task hierarchy:
aligned position encodings enable more accurate estimates of data. Heatmaps,
which use color value and hue, enable the least accurate estimates.</p>
</div>
</div>
<div id="some-opinionated-rules-of-thumb" class="section level3" number="7.1.4">
<h3>
<span class="header-section-number">7.1.4</span> Some Opinionated Rules of Thumb<a class="anchor" aria-label="anchor" href="#some-opinionated-rules-of-thumb"><i class="fas fa-link"></i></a>
</h3>
<p>Effective data visualization is challenging, and there are no simple solutions
or single approaches to doing it well. In this sense, data visualization is as
much an art as a science. This author has some rules of thumb he applies when
visualizing data:</p>
<ol style="list-style-type: decimal">
<li>
<strong>Visualize the same data in multiple ways.</strong> This will help ensure you
aren’t being unduly influenced by the quirks of certain encodings or inherent
visuo-cognitive biases.</li>
<li>
<strong>Perform precise statistical analysis to confirm findings identified when
inspecting plots.</strong> Examine the underlying data and devise statistical tests or
procedures to verify and quantify patterns you see.</li>
<li>
<strong>Informative is better than beautiful.</strong> Although it may be tempting to make
plots beautiful, this should not be at the expense of clarity. Make them clear
first, then make them beautiful.</li>
<li>
<strong>No plot is better than a useless plot.</strong> Some plots, while pretty, are
useless in a literal sense; the information contained within the plot is of no
use in either devising followup experiments or gaining insight into the data.</li>
<li>
<strong>Sometimes a table is the right way to present data.</strong> Tables may be
considered boring, but they are the most accurate representation of the data
(i.e. they ARE the data). Especially for relatively small datasets, a table may
do the job of communicating results better than any plot.</li>
<li>
<strong>If there is text on the plot, you should be able to read it.</strong> Many results
in biological data analysis have hundreds or thousands of individual points,
each of which has a label of some sort. If the font size for labels is too small
to read, or they substantially overlap each other making them illegible, either
the plot needs to change or the labels should not be drawn. Same goes for axis
labels.</li>
<li>
<strong>(Almost) every plot should have axis labels, tick marks and labels, a
legend, and a title.</strong> There are a few exceptions to this, but a plot should be
mostly understandable without any help from the text or caption. If it isn’t,
it’s probably a poor visualization strategy and can be improved.</li>
<li>
<strong>When using colors, be aware of color blindness.</strong> There are several known
forms of colorblindness in humans. Most default color schemes are “colorblind
friendly” but sometimes these color combinations can be counterinutitive. Check
sites like <a href="https://colorbrewer2.org/#type=sequential&amp;scheme=BuGn&amp;n=3">ColorBrewer</a>
to find color schemes that are likely to be colorblind friendly.</li>
<li>
<strong>Make differences visually appear as big as they mean.</strong> Scales on plots can
distort the meaning of the underlying data. For example, consider the following
two plots with identical data:</li>
</ol>
<div class="sourceCode" id="cb268"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com">patchwork</a></span><span class="op">)</span></span>
<span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html">tibble</a></span><span class="op">(</span></span>
<span>  percent<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">86</span>,<span class="fl">88</span>,<span class="fl">87</span>,<span class="fl">90</span>,<span class="fl">93</span>,<span class="fl">89</span><span class="op">)</span>,</span>
<span>  ID<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'A'</span>,<span class="st">'B'</span>,<span class="st">'C'</span>,<span class="st">'D'</span>,<span class="st">'E'</span>,<span class="st">'F'</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">g</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">data</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">ID</span>,y<span class="op">=</span><span class="va">percent</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html">geom_bar</a></span><span class="op">(</span>stat<span class="op">=</span><span class="st">"identity"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">g</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_cartesian.html">coord_cartesian</a></span><span class="op">(</span>ylim<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">85</span>,<span class="fl">95</span><span class="op">)</span><span class="op">)</span> <span class="op">|</span> <span class="va">g</span></span></code></pre></div>
<div class="inline-figure"><img src="main_files/viz%20misleading%20scales-1.png" width="672"></div>
<p>The left plot, which had its y axis limited to the values of the data, make the
differences between samples seem visually much larger than they actually are.
If a difference between 87 and 93 percent is in fact meaningful, then this might
be ok, but the context matters.</p>
<p>You will identify your own strategies over time as you practice data
visualization, and these rules of thumb might not work for you. There is no one
right way to do data visualization, and as long as you are dedicated to
representing your data accurately, there is no wrong way.</p>

</div>
</div>
<div id="dv-gg" class="section level2" number="7.2">
<h2>
<span class="header-section-number">7.2</span> Grammar of Graphics<a class="anchor" aria-label="anchor" href="#dv-gg"><i class="fas fa-link"></i></a>
</h2>
<p>The grammar of graphics is a system of rules that describes how data and
graphical aesthetics (e.g. color, size, shape, etc) are combined to form
graphics and plots. First popularized in the book <a href="https://www.google.com/books/edition/_/ZiwLCAAAQBAJ?hl=en">The Grammar of
Graphics</a> by Leland
Wilkinson and co-authors in 1999, this grammar is a major contribution to the
structural theory of statistical graphics. In 2005, Hadley Wickam wrote an
implementation of the grammar of graphics in R called
<a href="https://ggplot2.tidyverse.org/">ggplot2</a> (gg stands for grammar of graphics).</p>
<p>Under the grammar of graphics, every plot is the combination of three types of
information: <strong>data</strong>, <strong>geometry</strong>, and <strong>aesthetics</strong>. Data is the data we
wish to plot. Geometry is the type of geometry we wish to use to depict the data
(e.g. circles, squares, lines, etc). Aesthetics connect the data to the geometry
and defines how the data controls the way the selected geometry looks.</p>
<p>A simple example will help to explain. Consider the following made up sample
metadata tibble for a study of subjects who died with Alzheimer’s Disease (AD)
and neuropathologically normal controls:</p>
<div class="sourceCode" id="cb269"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ad_metadata</span></span></code></pre></div>
<pre><code>## # A tibble: 20 x 8
##    ID    age_at_death condition    tau  abeta   iba1   gfap braak_stage
##    &lt;chr&gt;        &lt;dbl&gt; &lt;fct&gt;      &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt; &lt;fct&gt;      
##  1 A1              81 AD        141017 230227  32959  26196 6          
##  2 A2              78 AD        141082 214944 204381  26739 6          
##  3 A3              80 AD         40788  46663      0  29308 2          
##  4 A4              85 AD         78770 136101  98074  41177 3          
##  5 A5              81 AD        110573  42893 140591  75334 5          
##  6 A6              79 AD        125934 199602 133705  91069 5          
##  7 A7              70 AD         32826  31016  34544  27905 1          
##  8 A8              76 AD         95281  92308 116275 143759 4          
##  9 A9              80 AD         55035 154453  62074 126360 2          
## 10 A10             94 AD         53040   9099  39297 137833 2          
## 11 C1              78 Control    35684      0  38523  59819 1          
## 12 C2              77 Control    62182  29663  73422  52276 3          
## 13 C3              73 Control    49062 106332      0  73822 2          
## 14 C4              70 Control    10123      0  13962  96704 0          
## 15 C5              74 Control     1530   2169   2002  83280 0          
## 16 C6              73 Control    25514  49980  25771  53798 1          
## 17 C7              81 Control    24367  48786  23961  17561 1          
## 18 C8              69 Control    43628  36442  19467  41970 2          
## 19 C9              78 Control    48923  64880  16367 110464 2          
## 20 C10             77 Control     9688   3818  12424  59021 0</code></pre>
<p>For context, <a href="https://en.wikipedia.org/wiki/Tau_protein">tau protein</a> and
<a href="https://en.wikipedia.org/wiki/Amyloid_beta">amyloid beta peptides from the amyloid precursor
protein</a>
aggregate into neurofibrillary tangles and A-beta plaques, respectively, the
brains of people with AD. Generally, the amount of both of these pathologies is
associated with more severe disease. <a href="https://en.wikipedia.org/wiki/Braak_staging">Braak
stage</a> is a neuropathological
assessment of the amount of pathology in a brain that is associated with the
severity of disease, where 0 indicates absence of pathology and 6 with
widespread involvement in multiple brain regions. Aggregation of tau is also a
<a href="https://pubmed.ncbi.nlm.nih.gov/8866683/">consequence of normal aging</a>, so must
accompany neurological symptoms such as dementia to indicate an AD diagnosis
post mortem. Note we have control samples as well as AD.</p>
<div class="float">
<img src="https://upload.wikimedia.org/wikipedia/commons/f/f7/Tauopathy_in_Alzheimer%27s_disease.jpg" alt="Tauopathy: tau protein accumulates in the cell bodies of affected neurons - Wikipedia"><div class="figcaption">Tauopathy: tau protein accumulates in the cell bodies of affected neurons -
Wikipedia</div>
</div>
<p>The histology measures <code>tau</code>, <code>abeta</code>, <code>iba1</code>, and <code>gfap</code> have been quantified
using digital microscopy, where brain sections are stained with
immunohistochemistry to identify the location and degree of pathology; the
measures in the table are the number of pixels of a 400 x 400 pixel image of a
piece of brain tissue that fluoresce when stained with the corresponding
antibody. Tau and A-beta antibodies are specialized to the types of aggregated
proteins mentioned above and provide a quantification of the level of overall
AD pathology. Ionized calcium binding adaptor molecule 1 (IBA1) is a marker of
activated <a href="https://en.wikipedia.org/wiki/Microglia">microglia</a>, the resident
macrophages of the brain, which is an
indication of neuroinflammation. Glial fibrillary acidic protein (GFAP) is a
marker for activated <a href="https://en.wikipedia.org/wiki/Astrocyte">astrocytes</a>,
specialized cells that derive from the neuron lineage, are critical for
maintaining the <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4292164/">blood brain
barrier</a>, and are also
involved in the neuroinflammatory response.</p>
<p>Let’s say we wished to visualize the relationship between age at death and the
amount of tau pathology. A scatter plot where each marker is a subject with <span class="math inline">\(x\)</span>
and <span class="math inline">\(y\)</span> position corresponding to <code>age_at_death</code> and <code>tau</code> respectively. The
following R code creates such a plot with ggplot2:</p>
<div class="sourceCode" id="cb271"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span>data<span class="op">=</span><span class="va">ad_metadata</span>, mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">age_at_death</span>, y<span class="op">=</span><span class="va">tau</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="main_files/unnamed-chunk-96-1.png" width="672"></div>
<p>All ggplot2 plots begin with the <code><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot()</a></code> function call, which is passed a
tibble with the <strong>data</strong> to be plotted. We then define the <strong>aesthetics</strong> are
defined by mapping the x coordinate to the <code>age_at_death</code> column and the y
coordinate to the <code>tau</code> column with <code>aes(x = age_at_death, y=tau)</code>. Finally, the
<strong>geometry</strong> as ‘point’ with <code><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point()</a></code>, meaning marks will be made at pairs
of x,y coordinates. The plot shows what we expect given our knowledge of the
relationship between age and amount of tau; the two look to be positively
correlated.</p>
<p>However, we are not capturing the whole story: we know that there are both AD
and Control subjects in this dataset. How does <code>condition</code> relate to this
relationship we see? We can layer on an additional aesthetic of color to add
this information to the plot:</p>
<div class="sourceCode" id="cb272"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span>data<span class="op">=</span><span class="va">ad_metadata</span>, mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">age_at_death</span>, y<span class="op">=</span><span class="va">tau</span>, color<span class="op">=</span><span class="va">condition</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="main_files/unnamed-chunk-97-1.png" width="672"></div>
<p>This looks a little clearer, showing that Control subjects generally have both
an earlier age at death and a lower amount of tau pathology. This might be a
problem, however, since if the age distributions of AD and Control groups are
different that might pose a problem with
<a href="https://en.wikipedia.org/wiki/Confounding">confounding</a>. We should investigate
this.</p>
<p>Instead of plotting age at death and tau against each other, we will examine the
distributions of each of these variables for AD and Control samples separately.
We will use the <a href="https://ggplot2.tidyverse.org/reference/geom_violin.html">violin
geometry</a> with
<code><a href="https://ggplot2.tidyverse.org/reference/geom_violin.html">geom_violin()</a></code> to look at the distributions of <code>age_at_death</code>:</p>
<div class="sourceCode" id="cb273"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span>data<span class="op">=</span><span class="va">ad_metadata</span>, mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">condition</span>, y<span class="op">=</span><span class="va">age_at_death</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_violin.html">geom_violin</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="main_files/unnamed-chunk-98-1.png" width="672"></div>
<p>We can see immediately that there are big differences between the age
distributions of the two groups. This is not ideal, but perhaps we can adjust
for these effects in downstream analyses. We’d like to look at the tau
distributions as well, but it would be nice to have these two plots side by side
in the same plot. To do that, we will use another library called
<a href="https://patchwork.data-imaginist.com/">patchwork</a>, which allows independent
ggplot2 plots to be arranged together with a simple expressive syntax:</p>
<div class="sourceCode" id="cb274"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com">patchwork</a></span><span class="op">)</span></span>
<span><span class="va">age_boxplot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span>data<span class="op">=</span><span class="va">ad_metadata</span>, mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">condition</span>, y<span class="op">=</span><span class="va">age_at_death</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html">geom_boxplot</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">tau_boxplot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span>data<span class="op">=</span><span class="va">ad_metadata</span>, mapping<span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">condition</span>, y<span class="op">=</span><span class="va">tau</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html">geom_boxplot</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">age_boxplot</span> <span class="op">|</span> <span class="va">tau_boxplot</span> <span class="co"># this puts the plots side by side</span></span></code></pre></div>
<div class="inline-figure"><img src="main_files/unnamed-chunk-99-1.png" width="672"></div>
<p>This confirms our suspicion, and also reveals a serious problem with our
samples: we have strong confounding of tau and age at death between AD and
Control samples. This means that if we look for differences between AD and
Control, we won’t know if the difference is due to the amount of tau pathology
or due to age of the subjects. With this sample set, we simply cannot
confidently answer that question. Just a few simple plots alerted us to this
problem; hopefully more expensive datasets have not already been generated for
these samples, so that hopefully different subjects are available that could
avoid this confounding.</p>
<p>This has been a biological data analysis oriented tutorial on plotting meant to
illustrate the principles of the grammar of graphics. Namely, every plot has
<strong>data</strong>, <strong>geometry</strong>, and <strong>aesthetics</strong> that can be independently controlled
to produce many types of plots. Many of these plots have names, like scatter
plots and boxplots, but as you compose different types of geometries and
aesthetics together you may find yourself generating plots that aren’t so easily
named.</p>
<p>The next sections of this chapter are a kind of “cook book” of different kinds
plots you can generate with data of different shapes. It is not intended to be
comprehensive, but a helpful guide when you are trying to decide how to
visualize your own datasets.</p>
<div class="box note">
<p>If you want to go directly to the comprehensive documentation of the many types
of ggplot2 plots, peruse the <a href="https://www.r-graph-gallery.com/">R Graph Gallery</a>
site.</p>
</div>
<div class="box readmore">
<ul>
<li>
<a href="https://ggplot2-book.org/index.html">ggplot2 - Elegant Graphics for Data Analysis</a>, by Hadley Wickam</li>
<li><a href="https://r4ds.had.co.nz/data-visualisation.html">R for Data Science - Data Vizualization</a></li>
<li><a href="https://www.r-graph-gallery.com/">R Graph Gallery</a></li>
<li><a href="https://patchwork.data-imaginist.com/">patchwork package</a></li>
</ul>
</div>
<div id="ggplot-mechanics" class="section level3" number="7.2.1">
<h3>
<span class="header-section-number">7.2.1</span> <code>ggplot</code> mechanics<a class="anchor" aria-label="anchor" href="#ggplot-mechanics"><i class="fas fa-link"></i></a>
</h3>
<p><code>ggplot</code> has two key concepts that give it great flexibility: <strong>layers</strong> and
**scales*.</p>
<p>Every plot has one or more layers that contain a type of geometry that
represents a data encoding. In general, each layer will only have one geometry
type, e.g. points or lines, but the geometry might be complex, e.g. density
plots. The layers added to a plot form a stack, where the layers added first are
beneath those added later. The geometry in each layer may draw from the same
data, or each may have its own. Each layer may also share the aesthetic mapping
from the <code><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot()</a></code> call, or may have its own. This is why both the <code><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot()</a></code>
function and each individual <code>geom_X()</code> function can accept data and aesthetic
mappings. The package comes with a large number of geometries described in its
<a href="https://ggplot2.tidyverse.org/reference/index.html#geoms">reference documentation</a>.</p>
<p>The geometry in each layer maps the data values to visual properties using
scales. A scale may map a data range to a pixel range, or to a color on a color
gradient, or one of a set of discrete colors or shapes. <code>ggplot</code> provides
reasonable default scales for each geometry type. You can override these
defaults by using the <a href="https://ggplot2.tidyverse.org/reference/index.html#scales"><code>scale_X</code>
functions</a>.</p>
<p>The <a href="https://ggplot2-book.org/">ggplot2 book</a> is an excellent resource for all
things ggplot2.</p>

</div>
</div>
<div id="plotting-one-dimension" class="section level2" number="7.3">
<h2>
<span class="header-section-number">7.3</span> Plotting One Dimension<a class="anchor" aria-label="anchor" href="#plotting-one-dimension"><i class="fas fa-link"></i></a>
</h2>
<p>The simplest plots involve plotting a single vector of numbers, or several such
vectors (e.g. for different samples). Each value in the vector typically
corresponds to a category or fixed value, for example the <code>tau</code> column from the
example above has pairs of <em>(ID, tau value)</em>. The order of these numbers can be
changed, but the vector remains one dimensional or 1-D.</p>
<div id="dv-bar" class="section level3" number="7.3.1">
<h3>
<span class="header-section-number">7.3.1</span> Bar chart<a class="anchor" aria-label="anchor" href="#dv-bar"><i class="fas fa-link"></i></a>
</h3>
<p>Bar charts map length (i.e. the height or width of a box) to the scalar value of
a number. The difference in visual length can help the viewer notice consistent
patterns in groups of bars, depending on how they are arranged:</p>
<div class="sourceCode" id="cb275"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">ad_metadata</span>, mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">ID</span>,y<span class="op">=</span><span class="va">tau</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html">geom_bar</a></span><span class="op">(</span>stat<span class="op">=</span><span class="st">"identity"</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="main_files/unnamed-chunk-101-1.png" width="672"></div>
<p>Note the <code>stat="identity"</code> argument is required because by default <code>geom_bar</code>
counts the number of values for each value of x, which in our case is only ever
one. This plot is not particularly helpful, so let’s change the fill color of
the bars based on condition:</p>
<div class="sourceCode" id="cb276"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">ad_metadata</span>, mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">ID</span>,y<span class="op">=</span><span class="va">tau</span>,fill<span class="op">=</span><span class="va">condition</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html">geom_bar</a></span><span class="op">(</span>stat<span class="op">=</span><span class="st">"identity"</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="main_files/unnamed-chunk-102-1.png" width="672"></div>
<p>Slightly better, but maybe we can see even more clearly if we sort our tibble
by tau first. Sorting elements in these 1-D charts is somewhat complicated, and
is explained in the [Reordering 1-D Data Elements] section below.</p>
<p>Bar charts can also plot negative numbers. In the following example, we center
the tau measurements by subtracting the mean from each value before plotting:</p>
<div class="sourceCode" id="cb277"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span><span class="va">ad_metadata</span>, tau_centered<span class="op">=</span><span class="op">(</span><span class="va">tau</span> <span class="op">-</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span><span class="op">(</span><span class="va">tau</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span>mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">ID</span>, y<span class="op">=</span><span class="va">tau_centered</span>, fill<span class="op">=</span><span class="va">condition</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html">geom_bar</a></span><span class="op">(</span>stat<span class="op">=</span><span class="st">"identity"</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="main_files/unnamed-chunk-103-1.png" width="672"></div>
</div>
<div id="lollipop-plots" class="section level3" number="7.3.2">
<h3>
<span class="header-section-number">7.3.2</span> Lollipop plots<a class="anchor" aria-label="anchor" href="#lollipop-plots"><i class="fas fa-link"></i></a>
</h3>
<p>Similar to <a href="data-visualization.html#dv-bar">bar charts</a>, so-called “lollipop plots” replace the bar
with a line segment and a circle. The length of the line segment is proportional
to the magnitude of the number, and the point marks the length of the segment as
a height on the y or length on the x axis, depending on orientation.</p>
<div class="sourceCode" id="cb278"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">ad_metadata</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span>mapping<span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">ID</span>, y<span class="op">=</span><span class="va">tau</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_segment.html">geom_segment</a></span><span class="op">(</span>mapping<span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">ID</span>, xend<span class="op">=</span><span class="va">ID</span>, y<span class="op">=</span><span class="fl">0</span>, yend<span class="op">=</span><span class="va">tau</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="main_files/unnamed-chunk-104-1.png" width="672"></div>
<p>Note that <code><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes()</a></code> mappings can be made on the <code><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot()</a></code> object or on each
individual geometry function call, to specify different mappings based on
geometry.</p>
</div>
<div id="stacked-area-charts" class="section level3" number="7.3.3">
<h3>
<span class="header-section-number">7.3.3</span> Stacked Area charts<a class="anchor" aria-label="anchor" href="#stacked-area-charts"><i class="fas fa-link"></i></a>
</h3>
<p>Stacked area charts can visualize multiple 1D data that share a common
categorical axis. The charts consist of one line per variable with vertices that
correspond to x and y values similar to a <a href="#bar-plots">bar</a> or
<a href="data-visualization.html#lollipop-plots">lollipop</a> plots. Each variable is plotted using the previous
one as a baseline, so that the height of the data points for each category is
proportional to their sum. The space between the lines for each variable and the
previous one are filled with a color. The following plot visualizes the amount
of marker stain for each of the four genes for each individaul:</p>
<div class="sourceCode" id="cb279"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">pivot_longer</span><span class="op">(</span></span>
<span>  <span class="va">ad_metadata</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">tau</span>,<span class="va">abeta</span>,<span class="va">iba1</span>,<span class="va">gfap</span><span class="op">)</span>,</span>
<span>  names_to<span class="op">=</span><span class="st">'Marker'</span>,</span>
<span>  values_to<span class="op">=</span><span class="st">'Intensity'</span></span>
<span>  <span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">ID</span>,y<span class="op">=</span><span class="va">Intensity</span>,group<span class="op">=</span><span class="va">Marker</span>,fill<span class="op">=</span><span class="va">Marker</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_ribbon.html">geom_area</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="main_files/unnamed-chunk-105-1.png" width="672"></div>
<p>We notice that subject A4 has the highest overall level of marker intensity,
followed by A1, A7, etc. The control samples overall have less intensity across
all markers. Certain samples, A2 and C5, have little to no <code>abeta</code> aggregation,
and C6 has little to no <code>tau</code>.</p>
<p>Stacked area plots require three pieces of data:</p>
<ul>
<li>
<strong>x</strong> - a numeric or categorical axis for vertical alignment</li>
<li>
<strong>y</strong> - a numeric axis to draw vertical proportions</li>
<li>
<strong>group</strong> - a categorical variable that indicates which (x,y) pairs correspond
to the same line</li>
</ul>
<p>In the example above, we needed to pivot our tibble so that the different
markers and their values were placed into columns <code>Marker</code> and <code>Intensity</code>,
respectively. Data for stacked bar charts will usually need to be in this ‘long’
format, as described in <a href="data-wrangling.html#rearranging-data">Rearranging Data</a>.</p>
<p>Sometimes it is more helpful to view the relative proportion of values in each
category rather than the actual values. The result is called a <em>proportional</em>
stacked area plots. While not a distinct plot type, we can create one by
preprocessing our data by dividing each value by the column sum:</p>
<div class="sourceCode" id="cb280"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">pivot_longer</span><span class="op">(</span></span>
<span>  <span class="va">ad_metadata</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">tau</span>,<span class="va">abeta</span>,<span class="va">iba1</span>,<span class="va">gfap</span><span class="op">)</span>,</span>
<span>  names_to<span class="op">=</span><span class="st">'Marker'</span>,</span>
<span>  values_to<span class="op">=</span><span class="st">'Intensity'</span></span>
<span>  <span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">ID</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="co"># we want to divide each subjects intensity values by the sum of all four markers</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span></span>
<span>    `Relative Intensity`<span class="op">=</span><span class="va">Intensity</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">Intensity</span><span class="op">)</span></span>
<span>  <span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">ungroup</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="co"># ungroup restores the tibble to its original number of rows after the transformation</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">ID</span>,y<span class="op">=</span><span class="va">`Relative Intensity`</span>,group<span class="op">=</span><span class="va">Marker</span>,fill<span class="op">=</span><span class="va">Marker</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_ribbon.html">geom_area</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="main_files/unnamed-chunk-106-1.png" width="672">
Now the values for each subject have been normalized to each sum to 1. In this
way, we might note that the relative proportion of <code>abeta</code> seems to be greater
in AD samples than Controls, but that may not be true of <code>tau</code>. These
observations may inspire us to ask these questions more rigorously than we have
done so far by inspection.</p>

</div>
</div>
<div id="visualizing-distributions" class="section level2" number="7.4">
<h2>
<span class="header-section-number">7.4</span> Visualizing Distributions<a class="anchor" aria-label="anchor" href="#visualizing-distributions"><i class="fas fa-link"></i></a>
</h2>
<p>The distribution is one of the most important properties of a set of numbers.
A distribution describes the general “shape” of the numbers, i.e. what is the
relative frequency of the values, or ranges of values, within the data.
Understanding the distribution of a data set is critical when choosing methods
to apply, since many methods are only appropriate when data is distributed in
certain ways, e.g. <a href="https://en.wikipedia.org/wiki/Linear_regression">linear
regression</a> assumes the
response variable is <a href="https://en.wikipedia.org/wiki/Normal_distribution">normally
distributed</a>, otherwise the
result of the model cannot be interpreted properly. Often, we don’t know how our
data are distributed when we obtain it and so we must examine the distribution
empirically. The visualizations in this section are all used for the purpose of
depicting the distribution of a set of numbers.</p>
<div id="histogram" class="section level3" number="7.4.1">
<h3>
<span class="header-section-number">7.4.1</span> Histogram<a class="anchor" aria-label="anchor" href="#histogram"><i class="fas fa-link"></i></a>
</h3>
<p>The most common way to plot the distribution of a 1-D set of data is the
<a href="https://en.wikipedia.org/wiki/Histogram">histogram</a>. The histogram divides up
the range of a dataset from minimum to maximum into bins usually of the same
width and tabulates the number of values that fall within each bin. Below is a
histogram of our <code>age_at_death</code> measurement for all samples:</p>
<div class="sourceCode" id="cb281"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">ad_metadata</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html">geom_histogram</a></span><span class="op">(</span>mapping<span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">age_at_death</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="main_files/unnamed-chunk-107-1.png" width="672"></div>
<p>Note that the histogram does not look very complete, because there are only 20
values in our data. We can mitigate this somewhat by increasing the number of
bins the data range is divided into:</p>
<div class="sourceCode" id="cb282"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">ad_metadata</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html">geom_histogram</a></span><span class="op">(</span>mapping<span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">age_at_death</span><span class="op">)</span>,bins<span class="op">=</span><span class="fl">10</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="main_files/unnamed-chunk-108-1.png" width="672"></div>
<p>This is a little bit better, but there are still some bins (76-79, 84-87) that
have no values. Compare this to the following synthetic dataset of 1000
normally distributed values:</p>
<div class="sourceCode" id="cb283"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html">tibble</a></span><span class="op">(</span></span>
<span>  x<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span><span class="op">(</span><span class="fl">1000</span><span class="op">)</span></span>
<span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html">geom_histogram</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">x</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="main_files/unnamed-chunk-109-1.png" width="672"></div>
<p>For distributions with a small number of samples, histograms might not be the
best visualization. We will continue with synthetic normally distributed dataset
for the remaining examples.</p>
<p>ggplot allows you to easily plot multiple distributions on the same plot:</p>
<div class="sourceCode" id="cb284"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html">tibble</a></span><span class="op">(</span></span>
<span>  x<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span><span class="op">(</span><span class="fl">1000</span><span class="op">)</span>,<span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span><span class="op">(</span><span class="fl">1000</span>,mean<span class="op">=</span><span class="fl">4</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  type<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="st">'A'</span>,<span class="fl">1000</span><span class="op">)</span>,<span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="st">'B'</span>,<span class="fl">1000</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">x</span>,fill<span class="op">=</span><span class="va">type</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html">geom_histogram</a></span><span class="op">(</span>bins<span class="op">=</span><span class="fl">30</span>, alpha<span class="op">=</span><span class="fl">0.6</span>, position<span class="op">=</span><span class="st">"identity"</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="main_files/unnamed-chunk-110-1.png" width="672"></div>
<p>The <code>alpha=0.6, position="identity"</code> arguments makes the bars partially
transparent so you can see the overlap more clearly.</p>
<div class="box readmore">
<ul>
<li><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html?q=histogram">ggplot2 geom_histogram reference</a></li>
<li><a href="https://www.r-graph-gallery.com/histogram.html">R Graph Gallery - histograms</a></li>
</ul>
</div>
</div>
<div id="density" class="section level3" number="7.4.2">
<h3>
<span class="header-section-number">7.4.2</span> Density<a class="anchor" aria-label="anchor" href="#density"><i class="fas fa-link"></i></a>
</h3>
<p>Another way to describe a distribution is with a <a href="https://www.r-graph-gallery.com/density-plot.html">density
plot</a>. Instead of binning the
values into intervals and drawing bars with height proportional to the number of
values in each bin, a density plot draws a smoothly interpolated line that
approximates the distribution instead. A key difference between a histogram and
a density plot is the density plot is always normalized so the integral under
the curve is approximately 1, whereas a histogram may be either counts or, if
the counts in each bin are divided by the total number of data points, a
proportion.</p>
<p>Compare the histogram and density plots
of the <code>age_at_death</code> variable from our example tibble:</p>
<div class="sourceCode" id="cb285"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com">patchwork</a></span><span class="op">)</span></span>
<span><span class="va">hist_g</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">ad_metadata</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html">geom_histogram</a></span><span class="op">(</span>mapping<span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">age_at_death</span><span class="op">)</span>,bins<span class="op">=</span><span class="fl">30</span><span class="op">)</span></span>
<span><span class="va">density_g</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">ad_metadata</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_density.html">geom_density</a></span><span class="op">(</span>mapping<span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">age_at_death</span><span class="op">)</span>,fill<span class="op">=</span><span class="st">"#c9a13daa"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">hist_g</span> <span class="op">|</span> <span class="va">density_g</span></span></code></pre></div>
<div class="inline-figure"><img src="main_files/unnamed-chunk-111-1.png" width="672"></div>
<p>Notice the overall shape of the two distributions is similar, with the highest
values in both around age 77. The density plot is a smoother representation of
a histogram, but its accuracy is still highly sensitive to the number of
measurements used to construct it. Compare the histogram and density plots of
two sets of 1000 normally distributed samples with different means:</p>
<div class="sourceCode" id="cb286"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com">patchwork</a></span><span class="op">)</span></span>
<span><span class="va">normal_samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html">tibble</a></span><span class="op">(</span></span>
<span>  x<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span><span class="op">(</span><span class="fl">1000</span><span class="op">)</span>,<span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span><span class="op">(</span><span class="fl">1000</span>,mean<span class="op">=</span><span class="fl">4</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  type<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="st">'A'</span>,<span class="fl">1000</span><span class="op">)</span>,<span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="st">'B'</span>,<span class="fl">1000</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">hist_g</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">normal_samples</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html">geom_histogram</a></span><span class="op">(</span></span>
<span>    mapping<span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">x</span>,fill<span class="op">=</span><span class="va">type</span><span class="op">)</span>,</span>
<span>    alpha<span class="op">=</span><span class="fl">0.6</span>,</span>
<span>    position<span class="op">=</span><span class="st">"identity"</span>,</span>
<span>    bins<span class="op">=</span><span class="fl">30</span></span>
<span><span class="op">)</span></span>
<span><span class="va">density_g</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">normal_samples</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_density.html">geom_density</a></span><span class="op">(</span></span>
<span>    mapping<span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">x</span>,fill<span class="op">=</span><span class="va">type</span><span class="op">)</span>,</span>
<span>    alpha<span class="op">=</span><span class="fl">0.6</span>,</span>
<span>    position<span class="op">=</span><span class="st">"identity"</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span><span class="va">hist_g</span> <span class="op">|</span> <span class="va">density_g</span></span></code></pre></div>
<div class="inline-figure"><img src="main_files/unnamed-chunk-112-1.png" width="672"></div>
<p>Again the two types of plots depict similar distributions, although they are
different enough to possibly suggest different interpretations. In general,
density plots might be preferable over histograms if the data are noisy or
sparse in that they produce cleaner plots, but potentially at the expense of
accuracy when the number of samples is low.</p>
<div class="box readmore">
<ul>
<li><a href="https://ggplot2.tidyverse.org/reference/geom_density.html?q=density">ggplot2 geom_density reference</a></li>
<li><a href="https://www.r-graph-gallery.com/density-plot.html">R Graph Gallery - density plots</a></li>
</ul>
</div>
</div>
<div id="boxplot" class="section level3" number="7.4.3">
<h3>
<span class="header-section-number">7.4.3</span> Boxplot<a class="anchor" aria-label="anchor" href="#boxplot"><i class="fas fa-link"></i></a>
</h3>
<p>Box plots, or box and whisker plots, are extremely common when used to describe
distributions. Below is a boxplot of age at death divided by condition:</p>
<div class="sourceCode" id="cb287"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">ad_metadata</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html">geom_boxplot</a></span><span class="op">(</span>mapping<span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">condition</span>,y<span class="op">=</span><span class="va">age_at_death</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="main_files/unnamed-chunk-113-1.png" width="672"></div>
<p>Boxplots are drawn assuming the data are
<a href="https://en.wikipedia.org/wiki/Unimodality">unimodal</a> (i.e. shaped like a hill,
possibly slanted to one side or the other), where the extents of the box
represent the 1st and 3rd quartile of the data, the central line is the median,
the whiskers are drawn as 1.5 times the value outside the 1st and 3rd quartiles.
Sometimes individual values more extreme than the whiskers are drawn
individually to identify them as outliers.</p>
<div class="float">
<img src="https://upload.wikimedia.org/wikipedia/commons/1/1a/Boxplot_vs_PDF.svg" alt="Boxplot anatomy. IQR stands for “inner quartile range”, the distance between the 1st and 3rd quartile - Wikipedia"><div class="figcaption">Boxplot anatomy. IQR stands for “inner quartile range”, the distance between
the 1st and 3rd quartile - Wikipedia</div>
</div>
<p>However, boxplots have some significant shortcomings. Primarily, the rectangle
of the inner quartile range does not describe the actual distribution of the
samples within it. Although the median can give a sense of skewness, if the
data are not unimodal this may be misleading. Consider the following
distributions plotted as boxplots or as violin plots (described in the next
section):</p>
<div class="sourceCode" id="cb288"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com">patchwork</a></span><span class="op">)</span></span>
<span><span class="va">normal_samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html">tibble</a></span><span class="op">(</span></span>
<span>  x<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span><span class="op">(</span><span class="fl">1000</span><span class="op">)</span>,<span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span><span class="op">(</span><span class="fl">1000</span>,<span class="fl">4</span><span class="op">)</span>,<span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span><span class="op">(</span><span class="fl">1000</span>,<span class="fl">2</span>,<span class="fl">3</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  type<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="st">'A'</span>,<span class="fl">2000</span><span class="op">)</span>,<span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="st">'B'</span>,<span class="fl">1000</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">g</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">normal_samples</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">type</span>,y<span class="op">=</span><span class="va">x</span>,fill<span class="op">=</span><span class="va">type</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">boxplot_g</span> <span class="op">&lt;-</span> <span class="va">g</span> <span class="op">+</span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html">geom_boxplot</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">violin_g</span> <span class="op">&lt;-</span> <span class="va">g</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_violin.html">geom_violin</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">boxplot_g</span> <span class="op">|</span> <span class="va">violin_g</span></span></code></pre></div>
<div class="inline-figure"><img src="main_files/unnamed-chunk-114-1.png" width="672"></div>
<p>The two distributions look almost identical in the boxplot figure; however they
are dramatically different when visualized using a method like a violin plot
where the contours of the entire distribution are depicted. Unless you are
certain that your data are unimodal, one of the other distribution visualization
methods in this section will likely more accurately depict your data than a
boxplot.</p>
<div class="box readmore">
<ul>
<li><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html">ggplot2 geom_boxplot reference</a></li>
<li><a href="https://www.r-graph-gallery.com/boxplot.html">R Graph Gallery - boxplots</a></li>
</ul>
</div>
</div>
<div id="violin-plot" class="section level3" number="7.4.4">
<h3>
<span class="header-section-number">7.4.4</span> Violin plot<a class="anchor" aria-label="anchor" href="#violin-plot"><i class="fas fa-link"></i></a>
</h3>
<p>As seen in the last section, a <a href="https://ggplot2.tidyverse.org/reference/geom_violin.html">violin
plot</a> is another way
to depict a distribution by producing a shape where the width is proportional to
the value along the x or y axis, depending on orientation. The “violin” shape is
similar in principle to a histogram or a density plot, in that it describes the
contour of all the data in the distribution, not just the quantiles and extents,
as in a box plot. Below is a violin plot of the <code>tau</code> measures from our example
tibble:</p>
<div class="sourceCode" id="cb289"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">ad_metadata</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_violin.html">geom_violin</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">condition</span>,y<span class="op">=</span><span class="va">tau</span>,fill<span class="op">=</span><span class="va">condition</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="main_files/unnamed-chunk-115-1.png" width="672"></div>
<p>The violin plot is both more and less descriptive than a boxplot; it does depict
the entire distribution of the data, but also doesn’t include features like
median by default.</p>
<div class="box readmore">
<ul>
<li><a href="https://ggplot2.tidyverse.org/reference/geom_violin.html">ggplot2 geom_violin reference</a></li>
<li><a href="https://www.r-graph-gallery.com/violin.html">R Graph Gallery - violin plots</a></li>
</ul>
</div>
</div>
<div id="beeswarm-plot" class="section level3" number="7.4.5">
<h3>
<span class="header-section-number">7.4.5</span> Beeswarm plot<a class="anchor" aria-label="anchor" href="#beeswarm-plot"><i class="fas fa-link"></i></a>
</h3>
<p>The <a href="https://r-charts.com/distribution/ggbeeswarm/">beeswarm plot</a> is similar to
a violin plot, but instead of plotting the contours of the data, it plots the
data itself as points like in a scatter plot. The individual values of the
distribution are organized vertically and spaced such that the points don’t
overlap. In this plot, the distribution of age at death is plotted for each kind
of sample and the markers are colored by the amount of tau:</p>
<div class="sourceCode" id="cb290"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/eclarke/ggbeeswarm">ggbeeswarm</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">ad_metadata</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/ggbeeswarm/man/geom_beeswarm.html">geom_beeswarm</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">condition</span>,y<span class="op">=</span><span class="va">age_at_death</span>,color<span class="op">=</span><span class="va">condition</span><span class="op">)</span>,cex<span class="op">=</span><span class="fl">2</span>,size<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="main_files/unnamed-chunk-116-1.png" width="672"></div>
<p>We may not have noticed before that our AD samples have a big gap in ages
between 74 and 81; since the beeswarm plot displays all the data, we can see it
easily here.</p>
<p>Beeswarm plots are typically only useful when the number of values is within a
range; not too many and not too few. The example above is close to having too
few values per group for this plot to be useful, but consider the following with
too many samples:</p>
<div class="sourceCode" id="cb291"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">normal_samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html">tibble</a></span><span class="op">(</span></span>
<span>  x<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span><span class="op">(</span><span class="fl">1000</span><span class="op">)</span>,<span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span><span class="op">(</span><span class="fl">1000</span>,<span class="fl">4</span><span class="op">)</span>,<span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span><span class="op">(</span><span class="fl">1000</span>,<span class="fl">2</span>,<span class="fl">3</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  type<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="st">'A'</span>,<span class="fl">2000</span><span class="op">)</span>,<span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="st">'B'</span>,<span class="fl">1000</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">normal_samples</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">type</span>,y<span class="op">=</span><span class="va">x</span>,color<span class="op">=</span><span class="va">type</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/ggbeeswarm/man/geom_beeswarm.html">geom_beeswarm</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="main_files/unnamed-chunk-117-1.png" width="672"></div>
<p>This plot likely has too many samples to be the right choice (it’s also ugly),
but it does give an idea of the distribution of the data.</p>
<p>In the previous examples the markers for each group also determined the color of
the group. This makes the chart a bit easier to read and more pleasing to the
eye, but is technically redundant. You can use however profitably however to
color markers by some other value that might be of interest. Consider this final
example where markers are colored by another randomly generated variable:</p>
<div class="sourceCode" id="cb292"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">normal_samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html">tibble</a></span><span class="op">(</span></span>
<span>  x<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span><span class="op">(</span><span class="fl">100</span><span class="op">)</span>,<span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span><span class="op">(</span><span class="fl">100</span>,<span class="fl">4</span><span class="op">)</span>,<span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span><span class="op">(</span><span class="fl">100</span>,<span class="fl">2</span>,<span class="fl">3</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  type<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="st">'A'</span>,<span class="fl">200</span><span class="op">)</span>,<span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="st">'B'</span>,<span class="fl">100</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  category<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'healthy'</span>,<span class="st">'disease'</span><span class="op">)</span>,<span class="fl">300</span>,replace<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">normal_samples</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">type</span>,y<span class="op">=</span><span class="va">x</span>,color<span class="op">=</span><span class="va">category</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/ggbeeswarm/man/geom_beeswarm.html">geom_beeswarm</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="main_files/unnamed-chunk-118-1.png" width="672"></div>
<p>We are now effectively visualizing three dimensions which may provide insight
into the data.</p>
<div class="box readmore">
<ul>
<li><a href="https://r-charts.com/distribution/ggbeeswarm/">beeswarm package reference</a></li>
</ul>
</div>
</div>
<div id="ridgeline" class="section level3" number="7.4.6">
<h3>
<span class="header-section-number">7.4.6</span> Ridgeline<a class="anchor" aria-label="anchor" href="#ridgeline"><i class="fas fa-link"></i></a>
</h3>
<p>If you have many non-trivial distributions that you would like the user to
compare, a good option is a <a href="https://www.r-graph-gallery.com/294-basic-ridgeline-plot.html">ridgeline
chart</a>. The
ridgeline plot is simply multiple density plots drawn for different variables
within the same plot. Like the beeswarm plot, ridgeline plots are provided by
another package outside ggplot2.</p>
<div class="sourceCode" id="cb293"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://wilkelab.org/ggridges/">ggridges</a></span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html">tibble</a></span><span class="op">(</span></span>
<span>  x<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span><span class="op">(</span><span class="fl">100</span><span class="op">)</span>,<span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span><span class="op">(</span><span class="fl">100</span>,<span class="fl">4</span><span class="op">)</span>,<span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span><span class="op">(</span><span class="fl">100</span>,<span class="fl">2</span>,<span class="fl">3</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  type<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="st">'A'</span>,<span class="fl">200</span><span class="op">)</span>,<span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="st">'B'</span>,<span class="fl">100</span><span class="op">)</span><span class="op">)</span>,</span>
<span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>y<span class="op">=</span><span class="va">type</span>,x<span class="op">=</span><span class="va">x</span>,fill<span class="op">=</span><span class="va">type</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://wilkelab.org/ggridges/reference/geom_density_ridges.html">geom_density_ridges</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="main_files/unnamed-chunk-119-1.png" width="672"></div>
<p>Many distributions may be plotted:</p>
<div class="sourceCode" id="cb294"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html">tibble</a></span><span class="op">(</span></span>
<span>  x<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span><span class="op">(</span><span class="fl">10000</span>,mean<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html">runif</a></span><span class="op">(</span><span class="fl">10</span>,<span class="fl">1</span>,<span class="fl">10</span><span class="op">)</span>,sd<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html">runif</a></span><span class="op">(</span><span class="fl">2</span>,<span class="fl">1</span>,<span class="fl">4</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  type<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"A"</span>,<span class="st">"B"</span>,<span class="st">"C"</span>,<span class="st">"D"</span>,<span class="st">"E"</span>,<span class="st">"F"</span>,<span class="st">"G"</span>,<span class="st">"H"</span>,<span class="st">"I"</span>,<span class="st">"J"</span><span class="op">)</span>,<span class="fl">1000</span><span class="op">)</span></span>
<span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>y<span class="op">=</span><span class="va">type</span>,x<span class="op">=</span><span class="va">x</span>,fill<span class="op">=</span><span class="va">type</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://wilkelab.org/ggridges/reference/geom_density_ridges.html">geom_density_ridges</a></span><span class="op">(</span>alpha<span class="op">=</span><span class="fl">0.6</span>,position<span class="op">=</span><span class="st">"identity"</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="main_files/unnamed-chunk-120-1.png" width="672"></div>
<div class="box readmore">
<ul>
<li><a href="https://www.r-graph-gallery.com/294-basic-ridgeline-plot.html">R Graph Gallery - ridgeline plots</a></li>
<li><a href="https://cran.r-project.org/web/packages/ggridges/vignettes/introduction.html">ggridges package vignette</a></li>
<li><a href="https://cran.r-project.org/web/packages/ggridges/index.html">ggridges package index on CRAN</a></li>
</ul>
</div>

</div>
</div>
<div id="plotting-two-or-more-dimensions" class="section level2" number="7.5">
<h2>
<span class="header-section-number">7.5</span> Plotting Two or More Dimensions<a class="anchor" aria-label="anchor" href="#plotting-two-or-more-dimensions"><i class="fas fa-link"></i></a>
</h2>
<p>Multi-dimensional plots allow relationships between different quantities to be
depicted visually. As mentioned in the [Elemental Perceptual Tasks] section,
this involves mapping data to visual elements using different encodings. The
plot types in this section are examples of these different approaches to
combining visual data encodings.</p>
<div id="scatter-plots" class="section level3" number="7.5.1">
<h3>
<span class="header-section-number">7.5.1</span> Scatter Plots<a class="anchor" aria-label="anchor" href="#scatter-plots"><i class="fas fa-link"></i></a>
</h3>
<p>Scatter plots visualize pairs of quantities, usually continuous, as points in
a two dimensional space. Most commonly, the points are plotted in <a href="https://en.wikipedia.org/wiki/Cartesian_coordinate_system">cartesian
coordinates</a>, but
<a href="https://en.wikipedia.org/wiki/Polar_coordinate_system">polar coordinates</a> or
other types of coordinate systems.</p>
<div class="sourceCode" id="cb295"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">ad_metadata</span>,mapping<span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">abeta</span>, y<span class="op">=</span><span class="va">tau</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span>size<span class="op">=</span><span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="main_files/viz%20dim%20scatter-1.png" width="672"></div>
<p>The <code><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point()</a></code> geometry draws a shape, circle by default, at pixel
coordinates proportional to each <span class="math inline">\((x,y)\)</span> pair. The shape may be changed by
specifying the <code>shape=N</code> parameter to <code>geom_point</code>, where <code>N</code> is an integer ID
of the shape. The following plot shows which IDs correspond to which shapes:</p>
<div class="sourceCode" id="cb296"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">g</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">x</span> <span class="kw">in</span> <span class="fl">0</span><span class="op">:</span><span class="fl">5</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="kw">for</span><span class="op">(</span><span class="va">y</span> <span class="kw">in</span> <span class="fl">0</span><span class="op">:</span><span class="fl">4</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="kw">if</span><span class="op">(</span><span class="va">x</span><span class="op">+</span><span class="va">y</span><span class="op">*</span><span class="fl">6</span> <span class="op">&lt;</span> <span class="fl">26</span><span class="op">)</span> <span class="op">{</span></span>
<span>      <span class="va">g</span> <span class="op">&lt;-</span> <span class="va">g</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">x</span>,y<span class="op">=</span><span class="va">y</span><span class="op">)</span>,<span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html">tibble</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">x</span>,y<span class="op">=</span><span class="va">y</span><span class="op">)</span>,shape<span class="op">=</span><span class="va">x</span><span class="op">+</span><span class="va">y</span><span class="op">*</span><span class="fl">6</span>,size<span class="op">=</span><span class="fl">8</span><span class="op">)</span> <span class="op">+</span></span>
<span>       <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_text.html">geom_label</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">x</span>,y<span class="op">=</span><span class="va">y</span><span class="op">+</span><span class="fl">0.5</span>,label<span class="op">=</span><span class="va">label</span><span class="op">)</span>,<span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html">tibble</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">x</span>,y<span class="op">=</span><span class="va">y</span>,label<span class="op">=</span><span class="va">x</span><span class="op">+</span><span class="va">y</span><span class="op">*</span><span class="fl">6</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="op">}</span></span>
<span>  <span class="op">}</span></span>
<span><span class="op">}</span></span>
<span><span class="va">g</span></span></code></pre></div>
<div class="inline-figure"><img src="main_files/point%20shapes-1.png" width="672"></div>
<p>However, <code>geom_points()</code> will choose up to six different shapes for you if you
map a categorical encoding to the <code>shape</code> aesthetic:</p>
<div class="sourceCode" id="cb297"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">ad_metadata</span>,mapping<span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">abeta</span>, y<span class="op">=</span><span class="va">tau</span>, shape<span class="op">=</span><span class="va">condition</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span>size<span class="op">=</span><span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<p><img src="main_files/scatter%20shape%20enc-1.png" width="672">
Controlling which shapes are mapped to which value can be accomplished with
<a href="https://ggplot2.tidyverse.org/reference/scale_manual.html"><code>scale_shape_manual()</code></a>:</p>
<div class="sourceCode" id="cb298"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">ad_metadata</span>,mapping<span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">abeta</span>, y<span class="op">=</span><span class="va">tau</span>, shape<span class="op">=</span><span class="va">condition</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span>size<span class="op">=</span><span class="fl">3</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_shape_manual</a></span><span class="op">(</span>values<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">3</span>,<span class="fl">9</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="main_files/scatter%20shape%20manual-1.png" width="672"></div>
<p>Color encodings can also be added using either a continuous or discrete values:</p>
<div class="sourceCode" id="cb299"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com">patchwork</a></span><span class="op">)</span></span>
<span><span class="va">g</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">ad_metadata</span><span class="op">)</span></span>
<span><span class="va">g_condition</span> <span class="op">&lt;-</span> <span class="va">g</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span>mapping<span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">abeta</span>, y<span class="op">=</span><span class="va">tau</span>, color<span class="op">=</span><span class="va">condition</span><span class="op">)</span>,size<span class="op">=</span><span class="fl">3</span><span class="op">)</span></span>
<span><span class="va">g_age</span> <span class="op">&lt;-</span> <span class="va">g</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span>mapping<span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">abeta</span>, y<span class="op">=</span><span class="va">tau</span>, color<span class="op">=</span><span class="va">age_at_death</span><span class="op">)</span>,size<span class="op">=</span><span class="fl">3</span><span class="op">)</span></span>
<span><span class="va">g_condition</span> <span class="op">/</span> <span class="va">g_age</span></span></code></pre></div>
<div class="inline-figure"><img src="main_files/scatter%20color-1.png" width="672"></div>
</div>
<div id="bubble-plots" class="section level3" number="7.5.2">
<h3>
<span class="header-section-number">7.5.2</span> Bubble Plots<a class="anchor" aria-label="anchor" href="#bubble-plots"><i class="fas fa-link"></i></a>
</h3>
<p>A close relative of the scatter plot is the bubble plot. A bubble plot is a
scatter plot where the area of the point markers is proportional to a third
continuous dimension. Consider the plot from above, with age at death encoded as
the area of each point:</p>
<div class="sourceCode" id="cb300"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">ad_metadata</span>,mapping<span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">abeta</span>, y<span class="op">=</span><span class="va">tau</span>, size<span class="op">=</span><span class="va">age_at_death</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span>alpha<span class="op">=</span><span class="fl">0.5</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure">
<img src="main_files/viz%20dim%20bubbl-1.png" width="672">
In this example, the markers have been given 50% transparency, so overlapping
bubbles may be discerned more easily.</div>
<div class="box warning">
<p>Recall from the <a href="data-visualization.html#human-visual-perception">Human Visual Perception</a> section that it is difficult for
humans to accurately assess the relative sizes of shapes that are not near each
other. Bubble plots may be useful in some instances, but should be used
carefully.</p>
</div>
</div>
<div id="connected-scatter-plots" class="section level3" number="7.5.3">
<h3>
<span class="header-section-number">7.5.3</span> Connected Scatter Plots<a class="anchor" aria-label="anchor" href="#connected-scatter-plots"><i class="fas fa-link"></i></a>
</h3>
<p>Another close relative of the scatter plot is the connected scatter plot. A
connected scatter plot is simply a scatter plot where certain pairs of points
are connected with a line in addition to the markers. Any pairs of points may
have lines connecting them, but most often the line indicates some dimension of
time, i.e. two points that are sequential in time are connected. Below is a
connected scatter plot with a line drawn from the youngest to oldest age at
death:</p>
<div class="sourceCode" id="cb301"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html">arrange</a></span><span class="op">(</span><span class="va">ad_metadata</span>,<span class="va">age_at_death</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span></span>
<span>    x<span class="op">=</span><span class="va">abeta</span>,</span>
<span>    xend<span class="op">=</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/lead-lag.html">lag</a></span><span class="op">(</span><span class="va">x</span>,<span class="fl">1</span><span class="op">)</span>,</span>
<span>    y<span class="op">=</span><span class="va">tau</span>,</span>
<span>    yend<span class="op">=</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/lead-lag.html">lag</a></span><span class="op">(</span><span class="va">y</span>,<span class="fl">1</span><span class="op">)</span></span>
<span>  <span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_segment.html">geom_segment</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">abeta</span>, xend<span class="op">=</span><span class="va">xend</span>, y<span class="op">=</span><span class="va">tau</span>, yend<span class="op">=</span><span class="va">yend</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">x</span>,y<span class="op">=</span><span class="va">y</span>,shape<span class="op">=</span><span class="va">condition</span>,color<span class="op">=</span><span class="va">condition</span><span class="op">)</span>,size<span class="op">=</span><span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="main_files/viz%20connected%20scatter-1.png" width="672"></div>
<p>The plot suggests that perhaps the AD and Control samples might be closer in age
within groups than between. We can confirm this using another type of plot, e.g.
density grouped by condition.</p>
</div>
<div id="line-plots" class="section level3" number="7.5.4">
<h3>
<span class="header-section-number">7.5.4</span> Line Plots<a class="anchor" aria-label="anchor" href="#line-plots"><i class="fas fa-link"></i></a>
</h3>
<p>Line plots connect pairs of points with a line without drawing a symbol at each
point. The <code><a href="https://ggplot2.tidyverse.org/reference/geom_path.html">geom_line()</a></code> function will draw lines between pairs of points sorted
by <span class="math inline">\(x\)</span> axis by default:</p>
<div class="sourceCode" id="cb302"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">ad_metadata</span>,mapping<span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">abeta</span>, y<span class="op">=</span><span class="va">tau</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html">geom_line</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="main_files/unnamed-chunk-122-1.png" width="672"></div>
<p>This may be useful when plotting multiple lines using the <code>group</code> aesthetic
mapping:</p>
<div class="sourceCode" id="cb303"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">pivot_longer</span><span class="op">(</span><span class="va">ad_metadata</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">tau</span>,<span class="va">abeta</span>,<span class="va">iba1</span>,<span class="va">gfap</span><span class="op">)</span>,</span>
<span>  names_to<span class="op">=</span><span class="st">'Marker'</span>,</span>
<span>  values_to<span class="op">=</span><span class="st">'Intensity'</span></span>
<span>  <span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">ad_metadata</span>,mapping<span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">ID</span>, y<span class="op">=</span><span class="va">Intensity</span>, group<span class="op">=</span><span class="va">Marker</span>, color<span class="op">=</span><span class="va">Marker</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html">geom_line</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="main_files/unnamed-chunk-123-1.png" width="672"></div>
<p>This plot is similar to a <a href="parallel-coordinate-plots">parallel coordinate plot</a>,
in that all four dimensions of each sample are displayed using a position
encoding.</p>
</div>
<div id="parallel-coordinate-plots" class="section level3" number="7.5.5">
<h3>
<span class="header-section-number">7.5.5</span> Parallel Coordinate Plots<a class="anchor" aria-label="anchor" href="#parallel-coordinate-plots"><i class="fas fa-link"></i></a>
</h3>
<p>Parallel coordinate plots are line plots where each of possibly many continuous
variables are given vertical position encodings and each sample has its own
line. Unlike the line plot above, each vertical axis can have a different scale,
thus allowing plots of variables with very different value ranges. The following
is a parallel coordinate plot of our AD metadata:</p>
<div class="sourceCode" id="cb304"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ggobi.github.io/ggally/">GGally</a></span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://ggobi.github.io/ggally/reference/ggparcoord.html">ggparcoord</a></span><span class="op">(</span><span class="va">ad_metadata</span>,</span>
<span>           columns<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">2</span>,<span class="fl">4</span><span class="op">:</span><span class="fl">8</span><span class="op">)</span>,</span>
<span>           groupColumn<span class="op">=</span><span class="fl">3</span>,</span>
<span>           showPoints<span class="op">=</span><span class="cn">TRUE</span></span>
<span>           <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_color_manual</a></span><span class="op">(</span>values<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"#bbbbbb"</span>, <span class="st">"#666666"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="main_files/viz%20pcp%20ad-1.png" width="672"></div>
<p>All the variables have been scaled and centered to have mean zero and standard
deviation 1, and the values plotted using a position encoding vertically long
each. Each line is a different subject, and have been colored based on the type
of subject, either AD or control. This encoding allows us to make a number of
observations about the samples, e.g. except for gfap, AD samples have
consistently higher values than Controls.</p>
<div class="box note">
<p>A common way to specify colors is with a six digit
<a href="https://en.wikipedia.org/wiki/Hexadecimal">hexadecimal</a> code, e.g. <code>#2ca9f3</code>.
These codes form a triplet of (<code>rr</code>,<code>gg</code>,<code>bb</code>), where <code>rr</code> is the value of the
color red, <code>gg</code> the value of green, and <code>bb</code> for blue, examples:</p>
<ul>
<li>
<code>#000000</code> is the color black, or a value of zero in all three colors</li>
<li>
<code>#ffffff</code> is the color white, with maximum value for all three colors</li>
<li>
<code>#ff0000</code> is the color red, with full red value and no value for green or blue</li>
<li>
<code>#0000ff</code> is the color blue, with full blue value and no value for red or green</li>
<li>
<code>#2ca9f3</code> is <span style="color: #2ca9f3">this color</span>, a combination of red, green, and blue</li>
</ul>
<p>All colors in tidyverse may be specified with these hexadecimal color codes. See
<a href="https://en.wikipedia.org/wiki/Web_colors">web colors</a> for more information on
how to specify colors in this way.</p>
</div>
</div>
<div id="heatmaps" class="section level3" number="7.5.6">
<h3>
<span class="header-section-number">7.5.6</span> Heatmaps<a class="anchor" aria-label="anchor" href="#heatmaps"><i class="fas fa-link"></i></a>
</h3>
<p>Heatmaps visualize values associated with a grid of points <span class="math inline">\((x,y,z)\)</span> as a grid
of colored rectangles, where <span class="math inline">\(x\)</span> and <span class="math inline">\(y\)</span> define the grid point coordinates and
<span class="math inline">\(z\)</span> is a continuous value. A common heatmap you might have seen is the <a href="https://en.wikipedia.org/wiki/Weather_map">weather
map</a>, which plots current or
predicted weather patterns on top of a geographic map:</p>
<div class="float">
<img src="weather_map.png" alt="Weather maps are heatmaps"><div class="figcaption">Weather maps are heatmaps</div>
</div>
<p>The color at each grid coordinate, which in this case are spaced closely enough
so we cannot distinguish the boundaries between the rectangles without zooming
in, is proportional to the type and intensity of weather at that location.
In the weather map, the weather values are mapped to a <em>color scales</em> that
describe the amount of precipitation as shown in the map legend.</p>
<p>In biology, heatmaps are more typically used to visualize matrices. The concept
is the same as with the weather map, except instead of geographic coordinates
the grid corresponds to the rows and columns of a data matrix, and the color of
each grid point is mapped to a color scale chosen by the designer. The grid of
rectangles is typically not overlayed over anything, as in the weather map, but
is instead the entire visualization.</p>
<p>We can create heatmap in R using the base R
<a href="https://www.rdocumentation.org/packages/stats/versions/3.6.2/topics/heatmap"><code>heatmap()</code></a>
function. Below is a heatmap visualization of our histological markers from the
AD example:</p>
<div class="sourceCode" id="cb305"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># heatmap() requires a R matrix, and cannot accept a tibble or a dataframe</span></span>
<span><span class="va">marker_matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span><span class="op">(</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="va">ad_metadata</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">tau</span>,<span class="va">abeta</span>,<span class="va">iba1</span>,<span class="va">gfap</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="co"># rownames of the matrix become y labels</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">marker_matrix</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">ad_metadata</span><span class="op">$</span><span class="va">ID</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/stats/heatmap.html">heatmap</a></span><span class="op">(</span><span class="va">marker_matrix</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="main_files/data%20viz%20marker%20heatmap-1.png" width="672"></div>
<p>The lower right portion of the plot is the heatmap, where the matrix values are
visualized on a color gradient from light yellow to dark red.</p>
<p>More precisely, the <code><a href="https://rdrr.io/r/stats/heatmap.html">heatmap()</a></code> function creates a <em>clustered heatmap</em>, where
the rows and columns have been <a href="data-science.html#hierarchical-clustering">hierarchically
clustered</a> separately and ordered according to how
similar they are to one another. The left and top line diagrams are
<a href="data-visualization.html#dendrograms">dendrograms</a>, which depict the similarity between rows and
columns as a tree, where the total branch length summed from one row/column to
another is proportional to the dissimilarity (i.e. numeric distance) of the two.</p>
<div class="box note">
<p>The base R <code><a href="https://rdrr.io/r/stats/heatmap.html">heatmap()</a></code> function performs many different operations on the input
matrix than just draw a grid of rectangles in proportion to the values. By
default, it also:</p>
<ul>
<li>performs <a href="data-science.html#hierarchical-clustering">hierarchical clustering</a> of the rows and
columns using a Euclidean distance function and orders them accordingly</li>
<li>draws <a href="#dendrogram">dendrograms</a> on the rows and columns according to the
clustering</li>
<li>scales the data in the rows to have mean zero and standard deviation 1</li>
</ul>
<p>Each of these defaults can be changed by passing arguments into the function
call. The following turns off all of the extra functionality and produces only a
heatmap of the matrix:</p>
<div class="sourceCode" id="cb306"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/stats/heatmap.html">heatmap</a></span><span class="op">(</span></span>
<span>  <span class="va">marker_matrix</span>,</span>
<span>  Rowv<span class="op">=</span><span class="cn">NA</span>,</span>
<span>  Colv<span class="op">=</span><span class="cn">NA</span>,</span>
<span>  scale<span class="op">=</span><span class="st">"none"</span>,</span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="main_files/data%20viz%20pure%20heatmap-1.png" width="672">
Note how the colors in this heatmap are very different for many of the
marker/sample pairs. This illustrates some of the dangers of using heatmaps,
which are described more fully in the <a href="data-visualization.html#how-to-use-heatmaps-responsibly">How To Use Heatmaps Responsibly</a> section
below.</p>
</div>
<p>The base R heatmap function has the major drawback that no color key is provided
to indicate how values map to colors. Another heatmap
<a href="https://www.rdocumentation.org/packages/gplots/versions/3.1.1/topics/heatmap.2"><code>heatmap.2</code></a>
function in the <a href="https://cran.r-project.org/web/packages/gplots/index.html"><code>gplots</code> package</a>
which has a similar interface to <code><a href="https://rdrr.io/r/stats/heatmap.html">heatmap()</a></code>, but allows provides more
parameters to control the behavior of the plot and includes a color key:</p>
<div class="sourceCode" id="cb307"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/talgalili/gplots">gplots</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/gplots/man/heatmap.2.html">heatmap.2</a></span><span class="op">(</span><span class="va">marker_matrix</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="main_files/heatmap.2-1.png" width="672"></div>
<p>The extra decoration (dashed and solid vertical lines called <em>traces</em> by the
package) provide another way to understand the magnitude of the value in each
cell. Refer to the <a href="https://www.rdocumentation.org/packages/gplots/versions/3.1.1/topics/heatmap.2"><code>gplots</code> documentation of
<code>heatmap2</code></a>
for more information on how to interpret the trace (hint: you may turn it off
with by passing the argument <code>trace = "none"</code> to the function call).</p>
<p>Both <code><a href="https://rdrr.io/r/stats/heatmap.html">heatmap()</a></code> and <code><a href="https://rdrr.io/pkg/gplots/man/heatmap.2.html">heatmap.2()</a></code> provide a useful method to annotate rows and
columns with a categorical variable. Each of our subjects either has AD or is a
control, and we can add a color bar along the margin of the plot to indicate the
status of each subject:</p>
<div class="sourceCode" id="cb308"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">condition_colors</span> <span class="op">&lt;-</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/transmute.html">transmute</a></span><span class="op">(</span></span>
<span>    <span class="va">ad_metadata</span>,</span>
<span>    color<span class="op">=</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/if_else.html">if_else</a></span><span class="op">(</span><span class="va">condition</span> <span class="op">==</span> <span class="st">"AD"</span>,<span class="st">"red"</span>,<span class="st">"blue"</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/stats/heatmap.html">heatmap</a></span><span class="op">(</span></span>
<span>  <span class="va">marker_matrix</span>,</span>
<span>  RowSideColors<span class="op">=</span><span class="va">condition_colors</span><span class="op">$</span><span class="va">color</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="main_files/heatmap%20margin-1.png" width="672"></div>
<p>And with <code>heatmap.2</code>:</p>
<div class="sourceCode" id="cb309"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/gplots/man/heatmap.2.html">heatmap.2</a></span><span class="op">(</span></span>
<span>  <span class="va">marker_matrix</span>,</span>
<span>  RowSideColors<span class="op">=</span><span class="va">condition_colors</span><span class="op">$</span><span class="va">color</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="main_files/heatmap.2%20margin-1.png" width="672"></div>
<p>Heatmaps may also be created using ggplot and the
<a href="https://ggplot2.tidyverse.org/reference/geom_tile.html"><code>geom_tile</code></a> geometry.
This method of generating heatmaps is more “manual” than the other functions
covered so far, and so are both flexible and require more work to obtain certain
functionality (e.g. clustering, color bars on side margins, etc). This geometry
requires the data to be in long format, with one column for <code>x</code>, <code>y</code>, and <code>z</code>
values:</p>
<div class="sourceCode" id="cb310"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">pivot_longer</span><span class="op">(</span></span>
<span>  <span class="va">ad_metadata</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">tau</span>,<span class="va">abeta</span>,<span class="va">iba1</span>,<span class="va">gfap</span><span class="op">)</span>,</span>
<span>  names_to<span class="op">=</span><span class="st">"Marker"</span>,</span>
<span>  values_to<span class="op">=</span><span class="st">"Intensity"</span></span>
<span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">Marker</span>,y<span class="op">=</span><span class="va">ID</span>,fill<span class="op">=</span><span class="va">Intensity</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_tile.html">geom_tile</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="main_files/heatmap%20geom%20tile-1.png" width="672"></div>
<p>Note this ggplot method does not scale or reorder rows or columns.</p>
<div id="specifying-heatmap-colors" class="section level4" number="7.5.6.1">
<h4>
<span class="header-section-number">7.5.6.1</span> Specifying Heatmap Colors<a class="anchor" aria-label="anchor" href="#specifying-heatmap-colors"><i class="fas fa-link"></i></a>
</h4>
<p>The colors of the heatmap may be changed by passing one of the <a href="https://stat.ethz.ch/R-manual/R-patched/library/grDevices/html/palettes.html">native color
palettes</a>
to the heatmap function with the <code>col</code> argument:</p>
<div class="sourceCode" id="cb311"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># native R colors are:</span></span>
<span><span class="co"># - rainbow(n, start=.7, end=.1)</span></span>
<span><span class="co"># - heat.colors(n)</span></span>
<span><span class="co"># - terrain.colors(n)</span></span>
<span><span class="co"># - topo.colors(n)</span></span>
<span><span class="co"># - cm.colors(n)</span></span>
<span><span class="co"># the n argument specifies the number of colors (i.e. resolution) of the colormap to return</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/stats/heatmap.html">heatmap</a></span><span class="op">(</span><span class="va">marker_matrix</span>,col<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/grDevices/palettes.html">cm.colors</a></span><span class="op">(</span><span class="fl">256</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="main_files/heatmap%20cm%20colors-1.png" width="672"></div>
<p>To change the color using ggplot and <code><a href="https://ggplot2.tidyverse.org/reference/geom_tile.html">geom_tile()</a></code>, use the
<a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html"><code>scale_fill_gradientn</code></a>
function to specify a different color palette:</p>
<div class="sourceCode" id="cb312"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">pivot_longer</span><span class="op">(</span></span>
<span>  <span class="va">ad_metadata</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">tau</span>,<span class="va">abeta</span>,<span class="va">iba1</span>,<span class="va">gfap</span><span class="op">)</span>,</span>
<span>  names_to<span class="op">=</span><span class="st">"Marker"</span>,</span>
<span>  values_to<span class="op">=</span><span class="st">"Intensity"</span></span>
<span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">Marker</span>,y<span class="op">=</span><span class="va">ID</span>,fill<span class="op">=</span><span class="va">Intensity</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_tile.html">geom_tile</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html">scale_fill_gradientn</a></span><span class="op">(</span>colors<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/grDevices/palettes.html">cm.colors</a></span><span class="op">(</span><span class="fl">256</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure">
<img src="main_files/heatmap%20geom%20tile%20color%20change-1.png" width="672">
The native color palettes in R look as follows:</div>
<div class="inline-figure"><img src="main_files/data%20viz%20heatmap%20native%20palettes-1.png" width="672"></div>
<p>Instead of these color palettes, you may use one of the
<a href="https://colorbrewer2.org/#type=sequential&amp;scheme=BuGn&amp;n=3">ColorBrewer</a>
palettes in the <a href="https://cran.r-project.org/web/packages/RColorBrewer/index.html">RColorBrewer
package</a></p>
<div class="sourceCode" id="cb313"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">RColorBrewer</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/RColorBrewer/man/ColorBrewer.html">display.brewer.all</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="main_files/heatmap%20colorbrewer%20palettes-1.png" width="672"></div>
<p>There are very many color palettes to choose from, and you may also create your
own palettes. Creating color palettes is <a href="https://www.r-bloggers.com/2016/07/creating-color-palettes-in-r/">somewhat
complicated</a>
in base R, but <code><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot()</a></code> provides <a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html">convenient
functions</a>
<code><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html">scale_fill_gradient()</a></code> and <code><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html">scale_fill_gradient2()</a></code> to produce gradient scales
with 2 or 3 colors, respectively. More discussion on how to choose colors is
described in the next section.</p>
</div>
</div>
<div id="how-to-use-heatmaps-responsibly" class="section level3" number="7.5.7">
<h3>
<span class="header-section-number">7.5.7</span> How To Use Heatmaps Responsibly<a class="anchor" aria-label="anchor" href="#how-to-use-heatmaps-responsibly"><i class="fas fa-link"></i></a>
</h3>
<p>While heatmaps may seem intuitive, they are actually quite complicated and can
be difficult to create in a way that accurately depicts the underlying matrix.
As described in the <a href="data-visualization.html#human-visual-perception">Human Visual Perception</a> section, our visual system
estimates the hue and value of colors based on adjacent colors. For large
heatmaps with many rows and columns, it is therefore very difficult to
accurately estimate relationships across wide sections of the heatmap. This may
not always be problematic, as heatmaps might only be needed to convey general
patterns in a dataset. However, the designer has a large degree of control over
how the heatmap utilizes hue and value when mapping to the underlying numbers,
and those choices have a large impact on the interpretation of the chart. Before
data can be effectively visualized in a heatmap, some preprocessing is usually
required to yield a useful plot.</p>
<p>There are four major factors that influence how a dataset can be visualized as
a heatmap:</p>
<ol style="list-style-type: decimal">
<li>The type of features, i.e. whether the features are continuous or discrete</li>
<li>The relative scales of the features</li>
<li>The total range of the data</li>
<li>Whether or not the data are centered</li>
</ol>
<div id="the-type-of-the-features" class="section level4" number="7.5.7.1">
<h4>
<span class="header-section-number">7.5.7.1</span> The Type of the Features<a class="anchor" aria-label="anchor" href="#the-type-of-the-features"><i class="fas fa-link"></i></a>
</h4>
<p>There are two broad types of features: discrete and continuous. The data
elements of a discrete feature take on one of a finite set of values. For
example, a discrete feature might include values like 0 and 1, <code>Case</code> and
<code>Control</code> or only the integers between 1 and 6. In general, we are primarily
interested to know whether samples with the same discrete value group together,
so ideally all colors that are mapped to discrete features will be easily
distinguishable from one another. As the number of possible values the feature
may take increases, finding a set of colors that are all distinguishable from
one another becomes increasingly difficult. Consider the following three
heatmaps that show different discrete data:</p>
<div class="sourceCode" id="cb314"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com">patchwork</a></span><span class="op">)</span></span>
<span><span class="va">random_letters</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">n</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">letters</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'A'</span>,<span class="st">'B'</span>,<span class="st">'C'</span>,<span class="st">'D'</span>,<span class="st">'E'</span>,<span class="st">'F'</span>,<span class="st">'G'</span>,<span class="st">'H'</span>,<span class="st">'I'</span>,<span class="st">'J'</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample</a></span><span class="op">(</span><span class="va">letters</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="va">n</span><span class="op">]</span>,<span class="fl">10</span>,replace<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="va">features</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">'F'</span>,<span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq</a></span><span class="op">(</span><span class="fl">10</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">g_binary</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html">tibble</a></span><span class="op">(</span></span>
<span>  feature<span class="op">=</span><span class="va">features</span>,</span>
<span>  a<span class="op">=</span><span class="fu">random_letters</span><span class="op">(</span><span class="fl">2</span><span class="op">)</span>,</span>
<span>  b<span class="op">=</span><span class="fu">random_letters</span><span class="op">(</span><span class="fl">2</span><span class="op">)</span>,</span>
<span>  c<span class="op">=</span><span class="fu">random_letters</span><span class="op">(</span><span class="fl">2</span><span class="op">)</span></span>
<span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">pivot_longer</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">a</span>,<span class="va">b</span>,<span class="va">c</span><span class="op">)</span><span class="op">)</span>  <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">name</span>,y<span class="op">=</span><span class="va">feature</span>,fill<span class="op">=</span><span class="va">value</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_tile.html">geom_tile</a></span><span class="op">(</span><span class="op">)</span></span>
<span>  </span>
<span><span class="va">g_trinary</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html">tibble</a></span><span class="op">(</span></span>
<span>  feature<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">'F'</span>,<span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq</a></span><span class="op">(</span><span class="fl">10</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  a<span class="op">=</span><span class="fu">random_letters</span><span class="op">(</span><span class="fl">3</span><span class="op">)</span>,</span>
<span>  b<span class="op">=</span><span class="fu">random_letters</span><span class="op">(</span><span class="fl">3</span><span class="op">)</span>,</span>
<span>  c<span class="op">=</span><span class="fu">random_letters</span><span class="op">(</span><span class="fl">3</span><span class="op">)</span></span>
<span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">pivot_longer</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">a</span>,<span class="va">b</span>,<span class="va">c</span><span class="op">)</span><span class="op">)</span>  <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">name</span>,y<span class="op">=</span><span class="va">feature</span>,fill<span class="op">=</span><span class="va">value</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_tile.html">geom_tile</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">g_unmanageable</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html">tibble</a></span><span class="op">(</span></span>
<span>  feature<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">'F'</span>,<span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq</a></span><span class="op">(</span><span class="fl">10</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  a<span class="op">=</span><span class="fu">random_letters</span><span class="op">(</span><span class="fl">10</span><span class="op">)</span>,</span>
<span>  b<span class="op">=</span><span class="fu">random_letters</span><span class="op">(</span><span class="fl">10</span><span class="op">)</span>,</span>
<span>  c<span class="op">=</span><span class="fu">random_letters</span><span class="op">(</span><span class="fl">10</span><span class="op">)</span></span>
<span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">pivot_longer</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">a</span>,<span class="va">b</span>,<span class="va">c</span><span class="op">)</span><span class="op">)</span>  <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">name</span>,y<span class="op">=</span><span class="va">feature</span>,fill<span class="op">=</span><span class="va">value</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_tile.html">geom_tile</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">g_binary</span> <span class="op">|</span> <span class="va">g_trinary</span> <span class="op">|</span> <span class="va">g_unmanageable</span></span></code></pre></div>
<div class="inline-figure"><img src="main_files/viz%20discrete%20heatmap-1.png" width="672"></div>
<p>The first and second heatmaps have 2 and 3 discrete values each, and the colors
are easy to tell apart. The third heatmap has 10 discrete values and it is very
difficult to distinguish between colors in different parts of the plot.</p>
<p>Continuous features can take any of a (possibly infinite) range of real values,
e.g. <span class="math inline">\(-3.252\)</span>, <span class="math inline">\(858,352\)</span>, or <span class="math inline">\(0\)</span>. These values will be mapped to colors of a
<em>gradient</em>, where two or more colors are blended together linearly from one
color to another, and values are mapped within this range of colors in
proportion to their relative magnitudes. Two gradients, one with two colors and
one with three, are shown below:</p>
<div class="sourceCode" id="cb315"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">g_two_up</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html">tibble</a></span><span class="op">(</span></span>
<span>  ID<span class="op">=</span><span class="st">'a'</span>,</span>
<span>  x<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">1</span>,<span class="fl">0.1</span><span class="op">)</span>,</span>
<span>  val<span class="op">=</span><span class="va">x</span></span>
<span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">x</span>,y<span class="op">=</span><span class="va">ID</span>,fill<span class="op">=</span><span class="va">val</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_tile.html">geom_tile</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html">scale_fill_gradient</a></span><span class="op">(</span>low<span class="op">=</span><span class="st">"#990000"</span>, high<span class="op">=</span><span class="st">"#ffdddd"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">g_two_down</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html">tibble</a></span><span class="op">(</span></span>
<span>  ID<span class="op">=</span><span class="st">'a'</span>,</span>
<span>  x<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">1</span>,<span class="fl">0.1</span><span class="op">)</span>,</span>
<span>  val<span class="op">=</span><span class="va">x</span></span>
<span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">x</span>,y<span class="op">=</span><span class="va">ID</span>,fill<span class="op">=</span><span class="va">val</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_tile.html">geom_tile</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html">scale_fill_gradient</a></span><span class="op">(</span>low<span class="op">=</span><span class="st">"#ffdddd"</span>, high<span class="op">=</span><span class="st">"#990000"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">g_three</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html">tibble</a></span><span class="op">(</span></span>
<span>  ID<span class="op">=</span><span class="st">'a'</span>,</span>
<span>  x<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">1</span>,<span class="fl">0.1</span><span class="op">)</span>,</span>
<span>  val<span class="op">=</span><span class="va">x</span></span>
<span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">x</span>,y<span class="op">=</span><span class="va">ID</span>,fill<span class="op">=</span><span class="va">val</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_tile.html">geom_tile</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html">scale_fill_gradient2</a></span><span class="op">(</span>low<span class="op">=</span><span class="st">"#990000"</span>, mid<span class="op">=</span><span class="st">"#ffffff"</span>, high<span class="op">=</span><span class="st">"#9999ff"</span>, midpoint<span class="op">=</span><span class="fl">0.5</span><span class="op">)</span></span>
<span></span>
<span><span class="va">g_two_up</span> <span class="op">/</span> <span class="va">g_two_down</span> <span class="op">/</span> <span class="va">g_three</span></span></code></pre></div>
<div class="inline-figure"><img src="main_files/viz%20heatmap%20continuous-1.png" width="672"></div>
<p>These three example gradients comprise the most common types of color mappings.
The first two with one color differ in which color (i.e. lighter or darker) is
mapped to the largest value. The third gradient is called a <em>diverging</em> palette,
because the maximum and minimum values show divergence from some central value,
which in this case is 0.5. For diverging color palettes,
the value in the middle of the gradient is often very important, as it defines
which values are visualized in which color. More on diverging color palettes is
in the following sections.</p>
<p>In general, the types of features, i.e. discrete or continuous, must all be the
same for a heatmap to be meaningful. Once this has been determined, the values
of each feature must be examined and possibly transformed, as described next.</p>
</div>
<div id="scales-of-features" class="section level4" number="7.5.7.2">
<h4>
<span class="header-section-number">7.5.7.2</span> Scales of Features<a class="anchor" aria-label="anchor" href="#scales-of-features"><i class="fas fa-link"></i></a>
</h4>
<p>The second important consideration when plotting a heatmap is to understand the
nature of the different features and how they relate to one another.
Specifically, the features must all be on a comparable scale, or transformed
appropriately to attain this, for the colors in the heatmap to be visible.
Consider the following heatmap with three features plotted as columns and 10
samples as rows:</p>
<div class="sourceCode" id="cb316"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html">tibble</a></span><span class="op">(</span></span>
<span>  ID<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">'F'</span>,<span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq</a></span><span class="op">(</span><span class="fl">10</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  a<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span><span class="op">(</span><span class="fl">10</span>,<span class="fl">0</span>,<span class="fl">1</span><span class="op">)</span>,</span>
<span>  b<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span><span class="op">(</span><span class="fl">10</span>,<span class="fl">100</span>,<span class="fl">20</span><span class="op">)</span>,</span>
<span>  c<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span><span class="op">(</span><span class="fl">10</span>,<span class="fl">20</span>,<span class="fl">5</span><span class="op">)</span></span>
<span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">pivot_longer</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">a</span>,<span class="va">b</span>,<span class="va">c</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">data</span>,<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">name</span>,y<span class="op">=</span><span class="va">ID</span>,fill<span class="op">=</span><span class="va">value</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_tile.html">geom_tile</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="main_files/viz%20heatmap%20mismatch%20scale-1.png" width="672"></div>
<p>The scales of the different features <span class="math inline">\(a\)</span>, <span class="math inline">\(b\)</span>, and <span class="math inline">\(c\)</span> are different, namely
they have means of 0, 100, and 20, respectively. This an be clearly seen when
the data are plotted as a <a href="beeswarm-plots">beeswarm plot</a>:</p>
<div class="sourceCode" id="cb317"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/eclarke/ggbeeswarm">ggbeeswarm</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/ggbeeswarm/man/geom_beeswarm.html">geom_beeswarm</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">name</span>,y<span class="op">=</span><span class="va">value</span>,color<span class="op">=</span><span class="va">name</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="main_files/viz%20heatmap%20violin-1.png" width="672"></div>
<p>As the data are on dramatically different scales, only the feature with the
largest spread (b) can be distinguished in the heatmap.</p>
<p>In order to visualize all these data meaningfully in the same plot, the features
must be <em>scaled</em> so they are all comparable. Typically this is done by
standardizing the features to have a mean of zero and a standard deviation of 1:</p>
<div class="sourceCode" id="cb318"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">pivot_wider</span><span class="op">(</span>id_cols<span class="op">=</span><span class="st">'ID'</span>,names_from<span class="op">=</span><span class="va">name</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/across.html">across</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">a</span>,<span class="va">b</span>,<span class="va">c</span><span class="op">)</span>,<span class="va">scale</span><span class="op">)</span></span>
<span>  <span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">pivot_longer</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">a</span>,<span class="va">b</span>,<span class="va">c</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">name</span>,y<span class="op">=</span><span class="va">ID</span>,fill<span class="op">=</span><span class="va">value</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_tile.html">geom_tile</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure">
<img src="main_files/viz%20heatmap%20standard-1.png" width="672">
Now each feature has a mean of zero and standard deviation of one:</div>
<div class="sourceCode" id="cb319"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">pivot_wider</span><span class="op">(</span>id_cols<span class="op">=</span><span class="st">'ID'</span>,names_from<span class="op">=</span><span class="va">name</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/across.html">across</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">a</span>,<span class="va">b</span>,<span class="va">c</span><span class="op">)</span>,<span class="va">scale</span><span class="op">)</span></span>
<span>  <span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">pivot_longer</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">a</span>,<span class="va">b</span>,<span class="va">c</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/ggbeeswarm/man/geom_beeswarm.html">geom_beeswarm</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">name</span>,y<span class="op">=</span><span class="va">value</span>,color<span class="op">=</span><span class="va">name</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="main_files/viz%20standard%20violin-1.png" width="672"></div>
<p>Now that the data are all on the same scale, each is mapped across the full
range of the color palette. In this case, we explicitly scaled each feature, but
it is possible to scale by each sample as well:</p>
<div class="sourceCode" id="cb320"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">pivot_wider</span><span class="op">(</span>id_cols<span class="op">=</span><span class="va">name</span>,names_from<span class="op">=</span><span class="va">ID</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/across.html">across</a></span><span class="op">(</span><span class="fu"><a href="https://tidyselect.r-lib.org/reference/starts_with.html">starts_with</a></span><span class="op">(</span><span class="st">'F'</span><span class="op">)</span>,<span class="va">scale</span><span class="op">)</span></span>
<span>  <span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">pivot_longer</span><span class="op">(</span><span class="fu"><a href="https://tidyselect.r-lib.org/reference/starts_with.html">starts_with</a></span><span class="op">(</span><span class="st">'F'</span><span class="op">)</span>,names_to<span class="op">=</span><span class="st">"ID"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">name</span>,y<span class="op">=</span><span class="va">ID</span>,fill<span class="op">=</span><span class="va">value</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_tile.html">geom_tile</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="main_files/viz%20heatmap%20scale%20row-1.png" width="672"></div>
<p>This result is dramatically different, and possibly not what you intended. In
this case the effect is somewhat obvious - all the values in b are the largest,
the values in c are the smallest, and c is in between. However, in general this
may not be the case, and care should be taken to understand which dimension of
the data has been scaled.</p>
<div class="box warning">
<p>Because features are typically not assumed to be on the same scale, the base R
<code><a href="https://rdrr.io/r/stats/heatmap.html">heatmap()</a></code> and <code>heatmap2()</code> functions <strong>automatically scale the rows of the
data prior to visualization</strong>.</p>
</div>
</div>
<div id="total-data-range" class="section level4" number="7.5.7.3">
<h4>
<span class="header-section-number">7.5.7.3</span> Total data range<a class="anchor" aria-label="anchor" href="#total-data-range"><i class="fas fa-link"></i></a>
</h4>
<p>As we saw in the previous section on scaling features, the color mapping is
highly sensitive to the range of all values in the data. Even if the features or
samples have been properly scaled, extreme values in any of the features can
render a heatmap uninformative. Consider the following heatmap that visualizes
data that have a single extreme outlier:</p>
<div class="sourceCode" id="cb321"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html">tibble</a></span><span class="op">(</span></span>
<span>  ID<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">'F'</span>,<span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq</a></span><span class="op">(</span><span class="fl">10</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  a<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span><span class="op">(</span><span class="fl">10</span>,<span class="fl">0</span>,<span class="fl">1</span><span class="op">)</span>,</span>
<span>  b<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span><span class="op">(</span><span class="fl">10</span>,<span class="fl">0</span>,<span class="fl">1</span><span class="op">)</span>,</span>
<span>  c<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span><span class="op">(</span><span class="fl">9</span>,<span class="fl">0</span>,<span class="fl">1</span><span class="op">)</span>,<span class="fl">1e9</span><span class="op">)</span></span>
<span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">pivot_longer</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">a</span>,<span class="va">b</span>,<span class="va">c</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">name</span>,y<span class="op">=</span><span class="va">ID</span>,fill<span class="op">=</span><span class="va">value</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_tile.html">geom_tile</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="main_files/viz%20outlier-1.png" width="672"></div>
<p>The single extreme value compressed all the other color values to the colors on
the lowest end of the color palette. This anomalous value must be removed before
differences between the other variables will be visible.</p>
<p>The issue above is caused by the fact that the default color gradient scales
assume there is a linear mapping between the values and the palette. This is
equivalent to assuming the data are relatively well distributed across a linear
range of values, as in a normal or uniform distribution. When the data do not
follow this assumption, any resulting heatmaps may suffer the same issue as
when there are extreme values. Consider the following heatmap with data randomly
sampled from an exponential distribution:</p>
<div class="sourceCode" id="cb322"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html">tibble</a></span><span class="op">(</span></span>
<span>  ID<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">'F'</span>,<span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq</a></span><span class="op">(</span><span class="fl">10</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  a<span class="op">=</span><span class="fl">10</span><span class="op">**</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span><span class="op">(</span><span class="fl">10</span>,<span class="fl">4</span>,<span class="fl">1</span><span class="op">)</span>,</span>
<span>  b<span class="op">=</span><span class="fl">10</span><span class="op">**</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span><span class="op">(</span><span class="fl">10</span>,<span class="fl">4</span>,<span class="fl">1</span><span class="op">)</span>,</span>
<span>  c<span class="op">=</span><span class="fl">10</span><span class="op">**</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span><span class="op">(</span><span class="fl">10</span>,<span class="fl">4</span>,<span class="fl">1</span><span class="op">)</span></span>
<span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">pivot_longer</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">a</span>,<span class="va">b</span>,<span class="va">c</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">data</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">name</span>,y<span class="op">=</span><span class="va">ID</span>,fill<span class="op">=</span><span class="va">value</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_tile.html">geom_tile</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="main_files/viz%20exponential-1.png" width="672"></div>
<p>The data are similarly compressed toward the lower end of the scale due to the
uneven distribution of data. In this case, we might consider taking a log
transformation of the data prior to drawing the heatmap:</p>
<div class="sourceCode" id="cb323"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span><span class="va">data</span>,</span>
<span>       value<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="va">value</span><span class="op">)</span></span>
<span>       <span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">name</span>,y<span class="op">=</span><span class="va">ID</span>,fill<span class="op">=</span><span class="va">value</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_tile.html">geom_tile</a></span><span class="op">(</span><span class="op">)</span> </span></code></pre></div>
<div class="inline-figure"><img src="main_files/viz%20heatmap%20log-1.png" width="672"></div>
<p>On a log scale, the heatmap visualizes the range of values linearly across the
gradient. In this case, we knew a log transform was appropriate to cast the
data onto an approximately linear scale, though in general this will need to be
explored for each dataset.</p>
</div>
<div id="whether-or-not-the-data-are-centered" class="section level4" number="7.5.7.4">
<h4>
<span class="header-section-number">7.5.7.4</span> Whether or not the data are centered<a class="anchor" aria-label="anchor" href="#whether-or-not-the-data-are-centered"><i class="fas fa-link"></i></a>
</h4>
<p>The final important consideration when creating a heatmap with continuous values
is determining if there is a meaningful center for each feature. For certain
kinds of data, data points above and below a certain value have a particular
meaning. For example, log2 fold change values from differential expression
greater or less than zero indicate a gene is increased or decreased,
respectively, in the comparison of interest. When visualizing these values as a
heatmap, it is therefore important to indicate this central value with a color,
usually white or black, appropriately, so that colors above and below this
value can be set accordingly, e.g. above zero is red, below zero blue.</p>
<p>To do this, first a 3-color diverging palette is required, i.e. a palette with
one color on one side, a different color on the other side, and a third color in
the middle, usually white or black. The following is a heatmap of normally
distributed values with mean 0 and standard deviation of 1 visualized on a
diverging color scale:</p>
<div class="sourceCode" id="cb324"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html">tibble</a></span><span class="op">(</span></span>
<span>  ID<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">'F'</span>,<span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq</a></span><span class="op">(</span><span class="fl">10</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  a<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span><span class="op">(</span><span class="fl">10</span>,<span class="fl">0</span>,<span class="fl">1</span><span class="op">)</span>,</span>
<span>  b<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span><span class="op">(</span><span class="fl">10</span>,<span class="fl">0</span>,<span class="fl">1</span><span class="op">)</span>,</span>
<span>  c<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span><span class="op">(</span><span class="fl">10</span>,<span class="fl">0</span>,<span class="fl">1</span><span class="op">)</span></span>
<span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">pivot_longer</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">a</span>,<span class="va">b</span>,<span class="va">c</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">name</span>,y<span class="op">=</span><span class="va">ID</span>,fill<span class="op">=</span><span class="va">value</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_tile.html">geom_tile</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html">scale_fill_gradient2</a></span><span class="op">(</span>low<span class="op">=</span><span class="st">"#000099"</span>, mid<span class="op">=</span><span class="st">"#ffffff"</span>, high<span class="op">=</span><span class="st">"#990000"</span>, midpoint<span class="op">=</span><span class="fl">0</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="main_files/viz-1.png" width="672"></div>
<p>Note the <code><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html">scale_fill_gradient2()</a></code> call, with <code>midpoint=0</code> argument. This
argument instructs the gradient to set the value of 0 to the middle color, which
is white. The midpoint of zero is default, but if your data are centered around
some other value, failing to specify this midpoint will produce misleading
visualizations:</p>
<div class="sourceCode" id="cb325"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html">tibble</a></span><span class="op">(</span></span>
<span>  ID<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">'F'</span>,<span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq</a></span><span class="op">(</span><span class="fl">10</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  a<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span><span class="op">(</span><span class="fl">10</span>,<span class="fl">1</span>,<span class="fl">1</span><span class="op">)</span>,</span>
<span>  b<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span><span class="op">(</span><span class="fl">10</span>,<span class="fl">1</span>,<span class="fl">1</span><span class="op">)</span>,</span>
<span>  c<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span><span class="op">(</span><span class="fl">10</span>,<span class="fl">1</span>,<span class="fl">1</span><span class="op">)</span></span>
<span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">pivot_longer</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">a</span>,<span class="va">b</span>,<span class="va">c</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">name</span>,y<span class="op">=</span><span class="va">ID</span>,fill<span class="op">=</span><span class="va">value</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_tile.html">geom_tile</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html">scale_fill_gradient2</a></span><span class="op">(</span>low<span class="op">=</span><span class="st">"#000099"</span>, mid<span class="op">=</span><span class="st">"#ffffff"</span>, high<span class="op">=</span><span class="st">"#990000"</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="main_files/viz%20improper%20midpoint-1.png" width="672"></div>
<p>Here our data are centered with a mean of 1, but the midpoint is still at its
default of zero. If we intend for 1 to be the midpoint corresponding to the
color white, we must specify it explicitly:</p>
<div class="sourceCode" id="cb326"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html">tibble</a></span><span class="op">(</span></span>
<span>  ID<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">'F'</span>,<span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq</a></span><span class="op">(</span><span class="fl">10</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  a<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span><span class="op">(</span><span class="fl">10</span>,<span class="fl">1</span>,<span class="fl">1</span><span class="op">)</span>,</span>
<span>  b<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span><span class="op">(</span><span class="fl">10</span>,<span class="fl">1</span>,<span class="fl">1</span><span class="op">)</span>,</span>
<span>  c<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span><span class="op">(</span><span class="fl">10</span>,<span class="fl">1</span>,<span class="fl">1</span><span class="op">)</span></span>
<span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">pivot_longer</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">a</span>,<span class="va">b</span>,<span class="va">c</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">name</span>,y<span class="op">=</span><span class="va">ID</span>,fill<span class="op">=</span><span class="va">value</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_tile.html">geom_tile</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html">scale_fill_gradient2</a></span><span class="op">(</span>low<span class="op">=</span><span class="st">"#000099"</span>, mid<span class="op">=</span><span class="st">"#ffffff"</span>, high<span class="op">=</span><span class="st">"#990000"</span>, midpoint<span class="op">=</span><span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="main_files/viz%20correct%20midpoint-1.png" width="672"></div>
<p>If your features do not have a meaningful center, then the color palette scale
should generally have only two colors that range from either light to dark or
dark to light:</p>
<div class="sourceCode" id="cb327"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html">tibble</a></span><span class="op">(</span></span>
<span>  ID<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">'F'</span>,<span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq</a></span><span class="op">(</span><span class="fl">10</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  a<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html">runif</a></span><span class="op">(</span><span class="fl">10</span>,<span class="fl">0</span>,<span class="fl">10</span><span class="op">)</span>,</span>
<span>  b<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html">runif</a></span><span class="op">(</span><span class="fl">10</span>,<span class="fl">0</span>,<span class="fl">10</span><span class="op">)</span>,</span>
<span>  c<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html">runif</a></span><span class="op">(</span><span class="fl">10</span>,<span class="fl">0</span>,<span class="fl">10</span><span class="op">)</span></span>
<span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">pivot_longer</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">a</span>,<span class="va">b</span>,<span class="va">c</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">g_up</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">data</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">name</span>,y<span class="op">=</span><span class="va">ID</span>,fill<span class="op">=</span><span class="va">value</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_tile.html">geom_tile</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html">scale_fill_gradient</a></span><span class="op">(</span>low<span class="op">=</span><span class="st">"#000099"</span>,high<span class="op">=</span><span class="st">"#eeeeff"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">g_dn</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">data</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">name</span>,y<span class="op">=</span><span class="va">ID</span>,fill<span class="op">=</span><span class="va">value</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_tile.html">geom_tile</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html">scale_fill_gradient</a></span><span class="op">(</span>low<span class="op">=</span><span class="st">"#ffffff"</span>,high<span class="op">=</span><span class="st">"#000099"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">g_up</span> <span class="op">|</span> <span class="va">g_dn</span></span></code></pre></div>
<div class="inline-figure"><img src="main_files/viz%20two%20color-1.png" width="672"></div>
<p>The direction of the color palette depends on the meaning of the data, and is
a subjective choice.</p>

</div>
</div>
</div>
<div id="other-kind-of-plots" class="section level2" number="7.6">
<h2>
<span class="header-section-number">7.6</span> Other Kind of Plots<a class="anchor" aria-label="anchor" href="#other-kind-of-plots"><i class="fas fa-link"></i></a>
</h2>
<div id="dendrograms" class="section level3" number="7.6.1">
<h3>
<span class="header-section-number">7.6.1</span> Dendrograms<a class="anchor" aria-label="anchor" href="#dendrograms"><i class="fas fa-link"></i></a>
</h3>
<p>Dendrograms are visualizations of tree structures (<em>dendron</em> means tree in
Greek). In the <a href="https://en.wikipedia.org/wiki/Graph_theory">graph theory</a> branch
of mathematics, a <a href="https://en.wikipedia.org/wiki/Tree_(graph_theory)">tree</a> is
defined as a graph composed of nodes and edges in which any two nodes are
connected by exactly one path. Such a graph is also called an <em>acyclic graph</em>,
where no paths exist that can be followed from a node back to iself. A full
treatment of trees and graph theory is beyond the scope of this book, but we
will discuss trees to the extent they are necessary to understand how to read a
dendrogram.</p>
<p>The concepts of a dendrogram are illustrated in the following figure:</p>
<div class="float">
<img src="dendrograms.png" alt="Dendrogram Illustration"><div class="figcaption">Dendrogram Illustration</div>
</div>
<p>The tree drawn in a typical dendrogram is created using the output of a
clustering algorithm, e.g. <a href="data-science.html#hierarchical-clustering">hierarchical clustering</a>
that groups data points together based on their dissimilarity. Briefly, in the
figure, the data are one dimensional scalars, so the distance between all points
is simply the absolute value of the difference between them. The two closest
points are then merged into a single group, and the pairwise distances are
recomputed for all remaining data points with a summary of this new group (the
group summarization method is specified by the user and is called the <a href="https://en.wikipedia.org/wiki/Hierarchical_clustering#Linkage_criteria">linkage
critera</a>).
Then the process repeats, where the two nearest data points or summarized groups
are merged into a new group, the new group is summarized, pairwise distances
recomputed, etc. until all the data points have been included in a group. In
this way, all data points are assigned to a <em>hierarchy</em> of groups. In general,
any hierarchical grouping of data can be used to generate a dendrogram.</p>
<p>Given a hierarchical clustering result, a dendrogram can be drawn in several
ways in R. Using our AD marker example data, we hierarchically cluster the
subjects based on the content of all their markers and visualize the result as
follows:</p>
<div class="sourceCode" id="cb328"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/andrie/ggdendro">ggdendro</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># produce a clustering of the data using the hclust for hierarchical clustering</span></span>
<span><span class="co"># and euclidean distance as the distance metric</span></span>
<span><span class="va">euc_dist</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/dist.html">dist</a></span><span class="op">(</span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="va">ad_metadata</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">tau</span>,<span class="va">abeta</span>,<span class="va">iba1</span>,<span class="va">gfap</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">hc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/hclust.html">hclust</a></span><span class="op">(</span><span class="va">euc_dist</span>, method<span class="op">=</span><span class="st">"ave"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># add ID as labels to the clustering object</span></span>
<span><span class="va">hc</span><span class="op">$</span><span class="va">labels</span> <span class="op">&lt;-</span> <span class="va">ad_metadata</span><span class="op">$</span><span class="va">ID</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ggdendro/man/ggdendrogram.html">ggdendrogram</a></span><span class="op">(</span><span class="va">hc</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="main_files/dendro%20full-1.png" width="672"></div>
<p>This dendrogram does not produce highly distinct clusters of AD and control
samples with all the data.</p>
<p>To illustrate what a dataset with strong clustering looks like, we draw
multivariate samples from two normal distributions and cluster the results:</p>
<div class="sourceCode" id="cb329"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com">patchwork</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/andrie/ggdendro">ggdendro</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">well_clustered_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html">tibble</a></span><span class="op">(</span></span>
<span>  ID<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_c.html">str_c</a></span><span class="op">(</span><span class="st">"A"</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span>,<span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_c.html">str_c</a></span><span class="op">(</span><span class="st">"B"</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  f1<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span><span class="op">(</span><span class="fl">10</span>,<span class="fl">0</span>,<span class="fl">1</span><span class="op">)</span>,<span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span><span class="op">(</span><span class="fl">10</span>,<span class="fl">10</span>,<span class="fl">1</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  f2<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span><span class="op">(</span><span class="fl">10</span>,<span class="fl">0</span>,<span class="fl">1</span><span class="op">)</span>,<span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span><span class="op">(</span><span class="fl">10</span>,<span class="fl">10</span>,<span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">scatter_g</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">well_clustered_data</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">f1</span>,y<span class="op">=</span><span class="va">f2</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># produce a clustering of the data using the hclust for hierarchical clustering</span></span>
<span><span class="co"># and euclidean distance as the distance metric</span></span>
<span><span class="va">euc_dist</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/dist.html">dist</a></span><span class="op">(</span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="va">well_clustered_data</span>,<span class="op">-</span><span class="va">ID</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">hc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/hclust.html">hclust</a></span><span class="op">(</span><span class="va">euc_dist</span>, method<span class="op">=</span><span class="st">"ave"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># add ID as labels to the clustering object</span></span>
<span><span class="va">hc</span><span class="op">$</span><span class="va">labels</span> <span class="op">&lt;-</span> <span class="va">well_clustered_data</span><span class="op">$</span><span class="va">ID</span></span>
<span></span>
<span><span class="va">dendro_g</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ggdendro/man/ggdendrogram.html">ggdendrogram</a></span><span class="op">(</span><span class="va">hc</span><span class="op">)</span></span>
<span><span class="va">scatter_g</span> <span class="op">|</span> <span class="va">dendro_g</span></span></code></pre></div>
<div class="inline-figure"><img src="main_files/dendro%20strong-1.png" width="672"></div>
<p>Here, all the A and B samples strongly cluster together, with a large distance
between clusters.</p>
<div class="box readmore">
<ul>
<li>
<a href="https://en.wikipedia.org/wiki/Dendrogram">Dendrograms</a> Wikipedia page</li>
<li><a href="https://cran.r-project.org/web/packages/ggdendro/vignettes/ggdendro.html"><code>ggdendro</code> package vignette</a></li>
</ul>
</div>
</div>
<div id="chord-diagrams-and-circos-plots" class="section level3" number="7.6.2">
<h3>
<span class="header-section-number">7.6.2</span> Chord Diagrams and Circos Plots<a class="anchor" aria-label="anchor" href="#chord-diagrams-and-circos-plots"><i class="fas fa-link"></i></a>
</h3>
<p><a href="http://circos.ca/">Circos</a> is a software package that creates circular
diagrams originally designed to depict genomic features and data. The software
is capable of making beautiful, information dense figures. The Circos software
itself is written in the <a href="https://www.perl.org/">Perl</a> programming language, but
the <a href="https://jokergoo.github.io/circlize_book/book/"><code>circlize</code> package</a> brings
these capabilities to R. The following is a basic circos plot created by the
<code>circlize</code> package:</p>
<div class="float">
<img src="circlize-glance-track-1-1.png" alt="Basic circlize plot"><div class="figcaption">Basic circlize plot</div>
</div>
<p>The circle is divided up into 8 <em>sectors</em> here labeled with letters. There is a
single track added to the plot, where each sector has a 2-dimensional random
dataset plotted within the arc of the sector’s track box. Many tracks may be
added to the same plot:</p>
<div class="float">
<img src="circlize-glance-track-links-1.png" alt="Multiple tracks in a circlize plot"><div class="figcaption">Multiple tracks in a circlize plot</div>
</div>
<p>The arches inside the circle are called <em>chords</em> and are drawn to indicate there
is a relationship between the coordinates of different sectors. Diagrams with
such chords between sectors are called <em>chord diagrams</em>, and some plots are
composed only of coords:</p>
<div class="float">
<img src="chord-diagram-basic-1.png" alt="circlize chord diagram"><div class="figcaption">circlize chord diagram</div>
</div>
<p>These chord diagrams are often used in genomics studies to illustrate
chromosomal rearrangements:</p>
<div class="float">
<img src="genomic-links-1.png" alt="genomic links example"><div class="figcaption">genomic links example</div>
</div>
<p>Read the <a href="https://jokergoo.github.io/circlize_book/book/">circlize documentation</a>
for more information on how to use this very feature rich package.</p>

</div>
</div>
<div id="multiple-plots" class="section level2" number="7.7">
<h2>
<span class="header-section-number">7.7</span> Multiple Plots<a class="anchor" aria-label="anchor" href="#multiple-plots"><i class="fas fa-link"></i></a>
</h2>
<p>Sometimes it is convenient to plot multiple sets of data at once. This can be
accomplished in two ways: facet wrapping to separate out different subsets of a
dataset into plots with shared axes, and adding multiple plots to the same
figure.</p>
<div id="facet-wrapping" class="section level3" number="7.7.1">
<h3>
<span class="header-section-number">7.7.1</span> Facet wrapping<a class="anchor" aria-label="anchor" href="#facet-wrapping"><i class="fas fa-link"></i></a>
</h3>
<p>Facet wrapping separates subsets of a dataset into plots with identical axes.
This can be useful if different groups of data points overlap and occlude one
another. Consider the following dataset with three groups of normally
distributed random samples:</p>
<div class="sourceCode" id="cb330"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://mvtnorm.R-forge.R-project.org">mvtnorm</a></span><span class="op">)</span> <span class="co"># package implementing multivariate normal distributions</span></span>
<span><span class="va">nsamp</span> <span class="op">&lt;-</span> <span class="fl">100</span></span>
<span><span class="va">data</span> <span class="op">&lt;-</span>  <span class="fu"><a href="https://rdrr.io/r/base/cbind.html">rbind</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/mvtnorm/man/Mvnorm.html">rmvnorm</a></span><span class="op">(</span><span class="va">nsamp</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">1</span><span class="op">)</span>,sigma<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">0.8</span>,<span class="fl">0.8</span>,<span class="fl">1</span><span class="op">)</span>,nrow<span class="op">=</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/mvtnorm/man/Mvnorm.html">rmvnorm</a></span><span class="op">(</span><span class="va">nsamp</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">1</span><span class="op">)</span>,sigma<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="op">-</span><span class="fl">0.8</span>,<span class="op">-</span><span class="fl">0.8</span>,<span class="fl">1</span><span class="op">)</span>,nrow<span class="op">=</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/mvtnorm/man/Mvnorm.html">rmvnorm</a></span><span class="op">(</span><span class="va">nsamp</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">1</span><span class="op">)</span>,sigma<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">0</span>,<span class="fl">0</span>,<span class="fl">1</span><span class="op">)</span>,nrow<span class="op">=</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'x'</span>,<span class="st">'y'</span><span class="op">)</span></span>
<span><span class="va">g_oneplot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html">as_tibble</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span></span>
<span>    sample_name<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="st">'A'</span>,<span class="va">nsamp</span><span class="op">)</span>,<span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="st">'B'</span>,<span class="va">nsamp</span><span class="op">)</span>,<span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="st">'C'</span>,<span class="va">nsamp</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">x</span>,y<span class="op">=</span><span class="va">y</span>,color<span class="op">=</span><span class="va">sample_name</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">g_oneplot</span></span></code></pre></div>
<div class="inline-figure"><img src="main_files/viz%20facet%20oneplot-1.png" width="672"></div>
<p>The three sample data points are plotted on top of each other, and although they
are colored differently, it is difficult to discern the patterns of each group
clearly. This same plot may be split into three, one for each sample group,
using the <a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html"><code>facet_wrap()</code></a>
function:</p>
<div class="sourceCode" id="cb331"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">g_oneplot</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/vars.html">vars</a></span><span class="op">(</span><span class="va">sample_name</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="main_files/viz%20facet%20wrap-1.png" width="672"></div>
<p>Now the trends of each group are clearly visible. The <code><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap()</a></code> function
accepts one required argument, which is the variable name in the data that
should be used to separate the groups decorated by the <code><a href="https://ggplot2.tidyverse.org/reference/vars.html">vars()</a></code> function. Any
geometry may be used with a facet wrap. For example, we could add a
<code><a href="https://ggplot2.tidyverse.org/reference/geom_smooth.html">geom_smooth()</a></code> to the plot to estimate a smoothing function within each facet:</p>
<div class="sourceCode" id="cb332"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">g_oneplot</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/vars.html">vars</a></span><span class="op">(</span><span class="va">sample_name</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_smooth.html">geom_smooth</a></span><span class="op">(</span>method<span class="op">=</span><span class="st">"loess"</span>, formula<span class="op">=</span><span class="va">y</span> <span class="op">~</span> <span class="va">x</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="main_files/viz%20facet%20wrap%20plus-1.png" width="672"></div>
<p>By default, facet wrapping fixes all the axes of each plot to be identical,
which enables the most accurate estimates of comparisons between plots according
to the <a href="data-visualization.html#elementary-perceptual-tasks">Elementary Perceptual Tasks</a> hierarchy. This and many other aspects of
the faceting may be adjusted using arguments as described in the
<a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">documentation</a>.</p>
</div>
<div id="multipanel-figures" class="section level3" number="7.7.2">
<h3>
<span class="header-section-number">7.7.2</span> Multipanel Figures<a class="anchor" aria-label="anchor" href="#multipanel-figures"><i class="fas fa-link"></i></a>
</h3>
<p>Facet wrapping is useful when all of the data for the separate plots are on the
same scale, and comparisons of the facet groups is meaningful. However, faceting
is less convenient when we desire to place plots with unrelated data in the same
figure. The <a href="https://patchwork.data-imaginist.com/"><code>patchwork</code></a> package allows
us to compose multiple plots together using an intuitive set of operators. For
example, we may put two plots next to each other by separating them with the <code>|</code>
operator:</p>
<div class="sourceCode" id="cb333"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html">tibble</a></span><span class="op">(</span></span>
<span>  a<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span><span class="op">(</span><span class="fl">100</span>,<span class="fl">0</span>,<span class="fl">1</span><span class="op">)</span>,</span>
<span>  b<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span><span class="op">(</span><span class="fl">100</span>,<span class="fl">3</span>,<span class="fl">2</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">g_scatter</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">data</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">a</span>, y<span class="op">=</span><span class="va">b</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">g_violin</span> <span class="op">&lt;-</span> <span class="fu">pivot_longer</span><span class="op">(</span><span class="va">data</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">a</span>,<span class="va">b</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">name</span>,y<span class="op">=</span><span class="va">value</span>,fill<span class="op">=</span><span class="va">name</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_violin.html">geom_violin</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">g_scatter</span> <span class="op">|</span> <span class="va">g_violin</span></span></code></pre></div>
<div class="inline-figure"><img src="main_files/viz%20patch%20simple-1.png" width="672"></div>
<p>Note that each plot was saved to its own variable and then composed with the
patchwork operator <code>|</code>, which puts the plots side-by-side. We could also place
one plot on top of the other with the <code>/</code> operator:</p>
<div class="sourceCode" id="cb334"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">g_scatter</span> <span class="op">/</span> <span class="va">g_violin</span></span></code></pre></div>
<div class="inline-figure"><img src="main_files/viz%20patchwork%20vert-1.png" width="672"></div>
<p>Plots may be composed into arbitrarily complex multipanel figures in this way
using <code>()</code> to group plots together:</p>
<div class="sourceCode" id="cb335"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">g_scatter</span> <span class="op">/</span> <span class="op">(</span> <span class="va">g_scatter</span> <span class="op">|</span> <span class="va">g_violin</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="main_files/patch%20complex-1.png" width="672"></div>
<p>Another possible composition:</p>
<div class="sourceCode" id="cb336"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">g_scatter</span> <span class="op">/</span> <span class="va">g_scatter</span> <span class="op">)</span> <span class="op">|</span> <span class="va">g_violin</span></span></code></pre></div>
<div class="inline-figure"><img src="main_files/viz%20complex%202-1.png" width="672"></div>
<p>The patchwork library is very flexible, and allows the designer to compose plots
in many different ways. These multipanel plots make excellent starting points
for producing <a href="publication-ready-plots">publication-quality figures</a>, as
described in the next section.</p>

</div>
</div>
<div id="publication-ready-plots" class="section level2" number="7.8">
<h2>
<span class="header-section-number">7.8</span> Publication Ready Plots<a class="anchor" aria-label="anchor" href="#publication-ready-plots"><i class="fas fa-link"></i></a>
</h2>
<p>ggplot is powerful and flexible tool for creating plots, but it falls a little
short of producing plots directly suitable for publication. This is primarily
due to the default styling provided by the package; standard ggplot formatting
is effective for examining plots on screens, but some elements, such as the
default grey plot background, are unnecessary and sometimes inappropriate for
polished plots in printed journals.</p>
<p>There are two ways to take ggplots the final step to publication quality
figures. The first is by using
<a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">themes</a> provided by the
ggplot package and customizing specific elements to taste. The second is to
export plots to file in <a href="https://en.wikipedia.org/wiki/Scalable_Vector_Graphics">Scalable Vector
Graphics</a> format, which
can then be imported into programs like <a href="https://www.adobe.com/products/illustrator.html">Adobe
Illustrator</a> or
<a href="https://inkscape.org/">Inkscape</a> to be edited further.</p>
<div id="ggplot-themes" class="section level3" number="7.8.1">
<h3>
<span class="header-section-number">7.8.1</span> <code>ggplot</code> Themes<a class="anchor" aria-label="anchor" href="#ggplot-themes"><i class="fas fa-link"></i></a>
</h3>
<p>Themes in ggplot are combinations of styling elements applied to the different
components of a chart that include everything except the display and styling of
data, which is controlled using geometry and aesthetics. The default theme
produces plots with a grey background, a white grid, no charting area frame, and
standard axis and tick labels:</p>
<div class="sourceCode" id="cb337"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">base_g</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html">tibble</a></span><span class="op">(</span></span>
<span>  x<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span><span class="op">(</span><span class="fl">100</span><span class="op">)</span>,</span>
<span>  y<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span><span class="op">(</span><span class="fl">100</span><span class="op">)</span></span>
<span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">x</span>, y<span class="op">=</span><span class="va">y</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">base_g</span></span></code></pre></div>
<div class="inline-figure"><img src="main_files/viz%20pub%20default-1.png" width="672"></div>
<p>This is fine for displaying on a screen, but if this plot were to be printed on
paper, the solid grey background would use unnecessary ink.</p>
<div class="box note">
<p>There is a helpful principle in plotting: <em>use the least amount of ink possible
to produce the desired plot.</em> This is not just for economic, ink-saving reasons,
but also because sparser plots without superfluous graphic elements allow the
reader to focus more on the data than the styling.</p>
</div>
<p><code>ggplot</code> comes with other themes that may be added to plots with
<a href="https://ggplot2.tidyverse.org/reference/ggtheme.html"><code>theme_X()</code></a> functions.
Another available theme is <code><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw()</a></code>:</p>
<div class="sourceCode" id="cb338"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">base_g</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="main_files/viz%20theme%20bw-1.png" width="672"></div>
<p>This theme replaces the grey background with white, draws light grey grid lines,
and adds a frame around the plotting area. Plots with this theme might be more
suitable for plots displayed with a projector, for example in presentations.</p>
<p>Another theme is <code><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_classic()</a></code>, which draws only the elements of the plot
that are strictly necessary:</p>
<div class="sourceCode" id="cb339"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">base_g</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_classic</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="main_files/viz%20classic-1.png" width="672"></div>
<p>There are many so-called <a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">“complete
themes”</a> included with
ggplot. If so desired, all style elements controlled by these complete themes
can be customized using the
<a href="https://ggplot2.tidyverse.org/reference/theme.html"><code>theme()</code></a> function. A full
treatment of themes is beyond the scope of this chapter, but you may read all
about it in the documentation.</p>
</div>
<div id="exporting-to-svg" class="section level3" number="7.8.2">
<h3>
<span class="header-section-number">7.8.2</span> Exporting to SVG<a class="anchor" aria-label="anchor" href="#exporting-to-svg"><i class="fas fa-link"></i></a>
</h3>
<p>Even the custom themes and <a href="multipanel-figures">multipanel figures</a> may not
produce figures that adhere to scientific journal guidelines or your own tastes.
In these instances, all ggplot plots may be saved in <a href="https://en.wikipedia.org/wiki/Scalable_Vector_Graphics">Scalable Vector
Graphics</a> (SVG) format.
In contrast with bitmapped file formats like PNG or JPG that store images as a
matrix of pixels, SVG format describes shapes using a mathematical
specification. For example, a circle is expressed in SVG format like:</p>
<div class="sourceCode" id="cb340"><pre class="sourceCode xml"><code class="sourceCode xml"><span id="cb340-1"><a href="data-visualization.html#cb340-1" tabindex="-1"></a>&lt;<span class="kw">circle</span><span class="ot"> cx=</span><span class="st">"50"</span><span class="ot"> cy=</span><span class="st">"50"</span><span class="ot"> r=</span><span class="st">"50"</span>/&gt;</span></code></pre></div>
<p>This <a href="https://developer.mozilla.org/en-US/docs/Web/SVG/Element/circle"><circle> tag</circle></a>
specifies that there should be a circle (ellipse) positioned with the center at
the coordinate (50, 50) with a radius of 50. Illustration software like <a href="https://www.adobe.com/products/illustrator.html">Adobe
Illustrator</a> or
<a href="https://inkscape.org/">Inkscape</a> can understand SVG instructions and generate
a visual representation of them. Text can also be included directly in SVG
files. Because these elements are represented mathematically, the illustration
programs that understand them can produce images of any size and resolution,
allowing the designer to create clear, high quality images as required by
journals.</p>
<p>Another benefit of SVG is that the illustration programs that understand them
can also be used to edit the individual elements of the entire plot. For
example, the position of text labels, the font, and the text itself, may be
arbitrarily modified. Consider the following multipanel figure:</p>
<div class="sourceCode" id="cb341"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com">patchwork</a></span><span class="op">)</span></span>
<span><span class="va">g1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html">tibble</a></span><span class="op">(</span></span>
<span>  x<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span><span class="op">(</span><span class="fl">100</span><span class="op">)</span>,</span>
<span>  y<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span><span class="op">(</span><span class="fl">100</span><span class="op">)</span></span>
<span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">x</span>, y<span class="op">=</span><span class="va">y</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">g2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html">tibble</a></span><span class="op">(</span></span>
<span>   x<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span><span class="op">(</span><span class="fl">100</span><span class="op">)</span>,</span>
<span>   y<span class="op">=</span><span class="fl">5</span><span class="op">*</span><span class="va">x</span><span class="op">+</span><span class="fl">3</span><span class="op">+</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span><span class="op">(</span><span class="fl">100</span>,<span class="fl">1</span>,<span class="fl">4</span><span class="op">)</span></span>
<span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">x</span>, y<span class="op">=</span><span class="va">y</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_smooth.html">geom_smooth</a></span><span class="op">(</span>method<span class="op">=</span><span class="st">"loess"</span>, formula<span class="op">=</span><span class="va">y</span> <span class="op">~</span> <span class="va">x</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">g3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html">tibble</a></span><span class="op">(</span></span>
<span>  group<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="st">'A'</span>,<span class="fl">100</span><span class="op">)</span>,<span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="st">'B'</span>,<span class="fl">100</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  val<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span><span class="op">(</span><span class="fl">100</span>,<span class="fl">3</span><span class="op">)</span>,<span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span><span class="op">(</span><span class="fl">100</span>,<span class="fl">5</span>,<span class="fl">3</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">group</span>, y<span class="op">=</span><span class="va">val</span>,fill<span class="op">=</span><span class="va">group</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>  </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_violin.html">geom_violin</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="op">(</span><span class="va">g1</span> <span class="op">/</span> <span class="va">g2</span><span class="op">)</span> <span class="op">|</span> <span class="va">g3</span></span></code></pre></div>
<div class="inline-figure"><img src="main_files/viz%20pub%20multipanel-1.png" width="672"></div>
<p>This figure may be saved to SVG format by using the
<a href="https://ggplot2.tidyverse.org/reference/ggsave.html"><code>ggsave()</code></a> function:</p>
<div class="sourceCode" id="cb342"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggsave.html">ggsave</a></span><span class="op">(</span><span class="st">'multipanel.svg'</span><span class="op">)</span></span></code></pre></div>
<p>The image edited in Adobe Illustrator to have a different font and annotations
is as follows:</p>
<div class="float">
<img src="multipanel.png" alt="Edited multipanel figure"><div class="figcaption">Edited multipanel figure</div>
</div>
<p>By combining themes in ggplot with the SVG+illustration software strategy, very
high quality and customized figures can be created with relatively little
effort.</p>

</div>
</div>
<div id="network-visualization" class="section level2" number="7.9">
<h2>
<span class="header-section-number">7.9</span> Network visualization<a class="anchor" aria-label="anchor" href="#network-visualization"><i class="fas fa-link"></i></a>
</h2>
<p>The most commonly used package to visualize network is <code>igraph</code>.</p>
<div class="sourceCode" id="cb343"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://igraph.org">igraph</a></span><span class="op">)</span></span></code></pre></div>
<p>First, let’s create a graph with 15 nodes.</p>
<div class="sourceCode" id="cb344"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">names</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Opal"</span>, <span class="st">"Tourmaline"</span>, <span class="st">"Emerald"</span>, <span class="st">"Citrine"</span>, <span class="st">"Jasper"</span>, <span class="st">"Moonstone"</span>, <span class="st">"Agate"</span>, <span class="st">"Chalcedony"</span>, <span class="st">"Crystal"</span>, <span class="st">"Obsidian"</span>, <span class="st">"Quartz"</span>, <span class="st">"Topaz"</span>, <span class="st">"Beryl"</span>, <span class="st">"Sugilite"</span>, <span class="st">"Sapphire"</span><span class="op">)</span></span></code></pre></div>
<p>Use <code><a href="https://rdrr.io/pkg/igraph/man/sample_gnm.html">sample_gnm()</a></code> to generate a random graph.</p>
<div class="sourceCode" id="cb345"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">10</span><span class="op">)</span></span>
<span><span class="va">samp_g</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/igraph/man/sample_gnm.html">sample_gnm</a></span><span class="op">(</span></span>
<span>  n <span class="op">=</span> <span class="fl">15</span>, <span class="co"># number of vertices</span></span>
<span>  m <span class="op">=</span> <span class="fl">20</span>, <span class="co"># total number of edges</span></span>
<span>  directed <span class="op">=</span> <span class="cn">FALSE</span> <span class="co"># if it's a directed network</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>To assign color to vertices, use <code><a href="https://rdrr.io/pkg/igraph/man/vertex_attr.html">vertex_attr()</a></code>. <code><a href="https://rdrr.io/pkg/igraph/man/gorder.html">gorder()</a></code> returns the number of vertices.</p>
<div class="sourceCode" id="cb346"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/igraph/man/vertex_attr.html">vertex_attr</a></span><span class="op">(</span><span class="va">samp_g</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>name <span class="op">=</span> <span class="va">names</span>, color <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="st">"slategray1"</span>, <span class="fu"><a href="https://rdrr.io/pkg/igraph/man/gorder.html">gorder</a></span><span class="op">(</span><span class="va">samp_g</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>To assign color to edges, use <code><a href="https://rdrr.io/pkg/igraph/man/vertex_attr.html">vertex_attr()</a></code>. <code><a href="https://rdrr.io/pkg/igraph/man/gsize.html">gsize()</a></code> returns the total number of edges. Here I am assigning random colors.</p>
<div class="sourceCode" id="cb347"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/igraph/man/edge_attr.html">edge_attr</a></span><span class="op">(</span><span class="va">samp_g</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>color <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample</a></span><span class="op">(</span></span>
<span>  x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"slateblue1"</span>, <span class="st">"lightpink1"</span>, <span class="st">"steelblue1"</span>, <span class="st">"palegreen3"</span><span class="op">)</span>,</span>
<span>  size <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/igraph/man/gsize.html">gsize</a></span><span class="op">(</span><span class="va">samp_g</span><span class="op">)</span>, replace <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>In real case, when you want to modify the color of a specific edge, you need to first find which edge you want to modify…</p>
<div class="sourceCode" id="cb348"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/igraph/man/edge.html">edges</a></span><span class="op">(</span><span class="va">samp_g</span><span class="op">)</span></span></code></pre></div>
<pre><code>## [[1]]
## IGRAPH 7328255 UN-- 15 20 -- Erdos renyi (gnm) graph
## + attr: name (g/c), type (g/c), loops (g/l), m (g/n), name (v/c), color
## | (v/c), color (e/c)
## + edges from 7328255 (vertex names):
##  [1] Emerald   --Citrine    Citrine   --Jasper     Tourmaline--Moonstone 
##  [4] Emerald   --Chalcedony Emerald   --Crystal    Opal      --Obsidian  
##  [7] Agate     --Obsidian   Opal      --Quartz     Jasper    --Quartz    
## [10] Agate     --Quartz     Obsidian  --Quartz     Obsidian  --Topaz     
## [13] Tourmaline--Beryl      Agate     --Beryl      Quartz    --Beryl     
## [16] Quartz    --Sugilite   Emerald   --Sapphire   Moonstone --Sapphire  
## [19] Agate     --Sapphire   Sugilite  --Sapphire  
## 
## attr(,"class")
## [1] "igraph.edge"</code></pre>
<p>Then modify it. For example, modify the 1st edge, which is “Emerald–Citrine”.</p>
<div class="sourceCode" id="cb350"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/igraph/man/edge_attr.html">edge_attr</a></span><span class="op">(</span><span class="va">samp_g</span>, name <span class="op">=</span> <span class="st">"color"</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"red"</span></span></code></pre></div>
<p>To visualize the network, just use <code><a href="https://rdrr.io/r/graphics/plot.default.html">plot()</a></code>!</p>
<div class="sourceCode" id="cb351"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">samp_g</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="main_files/unnamed-chunk-132-1.png" width="672"></div>
<p>In real-world, we don’t want a random network. Now let’s create our network from scratch.</p>
<p>There are many different ways to create a network. One of the most basic (and easy) ways is using <code><a href="https://rdrr.io/pkg/igraph/man/graph_from_data_frame.html">graph_from_data_frame()</a></code> function in <code>igraph</code>.<br>
1. Create 2 columns in the data frame: <code>from</code> and <code>to</code>. Each row represents an edge, and the entries are vertex names. For instance, as indicated in row 1 below, there is an edge “from Crystal to Moonstone”. In an undirected graph, you can put the two vertices in either way.<br>
2. (optional) Add another column <code>color</code> to represent the color to plot the edge.<br>
3. (optional) You can also add additional information for each edge. For example, add a column <code>weight</code> to specify the weight of this edge. The <code>weight</code> will become useful when you do network analysis later, for example, find the shortest path between two vertices. You can also add other attributes you want, like the “connection_type” here.</p>
<div class="sourceCode" id="cb352"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">4</span><span class="op">)</span></span>
<span><span class="va">my_gem_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span></span>
<span>  <span class="st">"from"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Crystal"</span>, <span class="st">"Crystal"</span>, <span class="st">"Crystal"</span>, <span class="st">"Crystal"</span>, <span class="st">"Topaz"</span>, <span class="st">"Obsidian"</span>, <span class="st">"Agate"</span>, <span class="st">"Topaz"</span><span class="op">)</span>,</span>
<span>  <span class="st">"to"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Moonstone"</span>, <span class="st">"Citrine"</span>, <span class="st">"Obsidian"</span>, <span class="st">"Agate"</span>, <span class="st">"Citrine"</span>, <span class="st">"Topaz"</span>, <span class="st">"Moonstone"</span>, <span class="st">"Moonstone"</span><span class="op">)</span>,</span>
<span>  <span class="st">"color"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"slateblue1"</span>, <span class="st">"slateblue1"</span>, <span class="st">"slateblue1"</span>, <span class="st">"steelblue1"</span>, <span class="st">"palegreen3"</span>, <span class="st">"steelblue1"</span>, <span class="st">"lightpink1"</span>, <span class="st">"palegreen3"</span><span class="op">)</span>,</span>
<span>  <span class="st">"connection_type"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">6</span><span class="op">)</span>, <span class="fl">2</span>, <span class="fl">2</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">my_gem_df</span><span class="op">$</span><span class="va">weight</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html">as.numeric</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html">as.factor</a></span><span class="op">(</span><span class="va">my_gem_df</span><span class="op">$</span><span class="va">color</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">my_gem_df</span></span></code></pre></div>
<pre><code>##       from        to      color connection_type weight
## 1  Crystal Moonstone slateblue1               1      3
## 2  Crystal   Citrine slateblue1               1      3
## 3  Crystal  Obsidian slateblue1               1      3
## 4  Crystal     Agate steelblue1               1      4
## 5    Topaz   Citrine palegreen3               1      2
## 6 Obsidian     Topaz steelblue1               1      4
## 7    Agate Moonstone lightpink1               2      1
## 8    Topaz Moonstone palegreen3               2      2</code></pre>
<p>Now create an igraph object from the data frame, and assign vertex color.</p>
<div class="sourceCode" id="cb354"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">g</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/igraph/man/graph_from_data_frame.html">graph_from_data_frame</a></span><span class="op">(</span><span class="va">my_gem_df</span>, directed <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/igraph/man/vertex_attr.html">vertex_attr</a></span><span class="op">(</span><span class="va">g</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span></span>
<span>  name <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Crystal"</span>, <span class="st">"Topaz"</span>, <span class="st">"Obsidian"</span>, <span class="st">"Agate"</span>, <span class="st">"Moonstone"</span>, <span class="st">"Citrine"</span><span class="op">)</span>,</span>
<span>  color <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"slategray1"</span>, <span class="st">"slategray1"</span>, <span class="st">"slategray1"</span>, <span class="st">"lightpink"</span>, <span class="st">"slategray1"</span>, <span class="st">"lightpink"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Again, simply use <code>plot(g)</code> to plot the network. We can also do some fancy customization to play around with the labels! <em>(Remember the <code>connection_type</code> I specified above in the data frame? In <code>edge.lty</code>, 1 is a solid line, and 2 is a dash line)</em></p>
<div class="sourceCode" id="cb355"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">4</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">g</span>,</span>
<span>  edge.width <span class="op">=</span> <span class="fl">1.5</span>,</span>
<span>  edge.lty <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/igraph/man/edge_attr.html">edge_attr</a></span><span class="op">(</span><span class="va">g</span>, <span class="st">"connection_type"</span><span class="op">)</span>,</span>
<span>  vertex.label.family <span class="op">=</span> <span class="st">"Trebuchet MS"</span>,</span>
<span>  vertex.label.font <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  vertex.label.cex <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  vertex.label.dist <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  vertex.label.degree <span class="op">=</span> <span class="va">pi</span> <span class="op">/</span> <span class="fl">2</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="main_files/unnamed-chunk-135-1.png" width="672"></div>
<p>To see the whole list of available arguments available when plotting your network, visit <a href="https://igraph.org/r/doc/plot.common.html">igraph plot manual</a>.</p>
<p>In the case of an directed graph, simply put <code>directed = TRUE</code> when creating the igraph object, and you will see the cute arrows.</p>
<div class="sourceCode" id="cb356"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">g</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/igraph/man/graph_from_data_frame.html">graph_from_data_frame</a></span><span class="op">(</span><span class="va">my_gem_df</span>, directed <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/igraph/man/vertex_attr.html">vertex_attr</a></span><span class="op">(</span><span class="va">g</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span></span>
<span>  name <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Crystal"</span>, <span class="st">"Topaz"</span>, <span class="st">"Obsidian"</span>, <span class="st">"Agate"</span>, <span class="st">"Moonstone"</span>, <span class="st">"Citrine"</span><span class="op">)</span>,</span>
<span>  color <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"slategray1"</span>, <span class="st">"slategray1"</span>, <span class="st">"slategray1"</span>, <span class="st">"lightpink"</span>, <span class="st">"slategray1"</span>, <span class="st">"lightpink"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">4</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">g</span>,</span>
<span>  edge.width <span class="op">=</span> <span class="fl">1.5</span>,</span>
<span>  edge.lty <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/igraph/man/edge_attr.html">edge_attr</a></span><span class="op">(</span><span class="va">g</span>, <span class="st">"connection_type"</span><span class="op">)</span>,</span>
<span>  edge.arrow.size <span class="op">=</span> <span class="fl">0.8</span>,</span>
<span>  vertex.label.family <span class="op">=</span> <span class="st">"Trebuchet MS"</span>,</span>
<span>  vertex.label.font <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  vertex.label.cex <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  vertex.label.dist <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  vertex.label.degree <span class="op">=</span> <span class="va">pi</span> <span class="op">/</span> <span class="fl">2</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="main_files/unnamed-chunk-136-1.png" width="672"></div>
<div id="interactive-network-plots" class="section level3" number="7.9.1">
<h3>
<span class="header-section-number">7.9.1</span> Interactive network plots<a class="anchor" aria-label="anchor" href="#interactive-network-plots"><i class="fas fa-link"></i></a>
</h3>
<p>Besides <code>igraph</code>, there are many other packages to create fancier networks. One example is <code>visNetwork</code>. It can create an interactive network, and it’s easy to use! Check <a href="https://cran.r-project.org/web/packages/visNetwork/vignettes/Introduction-to-visNetwork.html">example here</a>. It can be implemented in your HTML report knitted from R markdown and R Shiny app interface!<br>
One thing to keep in mind: as you can imagine, since it is interactive, <code>visNetwork</code> requires much more memory and time compared to a static graph. So, you may need to make a choice when you develop your web interface or R markdown HTML report.</p>

</div>
</div>
</div>
  <div class="chapter-nav">
<div class="prev"><a href="data-science.html"><span class="header-section-number">6</span> Data Science</a></div>
<div class="next"><a href="biology-bioinformatics.html"><span class="header-section-number">8</span> Biology &amp; Bioinformatics</a></div>
</div></main><div class="col-md-3 col-lg-2 d-none d-md-block sidebar sidebar-chapter">
    <nav id="toc" data-toggle="toc" aria-label="On this page"><h2>On this page</h2>
      <ul class="nav navbar-nav">
<li><a class="nav-link" href="#data-visualization"><span class="header-section-number">7</span> Data Visualization</a></li>
<li>
<a class="nav-link" href="#responsible-plotting"><span class="header-section-number">7.1</span> Responsible Plotting</a><ul class="nav navbar-nav">
<li><a class="nav-link" href="#visualization-principles"><span class="header-section-number">7.1.1</span> Visualization Principles</a></li>
<li><a class="nav-link" href="#human-visual-perception"><span class="header-section-number">7.1.2</span> Human Visual Perception</a></li>
<li><a class="nav-link" href="#visual-encodings"><span class="header-section-number">7.1.3</span> Visual Encodings</a></li>
<li><a class="nav-link" href="#some-opinionated-rules-of-thumb"><span class="header-section-number">7.1.4</span> Some Opinionated Rules of Thumb</a></li>
</ul>
</li>
<li>
<a class="nav-link" href="#dv-gg"><span class="header-section-number">7.2</span> Grammar of Graphics</a><ul class="nav navbar-nav"><li><a class="nav-link" href="#ggplot-mechanics"><span class="header-section-number">7.2.1</span> ggplot mechanics</a></li></ul>
</li>
<li>
<a class="nav-link" href="#plotting-one-dimension"><span class="header-section-number">7.3</span> Plotting One Dimension</a><ul class="nav navbar-nav">
<li><a class="nav-link" href="#dv-bar"><span class="header-section-number">7.3.1</span> Bar chart</a></li>
<li><a class="nav-link" href="#lollipop-plots"><span class="header-section-number">7.3.2</span> Lollipop plots</a></li>
<li><a class="nav-link" href="#stacked-area-charts"><span class="header-section-number">7.3.3</span> Stacked Area charts</a></li>
</ul>
</li>
<li>
<a class="nav-link" href="#visualizing-distributions"><span class="header-section-number">7.4</span> Visualizing Distributions</a><ul class="nav navbar-nav">
<li><a class="nav-link" href="#histogram"><span class="header-section-number">7.4.1</span> Histogram</a></li>
<li><a class="nav-link" href="#density"><span class="header-section-number">7.4.2</span> Density</a></li>
<li><a class="nav-link" href="#boxplot"><span class="header-section-number">7.4.3</span> Boxplot</a></li>
<li><a class="nav-link" href="#violin-plot"><span class="header-section-number">7.4.4</span> Violin plot</a></li>
<li><a class="nav-link" href="#beeswarm-plot"><span class="header-section-number">7.4.5</span> Beeswarm plot</a></li>
<li><a class="nav-link" href="#ridgeline"><span class="header-section-number">7.4.6</span> Ridgeline</a></li>
</ul>
</li>
<li>
<a class="nav-link" href="#plotting-two-or-more-dimensions"><span class="header-section-number">7.5</span> Plotting Two or More Dimensions</a><ul class="nav navbar-nav">
<li><a class="nav-link" href="#scatter-plots"><span class="header-section-number">7.5.1</span> Scatter Plots</a></li>
<li><a class="nav-link" href="#bubble-plots"><span class="header-section-number">7.5.2</span> Bubble Plots</a></li>
<li><a class="nav-link" href="#connected-scatter-plots"><span class="header-section-number">7.5.3</span> Connected Scatter Plots</a></li>
<li><a class="nav-link" href="#line-plots"><span class="header-section-number">7.5.4</span> Line Plots</a></li>
<li><a class="nav-link" href="#parallel-coordinate-plots"><span class="header-section-number">7.5.5</span> Parallel Coordinate Plots</a></li>
<li><a class="nav-link" href="#heatmaps"><span class="header-section-number">7.5.6</span> Heatmaps</a></li>
<li><a class="nav-link" href="#how-to-use-heatmaps-responsibly"><span class="header-section-number">7.5.7</span> How To Use Heatmaps Responsibly</a></li>
</ul>
</li>
<li>
<a class="nav-link" href="#other-kind-of-plots"><span class="header-section-number">7.6</span> Other Kind of Plots</a><ul class="nav navbar-nav">
<li><a class="nav-link" href="#dendrograms"><span class="header-section-number">7.6.1</span> Dendrograms</a></li>
<li><a class="nav-link" href="#chord-diagrams-and-circos-plots"><span class="header-section-number">7.6.2</span> Chord Diagrams and Circos Plots</a></li>
</ul>
</li>
<li>
<a class="nav-link" href="#multiple-plots"><span class="header-section-number">7.7</span> Multiple Plots</a><ul class="nav navbar-nav">
<li><a class="nav-link" href="#facet-wrapping"><span class="header-section-number">7.7.1</span> Facet wrapping</a></li>
<li><a class="nav-link" href="#multipanel-figures"><span class="header-section-number">7.7.2</span> Multipanel Figures</a></li>
</ul>
</li>
<li>
<a class="nav-link" href="#publication-ready-plots"><span class="header-section-number">7.8</span> Publication Ready Plots</a><ul class="nav navbar-nav">
<li><a class="nav-link" href="#ggplot-themes"><span class="header-section-number">7.8.1</span> ggplot Themes</a></li>
<li><a class="nav-link" href="#exporting-to-svg"><span class="header-section-number">7.8.2</span> Exporting to SVG</a></li>
</ul>
</li>
<li>
<a class="nav-link" href="#network-visualization"><span class="header-section-number">7.9</span> Network visualization</a><ul class="nav navbar-nav"><li><a class="nav-link" href="#interactive-network-plots"><span class="header-section-number">7.9.1</span> Interactive network plots</a></li></ul>
</li>
</ul>

      <div class="book-extra">
        <ul class="list-unstyled">
          
        </ul>
</div>
    </nav>
</div>

</div>
</div> <!-- .container -->

<footer class="bg-primary text-light mt-5"><div class="container"><div class="row">

  <div class="col-12 col-md-6 mt-3">
    <p>"<strong>BF591 - R for Biological Sciences</strong>" was written by . </p>
  </div>

  <div class="col-12 col-md-6 mt-3">
    <p>This book was built by the <a class="text-light" href="https://bookdown.org">bookdown</a> R package.</p>
  </div>

</div></div>
</footer><!-- dynamically load mathjax for compatibility with self-contained --><script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/latest.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script><script type="text/x-mathjax-config">const popovers = document.querySelectorAll('a.footnote-ref[data-toggle="popover"]');
for (let popover of popovers) {
  const div = document.createElement('div');
  div.setAttribute('style', 'position: absolute; top: 0, left:0; width:0, height:0, overflow: hidden; visibility: hidden;');
  div.innerHTML = popover.getAttribute('data-content');

  var has_math = div.querySelector("span.math");
  if (has_math) {
    document.body.appendChild(div);
    MathJax.Hub.Queue(["Typeset", MathJax.Hub, div]);
    MathJax.Hub.Queue(function() {
      popover.setAttribute('data-content', div.innerHTML);
      document.body.removeChild(div);
    })
  }
}
</script>
</body>
</html>
