<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta name="viewport" content="width=device-width, initial-scale=1" />
<meta property="og:title" content="6.4 Exploratory Data Analysis | BF591 - R for Biological Sciences" />
<meta property="og:type" content="book" />







<script type="text/x-mathjax-config">
MathJax.Hub.Config({
  TeX: { equationNumbers: { autoNumber: "AMS" } }
});
</script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<meta name="description" content="6.4 Exploratory Data Analysis | BF591 - R for Biological Sciences">

<title>6.4 Exploratory Data Analysis | BF591 - R for Biological Sciences</title>

<script src="libs/header-attrs-2.13/header-attrs.js"></script>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="libs/bootstrap-3.3.5/css/darkly.min.css" rel="stylesheet" />
<script src="libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="libs/navigation-1.1/tabsets.js"></script>


<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>


<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />

<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
/* show arrow before summary tag as in bootstrap
TODO: remove if boostrap in updated in html_document (rmarkdown#1485) */
details > summary {
  display: list-item;
  cursor: pointer;
}
</style>
</head>

<body>

<div class="container-fluid main-container">


<div class="row">
<div class="col-sm-12">
<div id="TOC">
<ul>
<li><a href="index.html#syllabus">Syllabus</a>
<ul>
<li><a href="course-schedule.html#course-schedule">Course Schedule</a></li>
<li><a href="instructors.html#instructors">Instructors</a></li>
<li><a href="office-hours.html#office-hours">Office Hours</a></li>
<li><a href="course-values-and-policies.html#course-values-and-policies">Course Values and Policies</a></li>
</ul></li>
<li><a href="introduction.html#introduction"><span class="toc-section-number">1</span> Introduction</a>
<ul>
<li><a href="who-this-book-is-for.html#who-this-book-is-for"><span class="toc-section-number">1.1</span> Who This Book Is For</a></li>
<li><a href="a-note-about-reinventing-the-wheel.html#a-note-about-reinventing-the-wheel"><span class="toc-section-number">1.2</span> A Note About Reinventing the Wheel</a></li>
<li><a href="sources-and-references.html#sources-and-references"><span class="toc-section-number">1.3</span> Sources and References</a></li>
</ul></li>
<li><a href="data-bio.html#data-bio"><span class="toc-section-number">2</span> Data in Biology</a>
<ul>
<li><a href="a-brief-history-of-data-in-molecular-biology.html#a-brief-history-of-data-in-molecular-biology"><span class="toc-section-number">2.1</span> A Brief History of Data in Molecular Biology</a></li>
<li><a href="biology-as-a-mature-data-science.html#biology-as-a-mature-data-science"><span class="toc-section-number">2.2</span> Biology as a Mature Data Science</a></li>
</ul></li>
<li><a href="preliminaries.html#preliminaries"><span class="toc-section-number">3</span> Preliminaries</a>
<ul>
<li><a href="prelim-r.html#prelim-r"><span class="toc-section-number">3.1</span> The R Language</a></li>
<li><a href="prelim-rstudio.html#prelim-rstudio"><span class="toc-section-number">3.2</span> RStudio</a></li>
<li><a href="prog-the-r-script.html#prog-the-r-script"><span class="toc-section-number">3.3</span> The R Script</a></li>
<li><a href="prog-workflow.html#prog-workflow"><span class="toc-section-number">3.4</span> The Scripting Workflow</a></li>
<li><a href="prelim-git.html#prelim-git"><span class="toc-section-number">3.5</span> git + github</a>
<ul>
<li><a href="prelim-git.html#motivation"><span class="toc-section-number">3.5.1</span> Motivation</a></li>
<li><a href="prelim-git.html#git"><span class="toc-section-number">3.5.2</span> git</a></li>
<li><a href="prelim-git.html#git-hosting-platforms-github"><span class="toc-section-number">3.5.3</span> Git hosting platforms (GitHub)</a></li>
</ul></li>
</ul></li>
<li><a href="prog-basics.html#prog-basics"><span class="toc-section-number">4</span> R Programming</a>
<ul>
<li><a href="before-you-begin.html#before-you-begin"><span class="toc-section-number">4.1</span> Before you begin</a></li>
<li><a href="introduction-1.html#introduction-1"><span class="toc-section-number">4.2</span> Introduction</a></li>
<li><a href="prog-r-syntax.html#prog-r-syntax"><span class="toc-section-number">4.3</span> R Syntax Basics</a></li>
<li><a href="prog-types.html#prog-types"><span class="toc-section-number">4.4</span> Basic Types of Values</a>
<ul>
<li><a href="prog-types.html#factors"><span class="toc-section-number">4.4.1</span> Factors</a></li>
</ul></li>
<li><a href="prog-struct.html#prog-struct"><span class="toc-section-number">4.5</span> Data Structures</a>
<ul>
<li><a href="prog-struct.html#vectors"><span class="toc-section-number">4.5.1</span> Vectors</a></li>
<li><a href="prog-struct.html#matrices"><span class="toc-section-number">4.5.2</span> Matrices</a></li>
<li><a href="prog-struct.html#lists-and-data-frames"><span class="toc-section-number">4.5.3</span> Lists and data frames</a></li>
</ul></li>
<li><a href="logical-tests-and-comparators.html#logical-tests-and-comparators"><span class="toc-section-number">4.6</span> Logical Tests and Comparators</a></li>
<li><a href="functions.html#functions"><span class="toc-section-number">4.7</span> Functions</a>
<ul>
<li><a href="functions.html#prog-dry"><span class="toc-section-number">4.7.1</span> DRY: Don’t Repeat Yourself</a></li>
<li><a href="functions.html#writing-your-own-functions"><span class="toc-section-number">4.7.2</span> Writing your own functions</a></li>
<li><a href="functions.html#scope"><span class="toc-section-number">4.7.3</span> Scope</a></li>
</ul></li>
<li><a href="iteration.html#iteration"><span class="toc-section-number">4.8</span> Iteration</a>
<ul>
<li><a href="iteration.html#vectorized-operations"><span class="toc-section-number">4.8.1</span> Vectorized operations</a></li>
<li><a href="iteration.html#functional-programming"><span class="toc-section-number">4.8.2</span> Functional programming</a></li>
<li><a href="iteration.html#apply-and-friends"><span class="toc-section-number">4.8.3</span> <code>apply()</code> and friends</a></li>
</ul></li>
<li><a href="installing-packages.html#installing-packages"><span class="toc-section-number">4.9</span> Installing Packages</a></li>
<li><a href="saving-and-loading-r-data.html#saving-and-loading-r-data"><span class="toc-section-number">4.10</span> Saving and Loading R Data</a></li>
<li><a href="prog-debugging.html#prog-debugging"><span class="toc-section-number">4.11</span> Troubleshooting and Debugging</a></li>
<li><a href="prog-style.html#prog-style"><span class="toc-section-number">4.12</span> Coding Style and Conventions</a>
<ul>
<li><a href="prog-style.html#is-my-code-correct"><span class="toc-section-number">4.12.1</span> Is my code correct?</a></li>
<li><a href="prog-style.html#does-my-code-follow-the-dry-principle"><span class="toc-section-number">4.12.2</span> Does my code follow the <span>DRY principle</span>?</a></li>
<li><a href="prog-style.html#did-i-choose-concise-but-descriptive-variable-and-function-names"><span class="toc-section-number">4.12.3</span> Did I choose concise but descriptive variable and function names?</a></li>
<li><a href="prog-style.html#did-i-use-indentation-and-naming-conventions-consistently-throughout-my-code"><span class="toc-section-number">4.12.4</span> Did I use indentation and naming conventions consistently throughout my code?</a></li>
<li><a href="prog-style.html#did-i-write-comments-especially-when-what-the-code-does-is-not-obvious"><span class="toc-section-number">4.12.5</span> Did I write comments, especially when what the code does is not obvious?</a></li>
<li><a href="prog-style.html#how-easy-would-it-be-for-someone-else-to-understand-my-code"><span class="toc-section-number">4.12.6</span> How easy would it be for someone else to understand my code?</a></li>
<li><a href="prog-style.html#is-my-code-easy-to-maintainchange"><span class="toc-section-number">4.12.7</span> Is my code easy to maintain/change?</a></li>
<li><a href="prog-style.html#the-styler-package"><span class="toc-section-number">4.12.8</span> The <code>styler</code> package</a></li>
</ul></li>
</ul></li>
<li><a href="data-wrangling.html#data-wrangling"><span class="toc-section-number">5</span> Data Wrangling</a>
<ul>
<li><a href="the-tidyverse.html#the-tidyverse"><span class="toc-section-number">5.1</span> The Tidyverse</a></li>
<li><a href="dw-basics.html#dw-basics"><span class="toc-section-number">5.2</span> Tidyverse Basics</a></li>
<li><a href="importing-data.html#importing-data"><span class="toc-section-number">5.3</span> Importing Data</a></li>
<li><a href="data-tibble.html#data-tibble"><span class="toc-section-number">5.4</span> The tibble</a></li>
<li><a href="tidy-data.html#tidy-data"><span class="toc-section-number">5.5</span> Tidy Data</a></li>
<li><a href="pipes.html#pipes"><span class="toc-section-number">5.6</span> pipes</a></li>
<li><a href="arranging-data.html#arranging-data"><span class="toc-section-number">5.7</span> Arranging Data</a>
<ul>
<li><a href="arranging-data.html#dplyrmutate---create-new-columns-using-other-columns"><span class="toc-section-number">5.7.1</span> <code>dplyr::mutate()</code> - Create new columns using other columns</a></li>
<li><a href="arranging-data.html#stringr---working-with-character-values"><span class="toc-section-number">5.7.2</span> <code>stringr</code> - Working with character values</a></li>
<li><a href="arranging-data.html#dplyrselect---subset-columns-by-name"><span class="toc-section-number">5.7.3</span> <code>dplyr::select()</code> - Subset Columns by Name</a></li>
<li><a href="arranging-data.html#dplyrfilter---pick-rows-out-of-a-data-set"><span class="toc-section-number">5.7.4</span> <code>dplyr::filter()</code> - Pick rows out of a data set</a></li>
<li><a href="arranging-data.html#dplyrarrange---order-rows-based-on-their-values"><span class="toc-section-number">5.7.5</span> <code>dplyr::arrange()</code> - Order rows based on their values</a></li>
<li><a href="arranging-data.html#putting-it-all-together"><span class="toc-section-number">5.7.6</span> Putting it all together</a></li>
</ul></li>
<li><a href="grouping-data.html#grouping-data"><span class="toc-section-number">5.8</span> Grouping Data</a></li>
<li><a href="rearranging-data.html#rearranging-data"><span class="toc-section-number">5.9</span> Rearranging Data</a></li>
<li><a href="relational-data.html#relational-data"><span class="toc-section-number">5.10</span> Relational Data</a>
<ul>
<li><a href="relational-data.html#dealing-with-multiple-matches"><span class="toc-section-number">5.10.1</span> Dealing with multiple matches</a></li>
</ul></li>
</ul></li>
<li><a href="data-science.html#data-science"><span class="toc-section-number">6</span> Data Science</a>
<ul>
<li><a href="data-modeling.html#data-modeling"><span class="toc-section-number">6.1</span> Data Modeling</a>
<ul>
<li><a href="data-modeling.html#a-worked-modeling-example"><span class="toc-section-number">6.1.1</span> A Worked Modeling Example</a></li>
<li><a href="data-modeling.html#data-summarization"><span class="toc-section-number">6.1.2</span> Data Summarization</a></li>
<li><a href="data-modeling.html#linear-models"><span class="toc-section-number">6.1.3</span> Linear Models</a></li>
<li><a href="data-modeling.html#flavors-of-linear-models"><span class="toc-section-number">6.1.4</span> Flavors of Linear Models</a></li>
<li><a href="data-modeling.html#assessing-model-accuracy-."><span class="toc-section-number">6.1.5</span> Assessing Model Accuracy .</a></li>
</ul></li>
<li><a href="statistical-distributions.html#statistical-distributions"><span class="toc-section-number">6.2</span> Statistical Distributions</a>
<ul>
<li><a href="statistical-distributions.html#random-variables"><span class="toc-section-number">6.2.1</span> Random Variables</a></li>
<li><a href="statistical-distributions.html#statistical-distribution-basics"><span class="toc-section-number">6.2.2</span> Statistical Distribution Basics</a></li>
<li><a href="statistical-distributions.html#distributions-in-r"><span class="toc-section-number">6.2.3</span> Distributions in R</a></li>
<li><a href="statistical-distributions.html#discrete-distributions"><span class="toc-section-number">6.2.4</span> Discrete Distributions</a></li>
<li><a href="statistical-distributions.html#continuous-distributions"><span class="toc-section-number">6.2.5</span> Continuous Distributions</a></li>
<li><a href="statistical-distributions.html#empirical-distributions"><span class="toc-section-number">6.2.6</span> Empirical Distributions</a></li>
</ul></li>
<li><a href="statistical-tests.html#statistical-tests"><span class="toc-section-number">6.3</span> Statistical Tests</a>
<ul>
<li><a href="statistical-tests.html#what-is-a-statistical-test"><span class="toc-section-number">6.3.1</span> What is a statistical test</a></li>
<li><a href="statistical-tests.html#common-statistical-tests"><span class="toc-section-number">6.3.2</span> Common Statistical Tests</a></li>
<li><a href="statistical-tests.html#choosing-a-test"><span class="toc-section-number">6.3.3</span> Choosing a Test</a></li>
<li><a href="statistical-tests.html#p-values"><span class="toc-section-number">6.3.4</span> P-values</a></li>
<li><a href="statistical-tests.html#multiple-hypothesis-testing-."><span class="toc-section-number">6.3.5</span> Multiple Hypothesis Testing .</a></li>
<li><a href="statistical-tests.html#statistical-power"><span class="toc-section-number">6.3.6</span> Statistical power</a></li>
</ul></li>
<li><a href="exploratory-data-analysis.html#exploratory-data-analysis"><span class="toc-section-number">6.4</span> Exploratory Data Analysis</a>
<ul>
<li><a href="exploratory-data-analysis.html#principal-component-analysis"><span class="toc-section-number">6.4.1</span> Principal Component Analysis</a></li>
<li><a href="exploratory-data-analysis.html#cluster-analysis"><span class="toc-section-number">6.4.2</span> Cluster Analysis</a></li>
<li><a href="exploratory-data-analysis.html#k-means-."><span class="toc-section-number">6.4.3</span> k-means .</a></li>
<li><a href="exploratory-data-analysis.html#others"><span class="toc-section-number">6.4.4</span> Others</a></li>
</ul></li>
<li><a href="network-analysis-..html#network-analysis-."><span class="toc-section-number">6.5</span> Network Analysis .</a>
<ul>
<li><a href="network-analysis-..html#basic-network-analysis"><span class="toc-section-number">6.5.1</span> Basic network analysis</a></li>
<li><a href="network-analysis-..html#create-your-own"><span class="toc-section-number">6.5.2</span> Create your own</a></li>
</ul></li>
</ul></li>
<li><a href="data-visualization.html#data-visualization"><span class="toc-section-number">7</span> Data Visualization</a>
<ul>
<li><a href="responsible-plotting.html#responsible-plotting"><span class="toc-section-number">7.1</span> Responsible Plotting</a>
<ul>
<li><a href="responsible-plotting.html#visualization-principles"><span class="toc-section-number">7.1.1</span> Visualization Principles</a></li>
<li><a href="responsible-plotting.html#human-visual-perception"><span class="toc-section-number">7.1.2</span> Human Visual Perception</a></li>
<li><a href="responsible-plotting.html#visual-encodings"><span class="toc-section-number">7.1.3</span> Visual Encodings</a></li>
<li><a href="responsible-plotting.html#some-opinionated-rules-of-thumb"><span class="toc-section-number">7.1.4</span> Some Opinionated Rules of Thumb</a></li>
</ul></li>
<li><a href="dv-gg.html#dv-gg"><span class="toc-section-number">7.2</span> Grammar of Graphics</a>
<ul>
<li><a href="dv-gg.html#ggplot-mechanics"><span class="toc-section-number">7.2.1</span> <code>ggplot</code> mechanics</a></li>
</ul></li>
<li><a href="plotting-one-dimension.html#plotting-one-dimension"><span class="toc-section-number">7.3</span> Plotting One Dimension</a>
<ul>
<li><a href="plotting-one-dimension.html#dv-bar"><span class="toc-section-number">7.3.1</span> Bar chart</a></li>
<li><a href="plotting-one-dimension.html#lollipop-plots"><span class="toc-section-number">7.3.2</span> Lollipop plots</a></li>
<li><a href="plotting-one-dimension.html#stacked-area-charts"><span class="toc-section-number">7.3.3</span> Stacked Area charts</a></li>
<li><a href="plotting-one-dimension.html#parallel-coordinate-plots"><span class="toc-section-number">7.3.4</span> Parallel Coordinate plots</a></li>
</ul></li>
<li><a href="visualizing-distributions.html#visualizing-distributions"><span class="toc-section-number">7.4</span> Visualizing Distributions</a>
<ul>
<li><a href="visualizing-distributions.html#histogram"><span class="toc-section-number">7.4.1</span> Histogram</a></li>
<li><a href="visualizing-distributions.html#density"><span class="toc-section-number">7.4.2</span> Density</a></li>
<li><a href="visualizing-distributions.html#boxplot"><span class="toc-section-number">7.4.3</span> Boxplot</a></li>
<li><a href="visualizing-distributions.html#violin-plot"><span class="toc-section-number">7.4.4</span> Violin plot</a></li>
<li><a href="visualizing-distributions.html#beeswarm-plot"><span class="toc-section-number">7.4.5</span> Beeswarm plot</a></li>
<li><a href="visualizing-distributions.html#ridgeline"><span class="toc-section-number">7.4.6</span> Ridgeline</a></li>
</ul></li>
<li><a href="plotting-two-or-more-dimensions.html#plotting-two-or-more-dimensions"><span class="toc-section-number">7.5</span> Plotting Two or More Dimensions</a>
<ul>
<li><a href="plotting-two-or-more-dimensions.html#scatter-plots"><span class="toc-section-number">7.5.1</span> Scatter Plots</a></li>
<li><a href="plotting-two-or-more-dimensions.html#bubble-plots"><span class="toc-section-number">7.5.2</span> Bubble Plots</a></li>
<li><a href="plotting-two-or-more-dimensions.html#connected-scatter-plots"><span class="toc-section-number">7.5.3</span> Connected Scatter Plots</a></li>
<li><a href="plotting-two-or-more-dimensions.html#line-plots"><span class="toc-section-number">7.5.4</span> Line Plots</a></li>
<li><a href="plotting-two-or-more-dimensions.html#parallel-coordinate-plots-1"><span class="toc-section-number">7.5.5</span> Parallel Coordinate Plots</a></li>
<li><a href="plotting-two-or-more-dimensions.html#heatmaps"><span class="toc-section-number">7.5.6</span> Heatmaps</a></li>
<li><a href="plotting-two-or-more-dimensions.html#how-to-use-heatmaps-responsibly"><span class="toc-section-number">7.5.7</span> How To Use Heatmaps Responsibly</a></li>
</ul></li>
<li><a href="other-kind-of-plots.html#other-kind-of-plots"><span class="toc-section-number">7.6</span> Other Kind of Plots</a>
<ul>
<li><a href="other-kind-of-plots.html#dendrograms"><span class="toc-section-number">7.6.1</span> Dendrograms</a></li>
<li><a href="other-kind-of-plots.html#chord-diagrams-and-circos-plots"><span class="toc-section-number">7.6.2</span> Chord Diagrams and Circos Plots</a></li>
</ul></li>
<li><a href="multiple-plots.html#multiple-plots"><span class="toc-section-number">7.7</span> Multiple Plots</a>
<ul>
<li><a href="multiple-plots.html#facet-wrapping"><span class="toc-section-number">7.7.1</span> Facet wrapping</a></li>
<li><a href="multiple-plots.html#multipanel-figures"><span class="toc-section-number">7.7.2</span> Multipanel Figures</a></li>
</ul></li>
<li><a href="publication-ready-plots.html#publication-ready-plots"><span class="toc-section-number">7.8</span> Publication Ready Plots</a>
<ul>
<li><a href="publication-ready-plots.html#ggplot-themes"><span class="toc-section-number">7.8.1</span> <code>ggplot</code> Themes</a></li>
<li><a href="publication-ready-plots.html#exporting-to-svg"><span class="toc-section-number">7.8.2</span> Exporting to SVG</a></li>
</ul></li>
<li><a href="network-visualization.html#network-visualization"><span class="toc-section-number">7.9</span> Network visualization</a>
<ul>
<li><a href="network-visualization.html#interactive-network-plots"><span class="toc-section-number">7.9.1</span> Interactive network plots</a></li>
</ul></li>
</ul></li>
<li><a href="biology-bioinformatics.html#biology-bioinformatics"><span class="toc-section-number">8</span> Biology &amp; Bioinformatics</a>
<ul>
<li><a href="bio-r.html#bio-r"><span class="toc-section-number">8.1</span> R in Biology</a></li>
<li><a href="biological-data-overview.html#biological-data-overview"><span class="toc-section-number">8.2</span> Biological Data Overview</a>
<ul>
<li><a href="biological-data-overview.html#bio-data-types"><span class="toc-section-number">8.2.1</span> Types of Biological Data</a></li>
<li><a href="biological-data-overview.html#csv-files"><span class="toc-section-number">8.2.2</span> CSV Files</a></li>
<li><a href="biological-data-overview.html#biological-data-is-not-tidy"><span class="toc-section-number">8.2.3</span> Biological data is NOT Tidy!</a></li>
</ul></li>
<li><a href="bio-bioconductor.html#bio-bioconductor"><span class="toc-section-number">8.3</span> Bioconductor</a></li>
<li><a href="microarrays.html#microarrays"><span class="toc-section-number">8.4</span> Microarrays</a></li>
<li><a href="high-throughput-sequencing.html#high-throughput-sequencing"><span class="toc-section-number">8.5</span> High Throughput Sequencing</a>
<ul>
<li><a href="high-throughput-sequencing.html#raw-hts-data"><span class="toc-section-number">8.5.1</span> Raw HTS Data</a></li>
<li><a href="high-throughput-sequencing.html#preliminary-hts-data-analysis"><span class="toc-section-number">8.5.2</span> Preliminary HTS Data Analysis</a></li>
<li><a href="high-throughput-sequencing.html#count-data"><span class="toc-section-number">8.5.3</span> Count Data</a></li>
</ul></li>
<li><a href="gene-identifiers.html#gene-identifiers"><span class="toc-section-number">8.6</span> Gene Identifiers</a>
<ul>
<li><a href="gene-identifiers.html#gene-identifier-systems"><span class="toc-section-number">8.6.1</span> Gene Identifier Systems</a></li>
<li><a href="gene-identifiers.html#mapping-between-identifier-systems"><span class="toc-section-number">8.6.2</span> Mapping Between Identifier Systems</a></li>
<li><a href="gene-identifiers.html#mapping-homologs"><span class="toc-section-number">8.6.3</span> Mapping Homologs</a></li>
</ul></li>
<li><a href="gene-expression.html#gene-expression"><span class="toc-section-number">8.7</span> Gene Expression</a>
<ul>
<li><a href="gene-expression.html#gene-expression-data-in-bioconductor"><span class="toc-section-number">8.7.1</span> Gene Expression Data in Bioconductor</a></li>
<li><a href="gene-expression.html#differential-expression-analysis"><span class="toc-section-number">8.7.2</span> Differential Expression Analysis</a></li>
<li><a href="gene-expression.html#microarray-gene-expression-data"><span class="toc-section-number">8.7.3</span> Microarray Gene Expression Data</a></li>
<li><a href="gene-expression.html#differential-expression-microarrays-limma"><span class="toc-section-number">8.7.4</span> Differential Expression: Microarrays (limma)</a></li>
<li><a href="gene-expression.html#rnaseq"><span class="toc-section-number">8.7.5</span> RNASeq</a></li>
<li><a href="gene-expression.html#rnaseq-gene-expression-data"><span class="toc-section-number">8.7.6</span> RNASeq Gene Expression Data</a></li>
<li><a href="gene-expression.html#filtering-counts"><span class="toc-section-number">8.7.7</span> Filtering Counts</a></li>
<li><a href="gene-expression.html#count-distributions"><span class="toc-section-number">8.7.8</span> Count Distributions</a></li>
<li><a href="gene-expression.html#differential-expression-rnaseq"><span class="toc-section-number">8.7.9</span> Differential Expression: RNASeq</a></li>
</ul></li>
<li><a href="gene-set-enrichment-analysis.html#gene-set-enrichment-analysis"><span class="toc-section-number">8.8</span> Gene Set Enrichment Analysis</a>
<ul>
<li><a href="gene-set-enrichment-analysis.html#gene-sets"><span class="toc-section-number">8.8.1</span> Gene Sets</a></li>
<li><a href="gene-set-enrichment-analysis.html#working-with-gene-sets-in-r"><span class="toc-section-number">8.8.2</span> Working with gene sets in R</a></li>
<li><a href="gene-set-enrichment-analysis.html#over-representation-analysis"><span class="toc-section-number">8.8.3</span> Over-representation Analysis</a></li>
<li><a href="gene-set-enrichment-analysis.html#rank-based-analysis"><span class="toc-section-number">8.8.4</span> Rank-based Analysis</a></li>
<li><a href="gene-set-enrichment-analysis.html#fgsea"><span class="toc-section-number">8.8.5</span> fgsea</a></li>
</ul></li>
<li><a href="biological-networks-..html#biological-networks-."><span class="toc-section-number">8.9</span> Biological Networks .</a>
<ul>
<li><a href="biological-networks-..html#biological-pathways-."><span class="toc-section-number">8.9.1</span> Biological Pathways .</a></li>
<li><a href="biological-networks-..html#gene-regulatory-networks-."><span class="toc-section-number">8.9.2</span> Gene Regulatory Networks .</a></li>
<li><a href="biological-networks-..html#protein-protein-interaction-network-."><span class="toc-section-number">8.9.3</span> Protein-Protein Interaction Network .</a></li>
<li><a href="biological-networks-..html#wgcna-."><span class="toc-section-number">8.9.4</span> WGCNA .</a></li>
</ul></li>
</ul></li>
<li><a href="engineering.html#engineering"><span class="toc-section-number">9</span> EngineeRing</a>
<ul>
<li><a href="unit-testing.html#unit-testing"><span class="toc-section-number">9.1</span> Unit Testing</a></li>
<li><a href="toolification.html#toolification"><span class="toc-section-number">9.2</span> Toolification</a>
<ul>
<li><a href="toolification.html#the-r-interpreter"><span class="toc-section-number">9.2.1</span> The R Interpreter</a></li>
<li><a href="toolification.html#rscript"><span class="toc-section-number">9.2.2</span> <code>Rscript</code></a></li>
<li><a href="toolification.html#commandargs"><span class="toc-section-number">9.2.3</span> <code>commandArgs()</code></a></li>
</ul></li>
<li><a href="parallel-processing.html#parallel-processing"><span class="toc-section-number">9.3</span> Parallel Processing</a>
<ul>
<li><a href="parallel-processing.html#brief-introduction-to-parallelization"><span class="toc-section-number">9.3.1</span> Brief Introduction to Parallelization</a></li>
<li><a href="parallel-processing.html#apply-and-friends-are-pleasingly-parallel"><span class="toc-section-number">9.3.2</span> <code>apply</code> and Friends Are Pleasingly Parallel</a></li>
<li><a href="parallel-processing.html#the-parallel-package"><span class="toc-section-number">9.3.3</span> The <code>parallel</code> package</a></li>
</ul></li>
<li><a href="object-orientation-in-r.html#object-orientation-in-r"><span class="toc-section-number">9.4</span> Object Orientation in R</a></li>
<li><a href="r-packages.html#r-packages"><span class="toc-section-number">9.5</span> R Packages</a></li>
</ul></li>
<li><a href="rshiny.html#rshiny"><span class="toc-section-number">10</span> RShiny</a>
<ul>
<li><a href="rshiny-introduction.html#rshiny-introduction"><span class="toc-section-number">10.1</span> RShiny Introduction</a>
<ul>
<li><a href="rshiny-introduction.html#how-does-a-web-page-work"><span class="toc-section-number">10.1.1</span> How does a web page work?</a></li>
</ul></li>
<li><a href="application-structure.html#application-structure"><span class="toc-section-number">10.2</span> Application Structure</a>
<ul>
<li><a href="application-structure.html#ui-design"><span class="toc-section-number">10.2.1</span> UI Design</a></li>
<li><a href="application-structure.html#server-functionality"><span class="toc-section-number">10.2.2</span> Server Functionality</a></li>
</ul></li>
<li><a href="reactivity.html#reactivity"><span class="toc-section-number">10.3</span> Reactivity</a></li>
<li><a href="publishing.html#publishing"><span class="toc-section-number">10.4</span> Publishing</a></li>
</ul></li>
<li><a href="communicating-with-r.html#communicating-with-r"><span class="toc-section-number">11</span> Communicating with R</a>
<ul>
<li><a href="rmarkdown-knitr.html#rmarkdown-knitr"><span class="toc-section-number">11.1</span> RMarkdown &amp; knitr</a></li>
<li><a href="bookdown.html#bookdown"><span class="toc-section-number">11.2</span> bookdown</a></li>
</ul></li>
<li><a href="contribution-guide.html#contribution-guide"><span class="toc-section-number">12</span> Contribution Guide</a>
<ul>
<li><a href="custom-blocks.html#custom-blocks"><span class="toc-section-number">12.1</span> Custom blocks</a></li>
</ul></li>
<li><a href="assignments.html#assignments">Assignments</a>
<ul>
<li><a href="assignment-format.html#assignment-format">Assignment Format</a></li>
<li><a href="starting-an-assignment.html#starting-an-assignment">Starting an Assignment</a>
<ul>
<li><a href="starting-an-assignment.html#starting-your-own-git-repository-and-cloning-it">Starting your own <code>git</code> repository and cloning it</a></li>
<li><a href="starting-an-assignment.html#contents">Contents</a></li>
<li><a href="starting-an-assignment.html#github-tutorial">GitHub tutorial</a></li>
<li><a href="starting-an-assignment.html#committing-and-pushing-with-r-and-without">Committing and Pushing with R (and without)</a></li>
<li><a href="starting-an-assignment.html#using-rstudio">Using RStudio</a></li>
<li><a href="starting-an-assignment.html#using-the-command-line">Using the command line</a></li>
</ul></li>
<li><a href="assignment-1.html#assignment-1">Assignment 1</a>
<ul>
<li><a href="assignment-1.html#problem-statement">Problem Statement</a></li>
<li><a href="assignment-1.html#learning-objectives-and-skill-list">Learning Objectives and Skill List</a></li>
<li><a href="assignment-1.html#instructions">Instructions</a></li>
<li><a href="assignment-1.html#hints">Hints</a></li>
</ul></li>
<li><a href="assignment-2.html#assignment-2">Assignment 2</a>
<ul>
<li><a href="assignment-2.html#problem-statement-1">Problem Statement</a></li>
<li><a href="assignment-2.html#learning-objectives">Learning Objectives</a></li>
<li><a href="assignment-2.html#skill-list">Skill List</a></li>
<li><a href="assignment-2.html#instructions-1">Instructions</a></li>
<li><a href="assignment-2.html#function-details">Function Details</a></li>
</ul></li>
<li><a href="assignment-3.html#assignment-3">Assignment 3</a>
<ul>
<li><a href="assignment-3.html#problem-statement-2">Problem Statement</a></li>
<li><a href="assignment-3.html#learning-objectives-1">Learning Objectives</a></li>
<li><a href="assignment-3.html#skill-list-1">Skill List</a></li>
<li><a href="assignment-3.html#background-on-microarrays">Background on Microarrays</a></li>
<li><a href="assignment-3.html#background-on-principal-component-analysis">Background on Principal Component Analysis</a></li>
<li><a href="assignment-3.html#marisa-et-al.-gene-expression-classification-of-colon-cancer-into-molecular-subtypes-characterization-validation-and-prognostic-value.-plos-medicine-may-2013.-pmid-23700391">Marisa et al. Gene Expression Classification of Colon Cancer into Molecular Subtypes: Characterization, Validation, and Prognostic Value. PLoS Medicine, May 2013. PMID: 23700391</a></li>
<li><a href="assignment-3.html#scaling-data-using-r-scale">Scaling data using R <code>scale()</code></a></li>
<li><a href="assignment-3.html#proportion-of-variance-explained">Proportion of variance explained</a></li>
<li><a href="assignment-3.html#plotting-and-visualization-of-pca">Plotting and visualization of PCA</a></li>
<li><a href="assignment-3.html#hierarchical-clustering-and-heatmaps">Hierarchical Clustering and Heatmaps</a></li>
<li><a href="assignment-3.html#references">References</a></li>
</ul></li>
<li><a href="assignment-4.html#assignment-4">Assignment 4</a>
<ul>
<li><a href="assignment-4.html#problem-statement-3">Problem Statement</a></li>
<li><a href="assignment-4.html#learning-objectives-2">Learning Objectives</a></li>
<li><a href="assignment-4.html#skill-list-2">Skill List</a></li>
<li><a href="assignment-4.html#instructions-2">Instructions</a></li>
<li><a href="assignment-4.html#deliverables">Deliverables</a></li>
<li><a href="assignment-4.html#function-details-1">Function Details</a></li>
<li><a href="assignment-4.html#hints-1">Hints</a></li>
</ul></li>
<li><a href="assignment-5.html#assignment-5">Assignment 5</a>
<ul>
<li><a href="assignment-5.html#problem-statement-4">Problem Statement</a></li>
<li><a href="assignment-5.html#learning-objectives-3">Learning Objectives</a></li>
<li><a href="assignment-5.html#skill-list-3">Skill List</a></li>
<li><a href="assignment-5.html#deseq2-background">DESeq2 Background</a></li>
<li><a href="assignment-5.html#generating-a-counts-matrix">Generating a counts matrix</a></li>
<li><a href="assignment-5.html#prefiltering-counts-matrix">Prefiltering Counts matrix</a></li>
<li><a href="assignment-5.html#median-of-ratios-normalization">Median-of-ratios normalization</a></li>
<li><a href="assignment-5.html#deseq2-preparation">DESeq2 preparation</a></li>
<li><a href="assignment-5.html#omeara-et-al.-transcriptional-reversion-of-cardiac-myocyte-fate-during-mammalian-cardiac-regeneration.-circ-res.-feb-2015.-pmid-25477501l">O’Meara et al. Transcriptional Reversion of Cardiac Myocyte Fate During Mammalian Cardiac Regeneration. Circ Res. Feb 2015. PMID: 25477501l</a></li>
<li><a href="assignment-5.html#reading-and-subsetting-the-data-from-verse_counts.tsv-and-sample_metadata.csv">1. Reading and subsetting the data from verse_counts.tsv and sample_metadata.csv</a></li>
<li><a href="assignment-5.html#running-deseq2">2. Running DESeq2</a></li>
<li><a href="assignment-5.html#annotating-results-to-construct-a-labeled-volcano-plot">3. Annotating results to construct a labeled volcano plot</a></li>
<li><a href="assignment-5.html#diagnostic-plot-of-the-raw-p-values-for-all-genes">4. Diagnostic plot of the raw p-values for all genes</a></li>
<li><a href="assignment-5.html#plotting-the-logfoldchanges-for-differentially-expressed-genes">5. Plotting the LogFoldChanges for differentially expressed genes</a></li>
<li><a href="assignment-5.html#the-choice-of-fdr-cutoff-depends-on-cost">The choice of FDR cutoff depends on cost</a></li>
<li><a href="assignment-5.html#plotting-the-normalized-counts-of-differentially-expressed-genes">6. Plotting the normalized counts of differentially expressed genes</a></li>
<li><a href="assignment-5.html#volcano-plot-to-visualize-differential-expression-results">7. Volcano Plot to visualize differential expression results</a></li>
<li><a href="assignment-5.html#running-fgsea-vignette">8. Running fgsea <span>vignette</span></a></li>
<li><a href="assignment-5.html#plotting-the-top-ten-positive-nes-and-top-ten-negative-nes-pathways">9. Plotting the top ten positive NES and top ten negative NES pathways</a></li>
<li><a href="assignment-5.html#references-1">References</a></li>
</ul></li>
<li><a href="assignment-6.html#assignment-6">Assignment 6</a>
<ul>
<li><a href="assignment-6.html#problem-statement-5">Problem Statement</a></li>
<li><a href="assignment-6.html#learning-objectives-4">Learning Objectives</a></li>
<li><a href="assignment-6.html#skill-list-4">Skill List</a></li>
<li><a href="assignment-6.html#instructions-3">Instructions</a></li>
<li><a href="assignment-6.html#function-details-2">Function Details</a></li>
</ul></li>
<li><a href="assignment-7.html#assignment-7">Assignment 7</a>
<ul>
<li><a href="assignment-7.html#problem-statement-6">Problem Statement</a></li>
<li><a href="assignment-7.html#learning-objectives-5">Learning Objectives</a></li>
<li><a href="assignment-7.html#skill-list-5">Skill List</a></li>
<li><a href="assignment-7.html#instructions-4">Instructions</a></li>
<li><a href="assignment-7.html#function-details-3">“Function” Details</a></li>
</ul></li>
</ul></li>
<li class="appendix"><span><b>Appendix</b></span></li>
<li><a href="class-outlines.html#class-outlines"><span class="toc-section-number">A</span> Class Outlines</a>
<ul>
<li><a href="week-1.html#week-1"><span class="toc-section-number">A.1</span> Week 1</a></li>
<li><a href="week-2.html#week-2"><span class="toc-section-number">A.2</span> Week 2</a></li>
<li><a href="week-3.html#week-3"><span class="toc-section-number">A.3</span> Week 3</a></li>
<li><a href="week-4.html#week-4"><span class="toc-section-number">A.4</span> Week 4</a></li>
<li><a href="week-5.html#week-5"><span class="toc-section-number">A.5</span> Week 5</a></li>
<li><a href="week-6.html#week-6"><span class="toc-section-number">A.6</span> Week 6</a></li>
<li><a href="week-7.html#week-7"><span class="toc-section-number">A.7</span> Week 7</a></li>
<li><a href="week-8.html#week-8"><span class="toc-section-number">A.8</span> Week 8</a></li>
<li><a href="week-9.html#week-9"><span class="toc-section-number">A.9</span> Week 9</a></li>
<li><a href="week-10.html#week-10"><span class="toc-section-number">A.10</span> Week 10</a></li>
<li><a href="week-11.html#week-11"><span class="toc-section-number">A.11</span> Week 11</a></li>
<li><a href="week-12.html#week-12"><span class="toc-section-number">A.12</span> Week 12</a></li>
<li><a href="week-13.html#week-13"><span class="toc-section-number">A.13</span> Week 13</a></li>
<li><a href="week-14.html#week-14"><span class="toc-section-number">A.14</span> Week 14</a></li>
</ul></li>
</ul>
</div>
</div>
</div>
<div class="row">
<div class="col-sm-12">
<div id="exploratory-data-analysis" class="section level2" number="6.4">
<h2><span class="header-section-number">6.4</span> Exploratory Data Analysis</h2>
<p>So far, we have discussed methods where we chose an explicit model to summarize
our data. However, sometimes we don’t have enough information about our data to
propose reasonable models, as we did earlier when exploring the distribution of
our imaginary gene expression dataset. There may be patterns in the data that
emerge when we compute different summaries and ask whether there is a non-random
structure to how the individual samples or features compare with one another.
The types of methods we use to take a dataset and examine it for structure
without a prefigured hypothesis is called <em>exploratory analysis</em> and is a key
approach to working with biological data.</p>
<div id="principal-component-analysis" class="section level3" number="6.4.1">
<h3><span class="header-section-number">6.4.1</span> Principal Component Analysis</h3>
<p>A very common method of exploratory analysis is <a href="https://en.wikipedia.org/wiki/Principal_component_analysis">principal component
analysis</a> or PCA.
PCA is a statistical procedure that identifies so called <em>directions of
orthogonal variance</em> that capture covariance between different dimensions in a
dataset. Because the approach captures this covariance between features of
arbitrary dimension, it is often used for so-called <a href="https://en.wikipedia.org/wiki/Dimensionality_reduction">dimensionality
reduction</a>, where a
large amount of variance in a dataset with a potentially large number of
dimensions may be expressed in terms of a set of <a href="https://en.wikipedia.org/wiki/Change_of_basis">basis
vectors</a> of smaller dimension.
The mathematical details of this approach are beyond the scope of this book, but
below we explain in general terms the intuition behind what PCA does, and
present an example of how it is used in a biological context.</p>
<p>PCA decomposes a dataset into a set of <a href="https://en.wikipedia.org/wiki/Orthonormal_basis">orthonormal basis
vectors</a> that collectively
capture all the variance in the dataset, where the first basis vector, called
the <em>first principal component</em> explains the largest fraction of the variance,
the second principal component explains the second largest fraction, and so on.
There are always an equal number of principal components as there are dimensions
in the dataset or the number of samples, whichever is smaller. Typically only a
small number of these components are needed to explain most of the variance.</p>
<p>Each principal component is a <span class="math inline">\(p\)</span>-dimensional <a href="https://en.wikipedia.org/wiki/Unit_vector">unit
vector</a> (i.e. a vector of magnitude
1), where <span class="math inline">\(p\)</span> is the number of features in the dataset, and the values are
weights that describe the component’s direction of variance. By multiplying this
component with the values in each sample, we obtain the <em>projection</em> of each
sample with respect to the basis of the component. The projections of each
sample made with each principal component produces a rotation of the dataset in
<span class="math inline">\(p\)</span> dimensional space. The figure below presents a geometric intuition of PCA.</p>
<div class="figure">
<img src="PCA.png" alt="" />
<p class="caption">Principal Component Analysis - Geometric Intuition Illustration</p>
</div>
<p>Many biological datasets, especially those that make genome-wide measurements
like with gene expression assays, have many thousands of features (e.g. genes)
and comparatively few samples. Since PCA can only determine a maximum number of
principal components as the smaller of the number of features or samples, we
will almost always only have as many components as samples. To demonstrate this,
we perform PCA using the
<a href="https://www.rdocumentation.org/packages/stats/versions/3.6.2/topics/prcomp"><code>stats::prcomp()</code></a>
function on an example microarray gene expression intensity dataset:</p>
<div class="sourceCode" id="cb228"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb228-1"><a href="exploratory-data-analysis.html#cb228-1" aria-hidden="true" tabindex="-1"></a><span class="co"># intensities contains microarray expression data for ~54k probesets x 20 samples</span></span>
<span id="cb228-2"><a href="exploratory-data-analysis.html#cb228-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb228-3"><a href="exploratory-data-analysis.html#cb228-3" aria-hidden="true" tabindex="-1"></a><span class="co"># transpose expression values so samples are rows</span></span>
<span id="cb228-4"><a href="exploratory-data-analysis.html#cb228-4" aria-hidden="true" tabindex="-1"></a>expr_mat <span class="ot">&lt;-</span> intensities <span class="sc">%&gt;%</span></span>
<span id="cb228-5"><a href="exploratory-data-analysis.html#cb228-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(<span class="sc">-</span><span class="fu">c</span>(probeset_id),<span class="at">names_to=</span><span class="st">&quot;sample&quot;</span>) <span class="sc">%&gt;%</span></span>
<span id="cb228-6"><a href="exploratory-data-analysis.html#cb228-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_wider</span>(<span class="at">names_from=</span>probeset_id)</span>
<span id="cb228-7"><a href="exploratory-data-analysis.html#cb228-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb228-8"><a href="exploratory-data-analysis.html#cb228-8" aria-hidden="true" tabindex="-1"></a><span class="co"># PCA expects all features (i.e. probesets) to be mean-centered,</span></span>
<span id="cb228-9"><a href="exploratory-data-analysis.html#cb228-9" aria-hidden="true" tabindex="-1"></a><span class="co"># convert to dataframe so we can use rownames</span></span>
<span id="cb228-10"><a href="exploratory-data-analysis.html#cb228-10" aria-hidden="true" tabindex="-1"></a>expr_mat_centered <span class="ot">&lt;-</span>  <span class="fu">as.data.frame</span>(</span>
<span id="cb228-11"><a href="exploratory-data-analysis.html#cb228-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">lapply</span>(dplyr<span class="sc">::</span><span class="fu">select</span>(expr_mat,<span class="sc">-</span><span class="fu">c</span>(sample)),<span class="cf">function</span>(x) x<span class="sc">-</span><span class="fu">mean</span>(x))</span>
<span id="cb228-12"><a href="exploratory-data-analysis.html#cb228-12" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb228-13"><a href="exploratory-data-analysis.html#cb228-13" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(expr_mat_centered) <span class="ot">&lt;-</span> expr_mat<span class="sc">$</span>sample</span>
<span id="cb228-14"><a href="exploratory-data-analysis.html#cb228-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb228-15"><a href="exploratory-data-analysis.html#cb228-15" aria-hidden="true" tabindex="-1"></a><span class="co"># prcomp performs PCA</span></span>
<span id="cb228-16"><a href="exploratory-data-analysis.html#cb228-16" aria-hidden="true" tabindex="-1"></a>pca <span class="ot">&lt;-</span> <span class="fu">prcomp</span>(</span>
<span id="cb228-17"><a href="exploratory-data-analysis.html#cb228-17" aria-hidden="true" tabindex="-1"></a>  expr_mat_centered,</span>
<span id="cb228-18"><a href="exploratory-data-analysis.html#cb228-18" aria-hidden="true" tabindex="-1"></a>  <span class="at">center=</span><span class="cn">FALSE</span>, <span class="co"># prcomp centers features to have mean zero by default, but we already did it</span></span>
<span id="cb228-19"><a href="exploratory-data-analysis.html#cb228-19" aria-hidden="true" tabindex="-1"></a>  <span class="at">scale=</span><span class="cn">TRUE</span> <span class="co"># prcomp scales each feature to have unit variance</span></span>
<span id="cb228-20"><a href="exploratory-data-analysis.html#cb228-20" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb228-21"><a href="exploratory-data-analysis.html#cb228-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb228-22"><a href="exploratory-data-analysis.html#cb228-22" aria-hidden="true" tabindex="-1"></a><span class="co"># the str() function prints the structure of its argument</span></span>
<span id="cb228-23"><a href="exploratory-data-analysis.html#cb228-23" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(pca)</span></code></pre></div>
<pre><code>## List of 5
##  $ sdev    : num [1:20] 101.5 81 71.2 61.9 57.9 ...
##  $ rotation: num [1:54675, 1:20] 0.00219 0.00173 -0.00313 0.00465 0.0034 ...
##   ..- attr(*, &quot;dimnames&quot;)=List of 2
##   .. ..$ : chr [1:54675] &quot;X1007_s_at&quot; &quot;X1053_at&quot; &quot;X117_at&quot; &quot;X121_at&quot; ...
##   .. ..$ : chr [1:20] &quot;PC1&quot; &quot;PC2&quot; &quot;PC3&quot; &quot;PC4&quot; ...
##  $ center  : logi FALSE
##  $ scale   : Named num [1:54675] 0.379 0.46 0.628 0.272 0.196 ...
##   ..- attr(*, &quot;names&quot;)= chr [1:54675] &quot;X1007_s_at&quot; &quot;X1053_at&quot; &quot;X117_at&quot; &quot;X121_at&quot; ...
##  $ x       : num [1:20, 1:20] -67.94 186.14 6.07 -70.72 9.58 ...
##   ..- attr(*, &quot;dimnames&quot;)=List of 2
##   .. ..$ : chr [1:20] &quot;A1&quot; &quot;A2&quot; &quot;A3&quot; &quot;A4&quot; ...
##   .. ..$ : chr [1:20] &quot;PC1&quot; &quot;PC2&quot; &quot;PC3&quot; &quot;PC4&quot; ...
##  - attr(*, &quot;class&quot;)= chr &quot;prcomp&quot;</code></pre>
<p>The result of <code>prcomp()</code> is a list with five members:</p>
<ul>
<li><code>sdev</code> - the standard deviation (i.e. the square root of the variance) for each component</li>
<li><code>rotation</code> - a matrix where the principal components are in columns</li>
<li><code>x</code> - the projections of the original data, aka the rotated data matrix</li>
<li><code>center</code> - if <code>center=TRUE</code> was passed, a vector of feature means</li>
<li><code>scale</code> - if <code>scale=TRUE</code> was passed, a vector of the feature variances</li>
</ul>
<p>Recall that each principal component explains a fraction of the overall variance
in the dataset. The <code>sdev</code> variable returned by <code>prcomp()</code> may be used to first
calculate the variance explained by each component by squaring it, then dividing
by the sum:</p>
<div class="sourceCode" id="cb230"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb230-1"><a href="exploratory-data-analysis.html#cb230-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(patchwork)</span>
<span id="cb230-2"><a href="exploratory-data-analysis.html#cb230-2" aria-hidden="true" tabindex="-1"></a>pca_var <span class="ot">&lt;-</span> <span class="fu">tibble</span>(</span>
<span id="cb230-3"><a href="exploratory-data-analysis.html#cb230-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">PC=</span><span class="fu">factor</span>(<span class="fu">str_c</span>(<span class="st">&quot;PC&quot;</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">20</span>),<span class="fu">str_c</span>(<span class="st">&quot;PC&quot;</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">20</span>)),</span>
<span id="cb230-4"><a href="exploratory-data-analysis.html#cb230-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">Variance=</span>pca<span class="sc">$</span>sdev<span class="sc">**</span><span class="dv">2</span>,</span>
<span id="cb230-5"><a href="exploratory-data-analysis.html#cb230-5" aria-hidden="true" tabindex="-1"></a>  <span class="st">`</span><span class="at">% Explained Variance</span><span class="st">`</span><span class="ot">=</span>Variance<span class="sc">/</span><span class="fu">sum</span>(Variance)<span class="sc">*</span><span class="dv">100</span>,</span>
<span id="cb230-6"><a href="exploratory-data-analysis.html#cb230-6" aria-hidden="true" tabindex="-1"></a>  <span class="st">`</span><span class="at">Cumulative % Explained Variance</span><span class="st">`</span><span class="ot">=</span><span class="fu">cumsum</span>(<span class="st">`</span><span class="at">% Explained Variance</span><span class="st">`</span>)</span>
<span id="cb230-7"><a href="exploratory-data-analysis.html#cb230-7" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb230-8"><a href="exploratory-data-analysis.html#cb230-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb230-9"><a href="exploratory-data-analysis.html#cb230-9" aria-hidden="true" tabindex="-1"></a>exp_g <span class="ot">&lt;-</span> pca_var <span class="sc">%&gt;%</span></span>
<span id="cb230-10"><a href="exploratory-data-analysis.html#cb230-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x=</span>PC,<span class="at">y=</span><span class="st">`</span><span class="at">% Explained Variance</span><span class="st">`</span>,<span class="at">group=</span><span class="dv">1</span>)) <span class="sc">+</span></span>
<span id="cb230-11"><a href="exploratory-data-analysis.html#cb230-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb230-12"><a href="exploratory-data-analysis.html#cb230-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_line</span>() <span class="sc">+</span></span>
<span id="cb230-13"><a href="exploratory-data-analysis.html#cb230-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.text.x=</span><span class="fu">element_text</span>(<span class="at">angle=</span><span class="dv">90</span>,<span class="at">hjust=</span><span class="dv">0</span>,<span class="at">vjust=</span><span class="fl">0.5</span>)) <span class="co"># make x labels rotate 90 degrees</span></span>
<span id="cb230-14"><a href="exploratory-data-analysis.html#cb230-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb230-15"><a href="exploratory-data-analysis.html#cb230-15" aria-hidden="true" tabindex="-1"></a>cum_g <span class="ot">&lt;-</span> pca_var <span class="sc">%&gt;%</span></span>
<span id="cb230-16"><a href="exploratory-data-analysis.html#cb230-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x=</span>PC,<span class="at">y=</span><span class="st">`</span><span class="at">Cumulative % Explained Variance</span><span class="st">`</span>,<span class="at">group=</span><span class="dv">1</span>)) <span class="sc">+</span></span>
<span id="cb230-17"><a href="exploratory-data-analysis.html#cb230-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb230-18"><a href="exploratory-data-analysis.html#cb230-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_line</span>() <span class="sc">+</span></span>
<span id="cb230-19"><a href="exploratory-data-analysis.html#cb230-19" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylim</span>(<span class="dv">0</span>,<span class="dv">100</span>) <span class="sc">+</span> <span class="co"># set y limits to [0,100]</span></span>
<span id="cb230-20"><a href="exploratory-data-analysis.html#cb230-20" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.text.x=</span><span class="fu">element_text</span>(<span class="at">angle=</span><span class="dv">90</span>,<span class="at">hjust=</span><span class="dv">0</span>,<span class="at">vjust=</span><span class="fl">0.5</span>))</span>
<span id="cb230-21"><a href="exploratory-data-analysis.html#cb230-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb230-22"><a href="exploratory-data-analysis.html#cb230-22" aria-hidden="true" tabindex="-1"></a>exp_g <span class="sc">|</span> cum_g</span></code></pre></div>
<p><img src="main_files/pca%20var-1.png" width="672" /></p>
<p>The first component explains nearly 20% of all variance in the dataset, followed
by the second component with about 12%, the third component 9%, and so on. The
cumulative variance plot on the right shows that the top 15 components are
required to capture 90% of the variance in the dataset. This suggests that each
sample contributes a significant amount of variance to the overall dataset, but
that there are still some features that covary among them.</p>
<p>An important use of PCA results is the identification of outlier samples. This
is accomplished by plotting the projections of each sample and examining the
result by eye to identify samples that are “far away” from the other samples.
This is usually done by inspection and outlier samples chosen subjectively;
there are no general rules to decide when a sample is an outlier by this method.
Below the projections for components 1 and 2 are plotted against each other as a
scatter plot:</p>
<div class="sourceCode" id="cb231"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb231-1"><a href="exploratory-data-analysis.html#cb231-1" aria-hidden="true" tabindex="-1"></a><span class="fu">as_tibble</span>(pca<span class="sc">$</span>x) <span class="sc">%&gt;%</span></span>
<span id="cb231-2"><a href="exploratory-data-analysis.html#cb231-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x=</span>PC1,<span class="at">y=</span>PC2)) <span class="sc">+</span></span>
<span id="cb231-3"><a href="exploratory-data-analysis.html#cb231-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>()</span></code></pre></div>
<p><img src="main_files/pca%20pairwise-1.png" width="672" /></p>
<p>By eye, no samples appear to be obvious outliers. However, this plot is just one
of many pairs of projections. We could plot all pairs of the first six
components using the <a href="https://ggobi.github.io/ggally/reference/ggpairs.html"><code>ggpairs() function</code></a> in the
<a href="https://ggobi.github.io/ggally/">GGally</a> package:</p>
<div class="sourceCode" id="cb232"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb232-1"><a href="exploratory-data-analysis.html#cb232-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(GGally)</span>
<span id="cb232-2"><a href="exploratory-data-analysis.html#cb232-2" aria-hidden="true" tabindex="-1"></a><span class="fu">as_tibble</span>(pca<span class="sc">$</span>x) <span class="sc">%&gt;%</span></span>
<span id="cb232-3"><a href="exploratory-data-analysis.html#cb232-3" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">select</span>(PC1<span class="sc">:</span>PC6) <span class="sc">%&gt;%</span></span>
<span id="cb232-4"><a href="exploratory-data-analysis.html#cb232-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggpairs</span>()</span></code></pre></div>
<p><img src="main_files/pca%20pairs-1.png" width="672" /></p>
<p>This is already nearly unintelligible and mostly uninformative. While it is very
common for projections for pairs of components to be plotted, an alternative way
to visualize projections across all samples is by plotting the distribution of
projections for each component using a <a href="#beeswarm-plots">beeswarm plot</a>:</p>
<div class="sourceCode" id="cb233"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb233-1"><a href="exploratory-data-analysis.html#cb233-1" aria-hidden="true" tabindex="-1"></a><span class="fu">as_tibble</span>(pca<span class="sc">$</span>x) <span class="sc">%&gt;%</span></span>
<span id="cb233-2"><a href="exploratory-data-analysis.html#cb233-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(<span class="fu">everything</span>(),<span class="at">names_to=</span><span class="st">&quot;PC&quot;</span>,<span class="at">values_to=</span><span class="st">&quot;projection&quot;</span>) <span class="sc">%&gt;%</span></span>
<span id="cb233-3"><a href="exploratory-data-analysis.html#cb233-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">PC=</span><span class="fu">fct_relevel</span>(PC,<span class="fu">str_c</span>(<span class="st">&quot;PC&quot;</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">20</span>))) <span class="sc">%&gt;%</span></span>
<span id="cb233-4"><a href="exploratory-data-analysis.html#cb233-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x=</span>PC,<span class="at">y=</span>projection)) <span class="sc">+</span></span>
<span id="cb233-5"><a href="exploratory-data-analysis.html#cb233-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_beeswarm</span>() <span class="sc">+</span> <span class="fu">labs</span>(<span class="at">title=</span><span class="st">&quot;PCA Projection Plot&quot;</span>) <span class="sc">+</span></span>
<span id="cb233-6"><a href="exploratory-data-analysis.html#cb233-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.text.x=</span><span class="fu">element_text</span>(<span class="at">angle=</span><span class="dv">90</span>,<span class="at">hjust=</span><span class="dv">0</span>,<span class="at">vjust=</span><span class="fl">0.5</span>))</span></code></pre></div>
<p><img src="main_files/pca%20proj%20beeswarm-1.png" width="672" /></p>
<p>Now we can see all the projections for all components in the same plot, although
we cannot see how they relate to one another.</p>
<p>These projection plots may become more useful if we layer on additional
information about our samples. There are two types of samples in our dataset (
labelled <code>A</code> and <code>C</code>. We can color our pairwise scatter plot by type like so:</p>
<div class="sourceCode" id="cb234"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb234-1"><a href="exploratory-data-analysis.html#cb234-1" aria-hidden="true" tabindex="-1"></a><span class="fu">as_tibble</span>(pca<span class="sc">$</span>x) <span class="sc">%&gt;%</span></span>
<span id="cb234-2"><a href="exploratory-data-analysis.html#cb234-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">type=</span>stringr<span class="sc">::</span><span class="fu">str_sub</span>(<span class="fu">rownames</span>(pca<span class="sc">$</span>x),<span class="dv">1</span>,<span class="dv">1</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb234-3"><a href="exploratory-data-analysis.html#cb234-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x=</span>PC1,<span class="at">y=</span>PC2,<span class="at">color=</span>type)) <span class="sc">+</span></span>
<span id="cb234-4"><a href="exploratory-data-analysis.html#cb234-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>()</span></code></pre></div>
<p><img src="main_files/pca%20color%20scatter-1.png" width="672" /></p>
<p>Little pattern is obvious from the plot, but we can plot pairs of components as
before but now with type information:</p>
<div class="sourceCode" id="cb235"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb235-1"><a href="exploratory-data-analysis.html#cb235-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(GGally)</span>
<span id="cb235-2"><a href="exploratory-data-analysis.html#cb235-2" aria-hidden="true" tabindex="-1"></a><span class="fu">as_tibble</span>(pca<span class="sc">$</span>x) <span class="sc">%&gt;%</span></span>
<span id="cb235-3"><a href="exploratory-data-analysis.html#cb235-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">type=</span>stringr<span class="sc">::</span><span class="fu">str_sub</span>(<span class="fu">rownames</span>(pca<span class="sc">$</span>x),<span class="dv">1</span>,<span class="dv">1</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb235-4"><a href="exploratory-data-analysis.html#cb235-4" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">select</span>(<span class="fu">c</span>(type,PC1<span class="sc">:</span>PC6)) <span class="sc">%&gt;%</span></span>
<span id="cb235-5"><a href="exploratory-data-analysis.html#cb235-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggpairs</span>(<span class="at">columns=</span><span class="dv">1</span><span class="sc">:</span><span class="dv">6</span>,<span class="at">mapping=</span><span class="fu">aes</span>(<span class="at">fill=</span>type))</span></code></pre></div>
<p><img src="main_files/pca%20pairs%20type-1.png" width="672" /></p>
<p>Examining PC3 and PC4, we now observe that there may indeed be some genes that
distinguish between types based on the separation of projection scores for the
two types. Finally, we can also color our beeswarm plot by type:</p>
<div class="sourceCode" id="cb236"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb236-1"><a href="exploratory-data-analysis.html#cb236-1" aria-hidden="true" tabindex="-1"></a><span class="fu">as_tibble</span>(pca<span class="sc">$</span>x) <span class="sc">%&gt;%</span></span>
<span id="cb236-2"><a href="exploratory-data-analysis.html#cb236-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(</span>
<span id="cb236-3"><a href="exploratory-data-analysis.html#cb236-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">sample=</span><span class="fu">rownames</span>(pca<span class="sc">$</span>x),</span>
<span id="cb236-4"><a href="exploratory-data-analysis.html#cb236-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">type=</span>stringr<span class="sc">::</span><span class="fu">str_sub</span>(sample,<span class="dv">1</span>,<span class="dv">1</span>)</span>
<span id="cb236-5"><a href="exploratory-data-analysis.html#cb236-5" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">%&gt;%</span></span>
<span id="cb236-6"><a href="exploratory-data-analysis.html#cb236-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(PC1<span class="sc">:</span>PC20,<span class="at">names_to=</span><span class="st">&quot;PC&quot;</span>,<span class="at">values_to=</span><span class="st">&quot;projection&quot;</span>) <span class="sc">%&gt;%</span></span>
<span id="cb236-7"><a href="exploratory-data-analysis.html#cb236-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">PC=</span><span class="fu">fct_relevel</span>(PC,<span class="fu">str_c</span>(<span class="st">&quot;PC&quot;</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">20</span>))) <span class="sc">%&gt;%</span></span>
<span id="cb236-8"><a href="exploratory-data-analysis.html#cb236-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x=</span>PC,<span class="at">y=</span>projection,<span class="at">color=</span>type)) <span class="sc">+</span></span>
<span id="cb236-9"><a href="exploratory-data-analysis.html#cb236-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_beeswarm</span>() <span class="sc">+</span> <span class="fu">labs</span>(<span class="at">title=</span><span class="st">&quot;PCA Projection Plot&quot;</span>) <span class="sc">+</span></span>
<span id="cb236-10"><a href="exploratory-data-analysis.html#cb236-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.text.x=</span><span class="fu">element_text</span>(<span class="at">angle=</span><span class="dv">90</span>,<span class="at">hjust=</span><span class="dv">0</span>,<span class="at">vjust=</span><span class="fl">0.5</span>))</span></code></pre></div>
<p><img src="main_files/pca%20proj%20beeswarm%20type-1.png" width="672" /></p>
<p>These approaches to plotting the results of a PCA are complementary, and all
may be useful in understanding the structure of a dataset.</p>
</div>
<div id="cluster-analysis" class="section level3" number="6.4.2">
<h3><span class="header-section-number">6.4.2</span> Cluster Analysis</h3>
<p>Cluster analysis or simple clustering is the task of grouping objects together
such that objects within the same group, or <em>cluster</em>, are more similar to each
other than to objects in other clusters. It is a type of <a href="https://en.wikipedia.org/wiki/Exploratory_data_analysis">exploratory data
analysis</a> that seeks to
identify structure or organization in a dataset without making strong
assumptions about the data or testing a specific hypothesis. Many different
<a href="https://en.wikipedia.org/wiki/Category:Cluster_analysis_algorithms">clustering algorithms</a>
have been developed that employ different computational strategies and are
designed for data with different types and properties. The choice of algorithm
is often dependent upon not only the type of data being clustered but also the
comparative performance of different clustering algorithms applied to the same
data. Different algorithms may identify different types of patterns, and so no
one clustering method is better than any other in general.</p>
<p>The goal of clustering may be easily illustrated with the following plot:</p>
<div class="sourceCode" id="cb237"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb237-1"><a href="exploratory-data-analysis.html#cb237-1" aria-hidden="true" tabindex="-1"></a>unclustered <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(well_clustered_data, <span class="fu">aes</span>(<span class="at">x=</span>f1,<span class="at">y=</span>f2)) <span class="sc">+</span></span>
<span id="cb237-2"><a href="exploratory-data-analysis.html#cb237-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb237-3"><a href="exploratory-data-analysis.html#cb237-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title=</span><span class="st">&quot;Unclustered&quot;</span>)</span>
<span id="cb237-4"><a href="exploratory-data-analysis.html#cb237-4" aria-hidden="true" tabindex="-1"></a>clustered <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(well_clustered_data, <span class="fu">aes</span>(<span class="at">x=</span>f1,<span class="at">y=</span>f2,<span class="at">color=</span>cluster,<span class="at">shape=</span>cluster)) <span class="sc">+</span></span>
<span id="cb237-5"><a href="exploratory-data-analysis.html#cb237-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb237-6"><a href="exploratory-data-analysis.html#cb237-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title=</span><span class="st">&quot;Clustered&quot;</span>)</span>
<span id="cb237-7"><a href="exploratory-data-analysis.html#cb237-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb237-8"><a href="exploratory-data-analysis.html#cb237-8" aria-hidden="true" tabindex="-1"></a>unclustered <span class="sc">|</span> clustered</span></code></pre></div>
<p><img src="main_files/clustering%20goal-1.png" width="672" /></p>
<p>That is, the goal of clustering is to take unlabeled data that has some
structure like on the left and label data points that are similar to one another
to be in the same cluster. A full treatment of cluster analysis is beyond the
scope of this book, but below we discuss a few of the most common and general
clustering algorithms used in biology and bioinformatics.</p>
<div id="hierarchical-clustering" class="section level4" number="6.4.2.1">
<h4><span class="header-section-number">6.4.2.1</span> Hierarchical Clustering</h4>
<p>Hierarchical clustering is a form of clustering that clusters data points
together in nested, or <em>hierarchical</em>, groups based on their dissimilarity from
one another. There are two broad strategies to accomplish this:</p>
<ul>
<li><strong>Agglomerative</strong>: all data points start in their own groups, and groups are
iteratively merged hierarchically into larger groups based on their similarity
until all data points have been added to a group</li>
<li><strong>Divisive</strong>: all data points start in the same group, and are recursively
split into smaller groups based on their dissimilarity</li>
</ul>
<p>Whichever approach is used, the critical step in clustering is choosing the
<a href="https://en.wikipedia.org/wiki/Metric_(mathematics)">distance function</a> that
quantifies how dissimilar two data points are. Distance functions, or <em>metrics</em>
are non-negative real numbers whose magnitude is proportional to some notion of
distance between a pair of points. There are <a href="https://en.wikipedia.org/wiki/Metric_(mathematics)#Generalized_metrics">many different distance
functions</a>
to choose from, and the choice is made based on the type of data being
clustered. For numeric data, <a href="https://en.wikipedia.org/wiki/Euclidean_distance">euclidean
distance</a>, which corresponds
to the length of a line segment drawn between two points in any <span class="math inline">\(n\)</span>-dimensional
<a href="https://en.wikipedia.org/wiki/Euclidean_space">Euclidean space</a> is often a
reasonable choice.</p>
<p>Because the divisive strategy is conceptually the inverse of
agglomerative clustering, we will limit our description to agglomerative
clustering. Once the distance function has been chosen, distances between all
pairs of points are computed and the two nearest points are merged into a group.
The new group of points is used for computing distances to all other points as a
set using a <em>linkage function</em>
which defines how the group is summarized into a single point for the purposes
of distance calculation. The linkage function is what distinguishes different
clustering algorithms, which include:</p>
<ul>
<li><a href="https://en.wikipedia.org/wiki/Single-linkage_clustering">Single-linkage</a> - distance between two groups is computed as the distance between the two nearest members of the groups</li>
<li><a href="https://en.wikipedia.org/wiki/Complete-linkage_clustering">Complete-linkage</a> - distance between two groups is computed as the distance between the two farthest members of the groups</li>
<li><a href="https://en.wikipedia.org/wiki/UPGMA">Unweighted Pair Group Method with Arithmetic mean (UPGMA)</a> - distance between two groups is the average distance of all pairs of points between groups</li>
<li><a href="https://en.wikipedia.org/wiki/WPGMA">Weighted Pair Group Method with Arithmetic mean (WPGMA)</a> - similar to UPGMA, but weights distances from pairs of groups evenly when merging</li>
</ul>
<p>The choice of linkage function (and therefore clustering algorithm) should be
determined based on knowledge of the dataset and assessment of clustering
performance. In general, there is no one linkage function that will always
perform well.</p>
<p>The following figure illustrates a simple example of clustering a 1 dimensional
set of points using WPGMA:</p>
<div class="figure">
<img src="hierarchical_clustering.png" alt="" />
<p class="caption">Conceptual illustration of agglomerative hierarchical clustering</p>
</div>
<p>Below we will cluster the synthetic dataset introduced above in R:</p>
<div class="sourceCode" id="cb238"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb238-1"><a href="exploratory-data-analysis.html#cb238-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(well_clustered_data, <span class="fu">aes</span>(<span class="at">x=</span>f1,<span class="at">y=</span>f2)) <span class="sc">+</span></span>
<span id="cb238-2"><a href="exploratory-data-analysis.html#cb238-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb238-3"><a href="exploratory-data-analysis.html#cb238-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title=</span><span class="st">&quot;Unclustered&quot;</span>)</span></code></pre></div>
<p><img src="main_files/hclust%20data-1.png" width="672" /></p>
<p>This synthetic dataset has two distinct groups of samples drawn from
multivariate normal samples. To hierarchically cluster these samples, we use the
<a href="https://www.rdocumentation.org/packages/stats/versions/3.6.2/topics/dist"><code>dist()</code></a>
and <a href="https://www.rdocumentation.org/packages/stats/versions/3.6.2/topics/hclust"><code>hclust()</code></a>
functions:</p>
<div class="sourceCode" id="cb239"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb239-1"><a href="exploratory-data-analysis.html#cb239-1" aria-hidden="true" tabindex="-1"></a><span class="co"># compute all pairwise distances using euclidean distance as the distance metric</span></span>
<span id="cb239-2"><a href="exploratory-data-analysis.html#cb239-2" aria-hidden="true" tabindex="-1"></a>euc_dist <span class="ot">&lt;-</span> <span class="fu">dist</span>(dplyr<span class="sc">::</span><span class="fu">select</span>(well_clustered_data,<span class="sc">-</span>ID))</span>
<span id="cb239-3"><a href="exploratory-data-analysis.html#cb239-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb239-4"><a href="exploratory-data-analysis.html#cb239-4" aria-hidden="true" tabindex="-1"></a><span class="co"># produce a clustering of the data using the hclust for hierarchical clustering</span></span>
<span id="cb239-5"><a href="exploratory-data-analysis.html#cb239-5" aria-hidden="true" tabindex="-1"></a>hc <span class="ot">&lt;-</span> <span class="fu">hclust</span>(euc_dist, <span class="at">method=</span><span class="st">&quot;ave&quot;</span>)</span>
<span id="cb239-6"><a href="exploratory-data-analysis.html#cb239-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb239-7"><a href="exploratory-data-analysis.html#cb239-7" aria-hidden="true" tabindex="-1"></a><span class="co"># add ID as labels to the clustering object</span></span>
<span id="cb239-8"><a href="exploratory-data-analysis.html#cb239-8" aria-hidden="true" tabindex="-1"></a>hc<span class="sc">$</span>labels <span class="ot">&lt;-</span> well_clustered_data<span class="sc">$</span>ID</span>
<span id="cb239-9"><a href="exploratory-data-analysis.html#cb239-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb239-10"><a href="exploratory-data-analysis.html#cb239-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb239-11"><a href="exploratory-data-analysis.html#cb239-11" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(hc)</span></code></pre></div>
<pre><code>## List of 7
##  $ merge      : int [1:59, 1:2] -50 -33 -48 -46 -44 -21 -11 -4 -36 -3 ...
##  $ height     : num [1:59] 0.00429 0.08695 0.23536 0.24789 0.25588 ...
##  $ order      : int [1:60] 8 9 7 6 19 18 11 16 4 14 ...
##  $ labels     : chr [1:60] &quot;A1&quot; &quot;A2&quot; &quot;A3&quot; &quot;A4&quot; ...
##  $ method     : chr &quot;average&quot;
##  $ call       : language hclust(d = euc_dist, method = &quot;ave&quot;)
##  $ dist.method: chr &quot;euclidean&quot;
##  - attr(*, &quot;class&quot;)= chr &quot;hclust&quot;</code></pre>
<p>The <code>hclust()</code> return object describes the clustering as a tree that can be
visualized using a <a href="other-kind-of-plots.html#dendrograms">dendrogram</a>:</p>
<div class="sourceCode" id="cb241"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb241-1"><a href="exploratory-data-analysis.html#cb241-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggdendro)</span>
<span id="cb241-2"><a href="exploratory-data-analysis.html#cb241-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ggdendrogram</span>(hc)</span></code></pre></div>
<p><img src="main_files/hclust%20dendro-1.png" width="672" />
Our data do indeed cluster well, where all samples from the same group cluster
together perfectly. See the <a href="other-kind-of-plots.html#dendrograms">dendrogram</a> section in the <a href="#data-viz">data
vizualization chapter</a> for more detail on how to plot clustering
results.</p>
<p>It is sometimes desirable to use split hierarchical clustering into groups based
on their pattern. In the above clustering, three discrete clusters corresponding
sample groups are clearly visible. If we wished to separate these three groups,
we can use the
<a href="https://www.rdocumentation.org/packages/stats/versions/3.6.2/topics/cutree"><code>cutree</code></a>
to divide the tree into three groups using <code>k=3</code>:</p>
<div class="sourceCode" id="cb242"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb242-1"><a href="exploratory-data-analysis.html#cb242-1" aria-hidden="true" tabindex="-1"></a>labels <span class="ot">&lt;-</span> <span class="fu">cutree</span>(hc,<span class="at">k=</span><span class="dv">3</span>)</span>
<span id="cb242-2"><a href="exploratory-data-analysis.html#cb242-2" aria-hidden="true" tabindex="-1"></a>labels</span></code></pre></div>
<pre><code>##  A1  A2  A3  A4  A5  A6  A7  A8  A9 A10 A11 A12 A13 A14 A15 A16 A17 A18 A19 A20 
##   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1 
##  B1  B2  B3  B4  B5  B6  B7  B8  B9 B10 B11 B12 B13 B14 B15 B16 B17 B18 B19 B20 
##   2   2   2   2   2   2   2   2   2   2   2   2   2   2   2   2   2   2   2   2 
##  C1  C2  C3  C4  C5  C6  C7  C8  C9 C10 C11 C12 C13 C14 C15 C16 C17 C18 C19 C20 
##   3   3   3   3   3   3   3   3   3   3   3   3   3   3   3   3   3   3   3   3</code></pre>
<p>We can then use these samples and labels to color our original plot as desired:</p>
<div class="sourceCode" id="cb244"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb244-1"><a href="exploratory-data-analysis.html#cb244-1" aria-hidden="true" tabindex="-1"></a><span class="co"># we turn our labels into a tibble so we can join them with the original tibble</span></span>
<span id="cb244-2"><a href="exploratory-data-analysis.html#cb244-2" aria-hidden="true" tabindex="-1"></a>well_clustered_data <span class="sc">%&gt;%</span></span>
<span id="cb244-3"><a href="exploratory-data-analysis.html#cb244-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">left_join</span>(</span>
<span id="cb244-4"><a href="exploratory-data-analysis.html#cb244-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">tibble</span>(</span>
<span id="cb244-5"><a href="exploratory-data-analysis.html#cb244-5" aria-hidden="true" tabindex="-1"></a>      <span class="at">ID=</span><span class="fu">names</span>(labels),</span>
<span id="cb244-6"><a href="exploratory-data-analysis.html#cb244-6" aria-hidden="true" tabindex="-1"></a>      <span class="at">label=</span><span class="fu">as.factor</span>(labels)</span>
<span id="cb244-7"><a href="exploratory-data-analysis.html#cb244-7" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb244-8"><a href="exploratory-data-analysis.html#cb244-8" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">%&gt;%</span></span>
<span id="cb244-9"><a href="exploratory-data-analysis.html#cb244-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x=</span>f1,<span class="at">y=</span>f2,<span class="at">color=</span>label)) <span class="sc">+</span></span>
<span id="cb244-10"><a href="exploratory-data-analysis.html#cb244-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb244-11"><a href="exploratory-data-analysis.html#cb244-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title=</span><span class="st">&quot;Clustered&quot;</span>)</span></code></pre></div>
<p><img src="main_files/hclust%20cutree%20color-1.png" width="672" /></p>
<p>We were able to recover the correct clustering because this dataset was easy to
cluster by construction. Real data are seldom so well behaved.</p>
</div>
</div>
<div id="k-means-." class="section level3" number="6.4.3">
<h3><span class="header-section-number">6.4.3</span> k-means .</h3>
</div>
<div id="others" class="section level3" number="6.4.4">
<h3><span class="header-section-number">6.4.4</span> Others</h3>

</div>
</div>
<p style="text-align: center;">
<a href="statistical-tests.html"><button class="btn btn-default">Previous</button></a>
<a href="network-analysis-..html"><button class="btn btn-default">Next</button></a>
</p>
</div>
</div>


</div>

<script>

// add bootstrap table styles to pandoc tables
$(document).ready(function () {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
});

</script>

</body>
</html>
