<!DOCTYPE html>
<html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Final Project | BF591 - R for Biological Sciences</title>
<meta name="generator" content="bookdown 0.24 with bs4_book()">
<meta property="og:title" content="Final Project | BF591 - R for Biological Sciences">
<meta property="og:type" content="book">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="Final Project | BF591 - R for Biological Sciences">
<!-- JS --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://kit.fontawesome.com/6ecbd6c532.js" crossorigin="anonymous"></script><script src="libs/header-attrs-2.13/header-attrs.js"></script><script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="libs/bootstrap-4.6.0/bootstrap.min.css" rel="stylesheet">
<script src="libs/bootstrap-4.6.0/bootstrap.bundle.min.js"></script><script src="libs/bs3compat-0.3.1/transition.js"></script><script src="libs/bs3compat-0.3.1/tabs.js"></script><script src="libs/bs3compat-0.3.1/bs3compat.js"></script><link href="libs/bs4_book-1.0.0/bs4_book.css" rel="stylesheet">
<script src="libs/bs4_book-1.0.0/bs4_book.js"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- CSS --><link rel="stylesheet" href="style.css">
<meta name="description" content="For the final project you will develop an R Shiny application that features multiple bioinformatics processes implemented in R. Your application must contain three mandatory components followed by...">
<meta property="og:description" content="For the final project you will develop an R Shiny application that features multiple bioinformatics processes implemented in R. Your application must contain three mandatory components followed by...">
<meta name="twitter:description" content="For the final project you will develop an R Shiny application that features multiple bioinformatics processes implemented in R. Your application must contain three mandatory components followed by...">
</head>
<body data-spy="scroll" data-target="#toc">

<div class="container-fluid">
<div class="row">
  <header class="col-sm-12 col-lg-3 sidebar sidebar-book"><a class="sr-only sr-only-focusable" href="#content">Skip to main content</a>

    <div class="d-flex align-items-start justify-content-between">
      <h1>
        <a href="index.html" title="">BF591 - R for Biological Sciences</a>
      </h1>
      <button class="btn btn-outline-primary d-lg-none ml-2 mt-1" type="button" data-toggle="collapse" data-target="#main-nav" aria-expanded="true" aria-controls="main-nav"><i class="fas fa-bars"></i><span class="sr-only">Show table of contents</span></button>
    </div>

    <div id="main-nav" class="collapse-lg">
      <form role="search">
        <input id="search" class="form-control" type="search" placeholder="Search" aria-label="Search">
</form>

      <nav aria-label="Table of contents"><h2>Table of contents</h2>
        <ul class="book-toc list-unstyled">
<li><a class="" href="index.html">R for Biological Sciences</a></li>
<li><a class="" href="introduction.html"><span class="header-section-number">1</span> Introduction</a></li>
<li><a class="" href="data-bio.html"><span class="header-section-number">2</span> Data in Biology</a></li>
<li><a class="" href="preliminaries.html"><span class="header-section-number">3</span> Preliminaries</a></li>
<li><a class="" href="prog-basics.html"><span class="header-section-number">4</span> R Programming</a></li>
<li><a class="" href="data-wrangling.html"><span class="header-section-number">5</span> Data Wrangling</a></li>
<li><a class="" href="data-science.html"><span class="header-section-number">6</span> Data Science</a></li>
<li><a class="" href="data-visualization.html"><span class="header-section-number">7</span> Data Visualization</a></li>
<li><a class="" href="biology-bioinformatics.html"><span class="header-section-number">8</span> Biology &amp; Bioinformatics</a></li>
<li><a class="" href="engineering.html"><span class="header-section-number">9</span> EngineeRing</a></li>
<li><a class="" href="rshiny.html"><span class="header-section-number">10</span> RShiny</a></li>
<li><a class="" href="communicating-with-r.html"><span class="header-section-number">11</span> Communicating with R</a></li>
<li><a class="" href="contribution-guide.html"><span class="header-section-number">12</span> Contribution Guide</a></li>
<li class="book-part">Assignments</li>
<li><a class="" href="starting-an-assignment.html">Starting an Assignment</a></li>
<li><a class="" href="assignment-1.html">Assignment 1</a></li>
<li><a class="" href="assignment-2.html">Assignment 2</a></li>
<li><a class="" href="assignment-3.html">Assignment 3</a></li>
<li><a class="" href="assignment-4.html">Assignment 4</a></li>
<li><a class="" href="assignment-5.html">Assignment 5</a></li>
<li><a class="" href="assignment-6.html">Assignment 6</a></li>
<li><a class="" href="assignment-7.html">Assignment 7</a></li>
<li><a class="active" href="final-project.html">Final Project</a></li>
<li class="book-part">Appendix</li>
<li><a class="" href="class-outlines.html"><span class="header-section-number">A</span> Class Outlines</a></li>
</ul>

        <div class="book-extra">
          
        </div>
      </nav>
</div>
  </header><main class="col-sm-12 col-md-9 col-lg-7" id="content"><div id="final-project" class="section level1 unnumbered">
<h1>Final Project<a class="anchor" aria-label="anchor" href="#final-project"><i class="fas fa-link"></i></a>
</h1>
<p>For the final project you will develop an R Shiny application that features
multiple bioinformatics processes implemented in R. Your application must
contain three mandatory components followed by a fourth component of your
choice, as described below. You may organize your application however you wish,
but it must exist as one singular application. Four separate applications will
not be acceptable. (Recommendation: put the four components in four different
tabs)</p>
<p>Part of this project is to create the data necessary to be loaded into this
application. We will provide sets of starting information, but you will need to
use the R code you’ve written for assignments one through seven to process the
raw data into an appropriate format to be loaded into your Shiny application.</p>
<p>Collaboration is encouraged but sharing code is forbidden. Your final submission
will include a GitHub repository containing the app and data as well as a five
minute pre-recorded screen grab video presentation offering a demonstration of
your application. We will view this presentation and grade your project based on
how well it fits the following criteria and the contents and quality of your
coding.</p>
<div id="general-guidelines" class="section level2" number="12.2">
<h2>
<span class="header-section-number">12.2</span> General Guidelines<a class="anchor" aria-label="anchor" href="#general-guidelines"><i class="fas fa-link"></i></a>
</h2>
<ul>
<li>Simple file input validation (check if it’s CSV or TSV that is well formatted,
error message otherwise)</li>
<li>Write comments in your code elaborating on functions, Shiny commands, or code
that is not immediately obvious to someone who isn’t familiar with your code</li>
<li>Your App should be user friendly. Buttons need to have labels on them. Text
descriptions are put in the UI to describe what each section and input is
doing.</li>
<li>(Optional) Data sharing across different tabs. E.g. After analysis is done in
the differential expression tab, proceed to the GSEA tab, which will
automatically load the results from the differential expression tab.</li>
</ul>
</div>
<div id="video-presentation-guidelines" class="section level2" number="12.3">
<h2>
<span class="header-section-number">12.3</span> Video Presentation Guidelines<a class="anchor" aria-label="anchor" href="#video-presentation-guidelines"><i class="fas fa-link"></i></a>
</h2>
<ul>
<li>A screen recorded presentation detailing the ins and outs of the Shiny application (Zoom can be used to record)</li>
<li>Narrate all steps of your presentation as you go</li>
<li>Presentations should be no more than 5 minutes and should include:
<ul>
<li>A brief introduction, tour, and demonstration of each tab in your application</li>
<li>Discuss some of the challenges and solutions you used while programming</li>
<li>A very brief explanation of the input data and how it was generated. For
example: “The data was comparing control and lung cancer samples from Paper et
al. I filtered the counts matrix to only contain genes where no more than half
of the samples within each condition had 0 counts. I then performed DE on the
counts matrix comparing A vs. B using DESeq2 with default parameters and
considered genes to be significant with a FDR &lt; .15”</li>
</ul>
</li>
<li>Quick walkthrough of Shiny functionalities checklist</li>
</ul>
<p>We recommend writing a script first to use during your presentation. The average
number of spoken words per minute in English is ~140, so you might use that as a
guide for length.</p>
</div>
<div id="required-components" class="section level2" number="12.4">
<h2>
<span class="header-section-number">12.4</span> Required Components<a class="anchor" aria-label="anchor" href="#required-components"><i class="fas fa-link"></i></a>
</h2>
<div id="sample-information-exploration" class="section level3" number="12.4.1">
<h3>
<span class="header-section-number">12.4.1</span> Sample Information Exploration<a class="anchor" aria-label="anchor" href="#sample-information-exploration"><i class="fas fa-link"></i></a>
</h3>
<p><strong>Inputs</strong>:</p>
<ul>
<li>Sample information matrix in CSV format</li>
</ul>
<p><strong>Shiny Functionality</strong>:</p>
<ul>
<li>
<p>Tab with a summary of the table that includes a summary of the type and values in each column, e.g.:
Number of rows: X
Number of columns: Y</p>
<table class="table table-sm">
<thead><tr class="header">
<th>Column Name</th>
<th>Type</th>
<th>Mean (sd) or Distinct Values</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>Age</td>
<td>double</td>
<td>61 (+/- 15)</td>
</tr>
<tr class="even">
<td>Condition</td>
<td>factor</td>
<td>AD, Control</td>
</tr>
<tr class="odd">
<td>Tau</td>
<td>double</td>
<td>1401 (+/- 310)</td>
</tr>
<tr class="even">
<td>etc</td>
<td></td>
<td></td>
</tr>
</tbody>
</table>
</li>
<li><p>Tab with a data table displaying the sample information, with sortable columns</p></li>
<li>
<p>Tab with histograms, density plots, or violin plots of continuous variables.</p>
<ul>
<li>If you want to make it fancy, allow the user to choose which column to plot
and another column to group by!</li>
</ul>
</li>
</ul>
</div>
<div id="counts-matrix-exploration" class="section level3" number="12.4.2">
<h3>
<span class="header-section-number">12.4.2</span> Counts Matrix Exploration<a class="anchor" aria-label="anchor" href="#counts-matrix-exploration"><i class="fas fa-link"></i></a>
</h3>
<p><strong>Inputs</strong>:</p>
<ul>
<li>Normalized counts matrix, by some method or other in CSV format</li>
</ul>
</div>
<div id="differential-expression" class="section level3" number="12.4.3">
<h3>
<span class="header-section-number">12.4.3</span> Differential Expression<a class="anchor" aria-label="anchor" href="#differential-expression"><i class="fas fa-link"></i></a>
</h3>
<p><strong>Inputs</strong>:</p>
<p>An RNA seq data set of soybeans with varying iron contamination levels from
bigPint library, reduced to 10% size.</p>
<p><strong>Shiny Functionality</strong>:</p>
<ul>
<li>Perform a differential expression analysis on RNA seq data using deseq2,
limma, edgeR from the provided counts file. You may use the package of your
choice to perform this analysis.</li>
<li>You must include a filtering step as a part of your analysis. All differential
expression libraries have filtering capabilities.</li>
<li>Create a tab in your shiny application that can load the counts file, analyze
it, and create a relevant plot or table to illustrate the differential
expression results. You may use a volcano plot or choose other relevant ways to
plot differentially expressed genes (such as the bigPint library).</li>
<li>In this tab, design a user interface to control the level of filtering that your
analysis performs on the count data.</li>
<li>Plots and or tables should be used to illustrate how this filtering can impact
the results of a dataset
<ul>
<li>Recommendation: Use a slide bar input to specify a FDR p-value cutoff, and
another slide bar input to specify another filter criteria, for example,
absolute log2 FC. After clicking a button, show the table of the top X
genes (ranked by p-value), and a volcano plot containing all the genes
with the significant genes (identified by the filter criteria selected)
colored differently.</li>
</ul>
</li>
</ul>
</div>
</div>
<div id="choose-your-own-adventure" class="section level2" number="12.5">
<h2>
<span class="header-section-number">12.5</span> Choose-your-own Adventure<a class="anchor" aria-label="anchor" href="#choose-your-own-adventure"><i class="fas fa-link"></i></a>
</h2>
<div id="gene-set-enrichment-analysis-1" class="section level3" number="12.5.1">
<h3>
<span class="header-section-number">12.5.1</span> Gene Set Enrichment Analysis<a class="anchor" aria-label="anchor" href="#gene-set-enrichment-analysis-1"><i class="fas fa-link"></i></a>
</h3>
<p><strong>Input</strong>:</p>
<p>Create a table of fgsea results from the differential expression data.</p>
<p>Choose an appropriate ranking metric (S2N, LogFoldChange, signed log(pvalue), etc.
Run fgsea with appropriate parameters (minsize and maxsize for gene sets, they already did it this way in assignment #5) against gene set of their choice
Unnest leading edge column (leading edge genes stored as list), save each in column delimited by something other than comma or tab.
Save the results to a CSV/TSV</p>
<p><strong>Shiny Functionalities</strong>:</p>
<ul>
<li>Tab 1
<ul>
<li>Sidebar
<ul>
<li>Upload from file (FileInput) - simple validation of CSV/TSV</li>
<li>Slider to adjust number of top pathways to plot (Reactive)</li>
</ul>
</li>
<li>Main Panel
<ul>
<li>Barplot of fgsea NES for top pathways selected by slider</li>
<li>Click on barplot for pathway to display table entry (On Click)</li>
</ul>
</li>
</ul>
</li>
<li>Tab 2
<ul>
<li>Sidebar
<ul>
<li>Slider to filter table by padj (Reactive)</li>
<li>Download button to export current filtered and displayed table results</li>
<li>Radio buttons to select all, positive or negative NES pathways (Reactive)</li>
</ul>
</li>
<li>Main panel
<ul>
<li>Sortable DT table</li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
<div id="network-analysis" class="section level3" number="12.5.2">
<h3>
<span class="header-section-number">12.5.2</span> Network Analysis<a class="anchor" aria-label="anchor" href="#network-analysis"><i class="fas fa-link"></i></a>
</h3>
<p><strong>Input</strong>:</p>
<p>Create a table (not in your Shiny app. do it in a general R session) that you
can draw a network from</p>
<p>Example: gene-gene correlation matrix (draw an edge between two vertices when
their correlation value reaches a certain threshold), top significantly
differentially expressed genes for several different phenotypes, or any other
matrices you like.</p>
<p>The number of vertices should be at least 10. Recommendation: around 15-30
vertices with a total number of edges around 2 to 5 times of the number of
vertices may be a good one to work with.</p>
<p><strong>Shiny Functionalities</strong>:</p>
<ul>
<li>Upload the table containing the information to draw a network.
<ul>
<li>Hint: there’s no requirement of the format of the table, but one of the
easiest way is to make it to have two columns named “from” and “to” which
are vertices to be used, and a column “weight” to indicate edge weight</li>
</ul>
</li>
<li>After clicking a button, draw the network. Basic igraph plot is okay.
<ul>
<li>Vertices should be colored NOT by the default color in igraph</li>
<li>Labels on vertices are optional. Make your own choice depending on the
density of your network.</li>
<li>(Optional) You can also explore other interactive plotting packages</li>
</ul>
</li>
<li>Choose at least one in this question. After clicking a button, show a table
output of the top 5 vertices with the highest… (one of the following) and
their value.
<ul>
<li>Degree</li>
<li>Closeness centrality</li>
<li>Betweenness centrality</li>
</ul>
</li>
<li>The user can select two vertices, click a button, and show the shortest path
between them. What required in this question are:
<ul>
<li>Two drop down menus (selectizeInput) to select two vertices</li>
<li>(Optional) It would be great if it prevents the user from selecting the same
vertices. You can either use ​updateSelectizeInput, or, implement a checking
step to make sure the two vertices are not the same.</li>
<li>A button</li>
<li>Show the shortest path as text output</li>
</ul>
</li>
</ul>
</div>
<div id="rshiny-test-suite" class="section level3" number="12.5.3">
<h3>
<span class="header-section-number">12.5.3</span> RShiny Test Suite<a class="anchor" aria-label="anchor" href="#rshiny-test-suite"><i class="fas fa-link"></i></a>
</h3>
<p>Instead of creating a fourth tab as a part of this project, you will instead
create a test application file for use with the “testthat” or “shinytest”
packages.</p>
<p>testthat can be used to create a headless (no visual elements) shiny application
where you can set inputs and use testthat functions like expect_equal() to
compare outputs. shinytest involves manually recording the state of the
application and comparing subsequent runs to this state.</p>
<p>Create a test file that uses either or both packages to test various parts of
your application. Each tab must have at least 5 distinct tests that measure
inputs and outputs of your application (so at least 15 total).</p>
<p>Tests should be relevant and should test parts of the application relevant to
the app’s running. For instance, a test confirming a counts file was filtered
correctly would be acceptable. Tests can measure the output of functions inside
your application or the output$
of the app itself.</p>
<p>Spend a portion of your presentation video running the test to show what is
working (or what doesn’t) and discuss briefly your approach to writing and
running the tests. Your test file will be reviewed with your project submission.</p>
</div>
<div id="differential-expression-exploratory-analysis" class="section level3" number="12.5.4">
<h3>
<span class="header-section-number">12.5.4</span> Differential Expression Exploratory Analysis<a class="anchor" aria-label="anchor" href="#differential-expression-exploratory-analysis"><i class="fas fa-link"></i></a>
</h3>
<p><strong>Input</strong>:
Differential Expression results
[Optional] Allow the program to access your previous differential expression results from the earlier section
Allow user to choose which kind of exploratory analysis to perform. Must have at
least 3 exploratory analysis options (PCA, heatmapping, clustering, sample
scatterplot, sample difference… Or any other appropriate plot. If you choose
analyses to perform outside of the aforementioned options, please explain what
they are and what they do on your app)</p>
<p>Appropriate interface which allows the user to set plot options. Suggestions
include filtering cutoffs, color options, plot title…</p>
<p><strong>Shiny Functionalities</strong>:</p>
<ul>
<li>Import data</li>
<li>Analysis selection with an appropriate method of selection</li>
<li>Once the analysis option has been selected, allow the user to edit the
analysis-relevant parameters in a way that is user friendly. Points will be
deducted if you try to emulate any of these in spirit or in exact
<ul>
<li>[Optional] If you allowed your user to select a subset of the data, display
the table of data used [in a different tab or otherwise]</li>
</ul>
</li>
<li>When parameters have been set, allow the user to click a button to enable
display of the desired plot</li>
</ul>
</div>
<div id="visualization-of-individual-gene-expressions" class="section level3" number="12.5.5">
<h3>
<span class="header-section-number">12.5.5</span> Visualization of Individual Gene Expression(s)<a class="anchor" aria-label="anchor" href="#visualization-of-individual-gene-expressions"><i class="fas fa-link"></i></a>
</h3>
<p><strong>Input</strong>:</p>
<p>Upload data of gene expression
Allow user to filter genes by specific criteria to find them easier (search/filter by name, p-value, id…
Allow user to select samples if necessary
[optional] Allow user to adjust format so they can display multiple genes across
a sample, specific gene expression across multiple samples, display multiple
genes in multiple samples…</p>
<p>Choosing to stick to one format instead is okay.</p>
<p><strong>Shiny Functionalities</strong>:
* Data upload and processing where necessary
* Upon completion of upload and processing, display a table of gene, sample
names with relevant information.
* Allow the user to filter, search, and select gene(s), sample(s) in a way that
makes sense to produce your desired output.
* Allow the user to edit display parameters for visual elements like colors.</p>
<p>Example usage from a journal article: (taken from Koch et al (2018) doi:
10.1165/rcmb.2017-0430TR )</p>
</div>
<div id="custom-option" class="section level3" number="12.5.6">
<h3>
<span class="header-section-number">12.5.6</span> Custom Option<a class="anchor" aria-label="anchor" href="#custom-option"><i class="fas fa-link"></i></a>
</h3>
<p>We have explored a number of different bioinformatics topics in this class and
BF528, but there are many more. If you would like to create your own fourth
component, draft a small proposal of what you would like your Shiny app to do
and what data it will use and send that to Adam or any of the TAs.</p>
<p>General requirements:
* Have 3 to 5 input elements (e.g. upload handler, radio selection input, text
input, drop down menu, slidebar input, color selection…)
* Have 2 to 3 output elements (e.g. plots, table, text…)
* Basic preprocessing of the data before input to Shiny app is required. You
don’t need to do a comprehensive analysis, but it should not be “download
something then use it directly.” You may refer to other sections above to see
the amount of pre-processing required.
* You may also adopt one of the pre-defined optional tab above (Gene Set
Enrichment Analysis, Network Analysis, RShiny Test Suite) with some slight
modifications (e.g. a different types of plot)</p>
<p>Input Data Sets
BF528 Project 2 fileset
Vanessa: I can provide an example matrix to be used for the network tab. My idea is to provide a pseudo gene-gene correlation matrix
bigPint soybean RNAseq?</p>

</div>
</div>
</div>



<div class="chapter-nav">
<div class="prev"><a href="assignment-7.html">Assignment 7</a></div>
<div class="next"><a href="class-outlines.html"><span class="header-section-number">A</span> Class Outlines</a></div>
</div></main><div class="col-md-3 col-lg-2 d-none d-md-block sidebar sidebar-chapter">
    <nav id="toc" data-toggle="toc" aria-label="On this page"><h2>On this page</h2>
      <ul class="nav navbar-nav">
<li><a class="nav-link" href="#final-project">Final Project</a></li>
<li><a class="nav-link" href="#general-guidelines"><span class="header-section-number">12.2</span> General Guidelines</a></li>
<li><a class="nav-link" href="#video-presentation-guidelines"><span class="header-section-number">12.3</span> Video Presentation Guidelines</a></li>
<li>
<a class="nav-link" href="#required-components"><span class="header-section-number">12.4</span> Required Components</a><ul class="nav navbar-nav">
<li><a class="nav-link" href="#sample-information-exploration"><span class="header-section-number">12.4.1</span> Sample Information Exploration</a></li>
<li><a class="nav-link" href="#counts-matrix-exploration"><span class="header-section-number">12.4.2</span> Counts Matrix Exploration</a></li>
<li><a class="nav-link" href="#differential-expression"><span class="header-section-number">12.4.3</span> Differential Expression</a></li>
</ul>
</li>
<li>
<a class="nav-link" href="#choose-your-own-adventure"><span class="header-section-number">12.5</span> Choose-your-own Adventure</a><ul class="nav navbar-nav">
<li><a class="nav-link" href="#gene-set-enrichment-analysis-1"><span class="header-section-number">12.5.1</span> Gene Set Enrichment Analysis</a></li>
<li><a class="nav-link" href="#network-analysis"><span class="header-section-number">12.5.2</span> Network Analysis</a></li>
<li><a class="nav-link" href="#rshiny-test-suite"><span class="header-section-number">12.5.3</span> RShiny Test Suite</a></li>
<li><a class="nav-link" href="#differential-expression-exploratory-analysis"><span class="header-section-number">12.5.4</span> Differential Expression Exploratory Analysis</a></li>
<li><a class="nav-link" href="#visualization-of-individual-gene-expressions"><span class="header-section-number">12.5.5</span> Visualization of Individual Gene Expression(s)</a></li>
<li><a class="nav-link" href="#custom-option"><span class="header-section-number">12.5.6</span> Custom Option</a></li>
</ul>
</li>
</ul>

      <div class="book-extra">
        <ul class="list-unstyled">
          
        </ul>
</div>
    </nav>
</div>
</div>
  

  

</div>
 <!-- .container -->

<footer class="bg-primary text-light mt-5"><div class="container"><div class="row">

  <div class="col-12 col-md-6 mt-3">
    <p>"<strong>BF591 - R for Biological Sciences</strong>" was written by . </p>
  </div>

  <div class="col-12 col-md-6 mt-3">
    <p>This book was built by the <a class="text-light" href="https://bookdown.org">bookdown</a> R package.</p>
  </div>

</div></div>
</footer><!-- dynamically load mathjax for compatibility with self-contained --><script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script><script type="text/x-mathjax-config">const popovers = document.querySelectorAll('a.footnote-ref[data-toggle="popover"]');
for (let popover of popovers) {
  const div = document.createElement('div');
  div.setAttribute('style', 'position: absolute; top: 0, left:0; width:0, height:0, overflow: hidden; visibility: hidden;');
  div.innerHTML = popover.getAttribute('data-content');

  var has_math = div.querySelector("span.math");
  if (has_math) {
    document.body.appendChild(div);
    MathJax.Hub.Queue(["Typeset", MathJax.Hub, div]);
    MathJax.Hub.Queue(function() {
      popover.setAttribute('data-content', div.innerHTML);
      document.body.removeChild(div);
    })
  }
}
</script>
</body>
</html>
