<!DOCTYPE html>
<html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Final Project | BF591 - R for Biological Sciences</title>
<meta name="generator" content="bookdown 0.40 with bs4_book()">
<meta property="og:title" content="Final Project | BF591 - R for Biological Sciences">
<meta property="og:type" content="book">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="Final Project | BF591 - R for Biological Sciences">
<!-- JS --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://kit.fontawesome.com/6ecbd6c532.js" crossorigin="anonymous"></script><script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="libs/bootstrap-4.6.0/bootstrap.min.css" rel="stylesheet">
<script src="libs/bootstrap-4.6.0/bootstrap.bundle.min.js"></script><script src="libs/bs3compat-0.8.0/transition.js"></script><script src="libs/bs3compat-0.8.0/tabs.js"></script><script src="libs/bs3compat-0.8.0/bs3compat.js"></script><link href="libs/bs4_book-1.0.0/bs4_book.css" rel="stylesheet">
<script src="libs/bs4_book-1.0.0/bs4_book.js"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- CSS --><style type="text/css">
    
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  </style>
<style type="text/css">
    /* Used with Pandoc 2.11+ new --citeproc when CSL is used */
    div.csl-bib-body { }
    div.csl-entry {
      clear: both;
          margin-bottom: 0em;
        }
    .hanging div.csl-entry {
      margin-left:2em;
      text-indent:-2em;
    }
    div.csl-left-margin {
      min-width:2em;
      float:left;
    }
    div.csl-right-inline {
      margin-left:2em;
      padding-left:1em;
    }
    div.csl-indent {
      margin-left: 2em;
    }
  </style>
<link rel="stylesheet" href="style.css">
<meta name="description" content="For the final project you will develop an R Shiny application that features multiple bioinformatics processes implemented in R. Your application must contain three mandatory components followed by...">
<meta property="og:description" content="For the final project you will develop an R Shiny application that features multiple bioinformatics processes implemented in R. Your application must contain three mandatory components followed by...">
<meta name="twitter:description" content="For the final project you will develop an R Shiny application that features multiple bioinformatics processes implemented in R. Your application must contain three mandatory components followed by...">
</head>
<body data-spy="scroll" data-target="#toc">

<div class="container-fluid">
<div class="row">
  <header class="col-sm-12 col-lg-3 sidebar sidebar-book"><a class="sr-only sr-only-focusable" href="#content">Skip to main content</a>

    <div class="d-flex align-items-start justify-content-between">
      <h1>
        <a href="index.html" title="">BF591 - R for Biological Sciences</a>
      </h1>
      <button class="btn btn-outline-primary d-lg-none ml-2 mt-1" type="button" data-toggle="collapse" data-target="#main-nav" aria-expanded="true" aria-controls="main-nav"><i class="fas fa-bars"></i><span class="sr-only">Show table of contents</span></button>
    </div>

    <div id="main-nav" class="collapse-lg">
      <form role="search">
        <input id="search" class="form-control" type="search" placeholder="Search" aria-label="Search">
</form>

      <nav aria-label="Table of contents"><h2>Table of contents</h2>
        <ul class="book-toc list-unstyled">
<li><a class="" href="index.html">R for Biological Sciences</a></li>
<li><a class="" href="detailed-class-schedule.html"><span class="header-section-number">1</span> Detailed Class Schedule</a></li>
<li><a class="" href="introduction.html"><span class="header-section-number">2</span> Introduction</a></li>
<li><a class="" href="data-bio.html"><span class="header-section-number">3</span> Data in Biology</a></li>
<li><a class="" href="preliminaries.html"><span class="header-section-number">4</span> Preliminaries</a></li>
<li><a class="" href="prog-basics.html"><span class="header-section-number">5</span> R Programming</a></li>
<li><a class="" href="data-wrangling.html"><span class="header-section-number">6</span> Data Wrangling</a></li>
<li><a class="" href="data-science.html"><span class="header-section-number">7</span> Data Science</a></li>
<li><a class="" href="data-visualization.html"><span class="header-section-number">8</span> Data Visualization</a></li>
<li><a class="" href="biology-bioinformatics.html"><span class="header-section-number">9</span> Biology &amp; Bioinformatics</a></li>
<li><a class="" href="engineering.html"><span class="header-section-number">10</span> EngineeRing</a></li>
<li><a class="" href="rshiny.html"><span class="header-section-number">11</span> RShiny</a></li>
<li><a class="" href="communicating-with-r.html"><span class="header-section-number">12</span> Communicating with R</a></li>
<li><a class="" href="contribution-guide.html"><span class="header-section-number">13</span> Contribution Guide</a></li>
<li class="book-part">Assignments</li>
<li><a class="" href="starting-an-assignment.html">Starting an Assignment</a></li>
<li><a class="" href="assignment-overview.html">Assignment Overview</a></li>
<li><a class="" href="assignment-0.html">Assignment 0</a></li>
<li><a class="" href="base-r.html">Base R</a></li>
<li><a class="" href="tidyverse-basics.html">Tidyverse Basics</a></li>
<li><a class="" href="bioinformatics-basics.html">Bioinformatics Basics</a></li>
<li><a class="" href="data-science-basics.html">Data Science Basics</a></li>
<li><a class="" href="counts-analysis.html">Counts Analysis</a></li>
<li><a class="" href="differential-expression-part-1.html">Differential Expression Part 1</a></li>
<li><a class="" href="differential-expression-part-2.html">Differential Expression Part 2</a></li>
<li><a class="" href="rshiny-basics.html">RShiny Basics</a></li>
<li><a class="active" href="final-project.html">Final Project</a></li>
<li class="book-part">Appendix</li>
<li><a class="" href="bibliography.html"><span class="header-section-number">A</span> Bibliography</a></li>
</ul>

        <div class="book-extra">
          
        </div>
      </nav>
</div>
  </header><main class="col-sm-12 col-md-9 col-lg-7" id="content"><div id="final-project" class="section level1 unnumbered">
<h1>Final Project<a class="anchor" aria-label="anchor" href="#final-project"><i class="fas fa-link"></i></a>
</h1>
<p>For the final project you will develop an R Shiny application that features
multiple bioinformatics processes implemented in R. Your application must
contain three mandatory components followed by a fourth component of your
choice, as described below. You may organize your application however you wish,
but it must be integrated into one singular application; four separate
applications will not be acceptable. However, it may be helpful to develop each
component as its own app, and then combine when all are complete. We suggest a
layout like the following:</p>
<div class="float">
<img src="final_project_structure.png" alt="Example final project app layout with nested tabs"><div class="figcaption">Example final project app layout with nested tabs</div>
</div>
<p>Note some components have subcomponents with their own input controls. You might
consider organizing those components as follows:</p>
<div class="float">
<img src="final_project_nested_tabs.png" alt="Nested tab input controls"><div class="figcaption">Nested tab input controls</div>
</div>
<p>Part of this project is to identify the data necessary to be loaded into this
application. We will provide sets of starting information, but you will need to
use the R code you’ve written for assignments one through seven to process the
raw data into an appropriate format to be loaded into your Shiny application.
Different processing steps and analyses will be required for each dataset, so
you must investigate each dataset closely to understand what is available and
what must be done yourself.</p>
<p><strong>You must submit your own individual app. This is not a group project.</strong>
However, you are encouraged to share ideas, troubleshoot and brainstorm together
but sharing code is forbidden. Your final submission will include a GitHub
repository containing the app and data as well as a five minute pre-recorded
screen grab video presentation offering a demonstration of your application. We
will view this presentation and grade your project based on how well it fits the
following criteria and the contents and quality of your coding.</p>
<div id="general-guidelines" class="section level2" number="13.2">
<h2>
<span class="header-section-number">13.2</span> General Guidelines<a class="anchor" aria-label="anchor" href="#general-guidelines"><i class="fas fa-link"></i></a>
</h2>
<ul>
<li>Simple file input validation (check if it’s CSV or TSV that is well formatted,
error message otherwise)</li>
<li>Write comments in your code elaborating on functions, Shiny commands, or code
that is not immediately obvious to someone who isn’t familiar with your code</li>
<li>Your App should be user friendly. Buttons need to have labels on them. Text
descriptions are put in the UI to describe what each section and input is
doing.</li>
<li>(Optional) Data sharing across different tabs. E.g. After analysis is done in
the differential expression tab, proceed to the GSEA tab, which will
automatically load the results from the differential expression tab.</li>
</ul>
</div>
<div id="video-presentation-guidelines" class="section level2" number="13.3">
<h2>
<span class="header-section-number">13.3</span> Video Presentation Guidelines<a class="anchor" aria-label="anchor" href="#video-presentation-guidelines"><i class="fas fa-link"></i></a>
</h2>
<ul>
<li>A screen recorded presentation detailing the ins and outs of the Shiny
application (Zoom can be used to record)</li>
<li>Narrate all steps of your presentation as you go</li>
<li>Presentations should be no more than 5 minutes and should include:
<ul>
<li>A brief introduction, tour, and demonstration of each tab in your application</li>
<li>Discuss some of the challenges and solutions you used while programming</li>
<li>A very brief explanation of the input data and how it was generated. For
example: “The data was comparing control and lung cancer samples from Paper et
al. I filtered the counts matrix to only contain genes where no more than half
of the samples within each condition had 0 counts. I then performed DE on the
counts matrix comparing A vs. B using DESeq2 with default parameters and
considered genes to be significant with a FDR &lt; .15”</li>
</ul>
</li>
<li>Quick walkthrough of Shiny functionalities checklist</li>
</ul>
<p>We recommend writing a script first to use during your presentation. The average
number of spoken words per minute in English is ~140, so you might use that as a
guide for length.</p>
</div>
<div id="input-data-sets" class="section level2" number="13.4">
<h2>
<span class="header-section-number">13.4</span> Input Data Sets<a class="anchor" aria-label="anchor" href="#input-data-sets"><i class="fas fa-link"></i></a>
</h2>
<p>We have identified several different datasets that you may use for your
projects. Each is a gene expression dataset generated with RNASeq data, and has
both sample information and a counts matrix available. You will need to identify
the data files containing this information with the references provided below
and perform any necessary pre-processing (e.g. counts normalization,
differential expression, etc.) to make the data suitable for your app. Each
dataset has a slightly different way of obtaining the data, so this should be
your first step. The app components described later on this page indicate what
the input should be using the language provided here.</p>
<p>Some hints:</p>
<ul>
<li>You will need to investigate the data available in each of the provided
sources to find sample info. For some of the datasets, it might be easier to
simply create a sample info matrix based on filenames, but others will need
more detailed (e.g. covariate) information that you will need to find and use.</li>
<li>You will need to closely examine the available counts matrices to understand
what they are and how they have been processed.</li>
<li>Some of the datasets have precomputed differential expression statistics. You
may use these analyses, but be sure you understand what comparison you are
using when describing your dataset and demonstrating your app.</li>
</ul>
<p>You may choose from the following datasets when building and demonstrating your
app:</p>
<p><strong>1. Soybean cotyledon gene expression across development</strong></p>
<p>This experiment assayed gene expression in the cotyledons, or embryonic
leaves, for soybean plants at different ages after planting.</p>
<ul>
<li><a href="https://rdrr.io/bioc/bigPint/man/se_soybean_cn_sub.html">bigPint R package</a></li>
</ul>
<p><strong>2. Post-mortem Huntington’s Disease prefrontal cortex compared with neurologically healthy controls</strong></p>
<p>This dataset profiled gene expression with RNASeq in post-mortem human
dorsolateral prefrontal cortex from patients who died from Huntington’s
Disease and age- and sex-matched neurologically healthy controls.</p>
<ul>
<li><a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE64810">mRNA-Seq Expression profiling of human post-mortem BA9 brain tissue for Huntington’s Disease and neurologically normal individuals</a></li>
</ul>
<p><strong>3. Transcriptional Reversion of Cardiac Myocyte Fate During Mammalian Cardiac Regeneration</strong></p>
<p>This dataset profiled gene expression with RNASeq in murine cardiac tissue
across different stages of development to identify genes associated with
the loss of these cell’s capacity to regenerate.</p>
<ul>
<li><a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE64403">Transcriptional Reversion of Cardiac Myocyte Fate During Mammalian Cardiac Regeneration</a></li>
</ul>
<p><strong>4. Molecular mechanisms underlying plasticity in a thermally varying environment</strong></p>
<p>This dataset profiled gene expression with RNASeq in drosophila melanogaster
at different stages of development exposed to constant or fluctuating
temperatures.</p>
<ul>
<li><a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE150450">Molecular mechanisms underlying plasticity in a thermally varying environment</a></li>
<li><a href="https://onlinelibrary.wiley.com/doi/10.1111/mec.16463">Salachan, Paul Vinu, and Jesper Givskov Sørensen. 2022. “Molecular Mechanisms Underlying Plasticity in a Thermally Varying Environment.” Molecular Ecology, April. https://doi.org/10.1111/mec.16463.</a></li>
</ul>
<p><strong>Note:</strong> This article is not open; you will need to find the article through
the library of your academic institution</p>
<p><strong>5. Your own idea</strong></p>
<p>You may identify a gene expression dataset of your own interest. The dataset
must have:</p>
<ul>
<li>A publication associated with it</li>
<li>A publicly available raw or normalized gene counts matrix with at least 6
samples and 5,000 genes</li>
<li>Available sample information that allows you to compute a differential
expression analysis</li>
</ul>
<p>Be sure to explain the background of the dataset in your demonstration video.</p>
</div>
<div id="required-components" class="section level2" number="13.5">
<h2>
<span class="header-section-number">13.5</span> Required Components<a class="anchor" aria-label="anchor" href="#required-components"><i class="fas fa-link"></i></a>
</h2>
<div id="sample-information-exploration" class="section level3" number="13.5.1">
<h3>
<span class="header-section-number">13.5.1</span> Sample Information Exploration<a class="anchor" aria-label="anchor" href="#sample-information-exploration"><i class="fas fa-link"></i></a>
</h3>
<p>The distinct values and distributions of sample information are important to
understand before conducting analysis of corresponding sample data. This
component allows the user to load and examine a sample information matrix.</p>
<p><strong>Inputs</strong>:</p>
<ul>
<li>Sample information matrix in CSV format</li>
</ul>
<p><strong>Shiny Functionalities</strong>:</p>
<ul>
<li>
<p>Tab with a summary of the table that includes a summary of the type and values in each column, e.g.:
Number of rows: X
Number of columns: Y</p>
<table class="table table-sm">
<thead><tr class="header">
<th>Column Name</th>
<th>Type</th>
<th>Mean (sd) or Distinct Values</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>Age</td>
<td>double</td>
<td>61 (+/- 15)</td>
</tr>
<tr class="even">
<td>Condition</td>
<td>factor</td>
<td>AD, Control</td>
</tr>
<tr class="odd">
<td>Tau</td>
<td>double</td>
<td>1401 (+/- 310)</td>
</tr>
<tr class="even">
<td>etc</td>
<td></td>
<td></td>
</tr>
</tbody>
</table>
</li>
<li><p>Tab with a data table displaying the sample information, with sortable columns</p></li>
<li>
<p>Tab with histograms, density plots, or violin plots of continuous variables.</p>
<ul>
<li>If you want to make it fancy, allow the user to choose which column to plot
and another column to group by!</li>
</ul>
</li>
</ul>
</div>
<div id="counts-matrix-exploration" class="section level3" number="13.5.2">
<h3>
<span class="header-section-number">13.5.2</span> Counts Matrix Exploration<a class="anchor" aria-label="anchor" href="#counts-matrix-exploration"><i class="fas fa-link"></i></a>
</h3>
<p>Exploring and visualizing counts matrices can aid in selecting gene count
filtering strategies and understanding counts data structure. This component
allows the user to choose different gene filtering thresholds and assess their
effects using diagnostic plots of the counts matrix.</p>
<p><strong>Inputs</strong>:</p>
<ul>
<li>Normalized counts matrix, by some method or other, in CSV format</li>
<li>Input controls that filter out genes based on their statistical properties:
<ul>
<li>Slider to include genes with at least X percentile of variance</li>
<li>Slider to include genes with at least X samples that are non-zero</li>
</ul>
</li>
</ul>
<p><strong>Shiny Functionalities</strong>:</p>
<ul>
<li>Tab with text or a table summarizing the effect of the filtering, including:
<ul>
<li>number of samples</li>
<li>total number of genes</li>
<li>number and % of genes passing current filter</li>
<li>number and % of genes not passing current filter</li>
</ul>
</li>
<li>Tab with diagnostic scatter plots, where genes passing filters are marked in
a darker color, and genes filtered out are lighter:
<ul>
<li>median count vs variance (consider log scale for plot)</li>
<li>median count vs number of zeros</li>
</ul>
</li>
<li>Tab with a clustered heatmap of counts remaining after filtering
<ul>
<li>consider enabling log-transforming counts for visualization</li>
<li>be sure to include a color bar in the legend</li>
</ul>
</li>
<li>Tab with a scatter plot of <a href="data-science.html#principal-component-analysis">principal component
analysis</a> projections. You may either:
<ul>
<li>allow the user to select which principal components to plot in a scatter
plot (e.g. PC1 vs PC2)</li>
<li>allow the user to plot the top <span class="math inline">\(N\)</span> principal components as a beeswarm plot</li>
<li>be sure to include the % variance explained by each component in the plot
labels</li>
</ul>
</li>
</ul>
</div>
<div id="differential-expression" class="section level3" number="13.5.3">
<h3>
<span class="header-section-number">13.5.3</span> Differential Expression<a class="anchor" aria-label="anchor" href="#differential-expression"><i class="fas fa-link"></i></a>
</h3>
<p>Differential expression identifies which genes, if any, are implicated in a
specific biological comparison. This component allows the user to load and
explore a differential expression dataset.</p>
<p><strong>Inputs</strong>:</p>
<ul>
<li>Results of a differential expression analysis in CSV format.
<ul>
<li>If results are already made available, you may use those</li>
<li>Otherwise perform a differential expression analysis using DESeq2, limma, or
edgeR from the provided counts file</li>
</ul>
</li>
</ul>
<p><strong>Shiny Functionalities</strong>:</p>
<ul>
<li>Tab with sortable table displaying differential expression results
<ul>
<li>Optional: enable gene name search functionality to filter rows of table</li>
</ul>
</li>
<li>Tab with content similar to that described in [Assignment 7]</li>
</ul>
</div>
</div>
<div id="choose-your-own-adventure" class="section level2" number="13.6">
<h2>
<span class="header-section-number">13.6</span> Choose-your-own Adventure<a class="anchor" aria-label="anchor" href="#choose-your-own-adventure"><i class="fas fa-link"></i></a>
</h2>
<p>Implement one or more of the following components as part of your app.</p>
<div id="gene-set-enrichment-analysis" class="section level3" number="13.6.1">
<h3>
<span class="header-section-number">13.6.1</span> Gene Set Enrichment Analysis<a class="anchor" aria-label="anchor" href="#gene-set-enrichment-analysis"><i class="fas fa-link"></i></a>
</h3>
<p>Use your differential gene expression results to compute <a href="final-project.html#gene-set-enrichment-analysis">gene set enrichment
analysis</a> with [fgsea]. You will need to identify
an appropriate gene set database that matches the organism studied in your
dataset.</p>
<p><strong>Input</strong>:</p>
<ul>
<li>a table of fgsea results from the differential expression data.
<ul>
<li>choose an appropriate ranking metric (log fold change, -log(pvalue), etc)
from your differential expression results</li>
<li>run fgsea with appropriate parameters against gene set database of your choice</li>
<li>save the results to a CSV/TSV</li>
</ul>
</li>
<li>file upload button</li>
</ul>
<p><strong>Shiny Functionalities</strong>:</p>
<ul>
<li>Tab 1
<ul>
<li>Sidebar
<ul>
<li>Slider to adjust number of top pathways to plot by adjusted p-value</li>
</ul>
</li>
<li>Main Panel
<ul>
<li>Barplot of fgsea NES for top pathways selected by slider</li>
<li>Optional: Click on barplot for pathway to display table entry</li>
</ul>
</li>
</ul>
</li>
<li>Tab 2
<ul>
<li>Sidebar
<ul>
<li>Slider to filter table by adjusted p-value (Reactive)</li>
<li>Radio buttons to select all, positive or negative NES pathways</li>
<li>Download button to export current filtered and displayed table results</li>
</ul>
</li>
<li>Main panel
<ul>
<li>Sortable data table displaying the results</li>
</ul>
</li>
</ul>
</li>
<li>Tab 3
<ul>
<li>Sidebar
<ul>
<li>Slider to filter table by adjusted p-value (Reactive)</li>
</ul>
</li>
<li>Main panel
<ul>
<li>Scatter plot of NES on x-axis and -log10 adjusted p-value on y-axis, with
gene sets below threshold in grey color</li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
<div id="correlation-network-analysis" class="section level3" number="13.6.2">
<h3>
<span class="header-section-number">13.6.2</span> Correlation Network Analysis<a class="anchor" aria-label="anchor" href="#correlation-network-analysis"><i class="fas fa-link"></i></a>
</h3>
<p>Create an app that computes pairwise gene expression correlation for a specific
set of input genes. The genes could be all genes in a particular pathway,
associated with disease, or perhaps the top <span class="math inline">\(N\)</span> differentially expressed genes.
Subset the gene expression matrix to include only the genes specified in the
input text area and compute all pairwise correlations. Your app should report if
any of the input genes are not found in the matrix. Also provide a slider input
that allows the user to set a minimum threshold for an edge between two genes to
be included (e.g. only draw edges for genes with a correlation &gt; 0.9). The
slider should be reactive, so the tables and plots update appropriately. Use the
[igraph] package as described in the <a href="data-science.html#network-analysis">Network Analysis</a> section to construct
your graph and to compute network statistics on each node.</p>
<p><strong>Input</strong>:</p>
<ul>
<li>Normalized counts matrix, by some method or other, in CSV format</li>
<li>A multi-line text box (hint: <a href="https://shiny.rstudio.com/reference/shiny/1.0.5/textAreaInput.html">textAreaInput</a>)
that accepts one gene name per line</li>
<li>A slider setting the minimum correlation for drawing an edge</li>
</ul>
<p><strong>Shiny Functionalities</strong>:</p>
<ul>
<li>Tab with a clustered heatmap of normalized counts corresponding to the genes
entered in the text area</li>
<li>Tab with a visualization of the correlation network
<ul>
<li>Vertices should be colored NOT by the default color in igraph</li>
<li>Vertices should be labeled with gene names</li>
<li>(Optional) The user can select two vertices, click a button, and show the shortest path
between them. What required in this question are:
<ul>
<li>Two drop down menus (selectizeInput) to select two vertices</li>
<li>Show the shortest path as text output</li>
</ul>
</li>
</ul>
</li>
<li>Tab with a table with the following metrics calculated for each gene in the input:
<ul>
<li>Degree</li>
<li>Closeness centrality</li>
<li>Betweenness centrality</li>
</ul>
</li>
</ul>
</div>
<div id="rshiny-test-suite" class="section level3" number="13.6.3">
<h3>
<span class="header-section-number">13.6.3</span> RShiny Test Suite<a class="anchor" aria-label="anchor" href="#rshiny-test-suite"><i class="fas fa-link"></i></a>
</h3>
<p>Instead of creating a fourth tab as a part of this project, you will instead
create a test application file for use with the “testthat” or “shinytest”
packages.</p>
<p>testthat can be used to create a headless (no visual elements) shiny application
where you can set inputs and use testthat functions like expect_equal() to
compare outputs. shinytest involves manually recording the state of the
application and comparing subsequent runs to this state.</p>
<p>Create a test file that uses either or both packages to test various parts of
your application. Each tab must have at least 5 distinct tests that measure
inputs and outputs of your application (so at least 15 total).</p>
<p>Tests should be relevant and should test parts of the application relevant to
the app’s running. For instance, a test confirming a counts file was filtered
correctly would be acceptable. Tests can measure the output of functions inside
your application or the <code>output$</code> of the app itself.</p>
<p>Spend a portion of your presentation video running the test to show what is
working (or what doesn’t) and discuss briefly your approach to writing and
running the tests. Your test file will be reviewed with your project submission.</p>
</div>
<div id="visualization-of-individual-gene-expressions" class="section level3" number="13.6.4">
<h3>
<span class="header-section-number">13.6.4</span> Visualization of Individual Gene Expression(s)<a class="anchor" aria-label="anchor" href="#visualization-of-individual-gene-expressions"><i class="fas fa-link"></i></a>
</h3>
<p>Visualizing individual gene counts is sometimes useful for examining or
verifying patterns identified by differential expression analysis. There are
many different ways of visualizing counts for a single gene. This app allows
counts from an arbitrary gene to be selected and visualized broken out by a
desired sample information variable.</p>
<p><strong>Input</strong>:</p>
<ul>
<li>Normalized counts matrix, by some method or other, in CSV format</li>
<li>Sample information matrix in CSV format</li>
<li>Input control that allows the user to choose one of the categorical fields
found in the sample information matrix file</li>
<li>Input control that allows the user to choose one of the genes found in the
counts matrix (hint: try implementing a <a href="https://stackoverflow.com/questions/47336114/searchbox-in-r-shiny">search box</a>)</li>
<li>Input control allowing the user to select one of <a href="#bar-chart">bar plot</a>,
<a href="data-visualization.html#boxplot">boxplot</a>, <a href="data-visualization.html#violin-plot">violin plot</a>, or <a href="data-visualization.html#beeswarm-plot">beeswarm plot</a>
</li>
<li>A button that makes the thing go</li>
</ul>
<p><strong>Shiny Functionalities</strong>:</p>
<ul>
<li>Content displaying a plot of the selected type with the normalized gene counts
for the selected gene split out by the categorical variable chosen</li>
</ul>
</div>
<div id="custom-option" class="section level3" number="13.6.5">
<h3>
<span class="header-section-number">13.6.5</span> Custom Option<a class="anchor" aria-label="anchor" href="#custom-option"><i class="fas fa-link"></i></a>
</h3>
<p>We have explored a number of different bioinformatics topics in this class and
BF528, but there are many more. If you would like to create your own fourth
component, draft a small proposal of what you would like your Shiny app to do
and what data it will use and send that to Adam or any of the TAs.</p>
<p>General requirements:
* Have 3 to 5 input elements (e.g. upload handler, radio selection input, text
input, drop down menu, slidebar input, color selection…)
* Have 2 to 3 output elements (e.g. plots, table, text…)
* Basic preprocessing of the data before input to Shiny app is required. You
don’t need to do a comprehensive analysis, but it should not be “download
something then use it directly”. You may refer to other sections above to see
the amount of pre-processing required.
* You may also adopt one of the pre-defined optional tab above (Gene Set
Enrichment Analysis, Network Analysis, RShiny Test Suite) with some slight
modifications (e.g. a different types of plot)</p>

</div>
</div>
</div>



<div class="chapter-nav">
<div class="prev"><a href="rshiny-basics.html">RShiny Basics</a></div>
<div class="next"><a href="bibliography.html"><span class="header-section-number">A</span> Bibliography</a></div>
</div></main><div class="col-md-3 col-lg-2 d-none d-md-block sidebar sidebar-chapter">
    <nav id="toc" data-toggle="toc" aria-label="On this page"><h2>On this page</h2>
      <ul class="nav navbar-nav">
<li><a class="nav-link" href="#final-project">Final Project</a></li>
<li><a class="nav-link" href="#general-guidelines"><span class="header-section-number">13.2</span> General Guidelines</a></li>
<li><a class="nav-link" href="#video-presentation-guidelines"><span class="header-section-number">13.3</span> Video Presentation Guidelines</a></li>
<li><a class="nav-link" href="#input-data-sets"><span class="header-section-number">13.4</span> Input Data Sets</a></li>
<li>
<a class="nav-link" href="#required-components"><span class="header-section-number">13.5</span> Required Components</a><ul class="nav navbar-nav">
<li><a class="nav-link" href="#sample-information-exploration"><span class="header-section-number">13.5.1</span> Sample Information Exploration</a></li>
<li><a class="nav-link" href="#counts-matrix-exploration"><span class="header-section-number">13.5.2</span> Counts Matrix Exploration</a></li>
<li><a class="nav-link" href="#differential-expression"><span class="header-section-number">13.5.3</span> Differential Expression</a></li>
</ul>
</li>
<li>
<a class="nav-link" href="#choose-your-own-adventure"><span class="header-section-number">13.6</span> Choose-your-own Adventure</a><ul class="nav navbar-nav">
<li><a class="nav-link" href="#gene-set-enrichment-analysis"><span class="header-section-number">13.6.1</span> Gene Set Enrichment Analysis</a></li>
<li><a class="nav-link" href="#correlation-network-analysis"><span class="header-section-number">13.6.2</span> Correlation Network Analysis</a></li>
<li><a class="nav-link" href="#rshiny-test-suite"><span class="header-section-number">13.6.3</span> RShiny Test Suite</a></li>
<li><a class="nav-link" href="#visualization-of-individual-gene-expressions"><span class="header-section-number">13.6.4</span> Visualization of Individual Gene Expression(s)</a></li>
<li><a class="nav-link" href="#custom-option"><span class="header-section-number">13.6.5</span> Custom Option</a></li>
</ul>
</li>
</ul>

      <div class="book-extra">
        <ul class="list-unstyled">
          
        </ul>
</div>
    </nav>
</div>
</div>
  

  

</div>
 <!-- .container -->

<footer class="bg-primary text-light mt-5"><div class="container"><div class="row">

  <div class="col-12 col-md-6 mt-3">
    <p>"<strong>BF591 - R for Biological Sciences</strong>" was written by . </p>
  </div>

  <div class="col-12 col-md-6 mt-3">
    <p>This book was built by the <a class="text-light" href="https://bookdown.org">bookdown</a> R package.</p>
  </div>

</div></div>
</footer><!-- dynamically load mathjax for compatibility with self-contained --><script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/latest.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script><script type="text/x-mathjax-config">const popovers = document.querySelectorAll('a.footnote-ref[data-toggle="popover"]');
for (let popover of popovers) {
  const div = document.createElement('div');
  div.setAttribute('style', 'position: absolute; top: 0, left:0; width:0, height:0, overflow: hidden; visibility: hidden;');
  div.innerHTML = popover.getAttribute('data-content');

  var has_math = div.querySelector("span.math");
  if (has_math) {
    document.body.appendChild(div);
    MathJax.Hub.Queue(["Typeset", MathJax.Hub, div]);
    MathJax.Hub.Queue(function() {
      popover.setAttribute('data-content', div.innerHTML);
      document.body.removeChild(div);
    })
  }
}
</script>
</body>
</html>
