<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta name="viewport" content="width=device-width, initial-scale=1" />
<meta property="og:title" content="8.7 Gene Expression | BF591 - R for Biological Sciences" />
<meta property="og:type" content="book" />







<script type="text/x-mathjax-config">
MathJax.Hub.Config({
  TeX: { equationNumbers: { autoNumber: "AMS" } }
});
</script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<meta name="description" content="8.7 Gene Expression | BF591 - R for Biological Sciences">

<title>8.7 Gene Expression | BF591 - R for Biological Sciences</title>

<script src="libs/header-attrs-2.13/header-attrs.js"></script>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="libs/bootstrap-3.3.5/css/darkly.min.css" rel="stylesheet" />
<script src="libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="libs/navigation-1.1/tabsets.js"></script>


<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>


<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />

<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
/* show arrow before summary tag as in bootstrap
TODO: remove if boostrap in updated in html_document (rmarkdown#1485) */
details > summary {
  display: list-item;
  cursor: pointer;
}
</style>
</head>

<body>

<div class="container-fluid main-container">


<div class="row">
<div class="col-sm-12">
<div id="TOC">
<ul>
<li><a href="index.html#syllabus">Syllabus</a>
<ul>
<li><a href="course-schedule.html#course-schedule">Course Schedule</a></li>
<li><a href="instructors.html#instructors">Instructors</a></li>
<li><a href="office-hours.html#office-hours">Office Hours</a></li>
<li><a href="course-values-and-policies.html#course-values-and-policies">Course Values and Policies</a></li>
</ul></li>
<li><a href="introduction.html#introduction"><span class="toc-section-number">1</span> Introduction</a>
<ul>
<li><a href="who-this-book-is-for.html#who-this-book-is-for"><span class="toc-section-number">1.1</span> Who This Book Is For</a></li>
<li><a href="a-note-about-reinventing-the-wheel.html#a-note-about-reinventing-the-wheel"><span class="toc-section-number">1.2</span> A Note About Reinventing the Wheel</a></li>
<li><a href="sources-and-references.html#sources-and-references"><span class="toc-section-number">1.3</span> Sources and References</a></li>
</ul></li>
<li><a href="data-bio.html#data-bio"><span class="toc-section-number">2</span> Data in Biology</a>
<ul>
<li><a href="a-brief-history-of-data-in-molecular-biology.html#a-brief-history-of-data-in-molecular-biology"><span class="toc-section-number">2.1</span> A Brief History of Data in Molecular Biology</a></li>
<li><a href="biology-as-a-mature-data-science.html#biology-as-a-mature-data-science"><span class="toc-section-number">2.2</span> Biology as a Mature Data Science</a></li>
</ul></li>
<li><a href="preliminaries.html#preliminaries"><span class="toc-section-number">3</span> Preliminaries</a>
<ul>
<li><a href="prelim-r.html#prelim-r"><span class="toc-section-number">3.1</span> The R Language</a></li>
<li><a href="prelim-rstudio.html#prelim-rstudio"><span class="toc-section-number">3.2</span> RStudio</a></li>
<li><a href="prog-the-r-script.html#prog-the-r-script"><span class="toc-section-number">3.3</span> The R Script</a></li>
<li><a href="prog-workflow.html#prog-workflow"><span class="toc-section-number">3.4</span> The Scripting Workflow</a></li>
<li><a href="prelim-git.html#prelim-git"><span class="toc-section-number">3.5</span> git + github</a>
<ul>
<li><a href="prelim-git.html#motivation"><span class="toc-section-number">3.5.1</span> Motivation</a></li>
<li><a href="prelim-git.html#git"><span class="toc-section-number">3.5.2</span> git</a></li>
<li><a href="prelim-git.html#git-hosting-platforms-github"><span class="toc-section-number">3.5.3</span> Git hosting platforms (GitHub)</a></li>
</ul></li>
</ul></li>
<li><a href="prog-basics.html#prog-basics"><span class="toc-section-number">4</span> R Programming</a>
<ul>
<li><a href="before-you-begin.html#before-you-begin"><span class="toc-section-number">4.1</span> Before you begin</a></li>
<li><a href="introduction-1.html#introduction-1"><span class="toc-section-number">4.2</span> Introduction</a></li>
<li><a href="prog-r-syntax.html#prog-r-syntax"><span class="toc-section-number">4.3</span> R Syntax Basics</a></li>
<li><a href="prog-types.html#prog-types"><span class="toc-section-number">4.4</span> Basic Types of Values</a>
<ul>
<li><a href="prog-types.html#factors"><span class="toc-section-number">4.4.1</span> Factors</a></li>
</ul></li>
<li><a href="prog-struct.html#prog-struct"><span class="toc-section-number">4.5</span> Data Structures</a>
<ul>
<li><a href="prog-struct.html#vectors"><span class="toc-section-number">4.5.1</span> Vectors</a></li>
<li><a href="prog-struct.html#matrices"><span class="toc-section-number">4.5.2</span> Matrices</a></li>
<li><a href="prog-struct.html#lists-and-data-frames"><span class="toc-section-number">4.5.3</span> Lists and data frames</a></li>
</ul></li>
<li><a href="logical-tests-and-comparators.html#logical-tests-and-comparators"><span class="toc-section-number">4.6</span> Logical Tests and Comparators</a></li>
<li><a href="functions.html#functions"><span class="toc-section-number">4.7</span> Functions</a>
<ul>
<li><a href="functions.html#prog-dry"><span class="toc-section-number">4.7.1</span> DRY: Don’t Repeat Yourself</a></li>
<li><a href="functions.html#writing-your-own-functions"><span class="toc-section-number">4.7.2</span> Writing your own functions</a></li>
<li><a href="functions.html#scope"><span class="toc-section-number">4.7.3</span> Scope</a></li>
</ul></li>
<li><a href="iteration.html#iteration"><span class="toc-section-number">4.8</span> Iteration</a>
<ul>
<li><a href="iteration.html#vectorized-operations"><span class="toc-section-number">4.8.1</span> Vectorized operations</a></li>
<li><a href="iteration.html#functional-programming"><span class="toc-section-number">4.8.2</span> Functional programming</a></li>
<li><a href="iteration.html#apply-and-friends"><span class="toc-section-number">4.8.3</span> <code>apply()</code> and friends</a></li>
</ul></li>
<li><a href="installing-packages.html#installing-packages"><span class="toc-section-number">4.9</span> Installing Packages</a></li>
<li><a href="saving-and-loading-r-data.html#saving-and-loading-r-data"><span class="toc-section-number">4.10</span> Saving and Loading R Data</a></li>
<li><a href="prog-debugging.html#prog-debugging"><span class="toc-section-number">4.11</span> Troubleshooting and Debugging</a></li>
<li><a href="prog-style.html#prog-style"><span class="toc-section-number">4.12</span> Coding Style and Conventions</a>
<ul>
<li><a href="prog-style.html#is-my-code-correct"><span class="toc-section-number">4.12.1</span> Is my code correct?</a></li>
<li><a href="prog-style.html#does-my-code-follow-the-dry-principle"><span class="toc-section-number">4.12.2</span> Does my code follow the <span>DRY principle</span>?</a></li>
<li><a href="prog-style.html#did-i-choose-concise-but-descriptive-variable-and-function-names"><span class="toc-section-number">4.12.3</span> Did I choose concise but descriptive variable and function names?</a></li>
<li><a href="prog-style.html#did-i-use-indentation-and-naming-conventions-consistently-throughout-my-code"><span class="toc-section-number">4.12.4</span> Did I use indentation and naming conventions consistently throughout my code?</a></li>
<li><a href="prog-style.html#did-i-write-comments-especially-when-what-the-code-does-is-not-obvious"><span class="toc-section-number">4.12.5</span> Did I write comments, especially when what the code does is not obvious?</a></li>
<li><a href="prog-style.html#how-easy-would-it-be-for-someone-else-to-understand-my-code"><span class="toc-section-number">4.12.6</span> How easy would it be for someone else to understand my code?</a></li>
<li><a href="prog-style.html#is-my-code-easy-to-maintainchange"><span class="toc-section-number">4.12.7</span> Is my code easy to maintain/change?</a></li>
<li><a href="prog-style.html#the-styler-package"><span class="toc-section-number">4.12.8</span> The <code>styler</code> package</a></li>
</ul></li>
</ul></li>
<li><a href="data-wrangling.html#data-wrangling"><span class="toc-section-number">5</span> Data Wrangling</a>
<ul>
<li><a href="the-tidyverse.html#the-tidyverse"><span class="toc-section-number">5.1</span> The Tidyverse</a></li>
<li><a href="dw-basics.html#dw-basics"><span class="toc-section-number">5.2</span> Tidyverse Basics</a></li>
<li><a href="importing-data.html#importing-data"><span class="toc-section-number">5.3</span> Importing Data</a></li>
<li><a href="data-tibble.html#data-tibble"><span class="toc-section-number">5.4</span> The tibble</a></li>
<li><a href="tidy-data.html#tidy-data"><span class="toc-section-number">5.5</span> Tidy Data</a></li>
<li><a href="pipes.html#pipes"><span class="toc-section-number">5.6</span> pipes</a></li>
<li><a href="arranging-data.html#arranging-data"><span class="toc-section-number">5.7</span> Arranging Data</a>
<ul>
<li><a href="arranging-data.html#dplyrmutate---create-new-columns-using-other-columns"><span class="toc-section-number">5.7.1</span> <code>dplyr::mutate()</code> - Create new columns using other columns</a></li>
<li><a href="arranging-data.html#stringr---working-with-character-values"><span class="toc-section-number">5.7.2</span> <code>stringr</code> - Working with character values</a></li>
<li><a href="arranging-data.html#dplyrselect---subset-columns-by-name"><span class="toc-section-number">5.7.3</span> <code>dplyr::select()</code> - Subset Columns by Name</a></li>
<li><a href="arranging-data.html#dplyrfilter---pick-rows-out-of-a-data-set"><span class="toc-section-number">5.7.4</span> <code>dplyr::filter()</code> - Pick rows out of a data set</a></li>
<li><a href="arranging-data.html#dplyrarrange---order-rows-based-on-their-values"><span class="toc-section-number">5.7.5</span> <code>dplyr::arrange()</code> - Order rows based on their values</a></li>
<li><a href="arranging-data.html#putting-it-all-together"><span class="toc-section-number">5.7.6</span> Putting it all together</a></li>
</ul></li>
<li><a href="grouping-data.html#grouping-data"><span class="toc-section-number">5.8</span> Grouping Data</a></li>
<li><a href="rearranging-data.html#rearranging-data"><span class="toc-section-number">5.9</span> Rearranging Data</a></li>
<li><a href="relational-data.html#relational-data"><span class="toc-section-number">5.10</span> Relational Data</a>
<ul>
<li><a href="relational-data.html#dealing-with-multiple-matches"><span class="toc-section-number">5.10.1</span> Dealing with multiple matches</a></li>
</ul></li>
</ul></li>
<li><a href="data-science.html#data-science"><span class="toc-section-number">6</span> Data Science</a>
<ul>
<li><a href="data-modeling.html#data-modeling"><span class="toc-section-number">6.1</span> Data Modeling</a>
<ul>
<li><a href="data-modeling.html#a-worked-modeling-example"><span class="toc-section-number">6.1.1</span> A Worked Modeling Example</a></li>
<li><a href="data-modeling.html#data-summarization"><span class="toc-section-number">6.1.2</span> Data Summarization</a></li>
<li><a href="data-modeling.html#linear-models"><span class="toc-section-number">6.1.3</span> Linear Models</a></li>
<li><a href="data-modeling.html#flavors-of-linear-models"><span class="toc-section-number">6.1.4</span> Flavors of Linear Models</a></li>
<li><a href="data-modeling.html#assessing-model-accuracy-."><span class="toc-section-number">6.1.5</span> Assessing Model Accuracy .</a></li>
</ul></li>
<li><a href="statistical-distributions.html#statistical-distributions"><span class="toc-section-number">6.2</span> Statistical Distributions</a>
<ul>
<li><a href="statistical-distributions.html#random-variables"><span class="toc-section-number">6.2.1</span> Random Variables</a></li>
<li><a href="statistical-distributions.html#statistical-distribution-basics"><span class="toc-section-number">6.2.2</span> Statistical Distribution Basics</a></li>
<li><a href="statistical-distributions.html#distributions-in-r"><span class="toc-section-number">6.2.3</span> Distributions in R</a></li>
<li><a href="statistical-distributions.html#discrete-distributions"><span class="toc-section-number">6.2.4</span> Discrete Distributions</a></li>
<li><a href="statistical-distributions.html#continuous-distributions"><span class="toc-section-number">6.2.5</span> Continuous Distributions</a></li>
<li><a href="statistical-distributions.html#empirical-distributions"><span class="toc-section-number">6.2.6</span> Empirical Distributions</a></li>
</ul></li>
<li><a href="statistical-tests.html#statistical-tests"><span class="toc-section-number">6.3</span> Statistical Tests</a>
<ul>
<li><a href="statistical-tests.html#what-is-a-statistical-test"><span class="toc-section-number">6.3.1</span> What is a statistical test</a></li>
<li><a href="statistical-tests.html#common-statistical-tests"><span class="toc-section-number">6.3.2</span> Common Statistical Tests</a></li>
<li><a href="statistical-tests.html#choosing-a-test"><span class="toc-section-number">6.3.3</span> Choosing a Test</a></li>
<li><a href="statistical-tests.html#p-values"><span class="toc-section-number">6.3.4</span> P-values</a></li>
<li><a href="statistical-tests.html#multiple-hypothesis-testing-."><span class="toc-section-number">6.3.5</span> Multiple Hypothesis Testing .</a></li>
<li><a href="statistical-tests.html#statistical-power"><span class="toc-section-number">6.3.6</span> Statistical power</a></li>
</ul></li>
<li><a href="exploratory-data-analysis.html#exploratory-data-analysis"><span class="toc-section-number">6.4</span> Exploratory Data Analysis</a>
<ul>
<li><a href="exploratory-data-analysis.html#principal-component-analysis"><span class="toc-section-number">6.4.1</span> Principal Component Analysis</a></li>
<li><a href="exploratory-data-analysis.html#cluster-analysis"><span class="toc-section-number">6.4.2</span> Cluster Analysis</a></li>
<li><a href="exploratory-data-analysis.html#k-means-."><span class="toc-section-number">6.4.3</span> k-means .</a></li>
<li><a href="exploratory-data-analysis.html#others"><span class="toc-section-number">6.4.4</span> Others</a></li>
</ul></li>
<li><a href="network-analysis-..html#network-analysis-."><span class="toc-section-number">6.5</span> Network Analysis .</a>
<ul>
<li><a href="network-analysis-..html#basic-network-analysis"><span class="toc-section-number">6.5.1</span> Basic network analysis</a></li>
<li><a href="network-analysis-..html#create-your-own"><span class="toc-section-number">6.5.2</span> Create your own</a></li>
</ul></li>
</ul></li>
<li><a href="data-visualization.html#data-visualization"><span class="toc-section-number">7</span> Data Visualization</a>
<ul>
<li><a href="responsible-plotting.html#responsible-plotting"><span class="toc-section-number">7.1</span> Responsible Plotting</a>
<ul>
<li><a href="responsible-plotting.html#visualization-principles"><span class="toc-section-number">7.1.1</span> Visualization Principles</a></li>
<li><a href="responsible-plotting.html#human-visual-perception"><span class="toc-section-number">7.1.2</span> Human Visual Perception</a></li>
<li><a href="responsible-plotting.html#visual-encodings"><span class="toc-section-number">7.1.3</span> Visual Encodings</a></li>
<li><a href="responsible-plotting.html#some-opinionated-rules-of-thumb"><span class="toc-section-number">7.1.4</span> Some Opinionated Rules of Thumb</a></li>
</ul></li>
<li><a href="dv-gg.html#dv-gg"><span class="toc-section-number">7.2</span> Grammar of Graphics</a>
<ul>
<li><a href="dv-gg.html#ggplot-mechanics"><span class="toc-section-number">7.2.1</span> <code>ggplot</code> mechanics</a></li>
</ul></li>
<li><a href="plotting-one-dimension.html#plotting-one-dimension"><span class="toc-section-number">7.3</span> Plotting One Dimension</a>
<ul>
<li><a href="plotting-one-dimension.html#dv-bar"><span class="toc-section-number">7.3.1</span> Bar chart</a></li>
<li><a href="plotting-one-dimension.html#lollipop-plots"><span class="toc-section-number">7.3.2</span> Lollipop plots</a></li>
<li><a href="plotting-one-dimension.html#stacked-area-charts"><span class="toc-section-number">7.3.3</span> Stacked Area charts</a></li>
<li><a href="plotting-one-dimension.html#parallel-coordinate-plots"><span class="toc-section-number">7.3.4</span> Parallel Coordinate plots</a></li>
</ul></li>
<li><a href="visualizing-distributions.html#visualizing-distributions"><span class="toc-section-number">7.4</span> Visualizing Distributions</a>
<ul>
<li><a href="visualizing-distributions.html#histogram"><span class="toc-section-number">7.4.1</span> Histogram</a></li>
<li><a href="visualizing-distributions.html#density"><span class="toc-section-number">7.4.2</span> Density</a></li>
<li><a href="visualizing-distributions.html#boxplot"><span class="toc-section-number">7.4.3</span> Boxplot</a></li>
<li><a href="visualizing-distributions.html#violin-plot"><span class="toc-section-number">7.4.4</span> Violin plot</a></li>
<li><a href="visualizing-distributions.html#beeswarm-plot"><span class="toc-section-number">7.4.5</span> Beeswarm plot</a></li>
<li><a href="visualizing-distributions.html#ridgeline"><span class="toc-section-number">7.4.6</span> Ridgeline</a></li>
</ul></li>
<li><a href="plotting-two-or-more-dimensions.html#plotting-two-or-more-dimensions"><span class="toc-section-number">7.5</span> Plotting Two or More Dimensions</a>
<ul>
<li><a href="plotting-two-or-more-dimensions.html#scatter-plots"><span class="toc-section-number">7.5.1</span> Scatter Plots</a></li>
<li><a href="plotting-two-or-more-dimensions.html#bubble-plots"><span class="toc-section-number">7.5.2</span> Bubble Plots</a></li>
<li><a href="plotting-two-or-more-dimensions.html#connected-scatter-plots"><span class="toc-section-number">7.5.3</span> Connected Scatter Plots</a></li>
<li><a href="plotting-two-or-more-dimensions.html#line-plots"><span class="toc-section-number">7.5.4</span> Line Plots</a></li>
<li><a href="plotting-two-or-more-dimensions.html#parallel-coordinate-plots-1"><span class="toc-section-number">7.5.5</span> Parallel Coordinate Plots</a></li>
<li><a href="plotting-two-or-more-dimensions.html#heatmaps"><span class="toc-section-number">7.5.6</span> Heatmaps</a></li>
<li><a href="plotting-two-or-more-dimensions.html#how-to-use-heatmaps-responsibly"><span class="toc-section-number">7.5.7</span> How To Use Heatmaps Responsibly</a></li>
</ul></li>
<li><a href="other-kind-of-plots.html#other-kind-of-plots"><span class="toc-section-number">7.6</span> Other Kind of Plots</a>
<ul>
<li><a href="other-kind-of-plots.html#dendrograms"><span class="toc-section-number">7.6.1</span> Dendrograms</a></li>
<li><a href="other-kind-of-plots.html#chord-diagrams-and-circos-plots"><span class="toc-section-number">7.6.2</span> Chord Diagrams and Circos Plots</a></li>
</ul></li>
<li><a href="multiple-plots.html#multiple-plots"><span class="toc-section-number">7.7</span> Multiple Plots</a>
<ul>
<li><a href="multiple-plots.html#facet-wrapping"><span class="toc-section-number">7.7.1</span> Facet wrapping</a></li>
<li><a href="multiple-plots.html#multipanel-figures"><span class="toc-section-number">7.7.2</span> Multipanel Figures</a></li>
</ul></li>
<li><a href="publication-ready-plots.html#publication-ready-plots"><span class="toc-section-number">7.8</span> Publication Ready Plots</a>
<ul>
<li><a href="publication-ready-plots.html#ggplot-themes"><span class="toc-section-number">7.8.1</span> <code>ggplot</code> Themes</a></li>
<li><a href="publication-ready-plots.html#exporting-to-svg"><span class="toc-section-number">7.8.2</span> Exporting to SVG</a></li>
</ul></li>
<li><a href="network-visualization.html#network-visualization"><span class="toc-section-number">7.9</span> Network visualization</a>
<ul>
<li><a href="network-visualization.html#interactive-network-plots"><span class="toc-section-number">7.9.1</span> Interactive network plots</a></li>
</ul></li>
</ul></li>
<li><a href="biology-bioinformatics.html#biology-bioinformatics"><span class="toc-section-number">8</span> Biology &amp; Bioinformatics</a>
<ul>
<li><a href="bio-r.html#bio-r"><span class="toc-section-number">8.1</span> R in Biology</a></li>
<li><a href="biological-data-overview.html#biological-data-overview"><span class="toc-section-number">8.2</span> Biological Data Overview</a>
<ul>
<li><a href="biological-data-overview.html#bio-data-types"><span class="toc-section-number">8.2.1</span> Types of Biological Data</a></li>
<li><a href="biological-data-overview.html#csv-files"><span class="toc-section-number">8.2.2</span> CSV Files</a></li>
<li><a href="biological-data-overview.html#biological-data-is-not-tidy"><span class="toc-section-number">8.2.3</span> Biological data is NOT Tidy!</a></li>
</ul></li>
<li><a href="bio-bioconductor.html#bio-bioconductor"><span class="toc-section-number">8.3</span> Bioconductor</a></li>
<li><a href="microarrays.html#microarrays"><span class="toc-section-number">8.4</span> Microarrays</a></li>
<li><a href="high-throughput-sequencing.html#high-throughput-sequencing"><span class="toc-section-number">8.5</span> High Throughput Sequencing</a>
<ul>
<li><a href="high-throughput-sequencing.html#raw-hts-data"><span class="toc-section-number">8.5.1</span> Raw HTS Data</a></li>
<li><a href="high-throughput-sequencing.html#preliminary-hts-data-analysis"><span class="toc-section-number">8.5.2</span> Preliminary HTS Data Analysis</a></li>
<li><a href="high-throughput-sequencing.html#count-data"><span class="toc-section-number">8.5.3</span> Count Data</a></li>
</ul></li>
<li><a href="gene-identifiers.html#gene-identifiers"><span class="toc-section-number">8.6</span> Gene Identifiers</a>
<ul>
<li><a href="gene-identifiers.html#gene-identifier-systems"><span class="toc-section-number">8.6.1</span> Gene Identifier Systems</a></li>
<li><a href="gene-identifiers.html#mapping-between-identifier-systems"><span class="toc-section-number">8.6.2</span> Mapping Between Identifier Systems</a></li>
<li><a href="gene-identifiers.html#mapping-homologs"><span class="toc-section-number">8.6.3</span> Mapping Homologs</a></li>
</ul></li>
<li><a href="gene-expression.html#gene-expression"><span class="toc-section-number">8.7</span> Gene Expression</a>
<ul>
<li><a href="gene-expression.html#gene-expression-data-in-bioconductor"><span class="toc-section-number">8.7.1</span> Gene Expression Data in Bioconductor</a></li>
<li><a href="gene-expression.html#differential-expression-analysis"><span class="toc-section-number">8.7.2</span> Differential Expression Analysis</a></li>
<li><a href="gene-expression.html#microarray-gene-expression-data"><span class="toc-section-number">8.7.3</span> Microarray Gene Expression Data</a></li>
<li><a href="gene-expression.html#differential-expression-microarrays-limma"><span class="toc-section-number">8.7.4</span> Differential Expression: Microarrays (limma)</a></li>
<li><a href="gene-expression.html#rnaseq"><span class="toc-section-number">8.7.5</span> RNASeq</a></li>
<li><a href="gene-expression.html#rnaseq-gene-expression-data"><span class="toc-section-number">8.7.6</span> RNASeq Gene Expression Data</a></li>
<li><a href="gene-expression.html#filtering-counts"><span class="toc-section-number">8.7.7</span> Filtering Counts</a></li>
<li><a href="gene-expression.html#count-distributions"><span class="toc-section-number">8.7.8</span> Count Distributions</a></li>
<li><a href="gene-expression.html#differential-expression-rnaseq"><span class="toc-section-number">8.7.9</span> Differential Expression: RNASeq</a></li>
</ul></li>
<li><a href="gene-set-enrichment-analysis.html#gene-set-enrichment-analysis"><span class="toc-section-number">8.8</span> Gene Set Enrichment Analysis</a>
<ul>
<li><a href="gene-set-enrichment-analysis.html#gene-sets"><span class="toc-section-number">8.8.1</span> Gene Sets</a></li>
<li><a href="gene-set-enrichment-analysis.html#working-with-gene-sets-in-r"><span class="toc-section-number">8.8.2</span> Working with gene sets in R</a></li>
<li><a href="gene-set-enrichment-analysis.html#over-representation-analysis"><span class="toc-section-number">8.8.3</span> Over-representation Analysis</a></li>
<li><a href="gene-set-enrichment-analysis.html#rank-based-analysis"><span class="toc-section-number">8.8.4</span> Rank-based Analysis</a></li>
<li><a href="gene-set-enrichment-analysis.html#fgsea"><span class="toc-section-number">8.8.5</span> fgsea</a></li>
</ul></li>
<li><a href="biological-networks-..html#biological-networks-."><span class="toc-section-number">8.9</span> Biological Networks .</a>
<ul>
<li><a href="biological-networks-..html#biological-pathways-."><span class="toc-section-number">8.9.1</span> Biological Pathways .</a></li>
<li><a href="biological-networks-..html#gene-regulatory-networks-."><span class="toc-section-number">8.9.2</span> Gene Regulatory Networks .</a></li>
<li><a href="biological-networks-..html#protein-protein-interaction-network-."><span class="toc-section-number">8.9.3</span> Protein-Protein Interaction Network .</a></li>
<li><a href="biological-networks-..html#wgcna-."><span class="toc-section-number">8.9.4</span> WGCNA .</a></li>
</ul></li>
</ul></li>
<li><a href="engineering.html#engineering"><span class="toc-section-number">9</span> EngineeRing</a>
<ul>
<li><a href="unit-testing.html#unit-testing"><span class="toc-section-number">9.1</span> Unit Testing</a></li>
<li><a href="toolification.html#toolification"><span class="toc-section-number">9.2</span> Toolification</a>
<ul>
<li><a href="toolification.html#the-r-interpreter"><span class="toc-section-number">9.2.1</span> The R Interpreter</a></li>
<li><a href="toolification.html#rscript"><span class="toc-section-number">9.2.2</span> <code>Rscript</code></a></li>
<li><a href="toolification.html#commandargs"><span class="toc-section-number">9.2.3</span> <code>commandArgs()</code></a></li>
</ul></li>
<li><a href="parallel-processing.html#parallel-processing"><span class="toc-section-number">9.3</span> Parallel Processing</a>
<ul>
<li><a href="parallel-processing.html#brief-introduction-to-parallelization"><span class="toc-section-number">9.3.1</span> Brief Introduction to Parallelization</a></li>
<li><a href="parallel-processing.html#apply-and-friends-are-pleasingly-parallel"><span class="toc-section-number">9.3.2</span> <code>apply</code> and Friends Are Pleasingly Parallel</a></li>
<li><a href="parallel-processing.html#the-parallel-package"><span class="toc-section-number">9.3.3</span> The <code>parallel</code> package</a></li>
</ul></li>
<li><a href="object-orientation-in-r.html#object-orientation-in-r"><span class="toc-section-number">9.4</span> Object Orientation in R</a></li>
<li><a href="r-packages.html#r-packages"><span class="toc-section-number">9.5</span> R Packages</a></li>
</ul></li>
<li><a href="rshiny.html#rshiny"><span class="toc-section-number">10</span> RShiny</a>
<ul>
<li><a href="rshiny-introduction.html#rshiny-introduction"><span class="toc-section-number">10.1</span> RShiny Introduction</a>
<ul>
<li><a href="rshiny-introduction.html#how-does-a-web-page-work"><span class="toc-section-number">10.1.1</span> How does a web page work?</a></li>
</ul></li>
<li><a href="application-structure.html#application-structure"><span class="toc-section-number">10.2</span> Application Structure</a>
<ul>
<li><a href="application-structure.html#ui-design"><span class="toc-section-number">10.2.1</span> UI Design</a></li>
<li><a href="application-structure.html#server-functionality"><span class="toc-section-number">10.2.2</span> Server Functionality</a></li>
</ul></li>
<li><a href="reactivity.html#reactivity"><span class="toc-section-number">10.3</span> Reactivity</a></li>
<li><a href="publishing.html#publishing"><span class="toc-section-number">10.4</span> Publishing</a></li>
</ul></li>
<li><a href="communicating-with-r.html#communicating-with-r"><span class="toc-section-number">11</span> Communicating with R</a>
<ul>
<li><a href="rmarkdown-knitr.html#rmarkdown-knitr"><span class="toc-section-number">11.1</span> RMarkdown &amp; knitr</a></li>
<li><a href="bookdown.html#bookdown"><span class="toc-section-number">11.2</span> bookdown</a></li>
</ul></li>
<li><a href="contribution-guide.html#contribution-guide"><span class="toc-section-number">12</span> Contribution Guide</a>
<ul>
<li><a href="custom-blocks.html#custom-blocks"><span class="toc-section-number">12.1</span> Custom blocks</a></li>
</ul></li>
<li><a href="assignments.html#assignments">Assignments</a>
<ul>
<li><a href="assignment-format.html#assignment-format">Assignment Format</a></li>
<li><a href="starting-an-assignment.html#starting-an-assignment">Starting an Assignment</a>
<ul>
<li><a href="starting-an-assignment.html#starting-your-own-git-repository-and-cloning-it">Starting your own <code>git</code> repository and cloning it</a></li>
<li><a href="starting-an-assignment.html#contents">Contents</a></li>
<li><a href="starting-an-assignment.html#github-tutorial">GitHub tutorial</a></li>
<li><a href="starting-an-assignment.html#committing-and-pushing-with-r-and-without">Committing and Pushing with R (and without)</a></li>
<li><a href="starting-an-assignment.html#using-rstudio">Using RStudio</a></li>
<li><a href="starting-an-assignment.html#using-the-command-line">Using the command line</a></li>
</ul></li>
<li><a href="assignment-1.html#assignment-1">Assignment 1</a>
<ul>
<li><a href="assignment-1.html#problem-statement">Problem Statement</a></li>
<li><a href="assignment-1.html#learning-objectives-and-skill-list">Learning Objectives and Skill List</a></li>
<li><a href="assignment-1.html#instructions">Instructions</a></li>
<li><a href="assignment-1.html#hints">Hints</a></li>
</ul></li>
<li><a href="assignment-2.html#assignment-2">Assignment 2</a>
<ul>
<li><a href="assignment-2.html#problem-statement-1">Problem Statement</a></li>
<li><a href="assignment-2.html#learning-objectives">Learning Objectives</a></li>
<li><a href="assignment-2.html#skill-list">Skill List</a></li>
<li><a href="assignment-2.html#instructions-1">Instructions</a></li>
<li><a href="assignment-2.html#function-details">Function Details</a></li>
</ul></li>
<li><a href="assignment-3.html#assignment-3">Assignment 3</a>
<ul>
<li><a href="assignment-3.html#problem-statement-2">Problem Statement</a></li>
<li><a href="assignment-3.html#learning-objectives-1">Learning Objectives</a></li>
<li><a href="assignment-3.html#skill-list-1">Skill List</a></li>
<li><a href="assignment-3.html#background-on-microarrays">Background on Microarrays</a></li>
<li><a href="assignment-3.html#background-on-principal-component-analysis">Background on Principal Component Analysis</a></li>
<li><a href="assignment-3.html#marisa-et-al.-gene-expression-classification-of-colon-cancer-into-molecular-subtypes-characterization-validation-and-prognostic-value.-plos-medicine-may-2013.-pmid-23700391">Marisa et al. Gene Expression Classification of Colon Cancer into Molecular Subtypes: Characterization, Validation, and Prognostic Value. PLoS Medicine, May 2013. PMID: 23700391</a></li>
<li><a href="assignment-3.html#scaling-data-using-r-scale">Scaling data using R <code>scale()</code></a></li>
<li><a href="assignment-3.html#proportion-of-variance-explained">Proportion of variance explained</a></li>
<li><a href="assignment-3.html#plotting-and-visualization-of-pca">Plotting and visualization of PCA</a></li>
<li><a href="assignment-3.html#hierarchical-clustering-and-heatmaps">Hierarchical Clustering and Heatmaps</a></li>
<li><a href="assignment-3.html#references">References</a></li>
</ul></li>
<li><a href="assignment-4.html#assignment-4">Assignment 4</a>
<ul>
<li><a href="assignment-4.html#problem-statement-3">Problem Statement</a></li>
<li><a href="assignment-4.html#learning-objectives-2">Learning Objectives</a></li>
<li><a href="assignment-4.html#skill-list-2">Skill List</a></li>
<li><a href="assignment-4.html#instructions-2">Instructions</a></li>
<li><a href="assignment-4.html#deliverables">Deliverables</a></li>
<li><a href="assignment-4.html#function-details-1">Function Details</a></li>
<li><a href="assignment-4.html#hints-1">Hints</a></li>
</ul></li>
<li><a href="assignment-5.html#assignment-5">Assignment 5</a>
<ul>
<li><a href="assignment-5.html#problem-statement-4">Problem Statement</a></li>
<li><a href="assignment-5.html#learning-objectives-3">Learning Objectives</a></li>
<li><a href="assignment-5.html#skill-list-3">Skill List</a></li>
<li><a href="assignment-5.html#deseq2-background">DESeq2 Background</a></li>
<li><a href="assignment-5.html#generating-a-counts-matrix">Generating a counts matrix</a></li>
<li><a href="assignment-5.html#prefiltering-counts-matrix">Prefiltering Counts matrix</a></li>
<li><a href="assignment-5.html#median-of-ratios-normalization">Median-of-ratios normalization</a></li>
<li><a href="assignment-5.html#deseq2-preparation">DESeq2 preparation</a></li>
<li><a href="assignment-5.html#omeara-et-al.-transcriptional-reversion-of-cardiac-myocyte-fate-during-mammalian-cardiac-regeneration.-circ-res.-feb-2015.-pmid-25477501l">O’Meara et al. Transcriptional Reversion of Cardiac Myocyte Fate During Mammalian Cardiac Regeneration. Circ Res. Feb 2015. PMID: 25477501l</a></li>
<li><a href="assignment-5.html#reading-and-subsetting-the-data-from-verse_counts.tsv-and-sample_metadata.csv">1. Reading and subsetting the data from verse_counts.tsv and sample_metadata.csv</a></li>
<li><a href="assignment-5.html#running-deseq2">2. Running DESeq2</a></li>
<li><a href="assignment-5.html#annotating-results-to-construct-a-labeled-volcano-plot">3. Annotating results to construct a labeled volcano plot</a></li>
<li><a href="assignment-5.html#diagnostic-plot-of-the-raw-p-values-for-all-genes">4. Diagnostic plot of the raw p-values for all genes</a></li>
<li><a href="assignment-5.html#plotting-the-logfoldchanges-for-differentially-expressed-genes">5. Plotting the LogFoldChanges for differentially expressed genes</a></li>
<li><a href="assignment-5.html#the-choice-of-fdr-cutoff-depends-on-cost">The choice of FDR cutoff depends on cost</a></li>
<li><a href="assignment-5.html#plotting-the-normalized-counts-of-differentially-expressed-genes">6. Plotting the normalized counts of differentially expressed genes</a></li>
<li><a href="assignment-5.html#volcano-plot-to-visualize-differential-expression-results">7. Volcano Plot to visualize differential expression results</a></li>
<li><a href="assignment-5.html#running-fgsea-vignette">8. Running fgsea <span>vignette</span></a></li>
<li><a href="assignment-5.html#plotting-the-top-ten-positive-nes-and-top-ten-negative-nes-pathways">9. Plotting the top ten positive NES and top ten negative NES pathways</a></li>
<li><a href="assignment-5.html#references-1">References</a></li>
</ul></li>
<li><a href="assignment-6.html#assignment-6">Assignment 6</a>
<ul>
<li><a href="assignment-6.html#problem-statement-5">Problem Statement</a></li>
<li><a href="assignment-6.html#learning-objectives-4">Learning Objectives</a></li>
<li><a href="assignment-6.html#skill-list-4">Skill List</a></li>
<li><a href="assignment-6.html#instructions-3">Instructions</a></li>
<li><a href="assignment-6.html#function-details-2">Function Details</a></li>
</ul></li>
<li><a href="assignment-7.html#assignment-7">Assignment 7</a>
<ul>
<li><a href="assignment-7.html#problem-statement-6">Problem Statement</a></li>
<li><a href="assignment-7.html#learning-objectives-5">Learning Objectives</a></li>
<li><a href="assignment-7.html#skill-list-5">Skill List</a></li>
<li><a href="assignment-7.html#instructions-4">Instructions</a></li>
<li><a href="assignment-7.html#function-details-3">“Function” Details</a></li>
</ul></li>
</ul></li>
<li class="appendix"><span><b>Appendix</b></span></li>
<li><a href="class-outlines.html#class-outlines"><span class="toc-section-number">A</span> Class Outlines</a>
<ul>
<li><a href="week-1.html#week-1"><span class="toc-section-number">A.1</span> Week 1</a></li>
<li><a href="week-2.html#week-2"><span class="toc-section-number">A.2</span> Week 2</a></li>
<li><a href="week-3.html#week-3"><span class="toc-section-number">A.3</span> Week 3</a></li>
<li><a href="week-4.html#week-4"><span class="toc-section-number">A.4</span> Week 4</a></li>
<li><a href="week-5.html#week-5"><span class="toc-section-number">A.5</span> Week 5</a></li>
<li><a href="week-6.html#week-6"><span class="toc-section-number">A.6</span> Week 6</a></li>
<li><a href="week-7.html#week-7"><span class="toc-section-number">A.7</span> Week 7</a></li>
<li><a href="week-8.html#week-8"><span class="toc-section-number">A.8</span> Week 8</a></li>
<li><a href="week-9.html#week-9"><span class="toc-section-number">A.9</span> Week 9</a></li>
<li><a href="week-10.html#week-10"><span class="toc-section-number">A.10</span> Week 10</a></li>
<li><a href="week-11.html#week-11"><span class="toc-section-number">A.11</span> Week 11</a></li>
<li><a href="week-12.html#week-12"><span class="toc-section-number">A.12</span> Week 12</a></li>
<li><a href="week-13.html#week-13"><span class="toc-section-number">A.13</span> Week 13</a></li>
<li><a href="week-14.html#week-14"><span class="toc-section-number">A.14</span> Week 14</a></li>
</ul></li>
</ul>
</div>
</div>
</div>
<div class="row">
<div class="col-sm-12">
<div id="gene-expression" class="section level2" number="8.7">
<h2><span class="header-section-number">8.7</span> Gene Expression</h2>
<p>Gene expression is the process by which information from a gene is used in the
synthesis of functional gene products that affect a phenotype. While gene
expression studies are often focused on protein coding genes, there are many
other functional molecules, e.g. <a href="https://www.genome.gov/genetics-glossary/Transfer-RNA">transfer
RNAs</a>,
<a href="https://en.wikipedia.org/wiki/Long_non-coding_RNA">lncRNAs</a>, etc, that are
produced by this process. The gene expression process has many steps and
intermediate products, as depicted in the following figure:</p>
<div class="figure">
<img src="gene_expression_nature.jpg" alt="" />
<p class="caption">The Gene Expression Process - An overview of the flow of information from DNA
to protein in a eukaryote</p>
</div>
<p>The specific parts of the genome that code for genes are copied, or
<em>transcribed</em> into RNA molecules called <em>transcripts</em>. In lower organisms like
bacteria, these RNA molecules are passed on directly to ribosomes, which
translate them into proteins. In most higher organisms like eukaryotes, these
initial transcripts, called <em>pre-messenger RNA (pre-mRNA) transcripts</em> are
further processed such that certain parts of the transcript, called <em>introns</em>,
are spliced out and the flanking sequences, called <em>exons</em>, are concatenated
together. After this splicing process is complete, the pre-mRNA transcripts
become mature <em>messenger RNA (mRNA) transcripts</em> which go on to be exported from
the nucleus and loaded into ribosomes in the cytoplasm to undergo translation
into proteins.</p>
<p>In gene expression studies, the relative abundance, or number of copies, of RNA
or mRNA transcripts in a sample is measured. The measurements are non-negative
numbers that are proportional to the relative abundance of a transcript with
respect to some reference, either another gene, or in the case of high
throughput assays like microarrays or high throughput sequencing, relative to
measurements of all distinct transcripts examined in the experiment.
Conceptually, the larger the magnitude of a transcript abundance measurement,
the more copies of that transcript were in the original sample.</p>
<p>There are many ways to measure the abundance of RNA transcripts; the following
are some of the common methods at the time of writing:</p>
<ul>
<li>Light absorbance - RNA absorbs ultraviolet light with wavelength 260 nm, which
can be used to determine RNA concentration in a sample using specialized equipment</li>
<li><a href="https://www.genome.gov/genetics-glossary/Northern-Blot">Northern blot</a> -
measures relative abundance of an RNA with a specific sequence</li>
<li><a href="https://www.thermofisher.com/blog/ask-a-scientist/what-is-qpcr/">quantitative polymerase chain reaction
(qPCR)</a> -
measures relative abundance of an RNA with a specific sequence using <a href="https://en.wikipedia.org/wiki/Polymerase_chain_reaction">PCR
amplification</a></li>
<li><a href="microarrays.html#microarrays">Oligonucleotide and microarrays</a> - measures relative abundance
of thousands of genes simultaneously using known DNA probe sequences and
fluorescently tagged RNA molecules</li>
<li><a href="gene-expression.html#rnaseq">High throughput RNA sequencing (RNASeq)</a> - measures thousands to
millions of RNA fragments simultaneously in proportion to their relative
abundance</li>
</ul>
<p>While any of the measurement methods above may be analyzed in R, the high
throughput methods (i.e. microarray, high throughput sequencing) are the primary
concern of this chapter. These methods generate measurements for thousands of
transcripts or genes simultaneously, requiring the power and flexibility of
programmatic analysis to process in a practical amount of time. The remainder of
this chapter is devoted to understanding the specific technologies, data types,
and analytical strategies involved in working with these data.</p>
<div class="box important">
<p>Gene expression measurements are almost always inherently relative, either due
to limitations of the measurement methods (e.g. microarrays, described below) or
because measuring all molecules in a sample will almost always be prohibitively
expensive, have diminishing returns, and it is very difficult if not impossible
to determine if all the molecules have been measured. This means we cannot in
general associate the numbers associated with the measurements with an absolute
molecular copy number. An important implication of the inherent relativity of
these measurements is: <strong>absence of evidence is not evidence of absence</strong>. In
other words, if a transcript has an abundance measurement of zero, this does
<em>not necessarily imply that the gene is not expressed</em>. It may be that the gene
is indeed expressed, but the copy number is sufficiently small that it was not
detected by the assay.</p>
</div>
<div id="gene-expression-data-in-bioconductor" class="section level3" number="8.7.1">
<h3><span class="header-section-number">8.7.1</span> Gene Expression Data in Bioconductor</h3>
<p>The <a href="https://bioconductor.org/packages/release/bioc/html/SummarizedExperiment.html">SummarizedExperiment
container</a>
is the standard way to load and work with gene expression data in Bioconductor.
This container requires the following information:</p>
<ul>
<li><code>assays</code> - one or more measurement assays (e.g. gene expression) in the form of a feature by sample matrix</li>
<li><code>colData</code> - metadata associated with the samples (i.e. columns) of the assays</li>
<li><code>rowData</code> - metadata associated with the features (i.e. rows) of the assays</li>
<li><code>exptData</code> - additional metadata about the experiment itself, like protocol, project name, etc</li>
</ul>
<p>The figure below illustrates how the SummarizedExperiment container is
structured and how the different data elements are accessed:</p>
<div class="figure">
<img src="https://media.springernature.com/full/springer-static/image/art%3A10.1038%2Fnmeth.3252/MediaObjects/41592_2015_Article_BFnmeth3252_Fig2_HTML.jpg?as=webp" alt="" />
<p class="caption">SummarizedExperiment Schematic - Huber, et al. 2015. “Orchestrating High-Throughput Genomic Analysis with Bioconductor.” Nature Methods 12 (2): 115–21.</p>
</div>
<p>Many Bioconductor packages for specific types of data, e.g.
<a href="#differential-expression--microarrays-limma">limma</a> create these
SummarizedExperiment objects for you, but you may also create your own by
assembling each of these data into data frames manually:</p>
<div class="sourceCode" id="cb367"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb367-1"><a href="gene-expression.html#cb367-1" aria-hidden="true" tabindex="-1"></a><span class="co"># microarray expression dataset intensities</span></span>
<span id="cb367-2"><a href="gene-expression.html#cb367-2" aria-hidden="true" tabindex="-1"></a>intensities <span class="ot">&lt;-</span> readr<span class="sc">::</span><span class="fu">read_delim</span>(<span class="st">&quot;example_intensity_data.csv&quot;</span>,<span class="at">delim=</span><span class="st">&quot; &quot;</span>)</span>
<span id="cb367-3"><a href="gene-expression.html#cb367-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb367-4"><a href="gene-expression.html#cb367-4" aria-hidden="true" tabindex="-1"></a><span class="co"># the first column of intensities tibble is the probesetId, extract to pass as rowData</span></span>
<span id="cb367-5"><a href="gene-expression.html#cb367-5" aria-hidden="true" tabindex="-1"></a>rowData <span class="ot">&lt;-</span> intensities[<span class="st">&quot;probeset_id&quot;</span>]</span>
<span id="cb367-6"><a href="gene-expression.html#cb367-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb367-7"><a href="gene-expression.html#cb367-7" aria-hidden="true" tabindex="-1"></a><span class="co"># remove probeset IDs from tibble and turn into a R dataframe so that we can assign rownames</span></span>
<span id="cb367-8"><a href="gene-expression.html#cb367-8" aria-hidden="true" tabindex="-1"></a><span class="co"># since tibbles don&#39;t support row names</span></span>
<span id="cb367-9"><a href="gene-expression.html#cb367-9" aria-hidden="true" tabindex="-1"></a>intensities <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(</span>
<span id="cb367-10"><a href="gene-expression.html#cb367-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(intensities, <span class="sc">-</span>probeset_id)</span>
<span id="cb367-11"><a href="gene-expression.html#cb367-11" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb367-12"><a href="gene-expression.html#cb367-12" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(intensities) <span class="ot">&lt;-</span> rowData<span class="sc">$</span>probeset_id</span>
<span id="cb367-13"><a href="gene-expression.html#cb367-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb367-14"><a href="gene-expression.html#cb367-14" aria-hidden="true" tabindex="-1"></a><span class="co"># these column data are made up, but you would have a sample metadata file to use</span></span>
<span id="cb367-15"><a href="gene-expression.html#cb367-15" aria-hidden="true" tabindex="-1"></a>colData <span class="ot">&lt;-</span> <span class="fu">tibble</span>(</span>
<span id="cb367-16"><a href="gene-expression.html#cb367-16" aria-hidden="true" tabindex="-1"></a>  <span class="at">sample_name=</span><span class="fu">colnames</span>(intensities),</span>
<span id="cb367-17"><a href="gene-expression.html#cb367-17" aria-hidden="true" tabindex="-1"></a>  <span class="at">condition=</span><span class="fu">sample</span>(<span class="fu">c</span>(<span class="st">&quot;case&quot;</span>,<span class="st">&quot;control&quot;</span>),<span class="fu">ncol</span>(intensities),<span class="at">replace=</span><span class="cn">TRUE</span>)</span>
<span id="cb367-18"><a href="gene-expression.html#cb367-18" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb367-19"><a href="gene-expression.html#cb367-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb367-20"><a href="gene-expression.html#cb367-20" aria-hidden="true" tabindex="-1"></a>se <span class="ot">&lt;-</span> <span class="fu">SummarizedExperiment</span>(</span>
<span id="cb367-21"><a href="gene-expression.html#cb367-21" aria-hidden="true" tabindex="-1"></a>   <span class="at">assays=</span><span class="fu">list</span>(<span class="at">intensities=</span>intensities),</span>
<span id="cb367-22"><a href="gene-expression.html#cb367-22" aria-hidden="true" tabindex="-1"></a>   <span class="at">colData=</span>colData,</span>
<span id="cb367-23"><a href="gene-expression.html#cb367-23" aria-hidden="true" tabindex="-1"></a>   <span class="at">rowData=</span>rowData</span>
<span id="cb367-24"><a href="gene-expression.html#cb367-24" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb367-25"><a href="gene-expression.html#cb367-25" aria-hidden="true" tabindex="-1"></a>se</span></code></pre></div>
<pre><code>## class: SummarizedExperiment
## dim: 54675 35
## metadata(0):
## assays(1): intensities
## rownames(54675): 1007_s_at 1053_at ... AFFX-TrpnX-5_at AFFX-TrpnX-M_at
## rowData names(1): probeset_id
## colnames(35): GSM972389 GSM972390 ... GSM972512 GSM972521
## colData names(2): sample_name condition</code></pre>
<p>Detailed documentation of how to create and use the SummarizedExperiment is available in the <a href="https://bioconductor.org/packages/release/bioc/vignettes/SummarizedExperiment/inst/doc/SummarizedExperiment.html#constructing-a-summarizedexperiment">SummarizedExperiment vignette</a>.</p>
<div class="box note">
<p>SummarizedExperiment is the successor to the older
<a href="https://www.bioconductor.org/packages/devel/bioc/vignettes/Biobase/inst/doc/ExpressionSetIntroduction.pdf">ExpressionSet</a>
container. Both are still used by Bioconductor packages, but
SummarizedExperiment is more modern and flexible, so it is suggested for use
whenever possible.</p>
</div>
</div>
<div id="differential-expression-analysis" class="section level3" number="8.7.2">
<h3><span class="header-section-number">8.7.2</span> Differential Expression Analysis</h3>
<p>Differential expression analysis seeks to identify to what extent gene
expression is associated with one or more variables of interest. For example,
we might be interested in genes that have higher expression in subjects with a
disease, or which genes change in response to a treatment.</p>
<p>Typically, gene expression analysis requires two types of data to run: an
expression matrix and a <em>design matrix</em>. The expression matrix will generally
have features (e.g. genes) as rows and samples as columns. The design matrix is
a numeric matrix that contains the variables we wish to model and any covariates
or confounders we wish to adjust for. The variables in the design matrix then
must be encoded in a way that statistical procedures can understand. The full
details of how design matrices are constructed is beyond the scope of this book.
Fortunately, R makes it very easy to construct these matrices from a tibble with
the <a href="https://www.rdocumentation.org/packages/stats/versions/3.6.2/topics/model.matrix"><code>model.matrix()</code>
function</a>.
Consider for the following imaginary sample information tibble that has 10 AD
and 10 Control subjects with different clinical and protein histology
measurements assayed on their brain tissue:</p>
<div class="sourceCode" id="cb369"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb369-1"><a href="gene-expression.html#cb369-1" aria-hidden="true" tabindex="-1"></a>ad_metadata</span></code></pre></div>
<pre><code>## # A tibble: 20 x 8
##    ID    age_at_death condition    tau  abeta   iba1   gfap braak_stage
##    &lt;chr&gt;        &lt;dbl&gt; &lt;fct&gt;      &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt; &lt;fct&gt;      
##  1 A1              81 AD         98543 166078  77299  75277 4          
##  2 A2              80 AD         32711  56919  49571 117441 1          
##  3 A3              91 AD        134427 295407 106724  67719 6          
##  4 A4              90 AD         81515 221801 171212  56346 4          
##  5 A5              80 AD         42442  97918   8321 109406 2          
##  6 A6              81 AD         54802  49358  82374  42797 2          
##  7 A7              77 AD        127166   8417  62675  85377 6          
##  8 A8              74 AD        103015  28478 131168 133892 5          
##  9 A9              82 AD        114270 188208 194818  61935 5          
## 10 A10             83 AD         90785  46279 121555 131853 4          
## 11 C1              79 Control      991    237   1062  64752 0          
## 12 C2              76 Control     2804   4423   3441  23384 0          
## 13 C3              73 Control     3120   7073   1402 135049 0          
## 14 C4              73 Control     7167    798   8365  18961 0          
## 15 C5              69 Control     6591   2590   4885  65094 0          
## 16 C6              77 Control    38697 103664  33430  23136 2          
## 17 C7              73 Control    58381  72071  16864  17295 3          
## 18 C8              75 Control     5102  10533   5321  89479 0          
## 19 C9              76 Control    59906  47967  19103 124651 3          
## 20 C10             81 Control    46553  49670  90763  52684 2</code></pre>
<p>We might be interested in identifying genes that are increased or decreased in
people with AD compared with Controls. We can create a model matrix for this as
follows:</p>
<div class="sourceCode" id="cb371"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb371-1"><a href="gene-expression.html#cb371-1" aria-hidden="true" tabindex="-1"></a><span class="fu">model.matrix</span>(<span class="sc">~</span> condition, <span class="at">data=</span>ad_metadata)</span></code></pre></div>
<pre><code>##    (Intercept) conditionAD
## 1            1           1
## 2            1           1
## 3            1           1
## 4            1           1
## 5            1           1
## 6            1           1
## 7            1           1
## 8            1           1
## 9            1           1
## 10           1           1
## 11           1           0
## 12           1           0
## 13           1           0
## 14           1           0
## 15           1           0
## 16           1           0
## 17           1           0
## 18           1           0
## 19           1           0
## 20           1           0
## attr(,&quot;assign&quot;)
## [1] 0 1
## attr(,&quot;contrasts&quot;)
## attr(,&quot;contrasts&quot;)$condition
## [1] &quot;contr.treatment&quot;</code></pre>
<p>The <code>~ condition</code> argument is an an R
<a href="https://rviews.rstudio.com/2017/02/01/the-r-formula-method-the-good-parts/">forumla</a>,
which is a concise description of how variable should be included in the model.
The general format of a formula is as follows:</p>
<pre><code># portions in [] are optional
[&lt;outcome variable&gt;] ~ &lt;predictor variable 1&gt; [+ &lt;predictor variable 2&gt;]...

# examples

# model Gene 3 expression as a function of disease status
`Gene 3` ~ condition

# model the amount of tau pathology as a function of abeta pathology,
# adjusting for age at death
tau ~ age_at_death + abeta

# create a model without an outcome variable that can be used to create the
# model matrix to test for differences with disease status adjusting for age at death
~ age_at_death + condition</code></pre>
<p>For most models, the design matrix will have an intercept term of all ones and
additional colums for the other variables in the model. We might wish to adjust
out the effect of age by including <code>age_at_death</code> as a covariate in the model:</p>
<div class="sourceCode" id="cb374"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb374-1"><a href="gene-expression.html#cb374-1" aria-hidden="true" tabindex="-1"></a><span class="fu">model.matrix</span>(<span class="sc">~</span> age_at_death <span class="sc">+</span> condition, <span class="at">data=</span>ad_metadata)</span></code></pre></div>
<pre><code>##    (Intercept) age_at_death conditionAD
## 1            1           81           1
## 2            1           80           1
## 3            1           91           1
## 4            1           90           1
## 5            1           80           1
## 6            1           81           1
## 7            1           77           1
## 8            1           74           1
## 9            1           82           1
## 10           1           83           1
## 11           1           79           0
## 12           1           76           0
## 13           1           73           0
## 14           1           73           0
## 15           1           69           0
## 16           1           77           0
## 17           1           73           0
## 18           1           75           0
## 19           1           76           0
## 20           1           81           0
## attr(,&quot;assign&quot;)
## [1] 0 1 2
## attr(,&quot;contrasts&quot;)
## attr(,&quot;contrasts&quot;)$condition
## [1] &quot;contr.treatment&quot;</code></pre>
<p>The model matrix now includes a column for age at death as well. This model
model matrix is now suitable to pass to differential expression software
packages to look for genes associated with disease status.</p>
<p>As mentioned above, modern gene expression assays measure thousands of genes
simultaneously. This means each gene must be tested for differential expression
individually. In general, each gene is tested using the same statistical model,
so a differential expression analysis package will perform something equivalent
to the following:</p>
<pre><code>Gene 1 ~ age_at_death + condition
Gene 2 ~ age_at_death + condition
Gene 3 ~ age_at_death + condition
...
Gene N ~ age_at_death + condition</code></pre>
<p>Each gene model will have its own statistics associated with it that we must
process and interpret after the analysis is complete.</p>
</div>
<div id="microarray-gene-expression-data" class="section level3" number="8.7.3">
<h3><span class="header-section-number">8.7.3</span> Microarray Gene Expression Data</h3>
<p>On a high level, there are four steps involved in analyzing gene expression
microarray data:</p>
<ol style="list-style-type: decimal">
<li><strong>Summarization of probes to probesets.</strong> Each gene is represented by
multiple probes with different sequences. Summarization is a statistical
procedure that computes a single value for each probeset from its corresponding
probes.</li>
<li><strong>Normalization.</strong> This includes removing background signal from individual
arrays as well as adjusting probeset intensities so that they are comparable
across multiple sample arrays.</li>
<li><strong>Quality control.</strong> Compare all normalized samples within a sample set to
identify, mitigate, or eliminate potential outlier samples.</li>
<li><strong>Analysis.</strong> Using the quality controlled expression data, Implement
statistical analysis to answer research questions.</li>
</ol>
<p>The full details of microarray analysis are beyond the scope of this book.
However in the following sections we cover some of the basic entry points to
performing these steps in R and Bioconductor.</p>
<p>The CEL data files from a set of microarrays can be loaded into R for analysis
using the <a href="https://www.bioconductor.org/packages/release/bioc/html/affy.html"><code>affy</code> Bioconductor
package</a>.
This package provides all the functions necessary for loading the data and
performing key preprocessing operations. Typically, two or more samples were
processed in this way, resulting in a set of CEL files that should be processed
together. These CEL files will typically be all stored in the same directory,
and may be loaded using the <code>affy::ReadAffy</code> function:</p>
<div class="sourceCode" id="cb377"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb377-1"><a href="gene-expression.html#cb377-1" aria-hidden="true" tabindex="-1"></a><span class="co"># read all CEL files in a single directory</span></span>
<span id="cb377-2"><a href="gene-expression.html#cb377-2" aria-hidden="true" tabindex="-1"></a>affy_batch <span class="ot">&lt;-</span> affy<span class="sc">::</span><span class="fu">ReadAffy</span>(<span class="at">celfile.path=</span><span class="st">&quot;directory/of/CELfiles&quot;</span>)</span>
<span id="cb377-3"><a href="gene-expression.html#cb377-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb377-4"><a href="gene-expression.html#cb377-4" aria-hidden="true" tabindex="-1"></a><span class="co"># or individual files in different directories</span></span>
<span id="cb377-5"><a href="gene-expression.html#cb377-5" aria-hidden="true" tabindex="-1"></a>cel_filenames <span class="ot">&lt;-</span> <span class="fu">c</span>(</span>
<span id="cb377-6"><a href="gene-expression.html#cb377-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">list.files</span>( <span class="at">path=</span><span class="st">&quot;CELfile_dir1&quot;</span>, <span class="at">full.names=</span><span class="cn">TRUE</span>, <span class="at">pattern=</span><span class="st">&quot;.CEL$&quot;</span> ),</span>
<span id="cb377-7"><a href="gene-expression.html#cb377-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">list.files</span>( <span class="at">path=</span><span class="st">&quot;CELfile_dir2&quot;</span>, <span class="at">full.names=</span><span class="cn">TRUE</span>, <span class="at">pattern=</span><span class="st">&quot;.CEL$&quot;</span> )</span>
<span id="cb377-8"><a href="gene-expression.html#cb377-8" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb377-9"><a href="gene-expression.html#cb377-9" aria-hidden="true" tabindex="-1"></a>affy_batch <span class="ot">&lt;-</span> affy<span class="sc">::</span><span class="fu">ReadAffy</span>(<span class="at">filenames=</span>cel_filenames)</span></code></pre></div>
<p>The <code>affy_batch</code> variable is a <code>AffyBatch</code> container, which stores information
on the probe definitions based on the type of microarray, probe-level intensity
for each sample, and other information about the experiment contained within the
CEL files.</p>
<p>The <code>affy</code> package provides functions to perform probe summarization and
normalization. The most popular method to accomplish this at the time of writing
is called Robust Multi-array Average or RMA <span class="citation">(<a href="#ref-Irizarry2003-hn" role="doc-biblioref">Irizarry et al. 2003</a>)</span>, which performs
summarization and normalization of multiple arrays simultaneously. Below, the
RMA algorithm is used to normalize an example dataset provided by the
<a href="https://bioconductor.org/packages/release/data/experiment/html/affydata.html"><code>affydata</code></a>
Bioconductor package.</p>
<div class="box note">
<p>Certain Bioconductor packages provide example datasets for use with companion
analysis packages. For example, the <code>affydata</code> package provides the <code>Dilution</code>
dataset, which was generated using two concentrations of cDNA from human liver
tissue and a central nervous system cell line. To load a data package into R,
first run <code>library(&lt;data package&gt;)</code> and then <code>data(&lt;dataset name&gt;)</code>.</p>
</div>
<div class="sourceCode" id="cb378"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb378-1"><a href="gene-expression.html#cb378-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(affy)</span>
<span id="cb378-2"><a href="gene-expression.html#cb378-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(affydata)</span></code></pre></div>
<pre><code>##      Package    LibPath                                     Item      
## [1,] &quot;affydata&quot; &quot;C:/Users/Adam/Documents/R/win-library/4.1&quot; &quot;Dilution&quot;
##      Title                        
## [1,] &quot;AffyBatch instance Dilution&quot;</code></pre>
<div class="sourceCode" id="cb380"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb380-1"><a href="gene-expression.html#cb380-1" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(Dilution)</span>
<span id="cb380-2"><a href="gene-expression.html#cb380-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb380-3"><a href="gene-expression.html#cb380-3" aria-hidden="true" tabindex="-1"></a><span class="co"># normalize the Dilution microarray expression values</span></span>
<span id="cb380-4"><a href="gene-expression.html#cb380-4" aria-hidden="true" tabindex="-1"></a><span class="co"># note Dilution is an AffyBatch object</span></span>
<span id="cb380-5"><a href="gene-expression.html#cb380-5" aria-hidden="true" tabindex="-1"></a>eset_rma <span class="ot">&lt;-</span> affy<span class="sc">::</span><span class="fu">rma</span>(Dilution,<span class="at">verbose=</span><span class="cn">FALSE</span>)</span>
<span id="cb380-6"><a href="gene-expression.html#cb380-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb380-7"><a href="gene-expression.html#cb380-7" aria-hidden="true" tabindex="-1"></a><span class="co"># plot distribution of non-normalized probes</span></span>
<span id="cb380-8"><a href="gene-expression.html#cb380-8" aria-hidden="true" tabindex="-1"></a><span class="co"># note rma normalization takes the log2 of the expression values,</span></span>
<span id="cb380-9"><a href="gene-expression.html#cb380-9" aria-hidden="true" tabindex="-1"></a><span class="co"># so we must do so on the raw data to compare</span></span>
<span id="cb380-10"><a href="gene-expression.html#cb380-10" aria-hidden="true" tabindex="-1"></a>raw_intensity <span class="ot">&lt;-</span> <span class="fu">as_tibble</span>(<span class="fu">exprs</span>(Dilution)) <span class="sc">%&gt;%</span></span>
<span id="cb380-11"><a href="gene-expression.html#cb380-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">probeset_id=</span><span class="fu">rownames</span>(<span class="fu">exprs</span>(Dilution))) <span class="sc">%&gt;%</span></span>
<span id="cb380-12"><a href="gene-expression.html#cb380-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(<span class="sc">-</span>probeset_id, <span class="at">names_to=</span><span class="st">&quot;Sample&quot;</span>, <span class="at">values_to =</span> <span class="st">&quot;Intensity&quot;</span>) <span class="sc">%&gt;%</span></span>
<span id="cb380-13"><a href="gene-expression.html#cb380-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(</span>
<span id="cb380-14"><a href="gene-expression.html#cb380-14" aria-hidden="true" tabindex="-1"></a>    <span class="st">`</span><span class="at">log2 Intensity</span><span class="st">`</span><span class="ot">=</span><span class="fu">log2</span>(Intensity),</span>
<span id="cb380-15"><a href="gene-expression.html#cb380-15" aria-hidden="true" tabindex="-1"></a>    <span class="at">Category=</span><span class="st">&quot;Before Normalization&quot;</span></span>
<span id="cb380-16"><a href="gene-expression.html#cb380-16" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb380-17"><a href="gene-expression.html#cb380-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb380-18"><a href="gene-expression.html#cb380-18" aria-hidden="true" tabindex="-1"></a><span class="co"># plot distribution of normalized probes</span></span>
<span id="cb380-19"><a href="gene-expression.html#cb380-19" aria-hidden="true" tabindex="-1"></a>rma_intensity <span class="ot">&lt;-</span> <span class="fu">as_tibble</span>(<span class="fu">exprs</span>(eset_rma)) <span class="sc">%&gt;%</span></span>
<span id="cb380-20"><a href="gene-expression.html#cb380-20" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">probesetid=</span><span class="fu">featureNames</span>(eset_rma)) <span class="sc">%&gt;%</span></span>
<span id="cb380-21"><a href="gene-expression.html#cb380-21" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(<span class="sc">-</span>probesetid, <span class="at">names_to=</span><span class="st">&quot;Sample&quot;</span>, <span class="at">values_to =</span> <span class="st">&quot;log2 Intensity&quot;</span>) <span class="sc">%&gt;%</span></span>
<span id="cb380-22"><a href="gene-expression.html#cb380-22" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">Category=</span><span class="st">&quot;RMA Normalized&quot;</span>)</span>
<span id="cb380-23"><a href="gene-expression.html#cb380-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb380-24"><a href="gene-expression.html#cb380-24" aria-hidden="true" tabindex="-1"></a>dplyr<span class="sc">::</span><span class="fu">bind_rows</span>(raw_intensity, rma_intensity) <span class="sc">%&gt;%</span></span>
<span id="cb380-25"><a href="gene-expression.html#cb380-25" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x=</span>Sample, <span class="at">y=</span><span class="st">`</span><span class="at">log2 Intensity</span><span class="st">`</span>)) <span class="sc">+</span></span>
<span id="cb380-26"><a href="gene-expression.html#cb380-26" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>() <span class="sc">+</span></span>
<span id="cb380-27"><a href="gene-expression.html#cb380-27" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(<span class="fu">vars</span>(Category))</span></code></pre></div>
<p><img src="main_files/microarray%20norm-1.png" width="672" /></p>
<p>Above, we first apply RMA to the <code>Dilution</code> microarray dataset using the
<code>affy::rma()</code> function. Then we plot the log 2 intensities of the probes in each
array, first using the raw intensities and then afysis seeking to
identify how associated gene expression is with one or more variables of
interest. For example,</p>
</div>
<div id="differential-expression-microarrays-limma" class="section level3" number="8.7.4">
<h3><span class="header-section-number">8.7.4</span> Differential Expression: Microarrays (limma)</h3>
<p><a href="https://bioconductor.org/packages/release/bioc/html/limma.html">limma</a>, which
is short for <code>li</code>near <code>m</code>odels of <code>mi</code>croarrays, is one of the <a href="https://bioconductor.org/packages/stats/">top most
downloaded</a> Bioconductor packages.
limma is utilized for analyzing microarray gene expression data, with a focus on
analyses using linear models to integrate <em>all</em> of the data from an experiment.
limma was developed for microarray analysis prior to the development of
sequencing based gene expression methods (i.e. <a href="gene-expression.html#rnaseq">RNASeq</a>) but has since added
functionality to analyze other types of gene expression data.</p>
<p>limma excels at analyzing these types of data as it can support arbitrarily
complex experimental designs while maintaining strong statistical power. An
experiment with a large number of conditions or predictors can still be analyze
even with small sample sizes. The method accomplishes this by using an
<a href="https://en.wikipedia.org/wiki/Empirical_Bayes_method">empirical Bayes</a> approach
that borrows information across all the genes in the dataset to better control
error in individual gene models. See <span class="citation"><a href="#ref-Ritchie2015-bj" role="doc-biblioref">Ritchie et al.</a> (<a href="#ref-Ritchie2015-bj" role="doc-biblioref">2015</a>)</span> for more details on how
limma works.</p>
<p>limma requres an expression matrix like that stored in an ExpressionSet or
SummarizedExperiment container and a design matrix:</p>
<div class="sourceCode" id="cb381"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb381-1"><a href="gene-expression.html#cb381-1" aria-hidden="true" tabindex="-1"></a><span class="co"># intensities is an example gene expression matrix that corresponds to our</span></span>
<span id="cb381-2"><a href="gene-expression.html#cb381-2" aria-hidden="true" tabindex="-1"></a><span class="co"># AD sample metadata of 10 AD and 10 Control individuals</span></span>
<span id="cb381-3"><a href="gene-expression.html#cb381-3" aria-hidden="true" tabindex="-1"></a>ad_se  <span class="ot">&lt;-</span> <span class="fu">SummarizedExperiment</span>(</span>
<span id="cb381-4"><a href="gene-expression.html#cb381-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">assays=</span><span class="fu">list</span>(<span class="at">intensities=</span>intensities),</span>
<span id="cb381-5"><a href="gene-expression.html#cb381-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">colData=</span>ad_metadata,</span>
<span id="cb381-6"><a href="gene-expression.html#cb381-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">rowData=</span><span class="fu">rownames</span>(intensities)</span>
<span id="cb381-7"><a href="gene-expression.html#cb381-7" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb381-8"><a href="gene-expression.html#cb381-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb381-9"><a href="gene-expression.html#cb381-9" aria-hidden="true" tabindex="-1"></a><span class="co"># define our design matrix for AD vs control, adjusting for age at death</span></span>
<span id="cb381-10"><a href="gene-expression.html#cb381-10" aria-hidden="true" tabindex="-1"></a>ad_vs_control_model <span class="ot">&lt;-</span> <span class="fu">model.matrix</span>(<span class="sc">~</span> age_at_death <span class="sc">+</span> condition, <span class="at">data=</span>ad_metadata)</span>
<span id="cb381-11"><a href="gene-expression.html#cb381-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb381-12"><a href="gene-expression.html#cb381-12" aria-hidden="true" tabindex="-1"></a><span class="co"># now run limma</span></span>
<span id="cb381-13"><a href="gene-expression.html#cb381-13" aria-hidden="true" tabindex="-1"></a><span class="co"># first fit all genes with the model</span></span>
<span id="cb381-14"><a href="gene-expression.html#cb381-14" aria-hidden="true" tabindex="-1"></a>fit <span class="ot">&lt;-</span> limma<span class="sc">::</span><span class="fu">lmFit</span>(</span>
<span id="cb381-15"><a href="gene-expression.html#cb381-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">assays</span>(se)<span class="sc">$</span>intensities,</span>
<span id="cb381-16"><a href="gene-expression.html#cb381-16" aria-hidden="true" tabindex="-1"></a>  ad_vs_control_model</span>
<span id="cb381-17"><a href="gene-expression.html#cb381-17" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb381-18"><a href="gene-expression.html#cb381-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb381-19"><a href="gene-expression.html#cb381-19" aria-hidden="true" tabindex="-1"></a><span class="co"># now better control fit errors with the empirical Bayes method</span></span>
<span id="cb381-20"><a href="gene-expression.html#cb381-20" aria-hidden="true" tabindex="-1"></a>fit <span class="ot">&lt;-</span> limma<span class="sc">::</span><span class="fu">eBayes</span>(fit)</span></code></pre></div>
<p>We have now conducted a differential expression analysis with limma. We can
extract out the results for our question of interest - which genes are
associated with AD - using the <code>limma::topTable()</code> function:</p>
<div class="sourceCode" id="cb382"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb382-1"><a href="gene-expression.html#cb382-1" aria-hidden="true" tabindex="-1"></a><span class="co"># the coefficient name conditionAD is the column name in the design matrix</span></span>
<span id="cb382-2"><a href="gene-expression.html#cb382-2" aria-hidden="true" tabindex="-1"></a><span class="co"># adjust=&quot;BH&quot; means perform Benjamini-Hochberg multiple testing adjustment</span></span>
<span id="cb382-3"><a href="gene-expression.html#cb382-3" aria-hidden="true" tabindex="-1"></a><span class="co"># on the nominal p-values from each gene test</span></span>
<span id="cb382-4"><a href="gene-expression.html#cb382-4" aria-hidden="true" tabindex="-1"></a><span class="co"># topTable only returns the top 10 results sorted by ascending p-value by default</span></span>
<span id="cb382-5"><a href="gene-expression.html#cb382-5" aria-hidden="true" tabindex="-1"></a><span class="fu">topTable</span>(fit, <span class="at">coef=</span><span class="st">&quot;conditionAD&quot;</span>, <span class="at">adjust=</span><span class="st">&quot;BH&quot;</span>)</span></code></pre></div>
<pre><code>##                   logFC  AveExpr         t      P.Value adj.P.Val         B
## 234942_s_at   1.1231942 8.397425  4.890779 9.613196e-05  0.770605 -2.987573
## 202423_at    -0.7499088 8.646687 -4.785281 1.221646e-04  0.770605 -3.025525
## 1557538_at   -0.8057310 4.348859 -4.697520 1.492072e-04  0.770605 -3.057737
## 1558290_a_at -0.8580750 7.278878 -4.658725 1.630238e-04  0.770605 -3.072162
## 234596_at     0.4925167 3.038448  4.564888 2.020432e-04  0.770605 -3.107523
## 225986_x_at  -0.6030385 6.958092 -4.502976 2.328354e-04  0.770605 -3.131216
## 203124_s_at  -1.2857166 8.565022 -4.488589 2.406451e-04  0.770605 -3.136763
## 209346_s_at   0.6578475 6.010219  4.421869 2.804637e-04  0.770605 -3.162688
## 224825_at     1.0266849 8.147104  4.421597 2.806391e-04  0.770605 -3.162795
## 205647_at    -0.8569475 4.399029 -4.399700 2.951151e-04  0.770605 -3.171376</code></pre>
<p>From the table, none of the probesets in this experiment are significant at FDR
&lt; 0.05.</p>
<div class="box readmore">
<ul>
<li><a href="https://bioinf.wehi.edu.au/limma/">Homepage</a></li>
<li><a href="https://bioconductor.org/packages/release/bioc/html/limma.html">BioConductor</a></li>
<li><a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4402510/">Publication</a><br />
</li>
</ul>
</div>
</div>
<div id="rnaseq" class="section level3" number="8.7.5">
<h3><span class="header-section-number">8.7.5</span> RNASeq</h3>
<p>RNA sequencing (RNASeq) is a HTS technology that measures the abundance of RNA
molecules in a sample. Briefly, RNA molecules are extracted from a sample using
a biochemical protocol that isolates RNA from other molecules (i.e. DNA,
proteins, etc) in the sample. Ribosomal RNA is removed from the sample (see Note
box below), and the remaining RNA molecules are size selected to retain only
short molecules (~15-300 nucleotides in length, depending on the protocol). The
size selected molecules are then reverse transcribed into cDNA and converted to
double stranded molecules. Sequencing libraries are prepared with these cDNA
using proprietary biochemical protocols to make them suitable for sequencing on
the sequencing instrument. After sequencing, FASTQ files containing millions of
reads for one or more samples is produced. Except for the initial RNA
extraction, all of these steps are typically performed by specialized
instrumentation facilities called <em>sequencing cores</em> who provide the data in
FASTQ format to the investigators.</p>
<div class="box note">
<p>80%-90% of the RNA mass in a cell is ribosomal RNA (rRNA), the RNA that makes up
ribosomes and is responsible for maintaining protein translation processes in
the cell. If total RNA was sequenced, the large majority of reads would
correspond to rRNAs and would therefore be of little or no use. For this reason,
rRNA molecules are removed from the RNA material that goes on to sequencing
using either a <em>poly-A tail enrichment</em> or <em>ribo-depletion</em> strategy. These two
methods maintain different populations of RNA and therefore sequencing datasets
generated with different strategies require different interpretations. The more
detail on this topic is beyond the scope of this book, but <span class="citation">(<a href="#ref-Zhao2018-ao" role="doc-biblioref">Zhao et al. 2018</a>)</span>
provide a good introduction and comparison of both approaches.</p>
</div>
<p>The reads produced in an RNASeq experiment represent RNA fragments from
transcripts. The number of reads that correspond to specific transcripts is
assumed to be proportional to the copy number of that transcript in the original
sample. The alignment process described in the <a href="high-throughput-sequencing.html#preliminary-hts-data-analysis">Preliminary HTS Data Analysis</a>
section produces alignments of reads against the genome or transcriptome, and
the number of reads that align against each gene or transcript is counted using
a <em>reference annotation</em> that defines the locations of every known gene in the
genome. The resulting counts are therefore estimates of the copy number of the
molecules for each gene or transcript in the original sample. Conceptually, the
more reads that map to a gene, the higher the abundance of transcripts for that
gene, and therefore we infer higher the expression of that gene.</p>
<div class="box important">
<p>Gene expression for different genes can vary by orders of magnitude within a
single cell, where highly expressed genes may have billions of transcripts and
others comparatively few. Therefore, the likelihood of obtaining a read for a
given transcript is proportional to the <em>relative abundance</em> of that transcript
compared with all other transcripts. Relatively rare transcripts have a low
probability of being sampled by the sequencing procedure, and extremely lowly
expressed genes may not be sampled at all. The library size (i.e. the total
number of reads) in the sequencing dataset for a sample determines the range of
gene expression the dataset is likely to detect, where datasets with few reads
are unlikely to sample lowly expressed transcripts even though they are truly
expressed. For this reason <strong>absence of evidence is not evidence of absence</strong> in
RNASeq data. If a read is not observed for any given gene, we cannot say that
the gene is not expressed; we can only say that the relative expression of that
gene is below the detection threshold of the dataset we generated, given the
number of sequenced reads.</p>
</div>
<p>For a single sample, the output of this read counting procedure is a vector of
read counts for each gene in the annotation. Genes with no reads mapping to them
will have a count of zero, and all others will have a read count of one or more;
as mentioned in the <a href="high-throughput-sequencing.html#count-data">Count Data</a> section, read counts are typically non-zero
integers. This procedure is typically repeated separately for a set of samples
that make up the experiment, and the counts for each sample are concatenated
together into a counts matrix. This count matrix is the primary result of
interest that is passed on to downstream analysis, most commonly differential
expression analysis.</p>
</div>
<div id="rnaseq-gene-expression-data" class="section level3" number="8.7.6">
<h3><span class="header-section-number">8.7.6</span> RNASeq Gene Expression Data</h3>
<p>The read counts for each gene or transcript in the genome are estimates of the
relative abundance of molecules in the original sample. The number of genes
counted depends on the organism being studied and the maturity of the gene
annotation used. Complex multicellular organisms have on the order of
thousands to tens of thousands of genes; for example, humans and mice have
20k-30k genes in their respective genomes. Each RNASeq experiment thus may yield
tens of thousands of measurements for each sample.</p>
<p>We shall consider an example RNASeq dataset generated by <span class="citation">(<a href="#ref-OMeara2015-mu" role="doc-biblioref">O’Meara et al. 2015</a>)</span>
generated to study how mouse cardiac cells can regenerate up until one week
of age, after which they lose the ability to do so. The dataset has eight
RNASeq samples, two for each of four time points across the developmental window
until adulthood. We will load in the expression data as a tibble:</p>
<div class="sourceCode" id="cb384"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb384-1"><a href="gene-expression.html#cb384-1" aria-hidden="true" tabindex="-1"></a>counts <span class="ot">&lt;-</span> <span class="fu">read_tsv</span>(<span class="st">&quot;verse_counts.tsv&quot;</span>)</span>
<span id="cb384-2"><a href="gene-expression.html#cb384-2" aria-hidden="true" tabindex="-1"></a>counts</span></code></pre></div>
<pre><code>## # A tibble: 55,416 x 9
##    gene                  P0_1  P0_2  P4_1  P4_2  P7_1  P7_2  Ad_1  Ad_2
##    &lt;chr&gt;                &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;
##  1 ENSMUSG00000102693.2     0     0     0     0     0     0     0     0
##  2 ENSMUSG00000064842.3     0     0     0     0     0     0     0     0
##  3 ENSMUSG00000051951.6    19    24    17    17    17    12     7     5
##  4 ENSMUSG00000102851.2     0     0     0     0     1     0     0     0
##  5 ENSMUSG00000103377.2     1     0     0     1     0     1     1     0
##  6 ENSMUSG00000104017.2     0     3     0     0     0     1     0     0
##  7 ENSMUSG00000103025.2     0     0     0     0     0     0     0     0
##  8 ENSMUSG00000089699.2     0     0     0     0     0     0     0     0
##  9 ENSMUSG00000103201.2     0     0     0     0     0     0     0     1
## 10 ENSMUSG00000103147.2     0     0     0     0     0     0     0     0
## # ... with 55,406 more rows</code></pre>
<p>The annotation counted reads for 55,416 annnotated Ensembl gene for the mouse
genome, as described in the <a href="gene-identifiers.html#gene-identifier-systems">Gene Identifier Systems</a> section.</p>
<p>There are typically three steps when analyzing a RNASeq count matrix:</p>
<ol style="list-style-type: decimal">
<li><strong>Filter genes that are unlikely to be differentially expressed.</strong></li>
<li><strong>Normalize filtered counts to make samples comparable to one another.</strong></li>
<li><strong>Differential expression analysis of filtered, normalized counts.</strong></li>
</ol>
<p>Each of these steps will be described in detail below.</p>
</div>
<div id="filtering-counts" class="section level3" number="8.7.7">
<h3><span class="header-section-number">8.7.7</span> Filtering Counts</h3>
<p>Filtering genes that are not likely to be differentially expressed is an
important step in differential expression analysis. The primary reason for this
is to reduce the number of genes tested for differential expression, thereby
reducing the multiple testing burden on the results. There are many approaches
and rationales for picking filters, and the choice is generally subjective and
highly dependent upon the specific dataset. In this section we discuss some of
the rationale and considerations for different filtering strategies.</p>
<p>Genes that were not detected in any sample are the easiest to filter, and often
eliminate many genes from consideration. Below we filter out all genes that have
zero counts in all samples:</p>
<div class="sourceCode" id="cb386"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb386-1"><a href="gene-expression.html#cb386-1" aria-hidden="true" tabindex="-1"></a>nonzero_genes <span class="ot">&lt;-</span> <span class="fu">rowSums</span>(counts[<span class="sc">-</span><span class="dv">1</span>])<span class="sc">!=</span><span class="dv">0</span></span>
<span id="cb386-2"><a href="gene-expression.html#cb386-2" aria-hidden="true" tabindex="-1"></a>filtered_counts <span class="ot">&lt;-</span> counts[nonzero_genes,]</span>
<span id="cb386-3"><a href="gene-expression.html#cb386-3" aria-hidden="true" tabindex="-1"></a>filtered_counts</span></code></pre></div>
<pre><code>## # A tibble: 32,613 x 9
##    gene                   P0_1  P0_2  P4_1  P4_2  P7_1  P7_2  Ad_1  Ad_2
##    &lt;chr&gt;                 &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;
##  1 ENSMUSG00000051951.6     19    24    17    17    17    12     7     5
##  2 ENSMUSG00000102851.2      0     0     0     0     1     0     0     0
##  3 ENSMUSG00000103377.2      1     0     0     1     0     1     1     0
##  4 ENSMUSG00000104017.2      0     3     0     0     0     1     0     0
##  5 ENSMUSG00000103201.2      0     0     0     0     0     0     0     1
##  6 ENSMUSG00000103161.2      0     0     1     0     0     0     0     0
##  7 ENSMUSG00000102331.2      5     8     2     6     6    11     4     3
##  8 ENSMUSG00000102343.2      0     0     0     1     0     0     0     0
##  9 ENSMUSG00000025900.14     4     3     6     7     7     3    36    35
## 10 ENSMUSG00000102948.2      1     0     0     0     0     0     0     1
## # ... with 32,603 more rows</code></pre>
<p>Only 32,613 genes remain after filtering out genes with all zeros, eliminating
nearly 23,000 genes, or more than 40% of the genes in the annotation. Filtering
genes with zeros in all samples as we have done above is somewhat liberal, as
there may be many genes with reads in only one sample. Genes with only a single
sample with a non-zero count also cannot be differentially expressed, so we can
easily decide to eliminate those as well:</p>
<div class="sourceCode" id="cb388"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb388-1"><a href="gene-expression.html#cb388-1" aria-hidden="true" tabindex="-1"></a>nonzero_genes <span class="ot">&lt;-</span> <span class="fu">rowSums</span>(counts[<span class="sc">-</span><span class="dv">1</span>])<span class="sc">&gt;=</span><span class="dv">2</span></span>
<span id="cb388-2"><a href="gene-expression.html#cb388-2" aria-hidden="true" tabindex="-1"></a>filtered_counts <span class="ot">&lt;-</span> counts[nonzero_genes,]</span>
<span id="cb388-3"><a href="gene-expression.html#cb388-3" aria-hidden="true" tabindex="-1"></a>filtered_counts</span></code></pre></div>
<pre><code>## # A tibble: 28,979 x 9
##    gene                   P0_1  P0_2  P4_1  P4_2  P7_1  P7_2  Ad_1  Ad_2
##    &lt;chr&gt;                 &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;
##  1 ENSMUSG00000051951.6     19    24    17    17    17    12     7     5
##  2 ENSMUSG00000103377.2      1     0     0     1     0     1     1     0
##  3 ENSMUSG00000104017.2      0     3     0     0     0     1     0     0
##  4 ENSMUSG00000102331.2      5     8     2     6     6    11     4     3
##  5 ENSMUSG00000025900.14     4     3     6     7     7     3    36    35
##  6 ENSMUSG00000102948.2      1     0     0     0     0     0     0     1
##  7 ENSMUSG00000025902.14   195   186   204   229   157   173   196   129
##  8 ENSMUSG00000104238.2      0     4     0     2     4     1     3     3
##  9 ENSMUSG00000102269.2      7    16     6     6     4     6     8     3
## 10 ENSMUSG00000118917.1      0     1     0     0     1     1     0     0
## # ... with 28,969 more rows</code></pre>
<p>We have now reduced our number of genes to 28,979 by including only genes with
at least two non-zero counts. We may now wonder how many genes we would include
if we required at least <span class="math inline">\(n\)</span> samples to have non-zero counts. Consider the
following bar chart of the number of genes detected in at least <span class="math inline">\(n\)</span> samples:</p>
<div class="sourceCode" id="cb390"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb390-1"><a href="gene-expression.html#cb390-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(patchwork)</span>
<span id="cb390-2"><a href="gene-expression.html#cb390-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb390-3"><a href="gene-expression.html#cb390-3" aria-hidden="true" tabindex="-1"></a>nonzero_counts <span class="ot">&lt;-</span> <span class="fu">mutate</span>(counts,<span class="st">`</span><span class="at">Number of samples</span><span class="st">`</span><span class="ot">=</span><span class="fu">rowSums</span>(counts[<span class="sc">-</span><span class="dv">1</span>]<span class="sc">!=</span><span class="dv">0</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb390-4"><a href="gene-expression.html#cb390-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(<span class="st">`</span><span class="at">Number of samples</span><span class="st">`</span>) <span class="sc">%&gt;%</span></span>
<span id="cb390-5"><a href="gene-expression.html#cb390-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarize</span>(<span class="st">`</span><span class="at">Number of genes</span><span class="st">`</span><span class="ot">=</span><span class="fu">n</span>() ) <span class="sc">%&gt;%</span></span>
<span id="cb390-6"><a href="gene-expression.html#cb390-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="st">`</span><span class="at">Cumulative number of genes</span><span class="st">`</span><span class="ot">=</span><span class="fu">sum</span>(<span class="st">`</span><span class="at">Number of genes</span><span class="st">`</span>)<span class="sc">-</span><span class="fu">lag</span>(<span class="fu">cumsum</span>(<span class="st">`</span><span class="at">Number of genes</span><span class="st">`</span>),<span class="dv">1</span>,<span class="at">default=</span><span class="dv">0</span>))</span>
<span id="cb390-7"><a href="gene-expression.html#cb390-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb390-8"><a href="gene-expression.html#cb390-8" aria-hidden="true" tabindex="-1"></a>sum_g <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(nonzero_counts) <span class="sc">+</span></span>
<span id="cb390-9"><a href="gene-expression.html#cb390-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_bar</span>(<span class="fu">aes</span>(<span class="at">x=</span><span class="st">`</span><span class="at">Number of samples</span><span class="st">`</span>,<span class="at">y=</span><span class="st">`</span><span class="at">Cumulative number of genes</span><span class="st">`</span>,<span class="at">fill=</span><span class="st">&quot;In at least n&quot;</span>),<span class="at">stat=</span><span class="st">&quot;identity&quot;</span>) <span class="sc">+</span></span>
<span id="cb390-10"><a href="gene-expression.html#cb390-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_bar</span>(<span class="fu">aes</span>(<span class="at">x=</span><span class="st">`</span><span class="at">Number of samples</span><span class="st">`</span>,<span class="at">y=</span><span class="st">`</span><span class="at">Number of genes</span><span class="st">`</span>,<span class="at">fill=</span><span class="st">&quot;In exactly n&quot;</span>),<span class="at">stat=</span><span class="st">&quot;identity&quot;</span>) <span class="sc">+</span></span>
<span id="cb390-11"><a href="gene-expression.html#cb390-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title=</span><span class="st">&quot;Number of genes detected in n samples&quot;</span>) <span class="sc">+</span></span>
<span id="cb390-12"><a href="gene-expression.html#cb390-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_discrete</span>(<span class="at">name=</span><span class="st">&quot;Nonzero in:&quot;</span>)</span>
<span id="cb390-13"><a href="gene-expression.html#cb390-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb390-14"><a href="gene-expression.html#cb390-14" aria-hidden="true" tabindex="-1"></a>sum_g</span></code></pre></div>
<p><img src="main_files/rnaseq%20nonzero%20samples-1.png" width="672" /></p>
<p>The largest number of genes is detected in none of the samples, followed by
genes detected in all samples, with many fewer in between. Based on the plot,
there is no obvious choice for how many nonzero samples are sufficient to use as
a filter.</p>
<p>We may gain additional insight into this problem by considering the distribution
of counts for genes present in at least <span class="math inline">\(n\)</span> samples as above, rather than just
the number of nonzero counts. Below is a plot similar to that above but, which
plots the distribution of mean counts across samples within each number of
nonzero samples:</p>
<div class="sourceCode" id="cb391"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb391-1"><a href="gene-expression.html#cb391-1" aria-hidden="true" tabindex="-1"></a><span class="co"># calculate the mean count for non-zero counts in genes with exactly n non-zero counts</span></span>
<span id="cb391-2"><a href="gene-expression.html#cb391-2" aria-hidden="true" tabindex="-1"></a><span class="fu">mutate</span>(counts,<span class="st">`</span><span class="at">Number of nonzero samples</span><span class="st">`</span><span class="ot">=</span><span class="fu">factor</span>(<span class="fu">rowSums</span>(counts[<span class="sc">-</span><span class="dv">1</span>]<span class="sc">!=</span><span class="dv">0</span>))) <span class="sc">%&gt;%</span></span>
<span id="cb391-3"><a href="gene-expression.html#cb391-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(<span class="sc">-</span><span class="fu">c</span>(gene,<span class="st">`</span><span class="at">Number of nonzero samples</span><span class="st">`</span>),<span class="at">names_to=</span><span class="st">&quot;Sample&quot;</span>,<span class="at">values_to=</span><span class="st">&quot;Count&quot;</span>) <span class="sc">%&gt;%</span></span>
<span id="cb391-4"><a href="gene-expression.html#cb391-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(<span class="st">`</span><span class="at">Number of nonzero samples</span><span class="st">`</span>) <span class="sc">%&gt;%</span></span>
<span id="cb391-5"><a href="gene-expression.html#cb391-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(gene, <span class="at">.add=</span><span class="cn">TRUE</span>) <span class="sc">%&gt;%</span> <span class="fu">summarize</span>(<span class="st">`</span><span class="at">Nonzero Count Mean</span><span class="st">`</span><span class="ot">=</span><span class="fu">mean</span>(Count[Count<span class="sc">!=</span><span class="dv">0</span>]),<span class="at">.groups=</span><span class="st">&quot;keep&quot;</span>) <span class="sc">%&gt;%</span></span>
<span id="cb391-6"><a href="gene-expression.html#cb391-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ungroup</span>() <span class="sc">%&gt;%</span></span>
<span id="cb391-7"><a href="gene-expression.html#cb391-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x=</span><span class="st">`</span><span class="at">Number of nonzero samples</span><span class="st">`</span>,<span class="at">y=</span><span class="st">`</span><span class="at">Nonzero Count Mean</span><span class="st">`</span>,<span class="at">fill=</span><span class="st">`</span><span class="at">Number of nonzero samples</span><span class="st">`</span>)) <span class="sc">+</span></span>
<span id="cb391-8"><a href="gene-expression.html#cb391-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_violin</span>(<span class="at">scale=</span><span class="st">&quot;width&quot;</span>) <span class="sc">+</span></span>
<span id="cb391-9"><a href="gene-expression.html#cb391-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_y_log10</span>() <span class="sc">+</span></span>
<span id="cb391-10"><a href="gene-expression.html#cb391-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title=</span><span class="st">&quot;Nonzero count mean for genes with exactly n nonzero counts&quot;</span>)</span></code></pre></div>
<p><img src="main_files/rnaseq%20nonzero%20dist-1.png" width="672" /></p>
<p>From this plot, we see that the mean count for genes with even a single sample
with a zero is substantially lower than most genes without any zeros. We might
therefore decide to filter out genes that have <em>any</em> zeros, with the rationale
that most genes of appreciable abundance would still be included. However, there
are some genes with a high average count for samples that are not zero that
would be eliminated if we chose to use only genes expressed in all 8 samples.
Again, there is no clear choice of filtering threshold, and a subjective
decision is necessary.</p>
<p>One important property of RNASeq expression data is that genes with higher mean
count also have higher variance. This mean-variance dependence means these count
data are <em>heteroskedastic</em>, a term used to describe data that has inconstant
variance across different values. We can visualize this dependence by
calculating the mean and variance for each of our normalized count genes and
plotting them against each other:</p>
<p><img src="main_files/rnaseq%20mean%20variance-1.png" width="672" /></p>
<p>Unlike in microarray data, where probesets can be filtered based on variance,
using low variance as a filter for count data would result in filtering out
genes with low mean, which would lead to undesirable bias toward highly
expressed genes.</p>
<p>Depending on the experiment, it may be meaningful if a gene is detected
at all in any of the samples. In these cases, genes with only 1 read in a single
sample might be kept, even though these genes would not result in confident
statistical inference if subjected to differential expression. The scientific
question asked by the experiment is an important consideration when deciding
thresholds.</p>
<p>When the goal of the RNASeq experiment is differential expression analysis, a
good rule of thumb when choosing a filtering threshold is to choose genes that
have sufficiently many non-zero counts such that the statistical method has
enough variance to perform reliable inference. For example, a reasonable
filtering threshold might be to include genes that have non-zero counts in at
least 50% of the samples. This threshold depends only upon the number of samples
being considered. In an experimental design with two or more sample groups, a
reasonable strategy might be to require at least 50% of samples <em>within at least
one group</em> or <em>within all groups</em> to be included. Taking the experimental design
into account when filtering can ensure the most interesting genes, e.g. those
expressed only in one condition, are captured.</p>
<p>Regardless of the filtering strategy used, <strong>there is no biological meaning to
any filtering threshold</strong>. The number of reads detected for each gene is
dependent primarily upon the library size, where larger libraries are more
likely to sample lowly abundant genes. Therefore, we cannot filter genes with
the intent of “removing lowly expressed genes”; instead we can only filter out
genes that are below the detection threshold afforded by the library size of the
dataset.</p>
</div>
<div id="count-distributions" class="section level3" number="8.7.8">
<h3><span class="header-section-number">8.7.8</span> Count Distributions</h3>
<p>The range of gene expression in a cell varies by orders of magnitude. The
following histogram plots the distribution of read count values for one of the
adult samples:</p>
<div class="sourceCode" id="cb392"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb392-1"><a href="gene-expression.html#cb392-1" aria-hidden="true" tabindex="-1"></a>dplyr<span class="sc">::</span><span class="fu">select</span>(filtered_counts, Ad_1) <span class="sc">%&gt;%</span></span>
<span id="cb392-2"><a href="gene-expression.html#cb392-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="st">`</span><span class="at">log10(counts+1)</span><span class="st">`</span><span class="ot">=</span><span class="fu">log10</span>(Ad_1<span class="sc">+</span><span class="dv">1</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb392-3"><a href="gene-expression.html#cb392-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x=</span><span class="st">`</span><span class="at">log10(counts+1)</span><span class="st">`</span>)) <span class="sc">+</span></span>
<span id="cb392-4"><a href="gene-expression.html#cb392-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">bins=</span><span class="dv">50</span>) <span class="sc">+</span></span>
<span id="cb392-5"><a href="gene-expression.html#cb392-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title=</span><span class="st">&#39;log10(counts) distribution for Adult mouse&#39;</span>)</span></code></pre></div>
<p><img src="main_files/rnaseq%20sample%20count%20dist-1.png" width="672" /></p>
<p>We add 1 count, called a <em>pseudocount</em>, to avoid taking the log of zero, therefore
all the genes at <span class="math inline">\(x=0\)</span> are genes that have no reads. From the plot, the largest
count is nearly <span class="math inline">\(10^6\)</span> while the smallest is 1, spanning nearly six orders
of magnitude. There are relatively few of these highly abundant genes, however,
and most genes are betwen the range of 100-10,000 reads. We see that this
distribution is similar for all the samples, when displayed as a <a href="visualizing-distributions.html#ridgeline">ridgeline
plot</a>:</p>
<div class="sourceCode" id="cb393"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb393-1"><a href="gene-expression.html#cb393-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggridges)</span>
<span id="cb393-2"><a href="gene-expression.html#cb393-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb393-3"><a href="gene-expression.html#cb393-3" aria-hidden="true" tabindex="-1"></a><span class="fu">pivot_longer</span>(filtered_counts,<span class="sc">-</span><span class="fu">c</span>(gene),<span class="at">names_to=</span><span class="st">&quot;Sample&quot;</span>,<span class="at">values_to=</span><span class="st">&quot;Count&quot;</span>) <span class="sc">%&gt;%</span></span>
<span id="cb393-4"><a href="gene-expression.html#cb393-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="st">`</span><span class="at">log10(Count+1)</span><span class="st">`</span><span class="ot">=</span><span class="fu">log10</span>(Count<span class="sc">+</span><span class="dv">1</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb393-5"><a href="gene-expression.html#cb393-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">y=</span>Sample,<span class="at">x=</span><span class="st">`</span><span class="at">log10(Count+1)</span><span class="st">`</span>,<span class="at">fill=</span>Sample)) <span class="sc">+</span></span>
<span id="cb393-6"><a href="gene-expression.html#cb393-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_density_ridges</span>(<span class="at">bandwidth=</span><span class="fl">0.132</span>) <span class="sc">+</span></span>
<span id="cb393-7"><a href="gene-expression.html#cb393-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title=</span><span class="st">&quot;log10(Count+1) Distributions&quot;</span>)</span></code></pre></div>
<p><img src="main_files/rnaseq%20ridgeline-1.png" width="672" /></p>
<p>The general shape of the distributions is similar, with a large mode of counts
at zero, representing genes that had zero counts in that sample, and a wider
mode between 100 and 10,000 where most consistently expressed genes fall.
However, we note that the peak count of these larger modes vary and also recall
that though the difference appears small, the log scale of the x axis means
the differences in counts may be substantial. We will discuss the implications
of this in the next section.</p>
<div id="count-normalization" class="section level4" number="8.7.8.1">
<h4><span class="header-section-number">8.7.8.1</span> Count Normalization</h4>
<p>The number of reads generated for each sample in an RNASeq dataset varies from
sample to sample due to randomness in the biochemical process that generates
the data. The relationship between the number of reads that maps to any given
gene and the relative abundance of the molecule in the sample is dependent upon
the library size of each sample. The direct number of counts, or the <em>raw
counts</em>, is therefore not directly comparable between samples. A simple example
should suffice to make this clear:</p>
<blockquote>
<p>Imagine two RNASeq datasets generated from the same RNA sample. One of the
samples was sequenced to 1000 reads, and the other to 2000 reads. Now imagine
that one gene makes up half of the RNA in the original sample. In the sample
with 1000 reads, we find this gene has 500 reads. In the sample with 2000 reads,
the gene has 1000 of those reads. In both cases the <em>fraction</em> of reads that map
to the gene is 0.5, but the absolute number of reads differs greatly. This is
why the raw counts for a gene are not directly comparable between samples with
different numbers of reads. Since in general every sample will have a unique
number of reads, these raw counts will never be directly comparable. The way we
mitigate this problem is with a statistical procedure called <em>count
normalization</em>.</p>
</blockquote>
<p>Count normalization is the process by which the number of raw counts in each
sample is scaled by a factor to make multiple samples comparable. Many
strategies have been proposed to do this, but the simplest is a family of
methods that divide by some factor of the library size. A common method in this
family is to compute <em>counts per million reads</em> or <em>CPM</em> normalization, which
scales each gene count by the number of millions of reads in the library:</p>
<p><span class="math display">\[
cpm_{s,i} =  \frac{c_{s,i}}{l_s} * 10^6
\]</span></p>
<p>Here, <span class="math inline">\(c_{s,i}\)</span> is the raw count of gene <span class="math inline">\(i\)</span> in sample <span class="math inline">\(s\)</span>, and <span class="math inline">\(l_s\)</span> is the
number of reads in the dataset for sample <span class="math inline">\(s\)</span>. If each sample is scaled by their
own library size, then in principle the proportion of read counts assigned to
each gene out of all reads for that sample will be comparable across samples.
When we have no additional knowledge of the distribution of genes in our system,
library size normalization methods are the most general and make the fewest
assumptions. Below is a ridgeline plot of the counts distribution after CPM
normalization:</p>
<p><img src="main_files/rnaseq%20ridgeline%20cpm-1.png" width="672" /></p>
<p>The distributions look quite different after CPM normalization, but the modes
now look more consistent than with raw counts.</p>
<p>One drawback of library size normalization is it is sensitive to extreme values;
individual genes with many more reads in one sample than in other samples. These
outliers can lead to pathological effects when comparing gene count values
across samples. However, if we are measuring gene expression in a single
biological system like mouse or human, as is often the case, other methods have
been developed to better use our understanding of the gene expression
distribution to perform more robust and biologically meaningful normalization.
A more robust approach is possible when we make the reasonable assumption that,
for any set of biological conditions in an experiment, <em>most genes are not
differentially expressed</em>. This is the assumption made by the <em>DESeq2
normalization method</em>.</p>
<p>The DESeq2 normalization method is a statistical procedure that uses the median
geometric mean computed across all samples to determine the scale factor for
each sample. The procedure is somewhat complicated and we encourage the reader
to examine it in detail in the DESeq2 publication <span class="citation">(<a href="#ref-Love2014-lb" role="doc-biblioref">Love, Huber, and Anders 2014</a>)</span>. The method has
proven to be consistent and reliable compared with other strategies
<span class="citation">(<a href="#ref-Dillies2013-rq" role="doc-biblioref">Dillies et al. 2013</a>)</span> and is the default normalization for the populare DESeq2
differential expression method, which will be described in a later section. The
DESeq2 normalization method may be implemented using the <a href="https://bioconductor.org/packages/release/bioc/html/DESeq2.html">DESeq2 Bioconductor
package</a>:</p>
<div class="sourceCode" id="cb394"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb394-1"><a href="gene-expression.html#cb394-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(DESeq2)</span>
<span id="cb394-2"><a href="gene-expression.html#cb394-2" aria-hidden="true" tabindex="-1"></a><span class="co"># DESeq2 requires a counts matrix, column data (sample information), and a formula</span></span>
<span id="cb394-3"><a href="gene-expression.html#cb394-3" aria-hidden="true" tabindex="-1"></a><span class="co"># the counts matrix *must be raw counts*</span></span>
<span id="cb394-4"><a href="gene-expression.html#cb394-4" aria-hidden="true" tabindex="-1"></a>count_mat <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(filtered_counts[<span class="sc">-</span><span class="dv">1</span>])</span>
<span id="cb394-5"><a href="gene-expression.html#cb394-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb394-6"><a href="gene-expression.html#cb394-6" aria-hidden="true" tabindex="-1"></a><span class="fu">row.names</span>(count_mat) <span class="ot">&lt;-</span> filtered_counts<span class="sc">$</span>gene</span>
<span id="cb394-7"><a href="gene-expression.html#cb394-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb394-8"><a href="gene-expression.html#cb394-8" aria-hidden="true" tabindex="-1"></a>dds <span class="ot">&lt;-</span> <span class="fu">DESeqDataSetFromMatrix</span>(</span>
<span id="cb394-9"><a href="gene-expression.html#cb394-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">countData=</span>count_mat,</span>
<span id="cb394-10"><a href="gene-expression.html#cb394-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">colData=</span><span class="fu">tibble</span>(<span class="at">sample_name=</span><span class="fu">colnames</span>(filtered_counts[<span class="sc">-</span><span class="dv">1</span>])),</span>
<span id="cb394-11"><a href="gene-expression.html#cb394-11" aria-hidden="true" tabindex="-1"></a>  <span class="at">design=</span><span class="sc">~</span><span class="dv">1</span> <span class="co"># no formula is needed for normalization, ~1 produces a trivial design matrix</span></span>
<span id="cb394-12"><a href="gene-expression.html#cb394-12" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb394-13"><a href="gene-expression.html#cb394-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb394-14"><a href="gene-expression.html#cb394-14" aria-hidden="true" tabindex="-1"></a><span class="co"># compute normalization factors</span></span>
<span id="cb394-15"><a href="gene-expression.html#cb394-15" aria-hidden="true" tabindex="-1"></a>dds <span class="ot">&lt;-</span> <span class="fu">estimateSizeFactors</span>(dds)</span>
<span id="cb394-16"><a href="gene-expression.html#cb394-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb394-17"><a href="gene-expression.html#cb394-17" aria-hidden="true" tabindex="-1"></a><span class="co"># extract the normalized counts</span></span>
<span id="cb394-18"><a href="gene-expression.html#cb394-18" aria-hidden="true" tabindex="-1"></a>deseq_norm_counts <span class="ot">&lt;-</span> <span class="fu">as_tibble</span>(<span class="fu">counts</span>(dds,<span class="at">normalized=</span><span class="cn">TRUE</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb394-19"><a href="gene-expression.html#cb394-19" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">gene=</span>filtered_counts<span class="sc">$</span>gene) <span class="sc">%&gt;%</span></span>
<span id="cb394-20"><a href="gene-expression.html#cb394-20" aria-hidden="true" tabindex="-1"></a>  <span class="fu">relocate</span>(gene)</span></code></pre></div>
<p>Using the DESeq2 package, we first construct a DESeq object using the raw counts
matrix, a sample information tibble with only a sample name column, and a
trivial design formula, which will be explained in more depth in the
<a href="gene-expression.html#deseq2edger">DESeq2/edgeR</a> section. The <code>estimateSizeFactors()</code> function performs the
normalization routine on the counts, and the <code>counts(dds,normalized=TRUE)</code> call
extracts the normalized counts matrix. When plotted as a ridgeline plot as
before, we see that the gene expression distribution modes are better behaved
compared to raw counts, and the shape of the distribution is preserved unlike
with CPM normalized counts:</p>
<div class="sourceCode" id="cb395"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb395-1"><a href="gene-expression.html#cb395-1" aria-hidden="true" tabindex="-1"></a><span class="fu">pivot_longer</span>(deseq_norm_counts,<span class="sc">-</span><span class="fu">c</span>(gene),<span class="at">names_to=</span><span class="st">&quot;Sample&quot;</span>,<span class="at">values_to=</span><span class="st">&quot;Count&quot;</span>) <span class="sc">%&gt;%</span></span>
<span id="cb395-2"><a href="gene-expression.html#cb395-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="st">`</span><span class="at">log10(Count+1)</span><span class="st">`</span><span class="ot">=</span><span class="fu">log10</span>(Count<span class="sc">+</span><span class="dv">1</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb395-3"><a href="gene-expression.html#cb395-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">y=</span>Sample,<span class="at">x=</span><span class="st">`</span><span class="at">log10(Count+1)</span><span class="st">`</span>,<span class="at">fill=</span>Sample)) <span class="sc">+</span></span>
<span id="cb395-4"><a href="gene-expression.html#cb395-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_density_ridges</span>(<span class="at">bandwidth=</span><span class="fl">0.132</span>) <span class="sc">+</span></span>
<span id="cb395-5"><a href="gene-expression.html#cb395-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title=</span><span class="st">&quot;log10(Count+1) Distributions&quot;</span>)</span></code></pre></div>
<p><img src="main_files/rnaseq%20ridgeline%20deseq2-1.png" width="672" /></p>
<p>The DESeq2 normalization procedure has two important considerations to keep in
mind:</p>
<ul>
<li><strong>The procedure borrows information across all samples.</strong> The geometric mean
of counts across all samples is computed as part of the normalization procedure.
This means that the size factors computed <em>depend on the samples in the
dataset</em>. The normalized counts for one sample will likely change if it is
normalized with a different set of samples.</li>
<li><strong>The procedure does not use genes with any zero counts.</strong> The geometric mean
calculation across all samples means that any sample with a zero count results
in the entire geometric mean being zero. For this reason <em>only genes with
nonzero counts in every sample are used to calculate the normalization factors</em>.
If one of the samples has a large number of zeros, for example due to a small
library size, this could dramatically influence the normalization factors for
the entire experiment.</li>
</ul>
<p>The CPM normalization procedure does not borrow information across all samples,
and therefore is not subject to these considerations.</p>
<p>These are only two of many possible normalization methods, DESeq2 being the most
popular at the time of writing. For a detailed explanation of many normalization
methods, see <span class="citation">(<a href="#ref-Dillies2013-rq" role="doc-biblioref">Dillies et al. 2013</a>)</span>.</p>
</div>
<div id="count-transformation" class="section level4" number="8.7.8.2">
<h4><span class="header-section-number">8.7.8.2</span> Count Transformation</h4>
<p>As mentioned in the <a href="high-throughput-sequencing.html#count-data">Count Data</a> section, one way to deal with the non-normality
of count data is to perform a data transformation that makes the counts data
follow a normal distribution. Transformation to normality allows common and
powerful statistical methods like linear regression to be used. The most basic
count transformation is to take the logarithm of the counts. However, this can
be problematic for genes with low counts, causing the values to spread further
apart than genes with high counts, increasing the likelihood of false positive
results. To address this issue, the DESeq2 package provides the <code>rlog()</code>
function that performs a regularized logarithmic transformation that adjusts the
log transform based on the mean count of each gene.</p>
<div class="box note">
<p>The rlog transformation also has the desirable effect of making the count data
homoskedastic, meaning the variance of a gene is not dependent on the mean. Some
statistical methods assume that a dataset has this property. The relationship of
mean count vs variance will be discussed in more detail in the next section.</p>
</div>
<p>We can clearlysee the effect of rlog on the counts distribution by plotting the
rlog’ed counts using a ridgeline plot as before:</p>
<div class="sourceCode" id="cb396"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb396-1"><a href="gene-expression.html#cb396-1" aria-hidden="true" tabindex="-1"></a><span class="co"># the dds is the DESeq2 object from earlier</span></span>
<span id="cb396-2"><a href="gene-expression.html#cb396-2" aria-hidden="true" tabindex="-1"></a>rld <span class="ot">&lt;-</span> <span class="fu">rlog</span>(dds)</span>
<span id="cb396-3"><a href="gene-expression.html#cb396-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb396-4"><a href="gene-expression.html#cb396-4" aria-hidden="true" tabindex="-1"></a><span class="co"># extract rlog values as a tibble</span></span>
<span id="cb396-5"><a href="gene-expression.html#cb396-5" aria-hidden="true" tabindex="-1"></a>rlog_counts <span class="ot">&lt;-</span> <span class="fu">as_tibble</span>(<span class="fu">assay</span>(rld))</span>
<span id="cb396-6"><a href="gene-expression.html#cb396-6" aria-hidden="true" tabindex="-1"></a>rlog_counts<span class="sc">$</span>gene <span class="ot">&lt;-</span> filtered_counts<span class="sc">$</span>gene</span>
<span id="cb396-7"><a href="gene-expression.html#cb396-7" aria-hidden="true" tabindex="-1"></a><span class="fu">pivot_longer</span>(rlog_counts,<span class="sc">-</span><span class="fu">c</span>(gene),<span class="at">names_to=</span><span class="st">&quot;Sample&quot;</span>,<span class="at">values_to=</span><span class="st">&quot;rlog count&quot;</span>) <span class="sc">%&gt;%</span></span>
<span id="cb396-8"><a href="gene-expression.html#cb396-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">y=</span>Sample,<span class="at">x=</span><span class="st">`</span><span class="at">rlog count</span><span class="st">`</span>,<span class="at">fill=</span>Sample)) <span class="sc">+</span></span>
<span id="cb396-9"><a href="gene-expression.html#cb396-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_density_ridges</span>(<span class="at">bandwidth=</span><span class="fl">0.132</span>) <span class="sc">+</span></span>
<span id="cb396-10"><a href="gene-expression.html#cb396-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title=</span><span class="st">&quot;rlog Distributions&quot;</span>)</span></code></pre></div>
<p><img src="main_files/rnaseq%20rlog-1.png" width="672" /></p>
<div class="box warning">
<p>It is generally accepted that count transformations have significant drawbacks
compared with statistical methods that model counts and errors explicitly, like
generalized linear models (GLM) such as Poisson and Negative Binomial
regressions <span class="citation">(<a href="#ref-St-Pierre2018-tc" role="doc-biblioref">St-Pierre, Shikon, and Schneider 2018</a>)</span>. Whenever possible, these GLM based methods
should be used instead of transformations. However, the transformations may
sometimes be necessary depending on the required analysis, so the rlog function
in DESeq2 may be useful for these cases.</p>
</div>
<div class="box readmore">
<ul>
<li><a href="https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#count-data-transformations">DESeq2 package vignette on Count data transformations</a></li>
</ul>
</div>
</div>
</div>
<div id="differential-expression-rnaseq" class="section level3" number="8.7.9">
<h3><span class="header-section-number">8.7.9</span> Differential Expression: RNASeq</h3>
<p>The read counts for each gene or transcript form the primary data used to
identify genes whose expression correlates with some condition of interest. The
counts matrix created by concatenating read counts for all genes across a set of
samples where rows are genes or transcripts and columns are samples is the most
common form of expression data used for this purpose. As mentioned in the
<a href="high-throughput-sequencing.html#count-data">Count Data</a> section, these counts are not normally distributed and therefore
require a statistical approach that can accommodate the counts distribution.</p>
<p>The strong relationship between the mean and variance is characteristic of
RNASeq expression data, and motivates the use of a generalized linear model
called <em>negative binomial regression</em>. Negative binomial regression models count
data using the <a href="https://en.wikipedia.org/wiki/Negative_binomial_distribution">negative binomial distribution</a>,
which allows the relationship of mean and variance of a counts distribution to
vary. The statistical details of the negative binomial distribution and negative
binomial regression are beyond the scope of this book, but below we will discuss
the aspects necessary to implement differential expression analysis.</p>
<div id="deseq2edger" class="section level4" number="8.7.9.1">
<h4><span class="header-section-number">8.7.9.1</span> DESeq2/EdgeR</h4>
<p><a href="https://bioconductor.org/packages/release/bioc/html/DESeq2.html">DESeq2</a> and
<a href="https://bioconductor.org/packages/release/bioc/html/edgeR.html">edgeR</a> are bioconductor packages that both implement negative binomial
regression to perform differential expression on RNASeq data. Both perform
raw counts normalization as part of their function, though they differ in
the normalization method (DESeq2 method is described above, while edgeR uses
trimmed mean of M-values (TMM), see <span class="citation">(<a href="#ref-Robinson2010-ik" role="doc-biblioref">Robinson and Oshlack 2010</a>)</span>). The interface for using
both packages is also similar, so we will focus on DESeq2 in this section.</p>
<p>As mentioned briefly above, DESeq2 requires three pieces of information to
perform differential expression:</p>
<ol style="list-style-type: decimal">
<li>A <strong>raw</strong> counts matrix with genes as rows and samples as columns</li>
<li>A data frame with metadata associated with the columns</li>
<li>A design formula for the differential expression model</li>
</ol>
<p>DESeq2 requires raw counts due to its normalization procedure, which borrows
information across samples when computing size factors, as described above in
the <a href="gene-expression.html#count-normalization">Count Normalization</a> section.</p>
<p>With these three pieces of information, we construct a <code>DESeq</code> object:</p>
<div class="sourceCode" id="cb397"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb397-1"><a href="gene-expression.html#cb397-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(DESeq2)</span>
<span id="cb397-2"><a href="gene-expression.html#cb397-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb397-3"><a href="gene-expression.html#cb397-3" aria-hidden="true" tabindex="-1"></a><span class="co"># filter counts to retain genes with at least 6 out of 8 samples with nonzero counts</span></span>
<span id="cb397-4"><a href="gene-expression.html#cb397-4" aria-hidden="true" tabindex="-1"></a>filtered_counts <span class="ot">&lt;-</span> counts[<span class="fu">rowSums</span>(counts[<span class="sc">-</span><span class="dv">1</span>]<span class="sc">!=</span><span class="dv">0</span>)<span class="sc">&gt;=</span><span class="dv">6</span>,]</span>
<span id="cb397-5"><a href="gene-expression.html#cb397-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb397-6"><a href="gene-expression.html#cb397-6" aria-hidden="true" tabindex="-1"></a><span class="co"># DESeq2 requires a counts matrix, column data (sample information), and a formula</span></span>
<span id="cb397-7"><a href="gene-expression.html#cb397-7" aria-hidden="true" tabindex="-1"></a><span class="co"># the counts matrix *must be raw counts*</span></span>
<span id="cb397-8"><a href="gene-expression.html#cb397-8" aria-hidden="true" tabindex="-1"></a>count_mat <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(filtered_counts[<span class="sc">-</span><span class="dv">1</span>])</span>
<span id="cb397-9"><a href="gene-expression.html#cb397-9" aria-hidden="true" tabindex="-1"></a><span class="fu">row.names</span>(count_mat) <span class="ot">&lt;-</span> filtered_counts<span class="sc">$</span>gene</span>
<span id="cb397-10"><a href="gene-expression.html#cb397-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb397-11"><a href="gene-expression.html#cb397-11" aria-hidden="true" tabindex="-1"></a><span class="co"># create a sample matrix from sample names</span></span>
<span id="cb397-12"><a href="gene-expression.html#cb397-12" aria-hidden="true" tabindex="-1"></a>sample_info <span class="ot">&lt;-</span> <span class="fu">tibble</span>(</span>
<span id="cb397-13"><a href="gene-expression.html#cb397-13" aria-hidden="true" tabindex="-1"></a>  <span class="at">sample_name=</span><span class="fu">colnames</span>(filtered_counts[<span class="sc">-</span><span class="dv">1</span>])</span>
<span id="cb397-14"><a href="gene-expression.html#cb397-14" aria-hidden="true" tabindex="-1"></a>) <span class="sc">%&gt;%</span></span>
<span id="cb397-15"><a href="gene-expression.html#cb397-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">separate</span>(sample_name,<span class="fu">c</span>(<span class="st">&quot;timepoint&quot;</span>,<span class="st">&quot;replicate&quot;</span>),<span class="at">sep=</span><span class="st">&quot;_&quot;</span>,<span class="at">remove=</span><span class="cn">FALSE</span>) <span class="sc">%&gt;%</span></span>
<span id="cb397-16"><a href="gene-expression.html#cb397-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(</span>
<span id="cb397-17"><a href="gene-expression.html#cb397-17" aria-hidden="true" tabindex="-1"></a>    <span class="at">timepoint=</span><span class="fu">factor</span>(timepoint,<span class="at">levels=</span><span class="fu">c</span>(<span class="st">&quot;Ad&quot;</span>,<span class="st">&quot;P0&quot;</span>,<span class="st">&quot;P4&quot;</span>,<span class="st">&quot;P7&quot;</span>))</span>
<span id="cb397-18"><a href="gene-expression.html#cb397-18" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb397-19"><a href="gene-expression.html#cb397-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb397-20"><a href="gene-expression.html#cb397-20" aria-hidden="true" tabindex="-1"></a>design <span class="ot">&lt;-</span> <span class="fu">formula</span>(<span class="sc">~</span> timepoint)</span>
<span id="cb397-21"><a href="gene-expression.html#cb397-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb397-22"><a href="gene-expression.html#cb397-22" aria-hidden="true" tabindex="-1"></a>sample_info</span></code></pre></div>
<pre><code>## # A tibble: 8 x 3
##   sample_name timepoint replicate
##   &lt;chr&gt;       &lt;fct&gt;     &lt;chr&gt;    
## 1 P0_1        P0        1        
## 2 P0_2        P0        2        
## 3 P4_1        P4        1        
## 4 P4_2        P4        2        
## 5 P7_1        P7        1        
## 6 P7_2        P7        2        
## 7 Ad_1        Ad        1        
## 8 Ad_2        Ad        2</code></pre>
<p>The design formula <code>~ timepoint</code> is similar to that supplied to the
<code>model.matrix()</code> function as described in the <a href="gene-expression.html#differential-expression-analysis">Differential Expression Analysis</a>
section, where DESeq2 implicitly creates the model matrix by combining the
column data with the formula. In this differential expression analysis, we wish
to identify genes that are different between the timepoints <code>P0</code>, <code>P4</code>, <code>P7</code>,
and <code>Ad</code>. With these three pieces of information prepared, we can create a DESeq
object and perform differential expression analysis:</p>
<div class="sourceCode" id="cb399"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb399-1"><a href="gene-expression.html#cb399-1" aria-hidden="true" tabindex="-1"></a>dds <span class="ot">&lt;-</span> <span class="fu">DESeqDataSetFromMatrix</span>(</span>
<span id="cb399-2"><a href="gene-expression.html#cb399-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">countData=</span>count_mat,</span>
<span id="cb399-3"><a href="gene-expression.html#cb399-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">colData=</span>sample_info,</span>
<span id="cb399-4"><a href="gene-expression.html#cb399-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">design=</span>design</span>
<span id="cb399-5"><a href="gene-expression.html#cb399-5" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb399-6"><a href="gene-expression.html#cb399-6" aria-hidden="true" tabindex="-1"></a>dds <span class="ot">&lt;-</span> <span class="fu">DESeq</span>(dds)</span>
<span id="cb399-7"><a href="gene-expression.html#cb399-7" aria-hidden="true" tabindex="-1"></a><span class="fu">resultsNames</span>(dds)</span></code></pre></div>
<pre><code>## [1] &quot;Intercept&quot;          &quot;timepoint_P0_vs_Ad&quot; &quot;timepoint_P4_vs_Ad&quot;
## [4] &quot;timepoint_P7_vs_Ad&quot;</code></pre>
<p>DESeq2 performed differential expression on each gene and estimated parameters
for three different comparisons - P0 vs Ad, P4 vs Ad, and P7 vs Ad. Here, Ad is
the <em>reference group</em> so that we may interpret the statistics as relative to the
adult animals. The choice of reference group does not change the magnitude of
the results, but may change the sign of the estimated coefficients, which are
log2 fold changes. To extract out the differential expression statistics for the
P0 vs Ad comparison, we can use the <code>results()</code> function:</p>
<div class="sourceCode" id="cb401"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb401-1"><a href="gene-expression.html#cb401-1" aria-hidden="true" tabindex="-1"></a>res <span class="ot">&lt;-</span> <span class="fu">results</span>(dds, <span class="at">name=</span><span class="st">&quot;timepoint_P0_vs_Ad&quot;</span>)</span>
<span id="cb401-2"><a href="gene-expression.html#cb401-2" aria-hidden="true" tabindex="-1"></a>p0_vs_Ad_de <span class="ot">&lt;-</span> <span class="fu">as_tibble</span>(res) <span class="sc">%&gt;%</span></span>
<span id="cb401-3"><a href="gene-expression.html#cb401-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">gene=</span><span class="fu">rownames</span>(res)) <span class="sc">%&gt;%</span></span>
<span id="cb401-4"><a href="gene-expression.html#cb401-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">relocate</span>(gene) <span class="sc">%&gt;%</span></span>
<span id="cb401-5"><a href="gene-expression.html#cb401-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(pvalue)</span>
<span id="cb401-6"><a href="gene-expression.html#cb401-6" aria-hidden="true" tabindex="-1"></a>p0_vs_Ad_de</span></code></pre></div>
<pre><code>## # A tibble: 21,213 x 7
##    gene                  baseMean log2FoldChange lfcSE  stat    pvalue      padj
##    &lt;chr&gt;                    &lt;dbl&gt;          &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;
##  1 ENSMUSG00000000031.17   19624.           6.75 0.190  35.6 5.70e-278 1.07e-273
##  2 ENSMUSG00000026418.17    9671.           9.84 0.285  34.5 2.29e-261 2.15e-257
##  3 ENSMUSG00000051855.16    3462.           5.35 0.157  34.1 2.84e-255 1.77e-251
##  4 ENSMUSG00000027750.17    9482.           4.05 0.125  32.4 2.26e-230 1.06e-226
##  5 ENSMUSG00000042828.13    3906.          -5.32 0.167 -31.8 1.67e-221 6.25e-218
##  6 ENSMUSG00000046598.16    1154.          -6.08 0.197 -30.8 4.77e-208 1.49e-204
##  7 ENSMUSG00000019817.19    1952.           5.15 0.171  30.1 1.43e-198 3.83e-195
##  8 ENSMUSG00000021622.4    17146.          -4.57 0.157 -29.0 4.41e-185 1.03e-181
##  9 ENSMUSG00000002500.16    2121.          -6.20 0.217 -28.5 6.69e-179 1.39e-175
## 10 ENSMUSG00000024598.10    1885.           5.90 0.209  28.2 5.56e-175 1.04e-171
## # ... with 21,203 more rows</code></pre>
<p>These are the results of the differential expression comparing P0 with Ad; the
other contrasts (P4 vs Ad and P7 vs Ad) may be extracted by specifying the
appropriate result names to the <code>name</code> argument to the <code>results()</code> function. The columns in the DESeq2 results are defined as follows:</p>
<ul>
<li><em>baseMean</em> - the mean normalized count of all samples for the gene</li>
<li><em>log2FoldChange</em> - the estimated coefficient (i.e. log2FoldChange) for the comparison of interest</li>
<li><em>lfcSE</em> - the standard error for the log2FoldChange estimate</li>
<li><em>stat</em> - the <a href="https://en.wikipedia.org/wiki/Wald_test">Wald statistic</a> associated with the log2FoldChange estimate</li>
<li><em>pvalue</em> - the nominal p-value of the Wald test (i.e. the signifiance of the association)</li>
<li><em>padj</em> - the multiple testing adjusted p-value (i.e. false discovery rate)</li>
</ul>
<p>In the P0 vs Ad comparison, we can check the number of significant results easily:</p>
<div class="sourceCode" id="cb403"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb403-1"><a href="gene-expression.html#cb403-1" aria-hidden="true" tabindex="-1"></a><span class="co"># number of genes significant at FDR &lt; 0.05</span></span>
<span id="cb403-2"><a href="gene-expression.html#cb403-2" aria-hidden="true" tabindex="-1"></a><span class="fu">filter</span>(p0_vs_Ad_de,padj<span class="sc">&lt;</span><span class="fl">0.05</span>)</span></code></pre></div>
<pre><code>## # A tibble: 7,557 x 7
##    gene                  baseMean log2FoldChange lfcSE  stat    pvalue      padj
##    &lt;chr&gt;                    &lt;dbl&gt;          &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;
##  1 ENSMUSG00000000031.17   19624.           6.75 0.190  35.6 5.70e-278 1.07e-273
##  2 ENSMUSG00000026418.17    9671.           9.84 0.285  34.5 2.29e-261 2.15e-257
##  3 ENSMUSG00000051855.16    3462.           5.35 0.157  34.1 2.84e-255 1.77e-251
##  4 ENSMUSG00000027750.17    9482.           4.05 0.125  32.4 2.26e-230 1.06e-226
##  5 ENSMUSG00000042828.13    3906.          -5.32 0.167 -31.8 1.67e-221 6.25e-218
##  6 ENSMUSG00000046598.16    1154.          -6.08 0.197 -30.8 4.77e-208 1.49e-204
##  7 ENSMUSG00000019817.19    1952.           5.15 0.171  30.1 1.43e-198 3.83e-195
##  8 ENSMUSG00000021622.4    17146.          -4.57 0.157 -29.0 4.41e-185 1.03e-181
##  9 ENSMUSG00000002500.16    2121.          -6.20 0.217 -28.5 6.69e-179 1.39e-175
## 10 ENSMUSG00000024598.10    1885.           5.90 0.209  28.2 5.56e-175 1.04e-171
## # ... with 7,547 more rows</code></pre>
<p>There were 7,557 significantly differentially expressed genes associated with
this comparison, which includes both up- and down-regulated genes (examine the
signs of the log2FoldChange column). A common diagnostic plot of differenial
expression results is the so-called <em>volcano plot</em> which plots the
log2FoldChange on the x-axis against -log10 adjusted p-value on the y-axis:</p>
<div class="sourceCode" id="cb405"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb405-1"><a href="gene-expression.html#cb405-1" aria-hidden="true" tabindex="-1"></a><span class="fu">mutate</span>(</span>
<span id="cb405-2"><a href="gene-expression.html#cb405-2" aria-hidden="true" tabindex="-1"></a>  p0_vs_Ad_de,</span>
<span id="cb405-3"><a href="gene-expression.html#cb405-3" aria-hidden="true" tabindex="-1"></a>  <span class="st">`</span><span class="at">-log10(adjusted p)</span><span class="st">`</span><span class="ot">=</span><span class="sc">-</span><span class="fu">log10</span>(padj),</span>
<span id="cb405-4"><a href="gene-expression.html#cb405-4" aria-hidden="true" tabindex="-1"></a>  <span class="st">`</span><span class="at">FDR&lt;0.05</span><span class="st">`</span><span class="ot">=</span>padj<span class="sc">&lt;</span><span class="fl">0.05</span></span>
<span id="cb405-5"><a href="gene-expression.html#cb405-5" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">%&gt;%</span></span>
<span id="cb405-6"><a href="gene-expression.html#cb405-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x=</span>log2FoldChange,<span class="at">y=</span><span class="st">`</span><span class="at">-log10(adjusted p)</span><span class="st">`</span>,<span class="at">color=</span><span class="st">`</span><span class="at">FDR&lt;0.05</span><span class="st">`</span>)) <span class="sc">+</span></span>
<span id="cb405-7"><a href="gene-expression.html#cb405-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>()</span></code></pre></div>
<p><img src="main_files/rnaseq%20deseq2%20volcano-1.png" width="672" /></p>
<p>The plot shows that there are significant genes at FDR &lt; 0.05 that are increased
and decreased in the comparison.</p>
<p>DESeq2 has many capabilities beyond this basic differential expression
functionality as described in its <a href="http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html">comprehensive documentation in its vignette</a>.</p>
<p>As mentioned, edgeR implements a very similar model to DESeq2 and also has
<a href="https://bioconductor.org/packages/release/bioc/html/edgeR.html">excellent
documentation</a>.</p>
</div>
<div id="limmavoom" class="section level4" number="8.7.9.2">
<h4><span class="header-section-number">8.7.9.2</span> limma/voom</h4>
<p>While the limma package was initially developed for microarray analysis, the
authors added support for RNASeq soon after the HTS technique became available.
limma accommodates count data by first performing a <a href="gene-expression.html#count-transformation">Count Transformation</a> using
its own
<a href="https://genomebiology.biomedcentral.com/articles/10.1186/gb-2014-15-2-r29">voom</a>
transformation procedure. voom transforms counts by first performing a counts
per million normalization, taking the logarithm of the CPM values, and finally
estimates the mean-variance relationship across genes for use in its empirical
Bayes statistical framework. The counts data being thus transformed, limma
performs statistical inference using its usual linear model framework to produce
differential expression results with arbitrarily complex statistical models.</p>
<p>This brief example is from the <a href="https://www.bioconductor.org/packages/devel/bioc/vignettes/limma/inst/doc/">limma User
Guide</a>
chapter 15, and covers loading
and processing data from an RNA-seq experiment. We go into more depth while
working with limma in assignment 6.</p>
<p>Without going into too much detail, the <code>design</code> variable is how we inform limma
of our experimental conditions, and where limma draws from to construct its
linear models correctly. This design is relatively simple, just four samples
belonging to two different conditions (the <code>swirls</code> here refer to the swirl of
zebra fish, you can just see them as a phenotypic difference)</p>
<div class="sourceCode" id="cb406"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb406-1"><a href="gene-expression.html#cb406-1" aria-hidden="true" tabindex="-1"></a>design <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">swirl =</span> <span class="fu">c</span>(<span class="st">&quot;swirl.1&quot;</span>, <span class="st">&quot;swirl.2&quot;</span>, <span class="st">&quot;swirl.3&quot;</span>, <span class="st">&quot;swirl.4&quot;</span>),</span>
<span id="cb406-2"><a href="gene-expression.html#cb406-2" aria-hidden="true" tabindex="-1"></a>                     <span class="at">condition =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="sc">-</span><span class="dv">1</span>, <span class="dv">1</span>, <span class="sc">-</span><span class="dv">1</span>))</span>
<span id="cb406-3"><a href="gene-expression.html#cb406-3" aria-hidden="true" tabindex="-1"></a>dge <span class="ot">&lt;-</span> <span class="fu">DGEList</span>(<span class="at">counts=</span>counts)</span>
<span id="cb406-4"><a href="gene-expression.html#cb406-4" aria-hidden="true" tabindex="-1"></a>keep <span class="ot">&lt;-</span> <span class="fu">filterByExpr</span>(dge, design)</span>
<span id="cb406-5"><a href="gene-expression.html#cb406-5" aria-hidden="true" tabindex="-1"></a>dge <span class="ot">&lt;-</span> dge[keep,,keep.lib.sizes<span class="ot">=</span><span class="cn">FALSE</span>]</span></code></pre></div>
<p><code>DGEList()</code> is a function from edgeR, of which limma borrows some loading and
filtering functions. This experiment filters by expression level, and uses
square bracket notation (<code>[]</code>) to reduce the number of rows.</p>
<p>Finally, the expression data is transformed into CPM, counts per million, and
a linear model is applied to the data with <code>lmFit()</code>. <code>topTable()</code> is used to view the most
differentially expressed data.</p>
<div class="sourceCode" id="cb407"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb407-1"><a href="gene-expression.html#cb407-1" aria-hidden="true" tabindex="-1"></a><span class="co"># limma trend</span></span>
<span id="cb407-2"><a href="gene-expression.html#cb407-2" aria-hidden="true" tabindex="-1"></a>logCPM <span class="ot">&lt;-</span> <span class="fu">cpm</span>(dge, <span class="at">log=</span><span class="cn">TRUE</span>, <span class="at">prior.count=</span><span class="dv">3</span>)</span>
<span id="cb407-3"><a href="gene-expression.html#cb407-3" aria-hidden="true" tabindex="-1"></a>fit <span class="ot">&lt;-</span> <span class="fu">lmFit</span>(logCPM, design)</span>
<span id="cb407-4"><a href="gene-expression.html#cb407-4" aria-hidden="true" tabindex="-1"></a>fit <span class="ot">&lt;-</span> <span class="fu">eBayes</span>(fit, <span class="at">trend=</span><span class="cn">TRUE</span>)</span>
<span id="cb407-5"><a href="gene-expression.html#cb407-5" aria-hidden="true" tabindex="-1"></a><span class="fu">topTable</span>(fit, <span class="at">coef=</span><span class="fu">ncol</span>(design))</span></code></pre></div>

</div>
</div>
</div>
<h3>References</h3>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-Dillies2013-rq" class="csl-entry">
Dillies, Marie-Agnès, Andrea Rau, Julie Aubert, Christelle Hennequet-Antier, Marine Jeanmougin, Nicolas Servant, Céline Keime, et al. 2013. <span>“A Comprehensive Evaluation of Normalization Methods for Illumina High-Throughput <span>RNA</span> Sequencing Data Analysis.”</span> <em>Brief. Bioinform.</em> 14 (6): 671–83.
</div>
<div id="ref-Irizarry2003-hn" class="csl-entry">
Irizarry, Rafael A, Bridget Hobbs, Francois Collin, Yasmin D Beazer-Barclay, Kristen J Antonellis, Uwe Scherf, and Terence P Speed. 2003. <span>“Exploration, Normalization, and Summaries of High Density Oligonucleotide Array Probe Level Data.”</span> <em>Biostatistics</em> 4 (2): 249–64.
</div>
<div id="ref-Love2014-lb" class="csl-entry">
Love, Michael I, Wolfgang Huber, and Simon Anders. 2014. <span>“Moderated Estimation of Fold Change and Dispersion for <span class="nocase">RNA-seq</span> Data with <span>DESeq2</span>.”</span> <em>Genome Biol.</em> 15 (12): 550.
</div>
<div id="ref-OMeara2015-mu" class="csl-entry">
O’Meara, Caitlin C, Joseph A Wamstad, Rachel A Gladstone, Gregory M Fomovsky, Vincent L Butty, Avanti Shrikumar, Joseph B Gannon, Laurie A Boyer, and Richard T Lee. 2015. <span>“Transcriptional Reversion of Cardiac Myocyte Fate During Mammalian Cardiac Regeneration.”</span> <em>Circ. Res.</em> 116 (5): 804–15.
</div>
<div id="ref-Ritchie2015-bj" class="csl-entry">
Ritchie, Matthew E, Belinda Phipson, Di Wu, Yifang Hu, Charity W Law, Wei Shi, and Gordon K Smyth. 2015. <span>“Limma Powers Differential Expression Analyses for <span class="nocase">RNA-sequencing</span> and Microarray Studies.”</span> <em>Nucleic Acids Res.</em>, January, gkv007.
</div>
<div id="ref-Robinson2010-ik" class="csl-entry">
Robinson, Mark D, and Alicia Oshlack. 2010. <span>“A Scaling Normalization Method for Differential Expression Analysis of <span class="nocase">RNA-seq</span> Data.”</span> <em>Genome Biol.</em> 11 (3): R25.
</div>
<div id="ref-St-Pierre2018-tc" class="csl-entry">
St-Pierre, Anne P, Violaine Shikon, and David C Schneider. 2018. <span>“Count Data in Biology-Data Transformation or Model Reformation?”</span> <em>Ecol. Evol.</em> 8 (6): 3077–85.
</div>
<div id="ref-Zhao2018-ao" class="csl-entry">
Zhao, Shanrong, Ying Zhang, Ramya Gamini, Baohong Zhang, and David von Schack. 2018. <span>“Evaluation of Two Main <span class="nocase">RNA-seq</span> Approaches for Gene Quantification in Clinical <span>RNA</span> Sequencing: polyA+ Selection Versus <span class="nocase">rRNA</span> Depletion.”</span> <em>Sci. Rep.</em> 8 (1): 4781.
</div>
</div>
<p style="text-align: center;">
<a href="gene-identifiers.html"><button class="btn btn-default">Previous</button></a>
<a href="gene-set-enrichment-analysis.html"><button class="btn btn-default">Next</button></a>
</p>
</div>
</div>


</div>

<script>

// add bootstrap table styles to pandoc tables
$(document).ready(function () {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
});

</script>

</body>
</html>
