<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta name="viewport" content="width=device-width, initial-scale=1" />
<meta property="og:title" content="8.6 Gene Identifiers | BF591 - R for Biological Sciences" />
<meta property="og:type" content="book" />







<script type="text/x-mathjax-config">
MathJax.Hub.Config({
  TeX: { equationNumbers: { autoNumber: "AMS" } }
});
</script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<meta name="description" content="8.6 Gene Identifiers | BF591 - R for Biological Sciences">

<title>8.6 Gene Identifiers | BF591 - R for Biological Sciences</title>

<script src="libs/header-attrs-2.13/header-attrs.js"></script>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="libs/bootstrap-3.3.5/css/darkly.min.css" rel="stylesheet" />
<script src="libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="libs/navigation-1.1/tabsets.js"></script>


<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>


<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />

<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
/* show arrow before summary tag as in bootstrap
TODO: remove if boostrap in updated in html_document (rmarkdown#1485) */
details > summary {
  display: list-item;
  cursor: pointer;
}
</style>
</head>

<body>

<div class="container-fluid main-container">


<div class="row">
<div class="col-sm-12">
<div id="TOC">
<ul>
<li><a href="index.html#syllabus">Syllabus</a>
<ul>
<li><a href="course-schedule.html#course-schedule">Course Schedule</a></li>
<li><a href="instructors.html#instructors">Instructors</a></li>
<li><a href="office-hours.html#office-hours">Office Hours</a></li>
<li><a href="course-values-and-policies.html#course-values-and-policies">Course Values and Policies</a></li>
</ul></li>
<li><a href="introduction.html#introduction"><span class="toc-section-number">1</span> Introduction</a>
<ul>
<li><a href="who-this-book-is-for.html#who-this-book-is-for"><span class="toc-section-number">1.1</span> Who This Book Is For</a></li>
<li><a href="a-note-about-reinventing-the-wheel.html#a-note-about-reinventing-the-wheel"><span class="toc-section-number">1.2</span> A Note About Reinventing the Wheel</a></li>
<li><a href="sources-and-references.html#sources-and-references"><span class="toc-section-number">1.3</span> Sources and References</a></li>
</ul></li>
<li><a href="data-bio.html#data-bio"><span class="toc-section-number">2</span> Data in Biology</a>
<ul>
<li><a href="a-brief-history-of-data-in-molecular-biology.html#a-brief-history-of-data-in-molecular-biology"><span class="toc-section-number">2.1</span> A Brief History of Data in Molecular Biology</a></li>
<li><a href="biology-as-a-mature-data-science.html#biology-as-a-mature-data-science"><span class="toc-section-number">2.2</span> Biology as a Mature Data Science</a></li>
</ul></li>
<li><a href="preliminaries.html#preliminaries"><span class="toc-section-number">3</span> Preliminaries</a>
<ul>
<li><a href="prelim-r.html#prelim-r"><span class="toc-section-number">3.1</span> The R Language</a></li>
<li><a href="prelim-rstudio.html#prelim-rstudio"><span class="toc-section-number">3.2</span> RStudio</a></li>
<li><a href="prog-the-r-script.html#prog-the-r-script"><span class="toc-section-number">3.3</span> The R Script</a></li>
<li><a href="prog-workflow.html#prog-workflow"><span class="toc-section-number">3.4</span> The Scripting Workflow</a></li>
<li><a href="prelim-git.html#prelim-git"><span class="toc-section-number">3.5</span> git + github</a>
<ul>
<li><a href="prelim-git.html#motivation"><span class="toc-section-number">3.5.1</span> Motivation</a></li>
<li><a href="prelim-git.html#git"><span class="toc-section-number">3.5.2</span> git</a></li>
<li><a href="prelim-git.html#git-hosting-platforms-github"><span class="toc-section-number">3.5.3</span> Git hosting platforms (GitHub)</a></li>
</ul></li>
</ul></li>
<li><a href="prog-basics.html#prog-basics"><span class="toc-section-number">4</span> R Programming</a>
<ul>
<li><a href="before-you-begin.html#before-you-begin"><span class="toc-section-number">4.1</span> Before you begin</a></li>
<li><a href="introduction-1.html#introduction-1"><span class="toc-section-number">4.2</span> Introduction</a></li>
<li><a href="prog-r-syntax.html#prog-r-syntax"><span class="toc-section-number">4.3</span> R Syntax Basics</a></li>
<li><a href="prog-types.html#prog-types"><span class="toc-section-number">4.4</span> Basic Types of Values</a>
<ul>
<li><a href="prog-types.html#factors"><span class="toc-section-number">4.4.1</span> Factors</a></li>
</ul></li>
<li><a href="prog-struct.html#prog-struct"><span class="toc-section-number">4.5</span> Data Structures</a>
<ul>
<li><a href="prog-struct.html#vectors"><span class="toc-section-number">4.5.1</span> Vectors</a></li>
<li><a href="prog-struct.html#matrices"><span class="toc-section-number">4.5.2</span> Matrices</a></li>
<li><a href="prog-struct.html#lists-and-data-frames"><span class="toc-section-number">4.5.3</span> Lists and data frames</a></li>
</ul></li>
<li><a href="logical-tests-and-comparators.html#logical-tests-and-comparators"><span class="toc-section-number">4.6</span> Logical Tests and Comparators</a></li>
<li><a href="functions.html#functions"><span class="toc-section-number">4.7</span> Functions</a>
<ul>
<li><a href="functions.html#prog-dry"><span class="toc-section-number">4.7.1</span> DRY: Don’t Repeat Yourself</a></li>
<li><a href="functions.html#writing-your-own-functions"><span class="toc-section-number">4.7.2</span> Writing your own functions</a></li>
<li><a href="functions.html#scope"><span class="toc-section-number">4.7.3</span> Scope</a></li>
</ul></li>
<li><a href="iteration.html#iteration"><span class="toc-section-number">4.8</span> Iteration</a>
<ul>
<li><a href="iteration.html#vectorized-operations"><span class="toc-section-number">4.8.1</span> Vectorized operations</a></li>
<li><a href="iteration.html#functional-programming"><span class="toc-section-number">4.8.2</span> Functional programming</a></li>
<li><a href="iteration.html#apply-and-friends"><span class="toc-section-number">4.8.3</span> <code>apply()</code> and friends</a></li>
</ul></li>
<li><a href="installing-packages.html#installing-packages"><span class="toc-section-number">4.9</span> Installing Packages</a></li>
<li><a href="saving-and-loading-r-data.html#saving-and-loading-r-data"><span class="toc-section-number">4.10</span> Saving and Loading R Data</a></li>
<li><a href="prog-debugging.html#prog-debugging"><span class="toc-section-number">4.11</span> Troubleshooting and Debugging</a></li>
<li><a href="prog-style.html#prog-style"><span class="toc-section-number">4.12</span> Coding Style and Conventions</a>
<ul>
<li><a href="prog-style.html#is-my-code-correct"><span class="toc-section-number">4.12.1</span> Is my code correct?</a></li>
<li><a href="prog-style.html#does-my-code-follow-the-dry-principle"><span class="toc-section-number">4.12.2</span> Does my code follow the <span>DRY principle</span>?</a></li>
<li><a href="prog-style.html#did-i-choose-concise-but-descriptive-variable-and-function-names"><span class="toc-section-number">4.12.3</span> Did I choose concise but descriptive variable and function names?</a></li>
<li><a href="prog-style.html#did-i-use-indentation-and-naming-conventions-consistently-throughout-my-code"><span class="toc-section-number">4.12.4</span> Did I use indentation and naming conventions consistently throughout my code?</a></li>
<li><a href="prog-style.html#did-i-write-comments-especially-when-what-the-code-does-is-not-obvious"><span class="toc-section-number">4.12.5</span> Did I write comments, especially when what the code does is not obvious?</a></li>
<li><a href="prog-style.html#how-easy-would-it-be-for-someone-else-to-understand-my-code"><span class="toc-section-number">4.12.6</span> How easy would it be for someone else to understand my code?</a></li>
<li><a href="prog-style.html#is-my-code-easy-to-maintainchange"><span class="toc-section-number">4.12.7</span> Is my code easy to maintain/change?</a></li>
<li><a href="prog-style.html#the-styler-package"><span class="toc-section-number">4.12.8</span> The <code>styler</code> package</a></li>
</ul></li>
</ul></li>
<li><a href="data-wrangling.html#data-wrangling"><span class="toc-section-number">5</span> Data Wrangling</a>
<ul>
<li><a href="the-tidyverse.html#the-tidyverse"><span class="toc-section-number">5.1</span> The Tidyverse</a></li>
<li><a href="dw-basics.html#dw-basics"><span class="toc-section-number">5.2</span> Tidyverse Basics</a></li>
<li><a href="importing-data.html#importing-data"><span class="toc-section-number">5.3</span> Importing Data</a></li>
<li><a href="data-tibble.html#data-tibble"><span class="toc-section-number">5.4</span> The tibble</a></li>
<li><a href="tidy-data.html#tidy-data"><span class="toc-section-number">5.5</span> Tidy Data</a></li>
<li><a href="pipes.html#pipes"><span class="toc-section-number">5.6</span> pipes</a></li>
<li><a href="arranging-data.html#arranging-data"><span class="toc-section-number">5.7</span> Arranging Data</a>
<ul>
<li><a href="arranging-data.html#dplyrmutate---create-new-columns-using-other-columns"><span class="toc-section-number">5.7.1</span> <code>dplyr::mutate()</code> - Create new columns using other columns</a></li>
<li><a href="arranging-data.html#stringr---working-with-character-values"><span class="toc-section-number">5.7.2</span> <code>stringr</code> - Working with character values</a></li>
<li><a href="arranging-data.html#dplyrselect---subset-columns-by-name"><span class="toc-section-number">5.7.3</span> <code>dplyr::select()</code> - Subset Columns by Name</a></li>
<li><a href="arranging-data.html#dplyrfilter---pick-rows-out-of-a-data-set"><span class="toc-section-number">5.7.4</span> <code>dplyr::filter()</code> - Pick rows out of a data set</a></li>
<li><a href="arranging-data.html#dplyrarrange---order-rows-based-on-their-values"><span class="toc-section-number">5.7.5</span> <code>dplyr::arrange()</code> - Order rows based on their values</a></li>
<li><a href="arranging-data.html#putting-it-all-together"><span class="toc-section-number">5.7.6</span> Putting it all together</a></li>
</ul></li>
<li><a href="grouping-data.html#grouping-data"><span class="toc-section-number">5.8</span> Grouping Data</a></li>
<li><a href="rearranging-data.html#rearranging-data"><span class="toc-section-number">5.9</span> Rearranging Data</a></li>
<li><a href="relational-data.html#relational-data"><span class="toc-section-number">5.10</span> Relational Data</a>
<ul>
<li><a href="relational-data.html#dealing-with-multiple-matches"><span class="toc-section-number">5.10.1</span> Dealing with multiple matches</a></li>
</ul></li>
</ul></li>
<li><a href="data-science.html#data-science"><span class="toc-section-number">6</span> Data Science</a>
<ul>
<li><a href="data-modeling.html#data-modeling"><span class="toc-section-number">6.1</span> Data Modeling</a>
<ul>
<li><a href="data-modeling.html#a-worked-modeling-example"><span class="toc-section-number">6.1.1</span> A Worked Modeling Example</a></li>
<li><a href="data-modeling.html#data-summarization"><span class="toc-section-number">6.1.2</span> Data Summarization</a></li>
<li><a href="data-modeling.html#linear-models"><span class="toc-section-number">6.1.3</span> Linear Models</a></li>
<li><a href="data-modeling.html#flavors-of-linear-models"><span class="toc-section-number">6.1.4</span> Flavors of Linear Models</a></li>
<li><a href="data-modeling.html#assessing-model-accuracy-."><span class="toc-section-number">6.1.5</span> Assessing Model Accuracy .</a></li>
</ul></li>
<li><a href="statistical-distributions.html#statistical-distributions"><span class="toc-section-number">6.2</span> Statistical Distributions</a>
<ul>
<li><a href="statistical-distributions.html#random-variables"><span class="toc-section-number">6.2.1</span> Random Variables</a></li>
<li><a href="statistical-distributions.html#statistical-distribution-basics"><span class="toc-section-number">6.2.2</span> Statistical Distribution Basics</a></li>
<li><a href="statistical-distributions.html#distributions-in-r"><span class="toc-section-number">6.2.3</span> Distributions in R</a></li>
<li><a href="statistical-distributions.html#discrete-distributions"><span class="toc-section-number">6.2.4</span> Discrete Distributions</a></li>
<li><a href="statistical-distributions.html#continuous-distributions"><span class="toc-section-number">6.2.5</span> Continuous Distributions</a></li>
<li><a href="statistical-distributions.html#empirical-distributions"><span class="toc-section-number">6.2.6</span> Empirical Distributions</a></li>
</ul></li>
<li><a href="statistical-tests.html#statistical-tests"><span class="toc-section-number">6.3</span> Statistical Tests</a>
<ul>
<li><a href="statistical-tests.html#what-is-a-statistical-test"><span class="toc-section-number">6.3.1</span> What is a statistical test</a></li>
<li><a href="statistical-tests.html#common-statistical-tests"><span class="toc-section-number">6.3.2</span> Common Statistical Tests</a></li>
<li><a href="statistical-tests.html#choosing-a-test"><span class="toc-section-number">6.3.3</span> Choosing a Test</a></li>
<li><a href="statistical-tests.html#p-values"><span class="toc-section-number">6.3.4</span> P-values</a></li>
<li><a href="statistical-tests.html#multiple-hypothesis-testing-."><span class="toc-section-number">6.3.5</span> Multiple Hypothesis Testing .</a></li>
<li><a href="statistical-tests.html#statistical-power"><span class="toc-section-number">6.3.6</span> Statistical power</a></li>
</ul></li>
<li><a href="exploratory-data-analysis.html#exploratory-data-analysis"><span class="toc-section-number">6.4</span> Exploratory Data Analysis</a>
<ul>
<li><a href="exploratory-data-analysis.html#principal-component-analysis"><span class="toc-section-number">6.4.1</span> Principal Component Analysis</a></li>
<li><a href="exploratory-data-analysis.html#cluster-analysis"><span class="toc-section-number">6.4.2</span> Cluster Analysis</a></li>
<li><a href="exploratory-data-analysis.html#k-means-."><span class="toc-section-number">6.4.3</span> k-means .</a></li>
<li><a href="exploratory-data-analysis.html#others"><span class="toc-section-number">6.4.4</span> Others</a></li>
</ul></li>
<li><a href="network-analysis-..html#network-analysis-."><span class="toc-section-number">6.5</span> Network Analysis .</a>
<ul>
<li><a href="network-analysis-..html#basic-network-analysis"><span class="toc-section-number">6.5.1</span> Basic network analysis</a></li>
<li><a href="network-analysis-..html#create-your-own"><span class="toc-section-number">6.5.2</span> Create your own</a></li>
</ul></li>
</ul></li>
<li><a href="data-visualization.html#data-visualization"><span class="toc-section-number">7</span> Data Visualization</a>
<ul>
<li><a href="responsible-plotting.html#responsible-plotting"><span class="toc-section-number">7.1</span> Responsible Plotting</a>
<ul>
<li><a href="responsible-plotting.html#visualization-principles"><span class="toc-section-number">7.1.1</span> Visualization Principles</a></li>
<li><a href="responsible-plotting.html#human-visual-perception"><span class="toc-section-number">7.1.2</span> Human Visual Perception</a></li>
<li><a href="responsible-plotting.html#visual-encodings"><span class="toc-section-number">7.1.3</span> Visual Encodings</a></li>
<li><a href="responsible-plotting.html#some-opinionated-rules-of-thumb"><span class="toc-section-number">7.1.4</span> Some Opinionated Rules of Thumb</a></li>
</ul></li>
<li><a href="dv-gg.html#dv-gg"><span class="toc-section-number">7.2</span> Grammar of Graphics</a>
<ul>
<li><a href="dv-gg.html#ggplot-mechanics"><span class="toc-section-number">7.2.1</span> <code>ggplot</code> mechanics</a></li>
</ul></li>
<li><a href="plotting-one-dimension.html#plotting-one-dimension"><span class="toc-section-number">7.3</span> Plotting One Dimension</a>
<ul>
<li><a href="plotting-one-dimension.html#dv-bar"><span class="toc-section-number">7.3.1</span> Bar chart</a></li>
<li><a href="plotting-one-dimension.html#lollipop-plots"><span class="toc-section-number">7.3.2</span> Lollipop plots</a></li>
<li><a href="plotting-one-dimension.html#stacked-area-charts"><span class="toc-section-number">7.3.3</span> Stacked Area charts</a></li>
<li><a href="plotting-one-dimension.html#parallel-coordinate-plots"><span class="toc-section-number">7.3.4</span> Parallel Coordinate plots</a></li>
</ul></li>
<li><a href="visualizing-distributions.html#visualizing-distributions"><span class="toc-section-number">7.4</span> Visualizing Distributions</a>
<ul>
<li><a href="visualizing-distributions.html#histogram"><span class="toc-section-number">7.4.1</span> Histogram</a></li>
<li><a href="visualizing-distributions.html#density"><span class="toc-section-number">7.4.2</span> Density</a></li>
<li><a href="visualizing-distributions.html#boxplot"><span class="toc-section-number">7.4.3</span> Boxplot</a></li>
<li><a href="visualizing-distributions.html#violin-plot"><span class="toc-section-number">7.4.4</span> Violin plot</a></li>
<li><a href="visualizing-distributions.html#beeswarm-plot"><span class="toc-section-number">7.4.5</span> Beeswarm plot</a></li>
<li><a href="visualizing-distributions.html#ridgeline"><span class="toc-section-number">7.4.6</span> Ridgeline</a></li>
</ul></li>
<li><a href="plotting-two-or-more-dimensions.html#plotting-two-or-more-dimensions"><span class="toc-section-number">7.5</span> Plotting Two or More Dimensions</a>
<ul>
<li><a href="plotting-two-or-more-dimensions.html#scatter-plots"><span class="toc-section-number">7.5.1</span> Scatter Plots</a></li>
<li><a href="plotting-two-or-more-dimensions.html#bubble-plots"><span class="toc-section-number">7.5.2</span> Bubble Plots</a></li>
<li><a href="plotting-two-or-more-dimensions.html#connected-scatter-plots"><span class="toc-section-number">7.5.3</span> Connected Scatter Plots</a></li>
<li><a href="plotting-two-or-more-dimensions.html#line-plots"><span class="toc-section-number">7.5.4</span> Line Plots</a></li>
<li><a href="plotting-two-or-more-dimensions.html#parallel-coordinate-plots-1"><span class="toc-section-number">7.5.5</span> Parallel Coordinate Plots</a></li>
<li><a href="plotting-two-or-more-dimensions.html#heatmaps"><span class="toc-section-number">7.5.6</span> Heatmaps</a></li>
<li><a href="plotting-two-or-more-dimensions.html#how-to-use-heatmaps-responsibly"><span class="toc-section-number">7.5.7</span> How To Use Heatmaps Responsibly</a></li>
</ul></li>
<li><a href="other-kind-of-plots.html#other-kind-of-plots"><span class="toc-section-number">7.6</span> Other Kind of Plots</a>
<ul>
<li><a href="other-kind-of-plots.html#dendrograms"><span class="toc-section-number">7.6.1</span> Dendrograms</a></li>
<li><a href="other-kind-of-plots.html#chord-diagrams-and-circos-plots"><span class="toc-section-number">7.6.2</span> Chord Diagrams and Circos Plots</a></li>
</ul></li>
<li><a href="multiple-plots.html#multiple-plots"><span class="toc-section-number">7.7</span> Multiple Plots</a>
<ul>
<li><a href="multiple-plots.html#facet-wrapping"><span class="toc-section-number">7.7.1</span> Facet wrapping</a></li>
<li><a href="multiple-plots.html#multipanel-figures"><span class="toc-section-number">7.7.2</span> Multipanel Figures</a></li>
</ul></li>
<li><a href="publication-ready-plots.html#publication-ready-plots"><span class="toc-section-number">7.8</span> Publication Ready Plots</a>
<ul>
<li><a href="publication-ready-plots.html#ggplot-themes"><span class="toc-section-number">7.8.1</span> <code>ggplot</code> Themes</a></li>
<li><a href="publication-ready-plots.html#exporting-to-svg"><span class="toc-section-number">7.8.2</span> Exporting to SVG</a></li>
</ul></li>
<li><a href="network-visualization.html#network-visualization"><span class="toc-section-number">7.9</span> Network visualization</a>
<ul>
<li><a href="network-visualization.html#interactive-network-plots"><span class="toc-section-number">7.9.1</span> Interactive network plots</a></li>
</ul></li>
</ul></li>
<li><a href="biology-bioinformatics.html#biology-bioinformatics"><span class="toc-section-number">8</span> Biology &amp; Bioinformatics</a>
<ul>
<li><a href="bio-r.html#bio-r"><span class="toc-section-number">8.1</span> R in Biology</a></li>
<li><a href="biological-data-overview.html#biological-data-overview"><span class="toc-section-number">8.2</span> Biological Data Overview</a>
<ul>
<li><a href="biological-data-overview.html#bio-data-types"><span class="toc-section-number">8.2.1</span> Types of Biological Data</a></li>
<li><a href="biological-data-overview.html#csv-files"><span class="toc-section-number">8.2.2</span> CSV Files</a></li>
<li><a href="biological-data-overview.html#biological-data-is-not-tidy"><span class="toc-section-number">8.2.3</span> Biological data is NOT Tidy!</a></li>
</ul></li>
<li><a href="bio-bioconductor.html#bio-bioconductor"><span class="toc-section-number">8.3</span> Bioconductor</a></li>
<li><a href="microarrays.html#microarrays"><span class="toc-section-number">8.4</span> Microarrays</a></li>
<li><a href="high-throughput-sequencing.html#high-throughput-sequencing"><span class="toc-section-number">8.5</span> High Throughput Sequencing</a>
<ul>
<li><a href="high-throughput-sequencing.html#raw-hts-data"><span class="toc-section-number">8.5.1</span> Raw HTS Data</a></li>
<li><a href="high-throughput-sequencing.html#preliminary-hts-data-analysis"><span class="toc-section-number">8.5.2</span> Preliminary HTS Data Analysis</a></li>
<li><a href="high-throughput-sequencing.html#count-data"><span class="toc-section-number">8.5.3</span> Count Data</a></li>
</ul></li>
<li><a href="gene-identifiers.html#gene-identifiers"><span class="toc-section-number">8.6</span> Gene Identifiers</a>
<ul>
<li><a href="gene-identifiers.html#gene-identifier-systems"><span class="toc-section-number">8.6.1</span> Gene Identifier Systems</a></li>
<li><a href="gene-identifiers.html#mapping-between-identifier-systems"><span class="toc-section-number">8.6.2</span> Mapping Between Identifier Systems</a></li>
<li><a href="gene-identifiers.html#mapping-homologs"><span class="toc-section-number">8.6.3</span> Mapping Homologs</a></li>
</ul></li>
<li><a href="gene-expression.html#gene-expression"><span class="toc-section-number">8.7</span> Gene Expression</a>
<ul>
<li><a href="gene-expression.html#gene-expression-data-in-bioconductor"><span class="toc-section-number">8.7.1</span> Gene Expression Data in Bioconductor</a></li>
<li><a href="gene-expression.html#differential-expression-analysis"><span class="toc-section-number">8.7.2</span> Differential Expression Analysis</a></li>
<li><a href="gene-expression.html#microarray-gene-expression-data"><span class="toc-section-number">8.7.3</span> Microarray Gene Expression Data</a></li>
<li><a href="gene-expression.html#differential-expression-microarrays-limma"><span class="toc-section-number">8.7.4</span> Differential Expression: Microarrays (limma)</a></li>
<li><a href="gene-expression.html#rnaseq"><span class="toc-section-number">8.7.5</span> RNASeq</a></li>
<li><a href="gene-expression.html#rnaseq-gene-expression-data"><span class="toc-section-number">8.7.6</span> RNASeq Gene Expression Data</a></li>
<li><a href="gene-expression.html#filtering-counts"><span class="toc-section-number">8.7.7</span> Filtering Counts</a></li>
<li><a href="gene-expression.html#count-distributions"><span class="toc-section-number">8.7.8</span> Count Distributions</a></li>
<li><a href="gene-expression.html#differential-expression-rnaseq"><span class="toc-section-number">8.7.9</span> Differential Expression: RNASeq</a></li>
</ul></li>
<li><a href="gene-set-enrichment-analysis.html#gene-set-enrichment-analysis"><span class="toc-section-number">8.8</span> Gene Set Enrichment Analysis</a>
<ul>
<li><a href="gene-set-enrichment-analysis.html#gene-sets"><span class="toc-section-number">8.8.1</span> Gene Sets</a></li>
<li><a href="gene-set-enrichment-analysis.html#working-with-gene-sets-in-r"><span class="toc-section-number">8.8.2</span> Working with gene sets in R</a></li>
<li><a href="gene-set-enrichment-analysis.html#over-representation-analysis"><span class="toc-section-number">8.8.3</span> Over-representation Analysis</a></li>
<li><a href="gene-set-enrichment-analysis.html#rank-based-analysis"><span class="toc-section-number">8.8.4</span> Rank-based Analysis</a></li>
<li><a href="gene-set-enrichment-analysis.html#fgsea"><span class="toc-section-number">8.8.5</span> fgsea</a></li>
</ul></li>
<li><a href="biological-networks-..html#biological-networks-."><span class="toc-section-number">8.9</span> Biological Networks .</a>
<ul>
<li><a href="biological-networks-..html#biological-pathways-."><span class="toc-section-number">8.9.1</span> Biological Pathways .</a></li>
<li><a href="biological-networks-..html#gene-regulatory-networks-."><span class="toc-section-number">8.9.2</span> Gene Regulatory Networks .</a></li>
<li><a href="biological-networks-..html#protein-protein-interaction-network-."><span class="toc-section-number">8.9.3</span> Protein-Protein Interaction Network .</a></li>
<li><a href="biological-networks-..html#wgcna-."><span class="toc-section-number">8.9.4</span> WGCNA .</a></li>
</ul></li>
</ul></li>
<li><a href="engineering.html#engineering"><span class="toc-section-number">9</span> EngineeRing</a>
<ul>
<li><a href="unit-testing.html#unit-testing"><span class="toc-section-number">9.1</span> Unit Testing</a></li>
<li><a href="toolification.html#toolification"><span class="toc-section-number">9.2</span> Toolification</a>
<ul>
<li><a href="toolification.html#the-r-interpreter"><span class="toc-section-number">9.2.1</span> The R Interpreter</a></li>
<li><a href="toolification.html#rscript"><span class="toc-section-number">9.2.2</span> <code>Rscript</code></a></li>
<li><a href="toolification.html#commandargs"><span class="toc-section-number">9.2.3</span> <code>commandArgs()</code></a></li>
</ul></li>
<li><a href="parallel-processing.html#parallel-processing"><span class="toc-section-number">9.3</span> Parallel Processing</a>
<ul>
<li><a href="parallel-processing.html#brief-introduction-to-parallelization"><span class="toc-section-number">9.3.1</span> Brief Introduction to Parallelization</a></li>
<li><a href="parallel-processing.html#apply-and-friends-are-pleasingly-parallel"><span class="toc-section-number">9.3.2</span> <code>apply</code> and Friends Are Pleasingly Parallel</a></li>
<li><a href="parallel-processing.html#the-parallel-package"><span class="toc-section-number">9.3.3</span> The <code>parallel</code> package</a></li>
</ul></li>
<li><a href="object-orientation-in-r.html#object-orientation-in-r"><span class="toc-section-number">9.4</span> Object Orientation in R</a></li>
<li><a href="r-packages.html#r-packages"><span class="toc-section-number">9.5</span> R Packages</a></li>
</ul></li>
<li><a href="rshiny.html#rshiny"><span class="toc-section-number">10</span> RShiny</a>
<ul>
<li><a href="rshiny-introduction.html#rshiny-introduction"><span class="toc-section-number">10.1</span> RShiny Introduction</a>
<ul>
<li><a href="rshiny-introduction.html#how-does-a-web-page-work"><span class="toc-section-number">10.1.1</span> How does a web page work?</a></li>
</ul></li>
<li><a href="application-structure.html#application-structure"><span class="toc-section-number">10.2</span> Application Structure</a>
<ul>
<li><a href="application-structure.html#ui-design"><span class="toc-section-number">10.2.1</span> UI Design</a></li>
<li><a href="application-structure.html#server-functionality"><span class="toc-section-number">10.2.2</span> Server Functionality</a></li>
</ul></li>
<li><a href="reactivity.html#reactivity"><span class="toc-section-number">10.3</span> Reactivity</a></li>
<li><a href="publishing.html#publishing"><span class="toc-section-number">10.4</span> Publishing</a></li>
</ul></li>
<li><a href="communicating-with-r.html#communicating-with-r"><span class="toc-section-number">11</span> Communicating with R</a>
<ul>
<li><a href="rmarkdown-knitr.html#rmarkdown-knitr"><span class="toc-section-number">11.1</span> RMarkdown &amp; knitr</a></li>
<li><a href="bookdown.html#bookdown"><span class="toc-section-number">11.2</span> bookdown</a></li>
</ul></li>
<li><a href="contribution-guide.html#contribution-guide"><span class="toc-section-number">12</span> Contribution Guide</a>
<ul>
<li><a href="custom-blocks.html#custom-blocks"><span class="toc-section-number">12.1</span> Custom blocks</a></li>
</ul></li>
<li><a href="assignments.html#assignments">Assignments</a>
<ul>
<li><a href="assignment-format.html#assignment-format">Assignment Format</a></li>
<li><a href="starting-an-assignment.html#starting-an-assignment">Starting an Assignment</a>
<ul>
<li><a href="starting-an-assignment.html#starting-your-own-git-repository-and-cloning-it">Starting your own <code>git</code> repository and cloning it</a></li>
<li><a href="starting-an-assignment.html#contents">Contents</a></li>
<li><a href="starting-an-assignment.html#github-tutorial">GitHub tutorial</a></li>
<li><a href="starting-an-assignment.html#committing-and-pushing-with-r-and-without">Committing and Pushing with R (and without)</a></li>
<li><a href="starting-an-assignment.html#using-rstudio">Using RStudio</a></li>
<li><a href="starting-an-assignment.html#using-the-command-line">Using the command line</a></li>
</ul></li>
<li><a href="assignment-1.html#assignment-1">Assignment 1</a>
<ul>
<li><a href="assignment-1.html#problem-statement">Problem Statement</a></li>
<li><a href="assignment-1.html#learning-objectives-and-skill-list">Learning Objectives and Skill List</a></li>
<li><a href="assignment-1.html#instructions">Instructions</a></li>
<li><a href="assignment-1.html#hints">Hints</a></li>
</ul></li>
<li><a href="assignment-2.html#assignment-2">Assignment 2</a>
<ul>
<li><a href="assignment-2.html#problem-statement-1">Problem Statement</a></li>
<li><a href="assignment-2.html#learning-objectives">Learning Objectives</a></li>
<li><a href="assignment-2.html#skill-list">Skill List</a></li>
<li><a href="assignment-2.html#instructions-1">Instructions</a></li>
<li><a href="assignment-2.html#function-details">Function Details</a></li>
</ul></li>
<li><a href="assignment-3.html#assignment-3">Assignment 3</a>
<ul>
<li><a href="assignment-3.html#problem-statement-2">Problem Statement</a></li>
<li><a href="assignment-3.html#learning-objectives-1">Learning Objectives</a></li>
<li><a href="assignment-3.html#skill-list-1">Skill List</a></li>
<li><a href="assignment-3.html#background-on-microarrays">Background on Microarrays</a></li>
<li><a href="assignment-3.html#background-on-principal-component-analysis">Background on Principal Component Analysis</a></li>
<li><a href="assignment-3.html#marisa-et-al.-gene-expression-classification-of-colon-cancer-into-molecular-subtypes-characterization-validation-and-prognostic-value.-plos-medicine-may-2013.-pmid-23700391">Marisa et al. Gene Expression Classification of Colon Cancer into Molecular Subtypes: Characterization, Validation, and Prognostic Value. PLoS Medicine, May 2013. PMID: 23700391</a></li>
<li><a href="assignment-3.html#scaling-data-using-r-scale">Scaling data using R <code>scale()</code></a></li>
<li><a href="assignment-3.html#proportion-of-variance-explained">Proportion of variance explained</a></li>
<li><a href="assignment-3.html#plotting-and-visualization-of-pca">Plotting and visualization of PCA</a></li>
<li><a href="assignment-3.html#hierarchical-clustering-and-heatmaps">Hierarchical Clustering and Heatmaps</a></li>
<li><a href="assignment-3.html#references">References</a></li>
</ul></li>
<li><a href="assignment-4.html#assignment-4">Assignment 4</a>
<ul>
<li><a href="assignment-4.html#problem-statement-3">Problem Statement</a></li>
<li><a href="assignment-4.html#learning-objectives-2">Learning Objectives</a></li>
<li><a href="assignment-4.html#skill-list-2">Skill List</a></li>
<li><a href="assignment-4.html#instructions-2">Instructions</a></li>
<li><a href="assignment-4.html#deliverables">Deliverables</a></li>
<li><a href="assignment-4.html#function-details-1">Function Details</a></li>
<li><a href="assignment-4.html#hints-1">Hints</a></li>
</ul></li>
<li><a href="assignment-5.html#assignment-5">Assignment 5</a>
<ul>
<li><a href="assignment-5.html#problem-statement-4">Problem Statement</a></li>
<li><a href="assignment-5.html#learning-objectives-3">Learning Objectives</a></li>
<li><a href="assignment-5.html#skill-list-3">Skill List</a></li>
<li><a href="assignment-5.html#deseq2-background">DESeq2 Background</a></li>
<li><a href="assignment-5.html#generating-a-counts-matrix">Generating a counts matrix</a></li>
<li><a href="assignment-5.html#prefiltering-counts-matrix">Prefiltering Counts matrix</a></li>
<li><a href="assignment-5.html#median-of-ratios-normalization">Median-of-ratios normalization</a></li>
<li><a href="assignment-5.html#deseq2-preparation">DESeq2 preparation</a></li>
<li><a href="assignment-5.html#omeara-et-al.-transcriptional-reversion-of-cardiac-myocyte-fate-during-mammalian-cardiac-regeneration.-circ-res.-feb-2015.-pmid-25477501l">O’Meara et al. Transcriptional Reversion of Cardiac Myocyte Fate During Mammalian Cardiac Regeneration. Circ Res. Feb 2015. PMID: 25477501l</a></li>
<li><a href="assignment-5.html#reading-and-subsetting-the-data-from-verse_counts.tsv-and-sample_metadata.csv">1. Reading and subsetting the data from verse_counts.tsv and sample_metadata.csv</a></li>
<li><a href="assignment-5.html#running-deseq2">2. Running DESeq2</a></li>
<li><a href="assignment-5.html#annotating-results-to-construct-a-labeled-volcano-plot">3. Annotating results to construct a labeled volcano plot</a></li>
<li><a href="assignment-5.html#diagnostic-plot-of-the-raw-p-values-for-all-genes">4. Diagnostic plot of the raw p-values for all genes</a></li>
<li><a href="assignment-5.html#plotting-the-logfoldchanges-for-differentially-expressed-genes">5. Plotting the LogFoldChanges for differentially expressed genes</a></li>
<li><a href="assignment-5.html#the-choice-of-fdr-cutoff-depends-on-cost">The choice of FDR cutoff depends on cost</a></li>
<li><a href="assignment-5.html#plotting-the-normalized-counts-of-differentially-expressed-genes">6. Plotting the normalized counts of differentially expressed genes</a></li>
<li><a href="assignment-5.html#volcano-plot-to-visualize-differential-expression-results">7. Volcano Plot to visualize differential expression results</a></li>
<li><a href="assignment-5.html#running-fgsea-vignette">8. Running fgsea <span>vignette</span></a></li>
<li><a href="assignment-5.html#plotting-the-top-ten-positive-nes-and-top-ten-negative-nes-pathways">9. Plotting the top ten positive NES and top ten negative NES pathways</a></li>
<li><a href="assignment-5.html#references-1">References</a></li>
</ul></li>
<li><a href="assignment-6.html#assignment-6">Assignment 6</a>
<ul>
<li><a href="assignment-6.html#problem-statement-5">Problem Statement</a></li>
<li><a href="assignment-6.html#learning-objectives-4">Learning Objectives</a></li>
<li><a href="assignment-6.html#skill-list-4">Skill List</a></li>
<li><a href="assignment-6.html#instructions-3">Instructions</a></li>
<li><a href="assignment-6.html#function-details-2">Function Details</a></li>
</ul></li>
<li><a href="assignment-7.html#assignment-7">Assignment 7</a>
<ul>
<li><a href="assignment-7.html#problem-statement-6">Problem Statement</a></li>
<li><a href="assignment-7.html#learning-objectives-5">Learning Objectives</a></li>
<li><a href="assignment-7.html#skill-list-5">Skill List</a></li>
<li><a href="assignment-7.html#instructions-4">Instructions</a></li>
<li><a href="assignment-7.html#function-details-3">“Function” Details</a></li>
</ul></li>
</ul></li>
<li class="appendix"><span><b>Appendix</b></span></li>
<li><a href="class-outlines.html#class-outlines"><span class="toc-section-number">A</span> Class Outlines</a>
<ul>
<li><a href="week-1.html#week-1"><span class="toc-section-number">A.1</span> Week 1</a></li>
<li><a href="week-2.html#week-2"><span class="toc-section-number">A.2</span> Week 2</a></li>
<li><a href="week-3.html#week-3"><span class="toc-section-number">A.3</span> Week 3</a></li>
<li><a href="week-4.html#week-4"><span class="toc-section-number">A.4</span> Week 4</a></li>
<li><a href="week-5.html#week-5"><span class="toc-section-number">A.5</span> Week 5</a></li>
<li><a href="week-6.html#week-6"><span class="toc-section-number">A.6</span> Week 6</a></li>
<li><a href="week-7.html#week-7"><span class="toc-section-number">A.7</span> Week 7</a></li>
<li><a href="week-8.html#week-8"><span class="toc-section-number">A.8</span> Week 8</a></li>
<li><a href="week-9.html#week-9"><span class="toc-section-number">A.9</span> Week 9</a></li>
<li><a href="week-10.html#week-10"><span class="toc-section-number">A.10</span> Week 10</a></li>
<li><a href="week-11.html#week-11"><span class="toc-section-number">A.11</span> Week 11</a></li>
<li><a href="week-12.html#week-12"><span class="toc-section-number">A.12</span> Week 12</a></li>
<li><a href="week-13.html#week-13"><span class="toc-section-number">A.13</span> Week 13</a></li>
<li><a href="week-14.html#week-14"><span class="toc-section-number">A.14</span> Week 14</a></li>
</ul></li>
</ul>
</div>
</div>
</div>
<div class="row">
<div class="col-sm-12">
<div id="gene-identifiers" class="section level2" number="8.6">
<h2><span class="header-section-number">8.6</span> Gene Identifiers</h2>
<div id="gene-identifier-systems" class="section level3" number="8.6.1">
<h3><span class="header-section-number">8.6.1</span> Gene Identifier Systems</h3>
<p>Since the first genetic sequence for a gene was determined in 1965 - <a href="https://pubmed.ncbi.nlm.nih.gov/14263761/">an alanine
tRNA in yeast</a><span class="citation">(<a href="#ref-Holley1965-tu" role="doc-biblioref">Holley et al. 1965</a>)</span> -
scientists have been trying to give them names. In 1979, <a href="https://pubmed.ncbi.nlm.nih.gov/295268/">formal guidelines for
the nomenclature used for human genes was
proposed</a><span class="citation">(<a href="#ref-Shows1979-iw" role="doc-biblioref">Shows et al. 1979</a>)</span> along with the
founding of the Human Gene Nomenclature Committee <span class="citation">(<a href="#ref-Bruford2020-zs" role="doc-biblioref">Bruford et al. 2020</a>)</span>, so that
researchers would have a common vocabulary to refer to genes in a consistent
way. In 1989, this committee was placed under the auspices of the Human Genome
Organisation (HUGO), becoming the HUGO Gene Nomenclature Committee (HGNC). The
HGNC has been the official body providing guidelines and gene naming authority
since, and as such HGNC gene names are often called <em>gene symbols</em>.</p>
<div class="box note">
<p>As per <span class="citation"><a href="#ref-Bruford2020-zs" role="doc-biblioref">Bruford et al.</a> (<a href="#ref-Bruford2020-zs" role="doc-biblioref">2020</a>)</span>, the official human gene symbol guidelines are as
follows:</p>
<ol style="list-style-type: decimal">
<li>Each gene is assigned a unique symbol, HGNC ID and descriptive name.</li>
<li>Symbols contain only uppercase Latin letters and Arabic numerals.</li>
<li>Symbols should not be the same as commonly used abbreviations</li>
<li>Nomenclature should not contain reference to any species or “G” for gene.</li>
<li>Nomenclature should not be offensive or pejorative.</li>
</ol>
</div>
<p>Gene symbols are the most human-readable system for naming genes. The gene
symbols <em>BRCA1</em>, <em>APOE</em>, and <em>ACE2</em> may be familiar to you as they are involved
in common diseases, namely breast cancer, Alzheimer’s Disease risk, and
SARS-COV-2, respectively. Typically, the gene symbol is an acronym that roughly
represents a label of what the gene is or does (or was originally found to be or
do, as many genes are subsequently discovered to be involved in entirely
separate processes as well), e.g. <em>APOE</em> represents the gene <a href="https://www.ensembl.org/Homo_sapiens/Gene/Summary?db=core;g=ENSG00000130203;r=19:44905791-44909393">apolipoprotein
E</a>.
This convention is convenient for humans when reading and identifying genes.
However, standardized though the symbol conventions may be, they are not always
easy for computers to work with, and ambiguities can cause mapping problems.</p>
<p>Other gene identifier systems developed along with other gene information
databases. In 2000, the <a href="https://www.ensembl.org/">Ensembl</a> genome browser was
launched by the <a href="https://wellcome.org/">Wellcome Trust</a>, a charitable foundation
based in the United Kingdom, with the goal of providing automated annotation of
the human genome. The Ensembl Project, which supports the genome browser,
recognized even before the publication of the human genome that manual
annotation and curation of genes is a slow and labor intensive process that
would not provide researchers around the world timely access to information. The
project therefore created a set of automated tools and pipelines to collect,
process, and publish rapid and consistent annotation of genes in the human
genome. Since its initial release, Ensembl now supports over 50,000 different
genomes.</p>
<p>Ensembl assigns a automatic, systematic ID called the Ensembl Gene ID to every
gene in its database. Human Ensembl Gene IDs take the form <code>ENSG</code> plus an 11
digit number, optionally followed by a period delimited version number. For
example, at the time of writing the BRCA1 gene has an Ensembl Gene ID of
<code>ENSG00000012048.23</code>. The stable ID portion (i.e. <code>ENSGNNNNNNNNNNN</code>) will remain
associated with the gene forever (unless the gene annotation <a href="http://useast.ensembl.org/Help/View?id=181">changes
“dramatically” in which case it is
retired</a>). The <code>.23</code> is the version
number of the gene annotation, meaning this gene has been updated 22 times (plus
its initial version) since addition to the database. The additional version
information is very important for reproducibility of biological analysis, since
conclusions drawn by results of these analyses are usually based on the most
current information about a gene which are continually updated over time.</p>
<div class="box note">
<p>Ensembl maintains annotations for many different organisms, and the gene
identifiers for each genome contain codes that indicate which organism the gene
is for. Here are the codes for genes in several different species:</p>
<table style="width:92%;">
<colgroup>
<col width="23%" />
<col width="36%" />
<col width="31%" />
</colgroup>
<thead>
<tr class="header">
<th>Gene ID Prefix</th>
<th>Organism</th>
<th>Example</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><code>ENSG</code></td>
<td>Homo sapiens</td>
<td><a href="https://useast.ensembl.org/Homo_sapiens/Gene/Summary?db=core;g=ENSG00000139618;r=13:32315086-32400268">ENSG00000139618</a></td>
</tr>
<tr class="even">
<td><code>ENSMUSG</code></td>
<td>Mus musculus (mouse)</td>
<td><a href="https://useast.ensembl.org/Mus_musculus/Gene/Summary?db=core;g=ENSMUSG00000017167;r=11:101061349-101081550">ENSMUSG00000017167</a></td>
</tr>
<tr class="odd">
<td><code>ENSDARG</code></td>
<td>Danio rerio (zebrafish)</td>
<td><a href="https://useast.ensembl.org/Danio_rerio/Gene/Summary?db=core;g=ENSDARG00000024771;r=18:5213338-5227420;t=ENSDART00000033574">ENSDARG00000024771</a></td>
</tr>
<tr class="even">
<td><code>ENSMMUG</code></td>
<td>Macaca mulata (Macaque)</td>
<td><a href="https://useast.ensembl.org/Macaca_mulatta/Gene/Summary?db=core;g=ENSMMUG00000020312;r=5:65376507-65505566">ENSMMUG00000020312</a></td>
</tr>
<tr class="odd">
<td><code>ENSVPAG</code></td>
<td>Vicugna pacos (Alpaca)</td>
<td><a href="https://useast.ensembl.org/Vicugna_pacos/Gene/Summary?db=core;g=ENSVPAG00000001990;r=GeneScaffold_2392:178775-179671;t=ENSVPAT00000001989">ENSVPAG00000001990</a></td>
</tr>
</tbody>
</table>
</div>
<p>Ensembl Gene IDs and gene symbols are the most commonly used gene identifiers.
The <a href="https://www.gencodegenes.org/">GENCODE project</a> which provides consistent
and stable genome annotation releases for human and mouse genomes, uses these
two types of identifiers with its official annotations.</p>
<div class="box note">
<p>Ensembl provides stable identifiers for transcripts as well as genes. Transcript
identifiers correspond to distinct patterns of exons/introns that have been
identified for a gene, and each gene has one or more distinct transcripts.
Ensembl Transcript IDs have the form <code>ENSTNNNNNNNNNNN.NN</code> similar to Gene IDs.</p>
</div>
<div class="box note">
<p>Ensembl is not the only gene identifier system besides gene symbol. Several
other databases have devised and maintain their own identifiers, most notably
[Entrez Gene IDs](<a href="">UCSC Gene IDs</a> used by the <a href="https://www.ncbi.nlm.nih.gov/gene">NCBI Gene
database</a> which assigns a unique integer to
each gene in each organism, and the <a href="https://www.omim.org/">Online Mendelian Inheritance in Man
(OMIM)</a> database, which has identifiers that look like
<code>OMIM:NNNNN</code>, where each OMIM ID refers to a unique gene or human phenotype.
However, the primary identifiers used in modern human biological research remain
Ensembl IDs and official HGNC gene symbols.</p>
</div>
</div>
<div id="mapping-between-identifier-systems" class="section level3" number="8.6.2">
<h3><span class="header-section-number">8.6.2</span> Mapping Between Identifier Systems</h3>
<p>A very common operation in biological analysis is to map between gene identifier
systems, e.g. you are given Ensembl Gene ID and want to map to the more
human-recognizable gene symbols. Ensembl provides a service called
<a href="https://www.ensembl.org/biomart/martview">BioMart</a> that allows you to download
annotation information for genes, transcripts, and other information maintained
in their databases. It also provides limited access to external sources of
information on their genes, including cross references to HGNC gene symbols and
some other gene identifier systems. The Ensembl website has <a href="https://useast.ensembl.org/info/data/biomart/index.html">helpful
documentation</a> on how
to use BioMart to download annotation data using the web interface.</p>
<p>In addition to downloading annotations as a CSV file from the web interface,
Ensembl also provides <a href="https://useast.ensembl.org/info/data/biomart/biomart_r_package.html">the biomaRt Bioconductor
package</a> to
allow programmatic access to the same information directly from R. There is <a href="https://useast.ensembl.org/info/genome/index.html">much
more information in the Ensembl
databases</a> than gene
annotation data that can be accessed via BioMart, but we will provide a brief
example of how to extract gene information here:</p>
<div class="sourceCode" id="cb364"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb364-1"><a href="gene-identifiers.html#cb364-1" aria-hidden="true" tabindex="-1"></a><span class="co"># this assumes biomaRt is already installed through bioconductor</span></span>
<span id="cb364-2"><a href="gene-identifiers.html#cb364-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(biomaRt)</span>
<span id="cb364-3"><a href="gene-identifiers.html#cb364-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb364-4"><a href="gene-identifiers.html#cb364-4" aria-hidden="true" tabindex="-1"></a><span class="co"># the human biomaRt database is named &quot;hsapiens_gene_ensembl&quot;</span></span>
<span id="cb364-5"><a href="gene-identifiers.html#cb364-5" aria-hidden="true" tabindex="-1"></a>ensembl <span class="ot">&lt;-</span> <span class="fu">useEnsembl</span>(<span class="at">biomart=</span><span class="st">&quot;ensembl&quot;</span>, <span class="at">dataset=</span><span class="st">&quot;hsapiens_gene_ensembl&quot;</span>)</span>
<span id="cb364-6"><a href="gene-identifiers.html#cb364-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb364-7"><a href="gene-identifiers.html#cb364-7" aria-hidden="true" tabindex="-1"></a><span class="co"># listAttributes() returns a data frame, turn into a tibble to help with formatting</span></span>
<span id="cb364-8"><a href="gene-identifiers.html#cb364-8" aria-hidden="true" tabindex="-1"></a><span class="fu">as_tibble</span>(<span class="fu">listAttributes</span>(ensembl))</span>
<span id="cb364-9"><a href="gene-identifiers.html#cb364-9" aria-hidden="true" tabindex="-1"></a><span class="co"># A tibble: 3,143 x 3</span></span>
<span id="cb364-10"><a href="gene-identifiers.html#cb364-10" aria-hidden="true" tabindex="-1"></a>   name                          description                  page        </span>
<span id="cb364-11"><a href="gene-identifiers.html#cb364-11" aria-hidden="true" tabindex="-1"></a>   <span class="sc">&lt;</span>chr<span class="sc">&gt;</span>                         <span class="er">&lt;</span>chr<span class="sc">&gt;</span>                        <span class="er">&lt;</span>chr<span class="sc">&gt;</span>       </span>
<span id="cb364-12"><a href="gene-identifiers.html#cb364-12" aria-hidden="true" tabindex="-1"></a> <span class="dv">1</span> ensembl_gene_id               Gene stable ID               feature_page</span>
<span id="cb364-13"><a href="gene-identifiers.html#cb364-13" aria-hidden="true" tabindex="-1"></a> <span class="dv">2</span> ensembl_gene_id_version       Gene stable ID version       feature_page</span>
<span id="cb364-14"><a href="gene-identifiers.html#cb364-14" aria-hidden="true" tabindex="-1"></a> <span class="dv">3</span> ensembl_transcript_id         Transcript stable ID         feature_page</span>
<span id="cb364-15"><a href="gene-identifiers.html#cb364-15" aria-hidden="true" tabindex="-1"></a> <span class="dv">4</span> ensembl_transcript_id_version Transcript stable ID version feature_page</span>
<span id="cb364-16"><a href="gene-identifiers.html#cb364-16" aria-hidden="true" tabindex="-1"></a> <span class="dv">5</span> ensembl_peptide_id            Protein stable ID            feature_page</span>
<span id="cb364-17"><a href="gene-identifiers.html#cb364-17" aria-hidden="true" tabindex="-1"></a> <span class="dv">6</span> ensembl_peptide_id_version    Protein stable ID version    feature_page</span>
<span id="cb364-18"><a href="gene-identifiers.html#cb364-18" aria-hidden="true" tabindex="-1"></a> <span class="dv">7</span> ensembl_exon_id               Exon stable ID               feature_page</span>
<span id="cb364-19"><a href="gene-identifiers.html#cb364-19" aria-hidden="true" tabindex="-1"></a> <span class="dv">8</span> description                   Gene description             feature_page</span>
<span id="cb364-20"><a href="gene-identifiers.html#cb364-20" aria-hidden="true" tabindex="-1"></a> <span class="dv">9</span> chromosome_name               Chromosome<span class="sc">/</span>scaffold name     feature_page</span>
<span id="cb364-21"><a href="gene-identifiers.html#cb364-21" aria-hidden="true" tabindex="-1"></a><span class="dv">10</span> start_position                Gene <span class="fu">start</span> (bp)              feature_page</span>
<span id="cb364-22"><a href="gene-identifiers.html#cb364-22" aria-hidden="true" tabindex="-1"></a><span class="co"># ... with 3,133 more rows</span></span></code></pre></div>
<p>The <code>name</code> column provides the programmatic name associated with the attribute
that can be used to retrieve that annotation information using the <code>getBM()</code>
function:</p>
<div class="sourceCode" id="cb365"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb365-1"><a href="gene-identifiers.html#cb365-1" aria-hidden="true" tabindex="-1"></a><span class="co"># create a tibble with ensembl gene ID, HGNC gene symbol, and gene description</span></span>
<span id="cb365-2"><a href="gene-identifiers.html#cb365-2" aria-hidden="true" tabindex="-1"></a>gene_map <span class="ot">&lt;-</span> <span class="fu">as_tibble</span>(</span>
<span id="cb365-3"><a href="gene-identifiers.html#cb365-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">getBM</span>(</span>
<span id="cb365-4"><a href="gene-identifiers.html#cb365-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">attributes=</span><span class="fu">c</span>(<span class="st">&quot;ensembl_gene_id&quot;</span>, <span class="st">&quot;hgnc_gene_symbol&quot;</span>, <span class="st">&quot;description&quot;</span>),</span>
<span id="cb365-5"><a href="gene-identifiers.html#cb365-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">mart=</span>ensembl</span>
<span id="cb365-6"><a href="gene-identifiers.html#cb365-6" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb365-7"><a href="gene-identifiers.html#cb365-7" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb365-8"><a href="gene-identifiers.html#cb365-8" aria-hidden="true" tabindex="-1"></a>gene_map</span>
<span id="cb365-9"><a href="gene-identifiers.html#cb365-9" aria-hidden="true" tabindex="-1"></a><span class="co"># A tibble: 68,012 x 3</span></span>
<span id="cb365-10"><a href="gene-identifiers.html#cb365-10" aria-hidden="true" tabindex="-1"></a>   ensembl_gene_id hgnc_symbol description                                                               </span>
<span id="cb365-11"><a href="gene-identifiers.html#cb365-11" aria-hidden="true" tabindex="-1"></a>   <span class="sc">&lt;</span>chr<span class="sc">&gt;</span>           <span class="er">&lt;</span>chr<span class="sc">&gt;</span>       <span class="er">&lt;</span>chr<span class="sc">&gt;</span>                                                                     </span>
<span id="cb365-12"><a href="gene-identifiers.html#cb365-12" aria-hidden="true" tabindex="-1"></a> <span class="dv">1</span> ENSG00000210049 MT<span class="sc">-</span>TF       mitochondrially encoded tRNA<span class="sc">-</span><span class="fu">Phe</span> (UUU<span class="sc">/</span>C) [Source<span class="sc">:</span>HGNC Symbol;Acc<span class="sc">:</span>HGNC<span class="sc">:</span><span class="dv">748</span><span class="sc">~</span></span>
<span id="cb365-13"><a href="gene-identifiers.html#cb365-13" aria-hidden="true" tabindex="-1"></a> <span class="dv">2</span> ENSG00000211459 MT<span class="sc">-</span>RNR1     mitochondrially encoded 12S rRNA [Source<span class="sc">:</span>HGNC Symbol;Acc<span class="sc">:</span>HGNC<span class="sc">:</span><span class="dv">7470</span>]       </span>
<span id="cb365-14"><a href="gene-identifiers.html#cb365-14" aria-hidden="true" tabindex="-1"></a> <span class="dv">3</span> ENSG00000210077 MT<span class="sc">-</span>TV       mitochondrially encoded tRNA<span class="sc">-</span><span class="fu">Val</span> (GUN) [Source<span class="sc">:</span>HGNC Symbol;Acc<span class="sc">:</span>HGNC<span class="sc">:</span><span class="dv">7500</span>]</span>
<span id="cb365-15"><a href="gene-identifiers.html#cb365-15" aria-hidden="true" tabindex="-1"></a> <span class="dv">4</span> ENSG00000210082 MT<span class="sc">-</span>RNR2     mitochondrially encoded 16S rRNA [Source<span class="sc">:</span>HGNC Symbol;Acc<span class="sc">:</span>HGNC<span class="sc">:</span><span class="dv">7471</span>]       </span>
<span id="cb365-16"><a href="gene-identifiers.html#cb365-16" aria-hidden="true" tabindex="-1"></a> <span class="dv">5</span> ENSG00000209082 MT<span class="sc">-</span>TL1      mitochondrially encoded tRNA<span class="sc">-</span><span class="fu">Leu</span> (UUA<span class="sc">/</span>G) <span class="dv">1</span> [Source<span class="sc">:</span>HGNC Symbol;Acc<span class="sc">:</span>HGNC<span class="sc">:</span><span class="dv">7</span><span class="sc">~</span></span>
<span id="cb365-17"><a href="gene-identifiers.html#cb365-17" aria-hidden="true" tabindex="-1"></a> <span class="dv">6</span> ENSG00000198888 MT<span class="sc">-</span>ND1      mitochondrially encoded NADH<span class="sc">:</span>ubiquinone oxidoreductase core subunit <span class="dv">1</span> [So<span class="sc">~</span></span>
<span id="cb365-18"><a href="gene-identifiers.html#cb365-18" aria-hidden="true" tabindex="-1"></a> <span class="dv">7</span> ENSG00000210100 MT<span class="sc">-</span>TI       mitochondrially encoded tRNA<span class="sc">-</span><span class="fu">Ile</span> (AUU<span class="sc">/</span>C) [Source<span class="sc">:</span>HGNC Symbol;Acc<span class="sc">:</span>HGNC<span class="sc">:</span><span class="dv">748</span><span class="sc">~</span></span>
<span id="cb365-19"><a href="gene-identifiers.html#cb365-19" aria-hidden="true" tabindex="-1"></a> <span class="dv">8</span> ENSG00000210107 MT<span class="sc">-</span>TQ       mitochondrially encoded tRNA<span class="sc">-</span><span class="fu">Gln</span> (CAA<span class="sc">/</span>G) [Source<span class="sc">:</span>HGNC Symbol;Acc<span class="sc">:</span>HGNC<span class="sc">:</span><span class="dv">749</span><span class="sc">~</span></span>
<span id="cb365-20"><a href="gene-identifiers.html#cb365-20" aria-hidden="true" tabindex="-1"></a> <span class="dv">9</span> ENSG00000210112 MT<span class="sc">-</span>TM       mitochondrially encoded tRNA<span class="sc">-</span><span class="fu">Met</span> (AUA<span class="sc">/</span>G) [Source<span class="sc">:</span>HGNC Symbol;Acc<span class="sc">:</span>HGNC<span class="sc">:</span><span class="dv">749</span><span class="sc">~</span></span>
<span id="cb365-21"><a href="gene-identifiers.html#cb365-21" aria-hidden="true" tabindex="-1"></a><span class="dv">10</span> ENSG00000198763 MT<span class="sc">-</span>ND2      mitochondrially encoded NADH<span class="sc">:</span>ubiquinone oxidoreductase core subunit <span class="dv">2</span> [So<span class="sc">~</span></span>
<span id="cb365-22"><a href="gene-identifiers.html#cb365-22" aria-hidden="true" tabindex="-1"></a><span class="co"># ... with 68,002 more rows</span></span></code></pre></div>
<p>With this Ensembl Gene ID to HGNC symbol mapping in hand, you can combine the
tibble above with other tibbles containing gene information by <a href="relational-data.html#relational-data">joining them
together</a>.</p>
<div class="box note">
<p>BioMart/biomaRt is not the only ways to map different gene identifiers. The
Bioconductor package
<a href="https://bioconductor.org/packages/release/bioc/html/AnnotationDbi.html">AnnotateDbi</a>
also provides this functionality in a flexible format independent of the Ensembl
databases. This package includes not only gene identifier mapping and
information, but also identifiers from technology platforms, e.g. probe set IDs
from <a href="microarrays.html#microarrays">microarrays</a>, that can help when working with these types of
data. The package also allows comprehensive and flexible <a href="gene-identifiers.html#mapping-homologs">homolog
mapping</a>. As with all Bioconductor packages, the
AnnotationDbi documentation is well written and comprehensive, though knowledge
of <a href="https://en.wikipedia.org/wiki/Relational_database">relational databases</a> is
helpful in understanding how the packages work.</p>
</div>
<div class="box readmore">
<ul>
<li><a href="https://www.ensembl.org/biomart/martview">Ensembl BioMart portal</a></li>
<li><a href="https://useast.ensembl.org/info/data/biomart/index.html">Ensembl BioMart web portal documentation</a></li>
<li><a href="http://www.bioconductor.org/packages/release/bioc/html/biomaRt.html">biomaRt Bioconductor documentation</a></li>
</ul>
</div>
</div>
<div id="mapping-homologs" class="section level3" number="8.6.3">
<h3><span class="header-section-number">8.6.3</span> Mapping Homologs</h3>
<p>Sometimes it is necessary to link datasets from different organisms together by
orthology. For example, in an experiment performed in mice, we might be
interested in comparing gene expression patterns observed in the mouse samples
to a publicly available human dataset. In these contexts, we must link gene
identifiers from one organism to their corresponding homologs in the other.
BioMart enables us to extract these linked identifiers by explicitly <a href="https://bioconductor.org/packages/release/bioc/vignettes/biomaRt/inst/doc/accessing_ensembl.html#given-the-human-gene-tp53-retrieve-the-human-chromosomal-location-of-this-gene-and-also-retrieve-the-chromosomal-location-and-refseq-id-of-its-homolog-in-mouse.">connecting
different biomaRt
databases</a>
with the <code>getLDS()</code> function:</p>
<div class="sourceCode" id="cb366"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb366-1"><a href="gene-identifiers.html#cb366-1" aria-hidden="true" tabindex="-1"></a>human_db <span class="ot">&lt;-</span> <span class="fu">useEnsembl</span>(<span class="st">&quot;ensembl&quot;</span>, <span class="at">dataset =</span> <span class="st">&quot;hsapiens_gene_ensembl&quot;</span>)</span>
<span id="cb366-2"><a href="gene-identifiers.html#cb366-2" aria-hidden="true" tabindex="-1"></a>mouse_db <span class="ot">&lt;-</span> <span class="fu">useEnsembl</span>(<span class="st">&quot;ensembl&quot;</span>, <span class="at">dataset =</span> <span class="st">&quot;mmusculus_gene_ensembl&quot;</span>)</span>
<span id="cb366-3"><a href="gene-identifiers.html#cb366-3" aria-hidden="true" tabindex="-1"></a>orth_map <span class="ot">&lt;-</span> <span class="fu">as_tibble</span>(</span>
<span id="cb366-4"><a href="gene-identifiers.html#cb366-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">getLDS</span>(<span class="at">attributes =</span> <span class="fu">c</span>(<span class="st">&quot;ensembl_gene_id&quot;</span>, <span class="st">&quot;hgnc_symbol&quot;</span>),</span>
<span id="cb366-5"><a href="gene-identifiers.html#cb366-5" aria-hidden="true" tabindex="-1"></a>       <span class="at">mart =</span> human_db,</span>
<span id="cb366-6"><a href="gene-identifiers.html#cb366-6" aria-hidden="true" tabindex="-1"></a>       <span class="at">attributesL =</span> <span class="fu">c</span>(<span class="st">&quot;ensembl_gene_id&quot;</span>, <span class="st">&quot;mgi_symbol&quot;</span>),</span>
<span id="cb366-7"><a href="gene-identifiers.html#cb366-7" aria-hidden="true" tabindex="-1"></a>       <span class="at">martL =</span> mouse_db</span>
<span id="cb366-8"><a href="gene-identifiers.html#cb366-8" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb366-9"><a href="gene-identifiers.html#cb366-9" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb366-10"><a href="gene-identifiers.html#cb366-10" aria-hidden="true" tabindex="-1"></a>orth_map</span>
<span id="cb366-11"><a href="gene-identifiers.html#cb366-11" aria-hidden="true" tabindex="-1"></a><span class="co"># A tibble: 26,390 x 4</span></span>
<span id="cb366-12"><a href="gene-identifiers.html#cb366-12" aria-hidden="true" tabindex="-1"></a>   Gene.stable.ID  HGNC.symbol Gene.stable.ID<span class="fl">.1</span>   MGI.symbol</span>
<span id="cb366-13"><a href="gene-identifiers.html#cb366-13" aria-hidden="true" tabindex="-1"></a>   <span class="sc">&lt;</span>chr<span class="sc">&gt;</span>           <span class="er">&lt;</span>chr<span class="sc">&gt;</span>       <span class="er">&lt;</span>chr<span class="sc">&gt;</span>              <span class="er">&lt;</span>chr<span class="sc">&gt;</span>     </span>
<span id="cb366-14"><a href="gene-identifiers.html#cb366-14" aria-hidden="true" tabindex="-1"></a> <span class="dv">1</span> ENSG00000198695 MT<span class="sc">-</span>ND6      ENSMUSG00000064368 <span class="st">&quot;mt-Nd6&quot;</span>  </span>
<span id="cb366-15"><a href="gene-identifiers.html#cb366-15" aria-hidden="true" tabindex="-1"></a> <span class="dv">2</span> ENSG00000212907 MT<span class="sc">-</span>ND4L     ENSMUSG00000065947 <span class="st">&quot;mt-Nd4l&quot;</span></span>
<span id="cb366-16"><a href="gene-identifiers.html#cb366-16" aria-hidden="true" tabindex="-1"></a> <span class="dv">3</span> ENSG00000279169 PRAMEF13    ENSMUSG00000094303 <span class="st">&quot;&quot;</span>        </span>
<span id="cb366-17"><a href="gene-identifiers.html#cb366-17" aria-hidden="true" tabindex="-1"></a> <span class="dv">4</span> ENSG00000279169 PRAMEF13    ENSMUSG00000094722 <span class="st">&quot;&quot;</span>        </span>
<span id="cb366-18"><a href="gene-identifiers.html#cb366-18" aria-hidden="true" tabindex="-1"></a> <span class="dv">5</span> ENSG00000279169 PRAMEF13    ENSMUSG00000095666 <span class="st">&quot;&quot;</span>        </span>
<span id="cb366-19"><a href="gene-identifiers.html#cb366-19" aria-hidden="true" tabindex="-1"></a> <span class="dv">6</span> ENSG00000279169 PRAMEF13    ENSMUSG00000094741 <span class="st">&quot;&quot;</span>        </span>
<span id="cb366-20"><a href="gene-identifiers.html#cb366-20" aria-hidden="true" tabindex="-1"></a> <span class="dv">7</span> ENSG00000279169 PRAMEF13    ENSMUSG00000094836 <span class="st">&quot;&quot;</span>        </span>
<span id="cb366-21"><a href="gene-identifiers.html#cb366-21" aria-hidden="true" tabindex="-1"></a> <span class="dv">8</span> ENSG00000279169 PRAMEF13    ENSMUSG00000074720 <span class="st">&quot;&quot;</span>        </span>
<span id="cb366-22"><a href="gene-identifiers.html#cb366-22" aria-hidden="true" tabindex="-1"></a> <span class="dv">9</span> ENSG00000279169 PRAMEF13    ENSMUSG00000096236 <span class="st">&quot;&quot;</span>        </span>
<span id="cb366-23"><a href="gene-identifiers.html#cb366-23" aria-hidden="true" tabindex="-1"></a><span class="dv">10</span> ENSG00000198763 MT<span class="sc">-</span>ND2      ENSMUSG00000064345 <span class="st">&quot;mt-Nd2&quot;</span>  </span>
<span id="cb366-24"><a href="gene-identifiers.html#cb366-24" aria-hidden="true" tabindex="-1"></a><span class="co"># ... with 26,380 more rows</span></span></code></pre></div>
<p>The <code>mgi_symbol</code> field refers to the gene symbol assigned in the <a href="https://www.jax.org/research-and-faculty/resources/mouse-genome-informatics">Mouse Genome
Informatics database</a>
maintained by <a href="https://www.jax.org/">The Jackson Laboratory</a>.</p>

</div>
</div>
<h3>References</h3>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-Bruford2020-zs" class="csl-entry">
Bruford, Elspeth A, Bryony Braschi, Paul Denny, Tamsin E M Jones, Ruth L Seal, and Susan Tweedie. 2020. <span>“Guidelines for Human Gene Nomenclature.”</span> <em>Nat. Genet.</em> 52 (8): 754–58.
</div>
<div id="ref-Holley1965-tu" class="csl-entry">
Holley, R W, J Apgar, G A Everett, J T Madison, M Marquisee, S H Merrill, J R Penswick, and A Zamir. 1965. <span>“<span>STRUCTURE</span> <span>OF</span> a <span>RIBONUCLEIC</span> <span>ACID</span>.”</span> <em>Science</em> 147 (3664): 1462–65.
</div>
<div id="ref-Shows1979-iw" class="csl-entry">
Shows, T B, C A Alper, D Bootsma, M Dorf, T Douglas, T Huisman, S Kit, et al. 1979. <span>“International System for Human Gene Nomenclature (1979) <span>ISGN</span> (1979).”</span> <em>Cytogenet. Cell Genet.</em> 25 (1-4): 96–116.
</div>
</div>
<p style="text-align: center;">
<a href="high-throughput-sequencing.html"><button class="btn btn-default">Previous</button></a>
<a href="gene-expression.html"><button class="btn btn-default">Next</button></a>
</p>
</div>
</div>


</div>

<script>

// add bootstrap table styles to pandoc tables
$(document).ready(function () {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
});

</script>

</body>
</html>
