<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>3 Preliminaries | BF591 - R for Biological Sciences</title>
  <meta name="description" content="3 Preliminaries | BF591 - R for Biological Sciences" />
  <meta name="generator" content="bookdown 0.24 and GitBook 2.6.7" />

  <meta property="og:title" content="3 Preliminaries | BF591 - R for Biological Sciences" />
  <meta property="og:type" content="book" />
  
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="3 Preliminaries | BF591 - R for Biological Sciences" />
  
  
  




  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="data-bio.html"/>
<link rel="next" href="prog-basics.html"/>
<script src="libs/header-attrs-2.11/header-attrs.js"></script>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.0.1/anchor-sections.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.0.1/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Syllabus</a>
<ul>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#course-schedule"><i class="fa fa-check"></i>Course Schedule</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#course-values-and-policies"><i class="fa fa-check"></i>Course Values and Policies</a></li>
</ul></li>
<li class="chapter" data-level="1" data-path="introduction.html"><a href="introduction.html"><i class="fa fa-check"></i><b>1</b> Introduction</a>
<ul>
<li class="chapter" data-level="1.1" data-path="introduction.html"><a href="introduction.html#who-this-book-is-for"><i class="fa fa-check"></i><b>1.1</b> Who This Book Is For</a></li>
<li class="chapter" data-level="1.2" data-path="introduction.html"><a href="introduction.html#a-note-about-reinventing-the-wheel"><i class="fa fa-check"></i><b>1.2</b> A Note About Reinventing the Wheel</a></li>
<li class="chapter" data-level="1.3" data-path="introduction.html"><a href="introduction.html#sources-and-references"><i class="fa fa-check"></i><b>1.3</b> Sources and References</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="data-bio.html"><a href="data-bio.html"><i class="fa fa-check"></i><b>2</b> Data in Biology</a>
<ul>
<li class="chapter" data-level="2.1" data-path="data-bio.html"><a href="data-bio.html#a-brief-history-of-data-in-molecular-biology"><i class="fa fa-check"></i><b>2.1</b> A Brief History of Data in Molecular Biology</a></li>
<li class="chapter" data-level="2.2" data-path="data-bio.html"><a href="data-bio.html#biology-as-a-mature-data-science"><i class="fa fa-check"></i><b>2.2</b> Biology as a Mature Data Science</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="preliminaries.html"><a href="preliminaries.html"><i class="fa fa-check"></i><b>3</b> Preliminaries</a>
<ul>
<li class="chapter" data-level="3.1" data-path="preliminaries.html"><a href="preliminaries.html#prelim-r"><i class="fa fa-check"></i><b>3.1</b> The R Language</a></li>
<li class="chapter" data-level="3.2" data-path="preliminaries.html"><a href="preliminaries.html#prelim-rstudio"><i class="fa fa-check"></i><b>3.2</b> RStudio</a></li>
<li class="chapter" data-level="3.3" data-path="preliminaries.html"><a href="preliminaries.html#prog-the-r-script"><i class="fa fa-check"></i><b>3.3</b> The R Script</a></li>
<li class="chapter" data-level="3.4" data-path="preliminaries.html"><a href="preliminaries.html#prog-workflow"><i class="fa fa-check"></i><b>3.4</b> The Scripting Workflow</a></li>
<li class="chapter" data-level="3.5" data-path="preliminaries.html"><a href="preliminaries.html#prelim-git"><i class="fa fa-check"></i><b>3.5</b> git + github</a>
<ul>
<li class="chapter" data-level="3.5.1" data-path="preliminaries.html"><a href="preliminaries.html#motivation"><i class="fa fa-check"></i><b>3.5.1</b> Motivation</a></li>
<li class="chapter" data-level="3.5.2" data-path="preliminaries.html"><a href="preliminaries.html#git"><i class="fa fa-check"></i><b>3.5.2</b> git</a></li>
<li class="chapter" data-level="3.5.3" data-path="preliminaries.html"><a href="preliminaries.html#git-hosting-platforms-github"><i class="fa fa-check"></i><b>3.5.3</b> Git hosting platforms (GitHub)</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="prog-basics.html"><a href="prog-basics.html"><i class="fa fa-check"></i><b>4</b> R Programming</a>
<ul>
<li class="chapter" data-level="4.1" data-path="prog-basics.html"><a href="prog-basics.html#before-you-begin"><i class="fa fa-check"></i><b>4.1</b> Before you begin</a></li>
<li class="chapter" data-level="4.2" data-path="prog-basics.html"><a href="prog-basics.html#introduction-1"><i class="fa fa-check"></i><b>4.2</b> Introduction</a></li>
<li class="chapter" data-level="4.3" data-path="prog-basics.html"><a href="prog-basics.html#prog-r-syntax"><i class="fa fa-check"></i><b>4.3</b> R Syntax Basics</a></li>
<li class="chapter" data-level="4.4" data-path="prog-basics.html"><a href="prog-basics.html#prog-types"><i class="fa fa-check"></i><b>4.4</b> Basic Types of Values</a></li>
<li class="chapter" data-level="4.5" data-path="prog-basics.html"><a href="prog-basics.html#prog-struct"><i class="fa fa-check"></i><b>4.5</b> Data Structures</a>
<ul>
<li class="chapter" data-level="4.5.1" data-path="prog-basics.html"><a href="prog-basics.html#vectors"><i class="fa fa-check"></i><b>4.5.1</b> Vectors</a></li>
<li class="chapter" data-level="4.5.2" data-path="prog-basics.html"><a href="prog-basics.html#matrices"><i class="fa fa-check"></i><b>4.5.2</b> Matrices</a></li>
<li class="chapter" data-level="4.5.3" data-path="prog-basics.html"><a href="prog-basics.html#lists-and-data-frames"><i class="fa fa-check"></i><b>4.5.3</b> Lists and data frames</a></li>
</ul></li>
<li class="chapter" data-level="4.6" data-path="prog-basics.html"><a href="prog-basics.html#logical-tests-and-comparators"><i class="fa fa-check"></i><b>4.6</b> Logical Tests and Comparators</a></li>
<li class="chapter" data-level="4.7" data-path="prog-basics.html"><a href="prog-basics.html#functions"><i class="fa fa-check"></i><b>4.7</b> Functions</a>
<ul>
<li class="chapter" data-level="4.7.1" data-path="prog-basics.html"><a href="prog-basics.html#prog-dry"><i class="fa fa-check"></i><b>4.7.1</b> DRY: Don’t Repeat Yourself</a></li>
<li class="chapter" data-level="4.7.2" data-path="prog-basics.html"><a href="prog-basics.html#writing-your-own-functions"><i class="fa fa-check"></i><b>4.7.2</b> Writing your own functions</a></li>
<li class="chapter" data-level="4.7.3" data-path="prog-basics.html"><a href="prog-basics.html#scope"><i class="fa fa-check"></i><b>4.7.3</b> Scope</a></li>
</ul></li>
<li class="chapter" data-level="4.8" data-path="prog-basics.html"><a href="prog-basics.html#iteration"><i class="fa fa-check"></i><b>4.8</b> Iteration</a>
<ul>
<li class="chapter" data-level="4.8.1" data-path="prog-basics.html"><a href="prog-basics.html#vectorized-operations"><i class="fa fa-check"></i><b>4.8.1</b> Vectorized operations</a></li>
<li class="chapter" data-level="4.8.2" data-path="prog-basics.html"><a href="prog-basics.html#functional-programming"><i class="fa fa-check"></i><b>4.8.2</b> Functional programming</a></li>
<li class="chapter" data-level="4.8.3" data-path="prog-basics.html"><a href="prog-basics.html#apply-and-friends"><i class="fa fa-check"></i><b>4.8.3</b> <code>apply()</code> and friends</a></li>
</ul></li>
<li class="chapter" data-level="4.9" data-path="prog-basics.html"><a href="prog-basics.html#installing-packages"><i class="fa fa-check"></i><b>4.9</b> Installing Packages</a></li>
<li class="chapter" data-level="4.10" data-path="prog-basics.html"><a href="prog-basics.html#prog-debugging"><i class="fa fa-check"></i><b>4.10</b> Troubleshooting and Debugging</a></li>
<li class="chapter" data-level="4.11" data-path="prog-basics.html"><a href="prog-basics.html#prog-style"><i class="fa fa-check"></i><b>4.11</b> Coding Style and Conventions</a>
<ul>
<li class="chapter" data-level="4.11.1" data-path="prog-basics.html"><a href="prog-basics.html#is-my-code-correct"><i class="fa fa-check"></i><b>4.11.1</b> Is my code correct?</a></li>
<li class="chapter" data-level="4.11.2" data-path="prog-basics.html"><a href="prog-basics.html#does-my-code-follow-the-dry-principle"><i class="fa fa-check"></i><b>4.11.2</b> Does my code follow the <span>DRY principle</span>?</a></li>
<li class="chapter" data-level="4.11.3" data-path="prog-basics.html"><a href="prog-basics.html#did-i-choose-concise-but-descriptive-variable-and-function-names"><i class="fa fa-check"></i><b>4.11.3</b> Did I choose concise but descriptive variable and function names?</a></li>
<li class="chapter" data-level="4.11.4" data-path="prog-basics.html"><a href="prog-basics.html#did-i-use-indentation-and-naming-conventions-consistently-throughout-my-code"><i class="fa fa-check"></i><b>4.11.4</b> Did I use indentation and naming conventions consistently throughout my code?</a></li>
<li class="chapter" data-level="4.11.5" data-path="prog-basics.html"><a href="prog-basics.html#did-i-write-comments-especially-when-what-the-code-does-is-not-obvious"><i class="fa fa-check"></i><b>4.11.5</b> Did I write comments, especially when what the code does is not obvious?</a></li>
<li class="chapter" data-level="4.11.6" data-path="prog-basics.html"><a href="prog-basics.html#how-easy-would-it-be-for-someone-else-to-understand-my-code"><i class="fa fa-check"></i><b>4.11.6</b> How easy would it be for someone else to understand my code?</a></li>
<li class="chapter" data-level="4.11.7" data-path="prog-basics.html"><a href="prog-basics.html#is-my-code-easy-to-maintainchange"><i class="fa fa-check"></i><b>4.11.7</b> Is my code easy to maintain/change?</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="5" data-path="data-wrangling.html"><a href="data-wrangling.html"><i class="fa fa-check"></i><b>5</b> Data Wrangling</a>
<ul>
<li class="chapter" data-level="5.1" data-path="data-wrangling.html"><a href="data-wrangling.html#the-tidyverse"><i class="fa fa-check"></i><b>5.1</b> The Tidyverse</a></li>
<li class="chapter" data-level="5.2" data-path="data-wrangling.html"><a href="data-wrangling.html#dw-basics"><i class="fa fa-check"></i><b>5.2</b> Tidyverse Basics</a></li>
<li class="chapter" data-level="5.3" data-path="data-wrangling.html"><a href="data-wrangling.html#importing-data"><i class="fa fa-check"></i><b>5.3</b> Importing Data</a></li>
<li class="chapter" data-level="5.4" data-path="data-wrangling.html"><a href="data-wrangling.html#data-tibble"><i class="fa fa-check"></i><b>5.4</b> The tibble</a></li>
<li class="chapter" data-level="5.5" data-path="data-wrangling.html"><a href="data-wrangling.html#pipes"><i class="fa fa-check"></i><b>5.5</b> pipes</a></li>
<li class="chapter" data-level="5.6" data-path="data-wrangling.html"><a href="data-wrangling.html#tidying-data"><i class="fa fa-check"></i><b>5.6</b> Tidying Data</a></li>
<li class="chapter" data-level="5.7" data-path="data-wrangling.html"><a href="data-wrangling.html#arranging-data"><i class="fa fa-check"></i><b>5.7</b> Arranging Data</a>
<ul>
<li class="chapter" data-level="5.7.1" data-path="data-wrangling.html"><a href="data-wrangling.html#dplyrmutate---create-new-columns-using-other-columns"><i class="fa fa-check"></i><b>5.7.1</b> <code>dplyr::mutate()</code> - Create new columns using other columns</a></li>
<li class="chapter" data-level="5.7.2" data-path="data-wrangling.html"><a href="data-wrangling.html#stringr---working-with-character-values"><i class="fa fa-check"></i><b>5.7.2</b> <code>stringr</code> - Working with character values</a></li>
<li class="chapter" data-level="5.7.3" data-path="data-wrangling.html"><a href="data-wrangling.html#dplyrselect---subset-columns-by-name"><i class="fa fa-check"></i><b>5.7.3</b> <code>dplyr::select()</code> - Subset Columns by Name</a></li>
<li class="chapter" data-level="5.7.4" data-path="data-wrangling.html"><a href="data-wrangling.html#dplyrfilter---pick-rows-out-of-a-data-set"><i class="fa fa-check"></i><b>5.7.4</b> <code>dplyr::filter()</code> - Pick rows out of a data set</a></li>
<li class="chapter" data-level="5.7.5" data-path="data-wrangling.html"><a href="data-wrangling.html#dplyrarrange---order-rows-based-on-their-values"><i class="fa fa-check"></i><b>5.7.5</b> <code>dplyr::arrange()</code> - Order rows based on their values</a></li>
<li class="chapter" data-level="5.7.6" data-path="data-wrangling.html"><a href="data-wrangling.html#putting-it-all-together"><i class="fa fa-check"></i><b>5.7.6</b> Putting it all together</a></li>
</ul></li>
<li class="chapter" data-level="5.8" data-path="data-wrangling.html"><a href="data-wrangling.html#grouping-data"><i class="fa fa-check"></i><b>5.8</b> Grouping Data</a></li>
<li class="chapter" data-level="5.9" data-path="data-wrangling.html"><a href="data-wrangling.html#relational-data"><i class="fa fa-check"></i><b>5.9</b> Relational Data</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="data-science.html"><a href="data-science.html"><i class="fa fa-check"></i><b>6</b> Data Science</a>
<ul>
<li class="chapter" data-level="6.1" data-path="data-science.html"><a href="data-science.html#ds-dists"><i class="fa fa-check"></i><b>6.1</b> Distributions &amp; Tests</a></li>
<li class="chapter" data-level="6.2" data-path="data-science.html"><a href="data-science.html#ds-summ"><i class="fa fa-check"></i><b>6.2</b> Data summarization &amp; cleaning</a></li>
<li class="chapter" data-level="6.3" data-path="data-science.html"><a href="data-science.html#ds-clust"><i class="fa fa-check"></i><b>6.3</b> Clustering</a></li>
<li class="chapter" data-level="6.4" data-path="data-science.html"><a href="data-science.html#ds-model"><i class="fa fa-check"></i><b>6.4</b> Data Modeling</a></li>
<li class="chapter" data-level="6.5" data-path="data-science.html"><a href="data-science.html#ds-networks"><i class="fa fa-check"></i><b>6.5</b> Network Analysis</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html"><i class="fa fa-check"></i><b>7</b> Biology &amp; Bioinformatics</a>
<ul>
<li class="chapter" data-level="7.1" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#bio-r"><i class="fa fa-check"></i><b>7.1</b> R in Biology</a></li>
<li class="chapter" data-level="7.2" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#rectangular-data-in-biology"><i class="fa fa-check"></i><b>7.2</b> Rectangular Data in Biology</a></li>
<li class="chapter" data-level="7.3" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#bio-bioconductor"><i class="fa fa-check"></i><b>7.3</b> Bioconductor</a></li>
<li class="chapter" data-level="7.4" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#bio-geneexp"><i class="fa fa-check"></i><b>7.4</b> Gene Expression</a></li>
<li class="chapter" data-level="7.5" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#bio-micro"><i class="fa fa-check"></i><b>7.5</b> Differential Expression: Microarrays</a></li>
<li class="chapter" data-level="7.6" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#bio-gsea"><i class="fa fa-check"></i><b>7.6</b> Gene Set Enrichment Analysis</a></li>
<li class="chapter" data-level="7.7" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#bio-hts"><i class="fa fa-check"></i><b>7.7</b> High Throughput Sequencing</a></li>
<li class="chapter" data-level="7.8" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#bio-rnaseq"><i class="fa fa-check"></i><b>7.8</b> Differential Expression: RNASeq</a></li>
<li class="chapter" data-level="7.9" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#bio-networks"><i class="fa fa-check"></i><b>7.9</b> Biological Network Analysis</a></li>
<li class="chapter" data-level="7.10" data-path="biology-bioinformatics.html"><a href="biology-bioinformatics.html#bio-scseq"><i class="fa fa-check"></i><b>7.10</b> Single Cell Sequencing Analysis</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="data-visualization.html"><a href="data-visualization.html"><i class="fa fa-check"></i><b>8</b> Data Visualization</a>
<ul>
<li class="chapter" data-level="8.1" data-path="data-visualization.html"><a href="data-visualization.html#dv-gg"><i class="fa fa-check"></i><b>8.1</b> Grammar of Graphics</a></li>
<li class="chapter" data-level="8.2" data-path="data-visualization.html"><a href="data-visualization.html#plotting-one-dimension"><i class="fa fa-check"></i><b>8.2</b> Plotting One Dimension</a></li>
<li class="chapter" data-level="8.3" data-path="data-visualization.html"><a href="data-visualization.html#plotting-two-or-more-dimensions"><i class="fa fa-check"></i><b>8.3</b> Plotting Two or More Dimensions</a></li>
<li class="chapter" data-level="8.4" data-path="data-visualization.html"><a href="data-visualization.html#visualizing-distributions"><i class="fa fa-check"></i><b>8.4</b> Visualizing Distributions</a></li>
<li class="chapter" data-level="8.5" data-path="data-visualization.html"><a href="data-visualization.html#other-kind-of-plots"><i class="fa fa-check"></i><b>8.5</b> Other Kind of Plots</a></li>
<li class="chapter" data-level="8.6" data-path="data-visualization.html"><a href="data-visualization.html#ds-network-viz"><i class="fa fa-check"></i><b>8.6</b> Network Viz</a></li>
<li class="chapter" data-level="8.7" data-path="data-visualization.html"><a href="data-visualization.html#tips-and-tricks"><i class="fa fa-check"></i><b>8.7</b> Tips and Tricks</a></li>
<li class="chapter" data-level="8.8" data-path="data-visualization.html"><a href="data-visualization.html#multiple-plots"><i class="fa fa-check"></i><b>8.8</b> Multiple Plots</a></li>
<li class="chapter" data-level="8.9" data-path="data-visualization.html"><a href="data-visualization.html#dv-resp"><i class="fa fa-check"></i><b>8.9</b> Responsible Plotting</a></li>
<li class="chapter" data-level="8.10" data-path="data-visualization.html"><a href="data-visualization.html#publication-ready-plots"><i class="fa fa-check"></i><b>8.10</b> Publication Ready Plots</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="engineering.html"><a href="engineering.html"><i class="fa fa-check"></i><b>9</b> EngineeRing</a>
<ul>
<li class="chapter" data-level="9.1" data-path="engineering.html"><a href="engineering.html#unit-testing"><i class="fa fa-check"></i><b>9.1</b> Unit Testing</a></li>
<li class="chapter" data-level="9.2" data-path="engineering.html"><a href="engineering.html#toolification"><i class="fa fa-check"></i><b>9.2</b> Toolification</a></li>
<li class="chapter" data-level="9.3" data-path="engineering.html"><a href="engineering.html#pipelines-workflows"><i class="fa fa-check"></i><b>9.3</b> Pipelines &amp; Workflows</a></li>
<li class="chapter" data-level="9.4" data-path="engineering.html"><a href="engineering.html#parallel-processing"><i class="fa fa-check"></i><b>9.4</b> Parallel Processing</a></li>
<li class="chapter" data-level="9.5" data-path="engineering.html"><a href="engineering.html#r-packages"><i class="fa fa-check"></i><b>9.5</b> R Packages</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="rshiny.html"><a href="rshiny.html"><i class="fa fa-check"></i><b>10</b> RShiny</a>
<ul>
<li class="chapter" data-level="10.1" data-path="rshiny.html"><a href="rshiny.html#overview"><i class="fa fa-check"></i><b>10.1</b> Overview</a></li>
<li class="chapter" data-level="10.2" data-path="rshiny.html"><a href="rshiny.html#learning-objectives"><i class="fa fa-check"></i><b>10.2</b> Learning Objectives</a></li>
<li class="chapter" data-level="10.3" data-path="rshiny.html"><a href="rshiny.html#skill-list"><i class="fa fa-check"></i><b>10.3</b> Skill List</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="communicating-with-r.html"><a href="communicating-with-r.html"><i class="fa fa-check"></i><b>11</b> Communicating with R</a>
<ul>
<li class="chapter" data-level="11.1" data-path="communicating-with-r.html"><a href="communicating-with-r.html#rmarkdown-knitr"><i class="fa fa-check"></i><b>11.1</b> RMarkdown &amp; knitr</a></li>
<li class="chapter" data-level="11.2" data-path="communicating-with-r.html"><a href="communicating-with-r.html#bookdown"><i class="fa fa-check"></i><b>11.2</b> bookdown</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="contribution-guide.html"><a href="contribution-guide.html"><i class="fa fa-check"></i><b>12</b> Contribution Guide</a>
<ul>
<li class="chapter" data-level="12.1" data-path="contribution-guide.html"><a href="contribution-guide.html#custom-blocks"><i class="fa fa-check"></i><b>12.1</b> Custom blocks</a></li>
</ul></li>
<li class="part"><span><b>I Assignments</b></span></li>
<li class="chapter" data-level="" data-path="assignment-format.html"><a href="assignment-format.html"><i class="fa fa-check"></i>Assignment Format</a></li>
<li class="chapter" data-level="" data-path="starting-an-assignment.html"><a href="starting-an-assignment.html"><i class="fa fa-check"></i>Starting an Assignment</a>
<ul>
<li><a href="starting-an-assignment.html#starting-your-own-git-repository-and-cloning-it">Starting your own <code>git</code> repository and cloning it</a>
<ul>
<li class="chapter" data-level="" data-path="starting-an-assignment.html"><a href="starting-an-assignment.html#contents"><i class="fa fa-check"></i>Contents</a></li>
<li class="chapter" data-level="" data-path="starting-an-assignment.html"><a href="starting-an-assignment.html#uhhh-various-github-classroom-details-idk.-todo-or-whatever."><i class="fa fa-check"></i>Uhhh various github classroom details, idk. TODO: or whatever.</a></li>
<li class="chapter" data-level="12.1.1" data-path="starting-an-assignment.html"><a href="starting-an-assignment.html#committing-and-pushing-with-r-and-without"><i class="fa fa-check"></i><b>12.1.1</b> Committing and Pushing with R (and without)</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="assignment-2.html"><a href="assignment-2.html"><i class="fa fa-check"></i>Assignment 2</a>
<ul>
<li class="chapter" data-level="" data-path="assignment-2.html"><a href="assignment-2.html#problem-statement"><i class="fa fa-check"></i>Problem Statement</a></li>
<li class="chapter" data-level="" data-path="assignment-2.html"><a href="assignment-2.html#learning-objectives-1"><i class="fa fa-check"></i>Learning Objectives</a></li>
<li class="chapter" data-level="" data-path="assignment-2.html"><a href="assignment-2.html#skill-list-1"><i class="fa fa-check"></i>Skill List</a></li>
<li class="chapter" data-level="" data-path="assignment-2.html"><a href="assignment-2.html#instructions"><i class="fa fa-check"></i>Instructions</a></li>
<li class="chapter" data-level="" data-path="assignment-2.html"><a href="assignment-2.html#function-details"><i class="fa fa-check"></i>Function Details</a>
<ul>
<li class="chapter" data-level="" data-path="assignment-2.html"><a href="assignment-2.html#bioconductor"><i class="fa fa-check"></i>1. Bioconductor</a></li>
<li><a href="assignment-2.html#load_expression">2. <code>load_expression()</code></a></li>
<li><a href="assignment-2.html#filter_15">3. <code>filter_15()</code></a></li>
<li><a href="assignment-2.html#affy_to_hgnc">4. <code>affy_to_hgnc()</code></a></li>
<li><a href="assignment-2.html#reduce_data">5. <code>reduce_data()</code></a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="assignment-5.html"><a href="assignment-5.html"><i class="fa fa-check"></i>Assignment 5</a>
<ul>
<li class="chapter" data-level="" data-path="assignment-5.html"><a href="assignment-5.html#problem-statement-1"><i class="fa fa-check"></i>Problem Statement</a></li>
<li class="chapter" data-level="" data-path="assignment-5.html"><a href="assignment-5.html#learning-objectives-2"><i class="fa fa-check"></i>Learning Objectives</a></li>
<li class="chapter" data-level="" data-path="assignment-5.html"><a href="assignment-5.html#skill-list-2"><i class="fa fa-check"></i>Skill List</a></li>
<li class="chapter" data-level="12.2" data-path="assignment-5.html"><a href="assignment-5.html#deseq2-background"><i class="fa fa-check"></i><b>12.2</b> DESeq2 Background</a></li>
<li class="chapter" data-level="12.3" data-path="assignment-5.html"><a href="assignment-5.html#generating-a-counts-matrix"><i class="fa fa-check"></i><b>12.3</b> Generating a counts matrix</a></li>
<li class="chapter" data-level="12.4" data-path="assignment-5.html"><a href="assignment-5.html#prefiltering-counts-matrix"><i class="fa fa-check"></i><b>12.4</b> Prefiltering Counts matrix</a></li>
<li class="chapter" data-level="12.5" data-path="assignment-5.html"><a href="assignment-5.html#median-of-ratios-normalization"><i class="fa fa-check"></i><b>12.5</b> Median-of-ratios normalization</a></li>
<li class="chapter" data-level="12.6" data-path="assignment-5.html"><a href="assignment-5.html#deseq2-preparation"><i class="fa fa-check"></i><b>12.6</b> DESeq2 preparation</a></li>
<li class="chapter" data-level="12.7" data-path="assignment-5.html"><a href="assignment-5.html#omeara-et-al.-transcriptional-reversion-of-cardiac-myocyte-fate-during-mammalian-cardiac-regeneration.-circ-res.-feb-2015.-pmid-25477501l"><i class="fa fa-check"></i><b>12.7</b> O’Meara et al. Transcriptional Reversion of Cardiac Myocyte Fate During Mammalian Cardiac Regeneration. Circ Res. Feb 2015. PMID: 25477501l</a></li>
<li class="chapter" data-level="12.8" data-path="assignment-5.html"><a href="assignment-5.html#factor-levels"><i class="fa fa-check"></i><b>12.8</b> Factor Levels</a></li>
<li class="chapter" data-level="12.9" data-path="assignment-5.html"><a href="assignment-5.html#performing-de-analysis"><i class="fa fa-check"></i><b>12.9</b> Performing DE Analysis</a></li>
<li class="chapter" data-level="12.10" data-path="assignment-5.html"><a href="assignment-5.html#p-value-distribution-and-multiple-test-correction"><i class="fa fa-check"></i><b>12.10</b> P-value distribution and multiple test correction</a></li>
<li class="chapter" data-level="12.11" data-path="assignment-5.html"><a href="assignment-5.html#choice-of-fdr-cutoff-depends-on-cost"><i class="fa fa-check"></i><b>12.11</b> Choice of FDR cutoff depends on cost</a></li>
<li class="chapter" data-level="12.12" data-path="assignment-5.html"><a href="assignment-5.html#histogram-of-log-fold-changes"><i class="fa fa-check"></i><b>12.12</b> Histogram of Log Fold Changes</a></li>
<li class="chapter" data-level="12.13" data-path="assignment-5.html"><a href="assignment-5.html#make-a-volcano-plot"><i class="fa fa-check"></i><b>12.13</b> Make a volcano plot</a></li>
<li class="chapter" data-level="12.14" data-path="assignment-5.html"><a href="assignment-5.html#a-warning-about-post-hoc-filtering-by-fold-change"><i class="fa fa-check"></i><b>12.14</b> A warning about <em>post hoc</em> filtering by fold change</a></li>
<li class="chapter" data-level="12.15" data-path="assignment-5.html"><a href="assignment-5.html#visualize-normalized-counts-for-top-ten-de-genes"><i class="fa fa-check"></i><b>12.15</b> Visualize normalized counts for top ten DE genes</a></li>
<li class="chapter" data-level="12.16" data-path="assignment-5.html"><a href="assignment-5.html#time-series-with-deseq2-lrt"><i class="fa fa-check"></i><b>12.16</b> Time series with DESeq2 LRT</a></li>
<li class="chapter" data-level="12.17" data-path="assignment-5.html"><a href="assignment-5.html#plotting-the-normalized-counts-over-time-for-the-top-de-gene-from-the-lrt-analysis"><i class="fa fa-check"></i><b>12.17</b> Plotting the normalized counts over time for the top DE gene from the LRT analysis</a></li>
<li class="chapter" data-level="12.18" data-path="assignment-5.html"><a href="assignment-5.html#extra-exercises"><i class="fa fa-check"></i><b>12.18</b> Extra exercises</a></li>
<li class="chapter" data-level="12.19" data-path="assignment-5.html"><a href="assignment-5.html#references"><i class="fa fa-check"></i><b>12.19</b> References</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="assignment-6.html"><a href="assignment-6.html"><i class="fa fa-check"></i>Assignment 6</a>
<ul>
<li class="chapter" data-level="" data-path="assignment-6.html"><a href="assignment-6.html#problem-statement-2"><i class="fa fa-check"></i>Problem Statement</a></li>
<li class="chapter" data-level="" data-path="assignment-6.html"><a href="assignment-6.html#learning-objectives-3"><i class="fa fa-check"></i>Learning Objectives</a></li>
<li class="chapter" data-level="" data-path="assignment-6.html"><a href="assignment-6.html#skill-list-3"><i class="fa fa-check"></i>Skill List</a></li>
<li class="chapter" data-level="" data-path="assignment-6.html"><a href="assignment-6.html#instructions-1"><i class="fa fa-check"></i>Instructions</a></li>
</ul></li>
<li class="appendix"><span><b>Appendix</b></span></li>
<li class="chapter" data-level="A" data-path="class-outlines.html"><a href="class-outlines.html"><i class="fa fa-check"></i><b>A</b> Class Outlines</a>
<ul>
<li class="chapter" data-level="A.1" data-path="class-outlines.html"><a href="class-outlines.html#week-1"><i class="fa fa-check"></i><b>A.1</b> Week 1</a></li>
<li class="chapter" data-level="A.2" data-path="class-outlines.html"><a href="class-outlines.html#week-2"><i class="fa fa-check"></i><b>A.2</b> Week 2</a></li>
</ul></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">BF591 - R for Biological Sciences</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="preliminaries" class="section level1" number="3">
<h1><span class="header-section-number">3</span> Preliminaries</h1>
<div id="prelim-r" class="section level2" number="3.1">
<h2><span class="header-section-number">3.1</span> The R Language</h2>
<p>R is a free programming language and environment where that language can be
used. More specifically, R is a <em>statistical</em> programming language, designed
for the express and exclusive purpose of conducting statistical analyses and
visualizing data. Said differently, <em>R is not a general purpose programming
language</em> unlike other languages such as python, Java, C, etc. As such, the
language’s real strengths are in the manipulation, analysis, and visualization
of data and statistical procedures, though it is often used for other purposes
(for example, web applications with <a href="rshiny.html#rshiny">RShiny</a>, or writing books like this one
with <a href="https://bookdown.org/yihui/bookdown/">bookdown</a>). You may download R for
free from the <a href="https://cloud.r-project.org/">Comprehensive R Archive Network</a>.</p>
<div class="box note">
<p>The effective biological analysis practitioner knows how to use multiple tools
for their appropriate purposes. The most common programming languages in this
field are <a href="https://www.python.org">python</a>, R, and scripting languages like
<a href="https://www.gnu.org/software/bash/">The Bourne Again Shell - bash</a>. While it is
beyond the scope of this book to cover which tools are best used where, R is
appropriate wherever data analysis and visualization are needed. Any operations
that do not involve these aspects (e.g. manipulating text files, programming web
servers, etc) are likely more suitable for other languages and software.</p>
</div>
<div class="box readmore">
<ul>
<li><a href="https://www.r-project.org/">R Project Home Page</a></li>
<li><a href="https://rstudio-education.github.io/hopr/starting.html">Hands-On Programming with R - Installing R and RStudio</a></li>
<li><a href="https://r4ds.had.co.nz/introduction.html#r">Section on R from R for Data Science</a></li>
</ul>
</div>
</div>
<div id="prelim-rstudio" class="section level2" number="3.2">
<h2><span class="header-section-number">3.2</span> RStudio</h2>
<p>This course assumes the learner is using the <a href="https://www.rstudio.com/">RStudio</a>
software platform for all analyses, unless otherwise noted. RStudio is a freely
available and fully featured integrated development environment (IDE) for R, and
has many convenient capabilities when learning R. RStudio may be downloaded and
installed for free from the site above.</p>
<p>All the examples and instructions in this book assume you have <a href="preliminaries.html#prelim-r">installed
R</a> are using <a href="preliminaries.html#prelim-rstudio">RStudio</a>. Be sure to <a href="preliminaries.html#prelim-rstudio">turn off automatic environment
saving in RStudio</a>! Because this is so important, here it is again:</p>
<div class="box important">
<p>By default, RStudio preserves your R environment when you shut it down and
restores it when you start it again. <strong>This is very bad practice!</strong></p>
<p>The state of your R environment, which includes the values stored in variables,
the R packages loaded, etc. from previously executed code is transient and may
not reflect the results your code produces when run alone.</p>
<p>Open the Tools &gt; Global Options… menu and:</p>
<ol style="list-style-type: decimal">
<li>Uncheck “Restore .RData into workspace at startup”</li>
<li>Set “Save workspace to .RData on exit:” to “Never”</li>
</ol>
<div class="figure">
<img src="rstudio_no_workspace_save.png" style="width:75.0%" alt="" />
<p class="caption">Never save workspace</p>
</div>
<p>The book R for Data Science has an <a href="https://r4ds.had.co.nz/workflow-projects.html#workflow-projects">excellent
chapter</a> on why
this is a problem and how to change the RStudio setting to avoid it.</p>
</div>
<div class="box readmore">
<ul>
<li><a href="https://rstudio-education.github.io/hopr/starting.html">Hands-On Programming with R - Installing R and RStudio</a></li>
<li><a href="https://education.rstudio.com/learn/beginner/">RStudio Education - Beginners Guide</a></li>
<li><a href="https://moderndive.netlify.app/1-getting-started.html">ModernDrive R Series - Getting Started</a></li>
<li><a href="https://r4ds.had.co.nz/workflow-projects.html">R for Data Science - What is real, and where does your analysis “live”?</a></li>
<li><a href="https://rstats.wtf/save-source.html#always-start-r-with-a-blank-slate">What They Forgot To Teach You About R - Always start R with a blank slate</a></li>
<li><a href="https://r4ds.had.co.nz/introduction.html#rstudio">R for Data Science - RStudio</a></li>
</ul>
</div>
</div>
<div id="prog-the-r-script" class="section level2" number="3.3">
<h2><span class="header-section-number">3.3</span> The R Script</h2>
<p>Before we cover the R language itself, we should talk about how you should run
your code and where it should live. <a href="preliminaries.html#prelim-r">As mentioned</a>, R is both a
programming language and an environment where you can run code written in that
language. The environment is a program (confusingly also called R) that allows
you to interact with it and run simple lines of code one at a time. This
environment is very useful for learning how the language works and
troubleshooting, but it is not suitable for recording and running large, complex
analyses that require many lines of code. Therefore, <strong>all important R code
should be written and saved in a file before you run it!</strong> The code may not be
correct, and the interactive R environment is helpful for debugging and
troubleshooting, but as soon as the code works it should be saved to the file
and rerun from there.</p>
<p>With this in mind, the basic unit of an R analysis is the <strong>R script</strong>. An R
script is a file that contains lines of R code that run sequentially as a unit
to complete one or more tasks. Every R script file has a name, which you choose
and should be descriptive but concise about what the script does; <code>script.R</code>,
<code>do_it.R</code>, and
<code>a_script_that_implements_my_very_cool_but_complicated_analysis_and_plots.R</code>
are generally poor names for scripts, whereas <code>analyze_gene_expression.R</code> might
be more suitable.</p>
<p>In RStudio, you can create a new script file in the current directory using the
<code>File -&gt; New File -&gt; R Script</code> menu item or the new R Script button at the top
of the screen:</p>
<div class="figure">
<img src="rstudio_new_r_script.png" style="width:25.0%" alt="" />
<p class="caption">New R Script</p>
</div>
<p>Your RStudio configuration should now enable you to write R code into the
(currently unsaved) file in the top left portion of the screen (labeled in the
figure as “File Editor”).</p>
<div class="figure">
<img src="rstudio_basic_interface.png" style="width:75.0%" alt="" />
<p class="caption">Basic RStudio Interface</p>
</div>
<p>You are now nearly ready to start coding in R!</p>
<div class="box readmore">
<ul>
<li><a href="https://speakerdeck.com/jennybc/how-to-name-files">How to name files</a> Some
useful and advanced tips on how to name files</li>
</ul>
</div>
</div>
<div id="prog-workflow" class="section level2" number="3.4">
<h2><span class="header-section-number">3.4</span> The Scripting Workflow</h2>
<p>But hold on, we’re still not quite ready to start coding. As mentioned above,
<strong>all important R code should be written and saved in a file before you run
it!</strong> Your scripts will very quickly contain many lines of code that are meant
to be run in sequential order. While developing your code it is very helpful to
run each individual line separately, building up your script incrementally over
time. To illustrate how to do this, we will begin with a simple R code that
stores the result of an arithmetic expression to a new variable:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="preliminaries.html#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="co"># stores the result of 1+1 into a variable named &#39;a&#39;</span></span>
<span id="cb1-2"><a href="preliminaries.html#cb1-2" aria-hidden="true" tabindex="-1"></a>a <span class="ot">&lt;-</span> <span class="dv">1</span><span class="sc">+</span><span class="dv">1</span></span></code></pre></div>
<p>The concepts in this line of code will be covered in greater depth later, but
for now an intuitive understanding will suffice to explain the development
workflow in RStudio.</p>
<p>When developing, this is the suggested sequence of operations:</p>
<ol style="list-style-type: decimal">
<li>Save your file (naming if necessary on the first save) with <code>Ctrl-s</code> on
Windows or <code>Cmd-s</code> on Mac</li>
<li>Execute the line or lines of code in your script you wish to evaluate using
<code>Ctrl-Enter</code> on Windows or <code>Cmd-Enter</code> on Mac. By default only the line with
the cursor is executed; you may click and drag with the mouse to select
multiple lines to execute if needed. <em>NB: you can press the up arrow key to
recall previously run commands on the console.</em></li>
<li>The executed code will be evaluated in the Console window, where you may
inspect the result and modify the code if necessary.</li>
<li>You may inspect the definitions of any variables you have declared in the
Environment tab at the upper right.</li>
<li>When you have verified that the code you executed does what you intend,
ensure the code in the file you started from is updated appropriately.</li>
<li>Go to step 1</li>
</ol>
<p>The above steps are depicted in the following figure:</p>
<div class="figure">
<img src="rstudio_workflow_1.png" style="width:75.0%" alt="" />
<p class="caption">RStudio workflow</p>
</div>
<p>Over time, you will gain comfort with this workflow and become more flexible
with how you use RStudio.</p>
<div class="box important">
<p>If you followed the <a href="prog-basics.html#before-you-begin">instructions</a> above and prevented RStudio
from saving your environment when you exit the program (which you should! Did I
mention you should?!), none of the results of code you previously ran will be
available upon starting a new RStudio session. Although this may seem
inconvenient, this is an excellent opportunity to verify that your script in its
current state does what you intend for it to do.</p>
<p><strong>It is extremely easy to ask R to do things you don’t mean for it to do!</strong></p>
<p>Rerunning your scripts from the beginning in a new RStudio session is an
excellent way to guard against this kind of error. This short page summarizes
this very well, you should read it:</p>
<ul>
<li><a href="https://rstats.wtf/save-source.html#always-start-r-with-a-blank-slate">What They Forgot To Teach You About R - Always start R with a blank
slate</a></li>
</ul>
</div>
<div class="box readmore">
<ul>
<li><a href="https://r4ds.had.co.nz/workflow-scripts.html">R for Data Science - Workflow: scripts</a></li>
<li><a href="https://www.rstudio.com/resources/cheatsheets/">RStudio IDE cheatsheet</a>
(scroll down the page to find the cheatsheet entitled “RStudio IDE cheatsheet”)</li>
</ul>
</div>
</div>
<div id="prelim-git" class="section level2" number="3.5">
<h2><span class="header-section-number">3.5</span> git + github</h2>
<div id="motivation" class="section level3" number="3.5.1">
<h3><span class="header-section-number">3.5.1</span> Motivation</h3>
<p>Biological analysis entails writing code, which changes over time as you
develop it, gain insight from your data, and identify new questions and
hypotheses. A common pattern when developing scripts is to make copies of older
code files to preserve them before making new changes to them. While it is a
good idea to maintain a record of all the code you previously ran, over time
this practice often leads to disorganized, cluttered, untidy analysis
directories.</p>
<p>For example, say you are working on a script named <code>my_R_script.R</code> and decide
you want to add a new analysis that substantially changes the code. You might be
tempted to make a copy of the current version of the code into a new file
named <code>my_R_script_v2.R</code> that you then make changes to, leaving your original
script intact and untouched going forward. You make your changes to your new
script, produce some stunning and fascinating plots, present the analysis at a
group meeting, only to discover later there was a critical bug in your code that
made the plots misleading and requires substantial redevelopment.</p>
<div class="box note">
<p><strong>Bugs happen.</strong> There are two types of bugs:</p>
<ol style="list-style-type: decimal">
<li>Syntax bugs: bugs due to incorrect language usage, which R will tell you
about and can (usually) be easily identified and fixed</li>
<li>Logic bugs: the code you write is syntactically correct, but does something
other than what you intend</li>
</ol>
<p><strong>Bugs are normal.</strong> The scenario described above, where you present results only to
discover your code wasn’t doing what you thought it was doing, is extremely
common and <em>it will happen to you</em>. This is normal, and finding a bug in your
code does not mean you are a bad programmer.</p>
</div>
<p>Rather than edit your version 2 of your script directly, you decide it is
sensible to copy the file to <code>my_R_script_v2_BAD.R</code> and edit the version 2
script to fix the bug. You are satisfied with your new version 2 script, and so
make a new copy <code>my_R_script_v2_final.R</code>. Upon review of your analysis, you
are asked to implement new changes to the script based on reviewer feedback. You
make a new copy of your script to <code>my_R_script_v2_final_revision.R</code> and make
the requested changes. Perhaps now your script is final, but in your directory
you now have five different versions of your analysis:</p>
<pre><code>my_R_script.R
my_R_script_v2.R
my_R_script_v2_BAD.R
my_R_script_v2_final.R
my_R_script_v2_final_revision.R</code></pre>
<p>When you write your code, you may know which scripts are which, and if you
follow <a href="prog-basics.html#prog-style">good programming practice</a> and carefully commented your
code you or your successors may be able to sleuth what was done. However, as
time passes, the intimate knowledge you thought you had about your code will be
replaced by other more immediately important things; eventually you may not even
understand or even recognize your own code, let alone someone else trying to
understand it. Not an ideal situation in any case. A better solution involves
recording changes to code over time in such a way that you can recover old code
if needed, but don’t clutter your analytical workspace with unneeded files.
<code>git</code> provides an efficient solution to this problem.</p>
</div>
<div id="git" class="section level3" number="3.5.2">
<h3><span class="header-section-number">3.5.2</span> git</h3>
<p><a href="https://git-scm.com/">git</a> is a free, open source <a href="https://en.wikipedia.org/wiki/Version_control">version
control</a> software program.
Version control software is used to track and record changes to code over time,
potentially by many developers working on the same software project
concurrently from different parts of the world. The base git software can be
used on the command line, or with <a href="https://git-scm.com/downloads/guis">graphical user interface
applications</a> for popular operating systems.</p>
<p>There are many excellent tutorials online (some linked below) that teach how to
use <code>git</code> but the basic concepts are described below. The command line
commands are listed, but the same operations apply in the graphical clients.</p>
<ul>
<li>A <strong>repository (or repo)</strong> is a collection of files in a directory that you
have asked git to track (run <code>git init</code> in a new directory)</li>
<li>Each file you wish to track must be explicitly added to the repo (run
<code>git add &lt;filename&gt;</code> from within the git repo directory)</li>
<li>When you modify a tracked file, git will notice those differences and show
them to you with <code>git status</code></li>
<li>You may tell git to track the changes to the explicit files that changed (also
run <code>git add &lt;filename&gt;</code> to record changes)</li>
<li>A set of tracked changes is stored in the repo by making a <strong>commit</strong>.
A commit takes a snapshot of all the tracked files in the repo at the
time the commit is made (run <code>git commit -c &lt;commit message&gt;</code> with a concise
commit message that describes what was done)</li>
<li>Each commit has a date and time associated with it. The files in the
repo can be reset to exactly the state they were in at any commit,
thus preserving all previous versions of code.</li>
</ul>
<p>For the vast majority of use cases, the <code>git init</code>, <code>git status</code>, <code>git add</code>, and <code>git commit</code> operations are all you will need to use git
effectively. Two more commands, <code>git push</code> and <code>git pull</code> are needed when
sharing your code with others as described in the next section.</p>
<div class="box readmore">
<ul>
<li><a href="https://git-scm.com/videos">Official git tutorial videos</a></li>
<li><a href="https://git-scm.com/book/en/v2">Official git book</a></li>
<li><a href="https://gitimmersion.com/">Git Immersion</a> - a guided tour through git
commands</li>
<li><a href="https://www.datacamp.com/courses/introduction-to-git">DataCamp - Git for data scientists</a></li>
</ul>
</div>
</div>
<div id="git-hosting-platforms-github" class="section level3" number="3.5.3">
<h3><span class="header-section-number">3.5.3</span> Git hosting platforms (GitHub)</h3>
<p>The basic git software only works on your local computer with local
repositories. To share this code with others, and receive others’ contributions,
a copy of the repo must be made available in a centralized location that
everyone can access. One such place is <a href="https://github.com">github.com</a>, which
is a free web application that hosts git repos.
<a href="https://bitbucket.org">bitbucket.org</a> is another popular free git repo hosting
service. These two services are practically the same, so we will focus on
GitHub.</p>
<div class="box note">
<p>There is no formal relationship between git and GitHub. git is an open source
software project maintained by hundreds of developers around the world (and is
<a href="https://github.com/git/git">hosted on GitHub</a>). GitHub is an independently
provided web service and application. The only connection between GitHub and git
is that GitHub hosts git repos.</p>
</div>
<p>As with git, there are many excellent tutorials on how to use GitHub, but the
basic concepts are described below.</p>
<ul>
<li>First you must create an account on GitHub if you don’t have one already</li>
<li>Then, create a new repo on GitHub that you wish to contain your code</li>
</ul>
<p>The next step depends on whether you have an existing local repo or not:</p>
<ul>
<li><em>If you do not already have a local git repo:</em> Follow the instructions on
GitHub to <code>clone</code> your GitHub repo and create a local copy that is connected
to the one on GitHub</li>
<li><em>If you already have a local git repo:</em> Follow the instructions on the GitHub
to connect your local repo to the GitHub one (this is called “adding a
remote”)</li>
</ul>
<p>Now, your local repo is connected to the same repo on GitHub, and the changes
you make to your local files can be sent, or <code>push</code>ed to the repo on GitHub:</p>
<ul>
<li>Make changes to your local files, and <code>git add</code> and <code>git commit</code> them as
above</li>
<li>Update the remote repo on GitHub by pushing your local commits with <code>git   push</code></li>
<li>Running <code>git status</code> will indicate whether your local repo is up to date
with your remote GitHub repo</li>
</ul>
<p>When you are working on a team of contributors to a GitHub repo, your local
files will become out of date as others push their changes. To ensure your local
repo is up to date with the GitHub repo, you must <code>pull</code> your changes from
GitHub with <code>git pull</code>.</p>
<div class="box note">
<p>git was designed to automatically combine changes made to a code base by
different developers whenever possible. However, if two people make changes to
the same parts of the same file, git may not be able to resolve those changes
on its own and the developers must communicate and decide what the code should
be. These instances are called <em>merge conflicts</em> and can be challenging to
resolve. Dealing with merge conflicts is beyond the scope of this book, but
some resources are linked below for further reading.</p>
</div>
<p>All the content and code for this book are <a href="https://github.com/bu-bioinfo/r_for_biological_sciences">stored and available on
GitHub</a>, as are the
<a href="#assignments">assignment code templates</a>.</p>
<div class="box readmore">
<ul>
<li><a href="https://docs.github.com/en/get-started/quickstart/hello-world">Official GitHub Tutorial</a></li>
<li><a href="https://www.freecodecamp.org/news/git-and-github-for-beginners/">FreeCodeCamp - Git and GitHub For Beginners</a></li>
<li><a href="https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/addressing-merge-conflicts/about-merge-conflicts">Official GitHub Tutorial on Merge Conflicts</a></li>
</ul>
</div>

</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="data-bio.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="prog-basics.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
