<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta name="viewport" content="width=device-width, initial-scale=1" />
<meta property="og:title" content="7.1 Responsible Plotting | BF591 - R for Biological Sciences" />
<meta property="og:type" content="book" />







<script type="text/x-mathjax-config">
MathJax.Hub.Config({
  TeX: { equationNumbers: { autoNumber: "AMS" } }
});
</script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<meta name="description" content="7.1 Responsible Plotting | BF591 - R for Biological Sciences">

<title>7.1 Responsible Plotting | BF591 - R for Biological Sciences</title>

<script src="libs/header-attrs-2.13/header-attrs.js"></script>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="libs/bootstrap-3.3.5/css/darkly.min.css" rel="stylesheet" />
<script src="libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="libs/navigation-1.1/tabsets.js"></script>


<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>


<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />

<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
/* show arrow before summary tag as in bootstrap
TODO: remove if boostrap in updated in html_document (rmarkdown#1485) */
details > summary {
  display: list-item;
  cursor: pointer;
}
</style>
</head>

<body>

<div class="container-fluid main-container">


<div class="row">
<div class="col-sm-12">
<div id="TOC">
<ul>
<li><a href="index.html#syllabus">Syllabus</a>
<ul>
<li><a href="course-schedule.html#course-schedule">Course Schedule</a></li>
<li><a href="instructors.html#instructors">Instructors</a></li>
<li><a href="office-hours.html#office-hours">Office Hours</a></li>
<li><a href="course-values-and-policies.html#course-values-and-policies">Course Values and Policies</a></li>
</ul></li>
<li><a href="introduction.html#introduction"><span class="toc-section-number">1</span> Introduction</a>
<ul>
<li><a href="who-this-book-is-for.html#who-this-book-is-for"><span class="toc-section-number">1.1</span> Who This Book Is For</a></li>
<li><a href="a-note-about-reinventing-the-wheel.html#a-note-about-reinventing-the-wheel"><span class="toc-section-number">1.2</span> A Note About Reinventing the Wheel</a></li>
<li><a href="sources-and-references.html#sources-and-references"><span class="toc-section-number">1.3</span> Sources and References</a></li>
</ul></li>
<li><a href="data-bio.html#data-bio"><span class="toc-section-number">2</span> Data in Biology</a>
<ul>
<li><a href="a-brief-history-of-data-in-molecular-biology.html#a-brief-history-of-data-in-molecular-biology"><span class="toc-section-number">2.1</span> A Brief History of Data in Molecular Biology</a></li>
<li><a href="biology-as-a-mature-data-science.html#biology-as-a-mature-data-science"><span class="toc-section-number">2.2</span> Biology as a Mature Data Science</a></li>
</ul></li>
<li><a href="preliminaries.html#preliminaries"><span class="toc-section-number">3</span> Preliminaries</a>
<ul>
<li><a href="prelim-r.html#prelim-r"><span class="toc-section-number">3.1</span> The R Language</a></li>
<li><a href="prelim-rstudio.html#prelim-rstudio"><span class="toc-section-number">3.2</span> RStudio</a></li>
<li><a href="prog-the-r-script.html#prog-the-r-script"><span class="toc-section-number">3.3</span> The R Script</a></li>
<li><a href="prog-workflow.html#prog-workflow"><span class="toc-section-number">3.4</span> The Scripting Workflow</a></li>
<li><a href="prelim-git.html#prelim-git"><span class="toc-section-number">3.5</span> git + github</a>
<ul>
<li><a href="prelim-git.html#motivation"><span class="toc-section-number">3.5.1</span> Motivation</a></li>
<li><a href="prelim-git.html#git"><span class="toc-section-number">3.5.2</span> git</a></li>
<li><a href="prelim-git.html#git-hosting-platforms-github"><span class="toc-section-number">3.5.3</span> Git hosting platforms (GitHub)</a></li>
</ul></li>
</ul></li>
<li><a href="prog-basics.html#prog-basics"><span class="toc-section-number">4</span> R Programming</a>
<ul>
<li><a href="before-you-begin.html#before-you-begin"><span class="toc-section-number">4.1</span> Before you begin</a></li>
<li><a href="introduction-1.html#introduction-1"><span class="toc-section-number">4.2</span> Introduction</a></li>
<li><a href="prog-r-syntax.html#prog-r-syntax"><span class="toc-section-number">4.3</span> R Syntax Basics</a></li>
<li><a href="prog-types.html#prog-types"><span class="toc-section-number">4.4</span> Basic Types of Values</a>
<ul>
<li><a href="prog-types.html#factors"><span class="toc-section-number">4.4.1</span> Factors</a></li>
</ul></li>
<li><a href="prog-struct.html#prog-struct"><span class="toc-section-number">4.5</span> Data Structures</a>
<ul>
<li><a href="prog-struct.html#vectors"><span class="toc-section-number">4.5.1</span> Vectors</a></li>
<li><a href="prog-struct.html#matrices"><span class="toc-section-number">4.5.2</span> Matrices</a></li>
<li><a href="prog-struct.html#lists-and-data-frames"><span class="toc-section-number">4.5.3</span> Lists and data frames</a></li>
</ul></li>
<li><a href="logical-tests-and-comparators.html#logical-tests-and-comparators"><span class="toc-section-number">4.6</span> Logical Tests and Comparators</a></li>
<li><a href="functions.html#functions"><span class="toc-section-number">4.7</span> Functions</a>
<ul>
<li><a href="functions.html#prog-dry"><span class="toc-section-number">4.7.1</span> DRY: Don’t Repeat Yourself</a></li>
<li><a href="functions.html#writing-your-own-functions"><span class="toc-section-number">4.7.2</span> Writing your own functions</a></li>
<li><a href="functions.html#scope"><span class="toc-section-number">4.7.3</span> Scope</a></li>
</ul></li>
<li><a href="iteration.html#iteration"><span class="toc-section-number">4.8</span> Iteration</a>
<ul>
<li><a href="iteration.html#vectorized-operations"><span class="toc-section-number">4.8.1</span> Vectorized operations</a></li>
<li><a href="iteration.html#functional-programming"><span class="toc-section-number">4.8.2</span> Functional programming</a></li>
<li><a href="iteration.html#apply-and-friends"><span class="toc-section-number">4.8.3</span> <code>apply()</code> and friends</a></li>
</ul></li>
<li><a href="installing-packages.html#installing-packages"><span class="toc-section-number">4.9</span> Installing Packages</a></li>
<li><a href="saving-and-loading-r-data.html#saving-and-loading-r-data"><span class="toc-section-number">4.10</span> Saving and Loading R Data</a></li>
<li><a href="prog-debugging.html#prog-debugging"><span class="toc-section-number">4.11</span> Troubleshooting and Debugging</a></li>
<li><a href="prog-style.html#prog-style"><span class="toc-section-number">4.12</span> Coding Style and Conventions</a>
<ul>
<li><a href="prog-style.html#is-my-code-correct"><span class="toc-section-number">4.12.1</span> Is my code correct?</a></li>
<li><a href="prog-style.html#does-my-code-follow-the-dry-principle"><span class="toc-section-number">4.12.2</span> Does my code follow the <span>DRY principle</span>?</a></li>
<li><a href="prog-style.html#did-i-choose-concise-but-descriptive-variable-and-function-names"><span class="toc-section-number">4.12.3</span> Did I choose concise but descriptive variable and function names?</a></li>
<li><a href="prog-style.html#did-i-use-indentation-and-naming-conventions-consistently-throughout-my-code"><span class="toc-section-number">4.12.4</span> Did I use indentation and naming conventions consistently throughout my code?</a></li>
<li><a href="prog-style.html#did-i-write-comments-especially-when-what-the-code-does-is-not-obvious"><span class="toc-section-number">4.12.5</span> Did I write comments, especially when what the code does is not obvious?</a></li>
<li><a href="prog-style.html#how-easy-would-it-be-for-someone-else-to-understand-my-code"><span class="toc-section-number">4.12.6</span> How easy would it be for someone else to understand my code?</a></li>
<li><a href="prog-style.html#is-my-code-easy-to-maintainchange"><span class="toc-section-number">4.12.7</span> Is my code easy to maintain/change?</a></li>
<li><a href="prog-style.html#the-styler-package"><span class="toc-section-number">4.12.8</span> The <code>styler</code> package</a></li>
</ul></li>
</ul></li>
<li><a href="data-wrangling.html#data-wrangling"><span class="toc-section-number">5</span> Data Wrangling</a>
<ul>
<li><a href="the-tidyverse.html#the-tidyverse"><span class="toc-section-number">5.1</span> The Tidyverse</a></li>
<li><a href="dw-basics.html#dw-basics"><span class="toc-section-number">5.2</span> Tidyverse Basics</a></li>
<li><a href="importing-data.html#importing-data"><span class="toc-section-number">5.3</span> Importing Data</a></li>
<li><a href="data-tibble.html#data-tibble"><span class="toc-section-number">5.4</span> The tibble</a></li>
<li><a href="tidy-data.html#tidy-data"><span class="toc-section-number">5.5</span> Tidy Data</a></li>
<li><a href="pipes.html#pipes"><span class="toc-section-number">5.6</span> pipes</a></li>
<li><a href="arranging-data.html#arranging-data"><span class="toc-section-number">5.7</span> Arranging Data</a>
<ul>
<li><a href="arranging-data.html#dplyrmutate---create-new-columns-using-other-columns"><span class="toc-section-number">5.7.1</span> <code>dplyr::mutate()</code> - Create new columns using other columns</a></li>
<li><a href="arranging-data.html#stringr---working-with-character-values"><span class="toc-section-number">5.7.2</span> <code>stringr</code> - Working with character values</a></li>
<li><a href="arranging-data.html#dplyrselect---subset-columns-by-name"><span class="toc-section-number">5.7.3</span> <code>dplyr::select()</code> - Subset Columns by Name</a></li>
<li><a href="arranging-data.html#dplyrfilter---pick-rows-out-of-a-data-set"><span class="toc-section-number">5.7.4</span> <code>dplyr::filter()</code> - Pick rows out of a data set</a></li>
<li><a href="arranging-data.html#dplyrarrange---order-rows-based-on-their-values"><span class="toc-section-number">5.7.5</span> <code>dplyr::arrange()</code> - Order rows based on their values</a></li>
<li><a href="arranging-data.html#putting-it-all-together"><span class="toc-section-number">5.7.6</span> Putting it all together</a></li>
</ul></li>
<li><a href="grouping-data.html#grouping-data"><span class="toc-section-number">5.8</span> Grouping Data</a></li>
<li><a href="rearranging-data.html#rearranging-data"><span class="toc-section-number">5.9</span> Rearranging Data</a></li>
<li><a href="relational-data.html#relational-data"><span class="toc-section-number">5.10</span> Relational Data</a>
<ul>
<li><a href="relational-data.html#dealing-with-multiple-matches"><span class="toc-section-number">5.10.1</span> Dealing with multiple matches</a></li>
</ul></li>
</ul></li>
<li><a href="data-science.html#data-science"><span class="toc-section-number">6</span> Data Science</a>
<ul>
<li><a href="data-modeling.html#data-modeling"><span class="toc-section-number">6.1</span> Data Modeling</a>
<ul>
<li><a href="data-modeling.html#a-worked-modeling-example"><span class="toc-section-number">6.1.1</span> A Worked Modeling Example</a></li>
<li><a href="data-modeling.html#data-summarization"><span class="toc-section-number">6.1.2</span> Data Summarization</a></li>
<li><a href="data-modeling.html#linear-models"><span class="toc-section-number">6.1.3</span> Linear Models</a></li>
<li><a href="data-modeling.html#flavors-of-linear-models"><span class="toc-section-number">6.1.4</span> Flavors of Linear Models</a></li>
<li><a href="data-modeling.html#assessing-model-accuracy-."><span class="toc-section-number">6.1.5</span> Assessing Model Accuracy .</a></li>
</ul></li>
<li><a href="statistical-distributions.html#statistical-distributions"><span class="toc-section-number">6.2</span> Statistical Distributions</a>
<ul>
<li><a href="statistical-distributions.html#random-variables"><span class="toc-section-number">6.2.1</span> Random Variables</a></li>
<li><a href="statistical-distributions.html#statistical-distribution-basics"><span class="toc-section-number">6.2.2</span> Statistical Distribution Basics</a></li>
<li><a href="statistical-distributions.html#distributions-in-r"><span class="toc-section-number">6.2.3</span> Distributions in R</a></li>
<li><a href="statistical-distributions.html#discrete-distributions"><span class="toc-section-number">6.2.4</span> Discrete Distributions</a></li>
<li><a href="statistical-distributions.html#continuous-distributions"><span class="toc-section-number">6.2.5</span> Continuous Distributions</a></li>
<li><a href="statistical-distributions.html#empirical-distributions"><span class="toc-section-number">6.2.6</span> Empirical Distributions</a></li>
</ul></li>
<li><a href="statistical-tests.html#statistical-tests"><span class="toc-section-number">6.3</span> Statistical Tests</a>
<ul>
<li><a href="statistical-tests.html#what-is-a-statistical-test"><span class="toc-section-number">6.3.1</span> What is a statistical test</a></li>
<li><a href="statistical-tests.html#common-statistical-tests"><span class="toc-section-number">6.3.2</span> Common Statistical Tests</a></li>
<li><a href="statistical-tests.html#choosing-a-test"><span class="toc-section-number">6.3.3</span> Choosing a Test</a></li>
<li><a href="statistical-tests.html#p-values"><span class="toc-section-number">6.3.4</span> P-values</a></li>
<li><a href="statistical-tests.html#multiple-hypothesis-testing-."><span class="toc-section-number">6.3.5</span> Multiple Hypothesis Testing .</a></li>
<li><a href="statistical-tests.html#statistical-power"><span class="toc-section-number">6.3.6</span> Statistical power</a></li>
</ul></li>
<li><a href="exploratory-data-analysis.html#exploratory-data-analysis"><span class="toc-section-number">6.4</span> Exploratory Data Analysis</a>
<ul>
<li><a href="exploratory-data-analysis.html#principal-component-analysis"><span class="toc-section-number">6.4.1</span> Principal Component Analysis</a></li>
<li><a href="exploratory-data-analysis.html#cluster-analysis"><span class="toc-section-number">6.4.2</span> Cluster Analysis</a></li>
<li><a href="exploratory-data-analysis.html#k-means-."><span class="toc-section-number">6.4.3</span> k-means .</a></li>
<li><a href="exploratory-data-analysis.html#others"><span class="toc-section-number">6.4.4</span> Others</a></li>
</ul></li>
<li><a href="network-analysis-..html#network-analysis-."><span class="toc-section-number">6.5</span> Network Analysis .</a>
<ul>
<li><a href="network-analysis-..html#basic-network-analysis"><span class="toc-section-number">6.5.1</span> Basic network analysis</a></li>
<li><a href="network-analysis-..html#create-your-own"><span class="toc-section-number">6.5.2</span> Create your own</a></li>
</ul></li>
</ul></li>
<li><a href="data-visualization.html#data-visualization"><span class="toc-section-number">7</span> Data Visualization</a>
<ul>
<li><a href="responsible-plotting.html#responsible-plotting"><span class="toc-section-number">7.1</span> Responsible Plotting</a>
<ul>
<li><a href="responsible-plotting.html#visualization-principles"><span class="toc-section-number">7.1.1</span> Visualization Principles</a></li>
<li><a href="responsible-plotting.html#human-visual-perception"><span class="toc-section-number">7.1.2</span> Human Visual Perception</a></li>
<li><a href="responsible-plotting.html#visual-encodings"><span class="toc-section-number">7.1.3</span> Visual Encodings</a></li>
<li><a href="responsible-plotting.html#some-opinionated-rules-of-thumb"><span class="toc-section-number">7.1.4</span> Some Opinionated Rules of Thumb</a></li>
</ul></li>
<li><a href="dv-gg.html#dv-gg"><span class="toc-section-number">7.2</span> Grammar of Graphics</a>
<ul>
<li><a href="dv-gg.html#ggplot-mechanics"><span class="toc-section-number">7.2.1</span> <code>ggplot</code> mechanics</a></li>
</ul></li>
<li><a href="plotting-one-dimension.html#plotting-one-dimension"><span class="toc-section-number">7.3</span> Plotting One Dimension</a>
<ul>
<li><a href="plotting-one-dimension.html#dv-bar"><span class="toc-section-number">7.3.1</span> Bar chart</a></li>
<li><a href="plotting-one-dimension.html#lollipop-plots"><span class="toc-section-number">7.3.2</span> Lollipop plots</a></li>
<li><a href="plotting-one-dimension.html#stacked-area-charts"><span class="toc-section-number">7.3.3</span> Stacked Area charts</a></li>
<li><a href="plotting-one-dimension.html#parallel-coordinate-plots"><span class="toc-section-number">7.3.4</span> Parallel Coordinate plots</a></li>
</ul></li>
<li><a href="visualizing-distributions.html#visualizing-distributions"><span class="toc-section-number">7.4</span> Visualizing Distributions</a>
<ul>
<li><a href="visualizing-distributions.html#histogram"><span class="toc-section-number">7.4.1</span> Histogram</a></li>
<li><a href="visualizing-distributions.html#density"><span class="toc-section-number">7.4.2</span> Density</a></li>
<li><a href="visualizing-distributions.html#boxplot"><span class="toc-section-number">7.4.3</span> Boxplot</a></li>
<li><a href="visualizing-distributions.html#violin-plot"><span class="toc-section-number">7.4.4</span> Violin plot</a></li>
<li><a href="visualizing-distributions.html#beeswarm-plot"><span class="toc-section-number">7.4.5</span> Beeswarm plot</a></li>
<li><a href="visualizing-distributions.html#ridgeline"><span class="toc-section-number">7.4.6</span> Ridgeline</a></li>
</ul></li>
<li><a href="plotting-two-or-more-dimensions.html#plotting-two-or-more-dimensions"><span class="toc-section-number">7.5</span> Plotting Two or More Dimensions</a>
<ul>
<li><a href="plotting-two-or-more-dimensions.html#scatter-plots"><span class="toc-section-number">7.5.1</span> Scatter Plots</a></li>
<li><a href="plotting-two-or-more-dimensions.html#bubble-plots"><span class="toc-section-number">7.5.2</span> Bubble Plots</a></li>
<li><a href="plotting-two-or-more-dimensions.html#connected-scatter-plots"><span class="toc-section-number">7.5.3</span> Connected Scatter Plots</a></li>
<li><a href="plotting-two-or-more-dimensions.html#line-plots"><span class="toc-section-number">7.5.4</span> Line Plots</a></li>
<li><a href="plotting-two-or-more-dimensions.html#parallel-coordinate-plots-1"><span class="toc-section-number">7.5.5</span> Parallel Coordinate Plots</a></li>
<li><a href="plotting-two-or-more-dimensions.html#heatmaps"><span class="toc-section-number">7.5.6</span> Heatmaps</a></li>
<li><a href="plotting-two-or-more-dimensions.html#how-to-use-heatmaps-responsibly"><span class="toc-section-number">7.5.7</span> How To Use Heatmaps Responsibly</a></li>
</ul></li>
<li><a href="other-kind-of-plots.html#other-kind-of-plots"><span class="toc-section-number">7.6</span> Other Kind of Plots</a>
<ul>
<li><a href="other-kind-of-plots.html#dendrograms"><span class="toc-section-number">7.6.1</span> Dendrograms</a></li>
<li><a href="other-kind-of-plots.html#chord-diagrams-and-circos-plots"><span class="toc-section-number">7.6.2</span> Chord Diagrams and Circos Plots</a></li>
</ul></li>
<li><a href="multiple-plots.html#multiple-plots"><span class="toc-section-number">7.7</span> Multiple Plots</a>
<ul>
<li><a href="multiple-plots.html#facet-wrapping"><span class="toc-section-number">7.7.1</span> Facet wrapping</a></li>
<li><a href="multiple-plots.html#multipanel-figures"><span class="toc-section-number">7.7.2</span> Multipanel Figures</a></li>
</ul></li>
<li><a href="publication-ready-plots.html#publication-ready-plots"><span class="toc-section-number">7.8</span> Publication Ready Plots</a>
<ul>
<li><a href="publication-ready-plots.html#ggplot-themes"><span class="toc-section-number">7.8.1</span> <code>ggplot</code> Themes</a></li>
<li><a href="publication-ready-plots.html#exporting-to-svg"><span class="toc-section-number">7.8.2</span> Exporting to SVG</a></li>
</ul></li>
<li><a href="network-visualization.html#network-visualization"><span class="toc-section-number">7.9</span> Network visualization</a>
<ul>
<li><a href="network-visualization.html#interactive-network-plots"><span class="toc-section-number">7.9.1</span> Interactive network plots</a></li>
</ul></li>
</ul></li>
<li><a href="biology-bioinformatics.html#biology-bioinformatics"><span class="toc-section-number">8</span> Biology &amp; Bioinformatics</a>
<ul>
<li><a href="bio-r.html#bio-r"><span class="toc-section-number">8.1</span> R in Biology</a></li>
<li><a href="biological-data-overview.html#biological-data-overview"><span class="toc-section-number">8.2</span> Biological Data Overview</a>
<ul>
<li><a href="biological-data-overview.html#bio-data-types"><span class="toc-section-number">8.2.1</span> Types of Biological Data</a></li>
<li><a href="biological-data-overview.html#csv-files"><span class="toc-section-number">8.2.2</span> CSV Files</a></li>
<li><a href="biological-data-overview.html#biological-data-is-not-tidy"><span class="toc-section-number">8.2.3</span> Biological data is NOT Tidy!</a></li>
</ul></li>
<li><a href="bio-bioconductor.html#bio-bioconductor"><span class="toc-section-number">8.3</span> Bioconductor</a></li>
<li><a href="microarrays.html#microarrays"><span class="toc-section-number">8.4</span> Microarrays</a></li>
<li><a href="high-throughput-sequencing.html#high-throughput-sequencing"><span class="toc-section-number">8.5</span> High Throughput Sequencing</a>
<ul>
<li><a href="high-throughput-sequencing.html#raw-hts-data"><span class="toc-section-number">8.5.1</span> Raw HTS Data</a></li>
<li><a href="high-throughput-sequencing.html#preliminary-hts-data-analysis"><span class="toc-section-number">8.5.2</span> Preliminary HTS Data Analysis</a></li>
<li><a href="high-throughput-sequencing.html#count-data"><span class="toc-section-number">8.5.3</span> Count Data</a></li>
</ul></li>
<li><a href="gene-identifiers.html#gene-identifiers"><span class="toc-section-number">8.6</span> Gene Identifiers</a>
<ul>
<li><a href="gene-identifiers.html#gene-identifier-systems"><span class="toc-section-number">8.6.1</span> Gene Identifier Systems</a></li>
<li><a href="gene-identifiers.html#mapping-between-identifier-systems"><span class="toc-section-number">8.6.2</span> Mapping Between Identifier Systems</a></li>
<li><a href="gene-identifiers.html#mapping-homologs"><span class="toc-section-number">8.6.3</span> Mapping Homologs</a></li>
</ul></li>
<li><a href="gene-expression.html#gene-expression"><span class="toc-section-number">8.7</span> Gene Expression</a>
<ul>
<li><a href="gene-expression.html#gene-expression-data-in-bioconductor"><span class="toc-section-number">8.7.1</span> Gene Expression Data in Bioconductor</a></li>
<li><a href="gene-expression.html#differential-expression-analysis"><span class="toc-section-number">8.7.2</span> Differential Expression Analysis</a></li>
<li><a href="gene-expression.html#microarray-gene-expression-data"><span class="toc-section-number">8.7.3</span> Microarray Gene Expression Data</a></li>
<li><a href="gene-expression.html#differential-expression-microarrays-limma"><span class="toc-section-number">8.7.4</span> Differential Expression: Microarrays (limma)</a></li>
<li><a href="gene-expression.html#rnaseq"><span class="toc-section-number">8.7.5</span> RNASeq</a></li>
<li><a href="gene-expression.html#rnaseq-gene-expression-data"><span class="toc-section-number">8.7.6</span> RNASeq Gene Expression Data</a></li>
<li><a href="gene-expression.html#filtering-counts"><span class="toc-section-number">8.7.7</span> Filtering Counts</a></li>
<li><a href="gene-expression.html#count-distributions"><span class="toc-section-number">8.7.8</span> Count Distributions</a></li>
<li><a href="gene-expression.html#differential-expression-rnaseq"><span class="toc-section-number">8.7.9</span> Differential Expression: RNASeq</a></li>
</ul></li>
<li><a href="gene-set-enrichment-analysis.html#gene-set-enrichment-analysis"><span class="toc-section-number">8.8</span> Gene Set Enrichment Analysis</a>
<ul>
<li><a href="gene-set-enrichment-analysis.html#gene-sets"><span class="toc-section-number">8.8.1</span> Gene Sets</a></li>
<li><a href="gene-set-enrichment-analysis.html#working-with-gene-sets-in-r"><span class="toc-section-number">8.8.2</span> Working with gene sets in R</a></li>
<li><a href="gene-set-enrichment-analysis.html#over-representation-analysis"><span class="toc-section-number">8.8.3</span> Over-representation Analysis</a></li>
<li><a href="gene-set-enrichment-analysis.html#rank-based-analysis"><span class="toc-section-number">8.8.4</span> Rank-based Analysis</a></li>
<li><a href="gene-set-enrichment-analysis.html#fgsea"><span class="toc-section-number">8.8.5</span> fgsea</a></li>
</ul></li>
<li><a href="biological-networks-..html#biological-networks-."><span class="toc-section-number">8.9</span> Biological Networks .</a>
<ul>
<li><a href="biological-networks-..html#biological-pathways-."><span class="toc-section-number">8.9.1</span> Biological Pathways .</a></li>
<li><a href="biological-networks-..html#gene-regulatory-networks-."><span class="toc-section-number">8.9.2</span> Gene Regulatory Networks .</a></li>
<li><a href="biological-networks-..html#protein-protein-interaction-network-."><span class="toc-section-number">8.9.3</span> Protein-Protein Interaction Network .</a></li>
<li><a href="biological-networks-..html#wgcna-."><span class="toc-section-number">8.9.4</span> WGCNA .</a></li>
</ul></li>
</ul></li>
<li><a href="engineering.html#engineering"><span class="toc-section-number">9</span> EngineeRing</a>
<ul>
<li><a href="unit-testing.html#unit-testing"><span class="toc-section-number">9.1</span> Unit Testing</a></li>
<li><a href="toolification.html#toolification"><span class="toc-section-number">9.2</span> Toolification</a>
<ul>
<li><a href="toolification.html#the-r-interpreter"><span class="toc-section-number">9.2.1</span> The R Interpreter</a></li>
<li><a href="toolification.html#rscript"><span class="toc-section-number">9.2.2</span> <code>Rscript</code></a></li>
<li><a href="toolification.html#commandargs"><span class="toc-section-number">9.2.3</span> <code>commandArgs()</code></a></li>
</ul></li>
<li><a href="parallel-processing.html#parallel-processing"><span class="toc-section-number">9.3</span> Parallel Processing</a>
<ul>
<li><a href="parallel-processing.html#brief-introduction-to-parallelization"><span class="toc-section-number">9.3.1</span> Brief Introduction to Parallelization</a></li>
<li><a href="parallel-processing.html#apply-and-friends-are-pleasingly-parallel"><span class="toc-section-number">9.3.2</span> <code>apply</code> and Friends Are Pleasingly Parallel</a></li>
<li><a href="parallel-processing.html#the-parallel-package"><span class="toc-section-number">9.3.3</span> The <code>parallel</code> package</a></li>
</ul></li>
<li><a href="object-orientation-in-r.html#object-orientation-in-r"><span class="toc-section-number">9.4</span> Object Orientation in R</a></li>
<li><a href="r-packages.html#r-packages"><span class="toc-section-number">9.5</span> R Packages</a></li>
</ul></li>
<li><a href="rshiny.html#rshiny"><span class="toc-section-number">10</span> RShiny</a>
<ul>
<li><a href="rshiny-introduction.html#rshiny-introduction"><span class="toc-section-number">10.1</span> RShiny Introduction</a>
<ul>
<li><a href="rshiny-introduction.html#how-does-a-web-page-work"><span class="toc-section-number">10.1.1</span> How does a web page work?</a></li>
</ul></li>
<li><a href="application-structure.html#application-structure"><span class="toc-section-number">10.2</span> Application Structure</a>
<ul>
<li><a href="application-structure.html#ui-design"><span class="toc-section-number">10.2.1</span> UI Design</a></li>
<li><a href="application-structure.html#server-functionality"><span class="toc-section-number">10.2.2</span> Server Functionality</a></li>
</ul></li>
<li><a href="reactivity.html#reactivity"><span class="toc-section-number">10.3</span> Reactivity</a></li>
<li><a href="publishing.html#publishing"><span class="toc-section-number">10.4</span> Publishing</a></li>
</ul></li>
<li><a href="communicating-with-r.html#communicating-with-r"><span class="toc-section-number">11</span> Communicating with R</a>
<ul>
<li><a href="rmarkdown-knitr.html#rmarkdown-knitr"><span class="toc-section-number">11.1</span> RMarkdown &amp; knitr</a></li>
<li><a href="bookdown.html#bookdown"><span class="toc-section-number">11.2</span> bookdown</a></li>
</ul></li>
<li><a href="contribution-guide.html#contribution-guide"><span class="toc-section-number">12</span> Contribution Guide</a>
<ul>
<li><a href="custom-blocks.html#custom-blocks"><span class="toc-section-number">12.1</span> Custom blocks</a></li>
</ul></li>
<li><a href="assignments.html#assignments">Assignments</a>
<ul>
<li><a href="assignment-format.html#assignment-format">Assignment Format</a></li>
<li><a href="starting-an-assignment.html#starting-an-assignment">Starting an Assignment</a>
<ul>
<li><a href="starting-an-assignment.html#starting-your-own-git-repository-and-cloning-it">Starting your own <code>git</code> repository and cloning it</a></li>
<li><a href="starting-an-assignment.html#contents">Contents</a></li>
<li><a href="starting-an-assignment.html#github-tutorial">GitHub tutorial</a></li>
<li><a href="starting-an-assignment.html#committing-and-pushing-with-r-and-without">Committing and Pushing with R (and without)</a></li>
<li><a href="starting-an-assignment.html#using-rstudio">Using RStudio</a></li>
<li><a href="starting-an-assignment.html#using-the-command-line">Using the command line</a></li>
</ul></li>
<li><a href="assignment-1.html#assignment-1">Assignment 1</a>
<ul>
<li><a href="assignment-1.html#problem-statement">Problem Statement</a></li>
<li><a href="assignment-1.html#learning-objectives-and-skill-list">Learning Objectives and Skill List</a></li>
<li><a href="assignment-1.html#instructions">Instructions</a></li>
<li><a href="assignment-1.html#hints">Hints</a></li>
</ul></li>
<li><a href="assignment-2.html#assignment-2">Assignment 2</a>
<ul>
<li><a href="assignment-2.html#problem-statement-1">Problem Statement</a></li>
<li><a href="assignment-2.html#learning-objectives">Learning Objectives</a></li>
<li><a href="assignment-2.html#skill-list">Skill List</a></li>
<li><a href="assignment-2.html#instructions-1">Instructions</a></li>
<li><a href="assignment-2.html#function-details">Function Details</a></li>
</ul></li>
<li><a href="assignment-3.html#assignment-3">Assignment 3</a>
<ul>
<li><a href="assignment-3.html#problem-statement-2">Problem Statement</a></li>
<li><a href="assignment-3.html#learning-objectives-1">Learning Objectives</a></li>
<li><a href="assignment-3.html#skill-list-1">Skill List</a></li>
<li><a href="assignment-3.html#background-on-microarrays">Background on Microarrays</a></li>
<li><a href="assignment-3.html#background-on-principal-component-analysis">Background on Principal Component Analysis</a></li>
<li><a href="assignment-3.html#marisa-et-al.-gene-expression-classification-of-colon-cancer-into-molecular-subtypes-characterization-validation-and-prognostic-value.-plos-medicine-may-2013.-pmid-23700391">Marisa et al. Gene Expression Classification of Colon Cancer into Molecular Subtypes: Characterization, Validation, and Prognostic Value. PLoS Medicine, May 2013. PMID: 23700391</a></li>
<li><a href="assignment-3.html#scaling-data-using-r-scale">Scaling data using R <code>scale()</code></a></li>
<li><a href="assignment-3.html#proportion-of-variance-explained">Proportion of variance explained</a></li>
<li><a href="assignment-3.html#plotting-and-visualization-of-pca">Plotting and visualization of PCA</a></li>
<li><a href="assignment-3.html#hierarchical-clustering-and-heatmaps">Hierarchical Clustering and Heatmaps</a></li>
<li><a href="assignment-3.html#references">References</a></li>
</ul></li>
<li><a href="assignment-4.html#assignment-4">Assignment 4</a>
<ul>
<li><a href="assignment-4.html#problem-statement-3">Problem Statement</a></li>
<li><a href="assignment-4.html#learning-objectives-2">Learning Objectives</a></li>
<li><a href="assignment-4.html#skill-list-2">Skill List</a></li>
<li><a href="assignment-4.html#instructions-2">Instructions</a></li>
<li><a href="assignment-4.html#deliverables">Deliverables</a></li>
<li><a href="assignment-4.html#function-details-1">Function Details</a></li>
<li><a href="assignment-4.html#hints-1">Hints</a></li>
</ul></li>
<li><a href="assignment-5.html#assignment-5">Assignment 5</a>
<ul>
<li><a href="assignment-5.html#problem-statement-4">Problem Statement</a></li>
<li><a href="assignment-5.html#learning-objectives-3">Learning Objectives</a></li>
<li><a href="assignment-5.html#skill-list-3">Skill List</a></li>
<li><a href="assignment-5.html#deseq2-background">DESeq2 Background</a></li>
<li><a href="assignment-5.html#generating-a-counts-matrix">Generating a counts matrix</a></li>
<li><a href="assignment-5.html#prefiltering-counts-matrix">Prefiltering Counts matrix</a></li>
<li><a href="assignment-5.html#median-of-ratios-normalization">Median-of-ratios normalization</a></li>
<li><a href="assignment-5.html#deseq2-preparation">DESeq2 preparation</a></li>
<li><a href="assignment-5.html#omeara-et-al.-transcriptional-reversion-of-cardiac-myocyte-fate-during-mammalian-cardiac-regeneration.-circ-res.-feb-2015.-pmid-25477501l">O’Meara et al. Transcriptional Reversion of Cardiac Myocyte Fate During Mammalian Cardiac Regeneration. Circ Res. Feb 2015. PMID: 25477501l</a></li>
<li><a href="assignment-5.html#reading-and-subsetting-the-data-from-verse_counts.tsv-and-sample_metadata.csv">1. Reading and subsetting the data from verse_counts.tsv and sample_metadata.csv</a></li>
<li><a href="assignment-5.html#running-deseq2">2. Running DESeq2</a></li>
<li><a href="assignment-5.html#annotating-results-to-construct-a-labeled-volcano-plot">3. Annotating results to construct a labeled volcano plot</a></li>
<li><a href="assignment-5.html#diagnostic-plot-of-the-raw-p-values-for-all-genes">4. Diagnostic plot of the raw p-values for all genes</a></li>
<li><a href="assignment-5.html#plotting-the-logfoldchanges-for-differentially-expressed-genes">5. Plotting the LogFoldChanges for differentially expressed genes</a></li>
<li><a href="assignment-5.html#the-choice-of-fdr-cutoff-depends-on-cost">The choice of FDR cutoff depends on cost</a></li>
<li><a href="assignment-5.html#plotting-the-normalized-counts-of-differentially-expressed-genes">6. Plotting the normalized counts of differentially expressed genes</a></li>
<li><a href="assignment-5.html#volcano-plot-to-visualize-differential-expression-results">7. Volcano Plot to visualize differential expression results</a></li>
<li><a href="assignment-5.html#running-fgsea-vignette">8. Running fgsea <span>vignette</span></a></li>
<li><a href="assignment-5.html#plotting-the-top-ten-positive-nes-and-top-ten-negative-nes-pathways">9. Plotting the top ten positive NES and top ten negative NES pathways</a></li>
<li><a href="assignment-5.html#references-1">References</a></li>
</ul></li>
<li><a href="assignment-6.html#assignment-6">Assignment 6</a>
<ul>
<li><a href="assignment-6.html#problem-statement-5">Problem Statement</a></li>
<li><a href="assignment-6.html#learning-objectives-4">Learning Objectives</a></li>
<li><a href="assignment-6.html#skill-list-4">Skill List</a></li>
<li><a href="assignment-6.html#instructions-3">Instructions</a></li>
<li><a href="assignment-6.html#function-details-2">Function Details</a></li>
</ul></li>
<li><a href="assignment-7.html#assignment-7">Assignment 7</a>
<ul>
<li><a href="assignment-7.html#problem-statement-6">Problem Statement</a></li>
<li><a href="assignment-7.html#learning-objectives-5">Learning Objectives</a></li>
<li><a href="assignment-7.html#skill-list-5">Skill List</a></li>
<li><a href="assignment-7.html#instructions-4">Instructions</a></li>
<li><a href="assignment-7.html#function-details-3">“Function” Details</a></li>
</ul></li>
</ul></li>
<li class="appendix"><span><b>Appendix</b></span></li>
<li><a href="class-outlines.html#class-outlines"><span class="toc-section-number">A</span> Class Outlines</a>
<ul>
<li><a href="week-1.html#week-1"><span class="toc-section-number">A.1</span> Week 1</a></li>
<li><a href="week-2.html#week-2"><span class="toc-section-number">A.2</span> Week 2</a></li>
<li><a href="week-3.html#week-3"><span class="toc-section-number">A.3</span> Week 3</a></li>
<li><a href="week-4.html#week-4"><span class="toc-section-number">A.4</span> Week 4</a></li>
<li><a href="week-5.html#week-5"><span class="toc-section-number">A.5</span> Week 5</a></li>
<li><a href="week-6.html#week-6"><span class="toc-section-number">A.6</span> Week 6</a></li>
<li><a href="week-7.html#week-7"><span class="toc-section-number">A.7</span> Week 7</a></li>
<li><a href="week-8.html#week-8"><span class="toc-section-number">A.8</span> Week 8</a></li>
<li><a href="week-9.html#week-9"><span class="toc-section-number">A.9</span> Week 9</a></li>
<li><a href="week-10.html#week-10"><span class="toc-section-number">A.10</span> Week 10</a></li>
<li><a href="week-11.html#week-11"><span class="toc-section-number">A.11</span> Week 11</a></li>
<li><a href="week-12.html#week-12"><span class="toc-section-number">A.12</span> Week 12</a></li>
<li><a href="week-13.html#week-13"><span class="toc-section-number">A.13</span> Week 13</a></li>
<li><a href="week-14.html#week-14"><span class="toc-section-number">A.14</span> Week 14</a></li>
</ul></li>
</ul>
</div>
</div>
</div>
<div class="row">
<div class="col-sm-12">
<div id="responsible-plotting" class="section level2" number="7.1">
<h2><span class="header-section-number">7.1</span> Responsible Plotting</h2>
<p>“Good plots empower us to ask good questions.” - Alberto Cairo, How Charts Lie</p>
<p>Plots are the primary vehicles of communicating scientific results. A very
common and useful approach to writing scientific papers is to decide how many
figures are needed to tell the story, and then create plots as necessary to
convey that story. The text plays a supporting role for the figures. Effective
data visualization has immense power to communicate the ideas (and beliefs) of
the chart creator. Consider the famous “hockey stick graph” depicting average
global temperatures created by Jerry Mahlman to describe the pattern of global
warming shown by Mann, Bradkely &amp; Hughes in 1999:</p>
<div class="figure">
<img src="hockey_stick_chart.jpg" alt="" />
<p class="caption">“Hockey stick graph” by Jerry Mahlman, showing the trend of temperature rise over the past 1000 years in the northern hemisphere</p>
</div>
<p>The data and models underlying the graph are very complex, but the story is
clear: the past 100 years has seen a dramatic increase in average temperature in
the northern hemisphere. This chart was one of the main drivers behind renewed
main stream awareness of climate change, in no small part because it is easy to
understand.</p>
<p>Given the power visualization has, and the increasing amount of data created by
and required for daily life, it is ever more important that data visualization
designers understand the principles and pitfalls of effective plotting. There is
no easy solution to accomplishing this, but Alberto Cairo and Enrico Bertini,
two journalists who specialize in data visualization and infographics, lay out
five qualities of an effective visualization, in Cairo’s book <a href="http://www.thefunctionalart.com/p/the-truthful-art-book.html">the truthful
art</a>:</p>
<ol style="list-style-type: decimal">
<li><strong>It is truthful</strong> - the message conveyed by the visualization expresses some
truth of the underlying data (to the extent that anything is “true”)</li>
<li><strong>It is functional</strong> - the visualization is accurate, and allows the reader
to do something useful with the insights they gain</li>
<li><strong>It is insightful</strong> - the visualization should reveal patterns in the data
that would otherwise be difficult or impossible to see and understand</li>
<li><strong>It is enlightening</strong> - the reader should leave the experience of engaging
with the visualization convinced of something, ideally what the designer intends</li>
<li><strong>It is beautiful</strong> - it is visually appealing to its target audience, which
helps the user more easily spend time and engage with the visualization</li>
</ol>
<p>A full treatment of these ideas are included in, among others, three excellent
books by Alberto Cairo, linked at the end of this section. The sections that
follow describe some of the ideas and concepts that make visualization
challenging and offer some guidance from the books mentioned on how to produce
effective visualizations.</p>
<div class="box readmore">
<p>All three of these books were written by Alberto Cairo:</p>
<ul>
<li><a href="http://www.thefunctionalart.com/p/reviews.html">How Charts Lie</a></li>
<li><a href="http://www.thefunctionalart.com/p/about-book.html">the funcitonal art</a></li>
<li><a href="http://www.thefunctionalart.com/p/the-truthful-art-book.html">the truthful art</a></li>
</ul>
</div>
<div id="visualization-principles" class="section level3" number="7.1.1">
<h3><span class="header-section-number">7.1.1</span> Visualization Principles</h3>
<p>The purpose of data visualization is to illustrate (from the latin <em>illustrare</em>,
meaning to light up or illuminate) relationships between quantities. Effective
data visualization takes advantage of our highly evolved <a href="https://en.wikipedia.org/wiki/Visual_perception">visual perception
system</a> to communicate these
relationships to our cognitive and decision making systems. Therefore, our
visualization strategies should take knowledge of our perceptual biases into
account when deciding how to present data. Poor data visualization choices may
make interpretation difficult, or worse lead us to incorrect inferences and
conclusions.</p>
</div>
<div id="human-visual-perception" class="section level3" number="7.1.2">
<h3><span class="header-section-number">7.1.2</span> Human Visual Perception</h3>
<p>The human visual perceptual system is intrinsically integrated with our
cognitive system. This makes it a <em>predictive</em> or <em>interpretive</em> system, where
patterns of light received by the eye are pre-processed before the relatively
slower executive function parts of our brains can make judgements.
If we compare our perception of the world around us to a cinematic film, this
makes our visual perception less raw video footage and more like a director’s
commentary, where we add our interpretations, motivations, experiences, and
anecdotal annotations to the images before we cognitively interpret them. Said
differently, our visuo-cognitive system performs <em>pattern recognition</em> on the
information we receive, and annotates the information with the patterns it
thinks it sees. Very often, our cognition predicts patterns that are accurate,
which therefore may be useful for us in making decisions, e.g. how to move out
of the trajectory of a falling tree branch. However, sometimes our pattern
recognition system, which is always <em>trying</em> to match patterns, sees patterns
that don’t exist. Some of the most well known and well studied examples of these
perceptual errors are <em>optical illusions.</em></p>
<p>Optical illusions are images or animations that reveal the errors and bias of
our perceptual system. Some illusions that are static images appear to move.
Others distort the true relations between lines or shapes, or lead to incorrect
or inconsistent perception of color. Consider the famous <a href="https://en.wikipedia.org/wiki/Checker_shadow_illusion">checkerboard shadow
illusion</a> created by
Edward H. Adelson in 1995:</p>
<div class="figure">
<img src="checker_illusion.png" alt="" />
<p class="caption">Checkerboard shadow illusion, the squares marked A and B are the same color</p>
</div>
<p>The image on the right clearly shows the two squares A and B are the same shade
of grey, though A appears darker than B in the left image. This is because our
perceptual system identifies color by <em>contrasting</em> adjacent colors. Since the
colors adjacent to the A and B squares differ, our perception of the colors also
differ. This example illustrates the contrast bias according to the <em>value</em>, or
lightness, of the color, but the principle also applies to different hues:</p>
<div class="figure">
<img src="gradient_illusion.png" alt="" />
<p class="caption">The three squares are drawn on top of and below the image and are the same color</p>
</div>
<p>An illusion that shows how our perceptual system distorts relationships between
geometry is as follows:</p>
<div class="figure">
<img src="wavy_grid.png" alt="" />
<p class="caption">The horizontal and vertical pattern appears to be “wavy” but in fact form straight lines</p>
</div>
<p>The pattern of light and dark squares intersecting with the grid makes the lines
appear “wavy” when all the horizontal and vertical lines are in fact straight,
parallel, and perpendicular.</p>
<p>A final example illusion that is relevant for data visualization demonstrates
a similar contextual bias of relative area:</p>
<div class="figure">
<img src="area_illusion.png" alt="" />
<p class="caption">The blue dots have the same diameter</p>
</div>
<p>The two blue dots in the image have the same radius, although the dot within
the large black circles appears smaller.</p>
<p>Not only are there illusions that distort our perception of shapes that are
there, but there are also illusions that convince us of the existence of shapes
that are <em>not</em> there. The following illusion illustrates the point:</p>
<div class="figure">
<img src="negative_shape.png" alt="" />
<p class="caption">The shapes are drawn to suggest a triangle in the middle, though no triangle is actually drawn</p>
</div>
<p>Seeing shapes when they aren’t there is nothing new to humans, of course.</p>
<div class="figure">
<img src="cloud_dog.png" alt="" />
<p class="caption">Clouds that resemble shapes - this author sees a running dog</p>
</div>
<p>Each of these perceptual biases relate to how we interpret many common data
visualization strategies. We cannot avoid making these interpretive errors via
our cognition alone, but we can use this knowledge in the design of our plots
to mitigate their effects. To quickly summarize the perceptual biases we have
demonstrated:</p>
<ol style="list-style-type: decimal">
<li>We perceive the value and hue of colors in contrast to adjacent colors</li>
<li>Certain types of geometry interfere with our ability to assess accurate
relationships between geometric shapes</li>
<li>We perceive the size (area) of shapes relative to nearby shapes</li>
<li>We may perceive shapes where there are none drawn, based on nearby shapes</li>
<li>We may perceive shapes that are similar to other familiar shapes, but have no
true relationship</li>
</ol>
<p>Each of these biases influence how we perceive and interpret plots, as we will
see in the rest of this chapter.</p>
</div>
<div id="visual-encodings" class="section level3" number="7.1.3">
<h3><span class="header-section-number">7.1.3</span> Visual Encodings</h3>
<p>Data visualization is the process of encoding numbers as positions, shapes, and
colors in a two- or sometimes three-dimensional space. There are many different
geometric properties we can use to create an encoding, including:</p>
<ol style="list-style-type: decimal">
<li>Length, width, or height</li>
<li>Position</li>
<li>Area</li>
<li>Angle/proportional area</li>
<li>Shape</li>
<li>Hue and value of colors</li>
</ol>
<p>Examples of each of these encodings is depicted for a fictitious dataset of
values between zero and one in the following figure:</p>
<div class="figure">
<img src="viz_encodings.png" alt="" />
<p class="caption">Different encodings for the same data. This figure was inspired by Alberto Cairo in his book the truthful art.</p>
</div>
<p>The example encodings in the figure are among the most common and recognizable
elements used in scientific plotting. Every time a reader examines a plot, they
must <em>decode</em> it to a (hopefully) accurate mental model of the quantities that
underlie it. Using familiar encodings that a reader will already know how to
interpret makes this decoding cognitively easier, thereby enabling the plot to
be more quickly understood.</p>
<p>Every quantity we wish to plot must be mapped to a visualization using one or
more of these geometric encodings. By combining different variables with
different encodings, we construct more complex plots. Consider the scatter plot,
which contains two continuous values per datum, each of which is mapped to a
visualization with a position encoding:</p>
<div class="sourceCode" id="cb262"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb262-1"><a href="responsible-plotting.html#cb262-1" aria-hidden="true" tabindex="-1"></a><span class="fu">tibble</span>(</span>
<span id="cb262-2"><a href="responsible-plotting.html#cb262-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">x=</span><span class="fu">rnorm</span>(<span class="dv">100</span>),</span>
<span id="cb262-3"><a href="responsible-plotting.html#cb262-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">y=</span><span class="fu">rnorm</span>(<span class="dv">100</span>,<span class="dv">0</span>,<span class="dv">5</span>)</span>
<span id="cb262-4"><a href="responsible-plotting.html#cb262-4" aria-hidden="true" tabindex="-1"></a>) <span class="sc">%&gt;%</span></span>
<span id="cb262-5"><a href="responsible-plotting.html#cb262-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x=</span>x,<span class="at">y=</span>y)) <span class="sc">+</span></span>
<span id="cb262-6"><a href="responsible-plotting.html#cb262-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>()</span></code></pre></div>
<p><img src="main_files/viz%20scatter-1.png" width="672" /></p>
<p>We might layer on another encoding by introducing a new variable that is the
absolute value of the product of <span class="math inline">\(x\)</span> and <span class="math inline">\(y\)</span> and make the area of each datum
proportional to that value:</p>
<div class="sourceCode" id="cb263"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb263-1"><a href="responsible-plotting.html#cb263-1" aria-hidden="true" tabindex="-1"></a>data <span class="ot">&lt;-</span>  <span class="fu">tibble</span>(</span>
<span id="cb263-2"><a href="responsible-plotting.html#cb263-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">x=</span><span class="fu">rnorm</span>(<span class="dv">100</span>),</span>
<span id="cb263-3"><a href="responsible-plotting.html#cb263-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">y=</span><span class="fu">rnorm</span>(<span class="dv">100</span>,<span class="dv">0</span>,<span class="dv">5</span>),</span>
<span id="cb263-4"><a href="responsible-plotting.html#cb263-4" aria-hidden="true" tabindex="-1"></a>  <span class="st">`</span><span class="at">x times y</span><span class="st">`</span><span class="ot">=</span><span class="fu">abs</span>(x<span class="sc">*</span>y)</span>
<span id="cb263-5"><a href="responsible-plotting.html#cb263-5" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb263-6"><a href="responsible-plotting.html#cb263-6" aria-hidden="true" tabindex="-1"></a>data <span class="sc">%&gt;%</span></span>
<span id="cb263-7"><a href="responsible-plotting.html#cb263-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x=</span>x,<span class="at">y=</span>y,<span class="at">size=</span><span class="st">`</span><span class="at">x times y</span><span class="st">`</span>)) <span class="sc">+</span></span>
<span id="cb263-8"><a href="responsible-plotting.html#cb263-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>()</span></code></pre></div>
<p><img src="main_files/viz%20scatter%202-1.png" width="672" /></p>
<p>We could layer yet more encodings to color each marker proportional to a value
between 0 and 1, and also change the shape based on a category:</p>
<div class="sourceCode" id="cb264"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb264-1"><a href="responsible-plotting.html#cb264-1" aria-hidden="true" tabindex="-1"></a>data <span class="ot">&lt;-</span> <span class="fu">mutate</span>(data,</span>
<span id="cb264-2"><a href="responsible-plotting.html#cb264-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">z=</span><span class="fu">runif</span>(<span class="dv">100</span>),</span>
<span id="cb264-3"><a href="responsible-plotting.html#cb264-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">category=</span><span class="fu">sample</span>(<span class="fu">c</span>(<span class="st">&#39;A&#39;</span>,<span class="st">&#39;B&#39;</span>,<span class="st">&#39;C&#39;</span>), <span class="dv">100</span>, <span class="at">replace=</span><span class="cn">TRUE</span>)</span>
<span id="cb264-4"><a href="responsible-plotting.html#cb264-4" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb264-5"><a href="responsible-plotting.html#cb264-5" aria-hidden="true" tabindex="-1"></a>data <span class="sc">%&gt;%</span></span>
<span id="cb264-6"><a href="responsible-plotting.html#cb264-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x=</span>x, <span class="at">y=</span>y, <span class="at">size=</span><span class="st">`</span><span class="at">x times y</span><span class="st">`</span>, <span class="at">color=</span>z, <span class="at">shape=</span>category)) <span class="sc">+</span></span>
<span id="cb264-7"><a href="responsible-plotting.html#cb264-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>()</span></code></pre></div>
<p><img src="main_files/viz%20scatter%203-1.png" width="672" /></p>
<p>Let’s say the sum of <span class="math inline">\(x\)</span> and <span class="math inline">\(y\)</span> is meaningful as well, and we are interested in
the order of the data along this axis. We can connect all the points with lines
by using a length encoding between adjacent data points after sorting by <span class="math inline">\(x+y\)</span>:</p>
<div class="sourceCode" id="cb265"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb265-1"><a href="responsible-plotting.html#cb265-1" aria-hidden="true" tabindex="-1"></a>data <span class="ot">&lt;-</span> <span class="fu">mutate</span>(data,</span>
<span id="cb265-2"><a href="responsible-plotting.html#cb265-2" aria-hidden="true" tabindex="-1"></a>  <span class="st">`</span><span class="at">x + y</span><span class="st">`</span><span class="ot">=</span>x<span class="sc">+</span>y,</span>
<span id="cb265-3"><a href="responsible-plotting.html#cb265-3" aria-hidden="true" tabindex="-1"></a>) <span class="sc">%&gt;%</span></span>
<span id="cb265-4"><a href="responsible-plotting.html#cb265-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(<span class="st">`</span><span class="at">x + y</span><span class="st">`</span>) <span class="sc">%&gt;%</span></span>
<span id="cb265-5"><a href="responsible-plotting.html#cb265-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(</span>
<span id="cb265-6"><a href="responsible-plotting.html#cb265-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">xend=</span><span class="fu">lag</span>(x,<span class="dv">1</span>),</span>
<span id="cb265-7"><a href="responsible-plotting.html#cb265-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">yend=</span><span class="fu">lag</span>(y,<span class="dv">1</span>)</span>
<span id="cb265-8"><a href="responsible-plotting.html#cb265-8" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb265-9"><a href="responsible-plotting.html#cb265-9" aria-hidden="true" tabindex="-1"></a>data <span class="sc">%&gt;%</span></span>
<span id="cb265-10"><a href="responsible-plotting.html#cb265-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb265-11"><a href="responsible-plotting.html#cb265-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_segment</span>(<span class="fu">aes</span>(<span class="at">x=</span>x,<span class="at">xend=</span>xend,<span class="at">y=</span>y,<span class="at">yend=</span>yend,<span class="at">alpha=</span><span class="fl">0.5</span>)) <span class="sc">+</span></span>
<span id="cb265-12"><a href="responsible-plotting.html#cb265-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">x=</span>x, <span class="at">y=</span>y, <span class="at">size=</span><span class="st">`</span><span class="at">x times y</span><span class="st">`</span>, <span class="at">color=</span>z, <span class="at">shape=</span>category))</span></code></pre></div>
<p><img src="main_files/viz%20scatter%204-1.png" width="672" /></p>
<p>The plot is becoming very busy and difficult to interpret because we have
encoded six different dimensions in the same plot:</p>
<ol style="list-style-type: decimal">
<li><span class="math inline">\(x\)</span> value has a positional encoding</li>
<li><span class="math inline">\(y\)</span> value has another positional encoding</li>
<li>The product <span class="math inline">\(xy\)</span> has an area encoding</li>
<li><span class="math inline">\(z\)</span> has a quantitative color value encoding</li>
<li>category has a categorical encoding to shape</li>
<li>The adjacency of coordinates along the <span class="math inline">\(x + y\)</span> axis with a length encoding</li>
</ol>
<p>Very complex plots may be built by thinking explicitly about how each dimension
of the data are encoded, but the designer (i.e. you) must carefully consider
whether the chosen encodings faithfully depict the relationships in the data.</p>
<p>These common encodings depicted in the earlier figure are not the only
possibilities for mapping visual properties to a set of values. Consider the
violin plot below:</p>
<div class="sourceCode" id="cb266"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb266-1"><a href="responsible-plotting.html#cb266-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">1337</span>)</span>
<span id="cb266-2"><a href="responsible-plotting.html#cb266-2" aria-hidden="true" tabindex="-1"></a><span class="fu">tibble</span>(</span>
<span id="cb266-3"><a href="responsible-plotting.html#cb266-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">value=</span><span class="fu">rnorm</span>(<span class="dv">100</span>),</span>
<span id="cb266-4"><a href="responsible-plotting.html#cb266-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">category=</span><span class="st">&#39;A&#39;</span></span>
<span id="cb266-5"><a href="responsible-plotting.html#cb266-5" aria-hidden="true" tabindex="-1"></a>) <span class="sc">%&gt;%</span></span>
<span id="cb266-6"><a href="responsible-plotting.html#cb266-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x=</span>category,<span class="at">y=</span>value,<span class="at">fill=</span>category)) <span class="sc">+</span></span>
<span id="cb266-7"><a href="responsible-plotting.html#cb266-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_violin</span>()</span></code></pre></div>
<p><img src="main_files/viz%20violin-1.png" width="672" /></p>
<p>Here, <code>geom_violin()</code> has preprocessed the data to map the number of values
across the full range of values into a density. The density is then encoded with
a width encoding, where the width of the violin across the range of values,
which is encoded with a position encoding.</p>
<p>Most common chart types are simply combinations of variables with different
encodings. The following table contains the encodings employed for different
types of plots:</p>
<table>
<colgroup>
<col width="21%" />
<col width="10%" />
<col width="28%" />
<col width="40%" />
</colgroup>
<thead>
<tr class="header">
<th>Plot</th>
<th><span class="math inline">\(x\)</span> encoding</th>
<th><span class="math inline">\(y\)</span> encoding</th>
<th>note</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><a href="scatter-plot">scatter</a></td>
<td>position</td>
<td>position</td>
<td></td>
</tr>
<tr class="even">
<td><a href="bar-chart">vertical bar</a></td>
<td>position</td>
<td>height</td>
<td></td>
</tr>
<tr class="odd">
<td><a href="bar-chart">horizontal bar</a></td>
<td>width</td>
<td>position</td>
<td></td>
</tr>
<tr class="even">
<td><a href="lollipop-plots">lollipop</a></td>
<td>position</td>
<td>height for line + position for “head”</td>
<td></td>
</tr>
<tr class="odd">
<td><a href="violin-plots">violin</a></td>
<td>position</td>
<td>width</td>
<td><span class="math inline">\(x\)</span> transformed to range, <span class="math inline">\(y\)</span> transformed to densities</td>
</tr>
</tbody>
</table>
<div id="elementary-perceptual-tasks" class="section level4" number="7.1.3.1">
<h4><span class="header-section-number">7.1.3.1</span> Elementary Perceptual Tasks</h4>
<p>The purpose of visualizations is to enable the reader to construct an accurate
mental model of the relationships or ideas contained within the data. As
discussed in the <a href="responsible-plotting.html#human-visual-perception">Human Visual Perception</a> section, not all visualizations allow
for easy and accurate judgments of this kind due to the biases of our biological
visuo-cognitive biases. However, some encodings of data allow <em>more accurate</em>
judgments than others. William Cleveland and Robert McGill, two statisticians,
proposed a hierarchy of visual encodings that span the spectrum of precision of
our ability to accurately interpret visual information <span class="citation">(<a href="#ref-Cleveland1984-ay" role="doc-biblioref">Cleveland and McGill 1984</a>)</span>. They
defined the act of mapping a specific type of encoding to a cognitive model an
<em>elementary perceptual task.</em> The hierarchy is depicted in the following figure:</p>
<div class="figure">
<img src="elementary_perceptual_tasks.png" alt="" />
<p class="caption">Elementary Perceptual Task Hierarchy - inspired by Alberto Cairo in the truthful art, which was inspired by Willian Cleveland and Robert McGill</p>
</div>
<p>The hierarchy is oriented from least to most cognitively difficult decoding
tasks, where the easier tasks lend themselves to easier comparisons and
judgements. For example, the easiest perceptual task is to compare visual
elements with a position encoding that are all aligned across all charts; the
positional decoding is the same for each chart, even if they use different
stylistic elements. Further down the hierarchy, more cognitive effort is
required to assess the precise relationships between elements of the
plot, therefore possibly reducing the accuracy of our judgements.</p>
<p>This is not to say that it is always better for the reader to make highly
accurate estimates. General trends and patterns may be more readily (and
beautifully) apparent with less accurate encodings. Consider the following
clustered heatmap:</p>
<div class="sourceCode" id="cb267"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb267-1"><a href="responsible-plotting.html#cb267-1" aria-hidden="true" tabindex="-1"></a><span class="co"># these data have two groups of samples with similar random data profiles</span></span>
<span id="cb267-2"><a href="responsible-plotting.html#cb267-2" aria-hidden="true" tabindex="-1"></a>data <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(</span>
<span id="cb267-3"><a href="responsible-plotting.html#cb267-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tibble</span>(</span>
<span id="cb267-4"><a href="responsible-plotting.html#cb267-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">A=</span><span class="fu">c</span>(<span class="fu">rnorm</span>(<span class="dv">10</span>),<span class="fu">rnorm</span>(<span class="dv">10</span>,<span class="dv">2</span>)),</span>
<span id="cb267-5"><a href="responsible-plotting.html#cb267-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">B=</span><span class="fu">c</span>(<span class="fu">rnorm</span>(<span class="dv">10</span>),<span class="fu">rnorm</span>(<span class="dv">10</span>,<span class="dv">2</span>)),</span>
<span id="cb267-6"><a href="responsible-plotting.html#cb267-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">C=</span><span class="fu">c</span>(<span class="fu">rnorm</span>(<span class="dv">10</span>),<span class="fu">rnorm</span>(<span class="dv">10</span>,<span class="dv">2</span>)),</span>
<span id="cb267-7"><a href="responsible-plotting.html#cb267-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">D=</span><span class="fu">c</span>(<span class="fu">rnorm</span>(<span class="dv">10</span>,<span class="dv">4</span>),<span class="fu">rnorm</span>(<span class="dv">10</span>,<span class="sc">-</span><span class="dv">1</span>)),</span>
<span id="cb267-8"><a href="responsible-plotting.html#cb267-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">E=</span><span class="fu">c</span>(<span class="fu">rnorm</span>(<span class="dv">10</span>,<span class="dv">4</span>),<span class="fu">rnorm</span>(<span class="dv">10</span>,<span class="sc">-</span><span class="dv">1</span>)),</span>
<span id="cb267-9"><a href="responsible-plotting.html#cb267-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">F=</span><span class="fu">c</span>(<span class="fu">rnorm</span>(<span class="dv">10</span>,<span class="dv">4</span>),<span class="fu">rnorm</span>(<span class="dv">10</span>,<span class="sc">-</span><span class="dv">1</span>))</span>
<span id="cb267-10"><a href="responsible-plotting.html#cb267-10" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb267-11"><a href="responsible-plotting.html#cb267-11" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb267-12"><a href="responsible-plotting.html#cb267-12" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(data) <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">&#39;G&#39;</span>,<span class="fu">seq</span>(<span class="fu">nrow</span>(data)))</span>
<span id="cb267-13"><a href="responsible-plotting.html#cb267-13" aria-hidden="true" tabindex="-1"></a><span class="fu">heatmap</span>(data)</span></code></pre></div>
<p><img src="main_files/heatmap%20perceptual%20task-1.png" width="672" /></p>
<p>In this toy example, it may be sufficient for the reader’s comprehension to show
that samples A, B, and C cluster together, as do D, E, and F, without requiring
them to know <em>precisely</em> how different they are. In this case, the clustering
suggested by the visualization should be presented in some other way that is
more precise as well, for example a statistical comparison with results reported
in a table, to confirm what our eyes tell us.</p>
<p>It is usually a good idea to plot data multiple ways, to ensure the patterns we
see with one encoding are also seen with another. We might consider plotting
the data again using a <a href="parallel-coordinate-plots">parallel coordinates plot</a>:</p>
<div class="sourceCode" id="cb268"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb268-1"><a href="responsible-plotting.html#cb268-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(GGally)</span>
<span id="cb268-2"><a href="responsible-plotting.html#cb268-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb268-3"><a href="responsible-plotting.html#cb268-3" aria-hidden="true" tabindex="-1"></a><span class="fu">as_tibble</span>(data) <span class="sc">%&gt;%</span> <span class="fu">mutate</span>(</span>
<span id="cb268-4"><a href="responsible-plotting.html#cb268-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">group=</span><span class="fu">c</span>(<span class="fu">rep</span>(<span class="st">&#39;Group 1&#39;</span>,<span class="dv">10</span>),<span class="fu">rep</span>(<span class="st">&#39;Group 2&#39;</span>,<span class="dv">10</span>))</span>
<span id="cb268-5"><a href="responsible-plotting.html#cb268-5" aria-hidden="true" tabindex="-1"></a>) <span class="sc">%&gt;%</span></span>
<span id="cb268-6"><a href="responsible-plotting.html#cb268-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggparcoord</span>(<span class="at">columns=</span><span class="dv">1</span><span class="sc">:</span><span class="dv">6</span>,<span class="at">groupColumn=</span><span class="dv">7</span>)</span></code></pre></div>
<p><img src="main_files/viz%20element%20pcp-1.png" width="672" /></p>
<p>For simplicity, we added a new variable <code>group</code> that is consistent with the
clustering we observed in the heatmap, which we might have extracted
computationally by cutting the row dendrogram into two flat clusters. The
parallel coordinate plot also visually confirms our interpretation of two
distinct clusters. We may now feel more comfortable including one or the other
plot in a figure, though this author finds the parallel coordinate plot more
accurate and enlightening. Recall the elementary perceptual task hierarchy:
aligned position encodings enable more accurate estimates of data. Heatmaps,
which use color value and hue, enable the least accurate estimates.</p>
</div>
</div>
<div id="some-opinionated-rules-of-thumb" class="section level3" number="7.1.4">
<h3><span class="header-section-number">7.1.4</span> Some Opinionated Rules of Thumb</h3>
<p>Effective data visualization is challenging, and there are no simple solutions
or single approaches to doing it well. In this sense, data visualization is as
much an art as a science. This author has some rules of thumb he applies when
visualizing data:</p>
<ol style="list-style-type: decimal">
<li><strong>Visualize the same data in multiple ways.</strong> This will help ensure you
aren’t being unduly influenced by the quirks of certain encodings or inherent
visuo-cognitive biases.</li>
<li><strong>Perform precise statistical analysis to confirm findings identified when
inspecting plots.</strong> Examine the underlying data and devise statistical tests or
procedures to verify and quantify patterns you see.</li>
<li><strong>Informative is better than beautiful.</strong> Although it may be tempting to make
plots beautiful, this should not be at the expense of clarity. Make them clear
first, then make them beautiful.</li>
<li><strong>No plot is better than a useless plot.</strong> Some plots, while pretty, are
useless in a literal sense; the information contained within the plot is of no
use in either devising followup experiments or gaining insight into the data.</li>
<li><strong>Sometimes a table is the right way to present data.</strong> Tables may be
considered boring, but they are the most accurate representation of the data
(i.e. they ARE the data). Especially for relatively small datasets, a table may
do the job of communicating results better than any plot.</li>
<li><strong>If there is text on the plot, you should be able to read it.</strong> Many results
in biological data analysis have hundreds or thousands of individual points,
each of which has a label of some sort. If the font size for labels is too small
to read, or they substantially overlap each other making them illegible, either
the plot needs to change or the labels should not be drawn. Same goes for axis
labels.</li>
<li><strong>(Almost) every plot should have axis labels, tick marks and labels, a
legend, and a title.</strong> There are a few exceptions to this, but a plot should be
mostly understandable without any help from the text or caption. If it isn’t,
it’s probably a poor visualization strategy and can be improved.</li>
<li><strong>When using colors, be aware of color blindness.</strong> There are several known
forms of colorblindness in humans. Most default color schemes are “colorblind
friendly” but sometimes these color combinations can be counterinutitive. Check
sites like <a href="https://colorbrewer2.org/#type=sequential&amp;scheme=BuGn&amp;n=3">ColorBrewer</a>
to find color schemes that are likely to be colorblind friendly.</li>
<li><strong>Make differences visually appear as big as they mean.</strong> Scales on plots can
distort the meaning of the underlying data. For example, consider the following
two plots with identical data:</li>
</ol>
<div class="sourceCode" id="cb269"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb269-1"><a href="responsible-plotting.html#cb269-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(patchwork)</span>
<span id="cb269-2"><a href="responsible-plotting.html#cb269-2" aria-hidden="true" tabindex="-1"></a>data <span class="ot">&lt;-</span> <span class="fu">tibble</span>(</span>
<span id="cb269-3"><a href="responsible-plotting.html#cb269-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">percent=</span><span class="fu">c</span>(<span class="dv">86</span>,<span class="dv">88</span>,<span class="dv">87</span>,<span class="dv">90</span>,<span class="dv">93</span>,<span class="dv">89</span>),</span>
<span id="cb269-4"><a href="responsible-plotting.html#cb269-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">ID=</span><span class="fu">c</span>(<span class="st">&#39;A&#39;</span>,<span class="st">&#39;B&#39;</span>,<span class="st">&#39;C&#39;</span>,<span class="st">&#39;D&#39;</span>,<span class="st">&#39;E&#39;</span>,<span class="st">&#39;F&#39;</span>)</span>
<span id="cb269-5"><a href="responsible-plotting.html#cb269-5" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb269-6"><a href="responsible-plotting.html#cb269-6" aria-hidden="true" tabindex="-1"></a>g <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(data, <span class="fu">aes</span>(<span class="at">x=</span>ID,<span class="at">y=</span>percent)) <span class="sc">+</span></span>
<span id="cb269-7"><a href="responsible-plotting.html#cb269-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_bar</span>(<span class="at">stat=</span><span class="st">&quot;identity&quot;</span>)</span>
<span id="cb269-8"><a href="responsible-plotting.html#cb269-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb269-9"><a href="responsible-plotting.html#cb269-9" aria-hidden="true" tabindex="-1"></a>g <span class="sc">+</span> <span class="fu">coord_cartesian</span>(<span class="at">ylim=</span><span class="fu">c</span>(<span class="dv">85</span>,<span class="dv">95</span>)) <span class="sc">|</span> g</span></code></pre></div>
<p><img src="main_files/viz%20misleading%20scales-1.png" width="672" /></p>
<p>The left plot, which had its y axis limited to the values of the data, make the
differences between samples seem visually much larger than they actually are.
If a difference between 87 and 93 percent is in fact meaningful, then this might
be ok, but the context matters.</p>
<p>You will identify your own strategies over time as you practice data
visualization, and these rules of thumb might not work for you. There is no one
right way to do data visualization, and as long as you are dedicated to
representing your data accurately, there is no wrong way.</p>

</div>
</div>
<h3>References</h3>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-Cleveland1984-ay" class="csl-entry">
Cleveland, William S, and Robert McGill. 1984. <span>“Graphical Perception: Theory, Experimentation, and Application to the Development of Graphical Methods.”</span> <em>J. Am. Stat. Assoc.</em> 79 (387): 531–54.
</div>
</div>
<p style="text-align: center;">
<a href="data-visualization.html"><button class="btn btn-default">Previous</button></a>
<a href="dv-gg.html"><button class="btn btn-default">Next</button></a>
</p>
</div>
</div>


</div>

<script>

// add bootstrap table styles to pandoc tables
$(document).ready(function () {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
});

</script>

</body>
</html>
