<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta name="viewport" content="width=device-width, initial-scale=1" />
<meta property="og:title" content="6.3 Statistical Tests | BF591 - R for Biological Sciences" />
<meta property="og:type" content="book" />







<script type="text/x-mathjax-config">
MathJax.Hub.Config({
  TeX: { equationNumbers: { autoNumber: "AMS" } }
});
</script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<meta name="description" content="6.3 Statistical Tests | BF591 - R for Biological Sciences">

<title>6.3 Statistical Tests | BF591 - R for Biological Sciences</title>

<script src="libs/header-attrs-2.13/header-attrs.js"></script>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="libs/bootstrap-3.3.5/css/darkly.min.css" rel="stylesheet" />
<script src="libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="libs/navigation-1.1/tabsets.js"></script>


<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>


<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />

<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
/* show arrow before summary tag as in bootstrap
TODO: remove if boostrap in updated in html_document (rmarkdown#1485) */
details > summary {
  display: list-item;
  cursor: pointer;
}
</style>
</head>

<body>

<div class="container-fluid main-container">


<div class="row">
<div class="col-sm-12">
<div id="TOC">
<ul>
<li><a href="index.html#syllabus">Syllabus</a>
<ul>
<li><a href="course-schedule.html#course-schedule">Course Schedule</a></li>
<li><a href="instructors.html#instructors">Instructors</a></li>
<li><a href="office-hours.html#office-hours">Office Hours</a></li>
<li><a href="course-values-and-policies.html#course-values-and-policies">Course Values and Policies</a></li>
</ul></li>
<li><a href="introduction.html#introduction"><span class="toc-section-number">1</span> Introduction</a>
<ul>
<li><a href="who-this-book-is-for.html#who-this-book-is-for"><span class="toc-section-number">1.1</span> Who This Book Is For</a></li>
<li><a href="a-note-about-reinventing-the-wheel.html#a-note-about-reinventing-the-wheel"><span class="toc-section-number">1.2</span> A Note About Reinventing the Wheel</a></li>
<li><a href="sources-and-references.html#sources-and-references"><span class="toc-section-number">1.3</span> Sources and References</a></li>
</ul></li>
<li><a href="data-bio.html#data-bio"><span class="toc-section-number">2</span> Data in Biology</a>
<ul>
<li><a href="a-brief-history-of-data-in-molecular-biology.html#a-brief-history-of-data-in-molecular-biology"><span class="toc-section-number">2.1</span> A Brief History of Data in Molecular Biology</a></li>
<li><a href="biology-as-a-mature-data-science.html#biology-as-a-mature-data-science"><span class="toc-section-number">2.2</span> Biology as a Mature Data Science</a></li>
</ul></li>
<li><a href="preliminaries.html#preliminaries"><span class="toc-section-number">3</span> Preliminaries</a>
<ul>
<li><a href="prelim-r.html#prelim-r"><span class="toc-section-number">3.1</span> The R Language</a></li>
<li><a href="prelim-rstudio.html#prelim-rstudio"><span class="toc-section-number">3.2</span> RStudio</a></li>
<li><a href="prog-the-r-script.html#prog-the-r-script"><span class="toc-section-number">3.3</span> The R Script</a></li>
<li><a href="prog-workflow.html#prog-workflow"><span class="toc-section-number">3.4</span> The Scripting Workflow</a></li>
<li><a href="prelim-git.html#prelim-git"><span class="toc-section-number">3.5</span> git + github</a>
<ul>
<li><a href="prelim-git.html#motivation"><span class="toc-section-number">3.5.1</span> Motivation</a></li>
<li><a href="prelim-git.html#git"><span class="toc-section-number">3.5.2</span> git</a></li>
<li><a href="prelim-git.html#git-hosting-platforms-github"><span class="toc-section-number">3.5.3</span> Git hosting platforms (GitHub)</a></li>
</ul></li>
</ul></li>
<li><a href="prog-basics.html#prog-basics"><span class="toc-section-number">4</span> R Programming</a>
<ul>
<li><a href="before-you-begin.html#before-you-begin"><span class="toc-section-number">4.1</span> Before you begin</a></li>
<li><a href="introduction-1.html#introduction-1"><span class="toc-section-number">4.2</span> Introduction</a></li>
<li><a href="prog-r-syntax.html#prog-r-syntax"><span class="toc-section-number">4.3</span> R Syntax Basics</a></li>
<li><a href="prog-types.html#prog-types"><span class="toc-section-number">4.4</span> Basic Types of Values</a>
<ul>
<li><a href="prog-types.html#factors"><span class="toc-section-number">4.4.1</span> Factors</a></li>
</ul></li>
<li><a href="prog-struct.html#prog-struct"><span class="toc-section-number">4.5</span> Data Structures</a>
<ul>
<li><a href="prog-struct.html#vectors"><span class="toc-section-number">4.5.1</span> Vectors</a></li>
<li><a href="prog-struct.html#matrices"><span class="toc-section-number">4.5.2</span> Matrices</a></li>
<li><a href="prog-struct.html#lists-and-data-frames"><span class="toc-section-number">4.5.3</span> Lists and data frames</a></li>
</ul></li>
<li><a href="logical-tests-and-comparators.html#logical-tests-and-comparators"><span class="toc-section-number">4.6</span> Logical Tests and Comparators</a></li>
<li><a href="functions.html#functions"><span class="toc-section-number">4.7</span> Functions</a>
<ul>
<li><a href="functions.html#prog-dry"><span class="toc-section-number">4.7.1</span> DRY: Don’t Repeat Yourself</a></li>
<li><a href="functions.html#writing-your-own-functions"><span class="toc-section-number">4.7.2</span> Writing your own functions</a></li>
<li><a href="functions.html#scope"><span class="toc-section-number">4.7.3</span> Scope</a></li>
</ul></li>
<li><a href="iteration.html#iteration"><span class="toc-section-number">4.8</span> Iteration</a>
<ul>
<li><a href="iteration.html#vectorized-operations"><span class="toc-section-number">4.8.1</span> Vectorized operations</a></li>
<li><a href="iteration.html#functional-programming"><span class="toc-section-number">4.8.2</span> Functional programming</a></li>
<li><a href="iteration.html#apply-and-friends"><span class="toc-section-number">4.8.3</span> <code>apply()</code> and friends</a></li>
</ul></li>
<li><a href="installing-packages.html#installing-packages"><span class="toc-section-number">4.9</span> Installing Packages</a></li>
<li><a href="saving-and-loading-r-data.html#saving-and-loading-r-data"><span class="toc-section-number">4.10</span> Saving and Loading R Data</a></li>
<li><a href="prog-debugging.html#prog-debugging"><span class="toc-section-number">4.11</span> Troubleshooting and Debugging</a></li>
<li><a href="prog-style.html#prog-style"><span class="toc-section-number">4.12</span> Coding Style and Conventions</a>
<ul>
<li><a href="prog-style.html#is-my-code-correct"><span class="toc-section-number">4.12.1</span> Is my code correct?</a></li>
<li><a href="prog-style.html#does-my-code-follow-the-dry-principle"><span class="toc-section-number">4.12.2</span> Does my code follow the <span>DRY principle</span>?</a></li>
<li><a href="prog-style.html#did-i-choose-concise-but-descriptive-variable-and-function-names"><span class="toc-section-number">4.12.3</span> Did I choose concise but descriptive variable and function names?</a></li>
<li><a href="prog-style.html#did-i-use-indentation-and-naming-conventions-consistently-throughout-my-code"><span class="toc-section-number">4.12.4</span> Did I use indentation and naming conventions consistently throughout my code?</a></li>
<li><a href="prog-style.html#did-i-write-comments-especially-when-what-the-code-does-is-not-obvious"><span class="toc-section-number">4.12.5</span> Did I write comments, especially when what the code does is not obvious?</a></li>
<li><a href="prog-style.html#how-easy-would-it-be-for-someone-else-to-understand-my-code"><span class="toc-section-number">4.12.6</span> How easy would it be for someone else to understand my code?</a></li>
<li><a href="prog-style.html#is-my-code-easy-to-maintainchange"><span class="toc-section-number">4.12.7</span> Is my code easy to maintain/change?</a></li>
<li><a href="prog-style.html#the-styler-package"><span class="toc-section-number">4.12.8</span> The <code>styler</code> package</a></li>
</ul></li>
</ul></li>
<li><a href="data-wrangling.html#data-wrangling"><span class="toc-section-number">5</span> Data Wrangling</a>
<ul>
<li><a href="the-tidyverse.html#the-tidyverse"><span class="toc-section-number">5.1</span> The Tidyverse</a></li>
<li><a href="dw-basics.html#dw-basics"><span class="toc-section-number">5.2</span> Tidyverse Basics</a></li>
<li><a href="importing-data.html#importing-data"><span class="toc-section-number">5.3</span> Importing Data</a></li>
<li><a href="data-tibble.html#data-tibble"><span class="toc-section-number">5.4</span> The tibble</a></li>
<li><a href="tidy-data.html#tidy-data"><span class="toc-section-number">5.5</span> Tidy Data</a></li>
<li><a href="pipes.html#pipes"><span class="toc-section-number">5.6</span> pipes</a></li>
<li><a href="arranging-data.html#arranging-data"><span class="toc-section-number">5.7</span> Arranging Data</a>
<ul>
<li><a href="arranging-data.html#dplyrmutate---create-new-columns-using-other-columns"><span class="toc-section-number">5.7.1</span> <code>dplyr::mutate()</code> - Create new columns using other columns</a></li>
<li><a href="arranging-data.html#stringr---working-with-character-values"><span class="toc-section-number">5.7.2</span> <code>stringr</code> - Working with character values</a></li>
<li><a href="arranging-data.html#dplyrselect---subset-columns-by-name"><span class="toc-section-number">5.7.3</span> <code>dplyr::select()</code> - Subset Columns by Name</a></li>
<li><a href="arranging-data.html#dplyrfilter---pick-rows-out-of-a-data-set"><span class="toc-section-number">5.7.4</span> <code>dplyr::filter()</code> - Pick rows out of a data set</a></li>
<li><a href="arranging-data.html#dplyrarrange---order-rows-based-on-their-values"><span class="toc-section-number">5.7.5</span> <code>dplyr::arrange()</code> - Order rows based on their values</a></li>
<li><a href="arranging-data.html#putting-it-all-together"><span class="toc-section-number">5.7.6</span> Putting it all together</a></li>
</ul></li>
<li><a href="grouping-data.html#grouping-data"><span class="toc-section-number">5.8</span> Grouping Data</a></li>
<li><a href="rearranging-data.html#rearranging-data"><span class="toc-section-number">5.9</span> Rearranging Data</a></li>
<li><a href="relational-data.html#relational-data"><span class="toc-section-number">5.10</span> Relational Data</a>
<ul>
<li><a href="relational-data.html#dealing-with-multiple-matches"><span class="toc-section-number">5.10.1</span> Dealing with multiple matches</a></li>
</ul></li>
</ul></li>
<li><a href="data-science.html#data-science"><span class="toc-section-number">6</span> Data Science</a>
<ul>
<li><a href="data-modeling.html#data-modeling"><span class="toc-section-number">6.1</span> Data Modeling</a>
<ul>
<li><a href="data-modeling.html#a-worked-modeling-example"><span class="toc-section-number">6.1.1</span> A Worked Modeling Example</a></li>
<li><a href="data-modeling.html#data-summarization"><span class="toc-section-number">6.1.2</span> Data Summarization</a></li>
<li><a href="data-modeling.html#linear-models"><span class="toc-section-number">6.1.3</span> Linear Models</a></li>
<li><a href="data-modeling.html#flavors-of-linear-models"><span class="toc-section-number">6.1.4</span> Flavors of Linear Models</a></li>
<li><a href="data-modeling.html#assessing-model-accuracy-."><span class="toc-section-number">6.1.5</span> Assessing Model Accuracy .</a></li>
</ul></li>
<li><a href="statistical-distributions.html#statistical-distributions"><span class="toc-section-number">6.2</span> Statistical Distributions</a>
<ul>
<li><a href="statistical-distributions.html#random-variables"><span class="toc-section-number">6.2.1</span> Random Variables</a></li>
<li><a href="statistical-distributions.html#statistical-distribution-basics"><span class="toc-section-number">6.2.2</span> Statistical Distribution Basics</a></li>
<li><a href="statistical-distributions.html#distributions-in-r"><span class="toc-section-number">6.2.3</span> Distributions in R</a></li>
<li><a href="statistical-distributions.html#discrete-distributions"><span class="toc-section-number">6.2.4</span> Discrete Distributions</a></li>
<li><a href="statistical-distributions.html#continuous-distributions"><span class="toc-section-number">6.2.5</span> Continuous Distributions</a></li>
<li><a href="statistical-distributions.html#empirical-distributions"><span class="toc-section-number">6.2.6</span> Empirical Distributions</a></li>
</ul></li>
<li><a href="statistical-tests.html#statistical-tests"><span class="toc-section-number">6.3</span> Statistical Tests</a>
<ul>
<li><a href="statistical-tests.html#what-is-a-statistical-test"><span class="toc-section-number">6.3.1</span> What is a statistical test</a></li>
<li><a href="statistical-tests.html#common-statistical-tests"><span class="toc-section-number">6.3.2</span> Common Statistical Tests</a></li>
<li><a href="statistical-tests.html#choosing-a-test"><span class="toc-section-number">6.3.3</span> Choosing a Test</a></li>
<li><a href="statistical-tests.html#p-values"><span class="toc-section-number">6.3.4</span> P-values</a></li>
<li><a href="statistical-tests.html#multiple-hypothesis-testing-."><span class="toc-section-number">6.3.5</span> Multiple Hypothesis Testing .</a></li>
<li><a href="statistical-tests.html#statistical-power"><span class="toc-section-number">6.3.6</span> Statistical power</a></li>
</ul></li>
<li><a href="exploratory-data-analysis.html#exploratory-data-analysis"><span class="toc-section-number">6.4</span> Exploratory Data Analysis</a>
<ul>
<li><a href="exploratory-data-analysis.html#principal-component-analysis"><span class="toc-section-number">6.4.1</span> Principal Component Analysis</a></li>
<li><a href="exploratory-data-analysis.html#cluster-analysis"><span class="toc-section-number">6.4.2</span> Cluster Analysis</a></li>
<li><a href="exploratory-data-analysis.html#k-means-."><span class="toc-section-number">6.4.3</span> k-means .</a></li>
<li><a href="exploratory-data-analysis.html#others"><span class="toc-section-number">6.4.4</span> Others</a></li>
</ul></li>
<li><a href="network-analysis-..html#network-analysis-."><span class="toc-section-number">6.5</span> Network Analysis .</a>
<ul>
<li><a href="network-analysis-..html#basic-network-analysis"><span class="toc-section-number">6.5.1</span> Basic network analysis</a></li>
<li><a href="network-analysis-..html#create-your-own"><span class="toc-section-number">6.5.2</span> Create your own</a></li>
</ul></li>
</ul></li>
<li><a href="data-visualization.html#data-visualization"><span class="toc-section-number">7</span> Data Visualization</a>
<ul>
<li><a href="responsible-plotting.html#responsible-plotting"><span class="toc-section-number">7.1</span> Responsible Plotting</a>
<ul>
<li><a href="responsible-plotting.html#visualization-principles"><span class="toc-section-number">7.1.1</span> Visualization Principles</a></li>
<li><a href="responsible-plotting.html#human-visual-perception"><span class="toc-section-number">7.1.2</span> Human Visual Perception</a></li>
<li><a href="responsible-plotting.html#visual-encodings"><span class="toc-section-number">7.1.3</span> Visual Encodings</a></li>
<li><a href="responsible-plotting.html#some-opinionated-rules-of-thumb"><span class="toc-section-number">7.1.4</span> Some Opinionated Rules of Thumb</a></li>
</ul></li>
<li><a href="dv-gg.html#dv-gg"><span class="toc-section-number">7.2</span> Grammar of Graphics</a>
<ul>
<li><a href="dv-gg.html#ggplot-mechanics"><span class="toc-section-number">7.2.1</span> <code>ggplot</code> mechanics</a></li>
</ul></li>
<li><a href="plotting-one-dimension.html#plotting-one-dimension"><span class="toc-section-number">7.3</span> Plotting One Dimension</a>
<ul>
<li><a href="plotting-one-dimension.html#dv-bar"><span class="toc-section-number">7.3.1</span> Bar chart</a></li>
<li><a href="plotting-one-dimension.html#lollipop-plots"><span class="toc-section-number">7.3.2</span> Lollipop plots</a></li>
<li><a href="plotting-one-dimension.html#stacked-area-charts"><span class="toc-section-number">7.3.3</span> Stacked Area charts</a></li>
<li><a href="plotting-one-dimension.html#parallel-coordinate-plots"><span class="toc-section-number">7.3.4</span> Parallel Coordinate plots</a></li>
</ul></li>
<li><a href="visualizing-distributions.html#visualizing-distributions"><span class="toc-section-number">7.4</span> Visualizing Distributions</a>
<ul>
<li><a href="visualizing-distributions.html#histogram"><span class="toc-section-number">7.4.1</span> Histogram</a></li>
<li><a href="visualizing-distributions.html#density"><span class="toc-section-number">7.4.2</span> Density</a></li>
<li><a href="visualizing-distributions.html#boxplot"><span class="toc-section-number">7.4.3</span> Boxplot</a></li>
<li><a href="visualizing-distributions.html#violin-plot"><span class="toc-section-number">7.4.4</span> Violin plot</a></li>
<li><a href="visualizing-distributions.html#beeswarm-plot"><span class="toc-section-number">7.4.5</span> Beeswarm plot</a></li>
<li><a href="visualizing-distributions.html#ridgeline"><span class="toc-section-number">7.4.6</span> Ridgeline</a></li>
</ul></li>
<li><a href="plotting-two-or-more-dimensions.html#plotting-two-or-more-dimensions"><span class="toc-section-number">7.5</span> Plotting Two or More Dimensions</a>
<ul>
<li><a href="plotting-two-or-more-dimensions.html#scatter-plots"><span class="toc-section-number">7.5.1</span> Scatter Plots</a></li>
<li><a href="plotting-two-or-more-dimensions.html#bubble-plots"><span class="toc-section-number">7.5.2</span> Bubble Plots</a></li>
<li><a href="plotting-two-or-more-dimensions.html#connected-scatter-plots"><span class="toc-section-number">7.5.3</span> Connected Scatter Plots</a></li>
<li><a href="plotting-two-or-more-dimensions.html#line-plots"><span class="toc-section-number">7.5.4</span> Line Plots</a></li>
<li><a href="plotting-two-or-more-dimensions.html#parallel-coordinate-plots-1"><span class="toc-section-number">7.5.5</span> Parallel Coordinate Plots</a></li>
<li><a href="plotting-two-or-more-dimensions.html#heatmaps"><span class="toc-section-number">7.5.6</span> Heatmaps</a></li>
<li><a href="plotting-two-or-more-dimensions.html#how-to-use-heatmaps-responsibly"><span class="toc-section-number">7.5.7</span> How To Use Heatmaps Responsibly</a></li>
</ul></li>
<li><a href="other-kind-of-plots.html#other-kind-of-plots"><span class="toc-section-number">7.6</span> Other Kind of Plots</a>
<ul>
<li><a href="other-kind-of-plots.html#dendrograms"><span class="toc-section-number">7.6.1</span> Dendrograms</a></li>
<li><a href="other-kind-of-plots.html#chord-diagrams-and-circos-plots"><span class="toc-section-number">7.6.2</span> Chord Diagrams and Circos Plots</a></li>
</ul></li>
<li><a href="multiple-plots.html#multiple-plots"><span class="toc-section-number">7.7</span> Multiple Plots</a>
<ul>
<li><a href="multiple-plots.html#facet-wrapping"><span class="toc-section-number">7.7.1</span> Facet wrapping</a></li>
<li><a href="multiple-plots.html#multipanel-figures"><span class="toc-section-number">7.7.2</span> Multipanel Figures</a></li>
</ul></li>
<li><a href="publication-ready-plots.html#publication-ready-plots"><span class="toc-section-number">7.8</span> Publication Ready Plots</a>
<ul>
<li><a href="publication-ready-plots.html#ggplot-themes"><span class="toc-section-number">7.8.1</span> <code>ggplot</code> Themes</a></li>
<li><a href="publication-ready-plots.html#exporting-to-svg"><span class="toc-section-number">7.8.2</span> Exporting to SVG</a></li>
</ul></li>
<li><a href="network-visualization.html#network-visualization"><span class="toc-section-number">7.9</span> Network visualization</a>
<ul>
<li><a href="network-visualization.html#interactive-network-plots"><span class="toc-section-number">7.9.1</span> Interactive network plots</a></li>
</ul></li>
</ul></li>
<li><a href="biology-bioinformatics.html#biology-bioinformatics"><span class="toc-section-number">8</span> Biology &amp; Bioinformatics</a>
<ul>
<li><a href="bio-r.html#bio-r"><span class="toc-section-number">8.1</span> R in Biology</a></li>
<li><a href="biological-data-overview.html#biological-data-overview"><span class="toc-section-number">8.2</span> Biological Data Overview</a>
<ul>
<li><a href="biological-data-overview.html#bio-data-types"><span class="toc-section-number">8.2.1</span> Types of Biological Data</a></li>
<li><a href="biological-data-overview.html#csv-files"><span class="toc-section-number">8.2.2</span> CSV Files</a></li>
<li><a href="biological-data-overview.html#biological-data-is-not-tidy"><span class="toc-section-number">8.2.3</span> Biological data is NOT Tidy!</a></li>
</ul></li>
<li><a href="bio-bioconductor.html#bio-bioconductor"><span class="toc-section-number">8.3</span> Bioconductor</a></li>
<li><a href="microarrays.html#microarrays"><span class="toc-section-number">8.4</span> Microarrays</a></li>
<li><a href="high-throughput-sequencing.html#high-throughput-sequencing"><span class="toc-section-number">8.5</span> High Throughput Sequencing</a>
<ul>
<li><a href="high-throughput-sequencing.html#raw-hts-data"><span class="toc-section-number">8.5.1</span> Raw HTS Data</a></li>
<li><a href="high-throughput-sequencing.html#preliminary-hts-data-analysis"><span class="toc-section-number">8.5.2</span> Preliminary HTS Data Analysis</a></li>
<li><a href="high-throughput-sequencing.html#count-data"><span class="toc-section-number">8.5.3</span> Count Data</a></li>
</ul></li>
<li><a href="gene-identifiers.html#gene-identifiers"><span class="toc-section-number">8.6</span> Gene Identifiers</a>
<ul>
<li><a href="gene-identifiers.html#gene-identifier-systems"><span class="toc-section-number">8.6.1</span> Gene Identifier Systems</a></li>
<li><a href="gene-identifiers.html#mapping-between-identifier-systems"><span class="toc-section-number">8.6.2</span> Mapping Between Identifier Systems</a></li>
<li><a href="gene-identifiers.html#mapping-homologs"><span class="toc-section-number">8.6.3</span> Mapping Homologs</a></li>
</ul></li>
<li><a href="gene-expression.html#gene-expression"><span class="toc-section-number">8.7</span> Gene Expression</a>
<ul>
<li><a href="gene-expression.html#gene-expression-data-in-bioconductor"><span class="toc-section-number">8.7.1</span> Gene Expression Data in Bioconductor</a></li>
<li><a href="gene-expression.html#differential-expression-analysis"><span class="toc-section-number">8.7.2</span> Differential Expression Analysis</a></li>
<li><a href="gene-expression.html#microarray-gene-expression-data"><span class="toc-section-number">8.7.3</span> Microarray Gene Expression Data</a></li>
<li><a href="gene-expression.html#differential-expression-microarrays-limma"><span class="toc-section-number">8.7.4</span> Differential Expression: Microarrays (limma)</a></li>
<li><a href="gene-expression.html#rnaseq"><span class="toc-section-number">8.7.5</span> RNASeq</a></li>
<li><a href="gene-expression.html#rnaseq-gene-expression-data"><span class="toc-section-number">8.7.6</span> RNASeq Gene Expression Data</a></li>
<li><a href="gene-expression.html#filtering-counts"><span class="toc-section-number">8.7.7</span> Filtering Counts</a></li>
<li><a href="gene-expression.html#count-distributions"><span class="toc-section-number">8.7.8</span> Count Distributions</a></li>
<li><a href="gene-expression.html#differential-expression-rnaseq"><span class="toc-section-number">8.7.9</span> Differential Expression: RNASeq</a></li>
</ul></li>
<li><a href="gene-set-enrichment-analysis.html#gene-set-enrichment-analysis"><span class="toc-section-number">8.8</span> Gene Set Enrichment Analysis</a>
<ul>
<li><a href="gene-set-enrichment-analysis.html#gene-sets"><span class="toc-section-number">8.8.1</span> Gene Sets</a></li>
<li><a href="gene-set-enrichment-analysis.html#working-with-gene-sets-in-r"><span class="toc-section-number">8.8.2</span> Working with gene sets in R</a></li>
<li><a href="gene-set-enrichment-analysis.html#over-representation-analysis"><span class="toc-section-number">8.8.3</span> Over-representation Analysis</a></li>
<li><a href="gene-set-enrichment-analysis.html#rank-based-analysis"><span class="toc-section-number">8.8.4</span> Rank-based Analysis</a></li>
<li><a href="gene-set-enrichment-analysis.html#fgsea"><span class="toc-section-number">8.8.5</span> fgsea</a></li>
</ul></li>
<li><a href="biological-networks-..html#biological-networks-."><span class="toc-section-number">8.9</span> Biological Networks .</a>
<ul>
<li><a href="biological-networks-..html#biological-pathways-."><span class="toc-section-number">8.9.1</span> Biological Pathways .</a></li>
<li><a href="biological-networks-..html#gene-regulatory-networks-."><span class="toc-section-number">8.9.2</span> Gene Regulatory Networks .</a></li>
<li><a href="biological-networks-..html#protein-protein-interaction-network-."><span class="toc-section-number">8.9.3</span> Protein-Protein Interaction Network .</a></li>
<li><a href="biological-networks-..html#wgcna-."><span class="toc-section-number">8.9.4</span> WGCNA .</a></li>
</ul></li>
</ul></li>
<li><a href="engineering.html#engineering"><span class="toc-section-number">9</span> EngineeRing</a>
<ul>
<li><a href="unit-testing.html#unit-testing"><span class="toc-section-number">9.1</span> Unit Testing</a></li>
<li><a href="toolification.html#toolification"><span class="toc-section-number">9.2</span> Toolification</a>
<ul>
<li><a href="toolification.html#the-r-interpreter"><span class="toc-section-number">9.2.1</span> The R Interpreter</a></li>
<li><a href="toolification.html#rscript"><span class="toc-section-number">9.2.2</span> <code>Rscript</code></a></li>
<li><a href="toolification.html#commandargs"><span class="toc-section-number">9.2.3</span> <code>commandArgs()</code></a></li>
</ul></li>
<li><a href="parallel-processing.html#parallel-processing"><span class="toc-section-number">9.3</span> Parallel Processing</a>
<ul>
<li><a href="parallel-processing.html#brief-introduction-to-parallelization"><span class="toc-section-number">9.3.1</span> Brief Introduction to Parallelization</a></li>
<li><a href="parallel-processing.html#apply-and-friends-are-pleasingly-parallel"><span class="toc-section-number">9.3.2</span> <code>apply</code> and Friends Are Pleasingly Parallel</a></li>
<li><a href="parallel-processing.html#the-parallel-package"><span class="toc-section-number">9.3.3</span> The <code>parallel</code> package</a></li>
</ul></li>
<li><a href="object-orientation-in-r.html#object-orientation-in-r"><span class="toc-section-number">9.4</span> Object Orientation in R</a></li>
<li><a href="r-packages.html#r-packages"><span class="toc-section-number">9.5</span> R Packages</a></li>
</ul></li>
<li><a href="rshiny.html#rshiny"><span class="toc-section-number">10</span> RShiny</a>
<ul>
<li><a href="rshiny-introduction.html#rshiny-introduction"><span class="toc-section-number">10.1</span> RShiny Introduction</a>
<ul>
<li><a href="rshiny-introduction.html#how-does-a-web-page-work"><span class="toc-section-number">10.1.1</span> How does a web page work?</a></li>
</ul></li>
<li><a href="application-structure.html#application-structure"><span class="toc-section-number">10.2</span> Application Structure</a>
<ul>
<li><a href="application-structure.html#ui-design"><span class="toc-section-number">10.2.1</span> UI Design</a></li>
<li><a href="application-structure.html#server-functionality"><span class="toc-section-number">10.2.2</span> Server Functionality</a></li>
</ul></li>
<li><a href="reactivity.html#reactivity"><span class="toc-section-number">10.3</span> Reactivity</a></li>
<li><a href="publishing.html#publishing"><span class="toc-section-number">10.4</span> Publishing</a></li>
</ul></li>
<li><a href="communicating-with-r.html#communicating-with-r"><span class="toc-section-number">11</span> Communicating with R</a>
<ul>
<li><a href="rmarkdown-knitr.html#rmarkdown-knitr"><span class="toc-section-number">11.1</span> RMarkdown &amp; knitr</a></li>
<li><a href="bookdown.html#bookdown"><span class="toc-section-number">11.2</span> bookdown</a></li>
</ul></li>
<li><a href="contribution-guide.html#contribution-guide"><span class="toc-section-number">12</span> Contribution Guide</a>
<ul>
<li><a href="custom-blocks.html#custom-blocks"><span class="toc-section-number">12.1</span> Custom blocks</a></li>
</ul></li>
<li><a href="assignments.html#assignments">Assignments</a>
<ul>
<li><a href="assignment-format.html#assignment-format">Assignment Format</a></li>
<li><a href="starting-an-assignment.html#starting-an-assignment">Starting an Assignment</a>
<ul>
<li><a href="starting-an-assignment.html#starting-your-own-git-repository-and-cloning-it">Starting your own <code>git</code> repository and cloning it</a></li>
<li><a href="starting-an-assignment.html#contents">Contents</a></li>
<li><a href="starting-an-assignment.html#github-tutorial">GitHub tutorial</a></li>
<li><a href="starting-an-assignment.html#committing-and-pushing-with-r-and-without">Committing and Pushing with R (and without)</a></li>
<li><a href="starting-an-assignment.html#using-rstudio">Using RStudio</a></li>
<li><a href="starting-an-assignment.html#using-the-command-line">Using the command line</a></li>
</ul></li>
<li><a href="assignment-1.html#assignment-1">Assignment 1</a>
<ul>
<li><a href="assignment-1.html#problem-statement">Problem Statement</a></li>
<li><a href="assignment-1.html#learning-objectives-and-skill-list">Learning Objectives and Skill List</a></li>
<li><a href="assignment-1.html#instructions">Instructions</a></li>
<li><a href="assignment-1.html#hints">Hints</a></li>
</ul></li>
<li><a href="assignment-2.html#assignment-2">Assignment 2</a>
<ul>
<li><a href="assignment-2.html#problem-statement-1">Problem Statement</a></li>
<li><a href="assignment-2.html#learning-objectives">Learning Objectives</a></li>
<li><a href="assignment-2.html#skill-list">Skill List</a></li>
<li><a href="assignment-2.html#instructions-1">Instructions</a></li>
<li><a href="assignment-2.html#function-details">Function Details</a></li>
</ul></li>
<li><a href="assignment-3.html#assignment-3">Assignment 3</a>
<ul>
<li><a href="assignment-3.html#problem-statement-2">Problem Statement</a></li>
<li><a href="assignment-3.html#learning-objectives-1">Learning Objectives</a></li>
<li><a href="assignment-3.html#skill-list-1">Skill List</a></li>
<li><a href="assignment-3.html#background-on-microarrays">Background on Microarrays</a></li>
<li><a href="assignment-3.html#background-on-principal-component-analysis">Background on Principal Component Analysis</a></li>
<li><a href="assignment-3.html#marisa-et-al.-gene-expression-classification-of-colon-cancer-into-molecular-subtypes-characterization-validation-and-prognostic-value.-plos-medicine-may-2013.-pmid-23700391">Marisa et al. Gene Expression Classification of Colon Cancer into Molecular Subtypes: Characterization, Validation, and Prognostic Value. PLoS Medicine, May 2013. PMID: 23700391</a></li>
<li><a href="assignment-3.html#scaling-data-using-r-scale">Scaling data using R <code>scale()</code></a></li>
<li><a href="assignment-3.html#proportion-of-variance-explained">Proportion of variance explained</a></li>
<li><a href="assignment-3.html#plotting-and-visualization-of-pca">Plotting and visualization of PCA</a></li>
<li><a href="assignment-3.html#hierarchical-clustering-and-heatmaps">Hierarchical Clustering and Heatmaps</a></li>
<li><a href="assignment-3.html#references">References</a></li>
</ul></li>
<li><a href="assignment-4.html#assignment-4">Assignment 4</a>
<ul>
<li><a href="assignment-4.html#problem-statement-3">Problem Statement</a></li>
<li><a href="assignment-4.html#learning-objectives-2">Learning Objectives</a></li>
<li><a href="assignment-4.html#skill-list-2">Skill List</a></li>
<li><a href="assignment-4.html#instructions-2">Instructions</a></li>
<li><a href="assignment-4.html#deliverables">Deliverables</a></li>
<li><a href="assignment-4.html#function-details-1">Function Details</a></li>
<li><a href="assignment-4.html#hints-1">Hints</a></li>
</ul></li>
<li><a href="assignment-5.html#assignment-5">Assignment 5</a>
<ul>
<li><a href="assignment-5.html#problem-statement-4">Problem Statement</a></li>
<li><a href="assignment-5.html#learning-objectives-3">Learning Objectives</a></li>
<li><a href="assignment-5.html#skill-list-3">Skill List</a></li>
<li><a href="assignment-5.html#deseq2-background">DESeq2 Background</a></li>
<li><a href="assignment-5.html#generating-a-counts-matrix">Generating a counts matrix</a></li>
<li><a href="assignment-5.html#prefiltering-counts-matrix">Prefiltering Counts matrix</a></li>
<li><a href="assignment-5.html#median-of-ratios-normalization">Median-of-ratios normalization</a></li>
<li><a href="assignment-5.html#deseq2-preparation">DESeq2 preparation</a></li>
<li><a href="assignment-5.html#omeara-et-al.-transcriptional-reversion-of-cardiac-myocyte-fate-during-mammalian-cardiac-regeneration.-circ-res.-feb-2015.-pmid-25477501l">O’Meara et al. Transcriptional Reversion of Cardiac Myocyte Fate During Mammalian Cardiac Regeneration. Circ Res. Feb 2015. PMID: 25477501l</a></li>
<li><a href="assignment-5.html#reading-and-subsetting-the-data-from-verse_counts.tsv-and-sample_metadata.csv">1. Reading and subsetting the data from verse_counts.tsv and sample_metadata.csv</a></li>
<li><a href="assignment-5.html#running-deseq2">2. Running DESeq2</a></li>
<li><a href="assignment-5.html#annotating-results-to-construct-a-labeled-volcano-plot">3. Annotating results to construct a labeled volcano plot</a></li>
<li><a href="assignment-5.html#diagnostic-plot-of-the-raw-p-values-for-all-genes">4. Diagnostic plot of the raw p-values for all genes</a></li>
<li><a href="assignment-5.html#plotting-the-logfoldchanges-for-differentially-expressed-genes">5. Plotting the LogFoldChanges for differentially expressed genes</a></li>
<li><a href="assignment-5.html#the-choice-of-fdr-cutoff-depends-on-cost">The choice of FDR cutoff depends on cost</a></li>
<li><a href="assignment-5.html#plotting-the-normalized-counts-of-differentially-expressed-genes">6. Plotting the normalized counts of differentially expressed genes</a></li>
<li><a href="assignment-5.html#volcano-plot-to-visualize-differential-expression-results">7. Volcano Plot to visualize differential expression results</a></li>
<li><a href="assignment-5.html#running-fgsea-vignette">8. Running fgsea <span>vignette</span></a></li>
<li><a href="assignment-5.html#plotting-the-top-ten-positive-nes-and-top-ten-negative-nes-pathways">9. Plotting the top ten positive NES and top ten negative NES pathways</a></li>
<li><a href="assignment-5.html#references-1">References</a></li>
</ul></li>
<li><a href="assignment-6.html#assignment-6">Assignment 6</a>
<ul>
<li><a href="assignment-6.html#problem-statement-5">Problem Statement</a></li>
<li><a href="assignment-6.html#learning-objectives-4">Learning Objectives</a></li>
<li><a href="assignment-6.html#skill-list-4">Skill List</a></li>
<li><a href="assignment-6.html#instructions-3">Instructions</a></li>
<li><a href="assignment-6.html#function-details-2">Function Details</a></li>
</ul></li>
<li><a href="assignment-7.html#assignment-7">Assignment 7</a>
<ul>
<li><a href="assignment-7.html#problem-statement-6">Problem Statement</a></li>
<li><a href="assignment-7.html#learning-objectives-5">Learning Objectives</a></li>
<li><a href="assignment-7.html#skill-list-5">Skill List</a></li>
<li><a href="assignment-7.html#instructions-4">Instructions</a></li>
<li><a href="assignment-7.html#function-details-3">“Function” Details</a></li>
</ul></li>
</ul></li>
<li class="appendix"><span><b>Appendix</b></span></li>
<li><a href="class-outlines.html#class-outlines"><span class="toc-section-number">A</span> Class Outlines</a>
<ul>
<li><a href="week-1.html#week-1"><span class="toc-section-number">A.1</span> Week 1</a></li>
<li><a href="week-2.html#week-2"><span class="toc-section-number">A.2</span> Week 2</a></li>
<li><a href="week-3.html#week-3"><span class="toc-section-number">A.3</span> Week 3</a></li>
<li><a href="week-4.html#week-4"><span class="toc-section-number">A.4</span> Week 4</a></li>
<li><a href="week-5.html#week-5"><span class="toc-section-number">A.5</span> Week 5</a></li>
<li><a href="week-6.html#week-6"><span class="toc-section-number">A.6</span> Week 6</a></li>
<li><a href="week-7.html#week-7"><span class="toc-section-number">A.7</span> Week 7</a></li>
<li><a href="week-8.html#week-8"><span class="toc-section-number">A.8</span> Week 8</a></li>
<li><a href="week-9.html#week-9"><span class="toc-section-number">A.9</span> Week 9</a></li>
<li><a href="week-10.html#week-10"><span class="toc-section-number">A.10</span> Week 10</a></li>
<li><a href="week-11.html#week-11"><span class="toc-section-number">A.11</span> Week 11</a></li>
<li><a href="week-12.html#week-12"><span class="toc-section-number">A.12</span> Week 12</a></li>
<li><a href="week-13.html#week-13"><span class="toc-section-number">A.13</span> Week 13</a></li>
<li><a href="week-14.html#week-14"><span class="toc-section-number">A.14</span> Week 14</a></li>
</ul></li>
</ul>
</div>
</div>
</div>
<div class="row">
<div class="col-sm-12">
<div id="statistical-tests" class="section level2" number="6.3">
<h2><span class="header-section-number">6.3</span> Statistical Tests</h2>
<div id="what-is-a-statistical-test" class="section level3" number="6.3.1">
<h3><span class="header-section-number">6.3.1</span> What is a statistical test</h3>
<p>A statistical test is an application of mathematics that we used to analyze
quantitative data. They can be described as a way to understand how the
independent variable(s) of the study affect the outcome of the dependent
variable. The kind of statistical test appropriate for a study depends on a
number of factors including variables characteristics, study design, and data
distribution. The independent variable is (are) the variable(s) being controlled
in the study to determine its effects on the dependent variable(s). Not all
studies have independent variable(s).</p>
<div id="study-design" class="section level4 unnumbered">
<h4>Study design</h4>
<p>The number of conditions a study has corresponds to the number of levels the
independent variable has, for example 1 condition = 1 level, 2 conditions = 2
levels, and so on. This is different from the number of independent variables a
study has. If a study has multiple independent variables, each will have its own
number of levels. Statistical tests can be categorized as to whether they can
handle 1, 2, or 2+ levels.</p>
<p>If you have 2 or more levels, then the type of grouping will need to be
considered: are they between-subjects* or repeated-measures**. A between-subject
design means that each participant undergoes one condition whereas a
repeated-measures design has each participant undergo all conditions. A third
type exists called matched groups, where different subjects undergo different
conditions but are matched to each other in some important way, and may be
treated as a repeated-measure design when selecting for a statistical test</p>
<p><small>
*Independent design = between-subjects = between-groups; not to be confused with the variable.
**Dependent design = repeated-measures = within-subjects = within-groups; also not to be confused with the variable.
</small></p>
</div>
<div id="variables" class="section level4 unnumbered">
<h4>Variables</h4>
<p>The characteristics of your variables, both independent (predictor) and
dependent (descriptor), will help you choose which statistical test to use.
Specifically, the number of each, their type, and the level of measurement, will
help you narrow your choices to select the appropriate test.</p>
<div id="number-of-variables" class="section level5 unnumbered">
<h5>Number of Variables</h5>
<p>The number of both independent and dependent variables will affect which test
you select. Test selection will differ whether you have 1 or 2+ dependent
variables and 0, 1, or 2+ independent variables.</p>
</div>
<div id="types-of-data-continuous-and-categorical" class="section level5 unnumbered">
<h5>Types of Data (Continuous and Categorical)</h5>
<p>Typically in bioinformatics, data is subdivided into 3 categories: Continuous,
Categorical, and Binary. Continuous data is data that can take real number
values within either a finite or infinite range. For example, height is a
continuous variable: 152.853288cm (60.1783 in), 182.9704cm (72.0356 in),
172.7cm(68in), 163cm(64.2in) and 145.647cm (57.3413 in) are all technically
valid heights. Categorical data is data that can be divided to categories or
groups such as (high, medium, and low) or (under10, 10-29, and 30+). Binary
data, data that can either be one thing or another (T/D, Y/N, 0/1, etc) falls
under the Categorical data umbrella though may get mentioned separately
elsewhere. Data that exists in sequential, positive integer form- such as number
of siblings- is called Discrete data (bonus 4th category!) but typically ends up
being treated as categorical data.</p>
</div>
<div id="levels-of-measurement" class="section level5 unnumbered">
<h5>Levels of Measurement</h5>
<p>The main levels of measurement in we use in statistics are Ratio, Interval,
Nominal, and Ordinal. Both Ratio and Interval levels have distance between
measurements defined; the biggest difference between the two is that Ratio
measurements have a meaningful zero value (and no negative numbers). Height in
inches or cm, Kelvin, and number of students in a class are all Ratio
measurements. Interval measurements do not have a meaningful zero. Celsius and
Fahrenheit both have arbitrary 0s- the freezing point of pure and (a specific
kind of) ocean water- making most standard temperature measurements type
Interval. Ordinal measurements have a meaningful order to their values but have
variable/imprecise distances between measurements, like socioeconomic status
(upper, middle, and lower) and days since diagnosis (under 10, 10-39, 40-99,
100+). Nominal measurements do not have meaningful order to their values, like
country of origin and yes/no. Ratio and Interval measurements are continuous
variables while Nominal and Ordinal measurements are categorical variables.</p>
</div>
</div>
<div id="data-distribution-parametric-vs-nonparametric" class="section level4 unnumbered">
<h4>Data Distribution (Parametric vs Nonparametric)</h4>
<p>This is the third factor to keep in mind for test selection and only applicable
for NON-nominal dependent variables. Parametric tests make the assumption that
the population the data is sourced from has a normal or Gaussian distribution.
These are powerful tests because of their data distribution assumption, with the
downside of only being able to be used in select cases. Nonparametric tests do
not assume a normal distribution and therefore can be used in a wide range of
cases, thought they are less likely to find significance in the results.</p>
</div>
</div>
<div id="common-statistical-tests" class="section level3" number="6.3.2">
<h3><span class="header-section-number">6.3.2</span> Common Statistical Tests</h3>
<p>Here are some common statistical tests and a quick overview as to how to run
them in R. If you would like more information about a specific test, links are
included in the descriptions</p>
<p>These are only some statistical tests. <a href="https://stats.oarc.ucla.edu/other/mult-pkg/whatstat/">Here’s a link to where you can find a
few more</a></p>
<p><a href="https://en.wikipedia.org/wiki/Category:Statistical_tests">And here’s a link to Wikipedia’s list of statistical tests if you want to
overload on statistical tests because it’s your new
hobby</a></p>
<div id="one-sample-t-test" class="section level4 unnumbered">
<h4><strong>One-Sample T-Test</strong></h4>
<p><em>Dependent Variables</em>: 1, continuous [ratio and interval]</p>
<p><em>Independent Variables</em>: 0 variables</p>
<p><em>Design</em>: 1 group, 1 level</p>
<p><em>Parametric</em>: Yes</p>
<div class="box readmore">
<p><a href="https://sphweb.bumc.bu.edu/otlt/MPH-Modules/PH717-QuantCore/PH717-Module7-T-tests/PH717-Module7-T-tests4.html">More t-test information shamelessly linked from SPH</a>
<a href="https://sphweb.bumc.bu.edu/otlt/mph-modules/bs/bs704_hypothesistest-means-proportions/BS704_HypothesisTest-Means-Proportions2.html#headingtaglink_1">Null and Alternate (Research) Hypotheses</a>
<a href="https://sphweb.bumc.bu.edu/otlt/mph-modules/bs/bs704_probability/bs704_probability9.html">Z-values</a>(not mentioned but are in texts that fully explain t-tests)
One-Sample t-tests come in 3 flavors: two-tailed test, one-tailed test (upper tail), and one-tailed test (lower tail).</p>
</div>
<div class="sourceCode" id="cb210"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb210-1"><a href="statistical-tests.html#cb210-1" aria-hidden="true" tabindex="-1"></a><span class="co"># One-sample t-test</span></span>
<span id="cb210-2"><a href="statistical-tests.html#cb210-2" aria-hidden="true" tabindex="-1"></a><span class="fu">t.test</span>(x, <span class="co">#numeric vector of data </span></span>
<span id="cb210-3"><a href="statistical-tests.html#cb210-3" aria-hidden="true" tabindex="-1"></a>       <span class="at">mu =</span> <span class="dv">0</span>, <span class="co">#True value of the mean. Defaults to 0</span></span>
<span id="cb210-4"><a href="statistical-tests.html#cb210-4" aria-hidden="true" tabindex="-1"></a>       <span class="at">alternative =</span> <span class="st">&quot;two.sided&quot;</span>, <span class="co"># (&quot;two.sided&quot;, &quot;less&quot;, &quot;greater&quot;) depending on which you want to use. Defaults to &quot;two.sided</span></span>
<span id="cb210-5"><a href="statistical-tests.html#cb210-5" aria-hidden="true" tabindex="-1"></a>       <span class="at">conf.level =</span> <span class="fl">0.95</span> <span class="co">#1-alpha. Defaults to 0.95</span></span>
<span id="cb210-6"><a href="statistical-tests.html#cb210-6" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div>
<p>The <code>t.test()</code> function allows for multiple kinds of t-tests. To use this
function to perform a one sample t-test, you will need a numeric vector of the
data you want to run the data on, the mean you want to compare it to, your
confidence level (1 - alpha), and your alternative hypothesis. The t-test, like
other kinds of t-tests, will compare the means by calculating the t-statistic,
p-value and confidence intervals of your data’s mean.</p>
<p><em>t</em>: The calculated t-statistic, uses formula : (mean(x) - mu)/(sd(x)*sqrt(n_samples))</p>
<p><em>df</em>: Degrees of Freedom. Calculated by: n_samples-1</p>
<p><em>p</em>: The p-value of the t-test as determined by the t-statistic and df on the <a href="https://sphweb.bumc.bu.edu/otlt/mph-modules/bs/bs704_hypothesistest-means-proportions/t-table.pdf">t-distribution table</a></p>
<p><em>(conf.level) percent confidence interval</em>: calculated interval where the true mean(x) is with conf.level % certainty.</p>
<p><em>sample estimates, mean of x</em>: The calculated mean of x</p>
<p>Accepting or rejecting the null hypothesis is a matter of determining if mean(x) is outside the percent confidence interval range, if it is then the p-value will determine the significance of the results.</p>
<div class="sourceCode" id="cb211"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb211-1"><a href="statistical-tests.html#cb211-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Example</span></span>
<span id="cb211-2"><a href="statistical-tests.html#cb211-2" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">10</span>)</span>
<span id="cb211-3"><a href="statistical-tests.html#cb211-3" aria-hidden="true" tabindex="-1"></a>x0 <span class="ot">&lt;-</span> <span class="fu">rnorm</span>(<span class="dv">100</span>, <span class="dv">0</span>, <span class="dv">1</span>) <span class="co">#rnorm(n_samples, true_mean, standard_deviation)</span></span>
<span id="cb211-4"><a href="statistical-tests.html#cb211-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb211-5"><a href="statistical-tests.html#cb211-5" aria-hidden="true" tabindex="-1"></a>ttestx0 <span class="ot">&lt;-</span> <span class="fu">t.test</span>(x0, <span class="at">mu =</span> <span class="dv">0</span>, <span class="at">alternative =</span> <span class="st">&quot;two.sided&quot;</span>, <span class="at">conf.level =</span> <span class="fl">0.95</span>) <span class="co">#actual running of the t-test</span></span>
<span id="cb211-6"><a href="statistical-tests.html#cb211-6" aria-hidden="true" tabindex="-1"></a><span class="co">#t.test(x0) yeilds the same results bc I used default values for mu, alternative, and conf.level</span></span>
<span id="cb211-7"><a href="statistical-tests.html#cb211-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb211-8"><a href="statistical-tests.html#cb211-8" aria-hidden="true" tabindex="-1"></a>tlessx0 <span class="ot">&lt;-</span> <span class="fu">t.test</span>(x0, <span class="at">mu =</span> <span class="dv">0</span>, <span class="at">alternative =</span> <span class="st">&quot;less&quot;</span>, <span class="at">conf.level =</span> <span class="fl">0.90</span>)</span>
<span id="cb211-9"><a href="statistical-tests.html#cb211-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb211-10"><a href="statistical-tests.html#cb211-10" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">knit_print</span>(ttestx0) <span class="co">#display t-test output via knitr</span></span></code></pre></div>
<pre><code>## 
##  One Sample t-test
## 
## data:  x0
## t = -1.4507, df = 99, p-value = 0.15
## alternative hypothesis: true mean is not equal to 0
## 95 percent confidence interval:
##  -0.32331056  0.05021268
## sample estimates:
##  mean of x 
## -0.1365489</code></pre>
<p>In <code>t.test_x0</code>, we are running a two-tailed t-test on a vector of 100 doubles
generated with <code>rnorm()</code> where the intended mean is 0. Since this is a
two-tailed test, the <em>alternate</em> hypothesis is <code>mean(x0) != 0</code> while the <em>null</em>
hypothesis is <code>mean(x0) = 0</code></p>
<p>Since mu=0 and 0 is within the range (-.323, 0.050), we have failed to reject
the null hypothesis with this set of data</p>
</div>
<div id="unpaired-t-test" class="section level4 unnumbered">
<h4><strong>Unpaired T-Test</strong></h4>
<p><em>Dependent Variables</em>: 1, continuous [ratio and interval]</p>
<p><em>Independent Variables</em>: 1</p>
<p><em>Design</em>: 2 groups, 2 levels</p>
<p><em>Parametric</em>: Yes</p>
<div class="sourceCode" id="cb213"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb213-1"><a href="statistical-tests.html#cb213-1" aria-hidden="true" tabindex="-1"></a><span class="co"># One-sample t-test</span></span>
<span id="cb213-2"><a href="statistical-tests.html#cb213-2" aria-hidden="true" tabindex="-1"></a><span class="fu">t.test</span>(x, <span class="co">#numeric vector of data 1</span></span>
<span id="cb213-3"><a href="statistical-tests.html#cb213-3" aria-hidden="true" tabindex="-1"></a>       y, <span class="co">#numeric vector of data 2</span></span>
<span id="cb213-4"><a href="statistical-tests.html#cb213-4" aria-hidden="true" tabindex="-1"></a>       <span class="at">alternative =</span> <span class="st">&quot;two.sided&quot;</span>, <span class="co"># (&quot;two.sided&quot;, &quot;less&quot;, &quot;greater&quot;) depending on which you want to use. Defaults to &quot;two.sided</span></span>
<span id="cb213-5"><a href="statistical-tests.html#cb213-5" aria-hidden="true" tabindex="-1"></a>       <span class="at">conf.level =</span> <span class="fl">0.95</span> <span class="co">#1-alpha. Defaults to 0.95</span></span>
<span id="cb213-6"><a href="statistical-tests.html#cb213-6" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div>
<p>The unpaired t-test functions similarly to the one-sample t-test except instead
of mu, it uses dataset y. Variance is assumed to be about equal between dataset
x and y.</p>
<div class="box readmore">
<p><a href="https://sphweb.bumc.bu.edu/otlt/MPH-Modules/PH717-QuantCore/PH717-Module7-T-tests/Module7-ttests6.html">Moer information about unpaired t-tests here</a>
<a href="https://sphweb.bumc.bu.edu/otlt/MPH-Modules/PH717-QuantCore/PH717-Module7-T-tests/Module7-ttests7.html">Another R guide for upaired t-tests</a></p>
</div>
<div class="sourceCode" id="cb214"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb214-1"><a href="statistical-tests.html#cb214-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Unpaired t-test example</span></span>
<span id="cb214-2"><a href="statistical-tests.html#cb214-2" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">10</span>)</span>
<span id="cb214-3"><a href="statistical-tests.html#cb214-3" aria-hidden="true" tabindex="-1"></a>x <span class="ot">&lt;-</span> <span class="fu">rnorm</span>(<span class="dv">100</span>, <span class="dv">0</span>, <span class="dv">1</span>)</span>
<span id="cb214-4"><a href="statistical-tests.html#cb214-4" aria-hidden="true" tabindex="-1"></a>y <span class="ot">&lt;-</span> <span class="fu">rnorm</span>(<span class="dv">100</span>, <span class="dv">3</span>, <span class="dv">1</span>)</span>
<span id="cb214-5"><a href="statistical-tests.html#cb214-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb214-6"><a href="statistical-tests.html#cb214-6" aria-hidden="true" tabindex="-1"></a>unpaired <span class="ot">&lt;-</span> <span class="fu">t.test</span>(x, y)</span>
<span id="cb214-7"><a href="statistical-tests.html#cb214-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb214-8"><a href="statistical-tests.html#cb214-8" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">knit_print</span>(unpaired)</span></code></pre></div>
<pre><code>## 
##  Welch Two Sample t-test
## 
## data:  x and y
## t = -22.51, df = 197.83, p-value &lt; 2.2e-16
## alternative hypothesis: true difference in means is not equal to 0
## 95 percent confidence interval:
##  -3.308051 -2.775122
## sample estimates:
##  mean of x  mean of y 
## -0.1365489  2.9050374</code></pre>
<p>Output for an unpaired two sample t-test is similar to the output for the
one-sample. Mu is typically 0 (though one could change it with ‘mu = n’ in the
function call) and the test uses the confidence interval for (mean(x)-mean(y)),
which is (-3.308, -2.775). Since mu does not exist within the confidence
interval, the null hypothesis can be rejected.</p>
</div>
<div id="paired-t-test" class="section level4 unnumbered">
<h4><strong>Paired T-Test</strong></h4>
<p><em>Dependent Variables</em>: 1, continuous [ratio and interval]</p>
<p><em>Independent Variables</em>: 1</p>
<p><em>Design</em>: 1 group, 2 levels</p>
<p><em>Parametric</em>: Yes</p>
<div class="sourceCode" id="cb216"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb216-1"><a href="statistical-tests.html#cb216-1" aria-hidden="true" tabindex="-1"></a><span class="co"># One-sample t-test</span></span>
<span id="cb216-2"><a href="statistical-tests.html#cb216-2" aria-hidden="true" tabindex="-1"></a><span class="fu">t.test</span>(x, <span class="co">#numeric vector of data 1</span></span>
<span id="cb216-3"><a href="statistical-tests.html#cb216-3" aria-hidden="true" tabindex="-1"></a>       y, <span class="co">#numeric vector of data 2</span></span>
<span id="cb216-4"><a href="statistical-tests.html#cb216-4" aria-hidden="true" tabindex="-1"></a>       <span class="at">paired =</span> <span class="cn">TRUE</span>, <span class="co">#defaults to FALSE</span></span>
<span id="cb216-5"><a href="statistical-tests.html#cb216-5" aria-hidden="true" tabindex="-1"></a>       <span class="at">alternative =</span> <span class="st">&quot;two.sided&quot;</span>, <span class="co"># (&quot;two.sided&quot;, &quot;less&quot;, &quot;greater&quot;) depending on which you want to use. Defaults to &quot;two.sided</span></span>
<span id="cb216-6"><a href="statistical-tests.html#cb216-6" aria-hidden="true" tabindex="-1"></a>       <span class="at">conf.level =</span> <span class="fl">0.95</span> <span class="co">#1-alpha. Defaults to 0.95</span></span>
<span id="cb216-7"><a href="statistical-tests.html#cb216-7" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div>
<p>The paired t-test works for matched and repeated-measures designs. Like the
unpaired t-test, it has a second data vector and does not need ‘mu.’ When
running a paired t-test make sure that you specify that <code>paired = TRUE</code> when
calling the function and that x and y have the same length.</p>
<p><a href="https://sphweb.bumc.bu.edu/otlt/MPH-Modules/PH717-QuantCore/PH717-Module7-T-tests/Module7-ttests8.html#headingtaglink_2">More information about paired t-tests
here</a></p>
<div class="sourceCode" id="cb217"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb217-1"><a href="statistical-tests.html#cb217-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Unpaired t-test example</span></span>
<span id="cb217-2"><a href="statistical-tests.html#cb217-2" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">10</span>)</span>
<span id="cb217-3"><a href="statistical-tests.html#cb217-3" aria-hidden="true" tabindex="-1"></a>x <span class="ot">&lt;-</span> <span class="fu">rnorm</span>(<span class="dv">100</span>, <span class="dv">0</span>, <span class="dv">13</span>)</span>
<span id="cb217-4"><a href="statistical-tests.html#cb217-4" aria-hidden="true" tabindex="-1"></a>y <span class="ot">&lt;-</span> <span class="fu">rnorm</span>(<span class="dv">100</span>, <span class="dv">3</span>, <span class="dv">1</span>)</span>
<span id="cb217-5"><a href="statistical-tests.html#cb217-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb217-6"><a href="statistical-tests.html#cb217-6" aria-hidden="true" tabindex="-1"></a>unpaired <span class="ot">&lt;-</span> <span class="fu">t.test</span>(x, y, <span class="at">paired =</span> <span class="cn">TRUE</span>)</span>
<span id="cb217-7"><a href="statistical-tests.html#cb217-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb217-8"><a href="statistical-tests.html#cb217-8" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">knit_print</span>(unpaired)</span></code></pre></div>
<pre><code>## 
##  Paired t-test
## 
## data:  x and y
## t = -3.7956, df = 99, p-value = 0.0002539
## alternative hypothesis: true difference in means is not equal to 0
## 95 percent confidence interval:
##  -7.126787 -2.233560
## sample estimates:
## mean of the differences 
##               -4.680174</code></pre>
<p>Similar to the unpaired t-test, the paired t-test looks at the differences
between x and y. Unlike the unpaired t-test, the statistic creates and runs its
test on a third set of data created from the differences of x and y. For an x
and y of length n, r would create a third data set of <code>new = ((x1-y1), (x2-y2),...(xn-1-yn-1), (xn-yn))</code> and run its testing with mean(new) and sd(new)</p>
</div>
<div id="chi-squared-test" class="section level4 unnumbered">
<h4><strong>Chi-Squared test</strong></h4>
<p><em>Dependent Variables</em>: 1, categorical</p>
<p><em>Independent Variables</em>: 1</p>
<p><em>Design</em>: 2 groups, 2 + levels</p>
<p><em>Parametric</em>: No</p>
<p><a href="https://sphweb.bumc.bu.edu/otlt/MPH-Modules/BS/BS704_HypothesisTesting-ChiSquare/BS704_HypothesisTesting-ChiSquare3.html">More Information on Chi-Squared tests</a></p>
<div class="sourceCode" id="cb219"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb219-1"><a href="statistical-tests.html#cb219-1" aria-hidden="true" tabindex="-1"></a><span class="co">#How to Run a chi-squared test</span></span>
<span id="cb219-2"><a href="statistical-tests.html#cb219-2" aria-hidden="true" tabindex="-1"></a><span class="fu">chisq.test</span>(x, <span class="co"># numeric vector, matrix, or factor</span></span>
<span id="cb219-3"><a href="statistical-tests.html#cb219-3" aria-hidden="true" tabindex="-1"></a>           y <span class="co"># numeric vector; ignored if x is a matrix; factor of the same length as x if x is a factor</span></span>
<span id="cb219-4"><a href="statistical-tests.html#cb219-4" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div>
</div>
<div id="chi-square-goodness-of-fit_-_-" class="section level4" number="6.3.2.1">
<h4><span class="header-section-number">6.3.2.1</span> __Chi-Square Goodness-Of-Fit_ _{-}</h4>
<p><em>Dependent Variables</em>: 1, categorical</p>
<p><em>Independent Variables</em>: 0</p>
<p><em>Design</em>: 1 group</p>
<p><em>Parametric</em>: No</p>
<p><a href="https://sphweb.bumc.bu.edu/otlt/MPH-Modules/BS/BS704_HypothesisTesting-ChiSquare/BS704_HypothesisTesting-ChiSquare2.html#headingtaglink_1">More Information on Chi-Squared test on one sample</a></p>
<div class="sourceCode" id="cb220"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb220-1"><a href="statistical-tests.html#cb220-1" aria-hidden="true" tabindex="-1"></a><span class="co">#How to Run a chi-square goodness of fit</span></span>
<span id="cb220-2"><a href="statistical-tests.html#cb220-2" aria-hidden="true" tabindex="-1"></a><span class="fu">chisq.test</span>(x, <span class="co"># numeric vector</span></span>
<span id="cb220-3"><a href="statistical-tests.html#cb220-3" aria-hidden="true" tabindex="-1"></a>           <span class="at">p =</span> <span class="co">#list of probabilities, like c(10, 10, 10, 70)/100), same length as x</span></span>
<span id="cb220-4"><a href="statistical-tests.html#cb220-4" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div>
<p>The chi-squared goodness-of-fit can compare one set of outcomes to a given set of probabilities to determine the likelihood of the test probabilities differing from the given. The set of given probabilities constitutes the null hypothesis. In the below example, the dataset is the tested outcome frequencies: the number of times x[1], x[2], x[3], and x[4] were observed. If my null hypothesis says that each of the 4 should be observed an equal number of times, then the resulting chi-square test should will have a p-value of p&lt;0.001</p>
<div class="sourceCode" id="cb221"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb221-1"><a href="statistical-tests.html#cb221-1" aria-hidden="true" tabindex="-1"></a>x <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">10</span>,<span class="dv">25</span>, <span class="dv">18</span>, <span class="dv">92</span>)</span>
<span id="cb221-2"><a href="statistical-tests.html#cb221-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb221-3"><a href="statistical-tests.html#cb221-3" aria-hidden="true" tabindex="-1"></a><span class="fu">chisq.test</span>(x, <span class="at">p=</span><span class="fu">c</span>(<span class="dv">25</span>, <span class="dv">25</span>, <span class="dv">25</span>, <span class="dv">25</span>)<span class="sc">/</span><span class="dv">100</span>)</span></code></pre></div>
<pre><code>## 
##  Chi-squared test for given probabilities
## 
## data:  x
## X-squared = 117.43, df = 3, p-value &lt; 2.2e-16</code></pre>
</div>
<div id="wilcoxon-mann-whitney-test" class="section level4 unnumbered">
<h4><strong>Wilcoxon-Mann Whitney test</strong></h4>
<p><em>Dependent Variables</em>: 1, (ordinal, interval, or ratio)</p>
<p><em>Independent Variables</em>: 1</p>
<p><em>Design</em>: 2 groups, 2 levels</p>
<p><em>Parametric</em>: No</p>
<p><a href="https://sphweb.bumc.bu.edu/otlt/mph-modules/bs/bs704_nonparametric/bs704_nonparametric4.html">More information here</a></p>
<div class="sourceCode" id="cb223"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb223-1"><a href="statistical-tests.html#cb223-1" aria-hidden="true" tabindex="-1"></a><span class="fu">wilcox.test</span>(x,<span class="co">#Numeric vector of data</span></span>
<span id="cb223-2"><a href="statistical-tests.html#cb223-2" aria-hidden="true" tabindex="-1"></a>            y <span class="co">#optional numeric vector of data)</span></span></code></pre></div>
</div>
<div id="wilcoxon-signed-ranks-test" class="section level4 unnumbered">
<h4><strong>Wilcoxon Signed Ranks test</strong></h4>
<p><em>Dependent Variables</em>: 1, (ordinal, interval, or ratio)</p>
<p><em>Independent Variables</em>: 1</p>
<p><em>Design</em>: 1 groups, 2 levels</p>
<p><em>Parametric</em>: No</p>
<p><a href="https://sphweb.bumc.bu.edu/otlt/mph-modules/bs/bs704_nonparametric/BS704_Nonparametric6.html">More information here</a></p>
<pre class="{r`eval=false}"><code>
wilcox.test(x,#Numeric vector of data
            y, #numeric vector of data from 2nd level- same length
            paired = TRUE)
</code></pre>
</div>
<div id="one-way-anova" class="section level4 unnumbered">
<h4><strong>One-Way ANOVA</strong></h4>
<p><em>Dependent Variables</em>: 1, continuous</p>
<p><em>Independent Variables</em>: 1</p>
<p><em>Design</em>: 2+ groups, 2+ levels</p>
<p><em>Parametric</em>: Yes</p>
<p><a href="https://sphweb.bumc.bu.edu/otlt/MPH-Modules/BS/BS704_HypothesisTesting-ANOVA/BS704_HypothesisTesting-ANOVA4.html#headingtaglink_3">More information here</a></p>
<div class="sourceCode" id="cb225"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb225-1"><a href="statistical-tests.html#cb225-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(<span class="fu">aov</span>(formula, <span class="co">#formula specifying the model</span></span>
<span id="cb225-2"><a href="statistical-tests.html#cb225-2" aria-hidden="true" tabindex="-1"></a>            data <span class="co">#data frame containing the variables in the formula))</span></span></code></pre></div>
</div>
<div id="regressions" class="section level4 unnumbered">
<h4><strong>Regressions</strong></h4>
</div>
<div id="factor-analysis" class="section level4 unnumbered">
<h4><strong>Factor Analysis</strong></h4>
</div>
</div>
<div id="choosing-a-test" class="section level3" number="6.3.3">
<h3><span class="header-section-number">6.3.3</span> Choosing a Test</h3>
<ol style="list-style-type: decimal">
<li>Look for your <em>Dependent Variable(s).</em> How many do you have? (1 or 2+)
<ul>
<li>Which ones exist on a continuum, where any fraction within the valid range is technically possible, though the actual data might have rounded? (Continuous vs Categorical)</li>
<li>For Categorical: If values do not have a definite order then it’s Nominal. Else it’s probably Ordinal</li>
</ul></li>
<li>Look for your <em>Independent Variable(s).</em> How many do you have? (0, 1, or 2+)
<ul>
<li>Define their types and level of measurement as described in 1a and then do 2b and 2c if you have 1 or more independent variables</li>
<li>How many conditions does each independent variable have? (= # of levels)</li>
<li>Does the same subject or (equivalent subjects) undergo each condition/level? (‘Yes’ -&gt; repeated-measures; ‘No’ -&gt; between-groups</li>
</ul></li>
<li>If you do not have nominal data, can you assume your data would have a normal distribution if you measured the general population? (‘Yes’-&gt; parametric; ‘No’ -&gt; non-parametric)</li>
</ol>
</div>
<div id="p-values" class="section level3" number="6.3.4">
<h3><span class="header-section-number">6.3.4</span> P-values</h3>
<p>In statistics, the p-value is the probability of obtaining the observed results assuming that the null hypothesis is correct. In other words, the p-value is the probability of the same results happening by chance selection from the normal population rather than the test condition.</p>
<p>For instance, imagine there is a game that randomly picks one out of four colors. The game developer decides to add a 5th color, yellow, so that each color now has 1/5 chance of winning and a 4/5 chance of losing. How many of its first games would yellow have to lose in a row before you suspect the dev forgot to code in the ability for yellow to win? If you want to wait to until there is a &lt; 5% chance of yellow losing every single game in a row under normal circumstances before pointing this out to the dev, then yellow would have to lose 14/14 games (a 4.4% likelihood).</p>
<p>To visualize how 14 losses in a row in a game of chance with 1/5 success rate would occur in less than 5% of all instances, we can run a simulation of 1000 instances of 14 games with 1/5 chance success and displaying the results with their frequencies of success:</p>
<div class="sourceCode" id="cb226"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb226-1"><a href="statistical-tests.html#cb226-1" aria-hidden="true" tabindex="-1"></a>roll <span class="ot">&lt;-</span> <span class="cf">function</span>(nrolls, noptions){</span>
<span id="cb226-2"><a href="statistical-tests.html#cb226-2" aria-hidden="true" tabindex="-1"></a> outcomes <span class="ot">&lt;-</span> <span class="fu">sample</span>(<span class="dv">1</span><span class="sc">:</span>noptions, nrolls, <span class="cn">TRUE</span>)</span>
<span id="cb226-3"><a href="statistical-tests.html#cb226-3" aria-hidden="true" tabindex="-1"></a> <span class="fu">return</span>(<span class="fu">length</span>(outcomes[outcomes <span class="sc">==</span> <span class="dv">1</span>]))</span>
<span id="cb226-4"><a href="statistical-tests.html#cb226-4" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb226-5"><a href="statistical-tests.html#cb226-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb226-6"><a href="statistical-tests.html#cb226-6" aria-hidden="true" tabindex="-1"></a>dataset <span class="ot">&lt;-</span> <span class="cf">function</span>(sample, nrolls, noptions){</span>
<span id="cb226-7"><a href="statistical-tests.html#cb226-7" aria-hidden="true" tabindex="-1"></a>  simulations <span class="ot">&lt;-</span> <span class="fu">c</span>()</span>
<span id="cb226-8"><a href="statistical-tests.html#cb226-8" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span>sample){simulations[i] <span class="ot">&lt;-</span> <span class="fu">roll</span>(nrolls, noptions)}</span>
<span id="cb226-9"><a href="statistical-tests.html#cb226-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(simulations)</span>
<span id="cb226-10"><a href="statistical-tests.html#cb226-10" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb226-11"><a href="statistical-tests.html#cb226-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb226-12"><a href="statistical-tests.html#cb226-12" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">18</span>)</span>
<span id="cb226-13"><a href="statistical-tests.html#cb226-13" aria-hidden="true" tabindex="-1"></a>Yellow_Wins <span class="ot">&lt;-</span> <span class="fu">dataset</span>(<span class="dv">1000</span>, <span class="dv">14</span>, <span class="dv">5</span>)</span>
<span id="cb226-14"><a href="statistical-tests.html#cb226-14" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(Yellow_Wins, <span class="at">freq=</span><span class="cn">FALSE</span>, <span class="at">right=</span><span class="cn">FALSE</span>)</span></code></pre></div>
<p><img src="main_files/unnamed-chunk-75-1.png" width="672" />
Here you see that slightly less than 5% of all run instances had 0 wins.</p>
<p>In this situation, we are comparing if yellow’s win rate is significantly lower than if yellow had an equal chance at winning. The research hypothesis or alternative hypothesis in this case is that yellow’s chance of winning is significantly lower than the rest of the colors while the null hypothesis would be that yellow’s chance of winning is the same as the rest of the colors.
The p-value of this test would be 0.043 because 43 out of 1000 instances of 14 rolls with a 1/5 chance of success totaled 0 successes, the number of successes we saw. If we instead were looking for 1 success or less out of 14 rolls, the p-value would be 0.19, because our simulation ran 147 instances of 1 success and 42 instances of 0 success.</p>
<p>Since p-values are calculated by the percentage of outcomes that are higher than the one observed, on a normal distribution they may look something like this:</p>
<div class="sourceCode" id="cb227"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb227-1"><a href="statistical-tests.html#cb227-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">2</span>)</span>
<span id="cb227-2"><a href="statistical-tests.html#cb227-2" aria-hidden="true" tabindex="-1"></a>norm <span class="ot">&lt;-</span> <span class="fu">rnorm</span>(<span class="at">n =</span> <span class="dv">50000</span>, <span class="at">mean =</span> <span class="dv">4</span>, <span class="at">sd =</span> <span class="dv">1</span>)</span>
<span id="cb227-3"><a href="statistical-tests.html#cb227-3" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(<span class="fu">density</span>(norm),</span>
<span id="cb227-4"><a href="statistical-tests.html#cb227-4" aria-hidden="true" tabindex="-1"></a>     <span class="at">main =</span> <span class="st">&quot;P-Values on a Normal Distribution&quot;</span>, <span class="at">xlab =</span> <span class="st">&quot;X-Values&quot;</span>,</span>
<span id="cb227-5"><a href="statistical-tests.html#cb227-5" aria-hidden="true" tabindex="-1"></a>     <span class="at">lwd =</span> <span class="dv">2</span>, <span class="at">col =</span> <span class="st">&quot;royalblue&quot;</span></span>
<span id="cb227-6"><a href="statistical-tests.html#cb227-6" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb227-7"><a href="statistical-tests.html#cb227-7" aria-hidden="true" tabindex="-1"></a><span class="fu">polygon</span>(</span>
<span id="cb227-8"><a href="statistical-tests.html#cb227-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">x =</span> <span class="fu">c</span>(<span class="fu">density</span>(norm)<span class="sc">$</span>x[<span class="fu">density</span>(norm)<span class="sc">$</span>x <span class="sc">&gt;=</span> <span class="dv">6</span>], <span class="dv">6</span>),</span>
<span id="cb227-9"><a href="statistical-tests.html#cb227-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">y =</span> <span class="fu">c</span>(<span class="fu">density</span>(norm)<span class="sc">$</span>y[<span class="fu">density</span>(norm)<span class="sc">$</span>x <span class="sc">&gt;=</span> <span class="dv">6</span>], <span class="dv">0</span>),</span>
<span id="cb227-10"><a href="statistical-tests.html#cb227-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">border =</span> <span class="st">&quot;royalblue&quot;</span>, <span class="at">col =</span> <span class="st">&quot;lightblue1&quot;</span></span>
<span id="cb227-11"><a href="statistical-tests.html#cb227-11" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb227-12"><a href="statistical-tests.html#cb227-12" aria-hidden="true" tabindex="-1"></a><span class="fu">segments</span>(<span class="at">x0 =</span> <span class="dv">6</span>, <span class="at">y0 =</span> <span class="dv">0</span>, <span class="at">y1 =</span> <span class="fl">0.4</span>, <span class="at">lwd =</span> <span class="dv">2</span>, <span class="at">lty =</span> <span class="dv">2</span>, <span class="at">col =</span> <span class="st">&quot;salmon&quot;</span>)</span>
<span id="cb227-13"><a href="statistical-tests.html#cb227-13" aria-hidden="true" tabindex="-1"></a><span class="fu">text</span>(<span class="dv">7</span>, <span class="fl">0.2</span>, <span class="st">&quot;Observed Value (@ 2sd)&quot;</span>)</span>
<span id="cb227-14"><a href="statistical-tests.html#cb227-14" aria-hidden="true" tabindex="-1"></a><span class="fu">text</span>(<span class="fl">3.9</span>, <span class="fl">0.01</span>, <span class="st">&quot;Instances higher than crit. val -&gt;&quot;</span>)</span></code></pre></div>
<p><img src="main_files/unnamed-chunk-76-1.png" width="672" /></p>
<p>Here the observed value is 6, which happens to be exactly 2 standard deviations above the mean. To compute the p-value of this instance, the area under the curve would need to be calculated which can be easily done by looking up the appropriate values on a z-table. <a href="https://sphweb.bumc.bu.edu/otlt/mph-modules/bs/sas/sas4-onesamplettest/sas4-onesamplettest3.html">Here is a link about finding the area under the normal curve</a></p>
<p><a href="https://sphweb.bumc.bu.edu/otlt/MPH-Modules/PH717-QuantCore/PH717-Module7-T-tests/PH717-Module7-T-tests3.html">SPH’s information about p-values here</a></p>
</div>
<div id="multiple-hypothesis-testing-." class="section level3" number="6.3.5">
<h3><span class="header-section-number">6.3.5</span> Multiple Hypothesis Testing .</h3>
<p>What is multiply hypothesis testing adjustment? Why is it important? What are
some common adjustment methods (Bonferroni (FWER) and Benjamini-Hochberg (FDR))?
How do we interpret adjusted p-values (depends on the adjustment method)? How
do we compute adjusted p-values in R?</p>
</div>
<div id="statistical-power" class="section level3" number="6.3.6">
<h3><span class="header-section-number">6.3.6</span> Statistical power</h3>
<p><strong>Statistical Power</strong> is the probability that the null hypothesis was actually false but the data did not have enough evidence to reject it. In short, it is the failure to reject the null hypothesis in the presence of a significant effect. This is a Type II error, also known as a false negative. Its counterpart, the Type I error, is the probability that a null hypothesis was rejected when there was no significant effect.</p>
<p>The chance of a Type I error is represented by alpha, the same alpha that we use to calculate significance level (1-alpha). Type II errors are represented by beta. The relationship between statistical power and beta are similar to the relationship between significance level and alpha: statistical power is calculated as 1-beta.</p>
<p><a href="https://sphweb.bumc.bu.edu/otlt/mph-modules/bs/bs704_hypothesistest-means-proportions/BS704_HypothesisTest-Means-Proportions3.html#headingtaglink_3">Read more about the relationship between Type II and Type I errors</a></p>
<p>To calculate beta, find the area under the curve of the research hypothesis distribution on the opposite side of the critical value line where alpha is calculated. To find the area in the tail of the distribution, you would need to reference a Z-score chart, which will not be expected of you in this course. If you wish to read more about it <a href="https://sphweb.bumc.bu.edu/otlt/mph-modules/bs/sas/sas4-onesamplettest/sas4-onesamplettest3.html">here is a link about finding the area under the normal curve with z-scores, same one as linked above</a></p>
<p><img src="main_files/unnamed-chunk-77-1.png" width="672" />
<small> Code sourced from:
caracal, “How do I find the probability of a type II error,” stats.stackexchange.com, Feb 19, 2011, <a href="https://stats.stackexchange.com/questions/7402/how-do-i-find-the-probability-of-a-type-ii-error">link</a>, Date Accessed: Feb 28, 2022 </small></p>
<p>The above graph shows the relationship of alpha and beta. Since the alternative hypothesis is looking for a higher mean, alpha is calculated by the area under the null hypothesis curve on right side of the critical value while beta is calculated by the area under the alternative hypothesis on the left of the critical value line. Should an experiment get a higher p-value than the critical value, then the Beta will increase in relationship to Alpha decreasing.</p>
<p><img src="main_files/unnamed-chunk-78-1.png" width="672" />
In this graph, the critical value was moved to 0.025 while the null and alternative hypotheses stayed the same. As the chance of alpha decreased, the chances of beta increased. The opposite is also true, as alpha increases, beta decreases</p>
<div id="applications-of-statistical-power" class="section level5 unnumbered">
<h5>Applications of Statistical Power</h5>
<p>One of the applications of statistical power, besides as a measure of Type II error probability, is to allow you to run a power analysis. A power analysis involves using the relationship between <strong>Effect Size</strong>, <strong>Sample Size</strong>, <strong>Significance</strong>, and <strong>Statistical Power</strong> when you have three of the four parts in order to find the value of the fourth. This is typically used to find the effect size of a study, as that is often the hardest to estimate, or to find a sample size that corresponds with a desired effect size when designing a study.
<a href="https://sphweb.bumc.bu.edu/otlt/MPH-Modules/BS/BS704_Power/BS704_Power6.html">Further reading on statistical power and power analysis</a></p>
</div>
<div id="effect-size" class="section level5 unnumbered">
<h5>Effect Size</h5>
<p>The effect size is a measurement of the magnitude of experimental effects. There are a number of ways to calculate effect size including but not limited to Cohen’s D, Pearson’s R, and Odds Ratio. If you would like to read about what BU’s SPH has to say about it, <a href="https://sphweb.bumc.bu.edu/otlt/mph-modules/bs/bs704_power/BS704_Power8.html">here is the SPH module link</a></p>
<p><small>
Citations:
Dorey FJ. Statistics in brief: Statistical power: what is it and when should it be used?. Clin Orthop Relat Res. 2011;469(2):619-620. <a href="doi:10.1007/s11999-010-1435-0" class="uri">doi:10.1007/s11999-010-1435-0</a></p>
<p>Parab S, Bhalerao S. Choosing statistical test. Int J Ayurveda Res. 2010;1(3):187-191. <a href="doi:10.4103/0974-7788.72494" class="uri">doi:10.4103/0974-7788.72494</a>
Introduction to SAS. UCLA: Statistical Consulting Group. from <a href="https://stats.idre.ucla.edu/sas/modules/sas-learning-moduleintroduction-to-the-features-of-sas/" class="uri">https://stats.idre.ucla.edu/sas/modules/sas-learning-moduleintroduction-to-the-features-of-sas/</a> (accessed February 24, 2022)</p>
<p>Ranganathan P, Gogtay NJ. An Introduction to Statistics - Data Types, Distributions and Summarizing Data. Indian J Crit Care Med. 2019 Jun;23(Suppl 2):S169-S170. doi: 10.5005/jp-journals-10071-23198. PMID: 31485129; PMCID: PMC6707495.</p>
<p></small></p>

</div>
</div>
</div>
<p style="text-align: center;">
<a href="statistical-distributions.html"><button class="btn btn-default">Previous</button></a>
<a href="exploratory-data-analysis.html"><button class="btn btn-default">Next</button></a>
</p>
</div>
</div>


</div>

<script>

// add bootstrap table styles to pandoc tables
$(document).ready(function () {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
});

</script>

</body>
</html>
