<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta name="viewport" content="width=device-width, initial-scale=1" />
<meta property="og:title" content="6.2 Statistical Distributions | BF591 - R for Biological Sciences" />
<meta property="og:type" content="book" />







<script type="text/x-mathjax-config">
MathJax.Hub.Config({
  TeX: { equationNumbers: { autoNumber: "AMS" } }
});
</script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<meta name="description" content="6.2 Statistical Distributions | BF591 - R for Biological Sciences">

<title>6.2 Statistical Distributions | BF591 - R for Biological Sciences</title>

<script src="libs/header-attrs-2.13/header-attrs.js"></script>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="libs/bootstrap-3.3.5/css/darkly.min.css" rel="stylesheet" />
<script src="libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="libs/navigation-1.1/tabsets.js"></script>


<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>


<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />

<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
/* show arrow before summary tag as in bootstrap
TODO: remove if boostrap in updated in html_document (rmarkdown#1485) */
details > summary {
  display: list-item;
  cursor: pointer;
}
</style>
</head>

<body>

<div class="container-fluid main-container">


<div class="row">
<div class="col-sm-12">
<div id="TOC">
<ul>
<li><a href="index.html#syllabus">Syllabus</a>
<ul>
<li><a href="course-schedule.html#course-schedule">Course Schedule</a></li>
<li><a href="instructors.html#instructors">Instructors</a></li>
<li><a href="office-hours.html#office-hours">Office Hours</a></li>
<li><a href="course-values-and-policies.html#course-values-and-policies">Course Values and Policies</a></li>
</ul></li>
<li><a href="introduction.html#introduction"><span class="toc-section-number">1</span> Introduction</a>
<ul>
<li><a href="who-this-book-is-for.html#who-this-book-is-for"><span class="toc-section-number">1.1</span> Who This Book Is For</a></li>
<li><a href="a-note-about-reinventing-the-wheel.html#a-note-about-reinventing-the-wheel"><span class="toc-section-number">1.2</span> A Note About Reinventing the Wheel</a></li>
<li><a href="sources-and-references.html#sources-and-references"><span class="toc-section-number">1.3</span> Sources and References</a></li>
</ul></li>
<li><a href="data-bio.html#data-bio"><span class="toc-section-number">2</span> Data in Biology</a>
<ul>
<li><a href="a-brief-history-of-data-in-molecular-biology.html#a-brief-history-of-data-in-molecular-biology"><span class="toc-section-number">2.1</span> A Brief History of Data in Molecular Biology</a></li>
<li><a href="biology-as-a-mature-data-science.html#biology-as-a-mature-data-science"><span class="toc-section-number">2.2</span> Biology as a Mature Data Science</a></li>
</ul></li>
<li><a href="preliminaries.html#preliminaries"><span class="toc-section-number">3</span> Preliminaries</a>
<ul>
<li><a href="prelim-r.html#prelim-r"><span class="toc-section-number">3.1</span> The R Language</a></li>
<li><a href="prelim-rstudio.html#prelim-rstudio"><span class="toc-section-number">3.2</span> RStudio</a></li>
<li><a href="prog-the-r-script.html#prog-the-r-script"><span class="toc-section-number">3.3</span> The R Script</a></li>
<li><a href="prog-workflow.html#prog-workflow"><span class="toc-section-number">3.4</span> The Scripting Workflow</a></li>
<li><a href="prelim-git.html#prelim-git"><span class="toc-section-number">3.5</span> git + github</a>
<ul>
<li><a href="prelim-git.html#motivation"><span class="toc-section-number">3.5.1</span> Motivation</a></li>
<li><a href="prelim-git.html#git"><span class="toc-section-number">3.5.2</span> git</a></li>
<li><a href="prelim-git.html#git-hosting-platforms-github"><span class="toc-section-number">3.5.3</span> Git hosting platforms (GitHub)</a></li>
</ul></li>
</ul></li>
<li><a href="prog-basics.html#prog-basics"><span class="toc-section-number">4</span> R Programming</a>
<ul>
<li><a href="before-you-begin.html#before-you-begin"><span class="toc-section-number">4.1</span> Before you begin</a></li>
<li><a href="introduction-1.html#introduction-1"><span class="toc-section-number">4.2</span> Introduction</a></li>
<li><a href="prog-r-syntax.html#prog-r-syntax"><span class="toc-section-number">4.3</span> R Syntax Basics</a></li>
<li><a href="prog-types.html#prog-types"><span class="toc-section-number">4.4</span> Basic Types of Values</a>
<ul>
<li><a href="prog-types.html#factors"><span class="toc-section-number">4.4.1</span> Factors</a></li>
</ul></li>
<li><a href="prog-struct.html#prog-struct"><span class="toc-section-number">4.5</span> Data Structures</a>
<ul>
<li><a href="prog-struct.html#vectors"><span class="toc-section-number">4.5.1</span> Vectors</a></li>
<li><a href="prog-struct.html#matrices"><span class="toc-section-number">4.5.2</span> Matrices</a></li>
<li><a href="prog-struct.html#lists-and-data-frames"><span class="toc-section-number">4.5.3</span> Lists and data frames</a></li>
</ul></li>
<li><a href="logical-tests-and-comparators.html#logical-tests-and-comparators"><span class="toc-section-number">4.6</span> Logical Tests and Comparators</a></li>
<li><a href="functions.html#functions"><span class="toc-section-number">4.7</span> Functions</a>
<ul>
<li><a href="functions.html#prog-dry"><span class="toc-section-number">4.7.1</span> DRY: Don’t Repeat Yourself</a></li>
<li><a href="functions.html#writing-your-own-functions"><span class="toc-section-number">4.7.2</span> Writing your own functions</a></li>
<li><a href="functions.html#scope"><span class="toc-section-number">4.7.3</span> Scope</a></li>
</ul></li>
<li><a href="iteration.html#iteration"><span class="toc-section-number">4.8</span> Iteration</a>
<ul>
<li><a href="iteration.html#vectorized-operations"><span class="toc-section-number">4.8.1</span> Vectorized operations</a></li>
<li><a href="iteration.html#functional-programming"><span class="toc-section-number">4.8.2</span> Functional programming</a></li>
<li><a href="iteration.html#apply-and-friends"><span class="toc-section-number">4.8.3</span> <code>apply()</code> and friends</a></li>
</ul></li>
<li><a href="installing-packages.html#installing-packages"><span class="toc-section-number">4.9</span> Installing Packages</a></li>
<li><a href="saving-and-loading-r-data.html#saving-and-loading-r-data"><span class="toc-section-number">4.10</span> Saving and Loading R Data</a></li>
<li><a href="prog-debugging.html#prog-debugging"><span class="toc-section-number">4.11</span> Troubleshooting and Debugging</a></li>
<li><a href="prog-style.html#prog-style"><span class="toc-section-number">4.12</span> Coding Style and Conventions</a>
<ul>
<li><a href="prog-style.html#is-my-code-correct"><span class="toc-section-number">4.12.1</span> Is my code correct?</a></li>
<li><a href="prog-style.html#does-my-code-follow-the-dry-principle"><span class="toc-section-number">4.12.2</span> Does my code follow the <span>DRY principle</span>?</a></li>
<li><a href="prog-style.html#did-i-choose-concise-but-descriptive-variable-and-function-names"><span class="toc-section-number">4.12.3</span> Did I choose concise but descriptive variable and function names?</a></li>
<li><a href="prog-style.html#did-i-use-indentation-and-naming-conventions-consistently-throughout-my-code"><span class="toc-section-number">4.12.4</span> Did I use indentation and naming conventions consistently throughout my code?</a></li>
<li><a href="prog-style.html#did-i-write-comments-especially-when-what-the-code-does-is-not-obvious"><span class="toc-section-number">4.12.5</span> Did I write comments, especially when what the code does is not obvious?</a></li>
<li><a href="prog-style.html#how-easy-would-it-be-for-someone-else-to-understand-my-code"><span class="toc-section-number">4.12.6</span> How easy would it be for someone else to understand my code?</a></li>
<li><a href="prog-style.html#is-my-code-easy-to-maintainchange"><span class="toc-section-number">4.12.7</span> Is my code easy to maintain/change?</a></li>
<li><a href="prog-style.html#the-styler-package"><span class="toc-section-number">4.12.8</span> The <code>styler</code> package</a></li>
</ul></li>
</ul></li>
<li><a href="data-wrangling.html#data-wrangling"><span class="toc-section-number">5</span> Data Wrangling</a>
<ul>
<li><a href="the-tidyverse.html#the-tidyverse"><span class="toc-section-number">5.1</span> The Tidyverse</a></li>
<li><a href="dw-basics.html#dw-basics"><span class="toc-section-number">5.2</span> Tidyverse Basics</a></li>
<li><a href="importing-data.html#importing-data"><span class="toc-section-number">5.3</span> Importing Data</a></li>
<li><a href="data-tibble.html#data-tibble"><span class="toc-section-number">5.4</span> The tibble</a></li>
<li><a href="tidy-data.html#tidy-data"><span class="toc-section-number">5.5</span> Tidy Data</a></li>
<li><a href="pipes.html#pipes"><span class="toc-section-number">5.6</span> pipes</a></li>
<li><a href="arranging-data.html#arranging-data"><span class="toc-section-number">5.7</span> Arranging Data</a>
<ul>
<li><a href="arranging-data.html#dplyrmutate---create-new-columns-using-other-columns"><span class="toc-section-number">5.7.1</span> <code>dplyr::mutate()</code> - Create new columns using other columns</a></li>
<li><a href="arranging-data.html#stringr---working-with-character-values"><span class="toc-section-number">5.7.2</span> <code>stringr</code> - Working with character values</a></li>
<li><a href="arranging-data.html#dplyrselect---subset-columns-by-name"><span class="toc-section-number">5.7.3</span> <code>dplyr::select()</code> - Subset Columns by Name</a></li>
<li><a href="arranging-data.html#dplyrfilter---pick-rows-out-of-a-data-set"><span class="toc-section-number">5.7.4</span> <code>dplyr::filter()</code> - Pick rows out of a data set</a></li>
<li><a href="arranging-data.html#dplyrarrange---order-rows-based-on-their-values"><span class="toc-section-number">5.7.5</span> <code>dplyr::arrange()</code> - Order rows based on their values</a></li>
<li><a href="arranging-data.html#putting-it-all-together"><span class="toc-section-number">5.7.6</span> Putting it all together</a></li>
</ul></li>
<li><a href="grouping-data.html#grouping-data"><span class="toc-section-number">5.8</span> Grouping Data</a></li>
<li><a href="rearranging-data.html#rearranging-data"><span class="toc-section-number">5.9</span> Rearranging Data</a></li>
<li><a href="relational-data.html#relational-data"><span class="toc-section-number">5.10</span> Relational Data</a>
<ul>
<li><a href="relational-data.html#dealing-with-multiple-matches"><span class="toc-section-number">5.10.1</span> Dealing with multiple matches</a></li>
</ul></li>
</ul></li>
<li><a href="data-science.html#data-science"><span class="toc-section-number">6</span> Data Science</a>
<ul>
<li><a href="data-modeling.html#data-modeling"><span class="toc-section-number">6.1</span> Data Modeling</a>
<ul>
<li><a href="data-modeling.html#a-worked-modeling-example"><span class="toc-section-number">6.1.1</span> A Worked Modeling Example</a></li>
<li><a href="data-modeling.html#data-summarization"><span class="toc-section-number">6.1.2</span> Data Summarization</a></li>
<li><a href="data-modeling.html#linear-models"><span class="toc-section-number">6.1.3</span> Linear Models</a></li>
<li><a href="data-modeling.html#flavors-of-linear-models"><span class="toc-section-number">6.1.4</span> Flavors of Linear Models</a></li>
<li><a href="data-modeling.html#assessing-model-accuracy-."><span class="toc-section-number">6.1.5</span> Assessing Model Accuracy .</a></li>
</ul></li>
<li><a href="statistical-distributions.html#statistical-distributions"><span class="toc-section-number">6.2</span> Statistical Distributions</a>
<ul>
<li><a href="statistical-distributions.html#random-variables"><span class="toc-section-number">6.2.1</span> Random Variables</a></li>
<li><a href="statistical-distributions.html#statistical-distribution-basics"><span class="toc-section-number">6.2.2</span> Statistical Distribution Basics</a></li>
<li><a href="statistical-distributions.html#distributions-in-r"><span class="toc-section-number">6.2.3</span> Distributions in R</a></li>
<li><a href="statistical-distributions.html#discrete-distributions"><span class="toc-section-number">6.2.4</span> Discrete Distributions</a></li>
<li><a href="statistical-distributions.html#continuous-distributions"><span class="toc-section-number">6.2.5</span> Continuous Distributions</a></li>
<li><a href="statistical-distributions.html#empirical-distributions"><span class="toc-section-number">6.2.6</span> Empirical Distributions</a></li>
</ul></li>
<li><a href="statistical-tests.html#statistical-tests"><span class="toc-section-number">6.3</span> Statistical Tests</a>
<ul>
<li><a href="statistical-tests.html#what-is-a-statistical-test"><span class="toc-section-number">6.3.1</span> What is a statistical test</a></li>
<li><a href="statistical-tests.html#common-statistical-tests"><span class="toc-section-number">6.3.2</span> Common Statistical Tests</a></li>
<li><a href="statistical-tests.html#choosing-a-test"><span class="toc-section-number">6.3.3</span> Choosing a Test</a></li>
<li><a href="statistical-tests.html#p-values"><span class="toc-section-number">6.3.4</span> P-values</a></li>
<li><a href="statistical-tests.html#multiple-hypothesis-testing-."><span class="toc-section-number">6.3.5</span> Multiple Hypothesis Testing .</a></li>
<li><a href="statistical-tests.html#statistical-power"><span class="toc-section-number">6.3.6</span> Statistical power</a></li>
</ul></li>
<li><a href="exploratory-data-analysis.html#exploratory-data-analysis"><span class="toc-section-number">6.4</span> Exploratory Data Analysis</a>
<ul>
<li><a href="exploratory-data-analysis.html#principal-component-analysis"><span class="toc-section-number">6.4.1</span> Principal Component Analysis</a></li>
<li><a href="exploratory-data-analysis.html#cluster-analysis"><span class="toc-section-number">6.4.2</span> Cluster Analysis</a></li>
<li><a href="exploratory-data-analysis.html#k-means-."><span class="toc-section-number">6.4.3</span> k-means .</a></li>
<li><a href="exploratory-data-analysis.html#others"><span class="toc-section-number">6.4.4</span> Others</a></li>
</ul></li>
<li><a href="network-analysis-..html#network-analysis-."><span class="toc-section-number">6.5</span> Network Analysis .</a>
<ul>
<li><a href="network-analysis-..html#basic-network-analysis"><span class="toc-section-number">6.5.1</span> Basic network analysis</a></li>
<li><a href="network-analysis-..html#create-your-own"><span class="toc-section-number">6.5.2</span> Create your own</a></li>
</ul></li>
</ul></li>
<li><a href="data-visualization.html#data-visualization"><span class="toc-section-number">7</span> Data Visualization</a>
<ul>
<li><a href="responsible-plotting.html#responsible-plotting"><span class="toc-section-number">7.1</span> Responsible Plotting</a>
<ul>
<li><a href="responsible-plotting.html#visualization-principles"><span class="toc-section-number">7.1.1</span> Visualization Principles</a></li>
<li><a href="responsible-plotting.html#human-visual-perception"><span class="toc-section-number">7.1.2</span> Human Visual Perception</a></li>
<li><a href="responsible-plotting.html#visual-encodings"><span class="toc-section-number">7.1.3</span> Visual Encodings</a></li>
<li><a href="responsible-plotting.html#some-opinionated-rules-of-thumb"><span class="toc-section-number">7.1.4</span> Some Opinionated Rules of Thumb</a></li>
</ul></li>
<li><a href="dv-gg.html#dv-gg"><span class="toc-section-number">7.2</span> Grammar of Graphics</a>
<ul>
<li><a href="dv-gg.html#ggplot-mechanics"><span class="toc-section-number">7.2.1</span> <code>ggplot</code> mechanics</a></li>
</ul></li>
<li><a href="plotting-one-dimension.html#plotting-one-dimension"><span class="toc-section-number">7.3</span> Plotting One Dimension</a>
<ul>
<li><a href="plotting-one-dimension.html#dv-bar"><span class="toc-section-number">7.3.1</span> Bar chart</a></li>
<li><a href="plotting-one-dimension.html#lollipop-plots"><span class="toc-section-number">7.3.2</span> Lollipop plots</a></li>
<li><a href="plotting-one-dimension.html#stacked-area-charts"><span class="toc-section-number">7.3.3</span> Stacked Area charts</a></li>
<li><a href="plotting-one-dimension.html#parallel-coordinate-plots"><span class="toc-section-number">7.3.4</span> Parallel Coordinate plots</a></li>
</ul></li>
<li><a href="visualizing-distributions.html#visualizing-distributions"><span class="toc-section-number">7.4</span> Visualizing Distributions</a>
<ul>
<li><a href="visualizing-distributions.html#histogram"><span class="toc-section-number">7.4.1</span> Histogram</a></li>
<li><a href="visualizing-distributions.html#density"><span class="toc-section-number">7.4.2</span> Density</a></li>
<li><a href="visualizing-distributions.html#boxplot"><span class="toc-section-number">7.4.3</span> Boxplot</a></li>
<li><a href="visualizing-distributions.html#violin-plot"><span class="toc-section-number">7.4.4</span> Violin plot</a></li>
<li><a href="visualizing-distributions.html#beeswarm-plot"><span class="toc-section-number">7.4.5</span> Beeswarm plot</a></li>
<li><a href="visualizing-distributions.html#ridgeline"><span class="toc-section-number">7.4.6</span> Ridgeline</a></li>
</ul></li>
<li><a href="plotting-two-or-more-dimensions.html#plotting-two-or-more-dimensions"><span class="toc-section-number">7.5</span> Plotting Two or More Dimensions</a>
<ul>
<li><a href="plotting-two-or-more-dimensions.html#scatter-plots"><span class="toc-section-number">7.5.1</span> Scatter Plots</a></li>
<li><a href="plotting-two-or-more-dimensions.html#bubble-plots"><span class="toc-section-number">7.5.2</span> Bubble Plots</a></li>
<li><a href="plotting-two-or-more-dimensions.html#connected-scatter-plots"><span class="toc-section-number">7.5.3</span> Connected Scatter Plots</a></li>
<li><a href="plotting-two-or-more-dimensions.html#line-plots"><span class="toc-section-number">7.5.4</span> Line Plots</a></li>
<li><a href="plotting-two-or-more-dimensions.html#parallel-coordinate-plots-1"><span class="toc-section-number">7.5.5</span> Parallel Coordinate Plots</a></li>
<li><a href="plotting-two-or-more-dimensions.html#heatmaps"><span class="toc-section-number">7.5.6</span> Heatmaps</a></li>
<li><a href="plotting-two-or-more-dimensions.html#how-to-use-heatmaps-responsibly"><span class="toc-section-number">7.5.7</span> How To Use Heatmaps Responsibly</a></li>
</ul></li>
<li><a href="other-kind-of-plots.html#other-kind-of-plots"><span class="toc-section-number">7.6</span> Other Kind of Plots</a>
<ul>
<li><a href="other-kind-of-plots.html#dendrograms"><span class="toc-section-number">7.6.1</span> Dendrograms</a></li>
<li><a href="other-kind-of-plots.html#chord-diagrams-and-circos-plots"><span class="toc-section-number">7.6.2</span> Chord Diagrams and Circos Plots</a></li>
</ul></li>
<li><a href="multiple-plots.html#multiple-plots"><span class="toc-section-number">7.7</span> Multiple Plots</a>
<ul>
<li><a href="multiple-plots.html#facet-wrapping"><span class="toc-section-number">7.7.1</span> Facet wrapping</a></li>
<li><a href="multiple-plots.html#multipanel-figures"><span class="toc-section-number">7.7.2</span> Multipanel Figures</a></li>
</ul></li>
<li><a href="publication-ready-plots.html#publication-ready-plots"><span class="toc-section-number">7.8</span> Publication Ready Plots</a>
<ul>
<li><a href="publication-ready-plots.html#ggplot-themes"><span class="toc-section-number">7.8.1</span> <code>ggplot</code> Themes</a></li>
<li><a href="publication-ready-plots.html#exporting-to-svg"><span class="toc-section-number">7.8.2</span> Exporting to SVG</a></li>
</ul></li>
<li><a href="network-visualization.html#network-visualization"><span class="toc-section-number">7.9</span> Network visualization</a>
<ul>
<li><a href="network-visualization.html#interactive-network-plots"><span class="toc-section-number">7.9.1</span> Interactive network plots</a></li>
</ul></li>
</ul></li>
<li><a href="biology-bioinformatics.html#biology-bioinformatics"><span class="toc-section-number">8</span> Biology &amp; Bioinformatics</a>
<ul>
<li><a href="bio-r.html#bio-r"><span class="toc-section-number">8.1</span> R in Biology</a></li>
<li><a href="biological-data-overview.html#biological-data-overview"><span class="toc-section-number">8.2</span> Biological Data Overview</a>
<ul>
<li><a href="biological-data-overview.html#bio-data-types"><span class="toc-section-number">8.2.1</span> Types of Biological Data</a></li>
<li><a href="biological-data-overview.html#csv-files"><span class="toc-section-number">8.2.2</span> CSV Files</a></li>
<li><a href="biological-data-overview.html#biological-data-is-not-tidy"><span class="toc-section-number">8.2.3</span> Biological data is NOT Tidy!</a></li>
</ul></li>
<li><a href="bio-bioconductor.html#bio-bioconductor"><span class="toc-section-number">8.3</span> Bioconductor</a></li>
<li><a href="microarrays.html#microarrays"><span class="toc-section-number">8.4</span> Microarrays</a></li>
<li><a href="high-throughput-sequencing.html#high-throughput-sequencing"><span class="toc-section-number">8.5</span> High Throughput Sequencing</a>
<ul>
<li><a href="high-throughput-sequencing.html#raw-hts-data"><span class="toc-section-number">8.5.1</span> Raw HTS Data</a></li>
<li><a href="high-throughput-sequencing.html#preliminary-hts-data-analysis"><span class="toc-section-number">8.5.2</span> Preliminary HTS Data Analysis</a></li>
<li><a href="high-throughput-sequencing.html#count-data"><span class="toc-section-number">8.5.3</span> Count Data</a></li>
</ul></li>
<li><a href="gene-identifiers.html#gene-identifiers"><span class="toc-section-number">8.6</span> Gene Identifiers</a>
<ul>
<li><a href="gene-identifiers.html#gene-identifier-systems"><span class="toc-section-number">8.6.1</span> Gene Identifier Systems</a></li>
<li><a href="gene-identifiers.html#mapping-between-identifier-systems"><span class="toc-section-number">8.6.2</span> Mapping Between Identifier Systems</a></li>
<li><a href="gene-identifiers.html#mapping-homologs"><span class="toc-section-number">8.6.3</span> Mapping Homologs</a></li>
</ul></li>
<li><a href="gene-expression.html#gene-expression"><span class="toc-section-number">8.7</span> Gene Expression</a>
<ul>
<li><a href="gene-expression.html#gene-expression-data-in-bioconductor"><span class="toc-section-number">8.7.1</span> Gene Expression Data in Bioconductor</a></li>
<li><a href="gene-expression.html#differential-expression-analysis"><span class="toc-section-number">8.7.2</span> Differential Expression Analysis</a></li>
<li><a href="gene-expression.html#microarray-gene-expression-data"><span class="toc-section-number">8.7.3</span> Microarray Gene Expression Data</a></li>
<li><a href="gene-expression.html#differential-expression-microarrays-limma"><span class="toc-section-number">8.7.4</span> Differential Expression: Microarrays (limma)</a></li>
<li><a href="gene-expression.html#rnaseq"><span class="toc-section-number">8.7.5</span> RNASeq</a></li>
<li><a href="gene-expression.html#rnaseq-gene-expression-data"><span class="toc-section-number">8.7.6</span> RNASeq Gene Expression Data</a></li>
<li><a href="gene-expression.html#filtering-counts"><span class="toc-section-number">8.7.7</span> Filtering Counts</a></li>
<li><a href="gene-expression.html#count-distributions"><span class="toc-section-number">8.7.8</span> Count Distributions</a></li>
<li><a href="gene-expression.html#differential-expression-rnaseq"><span class="toc-section-number">8.7.9</span> Differential Expression: RNASeq</a></li>
</ul></li>
<li><a href="gene-set-enrichment-analysis.html#gene-set-enrichment-analysis"><span class="toc-section-number">8.8</span> Gene Set Enrichment Analysis</a>
<ul>
<li><a href="gene-set-enrichment-analysis.html#gene-sets"><span class="toc-section-number">8.8.1</span> Gene Sets</a></li>
<li><a href="gene-set-enrichment-analysis.html#working-with-gene-sets-in-r"><span class="toc-section-number">8.8.2</span> Working with gene sets in R</a></li>
<li><a href="gene-set-enrichment-analysis.html#over-representation-analysis"><span class="toc-section-number">8.8.3</span> Over-representation Analysis</a></li>
<li><a href="gene-set-enrichment-analysis.html#rank-based-analysis"><span class="toc-section-number">8.8.4</span> Rank-based Analysis</a></li>
<li><a href="gene-set-enrichment-analysis.html#fgsea"><span class="toc-section-number">8.8.5</span> fgsea</a></li>
</ul></li>
<li><a href="biological-networks-..html#biological-networks-."><span class="toc-section-number">8.9</span> Biological Networks .</a>
<ul>
<li><a href="biological-networks-..html#biological-pathways-."><span class="toc-section-number">8.9.1</span> Biological Pathways .</a></li>
<li><a href="biological-networks-..html#gene-regulatory-networks-."><span class="toc-section-number">8.9.2</span> Gene Regulatory Networks .</a></li>
<li><a href="biological-networks-..html#protein-protein-interaction-network-."><span class="toc-section-number">8.9.3</span> Protein-Protein Interaction Network .</a></li>
<li><a href="biological-networks-..html#wgcna-."><span class="toc-section-number">8.9.4</span> WGCNA .</a></li>
</ul></li>
</ul></li>
<li><a href="engineering.html#engineering"><span class="toc-section-number">9</span> EngineeRing</a>
<ul>
<li><a href="unit-testing.html#unit-testing"><span class="toc-section-number">9.1</span> Unit Testing</a></li>
<li><a href="toolification.html#toolification"><span class="toc-section-number">9.2</span> Toolification</a>
<ul>
<li><a href="toolification.html#the-r-interpreter"><span class="toc-section-number">9.2.1</span> The R Interpreter</a></li>
<li><a href="toolification.html#rscript"><span class="toc-section-number">9.2.2</span> <code>Rscript</code></a></li>
<li><a href="toolification.html#commandargs"><span class="toc-section-number">9.2.3</span> <code>commandArgs()</code></a></li>
</ul></li>
<li><a href="parallel-processing.html#parallel-processing"><span class="toc-section-number">9.3</span> Parallel Processing</a>
<ul>
<li><a href="parallel-processing.html#brief-introduction-to-parallelization"><span class="toc-section-number">9.3.1</span> Brief Introduction to Parallelization</a></li>
<li><a href="parallel-processing.html#apply-and-friends-are-pleasingly-parallel"><span class="toc-section-number">9.3.2</span> <code>apply</code> and Friends Are Pleasingly Parallel</a></li>
<li><a href="parallel-processing.html#the-parallel-package"><span class="toc-section-number">9.3.3</span> The <code>parallel</code> package</a></li>
</ul></li>
<li><a href="object-orientation-in-r.html#object-orientation-in-r"><span class="toc-section-number">9.4</span> Object Orientation in R</a></li>
<li><a href="r-packages.html#r-packages"><span class="toc-section-number">9.5</span> R Packages</a></li>
</ul></li>
<li><a href="rshiny.html#rshiny"><span class="toc-section-number">10</span> RShiny</a>
<ul>
<li><a href="rshiny-introduction.html#rshiny-introduction"><span class="toc-section-number">10.1</span> RShiny Introduction</a>
<ul>
<li><a href="rshiny-introduction.html#how-does-a-web-page-work"><span class="toc-section-number">10.1.1</span> How does a web page work?</a></li>
</ul></li>
<li><a href="application-structure.html#application-structure"><span class="toc-section-number">10.2</span> Application Structure</a>
<ul>
<li><a href="application-structure.html#ui-design"><span class="toc-section-number">10.2.1</span> UI Design</a></li>
<li><a href="application-structure.html#server-functionality"><span class="toc-section-number">10.2.2</span> Server Functionality</a></li>
</ul></li>
<li><a href="reactivity.html#reactivity"><span class="toc-section-number">10.3</span> Reactivity</a></li>
<li><a href="publishing.html#publishing"><span class="toc-section-number">10.4</span> Publishing</a></li>
</ul></li>
<li><a href="communicating-with-r.html#communicating-with-r"><span class="toc-section-number">11</span> Communicating with R</a>
<ul>
<li><a href="rmarkdown-knitr.html#rmarkdown-knitr"><span class="toc-section-number">11.1</span> RMarkdown &amp; knitr</a></li>
<li><a href="bookdown.html#bookdown"><span class="toc-section-number">11.2</span> bookdown</a></li>
</ul></li>
<li><a href="contribution-guide.html#contribution-guide"><span class="toc-section-number">12</span> Contribution Guide</a>
<ul>
<li><a href="custom-blocks.html#custom-blocks"><span class="toc-section-number">12.1</span> Custom blocks</a></li>
</ul></li>
<li><a href="assignments.html#assignments">Assignments</a>
<ul>
<li><a href="assignment-format.html#assignment-format">Assignment Format</a></li>
<li><a href="starting-an-assignment.html#starting-an-assignment">Starting an Assignment</a>
<ul>
<li><a href="starting-an-assignment.html#starting-your-own-git-repository-and-cloning-it">Starting your own <code>git</code> repository and cloning it</a></li>
<li><a href="starting-an-assignment.html#contents">Contents</a></li>
<li><a href="starting-an-assignment.html#github-tutorial">GitHub tutorial</a></li>
<li><a href="starting-an-assignment.html#committing-and-pushing-with-r-and-without">Committing and Pushing with R (and without)</a></li>
<li><a href="starting-an-assignment.html#using-rstudio">Using RStudio</a></li>
<li><a href="starting-an-assignment.html#using-the-command-line">Using the command line</a></li>
</ul></li>
<li><a href="assignment-1.html#assignment-1">Assignment 1</a>
<ul>
<li><a href="assignment-1.html#problem-statement">Problem Statement</a></li>
<li><a href="assignment-1.html#learning-objectives-and-skill-list">Learning Objectives and Skill List</a></li>
<li><a href="assignment-1.html#instructions">Instructions</a></li>
<li><a href="assignment-1.html#hints">Hints</a></li>
</ul></li>
<li><a href="assignment-2.html#assignment-2">Assignment 2</a>
<ul>
<li><a href="assignment-2.html#problem-statement-1">Problem Statement</a></li>
<li><a href="assignment-2.html#learning-objectives">Learning Objectives</a></li>
<li><a href="assignment-2.html#skill-list">Skill List</a></li>
<li><a href="assignment-2.html#instructions-1">Instructions</a></li>
<li><a href="assignment-2.html#function-details">Function Details</a></li>
</ul></li>
<li><a href="assignment-3.html#assignment-3">Assignment 3</a>
<ul>
<li><a href="assignment-3.html#problem-statement-2">Problem Statement</a></li>
<li><a href="assignment-3.html#learning-objectives-1">Learning Objectives</a></li>
<li><a href="assignment-3.html#skill-list-1">Skill List</a></li>
<li><a href="assignment-3.html#background-on-microarrays">Background on Microarrays</a></li>
<li><a href="assignment-3.html#background-on-principal-component-analysis">Background on Principal Component Analysis</a></li>
<li><a href="assignment-3.html#marisa-et-al.-gene-expression-classification-of-colon-cancer-into-molecular-subtypes-characterization-validation-and-prognostic-value.-plos-medicine-may-2013.-pmid-23700391">Marisa et al. Gene Expression Classification of Colon Cancer into Molecular Subtypes: Characterization, Validation, and Prognostic Value. PLoS Medicine, May 2013. PMID: 23700391</a></li>
<li><a href="assignment-3.html#scaling-data-using-r-scale">Scaling data using R <code>scale()</code></a></li>
<li><a href="assignment-3.html#proportion-of-variance-explained">Proportion of variance explained</a></li>
<li><a href="assignment-3.html#plotting-and-visualization-of-pca">Plotting and visualization of PCA</a></li>
<li><a href="assignment-3.html#hierarchical-clustering-and-heatmaps">Hierarchical Clustering and Heatmaps</a></li>
<li><a href="assignment-3.html#references">References</a></li>
</ul></li>
<li><a href="assignment-4.html#assignment-4">Assignment 4</a>
<ul>
<li><a href="assignment-4.html#problem-statement-3">Problem Statement</a></li>
<li><a href="assignment-4.html#learning-objectives-2">Learning Objectives</a></li>
<li><a href="assignment-4.html#skill-list-2">Skill List</a></li>
<li><a href="assignment-4.html#instructions-2">Instructions</a></li>
<li><a href="assignment-4.html#deliverables">Deliverables</a></li>
<li><a href="assignment-4.html#function-details-1">Function Details</a></li>
<li><a href="assignment-4.html#hints-1">Hints</a></li>
</ul></li>
<li><a href="assignment-5.html#assignment-5">Assignment 5</a>
<ul>
<li><a href="assignment-5.html#problem-statement-4">Problem Statement</a></li>
<li><a href="assignment-5.html#learning-objectives-3">Learning Objectives</a></li>
<li><a href="assignment-5.html#skill-list-3">Skill List</a></li>
<li><a href="assignment-5.html#deseq2-background">DESeq2 Background</a></li>
<li><a href="assignment-5.html#generating-a-counts-matrix">Generating a counts matrix</a></li>
<li><a href="assignment-5.html#prefiltering-counts-matrix">Prefiltering Counts matrix</a></li>
<li><a href="assignment-5.html#median-of-ratios-normalization">Median-of-ratios normalization</a></li>
<li><a href="assignment-5.html#deseq2-preparation">DESeq2 preparation</a></li>
<li><a href="assignment-5.html#omeara-et-al.-transcriptional-reversion-of-cardiac-myocyte-fate-during-mammalian-cardiac-regeneration.-circ-res.-feb-2015.-pmid-25477501l">O’Meara et al. Transcriptional Reversion of Cardiac Myocyte Fate During Mammalian Cardiac Regeneration. Circ Res. Feb 2015. PMID: 25477501l</a></li>
<li><a href="assignment-5.html#reading-and-subsetting-the-data-from-verse_counts.tsv-and-sample_metadata.csv">1. Reading and subsetting the data from verse_counts.tsv and sample_metadata.csv</a></li>
<li><a href="assignment-5.html#running-deseq2">2. Running DESeq2</a></li>
<li><a href="assignment-5.html#annotating-results-to-construct-a-labeled-volcano-plot">3. Annotating results to construct a labeled volcano plot</a></li>
<li><a href="assignment-5.html#diagnostic-plot-of-the-raw-p-values-for-all-genes">4. Diagnostic plot of the raw p-values for all genes</a></li>
<li><a href="assignment-5.html#plotting-the-logfoldchanges-for-differentially-expressed-genes">5. Plotting the LogFoldChanges for differentially expressed genes</a></li>
<li><a href="assignment-5.html#the-choice-of-fdr-cutoff-depends-on-cost">The choice of FDR cutoff depends on cost</a></li>
<li><a href="assignment-5.html#plotting-the-normalized-counts-of-differentially-expressed-genes">6. Plotting the normalized counts of differentially expressed genes</a></li>
<li><a href="assignment-5.html#volcano-plot-to-visualize-differential-expression-results">7. Volcano Plot to visualize differential expression results</a></li>
<li><a href="assignment-5.html#running-fgsea-vignette">8. Running fgsea <span>vignette</span></a></li>
<li><a href="assignment-5.html#plotting-the-top-ten-positive-nes-and-top-ten-negative-nes-pathways">9. Plotting the top ten positive NES and top ten negative NES pathways</a></li>
<li><a href="assignment-5.html#references-1">References</a></li>
</ul></li>
<li><a href="assignment-6.html#assignment-6">Assignment 6</a>
<ul>
<li><a href="assignment-6.html#problem-statement-5">Problem Statement</a></li>
<li><a href="assignment-6.html#learning-objectives-4">Learning Objectives</a></li>
<li><a href="assignment-6.html#skill-list-4">Skill List</a></li>
<li><a href="assignment-6.html#instructions-3">Instructions</a></li>
<li><a href="assignment-6.html#function-details-2">Function Details</a></li>
</ul></li>
<li><a href="assignment-7.html#assignment-7">Assignment 7</a>
<ul>
<li><a href="assignment-7.html#problem-statement-6">Problem Statement</a></li>
<li><a href="assignment-7.html#learning-objectives-5">Learning Objectives</a></li>
<li><a href="assignment-7.html#skill-list-5">Skill List</a></li>
<li><a href="assignment-7.html#instructions-4">Instructions</a></li>
<li><a href="assignment-7.html#function-details-3">“Function” Details</a></li>
</ul></li>
</ul></li>
<li class="appendix"><span><b>Appendix</b></span></li>
<li><a href="class-outlines.html#class-outlines"><span class="toc-section-number">A</span> Class Outlines</a>
<ul>
<li><a href="week-1.html#week-1"><span class="toc-section-number">A.1</span> Week 1</a></li>
<li><a href="week-2.html#week-2"><span class="toc-section-number">A.2</span> Week 2</a></li>
<li><a href="week-3.html#week-3"><span class="toc-section-number">A.3</span> Week 3</a></li>
<li><a href="week-4.html#week-4"><span class="toc-section-number">A.4</span> Week 4</a></li>
<li><a href="week-5.html#week-5"><span class="toc-section-number">A.5</span> Week 5</a></li>
<li><a href="week-6.html#week-6"><span class="toc-section-number">A.6</span> Week 6</a></li>
<li><a href="week-7.html#week-7"><span class="toc-section-number">A.7</span> Week 7</a></li>
<li><a href="week-8.html#week-8"><span class="toc-section-number">A.8</span> Week 8</a></li>
<li><a href="week-9.html#week-9"><span class="toc-section-number">A.9</span> Week 9</a></li>
<li><a href="week-10.html#week-10"><span class="toc-section-number">A.10</span> Week 10</a></li>
<li><a href="week-11.html#week-11"><span class="toc-section-number">A.11</span> Week 11</a></li>
<li><a href="week-12.html#week-12"><span class="toc-section-number">A.12</span> Week 12</a></li>
<li><a href="week-13.html#week-13"><span class="toc-section-number">A.13</span> Week 13</a></li>
<li><a href="week-14.html#week-14"><span class="toc-section-number">A.14</span> Week 14</a></li>
</ul></li>
</ul>
</div>
</div>
</div>
<div class="row">
<div class="col-sm-12">
<div id="statistical-distributions" class="section level2" number="6.2">
<h2><span class="header-section-number">6.2</span> Statistical Distributions</h2>
<p>Biological data, like all data, is uncertain. Measurements always contain noise,
but a collection of measurements do not always contain signal. The field of
statistics grew from the recognition that mathematics can be used to quantify
uncertainty and help us reason about whether a signal exists within a dataset,
and how certain we are of that signal. At its core, statistics is about
separating the signal from the noise of a dataset in a rigorous and precise way.</p>
<p>One of the fundamental statistical tools used when estimating uncertainty is the
<em>statistical distribution</em>, or <em>probability distribution</em>, or simply
<em>distribution</em>. There are two broad classes of distributions: statistical, or
theoretical, distributions and empirical distributions. In this section we will
discuss some general properties of distributions, briefly describe some common
probability distributions, and explain how to specify and use these
distributions in R.</p>
<div id="random-variables" class="section level3" number="6.2.1">
<h3><span class="header-section-number">6.2.1</span> Random Variables</h3>
<p>A random variable is an object or quantity which depends upon random events and
that can have samples drawn from it. For example, a six-sided die is a random
variable with six possible outcomes, and a single roll of the die will yield one
of those outcomes with some probability (equal probability, if the die is fair).
A coin is also a random variable, with heads or tails as the outcome. A gene’s
expression in a RNASeq experiment is a random variable, where the possible
outcomes are any non-negative integer corresponding to the number of reads that
map to it. In these examples, the outcome is a simple category or real number,
but random variables can be associated with more complex outcomes as well,
including trees, sets, shapes, sequences, etc. The probability of each outcome
is specified by the distribution associated with the random variable.</p>
<p>Random variables are usually notated as capital letters, like <span class="math inline">\(X,Y,Z,\)</span> etc. A
sample drawn from a random variable is usually notated as a lowercase of the
same letter, e.g. <span class="math inline">\(x\)</span> is a sample drawn from the random variable <span class="math inline">\(X\)</span>. The
distribution of a random variable is usually described like “<span class="math inline">\(X\)</span> follows a
binomial distribution” or “<span class="math inline">\(Y\)</span> is a normally distributed random variable.”</p>
<p>The probability of a random variable taking one of its possible values is
usually written <span class="math inline">\(P(X = x)\)</span>. The value of <span class="math inline">\(P(X = x)\)</span> is defined by the
distribution of <span class="math inline">\(X\)</span>. As we will see later in the <a href="statistical-tests.html#p-values">p-values</a> section, sometimes
we are also interested in the probability of observing a value of <span class="math inline">\(x\)</span> or larger
(or smaller). These probabilities are written <span class="math inline">\(P(X &gt; x)\)</span> (or <span class="math inline">\(P(X &lt; x)\)</span>). How
these probabilities are computed is described in the next section.</p>
</div>
<div id="statistical-distribution-basics" class="section level3" number="6.2.2">
<h3><span class="header-section-number">6.2.2</span> Statistical Distribution Basics</h3>
<p>By definition, a statistical distribution is a function that maps <strong>the possible
values for a variable to how often they occur</strong>. Said differently, a statistical
distribution is used to compute the probability of seeing a single value, or a
range of values, relative to all other possible values, assuming the random
variable in question follows the statistical distribution. The following is a
visualization of the theoretical distribution of a normally distributed random
variable:</p>
<div class="sourceCode" id="cb184"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb184-1"><a href="statistical-distributions.html#cb184-1" aria-hidden="true" tabindex="-1"></a><span class="fu">tibble</span>(</span>
<span id="cb184-2"><a href="statistical-distributions.html#cb184-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">x =</span> <span class="fu">seq</span>(<span class="sc">-</span><span class="dv">4</span>,<span class="dv">4</span>,<span class="at">by=</span><span class="fl">0.1</span>),</span>
<span id="cb184-3"><a href="statistical-distributions.html#cb184-3" aria-hidden="true" tabindex="-1"></a>  <span class="st">`</span><span class="at">Probability Density</span><span class="st">`</span><span class="ot">=</span><span class="fu">dnorm</span>(x,<span class="dv">0</span>,<span class="dv">1</span>)</span>
<span id="cb184-4"><a href="statistical-distributions.html#cb184-4" aria-hidden="true" tabindex="-1"></a>) <span class="sc">%&gt;%</span></span>
<span id="cb184-5"><a href="statistical-distributions.html#cb184-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x=</span>x,<span class="at">y=</span><span class="st">`</span><span class="at">Probability Density</span><span class="st">`</span>)) <span class="sc">+</span></span>
<span id="cb184-6"><a href="statistical-distributions.html#cb184-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_line</span>() <span class="sc">+</span></span>
<span id="cb184-7"><a href="statistical-distributions.html#cb184-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title=</span><span class="st">&quot;Probability Density Function for a Normal Distribution&quot;</span>)</span></code></pre></div>
<p><img src="main_files/stat%20normal-1.png" width="672" /></p>
<p>The plot above depicts to the <em>probability density function</em> (PDF) of a normal
distribution with mean of zero and a standard deviation of one. The PDF defines
the probability associated with every possible value of <span class="math inline">\(x\)</span>, which for the
normal distribution is all real numbers. All PDFs have a closed mathematical
form. The PDF for the normal distribution is:</p>
<p><span class="math display">\[
P(X = x|\mu,\sigma) = \frac{1}{\sigma\sqrt{2\pi}}e^{\frac{-(x-\mu)^2}{2\sigma}}
\]</span></p>
<p>The notation <span class="math inline">\(P(X = x|\mu,\sigma)\)</span> is read like “the probability that the random
variable <span class="math inline">\(X\)</span> takes the value <span class="math inline">\(x\)</span>, given mean <span class="math inline">\(\mu\)</span> and standard deviation
<span class="math inline">\(\sigma\)</span>.” The normal distribution is an example of a <em>parametric distribution</em>
because its PDF requires two parameters - mean and standard deviation - to
compute probabilities. The choice of parameter values <span class="math inline">\(\mu\)</span> and <span class="math inline">\(\sigma\)</span> in the
normal distribution determine the probability of the value of <span class="math inline">\(x\)</span>.</p>
<div class="box note">
<p>Probability density functions are defined for continuous distributions only,
as described later in this section. Discrete distributions have a probability
mass function (PMF) instead of a PDF, since the probability distribution is
subdivided into a set of different categories. PMFs have different mathematical
characteristics than PDFs (e.g. they are not continuous and therefore are not
differentiable), but serve the same purpose.</p>
</div>
<p>In probability theory, if a plausible event has a probability of zero, this
<em>does not mean that event can never occur</em>. In fact, every specific value in a
distribution that supports all real numbers has a probability of zero (i.e. one
specific value out of an infinite number of values). Instead, the probability
distribution function allows us to reason about the <em>relative likelihood</em> of
observing values in one range of the distribution compared with the others.
While most values have extremely small relative probabilities, they never are
equal to zero. This is due to the asymptotic properties of probability
distributions, where every <a href="https://en.wikipedia.org/wiki/Support_(mathematics)">supported
value</a> of the distribution
has a non-zero value by definition, though many values may be very close to
zero.</p>
<p>The PDF provides the probability density of specific values within the
distribution, but sometimes we are interested in the probability of a value
being less than or equal to a particular value. We compute this using the
<em>cumulative distribution function</em> (CDF) or sometimes just <em>distribution
function</em>. In the plot below, both the PDF and CDF are plotted for a normal
distribution with mean zero and standard deviation of 1:</p>
<div class="sourceCode" id="cb185"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb185-1"><a href="statistical-distributions.html#cb185-1" aria-hidden="true" tabindex="-1"></a><span class="fu">tibble</span>(</span>
<span id="cb185-2"><a href="statistical-distributions.html#cb185-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">x =</span> <span class="fu">seq</span>(<span class="sc">-</span><span class="dv">4</span>,<span class="dv">4</span>,<span class="at">by=</span><span class="fl">0.1</span>),</span>
<span id="cb185-3"><a href="statistical-distributions.html#cb185-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">PDF=</span><span class="fu">dnorm</span>(x,<span class="dv">0</span>,<span class="dv">1</span>),</span>
<span id="cb185-4"><a href="statistical-distributions.html#cb185-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">CDF=</span><span class="fu">pnorm</span>(x,<span class="dv">0</span>,<span class="dv">1</span>)</span>
<span id="cb185-5"><a href="statistical-distributions.html#cb185-5" aria-hidden="true" tabindex="-1"></a>) <span class="sc">%&gt;%</span></span>
<span id="cb185-6"><a href="statistical-distributions.html#cb185-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb185-7"><a href="statistical-distributions.html#cb185-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_line</span>(<span class="fu">aes</span>(<span class="at">x=</span>x,<span class="at">y=</span>PDF,<span class="at">color=</span><span class="st">&quot;PDF&quot;</span>)) <span class="sc">+</span></span>
<span id="cb185-8"><a href="statistical-distributions.html#cb185-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_line</span>(<span class="fu">aes</span>(<span class="at">x=</span>x,<span class="at">y=</span>CDF,<span class="at">color=</span><span class="st">&quot;CDF&quot;</span>),<span class="at">linetype=</span><span class="st">&quot;dashed&quot;</span>)</span></code></pre></div>
<p><img src="main_files/stat%20pdf%20cdf-1.png" width="672" /></p>
<p>The value of the CDF corresponds to the area under the density curve up to the
corresponding value of <span class="math inline">\(x\)</span>; 1 minus that value is the area under the curve
greater than that value. The following figures illustrate this:</p>
<p><img src="main_files/stat%20cdfs-1.png" width="672" /></p>
<p>91% of the probability density is less than the (arbitrary) value of <span class="math inline">\(x=1.37\)</span>,
and likewise 9% is greater. This is how <a href="statistical-tests.html#p-values">p-values</a> are calculated, as described
later.</p>
<p>The CDF is also useful for generating samples from the distribution. In the
following plot, 1000 random samples were drawn from a uniform distribution in
the interval <span class="math inline">\((0,1)\)</span> and plotted using the inverse CDF function:</p>
<p><img src="main_files/stat%20inv%20cdf-1.png" width="672" /></p>
<p>The histograms on the margins show the distributions of the <span class="math inline">\(p\)</span> and <span class="math inline">\(x\)</span> values
for the scatter points. The <span class="math inline">\(p\)</span> coordinates are uniform, while the <span class="math inline">\(x\)</span>
distribution is normal with mean of zero and standard deviation of one. In this
way, we can draw samples from a normal distribution, or any other distribution
that has an invertible CDF.</p>
</div>
<div id="distributions-in-r" class="section level3" number="6.2.3">
<h3><span class="header-section-number">6.2.3</span> Distributions in R</h3>
<p>There are four key operations we performed with the normal distribution in the
previous section:</p>
<ol style="list-style-type: decimal">
<li>Calculate probabilities using the PDF</li>
<li>Calculate cumulative probabilities using the CDF</li>
<li>Calculate the value associated with a cumulative probability</li>
<li>Sample values from a parameterized distribution</li>
</ol>
<p>In R, each of these operations has a dedicated function for each different
distribution, prefixed by <code>d</code>, <code>p</code>, <code>q</code>, and <code>r</code>. For the normal distribution,
these functions are <code>dnorm</code>, <code>pnorm</code>, <code>qnorm</code>, and <code>rnorm</code>:</p>
<ol style="list-style-type: decimal">
<li><code>dnorm(x, mean=0, sd=1)</code> - PDF of the normal distribution</li>
<li><code>pnorm(q, mean=0, sd=1)</code> - CDF of the normal distribution</li>
<li><code>qnorm(p, mean=0, sd=1)</code> - inverse CDF; accepts quantiles between 0 and 1 and returns the
value of the distribution for those quantiles</li>
<li><code>rnorm(n, mean=0, sd=1)</code> - generate <code>n</code> samples from a normal distribution</li>
</ol>
<p>R uses this scheme for all of its base distributions, which include:</p>
<table style="width:71%;">
<colgroup>
<col width="27%" />
<col width="43%" />
</colgroup>
<thead>
<tr class="header">
<th>Distribution</th>
<th>Probability Density Function</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>Normal</td>
<td><code>dnorm(x,mean,sd)</code></td>
</tr>
<tr class="even">
<td>t Distribution</td>
<td><code>dt(x,df)</code></td>
</tr>
<tr class="odd">
<td>Poisson</td>
<td><code>dpois(n,lambda)</code></td>
</tr>
<tr class="even">
<td>Binomial</td>
<td><code>dbinom(x, size, prob)</code></td>
</tr>
<tr class="odd">
<td>Negative Binomial</td>
<td><code>dnbinom(x, size, prob, mu)</code></td>
</tr>
<tr class="even">
<td>Exponential</td>
<td><code>dexp(x, rate)</code></td>
</tr>
<tr class="odd">
<td><span class="math inline">\(\chi^2\)</span></td>
<td><code>dchisq(x, df)</code></td>
</tr>
</tbody>
</table>
<p>There are <a href="https://www.stat.umn.edu/geyer/old/5101/rlook.html">many more distributions</a>
implemented in R beyond these, and they all follow the same scheme.</p>
<p>The next two sections will cover examples of some of these distributions.
Generally, statistical distributions are divided into two categories: discrete
distributions and continuous distributions.</p>
</div>
<div id="discrete-distributions" class="section level3" number="6.2.4">
<h3><span class="header-section-number">6.2.4</span> Discrete Distributions</h3>
<p>Discrete distributions are defined over countable, possibly infinite sets.
Common discrete distributions include binomial (e.g. coin flips), multinomial
(e.g. dice rolls), Poisson (e.g. number of WGS sequencing reads that map to a
specific locus), etc. Below we describe a few of these in detail.</p>
<div id="bernoulli-random-trail-and-more" class="section level4" number="6.2.4.1">
<h4><span class="header-section-number">6.2.4.1</span> Bernoulli random trail and more</h4>
<p>One of the examples for discrete random variable distribution is the <strong>Bernoulli</strong>
function. A Bernoulli trail has only 2 outcomes with probability <span class="math inline">\(p\)</span> and <span class="math inline">\((1-p)\)</span>.
Consider flipping a fair coin, and the random variable X can take value 0 or 1
indicating you get a head or a tail. If it’s a fair coin, we would expect the
<span class="math inline">\(Pr(X = 0) = Pr(X = 1) = 0.5\)</span>. Or, if we throw a die and we record X = 1 when
we get a six, and X = 0 otherwise, then <span class="math inline">\(Pr(X = 0) = 5/6\)</span> and <span class="math inline">\(Pr(X = 1) = 1/6\)</span>.</p>
<p>Now consider a slightly complicated situation: what if we are throwing the die
<span class="math inline">\(n\)</span> times and we would like to analyze the total number of six, say <span class="math inline">\(x\)</span>, we get
during those <span class="math inline">\(n\)</span> throws? Now, this leads us to the <strong>binomial distribution</strong>.
If it’s a fair die, we would say our proportion parameter <span class="math inline">\(p = 1/6\)</span>, which means
the probability we are getting a six is 1/6 for each throw.</p>
<p><span class="math display">\[
\begin{equation}
f(x) = \frac {n!} {x!(n-x)!}p^x (1-p) ^{(n-x)}
\end{equation}
\]</span></p>
<p>The <strong>geometric</strong> random variable, similar to the binomial, is also from a
sequence of random Bernoulli trials with a constant probability parameter <span class="math inline">\(p\)</span>.
But this time, we define the random variable <span class="math inline">\(X\)</span> as the number of consecutive
failures before the first success. In this case, the probability of <span class="math inline">\(x\)</span>
consecutive failures followed by success on trial <span class="math inline">\(x+1\)</span> is:</p>
<p><span class="math display">\[
\begin{equation}
f(x) = p * (1-p)^x
\end{equation}
\]</span></p>
<p>The <strong>negative binomial</strong> distribution goes one step forward. This time, we are
still performing a sequence of independent Bernoulli random trials with a
constant probability of success equal to <span class="math inline">\(p\)</span>. But now we would like to record
the random variable <span class="math inline">\(Z\)</span> to be the total number of failures before we finally get
to the <span class="math inline">\(r^{th}\)</span> success. In other words, when we get to the <span class="math inline">\(r^{th}\)</span> success, we
had <span class="math inline">\(x+r\)</span> Bernoulli random trails, in which <span class="math inline">\(x\)</span> times failed and <span class="math inline">\(r\)</span> times
succeeded.</p>
<p><span class="math display">\[
\begin{equation}
f(x) = \frac {x+r-1} {r-1} p^r {(1-p)}^x
\end{equation}
\]</span></p>
</div>
<div id="poisson" class="section level4" number="6.2.4.2">
<h4><span class="header-section-number">6.2.4.2</span> Poisson</h4>
<p>The Poisson distribution is used to express the probability of a given number
of events occurring in a fixed interval of time or space, and these events
occur with a known constant mean rate and independently of the time since
the last event. <del>But no one understands this definition.</del></p>
<p>The formula for Poisson distribution is:</p>
<p><span class="math display">\[
\begin{equation}
f(k; \lambda) = Pr(X=k) = \frac {\lambda^k e^{-\lambda}} {k!}
\end{equation}
\]</span></p>
<ul>
<li>lambda is the expected value of the random variable X<br />
</li>
<li>k is the number of occurrences<br />
</li>
<li>e is Euler’s number (e=2.71828)<br />
<del>okay, the formula makes it even more confusing.</del></li>
</ul>
<p>Imagine you are working at a mail reception center, and your responsibility
is to receive incoming letters. Assume the number of incoming letters is not
affected by the day of the week or season of the year. You are expected to get
20 letters on average in a day. But, the actual number of letters you receive
each day will not be perfectly 20.</p>
<p>You recorded the number of letters you receive each day in a month (30 days).
In the following plot, each dot represents a day. The x-axis is calender day,
and y-axis is the number of letters you receive on that day. Although on average
you are receiving 20 letters each day, the actual number of letters each day
vary a lot.</p>
<div class="sourceCode" id="cb186"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb186-1"><a href="statistical-distributions.html#cb186-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">2</span>)</span>
<span id="cb186-2"><a href="statistical-distributions.html#cb186-2" aria-hidden="true" tabindex="-1"></a>my_letter <span class="ot">&lt;-</span> <span class="fu">rpois</span>(<span class="at">n =</span> <span class="dv">30</span>, <span class="at">lambda =</span> <span class="dv">20</span>)</span>
<span id="cb186-3"><a href="statistical-distributions.html#cb186-3" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(my_letter,</span>
<span id="cb186-4"><a href="statistical-distributions.html#cb186-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">main =</span> <span class="st">&quot;Letters received each day&quot;</span>,</span>
<span id="cb186-5"><a href="statistical-distributions.html#cb186-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">xlab =</span> <span class="st">&quot;day of the month&quot;</span>, <span class="at">ylab =</span> <span class="st">&quot;number of letters&quot;</span>,</span>
<span id="cb186-6"><a href="statistical-distributions.html#cb186-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">pch =</span> <span class="dv">19</span>, <span class="at">col =</span> <span class="st">&quot;royalblue&quot;</span></span>
<span id="cb186-7"><a href="statistical-distributions.html#cb186-7" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb186-8"><a href="statistical-distributions.html#cb186-8" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="at">a =</span> <span class="dv">20</span>, <span class="at">b =</span> <span class="dv">0</span>, <span class="at">lwd =</span> <span class="dv">2</span>, <span class="at">lty =</span> <span class="dv">3</span>, <span class="at">col =</span> <span class="st">&quot;salmon&quot;</span>)</span></code></pre></div>
<p><img src="main_files/unnamed-chunk-49-1.png" width="672" /></p>
<p>Now, let’s plot the density plot of our data. The <span class="math inline">\(x\)</span>-axis is the number of
letters on a single day, and the <span class="math inline">\(y\)</span>-axis is the probability.</p>
<div class="sourceCode" id="cb187"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb187-1"><a href="statistical-distributions.html#cb187-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(<span class="fu">density</span>(my_letter),</span>
<span id="cb187-2"><a href="statistical-distributions.html#cb187-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">lwd =</span> <span class="dv">2</span>, <span class="at">col =</span> <span class="st">&quot;royalblue&quot;</span>,</span>
<span id="cb187-3"><a href="statistical-distributions.html#cb187-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">main =</span> <span class="st">&quot;Probability of number of letters each day&quot;</span>,</span>
<span id="cb187-4"><a href="statistical-distributions.html#cb187-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">xlab =</span> <span class="st">&quot;number of letters&quot;</span></span>
<span id="cb187-5"><a href="statistical-distributions.html#cb187-5" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div>
<p><img src="main_files/unnamed-chunk-50-1.png" width="672" /></p>
<p>Since we only have 30 data points, it doesn’t look like a good curve. But,
after we worked at the mail reception for 5000 days, it becomes much closer
to the theoretical Poisson distribution with lambda = 20.</p>
<div class="sourceCode" id="cb188"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb188-1"><a href="statistical-distributions.html#cb188-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">3</span>)</span>
<span id="cb188-2"><a href="statistical-distributions.html#cb188-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(<span class="fu">density</span>(<span class="fu">rpois</span>(<span class="at">n =</span> <span class="dv">5000</span>, <span class="at">lambda =</span> <span class="dv">20</span>)),</span>
<span id="cb188-3"><a href="statistical-distributions.html#cb188-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">lwd =</span> <span class="dv">2</span>, <span class="at">col =</span> <span class="st">&quot;royalblue&quot;</span>,</span>
<span id="cb188-4"><a href="statistical-distributions.html#cb188-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">main =</span> <span class="st">&quot;Probability of number of letters each day&quot;</span>,</span>
<span id="cb188-5"><a href="statistical-distributions.html#cb188-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">xlab =</span> <span class="st">&quot;number of letters&quot;</span></span>
<span id="cb188-6"><a href="statistical-distributions.html#cb188-6" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div>
<p><img src="main_files/unnamed-chunk-51-1.png" width="672" /></p>
<p>Here is the theoretical Poisson distribution with lambda = 20:</p>
<div class="sourceCode" id="cb189"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb189-1"><a href="statistical-distributions.html#cb189-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(<span class="fu">dpois</span>(<span class="fu">c</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">40</span>), <span class="at">lambda =</span> <span class="dv">20</span>),</span>
<span id="cb189-2"><a href="statistical-distributions.html#cb189-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">lwd =</span> <span class="dv">2</span>, <span class="at">type =</span> <span class="st">&quot;l&quot;</span>, <span class="at">col =</span> <span class="st">&quot;royalblue&quot;</span>,</span>
<span id="cb189-3"><a href="statistical-distributions.html#cb189-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">ylab =</span> <span class="st">&quot;probability&quot;</span>, <span class="at">main =</span> <span class="st">&quot;Poisson lambda=20&quot;</span></span>
<span id="cb189-4"><a href="statistical-distributions.html#cb189-4" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div>
<p><img src="main_files/unnamed-chunk-52-1.png" width="672" /></p>
<p>If we want to know what’s the probability to receive, for example, 18 letters,<br />
we can use <code>dpois()</code> function.</p>
<div class="sourceCode" id="cb190"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb190-1"><a href="statistical-distributions.html#cb190-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dpois</span>(<span class="at">x =</span> <span class="dv">18</span>, <span class="at">lambda =</span> <span class="dv">20</span>)</span></code></pre></div>
<pre><code>## [1] 0.08439355</code></pre>
<p>If we want to know the probability of receiving 18 or less letters,
use <code>ppois()</code> function.</p>
<div class="sourceCode" id="cb192"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb192-1"><a href="statistical-distributions.html#cb192-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ppois</span>(<span class="at">q =</span> <span class="dv">18</span>, <span class="at">lambda =</span> <span class="dv">20</span>, <span class="at">lower.tail =</span> T)</span></code></pre></div>
<pre><code>## [1] 0.3814219</code></pre>
<p>It is the cumulative area colored in the following plot:</p>
<div class="sourceCode" id="cb194"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb194-1"><a href="statistical-distributions.html#cb194-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(<span class="fu">dpois</span>(<span class="fu">c</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">40</span>), <span class="at">lambda =</span> <span class="dv">20</span>),</span>
<span id="cb194-2"><a href="statistical-distributions.html#cb194-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">lwd =</span> <span class="dv">2</span>, <span class="at">type =</span> <span class="st">&quot;l&quot;</span>, <span class="at">col =</span> <span class="st">&quot;royalblue&quot;</span>,</span>
<span id="cb194-3"><a href="statistical-distributions.html#cb194-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">ylab =</span> <span class="st">&quot;probability&quot;</span>, <span class="at">main =</span> <span class="st">&quot;Poisson lambda=20&quot;</span></span>
<span id="cb194-4"><a href="statistical-distributions.html#cb194-4" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb194-5"><a href="statistical-distributions.html#cb194-5" aria-hidden="true" tabindex="-1"></a><span class="fu">polygon</span>(</span>
<span id="cb194-6"><a href="statistical-distributions.html#cb194-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">x =</span> <span class="fu">c</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">18</span>, <span class="dv">18</span>),</span>
<span id="cb194-7"><a href="statistical-distributions.html#cb194-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">y =</span> <span class="fu">c</span>(<span class="fu">dpois</span>(<span class="fu">c</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">18</span>), <span class="at">lambda =</span> <span class="dv">20</span>), <span class="dv">0</span>),</span>
<span id="cb194-8"><a href="statistical-distributions.html#cb194-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">border =</span> <span class="st">&quot;royalblue&quot;</span>, <span class="at">col =</span> <span class="st">&quot;lightblue1&quot;</span></span>
<span id="cb194-9"><a href="statistical-distributions.html#cb194-9" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb194-10"><a href="statistical-distributions.html#cb194-10" aria-hidden="true" tabindex="-1"></a><span class="fu">segments</span>(<span class="at">x0 =</span> <span class="dv">18</span>, <span class="at">y0 =</span> <span class="dv">0</span>, <span class="at">y1 =</span> <span class="fl">0.08439355</span>, <span class="at">lwd =</span> <span class="dv">2</span>, <span class="at">lty =</span> <span class="dv">2</span>, <span class="at">col =</span> <span class="st">&quot;salmon&quot;</span>)</span></code></pre></div>
<p><img src="main_files/unnamed-chunk-55-1.png" width="672" /></p>
<p><code>qpois()</code> is like the “reverse” of <code>ppois()</code>.</p>
<div class="sourceCode" id="cb195"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb195-1"><a href="statistical-distributions.html#cb195-1" aria-hidden="true" tabindex="-1"></a><span class="fu">qpois</span>(<span class="at">p =</span> <span class="fl">0.3814219</span>, <span class="at">lambda =</span> <span class="dv">20</span>)</span></code></pre></div>
<pre><code>## [1] 18</code></pre>
<p>Let’s review the definition of Poisson distribution.</p>
<ul>
<li>“these events occur in a fixed interval of time or space,” which is a day;<br />
</li>
<li>“these events occur with a known constant mean rate,” which is 20 letters;<br />
</li>
<li>“independently of the time since the last event,” which means the number of
letters you receive today is independent of the letter you receive tomorrow.
<del>more work today don’t guarantee less work tomorrow, just like in real life</del></li>
</ul>
<p>Now let’s re-visit the formula.</p>
<p><span class="math display">\[
\begin{equation}
f(k; \lambda) = Pr(X=k) = \frac {\lambda^k e^{-\lambda}} {k!}
\end{equation}
\]</span></p>
<ul>
<li>lambda is the expected value of <span class="math inline">\(X\)</span>, which is 20 letters in this example.<br />
</li>
<li><span class="math inline">\(k\)</span> is the number of occurrences, which is the number of letters you get on a
specific day.</li>
<li>e is Euler’s number (e=2.71828)</li>
</ul>
</div>
</div>
<div id="continuous-distributions" class="section level3" number="6.2.5">
<h3><span class="header-section-number">6.2.5</span> Continuous Distributions</h3>
<p>In contrast with discrete distributions, continuous distributions are defined
over infinite, possibly bounded domains, e.g. all real numbers. There are a
number of different continuous distributions. Here we will focus on the <strong>Normal
distribution (Gaussian distribution)</strong>. A lot of variables in real life are
normally distributed, common examples include people’s height, blood pressure,
and students’ exam score.</p>
<p>This is what a normal distribution with mean = 0 and standard deviation = 1
looks like:</p>
<div class="sourceCode" id="cb197"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb197-1"><a href="statistical-distributions.html#cb197-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">2</span>)</span>
<span id="cb197-2"><a href="statistical-distributions.html#cb197-2" aria-hidden="true" tabindex="-1"></a>norm <span class="ot">&lt;-</span> <span class="fu">rnorm</span>(<span class="at">n =</span> <span class="dv">50000</span>, <span class="at">mean =</span> <span class="dv">0</span>, <span class="at">sd =</span> <span class="dv">1</span>)</span>
<span id="cb197-3"><a href="statistical-distributions.html#cb197-3" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(<span class="fu">density</span>(norm),</span>
<span id="cb197-4"><a href="statistical-distributions.html#cb197-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">main =</span> <span class="st">&quot;A Normal distribution&quot;</span>, <span class="at">xlab =</span> <span class="st">&quot;x&quot;</span>,</span>
<span id="cb197-5"><a href="statistical-distributions.html#cb197-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">lwd =</span> <span class="dv">2</span>, <span class="at">col =</span> <span class="st">&quot;royalblue&quot;</span></span>
<span id="cb197-6"><a href="statistical-distributions.html#cb197-6" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div>
<p><img src="main_files/unnamed-chunk-57-1.png" width="672" /></p>
<p>Similar as the Poisson distribution above, there are several functions to work
with normal distribution, including <code>rnorm()</code>, <code>dnorm()</code>, <code>pnorm()</code>, and
<code>qnorm()</code>.</p>
<p><code>rnorm()</code> is used to draw random data points from a normal distribution with
a given mean and standard deviation.</p>
<div class="sourceCode" id="cb198"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb198-1"><a href="statistical-distributions.html#cb198-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">2</span>)</span>
<span id="cb198-2"><a href="statistical-distributions.html#cb198-2" aria-hidden="true" tabindex="-1"></a><span class="fu">rnorm</span>(</span>
<span id="cb198-3"><a href="statistical-distributions.html#cb198-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">n =</span> <span class="dv">6</span>, <span class="co"># number of data points to draw</span></span>
<span id="cb198-4"><a href="statistical-distributions.html#cb198-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">mean =</span> <span class="dv">0</span>, <span class="co"># mean</span></span>
<span id="cb198-5"><a href="statistical-distributions.html#cb198-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">sd =</span> <span class="dv">1</span> <span class="co"># standard deviation</span></span>
<span id="cb198-6"><a href="statistical-distributions.html#cb198-6" aria-hidden="true" tabindex="-1"></a>) </span></code></pre></div>
<pre><code>## [1] -0.89691455  0.18484918  1.58784533 -1.13037567 -0.08025176  0.13242028</code></pre>
<p><code>dnorm()</code> is the density at a given quantile. For instance, in the normal
distribution (mean=0, standard deviation=1) above, the probability density at
0.5 is roughly 0.35.</p>
<div class="sourceCode" id="cb200"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb200-1"><a href="statistical-distributions.html#cb200-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dnorm</span>(<span class="at">x =</span> <span class="fl">0.5</span>, <span class="at">mean =</span> <span class="dv">0</span>, <span class="at">sd =</span> <span class="dv">1</span>)</span></code></pre></div>
<pre><code>## [1] 0.3520653</code></pre>
<div class="sourceCode" id="cb202"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb202-1"><a href="statistical-distributions.html#cb202-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(<span class="fu">density</span>(norm),</span>
<span id="cb202-2"><a href="statistical-distributions.html#cb202-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">main =</span> <span class="st">&quot;A Normal distribution&quot;</span>, <span class="at">xlab =</span> <span class="st">&quot;x&quot;</span>,</span>
<span id="cb202-3"><a href="statistical-distributions.html#cb202-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">lwd =</span> <span class="dv">2</span>, <span class="at">col =</span> <span class="st">&quot;royalblue&quot;</span></span>
<span id="cb202-4"><a href="statistical-distributions.html#cb202-4" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb202-5"><a href="statistical-distributions.html#cb202-5" aria-hidden="true" tabindex="-1"></a><span class="fu">segments</span>(<span class="at">x0 =</span> <span class="fl">0.5</span>, <span class="at">y0 =</span> <span class="dv">0</span>, <span class="at">y1 =</span> <span class="fl">0.3520653</span>, <span class="at">lwd =</span> <span class="dv">2</span>, <span class="at">lty =</span> <span class="dv">3</span>, <span class="at">col =</span> <span class="st">&quot;salmon&quot;</span>)</span>
<span id="cb202-6"><a href="statistical-distributions.html#cb202-6" aria-hidden="true" tabindex="-1"></a><span class="fu">segments</span>(<span class="at">x0 =</span> <span class="sc">-</span><span class="dv">5</span>, <span class="at">y0 =</span> <span class="fl">0.3520653</span>, <span class="at">x1 =</span> <span class="fl">0.5</span>, <span class="at">lwd =</span> <span class="dv">2</span>, <span class="at">lty =</span> <span class="dv">3</span>, <span class="at">col =</span> <span class="st">&quot;salmon&quot;</span>)</span>
<span id="cb202-7"><a href="statistical-distributions.html#cb202-7" aria-hidden="true" tabindex="-1"></a><span class="fu">points</span>(<span class="at">x =</span> <span class="fl">0.5</span>, <span class="at">y =</span> <span class="fl">0.3520653</span>, <span class="at">cex =</span> <span class="fl">1.5</span>, <span class="at">lwd =</span> <span class="dv">2</span>, <span class="at">col =</span> <span class="st">&quot;red&quot;</span>)</span>
<span id="cb202-8"><a href="statistical-distributions.html#cb202-8" aria-hidden="true" tabindex="-1"></a><span class="fu">text</span>(<span class="at">x =</span> <span class="fl">1.1</span>, <span class="at">y =</span> <span class="fl">0.36</span>, <span class="at">labels =</span> <span class="st">&quot;0.3521&quot;</span>, <span class="at">col =</span> <span class="st">&quot;red2&quot;</span>)</span></code></pre></div>
<p><img src="main_files/unnamed-chunk-60-1.png" width="672" /></p>
<p><code>pnorm()</code> gives the distribution function. Or, you can think of it as the
cumulative of the left side of the density function until a given value,
which is the area colored in light blue in the following plot.</p>
<div class="sourceCode" id="cb203"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb203-1"><a href="statistical-distributions.html#cb203-1" aria-hidden="true" tabindex="-1"></a><span class="fu">pnorm</span>(<span class="at">q =</span> <span class="fl">0.5</span>, <span class="at">mean =</span> <span class="dv">0</span>, <span class="at">sd =</span> <span class="dv">1</span>)</span></code></pre></div>
<pre><code>## [1] 0.6914625</code></pre>
<div class="sourceCode" id="cb205"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb205-1"><a href="statistical-distributions.html#cb205-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(<span class="fu">density</span>(norm),</span>
<span id="cb205-2"><a href="statistical-distributions.html#cb205-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">main =</span> <span class="st">&quot;A Normal distribution&quot;</span>, <span class="at">xlab =</span> <span class="st">&quot;x&quot;</span>,</span>
<span id="cb205-3"><a href="statistical-distributions.html#cb205-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">lwd =</span> <span class="dv">2</span>, <span class="at">col =</span> <span class="st">&quot;royalblue&quot;</span></span>
<span id="cb205-4"><a href="statistical-distributions.html#cb205-4" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb205-5"><a href="statistical-distributions.html#cb205-5" aria-hidden="true" tabindex="-1"></a><span class="fu">polygon</span>(</span>
<span id="cb205-6"><a href="statistical-distributions.html#cb205-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">x =</span> <span class="fu">c</span>(<span class="fu">density</span>(norm)<span class="sc">$</span>x[<span class="fu">density</span>(norm)<span class="sc">$</span>x <span class="sc">&lt;=</span> <span class="fl">0.5</span>], <span class="fl">0.5</span>),</span>
<span id="cb205-7"><a href="statistical-distributions.html#cb205-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">y =</span> <span class="fu">c</span>(<span class="fu">density</span>(norm)<span class="sc">$</span>y[<span class="fu">density</span>(norm)<span class="sc">$</span>x <span class="sc">&lt;=</span> <span class="fl">0.5</span>], <span class="dv">0</span>),</span>
<span id="cb205-8"><a href="statistical-distributions.html#cb205-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">border =</span> <span class="st">&quot;royalblue&quot;</span>, <span class="at">col =</span> <span class="st">&quot;lightblue1&quot;</span></span>
<span id="cb205-9"><a href="statistical-distributions.html#cb205-9" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb205-10"><a href="statistical-distributions.html#cb205-10" aria-hidden="true" tabindex="-1"></a><span class="fu">segments</span>(<span class="at">x0 =</span> <span class="fl">0.5</span>, <span class="at">y0 =</span> <span class="dv">0</span>, <span class="at">y1 =</span> <span class="fl">0.3520653</span>, <span class="at">lwd =</span> <span class="dv">2</span>, <span class="at">lty =</span> <span class="dv">2</span>, <span class="at">col =</span> <span class="st">&quot;salmon&quot;</span>)</span></code></pre></div>
<p><img src="main_files/unnamed-chunk-62-1.png" width="672" /></p>
<p><code>qnorm()</code> gives the quantile function. You can think of it as the “reverse” of
<code>pnorm()</code>. For instance:</p>
<div class="sourceCode" id="cb206"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb206-1"><a href="statistical-distributions.html#cb206-1" aria-hidden="true" tabindex="-1"></a><span class="fu">pnorm</span>(<span class="at">q =</span> <span class="fl">0.5</span>, <span class="at">mean =</span> <span class="dv">0</span>, <span class="at">sd =</span> <span class="dv">1</span>)</span></code></pre></div>
<pre><code>## [1] 0.6914625</code></pre>
<div class="sourceCode" id="cb208"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb208-1"><a href="statistical-distributions.html#cb208-1" aria-hidden="true" tabindex="-1"></a><span class="fu">qnorm</span>(<span class="at">p =</span> <span class="fl">0.6914625</span>, <span class="at">mean =</span> <span class="dv">0</span>, <span class="at">sd =</span> <span class="dv">1</span>)</span></code></pre></div>
<pre><code>## [1] 0.5000001</code></pre>
<p>Similarly, other distributions such as chi-square distribution, they all have
the same set of functions: <code>rchisq()</code>, <code>dchisq()</code>, <code>pchisq()</code> and <code>qchisq()</code> etc.</p>
</div>
<div id="empirical-distributions" class="section level3" number="6.2.6">
<h3><span class="header-section-number">6.2.6</span> Empirical Distributions</h3>
<p>Empirical distributions describe the relative frequency of observed values in a
dataset. Empirical distributions may have any shape, and may be visualized using
any of the methods described in the <a href="visualizing-distributions.html#visualizing-distributions">Visualizing Distributions</a> section, e.g. a
<a href="#density-plots">density plot</a>. The following beeswarm and density plot both
visualize the same made-up dataset:</p>
<p><img src="main_files/stat%20hist-1.png" width="672" /></p>
<p>Note the <span class="math inline">\(y\)</span> axis in the plot on the right is scaled as a density, rather than a
count. A distribution plotted as a density ensures the values sum to 1, thus
making a probability distribution. This empirical distribution looks quite
complicated, and cannot be easily captured with a single mean and confidence
interval.</p>
</div>
</div>
<p style="text-align: center;">
<a href="data-modeling.html"><button class="btn btn-default">Previous</button></a>
<a href="statistical-tests.html"><button class="btn btn-default">Next</button></a>
</p>
</div>
</div>


</div>

<script>

// add bootstrap table styles to pandoc tables
$(document).ready(function () {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
});

</script>

</body>
</html>
