<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta name="viewport" content="width=device-width, initial-scale=1" />
<meta property="og:title" content="9.2 Toolification | BF591 - R for Biological Sciences" />
<meta property="og:type" content="book" />







<script type="text/x-mathjax-config">
MathJax.Hub.Config({
  TeX: { equationNumbers: { autoNumber: "AMS" } }
});
</script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<meta name="description" content="9.2 Toolification | BF591 - R for Biological Sciences">

<title>9.2 Toolification | BF591 - R for Biological Sciences</title>

<script src="libs/header-attrs-2.13/header-attrs.js"></script>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="libs/bootstrap-3.3.5/css/darkly.min.css" rel="stylesheet" />
<script src="libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="libs/navigation-1.1/tabsets.js"></script>


<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>


<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />

<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
/* show arrow before summary tag as in bootstrap
TODO: remove if boostrap in updated in html_document (rmarkdown#1485) */
details > summary {
  display: list-item;
  cursor: pointer;
}
</style>
</head>

<body>

<div class="container-fluid main-container">


<div class="row">
<div class="col-sm-12">
<div id="TOC">
<ul>
<li><a href="index.html#syllabus">Syllabus</a>
<ul>
<li><a href="course-schedule.html#course-schedule">Course Schedule</a></li>
<li><a href="instructors.html#instructors">Instructors</a></li>
<li><a href="office-hours.html#office-hours">Office Hours</a></li>
<li><a href="course-values-and-policies.html#course-values-and-policies">Course Values and Policies</a></li>
</ul></li>
<li><a href="introduction.html#introduction"><span class="toc-section-number">1</span> Introduction</a>
<ul>
<li><a href="who-this-book-is-for.html#who-this-book-is-for"><span class="toc-section-number">1.1</span> Who This Book Is For</a></li>
<li><a href="a-note-about-reinventing-the-wheel.html#a-note-about-reinventing-the-wheel"><span class="toc-section-number">1.2</span> A Note About Reinventing the Wheel</a></li>
<li><a href="sources-and-references.html#sources-and-references"><span class="toc-section-number">1.3</span> Sources and References</a></li>
</ul></li>
<li><a href="data-bio.html#data-bio"><span class="toc-section-number">2</span> Data in Biology</a>
<ul>
<li><a href="a-brief-history-of-data-in-molecular-biology.html#a-brief-history-of-data-in-molecular-biology"><span class="toc-section-number">2.1</span> A Brief History of Data in Molecular Biology</a></li>
<li><a href="biology-as-a-mature-data-science.html#biology-as-a-mature-data-science"><span class="toc-section-number">2.2</span> Biology as a Mature Data Science</a></li>
</ul></li>
<li><a href="preliminaries.html#preliminaries"><span class="toc-section-number">3</span> Preliminaries</a>
<ul>
<li><a href="prelim-r.html#prelim-r"><span class="toc-section-number">3.1</span> The R Language</a></li>
<li><a href="prelim-rstudio.html#prelim-rstudio"><span class="toc-section-number">3.2</span> RStudio</a></li>
<li><a href="prog-the-r-script.html#prog-the-r-script"><span class="toc-section-number">3.3</span> The R Script</a></li>
<li><a href="prog-workflow.html#prog-workflow"><span class="toc-section-number">3.4</span> The Scripting Workflow</a></li>
<li><a href="prelim-git.html#prelim-git"><span class="toc-section-number">3.5</span> git + github</a>
<ul>
<li><a href="prelim-git.html#motivation"><span class="toc-section-number">3.5.1</span> Motivation</a></li>
<li><a href="prelim-git.html#git"><span class="toc-section-number">3.5.2</span> git</a></li>
<li><a href="prelim-git.html#git-hosting-platforms-github"><span class="toc-section-number">3.5.3</span> Git hosting platforms (GitHub)</a></li>
</ul></li>
</ul></li>
<li><a href="prog-basics.html#prog-basics"><span class="toc-section-number">4</span> R Programming</a>
<ul>
<li><a href="before-you-begin.html#before-you-begin"><span class="toc-section-number">4.1</span> Before you begin</a></li>
<li><a href="introduction-1.html#introduction-1"><span class="toc-section-number">4.2</span> Introduction</a></li>
<li><a href="prog-r-syntax.html#prog-r-syntax"><span class="toc-section-number">4.3</span> R Syntax Basics</a></li>
<li><a href="prog-types.html#prog-types"><span class="toc-section-number">4.4</span> Basic Types of Values</a>
<ul>
<li><a href="prog-types.html#factors"><span class="toc-section-number">4.4.1</span> Factors</a></li>
</ul></li>
<li><a href="prog-struct.html#prog-struct"><span class="toc-section-number">4.5</span> Data Structures</a>
<ul>
<li><a href="prog-struct.html#vectors"><span class="toc-section-number">4.5.1</span> Vectors</a></li>
<li><a href="prog-struct.html#matrices"><span class="toc-section-number">4.5.2</span> Matrices</a></li>
<li><a href="prog-struct.html#lists-and-data-frames"><span class="toc-section-number">4.5.3</span> Lists and data frames</a></li>
</ul></li>
<li><a href="logical-tests-and-comparators.html#logical-tests-and-comparators"><span class="toc-section-number">4.6</span> Logical Tests and Comparators</a></li>
<li><a href="functions.html#functions"><span class="toc-section-number">4.7</span> Functions</a>
<ul>
<li><a href="functions.html#prog-dry"><span class="toc-section-number">4.7.1</span> DRY: Don’t Repeat Yourself</a></li>
<li><a href="functions.html#writing-your-own-functions"><span class="toc-section-number">4.7.2</span> Writing your own functions</a></li>
<li><a href="functions.html#scope"><span class="toc-section-number">4.7.3</span> Scope</a></li>
</ul></li>
<li><a href="iteration.html#iteration"><span class="toc-section-number">4.8</span> Iteration</a>
<ul>
<li><a href="iteration.html#vectorized-operations"><span class="toc-section-number">4.8.1</span> Vectorized operations</a></li>
<li><a href="iteration.html#functional-programming"><span class="toc-section-number">4.8.2</span> Functional programming</a></li>
<li><a href="iteration.html#apply-and-friends"><span class="toc-section-number">4.8.3</span> <code>apply()</code> and friends</a></li>
</ul></li>
<li><a href="installing-packages.html#installing-packages"><span class="toc-section-number">4.9</span> Installing Packages</a></li>
<li><a href="saving-and-loading-r-data.html#saving-and-loading-r-data"><span class="toc-section-number">4.10</span> Saving and Loading R Data</a></li>
<li><a href="prog-debugging.html#prog-debugging"><span class="toc-section-number">4.11</span> Troubleshooting and Debugging</a></li>
<li><a href="prog-style.html#prog-style"><span class="toc-section-number">4.12</span> Coding Style and Conventions</a>
<ul>
<li><a href="prog-style.html#is-my-code-correct"><span class="toc-section-number">4.12.1</span> Is my code correct?</a></li>
<li><a href="prog-style.html#does-my-code-follow-the-dry-principle"><span class="toc-section-number">4.12.2</span> Does my code follow the <span>DRY principle</span>?</a></li>
<li><a href="prog-style.html#did-i-choose-concise-but-descriptive-variable-and-function-names"><span class="toc-section-number">4.12.3</span> Did I choose concise but descriptive variable and function names?</a></li>
<li><a href="prog-style.html#did-i-use-indentation-and-naming-conventions-consistently-throughout-my-code"><span class="toc-section-number">4.12.4</span> Did I use indentation and naming conventions consistently throughout my code?</a></li>
<li><a href="prog-style.html#did-i-write-comments-especially-when-what-the-code-does-is-not-obvious"><span class="toc-section-number">4.12.5</span> Did I write comments, especially when what the code does is not obvious?</a></li>
<li><a href="prog-style.html#how-easy-would-it-be-for-someone-else-to-understand-my-code"><span class="toc-section-number">4.12.6</span> How easy would it be for someone else to understand my code?</a></li>
<li><a href="prog-style.html#is-my-code-easy-to-maintainchange"><span class="toc-section-number">4.12.7</span> Is my code easy to maintain/change?</a></li>
<li><a href="prog-style.html#the-styler-package"><span class="toc-section-number">4.12.8</span> The <code>styler</code> package</a></li>
</ul></li>
</ul></li>
<li><a href="data-wrangling.html#data-wrangling"><span class="toc-section-number">5</span> Data Wrangling</a>
<ul>
<li><a href="the-tidyverse.html#the-tidyverse"><span class="toc-section-number">5.1</span> The Tidyverse</a></li>
<li><a href="dw-basics.html#dw-basics"><span class="toc-section-number">5.2</span> Tidyverse Basics</a></li>
<li><a href="importing-data.html#importing-data"><span class="toc-section-number">5.3</span> Importing Data</a></li>
<li><a href="data-tibble.html#data-tibble"><span class="toc-section-number">5.4</span> The tibble</a></li>
<li><a href="tidy-data.html#tidy-data"><span class="toc-section-number">5.5</span> Tidy Data</a></li>
<li><a href="pipes.html#pipes"><span class="toc-section-number">5.6</span> pipes</a></li>
<li><a href="arranging-data.html#arranging-data"><span class="toc-section-number">5.7</span> Arranging Data</a>
<ul>
<li><a href="arranging-data.html#dplyrmutate---create-new-columns-using-other-columns"><span class="toc-section-number">5.7.1</span> <code>dplyr::mutate()</code> - Create new columns using other columns</a></li>
<li><a href="arranging-data.html#stringr---working-with-character-values"><span class="toc-section-number">5.7.2</span> <code>stringr</code> - Working with character values</a></li>
<li><a href="arranging-data.html#dplyrselect---subset-columns-by-name"><span class="toc-section-number">5.7.3</span> <code>dplyr::select()</code> - Subset Columns by Name</a></li>
<li><a href="arranging-data.html#dplyrfilter---pick-rows-out-of-a-data-set"><span class="toc-section-number">5.7.4</span> <code>dplyr::filter()</code> - Pick rows out of a data set</a></li>
<li><a href="arranging-data.html#dplyrarrange---order-rows-based-on-their-values"><span class="toc-section-number">5.7.5</span> <code>dplyr::arrange()</code> - Order rows based on their values</a></li>
<li><a href="arranging-data.html#putting-it-all-together"><span class="toc-section-number">5.7.6</span> Putting it all together</a></li>
</ul></li>
<li><a href="grouping-data.html#grouping-data"><span class="toc-section-number">5.8</span> Grouping Data</a></li>
<li><a href="rearranging-data.html#rearranging-data"><span class="toc-section-number">5.9</span> Rearranging Data</a></li>
<li><a href="relational-data.html#relational-data"><span class="toc-section-number">5.10</span> Relational Data</a>
<ul>
<li><a href="relational-data.html#dealing-with-multiple-matches"><span class="toc-section-number">5.10.1</span> Dealing with multiple matches</a></li>
</ul></li>
</ul></li>
<li><a href="data-science.html#data-science"><span class="toc-section-number">6</span> Data Science</a>
<ul>
<li><a href="data-modeling.html#data-modeling"><span class="toc-section-number">6.1</span> Data Modeling</a>
<ul>
<li><a href="data-modeling.html#a-worked-modeling-example"><span class="toc-section-number">6.1.1</span> A Worked Modeling Example</a></li>
<li><a href="data-modeling.html#data-summarization"><span class="toc-section-number">6.1.2</span> Data Summarization</a></li>
<li><a href="data-modeling.html#linear-models"><span class="toc-section-number">6.1.3</span> Linear Models</a></li>
<li><a href="data-modeling.html#flavors-of-linear-models"><span class="toc-section-number">6.1.4</span> Flavors of Linear Models</a></li>
<li><a href="data-modeling.html#assessing-model-accuracy-."><span class="toc-section-number">6.1.5</span> Assessing Model Accuracy .</a></li>
</ul></li>
<li><a href="statistical-distributions.html#statistical-distributions"><span class="toc-section-number">6.2</span> Statistical Distributions</a>
<ul>
<li><a href="statistical-distributions.html#random-variables"><span class="toc-section-number">6.2.1</span> Random Variables</a></li>
<li><a href="statistical-distributions.html#statistical-distribution-basics"><span class="toc-section-number">6.2.2</span> Statistical Distribution Basics</a></li>
<li><a href="statistical-distributions.html#distributions-in-r"><span class="toc-section-number">6.2.3</span> Distributions in R</a></li>
<li><a href="statistical-distributions.html#discrete-distributions"><span class="toc-section-number">6.2.4</span> Discrete Distributions</a></li>
<li><a href="statistical-distributions.html#continuous-distributions"><span class="toc-section-number">6.2.5</span> Continuous Distributions</a></li>
<li><a href="statistical-distributions.html#empirical-distributions"><span class="toc-section-number">6.2.6</span> Empirical Distributions</a></li>
</ul></li>
<li><a href="statistical-tests.html#statistical-tests"><span class="toc-section-number">6.3</span> Statistical Tests</a>
<ul>
<li><a href="statistical-tests.html#what-is-a-statistical-test"><span class="toc-section-number">6.3.1</span> What is a statistical test</a></li>
<li><a href="statistical-tests.html#common-statistical-tests"><span class="toc-section-number">6.3.2</span> Common Statistical Tests</a></li>
<li><a href="statistical-tests.html#choosing-a-test"><span class="toc-section-number">6.3.3</span> Choosing a Test</a></li>
<li><a href="statistical-tests.html#p-values"><span class="toc-section-number">6.3.4</span> P-values</a></li>
<li><a href="statistical-tests.html#multiple-hypothesis-testing-."><span class="toc-section-number">6.3.5</span> Multiple Hypothesis Testing .</a></li>
<li><a href="statistical-tests.html#statistical-power"><span class="toc-section-number">6.3.6</span> Statistical power</a></li>
</ul></li>
<li><a href="exploratory-data-analysis.html#exploratory-data-analysis"><span class="toc-section-number">6.4</span> Exploratory Data Analysis</a>
<ul>
<li><a href="exploratory-data-analysis.html#principal-component-analysis"><span class="toc-section-number">6.4.1</span> Principal Component Analysis</a></li>
<li><a href="exploratory-data-analysis.html#cluster-analysis"><span class="toc-section-number">6.4.2</span> Cluster Analysis</a></li>
<li><a href="exploratory-data-analysis.html#k-means-."><span class="toc-section-number">6.4.3</span> k-means .</a></li>
<li><a href="exploratory-data-analysis.html#others"><span class="toc-section-number">6.4.4</span> Others</a></li>
</ul></li>
<li><a href="network-analysis-..html#network-analysis-."><span class="toc-section-number">6.5</span> Network Analysis .</a>
<ul>
<li><a href="network-analysis-..html#basic-network-analysis"><span class="toc-section-number">6.5.1</span> Basic network analysis</a></li>
<li><a href="network-analysis-..html#create-your-own"><span class="toc-section-number">6.5.2</span> Create your own</a></li>
</ul></li>
</ul></li>
<li><a href="data-visualization.html#data-visualization"><span class="toc-section-number">7</span> Data Visualization</a>
<ul>
<li><a href="responsible-plotting.html#responsible-plotting"><span class="toc-section-number">7.1</span> Responsible Plotting</a>
<ul>
<li><a href="responsible-plotting.html#visualization-principles"><span class="toc-section-number">7.1.1</span> Visualization Principles</a></li>
<li><a href="responsible-plotting.html#human-visual-perception"><span class="toc-section-number">7.1.2</span> Human Visual Perception</a></li>
<li><a href="responsible-plotting.html#visual-encodings"><span class="toc-section-number">7.1.3</span> Visual Encodings</a></li>
<li><a href="responsible-plotting.html#some-opinionated-rules-of-thumb"><span class="toc-section-number">7.1.4</span> Some Opinionated Rules of Thumb</a></li>
</ul></li>
<li><a href="dv-gg.html#dv-gg"><span class="toc-section-number">7.2</span> Grammar of Graphics</a>
<ul>
<li><a href="dv-gg.html#ggplot-mechanics"><span class="toc-section-number">7.2.1</span> <code>ggplot</code> mechanics</a></li>
</ul></li>
<li><a href="plotting-one-dimension.html#plotting-one-dimension"><span class="toc-section-number">7.3</span> Plotting One Dimension</a>
<ul>
<li><a href="plotting-one-dimension.html#dv-bar"><span class="toc-section-number">7.3.1</span> Bar chart</a></li>
<li><a href="plotting-one-dimension.html#lollipop-plots"><span class="toc-section-number">7.3.2</span> Lollipop plots</a></li>
<li><a href="plotting-one-dimension.html#stacked-area-charts"><span class="toc-section-number">7.3.3</span> Stacked Area charts</a></li>
<li><a href="plotting-one-dimension.html#parallel-coordinate-plots"><span class="toc-section-number">7.3.4</span> Parallel Coordinate plots</a></li>
</ul></li>
<li><a href="visualizing-distributions.html#visualizing-distributions"><span class="toc-section-number">7.4</span> Visualizing Distributions</a>
<ul>
<li><a href="visualizing-distributions.html#histogram"><span class="toc-section-number">7.4.1</span> Histogram</a></li>
<li><a href="visualizing-distributions.html#density"><span class="toc-section-number">7.4.2</span> Density</a></li>
<li><a href="visualizing-distributions.html#boxplot"><span class="toc-section-number">7.4.3</span> Boxplot</a></li>
<li><a href="visualizing-distributions.html#violin-plot"><span class="toc-section-number">7.4.4</span> Violin plot</a></li>
<li><a href="visualizing-distributions.html#beeswarm-plot"><span class="toc-section-number">7.4.5</span> Beeswarm plot</a></li>
<li><a href="visualizing-distributions.html#ridgeline"><span class="toc-section-number">7.4.6</span> Ridgeline</a></li>
</ul></li>
<li><a href="plotting-two-or-more-dimensions.html#plotting-two-or-more-dimensions"><span class="toc-section-number">7.5</span> Plotting Two or More Dimensions</a>
<ul>
<li><a href="plotting-two-or-more-dimensions.html#scatter-plots"><span class="toc-section-number">7.5.1</span> Scatter Plots</a></li>
<li><a href="plotting-two-or-more-dimensions.html#bubble-plots"><span class="toc-section-number">7.5.2</span> Bubble Plots</a></li>
<li><a href="plotting-two-or-more-dimensions.html#connected-scatter-plots"><span class="toc-section-number">7.5.3</span> Connected Scatter Plots</a></li>
<li><a href="plotting-two-or-more-dimensions.html#line-plots"><span class="toc-section-number">7.5.4</span> Line Plots</a></li>
<li><a href="plotting-two-or-more-dimensions.html#parallel-coordinate-plots-1"><span class="toc-section-number">7.5.5</span> Parallel Coordinate Plots</a></li>
<li><a href="plotting-two-or-more-dimensions.html#heatmaps"><span class="toc-section-number">7.5.6</span> Heatmaps</a></li>
<li><a href="plotting-two-or-more-dimensions.html#how-to-use-heatmaps-responsibly"><span class="toc-section-number">7.5.7</span> How To Use Heatmaps Responsibly</a></li>
</ul></li>
<li><a href="other-kind-of-plots.html#other-kind-of-plots"><span class="toc-section-number">7.6</span> Other Kind of Plots</a>
<ul>
<li><a href="other-kind-of-plots.html#dendrograms"><span class="toc-section-number">7.6.1</span> Dendrograms</a></li>
<li><a href="other-kind-of-plots.html#chord-diagrams-and-circos-plots"><span class="toc-section-number">7.6.2</span> Chord Diagrams and Circos Plots</a></li>
</ul></li>
<li><a href="multiple-plots.html#multiple-plots"><span class="toc-section-number">7.7</span> Multiple Plots</a>
<ul>
<li><a href="multiple-plots.html#facet-wrapping"><span class="toc-section-number">7.7.1</span> Facet wrapping</a></li>
<li><a href="multiple-plots.html#multipanel-figures"><span class="toc-section-number">7.7.2</span> Multipanel Figures</a></li>
</ul></li>
<li><a href="publication-ready-plots.html#publication-ready-plots"><span class="toc-section-number">7.8</span> Publication Ready Plots</a>
<ul>
<li><a href="publication-ready-plots.html#ggplot-themes"><span class="toc-section-number">7.8.1</span> <code>ggplot</code> Themes</a></li>
<li><a href="publication-ready-plots.html#exporting-to-svg"><span class="toc-section-number">7.8.2</span> Exporting to SVG</a></li>
</ul></li>
<li><a href="network-visualization.html#network-visualization"><span class="toc-section-number">7.9</span> Network visualization</a>
<ul>
<li><a href="network-visualization.html#interactive-network-plots"><span class="toc-section-number">7.9.1</span> Interactive network plots</a></li>
</ul></li>
</ul></li>
<li><a href="biology-bioinformatics.html#biology-bioinformatics"><span class="toc-section-number">8</span> Biology &amp; Bioinformatics</a>
<ul>
<li><a href="bio-r.html#bio-r"><span class="toc-section-number">8.1</span> R in Biology</a></li>
<li><a href="biological-data-overview.html#biological-data-overview"><span class="toc-section-number">8.2</span> Biological Data Overview</a>
<ul>
<li><a href="biological-data-overview.html#bio-data-types"><span class="toc-section-number">8.2.1</span> Types of Biological Data</a></li>
<li><a href="biological-data-overview.html#csv-files"><span class="toc-section-number">8.2.2</span> CSV Files</a></li>
<li><a href="biological-data-overview.html#biological-data-is-not-tidy"><span class="toc-section-number">8.2.3</span> Biological data is NOT Tidy!</a></li>
</ul></li>
<li><a href="bio-bioconductor.html#bio-bioconductor"><span class="toc-section-number">8.3</span> Bioconductor</a></li>
<li><a href="microarrays.html#microarrays"><span class="toc-section-number">8.4</span> Microarrays</a></li>
<li><a href="high-throughput-sequencing.html#high-throughput-sequencing"><span class="toc-section-number">8.5</span> High Throughput Sequencing</a>
<ul>
<li><a href="high-throughput-sequencing.html#raw-hts-data"><span class="toc-section-number">8.5.1</span> Raw HTS Data</a></li>
<li><a href="high-throughput-sequencing.html#preliminary-hts-data-analysis"><span class="toc-section-number">8.5.2</span> Preliminary HTS Data Analysis</a></li>
<li><a href="high-throughput-sequencing.html#count-data"><span class="toc-section-number">8.5.3</span> Count Data</a></li>
</ul></li>
<li><a href="gene-identifiers.html#gene-identifiers"><span class="toc-section-number">8.6</span> Gene Identifiers</a>
<ul>
<li><a href="gene-identifiers.html#gene-identifier-systems"><span class="toc-section-number">8.6.1</span> Gene Identifier Systems</a></li>
<li><a href="gene-identifiers.html#mapping-between-identifier-systems"><span class="toc-section-number">8.6.2</span> Mapping Between Identifier Systems</a></li>
<li><a href="gene-identifiers.html#mapping-homologs"><span class="toc-section-number">8.6.3</span> Mapping Homologs</a></li>
</ul></li>
<li><a href="gene-expression.html#gene-expression"><span class="toc-section-number">8.7</span> Gene Expression</a>
<ul>
<li><a href="gene-expression.html#gene-expression-data-in-bioconductor"><span class="toc-section-number">8.7.1</span> Gene Expression Data in Bioconductor</a></li>
<li><a href="gene-expression.html#differential-expression-analysis"><span class="toc-section-number">8.7.2</span> Differential Expression Analysis</a></li>
<li><a href="gene-expression.html#microarray-gene-expression-data"><span class="toc-section-number">8.7.3</span> Microarray Gene Expression Data</a></li>
<li><a href="gene-expression.html#differential-expression-microarrays-limma"><span class="toc-section-number">8.7.4</span> Differential Expression: Microarrays (limma)</a></li>
<li><a href="gene-expression.html#rnaseq"><span class="toc-section-number">8.7.5</span> RNASeq</a></li>
<li><a href="gene-expression.html#rnaseq-gene-expression-data"><span class="toc-section-number">8.7.6</span> RNASeq Gene Expression Data</a></li>
<li><a href="gene-expression.html#filtering-counts"><span class="toc-section-number">8.7.7</span> Filtering Counts</a></li>
<li><a href="gene-expression.html#count-distributions"><span class="toc-section-number">8.7.8</span> Count Distributions</a></li>
<li><a href="gene-expression.html#differential-expression-rnaseq"><span class="toc-section-number">8.7.9</span> Differential Expression: RNASeq</a></li>
</ul></li>
<li><a href="gene-set-enrichment-analysis.html#gene-set-enrichment-analysis"><span class="toc-section-number">8.8</span> Gene Set Enrichment Analysis</a>
<ul>
<li><a href="gene-set-enrichment-analysis.html#gene-sets"><span class="toc-section-number">8.8.1</span> Gene Sets</a></li>
<li><a href="gene-set-enrichment-analysis.html#working-with-gene-sets-in-r"><span class="toc-section-number">8.8.2</span> Working with gene sets in R</a></li>
<li><a href="gene-set-enrichment-analysis.html#over-representation-analysis"><span class="toc-section-number">8.8.3</span> Over-representation Analysis</a></li>
<li><a href="gene-set-enrichment-analysis.html#rank-based-analysis"><span class="toc-section-number">8.8.4</span> Rank-based Analysis</a></li>
<li><a href="gene-set-enrichment-analysis.html#fgsea"><span class="toc-section-number">8.8.5</span> fgsea</a></li>
</ul></li>
<li><a href="biological-networks-..html#biological-networks-."><span class="toc-section-number">8.9</span> Biological Networks .</a>
<ul>
<li><a href="biological-networks-..html#biological-pathways-."><span class="toc-section-number">8.9.1</span> Biological Pathways .</a></li>
<li><a href="biological-networks-..html#gene-regulatory-networks-."><span class="toc-section-number">8.9.2</span> Gene Regulatory Networks .</a></li>
<li><a href="biological-networks-..html#protein-protein-interaction-network-."><span class="toc-section-number">8.9.3</span> Protein-Protein Interaction Network .</a></li>
<li><a href="biological-networks-..html#wgcna-."><span class="toc-section-number">8.9.4</span> WGCNA .</a></li>
</ul></li>
</ul></li>
<li><a href="engineering.html#engineering"><span class="toc-section-number">9</span> EngineeRing</a>
<ul>
<li><a href="unit-testing.html#unit-testing"><span class="toc-section-number">9.1</span> Unit Testing</a></li>
<li><a href="toolification.html#toolification"><span class="toc-section-number">9.2</span> Toolification</a>
<ul>
<li><a href="toolification.html#the-r-interpreter"><span class="toc-section-number">9.2.1</span> The R Interpreter</a></li>
<li><a href="toolification.html#rscript"><span class="toc-section-number">9.2.2</span> <code>Rscript</code></a></li>
<li><a href="toolification.html#commandargs"><span class="toc-section-number">9.2.3</span> <code>commandArgs()</code></a></li>
</ul></li>
<li><a href="parallel-processing.html#parallel-processing"><span class="toc-section-number">9.3</span> Parallel Processing</a>
<ul>
<li><a href="parallel-processing.html#brief-introduction-to-parallelization"><span class="toc-section-number">9.3.1</span> Brief Introduction to Parallelization</a></li>
<li><a href="parallel-processing.html#apply-and-friends-are-pleasingly-parallel"><span class="toc-section-number">9.3.2</span> <code>apply</code> and Friends Are Pleasingly Parallel</a></li>
<li><a href="parallel-processing.html#the-parallel-package"><span class="toc-section-number">9.3.3</span> The <code>parallel</code> package</a></li>
</ul></li>
<li><a href="object-orientation-in-r.html#object-orientation-in-r"><span class="toc-section-number">9.4</span> Object Orientation in R</a></li>
<li><a href="r-packages.html#r-packages"><span class="toc-section-number">9.5</span> R Packages</a></li>
</ul></li>
<li><a href="rshiny.html#rshiny"><span class="toc-section-number">10</span> RShiny</a>
<ul>
<li><a href="rshiny-introduction.html#rshiny-introduction"><span class="toc-section-number">10.1</span> RShiny Introduction</a>
<ul>
<li><a href="rshiny-introduction.html#how-does-a-web-page-work"><span class="toc-section-number">10.1.1</span> How does a web page work?</a></li>
</ul></li>
<li><a href="application-structure.html#application-structure"><span class="toc-section-number">10.2</span> Application Structure</a>
<ul>
<li><a href="application-structure.html#ui-design"><span class="toc-section-number">10.2.1</span> UI Design</a></li>
<li><a href="application-structure.html#server-functionality"><span class="toc-section-number">10.2.2</span> Server Functionality</a></li>
</ul></li>
<li><a href="reactivity.html#reactivity"><span class="toc-section-number">10.3</span> Reactivity</a></li>
<li><a href="publishing.html#publishing"><span class="toc-section-number">10.4</span> Publishing</a></li>
</ul></li>
<li><a href="communicating-with-r.html#communicating-with-r"><span class="toc-section-number">11</span> Communicating with R</a>
<ul>
<li><a href="rmarkdown-knitr.html#rmarkdown-knitr"><span class="toc-section-number">11.1</span> RMarkdown &amp; knitr</a></li>
<li><a href="bookdown.html#bookdown"><span class="toc-section-number">11.2</span> bookdown</a></li>
</ul></li>
<li><a href="contribution-guide.html#contribution-guide"><span class="toc-section-number">12</span> Contribution Guide</a>
<ul>
<li><a href="custom-blocks.html#custom-blocks"><span class="toc-section-number">12.1</span> Custom blocks</a></li>
</ul></li>
<li><a href="assignments.html#assignments">Assignments</a>
<ul>
<li><a href="assignment-format.html#assignment-format">Assignment Format</a></li>
<li><a href="starting-an-assignment.html#starting-an-assignment">Starting an Assignment</a>
<ul>
<li><a href="starting-an-assignment.html#starting-your-own-git-repository-and-cloning-it">Starting your own <code>git</code> repository and cloning it</a></li>
<li><a href="starting-an-assignment.html#contents">Contents</a></li>
<li><a href="starting-an-assignment.html#github-tutorial">GitHub tutorial</a></li>
<li><a href="starting-an-assignment.html#committing-and-pushing-with-r-and-without">Committing and Pushing with R (and without)</a></li>
<li><a href="starting-an-assignment.html#using-rstudio">Using RStudio</a></li>
<li><a href="starting-an-assignment.html#using-the-command-line">Using the command line</a></li>
</ul></li>
<li><a href="assignment-1.html#assignment-1">Assignment 1</a>
<ul>
<li><a href="assignment-1.html#problem-statement">Problem Statement</a></li>
<li><a href="assignment-1.html#learning-objectives-and-skill-list">Learning Objectives and Skill List</a></li>
<li><a href="assignment-1.html#instructions">Instructions</a></li>
<li><a href="assignment-1.html#hints">Hints</a></li>
</ul></li>
<li><a href="assignment-2.html#assignment-2">Assignment 2</a>
<ul>
<li><a href="assignment-2.html#problem-statement-1">Problem Statement</a></li>
<li><a href="assignment-2.html#learning-objectives">Learning Objectives</a></li>
<li><a href="assignment-2.html#skill-list">Skill List</a></li>
<li><a href="assignment-2.html#instructions-1">Instructions</a></li>
<li><a href="assignment-2.html#function-details">Function Details</a></li>
</ul></li>
<li><a href="assignment-3.html#assignment-3">Assignment 3</a>
<ul>
<li><a href="assignment-3.html#problem-statement-2">Problem Statement</a></li>
<li><a href="assignment-3.html#learning-objectives-1">Learning Objectives</a></li>
<li><a href="assignment-3.html#skill-list-1">Skill List</a></li>
<li><a href="assignment-3.html#background-on-microarrays">Background on Microarrays</a></li>
<li><a href="assignment-3.html#background-on-principal-component-analysis">Background on Principal Component Analysis</a></li>
<li><a href="assignment-3.html#marisa-et-al.-gene-expression-classification-of-colon-cancer-into-molecular-subtypes-characterization-validation-and-prognostic-value.-plos-medicine-may-2013.-pmid-23700391">Marisa et al. Gene Expression Classification of Colon Cancer into Molecular Subtypes: Characterization, Validation, and Prognostic Value. PLoS Medicine, May 2013. PMID: 23700391</a></li>
<li><a href="assignment-3.html#scaling-data-using-r-scale">Scaling data using R <code>scale()</code></a></li>
<li><a href="assignment-3.html#proportion-of-variance-explained">Proportion of variance explained</a></li>
<li><a href="assignment-3.html#plotting-and-visualization-of-pca">Plotting and visualization of PCA</a></li>
<li><a href="assignment-3.html#hierarchical-clustering-and-heatmaps">Hierarchical Clustering and Heatmaps</a></li>
<li><a href="assignment-3.html#references">References</a></li>
</ul></li>
<li><a href="assignment-4.html#assignment-4">Assignment 4</a>
<ul>
<li><a href="assignment-4.html#problem-statement-3">Problem Statement</a></li>
<li><a href="assignment-4.html#learning-objectives-2">Learning Objectives</a></li>
<li><a href="assignment-4.html#skill-list-2">Skill List</a></li>
<li><a href="assignment-4.html#instructions-2">Instructions</a></li>
<li><a href="assignment-4.html#deliverables">Deliverables</a></li>
<li><a href="assignment-4.html#function-details-1">Function Details</a></li>
<li><a href="assignment-4.html#hints-1">Hints</a></li>
</ul></li>
<li><a href="assignment-5.html#assignment-5">Assignment 5</a>
<ul>
<li><a href="assignment-5.html#problem-statement-4">Problem Statement</a></li>
<li><a href="assignment-5.html#learning-objectives-3">Learning Objectives</a></li>
<li><a href="assignment-5.html#skill-list-3">Skill List</a></li>
<li><a href="assignment-5.html#deseq2-background">DESeq2 Background</a></li>
<li><a href="assignment-5.html#generating-a-counts-matrix">Generating a counts matrix</a></li>
<li><a href="assignment-5.html#prefiltering-counts-matrix">Prefiltering Counts matrix</a></li>
<li><a href="assignment-5.html#median-of-ratios-normalization">Median-of-ratios normalization</a></li>
<li><a href="assignment-5.html#deseq2-preparation">DESeq2 preparation</a></li>
<li><a href="assignment-5.html#omeara-et-al.-transcriptional-reversion-of-cardiac-myocyte-fate-during-mammalian-cardiac-regeneration.-circ-res.-feb-2015.-pmid-25477501l">O’Meara et al. Transcriptional Reversion of Cardiac Myocyte Fate During Mammalian Cardiac Regeneration. Circ Res. Feb 2015. PMID: 25477501l</a></li>
<li><a href="assignment-5.html#reading-and-subsetting-the-data-from-verse_counts.tsv-and-sample_metadata.csv">1. Reading and subsetting the data from verse_counts.tsv and sample_metadata.csv</a></li>
<li><a href="assignment-5.html#running-deseq2">2. Running DESeq2</a></li>
<li><a href="assignment-5.html#annotating-results-to-construct-a-labeled-volcano-plot">3. Annotating results to construct a labeled volcano plot</a></li>
<li><a href="assignment-5.html#diagnostic-plot-of-the-raw-p-values-for-all-genes">4. Diagnostic plot of the raw p-values for all genes</a></li>
<li><a href="assignment-5.html#plotting-the-logfoldchanges-for-differentially-expressed-genes">5. Plotting the LogFoldChanges for differentially expressed genes</a></li>
<li><a href="assignment-5.html#the-choice-of-fdr-cutoff-depends-on-cost">The choice of FDR cutoff depends on cost</a></li>
<li><a href="assignment-5.html#plotting-the-normalized-counts-of-differentially-expressed-genes">6. Plotting the normalized counts of differentially expressed genes</a></li>
<li><a href="assignment-5.html#volcano-plot-to-visualize-differential-expression-results">7. Volcano Plot to visualize differential expression results</a></li>
<li><a href="assignment-5.html#running-fgsea-vignette">8. Running fgsea <span>vignette</span></a></li>
<li><a href="assignment-5.html#plotting-the-top-ten-positive-nes-and-top-ten-negative-nes-pathways">9. Plotting the top ten positive NES and top ten negative NES pathways</a></li>
<li><a href="assignment-5.html#references-1">References</a></li>
</ul></li>
<li><a href="assignment-6.html#assignment-6">Assignment 6</a>
<ul>
<li><a href="assignment-6.html#problem-statement-5">Problem Statement</a></li>
<li><a href="assignment-6.html#learning-objectives-4">Learning Objectives</a></li>
<li><a href="assignment-6.html#skill-list-4">Skill List</a></li>
<li><a href="assignment-6.html#instructions-3">Instructions</a></li>
<li><a href="assignment-6.html#function-details-2">Function Details</a></li>
</ul></li>
<li><a href="assignment-7.html#assignment-7">Assignment 7</a>
<ul>
<li><a href="assignment-7.html#problem-statement-6">Problem Statement</a></li>
<li><a href="assignment-7.html#learning-objectives-5">Learning Objectives</a></li>
<li><a href="assignment-7.html#skill-list-5">Skill List</a></li>
<li><a href="assignment-7.html#instructions-4">Instructions</a></li>
<li><a href="assignment-7.html#function-details-3">“Function” Details</a></li>
</ul></li>
</ul></li>
<li class="appendix"><span><b>Appendix</b></span></li>
<li><a href="class-outlines.html#class-outlines"><span class="toc-section-number">A</span> Class Outlines</a>
<ul>
<li><a href="week-1.html#week-1"><span class="toc-section-number">A.1</span> Week 1</a></li>
<li><a href="week-2.html#week-2"><span class="toc-section-number">A.2</span> Week 2</a></li>
<li><a href="week-3.html#week-3"><span class="toc-section-number">A.3</span> Week 3</a></li>
<li><a href="week-4.html#week-4"><span class="toc-section-number">A.4</span> Week 4</a></li>
<li><a href="week-5.html#week-5"><span class="toc-section-number">A.5</span> Week 5</a></li>
<li><a href="week-6.html#week-6"><span class="toc-section-number">A.6</span> Week 6</a></li>
<li><a href="week-7.html#week-7"><span class="toc-section-number">A.7</span> Week 7</a></li>
<li><a href="week-8.html#week-8"><span class="toc-section-number">A.8</span> Week 8</a></li>
<li><a href="week-9.html#week-9"><span class="toc-section-number">A.9</span> Week 9</a></li>
<li><a href="week-10.html#week-10"><span class="toc-section-number">A.10</span> Week 10</a></li>
<li><a href="week-11.html#week-11"><span class="toc-section-number">A.11</span> Week 11</a></li>
<li><a href="week-12.html#week-12"><span class="toc-section-number">A.12</span> Week 12</a></li>
<li><a href="week-13.html#week-13"><span class="toc-section-number">A.13</span> Week 13</a></li>
<li><a href="week-14.html#week-14"><span class="toc-section-number">A.14</span> Week 14</a></li>
</ul></li>
</ul>
</div>
</div>
</div>
<div class="row">
<div class="col-sm-12">
<div id="toolification" class="section level2" number="9.2">
<h2><span class="header-section-number">9.2</span> Toolification</h2>
<p>RStudio is a highly effective tool for developing R code, and most analyses we
conduct are suitable for running in its interactive environment. However, there
are certain circumstances where running R scripts in this interactive manner is
not optimal or even suitable:</p>
<ol style="list-style-type: decimal">
<li>The same script must be run on many different inputs</li>
<li>The data that is to be analyzed is stored on a system where interactive
RStudio session are not available</li>
<li>The script takes a very long time to run (e.g. days to weeks)</li>
<li>The computational demands of the analysis exceed the resources available on
the computer where RStudio is being run (e.g. a personal laptop has limited
memory)</li>
</ol>
<p>In these cases, it is necessary to run R scripts outside of RStudio, often on
a command line interface (CLI) like that found on linux clusters and cloud based
virtual machines. While most scripts can be run in an R interpreter on the
CLI and produce the same behavior, there are some steps that help
convert an R script into a tool that can be run in this way. This process of
transforming an R script in to a more generally usable tool that is run on the
CLI is termed <em>toolification</em> in this text. The following sections describe some
strategies for <em>toolifying</em> and running R scripts developed in RStudio on the
CLI.</p>
<div id="the-r-interpreter" class="section level3" number="9.2.1">
<h3><span class="header-section-number">9.2.1</span> The R Interpreter</h3>
<p>It bears mentioning that RStudio and R are related but independent programs.
Specifically, RStudio runs the R program behind its interface using what is
called the <a href="https://www.oreilly.com/library/view/r-in-a/9781449358204/ch05s09.html"><em>R interpreter</em></a>.
When you run R on a CLI, you are given an interactive interface where commands
written in the R language can be evaluated. In RStudio, the interpreter that is
running behind the scenes can be accessed in the Console tab at the bottom right:</p>
<div class="figure">
<img src="r_interpreter.png" alt="" />
<p class="caption">R Interpreter in both RStudio and a linux terminal</p>
</div>
<p>The interpreter by itself is not very useful, since most meaningful analyses
require many lines of code be run sequentially as a unit. The interpreter can be
helpful to test out individual lines of code and examine help documentation for
R functions.</p>
<p>The simplest way to run an R script from the command line is to first run an R
interpreter and then use the <a href="https://www.rdocumentation.org/packages/base/versions/3.6.2/topics/source"><code>source()</code> function</a>,
which accepts a filename for an R script as its first argument an executes all
lines of code in the script.</p>
<pre><code>$ cat simple_script.R
print(&#39;hello moon&#39;)
print(1+2)
print(str(c(1,2,3,4)))
$ R
&gt; source(&quot;simple_script.R&quot;)
[1] &quot;hello moon&quot;
[1] 3
 num [1:4] 1 2 3 4</code></pre>
<p>However, this method requires some interactivity, namely by running an R
interpreter first, so it is not sufficient to run in a non-interactive fashion.
We may need to run our script without any interactivity, for example when it is
a part of a computational pipeline.</p>
</div>
<div id="rscript" class="section level3" number="9.2.2">
<h3><span class="header-section-number">9.2.2</span> <code>Rscript</code></h3>
<p>The R program also includes the <a href="https://www.rdocumentation.org/packages/utils/versions/3.6.2/topics/Rscript"><code>Rscript</code> command</a>
that can be run on the command line:</p>
<pre><code>$ Rscript
Usage: /path/to/Rscript [--options] [-e expr [-e expr2 ...] | file] [args]

--options accepted are
  --help              Print usage and exit
  --version           Print version and exit
  --verbose           Print information on progress
  --default-packages=list
                      Where &#39;list&#39; is a comma-separated set
                        of package names, or &#39;NULL&#39;
or options to R, in addition to --no-echo --no-restore, such as
  --save              Do save workspace at the end of the session
  --no-environ        Don&#39;t read the site and user environment files
  --no-site-file      Don&#39;t read the site-wide Rprofile
  --no-init-file      Don&#39;t read the user R profile
  --restore           Do restore previously saved objects at startup
  --vanilla           Combine --no-save, --no-restore, --no-site-file
                        --no-init-file and --no-environ

&#39;file&#39; may contain spaces but not shell metacharacters
Expressions (one or more &#39;-e &lt;expr&gt;&#39;) may be used *instead* of &#39;file&#39;
See also  ?Rscript  from within R</code></pre>
<p>This CLI command accepts an R script as an input and executes the commands in
the file as if they had been passed to <code>source()</code>, for example:</p>
<pre><code>$ cat simple_script.R
print(&#39;hello moon&#39;)
print(1+2)
print(str(c(1,2,3,4)))
$ Rscript simple_script.R
[1] &quot;hello moon&quot;
[1] 3
 num [1:4] 1 2 3 4
$</code></pre>
<p>This is the simplest way to toolify an R script; simply run it on the command
line with <code>Rscript</code>. Toolifying simple R scripts that do not need to accept
inputs to execute on different data generally require no changes.</p>
<div class="box note">
<p><code>Rscript</code> is a convenience function that runs the following <code>R</code> command:</p>
<pre><code>$ Rscript simple_script.R # is equivalent to:
$ R --no-echo --no-restore --file=simple_script.R</code></pre>
<p>If for some reason you cannot run <code>Rscript</code> directly, you can use these
arguments with the <code>R</code> command to attain the same result.</p>
</div>
</div>
<div id="commandargs" class="section level3" number="9.2.3">
<h3><span class="header-section-number">9.2.3</span> <code>commandArgs()</code></h3>
<p>However, sometimes we may wish to control the behavior of a script directly from
the command line, rather than editing the script directly for every different
execution. To pass information into the script when it is run, we can pass
arguments with the <code>Rscript</code> command:</p>
<pre><code>$ Rscript simple_script.R abc
[1] &quot;hello moon&quot;
[1] 3
 num [1:4] 1 2 3 4</code></pre>
<p>Although we passed in the argument <code>abc</code>, the output of the script didn’t change
because the script wasn’t written to receive it. In order for a script to gain
access to the command line arguments, we must call the <a href="https://www.rdocumentation.org/packages/R.utils/versions/2.11.0/topics/commandArgs"><code>commandArgs()</code>
function</a>:</p>
<div class="sourceCode" id="cb472"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb472-1"><a href="toolification.html#cb472-1" aria-hidden="true" tabindex="-1"></a>args <span class="ot">&lt;-</span> <span class="fu">commandArgs</span>(<span class="at">trailingOnly=</span><span class="cn">TRUE</span>)</span></code></pre></div>
<p>Now when we execute the script, the arguments passed in are available in the
<code>args</code> variable:</p>
<pre><code>$ cat echo_args.R
print(commandArgs(trailingOnly=TRUE))
$ Rscript echo_args.R abc
[1] &quot;abc&quot;
$ Rscript echo_args.R abc 123
[1] &quot;abc&quot; &quot;123&quot;
$ Rscript echo_args.R # no args
character(0)</code></pre>
<p>In the last case when no arguments were passed, R is telling us the <code>args</code>
variable is a character vector of length zero.</p>
<div class="box note">
<p>By default, the <code>commandArgs()</code> function will return the full command that was
run, including the <code>Rscript</code> command itself and any additional arguments:</p>
<pre><code>$ Rscript -e &quot;commandArgs()&quot; abc 123
[1] &quot;/usr/bin/R&quot;
[2] &quot;--no-echo&quot;
[3] &quot;--no-restore&quot;
[4] &quot;-e&quot;
[5] &quot;commandArgs()&quot;
[6] &quot;--args&quot;
[7] &quot;abc&quot;
[8] &quot;123&quot;</code></pre>
<p>The <code>trailingOnly=TRUE</code> argument returns only the arguments provided at the end
of the command, after the <code>Rscript</code> portion:</p>
<pre><code>$ Rscript -e &quot;commandArgs(trailingOnly=TRUE)&quot; abc 123
[1] &quot;abc&quot; &quot;123&quot;</code></pre>
<p>Note you can provide individual commands instead of a script to <code>Rscript</code> with
the <code>-e</code> argument.</p>
</div>
<p>The <code>commandArgs()</code> function is all that is needed to toolify a R script.
Consider a simple script named <code>inspect_csv.R</code> that loads in any CSV file and
summarizes it as a tibble:</p>
<div class="sourceCode" id="cb476"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb476-1"><a href="toolification.html#cb476-1" aria-hidden="true" tabindex="-1"></a>args <span class="ot">&lt;-</span> <span class="fu">commandArgs</span>(<span class="at">trailingOnly=</span><span class="cn">TRUE</span>)</span>
<span id="cb476-2"><a href="toolification.html#cb476-2" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="fu">length</span>(args) <span class="sc">!=</span> <span class="dv">1</span>) {</span>
<span id="cb476-3"><a href="toolification.html#cb476-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">&quot;Usage: simple_script.R &lt;csv file&gt;</span><span class="sc">\n</span><span class="st">&quot;</span>) <span class="co"># cat() writes characters to the screen</span></span>
<span id="cb476-4"><a href="toolification.html#cb476-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">&quot;Provide exactly one argument that is a CSV filename</span><span class="sc">\n</span><span class="st">&quot;</span>)</span>
<span id="cb476-5"><a href="toolification.html#cb476-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">quit</span>(<span class="at">save=</span><span class="st">&quot;no&quot;</span>, <span class="at">status=</span><span class="dv">1</span>)</span>
<span id="cb476-6"><a href="toolification.html#cb476-6" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb476-7"><a href="toolification.html#cb476-7" aria-hidden="true" tabindex="-1"></a>fn <span class="ot">&lt;-</span> args[<span class="dv">1</span>]</span>
<span id="cb476-8"><a href="toolification.html#cb476-8" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb476-9"><a href="toolification.html#cb476-9" aria-hidden="true" tabindex="-1"></a><span class="fu">read_csv</span>(fn)</span></code></pre></div>
<p>We can now run the script with <code>Rscript</code> and give it the filename of a CSV file:</p>
<pre><code>$ cat data.csv
gene,sampleA,sampleB,sampleC
g1,1,35,20
g2,32,56,99
g3,392,583,444
g4,39853,16288,66928
$ Rscript inspect_csv.R data.csv
── Attaching packages ─────────────────────────────────────── tidyverse 1.3.1 ──
✔ ggplot2 3.3.5     ✔ purrr   0.3.4
✔ tibble  3.1.6     ✔ dplyr   1.0.7
✔ tidyr   1.2.0     ✔ stringr 1.4.0
✔ readr   2.1.2     ✔ forcats 0.5.1
── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
✖ dplyr::filter() masks stats::filter()
✖ dplyr::lag()    masks stats::lag()
Rows: 4 Columns: 4
── Column specification ────────────────────────────────────────────────────────
Delimiter: &quot;,&quot;
chr (1): gene
dbl (3): sampleA, sampleB, sampleC

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message   .
# A tibble: 4 × 4
  gene  sampleA sampleB sampleC
  &lt;chr&gt;   &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;
1 g1          1      35      20
2 g2         32      56      99
3 g3        392     583     444
4 g4      39853   16288   66928</code></pre>
<p>Note the input handling of the arguments, where the usage of the script and a
helpful error message is printed out before the script quits if there was not
exactly one argument provided.</p>
<p>In the example above, a filename was passed into the script as an argument. A
filename is encoded as a character string in this case, and <code>commandArgs()</code>
always produces a vector of strings. If we need to pass in arguments to control
numerical values, we will need to parse the arguments first. Consider the
following R implementation of the linux command <code>head</code>, which prints only the
top <span class="math inline">\(n\)</span> lines of a file to the screen:</p>
<div class="sourceCode" id="cb478"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb478-1"><a href="toolification.html#cb478-1" aria-hidden="true" tabindex="-1"></a>args <span class="ot">&lt;-</span> <span class="fu">commandArgs</span>(<span class="at">trailingOnly=</span><span class="cn">TRUE</span>)</span>
<span id="cb478-2"><a href="toolification.html#cb478-2" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="fu">length</span>(args) <span class="sc">!=</span> <span class="dv">2</span>) {</span>
<span id="cb478-3"><a href="toolification.html#cb478-3" aria-hidden="true" tabindex="-1"></a>      <span class="fu">cat</span>(<span class="st">&quot;Usage: head.R &lt;filename&gt; &lt;N&gt;</span><span class="sc">\n</span><span class="st">&quot;</span>)</span>
<span id="cb478-4"><a href="toolification.html#cb478-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">&quot;Provide exactly two arguments: a CSV filename and an integer N</span><span class="sc">\n</span><span class="st">&quot;</span>)</span>
<span id="cb478-5"><a href="toolification.html#cb478-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">quit</span>(<span class="at">save=</span><span class="st">&quot;no&quot;</span>, <span class="at">status=</span><span class="dv">1</span>)</span>
<span id="cb478-6"><a href="toolification.html#cb478-6" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb478-7"><a href="toolification.html#cb478-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb478-8"><a href="toolification.html#cb478-8" aria-hidden="true" tabindex="-1"></a><span class="co"># read in arguments</span></span>
<span id="cb478-9"><a href="toolification.html#cb478-9" aria-hidden="true" tabindex="-1"></a>fn <span class="ot">&lt;-</span> args[<span class="dv">1</span>]</span>
<span id="cb478-10"><a href="toolification.html#cb478-10" aria-hidden="true" tabindex="-1"></a>n <span class="ot">&lt;-</span> <span class="fu">as.integer</span>(args[<span class="dv">2</span>])</span>
<span id="cb478-11"><a href="toolification.html#cb478-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb478-12"><a href="toolification.html#cb478-12" aria-hidden="true" tabindex="-1"></a><span class="co"># the csv output will include a header row, so reduce n by 1</span></span>
<span id="cb478-13"><a href="toolification.html#cb478-13" aria-hidden="true" tabindex="-1"></a>n <span class="ot">&lt;-</span> n<span class="dv">-1</span></span>
<span id="cb478-14"><a href="toolification.html#cb478-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb478-15"><a href="toolification.html#cb478-15" aria-hidden="true" tabindex="-1"></a><span class="co"># suppressMessages() prevents messages like library loading text from being printed to the screen</span></span>
<span id="cb478-16"><a href="toolification.html#cb478-16" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressMessages</span>(</span>
<span id="cb478-17"><a href="toolification.html#cb478-17" aria-hidden="true" tabindex="-1"></a>    {</span>
<span id="cb478-18"><a href="toolification.html#cb478-18" aria-hidden="true" tabindex="-1"></a>        <span class="fu">library</span>(tidyverse, <span class="at">quietly=</span><span class="cn">TRUE</span>)</span>
<span id="cb478-19"><a href="toolification.html#cb478-19" aria-hidden="true" tabindex="-1"></a>        <span class="fu">read_csv</span>(fn) <span class="sc">%&gt;%</span></span>
<span id="cb478-20"><a href="toolification.html#cb478-20" aria-hidden="true" tabindex="-1"></a>            <span class="fu">slice_head</span>(<span class="at">n=</span>n) <span class="sc">%&gt;%</span></span>
<span id="cb478-21"><a href="toolification.html#cb478-21" aria-hidden="true" tabindex="-1"></a>            <span class="fu">write_csv</span>(<span class="fu">stdout</span>())</span>
<span id="cb478-22"><a href="toolification.html#cb478-22" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb478-23"><a href="toolification.html#cb478-23" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div>
<p>We again tested the number of arguments passed in for correct usage, and then
assigned the arguments to variables. The <code>n</code> argument is cast from a character
string to an integer in the process, enabling its use in the
<code>dplyr::slice_head()</code> function. We can print out the first three lines of a file
as follows:</p>
<pre><code>$ Rscript head.R data.csv 3
gene,sampleA,sampleB,sampleC
g1,1,35,20
g2,32,56,99</code></pre>
<p>Reading command line arguments into variables in a script can become tedious if
your script has a large number of arguments. Fortunately, the <a href="https://www.rdocumentation.org/packages/argparser/versions/0.7.1"><code>argparser</code>
package</a> can help handle many of the
repetitive operations, including specifying arguments, providing default values,
automatically casting to appropriate types like numbers, and printing out usage
information:</p>
<div class="sourceCode" id="cb480"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb480-1"><a href="toolification.html#cb480-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(argparser, <span class="at">quietly=</span><span class="cn">TRUE</span>)</span>
<span id="cb480-2"><a href="toolification.html#cb480-2" aria-hidden="true" tabindex="-1"></a>parser <span class="ot">&lt;-</span> <span class="fu">arg_parser</span>(<span class="st">&quot;R implementation of GNU coreutils head command&quot;</span>)</span>
<span id="cb480-3"><a href="toolification.html#cb480-3" aria-hidden="true" tabindex="-1"></a>parser <span class="ot">&lt;-</span> <span class="fu">add_argument</span>(parser, <span class="st">&quot;filename&quot;</span>, <span class="at">help=</span><span class="st">&quot;file to print lines from&quot;</span>)</span>
<span id="cb480-4"><a href="toolification.html#cb480-4" aria-hidden="true" tabindex="-1"></a>parser <span class="ot">&lt;-</span> <span class="fu">add_argument</span>(parser, <span class="st">&quot;-n&quot;</span>, <span class="at">help=</span><span class="st">&quot;number of lines to print&quot;</span>, <span class="at">type=</span><span class="st">&quot;numeric&quot;</span>, <span class="at">default=</span><span class="dv">10</span>)</span>
<span id="cb480-5"><a href="toolification.html#cb480-5" aria-hidden="true" tabindex="-1"></a>parser <span class="ot">&lt;-</span> <span class="fu">parse_args</span>(parser, <span class="fu">c</span>(<span class="st">&quot;-n&quot;</span>,<span class="dv">3</span>,<span class="st">&quot;data.csv&quot;</span>))</span>
<span id="cb480-6"><a href="toolification.html#cb480-6" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">paste</span>(<span class="st">&quot;printing from file:&quot;</span>,parser<span class="sc">$</span>filename))</span></code></pre></div>
<pre><code>## [1] &quot;printing from file: data.csv&quot;</code></pre>
<div class="sourceCode" id="cb482"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb482-1"><a href="toolification.html#cb482-1" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">paste</span>(<span class="st">&quot;printing top n:&quot;</span>,parser<span class="sc">$</span>n))</span></code></pre></div>
<pre><code>## [1] &quot;printing top n: 3&quot;</code></pre>
<p>With this library, we can rewrite our <code>head.R</code> script to be more concise:</p>
<div class="sourceCode" id="cb484"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb484-1"><a href="toolification.html#cb484-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(argparser, <span class="at">quietly=</span><span class="cn">TRUE</span>)</span>
<span id="cb484-2"><a href="toolification.html#cb484-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb484-3"><a href="toolification.html#cb484-3" aria-hidden="true" tabindex="-1"></a><span class="co"># instantiate parser</span></span>
<span id="cb484-4"><a href="toolification.html#cb484-4" aria-hidden="true" tabindex="-1"></a>parser <span class="ot">&lt;-</span> <span class="fu">arg_parser</span>(<span class="st">&quot;R implementation of GNU coreutils head command&quot;</span>)</span>
<span id="cb484-5"><a href="toolification.html#cb484-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb484-6"><a href="toolification.html#cb484-6" aria-hidden="true" tabindex="-1"></a><span class="co"># add arguments</span></span>
<span id="cb484-7"><a href="toolification.html#cb484-7" aria-hidden="true" tabindex="-1"></a>parser <span class="ot">&lt;-</span> <span class="fu">add_argument</span>(</span>
<span id="cb484-8"><a href="toolification.html#cb484-8" aria-hidden="true" tabindex="-1"></a>  parser,</span>
<span id="cb484-9"><a href="toolification.html#cb484-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">arg=</span><span class="st">&quot;filename&quot;</span>,</span>
<span id="cb484-10"><a href="toolification.html#cb484-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">help=</span><span class="st">&quot;file to print lines from&quot;</span></span>
<span id="cb484-11"><a href="toolification.html#cb484-11" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb484-12"><a href="toolification.html#cb484-12" aria-hidden="true" tabindex="-1"></a>parser <span class="ot">&lt;-</span> <span class="fu">add_argument</span>(</span>
<span id="cb484-13"><a href="toolification.html#cb484-13" aria-hidden="true" tabindex="-1"></a>  parser,</span>
<span id="cb484-14"><a href="toolification.html#cb484-14" aria-hidden="true" tabindex="-1"></a>  <span class="at">arg=</span><span class="st">&quot;--top&quot;</span>,</span>
<span id="cb484-15"><a href="toolification.html#cb484-15" aria-hidden="true" tabindex="-1"></a>  <span class="at">help=</span><span class="st">&quot;number of lines to print&quot;</span>,</span>
<span id="cb484-16"><a href="toolification.html#cb484-16" aria-hidden="true" tabindex="-1"></a>  <span class="at">type=</span><span class="st">&quot;numeric&quot;</span>,</span>
<span id="cb484-17"><a href="toolification.html#cb484-17" aria-hidden="true" tabindex="-1"></a>  <span class="at">default=</span><span class="dv">10</span>,</span>
<span id="cb484-18"><a href="toolification.html#cb484-18" aria-hidden="true" tabindex="-1"></a>  <span class="at">short=</span><span class="st">&#39;-n&#39;</span></span>
<span id="cb484-19"><a href="toolification.html#cb484-19" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb484-20"><a href="toolification.html#cb484-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb484-21"><a href="toolification.html#cb484-21" aria-hidden="true" tabindex="-1"></a>args <span class="ot">&lt;-</span> <span class="fu">parse_args</span>(parser)</span>
<span id="cb484-22"><a href="toolification.html#cb484-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb484-23"><a href="toolification.html#cb484-23" aria-hidden="true" tabindex="-1"></a>fn <span class="ot">&lt;-</span> args<span class="sc">$</span>filename</span>
<span id="cb484-24"><a href="toolification.html#cb484-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb484-25"><a href="toolification.html#cb484-25" aria-hidden="true" tabindex="-1"></a><span class="co"># the csv output will include a header row, so reduce n by 1</span></span>
<span id="cb484-26"><a href="toolification.html#cb484-26" aria-hidden="true" tabindex="-1"></a>n <span class="ot">&lt;-</span> args<span class="sc">$</span>top<span class="dv">-1</span></span>
<span id="cb484-27"><a href="toolification.html#cb484-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb484-28"><a href="toolification.html#cb484-28" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressMessages</span>(</span>
<span id="cb484-29"><a href="toolification.html#cb484-29" aria-hidden="true" tabindex="-1"></a>    {</span>
<span id="cb484-30"><a href="toolification.html#cb484-30" aria-hidden="true" tabindex="-1"></a>        <span class="fu">library</span>(tidyverse, <span class="at">quietly=</span><span class="cn">TRUE</span>)</span>
<span id="cb484-31"><a href="toolification.html#cb484-31" aria-hidden="true" tabindex="-1"></a>        <span class="fu">read_csv</span>(fn) <span class="sc">%&gt;%</span></span>
<span id="cb484-32"><a href="toolification.html#cb484-32" aria-hidden="true" tabindex="-1"></a>            <span class="fu">slice_head</span>(<span class="at">n=</span>n) <span class="sc">%&gt;%</span></span>
<span id="cb484-33"><a href="toolification.html#cb484-33" aria-hidden="true" tabindex="-1"></a>            <span class="fu">write_csv</span>(<span class="fu">stdout</span>())</span>
<span id="cb484-34"><a href="toolification.html#cb484-34" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb484-35"><a href="toolification.html#cb484-35" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div>
<p>Note we didn’t have to explicitly parse the <code>top</code> argument as an integer because
<code>type="numeric"</code> handled this for us. We can print out the top three lines of
our file like we did above using the new parser and arguments:</p>
<pre><code>$ Rscript head.R -n 3 data.csv
gene,sampleA,sampleB,sampleC
g1,1,35,20
g2,32,56,99</code></pre>
<p>We can also inspect the usage by passing the <code>-h</code> flag:</p>
<pre><code>$ Rscript head.R -h
usage: head.R [--] [--help] [--opts OPTS] [--top TOP] filename

R implementation of GNU coreutils head command

positional arguments:
  filename    file to print lines from

flags:
  -h, --help  show this help message and exit

optional arguments:
  -x, --opts  RDS file containing argument values
  -n, --top   number of lines to print [default: 10]</code></pre>
<p>These are the only tools required to toolify our R script.</p>
<div class="box important">
<p>Note that scripts developed in RStudio can be run on the command line, but
scripts written for CLI use with the strategies above cannot be easily run
inside RStudio! However, if you followed good practices and implemented your
script as a set of functions, you can easily write a CLI wapper script that
calls those functions, thereby enabling you to continue developing your code in
RStudio and maintain CLI tool functionality.</p>
</div>
</div>
</div>
<p style="text-align: center;">
<a href="unit-testing.html"><button class="btn btn-default">Previous</button></a>
<a href="parallel-processing.html"><button class="btn btn-default">Next</button></a>
</p>
</div>
</div>


</div>

<script>

// add bootstrap table styles to pandoc tables
$(document).ready(function () {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
});

</script>

</body>
</html>
