<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta name="viewport" content="width=device-width, initial-scale=1" />
<meta property="og:title" content="7.4 Visualizing Distributions | BF591 - R for Biological Sciences" />
<meta property="og:type" content="book" />







<script type="text/x-mathjax-config">
MathJax.Hub.Config({
  TeX: { equationNumbers: { autoNumber: "AMS" } }
});
</script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<meta name="description" content="7.4 Visualizing Distributions | BF591 - R for Biological Sciences">

<title>7.4 Visualizing Distributions | BF591 - R for Biological Sciences</title>

<script src="libs/header-attrs-2.13/header-attrs.js"></script>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="libs/bootstrap-3.3.5/css/darkly.min.css" rel="stylesheet" />
<script src="libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="libs/navigation-1.1/tabsets.js"></script>


<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>


<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />

<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
/* show arrow before summary tag as in bootstrap
TODO: remove if boostrap in updated in html_document (rmarkdown#1485) */
details > summary {
  display: list-item;
  cursor: pointer;
}
</style>
</head>

<body>

<div class="container-fluid main-container">


<div class="row">
<div class="col-sm-12">
<div id="TOC">
<ul>
<li><a href="index.html#syllabus">Syllabus</a>
<ul>
<li><a href="course-schedule.html#course-schedule">Course Schedule</a></li>
<li><a href="instructors.html#instructors">Instructors</a></li>
<li><a href="office-hours.html#office-hours">Office Hours</a></li>
<li><a href="course-values-and-policies.html#course-values-and-policies">Course Values and Policies</a></li>
</ul></li>
<li><a href="introduction.html#introduction"><span class="toc-section-number">1</span> Introduction</a>
<ul>
<li><a href="who-this-book-is-for.html#who-this-book-is-for"><span class="toc-section-number">1.1</span> Who This Book Is For</a></li>
<li><a href="a-note-about-reinventing-the-wheel.html#a-note-about-reinventing-the-wheel"><span class="toc-section-number">1.2</span> A Note About Reinventing the Wheel</a></li>
<li><a href="sources-and-references.html#sources-and-references"><span class="toc-section-number">1.3</span> Sources and References</a></li>
</ul></li>
<li><a href="data-bio.html#data-bio"><span class="toc-section-number">2</span> Data in Biology</a>
<ul>
<li><a href="a-brief-history-of-data-in-molecular-biology.html#a-brief-history-of-data-in-molecular-biology"><span class="toc-section-number">2.1</span> A Brief History of Data in Molecular Biology</a></li>
<li><a href="biology-as-a-mature-data-science.html#biology-as-a-mature-data-science"><span class="toc-section-number">2.2</span> Biology as a Mature Data Science</a></li>
</ul></li>
<li><a href="preliminaries.html#preliminaries"><span class="toc-section-number">3</span> Preliminaries</a>
<ul>
<li><a href="prelim-r.html#prelim-r"><span class="toc-section-number">3.1</span> The R Language</a></li>
<li><a href="prelim-rstudio.html#prelim-rstudio"><span class="toc-section-number">3.2</span> RStudio</a></li>
<li><a href="prog-the-r-script.html#prog-the-r-script"><span class="toc-section-number">3.3</span> The R Script</a></li>
<li><a href="prog-workflow.html#prog-workflow"><span class="toc-section-number">3.4</span> The Scripting Workflow</a></li>
<li><a href="prelim-git.html#prelim-git"><span class="toc-section-number">3.5</span> git + github</a>
<ul>
<li><a href="prelim-git.html#motivation"><span class="toc-section-number">3.5.1</span> Motivation</a></li>
<li><a href="prelim-git.html#git"><span class="toc-section-number">3.5.2</span> git</a></li>
<li><a href="prelim-git.html#git-hosting-platforms-github"><span class="toc-section-number">3.5.3</span> Git hosting platforms (GitHub)</a></li>
</ul></li>
</ul></li>
<li><a href="prog-basics.html#prog-basics"><span class="toc-section-number">4</span> R Programming</a>
<ul>
<li><a href="before-you-begin.html#before-you-begin"><span class="toc-section-number">4.1</span> Before you begin</a></li>
<li><a href="introduction-1.html#introduction-1"><span class="toc-section-number">4.2</span> Introduction</a></li>
<li><a href="prog-r-syntax.html#prog-r-syntax"><span class="toc-section-number">4.3</span> R Syntax Basics</a></li>
<li><a href="prog-types.html#prog-types"><span class="toc-section-number">4.4</span> Basic Types of Values</a>
<ul>
<li><a href="prog-types.html#factors"><span class="toc-section-number">4.4.1</span> Factors</a></li>
</ul></li>
<li><a href="prog-struct.html#prog-struct"><span class="toc-section-number">4.5</span> Data Structures</a>
<ul>
<li><a href="prog-struct.html#vectors"><span class="toc-section-number">4.5.1</span> Vectors</a></li>
<li><a href="prog-struct.html#matrices"><span class="toc-section-number">4.5.2</span> Matrices</a></li>
<li><a href="prog-struct.html#lists-and-data-frames"><span class="toc-section-number">4.5.3</span> Lists and data frames</a></li>
</ul></li>
<li><a href="logical-tests-and-comparators.html#logical-tests-and-comparators"><span class="toc-section-number">4.6</span> Logical Tests and Comparators</a></li>
<li><a href="functions.html#functions"><span class="toc-section-number">4.7</span> Functions</a>
<ul>
<li><a href="functions.html#prog-dry"><span class="toc-section-number">4.7.1</span> DRY: Don’t Repeat Yourself</a></li>
<li><a href="functions.html#writing-your-own-functions"><span class="toc-section-number">4.7.2</span> Writing your own functions</a></li>
<li><a href="functions.html#scope"><span class="toc-section-number">4.7.3</span> Scope</a></li>
</ul></li>
<li><a href="iteration.html#iteration"><span class="toc-section-number">4.8</span> Iteration</a>
<ul>
<li><a href="iteration.html#vectorized-operations"><span class="toc-section-number">4.8.1</span> Vectorized operations</a></li>
<li><a href="iteration.html#functional-programming"><span class="toc-section-number">4.8.2</span> Functional programming</a></li>
<li><a href="iteration.html#apply-and-friends"><span class="toc-section-number">4.8.3</span> <code>apply()</code> and friends</a></li>
</ul></li>
<li><a href="installing-packages.html#installing-packages"><span class="toc-section-number">4.9</span> Installing Packages</a></li>
<li><a href="saving-and-loading-r-data.html#saving-and-loading-r-data"><span class="toc-section-number">4.10</span> Saving and Loading R Data</a></li>
<li><a href="prog-debugging.html#prog-debugging"><span class="toc-section-number">4.11</span> Troubleshooting and Debugging</a></li>
<li><a href="prog-style.html#prog-style"><span class="toc-section-number">4.12</span> Coding Style and Conventions</a>
<ul>
<li><a href="prog-style.html#is-my-code-correct"><span class="toc-section-number">4.12.1</span> Is my code correct?</a></li>
<li><a href="prog-style.html#does-my-code-follow-the-dry-principle"><span class="toc-section-number">4.12.2</span> Does my code follow the <span>DRY principle</span>?</a></li>
<li><a href="prog-style.html#did-i-choose-concise-but-descriptive-variable-and-function-names"><span class="toc-section-number">4.12.3</span> Did I choose concise but descriptive variable and function names?</a></li>
<li><a href="prog-style.html#did-i-use-indentation-and-naming-conventions-consistently-throughout-my-code"><span class="toc-section-number">4.12.4</span> Did I use indentation and naming conventions consistently throughout my code?</a></li>
<li><a href="prog-style.html#did-i-write-comments-especially-when-what-the-code-does-is-not-obvious"><span class="toc-section-number">4.12.5</span> Did I write comments, especially when what the code does is not obvious?</a></li>
<li><a href="prog-style.html#how-easy-would-it-be-for-someone-else-to-understand-my-code"><span class="toc-section-number">4.12.6</span> How easy would it be for someone else to understand my code?</a></li>
<li><a href="prog-style.html#is-my-code-easy-to-maintainchange"><span class="toc-section-number">4.12.7</span> Is my code easy to maintain/change?</a></li>
<li><a href="prog-style.html#the-styler-package"><span class="toc-section-number">4.12.8</span> The <code>styler</code> package</a></li>
</ul></li>
</ul></li>
<li><a href="data-wrangling.html#data-wrangling"><span class="toc-section-number">5</span> Data Wrangling</a>
<ul>
<li><a href="the-tidyverse.html#the-tidyverse"><span class="toc-section-number">5.1</span> The Tidyverse</a></li>
<li><a href="dw-basics.html#dw-basics"><span class="toc-section-number">5.2</span> Tidyverse Basics</a></li>
<li><a href="importing-data.html#importing-data"><span class="toc-section-number">5.3</span> Importing Data</a></li>
<li><a href="data-tibble.html#data-tibble"><span class="toc-section-number">5.4</span> The tibble</a></li>
<li><a href="tidy-data.html#tidy-data"><span class="toc-section-number">5.5</span> Tidy Data</a></li>
<li><a href="pipes.html#pipes"><span class="toc-section-number">5.6</span> pipes</a></li>
<li><a href="arranging-data.html#arranging-data"><span class="toc-section-number">5.7</span> Arranging Data</a>
<ul>
<li><a href="arranging-data.html#dplyrmutate---create-new-columns-using-other-columns"><span class="toc-section-number">5.7.1</span> <code>dplyr::mutate()</code> - Create new columns using other columns</a></li>
<li><a href="arranging-data.html#stringr---working-with-character-values"><span class="toc-section-number">5.7.2</span> <code>stringr</code> - Working with character values</a></li>
<li><a href="arranging-data.html#dplyrselect---subset-columns-by-name"><span class="toc-section-number">5.7.3</span> <code>dplyr::select()</code> - Subset Columns by Name</a></li>
<li><a href="arranging-data.html#dplyrfilter---pick-rows-out-of-a-data-set"><span class="toc-section-number">5.7.4</span> <code>dplyr::filter()</code> - Pick rows out of a data set</a></li>
<li><a href="arranging-data.html#dplyrarrange---order-rows-based-on-their-values"><span class="toc-section-number">5.7.5</span> <code>dplyr::arrange()</code> - Order rows based on their values</a></li>
<li><a href="arranging-data.html#putting-it-all-together"><span class="toc-section-number">5.7.6</span> Putting it all together</a></li>
</ul></li>
<li><a href="grouping-data.html#grouping-data"><span class="toc-section-number">5.8</span> Grouping Data</a></li>
<li><a href="rearranging-data.html#rearranging-data"><span class="toc-section-number">5.9</span> Rearranging Data</a></li>
<li><a href="relational-data.html#relational-data"><span class="toc-section-number">5.10</span> Relational Data</a>
<ul>
<li><a href="relational-data.html#dealing-with-multiple-matches"><span class="toc-section-number">5.10.1</span> Dealing with multiple matches</a></li>
</ul></li>
</ul></li>
<li><a href="data-science.html#data-science"><span class="toc-section-number">6</span> Data Science</a>
<ul>
<li><a href="data-modeling.html#data-modeling"><span class="toc-section-number">6.1</span> Data Modeling</a>
<ul>
<li><a href="data-modeling.html#a-worked-modeling-example"><span class="toc-section-number">6.1.1</span> A Worked Modeling Example</a></li>
<li><a href="data-modeling.html#data-summarization"><span class="toc-section-number">6.1.2</span> Data Summarization</a></li>
<li><a href="data-modeling.html#linear-models"><span class="toc-section-number">6.1.3</span> Linear Models</a></li>
<li><a href="data-modeling.html#flavors-of-linear-models"><span class="toc-section-number">6.1.4</span> Flavors of Linear Models</a></li>
<li><a href="data-modeling.html#assessing-model-accuracy-."><span class="toc-section-number">6.1.5</span> Assessing Model Accuracy .</a></li>
</ul></li>
<li><a href="statistical-distributions.html#statistical-distributions"><span class="toc-section-number">6.2</span> Statistical Distributions</a>
<ul>
<li><a href="statistical-distributions.html#random-variables"><span class="toc-section-number">6.2.1</span> Random Variables</a></li>
<li><a href="statistical-distributions.html#statistical-distribution-basics"><span class="toc-section-number">6.2.2</span> Statistical Distribution Basics</a></li>
<li><a href="statistical-distributions.html#distributions-in-r"><span class="toc-section-number">6.2.3</span> Distributions in R</a></li>
<li><a href="statistical-distributions.html#discrete-distributions"><span class="toc-section-number">6.2.4</span> Discrete Distributions</a></li>
<li><a href="statistical-distributions.html#continuous-distributions"><span class="toc-section-number">6.2.5</span> Continuous Distributions</a></li>
<li><a href="statistical-distributions.html#empirical-distributions"><span class="toc-section-number">6.2.6</span> Empirical Distributions</a></li>
</ul></li>
<li><a href="statistical-tests.html#statistical-tests"><span class="toc-section-number">6.3</span> Statistical Tests</a>
<ul>
<li><a href="statistical-tests.html#what-is-a-statistical-test"><span class="toc-section-number">6.3.1</span> What is a statistical test</a></li>
<li><a href="statistical-tests.html#common-statistical-tests"><span class="toc-section-number">6.3.2</span> Common Statistical Tests</a></li>
<li><a href="statistical-tests.html#choosing-a-test"><span class="toc-section-number">6.3.3</span> Choosing a Test</a></li>
<li><a href="statistical-tests.html#p-values"><span class="toc-section-number">6.3.4</span> P-values</a></li>
<li><a href="statistical-tests.html#multiple-hypothesis-testing-."><span class="toc-section-number">6.3.5</span> Multiple Hypothesis Testing .</a></li>
<li><a href="statistical-tests.html#statistical-power"><span class="toc-section-number">6.3.6</span> Statistical power</a></li>
</ul></li>
<li><a href="exploratory-data-analysis.html#exploratory-data-analysis"><span class="toc-section-number">6.4</span> Exploratory Data Analysis</a>
<ul>
<li><a href="exploratory-data-analysis.html#principal-component-analysis"><span class="toc-section-number">6.4.1</span> Principal Component Analysis</a></li>
<li><a href="exploratory-data-analysis.html#cluster-analysis"><span class="toc-section-number">6.4.2</span> Cluster Analysis</a></li>
<li><a href="exploratory-data-analysis.html#k-means-."><span class="toc-section-number">6.4.3</span> k-means .</a></li>
<li><a href="exploratory-data-analysis.html#others"><span class="toc-section-number">6.4.4</span> Others</a></li>
</ul></li>
<li><a href="network-analysis-..html#network-analysis-."><span class="toc-section-number">6.5</span> Network Analysis .</a>
<ul>
<li><a href="network-analysis-..html#basic-network-analysis"><span class="toc-section-number">6.5.1</span> Basic network analysis</a></li>
<li><a href="network-analysis-..html#create-your-own"><span class="toc-section-number">6.5.2</span> Create your own</a></li>
</ul></li>
</ul></li>
<li><a href="data-visualization.html#data-visualization"><span class="toc-section-number">7</span> Data Visualization</a>
<ul>
<li><a href="responsible-plotting.html#responsible-plotting"><span class="toc-section-number">7.1</span> Responsible Plotting</a>
<ul>
<li><a href="responsible-plotting.html#visualization-principles"><span class="toc-section-number">7.1.1</span> Visualization Principles</a></li>
<li><a href="responsible-plotting.html#human-visual-perception"><span class="toc-section-number">7.1.2</span> Human Visual Perception</a></li>
<li><a href="responsible-plotting.html#visual-encodings"><span class="toc-section-number">7.1.3</span> Visual Encodings</a></li>
<li><a href="responsible-plotting.html#some-opinionated-rules-of-thumb"><span class="toc-section-number">7.1.4</span> Some Opinionated Rules of Thumb</a></li>
</ul></li>
<li><a href="dv-gg.html#dv-gg"><span class="toc-section-number">7.2</span> Grammar of Graphics</a>
<ul>
<li><a href="dv-gg.html#ggplot-mechanics"><span class="toc-section-number">7.2.1</span> <code>ggplot</code> mechanics</a></li>
</ul></li>
<li><a href="plotting-one-dimension.html#plotting-one-dimension"><span class="toc-section-number">7.3</span> Plotting One Dimension</a>
<ul>
<li><a href="plotting-one-dimension.html#dv-bar"><span class="toc-section-number">7.3.1</span> Bar chart</a></li>
<li><a href="plotting-one-dimension.html#lollipop-plots"><span class="toc-section-number">7.3.2</span> Lollipop plots</a></li>
<li><a href="plotting-one-dimension.html#stacked-area-charts"><span class="toc-section-number">7.3.3</span> Stacked Area charts</a></li>
<li><a href="plotting-one-dimension.html#parallel-coordinate-plots"><span class="toc-section-number">7.3.4</span> Parallel Coordinate plots</a></li>
</ul></li>
<li><a href="visualizing-distributions.html#visualizing-distributions"><span class="toc-section-number">7.4</span> Visualizing Distributions</a>
<ul>
<li><a href="visualizing-distributions.html#histogram"><span class="toc-section-number">7.4.1</span> Histogram</a></li>
<li><a href="visualizing-distributions.html#density"><span class="toc-section-number">7.4.2</span> Density</a></li>
<li><a href="visualizing-distributions.html#boxplot"><span class="toc-section-number">7.4.3</span> Boxplot</a></li>
<li><a href="visualizing-distributions.html#violin-plot"><span class="toc-section-number">7.4.4</span> Violin plot</a></li>
<li><a href="visualizing-distributions.html#beeswarm-plot"><span class="toc-section-number">7.4.5</span> Beeswarm plot</a></li>
<li><a href="visualizing-distributions.html#ridgeline"><span class="toc-section-number">7.4.6</span> Ridgeline</a></li>
</ul></li>
<li><a href="plotting-two-or-more-dimensions.html#plotting-two-or-more-dimensions"><span class="toc-section-number">7.5</span> Plotting Two or More Dimensions</a>
<ul>
<li><a href="plotting-two-or-more-dimensions.html#scatter-plots"><span class="toc-section-number">7.5.1</span> Scatter Plots</a></li>
<li><a href="plotting-two-or-more-dimensions.html#bubble-plots"><span class="toc-section-number">7.5.2</span> Bubble Plots</a></li>
<li><a href="plotting-two-or-more-dimensions.html#connected-scatter-plots"><span class="toc-section-number">7.5.3</span> Connected Scatter Plots</a></li>
<li><a href="plotting-two-or-more-dimensions.html#line-plots"><span class="toc-section-number">7.5.4</span> Line Plots</a></li>
<li><a href="plotting-two-or-more-dimensions.html#parallel-coordinate-plots-1"><span class="toc-section-number">7.5.5</span> Parallel Coordinate Plots</a></li>
<li><a href="plotting-two-or-more-dimensions.html#heatmaps"><span class="toc-section-number">7.5.6</span> Heatmaps</a></li>
<li><a href="plotting-two-or-more-dimensions.html#how-to-use-heatmaps-responsibly"><span class="toc-section-number">7.5.7</span> How To Use Heatmaps Responsibly</a></li>
</ul></li>
<li><a href="other-kind-of-plots.html#other-kind-of-plots"><span class="toc-section-number">7.6</span> Other Kind of Plots</a>
<ul>
<li><a href="other-kind-of-plots.html#dendrograms"><span class="toc-section-number">7.6.1</span> Dendrograms</a></li>
<li><a href="other-kind-of-plots.html#chord-diagrams-and-circos-plots"><span class="toc-section-number">7.6.2</span> Chord Diagrams and Circos Plots</a></li>
</ul></li>
<li><a href="multiple-plots.html#multiple-plots"><span class="toc-section-number">7.7</span> Multiple Plots</a>
<ul>
<li><a href="multiple-plots.html#facet-wrapping"><span class="toc-section-number">7.7.1</span> Facet wrapping</a></li>
<li><a href="multiple-plots.html#multipanel-figures"><span class="toc-section-number">7.7.2</span> Multipanel Figures</a></li>
</ul></li>
<li><a href="publication-ready-plots.html#publication-ready-plots"><span class="toc-section-number">7.8</span> Publication Ready Plots</a>
<ul>
<li><a href="publication-ready-plots.html#ggplot-themes"><span class="toc-section-number">7.8.1</span> <code>ggplot</code> Themes</a></li>
<li><a href="publication-ready-plots.html#exporting-to-svg"><span class="toc-section-number">7.8.2</span> Exporting to SVG</a></li>
</ul></li>
<li><a href="network-visualization.html#network-visualization"><span class="toc-section-number">7.9</span> Network visualization</a>
<ul>
<li><a href="network-visualization.html#interactive-network-plots"><span class="toc-section-number">7.9.1</span> Interactive network plots</a></li>
</ul></li>
</ul></li>
<li><a href="biology-bioinformatics.html#biology-bioinformatics"><span class="toc-section-number">8</span> Biology &amp; Bioinformatics</a>
<ul>
<li><a href="bio-r.html#bio-r"><span class="toc-section-number">8.1</span> R in Biology</a></li>
<li><a href="biological-data-overview.html#biological-data-overview"><span class="toc-section-number">8.2</span> Biological Data Overview</a>
<ul>
<li><a href="biological-data-overview.html#bio-data-types"><span class="toc-section-number">8.2.1</span> Types of Biological Data</a></li>
<li><a href="biological-data-overview.html#csv-files"><span class="toc-section-number">8.2.2</span> CSV Files</a></li>
<li><a href="biological-data-overview.html#biological-data-is-not-tidy"><span class="toc-section-number">8.2.3</span> Biological data is NOT Tidy!</a></li>
</ul></li>
<li><a href="bio-bioconductor.html#bio-bioconductor"><span class="toc-section-number">8.3</span> Bioconductor</a></li>
<li><a href="microarrays.html#microarrays"><span class="toc-section-number">8.4</span> Microarrays</a></li>
<li><a href="high-throughput-sequencing.html#high-throughput-sequencing"><span class="toc-section-number">8.5</span> High Throughput Sequencing</a>
<ul>
<li><a href="high-throughput-sequencing.html#raw-hts-data"><span class="toc-section-number">8.5.1</span> Raw HTS Data</a></li>
<li><a href="high-throughput-sequencing.html#preliminary-hts-data-analysis"><span class="toc-section-number">8.5.2</span> Preliminary HTS Data Analysis</a></li>
<li><a href="high-throughput-sequencing.html#count-data"><span class="toc-section-number">8.5.3</span> Count Data</a></li>
</ul></li>
<li><a href="gene-identifiers.html#gene-identifiers"><span class="toc-section-number">8.6</span> Gene Identifiers</a>
<ul>
<li><a href="gene-identifiers.html#gene-identifier-systems"><span class="toc-section-number">8.6.1</span> Gene Identifier Systems</a></li>
<li><a href="gene-identifiers.html#mapping-between-identifier-systems"><span class="toc-section-number">8.6.2</span> Mapping Between Identifier Systems</a></li>
<li><a href="gene-identifiers.html#mapping-homologs"><span class="toc-section-number">8.6.3</span> Mapping Homologs</a></li>
</ul></li>
<li><a href="gene-expression.html#gene-expression"><span class="toc-section-number">8.7</span> Gene Expression</a>
<ul>
<li><a href="gene-expression.html#gene-expression-data-in-bioconductor"><span class="toc-section-number">8.7.1</span> Gene Expression Data in Bioconductor</a></li>
<li><a href="gene-expression.html#differential-expression-analysis"><span class="toc-section-number">8.7.2</span> Differential Expression Analysis</a></li>
<li><a href="gene-expression.html#microarray-gene-expression-data"><span class="toc-section-number">8.7.3</span> Microarray Gene Expression Data</a></li>
<li><a href="gene-expression.html#differential-expression-microarrays-limma"><span class="toc-section-number">8.7.4</span> Differential Expression: Microarrays (limma)</a></li>
<li><a href="gene-expression.html#rnaseq"><span class="toc-section-number">8.7.5</span> RNASeq</a></li>
<li><a href="gene-expression.html#rnaseq-gene-expression-data"><span class="toc-section-number">8.7.6</span> RNASeq Gene Expression Data</a></li>
<li><a href="gene-expression.html#filtering-counts"><span class="toc-section-number">8.7.7</span> Filtering Counts</a></li>
<li><a href="gene-expression.html#count-distributions"><span class="toc-section-number">8.7.8</span> Count Distributions</a></li>
<li><a href="gene-expression.html#differential-expression-rnaseq"><span class="toc-section-number">8.7.9</span> Differential Expression: RNASeq</a></li>
</ul></li>
<li><a href="gene-set-enrichment-analysis.html#gene-set-enrichment-analysis"><span class="toc-section-number">8.8</span> Gene Set Enrichment Analysis</a>
<ul>
<li><a href="gene-set-enrichment-analysis.html#gene-sets"><span class="toc-section-number">8.8.1</span> Gene Sets</a></li>
<li><a href="gene-set-enrichment-analysis.html#working-with-gene-sets-in-r"><span class="toc-section-number">8.8.2</span> Working with gene sets in R</a></li>
<li><a href="gene-set-enrichment-analysis.html#over-representation-analysis"><span class="toc-section-number">8.8.3</span> Over-representation Analysis</a></li>
<li><a href="gene-set-enrichment-analysis.html#rank-based-analysis"><span class="toc-section-number">8.8.4</span> Rank-based Analysis</a></li>
<li><a href="gene-set-enrichment-analysis.html#fgsea"><span class="toc-section-number">8.8.5</span> fgsea</a></li>
</ul></li>
<li><a href="biological-networks-..html#biological-networks-."><span class="toc-section-number">8.9</span> Biological Networks .</a>
<ul>
<li><a href="biological-networks-..html#biological-pathways-."><span class="toc-section-number">8.9.1</span> Biological Pathways .</a></li>
<li><a href="biological-networks-..html#gene-regulatory-networks-."><span class="toc-section-number">8.9.2</span> Gene Regulatory Networks .</a></li>
<li><a href="biological-networks-..html#protein-protein-interaction-network-."><span class="toc-section-number">8.9.3</span> Protein-Protein Interaction Network .</a></li>
<li><a href="biological-networks-..html#wgcna-."><span class="toc-section-number">8.9.4</span> WGCNA .</a></li>
</ul></li>
</ul></li>
<li><a href="engineering.html#engineering"><span class="toc-section-number">9</span> EngineeRing</a>
<ul>
<li><a href="unit-testing.html#unit-testing"><span class="toc-section-number">9.1</span> Unit Testing</a></li>
<li><a href="toolification.html#toolification"><span class="toc-section-number">9.2</span> Toolification</a>
<ul>
<li><a href="toolification.html#the-r-interpreter"><span class="toc-section-number">9.2.1</span> The R Interpreter</a></li>
<li><a href="toolification.html#rscript"><span class="toc-section-number">9.2.2</span> <code>Rscript</code></a></li>
<li><a href="toolification.html#commandargs"><span class="toc-section-number">9.2.3</span> <code>commandArgs()</code></a></li>
</ul></li>
<li><a href="parallel-processing.html#parallel-processing"><span class="toc-section-number">9.3</span> Parallel Processing</a>
<ul>
<li><a href="parallel-processing.html#brief-introduction-to-parallelization"><span class="toc-section-number">9.3.1</span> Brief Introduction to Parallelization</a></li>
<li><a href="parallel-processing.html#apply-and-friends-are-pleasingly-parallel"><span class="toc-section-number">9.3.2</span> <code>apply</code> and Friends Are Pleasingly Parallel</a></li>
<li><a href="parallel-processing.html#the-parallel-package"><span class="toc-section-number">9.3.3</span> The <code>parallel</code> package</a></li>
</ul></li>
<li><a href="object-orientation-in-r.html#object-orientation-in-r"><span class="toc-section-number">9.4</span> Object Orientation in R</a></li>
<li><a href="r-packages.html#r-packages"><span class="toc-section-number">9.5</span> R Packages</a></li>
</ul></li>
<li><a href="rshiny.html#rshiny"><span class="toc-section-number">10</span> RShiny</a>
<ul>
<li><a href="rshiny-introduction.html#rshiny-introduction"><span class="toc-section-number">10.1</span> RShiny Introduction</a>
<ul>
<li><a href="rshiny-introduction.html#how-does-a-web-page-work"><span class="toc-section-number">10.1.1</span> How does a web page work?</a></li>
</ul></li>
<li><a href="application-structure.html#application-structure"><span class="toc-section-number">10.2</span> Application Structure</a>
<ul>
<li><a href="application-structure.html#ui-design"><span class="toc-section-number">10.2.1</span> UI Design</a></li>
<li><a href="application-structure.html#server-functionality"><span class="toc-section-number">10.2.2</span> Server Functionality</a></li>
</ul></li>
<li><a href="reactivity.html#reactivity"><span class="toc-section-number">10.3</span> Reactivity</a></li>
<li><a href="publishing.html#publishing"><span class="toc-section-number">10.4</span> Publishing</a></li>
</ul></li>
<li><a href="communicating-with-r.html#communicating-with-r"><span class="toc-section-number">11</span> Communicating with R</a>
<ul>
<li><a href="rmarkdown-knitr.html#rmarkdown-knitr"><span class="toc-section-number">11.1</span> RMarkdown &amp; knitr</a></li>
<li><a href="bookdown.html#bookdown"><span class="toc-section-number">11.2</span> bookdown</a></li>
</ul></li>
<li><a href="contribution-guide.html#contribution-guide"><span class="toc-section-number">12</span> Contribution Guide</a>
<ul>
<li><a href="custom-blocks.html#custom-blocks"><span class="toc-section-number">12.1</span> Custom blocks</a></li>
</ul></li>
<li><a href="assignments.html#assignments">Assignments</a>
<ul>
<li><a href="assignment-format.html#assignment-format">Assignment Format</a></li>
<li><a href="starting-an-assignment.html#starting-an-assignment">Starting an Assignment</a>
<ul>
<li><a href="starting-an-assignment.html#starting-your-own-git-repository-and-cloning-it">Starting your own <code>git</code> repository and cloning it</a></li>
<li><a href="starting-an-assignment.html#contents">Contents</a></li>
<li><a href="starting-an-assignment.html#github-tutorial">GitHub tutorial</a></li>
<li><a href="starting-an-assignment.html#committing-and-pushing-with-r-and-without">Committing and Pushing with R (and without)</a></li>
<li><a href="starting-an-assignment.html#using-rstudio">Using RStudio</a></li>
<li><a href="starting-an-assignment.html#using-the-command-line">Using the command line</a></li>
</ul></li>
<li><a href="assignment-1.html#assignment-1">Assignment 1</a>
<ul>
<li><a href="assignment-1.html#problem-statement">Problem Statement</a></li>
<li><a href="assignment-1.html#learning-objectives-and-skill-list">Learning Objectives and Skill List</a></li>
<li><a href="assignment-1.html#instructions">Instructions</a></li>
<li><a href="assignment-1.html#hints">Hints</a></li>
</ul></li>
<li><a href="assignment-2.html#assignment-2">Assignment 2</a>
<ul>
<li><a href="assignment-2.html#problem-statement-1">Problem Statement</a></li>
<li><a href="assignment-2.html#learning-objectives">Learning Objectives</a></li>
<li><a href="assignment-2.html#skill-list">Skill List</a></li>
<li><a href="assignment-2.html#instructions-1">Instructions</a></li>
<li><a href="assignment-2.html#function-details">Function Details</a></li>
</ul></li>
<li><a href="assignment-3.html#assignment-3">Assignment 3</a>
<ul>
<li><a href="assignment-3.html#problem-statement-2">Problem Statement</a></li>
<li><a href="assignment-3.html#learning-objectives-1">Learning Objectives</a></li>
<li><a href="assignment-3.html#skill-list-1">Skill List</a></li>
<li><a href="assignment-3.html#background-on-microarrays">Background on Microarrays</a></li>
<li><a href="assignment-3.html#background-on-principal-component-analysis">Background on Principal Component Analysis</a></li>
<li><a href="assignment-3.html#marisa-et-al.-gene-expression-classification-of-colon-cancer-into-molecular-subtypes-characterization-validation-and-prognostic-value.-plos-medicine-may-2013.-pmid-23700391">Marisa et al. Gene Expression Classification of Colon Cancer into Molecular Subtypes: Characterization, Validation, and Prognostic Value. PLoS Medicine, May 2013. PMID: 23700391</a></li>
<li><a href="assignment-3.html#scaling-data-using-r-scale">Scaling data using R <code>scale()</code></a></li>
<li><a href="assignment-3.html#proportion-of-variance-explained">Proportion of variance explained</a></li>
<li><a href="assignment-3.html#plotting-and-visualization-of-pca">Plotting and visualization of PCA</a></li>
<li><a href="assignment-3.html#hierarchical-clustering-and-heatmaps">Hierarchical Clustering and Heatmaps</a></li>
<li><a href="assignment-3.html#references">References</a></li>
</ul></li>
<li><a href="assignment-4.html#assignment-4">Assignment 4</a>
<ul>
<li><a href="assignment-4.html#problem-statement-3">Problem Statement</a></li>
<li><a href="assignment-4.html#learning-objectives-2">Learning Objectives</a></li>
<li><a href="assignment-4.html#skill-list-2">Skill List</a></li>
<li><a href="assignment-4.html#instructions-2">Instructions</a></li>
<li><a href="assignment-4.html#deliverables">Deliverables</a></li>
<li><a href="assignment-4.html#function-details-1">Function Details</a></li>
<li><a href="assignment-4.html#hints-1">Hints</a></li>
</ul></li>
<li><a href="assignment-5.html#assignment-5">Assignment 5</a>
<ul>
<li><a href="assignment-5.html#problem-statement-4">Problem Statement</a></li>
<li><a href="assignment-5.html#learning-objectives-3">Learning Objectives</a></li>
<li><a href="assignment-5.html#skill-list-3">Skill List</a></li>
<li><a href="assignment-5.html#deseq2-background">DESeq2 Background</a></li>
<li><a href="assignment-5.html#generating-a-counts-matrix">Generating a counts matrix</a></li>
<li><a href="assignment-5.html#prefiltering-counts-matrix">Prefiltering Counts matrix</a></li>
<li><a href="assignment-5.html#median-of-ratios-normalization">Median-of-ratios normalization</a></li>
<li><a href="assignment-5.html#deseq2-preparation">DESeq2 preparation</a></li>
<li><a href="assignment-5.html#omeara-et-al.-transcriptional-reversion-of-cardiac-myocyte-fate-during-mammalian-cardiac-regeneration.-circ-res.-feb-2015.-pmid-25477501l">O’Meara et al. Transcriptional Reversion of Cardiac Myocyte Fate During Mammalian Cardiac Regeneration. Circ Res. Feb 2015. PMID: 25477501l</a></li>
<li><a href="assignment-5.html#reading-and-subsetting-the-data-from-verse_counts.tsv-and-sample_metadata.csv">1. Reading and subsetting the data from verse_counts.tsv and sample_metadata.csv</a></li>
<li><a href="assignment-5.html#running-deseq2">2. Running DESeq2</a></li>
<li><a href="assignment-5.html#annotating-results-to-construct-a-labeled-volcano-plot">3. Annotating results to construct a labeled volcano plot</a></li>
<li><a href="assignment-5.html#diagnostic-plot-of-the-raw-p-values-for-all-genes">4. Diagnostic plot of the raw p-values for all genes</a></li>
<li><a href="assignment-5.html#plotting-the-logfoldchanges-for-differentially-expressed-genes">5. Plotting the LogFoldChanges for differentially expressed genes</a></li>
<li><a href="assignment-5.html#the-choice-of-fdr-cutoff-depends-on-cost">The choice of FDR cutoff depends on cost</a></li>
<li><a href="assignment-5.html#plotting-the-normalized-counts-of-differentially-expressed-genes">6. Plotting the normalized counts of differentially expressed genes</a></li>
<li><a href="assignment-5.html#volcano-plot-to-visualize-differential-expression-results">7. Volcano Plot to visualize differential expression results</a></li>
<li><a href="assignment-5.html#running-fgsea-vignette">8. Running fgsea <span>vignette</span></a></li>
<li><a href="assignment-5.html#plotting-the-top-ten-positive-nes-and-top-ten-negative-nes-pathways">9. Plotting the top ten positive NES and top ten negative NES pathways</a></li>
<li><a href="assignment-5.html#references-1">References</a></li>
</ul></li>
<li><a href="assignment-6.html#assignment-6">Assignment 6</a>
<ul>
<li><a href="assignment-6.html#problem-statement-5">Problem Statement</a></li>
<li><a href="assignment-6.html#learning-objectives-4">Learning Objectives</a></li>
<li><a href="assignment-6.html#skill-list-4">Skill List</a></li>
<li><a href="assignment-6.html#instructions-3">Instructions</a></li>
<li><a href="assignment-6.html#function-details-2">Function Details</a></li>
</ul></li>
<li><a href="assignment-7.html#assignment-7">Assignment 7</a>
<ul>
<li><a href="assignment-7.html#problem-statement-6">Problem Statement</a></li>
<li><a href="assignment-7.html#learning-objectives-5">Learning Objectives</a></li>
<li><a href="assignment-7.html#skill-list-5">Skill List</a></li>
<li><a href="assignment-7.html#instructions-4">Instructions</a></li>
<li><a href="assignment-7.html#function-details-3">“Function” Details</a></li>
</ul></li>
</ul></li>
<li class="appendix"><span><b>Appendix</b></span></li>
<li><a href="class-outlines.html#class-outlines"><span class="toc-section-number">A</span> Class Outlines</a>
<ul>
<li><a href="week-1.html#week-1"><span class="toc-section-number">A.1</span> Week 1</a></li>
<li><a href="week-2.html#week-2"><span class="toc-section-number">A.2</span> Week 2</a></li>
<li><a href="week-3.html#week-3"><span class="toc-section-number">A.3</span> Week 3</a></li>
<li><a href="week-4.html#week-4"><span class="toc-section-number">A.4</span> Week 4</a></li>
<li><a href="week-5.html#week-5"><span class="toc-section-number">A.5</span> Week 5</a></li>
<li><a href="week-6.html#week-6"><span class="toc-section-number">A.6</span> Week 6</a></li>
<li><a href="week-7.html#week-7"><span class="toc-section-number">A.7</span> Week 7</a></li>
<li><a href="week-8.html#week-8"><span class="toc-section-number">A.8</span> Week 8</a></li>
<li><a href="week-9.html#week-9"><span class="toc-section-number">A.9</span> Week 9</a></li>
<li><a href="week-10.html#week-10"><span class="toc-section-number">A.10</span> Week 10</a></li>
<li><a href="week-11.html#week-11"><span class="toc-section-number">A.11</span> Week 11</a></li>
<li><a href="week-12.html#week-12"><span class="toc-section-number">A.12</span> Week 12</a></li>
<li><a href="week-13.html#week-13"><span class="toc-section-number">A.13</span> Week 13</a></li>
<li><a href="week-14.html#week-14"><span class="toc-section-number">A.14</span> Week 14</a></li>
</ul></li>
</ul>
</div>
</div>
</div>
<div class="row">
<div class="col-sm-12">
<div id="visualizing-distributions" class="section level2" number="7.4">
<h2><span class="header-section-number">7.4</span> Visualizing Distributions</h2>
<p>The distribution is one of the most important properties of a set of numbers.
A distribution describes the general “shape” of the numbers, i.e. what is the
relative frequency of the values, or ranges of values, within the data.
Understanding the distribution of a data set is critical when choosing methods
to apply, since many methods are only appropriate when data is distributed in
certain ways, e.g. <a href="https://en.wikipedia.org/wiki/Linear_regression">linear
regression</a> assumes the
response variable is <a href="https://en.wikipedia.org/wiki/Normal_distribution">normally
distributed</a>, otherwise the
result of the model cannot be interpreted properly. Often, we don’t know how our
data are distributed when we obtain it and so we must examine the distribution
empirically. The visualizations in this section are all used for the purpose of
depicting the distribution of a set of numbers.</p>
<div id="histogram" class="section level3" number="7.4.1">
<h3><span class="header-section-number">7.4.1</span> Histogram</h3>
<p>The most common way to plot the distribution of a 1-D set of data is the
<a href="https://en.wikipedia.org/wiki/Histogram">histogram</a>. The histogram divides up
the range of a dataset from minimum to maximum into bins usually of the same
width and tabulates the number of values that fall within each bin. Below is a
histogram of our <code>age_at_death</code> measurement for all samples:</p>
<div class="sourceCode" id="cb282"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb282-1"><a href="visualizing-distributions.html#cb282-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(ad_metadata) <span class="sc">+</span></span>
<span id="cb282-2"><a href="visualizing-distributions.html#cb282-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">mapping=</span><span class="fu">aes</span>(<span class="at">x=</span>age_at_death))</span></code></pre></div>
<p><img src="main_files/unnamed-chunk-107-1.png" width="672" /></p>
<p>Note that the histogram does not look very complete, because there are only 20
values in our data. We can mitigate this somewhat by increasing the number of
bins the data range is divided into:</p>
<div class="sourceCode" id="cb283"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb283-1"><a href="visualizing-distributions.html#cb283-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(ad_metadata) <span class="sc">+</span></span>
<span id="cb283-2"><a href="visualizing-distributions.html#cb283-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">mapping=</span><span class="fu">aes</span>(<span class="at">x=</span>age_at_death),<span class="at">bins=</span><span class="dv">10</span>)</span></code></pre></div>
<p><img src="main_files/unnamed-chunk-108-1.png" width="672" /></p>
<p>This is a little bit better, but there are still some bins (76-79, 84-87) that
have no values. Compare this to the following synthetic dataset of 1000
normally distributed values:</p>
<div class="sourceCode" id="cb284"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb284-1"><a href="visualizing-distributions.html#cb284-1" aria-hidden="true" tabindex="-1"></a><span class="fu">tibble</span>(</span>
<span id="cb284-2"><a href="visualizing-distributions.html#cb284-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">x=</span><span class="fu">rnorm</span>(<span class="dv">1000</span>)</span>
<span id="cb284-3"><a href="visualizing-distributions.html#cb284-3" aria-hidden="true" tabindex="-1"></a>) <span class="sc">%&gt;%</span></span>
<span id="cb284-4"><a href="visualizing-distributions.html#cb284-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb284-5"><a href="visualizing-distributions.html#cb284-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="fu">aes</span>(<span class="at">x=</span>x))</span></code></pre></div>
<p><img src="main_files/unnamed-chunk-109-1.png" width="672" /></p>
<p>For distributions with a small number of samples, histograms might not be the
best visualization. We will continue with synthetic normally distributed dataset
for the remaining examples.</p>
<p>ggplot allows you to easily plot multiple distributions on the same plot:</p>
<div class="sourceCode" id="cb285"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb285-1"><a href="visualizing-distributions.html#cb285-1" aria-hidden="true" tabindex="-1"></a><span class="fu">tibble</span>(</span>
<span id="cb285-2"><a href="visualizing-distributions.html#cb285-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">x=</span><span class="fu">c</span>(<span class="fu">rnorm</span>(<span class="dv">1000</span>),<span class="fu">rnorm</span>(<span class="dv">1000</span>,<span class="at">mean=</span><span class="dv">4</span>)),</span>
<span id="cb285-3"><a href="visualizing-distributions.html#cb285-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">type=</span><span class="fu">c</span>(<span class="fu">rep</span>(<span class="st">&#39;A&#39;</span>,<span class="dv">1000</span>),<span class="fu">rep</span>(<span class="st">&#39;B&#39;</span>,<span class="dv">1000</span>))</span>
<span id="cb285-4"><a href="visualizing-distributions.html#cb285-4" aria-hidden="true" tabindex="-1"></a>) <span class="sc">%&gt;%</span></span>
<span id="cb285-5"><a href="visualizing-distributions.html#cb285-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x=</span>x,<span class="at">fill=</span>type)) <span class="sc">+</span></span>
<span id="cb285-6"><a href="visualizing-distributions.html#cb285-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">bins=</span><span class="dv">30</span>, <span class="at">alpha=</span><span class="fl">0.6</span>, <span class="at">position=</span><span class="st">&quot;identity&quot;</span>)</span></code></pre></div>
<p><img src="main_files/unnamed-chunk-110-1.png" width="672" /></p>
<p>The <code>alpha=0.6, position="identity"</code> arguments makes the bars partially
transparent so you can see the overlap more clearly.</p>
<div class="box readmore">
<ul>
<li><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html?q=histogram">ggplot2 geom_histogram reference</a></li>
<li><a href="https://www.r-graph-gallery.com/histogram.html">R Graph Gallery - histograms</a></li>
</ul>
</div>
</div>
<div id="density" class="section level3" number="7.4.2">
<h3><span class="header-section-number">7.4.2</span> Density</h3>
<p>Another way to describe a distribution is with a <a href="https://www.r-graph-gallery.com/density-plot.html">density
plot</a>. Instead of binning the
values into intervals and drawing bars with height proportional to the number of
values in each bin, a density plot draws a smoothly interpolated line that
approximates the distribution instead. A key difference between a histogram and
a density plot is the density plot is always normalized so the integral under
the curve is approximately 1, whereas a histogram may be either counts or, if
the counts in each bin are divided by the total number of data points, a
proportion.</p>
<p>Compare the histogram and density plots
of the <code>age_at_death</code> variable from our example tibble:</p>
<div class="sourceCode" id="cb286"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb286-1"><a href="visualizing-distributions.html#cb286-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(patchwork)</span>
<span id="cb286-2"><a href="visualizing-distributions.html#cb286-2" aria-hidden="true" tabindex="-1"></a>hist_g <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(ad_metadata) <span class="sc">+</span></span>
<span id="cb286-3"><a href="visualizing-distributions.html#cb286-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">mapping=</span><span class="fu">aes</span>(<span class="at">x=</span>age_at_death),<span class="at">bins=</span><span class="dv">30</span>)</span>
<span id="cb286-4"><a href="visualizing-distributions.html#cb286-4" aria-hidden="true" tabindex="-1"></a>density_g <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(ad_metadata) <span class="sc">+</span></span>
<span id="cb286-5"><a href="visualizing-distributions.html#cb286-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_density</span>(<span class="at">mapping=</span><span class="fu">aes</span>(<span class="at">x=</span>age_at_death),<span class="at">fill=</span><span class="st">&quot;#c9a13daa&quot;</span>)</span>
<span id="cb286-6"><a href="visualizing-distributions.html#cb286-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb286-7"><a href="visualizing-distributions.html#cb286-7" aria-hidden="true" tabindex="-1"></a>hist_g <span class="sc">|</span> density_g</span></code></pre></div>
<p><img src="main_files/unnamed-chunk-111-1.png" width="672" /></p>
<p>Notice the overall shape of the two distributions is similar, with the highest
values in both around age 77. The density plot is a smoother representation of
a histogram, but its accuracy is still highly sensitive to the number of
measurements used to construct it. Compare the histogram and density plots of
two sets of 1000 normally distributed samples with different means:</p>
<div class="sourceCode" id="cb287"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb287-1"><a href="visualizing-distributions.html#cb287-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(patchwork)</span>
<span id="cb287-2"><a href="visualizing-distributions.html#cb287-2" aria-hidden="true" tabindex="-1"></a>normal_samples <span class="ot">&lt;-</span> <span class="fu">tibble</span>(</span>
<span id="cb287-3"><a href="visualizing-distributions.html#cb287-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">x=</span><span class="fu">c</span>(<span class="fu">rnorm</span>(<span class="dv">1000</span>),<span class="fu">rnorm</span>(<span class="dv">1000</span>,<span class="at">mean=</span><span class="dv">4</span>)),</span>
<span id="cb287-4"><a href="visualizing-distributions.html#cb287-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">type=</span><span class="fu">c</span>(<span class="fu">rep</span>(<span class="st">&#39;A&#39;</span>,<span class="dv">1000</span>),<span class="fu">rep</span>(<span class="st">&#39;B&#39;</span>,<span class="dv">1000</span>))</span>
<span id="cb287-5"><a href="visualizing-distributions.html#cb287-5" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb287-6"><a href="visualizing-distributions.html#cb287-6" aria-hidden="true" tabindex="-1"></a>hist_g <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(normal_samples) <span class="sc">+</span></span>
<span id="cb287-7"><a href="visualizing-distributions.html#cb287-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(</span>
<span id="cb287-8"><a href="visualizing-distributions.html#cb287-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">mapping=</span><span class="fu">aes</span>(<span class="at">x=</span>x,<span class="at">fill=</span>type),</span>
<span id="cb287-9"><a href="visualizing-distributions.html#cb287-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">alpha=</span><span class="fl">0.6</span>,</span>
<span id="cb287-10"><a href="visualizing-distributions.html#cb287-10" aria-hidden="true" tabindex="-1"></a>    <span class="at">position=</span><span class="st">&quot;identity&quot;</span>,</span>
<span id="cb287-11"><a href="visualizing-distributions.html#cb287-11" aria-hidden="true" tabindex="-1"></a>    <span class="at">bins=</span><span class="dv">30</span></span>
<span id="cb287-12"><a href="visualizing-distributions.html#cb287-12" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb287-13"><a href="visualizing-distributions.html#cb287-13" aria-hidden="true" tabindex="-1"></a>density_g <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(normal_samples) <span class="sc">+</span></span>
<span id="cb287-14"><a href="visualizing-distributions.html#cb287-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_density</span>(</span>
<span id="cb287-15"><a href="visualizing-distributions.html#cb287-15" aria-hidden="true" tabindex="-1"></a>    <span class="at">mapping=</span><span class="fu">aes</span>(<span class="at">x=</span>x,<span class="at">fill=</span>type),</span>
<span id="cb287-16"><a href="visualizing-distributions.html#cb287-16" aria-hidden="true" tabindex="-1"></a>    <span class="at">alpha=</span><span class="fl">0.6</span>,</span>
<span id="cb287-17"><a href="visualizing-distributions.html#cb287-17" aria-hidden="true" tabindex="-1"></a>    <span class="at">position=</span><span class="st">&quot;identity&quot;</span></span>
<span id="cb287-18"><a href="visualizing-distributions.html#cb287-18" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb287-19"><a href="visualizing-distributions.html#cb287-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb287-20"><a href="visualizing-distributions.html#cb287-20" aria-hidden="true" tabindex="-1"></a>hist_g <span class="sc">|</span> density_g</span></code></pre></div>
<p><img src="main_files/unnamed-chunk-112-1.png" width="672" /></p>
<p>Again the two types of plots depict similar distributions, although they are
different enough to possibly suggest different interpretations. In general,
density plots might be preferable over histograms if the data are noisy or
sparse in that they produce cleaner plots, but potentially at the expense of
accuracy when the number of samples is low.</p>
<div class="box readmore">
<ul>
<li><a href="https://ggplot2.tidyverse.org/reference/geom_density.html?q=density">ggplot2 geom_density reference</a></li>
<li><a href="https://www.r-graph-gallery.com/density-plot.html">R Graph Gallery - density plots</a></li>
</ul>
</div>
</div>
<div id="boxplot" class="section level3" number="7.4.3">
<h3><span class="header-section-number">7.4.3</span> Boxplot</h3>
<p>Box plots, or box and whisker plots, are extremely common when used to describe
distributions. Below is a boxplot of age at death divided by condition:</p>
<div class="sourceCode" id="cb288"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb288-1"><a href="visualizing-distributions.html#cb288-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(ad_metadata) <span class="sc">+</span></span>
<span id="cb288-2"><a href="visualizing-distributions.html#cb288-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>(<span class="at">mapping=</span><span class="fu">aes</span>(<span class="at">x=</span>condition,<span class="at">y=</span>age_at_death))</span></code></pre></div>
<p><img src="main_files/unnamed-chunk-113-1.png" width="672" /></p>
<p>Boxplots are drawn assuming the data are
<a href="https://en.wikipedia.org/wiki/Unimodality">unimodal</a> (i.e. shaped like a hill,
possibly slanted to one side or the other), where the extents of the box
represent the 1st and 3rd quartile of the data, the central line is the median,
the whiskers are drawn as 1.5 times the value outside the 1st and 3rd quartiles.
Sometimes individual values more extreme than the whiskers are drawn
individually to identify them as outliers.</p>
<div class="figure">
<img src="https://upload.wikimedia.org/wikipedia/commons/1/1a/Boxplot_vs_PDF.svg" alt="" />
<p class="caption">Boxplot anatomy. IQR stands for “inner quartile range,” the distance between
the 1st and 3rd quartile - Wikipedia</p>
</div>
<p>However, boxplots have some significant shortcomings. Primarily, the rectangle
of the inner quartile range does not describe the actual distribution of the
samples within it. Although the median can give a sense of skewness, if the
data are not unimodal this may be misleading. Consider the following
distributions plotted as boxplots or as violin plots (described in the next
section):</p>
<div class="sourceCode" id="cb289"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb289-1"><a href="visualizing-distributions.html#cb289-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(patchwork)</span>
<span id="cb289-2"><a href="visualizing-distributions.html#cb289-2" aria-hidden="true" tabindex="-1"></a>normal_samples <span class="ot">&lt;-</span> <span class="fu">tibble</span>(</span>
<span id="cb289-3"><a href="visualizing-distributions.html#cb289-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">x=</span><span class="fu">c</span>(<span class="fu">rnorm</span>(<span class="dv">1000</span>),<span class="fu">rnorm</span>(<span class="dv">1000</span>,<span class="dv">4</span>),<span class="fu">rnorm</span>(<span class="dv">1000</span>,<span class="dv">2</span>,<span class="dv">3</span>)),</span>
<span id="cb289-4"><a href="visualizing-distributions.html#cb289-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">type=</span><span class="fu">c</span>(<span class="fu">rep</span>(<span class="st">&#39;A&#39;</span>,<span class="dv">2000</span>),<span class="fu">rep</span>(<span class="st">&#39;B&#39;</span>,<span class="dv">1000</span>))</span>
<span id="cb289-5"><a href="visualizing-distributions.html#cb289-5" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb289-6"><a href="visualizing-distributions.html#cb289-6" aria-hidden="true" tabindex="-1"></a>g <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(normal_samples, <span class="fu">aes</span>(<span class="at">x=</span>type,<span class="at">y=</span>x,<span class="at">fill=</span>type))</span>
<span id="cb289-7"><a href="visualizing-distributions.html#cb289-7" aria-hidden="true" tabindex="-1"></a>boxplot_g <span class="ot">&lt;-</span> g <span class="sc">+</span>  <span class="fu">geom_boxplot</span>()</span>
<span id="cb289-8"><a href="visualizing-distributions.html#cb289-8" aria-hidden="true" tabindex="-1"></a>violin_g <span class="ot">&lt;-</span> g <span class="sc">+</span> <span class="fu">geom_violin</span>()</span>
<span id="cb289-9"><a href="visualizing-distributions.html#cb289-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb289-10"><a href="visualizing-distributions.html#cb289-10" aria-hidden="true" tabindex="-1"></a>boxplot_g <span class="sc">|</span> violin_g</span></code></pre></div>
<p><img src="main_files/unnamed-chunk-114-1.png" width="672" /></p>
<p>The two distributions look almost identical in the boxplot figure; however they
are dramatically different when visualized using a method like a violin plot
where the contours of the entire distribution are depicted. Unless you are
certain that your data are unimodal, one of the other distribution visualization
methods in this section will likely more accurately depict your data than a
boxplot.</p>
<div class="box readmore">
<ul>
<li><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html">ggplot2 geom_boxplot reference</a></li>
<li><a href="https://www.r-graph-gallery.com/boxplot.html">R Graph Gallery - boxplots</a></li>
</ul>
</div>
</div>
<div id="violin-plot" class="section level3" number="7.4.4">
<h3><span class="header-section-number">7.4.4</span> Violin plot</h3>
<p>As seen in the last section, a <a href="https://ggplot2.tidyverse.org/reference/geom_violin.html">violin
plot</a> is another way
to depict a distribution by producing a shape where the width is proportional to
the value along the x or y axis, depending on orientation. The “violin” shape is
similar in principle to a histogram or a density plot, in that it describes the
contour of all the data in the distribution, not just the quantiles and extents,
as in a box plot. Below is a violin plot of the <code>tau</code> measures from our example
tibble:</p>
<div class="sourceCode" id="cb290"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb290-1"><a href="visualizing-distributions.html#cb290-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(ad_metadata) <span class="sc">+</span></span>
<span id="cb290-2"><a href="visualizing-distributions.html#cb290-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_violin</span>(<span class="fu">aes</span>(<span class="at">x=</span>condition,<span class="at">y=</span>tau,<span class="at">fill=</span>condition))</span></code></pre></div>
<p><img src="main_files/unnamed-chunk-115-1.png" width="672" /></p>
<p>The violin plot is both more and less descriptive than a boxplot; it does depict
the entire distribution of the data, but also doesn’t include features like
median by default.</p>
<div class="box readmore">
<ul>
<li><a href="https://ggplot2.tidyverse.org/reference/geom_violin.html">ggplot2 geom_violin reference</a></li>
<li><a href="https://www.r-graph-gallery.com/violin.html">R Graph Gallery - violin plots</a></li>
</ul>
</div>
</div>
<div id="beeswarm-plot" class="section level3" number="7.4.5">
<h3><span class="header-section-number">7.4.5</span> Beeswarm plot</h3>
<p>The <a href="https://r-charts.com/distribution/ggbeeswarm/">beeswarm plot</a> is similar to
a violin plot, but instead of plotting the contours of the data, it plots the
data itself as points like in a scatter plot. The individual values of the
distribution are organized vertically and spaced such that the points don’t
overlap. In this plot, the distribution of age at death is plotted for each kind
of sample and the markers are colored by the amount of tau:</p>
<div class="sourceCode" id="cb291"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb291-1"><a href="visualizing-distributions.html#cb291-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggbeeswarm)</span>
<span id="cb291-2"><a href="visualizing-distributions.html#cb291-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(ad_metadata) <span class="sc">+</span></span>
<span id="cb291-3"><a href="visualizing-distributions.html#cb291-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_beeswarm</span>(<span class="fu">aes</span>(<span class="at">x=</span>condition,<span class="at">y=</span>age_at_death,<span class="at">color=</span>condition),<span class="at">cex=</span><span class="dv">2</span>,<span class="at">size=</span><span class="dv">2</span>)</span></code></pre></div>
<p><img src="main_files/unnamed-chunk-116-1.png" width="672" /></p>
<p>We may not have noticed before that our AD samples have a big gap in ages
between 74 and 81; since the beeswarm plot displays all the data, we can see it
easily here.</p>
<p>Beeswarm plots are typically only useful when the number of values is within a
range; not too many and not too few. The example above is close to having too
few values per group for this plot to be useful, but consider the following with
too many samples:</p>
<div class="sourceCode" id="cb292"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb292-1"><a href="visualizing-distributions.html#cb292-1" aria-hidden="true" tabindex="-1"></a>normal_samples <span class="ot">&lt;-</span> <span class="fu">tibble</span>(</span>
<span id="cb292-2"><a href="visualizing-distributions.html#cb292-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">x=</span><span class="fu">c</span>(<span class="fu">rnorm</span>(<span class="dv">1000</span>),<span class="fu">rnorm</span>(<span class="dv">1000</span>,<span class="dv">4</span>),<span class="fu">rnorm</span>(<span class="dv">1000</span>,<span class="dv">2</span>,<span class="dv">3</span>)),</span>
<span id="cb292-3"><a href="visualizing-distributions.html#cb292-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">type=</span><span class="fu">c</span>(<span class="fu">rep</span>(<span class="st">&#39;A&#39;</span>,<span class="dv">2000</span>),<span class="fu">rep</span>(<span class="st">&#39;B&#39;</span>,<span class="dv">1000</span>))</span>
<span id="cb292-4"><a href="visualizing-distributions.html#cb292-4" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb292-5"><a href="visualizing-distributions.html#cb292-5" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(normal_samples, <span class="fu">aes</span>(<span class="at">x=</span>type,<span class="at">y=</span>x,<span class="at">color=</span>type)) <span class="sc">+</span></span>
<span id="cb292-6"><a href="visualizing-distributions.html#cb292-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_beeswarm</span>()</span></code></pre></div>
<p><img src="main_files/unnamed-chunk-117-1.png" width="672" /></p>
<p>This plot likely has too many samples to be the right choice (it’s also ugly),
but it does give an idea of the distribution of the data.</p>
<p>In the previous examples the markers for each group also determined the color of
the group. This makes the chart a bit easier to read and more pleasing to the
eye, but is technically redundant. You can use however profitably however to
color markers by some other value that might be of interest. Consider this final
example where markers are colored by another randomly generated variable:</p>
<div class="sourceCode" id="cb293"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb293-1"><a href="visualizing-distributions.html#cb293-1" aria-hidden="true" tabindex="-1"></a>normal_samples <span class="ot">&lt;-</span> <span class="fu">tibble</span>(</span>
<span id="cb293-2"><a href="visualizing-distributions.html#cb293-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">x=</span><span class="fu">c</span>(<span class="fu">rnorm</span>(<span class="dv">100</span>),<span class="fu">rnorm</span>(<span class="dv">100</span>,<span class="dv">4</span>),<span class="fu">rnorm</span>(<span class="dv">100</span>,<span class="dv">2</span>,<span class="dv">3</span>)),</span>
<span id="cb293-3"><a href="visualizing-distributions.html#cb293-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">type=</span><span class="fu">c</span>(<span class="fu">rep</span>(<span class="st">&#39;A&#39;</span>,<span class="dv">200</span>),<span class="fu">rep</span>(<span class="st">&#39;B&#39;</span>,<span class="dv">100</span>)),</span>
<span id="cb293-4"><a href="visualizing-distributions.html#cb293-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">category=</span><span class="fu">sample</span>(<span class="fu">c</span>(<span class="st">&#39;healthy&#39;</span>,<span class="st">&#39;disease&#39;</span>),<span class="dv">300</span>,<span class="at">replace=</span><span class="cn">TRUE</span>)</span>
<span id="cb293-5"><a href="visualizing-distributions.html#cb293-5" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb293-6"><a href="visualizing-distributions.html#cb293-6" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(normal_samples, <span class="fu">aes</span>(<span class="at">x=</span>type,<span class="at">y=</span>x,<span class="at">color=</span>category)) <span class="sc">+</span></span>
<span id="cb293-7"><a href="visualizing-distributions.html#cb293-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_beeswarm</span>()</span></code></pre></div>
<p><img src="main_files/unnamed-chunk-118-1.png" width="672" /></p>
<p>We are now effectively visualizing three dimensions which may provide insight
into the data.</p>
<div class="box readmore">
<ul>
<li><a href="https://r-charts.com/distribution/ggbeeswarm/">beeswarm package reference</a></li>
</ul>
</div>
</div>
<div id="ridgeline" class="section level3" number="7.4.6">
<h3><span class="header-section-number">7.4.6</span> Ridgeline</h3>
<p>If you have many non-trivial distributions that you would like the user to
compare, a good option is a <a href="https://www.r-graph-gallery.com/294-basic-ridgeline-plot.html">ridgeline
chart</a>. The
ridgeline plot is simply multiple density plots drawn for different variables
within the same plot. Like the beeswarm plot, ridgeline plots are provided by
another package outside ggplot2.</p>
<div class="sourceCode" id="cb294"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb294-1"><a href="visualizing-distributions.html#cb294-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggridges)</span>
<span id="cb294-2"><a href="visualizing-distributions.html#cb294-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb294-3"><a href="visualizing-distributions.html#cb294-3" aria-hidden="true" tabindex="-1"></a><span class="fu">tibble</span>(</span>
<span id="cb294-4"><a href="visualizing-distributions.html#cb294-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">x=</span><span class="fu">c</span>(<span class="fu">rnorm</span>(<span class="dv">100</span>),<span class="fu">rnorm</span>(<span class="dv">100</span>,<span class="dv">4</span>),<span class="fu">rnorm</span>(<span class="dv">100</span>,<span class="dv">2</span>,<span class="dv">3</span>)),</span>
<span id="cb294-5"><a href="visualizing-distributions.html#cb294-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">type=</span><span class="fu">c</span>(<span class="fu">rep</span>(<span class="st">&#39;A&#39;</span>,<span class="dv">200</span>),<span class="fu">rep</span>(<span class="st">&#39;B&#39;</span>,<span class="dv">100</span>)),</span>
<span id="cb294-6"><a href="visualizing-distributions.html#cb294-6" aria-hidden="true" tabindex="-1"></a>) <span class="sc">%&gt;%</span></span>
<span id="cb294-7"><a href="visualizing-distributions.html#cb294-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">y=</span>type,<span class="at">x=</span>x,<span class="at">fill=</span>type)) <span class="sc">+</span></span>
<span id="cb294-8"><a href="visualizing-distributions.html#cb294-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_density_ridges</span>()</span></code></pre></div>
<p><img src="main_files/unnamed-chunk-119-1.png" width="672" /></p>
<p>Many distributions may be plotted:</p>
<div class="sourceCode" id="cb295"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb295-1"><a href="visualizing-distributions.html#cb295-1" aria-hidden="true" tabindex="-1"></a><span class="fu">tibble</span>(</span>
<span id="cb295-2"><a href="visualizing-distributions.html#cb295-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">x=</span><span class="fu">rnorm</span>(<span class="dv">10000</span>,<span class="at">mean=</span><span class="fu">runif</span>(<span class="dv">10</span>,<span class="dv">1</span>,<span class="dv">10</span>),<span class="at">sd=</span><span class="fu">runif</span>(<span class="dv">2</span>,<span class="dv">1</span>,<span class="dv">4</span>)),</span>
<span id="cb295-3"><a href="visualizing-distributions.html#cb295-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">type=</span><span class="fu">rep</span>(<span class="fu">c</span>(<span class="st">&quot;A&quot;</span>,<span class="st">&quot;B&quot;</span>,<span class="st">&quot;C&quot;</span>,<span class="st">&quot;D&quot;</span>,<span class="st">&quot;E&quot;</span>,<span class="st">&quot;F&quot;</span>,<span class="st">&quot;G&quot;</span>,<span class="st">&quot;H&quot;</span>,<span class="st">&quot;I&quot;</span>,<span class="st">&quot;J&quot;</span>),<span class="dv">1000</span>)</span>
<span id="cb295-4"><a href="visualizing-distributions.html#cb295-4" aria-hidden="true" tabindex="-1"></a>) <span class="sc">%&gt;%</span></span>
<span id="cb295-5"><a href="visualizing-distributions.html#cb295-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">y=</span>type,<span class="at">x=</span>x,<span class="at">fill=</span>type)) <span class="sc">+</span></span>
<span id="cb295-6"><a href="visualizing-distributions.html#cb295-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_density_ridges</span>(<span class="at">alpha=</span><span class="fl">0.6</span>,<span class="at">position=</span><span class="st">&quot;identity&quot;</span>)</span></code></pre></div>
<p><img src="main_files/unnamed-chunk-120-1.png" width="672" /></p>
<div class="box readmore">
<ul>
<li><a href="https://www.r-graph-gallery.com/294-basic-ridgeline-plot.html">R Graph Gallery - ridgeline plots</a></li>
<li><a href="https://cran.r-project.org/web/packages/ggridges/vignettes/introduction.html">ggridges package vignette</a></li>
<li><a href="https://cran.r-project.org/web/packages/ggridges/index.html">ggridges package index on CRAN</a></li>
</ul>
</div>

</div>
</div>
<p style="text-align: center;">
<a href="plotting-one-dimension.html"><button class="btn btn-default">Previous</button></a>
<a href="plotting-two-or-more-dimensions.html"><button class="btn btn-default">Next</button></a>
</p>
</div>
</div>


</div>

<script>

// add bootstrap table styles to pandoc tables
$(document).ready(function () {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
});

</script>

</body>
</html>
